// typegpu/core/sampler/sampler.js
import { $gpuValueOf, $internal, $ownSnippet, $resolve } from "../../shared/symbols.js";
import { getName, setName } from "../../shared/meta.js";
import { snip } from "../../data/snippet.js";
import { inCodegenMode } from "../../execMode.js";
import { valueProxyHandler } from "../valueProxyUtils.js";
import { comparisonSampler, sampler } from "../../data/sampler.js";

//#region src/core/sampler/sampler.ts
function INTERNAL_createSampler(props, branch) {
	return new TgpuFixedSamplerImpl(sampler(), props, branch);
}
function INTERNAL_createComparisonSampler(props, branch) {
	return new TgpuFixedSamplerImpl(comparisonSampler(), props, branch);
}
function isSampler(resource) {
	const maybe = resource;
	return maybe?.resourceType === "sampler" && !!maybe[$internal];
}
function isComparisonSampler(resource) {
	const maybe = resource;
	return maybe?.resourceType === "sampler-comparison" && !!maybe[$internal];
}
var TgpuLaidOutSamplerImpl = class {
	[$internal] = { unwrap: void 0 };
	resourceType;
	schema;
	#membership;
	constructor(schema, membership) {
		this.schema = schema;
		this.#membership = membership;
		this.resourceType = schema.type === "sampler_comparison" ? "sampler-comparison" : "sampler";
		setName(this, membership.key);
	}
	[$resolve](ctx) {
		const id = ctx.makeUniqueIdentifier(getName(this), "global");
		const group = ctx.allocateLayoutEntry(this.#membership.layout);
		ctx.addDeclaration(`@group(${group}) @binding(${this.#membership.idx}) var ${id}: ${ctx.resolve(this.schema).value};`);
		return snip(id, this.schema, "handle");
	}
	get [$gpuValueOf]() {
		const schema = this.schema;
		return new Proxy({
			[$internal]: true,
			get [$ownSnippet]() {
				return snip(this, schema, "handle");
			},
			[$resolve]: (ctx) => ctx.resolve(this),
			toString: () => `${this.toString()}.$`
		}, valueProxyHandler);
	}
	get $() {
		if (inCodegenMode()) return this[$gpuValueOf];
		throw new Error("Direct access to sampler values is possible only as part of a compute dispatch or draw call.");
	}
	get value() {
		return this.$;
	}
	toString() {
		return `${this.resourceType}:${getName(this) ?? "<unnamed>"}`;
	}
};
var TgpuFixedSamplerImpl = class {
	[$internal];
	resourceType;
	schema;
	#filtering;
	#sampler = null;
	#props;
	#branch;
	constructor(schema, props, branch) {
		this.schema = schema;
		this.#props = props;
		this.#branch = branch;
		this.resourceType = schema.type === "sampler_comparison" ? "sampler-comparison" : "sampler";
		this[$internal] = { unwrap: () => {
			if (!this.#sampler) this.#sampler = this.#branch.device.createSampler({
				...this.#props,
				label: getName(this) ?? "<unnamed>"
			});
			return this.#sampler;
		} };
		this.#filtering = props.minFilter === "linear" || props.magFilter === "linear" || props.mipmapFilter === "linear";
	}
	[$resolve](ctx) {
		const id = ctx.makeUniqueIdentifier(getName(this), "global");
		const { group, binding } = ctx.allocateFixedEntry(this.schema.type === "sampler_comparison" ? { sampler: "comparison" } : { sampler: this.#filtering ? "filtering" : "non-filtering" }, this);
		ctx.addDeclaration(`@group(${group}) @binding(${binding}) var ${id}: ${ctx.resolve(this.schema).value};`);
		return snip(id, this.schema, "handle");
	}
	get [$gpuValueOf]() {
		const schema = this.schema;
		return new Proxy({
			[$internal]: true,
			get [$ownSnippet]() {
				return snip(this, schema, "handle");
			},
			[$resolve]: (ctx) => ctx.resolve(this),
			toString: () => `${this.toString()}.$`
		}, valueProxyHandler);
	}
	get $() {
		if (inCodegenMode()) return this[$gpuValueOf];
		throw new Error("Direct access to sampler values is possible only as part of a compute dispatch or draw call.");
	}
	get value() {
		return this.$;
	}
	$name(label) {
		setName(this, label);
		return this;
	}
	toString() {
		return `${this.resourceType}:${getName(this) ?? "<unnamed>"}`;
	}
};

//#endregion
export { INTERNAL_createComparisonSampler, INTERNAL_createSampler, TgpuLaidOutSamplerImpl, isComparisonSampler, isSampler };
