// typegpu/core/constant/tgpuConstant.js
import { $gpuValueOf, $internal, $ownSnippet, $resolve } from "../../shared/symbols.js";
import { getName, setName } from "../../shared/meta.js";
import { isData } from "../../data/dataTypes.js";
import { snip } from "../../data/snippet.js";
import { inCodegenMode } from "../../execMode.js";
import { valueProxyHandler } from "../valueProxyUtils.js";

//#region src/core/constant/tgpuConstant.ts
function constant(dataType, value) {
	if (!isData(dataType)) {
		if (!Array.isArray(value)) throw new Error(`Expected an array value for a partially-applied array schema, but received: ${typeof value}.`);
		return new TgpuConstImpl(dataType(value.length), value);
	}
	return new TgpuConstImpl(dataType, value);
}
function deepFreeze(object) {
	const propNames = Reflect.ownKeys(object);
	for (const name of propNames) {
		const value = object[name];
		if (value && typeof value === "object" || typeof value === "function") deepFreeze(value);
	}
	return Object.freeze(object);
}
var TgpuConstImpl = class {
	[$internal] = {};
	resourceType;
	dataType;
	#value;
	constructor(dataType, value) {
		this.resourceType = "const";
		this.dataType = dataType;
		this.#value = value && typeof value === "object" ? deepFreeze(value) : value;
	}
	$name(label) {
		setName(this, label);
		return this;
	}
	[$resolve](ctx) {
		const id = ctx.makeUniqueIdentifier(getName(this), "global");
		return ctx.gen.declareGlobalConst({
			id,
			dataType: this.dataType,
			init: snip(this.#value, this.dataType, "constant")
		});
	}
	toString() {
		return `const:${getName(this) ?? "<unnamed>"}`;
	}
	get [$gpuValueOf]() {
		const dataType = this.dataType;
		return new Proxy({
			[$internal]: true,
			get [$ownSnippet]() {
				return snip(this, dataType, "constant-immutable-def");
			},
			[$resolve]: (ctx) => ctx.resolve(this),
			toString: () => `const:${getName(this) ?? "<unnamed>"}.$`
		}, valueProxyHandler);
	}
	get $() {
		if (inCodegenMode()) return this[$gpuValueOf];
		return this.#value;
	}
	get value() {
		return this.$;
	}
};

//#endregion
export { constant };
