// typegpu/tgsl/forOfUtils.js
import { $gpuCallable } from "../shared/symbols.js";
import { isNaturallyEphemeral, isPtr, isVec, isWgslArray } from "../data/wgslTypes.js";
import { UnknownData } from "../data/dataTypes.js";
import { isAlias, snip } from "../data/snippet.js";
import { WgslTypeError, invariant } from "../errors.js";
import { stitch } from "../core/resolve/stitch.js";
import { createPtrFromOrigin, implicitFrom } from "../data/ptr.js";
import { i32, u32 } from "../data/numeric.js";
import { ArrayExpression, concretize } from "./generationHelpers.js";
import { accessIndex } from "./accessIndex.js";
import { arrayLength } from "../std/array.js";
import { isTgpuRange } from "../std/range.js";

//#region src/tgsl/forOfUtils.ts
function getLoopVarKind(elementSnippet) {
	return elementSnippet.origin === "constant-immutable-def" ? "const" : "let";
}
function getElementSnippet(iterableSnippet, index) {
	const elementSnippet = accessIndex(iterableSnippet, index);
	if (!elementSnippet) throw new WgslTypeError("`for ... of ...` loops only support array or vector iterables");
	return elementSnippet;
}
/**
* Determines the type of the element as accessible inside of the `for .. of` loop body
*/
function getElementType(elementSnippet, iterableSnippet) {
	let elementType = elementSnippet.dataType;
	if (elementType === UnknownData) throw new WgslTypeError(stitch`The elements in iterable ${iterableSnippet} are of unknown type`);
	if (isNaturallyEphemeral(elementSnippet.dataType) || elementSnippet.origin === "runtime" || elementSnippet.origin === "constant" || elementSnippet.origin === "constant-immutable-def" || elementSnippet.origin === "runtime-immutable-def") return elementType;
	if (!isPtr(elementType)) {
		const ptrType = createPtrFromOrigin(elementSnippet.origin, concretize(elementType));
		invariant(ptrType !== void 0, `Creating pointer type from origin ${elementSnippet.origin}`);
		elementType = ptrType;
	}
	return implicitFrom(elementType);
}
function getRangeSnippets(ctx, iterableSnippet, unroll = false) {
	const { value, dataType } = iterableSnippet;
	if (isTgpuRange(value)) {
		const { start, end, step } = value;
		const dataType$1 = [
			start,
			end,
			step
		].every((v) => v >= 0) ? u32 : i32;
		return {
			start: snip(start, dataType$1, "constant"),
			end: snip(end, dataType$1, "constant"),
			step: snip(step, dataType$1, "constant"),
			comparison: step < 0 ? ">" : "<"
		};
	}
	if (!unroll && !isAlias(iterableSnippet)) throw new Error(`\`for ... of ...\` loops only support std.range or iterables stored in variables.
-----
You can wrap iterable with \`tgpu.unroll(...)\`. If iterable is known at comptime, the loop will be unrolled.
-----`);
	const defaults = {
		start: snip(0, u32, "constant"),
		step: snip(1, u32, "constant"),
		comparison: "<"
	};
	if (isWgslArray(dataType)) return {
		...defaults,
		end: dataType.elementCount > 0 ? snip(dataType.elementCount, u32, "constant") : arrayLength[$gpuCallable].call(ctx, [iterableSnippet])
	};
	if (isVec(dataType)) return {
		...defaults,
		end: snip(dataType.componentCount, u32, "constant")
	};
	if (unroll) {
		if (Array.isArray(value)) return {
			...defaults,
			end: snip(value.length, u32, "constant")
		};
		if (value instanceof ArrayExpression) return {
			...defaults,
			end: snip(value.elements.length, u32, "constant")
		};
	}
	throw new WgslTypeError("`for ... of ...` loops only support array or vector iterables");
}

//#endregion
export { getElementSnippet, getElementType, getLoopVarKind, getRangeSnippets };
