// typegpu/tgsl/generationHelpers.js
import { $internal, $resolve } from "../shared/symbols.js";
import { WORKAROUND_getSchema, isMatInstance, isNaturallyEphemeral, isVecInstance } from "../data/wgslTypes.js";
import { UnknownData } from "../data/dataTypes.js";
import { isAlias, isSnippet, snip } from "../data/snippet.js";
import { WgslTypeError } from "../errors.js";
import { getOwnSnippet } from "../types.js";
import { isRef } from "../data/ref.js";
import { abstractFloat, abstractInt, bool, f32, i32 } from "../data/numeric.js";

//#region src/tgsl/generationHelpers.ts
function numericLiteralToSnippet(value) {
	if (value >= 2 ** 63 || value < -(2 ** 63)) return snip(value, abstractFloat, "constant", false);
	if (Number.isInteger(value)) {
		if (!Number.isSafeInteger(value)) console.warn(`The integer ${value} exceeds the safe integer range and may have lost precision.`);
		return snip(value, abstractInt, "constant", false);
	}
	return snip(value, abstractFloat, "constant", false);
}
function concretize(type) {
	if (type.type === "abstractFloat") return f32;
	if (type.type === "abstractInt") return i32;
	return type;
}
function concretizeSnippet(snippet) {
	return snip(snippet.value, concretize(snippet.dataType), snippet.origin);
}
function coerceToSnippet(value) {
	if (isSnippet(value)) return value;
	if (isRef(value)) throw new Error("Cannot use refs (d.ref(...)) from the outer scope.");
	const ownSnippet = getOwnSnippet(value);
	if (ownSnippet) return ownSnippet;
	if (isVecInstance(value) || isMatInstance(value)) return snip(value, WORKAROUND_getSchema(value), "constant", false);
	if (typeof value === "string" || typeof value === "function" || typeof value === "object" || typeof value === "symbol" || typeof value === "undefined" || value === null) return snip(value, UnknownData, "constant", false);
	if (typeof value === "number") return numericLiteralToSnippet(value);
	if (typeof value === "boolean") return snip(value, bool, "constant", false);
	return snip(value, UnknownData, "constant", false);
}
/**
* Intermediate representation for WGSL array expressions.
* Defers resolution. Stores array elements as snippets so the
* generator can access them when needed.
*/
var ArrayExpression = class {
	[$internal] = true;
	type;
	elements;
	constructor(type, elements) {
		this.type = type;
		this.elements = elements;
	}
	toString() {
		return "ArrayExpression";
	}
	[$resolve](ctx) {
		for (const elem of this.elements) if (isAlias(elem) && !isNaturallyEphemeral(elem.dataType)) {
			const snippetStr = ctx.resolve(elem.value, elem.dataType).value;
			const snippetType = ctx.resolve(concretize(elem.dataType)).value;
			throw new WgslTypeError(`'${snippetStr}' reference cannot be used in an array constructor.\n-----\nTry '${snippetType}(${snippetStr})' or 'arrayOf(${snippetType}, count)([...])' to copy the value instead.\n-----`);
		}
		return ctx.gen.typeInstantiation(this.type, this.elements);
	}
};

//#endregion
export { ArrayExpression, coerceToSnippet, concretize, concretizeSnippet, numericLiteralToSnippet };
