// typegpu/tgsl/infixDispatch.js
import { inCodegenMode } from "../execMode.js";
import { add, bitShiftLeft, bitShiftRight, div, mod, mul, sub } from "../std/operators.js";

//#region src/tgsl/infixDispatch.ts
const infixOperators = {
	add,
	sub,
	mul,
	div,
	mod,
	bitShiftLeft,
	bitShiftRight
};
/**
* InfixDispatch is only used in codegen mode.
* `lhs` may either be Numeric or Snippet.
* @example
* const external = d.vec2u(1);
* const fn = () => {
*    'use gpu';
*    external.mul(2); // lhs is Numeric
*    d.vec2u(1).mul(2) // lhs is a snippet
* }
*/
var InfixDispatch = class {
	lhs;
	operator;
	constructor(lhs, operator) {
		this.lhs = lhs;
		this.operator = operator;
	}
};
function isInfixDispatch(o) {
	return o instanceof InfixDispatch;
}
/**
* This function is used on vec/mat prototypes.
* This is done in runtime in order to avoid a circular dependency.
*/
function assignInfixOperator(base, operator, operatorSymbol) {
	const opImpl = infixOperators[operator];
	Object.defineProperty(base.prototype, operatorSymbol, { value: opImpl });
	function jsInfixDispatchFor(arg) {
		return opImpl(this, arg);
	}
	Object.defineProperty(base.prototype, operator, { get() {
		if (inCodegenMode()) return new InfixDispatch(this, opImpl);
		return jsInfixDispatchFor;
	} });
}

//#endregion
export { InfixDispatch, assignInfixOperator, infixOperators, isInfixDispatch };
