// typegpu/core/buffer/buffer.js
import { $internal } from "../../shared/symbols.js";
import { isWgslData } from "../../data/wgslTypes.js";
import { getName, setName } from "../../shared/meta.js";
import { isGPUBuffer } from "../../types.js";
import { sizeOf } from "../../data/sizeOf.js";
import { getCompiledWriter } from "../../data/compiledIO.js";
import { calculateOffsets, readFromArrayBuffer, writeToArrayBuffer } from "../../data/dataIO.js";
import { convertPartialToPatch, getPatchInstructions, patchArrayBuffer } from "../../data/partialIO.js";
import { mutable, readonly, uniform } from "./bufferUsage.js";

//#region src/core/buffer/buffer.ts
const usageToUsageConstructor = {
	uniform,
	mutable,
	readonly
};
function INTERNAL_createBuffer(group, typeSchema, initialOrBuffer) {
	if (!isWgslData(typeSchema)) return new TgpuBufferImpl(group, typeSchema, initialOrBuffer, ["storage", "uniform"]);
	return new TgpuBufferImpl(group, typeSchema, initialOrBuffer);
}
function isBuffer(value) {
	return value.resourceType === "buffer";
}
function isUsableAsVertex(buffer) {
	return !!buffer.usableAsVertex;
}
var TgpuBufferImpl = class {
	[$internal] = true;
	resourceType = "buffer";
	flags = GPUBufferUsage.COPY_DST | GPUBufferUsage.COPY_SRC;
	dataType;
	#device;
	#buffer = null;
	#ownBuffer;
	#destroyed = false;
	#internalBuffer;
	get #hostBuffer() {
		return this.#internalBuffer ??= new ArrayBuffer(sizeOf(this.dataType));
	}
	#mappedRange;
	#initialCallback;
	#disallowedUsages;
	initial;
	usableAsUniform = false;
	usableAsStorage = false;
	usableAsVertex = false;
	usableAsIndex = false;
	usableAsIndirect = false;
	constructor(root, dataType, initialOrBuffer, disallowedUsages) {
		this.dataType = dataType;
		this.#disallowedUsages = disallowedUsages;
		this.#device = root.device;
		if (isGPUBuffer(initialOrBuffer)) {
			this.#ownBuffer = false;
			this.#buffer = initialOrBuffer;
		} else {
			this.#ownBuffer = true;
			if (typeof initialOrBuffer === "function") this.#initialCallback = initialOrBuffer;
			else this.initial = initialOrBuffer;
		}
	}
	get buffer() {
		if (this.#destroyed) throw new Error("This buffer has been destroyed");
		if (!this.#buffer) {
			this.#buffer = this.#device.createBuffer({
				size: sizeOf(this.dataType),
				usage: this.flags,
				mappedAtCreation: !!this.initial || !!this.#initialCallback,
				label: getName(this) ?? "<unnamed>"
			});
			if (this.initial || this.#initialCallback) {
				if (this.#initialCallback) this.#initialCallback(this);
				else if (this.initial) writeToArrayBuffer(this.#getMappedRange(), this.dataType, this.initial);
				this.#unmapBuffer();
			}
		}
		return this.#buffer;
	}
	get destroyed() {
		return this.#destroyed;
	}
	get arrayBuffer() {
		if (this.buffer.mapState === "mapped") return this.#getMappedRange();
		return this.#hostBuffer;
	}
	#getMappedRange() {
		if (!this.#buffer || this.#buffer.mapState !== "mapped") throw new Error("Buffer is not mapped.");
		this.#mappedRange ??= this.#buffer.getMappedRange();
		return this.#mappedRange;
	}
	#unmapBuffer() {
		if (!this.#buffer || this.#buffer.mapState !== "mapped") return;
		this.#mappedRange = void 0;
		this.#buffer.unmap();
	}
	$name(label) {
		setName(this, label);
		if (this.#buffer) this.#buffer.label = label;
		return this;
	}
	$usage(...usages) {
		for (const usage of usages) {
			if (this.#disallowedUsages?.includes(usage)) throw new Error(`Buffer of type ${this.dataType.type} cannot be used as ${usage}`);
			this.flags |= usage === "uniform" ? GPUBufferUsage.UNIFORM : 0;
			this.flags |= usage === "storage" ? GPUBufferUsage.STORAGE : 0;
			this.flags |= usage === "vertex" ? GPUBufferUsage.VERTEX : 0;
			this.flags |= usage === "index" ? GPUBufferUsage.INDEX : 0;
			this.flags |= usage === "indirect" ? GPUBufferUsage.INDIRECT : 0;
			this.usableAsUniform = this.usableAsUniform || usage === "uniform";
			this.usableAsStorage = this.usableAsStorage || usage === "storage";
			this.usableAsVertex = this.usableAsVertex || usage === "vertex";
			this.usableAsIndex = this.usableAsIndex || usage === "index";
			this.usableAsIndirect = this.usableAsIndirect || usage === "indirect";
		}
		return this;
	}
	$addFlags(flags) {
		if (!this.#ownBuffer) throw new Error("Cannot add flags to a buffer that is not managed by TypeGPU.");
		if (flags & GPUBufferUsage.MAP_READ) {
			this.flags = GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ;
			return this;
		}
		if (flags & GPUBufferUsage.MAP_WRITE) {
			this.flags = GPUBufferUsage.COPY_SRC | GPUBufferUsage.MAP_WRITE;
			return this;
		}
		this.flags |= flags;
		return this;
	}
	compileWriter() {
		getCompiledWriter(this.dataType);
	}
	write(data, options) {
		const gpuBuffer = this.buffer;
		if (gpuBuffer.mapState === "mapped") {
			const mapped = this.#getMappedRange();
			if (data instanceof ArrayBuffer && data === mapped) return;
			writeToArrayBuffer(mapped, this.dataType, data, options);
			return;
		}
		if (!(data instanceof ArrayBuffer && data === this.#hostBuffer)) writeToArrayBuffer(this.#hostBuffer, this.dataType, data, options);
		const { startOffset, endOffset } = calculateOffsets(options, this.dataType, data);
		const size = endOffset - startOffset;
		this.#device.queue.writeBuffer(gpuBuffer, startOffset, this.#hostBuffer, startOffset, size);
	}
	/** @deprecated Use {@link patch} instead. */
	writePartial(data) {
		this.patch(convertPartialToPatch(this.dataType, data));
	}
	patch(data) {
		const gpuBuffer = this.buffer;
		if (gpuBuffer.mapState === "mapped") patchArrayBuffer(this.#getMappedRange(), this.dataType, data);
		else {
			const instructions = getPatchInstructions(this.dataType, data, this.#hostBuffer);
			for (const { data: data$1, gpuOffset } of instructions) this.#device.queue.writeBuffer(gpuBuffer, gpuOffset, data$1);
		}
	}
	clear() {
		const gpuBuffer = this.buffer;
		if (gpuBuffer.mapState === "mapped") {
			new Uint8Array(this.#getMappedRange()).fill(0);
			return;
		}
		const encoder = this.#device.createCommandEncoder();
		encoder.clearBuffer(gpuBuffer);
		this.#device.queue.submit([encoder.finish()]);
	}
	copyFrom(srcBuffer) {
		if (this.buffer.mapState === "mapped") throw new Error("Cannot copy to a mapped buffer.");
		const size = sizeOf(this.dataType);
		const encoder = this.#device.createCommandEncoder();
		encoder.copyBufferToBuffer(srcBuffer.buffer, 0, this.buffer, 0, size);
		this.#device.queue.submit([encoder.finish()]);
	}
	async read() {
		const gpuBuffer = this.buffer;
		if (gpuBuffer.mapState === "mapped") return readFromArrayBuffer(this.#getMappedRange(), this.dataType);
		if (gpuBuffer.usage & GPUBufferUsage.MAP_READ) {
			await gpuBuffer.mapAsync(GPUMapMode.READ);
			const res$1 = readFromArrayBuffer(this.#getMappedRange(), this.dataType);
			this.#unmapBuffer();
			return res$1;
		}
		const stagingBuffer = this.#device.createBuffer({
			size: sizeOf(this.dataType),
			usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ
		});
		const commandEncoder = this.#device.createCommandEncoder();
		commandEncoder.copyBufferToBuffer(gpuBuffer, 0, stagingBuffer, 0, sizeOf(this.dataType));
		this.#device.queue.submit([commandEncoder.finish()]);
		await stagingBuffer.mapAsync(GPUMapMode.READ, 0, sizeOf(this.dataType));
		const res = readFromArrayBuffer(stagingBuffer.getMappedRange(), this.dataType);
		stagingBuffer.unmap();
		stagingBuffer.destroy();
		return res;
	}
	as(usage) {
		return usageToUsageConstructor[usage]?.(this);
	}
	destroy() {
		if (this.#destroyed) return;
		this.#destroyed = true;
		this.#mappedRange = void 0;
		if (this.#ownBuffer) this.#buffer?.destroy();
	}
	toString() {
		return `buffer:${getName(this) ?? "<unnamed>"}`;
	}
};

//#endregion
export { INTERNAL_createBuffer, isBuffer, isUsableAsVertex };
