// typegpu/core/buffer/bufferUsage.js
import { $getNameForward, $gpuValueOf, $internal, $ownSnippet, $resolve } from "../../shared/symbols.js";
import { getName, setName } from "../../shared/meta.js";
import { snip } from "../../data/snippet.js";
import { IllegalBufferAccessError } from "../../errors.js";
import { getExecMode, inCodegenMode, isInsideTgpuFn } from "../../execMode.js";
import { schemaCallWrapper } from "../../data/schemaCallWrapper.js";
import { assertExhaustive } from "../../shared/utilityTypes.js";
import { valueProxyHandler } from "../valueProxyUtils.js";
import { isUsableAsStorage } from "../../extension.js";

//#region src/core/buffer/bufferUsage.ts
function isUsableAsUniform(buffer) {
	return !!buffer.usableAsUniform;
}
var TgpuFixedBufferImpl = class {
	resourceType = "buffer-usage";
	usage;
	buffer;
	[$internal];
	[$getNameForward];
	constructor(usage, buffer) {
		this.usage = usage;
		this.buffer = buffer;
		this[$internal] = { dataType: buffer.dataType };
		this[$getNameForward] = buffer;
	}
	$name(label) {
		setName(this, label);
		return this;
	}
	[$resolve](ctx) {
		const dataType = this.buffer.dataType;
		const id = ctx.makeUniqueIdentifier(getName(this), "global");
		const { group, binding } = ctx.allocateFixedEntry(this.usage === "uniform" ? { uniform: dataType } : {
			storage: dataType,
			access: this.usage
		}, this.buffer);
		return ctx.gen.declareGlobalVar({
			group,
			binding,
			scope: this.usage,
			id,
			dataType,
			init: void 0
		});
	}
	toString() {
		return `${this.usage}:${getName(this) ?? "<unnamed>"}`;
	}
	get [$gpuValueOf]() {
		const dataType = this.buffer.dataType;
		const usage = this.usage;
		return new Proxy({
			[$internal]: true,
			get [$ownSnippet]() {
				return snip(this, dataType, usage);
			},
			[$resolve]: (ctx) => ctx.resolve(this),
			toString: () => `${this.usage}:${getName(this) ?? "<unnamed>"}.$`
		}, valueProxyHandler);
	}
	get $() {
		const mode = getExecMode();
		const insideTgpuFn = isInsideTgpuFn();
		if (mode.type === "normal") throw new IllegalBufferAccessError(insideTgpuFn ? `Cannot access ${String(this.buffer)}. TypeGPU functions that depends on GPU resources need to be part of a compute dispatch, draw call or simulation` : ".$ is inaccessible during normal JS execution. Try `.read()`");
		if (mode.type === "codegen") return this[$gpuValueOf];
		if (mode.type === "simulate") {
			if (!mode.buffers.has(this.buffer)) mode.buffers.set(this.buffer, schemaCallWrapper(this.buffer.dataType, this.buffer.initial));
			return mode.buffers.get(this.buffer);
		}
		return assertExhaustive(mode, "bufferUsage.ts#TgpuFixedBufferImpl/$");
	}
	get value() {
		return this.$;
	}
	set $(value) {
		const mode = getExecMode();
		const insideTgpuFn = isInsideTgpuFn();
		if (mode.type === "normal") throw new IllegalBufferAccessError(insideTgpuFn ? `Cannot access ${String(this.buffer)}. TypeGPU functions that depends on GPU resources need to be part of a compute dispatch, draw call or simulation` : ".$ is inaccessible during normal JS execution. Try `.write()`");
		if (mode.type === "codegen") throw new Error("Unreachable bufferUsage.ts#TgpuFixedBufferImpl/$");
		if (mode.type === "simulate") {
			mode.buffers.set(this.buffer, value);
			return;
		}
		assertExhaustive(mode, "bufferUsage.ts#TgpuFixedBufferImpl/$");
	}
	set value(value) {
		this.$ = value;
	}
};
var TgpuLaidOutBufferImpl = class {
	[$internal];
	resourceType = "buffer-usage";
	usage;
	dataType;
	#membership;
	constructor(usage, dataType, membership) {
		this[$internal] = { dataType };
		this.usage = usage;
		this.dataType = dataType;
		this.#membership = membership;
		setName(this, membership.key);
	}
	[$resolve](ctx) {
		const id = ctx.makeUniqueIdentifier(getName(this), "global");
		const group = ctx.allocateLayoutEntry(this.#membership.layout);
		return ctx.gen.declareGlobalVar({
			group,
			binding: this.#membership.idx,
			scope: this.usage,
			id,
			dataType: this.dataType,
			init: void 0
		});
	}
	toString() {
		return `${this.usage}:${getName(this) ?? "<unnamed>"}`;
	}
	get [$gpuValueOf]() {
		const schema = this.dataType;
		const usage = this.usage;
		return new Proxy({
			[$internal]: true,
			get [$ownSnippet]() {
				return snip(this, schema, usage);
			},
			[$resolve]: (ctx) => ctx.resolve(this),
			toString: () => `${this.usage}:${getName(this) ?? "<unnamed>"}.$`
		}, valueProxyHandler);
	}
	get $() {
		if (inCodegenMode()) return this[$gpuValueOf];
		throw new Error("Direct access to buffer values is possible only as part of a compute dispatch or draw call. Try .read() or .write() instead");
	}
	get value() {
		return this.$;
	}
};
const mutableUsageMap = /* @__PURE__ */ new WeakMap();
function mutable(buffer) {
	if (!isUsableAsStorage(buffer)) throw new Error(`Cannot call as('mutable') on ${buffer}, as it is not allowed to be used as storage. To allow it, call .$usage('storage') when creating the buffer.`);
	let usage = mutableUsageMap.get(buffer);
	if (!usage) {
		usage = new TgpuFixedBufferImpl("mutable", buffer);
		mutableUsageMap.set(buffer, usage);
	}
	return usage;
}
const readonlyUsageMap = /* @__PURE__ */ new WeakMap();
function readonly(buffer) {
	if (!isUsableAsStorage(buffer)) throw new Error(`Cannot call as('readonly') on ${buffer}, as it is not allowed to be used as storage. To allow it, call .$usage('storage') when creating the buffer.`);
	let usage = readonlyUsageMap.get(buffer);
	if (!usage) {
		usage = new TgpuFixedBufferImpl("readonly", buffer);
		readonlyUsageMap.set(buffer, usage);
	}
	return usage;
}
const uniformUsageMap = /* @__PURE__ */ new WeakMap();
function uniform(buffer) {
	if (!isUsableAsUniform(buffer)) throw new Error(`Cannot call as('uniform') on ${buffer}, as it is not allowed to be used as a uniform. To allow it, call .$usage('uniform') when creating the buffer.`);
	let usage = uniformUsageMap.get(buffer);
	if (!usage) {
		usage = new TgpuFixedBufferImpl("uniform", buffer);
		uniformUsageMap.set(buffer, usage);
	}
	return usage;
}

//#endregion
export { TgpuLaidOutBufferImpl, isUsableAsUniform, mutable, readonly, uniform };
