// typegpu/core/buffer/bufferShorthand.js
import { $getNameForward, $gpuValueOf, $internal, $resolve } from "../../shared/symbols.js";
import { getName, setName } from "../../shared/meta.js";

//#region src/core/buffer/bufferShorthand.ts
function isBufferShorthand(value) {
	return value instanceof TgpuBufferShorthandImpl;
}
var TgpuBufferShorthandImpl = class {
	[$internal] = true;
	[$getNameForward];
	resourceType;
	buffer;
	#usage;
	constructor(resourceType, buffer) {
		this.resourceType = resourceType;
		this.buffer = buffer;
		this[$getNameForward] = buffer;
		this.#usage = this.buffer.as(this.resourceType);
	}
	$name(label) {
		setName(this, label);
		return this;
	}
	write(data, options) {
		this.buffer.write(data, options);
	}
	/** @deprecated Use {@link patch} instead. */
	writePartial(data) {
		this.buffer.writePartial(data);
	}
	patch(data) {
		this.buffer.patch(data);
	}
	read() {
		return this.buffer.read();
	}
	get [$gpuValueOf]() {
		return this.#usage.$;
	}
	get $() {
		return this.#usage.$;
	}
	get value() {
		return this.$;
	}
	toString() {
		return `${this.resourceType}BufferShorthand:${getName(this) ?? "<unnamed>"}`;
	}
	[$resolve](ctx) {
		return ctx.resolve(this.#usage);
	}
};

//#endregion
export { TgpuBufferShorthandImpl, isBufferShorthand };
