// shaders/dist/core/AngularBlur-CFsh0PlD.js
import { At as call, Ot as ZERO, Pt as floatE, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { c as transformPosition } from "./transformations-fqzznt4g.js";
var rawWeights = Array.from({ length: 32 }, (_, i) => {
	const t = (i / 31 - .5) * 2;
	return Math.exp(-.5 * t * t / .64);
});
var wSum = rawWeights.reduce((a, b) => a + b);
const BLUR_WEIGHTS = rawWeights.map((w) => w / wSum);
const TAP_COUNT = 32;
const angularBlurTapCoord = tgpu.fn([
	d.vec2f,
	d.f32,
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, intensity, uv, aspect, tapIndex) => {
	"use gpu";
	const centerPos = d.vec2f(center.x, __tsover_sub(1, center.y));
	const angleStep = __tsover_div(__tsover_mul(intensity, .005), 31);
	const angle = __tsover_mul(tapIndex, angleStep);
	const cosA = std.cos(angle);
	const sinA = std.sin(angle);
	const initX = __tsover_sub(uv.x, centerPos.x);
	const initY = __tsover_sub(uv.y, centerPos.y);
	const acX = __tsover_mul(initX, aspect);
	const acY = initY;
	const rotX = __tsover_sub(__tsover_mul(acX, cosA), __tsover_mul(acY, sinA));
	const rotY = __tsover_add(__tsover_mul(acX, sinA), __tsover_mul(acY, cosA));
	return d.vec2f(__tsover_add(__tsover_div(rotX, aspect), centerPos.x), __tsover_add(rotY, centerPos.y));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "tapIndex"
			}
		],
		"body": [0, [
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"center",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"angleStep",
				[
					1,
					[
						1,
						"intensity",
						"*",
						[5, "0.005"]
					],
					"/",
					[5, "31"]
				]
			],
			[
				13,
				"angle",
				[
					1,
					"tapIndex",
					"*",
					"angleStep"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angle"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angle"]
				]
			],
			[
				13,
				"initX",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"-",
					[
						7,
						"centerPos",
						"x"
					]
				]
			],
			[
				13,
				"initY",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					[
						7,
						"centerPos",
						"y"
					]
				]
			],
			[
				13,
				"acX",
				[
					1,
					"initX",
					"*",
					"aspect"
				]
			],
			[
				13,
				"acY",
				"initY"
			],
			[
				13,
				"rotX",
				[
					1,
					[
						1,
						"acX",
						"*",
						"cosA"
					],
					"-",
					[
						1,
						"acY",
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"rotY",
				[
					1,
					[
						1,
						"acX",
						"*",
						"sinA"
					],
					"+",
					[
						1,
						"acY",
						"*",
						"cosA"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						1,
						"rotX",
						"/",
						"aspect"
					],
					"+",
					[
						7,
						"centerPos",
						"x"
					]
				], [
					1,
					"rotY",
					"+",
					[
						7,
						"centerPos",
						"y"
					]
				]]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "AngularBlur",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Blurs",
	description: "Radial motion blur rotating around a center point",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 20,
			description: "Intensity of the angular blur effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Blur Intensity",
				group: "Effect"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the rotation",
			ui: {
				type: "position",
				label: "Center Position",
				group: "Position"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		let total;
		for (let i = 0; i < 32; i++) {
			const coord = call(angularBlurTapCoord, "angularBlurTapCoord", [
				uniforms.center,
				uniforms.intensity,
				ctx.uv,
				ctx.aspect,
				floatE(i)
			]);
			const weighted = texture.sample(coord).mul(BLUR_WEIGHTS[i]);
			total = total ? total.add(weighted) : weighted;
		}
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [total]);
	}
};
var AngularBlur_default = componentDefinition;
export { componentDefinition as a, angularBlurTapCoord as i, BLUR_WEIGHTS as n, TAP_COUNT as r, AngularBlur_default as t };
