// shaders/dist/core/Aurora-C0KIoJY5.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d, rt as noise_exports, tt as mxNoiseFloat3 } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
var TWO_PI = Math.PI * 2;
const auroraCurtain = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, layerIdx, layerTime, centerX, centerY, aspect, waviness, rayAmount, auroraHeight, seed, t) => {
	"use gpu";
	const tSlow = __tsover_add(__tsover_mul(t, .06), layerTime);
	const lSeed = __tsover_add(seed, __tsover_mul(layerIdx, 31.7));
	const wx = __tsover_mul(__tsover_mul(__tsover_sub(uv.x, centerX), aspect), 2);
	const ph1 = __tsover_mul(std.fract(__tsover_mul(std.sin(__tsover_mul(lSeed, 12.9)), 4758.5)), TWO_PI);
	const ph2 = __tsover_mul(std.fract(__tsover_mul(std.sin(__tsover_mul(lSeed, 78.2)), 2847.1)), TWO_PI);
	const ph3 = __tsover_mul(std.fract(__tsover_mul(std.sin(__tsover_mul(lSeed, 41.6)), 1593.7)), TWO_PI);
	const arc = __tsover_add(__tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_add(__tsover_mul(wx, .7), __tsover_mul(tSlow, .4)), ph1)), .15), __tsover_mul(std.sin(__tsover_add(__tsover_add(__tsover_mul(wx, 1.6), __tsover_mul(tSlow, .65)), ph2)), .08)), __tsover_mul(std.sin(__tsover_add(__tsover_add(__tsover_mul(wx, 2.9), __tsover_mul(tSlow, .3)), ph3)), .04));
	const noiseWarp = mxNoiseFloat3(d.vec3f(__tsover_mul(wx, .5), __tsover_mul(tSlow, .25), __tsover_mul(lSeed, .1)));
	const pathY = __tsover_add(__tsover_add(centerY, __tsover_mul(arc, waviness)), __tsover_mul(__tsover_mul(noiseWarp, .14), waviness));
	const warpedX = __tsover_add(wx, __tsover_mul(noiseWarp, 2.5));
	const rayFreq = std.mix(d.f32(6), d.f32(20), rayAmount);
	const rayPhase = __tsover_mul(std.fract(__tsover_mul(std.sin(__tsover_mul(lSeed, 53.7)), 3847.2)), TWO_PI);
	const rays = std.pow(__tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_add(__tsover_mul(warpedX, __tsover_mul(rayFreq, .15)), __tsover_mul(tSlow, .1)), rayPhase)), .5), .5), 1.5);
	const localHeight = __tsover_mul(auroraHeight, std.mix(.4, 1, rays));
	const distFromPath = __tsover_sub(uv.y, pathY);
	const bottomFade = std.smoothstep(-.04, .015, distFromPath);
	const topExtent = __tsover_mul(localHeight, __tsover_add(__tsover_mul(rays, .4), .6));
	const topFade = __tsover_sub(1, std.smoothstep(0, topExtent, distFromPath));
	const envelope = __tsover_mul(bottomFade, topFade);
	const sectionBright = std.mix(.4, 1, __tsover_add(__tsover_mul(noiseWarp, .5), .5));
	const colorT = std.clamp(__tsover_div(distFromPath, __tsover_add(topExtent, .001)), 0, 1);
	const coreD = __tsover_sub(colorT, .2);
	const coreBright = __tsover_add(__tsover_mul(std.exp(__tsover_mul(__tsover_mul(coreD, coreD), -6)), .3), 1);
	const finalAlpha = __tsover_mul(__tsover_mul(__tsover_mul(envelope, rays), sectionBright), coreBright);
	return d.vec2f(colorT, finalAlpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "layerIdx"
			},
			{
				"type": "i",
				"name": "layerTime"
			},
			{
				"type": "i",
				"name": "centerX"
			},
			{
				"type": "i",
				"name": "centerY"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "waviness"
			},
			{
				"type": "i",
				"name": "rayAmount"
			},
			{
				"type": "i",
				"name": "auroraHeight"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "t"
			}
		],
		"body": [0, [
			[
				13,
				"tSlow",
				[
					1,
					[
						1,
						"t",
						"*",
						[5, "0.06"]
					],
					"+",
					"layerTime"
				]
			],
			[
				13,
				"lSeed",
				[
					1,
					"seed",
					"+",
					[
						1,
						"layerIdx",
						"*",
						[5, "31.7"]
					]
				]
			],
			[
				13,
				"wx",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"-",
							"centerX"
						],
						"*",
						"aspect"
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"ph1",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									"lSeed",
									"*",
									[5, "12.9"]
								]]
							],
							"*",
							[5, "4758.5"]
						]]
					],
					"*",
					"TWO_PI"
				]
			],
			[
				13,
				"ph2",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									"lSeed",
									"*",
									[5, "78.2"]
								]]
							],
							"*",
							[5, "2847.1"]
						]]
					],
					"*",
					"TWO_PI"
				]
			],
			[
				13,
				"ph3",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									"lSeed",
									"*",
									[5, "41.6"]
								]]
							],
							"*",
							[5, "1593.7"]
						]]
					],
					"*",
					"TWO_PI"
				]
			],
			[
				13,
				"arc",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										1,
										[
											1,
											"wx",
											"*",
											[5, "0.7"]
										],
										"+",
										[
											1,
											"tSlow",
											"*",
											[5, "0.4"]
										]
									],
									"+",
									"ph1"
								]]
							],
							"*",
							[5, "0.15"]
						],
						"+",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										1,
										[
											1,
											"wx",
											"*",
											[5, "1.6"]
										],
										"+",
										[
											1,
											"tSlow",
											"*",
											[5, "0.65"]
										]
									],
									"+",
									"ph2"
								]]
							],
							"*",
							[5, "0.08"]
						]
					],
					"+",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									[
										1,
										"wx",
										"*",
										[5, "2.9"]
									],
									"+",
									[
										1,
										"tSlow",
										"*",
										[5, "0.3"]
									]
								],
								"+",
								"ph3"
							]]
						],
						"*",
						[5, "0.04"]
					]
				]
			],
			[
				13,
				"warpPos",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							"wx",
							"*",
							[5, "0.5"]
						],
						[
							1,
							"tSlow",
							"*",
							[5, "0.25"]
						],
						[
							1,
							"lSeed",
							"*",
							[5, "0.1"]
						]
					]
				]
			],
			[
				13,
				"noiseWarp",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					["warpPos"]
				]
			],
			[
				13,
				"pathY",
				[
					1,
					[
						1,
						"centerY",
						"+",
						[
							1,
							"arc",
							"*",
							"waviness"
						]
					],
					"+",
					[
						1,
						[
							1,
							"noiseWarp",
							"*",
							[5, "0.14"]
						],
						"*",
						"waviness"
					]
				]
			],
			[
				13,
				"warpedX",
				[
					1,
					"wx",
					"+",
					[
						1,
						"noiseWarp",
						"*",
						[5, "2.5"]
					]
				]
			],
			[
				13,
				"rayFreq",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "6"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "20"]]
						],
						"rayAmount"
					]
				]
			],
			[
				13,
				"rayPhase",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									"lSeed",
									"*",
									[5, "53.7"]
								]]
							],
							"*",
							[5, "3847.2"]
						]]
					],
					"*",
					"TWO_PI"
				]
			],
			[
				13,
				"rays",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										1,
										[
											1,
											"warpedX",
											"*",
											[
												1,
												"rayFreq",
												"*",
												[5, "0.15"]
											]
										],
										"+",
										[
											1,
											"tSlow",
											"*",
											[5, "0.1"]
										]
									],
									"+",
									"rayPhase"
								]]
							],
							"*",
							[5, "0.5"]
						],
						"+",
						[5, "0.5"]
					], [5, "1.5"]]
				]
			],
			[
				13,
				"localHeight",
				[
					1,
					"auroraHeight",
					"*",
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							[5, "0.4"],
							[5, "1"],
							"rays"
						]
					]
				]
			],
			[
				13,
				"distFromPath",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					"pathY"
				]
			],
			[
				13,
				"bottomFade",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							4,
							"-",
							[5, "0.04"]
						],
						[5, "0.015"],
						"distFromPath"
					]
				]
			],
			[
				13,
				"topExtent",
				[
					1,
					"localHeight",
					"*",
					[
						1,
						[
							1,
							"rays",
							"*",
							[5, "0.4"]
						],
						"+",
						[5, "0.6"]
					]
				]
			],
			[
				13,
				"topFade",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							"topExtent",
							"distFromPath"
						]
					]
				]
			],
			[
				13,
				"envelope",
				[
					1,
					"bottomFade",
					"*",
					"topFade"
				]
			],
			[
				13,
				"sectionBright",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "0.4"],
						[5, "1"],
						[
							1,
							[
								1,
								"noiseWarp",
								"*",
								[5, "0.5"]
							],
							"+",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"colorT",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"distFromPath",
							"/",
							[
								1,
								"topExtent",
								"+",
								[5, "0.001"]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"coreD",
				[
					1,
					"colorT",
					"-",
					[5, "0.2"]
				]
			],
			[
				13,
				"coreBright",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									1,
									"coreD",
									"*",
									"coreD"
								],
								"*",
								[
									4,
									"-",
									[5, "6"]
								]
							]]
						],
						"*",
						[5, "0.3"]
					],
					"+",
					[5, "1"]
				]
			],
			[
				13,
				"finalAlpha",
				[
					1,
					[
						1,
						[
							1,
							"envelope",
							"*",
							"rays"
						],
						"*",
						"sectionBright"
					],
					"*",
					"coreBright"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["colorT", "finalAlpha"]
			]]
		]],
		"externalNames": [
			"std",
			"TWO_PI",
			"d",
			"noise"
		]
	},
	externals: () => ({
		std,
		TWO_PI,
		d,
		noise: noise_exports
	})
}) && $.f)({}));
const auroraField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, animTime, seed, waviness, rayDensity, height, center, balance, intensity, curtainCount) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const centerX = center.x;
	const centerY = __tsover_sub(1, center.y);
	const wav = __tsover_mul(waviness, .01);
	const rayAmt = __tsover_mul(rayDensity, .01);
	const auroraHeight = __tsover_mul(height, .01);
	const intens = __tsover_mul(intensity, .018);
	const balanceExp = std.mix(2.5, .4, __tsover_mul(balance, .01));
	const t = animTime;
	const a0 = auroraCurtain(uv, d.f32(0), 0, centerX, centerY, aspect, wav, rayAmt, auroraHeight, seed, t);
	let weightedT = __tsover_mul(a0.x, a0.y);
	let totalAlpha = a0.y;
	const g1 = std.select(d.f32(0), d.f32(1), curtainCount >= 2);
	const a1 = auroraCurtain(uv, d.f32(1), 3.1, centerX, centerY, aspect, wav, rayAmt, auroraHeight, seed, t);
	const w1 = __tsover_mul(__tsover_mul(a1.y, .65), g1);
	weightedT = __tsover_add(weightedT, __tsover_mul(a1.x, w1));
	totalAlpha = __tsover_add(totalAlpha, w1);
	const g2 = std.select(d.f32(0), d.f32(1), curtainCount >= 3);
	const a2 = auroraCurtain(uv, d.f32(2), 6.7, centerX, centerY, aspect, wav, rayAmt, auroraHeight, seed, t);
	const w2 = __tsover_mul(__tsover_mul(a2.y, .45), g2);
	weightedT = __tsover_add(weightedT, __tsover_mul(a2.x, w2));
	totalAlpha = __tsover_add(totalAlpha, w2);
	const g3 = std.select(d.f32(0), d.f32(1), curtainCount >= 4);
	const a3 = auroraCurtain(uv, d.f32(3), 9.4, centerX, centerY, aspect, wav, rayAmt, auroraHeight, seed, t);
	const w3 = __tsover_mul(__tsover_mul(a3.y, .3), g3);
	weightedT = __tsover_add(weightedT, __tsover_mul(a3.x, w3));
	totalAlpha = __tsover_add(totalAlpha, w3);
	const avgT = std.pow(__tsover_div(weightedT, __tsover_add(totalAlpha, .001)), balanceExp);
	const lowToMid = std.smoothstep(0, .2, avgT);
	const midToHigh = std.smoothstep(.4, .85, avgT);
	const brightness = __tsover_mul(totalAlpha, intens);
	return d.vec3f(lowToMid, midToHigh, brightness);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "waviness"
			},
			{
				"type": "i",
				"name": "rayDensity"
			},
			{
				"type": "i",
				"name": "height"
			},
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "balance"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "curtainCount"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"centerX",
				[
					7,
					"center",
					"x"
				]
			],
			[
				13,
				"centerY",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"center",
						"y"
					]
				]
			],
			[
				13,
				"wav",
				[
					1,
					"waviness",
					"*",
					[5, "0.01"]
				]
			],
			[
				13,
				"rayAmt",
				[
					1,
					"rayDensity",
					"*",
					[5, "0.01"]
				]
			],
			[
				13,
				"auroraHeight",
				[
					1,
					"height",
					"*",
					[5, "0.01"]
				]
			],
			[
				13,
				"intens",
				[
					1,
					"intensity",
					"*",
					[5, "0.018"]
				]
			],
			[
				13,
				"balanceExp",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "2.5"],
						[5, "0.4"],
						[
							1,
							"balance",
							"*",
							[5, "0.01"]
						]
					]
				]
			],
			[
				13,
				"t",
				"animTime"
			],
			[
				13,
				"a0",
				[
					6,
					"auroraCurtain",
					[
						"uv",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[5, "0"],
						"centerX",
						"centerY",
						"aspect",
						"wav",
						"rayAmt",
						"auroraHeight",
						"seed",
						"t"
					]
				]
			],
			[
				12,
				"weightedT",
				[
					1,
					[
						7,
						"a0",
						"x"
					],
					"*",
					[
						7,
						"a0",
						"y"
					]
				]
			],
			[
				12,
				"totalAlpha",
				[
					7,
					"a0",
					"y"
				]
			],
			[
				13,
				"g1",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"curtainCount",
							">=",
							[5, "2"]
						]
					]
				]
			],
			[
				13,
				"a1",
				[
					6,
					"auroraCurtain",
					[
						"uv",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[5, "3.1"],
						"centerX",
						"centerY",
						"aspect",
						"wav",
						"rayAmt",
						"auroraHeight",
						"seed",
						"t"
					]
				]
			],
			[
				13,
				"w1",
				[
					1,
					[
						1,
						[
							7,
							"a1",
							"y"
						],
						"*",
						[5, "0.65"]
					],
					"*",
					"g1"
				]
			],
			[
				2,
				"weightedT",
				"=",
				[
					1,
					"weightedT",
					"+",
					[
						1,
						[
							7,
							"a1",
							"x"
						],
						"*",
						"w1"
					]
				]
			],
			[
				2,
				"totalAlpha",
				"=",
				[
					1,
					"totalAlpha",
					"+",
					"w1"
				]
			],
			[
				13,
				"g2",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"curtainCount",
							">=",
							[5, "3"]
						]
					]
				]
			],
			[
				13,
				"a2",
				[
					6,
					"auroraCurtain",
					[
						"uv",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "2"]]
						],
						[5, "6.7"],
						"centerX",
						"centerY",
						"aspect",
						"wav",
						"rayAmt",
						"auroraHeight",
						"seed",
						"t"
					]
				]
			],
			[
				13,
				"w2",
				[
					1,
					[
						1,
						[
							7,
							"a2",
							"y"
						],
						"*",
						[5, "0.45"]
					],
					"*",
					"g2"
				]
			],
			[
				2,
				"weightedT",
				"=",
				[
					1,
					"weightedT",
					"+",
					[
						1,
						[
							7,
							"a2",
							"x"
						],
						"*",
						"w2"
					]
				]
			],
			[
				2,
				"totalAlpha",
				"=",
				[
					1,
					"totalAlpha",
					"+",
					"w2"
				]
			],
			[
				13,
				"g3",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"curtainCount",
							">=",
							[5, "4"]
						]
					]
				]
			],
			[
				13,
				"a3",
				[
					6,
					"auroraCurtain",
					[
						"uv",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "3"]]
						],
						[5, "9.4"],
						"centerX",
						"centerY",
						"aspect",
						"wav",
						"rayAmt",
						"auroraHeight",
						"seed",
						"t"
					]
				]
			],
			[
				13,
				"w3",
				[
					1,
					[
						1,
						[
							7,
							"a3",
							"y"
						],
						"*",
						[5, "0.3"]
					],
					"*",
					"g3"
				]
			],
			[
				2,
				"weightedT",
				"=",
				[
					1,
					"weightedT",
					"+",
					[
						1,
						[
							7,
							"a3",
							"x"
						],
						"*",
						"w3"
					]
				]
			],
			[
				2,
				"totalAlpha",
				"=",
				[
					1,
					"totalAlpha",
					"+",
					"w3"
				]
			],
			[
				13,
				"avgT",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						1,
						"weightedT",
						"/",
						[
							1,
							"totalAlpha",
							"+",
							[5, "0.001"]
						]
					], "balanceExp"]
				]
			],
			[
				13,
				"lowToMid",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[5, "0.2"],
						"avgT"
					]
				]
			],
			[
				13,
				"midToHigh",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.4"],
						[5, "0.85"],
						"avgT"
					]
				]
			],
			[
				13,
				"brightness",
				[
					1,
					"totalAlpha",
					"*",
					"intens"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					"lowToMid",
					"midToHigh",
					"brightness"
				]
			]]
		]],
		"externalNames": [
			"std",
			"auroraCurtain",
			"d"
		]
	},
	externals: () => ({
		std,
		auroraCurtain,
		d
	})
}) && $.f)({}));
const auroraCompose = tgpu.fn([d.vec4f, d.f32], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, brightness) => {
	"use gpu";
	const col = std.clamp(color.xyz.mul(brightness), d.vec3f(0), d.vec3f(1));
	const alpha = std.clamp(__tsover_mul(brightness, .5), 0, 1);
	return d.vec4f(col, alpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}, {
			"type": "i",
			"name": "brightness"
		}],
		"body": [0, [
			[
				13,
				"col",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								[
									7,
									"color",
									"xyz"
								],
								"mul"
							],
							["brightness"]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[5, "1"]]
						]
					]
				]
			],
			[
				13,
				"alpha",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"brightness",
							"*",
							[5, "0.5"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				["col", "alpha"]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Aurora",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Mesmerizing aurora borealis with layered curtains, vertical rays, and flowing light.",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#a533f8",
			transform: transformColor,
			description: "Edge color at the curtain base",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#22ee88",
			transform: transformColor,
			description: "Core color in the bright center",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		colorC: {
			default: "#1694e8",
			transform: transformColor,
			description: "Tip color at the ray ends",
			ui: {
				type: "color",
				label: "Color C",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		balance: {
			default: 50,
			description: "Shifts color distribution across the curtain height",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Balance",
				group: "Colors"
			}
		},
		intensity: {
			default: 80,
			description: "Overall aurora brightness",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Aurora"
			}
		},
		curtainCount: {
			default: 4,
			compileTime: true,
			description: "Number of aurora curtain layers",
			ui: {
				type: "range",
				min: 1,
				max: 4,
				step: 1,
				label: "Curtains",
				group: "Aurora"
			}
		},
		speed: {
			default: 5,
			description: "Animation speed",
			ui: {
				type: "range",
				min: -10,
				max: 10,
				step: .1,
				label: "Speed",
				group: "Aurora"
			}
		},
		waviness: {
			default: 50,
			description: "How much the curtains undulate",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 200,
				step: 1,
				label: "Waviness",
				group: "Aurora"
			}
		},
		rayDensity: {
			default: 20,
			description: "Density of vertical ray structures",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Ray Detail",
				group: "Aurora"
			}
		},
		height: {
			default: 120,
			description: "How tall the aurora extends",
			ui: {
				type: ["range", "map"],
				min: 10,
				max: 200,
				step: 1,
				label: "Height",
				group: "Scene"
			}
		},
		center: {
			default: {
				x: .5,
				y: 0
			},
			transform: transformPosition,
			description: "Center position of the aurora",
			ui: {
				type: "position",
				label: "Center",
				group: "Scene"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for variation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Aurora"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const f = call(auroraField, "auroraField", [
			params.uvContext ?? ctx.uv,
			params.effectiveViewportSize ?? ctx.viewportSize,
			animatedTime(params),
			uniforms.seed,
			uniforms.waviness,
			uniforms.rayDensity,
			uniforms.height,
			uniforms.center,
			uniforms.balance,
			uniforms.intensity,
			uniforms.curtainCount
		]);
		const lowToMid = f.member("x");
		const midToHigh = f.member("y");
		const brightness = f.member("z");
		const variant = mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear;
		return call(auroraCompose, "auroraCompose", [call(variant, "mixColors", [
			call(variant, "mixColors", [
				uniforms.colorA,
				uniforms.colorB,
				lowToMid
			]),
			uniforms.colorC,
			midToHigh
		]), brightness]);
	}
};
var Aurora_default = componentDefinition;
export { componentDefinition as a, auroraField as i, auroraCompose as n, auroraCurtain as r, Aurora_default as t };
