// shaders/dist/core/BarShift-_mdK3UWW.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-CWOT0Wqh.js";
import { r as transformAngle, s as transformEdges } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
var DEG_TO_RAD = .0174533;
const barHash = tgpu.fn([d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((x) => {
	"use gpu";
	return __tsover_mul(__tsover_sub(std.fract(__tsover_mul(std.sin(__tsover_mul(x, 12.9898)), 43758.5453)), .5), 2);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "x"
		}],
		"body": [0, [[10, [
			1,
			[
				1,
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								"x",
								"*",
								[5, "12.9898"]
							]]
						],
						"*",
						[5, "43758.5453"]
					]]
				],
				"-",
				[5, "0.5"]
			],
			"*",
			[5, "2"]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const barSampleUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, angle, offset) => {
	"use gpu";
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const cosA = std.cos(angleRad);
	const sinA = std.sin(angleRad);
	const au = __tsover_mul(uv.x, aspect);
	const av = uv.y;
	const cu = __tsover_sub(au, __tsover_mul(aspect, .5));
	const cv = __tsover_sub(av, .5);
	const projectedU = __tsover_add(__tsover_mul(cu, cosA), __tsover_mul(cv, sinA));
	const projectedV = __tsover_add(__tsover_mul(__tsover_mul(cu, sinA), -1), __tsover_mul(cv, cosA));
	const distPV = __tsover_add(projectedV, offset);
	const dcu = __tsover_sub(__tsover_mul(projectedU, cosA), __tsover_mul(distPV, sinA));
	const dcv = __tsover_add(__tsover_mul(projectedU, sinA), __tsover_mul(distPV, cosA));
	return d.vec2f(__tsover_add(__tsover_div(dcu, aspect), .5), __tsover_add(dcv, .5));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "offset"
			}
		],
		"body": [0, [
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"au",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"av",
				[
					7,
					"uv",
					"y"
				]
			],
			[
				13,
				"cu",
				[
					1,
					"au",
					"-",
					[
						1,
						"aspect",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"cv",
				[
					1,
					"av",
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"projectedU",
				[
					1,
					[
						1,
						"cu",
						"*",
						"cosA"
					],
					"+",
					[
						1,
						"cv",
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"projectedV",
				[
					1,
					[
						1,
						[
							1,
							"cu",
							"*",
							"sinA"
						],
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					],
					"+",
					[
						1,
						"cv",
						"*",
						"cosA"
					]
				]
			],
			[
				13,
				"distPV",
				[
					1,
					"projectedV",
					"+",
					"offset"
				]
			],
			[
				13,
				"dcu",
				[
					1,
					[
						1,
						"projectedU",
						"*",
						"cosA"
					],
					"-",
					[
						1,
						"distPV",
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"dcv",
				[
					1,
					[
						1,
						"projectedU",
						"*",
						"sinA"
					],
					"+",
					[
						1,
						"distPV",
						"*",
						"cosA"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						1,
						"dcu",
						"/",
						"aspect"
					],
					"+",
					[5, "0.5"]
				], [
					1,
					"dcv",
					"+",
					[5, "0.5"]
				]]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"std",
			"d"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		std,
		d
	})
}) && $.f)({}));
const barShiftUV = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((count, angle, intensity, seed, animTime, uv, aspect) => {
	"use gpu";
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const cosA = std.cos(angleRad);
	const sinA = std.sin(angleRad);
	const longestAxis = std.max(aspect, 1);
	const au = __tsover_mul(uv.x, aspect);
	const av = uv.y;
	const cu = __tsover_sub(au, __tsover_mul(aspect, .5));
	const cv = __tsover_sub(av, .5);
	const projectedU = __tsover_add(__tsover_mul(cu, cosA), __tsover_mul(cv, sinA));
	const barCoord = __tsover_div(__tsover_mul(projectedU, count), longestAxis);
	const barIndex = std.floor(barCoord);
	const seedOffset = __tsover_mul(seed, .137);
	const staticOffset = barHash(__tsover_add(barIndex, seedOffset));
	const barSpeed = barHash(__tsover_add(__tsover_add(barIndex, seedOffset), 42.7));
	const animOffset = __tsover_mul(__tsover_mul(animTime, barSpeed), .2);
	return barSampleUV(uv, aspect, angle, __tsover_mul(__tsover_add(staticOffset, animOffset), intensity));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "count"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"longestAxis",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["aspect", [5, "1"]]
				]
			],
			[
				13,
				"au",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"av",
				[
					7,
					"uv",
					"y"
				]
			],
			[
				13,
				"cu",
				[
					1,
					"au",
					"-",
					[
						1,
						"aspect",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"cv",
				[
					1,
					"av",
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"projectedU",
				[
					1,
					[
						1,
						"cu",
						"*",
						"cosA"
					],
					"+",
					[
						1,
						"cv",
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"barCoord",
				[
					1,
					[
						1,
						"projectedU",
						"*",
						"count"
					],
					"/",
					"longestAxis"
				]
			],
			[
				13,
				"barIndex",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["barCoord"]
				]
			],
			[
				13,
				"seedOffset",
				[
					1,
					"seed",
					"*",
					[5, "0.137"]
				]
			],
			[
				13,
				"staticOffset",
				[
					6,
					"barHash",
					[[
						1,
						"barIndex",
						"+",
						"seedOffset"
					]]
				]
			],
			[
				13,
				"barSpeed",
				[
					6,
					"barHash",
					[[
						1,
						[
							1,
							"barIndex",
							"+",
							"seedOffset"
						],
						"+",
						[5, "42.7"]
					]]
				]
			],
			[
				13,
				"animOffset",
				[
					1,
					[
						1,
						"animTime",
						"*",
						"barSpeed"
					],
					"*",
					[5, "0.2"]
				]
			],
			[
				13,
				"offset",
				[
					1,
					[
						1,
						"staticOffset",
						"+",
						"animOffset"
					],
					"*",
					"intensity"
				]
			],
			[10, [
				6,
				"barSampleUV",
				[
					"uv",
					"aspect",
					"angle",
					"offset"
				]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"std",
			"barHash",
			"barSampleUV"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		std,
		barHash,
		barSampleUV
	})
}) && $.f)({}));
const componentDefinition = {
	name: "BarShift",
	category: "Distortions",
	description: "Slices content into parallel bars, each offset independently for a fractured or glitch-like effect",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	animatedTime: { speed: "speed" },
	props: {
		count: {
			default: 6,
			description: "Number of bars across the longest viewport dimension",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Count",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Angle of bar orientation in degrees (0 = vertical bars, 90 = horizontal bars)",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		intensity: {
			default: .15,
			description: "Maximum displacement per bar",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Intensity",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Randomization seed for per-bar offset variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		speed: {
			default: 0,
			description: "Animation speed — each bar drifts at its own rate and direction",
			ui: {
				type: "range",
				min: -2,
				max: 2,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, childNode, ctx, propValues, convertToTexture } = params;
		if (!childNode) {
			console.error("You must pass a child component into the BarShift shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		const t = animatedTime(params);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(barShiftUV, "barShiftUV", [
			uniforms.count,
			uniforms.angle,
			uniforms.intensity,
			uniforms.seed,
			t,
			ctx.uv,
			ctx.aspect
		]), (uv) => texture.sample(uv), edgeMode)]);
	},
	uvRemap: (ctx) => {
		const edgeMode = ctx.propValues.edges ?? 0;
		const t = animatedTime(ctx);
		return composeEdgeRemapExpr(call(barShiftUV, "barShiftUV", [
			ctx.uniforms.count,
			ctx.uniforms.angle,
			ctx.uniforms.intensity,
			ctx.uniforms.seed,
			t,
			ctx.uv,
			ctx.aspect
		]), ctx.mask, edgeMode);
	}
};
var BarShift_default = componentDefinition;
export { componentDefinition as a, barShiftUV as i, barHash as n, barSampleUV as r, BarShift_default as t };
