// shaders/dist/core/Beam-DRdKJKir.js
import { $t as mixPreconvertedLinear, At as call, It as vec4, Kt as convertP3ToMixSpaceCPU, R as std, Zt as mixColorsLinear, en as mixPreconvertedVariants, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
const beamField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, vp, start, end, startThickness, endThickness, startSoftness, endSoftness) => {
	"use gpu";
	const aspect = __tsover_div(vp.x, vp.y);
	const startPos = d.vec2f(start.x, __tsover_sub(1, start.y));
	const endPos = d.vec2f(end.x, __tsover_sub(1, end.y));
	const acStart = d.vec2f(__tsover_mul(startPos.x, aspect), startPos.y);
	const acEnd = d.vec2f(__tsover_mul(endPos.x, aspect), endPos.y);
	const acUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const lineVec = d.vec2f(__tsover_sub(acEnd.x, acStart.x), __tsover_sub(acEnd.y, acStart.y));
	const toPoint = d.vec2f(__tsover_sub(acUV.x, acStart.x), __tsover_sub(acUV.y, acStart.y));
	const dotProduct = __tsover_add(__tsover_mul(toPoint.x, lineVec.x), __tsover_mul(toPoint.y, lineVec.y));
	const lineLengthSq = __tsover_add(__tsover_mul(lineVec.x, lineVec.x), __tsover_mul(lineVec.y, lineVec.y));
	const t = std.clamp(__tsover_div(dotProduct, std.max(lineLengthSq, 1e-4)), 0, 1);
	const closestPoint = d.vec2f(__tsover_add(acStart.x, __tsover_mul(lineVec.x, t)), __tsover_add(acStart.y, __tsover_mul(lineVec.y, t)));
	const distVec = d.vec2f(__tsover_sub(acUV.x, closestPoint.x), __tsover_sub(acUV.y, closestPoint.y));
	const distance = std.sqrt(__tsover_add(__tsover_mul(distVec.x, distVec.x), __tsover_mul(distVec.y, distVec.y)));
	const startThick = __tsover_mul(startThickness, .25);
	const endThick = __tsover_mul(endThickness, .25);
	const thickness = std.mix(startThick, endThick, t);
	const softness = std.mix(startSoftness, endSoftness, t);
	const normalizedDist = __tsover_div(distance, std.max(thickness, 1e-4));
	const edgeEnd = __tsover_add(1, softness);
	const alphaBase = __tsover_sub(1, std.smoothstep(1, edgeEnd, normalizedDist));
	const softnessFactor = __tsover_add(1, __tsover_mul(softness, 1.5));
	const alpha = std.pow(alphaBase, softnessFactor);
	const colorGradientStart = __tsover_sub(1, softness);
	const colorGradientEnd = __tsover_add(1, softness);
	const colorT = std.smoothstep(colorGradientStart, colorGradientEnd, normalizedDist);
	return d.vec2f(colorT, alpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "vp"
			},
			{
				"type": "i",
				"name": "start"
			},
			{
				"type": "i",
				"name": "end"
			},
			{
				"type": "i",
				"name": "startThickness"
			},
			{
				"type": "i",
				"name": "endThickness"
			},
			{
				"type": "i",
				"name": "startSoftness"
			},
			{
				"type": "i",
				"name": "endSoftness"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"vp",
						"x"
					],
					"/",
					[
						7,
						"vp",
						"y"
					]
				]
			],
			[
				13,
				"startPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"start",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"start",
							"y"
						]
					]]
				]
			],
			[
				13,
				"endPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"end",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"end",
							"y"
						]
					]]
				]
			],
			[
				13,
				"acStart",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"startPos",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"startPos",
						"y"
					]]
				]
			],
			[
				13,
				"acEnd",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"endPos",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"endPos",
						"y"
					]]
				]
			],
			[
				13,
				"acUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"lineVec",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"acEnd",
							"x"
						],
						"-",
						[
							7,
							"acStart",
							"x"
						]
					], [
						1,
						[
							7,
							"acEnd",
							"y"
						],
						"-",
						[
							7,
							"acStart",
							"y"
						]
					]]
				]
			],
			[
				13,
				"toPoint",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"acUV",
							"x"
						],
						"-",
						[
							7,
							"acStart",
							"x"
						]
					], [
						1,
						[
							7,
							"acUV",
							"y"
						],
						"-",
						[
							7,
							"acStart",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dotProduct",
				[
					1,
					[
						1,
						[
							7,
							"toPoint",
							"x"
						],
						"*",
						[
							7,
							"lineVec",
							"x"
						]
					],
					"+",
					[
						1,
						[
							7,
							"toPoint",
							"y"
						],
						"*",
						[
							7,
							"lineVec",
							"y"
						]
					]
				]
			],
			[
				13,
				"lineLengthSq",
				[
					1,
					[
						1,
						[
							7,
							"lineVec",
							"x"
						],
						"*",
						[
							7,
							"lineVec",
							"x"
						]
					],
					"+",
					[
						1,
						[
							7,
							"lineVec",
							"y"
						],
						"*",
						[
							7,
							"lineVec",
							"y"
						]
					]
				]
			],
			[
				13,
				"t",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"dotProduct",
							"/",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["lineLengthSq", [5, "0.0001"]]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"closestPoint",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"acStart",
							"x"
						],
						"+",
						[
							1,
							[
								7,
								"lineVec",
								"x"
							],
							"*",
							"t"
						]
					], [
						1,
						[
							7,
							"acStart",
							"y"
						],
						"+",
						[
							1,
							[
								7,
								"lineVec",
								"y"
							],
							"*",
							"t"
						]
					]]
				]
			],
			[
				13,
				"distVec",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"acUV",
							"x"
						],
						"-",
						[
							7,
							"closestPoint",
							"x"
						]
					], [
						1,
						[
							7,
							"acUV",
							"y"
						],
						"-",
						[
							7,
							"closestPoint",
							"y"
						]
					]]
				]
			],
			[
				13,
				"distance",
				[
					6,
					[
						7,
						"std",
						"sqrt"
					],
					[[
						1,
						[
							1,
							[
								7,
								"distVec",
								"x"
							],
							"*",
							[
								7,
								"distVec",
								"x"
							]
						],
						"+",
						[
							1,
							[
								7,
								"distVec",
								"y"
							],
							"*",
							[
								7,
								"distVec",
								"y"
							]
						]
					]]
				]
			],
			[
				13,
				"startThick",
				[
					1,
					"startThickness",
					"*",
					[5, "0.25"]
				]
			],
			[
				13,
				"endThick",
				[
					1,
					"endThickness",
					"*",
					[5, "0.25"]
				]
			],
			[
				13,
				"thickness",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"startThick",
						"endThick",
						"t"
					]
				]
			],
			[
				13,
				"softness",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"startSoftness",
						"endSoftness",
						"t"
					]
				]
			],
			[
				13,
				"normalizedDist",
				[
					1,
					"distance",
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						["thickness", [5, "0.0001"]]
					]
				]
			],
			[
				13,
				"edgeEnd",
				[
					1,
					[5, "1"],
					"+",
					"softness"
				]
			],
			[
				13,
				"alphaBase",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "1"],
							"edgeEnd",
							"normalizedDist"
						]
					]
				]
			],
			[
				13,
				"softnessFactor",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						"softness",
						"*",
						[5, "1.5"]
					]
				]
			],
			[
				13,
				"alpha",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					["alphaBase", "softnessFactor"]
				]
			],
			[
				13,
				"colorGradientStart",
				[
					1,
					[5, "1"],
					"-",
					"softness"
				]
			],
			[
				13,
				"colorGradientEnd",
				[
					1,
					[5, "1"],
					"+",
					"softness"
				]
			],
			[
				13,
				"colorT",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						"colorGradientStart",
						"colorGradientEnd",
						"normalizedDist"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["colorT", "alpha"]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Beam",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "A beam of light from one point to another.",
	acceptsUVContext: true,
	extraFields: {
		convA: {
			schema: d.vec3f,
			initial: [
				0,
				0,
				0
			]
		},
		convB: {
			schema: d.vec3f,
			initial: [
				0,
				0,
				0
			]
		}
	},
	props: {
		startPosition: {
			default: {
				x: .2,
				y: .5
			},
			transform: transformPosition,
			description: "Starting point of the beam",
			ui: {
				type: "position",
				label: "Start Position",
				group: "Position"
			}
		},
		endPosition: {
			default: {
				x: .8,
				y: .5
			},
			transform: transformPosition,
			description: "Ending point of the beam",
			ui: {
				type: "position",
				label: "End Position",
				group: "Position"
			}
		},
		startThickness: {
			default: .2,
			description: "Thickness at the start of the beam",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "Start Thickness",
				group: "Effect"
			}
		},
		endThickness: {
			default: .2,
			description: "Thickness at the end of the beam",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "End Thickness",
				group: "Effect"
			}
		},
		startSoftness: {
			default: .5,
			description: "Edge softness at the start of the beam",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 50,
				step: .1,
				label: "Start Softness",
				group: "Effect"
			}
		},
		endSoftness: {
			default: .5,
			description: "Edge softness at the end of the beam",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .1,
				label: "End Softness",
				group: "Effect"
			}
		},
		insideColor: {
			default: "#FF0000",
			transform: transformColor,
			description: "Color at the center of the beam",
			ui: {
				type: "color",
				label: "Inside Color",
				group: "Colors"
			}
		},
		outsideColor: {
			default: "#0000FF",
			transform: transformColor,
			description: "Color at the edges of the beam",
			ui: {
				type: "color",
				label: "Outside Color",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const uv = params.uvContext ?? ctx.uv;
		const vp = params.effectiveViewportSize ?? ctx.viewportSize;
		const mode = propValues.colorSpace ?? 0;
		const field = call(beamField, "beamField", [
			uv,
			vp,
			uniforms.startPosition,
			uniforms.endPosition,
			uniforms.startThickness,
			uniforms.endThickness,
			uniforms.startSoftness,
			uniforms.endSoftness
		]);
		const colorT = field.member("x");
		const alpha = field.member("y");
		let beamColorRGB;
		if (mode !== 0) {
			let lastKey = "";
			params.onBeforeRender(() => {
				const a = params.getCpuValue("insideColor");
				const b = params.getCpuValue("outsideColor");
				if (!a || !b) return;
				const key = `${a.x},${a.y},${a.z}|${b.x},${b.y},${b.z}`;
				if (key === lastKey) return;
				lastKey = key;
				const ca = convertP3ToMixSpaceCPU(a.x, a.y, a.z, mode);
				const cb = convertP3ToMixSpaceCPU(b.x, b.y, b.z, mode);
				params.setExtraField("convA", [
					ca[0],
					ca[1],
					ca[2]
				]);
				params.setExtraField("convB", [
					cb[0],
					cb[1],
					cb[2]
				]);
			});
			beamColorRGB = call(mixPreconvertedVariants[mode] ?? mixPreconvertedLinear, "mixPreconvertedColors", [
				uniforms.convA,
				uniforms.convB,
				uniforms.insideColor.member("a"),
				uniforms.outsideColor.member("a"),
				colorT
			]).member("rgb");
		} else beamColorRGB = call(mixColorsLinear, "mixColors", [
			uniforms.insideColor,
			uniforms.outsideColor,
			colorT
		]).member("rgb");
		const avgAlpha = uniforms.insideColor.member("a").add(uniforms.outsideColor.member("a")).mul(.5);
		return vec4(beamColorRGB, avgAlpha.mul(alpha));
	}
};
var Beam_default = componentDefinition;
export { beamField as n, componentDefinition as r, Beam_default as t };
