// shaders/dist/core/BlockNoise-CA933kae.js
import { At as call, R as std, W as toneAndColorExpr, n as tgpu, on as d, rt as noise_exports, ut as value13 } from "./coords-BlwgBpJf.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
import { i as noiseToneProps, r as noiseSpeedProp, t as noiseColorProps } from "./noiseColor-Cr76BYLN.js";
const blockNoiseField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, scale, seed, animTime) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const pos = d.vec2f(__tsover_mul(uv.x, aspect), uv.y).mul(std.exp(scale)).add(seed);
	return value13(d.vec3f(pos.x, pos.y, __tsover_mul(animTime, .3)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"pos",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"aspectUV",
								"mul"
							],
							[[
								6,
								[
									7,
									"std",
									"exp"
								],
								["scale"]
							]]
						],
						"add"
					],
					["seed"]
				]
			],
			[10, [
				6,
				[
					7,
					"noise",
					"value13"
				],
				[[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"pos",
							"x"
						],
						[
							7,
							"pos",
							"y"
						],
						[
							1,
							"animTime",
							"*",
							[5, "0.3"]
						]
					]
				]]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"noise"
		]
	},
	externals: () => ({
		d,
		std,
		noise: noise_exports
	})
}) && $.f)({}));
const componentDefinition = {
	name: "BlockNoise",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Blocky value noise with soft cells that morph over time",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		...noiseColorProps(),
		scale: {
			default: 2,
			description: "Pattern scale (higher = larger patterns)",
			ui: {
				type: "range",
				min: -2,
				max: 5,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		...noiseToneProps(),
		...noiseSpeedProp()
	},
	fragment: (params) => {
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		return toneAndColorExpr(params, call(blockNoiseField, "blockNoiseField", [
			uv,
			viewport,
			params.uniforms.scale,
			params.uniforms.seed,
			t
		]));
	}
};
var BlockNoise_default = componentDefinition;
export { blockNoiseField as n, componentDefinition as r, BlockNoise_default as t };
