// shaders/dist/core/BlueNoise-D-WjLdgK.js
import { At as call, G as blue12, R as std, W as toneAndColorExpr, n as tgpu, on as d, rt as noise_exports } from "./coords-CWOT0Wqh.js";
import { i as noiseToneProps, t as noiseColorProps } from "./noiseColor-WSY8YuLZ.js";
const blueNoiseField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, grain, seed) => {
	"use gpu";
	return blue12(std.floor(uv.mul(viewport).div(grain)).add(seed));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "grain"
			},
			{
				"type": "i",
				"name": "seed"
			}
		],
		"body": [0, [[
			13,
			"cell",
			[
				6,
				[
					7,
					[
						6,
						[
							7,
							"std",
							"floor"
						],
						[[
							6,
							[
								7,
								[
									6,
									[
										7,
										"uv",
										"mul"
									],
									["viewport"]
								],
								"div"
							],
							["grain"]
						]]
					],
					"add"
				],
				["seed"]
			]
		], [10, [
			6,
			[
				7,
				"noise",
				"blue12"
			],
			["cell"]
		]]]],
		"externalNames": ["std", "noise"]
	},
	externals: () => ({
		std,
		noise: noise_exports
	})
}) && $.f)({}));
const componentDefinition = {
	name: "BlueNoise",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "High-frequency blue noise — even, grainy speckle ideal for dithering",
	acceptsUVContext: true,
	props: {
		...noiseColorProps(),
		grain: {
			default: 1,
			description: "Grain size in pixels (1 = per-pixel, higher = chunkier speckle)",
			ui: {
				type: "range",
				min: 1,
				max: 32,
				step: 1,
				label: "Grain Size",
				group: "Effect"
			}
		},
		...noiseToneProps()
	},
	fragment: (params) => {
		return toneAndColorExpr(params, call(blueNoiseField, "blueNoiseField", [
			params.uvContext ?? params.ctx.uv,
			params.effectiveViewportSize ?? params.ctx.viewportSize,
			params.uniforms.grain,
			params.uniforms.seed
		]));
	}
};
var BlueNoise_default = componentDefinition;
export { blueNoiseField as n, componentDefinition as r, BlueNoise_default as t };
