// shaders/dist/core/Blur-l7HgDuUW.js
import { At as call, B as createGuardedCompute, I as createGaussianBlurCompute, It as vec4, L as createVariableGaussianBlurCompute, N as BLUR_MAP_FORMAT, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
const INTENSITY_TO_RADIUS = .36;
const intensityToRadius = (intensity) => intensity * INTENSITY_TO_RADIUS;
function fillMapParams() {
	return d.struct({
		inputWidth: d.f32,
		inputHeight: d.f32,
		inputMin: d.f32,
		inputMax: d.f32,
		outputMin: d.f32,
		outputMax: d.f32,
		curve: d.f32
	});
}
function buildFillBlurMapGraph(computeWidth, computeHeight, channel) {
	const Params = fillMapParams();
	const layout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
		source: { texture: d.texture2d(d.f32) },
		blurMap: { storageTexture: d.textureStorage2d(BLUR_MAP_FORMAT, "write-only") },
		params: { uniform: Params }
	}), "layout");
	return {
		layout,
		kernel: tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			const p = layout.$.params;
			const u = __tsover_div(__tsover_add(d.f32(cx), .5), computeWidth);
			const v = __tsover_div(__tsover_add(d.f32(cy), .5), computeHeight);
			const srcX = std.clamp(d.i32(std.round(__tsover_mul(u, p.inputWidth))), 0, d.i32(__tsover_sub(p.inputWidth, 1)));
			const srcY = std.clamp(d.i32(std.round(__tsover_mul(v, p.inputHeight))), 0, d.i32(__tsover_sub(p.inputHeight, 1)));
			const sample = std.textureLoad(layout.$.source, d.vec2u(d.u32(srcX), d.u32(srcY)), 0);
			const raw = channel === "alpha" ? sample.w : channel === "alphaInverted" ? __tsover_sub(1, sample.w) : channel === "luminanceInverted" ? __tsover_sub(1, std.dot(sample.xyz, d.vec3f(.2126, .7152, .0722))) : std.dot(sample.xyz, d.vec3f(.2126, .7152, .0722));
			const inputRange = std.max(__tsover_sub(p.inputMax, p.inputMin), 1e-4);
			const normalised = std.clamp(__tsover_div(__tsover_sub(raw, p.inputMin), inputRange), 0, 1);
			const exponent = std.pow(2, __tsover_mul(-p.curve, 2));
			const eased = std.pow(normalised, exponent);
			const remapped = std.mix(p.outputMin, p.outputMax, eased);
			const radius = __tsover_mul(remapped, .36);
			std.textureStore(layout.$.blurMap, d.vec2u(cx, cy), d.vec4f(radius, 0, 0, 1));
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [
					[
						13,
						"p",
						[
							7,
							[
								7,
								"layout",
								"$"
							],
							"params"
						]
					],
					[
						13,
						"u",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									["cx"]
								],
								"+",
								[5, "0.5"]
							],
							"/",
							"computeWidth"
						]
					],
					[
						13,
						"v",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									["cy"]
								],
								"+",
								[5, "0.5"]
							],
							"/",
							"computeHeight"
						]
					],
					[
						13,
						"srcX",
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								[
									6,
									[
										7,
										"d",
										"i32"
									],
									[[
										6,
										[
											7,
											"std",
											"round"
										],
										[[
											1,
											"u",
											"*",
											[
												7,
												"p",
												"inputWidth"
											]
										]]
									]]
								],
								[5, "0"],
								[
									6,
									[
										7,
										"d",
										"i32"
									],
									[[
										1,
										[
											7,
											"p",
											"inputWidth"
										],
										"-",
										[5, "1"]
									]]
								]
							]
						]
					],
					[
						13,
						"srcY",
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								[
									6,
									[
										7,
										"d",
										"i32"
									],
									[[
										6,
										[
											7,
											"std",
											"round"
										],
										[[
											1,
											"v",
											"*",
											[
												7,
												"p",
												"inputHeight"
											]
										]]
									]]
								],
								[5, "0"],
								[
									6,
									[
										7,
										"d",
										"i32"
									],
									[[
										1,
										[
											7,
											"p",
											"inputHeight"
										],
										"-",
										[5, "1"]
									]]
								]
							]
						]
					],
					[
						13,
						"sample",
						[
							6,
							[
								7,
								"std",
								"textureLoad"
							],
							[
								[
									7,
									[
										7,
										"layout",
										"$"
									],
									"source"
								],
								[
									6,
									[
										7,
										"d",
										"vec2u"
									],
									[[
										6,
										[
											7,
											"d",
											"u32"
										],
										["srcX"]
									], [
										6,
										[
											7,
											"d",
											"u32"
										],
										["srcY"]
									]]
								],
								[5, "0"]
							]
						]
					],
					[
						13,
						"raw",
						[
							105,
							[
								1,
								"channel",
								"===",
								[103, "alpha"]
							],
							[
								7,
								"sample",
								"w"
							],
							[
								105,
								[
									1,
									"channel",
									"===",
									[103, "alphaInverted"]
								],
								[
									1,
									[5, "1"],
									"-",
									[
										7,
										"sample",
										"w"
									]
								],
								[
									105,
									[
										1,
										"channel",
										"===",
										[103, "luminanceInverted"]
									],
									[
										1,
										[5, "1"],
										"-",
										[
											6,
											[
												7,
												"std",
												"dot"
											],
											[[
												7,
												"sample",
												"xyz"
											], [
												6,
												[
													7,
													"d",
													"vec3f"
												],
												[
													[5, "0.2126"],
													[5, "0.7152"],
													[5, "0.0722"]
												]
											]]
										]
									],
									[
										6,
										[
											7,
											"std",
											"dot"
										],
										[[
											7,
											"sample",
											"xyz"
										], [
											6,
											[
												7,
												"d",
												"vec3f"
											],
											[
												[5, "0.2126"],
												[5, "0.7152"],
												[5, "0.0722"]
											]
										]]
									]
								]
							]
						]
					],
					[
						13,
						"inputRange",
						[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								1,
								[
									7,
									"p",
									"inputMax"
								],
								"-",
								[
									7,
									"p",
									"inputMin"
								]
							], [5, "0.0001"]]
						]
					],
					[
						13,
						"normalised",
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								[
									1,
									[
										1,
										"raw",
										"-",
										[
											7,
											"p",
											"inputMin"
										]
									],
									"/",
									"inputRange"
								],
								[5, "0"],
								[5, "1"]
							]
						]
					],
					[
						13,
						"exponent",
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							[[5, "2"], [
								1,
								[
									4,
									"-",
									[
										7,
										"p",
										"curve"
									]
								],
								"*",
								[5, "2"]
							]]
						]
					],
					[
						13,
						"eased",
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							["normalised", "exponent"]
						]
					],
					[
						13,
						"remapped",
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									7,
									"p",
									"outputMin"
								],
								[
									7,
									"p",
									"outputMax"
								],
								"eased"
							]
						]
					],
					[
						13,
						"radius",
						[
							1,
							"remapped",
							"*",
							[5, "0.36"]
						]
					],
					[
						6,
						[
							7,
							"std",
							"textureStore"
						],
						[
							[
								7,
								[
									7,
									"layout",
									"$"
								],
								"blurMap"
							],
							[
								6,
								[
									7,
									"d",
									"vec2u"
								],
								["cx", "cy"]
							],
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									"radius",
									[5, "0"],
									[5, "0"],
									[5, "1"]
								]
							]
						]
					]
				]],
				"externalNames": [
					"layout",
					"d",
					"computeWidth",
					"computeHeight",
					"std",
					"channel"
				]
			},
			externals: () => ({
				layout,
				d,
				computeWidth,
				computeHeight,
				std,
				channel
			})
		}) && $.f)({})).$name("fillBlurMap"),
		Params
	};
}
const componentDefinition = {
	name: "Blur",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Blurs",
	description: "A simple Gaussian blur effect",
	requiresRTT: true,
	requiresChild: true,
	props: { intensity: {
		default: 50,
		description: "Intensity of the blur effect",
		ui: {
			type: ["range", "map"],
			min: 0,
			max: 200,
			step: 1,
			label: "Intensity",
			group: "Effect"
		}
	} },
	compute: (params) => {
		const { childNode, gpu, convertToTexture, registerComputeTexture, getCpuValue, getMapInfo, onCleanup, onResize, dimensions } = params;
		if (!childNode) return null;
		const root = gpu?.root;
		if (!root) return null;
		const childTexture = convertToTexture(childNode);
		let curWidth = Math.max(1, Math.round(dimensions.width));
		let curHeight = Math.max(1, Math.round(dimensions.height));
		const mapInfo = getMapInfo("intensity");
		if (mapInfo) {
			const variable = createVariableGaussianBlurCompute(root, null, curWidth, curHeight, onCleanup);
			const blurredTexture$1 = registerComputeTexture(variable.outputTexture);
			const graph = buildFillBlurMapGraph(variable.computeWidth, variable.computeHeight, mapInfo.channel);
			const fillParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(graph.Params), "fillParams");
			let fillStep = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
				"use gpu";
				graph.kernel(cx, cy);
			}), {
				v: 1,
				name: void 0,
				ast: {
					"params": [{
						"type": "i",
						"name": "cx"
					}, {
						"type": "i",
						"name": "cy"
					}],
					"body": [0, [[
						6,
						[
							7,
							"graph",
							"kernel"
						],
						["cx", "cy"]
					]]],
					"externalNames": ["graph"]
				},
				externals: () => ({ graph })
			}) && $.f)({}), { size: [variable.computeWidth, variable.computeHeight] });
			onResize(({ width, height }) => {
				curWidth = Math.max(1, Math.round(width));
				curHeight = Math.max(1, Math.round(height));
				variable.setInputDimensions(curWidth, curHeight);
			});
			return {
				outputs: {
					childTexture,
					blurredTexture: blurredTexture$1
				},
				bindInputs: (resolve) => {
					const src = resolve(childTexture.key);
					if (src) variable.setInputTexture(src.texture);
					const mapSrc = resolve(mapInfo.sourceTexture.key);
					if (mapSrc) {
						const bg = root.createBindGroup(graph.layout, {
							source: mapSrc.texture,
							blurMap: variable.blurMapTexture,
							params: fillParams.buffer
						});
						fillStep = fillStep.with(bg);
					}
				},
				getComputeNodes: () => {
					const w = mapInfo.window();
					fillParams.write({
						inputWidth: curWidth,
						inputHeight: curHeight,
						inputMin: w.inputMin,
						inputMax: w.inputMax,
						outputMin: w.outputMin,
						outputMax: w.outputMax,
						curve: w.curve
					});
					return [fillStep, ...variable.computeSteps];
				}
			};
		}
		const gaussian = createGaussianBlurCompute(root, null, curWidth, curHeight, onCleanup);
		const blurredTexture = registerComputeTexture(gaussian.outputTexture);
		onResize(({ width, height }) => gaussian.setInputDimensions(Math.max(1, Math.round(width)), Math.max(1, Math.round(height))));
		return {
			outputs: {
				childTexture,
				blurredTexture
			},
			bindInputs: (resolve) => {
				const src = resolve(childTexture.key);
				if (src) gaussian.setInputTexture(src.texture);
			},
			getComputeNodes: () => {
				const raw = getCpuValue("intensity");
				if (typeof raw === "number") gaussian.updateRadius(intensityToRadius(raw));
				return gaussian.computeSteps;
			}
		};
	},
	fragment: ({ childNode, computeOutputs, ctx, convertToTexture }) => {
		if (!childNode) return ZERO;
		const blurred = computeOutputs?.blurredTexture;
		const sharp = computeOutputs?.childTexture;
		if (!blurred || !sharp) return call(unpremultiplyAlpha, "unpremultiplyAlpha", [convertToTexture(childNode).sample(ctx.uv)]);
		const sharpSample = sharp.sample(ctx.uv);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [vec4(blurred.sample(ctx.uv).member("rgb"), sharpSample.member("a"))]);
	}
};
var Blur_default = componentDefinition;
export { intensityToRadius as a, componentDefinition as i, INTENSITY_TO_RADIUS as n, buildFillBlurMapGraph as r, Blur_default as t };
