// shaders/dist/core/BrickPattern-DPez_hvr.js
import { At as call, It as vec4, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
var DEG_TO_RAD = .0174533;
const brickField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, cellsX, cellsY, mortar, softness, variation, rotation, offset, speedVariance, seed, animTime) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const au = __tsover_mul(uv.x, aspect);
	const av = uv.y;
	const cu = __tsover_sub(au, __tsover_mul(aspect, .5));
	const cv = __tsover_sub(av, .5);
	const rotRad = __tsover_mul(rotation, DEG_TO_RAD);
	const cosR = std.cos(rotRad);
	const sinR = std.sin(rotRad);
	const ru = __tsover_add(__tsover_mul(cu, cosR), __tsover_mul(cv, sinR));
	const rv = __tsover_add(-__tsover_mul(cu, sinR), __tsover_mul(cv, cosR));
	const u = __tsover_add(__tsover_div(ru, aspect), .5);
	const v = __tsover_add(rv, .5);
	const mortarU = __tsover_mul(mortar, .3);
	const mortarV = __tsover_div(__tsover_mul(__tsover_mul(mortarU, aspect), cellsY), cellsX);
	const row = std.floor(__tsover_mul(v, cellsY));
	const rowMod2 = __tsover_sub(row, __tsover_mul(2, std.floor(__tsover_div(row, 2))));
	const rowOffset = __tsover_mul(rowMod2, .5);
	const rowDir = __tsover_sub(__tsover_mul(rowMod2, 2), 1);
	const rowHash = __tsover_mul(__tsover_sub(std.fract(__tsover_mul(std.sin(__tsover_add(row, __tsover_mul(seed, .137))), 43758.5453)), .5), 2);
	const rowSpeedMult = std.mix(rowDir, rowHash, speedVariance);
	const brickArg = __tsover_add(__tsover_add(__tsover_add(__tsover_mul(u, cellsX), rowOffset), __tsover_mul(animTime, rowSpeedMult)), offset);
	const vScaled = __tsover_mul(v, cellsY);
	const cellU = std.fract(brickArg);
	const cellV = std.fract(vScaled);
	const soft = __tsover_mul(softness, .5);
	const aaU = __tsover_add(std.fwidth(brickArg), soft);
	const aaV = __tsover_add(std.fwidth(vScaled), soft);
	const brickU = __tsover_mul(std.smoothstep(__tsover_sub(mortarU, aaU), __tsover_add(mortarU, aaU), cellU), std.smoothstep(__tsover_sub(mortarU, aaU), __tsover_add(mortarU, aaU), __tsover_sub(1, cellU)));
	const brickV = __tsover_mul(std.smoothstep(__tsover_sub(mortarV, aaV), __tsover_add(mortarV, aaV), cellV), std.smoothstep(__tsover_sub(mortarV, aaV), __tsover_add(mortarV, aaV), __tsover_sub(1, cellV)));
	const brickMask = __tsover_mul(brickU, brickV);
	const brickId = d.vec2f(std.floor(brickArg), row);
	const q0 = std.fract(brickId.mul(d.vec2f(123.34, 345.45)));
	const q1 = q0.add(std.dot(q0, q0.add(34.345)));
	const rand = std.fract(__tsover_mul(q1.x, q1.y));
	const variedFactor = std.max(__tsover_add(1, __tsover_mul(__tsover_mul(__tsover_sub(rand, .5), 2), variation)), 0);
	return d.vec2f(brickMask, variedFactor);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "cellsX"
			},
			{
				"type": "i",
				"name": "cellsY"
			},
			{
				"type": "i",
				"name": "mortar"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "variation"
			},
			{
				"type": "i",
				"name": "rotation"
			},
			{
				"type": "i",
				"name": "offset"
			},
			{
				"type": "i",
				"name": "speedVariance"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"au",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"av",
				[
					7,
					"uv",
					"y"
				]
			],
			[
				13,
				"cu",
				[
					1,
					"au",
					"-",
					[
						1,
						"aspect",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"cv",
				[
					1,
					"av",
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"rotRad",
				[
					1,
					"rotation",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["rotRad"]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["rotRad"]
				]
			],
			[
				13,
				"ru",
				[
					1,
					[
						1,
						"cu",
						"*",
						"cosR"
					],
					"+",
					[
						1,
						"cv",
						"*",
						"sinR"
					]
				]
			],
			[
				13,
				"rv",
				[
					1,
					[
						4,
						"-",
						[
							1,
							"cu",
							"*",
							"sinR"
						]
					],
					"+",
					[
						1,
						"cv",
						"*",
						"cosR"
					]
				]
			],
			[
				13,
				"u",
				[
					1,
					[
						1,
						"ru",
						"/",
						"aspect"
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"v",
				[
					1,
					"rv",
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"mortarU",
				[
					1,
					"mortar",
					"*",
					[5, "0.3"]
				]
			],
			[
				13,
				"mortarV",
				[
					1,
					[
						1,
						[
							1,
							"mortarU",
							"*",
							"aspect"
						],
						"*",
						"cellsY"
					],
					"/",
					"cellsX"
				]
			],
			[
				13,
				"row",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						1,
						"v",
						"*",
						"cellsY"
					]]
				]
			],
			[
				13,
				"rowMod2",
				[
					1,
					"row",
					"-",
					[
						1,
						[5, "2"],
						"*",
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								1,
								"row",
								"/",
								[5, "2"]
							]]
						]
					]
				]
			],
			[
				13,
				"rowOffset",
				[
					1,
					"rowMod2",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"rowDir",
				[
					1,
					[
						1,
						"rowMod2",
						"*",
						[5, "2"]
					],
					"-",
					[5, "1"]
				]
			],
			[
				13,
				"rowHash",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										"row",
										"+",
										[
											1,
											"seed",
											"*",
											[5, "0.137"]
										]
									]]
								],
								"*",
								[5, "43758.5453"]
							]]
						],
						"-",
						[5, "0.5"]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"rowSpeedMult",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"rowDir",
						"rowHash",
						"speedVariance"
					]
				]
			],
			[
				13,
				"brickArg",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								"u",
								"*",
								"cellsX"
							],
							"+",
							"rowOffset"
						],
						"+",
						[
							1,
							"animTime",
							"*",
							"rowSpeedMult"
						]
					],
					"+",
					"offset"
				]
			],
			[
				13,
				"vScaled",
				[
					1,
					"v",
					"*",
					"cellsY"
				]
			],
			[
				13,
				"cellU",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["brickArg"]
				]
			],
			[
				13,
				"cellV",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["vScaled"]
				]
			],
			[
				13,
				"soft",
				[
					1,
					"softness",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"aaU",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fwidth"
						],
						["brickArg"]
					],
					"+",
					"soft"
				]
			],
			[
				13,
				"aaV",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fwidth"
						],
						["vScaled"]
					],
					"+",
					"soft"
				]
			],
			[
				13,
				"brickU",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"mortarU",
								"-",
								"aaU"
							],
							[
								1,
								"mortarU",
								"+",
								"aaU"
							],
							"cellU"
						]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"mortarU",
								"-",
								"aaU"
							],
							[
								1,
								"mortarU",
								"+",
								"aaU"
							],
							[
								1,
								[5, "1"],
								"-",
								"cellU"
							]
						]
					]
				]
			],
			[
				13,
				"brickV",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"mortarV",
								"-",
								"aaV"
							],
							[
								1,
								"mortarV",
								"+",
								"aaV"
							],
							"cellV"
						]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"mortarV",
								"-",
								"aaV"
							],
							[
								1,
								"mortarV",
								"+",
								"aaV"
							],
							[
								1,
								[5, "1"],
								"-",
								"cellV"
							]
						]
					]
				]
			],
			[
				13,
				"brickMask",
				[
					1,
					"brickU",
					"*",
					"brickV"
				]
			],
			[
				13,
				"brickId",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"floor"
						],
						["brickArg"]
					], "row"]
				]
			],
			[
				13,
				"q0",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						6,
						[
							7,
							"brickId",
							"mul"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "123.34"], [5, "345.45"]]
						]]
					]]
				]
			],
			[
				13,
				"q1",
				[
					6,
					[
						7,
						"q0",
						"add"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["q0", [
							6,
							[
								7,
								"q0",
								"add"
							],
							[[5, "34.345"]]
						]]
					]]
				]
			],
			[
				13,
				"rand",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							7,
							"q1",
							"x"
						],
						"*",
						[
							7,
							"q1",
							"y"
						]
					]]
				]
			],
			[
				13,
				"variedFactor",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[5, "1"],
						"+",
						[
							1,
							[
								1,
								[
									1,
									"rand",
									"-",
									[5, "0.5"]
								],
								"*",
								[5, "2"]
							],
							"*",
							"variation"
						]
					], [5, "0"]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["brickMask", "variedFactor"]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"std",
			"d"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "BrickPattern",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Classic brick wall pattern with alternating rows and mortar gaps",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorBrick: {
			default: "#000000",
			transform: transformColor,
			description: "Brick color",
			ui: {
				type: "color",
				label: "Brick Color",
				group: "Colors"
			}
		},
		colorMortar: {
			default: "#ffffff",
			transform: transformColor,
			description: "Mortar / gap color",
			ui: {
				type: "color",
				label: "Mortar Color",
				group: "Colors"
			}
		},
		cellsX: {
			default: 8,
			description: "Number of bricks per row",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Bricks per Row",
				group: "Effect"
			}
		},
		cellsY: {
			default: 10,
			description: "Number of brick rows",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Rows",
				group: "Effect"
			}
		},
		mortar: {
			default: .05,
			description: "Width of mortar gaps — equal pixel thickness in both directions",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Mortar Gap",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Softness of the brick edges (0 = crisp, 1 = very soft)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		variation: {
			default: 0,
			description: "Per-brick random lightening/darkening of the brick color",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Variation",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			transform: transformAngle,
			description: "Rotation of the pattern in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		speed: {
			default: 0,
			description: "Animation speed",
			ui: {
				type: "range",
				min: -2,
				max: 2,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		offset: {
			default: 0,
			description: "Static horizontal offset — shifts the brick pattern without animating",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Offset",
				group: "Animation"
			}
		},
		speedVariance: {
			default: 0,
			description: "How much each row's speed varies — at high values rows move at different speeds and directions",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Speed Variance",
				group: "Animation"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for per-row speed variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Animation"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, propValues } = params;
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const animTime = animatedTime(params);
		const field = call(brickField, "brickField", [
			uv,
			viewport,
			uniforms.cellsX,
			uniforms.cellsY,
			uniforms.mortar,
			uniforms.softness,
			uniforms.variation,
			uniforms.rotation,
			uniforms.offset,
			uniforms.speedVariance,
			uniforms.seed,
			animTime
		]);
		const brickMask = field.member("x");
		const variedFactor = field.member("y");
		const variedBrick = vec4(uniforms.colorBrick.member("rgb").mul(variedFactor), uniforms.colorBrick.member("a"));
		return call(mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			uniforms.colorMortar,
			variedBrick,
			brickMask
		]);
	}
};
var BrickPattern_default = componentDefinition;
export { brickField as n, componentDefinition as r, BrickPattern_default as t };
