// shaders/dist/core/BrightnessContrast-XvDV1ixv.js
import { At as call, Ot as ZERO, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
const brightnessContrastColor = tgpu.fn([
	d.vec4f,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, brightness, contrastRaw) => {
	"use gpu";
	const contrast = __tsover_add(contrastRaw, 1);
	const half = d.vec3f(.5, .5, .5);
	const result = color.xyz.sub(half).mul(contrast).add(half).add(d.vec3f(brightness, brightness, brightness));
	return d.vec4f(result, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "color"
			},
			{
				"type": "i",
				"name": "brightness"
			},
			{
				"type": "i",
				"name": "contrastRaw"
			}
		],
		"body": [0, [
			[
				13,
				"contrast",
				[
					1,
					"contrastRaw",
					"+",
					[5, "1"]
				]
			],
			[
				13,
				"half",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0.5"],
						[5, "0.5"],
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"contrasted",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										[
											7,
											"color",
											"xyz"
										],
										"sub"
									],
									["half"]
								],
								"mul"
							],
							["contrast"]
						],
						"add"
					],
					["half"]
				]
			],
			[
				13,
				"result",
				[
					6,
					[
						7,
						"contrasted",
						"add"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"brightness",
							"brightness",
							"brightness"
						]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				["result", [
					7,
					"color",
					"w"
				]]
			]]
		]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const componentDefinition = {
	name: "BrightnessContrast",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Adjust brightness and contrast of the image",
	requiresChild: true,
	props: {
		brightness: {
			default: 0,
			description: "Brightness adjustment (-1 to 1)",
			ui: {
				type: ["range", "map"],
				min: -1,
				max: 1,
				step: .01,
				label: "Brightness",
				group: "Adjustments"
			}
		},
		contrast: {
			default: 0,
			description: "Contrast adjustment (-1 to 1)",
			ui: {
				type: ["range", "map"],
				min: -1,
				max: 1,
				step: .01,
				label: "Contrast",
				group: "Adjustments"
			}
		}
	},
	fragment: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Brightness Contrast shader.");
			return ZERO;
		}
		return call(brightnessContrastColor, "brightnessContrastColor", [
			childNode,
			uniforms.brightness,
			uniforms.contrast
		]);
	}
};
var BrightnessContrast_default = componentDefinition;
export { brightnessContrastColor as n, componentDefinition as r, BrightnessContrast_default as t };
