// shaders/dist/core/Bulge-d93jKEwv.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-BlwgBpJf.js";
import { c as transformPosition, s as transformEdges } from "./transformations-fqzznt4g.js";
const bulgeUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, strength, radius, falloff, uv, aspect) => {
	"use gpu";
	const aspectCorrectedUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y));
	const delta = aspectCorrectedUV.sub(centerPos);
	const distance = std.length(delta);
	const effectRadius = __tsover_mul(radius, .5);
	const innerRadius = __tsover_mul(effectRadius, std.max(__tsover_sub(__tsover_sub(1, falloff), .001), 0));
	const smoothFalloff = __tsover_sub(1, std.smoothstep(innerRadius, effectRadius, distance));
	const normalizedDist = __tsover_div(distance, effectRadius);
	const distSq = __tsover_mul(normalizedDist, normalizedDist);
	const quadraticFalloff = std.max(0, __tsover_sub(1, distSq));
	const falloffTotal = __tsover_mul(smoothFalloff, quadraticFalloff);
	const negStrength = __tsover_mul(strength, -1);
	const displacementAmount = __tsover_mul(negStrength, falloffTotal);
	const scaleFactor = __tsover_add(1, displacementAmount);
	const bulgedDelta = delta.mul(scaleFactor);
	const bulgedUV = centerPos.add(bulgedDelta);
	return d.vec2f(__tsover_div(bulgedUV.x, aspect), bulgedUV.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "radius"
			},
			{
				"type": "i",
				"name": "falloff"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"aspectCorrectedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"delta",
				[
					6,
					[
						7,
						"aspectCorrectedUV",
						"sub"
					],
					["centerPos"]
				]
			],
			[
				13,
				"distance",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["delta"]
				]
			],
			[
				13,
				"effectRadius",
				[
					1,
					"radius",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"innerRadius",
				[
					1,
					"effectRadius",
					"*",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							1,
							[
								1,
								[5, "1"],
								"-",
								"falloff"
							],
							"-",
							[5, "0.001"]
						], [5, "0"]]
					]
				]
			],
			[
				13,
				"smoothFalloff",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							"innerRadius",
							"effectRadius",
							"distance"
						]
					]
				]
			],
			[
				13,
				"normalizedDist",
				[
					1,
					"distance",
					"/",
					"effectRadius"
				]
			],
			[
				13,
				"distSq",
				[
					1,
					"normalizedDist",
					"*",
					"normalizedDist"
				]
			],
			[
				13,
				"quadraticFalloff",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[5, "0"], [
						1,
						[5, "1"],
						"-",
						"distSq"
					]]
				]
			],
			[
				13,
				"falloffTotal",
				[
					1,
					"smoothFalloff",
					"*",
					"quadraticFalloff"
				]
			],
			[
				13,
				"negStrength",
				[
					1,
					"strength",
					"*",
					[
						4,
						"-",
						[5, "1"]
					]
				]
			],
			[
				13,
				"displacementAmount",
				[
					1,
					"negStrength",
					"*",
					"falloffTotal"
				]
			],
			[
				13,
				"scaleFactor",
				[
					1,
					[5, "1"],
					"+",
					"displacementAmount"
				]
			],
			[
				13,
				"bulgedDelta",
				[
					6,
					[
						7,
						"delta",
						"mul"
					],
					["scaleFactor"]
				]
			],
			[
				13,
				"bulgedUV",
				[
					6,
					[
						7,
						"centerPos",
						"add"
					],
					["bulgedDelta"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						7,
						"bulgedUV",
						"x"
					],
					"/",
					"aspect"
				], [
					7,
					"bulgedUV",
					"y"
				]]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Bulge",
	category: "Distortions",
	description: "Magnify or pinch content around a center point",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the bulge effect",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		strength: {
			default: 1,
			description: "The intensity of the bulge effect (positive = bulge out, negative = pinch in)",
			ui: {
				type: ["range", "map"],
				min: -1,
				max: 1,
				step: .01,
				label: "Strength",
				group: "Effect"
			}
		},
		radius: {
			default: 1,
			description: "The radius of the bulge effect area",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		falloff: {
			default: .5,
			description: "Controls the smoothness of the transition (0 = hard edge, 1 = very smooth)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Falloff",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, propValues, convertToTexture }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Bulge shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(bulgeUV, "bulgeUV", [
			uniforms.center,
			uniforms.strength,
			uniforms.radius,
			uniforms.falloff,
			ctx.uv,
			ctx.aspect
		]), (uv) => texture.sample(uv), edgeMode)]);
	},
	uvRemap: ({ uv, mask, uniforms, aspect, propValues }) => {
		const edgeMode = propValues.edges ?? 0;
		return composeEdgeRemapExpr(call(bulgeUV, "bulgeUV", [
			uniforms.center,
			uniforms.strength,
			uniforms.radius,
			uniforms.falloff,
			uv,
			aspect
		]), mask, edgeMode);
	}
};
var Bulge_default = componentDefinition;
export { bulgeUV as n, componentDefinition as r, Bulge_default as t };
