// shaders/dist/core/CarbonFiber-G7vIOz5W.js
import { At as call, It as vec4, Pt as floatE, R as std, T as sdf_exports, d as sdf3d_exports, et as mxNoiseFloat2, n as tgpu, on as d, rt as noise_exports } from "./coords-BlwgBpJf.js";
import { n as glass_exports } from "./glass-DdgMBWkF.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
import { t as shapeEffectBoundingBoxDeclaration } from "./shapeEffectBounds-BlxxsYMN.js";
var { sdfSpaceUV, offsetUV } = glass_exports;
var { createSvgSdfSampler, createAnalyticSdfSampler, driveAnalyticSubProps, ANALYTIC_SDF_EXTRA_FIELDS } = sdf_exports;
var { createVolumetricFieldComputeNode, buildVolumetricFieldSampler, VOLUMETRIC_FIELD_EXTRA_FIELDS, is3dShapeType, isVolumetricShapeType } = sdf3d_exports;
var DEG_TO_RAD = .017453292519943295;
var PI = Math.PI;
var TWO_PI_SEVEN = 2 * Math.PI * 7;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
function resolveShapeType(shapeType, shape) {
	if (shapeType) return shapeType;
	if (shape && typeof shape === "object") return shape.type || "circleSDF";
	if (typeof shape === "string") try {
		return JSON.parse(shape).type || "circleSDF";
	} catch {
		return "circleSDF";
	}
	return "circleSDF";
}
function buildCarbonFiberComposite(weaveStyle) {
	return tgpu.fn([
		d.vec4f,
		d.vec4f,
		d.vec4f,
		d.vec2f,
		d.vec2f,
		d.vec2f,
		d.vec3f,
		d.vec3f,
		d.vec4f,
		d.vec4f,
		d.vec4f
	], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((surf0, surfX, surfY, sdfUV, uv, viewportSize, lightColor, darkColor, p0, p1, p2) => {
		"use gpu";
		const scl = p0.x;
		const edgeSoftness = p0.y;
		const weaveScale = p0.z;
		const weaveAngle = p0.w;
		const relief = p1.x;
		const fiberSheen = p1.y;
		const roughness = p1.z;
		const clearcoat = p1.w;
		const environment = p2.x;
		const envRotation = p2.y;
		const lightAngle = p2.z;
		const volumetric = p2.w;
		const aspect = __tsover_div(viewportSize.x, viewportSize.y);
		const sharpEdge = std.max(__tsover_mul(edgeSoftness, .5), .001);
		const pxH = __tsover_div(1, viewportSize.y);
		const sdfRaw = surf0.x;
		const sdf = __tsover_div(sdfRaw, scl);
		if (sdf > __tsover_mul(pxH, 2)) return d.vec4f(0, 0, 0, 0);
		const lx = std.cos(__tsover_mul(lightAngle, DEG_TO_RAD));
		const ly = std.sin(__tsover_mul(lightAngle, DEG_TO_RAD));
		const wtx = std.cos(__tsover_mul(weaveAngle, DEG_TO_RAD));
		const wty = std.sin(__tsover_mul(weaveAngle, DEG_TO_RAD));
		const envC = std.cos(__tsover_mul(envRotation, DEG_TO_RAD));
		const envS = std.sin(__tsover_mul(envRotation, DEG_TO_RAD));
		const EPS = .01;
		const gradX = __tsover_div(__tsover_sub(surfX.x, sdfRaw), EPS);
		const gradY = __tsover_div(__tsover_sub(surfY.x, sdfRaw), EPS);
		const ddx = std.clamp(__tsover_div(__tsover_sub(surfX.w, surf0.w), EPS), -5, 5);
		const ddy = std.clamp(__tsover_div(__tsover_sub(surfY.w, surf0.w), EPS), -5, 5);
		const nVol = std.normalize(d.vec3f(ddx, ddy, -1));
		const rimT = std.clamp(__tsover_div(__tsover_mul(sdf, -1), .05), 0, 1);
		const rimBevel = __tsover_mul(__tsover_sub(1, rimT), 1.6);
		const nFlat = std.normalize(d.vec3f(__tsover_mul(gradX, rimBevel), __tsover_mul(gradY, rimBevel), -1));
		const nGeo = std.select(nFlat, nVol, volumetric > .5);
		const pdx = std.select(__tsover_sub(sdfUV.x, .5), surf0.y, volumetric > .5);
		const pdy = std.select(__tsover_sub(sdfUV.y, .5), surf0.z, volumetric > .5);
		const wbx = __tsover_mul(wty, -1);
		const wby = wtx;
		const warp = __tsover_add(__tsover_mul(pdx, wtx), __tsover_mul(pdy, wty));
		const weft = __tsover_add(__tsover_mul(pdx, wbx), __tsover_mul(pdy, wby));
		const wScale = weaveScale;
		const wA = __tsover_mul(warp, wScale);
		const wB = __tsover_mul(weft, wScale);
		const cellX = std.floor(wA);
		const cellY = std.floor(wB);
		const fA = __tsover_sub(std.fract(wA), .5);
		const fB = __tsover_sub(std.fract(wB), .5);
		let parity = d.f32(0);
		if (weaveStyle === "plain") {
			const s2 = __tsover_add(cellX, cellY);
			parity = __tsover_sub(s2, __tsover_mul(2, std.floor(__tsover_div(s2, 2))));
		} else {
			const s4 = __tsover_add(__tsover_sub(cellX, cellY), 64);
			const m4 = __tsover_sub(s4, __tsover_mul(4, std.floor(__tsover_div(s4, 4))));
			parity = std.select(d.f32(0), d.f32(1), m4 >= 2);
		}
		const fiberX = std.mix(wtx, wbx, parity);
		const fiberY = std.mix(wty, wby, parity);
		const crossX = std.mix(wbx, wtx, parity);
		const crossY = std.mix(wby, wty, parity);
		const acrossLocal = std.mix(fB, fA, parity);
		const alongLocal = std.mix(warp, weft, parity);
		const reliefAmt = __tsover_mul(relief, .32);
		const bulge = __tsover_mul(std.sin(__tsover_mul(acrossLocal, PI)), reliefAmt);
		const filJit = __tsover_mul(mxNoiseFloat2(d.vec2f(__tsover_mul(alongLocal, 2), __tsover_mul(acrossLocal, 40))), .18);
		const filament = __tsover_mul(__tsover_add(std.sin(__tsover_mul(acrossLocal, TWO_PI_SEVEN)), filJit), .13);
		const crossTilt = __tsover_add(bulge, filament);
		const n = std.normalize(d.vec3f(__tsover_add(nGeo.x, __tsover_mul(crossX, crossTilt)), __tsover_add(nGeo.y, __tsover_mul(crossY, crossTilt)), nGeo.z));
		const edge = std.min(__tsover_sub(.5, std.abs(fA)), __tsover_sub(.5, std.abs(fB)));
		const ao = std.mix(.5, 1, std.smoothstep(0, .05, edge));
		const fov = .6;
		const viewI = std.normalize(d.vec3f(__tsover_mul(__tsover_mul(__tsover_sub(uv.x, .5), aspect), fov), __tsover_mul(__tsover_sub(uv.y, .5), fov), 1));
		const cosView = std.clamp(__tsover_mul(std.dot(n, viewI), -1), 0, 1);
		const grazing = __tsover_sub(1, cosView);
		const R = std.reflect(viewI, n);
		const rough = std.clamp(roughness, 0, 1);
		const t0 = d.vec3f(fiberX, fiberY, 0);
		const T3 = std.normalize(t0.sub(n.mul(std.dot(n, t0))));
		const B3 = d.vec3f(__tsover_sub(__tsover_mul(n.y, T3.z), __tsover_mul(n.z, T3.y)), __tsover_sub(__tsover_mul(n.z, T3.x), __tsover_mul(n.x, T3.z)), __tsover_sub(__tsover_mul(n.x, T3.y), __tsover_mul(n.y, T3.x)));
		const L = std.normalize(d.vec3f(lx, ly, -.7));
		const Hh = std.normalize(L.sub(viewI));
		const hT = std.dot(Hh, T3);
		const hB = std.dot(Hh, B3);
		const hN = std.clamp(std.abs(std.dot(Hh, n)), 1e-4, 1);
		const baseA = std.mix(.06, .5, rough);
		const aAlong = __tsover_mul(baseA, 6);
		const aAcross = baseA;
		const expo = __tsover_div(__tsover_add(__tsover_mul(__tsover_div(hT, aAlong), __tsover_div(hT, aAlong)), __tsover_mul(__tsover_div(hB, aAcross), __tsover_div(hB, aAcross))), __tsover_mul(__tsover_add(hN, 1), .5));
		const sheen = __tsover_mul(std.exp(__tsover_mul(expo, -1)), fiberSheen);
		const ndl = std.clamp(__tsover_add(__tsover_mul(std.dot(n, L), .5), .5), 0, 1);
		const bodyLum = std.clamp(__tsover_add(__tsover_add(.035, __tsover_mul(ndl, .1)), sheen), 0, 1.3);
		const body = std.mix(darkColor, lightColor, d.vec3f(bodyLum)).mul(ao);
		const envStr = environment;
		const keyRad = d.f32(.16);
		const ex = __tsover_sub(__tsover_mul(R.x, envC), __tsover_mul(R.y, envS));
		const ey = __tsover_add(__tsover_mul(R.x, envS), __tsover_mul(R.y, envC));
		const p = d.vec2f(ex, ey);
		const sky = std.smoothstep(-.7, .85, ey);
		let coatLum = __tsover_mul(__tsover_mul(sky, sky), .3);
		const r0 = __tsover_mul(keyRad, 1);
		coatLum = __tsover_add(coatLum, __tsover_mul(__tsover_mul(std.smoothstep(r0, __tsover_mul(r0, .2), std.length(p.sub(d.vec2f(0, .55)))), 1.7), envStr));
		const r1 = __tsover_mul(keyRad, 1.15);
		coatLum = __tsover_add(coatLum, __tsover_mul(__tsover_mul(std.smoothstep(r1, __tsover_mul(r1, .2), std.length(p.sub(d.vec2f(-.55, .18)))), 1.05), envStr));
		const r2 = __tsover_mul(keyRad, 1);
		coatLum = __tsover_add(coatLum, __tsover_mul(__tsover_mul(std.smoothstep(r2, __tsover_mul(r2, .2), std.length(p.sub(d.vec2f(.55, .1)))), 1), envStr));
		const r3 = __tsover_mul(keyRad, 1.3);
		coatLum = __tsover_add(coatLum, __tsover_mul(__tsover_mul(std.smoothstep(r3, __tsover_mul(r3, .2), std.length(p.sub(d.vec2f(.2, -.5)))), .7), envStr));
		const r4 = __tsover_mul(keyRad, 1.1);
		coatLum = __tsover_add(coatLum, __tsover_mul(__tsover_mul(std.smoothstep(r4, __tsover_mul(r4, .2), std.length(p.sub(d.vec2f(-.35, -.25)))), .85), envStr));
		const coatF = __tsover_add(.05, __tsover_mul(std.pow(grazing, d.f32(5)), .95));
		const coat = __tsover_mul(__tsover_mul(coatLum, coatF), clearcoat);
		const rgb = body.add(d.vec3f(coat, coat, coat));
		const transition = std.smoothstep(0, 1, std.clamp(__tsover_mul(__tsover_div(__tsover_mul(sdf, -1), sharpEdge), 32), 0, 1));
		return d.vec4f(rgb.x, rgb.y, rgb.z, transition);
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [
				{
					"type": "i",
					"name": "surf0"
				},
				{
					"type": "i",
					"name": "surfX"
				},
				{
					"type": "i",
					"name": "surfY"
				},
				{
					"type": "i",
					"name": "sdfUV"
				},
				{
					"type": "i",
					"name": "uv"
				},
				{
					"type": "i",
					"name": "viewportSize"
				},
				{
					"type": "i",
					"name": "lightColor"
				},
				{
					"type": "i",
					"name": "darkColor"
				},
				{
					"type": "i",
					"name": "p0"
				},
				{
					"type": "i",
					"name": "p1"
				},
				{
					"type": "i",
					"name": "p2"
				}
			],
			"body": [0, [
				[
					13,
					"scl",
					[
						7,
						"p0",
						"x"
					]
				],
				[
					13,
					"edgeSoftness",
					[
						7,
						"p0",
						"y"
					]
				],
				[
					13,
					"weaveScale",
					[
						7,
						"p0",
						"z"
					]
				],
				[
					13,
					"weaveAngle",
					[
						7,
						"p0",
						"w"
					]
				],
				[
					13,
					"relief",
					[
						7,
						"p1",
						"x"
					]
				],
				[
					13,
					"fiberSheen",
					[
						7,
						"p1",
						"y"
					]
				],
				[
					13,
					"roughness",
					[
						7,
						"p1",
						"z"
					]
				],
				[
					13,
					"clearcoat",
					[
						7,
						"p1",
						"w"
					]
				],
				[
					13,
					"environment",
					[
						7,
						"p2",
						"x"
					]
				],
				[
					13,
					"envRotation",
					[
						7,
						"p2",
						"y"
					]
				],
				[
					13,
					"lightAngle",
					[
						7,
						"p2",
						"z"
					]
				],
				[
					13,
					"volumetric",
					[
						7,
						"p2",
						"w"
					]
				],
				[
					13,
					"aspect",
					[
						1,
						[
							7,
							"viewportSize",
							"x"
						],
						"/",
						[
							7,
							"viewportSize",
							"y"
						]
					]
				],
				[
					13,
					"sharpEdge",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							1,
							"edgeSoftness",
							"*",
							[5, "0.5"]
						], [5, "0.001"]]
					]
				],
				[
					13,
					"pxH",
					[
						1,
						[5, "1"],
						"/",
						[
							7,
							"viewportSize",
							"y"
						]
					]
				],
				[
					13,
					"sdfRaw",
					[
						7,
						"surf0",
						"x"
					]
				],
				[
					13,
					"sdf",
					[
						1,
						"sdfRaw",
						"/",
						"scl"
					]
				],
				[
					11,
					[
						1,
						"sdf",
						">",
						[
							1,
							"pxH",
							"*",
							[5, "2"]
						]
					],
					[10, [
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							[5, "0"],
							[5, "0"],
							[5, "0"],
							[5, "0"]
						]
					]]
				],
				[
					13,
					"lx",
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						[[
							1,
							"lightAngle",
							"*",
							"DEG_TO_RAD"
						]]
					]
				],
				[
					13,
					"ly",
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						[[
							1,
							"lightAngle",
							"*",
							"DEG_TO_RAD"
						]]
					]
				],
				[
					13,
					"wtx",
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						[[
							1,
							"weaveAngle",
							"*",
							"DEG_TO_RAD"
						]]
					]
				],
				[
					13,
					"wty",
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						[[
							1,
							"weaveAngle",
							"*",
							"DEG_TO_RAD"
						]]
					]
				],
				[
					13,
					"envC",
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						[[
							1,
							"envRotation",
							"*",
							"DEG_TO_RAD"
						]]
					]
				],
				[
					13,
					"envS",
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						[[
							1,
							"envRotation",
							"*",
							"DEG_TO_RAD"
						]]
					]
				],
				[
					13,
					"EPS",
					[5, "0.01"]
				],
				[
					13,
					"gradX",
					[
						1,
						[
							1,
							[
								7,
								"surfX",
								"x"
							],
							"-",
							"sdfRaw"
						],
						"/",
						"EPS"
					]
				],
				[
					13,
					"gradY",
					[
						1,
						[
							1,
							[
								7,
								"surfY",
								"x"
							],
							"-",
							"sdfRaw"
						],
						"/",
						"EPS"
					]
				],
				[
					13,
					"ddx",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									[
										7,
										"surfX",
										"w"
									],
									"-",
									[
										7,
										"surf0",
										"w"
									]
								],
								"/",
								"EPS"
							],
							[
								4,
								"-",
								[5, "5"]
							],
							[5, "5"]
						]
					]
				],
				[
					13,
					"ddy",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									[
										7,
										"surfY",
										"w"
									],
									"-",
									[
										7,
										"surf0",
										"w"
									]
								],
								"/",
								"EPS"
							],
							[
								4,
								"-",
								[5, "5"]
							],
							[5, "5"]
						]
					]
				],
				[
					13,
					"nVol",
					[
						6,
						[
							7,
							"std",
							"normalize"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								"ddx",
								"ddy",
								[
									4,
									"-",
									[5, "1"]
								]
							]
						]]
					]
				],
				[
					13,
					"rimT",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									"sdf",
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								],
								"/",
								[5, "0.05"]
							],
							[5, "0"],
							[5, "1"]
						]
					]
				],
				[
					13,
					"rimBevel",
					[
						1,
						[
							1,
							[5, "1"],
							"-",
							"rimT"
						],
						"*",
						[5, "1.6"]
					]
				],
				[
					13,
					"nFlat",
					[
						6,
						[
							7,
							"std",
							"normalize"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									"gradX",
									"*",
									"rimBevel"
								],
								[
									1,
									"gradY",
									"*",
									"rimBevel"
								],
								[
									4,
									"-",
									[5, "1"]
								]
							]
						]]
					]
				],
				[
					13,
					"nGeo",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							"nFlat",
							"nVol",
							[
								1,
								"volumetric",
								">",
								[5, "0.5"]
							]
						]
					]
				],
				[
					13,
					"pdx",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								1,
								[
									7,
									"sdfUV",
									"x"
								],
								"-",
								[5, "0.5"]
							],
							[
								7,
								"surf0",
								"y"
							],
							[
								1,
								"volumetric",
								">",
								[5, "0.5"]
							]
						]
					]
				],
				[
					13,
					"pdy",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								1,
								[
									7,
									"sdfUV",
									"y"
								],
								"-",
								[5, "0.5"]
							],
							[
								7,
								"surf0",
								"z"
							],
							[
								1,
								"volumetric",
								">",
								[5, "0.5"]
							]
						]
					]
				],
				[
					13,
					"wbx",
					[
						1,
						"wty",
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					]
				],
				[
					13,
					"wby",
					"wtx"
				],
				[
					13,
					"warp",
					[
						1,
						[
							1,
							"pdx",
							"*",
							"wtx"
						],
						"+",
						[
							1,
							"pdy",
							"*",
							"wty"
						]
					]
				],
				[
					13,
					"weft",
					[
						1,
						[
							1,
							"pdx",
							"*",
							"wbx"
						],
						"+",
						[
							1,
							"pdy",
							"*",
							"wby"
						]
					]
				],
				[
					13,
					"wScale",
					"weaveScale"
				],
				[
					13,
					"wA",
					[
						1,
						"warp",
						"*",
						"wScale"
					]
				],
				[
					13,
					"wB",
					[
						1,
						"weft",
						"*",
						"wScale"
					]
				],
				[
					13,
					"cellX",
					[
						6,
						[
							7,
							"std",
							"floor"
						],
						["wA"]
					]
				],
				[
					13,
					"cellY",
					[
						6,
						[
							7,
							"std",
							"floor"
						],
						["wB"]
					]
				],
				[
					13,
					"fA",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							["wA"]
						],
						"-",
						[5, "0.5"]
					]
				],
				[
					13,
					"fB",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							["wB"]
						],
						"-",
						[5, "0.5"]
					]
				],
				[
					12,
					"parity",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					11,
					[
						1,
						"weaveStyle",
						"===",
						[103, "plain"]
					],
					[0, [[
						13,
						"s2",
						[
							1,
							"cellX",
							"+",
							"cellY"
						]
					], [
						2,
						"parity",
						"=",
						[
							1,
							"s2",
							"-",
							[
								1,
								[5, "2"],
								"*",
								[
									6,
									[
										7,
										"std",
										"floor"
									],
									[[
										1,
										"s2",
										"/",
										[5, "2"]
									]]
								]
							]
						]
					]]],
					[0, [
						[
							13,
							"s4",
							[
								1,
								[
									1,
									"cellX",
									"-",
									"cellY"
								],
								"+",
								[5, "64"]
							]
						],
						[
							13,
							"m4",
							[
								1,
								"s4",
								"-",
								[
									1,
									[5, "4"],
									"*",
									[
										6,
										[
											7,
											"std",
											"floor"
										],
										[[
											1,
											"s4",
											"/",
											[5, "4"]
										]]
									]
								]
							]
						],
						[
							2,
							"parity",
							"=",
							[
								6,
								[
									7,
									"std",
									"select"
								],
								[
									[
										6,
										[
											7,
											"d",
											"f32"
										],
										[[5, "0"]]
									],
									[
										6,
										[
											7,
											"d",
											"f32"
										],
										[[5, "1"]]
									],
									[
										1,
										"m4",
										">=",
										[5, "2"]
									]
								]
							]
						]
					]]
				],
				[
					13,
					"fiberX",
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"wtx",
							"wbx",
							"parity"
						]
					]
				],
				[
					13,
					"fiberY",
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"wty",
							"wby",
							"parity"
						]
					]
				],
				[
					13,
					"crossX",
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"wbx",
							"wtx",
							"parity"
						]
					]
				],
				[
					13,
					"crossY",
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"wby",
							"wty",
							"parity"
						]
					]
				],
				[
					13,
					"acrossLocal",
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"fB",
							"fA",
							"parity"
						]
					]
				],
				[
					13,
					"alongLocal",
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"warp",
							"weft",
							"parity"
						]
					]
				],
				[
					13,
					"reliefAmt",
					[
						1,
						"relief",
						"*",
						[5, "0.32"]
					]
				],
				[
					13,
					"bulge",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								"acrossLocal",
								"*",
								"PI"
							]]
						],
						"*",
						"reliefAmt"
					]
				],
				[
					13,
					"filJit",
					[
						1,
						[
							6,
							[
								7,
								"noise",
								"mxNoiseFloat2"
							],
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									1,
									"alongLocal",
									"*",
									[5, "2"]
								], [
									1,
									"acrossLocal",
									"*",
									[5, "40"]
								]]
							]]
						],
						"*",
						[5, "0.18"]
					]
				],
				[
					13,
					"filament",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									"acrossLocal",
									"*",
									"TWO_PI_SEVEN"
								]]
							],
							"+",
							"filJit"
						],
						"*",
						[5, "0.13"]
					]
				],
				[
					13,
					"crossTilt",
					[
						1,
						"bulge",
						"+",
						"filament"
					]
				],
				[
					13,
					"n",
					[
						6,
						[
							7,
							"std",
							"normalize"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									[
										7,
										"nGeo",
										"x"
									],
									"+",
									[
										1,
										"crossX",
										"*",
										"crossTilt"
									]
								],
								[
									1,
									[
										7,
										"nGeo",
										"y"
									],
									"+",
									[
										1,
										"crossY",
										"*",
										"crossTilt"
									]
								],
								[
									7,
									"nGeo",
									"z"
								]
							]
						]]
					]
				],
				[
					13,
					"edge",
					[
						6,
						[
							7,
							"std",
							"min"
						],
						[[
							1,
							[5, "0.5"],
							"-",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								["fA"]
							]
						], [
							1,
							[5, "0.5"],
							"-",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								["fB"]
							]
						]]
					]
				],
				[
					13,
					"ao",
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							[5, "0.5"],
							[5, "1"],
							[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									[5, "0"],
									[5, "0.05"],
									"edge"
								]
							]
						]
					]
				],
				[
					13,
					"fov",
					[5, "0.6"]
				],
				[
					13,
					"viewI",
					[
						6,
						[
							7,
							"std",
							"normalize"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"uv",
												"x"
											],
											"-",
											[5, "0.5"]
										],
										"*",
										"aspect"
									],
									"*",
									"fov"
								],
								[
									1,
									[
										1,
										[
											7,
											"uv",
											"y"
										],
										"-",
										[5, "0.5"]
									],
									"*",
									"fov"
								],
								[5, "1"]
							]
						]]
					]
				],
				[
					13,
					"cosView",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									6,
									[
										7,
										"std",
										"dot"
									],
									["n", "viewI"]
								],
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							[5, "0"],
							[5, "1"]
						]
					]
				],
				[
					13,
					"grazing",
					[
						1,
						[5, "1"],
						"-",
						"cosView"
					]
				],
				[
					13,
					"R",
					[
						6,
						[
							7,
							"std",
							"reflect"
						],
						["viewI", "n"]
					]
				],
				[
					13,
					"rough",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							"roughness",
							[5, "0"],
							[5, "1"]
						]
					]
				],
				[
					13,
					"t0",
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"fiberX",
							"fiberY",
							[5, "0"]
						]
					]
				],
				[
					13,
					"T3",
					[
						6,
						[
							7,
							"std",
							"normalize"
						],
						[[
							6,
							[
								7,
								"t0",
								"sub"
							],
							[[
								6,
								[
									7,
									"n",
									"mul"
								],
								[[
									6,
									[
										7,
										"std",
										"dot"
									],
									["n", "t0"]
								]]
							]]
						]]
					]
				],
				[
					13,
					"B3",
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								[
									1,
									[
										7,
										"n",
										"y"
									],
									"*",
									[
										7,
										"T3",
										"z"
									]
								],
								"-",
								[
									1,
									[
										7,
										"n",
										"z"
									],
									"*",
									[
										7,
										"T3",
										"y"
									]
								]
							],
							[
								1,
								[
									1,
									[
										7,
										"n",
										"z"
									],
									"*",
									[
										7,
										"T3",
										"x"
									]
								],
								"-",
								[
									1,
									[
										7,
										"n",
										"x"
									],
									"*",
									[
										7,
										"T3",
										"z"
									]
								]
							],
							[
								1,
								[
									1,
									[
										7,
										"n",
										"x"
									],
									"*",
									[
										7,
										"T3",
										"y"
									]
								],
								"-",
								[
									1,
									[
										7,
										"n",
										"y"
									],
									"*",
									[
										7,
										"T3",
										"x"
									]
								]
							]
						]
					]
				],
				[
					13,
					"L",
					[
						6,
						[
							7,
							"std",
							"normalize"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								"lx",
								"ly",
								[
									4,
									"-",
									[5, "0.7"]
								]
							]
						]]
					]
				],
				[
					13,
					"Hh",
					[
						6,
						[
							7,
							"std",
							"normalize"
						],
						[[
							6,
							[
								7,
								"L",
								"sub"
							],
							["viewI"]
						]]
					]
				],
				[
					13,
					"hT",
					[
						6,
						[
							7,
							"std",
							"dot"
						],
						["Hh", "T3"]
					]
				],
				[
					13,
					"hB",
					[
						6,
						[
							7,
							"std",
							"dot"
						],
						["Hh", "B3"]
					]
				],
				[
					13,
					"hN",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									6,
									[
										7,
										"std",
										"dot"
									],
									["Hh", "n"]
								]]
							],
							[5, "0.0001"],
							[5, "1"]
						]
					]
				],
				[
					13,
					"baseA",
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							[5, "0.06"],
							[5, "0.5"],
							"rough"
						]
					]
				],
				[
					13,
					"aAlong",
					[
						1,
						"baseA",
						"*",
						[5, "6"]
					]
				],
				[
					13,
					"aAcross",
					"baseA"
				],
				[
					13,
					"expo",
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									"hT",
									"/",
									"aAlong"
								],
								"*",
								[
									1,
									"hT",
									"/",
									"aAlong"
								]
							],
							"+",
							[
								1,
								[
									1,
									"hB",
									"/",
									"aAcross"
								],
								"*",
								[
									1,
									"hB",
									"/",
									"aAcross"
								]
							]
						],
						"/",
						[
							1,
							[
								1,
								"hN",
								"+",
								[5, "1"]
							],
							"*",
							[5, "0.5"]
						]
					]
				],
				[
					13,
					"sheen",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								"expo",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]]
						],
						"*",
						"fiberSheen"
					]
				],
				[
					13,
					"ndl",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									[
										6,
										[
											7,
											"std",
											"dot"
										],
										["n", "L"]
									],
									"*",
									[5, "0.5"]
								],
								"+",
								[5, "0.5"]
							],
							[5, "0"],
							[5, "1"]
						]
					]
				],
				[
					13,
					"bodyLum",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									[5, "0.035"],
									"+",
									[
										1,
										"ndl",
										"*",
										[5, "0.1"]
									]
								],
								"+",
								"sheen"
							],
							[5, "0"],
							[5, "1.3"]
						]
					]
				],
				[
					13,
					"body",
					[
						6,
						[
							7,
							[
								6,
								[
									7,
									"std",
									"mix"
								],
								[
									"darkColor",
									"lightColor",
									[
										6,
										[
											7,
											"d",
											"vec3f"
										],
										["bodyLum"]
									]
								]
							],
							"mul"
						],
						["ao"]
					]
				],
				[
					13,
					"envStr",
					"environment"
				],
				[
					13,
					"keyRad",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0.16"]]
					]
				],
				[
					13,
					"ex",
					[
						1,
						[
							1,
							[
								7,
								"R",
								"x"
							],
							"*",
							"envC"
						],
						"-",
						[
							1,
							[
								7,
								"R",
								"y"
							],
							"*",
							"envS"
						]
					]
				],
				[
					13,
					"ey",
					[
						1,
						[
							1,
							[
								7,
								"R",
								"x"
							],
							"*",
							"envS"
						],
						"+",
						[
							1,
							[
								7,
								"R",
								"y"
							],
							"*",
							"envC"
						]
					]
				],
				[
					13,
					"p",
					[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						["ex", "ey"]
					]
				],
				[
					13,
					"sky",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								4,
								"-",
								[5, "0.7"]
							],
							[5, "0.85"],
							"ey"
						]
					]
				],
				[
					12,
					"coatLum",
					[
						1,
						[
							1,
							"sky",
							"*",
							"sky"
						],
						"*",
						[5, "0.3"]
					]
				],
				[
					13,
					"r0",
					[
						1,
						"keyRad",
						"*",
						[5, "1"]
					]
				],
				[
					2,
					"coatLum",
					"=",
					[
						1,
						"coatLum",
						"+",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"smoothstep"
									],
									[
										"r0",
										[
											1,
											"r0",
											"*",
											[5, "0.2"]
										],
										[
											6,
											[
												7,
												"std",
												"length"
											],
											[[
												6,
												[
													7,
													"p",
													"sub"
												],
												[[
													6,
													[
														7,
														"d",
														"vec2f"
													],
													[[5, "0"], [5, "0.55"]]
												]]
											]]
										]
									]
								],
								"*",
								[5, "1.7"]
							],
							"*",
							"envStr"
						]
					]
				],
				[
					13,
					"r1",
					[
						1,
						"keyRad",
						"*",
						[5, "1.15"]
					]
				],
				[
					2,
					"coatLum",
					"=",
					[
						1,
						"coatLum",
						"+",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"smoothstep"
									],
									[
										"r1",
										[
											1,
											"r1",
											"*",
											[5, "0.2"]
										],
										[
											6,
											[
												7,
												"std",
												"length"
											],
											[[
												6,
												[
													7,
													"p",
													"sub"
												],
												[[
													6,
													[
														7,
														"d",
														"vec2f"
													],
													[[
														4,
														"-",
														[5, "0.55"]
													], [5, "0.18"]]
												]]
											]]
										]
									]
								],
								"*",
								[5, "1.05"]
							],
							"*",
							"envStr"
						]
					]
				],
				[
					13,
					"r2",
					[
						1,
						"keyRad",
						"*",
						[5, "1"]
					]
				],
				[
					2,
					"coatLum",
					"=",
					[
						1,
						"coatLum",
						"+",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"smoothstep"
									],
									[
										"r2",
										[
											1,
											"r2",
											"*",
											[5, "0.2"]
										],
										[
											6,
											[
												7,
												"std",
												"length"
											],
											[[
												6,
												[
													7,
													"p",
													"sub"
												],
												[[
													6,
													[
														7,
														"d",
														"vec2f"
													],
													[[5, "0.55"], [5, "0.1"]]
												]]
											]]
										]
									]
								],
								"*",
								[5, "1"]
							],
							"*",
							"envStr"
						]
					]
				],
				[
					13,
					"r3",
					[
						1,
						"keyRad",
						"*",
						[5, "1.3"]
					]
				],
				[
					2,
					"coatLum",
					"=",
					[
						1,
						"coatLum",
						"+",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"smoothstep"
									],
									[
										"r3",
										[
											1,
											"r3",
											"*",
											[5, "0.2"]
										],
										[
											6,
											[
												7,
												"std",
												"length"
											],
											[[
												6,
												[
													7,
													"p",
													"sub"
												],
												[[
													6,
													[
														7,
														"d",
														"vec2f"
													],
													[[5, "0.2"], [
														4,
														"-",
														[5, "0.5"]
													]]
												]]
											]]
										]
									]
								],
								"*",
								[5, "0.7"]
							],
							"*",
							"envStr"
						]
					]
				],
				[
					13,
					"r4",
					[
						1,
						"keyRad",
						"*",
						[5, "1.1"]
					]
				],
				[
					2,
					"coatLum",
					"=",
					[
						1,
						"coatLum",
						"+",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"smoothstep"
									],
									[
										"r4",
										[
											1,
											"r4",
											"*",
											[5, "0.2"]
										],
										[
											6,
											[
												7,
												"std",
												"length"
											],
											[[
												6,
												[
													7,
													"p",
													"sub"
												],
												[[
													6,
													[
														7,
														"d",
														"vec2f"
													],
													[[
														4,
														"-",
														[5, "0.35"]
													], [
														4,
														"-",
														[5, "0.25"]
													]]
												]]
											]]
										]
									]
								],
								"*",
								[5, "0.85"]
							],
							"*",
							"envStr"
						]
					]
				],
				[
					13,
					"coatF",
					[
						1,
						[5, "0.05"],
						"+",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"pow"
								],
								["grazing", [
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "5"]]
								]]
							],
							"*",
							[5, "0.95"]
						]
					]
				],
				[
					13,
					"coat",
					[
						1,
						[
							1,
							"coatLum",
							"*",
							"coatF"
						],
						"*",
						"clearcoat"
					]
				],
				[
					13,
					"rgb",
					[
						6,
						[
							7,
							"body",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								"coat",
								"coat",
								"coat"
							]
						]]
					]
				],
				[
					13,
					"transition",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							[5, "1"],
							[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									[
										1,
										[
											1,
											[
												1,
												"sdf",
												"*",
												[
													4,
													"-",
													[5, "1"]
												]
											],
											"/",
											"sharpEdge"
										],
										"*",
										[5, "32"]
									],
									[5, "0"],
									[5, "1"]
								]
							]
						]
					]
				],
				[10, [
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							7,
							"rgb",
							"x"
						],
						[
							7,
							"rgb",
							"y"
						],
						[
							7,
							"rgb",
							"z"
						],
						"transition"
					]
				]]
			]],
			"externalNames": [
				"std",
				"d",
				"DEG_TO_RAD",
				"weaveStyle",
				"PI",
				"noise",
				"TWO_PI_SEVEN"
			]
		},
		externals: () => ({
			std,
			d,
			DEG_TO_RAD,
			weaveStyle,
			PI,
			noise: noise_exports,
			TWO_PI_SEVEN
		})
	}) && $.f)({})).$name(`carbonFiberComposite_${weaveStyle}`);
}
const componentDefinition = {
	name: "CarbonFiber",
	boundingBoxDeclaration: shapeEffectBoundingBoxDeclaration,
	category: "Shape Effects",
	description: "Photorealistic woven carbon fibre — interlaced tows whose anisotropic sheen flips ninety degrees cell to cell, raised into a quilted weave and finished with a glossy clearcoat that mirrors the studio. Plain or twill, in any shape.",
	extraFields: {
		...VOLUMETRIC_FIELD_EXTRA_FIELDS,
		...ANALYTIC_SDF_EXTRA_FIELDS
	},
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the carbon shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		scale: {
			default: 1,
			description: "Scale of the carbon shape (1 = default size)",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Scale",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the carbon shape in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		},
		lightColor: {
			default: "#b9bdc6",
			transform: transformColor,
			description: "The bright tone of the fibre sheen — recolor for forged/coloured carbon (red, blue, bronze)",
			ui: {
				type: "color",
				label: "Fiber Sheen",
				group: "Carbon"
			}
		},
		darkColor: {
			default: "#0b0c0e",
			transform: transformColor,
			description: "The deep resin tone the weave falls to in shadow — near-black for classic carbon",
			ui: {
				type: "color",
				label: "Resin Color",
				group: "Carbon"
			}
		},
		weaveStyle: {
			default: "twill",
			compileTime: true,
			description: "Weave pattern — plain (checkerboard) or 2×2 twill (the diagonal supercar weave)",
			ui: {
				type: "select",
				options: [{
					label: "Twill (2×2)",
					value: "twill"
				}, {
					label: "Plain",
					value: "plain"
				}],
				label: "Weave",
				group: "Carbon"
			}
		},
		weaveScale: {
			default: 30,
			description: "Number of woven tows across the shape — higher = finer, denser weave",
			ui: {
				type: "range",
				min: 2,
				max: 50,
				step: .5,
				label: "Weave Scale",
				group: "Carbon"
			}
		},
		weaveAngle: {
			default: 0,
			description: "Rotates the whole weave grid in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Weave Angle",
				group: "Carbon"
			}
		},
		relief: {
			default: .5,
			description: "Depth of the quilted weave — how far each tow bulges, catching light on its crown and shadowing the valleys between",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Weave Depth",
				group: "Carbon"
			}
		},
		fiberSheen: {
			default: 1.2,
			description: "Strength of the anisotropic fibre sheen — the directional satin glint running along each tow",
			ui: {
				type: "range",
				min: 0,
				max: 1.5,
				step: .01,
				label: "Fiber Sheen",
				group: "Carbon"
			}
		},
		roughness: {
			default: .5,
			description: "Softness of the fibre sheen — 0 = tight crisp glints, 1 = soft matte satin",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Roughness",
				group: "Surface"
			}
		},
		clearcoat: {
			default: 1.2,
			description: "Strength of the glossy lacquer coat — the sharp studio reflections of the wet clearcoat over the weave",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Clearcoat",
				group: "Surface"
			}
		},
		environment: {
			default: 2,
			description: "Strength of the reflected studio lighting seen in the clearcoat",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Environment",
				group: "Surface"
			}
		},
		envRotation: {
			default: 0,
			description: "Rotates the whole reflected studio — spins where the bright reflections fall",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Env Rotation",
				group: "Surface"
			}
		},
		lightAngle: {
			default: 215,
			description: "Direction of the key light for the fibre sheen, in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Light Angle",
				group: "Surface"
			}
		},
		edgeSoftness: {
			default: .05,
			description: "Softness of the shape boundary edge",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Softness",
				group: "Surface"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type"
		}
	},
	compute: (params) => {
		const vf = createVolumetricFieldComputeNode(params, () => params.getCpuValue("shape"), "raw");
		return vf ? {
			outputs: vf.outputs,
			getComputeNodes: (fp) => vf.getComputeNodes(fp)
		} : null;
	},
	fragment: (params) => {
		const { computeOutputs, uniforms, propValues, ctx, getCpuValue } = params;
		let sampler;
		let gradSampler;
		let volumetric = false;
		const fieldTex = computeOutputs?.volumetricFieldTexture;
		if (fieldTex) {
			const s = buildVolumetricFieldSampler(fieldTex, {
				originX: uniforms._vfOriginX,
				originY: uniforms._vfOriginY,
				spanX: uniforms._vfSpanX,
				spanY: uniforms._vfSpanY,
				activeResF: uniforms._vfActiveRes,
				rBound: uniforms._vfRBound
			});
			sampler = s.volumetricFieldSampler;
			gradSampler = s.volumetricFieldSamplerFast;
			volumetric = true;
		} else {
			const shapeSdfUrl = propValues.shapeSdfUrl || "";
			const shapeType = propValues.shapeType || "";
			if (shapeSdfUrl) {
				sampler = createSvgSdfSampler(params, shapeSdfUrl);
				gradSampler = sampler;
				volumetric = isVolumetricShapeType(shapeType);
			} else {
				const st = resolveShapeType(shapeType, getCpuValue("shape"));
				sampler = createAnalyticSdfSampler(st, driveAnalyticSubProps(params, () => getCpuValue("shape")), "raw");
				gradSampler = sampler;
				volumetric = is3dShapeType(st);
			}
		}
		const weaveStyle = (propValues.weaveStyle || "twill") === "plain" ? "plain" : "twill";
		const composite = buildCarbonFiberComposite(weaveStyle);
		const sdfUV = call(sdfSpaceUV, "sdfSpaceUV", [
			uniforms.center,
			uniforms.scale,
			uniforms.rotation,
			ctx.uv,
			ctx.aspect
		]);
		const surf0 = sampler(sdfUV);
		const surfX = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(.01),
			floatE(0)
		]));
		const surfY = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(0),
			floatE(.01)
		]));
		const volFlag = floatE(volumetric ? 1 : 0);
		const p0 = vec4(uniforms.scale, uniforms.edgeSoftness, uniforms.weaveScale, uniforms.weaveAngle);
		const p1 = vec4(uniforms.relief, uniforms.fiberSheen, uniforms.roughness, uniforms.clearcoat);
		const p2 = vec4(uniforms.environment, uniforms.envRotation, uniforms.lightAngle, volFlag);
		return call(composite, `carbonFiberComposite_${weaveStyle}`, [
			surf0,
			surfX,
			surfY,
			sdfUV,
			ctx.uv,
			ctx.viewportSize,
			uniforms.lightColor.member("rgb"),
			uniforms.darkColor.member("rgb"),
			p0,
			p1,
			p2
		]);
	}
};
var CarbonFiber_default = componentDefinition;
export { buildCarbonFiberComposite as n, componentDefinition as r, CarbonFiber_default as t };
