// shaders/dist/core/Chalkboard-C9PtcnYN.js
import { At as call, It as vec4, Ot as ZERO, Pt as floatE, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor } from "./transformations-fqzznt4g.js";
var SIN_045 = Math.sin(.785);
var COS_045 = Math.cos(.785);
var SIN_N45 = Math.sin(-.785);
var COS_N45 = Math.cos(-.785);
var chalkHash21 = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	return std.fract(__tsover_mul(std.sin(std.dot(p, d.vec2f(12.9898, 78.233))), 43758.5453));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				1,
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["p", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "12.9898"], [5, "78.233"]]
						]]
					]]
				],
				"*",
				[5, "43758.5453"]
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
var chalkVnoise = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	const i = std.floor(p);
	const f = std.fract(p);
	const u = f.mul(f).mul(f.mul(-2).add(3));
	const a = chalkHash21(i);
	const b = chalkHash21(d.vec2f(__tsover_add(i.x, 1), i.y));
	const c = chalkHash21(d.vec2f(i.x, __tsover_add(i.y, 1)));
	const dd = chalkHash21(d.vec2f(__tsover_add(i.x, 1), __tsover_add(i.y, 1)));
	return std.mix(std.mix(a, b, u.x), std.mix(c, dd, u.x), u.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [
			[
				13,
				"i",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["p"]
				]
			],
			[
				13,
				"f",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["p"]
				]
			],
			[
				13,
				"u",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"f",
								"mul"
							],
							["f"]
						],
						"mul"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"f",
									"mul"
								],
								[[
									4,
									"-",
									[5, "2"]
								]]
							],
							"add"
						],
						[[5, "3"]]
					]]
				]
			],
			[
				13,
				"a",
				[
					6,
					"chalkHash21",
					["i"]
				]
			],
			[
				13,
				"b",
				[
					6,
					"chalkHash21",
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								7,
								"i",
								"x"
							],
							"+",
							[5, "1"]
						], [
							7,
							"i",
							"y"
						]]
					]]
				]
			],
			[
				13,
				"c",
				[
					6,
					"chalkHash21",
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							7,
							"i",
							"x"
						], [
							1,
							[
								7,
								"i",
								"y"
							],
							"+",
							[5, "1"]
						]]
					]]
				]
			],
			[
				13,
				"dd",
				[
					6,
					"chalkHash21",
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								7,
								"i",
								"x"
							],
							"+",
							[5, "1"]
						], [
							1,
							[
								7,
								"i",
								"y"
							],
							"+",
							[5, "1"]
						]]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"mix"
				],
				[
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"a",
							"b",
							[
								7,
								"u",
								"x"
							]
						]
					],
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"c",
							"dd",
							[
								7,
								"u",
								"x"
							]
						]
					],
					[
						7,
						"u",
						"y"
					]
				]
			]]
		]],
		"externalNames": [
			"std",
			"chalkHash21",
			"d"
		]
	},
	externals: () => ({
		std,
		chalkHash21,
		d
	})
}) && $.f)({}));
var chalkLuma = tgpu.fn([d.vec4f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color) => {
	"use gpu";
	return std.dot(d.vec3f(color.x, color.y, color.z), d.vec3f(.299, .587, .114));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"dot"
			],
			[[
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[
						7,
						"color",
						"x"
					],
					[
						7,
						"color",
						"y"
					],
					[
						7,
						"color",
						"z"
					]
				]
			], [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[5, "0.299"],
					[5, "0.587"],
					[5, "0.114"]
				]
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const chalkOffsetUV = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, edgeThickness, ox, oy) => {
	"use gpu";
	const texelX = __tsover_div(1, viewport.x);
	const texelY = __tsover_div(1, viewport.y);
	return d.vec2f(__tsover_add(uv.x, __tsover_mul(__tsover_mul(texelX, edgeThickness), ox)), __tsover_add(uv.y, __tsover_mul(__tsover_mul(texelY, edgeThickness), oy)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "edgeThickness"
			},
			{
				"type": "i",
				"name": "ox"
			},
			{
				"type": "i",
				"name": "oy"
			}
		],
		"body": [0, [
			[
				13,
				"texelX",
				[
					1,
					[5, "1"],
					"/",
					[
						7,
						"viewport",
						"x"
					]
				]
			],
			[
				13,
				"texelY",
				[
					1,
					[5, "1"],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						7,
						"uv",
						"x"
					],
					"+",
					[
						1,
						[
							1,
							"texelX",
							"*",
							"edgeThickness"
						],
						"*",
						"ox"
					]
				], [
					1,
					[
						7,
						"uv",
						"y"
					],
					"+",
					[
						1,
						[
							1,
							"texelY",
							"*",
							"edgeThickness"
						],
						"*",
						"oy"
					]
				]]
			]]
		]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
var chalkHatchLine = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((P, hatchScale, shade, sa, ca, thr) => {
	"use gpu";
	const ry = __tsover_add(__tsover_mul(P.x, sa), __tsover_mul(P.y, ca));
	const hl = __tsover_mul(std.abs(__tsover_sub(std.fract(__tsover_div(ry, hatchScale)), .5)), 2);
	const line = __tsover_sub(1, std.smoothstep(0, .35, hl));
	return __tsover_mul(line, std.step(thr, shade));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "P"
			},
			{
				"type": "i",
				"name": "hatchScale"
			},
			{
				"type": "i",
				"name": "shade"
			},
			{
				"type": "i",
				"name": "sa"
			},
			{
				"type": "i",
				"name": "ca"
			},
			{
				"type": "i",
				"name": "thr"
			}
		],
		"body": [0, [
			[
				13,
				"ry",
				[
					1,
					[
						1,
						[
							7,
							"P",
							"x"
						],
						"*",
						"sa"
					],
					"+",
					[
						1,
						[
							7,
							"P",
							"y"
						],
						"*",
						"ca"
					]
				]
			],
			[
				13,
				"hl",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"fract"
								],
								[[
									1,
									"ry",
									"/",
									"hatchScale"
								]]
							],
							"-",
							[5, "0.5"]
						]]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"line",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							[5, "0.35"],
							"hl"
						]
					]
				]
			],
			[10, [
				1,
				"line",
				"*",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					["thr", "shade"]
				]
			]]
		]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const chalkboardCompose = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec2f,
	d.vec2f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((sobelA, sobelB, uv, viewport, misc, params, boardColor, chalkColor) => {
	"use gpu";
	const tl = sobelA.x;
	const t = sobelA.y;
	const tr = sobelA.z;
	const l = sobelA.w;
	const r = sobelB.x;
	const bl = sobelB.y;
	const bo = sobelB.z;
	const br = sobelB.w;
	const centerLum = misc.x;
	const centerAlpha = misc.y;
	const edgeSensitivity = params.x;
	const hatchScale = params.y;
	const shading = params.z;
	const grain = params.w;
	const gx = __tsover_add(__tsover_add(__tsover_add(__tsover_sub(__tsover_sub(-tl, __tsover_mul(l, 2)), bl), tr), __tsover_mul(r, 2)), br);
	const gy = __tsover_add(__tsover_add(__tsover_add(__tsover_sub(__tsover_sub(-tl, __tsover_mul(t, 2)), tr), bl), __tsover_mul(bo, 2)), br);
	const edgeMag = std.sqrt(__tsover_add(__tsover_mul(gx, gx), __tsover_mul(gy, gy)));
	const edgeHigh = std.mix(d.f32(1), .08, edgeSensitivity);
	const edge = std.smoothstep(__tsover_mul(edgeHigh, .4), edgeHigh, edgeMag);
	const shade = __tsover_sub(1, centerLum);
	const P = uv.mul(viewport);
	const hatch = __tsover_mul(std.max(std.max(chalkHatchLine(P, hatchScale, shade, SIN_045, COS_045, .22), chalkHatchLine(P, hatchScale, shade, SIN_N45, COS_N45, .5)), chalkHatchLine(P, hatchScale, shade, 0, 1, .78)), shading);
	const dust = std.mix(__tsover_sub(1, grain), d.f32(1), chalkVnoise(P.mul(.5)));
	const chalkMask = std.clamp(__tsover_mul(std.max(edge, hatch), dust), 0, 1);
	const boardGrain = __tsover_mul(__tsover_sub(chalkVnoise(P.mul(.9)), .5), .04);
	const boardRGB = boardColor.xyz.add(boardGrain);
	const col = std.mix(boardRGB, chalkColor.xyz, d.vec3f(chalkMask));
	return d.vec4f(col, centerAlpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "sobelA"
			},
			{
				"type": "i",
				"name": "sobelB"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "misc"
			},
			{
				"type": "i",
				"name": "params"
			},
			{
				"type": "i",
				"name": "boardColor"
			},
			{
				"type": "i",
				"name": "chalkColor"
			}
		],
		"body": [0, [
			[
				13,
				"tl",
				[
					7,
					"sobelA",
					"x"
				]
			],
			[
				13,
				"t",
				[
					7,
					"sobelA",
					"y"
				]
			],
			[
				13,
				"tr",
				[
					7,
					"sobelA",
					"z"
				]
			],
			[
				13,
				"l",
				[
					7,
					"sobelA",
					"w"
				]
			],
			[
				13,
				"r",
				[
					7,
					"sobelB",
					"x"
				]
			],
			[
				13,
				"bl",
				[
					7,
					"sobelB",
					"y"
				]
			],
			[
				13,
				"bo",
				[
					7,
					"sobelB",
					"z"
				]
			],
			[
				13,
				"br",
				[
					7,
					"sobelB",
					"w"
				]
			],
			[
				13,
				"centerLum",
				[
					7,
					"misc",
					"x"
				]
			],
			[
				13,
				"centerAlpha",
				[
					7,
					"misc",
					"y"
				]
			],
			[
				13,
				"edgeSensitivity",
				[
					7,
					"params",
					"x"
				]
			],
			[
				13,
				"hatchScale",
				[
					7,
					"params",
					"y"
				]
			],
			[
				13,
				"shading",
				[
					7,
					"params",
					"z"
				]
			],
			[
				13,
				"grain",
				[
					7,
					"params",
					"w"
				]
			],
			[
				13,
				"gx",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									[
										4,
										"-",
										"tl"
									],
									"-",
									[
										1,
										"l",
										"*",
										[5, "2"]
									]
								],
								"-",
								"bl"
							],
							"+",
							"tr"
						],
						"+",
						[
							1,
							"r",
							"*",
							[5, "2"]
						]
					],
					"+",
					"br"
				]
			],
			[
				13,
				"gy",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									[
										4,
										"-",
										"tl"
									],
									"-",
									[
										1,
										"t",
										"*",
										[5, "2"]
									]
								],
								"-",
								"tr"
							],
							"+",
							"bl"
						],
						"+",
						[
							1,
							"bo",
							"*",
							[5, "2"]
						]
					],
					"+",
					"br"
				]
			],
			[
				13,
				"edgeMag",
				[
					6,
					[
						7,
						"std",
						"sqrt"
					],
					[[
						1,
						[
							1,
							"gx",
							"*",
							"gx"
						],
						"+",
						[
							1,
							"gy",
							"*",
							"gy"
						]
					]]
				]
			],
			[
				13,
				"edgeHigh",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[5, "0.08"],
						"edgeSensitivity"
					]
				]
			],
			[
				13,
				"edge",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"edgeHigh",
							"*",
							[5, "0.4"]
						],
						"edgeHigh",
						"edgeMag"
					]
				]
			],
			[
				13,
				"shade",
				[
					1,
					[5, "1"],
					"-",
					"centerLum"
				]
			],
			[
				13,
				"P",
				[
					6,
					[
						7,
						"uv",
						"mul"
					],
					["viewport"]
				]
			],
			[
				13,
				"hatch",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								6,
								"chalkHatchLine",
								[
									"P",
									"hatchScale",
									"shade",
									"SIN_045",
									"COS_045",
									[5, "0.22"]
								]
							], [
								6,
								"chalkHatchLine",
								[
									"P",
									"hatchScale",
									"shade",
									"SIN_N45",
									"COS_N45",
									[5, "0.5"]
								]
							]]
						], [
							6,
							"chalkHatchLine",
							[
								"P",
								"hatchScale",
								"shade",
								[5, "0"],
								[5, "1"],
								[5, "0.78"]
							]
						]]
					],
					"*",
					"shading"
				]
			],
			[
				13,
				"dust",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							1,
							[5, "1"],
							"-",
							"grain"
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							6,
							"chalkVnoise",
							[[
								6,
								[
									7,
									"P",
									"mul"
								],
								[[5, "0.5"]]
							]]
						]
					]
				]
			],
			[
				13,
				"chalkMask",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["edge", "hatch"]
							],
							"*",
							"dust"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"boardGrain",
				[
					1,
					[
						1,
						[
							6,
							"chalkVnoise",
							[[
								6,
								[
									7,
									"P",
									"mul"
								],
								[[5, "0.9"]]
							]]
						],
						"-",
						[5, "0.5"]
					],
					"*",
					[5, "0.04"]
				]
			],
			[
				13,
				"boardRGB",
				[
					6,
					[
						7,
						[
							7,
							"boardColor",
							"xyz"
						],
						"add"
					],
					["boardGrain"]
				]
			],
			[
				13,
				"col",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"boardRGB",
						[
							7,
							"chalkColor",
							"xyz"
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["chalkMask"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				["col", "centerAlpha"]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"chalkHatchLine",
			"SIN_045",
			"COS_045",
			"SIN_N45",
			"COS_N45",
			"chalkVnoise"
		]
	},
	externals: () => ({
		std,
		d,
		chalkHatchLine,
		SIN_045,
		COS_045,
		SIN_N45,
		COS_N45,
		chalkVnoise
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Chalkboard",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Renders content as a chalk drawing on a blackboard, with edge strokes and cross-hatch shading",
	requiresRTT: true,
	requiresChild: true,
	props: {
		boardColor: {
			default: "#000000",
			transform: transformColor,
			description: "Background board color",
			ui: {
				type: "color",
				label: "Board",
				group: "Colors"
			}
		},
		chalkColor: {
			default: "#eceadb",
			transform: transformColor,
			description: "Chalk stroke color",
			ui: {
				type: "color",
				label: "Chalk",
				group: "Colors"
			}
		},
		edgeSensitivity: {
			default: .5,
			description: "How readily edges are detected and drawn as chalk strokes",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Sensitivity",
				group: "Effect"
			}
		},
		edgeThickness: {
			default: 1.5,
			description: "Thickness of the chalk outline strokes",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 4,
				step: .1,
				label: "Edge Thickness",
				group: "Effect"
			}
		},
		shading: {
			default: .05,
			description: "Amount of cross-hatch shading filling darker regions",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Shading",
				group: "Effect"
			}
		},
		hatchScale: {
			default: 16,
			description: "Spacing of the cross-hatch lines in pixels",
			ui: {
				type: ["range", "map"],
				min: 3,
				max: 24,
				step: .5,
				label: "Hatch Scale",
				group: "Effect"
			}
		},
		grain: {
			default: .45,
			description: "Chalk dustiness / grain that breaks up the strokes",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Grain",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const viewport = ctx.viewportSize;
		const lumAt = (ox, oy) => {
			const uv = call(chalkOffsetUV, "chalkOffsetUV", [
				ctx.uv,
				viewport,
				uniforms.edgeThickness,
				floatE(ox),
				floatE(oy)
			]);
			return call(chalkLuma, "chalkLuma", [call(unpremultiplyAlpha, "unpremultiplyAlpha", [texture.sample(uv)])]);
		};
		const tl = lumAt(-1, -1);
		const t = lumAt(0, -1);
		const tr = lumAt(1, -1);
		const l = lumAt(-1, 0);
		const r = lumAt(1, 0);
		const bl = lumAt(-1, 1);
		const bo = lumAt(0, 1);
		const br = lumAt(1, 1);
		const center = call(unpremultiplyAlpha, "unpremultiplyAlpha", [texture.sample(ctx.uv)]);
		const centerLum = call(chalkLuma, "chalkLuma", [center]);
		const sobelA = vec4(tl, t, tr, l);
		const sobelB = vec4(r, bl, bo, br);
		const misc = vec4(centerLum, center.member("a"), 0, 0);
		const params = vec4(uniforms.edgeSensitivity, uniforms.hatchScale, uniforms.shading, uniforms.grain);
		return call(chalkboardCompose, "chalkboardCompose", [
			sobelA,
			sobelB,
			ctx.uv,
			viewport,
			misc,
			params,
			uniforms.boardColor,
			uniforms.chalkColor
		]);
	}
};
var Chalkboard_default = componentDefinition;
export { componentDefinition as i, chalkOffsetUV as n, chalkboardCompose as r, Chalkboard_default as t };
