// shaders/dist/core/ChannelBlur-B1bY63Fb.js
import { At as call, I as createGaussianBlurCompute, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
const INTENSITY_TO_RADIUS = .1;
const channelIntensityToRadius = (intensity) => intensity * INTENSITY_TO_RADIUS;
const channelBlurCompose = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.f32,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((sharp, blurred, redIntensity, greenIntensity, blueIntensity) => {
	"use gpu";
	const redRadius = __tsover_mul(redIntensity, .1);
	const greenRadius = __tsover_mul(greenIntensity, .1);
	const blueRadius = __tsover_mul(blueIntensity, .1);
	const maxRadius = std.max(std.max(std.max(redRadius, greenRadius), blueRadius), .01);
	const rMix = __tsover_div(redRadius, maxRadius);
	const gMix = __tsover_div(greenRadius, maxRadius);
	const bMix = __tsover_div(blueRadius, maxRadius);
	const finalR = std.mix(sharp.x, blurred.x, rMix);
	const finalG = std.mix(sharp.y, blurred.y, gMix);
	const finalB = std.mix(sharp.z, blurred.z, bMix);
	return d.vec4f(finalR, finalG, finalB, sharp.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "sharp"
			},
			{
				"type": "i",
				"name": "blurred"
			},
			{
				"type": "i",
				"name": "redIntensity"
			},
			{
				"type": "i",
				"name": "greenIntensity"
			},
			{
				"type": "i",
				"name": "blueIntensity"
			}
		],
		"body": [0, [
			[
				13,
				"redRadius",
				[
					1,
					"redIntensity",
					"*",
					[5, "0.1"]
				]
			],
			[
				13,
				"greenRadius",
				[
					1,
					"greenIntensity",
					"*",
					[5, "0.1"]
				]
			],
			[
				13,
				"blueRadius",
				[
					1,
					"blueIntensity",
					"*",
					[5, "0.1"]
				]
			],
			[
				13,
				"maxRadius",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							6,
							[
								7,
								"std",
								"max"
							],
							["redRadius", "greenRadius"]
						], "blueRadius"]
					], [5, "0.01"]]
				]
			],
			[
				13,
				"rMix",
				[
					1,
					"redRadius",
					"/",
					"maxRadius"
				]
			],
			[
				13,
				"gMix",
				[
					1,
					"greenRadius",
					"/",
					"maxRadius"
				]
			],
			[
				13,
				"bMix",
				[
					1,
					"blueRadius",
					"/",
					"maxRadius"
				]
			],
			[
				13,
				"finalR",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							7,
							"sharp",
							"x"
						],
						[
							7,
							"blurred",
							"x"
						],
						"rMix"
					]
				]
			],
			[
				13,
				"finalG",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							7,
							"sharp",
							"y"
						],
						[
							7,
							"blurred",
							"y"
						],
						"gMix"
					]
				]
			],
			[
				13,
				"finalB",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							7,
							"sharp",
							"z"
						],
						[
							7,
							"blurred",
							"z"
						],
						"bMix"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					"finalR",
					"finalG",
					"finalB",
					[
						7,
						"sharp",
						"w"
					]
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({})).$name("channelBlurCompose");
const componentDefinition = {
	name: "ChannelBlur",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Blurs",
	description: "Independent blur for red, green, and blue channels",
	requiresRTT: true,
	requiresChild: true,
	props: {
		redIntensity: {
			default: 0,
			description: "Blur intensity for red channel",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Red Intensity",
				group: "Effect"
			}
		},
		greenIntensity: {
			default: 20,
			description: "Blur intensity for green channel",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Green Intensity",
				group: "Effect"
			}
		},
		blueIntensity: {
			default: 40,
			description: "Blur intensity for blue channel",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Blue Intensity",
				group: "Effect"
			}
		}
	},
	compute: (params) => {
		const { childNode, gpu, convertToTexture, registerComputeTexture, getCpuValue, onCleanup, onResize, dimensions } = params;
		if (!childNode) return null;
		const root = gpu?.root;
		if (!root) return null;
		const childTexture = convertToTexture(childNode);
		const gaussian = createGaussianBlurCompute(root, null, Math.max(1, Math.round(dimensions.width)), Math.max(1, Math.round(dimensions.height)), onCleanup);
		const blurredTexture = registerComputeTexture(gaussian.outputTexture);
		onResize(({ width, height }) => gaussian.setInputDimensions(Math.max(1, Math.round(width)), Math.max(1, Math.round(height))));
		return {
			outputs: {
				childTexture,
				blurredTexture
			},
			bindInputs: (resolve) => {
				const src = resolve(childTexture.key);
				if (src) gaussian.setInputTexture(src.texture);
			},
			getComputeNodes: () => {
				const r = (getCpuValue("redIntensity") ?? 0) * .1;
				const g = (getCpuValue("greenIntensity") ?? 0) * .1;
				const b = (getCpuValue("blueIntensity") ?? 0) * .1;
				gaussian.updateRadius(Math.max(r, g, b, .01));
				return gaussian.computeSteps;
			}
		};
	},
	fragment: ({ childNode, computeOutputs, uniforms, ctx, convertToTexture }) => {
		if (!childNode) return ZERO;
		const blurred = computeOutputs?.blurredTexture;
		const sharp = computeOutputs?.childTexture;
		if (!blurred || !sharp) return call(unpremultiplyAlpha, "unpremultiplyAlpha", [convertToTexture(childNode).sample(ctx.uv)]);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [call(channelBlurCompose, "channelBlurCompose", [
			sharp.sample(ctx.uv),
			blurred.sample(ctx.uv),
			uniforms.redIntensity,
			uniforms.greenIntensity,
			uniforms.blueIntensity
		])]);
	}
};
var ChannelBlur_default = componentDefinition;
export { componentDefinition as a, channelIntensityToRadius as i, INTENSITY_TO_RADIUS as n, channelBlurCompose as r, ChannelBlur_default as t };
