// shaders/dist/core/Checkerboard-DbjFO-fF.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
const checkerCoord = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, cells) => {
	"use gpu";
	return d.vec2f(__tsover_mul(__tsover_mul(uv.x, aspect), cells), __tsover_mul(__tsover_sub(1, uv.y), cells));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "cells"
			}
		],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				],
				"*",
				"cells"
			], [
				1,
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"uv",
						"y"
					]
				],
				"*",
				"cells"
			]]
		]]]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const checkerFilterValue = tgpu.fn([d.vec2f, d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, w) => {
	"use gpu";
	const aHalf = d.vec2f(__tsover_mul(w.x, .5), __tsover_mul(w.y, .5));
	const minusSideX = std.abs(__tsover_sub(std.fract(__tsover_mul(__tsover_sub(p.x, aHalf.x), .5)), .5));
	const minusSideY = std.abs(__tsover_sub(std.fract(__tsover_mul(__tsover_sub(p.y, aHalf.y), .5)), .5));
	const plusSideX = std.abs(__tsover_sub(std.fract(__tsover_mul(__tsover_add(p.x, aHalf.x), .5)), .5));
	const plusSideY = std.abs(__tsover_sub(std.fract(__tsover_mul(__tsover_add(p.y, aHalf.y), .5)), .5));
	const iX = __tsover_div(__tsover_mul(__tsover_sub(minusSideX, plusSideX), 2), w.x);
	const iY = __tsover_div(__tsover_mul(__tsover_sub(minusSideY, plusSideY), 2), w.y);
	return std.clamp(__tsover_sub(.5, __tsover_mul(__tsover_mul(iX, iY), .5)), 0, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}, {
			"type": "i",
			"name": "w"
		}],
		"body": [0, [
			[
				13,
				"aHalf",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"w",
							"x"
						],
						"*",
						[5, "0.5"]
					], [
						1,
						[
							7,
							"w",
							"y"
						],
						"*",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"minusSideX",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								1,
								[
									1,
									[
										7,
										"p",
										"x"
									],
									"-",
									[
										7,
										"aHalf",
										"x"
									]
								],
								"*",
								[5, "0.5"]
							]]
						],
						"-",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"minusSideY",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								1,
								[
									1,
									[
										7,
										"p",
										"y"
									],
									"-",
									[
										7,
										"aHalf",
										"y"
									]
								],
								"*",
								[5, "0.5"]
							]]
						],
						"-",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"plusSideX",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								1,
								[
									1,
									[
										7,
										"p",
										"x"
									],
									"+",
									[
										7,
										"aHalf",
										"x"
									]
								],
								"*",
								[5, "0.5"]
							]]
						],
						"-",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"plusSideY",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								1,
								[
									1,
									[
										7,
										"p",
										"y"
									],
									"+",
									[
										7,
										"aHalf",
										"y"
									]
								],
								"*",
								[5, "0.5"]
							]]
						],
						"-",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"iX",
				[
					1,
					[
						1,
						[
							1,
							"minusSideX",
							"-",
							"plusSideX"
						],
						"*",
						[5, "2"]
					],
					"/",
					[
						7,
						"w",
						"x"
					]
				]
			],
			[
				13,
				"iY",
				[
					1,
					[
						1,
						[
							1,
							"minusSideY",
							"-",
							"plusSideY"
						],
						"*",
						[5, "2"]
					],
					"/",
					[
						7,
						"w",
						"y"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"clamp"
				],
				[
					[
						1,
						[5, "0.5"],
						"-",
						[
							1,
							[
								1,
								"iX",
								"*",
								"iY"
							],
							"*",
							[5, "0.5"]
						]
					],
					[5, "0"],
					[5, "1"]
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const checkerBlend = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, cells, softness) => {
	"use gpu";
	const p = checkerCoord(uv, __tsover_div(viewport.x, std.max(viewport.y, 1e-6)), cells);
	const dpx = std.dpdx(p);
	const dpy = std.dpdy(p);
	const baseW = d.vec2f(std.max(std.abs(dpx.x), std.abs(dpy.x)), std.max(std.abs(dpx.y), std.abs(dpy.y)));
	return checkerFilterValue(p, d.vec2f(std.max(__tsover_add(baseW.x, softness), 1e-5), std.max(__tsover_add(baseW.y, softness), 1e-5)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "cells"
			},
			{
				"type": "i",
				"name": "softness"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							7,
							"viewport",
							"y"
						], [5, "0.000001"]]
					]
				]
			],
			[
				13,
				"p",
				[
					6,
					"checkerCoord",
					[
						"uv",
						"aspect",
						"cells"
					]
				]
			],
			[
				13,
				"dpx",
				[
					6,
					[
						7,
						"std",
						"dpdx"
					],
					["p"]
				]
			],
			[
				13,
				"dpy",
				[
					6,
					[
						7,
						"std",
						"dpdy"
					],
					["p"]
				]
			],
			[
				13,
				"baseW",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								7,
								"dpx",
								"x"
							]]
						], [
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								7,
								"dpy",
								"x"
							]]
						]]
					], [
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								7,
								"dpx",
								"y"
							]]
						], [
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								7,
								"dpy",
								"y"
							]]
						]]
					]]
				]
			],
			[
				13,
				"w",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							1,
							[
								7,
								"baseW",
								"x"
							],
							"+",
							"softness"
						], [5, "0.00001"]]
					], [
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							1,
							[
								7,
								"baseW",
								"y"
							],
							"+",
							"softness"
						], [5, "0.00001"]]
					]]
				]
			],
			[10, [
				6,
				"checkerFilterValue",
				["p", "w"]
			]]
		]],
		"externalNames": [
			"std",
			"checkerCoord",
			"d",
			"checkerFilterValue"
		]
	},
	externals: () => ({
		std,
		checkerCoord,
		d,
		checkerFilterValue
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Checkerboard",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Classic checkerboard pattern with two alternating colors",
	acceptsUVContext: true,
	props: {
		colorA: {
			default: "#cccccc",
			transform: transformColor,
			description: "First color of the checkerboard pattern",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#999999",
			transform: transformColor,
			description: "Second color of the checkerboard pattern",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		cells: {
			default: 8,
			description: "Number of cells along the canvas height (creates square cells). Switch the unit to px in the editor to set an absolute cell size instead.",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 50,
				step: 1,
				label: "Cells",
				group: "Effect",
				dimensional: "count-canvas-height"
			}
		},
		softness: {
			default: 0,
			description: "Smoothness of the transition between colors (0 = hard edges, 1 = very soft)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Softness",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const blendFactor = call(checkerBlend, "checkerBlend", [
			params.uvContext ?? params.ctx.uv,
			params.effectiveViewportSize ?? params.ctx.viewportSize,
			params.uniforms.cells,
			params.uniforms.softness
		]);
		return call(mixColorsVariants[params.propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			params.uniforms.colorA,
			params.uniforms.colorB,
			blendFactor
		]);
	}
};
var Checkerboard_default = componentDefinition;
export { componentDefinition as a, checkerFilterValue as i, checkerBlend as n, checkerCoord as r, Checkerboard_default as t };
