// shaders/dist/core/Chevron-D5-x4Wwd.js
import { At as call, Ft as mixExpr, It as vec4, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
var DEG_TO_RAD = .0174533;
const chevronMask = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, count, angle, balance, softness, offset, animTime) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const au = __tsover_mul(uv.x, aspect);
	const av = uv.y;
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const cosA = std.cos(angleRad);
	const sinA = std.sin(angleRad);
	const cu = __tsover_sub(au, __tsover_mul(aspect, .5));
	const cv = __tsover_sub(av, .5);
	const rotU = __tsover_add(__tsover_mul(cu, cosA), __tsover_mul(cv, sinA));
	const rotV = __tsover_add(-__tsover_mul(cu, sinA), __tsover_mul(cv, cosA));
	const zigzag = __tsover_mul(std.abs(__tsover_sub(std.fract(__tsover_mul(rotV, count)), .5)), 2);
	const stripeCoord = __tsover_add(__tsover_add(__tsover_add(__tsover_mul(rotU, count), __tsover_mul(zigzag, .5)), animTime), offset);
	const pattern = std.fract(stripeCoord);
	const fw = std.fwidth(stripeCoord);
	return std.smoothstep(__tsover_sub(__tsover_sub(balance, softness), fw), __tsover_add(__tsover_add(balance, softness), fw), pattern);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "count"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "balance"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "offset"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"au",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"av",
				[
					7,
					"uv",
					"y"
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"cu",
				[
					1,
					"au",
					"-",
					[
						1,
						"aspect",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"cv",
				[
					1,
					"av",
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"rotU",
				[
					1,
					[
						1,
						"cu",
						"*",
						"cosA"
					],
					"+",
					[
						1,
						"cv",
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"rotV",
				[
					1,
					[
						4,
						"-",
						[
							1,
							"cu",
							"*",
							"sinA"
						]
					],
					"+",
					[
						1,
						"cv",
						"*",
						"cosA"
					]
				]
			],
			[
				13,
				"zigzag",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"fract"
								],
								[[
									1,
									"rotV",
									"*",
									"count"
								]]
							],
							"-",
							[5, "0.5"]
						]]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"stripeCoord",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								"rotU",
								"*",
								"count"
							],
							"+",
							[
								1,
								"zigzag",
								"*",
								[5, "0.5"]
							]
						],
						"+",
						"animTime"
					],
					"+",
					"offset"
				]
			],
			[
				13,
				"pattern",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["stripeCoord"]
				]
			],
			[
				13,
				"fw",
				[
					6,
					[
						7,
						"std",
						"fwidth"
					],
					["stripeCoord"]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"smoothstep"
				],
				[
					[
						1,
						[
							1,
							"balance",
							"-",
							"softness"
						],
						"-",
						"fw"
					],
					[
						1,
						[
							1,
							"balance",
							"+",
							"softness"
						],
						"+",
						"fw"
					],
					"pattern"
				]
			]]
		]],
		"externalNames": ["DEG_TO_RAD", "std"]
	},
	externals: () => ({
		DEG_TO_RAD,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Chevron",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Animated chevron / zigzag stripe pattern",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "First color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Second color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		count: {
			default: 5,
			description: "Number of chevron pairs visible",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Count",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Rotation angle of the chevrons",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		balance: {
			default: .5,
			description: "Ratio of the two colors",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Balance",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Edge softness",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		speed: {
			default: 0,
			description: "Animation speed",
			ui: {
				type: "range",
				min: -2,
				max: 2,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		offset: {
			default: 0,
			description: "Phase offset for pattern positioning",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Offset",
				group: "Animation"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, propValues } = params;
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const animTime = animatedTime(params);
		const mask = call(chevronMask, "chevronMask", [
			uv,
			viewport,
			uniforms.count,
			uniforms.angle,
			uniforms.balance,
			uniforms.softness,
			uniforms.offset,
			animTime
		]);
		return vec4(call(mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			mask
		]).member("rgb"), mixExpr(uniforms.colorA.member("a"), uniforms.colorB.member("a"), mask));
	}
};
var Chevron_default = componentDefinition;
export { chevronMask as n, componentDefinition as r, Chevron_default as t };
