// shaders/dist/core/ChromaFlow-CuLYMObY.js
import { At as call, M as toHalfFloat, Pt as floatE, R as std, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor } from "./transformations-fqzznt4g.js";
var GRID_SIZE = 128;
var num = (v, f) => typeof v === "number" ? v : f;
const chromaFlowTapUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, dx, dy) => {
	"use gpu";
	return d.vec2f(__tsover_add(uv.x, dx), __tsover_add(uv.y, dy));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "dx"
			},
			{
				"type": "i",
				"name": "dy"
			}
		],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					7,
					"uv",
					"x"
				],
				"+",
				"dx"
			], [
				1,
				[
					7,
					"uv",
					"y"
				],
				"+",
				"dy"
			]]
		]]]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({})).$name("chromaFlowTapUV");
const chromaFlowColor = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((s0, s1, s2, s3, s4, base, up, down, left, right) => {
	"use gpu";
	const smoothedLiquid = __tsover_mul(__tsover_add(__tsover_add(__tsover_add(__tsover_add(s0.z, s1.z), s2.z), s3.z), s4.z), .2);
	const liquidIntensity = std.smoothstep(0, .1, smoothedLiquid);
	const flow = s0.xy.add(s1.xy).add(s2.xy).add(s3.xy).add(s4.xy).mul(.2);
	const flowMagnitude = std.length(flow);
	const hasFlow = std.smoothstep(.01, .1, flowMagnitude);
	const normalizedX = __tsover_div(flow.x, __tsover_add(flowMagnitude, .001));
	const normalizedY = __tsover_div(flow.y, __tsover_add(flowMagnitude, .001));
	const rightAmount = std.smoothstep(0, .7, std.max(normalizedX, 0));
	const leftAmount = std.smoothstep(0, .7, std.max(__tsover_mul(normalizedX, -1), 0));
	const upAmount = std.smoothstep(0, .7, std.max(normalizedY, 0));
	const downAmount = std.smoothstep(0, .7, std.max(__tsover_mul(normalizedY, -1), 0));
	const horizontalColor = left.mul(leftAmount).add(right.mul(rightAmount));
	const verticalColor = down.mul(downAmount).add(up.mul(upAmount));
	const horizontalWeight = __tsover_add(leftAmount, rightAmount);
	const verticalWeight = __tsover_add(upAmount, downAmount);
	const totalWeight = __tsover_add(__tsover_add(horizontalWeight, verticalWeight), .001);
	const directionalColor = horizontalColor.mul(__tsover_div(horizontalWeight, totalWeight)).add(verticalColor.mul(__tsover_div(verticalWeight, totalWeight)));
	return std.mix(base, directionalColor, d.vec4f(hasFlow)).mul(liquidIntensity);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "s0"
			},
			{
				"type": "i",
				"name": "s1"
			},
			{
				"type": "i",
				"name": "s2"
			},
			{
				"type": "i",
				"name": "s3"
			},
			{
				"type": "i",
				"name": "s4"
			},
			{
				"type": "i",
				"name": "base"
			},
			{
				"type": "i",
				"name": "up"
			},
			{
				"type": "i",
				"name": "down"
			},
			{
				"type": "i",
				"name": "left"
			},
			{
				"type": "i",
				"name": "right"
			}
		],
		"body": [0, [
			[
				13,
				"smoothedLiquid",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									[
										7,
										"s0",
										"z"
									],
									"+",
									[
										7,
										"s1",
										"z"
									]
								],
								"+",
								[
									7,
									"s2",
									"z"
								]
							],
							"+",
							[
								7,
								"s3",
								"z"
							]
						],
						"+",
						[
							7,
							"s4",
							"z"
						]
					],
					"*",
					[5, "0.2"]
				]
			],
			[
				13,
				"liquidIntensity",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[5, "0.1"],
						"smoothedLiquid"
					]
				]
			],
			[
				13,
				"flow",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										[
											6,
											[
												7,
												[
													6,
													[
														7,
														[
															7,
															"s0",
															"xy"
														],
														"add"
													],
													[[
														7,
														"s1",
														"xy"
													]]
												],
												"add"
											],
											[[
												7,
												"s2",
												"xy"
											]]
										],
										"add"
									],
									[[
										7,
										"s3",
										"xy"
									]]
								],
								"add"
							],
							[[
								7,
								"s4",
								"xy"
							]]
						],
						"mul"
					],
					[[5, "0.2"]]
				]
			],
			[
				13,
				"flowMagnitude",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["flow"]
				]
			],
			[
				13,
				"hasFlow",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.01"],
						[5, "0.1"],
						"flowMagnitude"
					]
				]
			],
			[
				13,
				"normalizedX",
				[
					1,
					[
						7,
						"flow",
						"x"
					],
					"/",
					[
						1,
						"flowMagnitude",
						"+",
						[5, "0.001"]
					]
				]
			],
			[
				13,
				"normalizedY",
				[
					1,
					[
						7,
						"flow",
						"y"
					],
					"/",
					[
						1,
						"flowMagnitude",
						"+",
						[5, "0.001"]
					]
				]
			],
			[
				13,
				"rightAmount",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[5, "0.7"],
						[
							6,
							[
								7,
								"std",
								"max"
							],
							["normalizedX", [5, "0"]]
						]
					]
				]
			],
			[
				13,
				"leftAmount",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[5, "0.7"],
						[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								1,
								"normalizedX",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							], [5, "0"]]
						]
					]
				]
			],
			[
				13,
				"upAmount",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[5, "0.7"],
						[
							6,
							[
								7,
								"std",
								"max"
							],
							["normalizedY", [5, "0"]]
						]
					]
				]
			],
			[
				13,
				"downAmount",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[5, "0.7"],
						[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								1,
								"normalizedY",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							], [5, "0"]]
						]
					]
				]
			],
			[
				13,
				"horizontalColor",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"left",
								"mul"
							],
							["leftAmount"]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"right",
							"mul"
						],
						["rightAmount"]
					]]
				]
			],
			[
				13,
				"verticalColor",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"down",
								"mul"
							],
							["downAmount"]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"up",
							"mul"
						],
						["upAmount"]
					]]
				]
			],
			[
				13,
				"horizontalWeight",
				[
					1,
					"leftAmount",
					"+",
					"rightAmount"
				]
			],
			[
				13,
				"verticalWeight",
				[
					1,
					"upAmount",
					"+",
					"downAmount"
				]
			],
			[
				13,
				"totalWeight",
				[
					1,
					[
						1,
						"horizontalWeight",
						"+",
						"verticalWeight"
					],
					"+",
					[5, "0.001"]
				]
			],
			[
				13,
				"directionalColor",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"horizontalColor",
								"mul"
							],
							[[
								1,
								"horizontalWeight",
								"/",
								"totalWeight"
							]]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"verticalColor",
							"mul"
						],
						[[
							1,
							"verticalWeight",
							"/",
							"totalWeight"
						]]
					]]
				]
			],
			[
				13,
				"finalColor",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"base",
						"directionalColor",
						[
							6,
							[
								7,
								"d",
								"vec4f"
							],
							["hasFlow"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"finalColor",
					"mul"
				],
				["liquidIntensity"]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({})).$name("chromaFlowColor");
const componentDefinition = {
	name: "ChromaFlow",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Interactive",
	description: "Interactive liquid flow effect that follows your cursor",
	usesPointer: true,
	props: {
		baseColor: {
			default: "#0066ff",
			transform: transformColor,
			description: "Base liquid color",
			ui: {
				type: "color",
				label: "Base Color",
				group: "Colors"
			}
		},
		upColor: {
			default: "#00ff00",
			transform: transformColor,
			description: "Color for upward movement",
			ui: {
				type: "color",
				label: "Up Color",
				group: "Colors"
			}
		},
		downColor: {
			default: "#ff0000",
			transform: transformColor,
			description: "Color for downward movement",
			ui: {
				type: "color",
				label: "Down Color",
				group: "Colors"
			}
		},
		leftColor: {
			default: "#0000ff",
			transform: transformColor,
			description: "Color for leftward movement",
			ui: {
				type: "color",
				label: "Left Color",
				group: "Colors"
			}
		},
		rightColor: {
			default: "#ffff00",
			transform: transformColor,
			description: "Color for rightward movement",
			ui: {
				type: "color",
				label: "Right Color",
				group: "Colors"
			}
		},
		intensity: {
			default: 1,
			description: "Strength of the liquid effect",
			ui: {
				type: "range",
				min: .5,
				max: 1.5,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		radius: {
			default: 3,
			description: "Radius of the liquid effect",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		momentum: {
			default: 30,
			description: "How much momentum colors retain in their flow direction",
			ui: {
				type: "range",
				min: 10,
				max: 60,
				step: 1,
				label: "Momentum",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, uvContext, onBeforeRender, onCleanup, dimensions, createDataTexture, registerMediaTexture, getCpuValue } = params;
		const fieldData = new Float32Array(GRID_SIZE * GRID_SIZE * 4);
		const tempFieldData = new Float32Array(GRID_SIZE * GRID_SIZE * 4);
		const texData = new Uint16Array(GRID_SIZE * GRID_SIZE * 4);
		const fieldTex = createDataTexture({
			width: GRID_SIZE,
			height: GRID_SIZE,
			format: "rgba16float",
			data: texData,
			label: "chromaflow-field"
		});
		onCleanup(() => fieldTex.destroy());
		const fieldKit = registerMediaTexture(() => fieldTex.texture);
		let mouseVelX = 0, mouseVelY = 0;
		let prevX = .5, prevY = .5;
		let lastTime = Date.now();
		let fieldMax = 0;
		onBeforeRender((fp) => {
			const { pointer } = fp;
			const currentTime = Date.now();
			const dt = Math.min((currentTime - lastTime) / 1e3, .016);
			lastTime = currentTime;
			const aspect = dimensions.width / Math.max(1, dimensions.height);
			const velX = dt > 0 ? (pointer.x - prevX) / dt : 0;
			const velY = dt > 0 ? (pointer.y - prevY) / dt : 0;
			mouseVelX = mouseVelX * .85 + velX * .15;
			mouseVelY = mouseVelY * .85 + velY * .15;
			prevX = pointer.x;
			prevY = pointer.y;
			const injecting = Math.abs(velX) + Math.abs(velY) > .01;
			if (!injecting && fieldMax < 1e-4) return;
			const intensity = num(getCpuValue("intensity"), 1);
			const radius = num(getCpuValue("radius"), 3) * .05;
			const momentum = num(getCpuValue("momentum"), 30);
			const flowFadeRate = 1 - dt / Math.max(.1, 1);
			const liquidFadeRate = 1 - dt;
			const flowSpeed = momentum * 50 * dt;
			let newMax = 0;
			for (let i = 0; i < GRID_SIZE; i++) for (let j = 0; j < GRID_SIZE; j++) {
				const idx = (i * GRID_SIZE + j) * 4;
				const fx0 = fieldData[idx];
				const fy0 = fieldData[idx + 1];
				tempFieldData[idx] = fx0 * flowFadeRate;
				tempFieldData[idx + 1] = fy0 * flowFadeRate;
				tempFieldData[idx + 2] = fieldData[idx + 2] * liquidFadeRate;
				if (Math.abs(fx0) > .001 || Math.abs(fy0) > .001) {
					const advectX = j - fx0 * flowSpeed;
					const advectY = i - fy0 * flowSpeed;
					const x0 = Math.floor(advectX);
					const y0 = Math.floor(advectY);
					const x1 = x0 + 1;
					const y1 = y0 + 1;
					if (x0 >= 0 && y0 >= 0 && x1 < GRID_SIZE && y1 < GRID_SIZE) {
						const fx = advectX - x0;
						const fy = advectY - y0;
						const idx00 = (y0 * GRID_SIZE + x0) * 4;
						const idx01 = (y0 * GRID_SIZE + x1) * 4;
						const idx10 = (y1 * GRID_SIZE + x0) * 4;
						const idx11 = (y1 * GRID_SIZE + x1) * 4;
						const sampledLiquid = fieldData[idx00 + 2] * (1 - fx) * (1 - fy) + fieldData[idx01 + 2] * fx * (1 - fy) + fieldData[idx10 + 2] * (1 - fx) * fy + fieldData[idx11 + 2] * fx * fy;
						tempFieldData[idx + 2] = sampledLiquid * liquidFadeRate;
					}
				}
				const m = Math.max(Math.abs(tempFieldData[idx]), Math.abs(tempFieldData[idx + 1]), tempFieldData[idx + 2]);
				if (m > newMax) newMax = m;
			}
			if (injecting) {
				const speed = Math.sqrt(mouseVelX * mouseVelX + mouseVelY * mouseVelY);
				const effectiveRadius = radius * Math.min(speed * speed * 20, 1);
				const maxDistSq = effectiveRadius * 2 * (effectiveRadius * 2);
				const radSq = effectiveRadius * effectiveRadius;
				const addScale = intensity * 100 * dt * .01;
				const speedMultiplier = Math.min(speed * 10, 1);
				for (let i = 0; i < GRID_SIZE; i++) for (let j = 0; j < GRID_SIZE; j++) {
					const cellX = (j + .5) / GRID_SIZE;
					const cellY = (i + .5) / GRID_SIZE;
					const dx = aspect >= 1 ? (cellX - pointer.x) * aspect : cellX - pointer.x;
					const dy = aspect >= 1 ? cellY - pointer.y : (cellY - pointer.y) / aspect;
					const distSq = dx * dx + dy * dy;
					if (distSq < maxDistSq) {
						const idx = (i * GRID_SIZE + j) * 4;
						const influence = Math.exp(-distSq / radSq);
						tempFieldData[idx] += mouseVelX * influence * addScale;
						tempFieldData[idx + 1] += mouseVelY * influence * addScale;
						tempFieldData[idx + 2] += influence * addScale * speedMultiplier;
						tempFieldData[idx] = Math.max(-1, Math.min(1, tempFieldData[idx]));
						tempFieldData[idx + 1] = Math.max(-1, Math.min(1, tempFieldData[idx + 1]));
						tempFieldData[idx + 2] = Math.max(0, Math.min(1, tempFieldData[idx + 2]));
					}
				}
			}
			fieldMax = injecting ? 1 : newMax;
			fieldData.set(tempFieldData);
			for (let k = 0; k < texData.length; k++) texData[k] = toHalfFloat(fieldData[k]);
			fieldTex.write(texData);
		});
		const uv = uvContext ?? ctx.uv;
		const px = 1 / GRID_SIZE;
		const clamp = "linearClamp";
		const tap = (dx, dy) => fieldKit.sample(call(chromaFlowTapUV, "chromaFlowTapUV", [
			uv,
			floatE(dx),
			floatE(dy)
		]), clamp);
		return call(chromaFlowColor, "chromaFlowColor", [
			tap(0, 0),
			tap(px, 0),
			tap(0, px),
			tap(-px, 0),
			tap(0, -px),
			uniforms.baseColor,
			uniforms.upColor,
			uniforms.downColor,
			uniforms.leftColor,
			uniforms.rightColor
		]);
	}
};
var ChromaFlow_default = componentDefinition;
export { componentDefinition as i, chromaFlowColor as n, chromaFlowTapUV as r, ChromaFlow_default as t };
