// shaders/dist/core/ChromaticAberration-BrGXd7kq.js
import { At as call, It as vec4, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { r as transformAngle } from "./transformations-fqzznt4g.js";
const chromaticOffsetUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, angle, strength, aspect, channelOffset) => {
	"use gpu";
	const angleRad = std.radians(angle);
	const direction = d.vec2f(__tsover_div(std.cos(angleRad), aspect), std.sin(angleRad));
	const scaledStrength = __tsover_mul(strength, .1);
	const offsetBase = direction.mul(scaledStrength);
	return uv.add(offsetBase.mul(channelOffset));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "channelOffset"
			}
		],
		"body": [0, [
			[
				13,
				"angleRad",
				[
					6,
					[
						7,
						"std",
						"radians"
					],
					["angle"]
				]
			],
			[
				13,
				"direction",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"cos"
							],
							["angleRad"]
						],
						"/",
						"aspect"
					], [
						6,
						[
							7,
							"std",
							"sin"
						],
						["angleRad"]
					]]
				]
			],
			[
				13,
				"scaledStrength",
				[
					1,
					"strength",
					"*",
					[5, "0.1"]
				]
			],
			[
				13,
				"offsetBase",
				[
					6,
					[
						7,
						"direction",
						"mul"
					],
					["scaledStrength"]
				]
			],
			[10, [
				6,
				[
					7,
					"uv",
					"add"
				],
				[[
					6,
					[
						7,
						"offsetBase",
						"mul"
					],
					["channelOffset"]
				]]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "ChromaticAberration",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Separate RGB channels for a prismatic distortion effect",
	requiresRTT: true,
	requiresChild: true,
	props: {
		strength: {
			default: .2,
			description: "Overall strength of the chromatic aberration effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Strength",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Direction of the chromatic aberration in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		redOffset: {
			default: -1,
			description: "Red channel offset multiplier",
			ui: {
				type: ["range", "map"],
				min: -2,
				max: 2,
				step: .1,
				label: "Red Offset",
				group: "Effect"
			}
		},
		greenOffset: {
			default: 0,
			description: "Green channel offset multiplier",
			ui: {
				type: ["range", "map"],
				min: -2,
				max: 2,
				step: .1,
				label: "Green Offset",
				group: "Effect"
			}
		},
		blueOffset: {
			default: 1,
			description: "Blue channel offset multiplier",
			ui: {
				type: ["range", "map"],
				min: -2,
				max: 2,
				step: .1,
				label: "Blue Offset",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const redUV = call(chromaticOffsetUV, "chromaticOffsetUV", [
			ctx.uv,
			uniforms.angle,
			uniforms.strength,
			ctx.aspect,
			uniforms.redOffset
		]);
		const greenUV = call(chromaticOffsetUV, "chromaticOffsetUV", [
			ctx.uv,
			uniforms.angle,
			uniforms.strength,
			ctx.aspect,
			uniforms.greenOffset
		]);
		const blueUV = call(chromaticOffsetUV, "chromaticOffsetUV", [
			ctx.uv,
			uniforms.angle,
			uniforms.strength,
			ctx.aspect,
			uniforms.blueOffset
		]);
		const redSample = texture.sample(redUV);
		const greenSample = texture.sample(greenUV);
		const blueSample = texture.sample(blueUV);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [vec4(redSample.member("r"), greenSample.member("g"), blueSample.member("b"), greenSample.member("a"))]);
	}
};
var ChromaticAberration_default = componentDefinition;
export { chromaticOffsetUV as n, componentDefinition as r, ChromaticAberration_default as t };
