// shaders/dist/core/Circle-BLNjdX3R.js
import { At as call, It as vec4, O as strokeMaskFromSdf, Qt as mixColorsVariants, T as sdf_exports, Zt as mixColorsLinear, h as circleSdf, n as tgpu, on as d, p as ShapeMask } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition, l as transformStrokePosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
const circleShape = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32
], ShapeMask)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, radius, softness, strokeThickness, strokePosition, uv, aspect) => {
	"use gpu";
	const aspectUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y));
	return strokeMaskFromSdf(circleSdf(__tsover_sub(aspectUV.x, centerPos.x), __tsover_sub(aspectUV.y, centerPos.y), __tsover_mul(radius, .5)), softness, strokeThickness, strokePosition);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "radius"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "strokeThickness"
			},
			{
				"type": "i",
				"name": "strokePosition"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						7,
						"aspectUV",
						"x"
					],
					"-",
					[
						7,
						"centerPos",
						"x"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"aspectUV",
						"y"
					],
					"-",
					[
						7,
						"centerPos",
						"y"
					]
				]
			],
			[
				13,
				"circleEdge",
				[
					1,
					"radius",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"sdf",
						"circleSdf"
					],
					[
						"dx",
						"dy",
						"circleEdge"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"sdf",
					"strokeMaskFromSdf"
				],
				[
					"dist",
					"softness",
					"strokeThickness",
					"strokePosition"
				]
			]]
		]],
		"externalNames": ["d", "sdf"]
	},
	externals: () => ({
		d,
		sdf: sdf_exports
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Circle",
	category: "Shapes",
	description: "Generate a circle with adjustable size and softness",
	acceptsUVContext: true,
	boundingBoxDeclaration: { propBindings: {
		x: {
			prop: "center",
			as: "position-x"
		},
		y: {
			prop: "center",
			as: "position-y"
		},
		width: {
			prop: "radius",
			as: "canvas-height"
		},
		height: {
			prop: "radius",
			as: "canvas-height"
		}
	} },
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		color: {
			default: "#ffffff",
			transform: transformColor,
			description: "The color of the circle",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		radius: {
			default: 1,
			description: "The radius of the circle. A value of one (1) is sets the circle to fit the canvas.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .01,
				label: "Radius",
				group: "Effect",
				units: ["%", "px"]
			}
		},
		softness: {
			default: 0,
			description: "Edge softness. Lower values like zero (0) are sharp, higher values like one (1) are softer.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect",
				dimensional: "canvas-height"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the circle",
			ui: {
				type: "position",
				label: "Center Position",
				group: "Position",
				units: ["%", "px"]
			}
		},
		strokeThickness: {
			default: 0,
			description: "The thickness of the stroke outline. Zero (0) means no stroke.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .5,
				step: .01,
				label: "Stroke Thickness",
				group: "Stroke"
			}
		},
		strokeColor: {
			default: "#000000",
			transform: transformColor,
			description: "The color of the stroke outline",
			ui: {
				type: "color",
				label: "Stroke Color",
				group: "Colors"
			}
		},
		strokePosition: {
			default: "center",
			transform: transformStrokePosition,
			description: "Position of the stroke relative to the circle edge",
			ui: {
				type: "select",
				options: [
					{
						label: "Outside",
						value: "outside"
					},
					{
						label: "Center",
						value: "center"
					},
					{
						label: "Inside",
						value: "inside"
					}
				],
				label: "Stroke Position",
				group: "Stroke"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for blending fill and stroke colors in soft edges",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Blending",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const uv = params.uvContext ?? ctx.uv;
		const colorSpaceMode = propValues.colorSpace ?? 0;
		const mask = call(circleShape, "circleShape", [
			uniforms.center,
			uniforms.radius,
			uniforms.softness,
			uniforms.strokeThickness,
			uniforms.strokePosition,
			uv,
			ctx.aspect
		]);
		const blended = call(mixColorsVariants[colorSpaceMode] ?? mixColorsLinear, "mixColors", [
			uniforms.color,
			uniforms.strokeColor,
			mask.member("strokeBlend")
		]);
		return vec4(blended.member("rgb"), blended.member("a").mul(mask.member("overallMask")));
	}
};
var Circle_default = componentDefinition;
export { circleShape as n, componentDefinition as r, Circle_default as t };
