// shaders/dist/core/ColorWheel-CI_qR6vL.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
var DEG_TO_RAD = .0174533;
const colorWheelT = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, angle, scale, animTime) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const cosA = std.cos(angleRad);
	const sinA = std.sin(angleRad);
	const projectedCoord = __tsover_add(__tsover_mul(__tsover_mul(uv.x, aspect), cosA), __tsover_mul(uv.y, sinA));
	return std.fract(__tsover_add(__tsover_mul(__tsover_mul(projectedCoord, scale), .2), animTime));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"projectedCoord",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"*",
							"aspect"
						],
						"*",
						"cosA"
					],
					"+",
					[
						1,
						[
							7,
							"uv",
							"y"
						],
						"*",
						"sinA"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"fract"
				],
				[[
					1,
					[
						1,
						[
							1,
							"projectedCoord",
							"*",
							"scale"
						],
						"*",
						[5, "0.2"]
					],
					"+",
					"animTime"
				]]
			]]
		]],
		"externalNames": ["DEG_TO_RAD", "std"]
	},
	externals: () => ({
		DEG_TO_RAD,
		std
	})
}) && $.f)({}));
const colorWheelRainbow = tgpu.fn([d.f32], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((t) => {
	"use gpu";
	const h6 = __tsover_mul(t, 6);
	const r = std.clamp(__tsover_sub(std.abs(__tsover_sub(h6, 3)), 1), 0, 1);
	const g = std.clamp(__tsover_sub(2, std.abs(__tsover_sub(h6, 2))), 0, 1);
	const b = std.clamp(__tsover_sub(2, std.abs(__tsover_sub(h6, 4))), 0, 1);
	return d.vec4f(r, g, b, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "t"
		}],
		"body": [0, [
			[
				13,
				"h6",
				[
					1,
					"t",
					"*",
					[5, "6"]
				]
			],
			[
				13,
				"r",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"h6",
									"-",
									[5, "3"]
								]]
							],
							"-",
							[5, "1"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"g",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[5, "2"],
							"-",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"h6",
									"-",
									[5, "2"]
								]]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"b",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[5, "2"],
							"-",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"h6",
									"-",
									[5, "4"]
								]]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					"r",
					"g",
					"b",
					[5, "1"]
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const colorWheelPhase = tgpu.fn([d.f32], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((t) => {
	"use gpu";
	const t3 = __tsover_mul(t, 3);
	return d.vec2f(std.fract(t3), std.floor(t3));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "t"
		}],
		"body": [0, [[
			13,
			"t3",
			[
				1,
				"t",
				"*",
				[5, "3"]
			]
		], [10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				6,
				[
					7,
					"std",
					"fract"
				],
				["t3"]
			], [
				6,
				[
					7,
					"std",
					"floor"
				],
				["t3"]
			]]
		]]]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const colorWheelSelect = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((c01, c12, c20, floorT3) => {
	"use gpu";
	const seg1 = std.mix(c20, c12, std.step(floorT3, 1.5));
	return std.mix(seg1, c01, std.step(floorT3, .5));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "c01"
			},
			{
				"type": "i",
				"name": "c12"
			},
			{
				"type": "i",
				"name": "c20"
			},
			{
				"type": "i",
				"name": "floorT3"
			}
		],
		"body": [0, [[
			13,
			"seg1",
			[
				6,
				[
					7,
					"std",
					"mix"
				],
				[
					"c20",
					"c12",
					[
						6,
						[
							7,
							"std",
							"step"
						],
						["floorT3", [5, "1.5"]]
					]
				]
			]
		], [10, [
			6,
			[
				7,
				"std",
				"mix"
			],
			[
				"seg1",
				"c01",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					["floorT3", [5, "0.5"]]
				]
			]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const componentDefinition = {
	name: "ColorWheel",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "A directional gradient that smoothly cycles through rainbow colors or a custom set of three colors",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		mode: {
			default: "rainbow",
			compileTime: true,
			description: "Rainbow cycles through the full spectrum; Custom loops through your three chosen colors",
			ui: {
				type: "select",
				options: [{
					label: "Rainbow",
					value: "rainbow"
				}, {
					label: "Custom",
					value: "custom"
				}],
				label: "Mode",
				group: "Colors"
			}
		},
		colorA: {
			default: "#ff0000",
			transform: transformColor,
			description: "First color in the cycle",
			ui: {
				type: "color",
				label: "Color 1",
				group: "Colors",
				condition: { mode: "custom" }
			}
		},
		colorB: {
			default: "#00ff88",
			transform: transformColor,
			description: "Second color in the cycle",
			ui: {
				type: "color",
				label: "Color 2",
				group: "Colors",
				condition: { mode: "custom" }
			}
		},
		colorC: {
			default: "#0066ff",
			transform: transformColor,
			description: "Third color in the cycle",
			ui: {
				type: "color",
				label: "Color 3",
				group: "Colors",
				condition: { mode: "custom" }
			}
		},
		scale: {
			default: 1,
			description: "Number of color cycles across the viewport",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 10,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Direction the gradient flows",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		speed: {
			default: .05,
			description: "Speed at which the gradient cycles",
			ui: {
				type: "range",
				min: -1,
				max: 1,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		colorSpace: {
			default: "oklch",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for blending between custom colors",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors",
				condition: { mode: "custom" }
			}
		}
	},
	fragment: (params) => {
		const { uniforms, propValues } = params;
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const animTime = animatedTime(params);
		const t = call(colorWheelT, "colorWheelT", [
			uv,
			viewport,
			uniforms.angle,
			uniforms.scale,
			animTime
		]);
		if ((propValues.mode ?? "rainbow") === "rainbow") return call(colorWheelRainbow, "colorWheelRainbow", [t]);
		const phase = call(colorWheelPhase, "colorWheelPhase", [t]);
		const si = phase.member("x");
		const floorT3 = phase.member("y");
		const variant = mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear;
		return call(colorWheelSelect, "colorWheelSelect", [
			call(variant, "mixColors", [
				uniforms.colorA,
				uniforms.colorB,
				si
			]),
			call(variant, "mixColors", [
				uniforms.colorB,
				uniforms.colorC,
				si
			]),
			call(variant, "mixColors", [
				uniforms.colorC,
				uniforms.colorA,
				si
			]),
			floorT3
		]);
	}
};
var ColorWheel_default = componentDefinition;
export { colorWheelT as a, colorWheelSelect as i, colorWheelPhase as n, componentDefinition as o, colorWheelRainbow as r, ColorWheel_default as t };
