// shaders/dist/core/ConcentricSpin-FJuBPPHi.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-BlwgBpJf.js";
import { c as transformPosition, s as transformEdges } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
var PI = Math.PI;
var TAU = Math.PI * 2;
const csHash = tgpu.fn([d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((x) => {
	"use gpu";
	return __tsover_mul(__tsover_sub(std.fract(__tsover_mul(std.sin(__tsover_mul(x, 12.9898)), 43758.5453)), .5), 2);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "x"
		}],
		"body": [0, [[10, [
			1,
			[
				1,
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								"x",
								"*",
								[5, "12.9898"]
							]]
						],
						"*",
						[5, "43758.5453"]
					]]
				],
				"-",
				[5, "0.5"]
			],
			"*",
			[5, "2"]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const csRotateUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, center, angle) => {
	"use gpu";
	const centerPos = d.vec2f(center.x, __tsover_sub(1, center.y));
	const dx = __tsover_mul(__tsover_sub(uv.x, centerPos.x), aspect);
	const dy = __tsover_sub(uv.y, centerPos.y);
	const cosA = std.cos(angle);
	const sinA = std.sin(angle);
	const rotX = __tsover_sub(__tsover_mul(dx, cosA), __tsover_mul(dy, sinA));
	const rotY = __tsover_add(__tsover_mul(dx, sinA), __tsover_mul(dy, cosA));
	return d.vec2f(__tsover_add(__tsover_div(rotX, aspect), centerPos.x), __tsover_add(rotY, centerPos.y));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "angle"
			}
		],
		"body": [0, [
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"center",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"-",
						[
							7,
							"centerPos",
							"x"
						]
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					[
						7,
						"centerPos",
						"y"
					]
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angle"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angle"]
				]
			],
			[
				13,
				"rotX",
				[
					1,
					[
						1,
						"dx",
						"*",
						"cosA"
					],
					"-",
					[
						1,
						"dy",
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"rotY",
				[
					1,
					[
						1,
						"dx",
						"*",
						"sinA"
					],
					"+",
					[
						1,
						"dy",
						"*",
						"cosA"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						1,
						"rotX",
						"/",
						"aspect"
					],
					"+",
					[
						7,
						"centerPos",
						"x"
					]
				], [
					1,
					"rotY",
					"+",
					[
						7,
						"centerPos",
						"y"
					]
				]]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const concentricSpinUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, intensity, rings, smoothness, seed, speedRandomness, animTime, uv, aspect) => {
	"use gpu";
	const centerPos = d.vec2f(center.x, __tsover_sub(1, center.y));
	const dx = __tsover_mul(__tsover_sub(uv.x, centerPos.x), aspect);
	const dy = __tsover_sub(uv.y, centerPos.y);
	const dist = std.length(d.vec2f(dx, dy));
	const ringCoord = __tsover_mul(dist, rings);
	const ringIndex = std.floor(ringCoord);
	const ringFrac = std.fract(ringCoord);
	const maxAngle = __tsover_mul(intensity, .01745);
	const seedOffset = __tsover_mul(seed, .137);
	const staticA = __tsover_mul(csHash(__tsover_add(ringIndex, seedOffset)), maxAngle);
	const staticB = __tsover_mul(csHash(__tsover_add(__tsover_add(ringIndex, 1), seedOffset)), maxAngle);
	const ringSpeedA = std.mix(1, csHash(__tsover_add(ringIndex, 42.7)), speedRandomness);
	const ringSpeedB = std.mix(1, csHash(__tsover_add(__tsover_add(ringIndex, 1), 42.7)), speedRandomness);
	const animA = __tsover_mul(__tsover_mul(animTime, .25), ringSpeedA);
	const animB = __tsover_mul(__tsover_mul(animTime, .25), ringSpeedB);
	const totalA = __tsover_add(staticA, animA);
	const totalB = __tsover_add(staticB, animB);
	const wrapArg = __tsover_add(__tsover_sub(totalB, totalA), PI);
	const diff = __tsover_sub(__tsover_sub(wrapArg, __tsover_mul(TAU, std.floor(__tsover_div(wrapArg, TAU)))), PI);
	const halfEdge = __tsover_add(__tsover_mul(smoothness, .5), .001);
	const blend = std.smoothstep(__tsover_sub(.5, halfEdge), __tsover_add(.5, halfEdge), ringFrac);
	return csRotateUV(uv, aspect, center, __tsover_add(totalA, __tsover_mul(diff, blend)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "rings"
			},
			{
				"type": "i",
				"name": "smoothness"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "speedRandomness"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"center",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"-",
						[
							7,
							"centerPos",
							"x"
						]
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					[
						7,
						"centerPos",
						"y"
					]
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						["dx", "dy"]
					]]
				]
			],
			[
				13,
				"ringCoord",
				[
					1,
					"dist",
					"*",
					"rings"
				]
			],
			[
				13,
				"ringIndex",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["ringCoord"]
				]
			],
			[
				13,
				"ringFrac",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["ringCoord"]
				]
			],
			[
				13,
				"maxAngle",
				[
					1,
					"intensity",
					"*",
					[5, "0.01745"]
				]
			],
			[
				13,
				"seedOffset",
				[
					1,
					"seed",
					"*",
					[5, "0.137"]
				]
			],
			[
				13,
				"staticA",
				[
					1,
					[
						6,
						"csHash",
						[[
							1,
							"ringIndex",
							"+",
							"seedOffset"
						]]
					],
					"*",
					"maxAngle"
				]
			],
			[
				13,
				"staticB",
				[
					1,
					[
						6,
						"csHash",
						[[
							1,
							[
								1,
								"ringIndex",
								"+",
								[5, "1"]
							],
							"+",
							"seedOffset"
						]]
					],
					"*",
					"maxAngle"
				]
			],
			[
				13,
				"ringSpeedA",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "1"],
						[
							6,
							"csHash",
							[[
								1,
								"ringIndex",
								"+",
								[5, "42.7"]
							]]
						],
						"speedRandomness"
					]
				]
			],
			[
				13,
				"ringSpeedB",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "1"],
						[
							6,
							"csHash",
							[[
								1,
								[
									1,
									"ringIndex",
									"+",
									[5, "1"]
								],
								"+",
								[5, "42.7"]
							]]
						],
						"speedRandomness"
					]
				]
			],
			[
				13,
				"animA",
				[
					1,
					[
						1,
						"animTime",
						"*",
						[5, "0.25"]
					],
					"*",
					"ringSpeedA"
				]
			],
			[
				13,
				"animB",
				[
					1,
					[
						1,
						"animTime",
						"*",
						[5, "0.25"]
					],
					"*",
					"ringSpeedB"
				]
			],
			[
				13,
				"totalA",
				[
					1,
					"staticA",
					"+",
					"animA"
				]
			],
			[
				13,
				"totalB",
				[
					1,
					"staticB",
					"+",
					"animB"
				]
			],
			[
				13,
				"wrapArg",
				[
					1,
					[
						1,
						"totalB",
						"-",
						"totalA"
					],
					"+",
					"PI"
				]
			],
			[
				13,
				"diff",
				[
					1,
					[
						1,
						"wrapArg",
						"-",
						[
							1,
							"TAU",
							"*",
							[
								6,
								[
									7,
									"std",
									"floor"
								],
								[[
									1,
									"wrapArg",
									"/",
									"TAU"
								]]
							]
						]
					],
					"-",
					"PI"
				]
			],
			[
				13,
				"halfEdge",
				[
					1,
					[
						1,
						"smoothness",
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.001"]
				]
			],
			[
				13,
				"blend",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							[5, "0.5"],
							"-",
							"halfEdge"
						],
						[
							1,
							[5, "0.5"],
							"+",
							"halfEdge"
						],
						"ringFrac"
					]
				]
			],
			[
				13,
				"angle",
				[
					1,
					"totalA",
					"+",
					[
						1,
						"diff",
						"*",
						"blend"
					]
				]
			],
			[10, [
				6,
				"csRotateUV",
				[
					"uv",
					"aspect",
					"center",
					"angle"
				]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"csHash",
			"PI",
			"TAU",
			"csRotateUV"
		]
	},
	externals: () => ({
		d,
		std,
		csHash,
		PI,
		TAU,
		csRotateUV
	})
}) && $.f)({}));
const componentDefinition = {
	name: "ConcentricSpin",
	category: "Distortions",
	description: "Concentric rings that each rotate the underlying image by different amounts",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	animatedTime: { speed: "speed" },
	props: {
		intensity: {
			default: 20,
			description: "Maximum rotation angle per ring",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Effect"
			}
		},
		rings: {
			default: 8,
			description: "Number of concentric rings",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Rings",
				group: "Effect"
			}
		},
		smoothness: {
			default: .03,
			description: "Softness of transitions between rings",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Smoothness",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Randomization seed for per-ring rotation variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		speed: {
			default: .1,
			description: "Speed of continuous ring rotation",
			ui: {
				type: "range",
				min: -5,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		speedRandomness: {
			default: .5,
			description: "How much each ring varies in rotation speed and direction",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Speed Randomness",
				group: "Animation"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the concentric rings",
			ui: {
				type: "position",
				label: "Center Position",
				group: "Position"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, childNode, ctx, propValues, convertToTexture } = params;
		if (!childNode) {
			console.error("You must pass a child component into the ConcentricSpin shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		const t = animatedTime(params);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(concentricSpinUV, "concentricSpinUV", [
			uniforms.center,
			uniforms.intensity,
			uniforms.rings,
			uniforms.smoothness,
			uniforms.seed,
			uniforms.speedRandomness,
			t,
			ctx.uv,
			ctx.aspect
		]), (uv) => texture.sample(uv), edgeMode)]);
	},
	uvRemap: (ctx) => {
		const edgeMode = ctx.propValues.edges ?? 0;
		const t = animatedTime(ctx);
		return composeEdgeRemapExpr(call(concentricSpinUV, "concentricSpinUV", [
			ctx.uniforms.center,
			ctx.uniforms.intensity,
			ctx.uniforms.rings,
			ctx.uniforms.smoothness,
			ctx.uniforms.seed,
			ctx.uniforms.speedRandomness,
			t,
			ctx.uv,
			ctx.aspect
		]), ctx.mask, edgeMode);
	}
};
var ConcentricSpin_default = componentDefinition;
export { csRotateUV as a, csHash as i, componentDefinition as n, concentricSpinUV as r, ConcentricSpin_default as t };
