// shaders/dist/core/ConicGradient-C3wD1h2z.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, gt as mixColorStopsRuntime, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig } from "./porters-Dg80Yg90.js";
var TAU = 6.283185307179586;
var INV_360 = 1 / 360;
const conicFinalT = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.vec2f,
	d.vec2f
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, rotation, repeat, uv, viewport) => {
	"use gpu";
	const safeViewportY = std.max(viewport.y, 1e-6);
	const aspect = __tsover_div(viewport.x, safeViewportY);
	const aspectUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y));
	const dx = __tsover_sub(aspectUV.x, centerPos.x);
	const dy = __tsover_sub(aspectUV.y, centerPos.y);
	const angle = std.atan2(dy, dx);
	const normalized = __tsover_sub(.5, __tsover_div(angle, TAU));
	const rotationFraction = __tsover_mul(rotation, INV_360);
	return std.fract(__tsover_mul(__tsover_add(normalized, rotationFraction), repeat));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "rotation"
			},
			{
				"type": "i",
				"name": "repeat"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			}
		],
		"body": [0, [
			[
				13,
				"safeViewportY",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						7,
						"viewport",
						"y"
					], [5, "0.000001"]]
				]
			],
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					"safeViewportY"
				]
			],
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						7,
						"aspectUV",
						"x"
					],
					"-",
					[
						7,
						"centerPos",
						"x"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"aspectUV",
						"y"
					],
					"-",
					[
						7,
						"centerPos",
						"y"
					]
				]
			],
			[
				13,
				"angle",
				[
					6,
					[
						7,
						"std",
						"atan2"
					],
					["dy", "dx"]
				]
			],
			[
				13,
				"normalized",
				[
					1,
					[5, "0.5"],
					"-",
					[
						1,
						"angle",
						"/",
						"TAU"
					]
				]
			],
			[
				13,
				"rotationFraction",
				[
					1,
					"rotation",
					"*",
					"INV_360"
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"fract"
				],
				[[
					1,
					[
						1,
						"normalized",
						"+",
						"rotationFraction"
					],
					"*",
					"repeat"
				]]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"TAU",
			"INV_360"
		]
	},
	externals: () => ({
		std,
		d,
		TAU,
		INV_360
	})
}) && $.f)({}));
function gradientColor(params, finalT) {
	const { uniforms, propValues } = params;
	const colorSpaceMode = propValues.colorSpace ?? 0;
	if ((propValues.stopCount ?? 0) > 1) return mixColorStopsRuntime(finalT, {
		colorsArray: uniforms.colorsArray,
		positionsArray: uniforms.positionsArray,
		convertedColorsArray: uniforms.convertedColorsArray,
		stopCount: uniforms.stopCount
	}, colorSpaceMode);
	return call(mixColorsVariants[colorSpaceMode] ?? mixColorsLinear, "mixColors", [
		uniforms.colorA,
		uniforms.colorB,
		finalT
	]);
}
const componentDefinition = {
	name: "ConicGradient",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Colors sweep in a full circle around a center point, like a color wheel",
	acceptsUVContext: true,
	props: {
		colorA: {
			default: "#FF0080",
			transform: transformColor,
			description: "Starting color of the sweep",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#00BFFF",
			transform: transformColor,
			description: "Ending color of the sweep (wraps back to Color A)",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		stops: colorStopsPropConfig(),
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the sweep",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation offset in degrees — shifts where Color A begins",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		repeat: {
			default: 1,
			description: "Number of times the gradient repeats around the circle. Values above 1 create a starburst pattern.",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 24,
				step: 1,
				label: "Repeat",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		return gradientColor(params, call(conicFinalT, "conicFinalT", [
			params.uniforms.center,
			params.uniforms.rotation,
			params.uniforms.repeat,
			uv,
			viewport
		]));
	}
};
var ConicGradient_default = componentDefinition;
export { componentDefinition as n, conicFinalT as r, ConicGradient_default as t };
