// shaders/dist/core/ContourLines-C3Hwaw1Y.js
import { At as call, It as vec4, Ot as ZERO, Pt as floatE, R as std, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, i as transformBoolean } from "./transformations-fqzznt4g.js";
var transformSource = (value) => value === "alpha" ? 1 : 0;
var transformColorMode = (value) => value === "custom" ? 1 : 0;
var sourceModeOf = (raw) => typeof raw === "number" ? raw : raw === "alpha" ? 1 : 0;
var colorModeOf = (raw) => typeof raw === "number" ? raw : raw === "custom" ? 1 : 0;
const contourLuma = tgpu.fn([d.vec4f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color) => {
	"use gpu";
	return std.dot(d.vec3f(color.x, color.y, color.z), d.vec3f(.2126, .7152, .0722));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"dot"
			],
			[[
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[
						7,
						"color",
						"x"
					],
					[
						7,
						"color",
						"y"
					],
					[
						7,
						"color",
						"z"
					]
				]
			], [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[5, "0.2126"],
					[5, "0.7152"],
					[5, "0.0722"]
				]
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const contourLineMask = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((sourceValue, levels, lineWidth, softness, gamma, invert) => {
	"use gpu";
	const invertedValue = std.select(sourceValue, __tsover_sub(1, sourceValue), invert > .5);
	const sv = std.pow(invertedValue, gamma);
	const scaledValue = __tsover_add(__tsover_mul(sv, levels), .5);
	const fw = std.max(std.fwidth(scaledValue), 1e-4);
	const distFromBoundary = std.abs(__tsover_sub(std.fract(__tsover_add(scaledValue, .5)), .5));
	const normalizedDist = __tsover_div(distFromBoundary, fw);
	const transitionFactor = __tsover_add(__tsover_mul(softness, .99), .01);
	const innerEdge = __tsover_mul(lineWidth, __tsover_sub(1, transitionFactor));
	return __tsover_sub(1, std.smoothstep(innerEdge, lineWidth, normalizedDist));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "sourceValue"
			},
			{
				"type": "i",
				"name": "levels"
			},
			{
				"type": "i",
				"name": "lineWidth"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "gamma"
			},
			{
				"type": "i",
				"name": "invert"
			}
		],
		"body": [0, [
			[
				13,
				"invertedValue",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"sourceValue",
						[
							1,
							[5, "1"],
							"-",
							"sourceValue"
						],
						[
							1,
							"invert",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"sv",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					["invertedValue", "gamma"]
				]
			],
			[
				13,
				"scaledValue",
				[
					1,
					[
						1,
						"sv",
						"*",
						"levels"
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"fw",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"fwidth"
						],
						["scaledValue"]
					], [5, "0.0001"]]
				]
			],
			[
				13,
				"distFromBoundary",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								1,
								"scaledValue",
								"+",
								[5, "0.5"]
							]]
						],
						"-",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"normalizedDist",
				[
					1,
					"distFromBoundary",
					"/",
					"fw"
				]
			],
			[
				13,
				"transitionFactor",
				[
					1,
					[
						1,
						"softness",
						"*",
						[5, "0.99"]
					],
					"+",
					[5, "0.01"]
				]
			],
			[
				13,
				"innerEdge",
				[
					1,
					"lineWidth",
					"*",
					[
						1,
						[5, "1"],
						"-",
						"transitionFactor"
					]
				]
			],
			[10, [
				1,
				[5, "1"],
				"-",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						"innerEdge",
						"lineWidth",
						"normalizedDist"
					]
				]
			]]
		]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const contourCompose = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((bg, line, mask) => {
	"use gpu";
	const rgb = std.mix(d.vec3f(bg.x, bg.y, bg.z), d.vec3f(line.x, line.y, line.z), d.vec3f(mask));
	const alpha = std.mix(bg.w, line.w, mask);
	return d.vec4f(rgb.x, rgb.y, rgb.z, alpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "bg"
			},
			{
				"type": "i",
				"name": "line"
			},
			{
				"type": "i",
				"name": "mask"
			}
		],
		"body": [0, [
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									7,
									"bg",
									"x"
								],
								[
									7,
									"bg",
									"y"
								],
								[
									7,
									"bg",
									"z"
								]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									7,
									"line",
									"x"
								],
								[
									7,
									"line",
									"y"
								],
								[
									7,
									"line",
									"z"
								]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["mask"]
						]
					]
				]
			],
			[
				13,
				"alpha",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							7,
							"bg",
							"w"
						],
						[
							7,
							"line",
							"w"
						],
						"mask"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"rgb",
						"x"
					],
					[
						7,
						"rgb",
						"y"
					],
					[
						7,
						"rgb",
						"z"
					],
					"alpha"
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "ContourLines",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Draw topographical contour lines based on luminance or alpha",
	requiresChild: true,
	blendWithChildren: false,
	props: {
		levels: {
			default: 5,
			description: "Number of contour levels",
			ui: {
				type: ["range", "map"],
				min: 2,
				max: 30,
				step: 1,
				label: "Levels",
				group: "Effect"
			}
		},
		lineWidth: {
			default: 2,
			description: "Width of the contour lines in pixels",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 5,
				step: .1,
				label: "Line Width",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Edge softness of the lines (0 = sharp, 1 = soft)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		gamma: {
			default: .5,
			description: "Contour distribution. <1 clusters in bright, >1 clusters in dark",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 2,
				step: .01,
				label: "Gamma",
				group: "Effect"
			}
		},
		invert: {
			default: false,
			description: "Invert the source values",
			transform: transformBoolean,
			ui: {
				type: "checkbox",
				label: "Invert",
				group: "Effect"
			}
		},
		source: {
			default: "luminance",
			description: "Use luminance or alpha channel for contours",
			transform: transformSource,
			compileTime: true,
			ui: {
				type: "select",
				options: [{
					label: "Luminance",
					value: "luminance"
				}, {
					label: "Alpha",
					value: "alpha"
				}],
				label: "Source",
				group: "Effect"
			}
		},
		colorMode: {
			default: "source",
			description: "Use source image colors or custom colors",
			transform: transformColorMode,
			compileTime: true,
			ui: {
				type: "select",
				options: [{
					label: "Source",
					value: "source"
				}, {
					label: "Custom",
					value: "custom"
				}],
				label: "Color Mode",
				group: "Colors"
			}
		},
		lineColor: {
			default: "#000000",
			description: "Color of the contour lines (custom mode)",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Line Color",
				group: "Colors",
				condition: { colorMode: "custom" }
			}
		},
		backgroundColor: {
			default: "transparent",
			description: "Background color (custom mode)",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Background Color",
				group: "Colors",
				condition: { colorMode: "custom" }
			}
		}
	},
	fragment: ({ childNode, uniforms, propValues }) => {
		if (!childNode) return ZERO;
		const lineMask = call(contourLineMask, "contourLineMask", [
			sourceModeOf(propValues.source) === 1 ? childNode.member("a") : call(contourLuma, "contourLuma", [childNode]),
			uniforms.levels,
			uniforms.lineWidth,
			uniforms.softness,
			uniforms.gamma,
			uniforms.invert
		]);
		let bg;
		let line;
		if (colorModeOf(propValues.colorMode) === 1) {
			bg = uniforms.backgroundColor;
			line = uniforms.lineColor;
		} else {
			bg = ZERO;
			line = vec4(childNode.member("rgb"), floatE(1));
		}
		return call(contourCompose, "contourCompose", [
			bg,
			line,
			lineMask
		]);
	}
};
var ContourLines_default = componentDefinition;
export { contourLuma as a, contourLineMask as i, componentDefinition as n, contourCompose as r, ContourLines_default as t };
