// shaders/dist/core/CRTScreen-CSBjpbn4.js
import { At as call, Ot as ZERO, Pt as floatE, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
const crtSampleUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, colorShift, dir) => {
	"use gpu";
	const scaledColorShift = __tsover_mul(colorShift, .002);
	return d.vec2f(__tsover_add(uv.x, __tsover_mul(scaledColorShift, dir)), uv.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "colorShift"
			},
			{
				"type": "i",
				"name": "dir"
			}
		],
		"body": [0, [[
			13,
			"scaledColorShift",
			[
				1,
				"colorShift",
				"*",
				[5, "0.002"]
			]
		], [10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					7,
					"uv",
					"x"
				],
				"+",
				[
					1,
					"scaledColorShift",
					"*",
					"dir"
				]
			], [
				7,
				"uv",
				"y"
			]]
		]]]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const crtCompose = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((redSample, greenSample, blueSample, uv, aspect, pixelSize, scanlineIntensity, scanlineFrequency, brightness, contrast, vignetteIntensity, vignetteRadius) => {
	"use gpu";
	const adjustedColor = d.vec3f(redSample.x, greenSample.y, blueSample.z).sub(d.vec3f(.5)).mul(contrast).add(d.vec3f(.5)).mul(brightness);
	const scanlinePosition = __tsover_mul(uv.y, scanlineFrequency);
	const scanlineWave = std.sin(__tsover_mul(scanlinePosition, __tsover_mul(3.14159, 2)));
	const scanlineEffect = __tsover_sub(1, __tsover_mul(scanlineIntensity, __tsover_add(__tsover_mul(scanlineWave, .5), .5)));
	const scanlinedColor = adjustedColor.mul(scanlineEffect);
	const phosphorScale = __tsover_mul(pixelSize, .5);
	const pixelPhase = std.fract(uv.mul(phosphorScale));
	const redMask = __tsover_add(__tsover_mul(std.sin(__tsover_mul(pixelPhase.x, 6.28318)), .1), .95);
	const greenMask = __tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_mul(pixelPhase.x, 6.28318), 2.09)), .1), .95);
	const blueMask = __tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_mul(pixelPhase.x, 6.28318), 4.18)), .1), .95);
	const phosphorColor = d.vec3f(__tsover_mul(scanlinedColor.x, redMask), __tsover_mul(scanlinedColor.y, greenMask), __tsover_mul(scanlinedColor.z, blueMask));
	const centeredUV = uv.sub(d.vec2f(.5, .5));
	const aspectCorrectedUV = d.vec2f(__tsover_mul(centeredUV.x, aspect), centeredUV.y);
	const vignetteDistance = std.length(aspectCorrectedUV);
	const innerEdge = __tsover_sub(1, __tsover_mul(vignetteRadius, .9));
	const outerEdge = __tsover_add(innerEdge, .5);
	const vignetteValue = __tsover_sub(1, std.smoothstep(innerEdge, outerEdge, vignetteDistance));
	const vignette = std.mix(1, vignetteValue, vignetteIntensity);
	const finalColor = phosphorColor.mul(vignette);
	return d.vec4f(finalColor.x, finalColor.y, finalColor.z, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "redSample"
			},
			{
				"type": "i",
				"name": "greenSample"
			},
			{
				"type": "i",
				"name": "blueSample"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "pixelSize"
			},
			{
				"type": "i",
				"name": "scanlineIntensity"
			},
			{
				"type": "i",
				"name": "scanlineFrequency"
			},
			{
				"type": "i",
				"name": "brightness"
			},
			{
				"type": "i",
				"name": "contrast"
			},
			{
				"type": "i",
				"name": "vignetteIntensity"
			},
			{
				"type": "i",
				"name": "vignetteRadius"
			}
		],
		"body": [0, [
			[
				13,
				"baseColor",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"redSample",
							"x"
						],
						[
							7,
							"greenSample",
							"y"
						],
						[
							7,
							"blueSample",
							"z"
						]
					]
				]
			],
			[
				13,
				"adjustedColor",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										[
											6,
											[
												7,
												"baseColor",
												"sub"
											],
											[[
												6,
												[
													7,
													"d",
													"vec3f"
												],
												[[5, "0.5"]]
											]]
										],
										"mul"
									],
									["contrast"]
								],
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[[5, "0.5"]]
							]]
						],
						"mul"
					],
					["brightness"]
				]
			],
			[
				13,
				"scanlinePosition",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"*",
					"scanlineFrequency"
				]
			],
			[
				13,
				"scanlineWave",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						1,
						"scanlinePosition",
						"*",
						[
							1,
							[5, "3.14159"],
							"*",
							[5, "2"]
						]
					]]
				]
			],
			[
				13,
				"scanlineEffect",
				[
					1,
					[5, "1"],
					"-",
					[
						1,
						"scanlineIntensity",
						"*",
						[
							1,
							[
								1,
								"scanlineWave",
								"*",
								[5, "0.5"]
							],
							"+",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"scanlinedColor",
				[
					6,
					[
						7,
						"adjustedColor",
						"mul"
					],
					["scanlineEffect"]
				]
			],
			[
				13,
				"phosphorScale",
				[
					1,
					"pixelSize",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"pixelPhase",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						6,
						[
							7,
							"uv",
							"mul"
						],
						["phosphorScale"]
					]]
				]
			],
			[
				13,
				"redMask",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									7,
									"pixelPhase",
									"x"
								],
								"*",
								[5, "6.28318"]
							]]
						],
						"*",
						[5, "0.1"]
					],
					"+",
					[5, "0.95"]
				]
			],
			[
				13,
				"greenMask",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									[
										7,
										"pixelPhase",
										"x"
									],
									"*",
									[5, "6.28318"]
								],
								"+",
								[5, "2.09"]
							]]
						],
						"*",
						[5, "0.1"]
					],
					"+",
					[5, "0.95"]
				]
			],
			[
				13,
				"blueMask",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									[
										7,
										"pixelPhase",
										"x"
									],
									"*",
									[5, "6.28318"]
								],
								"+",
								[5, "4.18"]
							]]
						],
						"*",
						[5, "0.1"]
					],
					"+",
					[5, "0.95"]
				]
			],
			[
				13,
				"phosphorColor",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								7,
								"scanlinedColor",
								"x"
							],
							"*",
							"redMask"
						],
						[
							1,
							[
								7,
								"scanlinedColor",
								"y"
							],
							"*",
							"greenMask"
						],
						[
							1,
							[
								7,
								"scanlinedColor",
								"z"
							],
							"*",
							"blueMask"
						]
					]
				]
			],
			[
				13,
				"centeredUV",
				[
					6,
					[
						7,
						"uv",
						"sub"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "0.5"], [5, "0.5"]]
					]]
				]
			],
			[
				13,
				"aspectCorrectedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"centeredUV",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"centeredUV",
						"y"
					]]
				]
			],
			[
				13,
				"vignetteDistance",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["aspectCorrectedUV"]
				]
			],
			[
				13,
				"innerEdge",
				[
					1,
					[5, "1"],
					"-",
					[
						1,
						"vignetteRadius",
						"*",
						[5, "0.9"]
					]
				]
			],
			[
				13,
				"outerEdge",
				[
					1,
					"innerEdge",
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"vignetteValue",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							"innerEdge",
							"outerEdge",
							"vignetteDistance"
						]
					]
				]
			],
			[
				13,
				"vignette",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "1"],
						"vignetteValue",
						"vignetteIntensity"
					]
				]
			],
			[
				13,
				"finalColor",
				[
					6,
					[
						7,
						"phosphorColor",
						"mul"
					],
					["vignette"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"finalColor",
						"x"
					],
					[
						7,
						"finalColor",
						"y"
					],
					[
						7,
						"finalColor",
						"z"
					],
					[5, "1"]
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "CRTScreen",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Retro CRT monitor simulation with scanlines",
	requiresRTT: true,
	requiresChild: true,
	props: {
		pixelSize: {
			default: 128,
			description: "Size of individual TV pixels (lower = more pixels)",
			ui: {
				type: ["range", "map"],
				min: 8,
				max: 128,
				step: 1,
				label: "Pixel Size",
				group: "Effect"
			}
		},
		colorShift: {
			default: 1,
			description: "Chromatic aberration amount",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 10,
				step: .1,
				label: "Color Shift",
				group: "Effect"
			}
		},
		scanlineIntensity: {
			default: .3,
			description: "Strength of horizontal scanlines",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Scanline Intensity",
				group: "Effect"
			}
		},
		scanlineFrequency: {
			default: 200,
			description: "Number of scanlines across screen",
			ui: {
				type: ["range", "map"],
				min: 100,
				max: 800,
				step: 10,
				label: "Scanline Frequency",
				group: "Effect"
			}
		},
		brightness: {
			default: 1,
			description: "Screen brightness boost",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 2,
				step: .1,
				label: "Brightness",
				group: "Adjustments"
			}
		},
		contrast: {
			default: 1,
			description: "Screen contrast enhancement",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 2,
				step: .1,
				label: "Contrast",
				group: "Adjustments"
			}
		},
		vignetteIntensity: {
			default: 1,
			description: "Strength of corner darkening effect (0 = off)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Vignette Intensity",
				group: "Effect"
			}
		},
		vignetteRadius: {
			default: .5,
			description: "How far the vignette extends inward (0 = edges only, 1 = reaches center)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Vignette Radius",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const redUV = call(crtSampleUV, "crtSampleUV", [
			ctx.uv,
			uniforms.colorShift,
			floatE(1)
		]);
		const blueUV = call(crtSampleUV, "crtSampleUV", [
			ctx.uv,
			uniforms.colorShift,
			floatE(-1)
		]);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [call(crtCompose, "crtCompose", [
			texture.sample(redUV),
			texture.sample(ctx.uv),
			texture.sample(blueUV),
			ctx.uv,
			ctx.aspect,
			uniforms.pixelSize,
			uniforms.scanlineIntensity,
			uniforms.scanlineFrequency,
			uniforms.brightness,
			uniforms.contrast,
			uniforms.vignetteIntensity,
			uniforms.vignetteRadius
		])]);
	}
};
var CRTScreen_default = componentDefinition;
export { crtSampleUV as i, componentDefinition as n, crtCompose as r, CRTScreen_default as t };
