// shaders/dist/core/Crystal-CC7d_6kZ.js
import { At as call, It as vec4, Ot as ZERO, Pt as floatE, R as std, T as sdf_exports, Ut as blend_exports, d as sdf3d_exports, n as tgpu, nt as mxWorleyNoiseFloat2Pub, on as d, rt as noise_exports } from "./coords-CWOT0Wqh.js";
import { n as glass_exports } from "./glass-hUL2pCYG.js";
import { a as transformColor, c as transformPosition, i as transformBoolean } from "./transformations-fqzznt4g.js";
import { t as shapeEffectBoundingBoxDeclaration } from "./shapeEffectBounds-BlxxsYMN.js";
var { sdfSpaceUV, offsetUV } = glass_exports;
var { unpremultiplyAlpha } = blend_exports;
var { createSvgSdfSampler, createAnalyticSdfSampler, driveAnalyticSubProps, ANALYTIC_SDF_EXTRA_FIELDS } = sdf_exports;
var { createVolumetricFieldComputeNode, buildVolumetricFieldSampler, VOLUMETRIC_FIELD_EXTRA_FIELDS, is3dShapeType, isVolumetricShapeType } = sdf3d_exports;
var DEG_TO_RAD = .017453292519943295;
var PI = Math.PI;
var TWO_PI = 6.283185307179586;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "polygonSDF",
	radius: .35,
	sides: 10
});
function resolveShapeType(shapeType, shape) {
	if (shapeType) return shapeType;
	if (shape && typeof shape === "object") return shape.type || "circleSDF";
	if (typeof shape === "string") try {
		return JSON.parse(shape).type || "circleSDF";
	} catch {
		return "circleSDF";
	}
	return "circleSDF";
}
const crystalKaleidoFold = tgpu.fn([
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uvX, uvY, sectorAngle) => {
	"use gpu";
	const dx = __tsover_sub(uvX, .5);
	const dy = __tsover_sub(uvY, .5);
	const r = std.sqrt(__tsover_add(__tsover_mul(dx, dx), __tsover_mul(dy, dy)));
	const theta = __tsover_add(std.atan2(dy, dx), PI);
	const inSector = __tsover_mul(std.fract(__tsover_div(theta, sectorAngle)), sectorAngle);
	const folded = std.min(inSector, __tsover_sub(sectorAngle, inSector));
	return d.vec2f(__tsover_mul(r, std.cos(folded)), __tsover_mul(r, std.sin(folded)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uvX"
			},
			{
				"type": "i",
				"name": "uvY"
			},
			{
				"type": "i",
				"name": "sectorAngle"
			}
		],
		"body": [0, [
			[
				13,
				"dx",
				[
					1,
					"uvX",
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"dy",
				[
					1,
					"uvY",
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"r",
				[
					6,
					[
						7,
						"std",
						"sqrt"
					],
					[[
						1,
						[
							1,
							"dx",
							"*",
							"dx"
						],
						"+",
						[
							1,
							"dy",
							"*",
							"dy"
						]
					]]
				]
			],
			[
				13,
				"theta",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"atan2"
						],
						["dy", "dx"]
					],
					"+",
					"PI"
				]
			],
			[
				13,
				"inSector",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						[[
							1,
							"theta",
							"/",
							"sectorAngle"
						]]
					],
					"*",
					"sectorAngle"
				]
			],
			[
				13,
				"folded",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					["inSector", [
						1,
						"sectorAngle",
						"-",
						"inSector"
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					"r",
					"*",
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						["folded"]
					]
				], [
					1,
					"r",
					"*",
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						["folded"]
					]
				]]
			]]
		]],
		"externalNames": [
			"std",
			"PI",
			"d"
		]
	},
	externals: () => ({
		std,
		PI,
		d
	})
}) && $.f)({}));
const CrystalTrace = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
	rUV: d.vec2f,
	lensUV: d.vec2f,
	bUV: d.vec2f,
	innerUV: d.vec2f,
	outerUV: d.vec2f,
	factors: d.vec4f,
	transition: d.f32,
	fresnelContrib: d.vec3f
}), "CrystalTrace");
const crystalTrace = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f
], CrystalTrace)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((surf0, surfX, surfY, sdfUV, uv, viewportSize, center, fresnelColor, p0, p1, p2, p3) => {
	"use gpu";
	const scale = p0.x;
	const edgeSoftness = p0.y;
	const refraction = p0.z;
	const dispersion = p0.w;
	const facets = p1.x;
	const innerZoom = p1.y;
	const lightAngle = p1.z;
	const highlights = p1.w;
	const shadows = p2.x;
	const brightness = p2.y;
	const fresnel = p2.z;
	const fresnelSoftness = p2.w;
	const volumetric = p3.x;
	const aspect = __tsover_div(viewportSize.x, viewportSize.y);
	const sharp = std.max(__tsover_mul(edgeSoftness, .5), .001);
	const pxH = __tsover_div(1, viewportSize.y);
	const centerPosX = center.x;
	const centerPosY = __tsover_sub(1, center.y);
	const sdfRaw = surf0.x;
	const sdf = __tsover_div(sdfRaw, scale);
	const EPS = .01;
	const gradX = __tsover_div(__tsover_sub(surfX.x, sdfRaw), EPS);
	const gradY = __tsover_div(__tsover_sub(surfY.x, sdfRaw), EPS);
	const rb1 = std.clamp(__tsover_mul(__tsover_div(__tsover_mul(sdf, -1), sharp), 32), 0, 1);
	const lightRad = __tsover_mul(lightAngle, DEG_TO_RAD);
	const lx = __tsover_add(std.cos(lightRad), __tsover_mul(__tsover_sub(centerPosX, .5), 3));
	const ly = __tsover_add(std.sin(lightRad), __tsover_mul(__tsover_sub(centerPosY, .5), 3));
	const L = std.normalize(d.vec3f(lx, ly, -.75));
	const Vv = d.vec3f(0, 0, -1);
	const H = std.normalize(L.add(Vv));
	const L2 = std.normalize(d.vec3f(__tsover_mul(ly, -1), lx, -.4));
	const H2 = std.normalize(L2.add(Vv));
	const depthNorm = std.clamp(__tsover_div(__tsover_mul(sdf, -1), .15), 0, 1);
	const outerWarp = __tsover_mul(__tsover_sub(1, std.smoothstep(0, .03, sdf)), .4);
	const refrScale = __tsover_mul(refraction, .15);
	const haloOffX = __tsover_div(__tsover_mul(__tsover_mul(__tsover_mul(gradX, -1), refrScale), outerWarp), aspect);
	const haloOffY = __tsover_mul(__tsover_mul(__tsover_mul(gradY, -1), refrScale), outerWarp);
	const pdx = std.select(__tsover_sub(sdfUV.x, .5), surf0.y, volumetric > .5);
	const pdy = std.select(__tsover_sub(sdfUV.y, .5), surf0.z, volumetric > .5);
	const sAngle = __tsover_div(TWO_PI, facets);
	const fUv = crystalKaleidoFold(__tsover_add(pdx, .5), __tsover_add(pdy, .5), sAngle);
	const crystalRNorm = __tsover_div(std.sqrt(__tsover_add(__tsover_mul(pdx, pdx), __tsover_mul(pdy, pdy))), .35);
	const tableMask = __tsover_sub(1, std.smoothstep(.22, .32, crystalRNorm));
	const radialDensity = __tsover_add(1, __tsover_mul(std.clamp(crystalRNorm, 0, 1), .6));
	const tableSuppress = __tsover_sub(1, __tsover_mul(tableMask, .9));
	const ddx = std.clamp(__tsover_div(__tsover_sub(surfX.w, surf0.w), EPS), -4, 4);
	const ddy = std.clamp(__tsover_div(__tsover_sub(surfY.w, surf0.w), EPS), -4, 4);
	const nSurfVol = std.normalize(d.vec3f(ddx, ddy, -1));
	const bevelSlope = __tsover_mul(__tsover_mul(__tsover_sub(1, std.smoothstep(0, 1, depthNorm)), 1.7), __tsover_sub(1, tableMask));
	const nSurfFlat = std.normalize(d.vec3f(__tsover_mul(gradX, bevelSlope), __tsover_mul(gradY, bevelSlope), -1));
	const nSurf = std.select(nSurfFlat, nSurfVol, volumetric > .5);
	const cosView = std.clamp(__tsover_mul(nSurf.z, -1), 0, 1);
	const grazing = __tsover_sub(1, cosView);
	const vScale1 = __tsover_mul(__tsover_mul(facets, 1.2), radialDensity);
	const fUvS1 = fUv.mul(vScale1);
	const feps = .02;
	const f1center = mxWorleyNoiseFloat2Pub(fUvS1, 1);
	const gx1 = __tsover_div(__tsover_sub(mxWorleyNoiseFloat2Pub(fUvS1.add(d.vec2f(feps, 0)), 1), f1center), feps);
	const gy1 = __tsover_div(__tsover_sub(mxWorleyNoiseFloat2Pub(fUvS1.add(d.vec2f(0, feps)), 1), f1center), feps);
	const gLen1 = std.max(std.sqrt(__tsover_add(__tsover_mul(gx1, gx1), __tsover_mul(gy1, gy1))), .001);
	const nx1 = __tsover_div(gx1, gLen1);
	const ny1 = __tsover_div(gy1, gLen1);
	const cellId1 = std.floor(fUvS1);
	const cellRand = std.fract(__tsover_mul(std.sin(__tsover_add(__tsover_mul(cellId1.x, 12.9898), __tsover_mul(cellId1.y, 78.233))), 43758.5453));
	const nF1 = std.normalize(nSurf.add(d.vec3f(__tsover_mul(nx1, .55), __tsover_mul(ny1, .55), 0)));
	const ndlS = std.dot(nSurf, L);
	const ndl1 = __tsover_mul(__tsover_sub(std.dot(nF1, L), ndlS), 2.2);
	const fUvS2 = fUv.mul(__tsover_mul(vScale1, 1.7)).add(d.vec2f(3.7, 1.2));
	const c2 = mxWorleyNoiseFloat2Pub(fUvS2, 1);
	const gx2 = __tsover_div(__tsover_sub(mxWorleyNoiseFloat2Pub(fUvS2.add(d.vec2f(feps, 0)), 1), c2), feps);
	const gy2 = __tsover_div(__tsover_sub(mxWorleyNoiseFloat2Pub(fUvS2.add(d.vec2f(0, feps)), 1), c2), feps);
	const gLen2 = std.max(std.sqrt(__tsover_add(__tsover_mul(gx2, gx2), __tsover_mul(gy2, gy2))), .001);
	const nx2 = __tsover_div(gx2, gLen2);
	const ny2 = __tsover_div(gy2, gLen2);
	const nF2 = std.normalize(nSurf.add(d.vec3f(__tsover_mul(nx2, .4), __tsover_mul(ny2, .4), 0)));
	const ndl2 = __tsover_mul(__tsover_sub(std.dot(nF2, L2), std.dot(nSurf, L2)), 2.2);
	const I = d.vec3f(0, 0, 1);
	const facetRefrAmp = std.clamp(__tsover_mul(refraction, .8), 0, 1.6);
	const nRefr = std.normalize(nSurf.add(d.vec3f(__tsover_mul(__tsover_mul(__tsover_mul(nx1, .75), facetRefrAmp), tableSuppress), __tsover_mul(__tsover_mul(__tsover_mul(ny1, .75), facetRefrAmp), tableSuppress), 0)).add(d.vec3f(__tsover_mul(__tsover_mul(nx2, .55), facetRefrAmp), __tsover_mul(__tsover_mul(ny2, .55), facetRefrAmp), 0)));
	const eta = __tsover_div(1, __tsover_add(1, __tsover_mul(refraction, .55)));
	const refr = std.refract(I, nRefr, eta);
	const thickVol = std.clamp(__tsover_mul(sdfRaw, -1), 0, .6);
	const thickFlat = __tsover_mul(std.clamp(__tsover_mul(sdfRaw, -1), 0, .12), 1.6);
	const thickEff = std.select(thickFlat, thickVol, volumetric > .5);
	const thickFacet = __tsover_mul(thickEff, __tsover_add(.55, __tsover_mul(cellRand, .9)));
	const offMag = __tsover_mul(__tsover_mul(thickFacet, scale), rb1);
	const totalOffX = __tsover_add(__tsover_div(__tsover_mul(__tsover_div(refr.x, std.max(refr.z, .25)), offMag), aspect), haloOffX);
	const totalOffY = __tsover_add(__tsover_mul(__tsover_div(refr.y, std.max(refr.z, .25)), offMag), haloOffY);
	const nBack = std.normalize(d.vec3f(__tsover_mul(nx2, .7), __tsover_mul(ny2, .7), -1));
	const bounce = std.reflect(refr, nBack);
	const bounceMag = __tsover_mul(__tsover_mul(__tsover_mul(thickEff, scale), rb1), .8);
	const bounceOffX = __tsover_div(__tsover_mul(__tsover_div(bounce.x, std.max(std.abs(bounce.z), .3)), bounceMag), aspect);
	const bounceOffY = __tsover_mul(__tsover_div(bounce.y, std.max(std.abs(bounce.z), .3)), bounceMag);
	const innerW = __tsover_mul(__tsover_mul(__tsover_add(__tsover_mul(grazing, .5), .18), __tsover_add(__tsover_mul(cellRand, .7), .3)), rb1);
	const pbx = __tsover_add(pdx, __tsover_mul(__tsover_mul(__tsover_div(refr.x, std.max(refr.z, .25)), thickEff), 1.2));
	const pby = __tsover_add(pdy, __tsover_mul(__tsover_mul(__tsover_div(refr.y, std.max(refr.z, .25)), thickEff), 1.2));
	const fUvS3 = crystalKaleidoFold(__tsover_add(pbx, .5), __tsover_add(pby, .5), sAngle).mul(__tsover_mul(vScale1, 1.4)).add(d.vec2f(9.1, 4.7));
	const c3 = mxWorleyNoiseFloat2Pub(fUvS3, 1);
	const gx3 = __tsover_div(__tsover_sub(mxWorleyNoiseFloat2Pub(fUvS3.add(d.vec2f(feps, 0)), 1), c3), feps);
	const gy3 = __tsover_div(__tsover_sub(mxWorleyNoiseFloat2Pub(fUvS3.add(d.vec2f(0, feps)), 1), c3), feps);
	const gLen3 = std.max(std.sqrt(__tsover_add(__tsover_mul(gx3, gx3), __tsover_mul(gy3, gy3))), .001);
	const nx3 = __tsover_div(gx3, gLen3);
	const ny3 = __tsover_div(gy3, gLen3);
	const cellId3 = std.floor(fUvS3);
	const cellRand3 = std.fract(__tsover_mul(std.sin(__tsover_add(__tsover_mul(cellId3.x, 31.727), __tsover_mul(cellId3.y, 57.131))), 43758.5453));
	const w3 = __tsover_mul(__tsover_mul(std.clamp(__tsover_mul(thickEff, 2.5), 0, 1), std.clamp(__tsover_mul(refraction, 1.2), 0, 1)), rb1);
	const wedgeLit = __tsover_add(__tsover_mul(__tsover_add(__tsover_mul(nx3, lx), __tsover_mul(ny3, ly)), .5), .5);
	const mosaicLevel = __tsover_mul(__tsover_add(__tsover_mul(cellRand3, .9), .55), __tsover_add(__tsover_mul(wedgeLit, .7), .65));
	const edge3 = __tsover_add(__tsover_mul(std.smoothstep(.15, .7, gLen3), .45), .55);
	const mosaicMul = __tsover_mul(mosaicLevel, edge3);
	const mosaicFire = __tsover_mul(__tsover_mul(std.smoothstep(.75, 1, __tsover_mul(wedgeLit, __tsover_add(cellRand3, .4))), w3), .8);
	const zoomedX = __tsover_add(centerPosX, __tsover_div(__tsover_sub(uv.x, centerPosX), innerZoom));
	const zoomedY = __tsover_add(centerPosY, __tsover_div(__tsover_sub(uv.y, centerPosY), innerZoom));
	const lensX = __tsover_add(zoomedX, totalOffX);
	const lensY = __tsover_add(zoomedY, totalOffY);
	const lit1 = std.clamp(ndl1, 0, 1);
	const shadow1 = std.clamp(__tsover_mul(ndl1, -1), 0, 1);
	const bright1 = std.max(__tsover_sub(__tsover_add(1, __tsover_mul(lit1, highlights)), __tsover_mul(shadow1, shadows)), .02);
	const lit2 = std.clamp(ndl2, 0, 1);
	const shadow2 = std.clamp(__tsover_mul(ndl2, -1), 0, 1);
	const bright2 = __tsover_sub(__tsover_add(1, __tsover_mul(lit2, __tsover_mul(highlights, .5))), __tsover_mul(shadow2, __tsover_mul(shadows, .5)));
	const surfBright = __tsover_sub(__tsover_add(1, __tsover_mul(__tsover_mul(std.clamp(ndlS, 0, 1), highlights), .4)), __tsover_mul(__tsover_mul(std.clamp(__tsover_mul(ndlS, -1), 0, 1), shadows), .6));
	const combinedBright = __tsover_mul(__tsover_mul(bright1, bright2), surfBright);
	const chrScale = __tsover_mul(dispersion, .12);
	const chrOffX = __tsover_mul(totalOffX, chrScale);
	const chrOffY = __tsover_mul(totalOffY, chrScale);
	const baseEdge = std.smoothstep(.2, .8, gLen1);
	const edgeDarken = __tsover_add(__tsover_mul(std.mix(baseEdge, 1, tableMask), .4), .6);
	const tableBoost = __tsover_mul(tableMask, .25);
	const facetMul = __tsover_mul(__tsover_mul(__tsover_mul(__tsover_add(combinedBright, tableBoost), edgeDarken), std.mix(d.f32(1), mosaicMul, w3)), brightness);
	const ndh1 = std.clamp(std.dot(nF1, H), 0, 1);
	const ndh2 = std.clamp(std.dot(nF2, H2), 0, 1);
	const spec1 = __tsover_mul(std.pow(ndh1, d.f32(110)), 2.2);
	const spec2 = __tsover_mul(std.pow(ndh2, d.f32(56)), .9);
	const glint = __tsover_mul(std.smoothstep(.965, .995, ndh1), 3);
	const flashContrib = __tsover_mul(__tsover_mul(__tsover_add(__tsover_add(__tsover_add(spec1, spec2), glint), mosaicFire), highlights), brightness);
	const fresnelExp = std.mix(6, 1.5, std.clamp(__tsover_mul(fresnelSoftness, .5), 0, 1));
	const fresnelRim = __tsover_mul(__tsover_mul(__tsover_mul(std.pow(grazing, fresnelExp), fresnel), 2), rb1);
	const fresnelContrib = fresnelColor.mul(fresnelRim);
	const normalDotLight = __tsover_add(__tsover_mul(gradX, lx), __tsover_mul(gradY, ly));
	const lightFacing = __tsover_add(__tsover_mul(normalDotLight, .5), .5);
	const sdfOuter = __tsover_sub(sdf, pxH);
	const rb2base = __tsover_sub(std.clamp(__tsover_mul(__tsover_div(__tsover_mul(sdfOuter, -1), sharp), 16), 0, 1), std.clamp(__tsover_mul(__tsover_div(__tsover_mul(sdf, -1), sharp), 16), 0, 1));
	const rb2 = __tsover_mul(__tsover_mul(rb2base, lightFacing), .35);
	const transition = std.smoothstep(0, 1, rb1);
	const outerUV = d.vec2f(__tsover_add(uv.x, haloOffX), __tsover_add(uv.y, haloOffY));
	return CrystalTrace({
		rUV: d.vec2f(__tsover_add(lensX, chrOffX), __tsover_add(lensY, chrOffY)),
		lensUV: d.vec2f(lensX, lensY),
		bUV: d.vec2f(__tsover_sub(lensX, chrOffX), __tsover_sub(lensY, chrOffY)),
		innerUV: d.vec2f(__tsover_add(lensX, bounceOffX), __tsover_add(lensY, bounceOffY)),
		outerUV,
		factors: d.vec4f(facetMul, flashContrib, innerW, rb2),
		transition,
		fresnelContrib
	});
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "surf0"
			},
			{
				"type": "i",
				"name": "surfX"
			},
			{
				"type": "i",
				"name": "surfY"
			},
			{
				"type": "i",
				"name": "sdfUV"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewportSize"
			},
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "fresnelColor"
			},
			{
				"type": "i",
				"name": "p0"
			},
			{
				"type": "i",
				"name": "p1"
			},
			{
				"type": "i",
				"name": "p2"
			},
			{
				"type": "i",
				"name": "p3"
			}
		],
		"body": [0, [
			[
				13,
				"scale",
				[
					7,
					"p0",
					"x"
				]
			],
			[
				13,
				"edgeSoftness",
				[
					7,
					"p0",
					"y"
				]
			],
			[
				13,
				"refraction",
				[
					7,
					"p0",
					"z"
				]
			],
			[
				13,
				"dispersion",
				[
					7,
					"p0",
					"w"
				]
			],
			[
				13,
				"facets",
				[
					7,
					"p1",
					"x"
				]
			],
			[
				13,
				"innerZoom",
				[
					7,
					"p1",
					"y"
				]
			],
			[
				13,
				"lightAngle",
				[
					7,
					"p1",
					"z"
				]
			],
			[
				13,
				"highlights",
				[
					7,
					"p1",
					"w"
				]
			],
			[
				13,
				"shadows",
				[
					7,
					"p2",
					"x"
				]
			],
			[
				13,
				"brightness",
				[
					7,
					"p2",
					"y"
				]
			],
			[
				13,
				"fresnel",
				[
					7,
					"p2",
					"z"
				]
			],
			[
				13,
				"fresnelSoftness",
				[
					7,
					"p2",
					"w"
				]
			],
			[
				13,
				"volumetric",
				[
					7,
					"p3",
					"x"
				]
			],
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewportSize",
						"x"
					],
					"/",
					[
						7,
						"viewportSize",
						"y"
					]
				]
			],
			[
				13,
				"sharp",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"edgeSoftness",
						"*",
						[5, "0.5"]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"pxH",
				[
					1,
					[5, "1"],
					"/",
					[
						7,
						"viewportSize",
						"y"
					]
				]
			],
			[
				13,
				"centerPosX",
				[
					7,
					"center",
					"x"
				]
			],
			[
				13,
				"centerPosY",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"center",
						"y"
					]
				]
			],
			[
				13,
				"sdfRaw",
				[
					7,
					"surf0",
					"x"
				]
			],
			[
				13,
				"sdf",
				[
					1,
					"sdfRaw",
					"/",
					"scale"
				]
			],
			[
				13,
				"EPS",
				[5, "0.01"]
			],
			[
				13,
				"gradX",
				[
					1,
					[
						1,
						[
							7,
							"surfX",
							"x"
						],
						"-",
						"sdfRaw"
					],
					"/",
					"EPS"
				]
			],
			[
				13,
				"gradY",
				[
					1,
					[
						1,
						[
							7,
							"surfY",
							"x"
						],
						"-",
						"sdfRaw"
					],
					"/",
					"EPS"
				]
			],
			[
				13,
				"rb1",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									"sdf",
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								],
								"/",
								"sharp"
							],
							"*",
							[5, "32"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"lightRad",
				[
					1,
					"lightAngle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"lx",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						["lightRad"]
					],
					"+",
					[
						1,
						[
							1,
							"centerPosX",
							"-",
							[5, "0.5"]
						],
						"*",
						[5, "3"]
					]
				]
			],
			[
				13,
				"ly",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						["lightRad"]
					],
					"+",
					[
						1,
						[
							1,
							"centerPosY",
							"-",
							[5, "0.5"]
						],
						"*",
						[5, "3"]
					]
				]
			],
			[
				13,
				"L",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"lx",
							"ly",
							[
								4,
								"-",
								[5, "0.75"]
							]
						]
					]]
				]
			],
			[
				13,
				"Vv",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0"],
						[5, "0"],
						[
							4,
							"-",
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"H",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"L",
							"add"
						],
						["Vv"]
					]]
				]
			],
			[
				13,
				"L2",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								"ly",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"lx",
							[
								4,
								"-",
								[5, "0.4"]
							]
						]
					]]
				]
			],
			[
				13,
				"H2",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"L2",
							"add"
						],
						["Vv"]
					]]
				]
			],
			[
				13,
				"depthNorm",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"sdf",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"/",
							[5, "0.15"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"outerWarp",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[5, "0"],
								[5, "0.03"],
								"sdf"
							]
						]
					],
					"*",
					[5, "0.4"]
				]
			],
			[
				13,
				"refrScale",
				[
					1,
					"refraction",
					"*",
					[5, "0.15"]
				]
			],
			[
				13,
				"haloOffX",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								"gradX",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"*",
							"refrScale"
						],
						"*",
						"outerWarp"
					],
					"/",
					"aspect"
				]
			],
			[
				13,
				"haloOffY",
				[
					1,
					[
						1,
						[
							1,
							"gradY",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"*",
						"refrScale"
					],
					"*",
					"outerWarp"
				]
			],
			[
				13,
				"pdx",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							1,
							[
								7,
								"sdfUV",
								"x"
							],
							"-",
							[5, "0.5"]
						],
						[
							7,
							"surf0",
							"y"
						],
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"pdy",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							1,
							[
								7,
								"sdfUV",
								"y"
							],
							"-",
							[5, "0.5"]
						],
						[
							7,
							"surf0",
							"z"
						],
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"sAngle",
				[
					1,
					"TWO_PI",
					"/",
					"facets"
				]
			],
			[
				13,
				"fUv",
				[
					6,
					"crystalKaleidoFold",
					[
						[
							1,
							"pdx",
							"+",
							[5, "0.5"]
						],
						[
							1,
							"pdy",
							"+",
							[5, "0.5"]
						],
						"sAngle"
					]
				]
			],
			[
				13,
				"crystalRNorm",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"pdx",
								"*",
								"pdx"
							],
							"+",
							[
								1,
								"pdy",
								"*",
								"pdy"
							]
						]]
					],
					"/",
					[5, "0.35"]
				]
			],
			[
				13,
				"tableMask",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0.22"],
							[5, "0.32"],
							"crystalRNorm"
						]
					]
				]
			],
			[
				13,
				"radialDensity",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								"crystalRNorm",
								[5, "0"],
								[5, "1"]
							]
						],
						"*",
						[5, "0.6"]
					]
				]
			],
			[
				13,
				"tableSuppress",
				[
					1,
					[5, "1"],
					"-",
					[
						1,
						"tableMask",
						"*",
						[5, "0.9"]
					]
				]
			],
			[
				13,
				"ddx",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"surfX",
									"w"
								],
								"-",
								[
									7,
									"surf0",
									"w"
								]
							],
							"/",
							"EPS"
						],
						[
							4,
							"-",
							[5, "4"]
						],
						[5, "4"]
					]
				]
			],
			[
				13,
				"ddy",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"surfY",
									"w"
								],
								"-",
								[
									7,
									"surf0",
									"w"
								]
							],
							"/",
							"EPS"
						],
						[
							4,
							"-",
							[5, "4"]
						],
						[5, "4"]
					]
				]
			],
			[
				13,
				"nSurfVol",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"ddx",
							"ddy",
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]]
				]
			],
			[
				13,
				"bevelSlope",
				[
					1,
					[
						1,
						[
							1,
							[5, "1"],
							"-",
							[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									[5, "0"],
									[5, "1"],
									"depthNorm"
								]
							]
						],
						"*",
						[5, "1.7"]
					],
					"*",
					[
						1,
						[5, "1"],
						"-",
						"tableMask"
					]
				]
			],
			[
				13,
				"nSurfFlat",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								"gradX",
								"*",
								"bevelSlope"
							],
							[
								1,
								"gradY",
								"*",
								"bevelSlope"
							],
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]]
				]
			],
			[
				13,
				"nSurf",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"nSurfFlat",
						"nSurfVol",
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"cosView",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								7,
								"nSurf",
								"z"
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"grazing",
				[
					1,
					[5, "1"],
					"-",
					"cosView"
				]
			],
			[
				13,
				"vScale1",
				[
					1,
					[
						1,
						"facets",
						"*",
						[5, "1.2"]
					],
					"*",
					"radialDensity"
				]
			],
			[
				13,
				"fUvS1",
				[
					6,
					[
						7,
						"fUv",
						"mul"
					],
					["vScale1"]
				]
			],
			[
				13,
				"feps",
				[5, "0.02"]
			],
			[
				13,
				"f1center",
				[
					6,
					[
						7,
						"noise",
						"mxWorleyNoiseFloat2Pub"
					],
					["fUvS1", [5, "1"]]
				]
			],
			[
				13,
				"gx1",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"noise",
								"mxWorleyNoiseFloat2Pub"
							],
							[[
								6,
								[
									7,
									"fUvS1",
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									["feps", [5, "0"]]
								]]
							], [5, "1"]]
						],
						"-",
						"f1center"
					],
					"/",
					"feps"
				]
			],
			[
				13,
				"gy1",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"noise",
								"mxWorleyNoiseFloat2Pub"
							],
							[[
								6,
								[
									7,
									"fUvS1",
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									[[5, "0"], "feps"]
								]]
							], [5, "1"]]
						],
						"-",
						"f1center"
					],
					"/",
					"feps"
				]
			],
			[
				13,
				"gLen1",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"gx1",
								"*",
								"gx1"
							],
							"+",
							[
								1,
								"gy1",
								"*",
								"gy1"
							]
						]]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"nx1",
				[
					1,
					"gx1",
					"/",
					"gLen1"
				]
			],
			[
				13,
				"ny1",
				[
					1,
					"gy1",
					"/",
					"gLen1"
				]
			],
			[
				13,
				"cellId1",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["fUvS1"]
				]
			],
			[
				13,
				"cellRand",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									[
										7,
										"cellId1",
										"x"
									],
									"*",
									[5, "12.9898"]
								],
								"+",
								[
									1,
									[
										7,
										"cellId1",
										"y"
									],
									"*",
									[5, "78.233"]
								]
							]]
						],
						"*",
						[5, "43758.5453"]
					]]
				]
			],
			[
				13,
				"nF1",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"nSurf",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									"nx1",
									"*",
									[5, "0.55"]
								],
								[
									1,
									"ny1",
									"*",
									[5, "0.55"]
								],
								[5, "0"]
							]
						]]
					]]
				]
			],
			[
				13,
				"ndlS",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["nSurf", "L"]
				]
			],
			[
				13,
				"ndl1",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["nF1", "L"]
						],
						"-",
						"ndlS"
					],
					"*",
					[5, "2.2"]
				]
			],
			[
				13,
				"fUvS2",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"fUv",
								"mul"
							],
							[[
								1,
								"vScale1",
								"*",
								[5, "1.7"]
							]]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "3.7"], [5, "1.2"]]
					]]
				]
			],
			[
				13,
				"c2",
				[
					6,
					[
						7,
						"noise",
						"mxWorleyNoiseFloat2Pub"
					],
					["fUvS2", [5, "1"]]
				]
			],
			[
				13,
				"gx2",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"noise",
								"mxWorleyNoiseFloat2Pub"
							],
							[[
								6,
								[
									7,
									"fUvS2",
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									["feps", [5, "0"]]
								]]
							], [5, "1"]]
						],
						"-",
						"c2"
					],
					"/",
					"feps"
				]
			],
			[
				13,
				"gy2",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"noise",
								"mxWorleyNoiseFloat2Pub"
							],
							[[
								6,
								[
									7,
									"fUvS2",
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									[[5, "0"], "feps"]
								]]
							], [5, "1"]]
						],
						"-",
						"c2"
					],
					"/",
					"feps"
				]
			],
			[
				13,
				"gLen2",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"gx2",
								"*",
								"gx2"
							],
							"+",
							[
								1,
								"gy2",
								"*",
								"gy2"
							]
						]]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"nx2",
				[
					1,
					"gx2",
					"/",
					"gLen2"
				]
			],
			[
				13,
				"ny2",
				[
					1,
					"gy2",
					"/",
					"gLen2"
				]
			],
			[
				13,
				"nF2",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"nSurf",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									"nx2",
									"*",
									[5, "0.4"]
								],
								[
									1,
									"ny2",
									"*",
									[5, "0.4"]
								],
								[5, "0"]
							]
						]]
					]]
				]
			],
			[
				13,
				"ndl2",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["nF2", "L2"]
						],
						"-",
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["nSurf", "L2"]
						]
					],
					"*",
					[5, "2.2"]
				]
			],
			[
				13,
				"I",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"facetRefrAmp",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"refraction",
							"*",
							[5, "0.8"]
						],
						[5, "0"],
						[5, "1.6"]
					]
				]
			],
			[
				13,
				"nRefr",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"nSurf",
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec3f"
									],
									[
										[
											1,
											[
												1,
												[
													1,
													"nx1",
													"*",
													[5, "0.75"]
												],
												"*",
												"facetRefrAmp"
											],
											"*",
											"tableSuppress"
										],
										[
											1,
											[
												1,
												[
													1,
													"ny1",
													"*",
													[5, "0.75"]
												],
												"*",
												"facetRefrAmp"
											],
											"*",
											"tableSuppress"
										],
										[5, "0"]
									]
								]]
							],
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									[
										1,
										"nx2",
										"*",
										[5, "0.55"]
									],
									"*",
									"facetRefrAmp"
								],
								[
									1,
									[
										1,
										"ny2",
										"*",
										[5, "0.55"]
									],
									"*",
									"facetRefrAmp"
								],
								[5, "0"]
							]
						]]
					]]
				]
			],
			[
				13,
				"eta",
				[
					1,
					[5, "1"],
					"/",
					[
						1,
						[5, "1"],
						"+",
						[
							1,
							"refraction",
							"*",
							[5, "0.55"]
						]
					]
				]
			],
			[
				13,
				"refr",
				[
					6,
					[
						7,
						"std",
						"refract"
					],
					[
						"I",
						"nRefr",
						"eta"
					]
				]
			],
			[
				13,
				"thickVol",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"sdfRaw",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						[5, "0"],
						[5, "0.6"]
					]
				]
			],
			[
				13,
				"thickFlat",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								"sdfRaw",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							[5, "0"],
							[5, "0.12"]
						]
					],
					"*",
					[5, "1.6"]
				]
			],
			[
				13,
				"thickField",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"thickFlat",
						"thickVol",
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"thickEff",
				"thickField"
			],
			[
				13,
				"thickFacet",
				[
					1,
					"thickEff",
					"*",
					[
						1,
						[5, "0.55"],
						"+",
						[
							1,
							"cellRand",
							"*",
							[5, "0.9"]
						]
					]
				]
			],
			[
				13,
				"offMag",
				[
					1,
					[
						1,
						"thickFacet",
						"*",
						"scale"
					],
					"*",
					"rb1"
				]
			],
			[
				13,
				"totalOffX",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									7,
									"refr",
									"x"
								],
								"/",
								[
									6,
									[
										7,
										"std",
										"max"
									],
									[[
										7,
										"refr",
										"z"
									], [5, "0.25"]]
								]
							],
							"*",
							"offMag"
						],
						"/",
						"aspect"
					],
					"+",
					"haloOffX"
				]
			],
			[
				13,
				"totalOffY",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"refr",
								"y"
							],
							"/",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								[[
									7,
									"refr",
									"z"
								], [5, "0.25"]]
							]
						],
						"*",
						"offMag"
					],
					"+",
					"haloOffY"
				]
			],
			[
				13,
				"nBack",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								"nx2",
								"*",
								[5, "0.7"]
							],
							[
								1,
								"ny2",
								"*",
								[5, "0.7"]
							],
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]]
				]
			],
			[
				13,
				"bounce",
				[
					6,
					[
						7,
						"std",
						"reflect"
					],
					["refr", "nBack"]
				]
			],
			[
				13,
				"bounceMag",
				[
					1,
					[
						1,
						[
							1,
							"thickEff",
							"*",
							"scale"
						],
						"*",
						"rb1"
					],
					"*",
					[5, "0.8"]
				]
			],
			[
				13,
				"bounceOffX",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"bounce",
								"x"
							],
							"/",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								[[
									6,
									[
										7,
										"std",
										"abs"
									],
									[[
										7,
										"bounce",
										"z"
									]]
								], [5, "0.3"]]
							]
						],
						"*",
						"bounceMag"
					],
					"/",
					"aspect"
				]
			],
			[
				13,
				"bounceOffY",
				[
					1,
					[
						1,
						[
							7,
							"bounce",
							"y"
						],
						"/",
						[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									7,
									"bounce",
									"z"
								]]
							], [5, "0.3"]]
						]
					],
					"*",
					"bounceMag"
				]
			],
			[
				13,
				"innerW",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								"grazing",
								"*",
								[5, "0.5"]
							],
							"+",
							[5, "0.18"]
						],
						"*",
						[
							1,
							[
								1,
								"cellRand",
								"*",
								[5, "0.7"]
							],
							"+",
							[5, "0.3"]
						]
					],
					"*",
					"rb1"
				]
			],
			[
				13,
				"pbx",
				[
					1,
					"pdx",
					"+",
					[
						1,
						[
							1,
							[
								1,
								[
									7,
									"refr",
									"x"
								],
								"/",
								[
									6,
									[
										7,
										"std",
										"max"
									],
									[[
										7,
										"refr",
										"z"
									], [5, "0.25"]]
								]
							],
							"*",
							"thickEff"
						],
						"*",
						[5, "1.2"]
					]
				]
			],
			[
				13,
				"pby",
				[
					1,
					"pdy",
					"+",
					[
						1,
						[
							1,
							[
								1,
								[
									7,
									"refr",
									"y"
								],
								"/",
								[
									6,
									[
										7,
										"std",
										"max"
									],
									[[
										7,
										"refr",
										"z"
									], [5, "0.25"]]
								]
							],
							"*",
							"thickEff"
						],
						"*",
						[5, "1.2"]
					]
				]
			],
			[
				13,
				"fUvB",
				[
					6,
					"crystalKaleidoFold",
					[
						[
							1,
							"pbx",
							"+",
							[5, "0.5"]
						],
						[
							1,
							"pby",
							"+",
							[5, "0.5"]
						],
						"sAngle"
					]
				]
			],
			[
				13,
				"fUvS3",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"fUvB",
								"mul"
							],
							[[
								1,
								"vScale1",
								"*",
								[5, "1.4"]
							]]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "9.1"], [5, "4.7"]]
					]]
				]
			],
			[
				13,
				"c3",
				[
					6,
					[
						7,
						"noise",
						"mxWorleyNoiseFloat2Pub"
					],
					["fUvS3", [5, "1"]]
				]
			],
			[
				13,
				"gx3",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"noise",
								"mxWorleyNoiseFloat2Pub"
							],
							[[
								6,
								[
									7,
									"fUvS3",
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									["feps", [5, "0"]]
								]]
							], [5, "1"]]
						],
						"-",
						"c3"
					],
					"/",
					"feps"
				]
			],
			[
				13,
				"gy3",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"noise",
								"mxWorleyNoiseFloat2Pub"
							],
							[[
								6,
								[
									7,
									"fUvS3",
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									[[5, "0"], "feps"]
								]]
							], [5, "1"]]
						],
						"-",
						"c3"
					],
					"/",
					"feps"
				]
			],
			[
				13,
				"gLen3",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"gx3",
								"*",
								"gx3"
							],
							"+",
							[
								1,
								"gy3",
								"*",
								"gy3"
							]
						]]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"nx3",
				[
					1,
					"gx3",
					"/",
					"gLen3"
				]
			],
			[
				13,
				"ny3",
				[
					1,
					"gy3",
					"/",
					"gLen3"
				]
			],
			[
				13,
				"cellId3",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["fUvS3"]
				]
			],
			[
				13,
				"cellRand3",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									[
										7,
										"cellId3",
										"x"
									],
									"*",
									[5, "31.727"]
								],
								"+",
								[
									1,
									[
										7,
										"cellId3",
										"y"
									],
									"*",
									[5, "57.131"]
								]
							]]
						],
						"*",
						[5, "43758.5453"]
					]]
				]
			],
			[
				13,
				"w3",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								[
									1,
									"thickEff",
									"*",
									[5, "2.5"]
								],
								[5, "0"],
								[5, "1"]
							]
						],
						"*",
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								[
									1,
									"refraction",
									"*",
									[5, "1.2"]
								],
								[5, "0"],
								[5, "1"]
							]
						]
					],
					"*",
					"rb1"
				]
			],
			[
				13,
				"wedgeLit",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								"nx3",
								"*",
								"lx"
							],
							"+",
							[
								1,
								"ny3",
								"*",
								"ly"
							]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"mosaicLevel",
				[
					1,
					[
						1,
						[
							1,
							"cellRand3",
							"*",
							[5, "0.9"]
						],
						"+",
						[5, "0.55"]
					],
					"*",
					[
						1,
						[
							1,
							"wedgeLit",
							"*",
							[5, "0.7"]
						],
						"+",
						[5, "0.65"]
					]
				]
			],
			[
				13,
				"edge3",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[5, "0.15"],
								[5, "0.7"],
								"gLen3"
							]
						],
						"*",
						[5, "0.45"]
					],
					"+",
					[5, "0.55"]
				]
			],
			[
				13,
				"mosaicMul",
				[
					1,
					"mosaicLevel",
					"*",
					"edge3"
				]
			],
			[
				13,
				"mosaicFire",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[5, "0.75"],
								[5, "1"],
								[
									1,
									"wedgeLit",
									"*",
									[
										1,
										"cellRand3",
										"+",
										[5, "0.4"]
									]
								]
							]
						],
						"*",
						"w3"
					],
					"*",
					[5, "0.8"]
				]
			],
			[
				13,
				"zoomedX",
				[
					1,
					"centerPosX",
					"+",
					[
						1,
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"-",
							"centerPosX"
						],
						"/",
						"innerZoom"
					]
				]
			],
			[
				13,
				"zoomedY",
				[
					1,
					"centerPosY",
					"+",
					[
						1,
						[
							1,
							[
								7,
								"uv",
								"y"
							],
							"-",
							"centerPosY"
						],
						"/",
						"innerZoom"
					]
				]
			],
			[
				13,
				"lensX",
				[
					1,
					"zoomedX",
					"+",
					"totalOffX"
				]
			],
			[
				13,
				"lensY",
				[
					1,
					"zoomedY",
					"+",
					"totalOffY"
				]
			],
			[
				13,
				"lit1",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"ndl1",
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"shadow1",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"ndl1",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"bright1",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[
							1,
							[5, "1"],
							"+",
							[
								1,
								"lit1",
								"*",
								"highlights"
							]
						],
						"-",
						[
							1,
							"shadow1",
							"*",
							"shadows"
						]
					], [5, "0.02"]]
				]
			],
			[
				13,
				"lit2",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"ndl2",
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"shadow2",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"ndl2",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"bright2",
				[
					1,
					[
						1,
						[5, "1"],
						"+",
						[
							1,
							"lit2",
							"*",
							[
								1,
								"highlights",
								"*",
								[5, "0.5"]
							]
						]
					],
					"-",
					[
						1,
						"shadow2",
						"*",
						[
							1,
							"shadows",
							"*",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"surfBright",
				[
					1,
					[
						1,
						[5, "1"],
						"+",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										"ndlS",
										[5, "0"],
										[5, "1"]
									]
								],
								"*",
								"highlights"
							],
							"*",
							[5, "0.4"]
						]
					],
					"-",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									[
										1,
										"ndlS",
										"*",
										[
											4,
											"-",
											[5, "1"]
										]
									],
									[5, "0"],
									[5, "1"]
								]
							],
							"*",
							"shadows"
						],
						"*",
						[5, "0.6"]
					]
				]
			],
			[
				13,
				"combinedBright",
				[
					1,
					[
						1,
						"bright1",
						"*",
						"bright2"
					],
					"*",
					"surfBright"
				]
			],
			[
				13,
				"chrScale",
				[
					1,
					"dispersion",
					"*",
					[5, "0.12"]
				]
			],
			[
				13,
				"chrOffX",
				[
					1,
					"totalOffX",
					"*",
					"chrScale"
				]
			],
			[
				13,
				"chrOffY",
				[
					1,
					"totalOffY",
					"*",
					"chrScale"
				]
			],
			[
				13,
				"baseEdge",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.2"],
						[5, "0.8"],
						"gLen1"
					]
				]
			],
			[
				13,
				"edgeDarken",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								"baseEdge",
								[5, "1"],
								"tableMask"
							]
						],
						"*",
						[5, "0.4"]
					],
					"+",
					[5, "0.6"]
				]
			],
			[
				13,
				"tableBoost",
				[
					1,
					"tableMask",
					"*",
					[5, "0.25"]
				]
			],
			[
				13,
				"facetMul",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								"combinedBright",
								"+",
								"tableBoost"
							],
							"*",
							"edgeDarken"
						],
						"*",
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "1"]]
								],
								"mosaicMul",
								"w3"
							]
						]
					],
					"*",
					"brightness"
				]
			],
			[
				13,
				"ndh1",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["nF1", "H"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"ndh2",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["nF2", "H2"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"spec1",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"pow"
						],
						["ndh1", [
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "110"]]
						]]
					],
					"*",
					[5, "2.2"]
				]
			],
			[
				13,
				"spec2",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"pow"
						],
						["ndh2", [
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "56"]]
						]]
					],
					"*",
					[5, "0.9"]
				]
			],
			[
				13,
				"glint",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0.965"],
							[5, "0.995"],
							"ndh1"
						]
					],
					"*",
					[5, "3"]
				]
			],
			[
				13,
				"flashContrib",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									"spec1",
									"+",
									"spec2"
								],
								"+",
								"glint"
							],
							"+",
							"mosaicFire"
						],
						"*",
						"highlights"
					],
					"*",
					"brightness"
				]
			],
			[
				13,
				"fresnelExp",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "6"],
						[5, "1.5"],
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								[
									1,
									"fresnelSoftness",
									"*",
									[5, "0.5"]
								],
								[5, "0"],
								[5, "1"]
							]
						]
					]
				]
			],
			[
				13,
				"fresnelRim",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"pow"
								],
								["grazing", "fresnelExp"]
							],
							"*",
							"fresnel"
						],
						"*",
						[5, "2"]
					],
					"*",
					"rb1"
				]
			],
			[
				13,
				"fresnelContrib",
				[
					6,
					[
						7,
						"fresnelColor",
						"mul"
					],
					["fresnelRim"]
				]
			],
			[
				13,
				"normalDotLight",
				[
					1,
					[
						1,
						"gradX",
						"*",
						"lx"
					],
					"+",
					[
						1,
						"gradY",
						"*",
						"ly"
					]
				]
			],
			[
				13,
				"lightFacing",
				[
					1,
					[
						1,
						"normalDotLight",
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"sdfOuter",
				[
					1,
					"sdf",
					"-",
					"pxH"
				]
			],
			[
				13,
				"rb2base",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									[
										1,
										"sdfOuter",
										"*",
										[
											4,
											"-",
											[5, "1"]
										]
									],
									"/",
									"sharp"
								],
								"*",
								[5, "16"]
							],
							[5, "0"],
							[5, "1"]
						]
					],
					"-",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									[
										1,
										"sdf",
										"*",
										[
											4,
											"-",
											[5, "1"]
										]
									],
									"/",
									"sharp"
								],
								"*",
								[5, "16"]
							],
							[5, "0"],
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"rb2",
				[
					1,
					[
						1,
						"rb2base",
						"*",
						"lightFacing"
					],
					"*",
					[5, "0.35"]
				]
			],
			[
				13,
				"transition",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[5, "1"],
						"rb1"
					]
				]
			],
			[
				13,
				"outerUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"+",
						"haloOffX"
					], [
						1,
						[
							7,
							"uv",
							"y"
						],
						"+",
						"haloOffY"
					]]
				]
			],
			[10, [
				6,
				"CrystalTrace",
				[[104, {
					"rUV": [
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							"lensX",
							"+",
							"chrOffX"
						], [
							1,
							"lensY",
							"+",
							"chrOffY"
						]]
					],
					"lensUV": [
						6,
						[
							7,
							"d",
							"vec2f"
						],
						["lensX", "lensY"]
					],
					"bUV": [
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							"lensX",
							"-",
							"chrOffX"
						], [
							1,
							"lensY",
							"-",
							"chrOffY"
						]]
					],
					"innerUV": [
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							"lensX",
							"+",
							"bounceOffX"
						], [
							1,
							"lensY",
							"+",
							"bounceOffY"
						]]
					],
					"outerUV": "outerUV",
					"factors": [
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							"facetMul",
							"flashContrib",
							"innerW",
							"rb2"
						]
					],
					"transition": "transition",
					"fresnelContrib": "fresnelContrib"
				}]]
			]]
		]],
		"externalNames": [
			"std",
			"DEG_TO_RAD",
			"d",
			"TWO_PI",
			"crystalKaleidoFold",
			"noise",
			"CrystalTrace"
		]
	},
	externals: () => ({
		std,
		DEG_TO_RAD,
		d,
		TWO_PI,
		crystalKaleidoFold,
		noise: noise_exports,
		CrystalTrace
	})
}) && $.f)({}));
const crystalCompose = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec3f,
	d.vec3f,
	d.vec4f,
	d.vec4f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((sampledR, sampledG, sampledB, innerSample, baseColor, fresnelContrib, tintColor, factors, misc) => {
	"use gpu";
	const facetMul = factors.x;
	const flashContrib = factors.y;
	const innerW = factors.z;
	const rb2 = factors.w;
	const transition = misc.x;
	const tintIntensity = misc.y;
	const preserveLum = misc.z;
	const cutout = misc.w;
	const primary = d.vec3f(sampledR.x, sampledG.y, sampledB.z);
	const innerRgb = d.vec3f(innerSample.x, innerSample.y, innerSample.z);
	const refractedRgb = std.mix(primary, innerRgb, d.vec3f(innerW));
	const refractedA = sampledG.w;
	const litRgb = refractedRgb.mul(facetMul);
	const lumWeights = d.vec3f(.299, .587, .114);
	const origLum = std.dot(litRgb, lumWeights);
	const tinted = std.mix(litRgb, tintColor, d.vec3f(tintIntensity));
	const tintedLum = std.dot(tinted, lumWeights);
	const lumPreserved = tinted.mul(__tsover_div(origLum, std.max(tintedLum, 1e-4)));
	const finalRgb = std.select(tinted, lumPreserved, preserveLum > .5).add(d.vec3f(flashContrib, flashContrib, flashContrib)).add(fresnelContrib).add(d.vec3f(rb2, rb2, rb2));
	const lighting = d.vec4f(finalRgb.x, finalRgb.y, finalRgb.z, refractedA);
	const composited = std.mix(baseColor, lighting, d.vec4f(transition));
	const cutoutAlpha = std.select(d.f32(1), transition, cutout > .5);
	return d.vec4f(composited.x, composited.y, composited.z, __tsover_mul(composited.w, cutoutAlpha));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "sampledR"
			},
			{
				"type": "i",
				"name": "sampledG"
			},
			{
				"type": "i",
				"name": "sampledB"
			},
			{
				"type": "i",
				"name": "innerSample"
			},
			{
				"type": "i",
				"name": "baseColor"
			},
			{
				"type": "i",
				"name": "fresnelContrib"
			},
			{
				"type": "i",
				"name": "tintColor"
			},
			{
				"type": "i",
				"name": "factors"
			},
			{
				"type": "i",
				"name": "misc"
			}
		],
		"body": [0, [
			[
				13,
				"facetMul",
				[
					7,
					"factors",
					"x"
				]
			],
			[
				13,
				"flashContrib",
				[
					7,
					"factors",
					"y"
				]
			],
			[
				13,
				"innerW",
				[
					7,
					"factors",
					"z"
				]
			],
			[
				13,
				"rb2",
				[
					7,
					"factors",
					"w"
				]
			],
			[
				13,
				"transition",
				[
					7,
					"misc",
					"x"
				]
			],
			[
				13,
				"tintIntensity",
				[
					7,
					"misc",
					"y"
				]
			],
			[
				13,
				"preserveLum",
				[
					7,
					"misc",
					"z"
				]
			],
			[
				13,
				"cutout",
				[
					7,
					"misc",
					"w"
				]
			],
			[
				13,
				"primary",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"sampledR",
							"x"
						],
						[
							7,
							"sampledG",
							"y"
						],
						[
							7,
							"sampledB",
							"z"
						]
					]
				]
			],
			[
				13,
				"innerRgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"innerSample",
							"x"
						],
						[
							7,
							"innerSample",
							"y"
						],
						[
							7,
							"innerSample",
							"z"
						]
					]
				]
			],
			[
				13,
				"refractedRgb",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"primary",
						"innerRgb",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["innerW"]
						]
					]
				]
			],
			[
				13,
				"refractedA",
				[
					7,
					"sampledG",
					"w"
				]
			],
			[
				13,
				"litRgb",
				[
					6,
					[
						7,
						"refractedRgb",
						"mul"
					],
					["facetMul"]
				]
			],
			[
				13,
				"lumWeights",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0.299"],
						[5, "0.587"],
						[5, "0.114"]
					]
				]
			],
			[
				13,
				"origLum",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["litRgb", "lumWeights"]
				]
			],
			[
				13,
				"tinted",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"litRgb",
						"tintColor",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["tintIntensity"]
						]
					]
				]
			],
			[
				13,
				"tintedLum",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["tinted", "lumWeights"]
				]
			],
			[
				13,
				"lumPreserved",
				[
					6,
					[
						7,
						"tinted",
						"mul"
					],
					[[
						1,
						"origLum",
						"/",
						[
							6,
							[
								7,
								"std",
								"max"
							],
							["tintedLum", [5, "0.0001"]]
						]
					]]
				]
			],
			[
				13,
				"crystalRgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"std",
								"select"
							],
							[
								"tinted",
								"lumPreserved",
								[
									1,
									"preserveLum",
									">",
									[5, "0.5"]
								]
							]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"flashContrib",
							"flashContrib",
							"flashContrib"
						]
					]]
				]
			],
			[
				13,
				"finalRgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"crystalRgb",
								"add"
							],
							["fresnelContrib"]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"rb2",
							"rb2",
							"rb2"
						]
					]]
				]
			],
			[
				13,
				"lighting",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							7,
							"finalRgb",
							"x"
						],
						[
							7,
							"finalRgb",
							"y"
						],
						[
							7,
							"finalRgb",
							"z"
						],
						"refractedA"
					]
				]
			],
			[
				13,
				"composited",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"baseColor",
						"lighting",
						[
							6,
							[
								7,
								"d",
								"vec4f"
							],
							["transition"]
						]
					]
				]
			],
			[
				13,
				"cutoutAlpha",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						"transition",
						[
							1,
							"cutout",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"composited",
						"x"
					],
					[
						7,
						"composited",
						"y"
					],
					[
						7,
						"composited",
						"z"
					],
					[
						1,
						[
							7,
							"composited",
							"w"
						],
						"*",
						"cutoutAlpha"
					]
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Crystal",
	boundingBoxDeclaration: shapeEffectBoundingBoxDeclaration,
	category: "Shape Effects",
	description: "Diamond-like crystal lens with faceted refraction.",
	requiresRTT: true,
	requiresChild: true,
	extraFields: {
		...VOLUMETRIC_FIELD_EXTRA_FIELDS,
		...ANALYTIC_SDF_EXTRA_FIELDS
	},
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the crystal shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		scale: {
			default: 1,
			description: "Scale of the crystal shape (1 = default size)",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Scale",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the crystal shape in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		},
		cutout: {
			default: false,
			transform: transformBoolean,
			description: "Cut out alpha outside the crystal shape",
			compileTime: true,
			ui: {
				type: "checkbox",
				label: "Cutout",
				group: "Crystal"
			}
		},
		refraction: {
			default: .5,
			description: "How strongly the crystal refracts content beneath",
			ui: {
				type: "range",
				min: 0,
				max: 3,
				step: .01,
				label: "Refraction",
				group: "Crystal"
			}
		},
		dispersion: {
			default: .5,
			description: "Prismatic rainbow dispersion — splits light into spectral colors",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Dispersion",
				group: "Crystal"
			}
		},
		facets: {
			default: 5,
			description: "Symmetry order — how many times the facet pattern repeats around the center",
			ui: {
				type: "range",
				min: 3,
				max: 24,
				step: 1,
				label: "Facets",
				group: "Crystal"
			}
		},
		fresnel: {
			default: .05,
			description: "Fresnel rim glow intensity around the crystal boundary",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Fresnel",
				group: "Fresnel"
			}
		},
		fresnelSoftness: {
			default: 1,
			description: "Fresnel rim width — higher values spread the glow further inward",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Fresnel Softness",
				group: "Fresnel"
			}
		},
		fresnelColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color of the fresnel rim glow",
			ui: {
				type: "color",
				label: "Fresnel Color",
				group: "Fresnel"
			}
		},
		edgeSoftness: {
			default: 0,
			description: "Softness of the crystal boundary edge",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Softness",
				group: "Crystal"
			}
		},
		innerZoom: {
			default: 1.5,
			description: "Magnification of content seen through the crystal",
			ui: {
				type: "range",
				min: .5,
				max: 3,
				step: .01,
				label: "Inner Zoom",
				group: "Crystal"
			}
		},
		lightAngle: {
			default: 270,
			description: "Light direction angle in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Light Angle",
				group: "Lighting"
			}
		},
		highlights: {
			default: .5,
			description: "Additive brightness on light-facing facets — never darkens",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Highlights",
				group: "Lighting"
			}
		},
		shadows: {
			default: .3,
			description: "Darkening on shadow-facing facets — never brightens",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Shadows",
				group: "Lighting"
			}
		},
		brightness: {
			default: 1.2,
			description: "Overall crystal brightness — higher values push facets toward brilliant white",
			ui: {
				type: "range",
				min: .5,
				max: 3,
				step: .01,
				label: "Brightness",
				group: "Lighting"
			}
		},
		tintColor: {
			default: "#e8e0ff",
			transform: transformColor,
			description: "Crystal body tint color",
			ui: {
				type: "color",
				label: "Tint Color",
				group: "Tint"
			}
		},
		tintIntensity: {
			default: 0,
			description: "How much tint color is applied to the crystal interior",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Tint Intensity",
				group: "Tint"
			}
		},
		tintPreserveLuminosity: {
			default: true,
			transform: transformBoolean,
			description: "Preserve original brightness when tinting",
			ui: {
				type: "checkbox",
				label: "Preserve Luminosity",
				group: "Tint"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type"
		}
	},
	compute: (params) => {
		const vf = createVolumetricFieldComputeNode(params, () => params.getCpuValue("shape"), "triplanar");
		return vf ? {
			outputs: vf.outputs,
			getComputeNodes: (fp) => vf.getComputeNodes(fp)
		} : null;
	},
	fragment: (params) => {
		const { childNode, computeOutputs, convertToTexture, uniforms, propValues, ctx, getCpuValue } = params;
		if (!childNode) return ZERO;
		const childTexture = convertToTexture(childNode);
		let sampler;
		let gradSampler;
		let volumetric = false;
		const fieldTex = computeOutputs?.volumetricFieldTexture;
		if (fieldTex) {
			const s = buildVolumetricFieldSampler(fieldTex, {
				originX: uniforms._vfOriginX,
				originY: uniforms._vfOriginY,
				spanX: uniforms._vfSpanX,
				spanY: uniforms._vfSpanY,
				activeResF: uniforms._vfActiveRes,
				rBound: uniforms._vfRBound
			});
			sampler = s.volumetricFieldSampler;
			gradSampler = s.volumetricFieldSamplerFast;
			volumetric = true;
		} else {
			const shapeSdfUrl = propValues.shapeSdfUrl || "";
			const shapeType = propValues.shapeType || "";
			if (shapeSdfUrl) {
				sampler = createSvgSdfSampler(params, shapeSdfUrl);
				gradSampler = sampler;
				volumetric = isVolumetricShapeType(shapeType);
			} else {
				const st = resolveShapeType(shapeType, getCpuValue("shape"));
				sampler = createAnalyticSdfSampler(st, driveAnalyticSubProps(params, () => getCpuValue("shape")), "triplanar");
				gradSampler = sampler;
				volumetric = is3dShapeType(st);
			}
		}
		const sdfUV = call(sdfSpaceUV, "sdfSpaceUV", [
			uniforms.center,
			uniforms.scale,
			uniforms.rotation,
			ctx.uv,
			ctx.aspect
		]);
		const surf0 = sampler(sdfUV);
		const surfX = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(.01),
			floatE(0)
		]));
		const surfY = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(0),
			floatE(.01)
		]));
		const volFlag = floatE(volumetric ? 1 : 0);
		const trace = call(crystalTrace, "crystalTrace", [
			surf0,
			surfX,
			surfY,
			sdfUV,
			ctx.uv,
			ctx.viewportSize,
			uniforms.center,
			uniforms.fresnelColor.member("rgb"),
			vec4(uniforms.scale, uniforms.edgeSoftness, uniforms.refraction, uniforms.dispersion),
			vec4(uniforms.facets, uniforms.innerZoom, uniforms.lightAngle, uniforms.highlights),
			vec4(uniforms.shadows, uniforms.brightness, uniforms.fresnel, uniforms.fresnelSoftness),
			vec4(volFlag, floatE(0), floatE(0), floatE(0))
		]);
		const baseSampleUV = Number(propValues.cutout) > 0 ? ctx.uv : trace.member("outerUV");
		return call(crystalCompose, "crystalCompose", [
			call(unpremultiplyAlpha, "unpremultiplyAlpha", [childTexture.sample(trace.member("rUV"))]),
			call(unpremultiplyAlpha, "unpremultiplyAlpha", [childTexture.sample(trace.member("lensUV"))]),
			call(unpremultiplyAlpha, "unpremultiplyAlpha", [childTexture.sample(trace.member("bUV"))]),
			call(unpremultiplyAlpha, "unpremultiplyAlpha", [childTexture.sample(trace.member("innerUV"))]),
			call(unpremultiplyAlpha, "unpremultiplyAlpha", [childTexture.sample(baseSampleUV)]),
			trace.member("fresnelContrib"),
			uniforms.tintColor.member("rgb"),
			trace.member("factors"),
			vec4(trace.member("transition"), uniforms.tintIntensity, uniforms.tintPreserveLuminosity, uniforms.cutout)
		]);
	}
};
var Crystal_default = componentDefinition;
export { crystalKaleidoFold as a, crystalCompose as i, Crystal_default as n, crystalTrace as o, componentDefinition as r, CrystalTrace as t };
