// shaders/dist/core/CursorRipples-CLYoq-BF.js
import { At as call, B as createGuardedCompute, It as vec4, Nt as expr, Ot as ZERO, R as std, V as createStateBuffer, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { s as transformEdges } from "./transformations-fqzznt4g.js";
var GRID_SIZE = 128;
var CELL_COUNT = GRID_SIZE * GRID_SIZE;
var GRID_MAX = GRID_SIZE - 1;
var STATE_FORMAT = "rgba16float";
var propagateLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	readBuf: {
		storage: d.arrayOf(d.f32, CELL_COUNT),
		access: "readonly"
	},
	writeBuf: {
		storage: d.arrayOf(d.f32, CELL_COUNT),
		access: "mutable"
	},
	params: { uniform: /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
		cursorX: d.f32,
		cursorY: d.f32,
		mouseSpeed: d.f32,
		dt: d.f32,
		decay: d.f32,
		radius: d.f32,
		aspect: d.f32
	}), "uniform") }
}), "propagateLayout");
var gradientLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	srcBuf: {
		storage: d.arrayOf(d.f32, CELL_COUNT),
		access: "readonly"
	},
	dispTex: { storageTexture: d.textureStorage2d(STATE_FORMAT, "write-only") }
}), "gradientLayout");
const propagateKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	if (cx > d.u32(0) && cx < d.u32(GRID_MAX) && cy > d.u32(0) && cy < d.u32(GRID_MAX)) {
		const p = propagateLayout.$.params;
		const gs = d.u32(GRID_SIZE);
		const idx = __tsover_add(__tsover_mul(cy, gs), cx);
		const left = propagateLayout.$.readBuf[__tsover_sub(idx, d.u32(1))];
		const right = propagateLayout.$.readBuf[__tsover_add(idx, d.u32(1))];
		const up = propagateLayout.$.readBuf[__tsover_sub(idx, gs)];
		const down = propagateLayout.$.readBuf[__tsover_add(idx, gs)];
		const avg = __tsover_mul(__tsover_add(__tsover_add(__tsover_add(left, right), up), down), .5);
		const prev = propagateLayout.$.writeBuf[idx];
		const dampFactor = __tsover_sub(1, __tsover_mul(p.decay, .004));
		let next = __tsover_mul(__tsover_sub(avg, prev), dampFactor);
		const cellX = __tsover_div(__tsover_add(d.f32(cx), .5), d.f32(GRID_SIZE));
		const cellY = __tsover_div(__tsover_add(d.f32(cy), .5), d.f32(GRID_SIZE));
		const ddx = std.select(__tsover_sub(cellX, p.cursorX), __tsover_mul(__tsover_sub(cellX, p.cursorX), p.aspect), p.aspect >= 1);
		const ddy = std.select(__tsover_div(__tsover_sub(cellY, p.cursorY), p.aspect), __tsover_sub(cellY, p.cursorY), p.aspect >= 1);
		const distSq = __tsover_add(__tsover_mul(ddx, ddx), __tsover_mul(ddy, ddy));
		const influenceRadius = __tsover_mul(p.radius, 3);
		if (distSq < __tsover_mul(influenceRadius, influenceRadius) && p.mouseSpeed > .01) {
			const influence = std.exp(__tsover_mul(__tsover_div(distSq, __tsover_mul(p.radius, p.radius)), -1));
			next = __tsover_sub(next, __tsover_mul(__tsover_mul(__tsover_mul(influence, p.mouseSpeed), p.dt), 3));
		}
		propagateLayout.$.writeBuf[idx] = next;
	}
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [[
			11,
			[
				3,
				[
					3,
					[
						3,
						[
							1,
							"cx",
							">",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "0"]]
							]
						],
						"&&",
						[
							1,
							"cx",
							"<",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								["GRID_MAX"]
							]
						]
					],
					"&&",
					[
						1,
						"cy",
						">",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							[[5, "0"]]
						]
					]
				],
				"&&",
				[
					1,
					"cy",
					"<",
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						["GRID_MAX"]
					]
				]
			],
			[0, [
				[
					13,
					"p",
					[
						7,
						[
							7,
							"propagateLayout",
							"$"
						],
						"params"
					]
				],
				[
					13,
					"gs",
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						["GRID_SIZE"]
					]
				],
				[
					13,
					"idx",
					[
						1,
						[
							1,
							"cy",
							"*",
							"gs"
						],
						"+",
						"cx"
					]
				],
				[
					13,
					"left",
					[
						8,
						[
							7,
							[
								7,
								"propagateLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							"idx",
							"-",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							]
						]
					]
				],
				[
					13,
					"right",
					[
						8,
						[
							7,
							[
								7,
								"propagateLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							"idx",
							"+",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							]
						]
					]
				],
				[
					13,
					"up",
					[
						8,
						[
							7,
							[
								7,
								"propagateLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							"idx",
							"-",
							"gs"
						]
					]
				],
				[
					13,
					"down",
					[
						8,
						[
							7,
							[
								7,
								"propagateLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							"idx",
							"+",
							"gs"
						]
					]
				],
				[
					13,
					"avg",
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									"left",
									"+",
									"right"
								],
								"+",
								"up"
							],
							"+",
							"down"
						],
						"*",
						[5, "0.5"]
					]
				],
				[
					13,
					"prev",
					[
						8,
						[
							7,
							[
								7,
								"propagateLayout",
								"$"
							],
							"writeBuf"
						],
						"idx"
					]
				],
				[
					13,
					"dampFactor",
					[
						1,
						[5, "1"],
						"-",
						[
							1,
							[
								7,
								"p",
								"decay"
							],
							"*",
							[5, "0.004"]
						]
					]
				],
				[
					12,
					"next",
					[
						1,
						[
							1,
							"avg",
							"-",
							"prev"
						],
						"*",
						"dampFactor"
					]
				],
				[
					13,
					"cellX",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cx"]
							],
							"+",
							[5, "0.5"]
						],
						"/",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["GRID_SIZE"]
						]
					]
				],
				[
					13,
					"cellY",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cy"]
							],
							"+",
							[5, "0.5"]
						],
						"/",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["GRID_SIZE"]
						]
					]
				],
				[
					13,
					"ddx",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								1,
								"cellX",
								"-",
								[
									7,
									"p",
									"cursorX"
								]
							],
							[
								1,
								[
									1,
									"cellX",
									"-",
									[
										7,
										"p",
										"cursorX"
									]
								],
								"*",
								[
									7,
									"p",
									"aspect"
								]
							],
							[
								1,
								[
									7,
									"p",
									"aspect"
								],
								">=",
								[5, "1"]
							]
						]
					]
				],
				[
					13,
					"ddy",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								1,
								[
									1,
									"cellY",
									"-",
									[
										7,
										"p",
										"cursorY"
									]
								],
								"/",
								[
									7,
									"p",
									"aspect"
								]
							],
							[
								1,
								"cellY",
								"-",
								[
									7,
									"p",
									"cursorY"
								]
							],
							[
								1,
								[
									7,
									"p",
									"aspect"
								],
								">=",
								[5, "1"]
							]
						]
					]
				],
				[
					13,
					"distSq",
					[
						1,
						[
							1,
							"ddx",
							"*",
							"ddx"
						],
						"+",
						[
							1,
							"ddy",
							"*",
							"ddy"
						]
					]
				],
				[
					13,
					"influenceRadius",
					[
						1,
						[
							7,
							"p",
							"radius"
						],
						"*",
						[5, "3"]
					]
				],
				[
					11,
					[
						3,
						[
							1,
							"distSq",
							"<",
							[
								1,
								"influenceRadius",
								"*",
								"influenceRadius"
							]
						],
						"&&",
						[
							1,
							[
								7,
								"p",
								"mouseSpeed"
							],
							">",
							[5, "0.01"]
						]
					],
					[0, [[
						13,
						"influence",
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									1,
									"distSq",
									"/",
									[
										1,
										[
											7,
											"p",
											"radius"
										],
										"*",
										[
											7,
											"p",
											"radius"
										]
									]
								],
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]]
						]
					], [
						2,
						"next",
						"=",
						[
							1,
							"next",
							"-",
							[
								1,
								[
									1,
									[
										1,
										"influence",
										"*",
										[
											7,
											"p",
											"mouseSpeed"
										]
									],
									"*",
									[
										7,
										"p",
										"dt"
									]
								],
								"*",
								[5, "3"]
							]
						]
					]]]
				],
				[
					2,
					[
						8,
						[
							7,
							[
								7,
								"propagateLayout",
								"$"
							],
							"writeBuf"
						],
						"idx"
					],
					"=",
					"next"
				]
			]]
		]]],
		"externalNames": [
			"d",
			"GRID_MAX",
			"propagateLayout",
			"GRID_SIZE",
			"std"
		]
	},
	externals: () => ({
		d,
		GRID_MAX,
		propagateLayout,
		GRID_SIZE,
		std
	})
}) && $.f)({})).$name("cursorRipplesPropagate");
const gradientKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	if (cx > d.u32(0) && cx < d.u32(GRID_MAX) && cy > d.u32(0) && cy < d.u32(GRID_MAX)) {
		const gs = d.u32(GRID_SIZE);
		const idx = __tsover_add(__tsover_mul(cy, gs), cx);
		const right = gradientLayout.$.srcBuf[__tsover_add(idx, d.u32(1))];
		const left = gradientLayout.$.srcBuf[__tsover_sub(idx, d.u32(1))];
		const down = gradientLayout.$.srcBuf[__tsover_add(idx, gs)];
		const up = gradientLayout.$.srcBuf[__tsover_sub(idx, gs)];
		const dx = __tsover_mul(__tsover_sub(right, left), .5);
		const dy = __tsover_mul(__tsover_sub(down, up), .5);
		std.textureStore(gradientLayout.$.dispTex, d.vec2u(cx, cy), d.vec4f(dx, dy, 0, 1));
	}
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [[
			11,
			[
				3,
				[
					3,
					[
						3,
						[
							1,
							"cx",
							">",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "0"]]
							]
						],
						"&&",
						[
							1,
							"cx",
							"<",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								["GRID_MAX"]
							]
						]
					],
					"&&",
					[
						1,
						"cy",
						">",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							[[5, "0"]]
						]
					]
				],
				"&&",
				[
					1,
					"cy",
					"<",
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						["GRID_MAX"]
					]
				]
			],
			[0, [
				[
					13,
					"gs",
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						["GRID_SIZE"]
					]
				],
				[
					13,
					"idx",
					[
						1,
						[
							1,
							"cy",
							"*",
							"gs"
						],
						"+",
						"cx"
					]
				],
				[
					13,
					"right",
					[
						8,
						[
							7,
							[
								7,
								"gradientLayout",
								"$"
							],
							"srcBuf"
						],
						[
							1,
							"idx",
							"+",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							]
						]
					]
				],
				[
					13,
					"left",
					[
						8,
						[
							7,
							[
								7,
								"gradientLayout",
								"$"
							],
							"srcBuf"
						],
						[
							1,
							"idx",
							"-",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							]
						]
					]
				],
				[
					13,
					"down",
					[
						8,
						[
							7,
							[
								7,
								"gradientLayout",
								"$"
							],
							"srcBuf"
						],
						[
							1,
							"idx",
							"+",
							"gs"
						]
					]
				],
				[
					13,
					"up",
					[
						8,
						[
							7,
							[
								7,
								"gradientLayout",
								"$"
							],
							"srcBuf"
						],
						[
							1,
							"idx",
							"-",
							"gs"
						]
					]
				],
				[
					13,
					"dx",
					[
						1,
						[
							1,
							"right",
							"-",
							"left"
						],
						"*",
						[5, "0.5"]
					]
				],
				[
					13,
					"dy",
					[
						1,
						[
							1,
							"down",
							"-",
							"up"
						],
						"*",
						[5, "0.5"]
					]
				],
				[
					6,
					[
						7,
						"std",
						"textureStore"
					],
					[
						[
							7,
							[
								7,
								"gradientLayout",
								"$"
							],
							"dispTex"
						],
						[
							6,
							[
								7,
								"d",
								"vec2u"
							],
							["cx", "cy"]
						],
						[
							6,
							[
								7,
								"d",
								"vec4f"
							],
							[
								"dx",
								"dy",
								[5, "0"],
								[5, "1"]
							]
						]
					]
				]
			]]
		]]],
		"externalNames": [
			"d",
			"GRID_MAX",
			"GRID_SIZE",
			"gradientLayout",
			"std"
		]
	},
	externals: () => ({
		d,
		GRID_MAX,
		GRID_SIZE,
		gradientLayout,
		std
	})
}) && $.f)({})).$name("cursorRipplesGradient");
var RippleUVs = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
	rUV: d.vec2f,
	gUV: d.vec2f,
	bUV: d.vec2f
}), "RippleUVs");
const cursorRipplesUVs = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32
], RippleUVs)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, disp, intensity, chromaticSplit) => {
	"use gpu";
	const scaled = disp.mul(__tsover_mul(intensity, .1));
	const maxDisp = .15;
	const clamped = std.clamp(scaled, d.vec2f(-maxDisp, -maxDisp), d.vec2f(maxDisp, maxDisp));
	const chromaticScale = __tsover_mul(chromaticSplit, .1);
	return RippleUVs({
		rUV: uv.sub(clamped.mul(__tsover_add(1, chromaticScale))),
		gUV: uv.sub(clamped),
		bUV: uv.sub(clamped.mul(__tsover_sub(1, chromaticScale)))
	});
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "disp"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "chromaticSplit"
			}
		],
		"body": [0, [
			[
				13,
				"scaled",
				[
					6,
					[
						7,
						"disp",
						"mul"
					],
					[[
						1,
						"intensity",
						"*",
						[5, "0.1"]
					]]
				]
			],
			[
				13,
				"maxDisp",
				[5, "0.15"]
			],
			[
				13,
				"clamped",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"scaled",
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								4,
								"-",
								"maxDisp"
							], [
								4,
								"-",
								"maxDisp"
							]]
						],
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["maxDisp", "maxDisp"]
						]
					]
				]
			],
			[
				13,
				"chromaticScale",
				[
					1,
					"chromaticSplit",
					"*",
					[5, "0.1"]
				]
			],
			[
				13,
				"rUV",
				[
					6,
					[
						7,
						"uv",
						"sub"
					],
					[[
						6,
						[
							7,
							"clamped",
							"mul"
						],
						[[
							1,
							[5, "1"],
							"+",
							"chromaticScale"
						]]
					]]
				]
			],
			[
				13,
				"gUV",
				[
					6,
					[
						7,
						"uv",
						"sub"
					],
					["clamped"]
				]
			],
			[
				13,
				"bUV",
				[
					6,
					[
						7,
						"uv",
						"sub"
					],
					[[
						6,
						[
							7,
							"clamped",
							"mul"
						],
						[[
							1,
							[5, "1"],
							"-",
							"chromaticScale"
						]]
					]]
				]
			],
			[10, [
				6,
				"RippleUVs",
				[[104, {
					"rUV": "rUV",
					"gUV": "gUV",
					"bUV": "bUV"
				}]]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"RippleUVs"
		]
	},
	externals: () => ({
		std,
		d,
		RippleUVs
	})
}) && $.f)({}));
const componentDefinition = {
	name: "CursorRipples",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Interactive",
	description: "Fluid-like ripple distortion",
	requiresRTT: true,
	requiresChild: true,
	usesPointer: true,
	props: {
		intensity: {
			default: 10,
			description: "Strength of the ripple distortion",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		decay: {
			default: 10,
			description: "How quickly ripples fade (higher = faster)",
			ui: {
				type: "range",
				min: 0,
				max: 20,
				step: .1,
				label: "Decay",
				group: "Effect"
			}
		},
		radius: {
			default: .5,
			description: "Radius of cursor influence",
			ui: {
				type: "range",
				min: .1,
				max: 1,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		chromaticSplit: {
			default: 1,
			description: "RGB channel separation along ripple edges",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 3,
				step: .1,
				label: "Chromatic Split",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	compute: (params) => {
		const { childNode, gpu, registerComputeTexture, getCpuValue, onCleanup, onResize, dimensions } = params;
		if (!childNode) return null;
		const root = gpu?.root;
		if (!root) return null;
		const bufferA = createStateBuffer(root, d.f32, CELL_COUNT);
		const bufferB = createStateBuffer(root, d.f32, CELL_COUNT);
		const dispTex = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
			size: [GRID_SIZE, GRID_SIZE],
			format: STATE_FORMAT
		}).$usage("storage", "sampled"), "dispTex");
		onCleanup(() => dispTex.destroy());
		const displacement = registerComputeTexture(dispTex);
		const propParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(d.struct({
			cursorX: d.f32,
			cursorY: d.f32,
			mouseSpeed: d.f32,
			dt: d.f32,
			decay: d.f32,
			radius: d.f32,
			aspect: d.f32
		})), "propParams");
		let curW = Math.max(1, Math.round(dimensions.width));
		let curH = Math.max(1, Math.round(dimensions.height));
		onResize(({ width, height }) => {
			curW = Math.max(1, Math.round(width));
			curH = Math.max(1, Math.round(height));
		});
		const size = [GRID_SIZE, GRID_SIZE];
		const propBgAB = root.createBindGroup(propagateLayout, {
			readBuf: bufferA,
			writeBuf: bufferB,
			params: propParams.buffer
		});
		const propBgBA = root.createBindGroup(propagateLayout, {
			readBuf: bufferB,
			writeBuf: bufferA,
			params: propParams.buffer
		});
		const propagate = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			propagateKernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"propagateKernel",
					["cx", "cy"]
				]]],
				"externalNames": ["propagateKernel"]
			},
			externals: () => ({ propagateKernel })
		}) && $.f)({}), { size });
		const gradBgB = root.createBindGroup(gradientLayout, {
			srcBuf: bufferB,
			dispTex
		});
		const gradBgA = root.createBindGroup(gradientLayout, {
			srcBuf: bufferA,
			dispTex
		});
		const gradient = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			gradientKernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"gradientKernel",
					["cx", "cy"]
				]]],
				"externalNames": ["gradientKernel"]
			},
			externals: () => ({ gradientKernel })
		}) && $.f)({}), { size });
		let direction = 0;
		let prevX = .5;
		let prevY = .5;
		let lastActiveTime = Date.now();
		return {
			outputs: { displacement },
			getComputeNodes: (frameParams) => {
				const fp = frameParams;
				const currentTime = Date.now();
				const dt = Math.min(fp.deltaTime ?? 0, .016);
				const aspect = Math.max(1, fp.dimensions?.width ?? curW) / Math.max(1, fp.dimensions?.height ?? curH);
				const decay = getCpuValue("decay") ?? 10;
				const radius = (getCpuValue("radius") ?? .5) * .05;
				const px = fp.pointer?.x ?? .5;
				const py = fp.pointer?.y ?? .5;
				const rawVelX = dt > 0 ? (px - prevX) / dt : 0;
				const rawVelY = dt > 0 ? (py - prevY) / dt : 0;
				const mouseSpeed = Math.min(Math.sqrt(rawVelX * rawVelX + rawVelY * rawVelY), 2);
				prevX = px;
				prevY = py;
				if (mouseSpeed > .01) lastActiveTime = currentTime;
				const dampFactor = 1 - decay * .004;
				const settleTime = dampFactor >= 1 ? Infinity : Math.min(3e4, Math.log(1e-6) / Math.log(Math.max(dampFactor, .001)) * 16.67);
				if (currentTime - lastActiveTime > settleTime) return null;
				propParams.write({
					cursorX: px,
					cursorY: py,
					mouseSpeed,
					dt,
					decay,
					radius,
					aspect
				});
				const nodes = direction === 0 ? [propagate.with(propBgAB), gradient.with(gradBgB)] : [propagate.with(propBgBA), gradient.with(gradBgA)];
				direction = 1 - direction;
				return nodes;
			}
		};
	},
	fragment: ({ childNode, ctx, computeOutputs, propValues, convertToTexture, uniforms }) => {
		if (!childNode) return ZERO;
		const childTex = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		const dispTex = computeOutputs?.displacement;
		const disp = dispTex ? dispTex.sample(ctx.uv, "linearClamp").member("xy") : expr("vec2f(0.0, 0.0)");
		const uvs = call(cursorRipplesUVs, "cursorRipplesUVs", [
			ctx.uv,
			disp,
			uniforms.intensity,
			uniforms.chromaticSplit
		]);
		const sampleChild = (uv) => childTex.sample(uv);
		const rSample = applyEdgeHandlingExpr(uvs.member("rUV"), sampleChild, edgeMode);
		const gSample = applyEdgeHandlingExpr(uvs.member("gUV"), sampleChild, edgeMode);
		const bSample = applyEdgeHandlingExpr(uvs.member("bUV"), sampleChild, edgeMode);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [vec4(rSample.member("r"), gSample.member("g"), bSample.member("b"), gSample.member("a"))]);
	}
};
var CursorRipples_default = componentDefinition;
export { propagateKernel as a, gradientKernel as i, componentDefinition as n, cursorRipplesUVs as r, CursorRipples_default as t };
