// shaders/dist/core/CursorTrail-CkNn85vn.js
import { At as call, B as createGuardedCompute, Gt as colorMixing_exports, It as vec4, Ot as ZERO, Pt as floatE, R as std, V as createStateBuffer, gt as mixColorStopsRuntime, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig } from "./porters-Dg80Yg90.js";
var GRID_SIZE = 128;
var CELL_COUNT = GRID_SIZE * GRID_SIZE;
var STATE_FORMAT = "rgba16float";
var cm = colorMixing_exports;
var decayLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	trailBuf: {
		storage: d.arrayOf(d.vec4f, CELL_COUNT),
		access: "mutable"
	},
	params: { uniform: /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
		fadeRate: d.f32,
		ageRate: d.f32,
		shrink: d.f32
	}), "uniform") }
}), "decayLayout");
var stampLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	trailBuf: {
		storage: d.arrayOf(d.vec4f, CELL_COUNT),
		access: "mutable"
	},
	params: { uniform: /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
		cursorX: d.f32,
		cursorY: d.f32,
		radius: d.f32,
		amount: d.f32,
		aspect: d.f32
	}), "uniform") }
}), "stampLayout");
var outputLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	trailBuf: {
		storage: d.arrayOf(d.vec4f, CELL_COUNT),
		access: "readonly"
	},
	trailTex: { storageTexture: d.textureStorage2d(STATE_FORMAT, "write-only") }
}), "outputLayout");
const decayKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = decayLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(GRID_SIZE)), cx);
	const cell = decayLayout.$.trailBuf[idx];
	const intensity = __tsover_mul(cell.x, p.fadeRate);
	const size = __tsover_sub(1, __tsover_mul(__tsover_sub(1, intensity), p.shrink));
	const age = std.min(__tsover_add(cell.z, p.ageRate), 1);
	decayLayout.$.trailBuf[idx] = d.vec4f(intensity, size, age, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"decayLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["GRID_SIZE"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"cell",
				[
					8,
					[
						7,
						[
							7,
							"decayLayout",
							"$"
						],
						"trailBuf"
					],
					"idx"
				]
			],
			[
				13,
				"intensity",
				[
					1,
					[
						7,
						"cell",
						"x"
					],
					"*",
					[
						7,
						"p",
						"fadeRate"
					]
				]
			],
			[
				13,
				"size",
				[
					1,
					[5, "1"],
					"-",
					[
						1,
						[
							1,
							[5, "1"],
							"-",
							"intensity"
						],
						"*",
						[
							7,
							"p",
							"shrink"
						]
					]
				]
			],
			[
				13,
				"age",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					[[
						1,
						[
							7,
							"cell",
							"z"
						],
						"+",
						[
							7,
							"p",
							"ageRate"
						]
					], [5, "1"]]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"decayLayout",
							"$"
						],
						"trailBuf"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						"intensity",
						"size",
						"age",
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"decayLayout",
			"d",
			"GRID_SIZE",
			"std"
		]
	},
	externals: () => ({
		decayLayout,
		d,
		GRID_SIZE,
		std
	})
}) && $.f)({})).$name("cursorTrailDecay");
const stampKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = stampLayout.$.params;
	const cellX = __tsover_div(__tsover_add(d.f32(cx), .5), d.f32(GRID_SIZE));
	const cellY = __tsover_div(__tsover_add(d.f32(cy), .5), d.f32(GRID_SIZE));
	const ddx = std.select(__tsover_sub(cellX, p.cursorX), __tsover_mul(__tsover_sub(cellX, p.cursorX), p.aspect), p.aspect >= 1);
	const ddy = std.select(__tsover_div(__tsover_sub(cellY, p.cursorY), p.aspect), __tsover_sub(cellY, p.cursorY), p.aspect >= 1);
	const distSq = __tsover_add(__tsover_mul(ddx, ddx), __tsover_mul(ddy, ddy));
	const influenceRadius = __tsover_mul(p.radius, 3);
	if (distSq < __tsover_mul(influenceRadius, influenceRadius)) {
		const idx = __tsover_add(__tsover_mul(cy, d.u32(GRID_SIZE)), cx);
		const influence = std.exp(__tsover_mul(__tsover_div(distSq, __tsover_mul(p.radius, p.radius)), -1));
		const cell = stampLayout.$.trailBuf[idx];
		const newIntensity = std.min(__tsover_add(cell.x, __tsover_mul(influence, p.amount)), 1);
		stampLayout.$.trailBuf[idx] = d.vec4f(newIntensity, 1, 0, 0);
	}
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"stampLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"cellX",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["cx"]
						],
						"+",
						[5, "0.5"]
					],
					"/",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["GRID_SIZE"]
					]
				]
			],
			[
				13,
				"cellY",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["cy"]
						],
						"+",
						[5, "0.5"]
					],
					"/",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["GRID_SIZE"]
					]
				]
			],
			[
				13,
				"ddx",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							1,
							"cellX",
							"-",
							[
								7,
								"p",
								"cursorX"
							]
						],
						[
							1,
							[
								1,
								"cellX",
								"-",
								[
									7,
									"p",
									"cursorX"
								]
							],
							"*",
							[
								7,
								"p",
								"aspect"
							]
						],
						[
							1,
							[
								7,
								"p",
								"aspect"
							],
							">=",
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"ddy",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							1,
							[
								1,
								"cellY",
								"-",
								[
									7,
									"p",
									"cursorY"
								]
							],
							"/",
							[
								7,
								"p",
								"aspect"
							]
						],
						[
							1,
							"cellY",
							"-",
							[
								7,
								"p",
								"cursorY"
							]
						],
						[
							1,
							[
								7,
								"p",
								"aspect"
							],
							">=",
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"distSq",
				[
					1,
					[
						1,
						"ddx",
						"*",
						"ddx"
					],
					"+",
					[
						1,
						"ddy",
						"*",
						"ddy"
					]
				]
			],
			[
				13,
				"influenceRadius",
				[
					1,
					[
						7,
						"p",
						"radius"
					],
					"*",
					[5, "3"]
				]
			],
			[
				11,
				[
					1,
					"distSq",
					"<",
					[
						1,
						"influenceRadius",
						"*",
						"influenceRadius"
					]
				],
				[0, [
					[
						13,
						"idx",
						[
							1,
							[
								1,
								"cy",
								"*",
								[
									6,
									[
										7,
										"d",
										"u32"
									],
									["GRID_SIZE"]
								]
							],
							"+",
							"cx"
						]
					],
					[
						13,
						"influence",
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									1,
									"distSq",
									"/",
									[
										1,
										[
											7,
											"p",
											"radius"
										],
										"*",
										[
											7,
											"p",
											"radius"
										]
									]
								],
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]]
						]
					],
					[
						13,
						"cell",
						[
							8,
							[
								7,
								[
									7,
									"stampLayout",
									"$"
								],
								"trailBuf"
							],
							"idx"
						]
					],
					[
						13,
						"newIntensity",
						[
							6,
							[
								7,
								"std",
								"min"
							],
							[[
								1,
								[
									7,
									"cell",
									"x"
								],
								"+",
								[
									1,
									"influence",
									"*",
									[
										7,
										"p",
										"amount"
									]
								]
							], [5, "1"]]
						]
					],
					[
						2,
						[
							8,
							[
								7,
								[
									7,
									"stampLayout",
									"$"
								],
								"trailBuf"
							],
							"idx"
						],
						"=",
						[
							6,
							[
								7,
								"d",
								"vec4f"
							],
							[
								"newIntensity",
								[5, "1"],
								[5, "0"],
								[5, "0"]
							]
						]
					]
				]]
			]
		]],
		"externalNames": [
			"stampLayout",
			"d",
			"GRID_SIZE",
			"std"
		]
	},
	externals: () => ({
		stampLayout,
		d,
		GRID_SIZE,
		std
	})
}) && $.f)({})).$name("cursorTrailStamp");
const outputKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(GRID_SIZE)), cx);
	const cell = outputLayout.$.trailBuf[idx];
	std.textureStore(outputLayout.$.trailTex, d.vec2u(cx, cy), cell);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["GRID_SIZE"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"cell",
				[
					8,
					[
						7,
						[
							7,
							"outputLayout",
							"$"
						],
						"trailBuf"
					],
					"idx"
				]
			],
			[
				6,
				[
					7,
					"std",
					"textureStore"
				],
				[
					[
						7,
						[
							7,
							"outputLayout",
							"$"
						],
						"trailTex"
					],
					[
						6,
						[
							7,
							"d",
							"vec2u"
						],
						["cx", "cy"]
					],
					"cell"
				]
			]
		]],
		"externalNames": [
			"d",
			"GRID_SIZE",
			"outputLayout",
			"std"
		]
	},
	externals: () => ({
		d,
		GRID_SIZE,
		outputLayout,
		std
	})
}) && $.f)({})).$name("cursorTrailOutput");
const cursorTrailTapUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, ox, oy) => {
	"use gpu";
	return d.vec2f(__tsover_add(uv.x, ox), __tsover_add(uv.y, oy));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "ox"
			},
			{
				"type": "i",
				"name": "oy"
			}
		],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					7,
					"uv",
					"x"
				],
				"+",
				"ox"
			], [
				1,
				[
					7,
					"uv",
					"y"
				],
				"+",
				"oy"
			]]
		]]]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
var TrailField = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
	age: d.f32,
	mask: d.f32
}), "TrailField");
const cursorTrailField = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f
], TrailField)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((c, s1, s2, s3, s4) => {
	"use gpu";
	const intensity = __tsover_add(__tsover_add(__tsover_add(__tsover_add(__tsover_mul(c.x, .5), __tsover_mul(s1.x, .125)), __tsover_mul(s2.x, .125)), __tsover_mul(s3.x, .125)), __tsover_mul(s4.x, .125));
	return TrailField({
		age: __tsover_add(__tsover_add(__tsover_add(__tsover_add(__tsover_mul(c.z, .5), __tsover_mul(s1.z, .125)), __tsover_mul(s2.z, .125)), __tsover_mul(s3.z, .125)), __tsover_mul(s4.z, .125)),
		mask: std.step(.01, intensity)
	});
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "c"
			},
			{
				"type": "i",
				"name": "s1"
			},
			{
				"type": "i",
				"name": "s2"
			},
			{
				"type": "i",
				"name": "s3"
			},
			{
				"type": "i",
				"name": "s4"
			}
		],
		"body": [0, [
			[
				13,
				"intensity",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									[
										7,
										"c",
										"x"
									],
									"*",
									[5, "0.5"]
								],
								"+",
								[
									1,
									[
										7,
										"s1",
										"x"
									],
									"*",
									[5, "0.125"]
								]
							],
							"+",
							[
								1,
								[
									7,
									"s2",
									"x"
								],
								"*",
								[5, "0.125"]
							]
						],
						"+",
						[
							1,
							[
								7,
								"s3",
								"x"
							],
							"*",
							[5, "0.125"]
						]
					],
					"+",
					[
						1,
						[
							7,
							"s4",
							"x"
						],
						"*",
						[5, "0.125"]
					]
				]
			],
			[
				13,
				"age",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									[
										7,
										"c",
										"z"
									],
									"*",
									[5, "0.5"]
								],
								"+",
								[
									1,
									[
										7,
										"s1",
										"z"
									],
									"*",
									[5, "0.125"]
								]
							],
							"+",
							[
								1,
								[
									7,
									"s2",
									"z"
								],
								"*",
								[5, "0.125"]
							]
						],
						"+",
						[
							1,
							[
								7,
								"s3",
								"z"
							],
							"*",
							[5, "0.125"]
						]
					],
					"+",
					[
						1,
						[
							7,
							"s4",
							"z"
						],
						"*",
						[5, "0.125"]
					]
				]
			],
			[
				13,
				"mask",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					[[5, "0.01"], "intensity"]
				]
			],
			[10, [
				6,
				"TrailField",
				[[104, {
					"age": "age",
					"mask": "mask"
				}]]
			]]
		]],
		"externalNames": ["std", "TrailField"]
	},
	externals: () => ({
		std,
		TrailField
	})
}) && $.f)({}));
const componentDefinition = {
	name: "CursorTrail",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Interactive",
	description: "Animated trail effect that tracks cursor movement",
	usesPointer: true,
	props: {
		colorA: {
			default: "#00aaff",
			transform: transformColor,
			description: "Color of fresh trails",
			ui: {
				type: "color",
				label: "Start Color",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ff00aa",
			transform: transformColor,
			description: "Color trails transition to as they fade",
			ui: {
				type: "color",
				label: "End Color",
				group: "Colors"
			}
		},
		stops: colorStopsPropConfig(),
		radius: {
			default: .5,
			description: "Base radius of trail circles",
			ui: {
				type: "range",
				min: .5,
				max: 2,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		length: {
			default: .5,
			description: "How long trail circles persist (in seconds)",
			ui: {
				type: "range",
				min: .1,
				max: 2,
				step: .1,
				label: "Trail Length",
				group: "Animation"
			}
		},
		shrink: {
			default: 1,
			description: "How much circles shrink as they fade out (0 = no shrink, 1 = full shrink)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .1,
				label: "Shrink Amount",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	compute: (params) => {
		const { gpu, registerComputeTexture, getCpuValue, onCleanup, onResize, dimensions } = params;
		const root = gpu?.root;
		if (!root) return null;
		const trailBuf = createStateBuffer(root, d.vec4f, CELL_COUNT);
		const trailTex = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
			size: [GRID_SIZE, GRID_SIZE],
			format: STATE_FORMAT
		}).$usage("storage", "sampled"), "trailTex");
		onCleanup(() => trailTex.destroy());
		const trailTexture = registerComputeTexture(trailTex);
		const decayParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(d.struct({
			fadeRate: d.f32,
			ageRate: d.f32,
			shrink: d.f32
		})), "decayParams");
		const stampParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(d.struct({
			cursorX: d.f32,
			cursorY: d.f32,
			radius: d.f32,
			amount: d.f32,
			aspect: d.f32
		})), "stampParams");
		let curW = Math.max(1, Math.round(dimensions.width));
		let curH = Math.max(1, Math.round(dimensions.height));
		onResize(({ width, height }) => {
			curW = Math.max(1, Math.round(width));
			curH = Math.max(1, Math.round(height));
		});
		const size = [GRID_SIZE, GRID_SIZE];
		const decayBg = root.createBindGroup(decayLayout, {
			trailBuf,
			params: decayParams.buffer
		});
		const stampBg = root.createBindGroup(stampLayout, {
			trailBuf,
			params: stampParams.buffer
		});
		const outputBg = root.createBindGroup(outputLayout, {
			trailBuf,
			trailTex
		});
		const decay = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			decayKernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"decayKernel",
					["cx", "cy"]
				]]],
				"externalNames": ["decayKernel"]
			},
			externals: () => ({ decayKernel })
		}) && $.f)({}), {
			size,
			bindGroup: decayBg
		});
		const stamp = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			stampKernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"stampKernel",
					["cx", "cy"]
				]]],
				"externalNames": ["stampKernel"]
			},
			externals: () => ({ stampKernel })
		}) && $.f)({}), {
			size,
			bindGroup: stampBg
		});
		const output = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			outputKernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"outputKernel",
					["cx", "cy"]
				]]],
				"externalNames": ["outputKernel"]
			},
			externals: () => ({ outputKernel })
		}) && $.f)({}), {
			size,
			bindGroup: outputBg
		});
		let prevX = .5;
		let prevY = .5;
		let lastActiveTime = Date.now();
		return {
			outputs: { trailTexture },
			getComputeNodes: (frameParams) => {
				const fp = frameParams;
				const currentTime = Date.now();
				const dt = Math.min(fp.deltaTime ?? 0, .016);
				const aspect = Math.max(1, fp.dimensions?.width ?? curW) / Math.max(1, fp.dimensions?.height ?? curH);
				const radius = (getCpuValue("radius") ?? .5) * .1;
				const length = getCpuValue("length") ?? .5;
				const shrink = getCpuValue("shrink") ?? 1;
				const px = fp.pointer?.x ?? .5;
				const py = fp.pointer?.y ?? .5;
				const dx = px - prevX;
				const dy = py - prevY;
				const speed = Math.sqrt(dx * dx + dy * dy);
				const oldX = prevX;
				const oldY = prevY;
				prevX = px;
				prevY = py;
				if (speed > .001) lastActiveTime = currentTime;
				const settleMs = Math.min(3e4, Math.max(3e3, length * 5e3));
				if (currentTime - lastActiveTime > settleMs) return null;
				decayParams.write({
					fadeRate: 1 - dt / Math.max(.1, length),
					ageRate: dt / Math.max(.1, length),
					shrink
				});
				const nodes = [decay];
				if (speed > .001) {
					const stepSize = Math.max(.005, radius * .5);
					const numSteps = Math.min(20, Math.max(1, Math.ceil(speed / stepSize)));
					const amountPerStep = speed / numSteps * 50 * dt;
					for (let s = 0; s < numSteps; s++) {
						const t = (s + .5) / numSteps;
						const cursorX = oldX + dx * t;
						const cursorY = oldY + dy * t;
						nodes.push(() => stampParams.write({
							cursorX,
							cursorY,
							radius,
							amount: amountPerStep,
							aspect
						}));
						nodes.push(stamp);
					}
				}
				nodes.push(output);
				return nodes;
			}
		};
	},
	fragment: ({ ctx, computeOutputs, propValues, uniforms }) => {
		const trailTex = computeOutputs?.trailTexture;
		if (!trailTex) return ZERO;
		const px = 1 / GRID_SIZE;
		const tap = (ox, oy) => trailTex.sample(call(cursorTrailTapUV, "cursorTrailTapUV", [
			ctx.uv,
			floatE(ox),
			floatE(oy)
		]), "linearClamp");
		const field = call(cursorTrailField, "cursorTrailField", [
			trailTex.sample(ctx.uv, "linearClamp"),
			tap(px, 0),
			tap(0, px),
			tap(-px, 0),
			tap(0, -px)
		]);
		const trailAge = field.member("age");
		const circleMask = field.member("mask");
		const colorSpaceMode = propValues.colorSpace ?? 0;
		const stopCount = propValues.stopCount ?? 0;
		let trailColor;
		if (stopCount > 1) trailColor = mixColorStopsRuntime(trailAge, {
			colorsArray: uniforms.colorsArray,
			positionsArray: uniforms.positionsArray,
			convertedColorsArray: uniforms.convertedColorsArray,
			stopCount: uniforms.stopCount
		}, colorSpaceMode);
		else trailColor = call(cm.mixColorsVariants[colorSpaceMode] ?? cm.mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			trailAge
		]);
		return vec4(trailColor.member("rgb"), trailColor.member("a").mul(circleMask));
	}
};
var CursorTrail_default = componentDefinition;
export { decayKernel as a, cursorTrailTapUV as i, componentDefinition as n, outputKernel as o, cursorTrailField as r, stampKernel as s, CursorTrail_default as t };
