// shaders/dist/core/DiamondGradient-HmZwnTdD.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, gt as mixColorStopsRuntime, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig } from "./porters-Dg80Yg90.js";
var DEG_TO_RAD = .017453292519943295;
const diamondFinalT = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.vec2f
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, size, rotation, repeat, roundness, uv, viewport) => {
	"use gpu";
	const safeViewportY = std.max(viewport.y, 1e-6);
	const aspect = __tsover_div(viewport.x, safeViewportY);
	const aspectUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y));
	const dx = __tsover_sub(aspectUV.x, centerPos.x);
	const dy = __tsover_sub(aspectUV.y, centerPos.y);
	const rotRad = __tsover_mul(rotation, DEG_TO_RAD);
	const cosR = std.cos(rotRad);
	const sinR = std.sin(rotRad);
	const rdx = __tsover_add(__tsover_mul(dx, cosR), __tsover_mul(dy, sinR));
	const rdy = __tsover_sub(__tsover_mul(dy, cosR), __tsover_mul(dx, sinR));
	const absDx = std.abs(rdx);
	const absDy = std.abs(rdy);
	const l1 = __tsover_add(absDx, absDy);
	const linf = std.max(absDx, absDy);
	const dist = std.mix(l1, linf, roundness);
	const t = __tsover_mul(__tsover_div(dist, size), repeat);
	return std.select(std.clamp(t, 0, 1), std.fract(t), repeat > 1.0001);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "size"
			},
			{
				"type": "i",
				"name": "rotation"
			},
			{
				"type": "i",
				"name": "repeat"
			},
			{
				"type": "i",
				"name": "roundness"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			}
		],
		"body": [0, [
			[
				13,
				"safeViewportY",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						7,
						"viewport",
						"y"
					], [5, "0.000001"]]
				]
			],
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					"safeViewportY"
				]
			],
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						7,
						"aspectUV",
						"x"
					],
					"-",
					[
						7,
						"centerPos",
						"x"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"aspectUV",
						"y"
					],
					"-",
					[
						7,
						"centerPos",
						"y"
					]
				]
			],
			[
				13,
				"rotRad",
				[
					1,
					"rotation",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["rotRad"]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["rotRad"]
				]
			],
			[
				13,
				"rdx",
				[
					1,
					[
						1,
						"dx",
						"*",
						"cosR"
					],
					"+",
					[
						1,
						"dy",
						"*",
						"sinR"
					]
				]
			],
			[
				13,
				"rdy",
				[
					1,
					[
						1,
						"dy",
						"*",
						"cosR"
					],
					"-",
					[
						1,
						"dx",
						"*",
						"sinR"
					]
				]
			],
			[
				13,
				"absDx",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					["rdx"]
				]
			],
			[
				13,
				"absDy",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					["rdy"]
				]
			],
			[
				13,
				"l1",
				[
					1,
					"absDx",
					"+",
					"absDy"
				]
			],
			[
				13,
				"linf",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["absDx", "absDy"]
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"l1",
						"linf",
						"roundness"
					]
				]
			],
			[
				13,
				"t",
				[
					1,
					[
						1,
						"dist",
						"/",
						"size"
					],
					"*",
					"repeat"
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"select"
				],
				[
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							"t",
							[5, "0"],
							[5, "1"]
						]
					],
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						["t"]
					],
					[
						1,
						"repeat",
						">",
						[5, "1.0001"]
					]
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"DEG_TO_RAD"
		]
	},
	externals: () => ({
		std,
		d,
		DEG_TO_RAD
	})
}) && $.f)({}));
function gradientColor(params, finalT) {
	const { uniforms, propValues } = params;
	const colorSpaceMode = propValues.colorSpace ?? 0;
	if ((propValues.stopCount ?? 0) > 1) return mixColorStopsRuntime(finalT, {
		colorsArray: uniforms.colorsArray,
		positionsArray: uniforms.positionsArray,
		convertedColorsArray: uniforms.convertedColorsArray,
		stopCount: uniforms.stopCount
	}, colorSpaceMode);
	return call(mixColorsVariants[colorSpaceMode] ?? mixColorsLinear, "mixColors", [
		uniforms.colorA,
		uniforms.colorB,
		finalT
	]);
}
const componentDefinition = {
	name: "DiamondGradient",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Diamond-shaped gradient radiating from a center point using Manhattan distance",
	acceptsUVContext: true,
	props: {
		colorA: {
			default: "#4ffb4a",
			transform: transformColor,
			description: "Color at the center of the diamond",
			ui: {
				type: "color",
				label: "Color A (Center)",
				group: "Colors"
			}
		},
		colorB: {
			default: "#4f1238",
			transform: transformColor,
			description: "Color at the outer edges of the diamond",
			ui: {
				type: "color",
				label: "Color B (Edge)",
				group: "Colors"
			}
		},
		stops: colorStopsPropConfig(),
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the diamond",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		size: {
			default: .7,
			description: "Extent of the gradient — controls how far Color A reaches before transitioning to Color B",
			ui: {
				type: ["range", "map"],
				min: .01,
				max: 2,
				step: .01,
				label: "Size",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation in degrees — tilts the diamond into a rhombus",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		repeat: {
			default: 1,
			description: "Number of times the gradient repeats outward. Values above 1 create concentric diamond or square bands.",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 16,
				step: .5,
				label: "Repeat",
				group: "Effect"
			}
		},
		roundness: {
			default: 0,
			description: "Morphs from a sharp diamond (0) to a square (1)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Roundness",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		return gradientColor(params, call(diamondFinalT, "diamondFinalT", [
			params.uniforms.center,
			params.uniforms.size,
			params.uniforms.rotation,
			params.uniforms.repeat,
			params.uniforms.roundness,
			uv,
			viewport
		]));
	}
};
var DiamondGradient_default = componentDefinition;
export { componentDefinition as n, diamondFinalT as r, DiamondGradient_default as t };
