// shaders/dist/core/DiffuseBlur-uXE0w0DD.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { s as transformEdges } from "./transformations-fqzznt4g.js";
const diffuseBlurUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.vec2f
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, intensity, viewport) => {
	"use gpu";
	const seed = uv.mul(1e3);
	const sinVal = std.sin(std.dot(seed, d.vec2f(25.9796, 156.466)));
	const offset = std.fract(d.vec2f(__tsover_mul(sinVal, 43758.5453), __tsover_mul(sinVal, 28461.7231))).mul(2).sub(d.vec2f(1, 1)).mul(intensity).div(viewport);
	return uv.add(offset);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "viewport"
			}
		],
		"body": [0, [
			[
				13,
				"seed",
				[
					6,
					[
						7,
						"uv",
						"mul"
					],
					[[5, "1000"]]
				]
			],
			[
				13,
				"sinVal",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["seed", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "25.9796"], [5, "156.466"]]
						]]
					]]
				]
			],
			[
				13,
				"randVec",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							"sinVal",
							"*",
							[5, "43758.5453"]
						], [
							1,
							"sinVal",
							"*",
							[5, "28461.7231"]
						]]
					]]
				]
			],
			[
				13,
				"offset",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										[
											6,
											[
												7,
												"randVec",
												"mul"
											],
											[[5, "2"]]
										],
										"sub"
									],
									[[
										6,
										[
											7,
											"d",
											"vec2f"
										],
										[[5, "1"], [5, "1"]]
									]]
								],
								"mul"
							],
							["intensity"]
						],
						"div"
					],
					["viewport"]
				]
			],
			[10, [
				6,
				[
					7,
					"uv",
					"add"
				],
				["offset"]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "DiffuseBlur",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Blurs",
	description: "Grain-like pixel displacement at random",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 30,
			description: "Intensity of the diffuse blur effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, propValues, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(diffuseBlurUV, "diffuseBlurUV", [
			ctx.uv,
			uniforms.intensity,
			ctx.viewportSize
		]), (uv) => texture.sample(uv), edgeMode)]);
	}
};
var DiffuseBlur_default = componentDefinition;
export { componentDefinition as n, diffuseBlurUV as r, DiffuseBlur_default as t };
