// shaders/dist/core/DOMTexture-Zxx_WTGZ.js
import { At as call, Ht as srgbToLinear, It as vec4 } from "./coords-BlwgBpJf.js";
const componentDefinition = {
	name: "DOMTexture",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Textures",
	description: "Render live HTML/DOM content as a WebGPU texture layer via the html-in-canvas API. Requires Chrome Canary with chrome://flags/#canvas-draw-element enabled.",
	capturesDOM: true,
	experimental: {
		message: "Requires Chrome Canary with chrome://flags/#canvas-draw-element enabled. Do not use for production use.",
		docs: "This component is powered by the WICG html-in-canvas proposal, which is currently only available in Chrome Canary behind a feature flag. It is not suitable for production use and may change as the specification evolves.",
		docsLink: {
			url: "https://github.com/WICG/html-in-canvas",
			label: "View the spec"
		}
	},
	props: {},
	fragment: ({ ctx, gpu, domCanvas, onCleanup, onBeforeRender, onResize, createMediaTexture, registerMediaTexture }) => {
		const device = gpu.device;
		const initialDpr = Math.min(window.devicePixelRatio, 2);
		let current = createMediaTexture({
			width: Math.max(1, Math.round(window.innerWidth * initialDpr)),
			height: Math.max(1, Math.round(window.innerHeight * initialDpr)),
			label: "DOMTexture"
		});
		let isDisposed = false;
		let contentElement = null;
		let hasLoggedError = false;
		let paintSetup = false;
		let hasFreshSnapshot = false;
		const ensureTexSize = (w, h) => {
			const nw = Math.max(1, w);
			const nh = Math.max(1, h);
			if (isDisposed || current.width === nw && current.height === nh) return;
			const next = createMediaTexture({
				width: nw,
				height: nh,
				label: "DOMTexture"
			});
			const old = current;
			current = next;
			old.destroy();
		};
		onBeforeRender(() => {
			if (!domCanvas || isDisposed) return;
			if (!contentElement) {
				contentElement = domCanvas.firstElementChild;
				if (contentElement) {
					const rect = contentElement.getBoundingClientRect();
					const physW = Math.round(rect.width * window.devicePixelRatio);
					const physH = Math.round(rect.height * window.devicePixelRatio);
					if (physW > 0 && physH > 0) ensureTexSize(physW, physH);
				}
			}
			if (!contentElement) return;
			if (!paintSetup) {
				domCanvas.getContext("2d");
				domCanvas.onpaint = () => {
					hasFreshSnapshot = true;
				};
				domCanvas.requestPaint?.();
				paintSetup = true;
				return;
			}
			if (hasFreshSnapshot && contentElement) {
				const copyFn = device.queue.copyElementImageToTexture;
				if (typeof copyFn !== "function") {
					hasFreshSnapshot = false;
					return;
				}
				try {
					copyFn.call(device.queue, contentElement, { texture: current.unwrap() });
					hasFreshSnapshot = false;
				} catch (e) {
					if (!hasLoggedError) {
						console.warn("[DOMTexture] copyElementImageToTexture failed.\nRequires Chrome Canary with chrome://flags/#canvas-draw-element enabled.\nThe element must be a direct child of a <canvas layoutsubtree>.", e);
						hasLoggedError = true;
					}
				}
			}
		});
		onResize(() => {
			const newDpr = Math.min(window.devicePixelRatio, 2);
			ensureTexSize(Math.round(window.innerWidth * newDpr), Math.round(window.innerHeight * newDpr));
			contentElement = null;
			hasFreshSnapshot = false;
			if (domCanvas) domCanvas.requestPaint?.();
		});
		onCleanup(() => {
			isDisposed = true;
			if (domCanvas) domCanvas.onpaint = null;
			current.destroy();
		});
		const sampled = registerMediaTexture(() => current.texture).sample(ctx.uv);
		return vec4(call(srgbToLinear, "srgbToLinear", [sampled.member("rgb")]), sampled.member("a"));
	}
};
var DOMTexture_default = componentDefinition;
export { componentDefinition as n, DOMTexture_default as t };
