// shaders/dist/core/DotGrid-ioUj-Lim.js
import { At as call, It as vec4, R as std, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
const dotGridCellCenterUV = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, density) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const correctedX = __tsover_mul(uv.x, aspect);
	const correctedY = __tsover_sub(1, uv.y);
	const cellCenterCorrectedX = __tsover_div(__tsover_add(std.floor(__tsover_mul(correctedX, density)), .5), density);
	const cellCenterCorrectedY = __tsover_div(__tsover_add(std.floor(__tsover_mul(correctedY, density)), .5), density);
	return d.vec2f(__tsover_div(cellCenterCorrectedX, aspect), __tsover_sub(1, cellCenterCorrectedY));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "density"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"correctedX",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"correctedY",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"uv",
						"y"
					]
				]
			],
			[
				13,
				"cellCenterCorrectedX",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								1,
								"correctedX",
								"*",
								"density"
							]]
						],
						"+",
						[5, "0.5"]
					],
					"/",
					"density"
				]
			],
			[
				13,
				"cellCenterCorrectedY",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								1,
								"correctedY",
								"*",
								"density"
							]]
						],
						"+",
						[5, "0.5"]
					],
					"/",
					"density"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					"cellCenterCorrectedX",
					"/",
					"aspect"
				], [
					1,
					[5, "1"],
					"-",
					"cellCenterCorrectedY"
				]]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const dotGridAlpha = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, density, dotSize, offset, speedVariance, twinkle, animTime, time) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const correctedUV = d.vec2f(__tsover_mul(uv.x, aspect), __tsover_sub(1, uv.y));
	const gridUV = correctedUV.mul(density);
	const rowIndex = std.floor(gridUV.y);
	const rowHash = std.fract(__tsover_mul(std.sin(__tsover_mul(rowIndex, 127.1)), 43758.5453));
	const speedMod = __tsover_add(1, __tsover_mul(__tsover_mul(__tsover_sub(rowHash, .5), 4), speedVariance));
	const drift = __tsover_mul(animTime, speedMod);
	const staticStagger = __tsover_mul(offset, __tsover_sub(rowIndex, __tsover_mul(2, std.floor(__tsover_div(rowIndex, 2)))));
	const shiftedGrid = d.vec2f(__tsover_sub(__tsover_sub(gridUV.x, staticStagger), drift), gridUV.y);
	const cellUV = shiftedGrid.sub(std.floor(shiftedGrid));
	const centerDistance = std.length(cellUV.sub(.5));
	const pixelSize = std.length(std.fwidth(correctedUV.mul(density)));
	const coverage = __tsover_sub(1, std.smoothstep(__tsover_sub(__tsover_mul(dotSize, .5), __tsover_mul(pixelSize, .5)), __tsover_mul(dotSize, .5), centerDistance));
	const gridPosition = std.floor(shiftedGrid);
	const dotPhase = std.fract(__tsover_mul(std.sin(__tsover_add(__tsover_mul(gridPosition.x, 12.9898), __tsover_mul(gridPosition.y, 78.233))), 43758.5453));
	const twinkleTime = __tsover_add(__tsover_mul(time, 2), __tsover_mul(dotPhase, 6.28318));
	const twinkleValue = __tsover_add(__tsover_mul(std.sin(twinkleTime), .5), .5);
	const twinkleModifier = std.mix(1, twinkleValue, twinkle);
	return __tsover_mul(coverage, twinkleModifier);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "density"
			},
			{
				"type": "i",
				"name": "dotSize"
			},
			{
				"type": "i",
				"name": "offset"
			},
			{
				"type": "i",
				"name": "speedVariance"
			},
			{
				"type": "i",
				"name": "twinkle"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "time"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"correctedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"uv",
							"y"
						]
					]]
				]
			],
			[
				13,
				"gridUV",
				[
					6,
					[
						7,
						"correctedUV",
						"mul"
					],
					["density"]
				]
			],
			[
				13,
				"rowIndex",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						7,
						"gridUV",
						"y"
					]]
				]
			],
			[
				13,
				"rowHash",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								"rowIndex",
								"*",
								[5, "127.1"]
							]]
						],
						"*",
						[5, "43758.5453"]
					]]
				]
			],
			[
				13,
				"speedMod",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						[
							1,
							[
								1,
								"rowHash",
								"-",
								[5, "0.5"]
							],
							"*",
							[5, "4"]
						],
						"*",
						"speedVariance"
					]
				]
			],
			[
				13,
				"drift",
				[
					1,
					"animTime",
					"*",
					"speedMod"
				]
			],
			[
				13,
				"staticStagger",
				[
					1,
					"offset",
					"*",
					[
						1,
						"rowIndex",
						"-",
						[
							1,
							[5, "2"],
							"*",
							[
								6,
								[
									7,
									"std",
									"floor"
								],
								[[
									1,
									"rowIndex",
									"/",
									[5, "2"]
								]]
							]
						]
					]
				]
			],
			[
				13,
				"shiftedGrid",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								7,
								"gridUV",
								"x"
							],
							"-",
							"staticStagger"
						],
						"-",
						"drift"
					], [
						7,
						"gridUV",
						"y"
					]]
				]
			],
			[
				13,
				"cellUV",
				[
					6,
					[
						7,
						"shiftedGrid",
						"sub"
					],
					[[
						6,
						[
							7,
							"std",
							"floor"
						],
						["shiftedGrid"]
					]]
				]
			],
			[
				13,
				"centerDistance",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							"cellUV",
							"sub"
						],
						[[5, "0.5"]]
					]]
				]
			],
			[
				13,
				"pixelSize",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							"std",
							"fwidth"
						],
						[[
							6,
							[
								7,
								"correctedUV",
								"mul"
							],
							["density"]
						]]
					]]
				]
			],
			[
				13,
				"coverage",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								[
									1,
									"dotSize",
									"*",
									[5, "0.5"]
								],
								"-",
								[
									1,
									"pixelSize",
									"*",
									[5, "0.5"]
								]
							],
							[
								1,
								"dotSize",
								"*",
								[5, "0.5"]
							],
							"centerDistance"
						]
					]
				]
			],
			[
				13,
				"gridPosition",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["shiftedGrid"]
				]
			],
			[
				13,
				"dotPhase",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									[
										7,
										"gridPosition",
										"x"
									],
									"*",
									[5, "12.9898"]
								],
								"+",
								[
									1,
									[
										7,
										"gridPosition",
										"y"
									],
									"*",
									[5, "78.233"]
								]
							]]
						],
						"*",
						[5, "43758.5453"]
					]]
				]
			],
			[
				13,
				"twinkleTime",
				[
					1,
					[
						1,
						"time",
						"*",
						[5, "2"]
					],
					"+",
					[
						1,
						"dotPhase",
						"*",
						[5, "6.28318"]
					]
				]
			],
			[
				13,
				"twinkleValue",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							["twinkleTime"]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"twinkleModifier",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "1"],
						"twinkleValue",
						"twinkle"
					]
				]
			],
			[10, [
				1,
				"coverage",
				"*",
				"twinkleModifier"
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "DotGrid",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Grid of dots with optional twinkling animation",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		color: {
			default: "#ffffff",
			description: "The color of the dot",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		density: {
			default: 30,
			description: "The number of dots on the longest canvas edge",
			ui: {
				type: "range",
				min: 1,
				max: 200,
				step: 1,
				label: "Density",
				group: "Effect"
			}
		},
		dotSize: {
			default: .3,
			description: "The size of each dot, zero (0) being invisible, one (1) filled the grid with no gaps",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Dot Size",
				group: "Effect"
			}
		},
		offset: {
			default: 0,
			description: "Horizontal stagger of alternating rows (0.5 = classic polka-dot brick offset)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Row Offset",
				group: "Effect"
			}
		},
		speed: {
			default: 0,
			description: "Animates the rows drifting horizontally (0 = static)",
			ui: {
				type: "range",
				min: -3,
				max: 3,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		speedVariance: {
			default: .3,
			description: "Per-row random speed variance for irregular drifting motion",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Speed Variance",
				group: "Animation"
			}
		},
		twinkle: {
			default: 0,
			description: "Intensity of the twinkle effect (0 = off, 1 = full twinkle)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Twinkle",
				group: "Animation"
			}
		}
	},
	mapSampleUVs: ({ uniforms, ctx }) => {
		const cellCenter = call(dotGridCellCenterUV, "dotGridCellCenterUV", [
			ctx.uv,
			ctx.viewportSize,
			uniforms.density
		]);
		return {
			dotSize: cellCenter,
			twinkle: cellCenter
		};
	},
	fragment: (params) => {
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const animTime = animatedTime(params);
		const alpha = call(dotGridAlpha, "dotGridAlpha", [
			uv,
			viewport,
			params.uniforms.density,
			params.uniforms.dotSize,
			params.uniforms.offset,
			params.uniforms.speedVariance,
			params.uniforms.twinkle,
			animTime,
			params.ctx.time
		]);
		return vec4(params.uniforms.color.member("rgb"), alpha);
	}
};
var DotGrid_default = componentDefinition;
export { dotGridCellCenterUV as i, componentDefinition as n, dotGridAlpha as r, DotGrid_default as t };
