// shaders/dist/core/DropShadow-C7KAuipC.js
import { At as call, It as vec4, Ot as ZERO, Pt as floatE, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, i as transformBoolean } from "./transformations-fqzznt4g.js";
var DEG_TO_RAD = .017453292519943295;
var WEIGHTS = [
	.056,
	.135,
	.278,
	.487,
	.726,
	.923,
	1,
	.923,
	.726,
	.487,
	.278,
	.135,
	.056
];
var WEIGHT_SUM = 6.214;
const dropShadowUV = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, angle, distance) => {
	"use gpu";
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const dist = __tsover_mul(distance, .2);
	const aspect = __tsover_div(viewport.x, viewport.y);
	const ox = __tsover_div(__tsover_mul(std.sin(angleRad), dist), aspect);
	const oy = __tsover_mul(__tsover_mul(std.cos(angleRad), -1), dist);
	return d.vec2f(__tsover_sub(uv.x, ox), __tsover_sub(uv.y, oy));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "distance"
			}
		],
		"body": [0, [
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"dist",
				[
					1,
					"distance",
					"*",
					[5, "0.2"]
				]
			],
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"ox",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							["angleRad"]
						],
						"*",
						"dist"
					],
					"/",
					"aspect"
				]
			],
			[
				13,
				"oy",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"cos"
							],
							["angleRad"]
						],
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					],
					"*",
					"dist"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						7,
						"uv",
						"x"
					],
					"-",
					"ox"
				], [
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					"oy"
				]]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"std",
			"d"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		std,
		d
	})
}) && $.f)({}));
const dropShadowTapUV = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, blurRadius, ox, oy) => {
	"use gpu";
	return d.vec2f(__tsover_add(uv.x, __tsover_div(__tsover_mul(ox, blurRadius), viewport.x)), __tsover_add(uv.y, __tsover_div(__tsover_mul(oy, blurRadius), viewport.y)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "blurRadius"
			},
			{
				"type": "i",
				"name": "ox"
			},
			{
				"type": "i",
				"name": "oy"
			}
		],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					7,
					"uv",
					"x"
				],
				"+",
				[
					1,
					[
						1,
						"ox",
						"*",
						"blurRadius"
					],
					"/",
					[
						7,
						"viewport",
						"x"
					]
				]
			], [
				1,
				[
					7,
					"uv",
					"y"
				],
				"+",
				[
					1,
					[
						1,
						"oy",
						"*",
						"blurRadius"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			]]
		]]]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const dropShadowComposite = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((originalColor, shadowColor, shadowAlpha, cutout) => {
	"use gpu";
	const shadowRgb = d.vec3f(shadowColor.x, shadowColor.y, shadowColor.z);
	const shadowA = __tsover_mul(shadowColor.w, shadowAlpha);
	const cutoutA = std.clamp(__tsover_sub(shadowA, originalColor.w), 0, 1);
	const cutoutRgb = shadowRgb.mul(shadowColor.w).mul(cutoutA);
	const shadowPremulRgb = shadowRgb.mul(shadowColor.w).mul(shadowAlpha);
	const oneMinusOrigA = __tsover_sub(1, originalColor.w);
	const normalRgb = d.vec3f(originalColor.x, originalColor.y, originalColor.z).add(shadowPremulRgb.mul(oneMinusOrigA));
	const normalA = std.max(__tsover_add(originalColor.w, __tsover_mul(shadowA, oneMinusOrigA)), originalColor.w);
	const useCut = cutout > .5;
	const rgb = std.select(normalRgb, cutoutRgb, useCut);
	const a = std.select(normalA, cutoutA, useCut);
	return d.vec4f(rgb.x, rgb.y, rgb.z, a);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "originalColor"
			},
			{
				"type": "i",
				"name": "shadowColor"
			},
			{
				"type": "i",
				"name": "shadowAlpha"
			},
			{
				"type": "i",
				"name": "cutout"
			}
		],
		"body": [0, [
			[
				13,
				"shadowRgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"shadowColor",
							"x"
						],
						[
							7,
							"shadowColor",
							"y"
						],
						[
							7,
							"shadowColor",
							"z"
						]
					]
				]
			],
			[
				13,
				"shadowA",
				[
					1,
					[
						7,
						"shadowColor",
						"w"
					],
					"*",
					"shadowAlpha"
				]
			],
			[
				13,
				"cutoutA",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"shadowA",
							"-",
							[
								7,
								"originalColor",
								"w"
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"cutoutRgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"shadowRgb",
								"mul"
							],
							[[
								7,
								"shadowColor",
								"w"
							]]
						],
						"mul"
					],
					["cutoutA"]
				]
			],
			[
				13,
				"shadowPremulRgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"shadowRgb",
								"mul"
							],
							[[
								7,
								"shadowColor",
								"w"
							]]
						],
						"mul"
					],
					["shadowAlpha"]
				]
			],
			[
				13,
				"oneMinusOrigA",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"originalColor",
						"w"
					]
				]
			],
			[
				13,
				"normalRgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									7,
									"originalColor",
									"x"
								],
								[
									7,
									"originalColor",
									"y"
								],
								[
									7,
									"originalColor",
									"z"
								]
							]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"shadowPremulRgb",
							"mul"
						],
						["oneMinusOrigA"]
					]]
				]
			],
			[
				13,
				"normalA",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[
							7,
							"originalColor",
							"w"
						],
						"+",
						[
							1,
							"shadowA",
							"*",
							"oneMinusOrigA"
						]
					], [
						7,
						"originalColor",
						"w"
					]]
				]
			],
			[
				13,
				"useCut",
				[
					1,
					"cutout",
					">",
					[5, "0.5"]
				]
			],
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"normalRgb",
						"cutoutRgb",
						"useCut"
					]
				]
			],
			[
				13,
				"a",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"normalA",
						"cutoutA",
						"useCut"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"rgb",
						"x"
					],
					[
						7,
						"rgb",
						"y"
					],
					[
						7,
						"rgb",
						"z"
					],
					"a"
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "DropShadow",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Adds a soft shadow behind the child content based on its alpha silhouette",
	requiresRTT: true,
	requiresChild: true,
	props: {
		color: {
			default: "#000000",
			transform: transformColor,
			description: "Shadow color",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		distance: {
			default: .1,
			description: "How far the shadow is offset from the content",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .005,
				label: "Distance",
				group: "Effect"
			}
		},
		angle: {
			default: 135,
			description: "Direction the shadow is cast (compass degrees: 0=up, 90=right, 135=lower-right, 180=down)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		blur: {
			default: 5,
			description: "Shadow softness (blur radius in pixels)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .5,
				label: "Blur",
				group: "Effect"
			}
		},
		intensity: {
			default: .5,
			description: "Shadow intensity — how strong/visible the shadow is",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Intensity",
				group: "Effect"
			}
		},
		cutout: {
			default: false,
			transform: transformBoolean,
			compileTime: true,
			description: "Hide the original layer and show only the shadow",
			ui: {
				type: "checkbox",
				label: "Cutout",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { childNode, convertToTexture, uniforms, propValues, ctx } = params;
		if (!childNode) return ZERO;
		const childTexture = convertToTexture(childNode);
		const originalColor = childTexture.sample(ctx.uv);
		const shadowUV = call(dropShadowUV, "dropShadowUV", [
			ctx.uv,
			ctx.viewportSize,
			uniforms.angle,
			uniforms.distance
		]);
		const hTap = (x) => childTexture.sample(call(dropShadowTapUV, "dropShadowTapUV", [
			shadowUV,
			ctx.viewportSize,
			uniforms.blur,
			floatE(x),
			floatE(0)
		])).member("a").mul(WEIGHTS[x + 6] / WEIGHT_SUM);
		let hTotal = hTap(-6);
		for (let x = -5; x <= 6; x++) hTotal = hTotal.add(hTap(x));
		const horizontalTexture = convertToTexture(vec4(hTotal, floatE(0), floatE(0), floatE(1)));
		const vTap = (y) => horizontalTexture.sample(call(dropShadowTapUV, "dropShadowTapUV", [
			ctx.uv,
			ctx.viewportSize,
			uniforms.blur,
			floatE(0),
			floatE(y)
		])).member("r").mul(WEIGHTS[y + 6] / WEIGHT_SUM);
		let vTotal = vTap(-6);
		for (let y = -5; y <= 6; y++) vTotal = vTotal.add(vTap(y));
		const shadowAlpha = vTotal.mul(uniforms.intensity);
		const cutout = propValues.cutout === 1;
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [call(dropShadowComposite, "dropShadowComposite", [
			originalColor,
			uniforms.color,
			shadowAlpha,
			floatE(cutout ? 1 : 0)
		])]);
	}
};
var DropShadow_default = componentDefinition;
export { dropShadowUV as a, dropShadowTapUV as i, componentDefinition as n, dropShadowComposite as r, DropShadow_default as t };
