// shaders/dist/core/Duotone-CuJunWJ0.js
import { At as call, It as vec4, Ot as ZERO, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
const duotoneFactor = tgpu.fn([d.vec4f, d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, blend) => {
	"use gpu";
	const luminance = std.dot(color.xyz, d.vec3f(.299, .587, .114));
	return std.smoothstep(__tsover_sub(blend, .5), __tsover_add(blend, .5), luminance);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}, {
			"type": "i",
			"name": "blend"
		}],
		"body": [0, [[
			13,
			"luminance",
			[
				6,
				[
					7,
					"std",
					"dot"
				],
				[[
					7,
					"color",
					"xyz"
				], [
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0.299"],
						[5, "0.587"],
						[5, "0.114"]
					]
				]]
			]
		], [10, [
			6,
			[
				7,
				"std",
				"smoothstep"
			],
			[
				[
					1,
					"blend",
					"-",
					[5, "0.5"]
				],
				[
					1,
					"blend",
					"+",
					[5, "0.5"]
				],
				"luminance"
			]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Duotone",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Map colors to two tones based on luminance",
	requiresChild: true,
	props: {
		colorA: {
			default: "#ff0000",
			transform: transformColor,
			description: "First color (used for darker areas)",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#023af4",
			transform: transformColor,
			description: "Second color (used for brighter areas)",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		blend: {
			default: .5,
			description: "Blend point between the two colors",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Blend",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: ({ uniforms, childNode, propValues }) => {
		if (!childNode) return ZERO;
		const t = call(duotoneFactor, "duotoneFactor", [childNode, uniforms.blend]);
		return vec4(call(mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			t
		]).member("rgb"), childNode.member("a"));
	}
};
var Duotone_default = componentDefinition;
export { componentDefinition as n, duotoneFactor as r, Duotone_default as t };
