// shaders/dist/core/Engraving-fQ1zZJ89.js
import { At as call, It as vec4, Ot as ZERO, Pt as floatE, R as std, Wt as unpremultiplyAlpha, et as mxNoiseFloat2, n as tgpu, on as d, rt as noise_exports } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition, r as transformAngle } from "./transformations-fqzznt4g.js";
var TAU = 6.283185307179586;
var DEG_TO_RAD = .017453292519943295;
var transformStyle = (value) => value === "spiral" ? 2 : value === "line" ? 0 : 1;
var styleModeOf = (raw) => typeof raw === "number" ? raw : transformStyle(String(raw));
const engravingPlate = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, angleDeg, frequency, reliefPhase, level, aa) => {
	"use gpu";
	const angleRad = __tsover_mul(angleDeg, DEG_TO_RAD);
	const c = std.cos(angleRad);
	const s = std.sin(angleRad);
	const w = __tsover_add(__tsover_mul(p.x, __tsover_mul(s, -1)), __tsover_mul(p.y, c));
	const phase = __tsover_add(__tsover_mul(__tsover_mul(w, frequency), TAU), reliefPhase);
	const v = std.cos(phase);
	const thr = std.mix(-1.15, 1.15, std.clamp(level, 0, 1));
	return std.smoothstep(__tsover_sub(thr, aa), __tsover_add(thr, aa), v);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "p"
			},
			{
				"type": "i",
				"name": "angleDeg"
			},
			{
				"type": "i",
				"name": "frequency"
			},
			{
				"type": "i",
				"name": "reliefPhase"
			},
			{
				"type": "i",
				"name": "level"
			},
			{
				"type": "i",
				"name": "aa"
			}
		],
		"body": [0, [
			[
				13,
				"angleRad",
				[
					1,
					"angleDeg",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"c",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"s",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"w",
				[
					1,
					[
						1,
						[
							7,
							"p",
							"x"
						],
						"*",
						[
							1,
							"s",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						]
					],
					"+",
					[
						1,
						[
							7,
							"p",
							"y"
						],
						"*",
						"c"
					]
				]
			],
			[
				13,
				"phase",
				[
					1,
					[
						1,
						[
							1,
							"w",
							"*",
							"frequency"
						],
						"*",
						"TAU"
					],
					"+",
					"reliefPhase"
				]
			],
			[
				13,
				"v",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["phase"]
				]
			],
			[
				13,
				"thr",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							4,
							"-",
							[5, "1.15"]
						],
						[5, "1.15"],
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								"level",
								[5, "0"],
								[5, "1"]
							]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"smoothstep"
				],
				[
					[
						1,
						"thr",
						"-",
						"aa"
					],
					[
						1,
						"thr",
						"+",
						"aa"
					],
					"v"
				]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"std",
			"TAU"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		std,
		TAU
	})
}) && $.f)({}));
const engravingSpiralPlate = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, c, frequency, reliefPhase, level, aa) => {
	"use gpu";
	const dv = p.sub(c);
	const r = std.length(dv);
	const theta = std.atan2(dv.y, dv.x);
	const phase = __tsover_add(__tsover_mul(__tsover_sub(__tsover_mul(r, frequency), __tsover_mul(theta, __tsover_div(1, TAU))), TAU), reliefPhase);
	const v = std.cos(phase);
	const thr = std.mix(-1.15, 1.15, std.clamp(level, 0, 1));
	return std.smoothstep(__tsover_sub(thr, aa), __tsover_add(thr, aa), v);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "p"
			},
			{
				"type": "i",
				"name": "c"
			},
			{
				"type": "i",
				"name": "frequency"
			},
			{
				"type": "i",
				"name": "reliefPhase"
			},
			{
				"type": "i",
				"name": "level"
			},
			{
				"type": "i",
				"name": "aa"
			}
		],
		"body": [0, [
			[
				13,
				"dv",
				[
					6,
					[
						7,
						"p",
						"sub"
					],
					["c"]
				]
			],
			[
				13,
				"r",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["dv"]
				]
			],
			[
				13,
				"theta",
				[
					6,
					[
						7,
						"std",
						"atan2"
					],
					[[
						7,
						"dv",
						"y"
					], [
						7,
						"dv",
						"x"
					]]
				]
			],
			[
				13,
				"phase",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								"r",
								"*",
								"frequency"
							],
							"-",
							[
								1,
								"theta",
								"*",
								[
									1,
									[5, "1"],
									"/",
									"TAU"
								]
							]
						],
						"*",
						"TAU"
					],
					"+",
					"reliefPhase"
				]
			],
			[
				13,
				"v",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["phase"]
				]
			],
			[
				13,
				"thr",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							4,
							"-",
							[5, "1.15"]
						],
						[5, "1.15"],
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								"level",
								[5, "0"],
								[5, "1"]
							]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"smoothstep"
				],
				[
					[
						1,
						"thr",
						"-",
						"aa"
					],
					[
						1,
						"thr",
						"+",
						"aa"
					],
					"v"
				]
			]]
		]],
		"externalNames": ["std", "TAU"]
	},
	externals: () => ({
		std,
		TAU
	})
}) && $.f)({}));
function makeEngravingComposite(style) {
	return tgpu.fn([
		d.vec4f,
		d.vec2f,
		d.f32,
		d.vec4f,
		d.vec4f,
		d.vec2f,
		d.vec4f,
		d.vec4f
	], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((childColor, uv, aspect, paper, ink, center, p0, p1) => {
		"use gpu";
		const angleDeg = p0.x;
		const frequency = p0.y;
		const relief = p0.z;
		const waviness = p0.w;
		const contrast = p1.x;
		const viewportY = p1.y;
		const lum = std.dot(d.vec3f(childColor.x, childColor.y, childColor.z), d.vec3f(.299, .587, .114));
		const level = std.clamp(__tsover_add(.5, __tsover_mul(__tsover_sub(lum, .5), contrast)), 0, 1);
		const pa = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
		const nx = mxNoiseFloat2(pa.mul(3.1));
		const ny = mxNoiseFloat2(pa.mul(3.1).add(d.vec2f(7.31, 3.77)));
		const p = pa.add(d.vec2f(nx, ny).mul(__tsover_mul(waviness, .06)));
		const aa = std.clamp(__tsover_div(__tsover_mul(frequency, TAU), std.max(viewportY, 1)), .02, 1);
		const reliefPhase = __tsover_mul(__tsover_mul(relief, level), 9.42);
		let cover = d.f32(0);
		if (style === 2) cover = engravingSpiralPlate(p, d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y)), frequency, reliefPhase, level, aa);
		else if (style === 1) {
			const ink1 = engravingPlate(p, angleDeg, frequency, reliefPhase, level, aa);
			const level2 = std.clamp(__tsover_div(level, .55), 0, 1);
			const ink2 = engravingPlate(p, __tsover_add(angleDeg, 72), __tsover_mul(frequency, .92), __tsover_mul(reliefPhase, .7), level2, aa);
			const level3 = std.clamp(__tsover_div(level, .28), 0, 1);
			const ink3 = engravingPlate(p, __tsover_sub(angleDeg, 38), __tsover_mul(frequency, 1.13), __tsover_mul(reliefPhase, .5), level3, aa);
			cover = std.max(ink1, std.max(ink2, ink3));
		} else cover = engravingPlate(p, angleDeg, frequency, reliefPhase, level, aa);
		const rgb = std.mix(d.vec3f(paper.x, paper.y, paper.z), d.vec3f(ink.x, ink.y, ink.z), d.vec3f(cover));
		const alpha = __tsover_mul(childColor.w, std.mix(paper.w, ink.w, cover));
		return d.vec4f(rgb.x, rgb.y, rgb.z, alpha);
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [
				{
					"type": "i",
					"name": "childColor"
				},
				{
					"type": "i",
					"name": "uv"
				},
				{
					"type": "i",
					"name": "aspect"
				},
				{
					"type": "i",
					"name": "paper"
				},
				{
					"type": "i",
					"name": "ink"
				},
				{
					"type": "i",
					"name": "center"
				},
				{
					"type": "i",
					"name": "p0"
				},
				{
					"type": "i",
					"name": "p1"
				}
			],
			"body": [0, [
				[
					13,
					"angleDeg",
					[
						7,
						"p0",
						"x"
					]
				],
				[
					13,
					"frequency",
					[
						7,
						"p0",
						"y"
					]
				],
				[
					13,
					"relief",
					[
						7,
						"p0",
						"z"
					]
				],
				[
					13,
					"waviness",
					[
						7,
						"p0",
						"w"
					]
				],
				[
					13,
					"contrast",
					[
						7,
						"p1",
						"x"
					]
				],
				[
					13,
					"viewportY",
					[
						7,
						"p1",
						"y"
					]
				],
				[
					13,
					"lum",
					[
						6,
						[
							7,
							"std",
							"dot"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									7,
									"childColor",
									"x"
								],
								[
									7,
									"childColor",
									"y"
								],
								[
									7,
									"childColor",
									"z"
								]
							]
						], [
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0.299"],
								[5, "0.587"],
								[5, "0.114"]
							]
						]]
					]
				],
				[
					13,
					"level",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[5, "0.5"],
								"+",
								[
									1,
									[
										1,
										"lum",
										"-",
										[5, "0.5"]
									],
									"*",
									"contrast"
								]
							],
							[5, "0"],
							[5, "1"]
						]
					]
				],
				[
					13,
					"pa",
					[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								7,
								"uv",
								"x"
							],
							"*",
							"aspect"
						], [
							7,
							"uv",
							"y"
						]]
					]
				],
				[
					13,
					"nx",
					[
						6,
						[
							7,
							"noise",
							"mxNoiseFloat2"
						],
						[[
							6,
							[
								7,
								"pa",
								"mul"
							],
							[[5, "3.1"]]
						]]
					]
				],
				[
					13,
					"ny",
					[
						6,
						[
							7,
							"noise",
							"mxNoiseFloat2"
						],
						[[
							6,
							[
								7,
								[
									6,
									[
										7,
										"pa",
										"mul"
									],
									[[5, "3.1"]]
								],
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[5, "7.31"], [5, "3.77"]]
							]]
						]]
					]
				],
				[
					13,
					"p",
					[
						6,
						[
							7,
							"pa",
							"add"
						],
						[[
							6,
							[
								7,
								[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									["nx", "ny"]
								],
								"mul"
							],
							[[
								1,
								"waviness",
								"*",
								[5, "0.06"]
							]]
						]]
					]
				],
				[
					13,
					"aa",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									"frequency",
									"*",
									"TAU"
								],
								"/",
								[
									6,
									[
										7,
										"std",
										"max"
									],
									["viewportY", [5, "1"]]
								]
							],
							[5, "0.02"],
							[5, "1"]
						]
					]
				],
				[
					13,
					"reliefPhase",
					[
						1,
						[
							1,
							"relief",
							"*",
							"level"
						],
						"*",
						[5, "9.42"]
					]
				],
				[
					12,
					"cover",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					11,
					[
						1,
						"style",
						"===",
						[5, "2"]
					],
					[0, [[
						13,
						"sc",
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								[
									7,
									"center",
									"x"
								],
								"*",
								"aspect"
							], [
								1,
								[5, "1"],
								"-",
								[
									7,
									"center",
									"y"
								]
							]]
						]
					], [
						2,
						"cover",
						"=",
						[
							6,
							"engravingSpiralPlate",
							[
								"p",
								"sc",
								"frequency",
								"reliefPhase",
								"level",
								"aa"
							]
						]
					]]],
					[
						11,
						[
							1,
							"style",
							"===",
							[5, "1"]
						],
						[0, [
							[
								13,
								"ink1",
								[
									6,
									"engravingPlate",
									[
										"p",
										"angleDeg",
										"frequency",
										"reliefPhase",
										"level",
										"aa"
									]
								]
							],
							[
								13,
								"level2",
								[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										[
											1,
											"level",
											"/",
											[5, "0.55"]
										],
										[5, "0"],
										[5, "1"]
									]
								]
							],
							[
								13,
								"ink2",
								[
									6,
									"engravingPlate",
									[
										"p",
										[
											1,
											"angleDeg",
											"+",
											[5, "72"]
										],
										[
											1,
											"frequency",
											"*",
											[5, "0.92"]
										],
										[
											1,
											"reliefPhase",
											"*",
											[5, "0.7"]
										],
										"level2",
										"aa"
									]
								]
							],
							[
								13,
								"level3",
								[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										[
											1,
											"level",
											"/",
											[5, "0.28"]
										],
										[5, "0"],
										[5, "1"]
									]
								]
							],
							[
								13,
								"ink3",
								[
									6,
									"engravingPlate",
									[
										"p",
										[
											1,
											"angleDeg",
											"-",
											[5, "38"]
										],
										[
											1,
											"frequency",
											"*",
											[5, "1.13"]
										],
										[
											1,
											"reliefPhase",
											"*",
											[5, "0.5"]
										],
										"level3",
										"aa"
									]
								]
							],
							[
								2,
								"cover",
								"=",
								[
									6,
									[
										7,
										"std",
										"max"
									],
									["ink1", [
										6,
										[
											7,
											"std",
											"max"
										],
										["ink2", "ink3"]
									]]
								]
							]
						]],
						[0, [[
							2,
							"cover",
							"=",
							[
								6,
								"engravingPlate",
								[
									"p",
									"angleDeg",
									"frequency",
									"reliefPhase",
									"level",
									"aa"
								]
							]
						]]]
					]
				],
				[
					13,
					"rgb",
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										7,
										"paper",
										"x"
									],
									[
										7,
										"paper",
										"y"
									],
									[
										7,
										"paper",
										"z"
									]
								]
							],
							[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										7,
										"ink",
										"x"
									],
									[
										7,
										"ink",
										"y"
									],
									[
										7,
										"ink",
										"z"
									]
								]
							],
							[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								["cover"]
							]
						]
					]
				],
				[
					13,
					"alpha",
					[
						1,
						[
							7,
							"childColor",
							"w"
						],
						"*",
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									7,
									"paper",
									"w"
								],
								[
									7,
									"ink",
									"w"
								],
								"cover"
							]
						]
					]
				],
				[10, [
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							7,
							"rgb",
							"x"
						],
						[
							7,
							"rgb",
							"y"
						],
						[
							7,
							"rgb",
							"z"
						],
						"alpha"
					]
				]]
			]],
			"externalNames": [
				"std",
				"d",
				"noise",
				"TAU",
				"style",
				"engravingSpiralPlate",
				"engravingPlate"
			]
		},
		externals: () => ({
			std,
			d,
			noise: noise_exports,
			TAU,
			style,
			engravingSpiralPlate,
			engravingPlate
		})
	}) && $.f)({})).$name(`engravingComposite_${style}`);
}
const componentDefinition = {
	name: "Engraving",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Copper-plate line engraving — the image is redrawn as flowing line work whose weight swells with darkness, lines displaced by the form like a banknote portrait: a single plate, cross-hatched shadow plates, or one continuous spiral cut",
	requiresRTT: true,
	requiresChild: true,
	blendWithChildren: false,
	props: {
		style: {
			default: "crosshatch",
			transform: transformStyle,
			compileTime: true,
			description: "The engraving cut: a single line plate, cross-hatched plates building up the shadows, or one continuous spiral",
			ui: {
				type: "select",
				options: [
					{
						label: "Line",
						value: "line"
					},
					{
						label: "Cross-Hatch",
						value: "crosshatch"
					},
					{
						label: "Spiral",
						value: "spiral"
					}
				],
				label: "Style",
				group: "Lines"
			}
		},
		frequency: {
			default: 90,
			description: "Line density — how many engraved lines span the height of the canvas",
			ui: {
				type: ["range", "map"],
				min: 20,
				max: 300,
				step: 1,
				label: "Frequency",
				group: "Lines"
			}
		},
		angle: {
			default: 8,
			transform: transformAngle,
			description: "Direction of the base line work (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Lines",
				condition: { style: "line" }
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center the spiral coils outward from",
			ui: {
				type: "position",
				label: "Center",
				group: "Lines",
				condition: { style: "spiral" }
			}
		},
		relief: {
			default: .6,
			description: "How strongly the image brightness displaces the lines — the engraved 'lines climb over the form' effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .01,
				label: "Relief",
				group: "Lines"
			}
		},
		waviness: {
			default: .35,
			description: "Organic meander of the line work, like a hand-pulled burin stroke",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Waviness",
				group: "Lines"
			}
		},
		contrast: {
			default: 1.15,
			description: "Tonal contrast applied before the lines are cut — higher pushes mid-tones toward pure line or pure paper",
			ui: {
				type: ["range", "map"],
				min: .25,
				max: 3,
				step: .01,
				label: "Contrast",
				group: "Tone"
			}
		},
		inkColor: {
			default: "#1a1410",
			transform: transformColor,
			description: "Color of the engraved ink",
			ui: {
				type: "color",
				label: "Ink",
				group: "Colors"
			}
		},
		paperColor: {
			default: "#f4eee2",
			transform: transformColor,
			description: "Paper color shown between the lines",
			ui: {
				type: "color",
				label: "Paper",
				group: "Colors"
			}
		}
	},
	fragment: ({ childNode, ctx, uniforms, propValues, convertToTexture }) => {
		if (!childNode) return ZERO;
		const childColor = call(unpremultiplyAlpha, "unpremultiplyAlpha", [convertToTexture(childNode).sample(ctx.uv)]);
		const style = styleModeOf(propValues.style);
		const composite = makeEngravingComposite(style);
		const p0 = vec4(uniforms.angle, uniforms.frequency, uniforms.relief, uniforms.waviness);
		const p1 = vec4(uniforms.contrast, ctx.viewportSize.member("y"), floatE(0), floatE(0));
		return call(composite, `engravingComposite_${style}`, [
			childColor,
			ctx.uv,
			ctx.aspect,
			uniforms.paperColor,
			uniforms.inkColor,
			uniforms.center,
			p0,
			p1
		]);
	}
};
var Engraving_default = componentDefinition;
export { makeEngravingComposite as a, engravingSpiralPlate as i, componentDefinition as n, engravingPlate as r, Engraving_default as t };
