// shaders/dist/core/ErosionNoise-ZrJulbtK.js
import { At as call, J as erosion12, R as std, W as toneAndColorExpr, n as tgpu, on as d, rt as noise_exports } from "./coords-BlwgBpJf.js";
import { i as noiseToneProps, t as noiseColorProps } from "./noiseColor-Cr76BYLN.js";
const erosionNoiseField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, scale, seed) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const pos = d.vec2f(__tsover_mul(uv.x, aspect), uv.y).mul(std.exp(scale)).add(seed);
	return __tsover_add(__tsover_mul(erosion12(pos).x, .5), .5);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "seed"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"pos",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"aspectUV",
								"mul"
							],
							[[
								6,
								[
									7,
									"std",
									"exp"
								],
								["scale"]
							]]
						],
						"add"
					],
					["seed"]
				]
			],
			[10, [
				1,
				[
					1,
					[
						7,
						[
							6,
							[
								7,
								"noise",
								"erosion12"
							],
							["pos"]
						],
						"x"
					],
					"*",
					[5, "0.5"]
				],
				"+",
				[5, "0.5"]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"noise"
		]
	},
	externals: () => ({
		d,
		std,
		noise: noise_exports
	})
}) && $.f)({}));
const componentDefinition = {
	name: "ErosionNoise",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Branching, hydraulic-erosion ridges carved into noise",
	acceptsUVContext: true,
	props: {
		...noiseColorProps(),
		scale: {
			default: 1.5,
			description: "Pattern scale (higher = larger patterns)",
			ui: {
				type: "range",
				min: -2,
				max: 5,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		...noiseToneProps()
	},
	fragment: (params) => {
		return toneAndColorExpr(params, call(erosionNoiseField, "erosionNoiseField", [
			params.uvContext ?? params.ctx.uv,
			params.effectiveViewportSize ?? params.ctx.viewportSize,
			params.uniforms.scale,
			params.uniforms.seed
		]));
	}
};
var ErosionNoise_default = componentDefinition;
export { componentDefinition as n, erosionNoiseField as r, ErosionNoise_default as t };
