// shaders/dist/core/Exposure-D8apJapP.js
import { At as call, Ot as ZERO, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
const exposureColor = tgpu.fn([d.vec4f, d.f32], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, gain) => {
	"use gpu";
	const rgb = color.xyz.mul(gain);
	return d.vec4f(rgb, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}, {
			"type": "i",
			"name": "gain"
		}],
		"body": [0, [[
			13,
			"rgb",
			[
				6,
				[
					7,
					[
						7,
						"color",
						"xyz"
					],
					"mul"
				],
				["gain"]
			]
		], [10, [
			6,
			[
				7,
				"d",
				"vec4f"
			],
			["rgb", [
				7,
				"color",
				"w"
			]]
		]]]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const componentDefinition = {
	name: "Exposure",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Multiplicative exposure (gain) on the child. Unlike additive brightness, it scales values around black — blacks stay black while highlights can be pushed past 1.0 to feed HDR bloom and tone-mapping roll-off. Pointwise, so no extra render pass is needed.",
	requiresChild: true,
	props: { exposure: {
		default: 1,
		description: "Brightness multiplier (gain). 1 = unchanged, >1 pushes highlights past white for HDR bloom, <1 darkens.",
		ui: {
			type: ["range", "map"],
			min: 0,
			max: 8,
			step: .1,
			label: "Exposure",
			group: "Adjustments"
		}
	} },
	fragment: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Exposure shader.");
			return ZERO;
		}
		return call(exposureColor, "exposureColor", [childNode, uniforms.exposure]);
	}
};
var Exposure_default = componentDefinition;
export { componentDefinition as n, exposureColor as r, Exposure_default as t };
