// shaders/dist/core/FallingLines-Bv8X8gRy.js
import { At as call, It as vec4, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
var DEG_TO_RAD = Math.PI / 180;
var colHash = tgpu.fn([d.f32], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((idx) => {
	"use gpu";
	return d.vec2f(std.fract(__tsover_mul(std.sin(__tsover_mul(idx, 127.1)), 43758.5453123)), std.fract(__tsover_mul(std.sin(__tsover_mul(idx, 311.7)), 38291.3421)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "idx"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				6,
				[
					7,
					"std",
					"fract"
				],
				[[
					1,
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						[[
							1,
							"idx",
							"*",
							[5, "127.1"]
						]]
					],
					"*",
					[5, "43758.5453123"]
				]]
			], [
				6,
				[
					7,
					"std",
					"fract"
				],
				[[
					1,
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						[[
							1,
							"idx",
							"*",
							[5, "311.7"]
						]]
					],
					"*",
					[5, "38291.3421"]
				]]
			]]
		]]]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const fallingLinesField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, vp, angle, density, speedVariance, trailLength, strokeWidth, rounding, balance, animTime) => {
	"use gpu";
	const aspect = __tsover_div(vp.x, vp.y);
	const centeredUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y).sub(d.vec2f(__tsover_mul(aspect, .5), .5));
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const cosA = std.cos(angleRad);
	const sinA = std.sin(angleRad);
	const u = __tsover_sub(__tsover_mul(centeredUV.y, cosA), __tsover_mul(centeredUV.x, sinA));
	const v = __tsover_add(__tsover_mul(centeredUV.x, cosA), __tsover_mul(centeredUV.y, sinA));
	const colF = __tsover_mul(u, density);
	const colIdx = std.floor(colF);
	const colLocalU = std.fract(colF);
	const h = colHash(colIdx);
	const speedMod = __tsover_add(1, __tsover_mul(__tsover_mul(__tsover_sub(h.x, .5), speedVariance), 2));
	const phase = h.y;
	const animOffset = std.fract(__tsover_mul(animTime, speedMod));
	const vPeriodic = std.fract(__tsover_sub(__tsover_sub(v, animOffset), phase));
	const trailLen = trailLength;
	const withinStreak = std.clamp(__tsover_div(vPeriodic, trailLen), 0, 1);
	const fwV = std.clamp(std.fwidth(v), .001, .1);
	const fwU = std.clamp(__tsover_mul(std.fwidth(u), density), .001, .1);
	const streakMask = __tsover_sub(1, std.smoothstep(__tsover_sub(trailLen, fwV), __tsover_add(trailLen, fwV), vPeriodic));
	const hw = __tsover_mul(strokeWidth, .5);
	const distFromCenter = std.abs(__tsover_sub(colLocalU, .5));
	const widthMask = __tsover_sub(1, std.smoothstep(__tsover_sub(hw, fwU), __tsover_add(hw, fwU), distFromCenter));
	const rectMask = __tsover_mul(streakMask, widthMask);
	const maxDepth = __tsover_mul(hw, rounding);
	const gapDist = __tsover_mul(std.max(__tsover_sub(vPeriodic, trailLen), 0), density);
	const tcap = std.clamp(__tsover_div(gapDist, std.max(maxDepth, 1e-4)), 0, 1);
	const capHW = __tsover_mul(hw, std.sqrt(std.max(__tsover_sub(1, __tsover_mul(tcap, tcap)), 0)));
	const capAlpha = __tsover_mul(__tsover_sub(1, std.smoothstep(std.max(__tsover_sub(capHW, fwU), 0), std.max(__tsover_add(capHW, fwU), .001), distFromCenter)), std.smoothstep(0, fwU, capHW));
	const finalMask = std.max(rectMask, capAlpha);
	const balanceNum = __tsover_sub(__tsover_mul(2, balance), 1);
	const balanceDen = std.max(__tsover_mul(2, __tsover_sub(1, balance)), .001);
	const balancedT = std.clamp(__tsover_div(__tsover_add(withinStreak, balanceNum), balanceDen), 0, 1);
	return d.vec2f(finalMask, balancedT);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "vp"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "density"
			},
			{
				"type": "i",
				"name": "speedVariance"
			},
			{
				"type": "i",
				"name": "trailLength"
			},
			{
				"type": "i",
				"name": "strokeWidth"
			},
			{
				"type": "i",
				"name": "rounding"
			},
			{
				"type": "i",
				"name": "balance"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"vp",
						"x"
					],
					"/",
					[
						7,
						"vp",
						"y"
					]
				]
			],
			[
				13,
				"corrUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centeredUV",
				[
					6,
					[
						7,
						"corrUV",
						"sub"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							"aspect",
							"*",
							[5, "0.5"]
						], [5, "0.5"]]
					]]
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"u",
				[
					1,
					[
						1,
						[
							7,
							"centeredUV",
							"y"
						],
						"*",
						"cosA"
					],
					"-",
					[
						1,
						[
							7,
							"centeredUV",
							"x"
						],
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"v",
				[
					1,
					[
						1,
						[
							7,
							"centeredUV",
							"x"
						],
						"*",
						"cosA"
					],
					"+",
					[
						1,
						[
							7,
							"centeredUV",
							"y"
						],
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"colF",
				[
					1,
					"u",
					"*",
					"density"
				]
			],
			[
				13,
				"colIdx",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["colF"]
				]
			],
			[
				13,
				"colLocalU",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["colF"]
				]
			],
			[
				13,
				"h",
				[
					6,
					"colHash",
					["colIdx"]
				]
			],
			[
				13,
				"speedMod",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						[
							1,
							[
								1,
								[
									7,
									"h",
									"x"
								],
								"-",
								[5, "0.5"]
							],
							"*",
							"speedVariance"
						],
						"*",
						[5, "2"]
					]
				]
			],
			[
				13,
				"phase",
				[
					7,
					"h",
					"y"
				]
			],
			[
				13,
				"animOffset",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						"animTime",
						"*",
						"speedMod"
					]]
				]
			],
			[
				13,
				"vPeriodic",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							1,
							"v",
							"-",
							"animOffset"
						],
						"-",
						"phase"
					]]
				]
			],
			[
				13,
				"trailLen",
				"trailLength"
			],
			[
				13,
				"withinStreak",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"vPeriodic",
							"/",
							"trailLen"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"fwV",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"fwidth"
							],
							["v"]
						],
						[5, "0.001"],
						[5, "0.1"]
					]
				]
			],
			[
				13,
				"fwU",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"fwidth"
								],
								["u"]
							],
							"*",
							"density"
						],
						[5, "0.001"],
						[5, "0.1"]
					]
				]
			],
			[
				13,
				"streakMask",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"trailLen",
								"-",
								"fwV"
							],
							[
								1,
								"trailLen",
								"+",
								"fwV"
							],
							"vPeriodic"
						]
					]
				]
			],
			[
				13,
				"hw",
				[
					1,
					"strokeWidth",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"distFromCenter",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						1,
						"colLocalU",
						"-",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"widthMask",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"hw",
								"-",
								"fwU"
							],
							[
								1,
								"hw",
								"+",
								"fwU"
							],
							"distFromCenter"
						]
					]
				]
			],
			[
				13,
				"rectMask",
				[
					1,
					"streakMask",
					"*",
					"widthMask"
				]
			],
			[
				13,
				"maxDepth",
				[
					1,
					"hw",
					"*",
					"rounding"
				]
			],
			[
				13,
				"gapDist",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							1,
							"vPeriodic",
							"-",
							"trailLen"
						], [5, "0"]]
					],
					"*",
					"density"
				]
			],
			[
				13,
				"tcap",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"gapDist",
							"/",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["maxDepth", [5, "0.0001"]]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"capHW",
				[
					1,
					"hw",
					"*",
					[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								1,
								[5, "1"],
								"-",
								[
									1,
									"tcap",
									"*",
									"tcap"
								]
							], [5, "0"]]
						]]
					]
				]
			],
			[
				13,
				"capAlpha",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[
									6,
									[
										7,
										"std",
										"max"
									],
									[[
										1,
										"capHW",
										"-",
										"fwU"
									], [5, "0"]]
								],
								[
									6,
									[
										7,
										"std",
										"max"
									],
									[[
										1,
										"capHW",
										"+",
										"fwU"
									], [5, "0.001"]]
								],
								"distFromCenter"
							]
						]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							"fwU",
							"capHW"
						]
					]
				]
			],
			[
				13,
				"finalMask",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["rectMask", "capAlpha"]
				]
			],
			[
				13,
				"balanceNum",
				[
					1,
					[
						1,
						[5, "2"],
						"*",
						"balance"
					],
					"-",
					[5, "1"]
				]
			],
			[
				13,
				"balanceDen",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[5, "2"],
						"*",
						[
							1,
							[5, "1"],
							"-",
							"balance"
						]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"balancedT",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"withinStreak",
								"+",
								"balanceNum"
							],
							"/",
							"balanceDen"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["finalMask", "balancedT"]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std",
			"colHash"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std,
		colHash
	})
}) && $.f)({}));
const componentDefinition = {
	name: "FallingLines",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Directional falling lines with a leading-to-trailing color fade",
	requiresRTT: false,
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color at the leading edge of each line",
			ui: {
				type: "color",
				label: "Lead Color",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff00",
			transform: transformColor,
			description: "Color at the trailing edge (transparent by default)",
			ui: {
				type: "color",
				label: "Trail Color",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for interpolation between lead and trail colors",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		angle: {
			default: 90,
			description: "Direction of movement in degrees (90=down, 270=up, 0=right, 180=left)",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		speed: {
			default: .5,
			description: "Movement speed",
			ui: {
				type: "range",
				min: 0,
				max: 3,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		speedVariance: {
			default: .3,
			description: "Per-line speed variance (0=uniform, 1=high variance)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Speed Variance",
				group: "Animation"
			}
		},
		density: {
			default: 15,
			description: "Number of line columns across the canvas",
			ui: {
				type: "range",
				min: 1,
				max: 60,
				step: 1,
				label: "Density",
				group: "Effect"
			}
		},
		trailLength: {
			default: .35,
			description: "Streak length relative to spacing (0=point, 1=continuous)",
			ui: {
				type: ["range", "map"],
				min: .01,
				max: 1,
				step: .01,
				label: "Trail Length",
				group: "Effect"
			}
		},
		balance: {
			default: .5,
			description: "Color mix midpoint (0.5=linear, 0=all trailing/colorB, 1=all leading/colorA)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Balance",
				group: "Colors"
			}
		},
		strokeWidth: {
			default: .15,
			description: "Line thickness as fraction of column width (0=hairline, 1=full width)",
			ui: {
				type: ["range", "map"],
				min: .02,
				max: 1,
				step: .02,
				label: "Stroke Width",
				group: "Effect"
			}
		},
		rounding: {
			default: 1,
			description: "Rounds the leading edge (0=flat/square, 1=fully rounded cap)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Rounding",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const uv = params.uvContext ?? ctx.uv;
		const vp = params.effectiveViewportSize ?? ctx.viewportSize;
		const animTime = animatedTime(params);
		const field = call(fallingLinesField, "fallingLinesField", [
			uv,
			vp,
			uniforms.angle,
			uniforms.density,
			uniforms.speedVariance,
			uniforms.trailLength,
			uniforms.strokeWidth,
			uniforms.rounding,
			uniforms.balance,
			animTime
		]);
		const finalMask = field.member("x");
		const balancedT = field.member("y");
		const mixedColor = call(mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			uniforms.colorB,
			uniforms.colorA,
			balancedT
		]);
		return vec4(mixedColor.member("rgb"), mixedColor.member("a").mul(finalMask));
	}
};
var FallingLines_default = componentDefinition;
export { componentDefinition as n, fallingLinesField as r, FallingLines_default as t };
