// shaders/dist/core/FilmGrain-CoW4iHsT.js
import { At as call, Ot as ZERO, R as std, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
const filmGrainColor = tgpu.fn([
	d.vec4f,
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, uv, viewport, strength, bias, animTime) => {
	"use gpu";
	const pixelCoord = uv.mul(viewport);
	const noiseInput = __tsover_add(std.dot(pixelCoord, d.vec2f(12.9898, 78.233)), animTime);
	const noise = std.fract(__tsover_mul(std.sin(noiseInput), 43758.5453));
	const grain = __tsover_sub(__tsover_mul(noise, 2), 1);
	const brightness = std.clamp(std.dot(color.xyz, d.vec3f(.2126, .7152, .0722)), 0, 1);
	const darkness = __tsover_sub(1, brightness);
	const darkFactor = std.pow(__tsover_add(darkness, 1e-6), bias);
	const grainIntensity = __tsover_mul(__tsover_mul(grain, darkFactor), __tsover_mul(strength, .1));
	const grainedColor = color.xyz.add(grainIntensity);
	return d.vec4f(grainedColor, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "color"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "bias"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"pixelCoord",
				[
					6,
					[
						7,
						"uv",
						"mul"
					],
					["viewport"]
				]
			],
			[
				13,
				"noiseInput",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"dot"
						],
						["pixelCoord", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "12.9898"], [5, "78.233"]]
						]]
					],
					"+",
					"animTime"
				]
			],
			[
				13,
				"noise",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							["noiseInput"]
						],
						"*",
						[5, "43758.5453"]
					]]
				]
			],
			[
				13,
				"grain",
				[
					1,
					[
						1,
						"noise",
						"*",
						[5, "2"]
					],
					"-",
					[5, "1"]
				]
			],
			[
				13,
				"brightness",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							[[
								7,
								"color",
								"xyz"
							], [
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[5, "0.2126"],
									[5, "0.7152"],
									[5, "0.0722"]
								]
							]]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"darkness",
				[
					1,
					[5, "1"],
					"-",
					"brightness"
				]
			],
			[
				13,
				"darkFactor",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						1,
						"darkness",
						"+",
						[5, "0.000001"]
					], "bias"]
				]
			],
			[
				13,
				"grainIntensity",
				[
					1,
					[
						1,
						"grain",
						"*",
						"darkFactor"
					],
					"*",
					[
						1,
						"strength",
						"*",
						[5, "0.1"]
					]
				]
			],
			[
				13,
				"grainedColor",
				[
					6,
					[
						7,
						[
							7,
							"color",
							"xyz"
						],
						"add"
					],
					["grainIntensity"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				["grainedColor", [
					7,
					"color",
					"w"
				]]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "FilmGrain",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Analog film grain texture overlay, weighted toward darker areas",
	requiresChild: true,
	extraFields: { animTime: {
		schema: d.f32,
		initial: 0
	} },
	props: {
		strength: {
			default: .5,
			description: "Intensity of the film grain noise",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .025,
				label: "Strength",
				group: "Effect"
			}
		},
		bias: {
			default: 2,
			description: "Concentrates grain in darker areas. Higher values focus grain more heavily on shadows; 0 applies grain uniformly.",
			ui: {
				type: "range",
				min: 0,
				max: 10,
				step: .1,
				label: "Bias",
				group: "Effect"
			}
		},
		animated: {
			default: false,
			description: "When enabled, the grain pattern changes each frame for a dynamic film effect",
			ui: {
				type: "checkbox",
				label: "Animated",
				group: "Effect"
			},
			transform: (value) => value ? 1 : 0
		}
	},
	fragment: (params) => {
		const { childNode, uniforms, ctx } = params;
		if (!childNode) return ZERO;
		let animTimeAcc = 0;
		params.onBeforeRender(({ deltaTime }) => {
			const on = params.getCpuValue("animated") > 0 ? 1 : 0;
			animTimeAcc += deltaTime * on * 10;
			params.setExtraField("animTime", animTimeAcc);
		});
		return call(filmGrainColor, "filmGrainColor", [
			childNode,
			ctx.uv,
			ctx.viewportSize,
			uniforms.strength,
			uniforms.bias,
			uniforms.animTime
		]);
	}
};
var FilmGrain_default = componentDefinition;
export { componentDefinition as n, filmGrainColor as r, FilmGrain_default as t };
