// shaders/dist/core/Flip-Dw194RY6.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { i as transformBoolean } from "./transformations-fqzznt4g.js";
const flipUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, flipX, flipY) => {
	"use gpu";
	const x = std.select(uv.x, __tsover_sub(1, uv.x), flipX > 0);
	const y = std.select(uv.y, __tsover_sub(1, uv.y), flipY > 0);
	return d.vec2f(x, y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "flipX"
			},
			{
				"type": "i",
				"name": "flipY"
			}
		],
		"body": [0, [
			[
				13,
				"x",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							7,
							"uv",
							"x"
						],
						[
							1,
							[5, "1"],
							"-",
							[
								7,
								"uv",
								"x"
							]
						],
						[
							1,
							"flipX",
							">",
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"y",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							7,
							"uv",
							"y"
						],
						[
							1,
							[5, "1"],
							"-",
							[
								7,
								"uv",
								"y"
							]
						],
						[
							1,
							"flipY",
							">",
							[5, "0"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["x", "y"]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Flip",
	category: "Distortions",
	description: "Mirror content horizontally, vertically, or both",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		flipX: {
			default: false,
			transform: transformBoolean,
			description: "Mirror the content horizontally (left ↔ right)",
			ui: {
				type: "checkbox",
				label: "Flip X",
				group: "Effect"
			}
		},
		flipY: {
			default: false,
			transform: transformBoolean,
			description: "Mirror the content vertically (top ↔ bottom)",
			ui: {
				type: "checkbox",
				label: "Flip Y",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, convertToTexture }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Flip shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const flipped = call(flipUV, "flipUV", [
			ctx.uv,
			uniforms.flipX,
			uniforms.flipY
		]);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [texture.sample(flipped)]);
	},
	uvRemap: ({ uv, mask, uniforms }) => {
		return {
			uv: call(flipUV, "flipUV", [
				uv,
				uniforms.flipX,
				uniforms.flipY
			]),
			mask
		};
	}
};
var Flip_default = componentDefinition;
export { componentDefinition as n, flipUV as r, Flip_default as t };
