// shaders/dist/core/Flower-Cdg8fOzZ.js
import { At as call, E as shapeLocalCoords, It as vec4, O as strokeMaskFromSdf, Qt as mixColorsVariants, T as sdf_exports, Zt as mixColorsLinear, n as tgpu, on as d, p as ShapeMask, y as flowerSdf } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition, l as transformStrokePosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
const flowerShape = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32
], ShapeMask)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, radius, sides, innerRatio, rotation, softness, strokeThickness, strokePosition, uv, aspect) => {
	"use gpu";
	const local = shapeLocalCoords(center, rotation, uv, aspect);
	return strokeMaskFromSdf(flowerSdf(local.x, local.y, radius, sides, innerRatio), softness, strokeThickness, strokePosition);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "radius"
			},
			{
				"type": "i",
				"name": "sides"
			},
			{
				"type": "i",
				"name": "innerRatio"
			},
			{
				"type": "i",
				"name": "rotation"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "strokeThickness"
			},
			{
				"type": "i",
				"name": "strokePosition"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"local",
				[
					6,
					[
						7,
						"sdf",
						"shapeLocalCoords"
					],
					[
						"center",
						"rotation",
						"uv",
						"aspect"
					]
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"sdf",
						"flowerSdf"
					],
					[
						[
							7,
							"local",
							"x"
						],
						[
							7,
							"local",
							"y"
						],
						"radius",
						"sides",
						"innerRatio"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"sdf",
					"strokeMaskFromSdf"
				],
				[
					"dist",
					"softness",
					"strokeThickness",
					"strokePosition"
				]
			]]
		]],
		"externalNames": ["sdf"]
	},
	externals: () => ({ sdf: sdf_exports })
}) && $.f)({}));
const componentDefinition = {
	name: "Flower",
	boundingBoxDeclaration: { propBindings: {
		x: {
			prop: "center",
			as: "position-x"
		},
		y: {
			prop: "center",
			as: "position-y"
		},
		width: {
			prop: "radius",
			as: "half-canvas-height"
		},
		height: {
			prop: "radius",
			as: "half-canvas-height"
		},
		rotation: {
			prop: "rotation",
			as: "degrees"
		}
	} },
	category: "Shapes",
	description: "Petal shape with N lobes and adjustable inner-to-outer radius ratio",
	acceptsUVContext: true,
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		color: {
			default: "#ffffff",
			transform: transformColor,
			description: "Fill color of the flower",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the flower",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		radius: {
			default: .4,
			description: "Outer petal tip radius in UV space",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Radius",
				group: "Shape"
			}
		},
		sides: {
			default: 5,
			description: "Number of petals",
			ui: {
				type: "range",
				min: 3,
				max: 12,
				step: 1,
				label: "Petals",
				group: "Shape"
			}
		},
		innerRatio: {
			default: .4,
			description: "Inner valley radius as a ratio of outer radius — lower values make deeper notches",
			ui: {
				type: "range",
				min: .1,
				max: .95,
				step: .01,
				label: "Inner Ratio",
				group: "Shape"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Shape"
			}
		},
		softness: {
			default: 0,
			description: "Edge softness for antialiasing",
			ui: {
				type: "range",
				min: 0,
				max: .1,
				step: .001,
				label: "Softness",
				group: "Effect"
			}
		},
		strokeThickness: {
			default: 0,
			description: "Stroke thickness. Zero means no stroke.",
			ui: {
				type: "range",
				min: 0,
				max: .2,
				step: .005,
				label: "Stroke Thickness",
				group: "Stroke"
			}
		},
		strokeColor: {
			default: "#000000",
			transform: transformColor,
			description: "Color of the stroke outline",
			ui: {
				type: "color",
				label: "Stroke Color",
				group: "Colors"
			}
		},
		strokePosition: {
			default: "center",
			transform: transformStrokePosition,
			description: "Position of the stroke relative to the shape edge",
			ui: {
				type: "select",
				options: [
					{
						label: "Outside",
						value: "outside"
					},
					{
						label: "Center",
						value: "center"
					},
					{
						label: "Inside",
						value: "inside"
					}
				],
				label: "Stroke Position",
				group: "Stroke"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for blending fill and stroke colors",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Blending",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const uv = params.uvContext ?? ctx.uv;
		const colorSpaceMode = propValues.colorSpace ?? 0;
		const mask = call(flowerShape, "flowerShape", [
			uniforms.center,
			uniforms.radius,
			uniforms.sides,
			uniforms.innerRatio,
			uniforms.rotation,
			uniforms.softness,
			uniforms.strokeThickness,
			uniforms.strokePosition,
			uv,
			ctx.aspect
		]);
		const blended = call(mixColorsVariants[colorSpaceMode] ?? mixColorsLinear, "mixColors", [
			uniforms.color,
			uniforms.strokeColor,
			mask.member("strokeBlend")
		]);
		return vec4(blended.member("rgb"), blended.member("a").mul(mask.member("overallMask")));
	}
};
var Flower_default = componentDefinition;
export { componentDefinition as n, flowerShape as r, Flower_default as t };
