// shaders/dist/core/FlowingGradient-DdFrloZC.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d, rt as noise_exports, tt as mxNoiseFloat3 } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
var FOUR_PI = 4 * Math.PI;
const flowingFactors = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((baseUV, viewport, distortion, seed, animTime) => {
	"use gpu";
	const t = __tsover_mul(animTime, .075);
	const aspect = __tsover_div(viewport.x, viewport.y);
	const uv0 = baseUV.sub(.5);
	const seedOff = __tsover_mul(seed, .17);
	const noiseVal = mxNoiseFloat3(d.vec3f(__tsover_add(__tsover_mul(uv0.x, .8), __tsover_mul(t, .4)), __tsover_sub(__tsover_mul(uv0.y, .8), __tsover_mul(t, .25)), seedOff));
	const angle = __tsover_mul(__tsover_mul(noiseVal, FOUR_PI), distortion);
	const uvCorr = d.vec2f(uv0.x, __tsover_div(uv0.y, aspect));
	const ca = std.cos(angle);
	const sa = std.sin(angle);
	const rotX = __tsover_sub(__tsover_mul(uvCorr.x, ca), __tsover_mul(uvCorr.y, sa));
	const rotY = __tsover_mul(__tsover_add(__tsover_mul(uvCorr.x, sa), __tsover_mul(uvCorr.y, ca)), aspect);
	const ws = __tsover_mul(t, 2);
	const warpX = __tsover_add(rotX, __tsover_mul(__tsover_div(std.sin(__tsover_add(__tsover_mul(rotY, 5), ws)), 50), distortion));
	const warpY = __tsover_add(rotY, __tsover_mul(__tsover_div(std.sin(__tsover_add(__tsover_mul(warpX, 7.5), ws)), 25), distortion));
	const wp1 = d.vec3f(__tsover_mul(warpX, 2), __tsover_mul(warpY, 2), __tsover_add(__tsover_mul(t, .5), seedOff));
	const wx1 = mxNoiseFloat3(wp1.add(d.vec3f(5.2, 1.3, 0)));
	const wy1 = mxNoiseFloat3(wp1.add(d.vec3f(1.7, 9.2, 0)));
	const dw1X = __tsover_add(warpX, __tsover_mul(__tsover_mul(wx1, .25), distortion));
	const dw1Y = __tsover_add(warpY, __tsover_mul(__tsover_mul(wy1, .25), distortion));
	const wp2 = d.vec3f(__tsover_mul(dw1X, 1.5), __tsover_mul(dw1Y, 1.5), __tsover_add(__tsover_mul(t, .3), seedOff));
	const wx2 = mxNoiseFloat3(wp2.add(d.vec3f(8.3, 2.8, 0)));
	const wy2 = mxNoiseFloat3(wp2.add(d.vec3f(3.1, 7.4, 0)));
	const dw2X = __tsover_add(dw1X, __tsover_mul(__tsover_mul(wx2, .2), distortion));
	const dw2Y = __tsover_add(dw1Y, __tsover_mul(__tsover_mul(wy2, .2), distortion));
	const zoneNoise = mxNoiseFloat3(d.vec3f(__tsover_mul(dw2X, 1.8), __tsover_mul(dw2Y, 1.8), __tsover_add(__tsover_mul(t, .3), seedOff)));
	const edgeHardness = std.pow(__tsover_add(__tsover_mul(wx1, .5), .5), 1.5);
	const edgeWidth = std.mix(.5, .008, edgeHardness);
	const t1 = std.smoothstep(__tsover_mul(edgeWidth, -1), edgeWidth, zoneNoise);
	const t2 = std.smoothstep(__tsover_mul(edgeWidth, -1), edgeWidth, wy1);
	const foldMask1 = __tsover_sub(1, std.abs(__tsover_sub(__tsover_mul(t1, 2), 1)));
	const foldMask2 = __tsover_sub(1, std.abs(__tsover_sub(__tsover_mul(t2, 2), 1)));
	const foldHighlight = __tsover_mul(std.max(foldMask1, foldMask2), edgeHardness);
	const surfaceLight = __tsover_add(__tsover_mul(__tsover_add(__tsover_mul(wx2, .5), .5), .12), .94);
	const litScale = __tsover_mul(surfaceLight, __tsover_add(1, __tsover_mul(foldHighlight, .35)));
	return d.vec3f(t1, t2, litScale);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "baseUV"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "distortion"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"t",
				[
					1,
					"animTime",
					"*",
					[5, "0.075"]
				]
			],
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"uv0",
				[
					6,
					[
						7,
						"baseUV",
						"sub"
					],
					[[5, "0.5"]]
				]
			],
			[
				13,
				"seedOff",
				[
					1,
					"seed",
					"*",
					[5, "0.17"]
				]
			],
			[
				13,
				"noiseVal",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								[
									1,
									[
										7,
										"uv0",
										"x"
									],
									"*",
									[5, "0.8"]
								],
								"+",
								[
									1,
									"t",
									"*",
									[5, "0.4"]
								]
							],
							[
								1,
								[
									1,
									[
										7,
										"uv0",
										"y"
									],
									"*",
									[5, "0.8"]
								],
								"-",
								[
									1,
									"t",
									"*",
									[5, "0.25"]
								]
							],
							"seedOff"
						]
					]]
				]
			],
			[
				13,
				"angle",
				[
					1,
					[
						1,
						"noiseVal",
						"*",
						"FOUR_PI"
					],
					"*",
					"distortion"
				]
			],
			[
				13,
				"uvCorr",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"uv0",
						"x"
					], [
						1,
						[
							7,
							"uv0",
							"y"
						],
						"/",
						"aspect"
					]]
				]
			],
			[
				13,
				"ca",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angle"]
				]
			],
			[
				13,
				"sa",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angle"]
				]
			],
			[
				13,
				"rotX",
				[
					1,
					[
						1,
						[
							7,
							"uvCorr",
							"x"
						],
						"*",
						"ca"
					],
					"-",
					[
						1,
						[
							7,
							"uvCorr",
							"y"
						],
						"*",
						"sa"
					]
				]
			],
			[
				13,
				"rotY",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"uvCorr",
								"x"
							],
							"*",
							"sa"
						],
						"+",
						[
							1,
							[
								7,
								"uvCorr",
								"y"
							],
							"*",
							"ca"
						]
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"ws",
				[
					1,
					"t",
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"warpX",
				[
					1,
					"rotX",
					"+",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										1,
										"rotY",
										"*",
										[5, "5"]
									],
									"+",
									"ws"
								]]
							],
							"/",
							[5, "50"]
						],
						"*",
						"distortion"
					]
				]
			],
			[
				13,
				"warpY",
				[
					1,
					"rotY",
					"+",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										1,
										"warpX",
										"*",
										[5, "7.5"]
									],
									"+",
									"ws"
								]]
							],
							"/",
							[5, "25"]
						],
						"*",
						"distortion"
					]
				]
			],
			[
				13,
				"wp1",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							"warpX",
							"*",
							[5, "2"]
						],
						[
							1,
							"warpY",
							"*",
							[5, "2"]
						],
						[
							1,
							[
								1,
								"t",
								"*",
								[5, "0.5"]
							],
							"+",
							"seedOff"
						]
					]
				]
			],
			[
				13,
				"wx1",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					[[
						6,
						[
							7,
							"wp1",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "5.2"],
								[5, "1.3"],
								[5, "0"]
							]
						]]
					]]
				]
			],
			[
				13,
				"wy1",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					[[
						6,
						[
							7,
							"wp1",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1.7"],
								[5, "9.2"],
								[5, "0"]
							]
						]]
					]]
				]
			],
			[
				13,
				"dw1X",
				[
					1,
					"warpX",
					"+",
					[
						1,
						[
							1,
							"wx1",
							"*",
							[5, "0.25"]
						],
						"*",
						"distortion"
					]
				]
			],
			[
				13,
				"dw1Y",
				[
					1,
					"warpY",
					"+",
					[
						1,
						[
							1,
							"wy1",
							"*",
							[5, "0.25"]
						],
						"*",
						"distortion"
					]
				]
			],
			[
				13,
				"wp2",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							"dw1X",
							"*",
							[5, "1.5"]
						],
						[
							1,
							"dw1Y",
							"*",
							[5, "1.5"]
						],
						[
							1,
							[
								1,
								"t",
								"*",
								[5, "0.3"]
							],
							"+",
							"seedOff"
						]
					]
				]
			],
			[
				13,
				"wx2",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					[[
						6,
						[
							7,
							"wp2",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "8.3"],
								[5, "2.8"],
								[5, "0"]
							]
						]]
					]]
				]
			],
			[
				13,
				"wy2",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					[[
						6,
						[
							7,
							"wp2",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "3.1"],
								[5, "7.4"],
								[5, "0"]
							]
						]]
					]]
				]
			],
			[
				13,
				"dw2X",
				[
					1,
					"dw1X",
					"+",
					[
						1,
						[
							1,
							"wx2",
							"*",
							[5, "0.2"]
						],
						"*",
						"distortion"
					]
				]
			],
			[
				13,
				"dw2Y",
				[
					1,
					"dw1Y",
					"+",
					[
						1,
						[
							1,
							"wy2",
							"*",
							[5, "0.2"]
						],
						"*",
						"distortion"
					]
				]
			],
			[
				13,
				"zoneNoise",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								"dw2X",
								"*",
								[5, "1.8"]
							],
							[
								1,
								"dw2Y",
								"*",
								[5, "1.8"]
							],
							[
								1,
								[
									1,
									"t",
									"*",
									[5, "0.3"]
								],
								"+",
								"seedOff"
							]
						]
					]]
				]
			],
			[
				13,
				"edgeHardness",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						1,
						[
							1,
							"wx1",
							"*",
							[5, "0.5"]
						],
						"+",
						[5, "0.5"]
					], [5, "1.5"]]
				]
			],
			[
				13,
				"edgeWidth",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "0.5"],
						[5, "0.008"],
						"edgeHardness"
					]
				]
			],
			[
				13,
				"t1",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"edgeWidth",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"edgeWidth",
						"zoneNoise"
					]
				]
			],
			[
				13,
				"t2",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"edgeWidth",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"edgeWidth",
						"wy1"
					]
				]
			],
			[
				13,
				"foldMask1",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							1,
							[
								1,
								"t1",
								"*",
								[5, "2"]
							],
							"-",
							[5, "1"]
						]]
					]
				]
			],
			[
				13,
				"foldMask2",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							1,
							[
								1,
								"t2",
								"*",
								[5, "2"]
							],
							"-",
							[5, "1"]
						]]
					]
				]
			],
			[
				13,
				"foldHighlight",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"max"
						],
						["foldMask1", "foldMask2"]
					],
					"*",
					"edgeHardness"
				]
			],
			[
				13,
				"surfaceLight",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								"wx2",
								"*",
								[5, "0.5"]
							],
							"+",
							[5, "0.5"]
						],
						"*",
						[5, "0.12"]
					],
					"+",
					[5, "0.94"]
				]
			],
			[
				13,
				"litScale",
				[
					1,
					"surfaceLight",
					"*",
					[
						1,
						[5, "1"],
						"+",
						[
							1,
							"foldHighlight",
							"*",
							[5, "0.35"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					"t1",
					"t2",
					"litScale"
				]
			]]
		]],
		"externalNames": [
			"noise",
			"d",
			"FOUR_PI",
			"std"
		]
	},
	externals: () => ({
		noise: noise_exports,
		d,
		FOUR_PI,
		std
	})
}) && $.f)({}));
const flowingCompose = tgpu.fn([d.vec4f, d.f32], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((base, litScale) => {
	"use gpu";
	const lit = std.clamp(base.xyz.mul(litScale), d.vec3f(0), d.vec3f(1));
	return d.vec4f(lit, base.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "base"
		}, {
			"type": "i",
			"name": "litScale"
		}],
		"body": [0, [[
			13,
			"lit",
			[
				6,
				[
					7,
					"std",
					"clamp"
				],
				[
					[
						6,
						[
							7,
							[
								7,
								"base",
								"xyz"
							],
							"mul"
						],
						["litScale"]
					],
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[[5, "0"]]
					],
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[[5, "1"]]
					]
				]
			]
		], [10, [
			6,
			[
				7,
				"d",
				"vec4f"
			],
			["lit", [
				7,
				"base",
				"w"
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "FlowingGradient",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Liquid silk gradient with organic flowing color bands",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#0a0015",
			transform: transformColor,
			description: "Deep background color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#6b17e6",
			transform: transformColor,
			description: "Primary accent color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		colorC: {
			default: "#ff4d6a",
			transform: transformColor,
			description: "Secondary accent color",
			ui: {
				type: "color",
				label: "Color C",
				group: "Colors"
			}
		},
		colorD: {
			default: "#ff6b35",
			transform: transformColor,
			description: "Tertiary accent color",
			ui: {
				type: "color",
				label: "Color D",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "oklch",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 10,
				step: .1,
				label: "Speed",
				group: "Effect"
			}
		},
		distortion: {
			default: .5,
			description: "Organic distortion intensity",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "Distortion",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for variation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const uv = params.uvContext ?? ctx.uv;
		const viewport = params.effectiveViewportSize ?? ctx.viewportSize;
		const t = animatedTime(params);
		const f = call(flowingFactors, "flowingFactors", [
			uv,
			viewport,
			uniforms.distortion,
			uniforms.seed,
			t
		]);
		const t1 = f.member("x");
		const t2 = f.member("y");
		const litScale = f.member("z");
		const variant = mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear;
		return call(flowingCompose, "flowingCompose", [call(variant, "mixColors", [
			call(variant, "mixColors", [
				uniforms.colorA,
				uniforms.colorB,
				t1
			]),
			call(variant, "mixColors", [
				uniforms.colorC,
				uniforms.colorD,
				t1
			]),
			t2
		]), litScale]);
	}
};
var FlowingGradient_default = componentDefinition;
export { flowingFactors as i, componentDefinition as n, flowingCompose as r, FlowingGradient_default as t };
