// shaders/dist/core/FlowField-CwU9iWbf.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, rt as noise_exports, tt as mxNoiseFloat3, xt as composeEdgeRemapExpr } from "./coords-CWOT0Wqh.js";
import { s as transformEdges } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
var TWO_PI = Math.PI * 2;
const flowFieldUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, scale, strength, time, evolutionTime) => {
	"use gpu";
	const aspectCorrectedUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const drift1 = __tsover_mul(time, .7);
	const evolve1 = __tsover_mul(evolutionTime, .2);
	const angle1 = __tsover_mul(mxNoiseFloat3(d.vec3f(__tsover_add(__tsover_mul(aspectCorrectedUV.x, scale), __tsover_mul(drift1, .3)), __tsover_add(__tsover_mul(aspectCorrectedUV.y, scale), __tsover_mul(drift1, .5)), evolve1)), TWO_PI);
	const flow1 = d.vec2f(std.cos(angle1), std.sin(angle1));
	const drift2 = __tsover_mul(time, .5);
	const evolve2 = __tsover_mul(evolutionTime, .15);
	const angle2 = __tsover_mul(mxNoiseFloat3(d.vec3f(__tsover_add(__tsover_add(__tsover_mul(aspectCorrectedUV.x, __tsover_mul(scale, 2.3)), __tsover_mul(drift2, -.4)), 173.5), __tsover_add(__tsover_add(__tsover_mul(aspectCorrectedUV.y, __tsover_mul(scale, 2.3)), __tsover_mul(drift2, .3)), 291.7), __tsover_add(evolve2, 50))), TWO_PI);
	const flow2 = d.vec2f(std.cos(angle2), std.sin(angle2));
	const drift3 = __tsover_mul(time, .3);
	const evolve3 = __tsover_mul(evolutionTime, .1);
	const angle3 = __tsover_mul(mxNoiseFloat3(d.vec3f(__tsover_add(__tsover_add(__tsover_mul(aspectCorrectedUV.x, __tsover_mul(scale, 4.7)), __tsover_mul(drift3, .2)), 527.3), __tsover_add(__tsover_add(__tsover_mul(aspectCorrectedUV.y, __tsover_mul(scale, 4.7)), __tsover_mul(drift3, -.6)), 839.1), __tsover_add(evolve3, 100))), TWO_PI);
	const flow3 = d.vec2f(std.cos(angle3), std.sin(angle3));
	const combinedFlow = flow1.add(flow2.mul(.5)).add(flow3.mul(.25));
	const flowLength = std.max(std.length(combinedFlow), .001);
	const normalizedFlow = combinedFlow.div(flowLength);
	const finalFlow = d.vec2f(__tsover_div(__tsover_mul(normalizedFlow.x, strength), aspect), __tsover_mul(normalizedFlow.y, strength));
	return d.vec2f(__tsover_add(uv.x, finalFlow.x), __tsover_add(uv.y, finalFlow.y));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "time"
			},
			{
				"type": "i",
				"name": "evolutionTime"
			}
		],
		"body": [0, [
			[
				13,
				"aspectCorrectedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"drift1",
				[
					1,
					"time",
					"*",
					[5, "0.7"]
				]
			],
			[
				13,
				"evolve1",
				[
					1,
					"evolutionTime",
					"*",
					[5, "0.2"]
				]
			],
			[
				13,
				"angle1",
				[
					1,
					[
						6,
						[
							7,
							"noise",
							"mxNoiseFloat3"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									[
										1,
										[
											7,
											"aspectCorrectedUV",
											"x"
										],
										"*",
										"scale"
									],
									"+",
									[
										1,
										"drift1",
										"*",
										[5, "0.3"]
									]
								],
								[
									1,
									[
										1,
										[
											7,
											"aspectCorrectedUV",
											"y"
										],
										"*",
										"scale"
									],
									"+",
									[
										1,
										"drift1",
										"*",
										[5, "0.5"]
									]
								],
								"evolve1"
							]
						]]
					],
					"*",
					"TWO_PI"
				]
			],
			[
				13,
				"flow1",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"cos"
						],
						["angle1"]
					], [
						6,
						[
							7,
							"std",
							"sin"
						],
						["angle1"]
					]]
				]
			],
			[
				13,
				"drift2",
				[
					1,
					"time",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"evolve2",
				[
					1,
					"evolutionTime",
					"*",
					[5, "0.15"]
				]
			],
			[
				13,
				"angle2",
				[
					1,
					[
						6,
						[
							7,
							"noise",
							"mxNoiseFloat3"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"aspectCorrectedUV",
												"x"
											],
											"*",
											[
												1,
												"scale",
												"*",
												[5, "2.3"]
											]
										],
										"+",
										[
											1,
											"drift2",
											"*",
											[
												4,
												"-",
												[5, "0.4"]
											]
										]
									],
									"+",
									[5, "173.5"]
								],
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"aspectCorrectedUV",
												"y"
											],
											"*",
											[
												1,
												"scale",
												"*",
												[5, "2.3"]
											]
										],
										"+",
										[
											1,
											"drift2",
											"*",
											[5, "0.3"]
										]
									],
									"+",
									[5, "291.7"]
								],
								[
									1,
									"evolve2",
									"+",
									[5, "50"]
								]
							]
						]]
					],
					"*",
					"TWO_PI"
				]
			],
			[
				13,
				"flow2",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"cos"
						],
						["angle2"]
					], [
						6,
						[
							7,
							"std",
							"sin"
						],
						["angle2"]
					]]
				]
			],
			[
				13,
				"drift3",
				[
					1,
					"time",
					"*",
					[5, "0.3"]
				]
			],
			[
				13,
				"evolve3",
				[
					1,
					"evolutionTime",
					"*",
					[5, "0.1"]
				]
			],
			[
				13,
				"angle3",
				[
					1,
					[
						6,
						[
							7,
							"noise",
							"mxNoiseFloat3"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"aspectCorrectedUV",
												"x"
											],
											"*",
											[
												1,
												"scale",
												"*",
												[5, "4.7"]
											]
										],
										"+",
										[
											1,
											"drift3",
											"*",
											[5, "0.2"]
										]
									],
									"+",
									[5, "527.3"]
								],
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"aspectCorrectedUV",
												"y"
											],
											"*",
											[
												1,
												"scale",
												"*",
												[5, "4.7"]
											]
										],
										"+",
										[
											1,
											"drift3",
											"*",
											[
												4,
												"-",
												[5, "0.6"]
											]
										]
									],
									"+",
									[5, "839.1"]
								],
								[
									1,
									"evolve3",
									"+",
									[5, "100"]
								]
							]
						]]
					],
					"*",
					"TWO_PI"
				]
			],
			[
				13,
				"flow3",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"cos"
						],
						["angle3"]
					], [
						6,
						[
							7,
							"std",
							"sin"
						],
						["angle3"]
					]]
				]
			],
			[
				13,
				"combinedFlow",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"flow1",
								"add"
							],
							[[
								6,
								[
									7,
									"flow2",
									"mul"
								],
								[[5, "0.5"]]
							]]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"flow3",
							"mul"
						],
						[[5, "0.25"]]
					]]
				]
			],
			[
				13,
				"flowLength",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"length"
						],
						["combinedFlow"]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"normalizedFlow",
				[
					6,
					[
						7,
						"combinedFlow",
						"div"
					],
					["flowLength"]
				]
			],
			[
				13,
				"finalFlow",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								7,
								"normalizedFlow",
								"x"
							],
							"*",
							"strength"
						],
						"/",
						"aspect"
					], [
						1,
						[
							7,
							"normalizedFlow",
							"y"
						],
						"*",
						"strength"
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						7,
						"uv",
						"x"
					],
					"+",
					[
						7,
						"finalFlow",
						"x"
					]
				], [
					1,
					[
						7,
						"uv",
						"y"
					],
					"+",
					[
						7,
						"finalFlow",
						"y"
					]
				]]
			]]
		]],
		"externalNames": [
			"d",
			"noise",
			"TWO_PI",
			"std"
		]
	},
	externals: () => ({
		d,
		noise: noise_exports,
		TWO_PI,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "FlowField",
	category: "Distortions",
	description: "Fluid-like distortion with constant smooth motion",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	animatedTime: { speed: "speed" },
	extraAnimatedTimes: { evolution: "evolutionSpeed" },
	props: {
		strength: {
			default: .15,
			description: "Intensity of the flow distortion",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .5,
				step: .01,
				label: "Strength",
				group: "Effect"
			}
		},
		detail: {
			default: 2,
			description: "Scale of the flow patterns",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 5,
				step: .1,
				label: "Detail",
				group: "Effect"
			}
		},
		speed: {
			default: 0,
			description: "Speed of the flow",
			ui: {
				type: "range",
				min: 0,
				max: 20,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		evolutionSpeed: {
			default: 0,
			description: "How fast the flow field pattern reshapes over time",
			ui: {
				type: "range",
				min: 0,
				max: 20,
				step: .1,
				label: "Evolution Speed",
				group: "Animation"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for flow pattern variation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, childNode, ctx, propValues, convertToTexture } = params;
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		const time = animatedTime(params, "seed").mul(.1);
		const evolutionTime = animatedTime(params, "seed", "_animTime_evolution").mul(.1);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(flowFieldUV, "flowFieldUV", [
			ctx.uv,
			ctx.aspect,
			uniforms.detail,
			uniforms.strength,
			time,
			evolutionTime
		]), (uv) => texture.sample(uv), edgeMode)]);
	},
	uvRemap: (ctx) => {
		const edgeMode = ctx.propValues.edges ?? 0;
		const time = animatedTime(ctx, "seed").mul(.1);
		const evolutionTime = animatedTime(ctx, "seed", "_animTime_evolution").mul(.1);
		return composeEdgeRemapExpr(call(flowFieldUV, "flowFieldUV", [
			ctx.uv,
			ctx.aspect,
			ctx.uniforms.detail,
			ctx.uniforms.strength,
			time,
			evolutionTime
		]), ctx.mask, edgeMode);
	}
};
var FlowField_default = componentDefinition;
export { componentDefinition as n, flowFieldUV as r, FlowField_default as t };
