// shaders/dist/core/FlutedGlass-B9652Bh4.js
import { At as call, Ct as edgeMirrorUV, It as vec4, Nt as expr, Ot as ZERO, R as std, Rt as formatFloat, St as edgeClampUV, Tt as edgeWrapUV, Wt as unpremultiplyAlpha, n as tgpu, on as d, wt as edgeTransparentMask } from "./coords-CWOT0Wqh.js";
import { a as transformColor, s as transformEdges } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
var PI = Math.PI;
const FlutedGeom = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
	refractedUV: d.vec2f,
	chrOff: d.vec2f,
	spec: d.f32
}), "FlutedGeom");
const flutedGlassGeom = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec3f
], FlutedGeom)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, t, angleDeg, frequency, softness, waveAmp, waveFreq, refraction, aberration, lightAngleDeg, highlight, highlightSoftness, shapeParams) => {
	"use gpu";
	const expHi = shapeParams.x;
	const expLo = shapeParams.y;
	const wavesFlag = shapeParams.z;
	const r = __tsover_div(__tsover_mul(angleDeg, PI), 180);
	const cosA = std.cos(r);
	const sinA = std.sin(r);
	const aspCorr = d.vec2f(__tsover_mul(__tsover_sub(uv.x, .5), aspect), __tsover_sub(uv.y, .5));
	const u = __tsover_add(__tsover_mul(aspCorr.x, cosA), __tsover_mul(aspCorr.y, sinA));
	const v = __tsover_add(__tsover_mul(-aspCorr.x, sinA), __tsover_mul(aspCorr.y, cosA));
	const waveTerm = __tsover_mul(std.sin(__tsover_add(__tsover_mul(__tsover_mul(v, waveFreq), __tsover_mul(PI, 2)), __tsover_mul(t, 2))), waveAmp);
	const fluteU = __tsover_add(u, __tsover_mul(wavesFlag, waveTerm));
	const flutePos = __tsover_add(__tsover_mul(fluteU, frequency), t);
	const cellPos = __tsover_mul(__tsover_sub(std.fract(flutePos), .5), 2);
	const absCell = std.max(std.abs(cellPos), 1e-4);
	const exponent = std.mix(expHi, expLo, softness);
	const slope = __tsover_mul(std.sign(cellPos), std.pow(absCell, exponent));
	const halfCell = __tsover_div(.5, std.max(frequency, .001));
	const refrU = -__tsover_mul(__tsover_mul(slope, refraction), halfCell);
	const refractedUV = d.vec2f(__tsover_add(uv.x, __tsover_div(__tsover_mul(refrU, cosA), aspect)), __tsover_add(uv.y, __tsover_mul(refrU, sinA)));
	const chrU = __tsover_mul(__tsover_mul(refrU, aberration), .5);
	const chrOff = d.vec2f(__tsover_div(__tsover_mul(chrU, cosA), aspect), __tsover_mul(chrU, sinA));
	const slopeSq = std.min(__tsover_mul(slope, slope), 1);
	const nz = std.sqrt(__tsover_sub(1, slopeSq));
	const half = __tsover_div(__tsover_mul(lightAngleDeg, PI), 360);
	const hx = std.sin(half);
	const hy = std.cos(half);
	const nDotH = std.max(__tsover_add(__tsover_mul(slope, hx), __tsover_mul(nz, hy)), 0);
	const shininess = std.exp2(__tsover_sub(8, __tsover_mul(highlightSoftness, 7)));
	const fresnel = std.pow(__tsover_sub(1, nz), 5);
	const fresnelMix = __tsover_add(.04, __tsover_mul(.96, fresnel));
	return FlutedGeom({
		refractedUV,
		chrOff,
		spec: __tsover_mul(__tsover_mul(std.pow(nDotH, shininess), fresnelMix), highlight)
	});
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "t"
			},
			{
				"type": "i",
				"name": "angleDeg"
			},
			{
				"type": "i",
				"name": "frequency"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "waveAmp"
			},
			{
				"type": "i",
				"name": "waveFreq"
			},
			{
				"type": "i",
				"name": "refraction"
			},
			{
				"type": "i",
				"name": "aberration"
			},
			{
				"type": "i",
				"name": "lightAngleDeg"
			},
			{
				"type": "i",
				"name": "highlight"
			},
			{
				"type": "i",
				"name": "highlightSoftness"
			},
			{
				"type": "i",
				"name": "shapeParams"
			}
		],
		"body": [0, [
			[
				13,
				"expHi",
				[
					7,
					"shapeParams",
					"x"
				]
			],
			[
				13,
				"expLo",
				[
					7,
					"shapeParams",
					"y"
				]
			],
			[
				13,
				"wavesFlag",
				[
					7,
					"shapeParams",
					"z"
				]
			],
			[
				13,
				"r",
				[
					1,
					[
						1,
						"angleDeg",
						"*",
						"PI"
					],
					"/",
					[5, "180"]
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["r"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["r"]
				]
			],
			[
				13,
				"aspCorr",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"-",
							[5, "0.5"]
						],
						"*",
						"aspect"
					], [
						1,
						[
							7,
							"uv",
							"y"
						],
						"-",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"u",
				[
					1,
					[
						1,
						[
							7,
							"aspCorr",
							"x"
						],
						"*",
						"cosA"
					],
					"+",
					[
						1,
						[
							7,
							"aspCorr",
							"y"
						],
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"v",
				[
					1,
					[
						1,
						[
							4,
							"-",
							[
								7,
								"aspCorr",
								"x"
							]
						],
						"*",
						"sinA"
					],
					"+",
					[
						1,
						[
							7,
							"aspCorr",
							"y"
						],
						"*",
						"cosA"
					]
				]
			],
			[
				13,
				"waveTerm",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						[[
							1,
							[
								1,
								[
									1,
									"v",
									"*",
									"waveFreq"
								],
								"*",
								[
									1,
									"PI",
									"*",
									[5, "2"]
								]
							],
							"+",
							[
								1,
								"t",
								"*",
								[5, "2"]
							]
						]]
					],
					"*",
					"waveAmp"
				]
			],
			[
				13,
				"fluteU",
				[
					1,
					"u",
					"+",
					[
						1,
						"wavesFlag",
						"*",
						"waveTerm"
					]
				]
			],
			[
				13,
				"flutePos",
				[
					1,
					[
						1,
						"fluteU",
						"*",
						"frequency"
					],
					"+",
					"t"
				]
			],
			[
				13,
				"cellPos",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							["flutePos"]
						],
						"-",
						[5, "0.5"]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"absCell",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"abs"
						],
						["cellPos"]
					], [5, "0.0001"]]
				]
			],
			[
				13,
				"exponent",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"expHi",
						"expLo",
						"softness"
					]
				]
			],
			[
				13,
				"slope",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"sign"
						],
						["cellPos"]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"pow"
						],
						["absCell", "exponent"]
					]
				]
			],
			[
				13,
				"halfCell",
				[
					1,
					[5, "0.5"],
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						["frequency", [5, "0.001"]]
					]
				]
			],
			[
				13,
				"refrU",
				[
					4,
					"-",
					[
						1,
						[
							1,
							"slope",
							"*",
							"refraction"
						],
						"*",
						"halfCell"
					]
				]
			],
			[
				13,
				"refractedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"+",
						[
							1,
							[
								1,
								"refrU",
								"*",
								"cosA"
							],
							"/",
							"aspect"
						]
					], [
						1,
						[
							7,
							"uv",
							"y"
						],
						"+",
						[
							1,
							"refrU",
							"*",
							"sinA"
						]
					]]
				]
			],
			[
				13,
				"chrU",
				[
					1,
					[
						1,
						"refrU",
						"*",
						"aberration"
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"chrOff",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							"chrU",
							"*",
							"cosA"
						],
						"/",
						"aspect"
					], [
						1,
						"chrU",
						"*",
						"sinA"
					]]
				]
			],
			[
				13,
				"slopeSq",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					[[
						1,
						"slope",
						"*",
						"slope"
					], [5, "1"]]
				]
			],
			[
				13,
				"nz",
				[
					6,
					[
						7,
						"std",
						"sqrt"
					],
					[[
						1,
						[5, "1"],
						"-",
						"slopeSq"
					]]
				]
			],
			[
				13,
				"half",
				[
					1,
					[
						1,
						"lightAngleDeg",
						"*",
						"PI"
					],
					"/",
					[5, "360"]
				]
			],
			[
				13,
				"hx",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["half"]
				]
			],
			[
				13,
				"hy",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["half"]
				]
			],
			[
				13,
				"nDotH",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[
							1,
							"slope",
							"*",
							"hx"
						],
						"+",
						[
							1,
							"nz",
							"*",
							"hy"
						]
					], [5, "0"]]
				]
			],
			[
				13,
				"shininess",
				[
					6,
					[
						7,
						"std",
						"exp2"
					],
					[[
						1,
						[5, "8"],
						"-",
						[
							1,
							"highlightSoftness",
							"*",
							[5, "7"]
						]
					]]
				]
			],
			[
				13,
				"fresnel",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						1,
						[5, "1"],
						"-",
						"nz"
					], [5, "5"]]
				]
			],
			[
				13,
				"fresnelMix",
				[
					1,
					[5, "0.04"],
					"+",
					[
						1,
						[5, "0.96"],
						"*",
						"fresnel"
					]
				]
			],
			[
				13,
				"spec",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							["nDotH", "shininess"]
						],
						"*",
						"fresnelMix"
					],
					"*",
					"highlight"
				]
			],
			[10, [
				6,
				"FlutedGeom",
				[[104, {
					"refractedUV": "refractedUV",
					"chrOff": "chrOff",
					"spec": "spec"
				}]]
			]]
		]],
		"externalNames": [
			"PI",
			"std",
			"d",
			"FlutedGeom"
		]
	},
	externals: () => ({
		PI,
		std,
		d,
		FlutedGeom
	})
}) && $.f)({}));
const componentDefinition = {
	name: "FlutedGlass",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Distortions",
	description: "Full-screen fluted glass effect — refracts content through repeating cylindrical bars",
	requiresRTT: true,
	requiresChild: true,
	animatedTime: { speed: "speed" },
	props: {
		shape: {
			default: "bars",
			description: "Cross-section shape of each flute",
			compileTime: true,
			transform: (value) => ({
				"bars": 0,
				"rounded": 1,
				"waves": 2
			})[value] ?? 0,
			ui: {
				type: "select",
				options: [
					{
						label: "Bars",
						value: "bars"
					},
					{
						label: "Rounded",
						value: "rounded"
					},
					{
						label: "Waves",
						value: "waves"
					}
				],
				label: "Shape",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			description: "Direction of the flutes in degrees (0 = vertical bars)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		frequency: {
			default: 10,
			description: "Number of flutes across the longest viewport axis",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 20,
				step: 1,
				label: "Frequency",
				group: "Effect"
			}
		},
		softness: {
			default: .5,
			description: "How smoothly distortion fades from each flute centre to its edge (0 = flat middle / sharp seams, 1 = gentle curve)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		waveAmplitude: {
			default: .06,
			description: "How far each flute sways horizontally as it travels (Waves shape only)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .5,
				step: .01,
				label: "Wave Amplitude",
				group: "Effect",
				condition: { shape: "waves" }
			}
		},
		waveFrequency: {
			default: 1.5,
			description: "How many sways fit along each flute (Waves shape only)",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 10,
				step: .1,
				label: "Wave Frequency",
				group: "Effect",
				condition: { shape: "waves" }
			}
		},
		speed: {
			default: 0,
			description: "Animation speed — drifts the flute pattern over time and flows wave perturbations",
			ui: {
				type: "range",
				min: -1,
				max: 1,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		refraction: {
			default: 1.5,
			description: "How aggressively each flute bends content beneath it",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 4,
				step: .01,
				label: "Refraction",
				group: "Effect"
			}
		},
		aberration: {
			default: .2,
			description: "Chromatic aberration — splits RGB along the refraction direction at flute seams",
			compileTimeWhen: (prev, next) => prev > 0 !== next > 0,
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Aberration",
				group: "Effect"
			}
		},
		lightAngle: {
			default: 30,
			description: "Direction the light source is coming from (0 = head-on, 90 = grazing)",
			ui: {
				type: "range",
				min: -90,
				max: 90,
				step: 1,
				label: "Light Angle",
				group: "Highlight"
			}
		},
		highlight: {
			default: .2,
			description: "Strength of the specular reflection on each flute",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .01,
				label: "Highlight",
				group: "Highlight"
			}
		},
		highlightSoftness: {
			default: .3,
			description: "Spread of the specular peak (0 = pin-tight, 1 = broad sheen)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Highlight Softness",
				group: "Highlight"
			}
		},
		highlightColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color of the specular highlight",
			ui: {
				type: "color",
				label: "Highlight Color",
				group: "Highlight"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when distortion samples beyond the canvas",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, childNode, ctx, propValues, convertToTexture } = params;
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const shapeId = propValues.shape ?? 0;
		const expHi = shapeId === 0 ? 16 : 8;
		const expLo = shapeId === 0 ? 4 : 3;
		const wavesFlag = shapeId === 2 ? 1 : 0;
		const edgeMode = propValues.edges ?? 2;
		const aberrationEnabled = (propValues.aberration ?? .2) > 0;
		const t = animatedTime(params).mul(-1);
		const geom = call(flutedGlassGeom, "flutedGlassGeom", [
			ctx.uv,
			ctx.aspect,
			t,
			uniforms.angle,
			uniforms.frequency,
			uniforms.softness,
			uniforms.waveAmplitude,
			uniforms.waveFrequency,
			uniforms.refraction,
			uniforms.aberration,
			uniforms.lightAngle,
			uniforms.highlight,
			uniforms.highlightSoftness,
			expr(`vec3f(${formatFloat(expHi)}, ${formatFloat(expLo)}, ${formatFloat(wavesFlag)})`)
		]);
		const sampleAt = (uvE) => {
			if (edgeMode === 1) return texture.sample(uvE).mul(call(edgeTransparentMask, "edgeTransparentMask", [uvE]));
			if (edgeMode === 2) return texture.sample(call(edgeMirrorUV, "edgeMirrorUV", [uvE]));
			if (edgeMode === 3) return texture.sample(call(edgeWrapUV, "edgeWrapUV", [uvE]));
			return texture.sample(call(edgeClampUV, "edgeClampUV", [uvE]));
		};
		const refractedUV = geom.member("refractedUV");
		let sampled;
		if (!aberrationEnabled) sampled = sampleAt(refractedUV);
		else {
			const chrOff = geom.member("chrOff");
			const rS = sampleAt(refractedUV.add(chrOff));
			const gS = sampleAt(refractedUV);
			const bS = sampleAt(refractedUV.add(chrOff.mul(-1)));
			sampled = vec4(rS.member("r"), gS.member("g"), bS.member("b"), gS.member("a"));
		}
		const straight = call(unpremultiplyAlpha, "unpremultiplyAlpha", [sampled]);
		const specA = geom.member("spec").mul(straight.member("a"));
		return vec4(straight.member("rgb").add(uniforms.highlightColor.member("rgb").mul(specA)), straight.member("a"));
	}
};
var FlutedGlass_default = componentDefinition;
export { flutedGlassGeom as i, FlutedGlass_default as n, componentDefinition as r, FlutedGeom as t };
