// shaders/dist/core/Fog-ClrsfQfj.js
import { At as call, B as createGuardedCompute, Gt as colorMixing_exports, It as vec4, Ot as ZERO, R as std, V as createStateBuffer, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
var N = 256;
var COUNT = N * N;
var JACOBI_ITERS = 10;
var WARM_JACOBI_ITERS = 6;
var WARM_STEPS = 50;
var PRESSURE_DISSIPATION = .8;
var VELOCITY_DISSIPATION = .15;
var INTERNAL_MAX_VEL = N * .6;
var TAU = Math.PI * 2;
var STATE_FORMAT = "rgba16float";
var cm = colorMixing_exports;
var FogParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
	dt: d.f32,
	time: d.f32,
	seed: d.f32,
	turbulence: d.f32,
	curlStrength: d.f32,
	blending: d.f32,
	cursorX: d.f32,
	cursorY: d.f32,
	cursorVelX: d.f32,
	cursorVelY: d.f32,
	mouseActive: d.f32,
	mouseRadSq: d.f32
}), "FogParams");
var fluidLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	velA: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "mutable"
	},
	velB: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "mutable"
	},
	densA: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "mutable"
	},
	densB: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "mutable"
	},
	pressure: {
		storage: d.arrayOf(d.f32, COUNT),
		access: "mutable"
	},
	divergence: {
		storage: d.arrayOf(d.f32, COUNT),
		access: "mutable"
	},
	params: { uniform: FogParams }
}), "fluidLayout");
var outputLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	densA: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "readonly"
	},
	outTex: { storageTexture: d.textureStorage2d(STATE_FORMAT, "write-only") }
}), "outputLayout");
var nw = tgpu.fn([
	d.i32,
	d.i32,
	d.i32,
	d.i32
], d.u32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((ii, ji, di, dj) => {
	"use gpu";
	const row = __tsover_mod(__tsover_add(__tsover_add(ii, di), N), N);
	const col = __tsover_mod(__tsover_add(__tsover_add(ji, dj), N), N);
	return d.u32(__tsover_add(__tsover_mul(row, N), col));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "ii"
			},
			{
				"type": "i",
				"name": "ji"
			},
			{
				"type": "i",
				"name": "di"
			},
			{
				"type": "i",
				"name": "dj"
			}
		],
		"body": [0, [
			[
				13,
				"row",
				[
					1,
					[
						1,
						[
							1,
							"ii",
							"+",
							"di"
						],
						"+",
						"N"
					],
					"%",
					"N"
				]
			],
			[
				13,
				"col",
				[
					1,
					[
						1,
						[
							1,
							"ji",
							"+",
							"dj"
						],
						"+",
						"N"
					],
					"%",
					"N"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"u32"
				],
				[[
					1,
					[
						1,
						"row",
						"*",
						"N"
					],
					"+",
					"col"
				]]
			]]
		]],
		"externalNames": ["N", "d"]
	},
	externals: () => ({
		N,
		d
	})
}) && $.f)({}));
var hashFog = tgpu.fn([d.f32, d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((px, py) => {
	"use gpu";
	return std.fract(__tsover_mul(std.sin(__tsover_add(__tsover_mul(px, 127.1), __tsover_mul(py, 311.7))), 43758.5453));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "px"
		}, {
			"type": "i",
			"name": "py"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				1,
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						1,
						[
							1,
							"px",
							"*",
							[5, "127.1"]
						],
						"+",
						[
							1,
							"py",
							"*",
							[5, "311.7"]
						]
					]]
				],
				"*",
				[5, "43758.5453"]
			]]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
var valueNoise2D = tgpu.fn([d.f32, d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((nx, ny) => {
	"use gpu";
	const ix = std.floor(nx);
	const iy = std.floor(ny);
	const sx = std.fract(nx);
	const sy = std.fract(ny);
	const ux = __tsover_mul(__tsover_mul(sx, sx), __tsover_sub(3, __tsover_mul(sx, 2)));
	const uy = __tsover_mul(__tsover_mul(sy, sy), __tsover_sub(3, __tsover_mul(sy, 2)));
	return std.mix(std.mix(hashFog(ix, iy), hashFog(__tsover_add(ix, 1), iy), ux), std.mix(hashFog(ix, __tsover_add(iy, 1)), hashFog(__tsover_add(ix, 1), __tsover_add(iy, 1)), ux), uy);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "nx"
		}, {
			"type": "i",
			"name": "ny"
		}],
		"body": [0, [
			[
				13,
				"ix",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["nx"]
				]
			],
			[
				13,
				"iy",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["ny"]
				]
			],
			[
				13,
				"sx",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["nx"]
				]
			],
			[
				13,
				"sy",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["ny"]
				]
			],
			[
				13,
				"ux",
				[
					1,
					[
						1,
						"sx",
						"*",
						"sx"
					],
					"*",
					[
						1,
						[5, "3"],
						"-",
						[
							1,
							"sx",
							"*",
							[5, "2"]
						]
					]
				]
			],
			[
				13,
				"uy",
				[
					1,
					[
						1,
						"sy",
						"*",
						"sy"
					],
					"*",
					[
						1,
						[5, "3"],
						"-",
						[
							1,
							"sy",
							"*",
							[5, "2"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"mix"
				],
				[
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							[
								6,
								"hashFog",
								["ix", "iy"]
							],
							[
								6,
								"hashFog",
								[[
									1,
									"ix",
									"+",
									[5, "1"]
								], "iy"]
							],
							"ux"
						]
					],
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							[
								6,
								"hashFog",
								["ix", [
									1,
									"iy",
									"+",
									[5, "1"]
								]]
							],
							[
								6,
								"hashFog",
								[[
									1,
									"ix",
									"+",
									[5, "1"]
								], [
									1,
									"iy",
									"+",
									[5, "1"]
								]]
							],
							"ux"
						]
					],
					"uy"
				]
			]]
		]],
		"externalNames": ["std", "hashFog"]
	},
	externals: () => ({
		std,
		hashFog
	})
}) && $.f)({}));
var fbm = tgpu.fn([d.f32, d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((nx, ny) => {
	"use gpu";
	return __tsover_add(__tsover_add(__tsover_mul(valueNoise2D(nx, ny), .5), __tsover_mul(valueNoise2D(__tsover_add(__tsover_mul(nx, 2), 5.37), __tsover_add(__tsover_mul(ny, 2), 9.13)), .3)), __tsover_mul(valueNoise2D(__tsover_add(__tsover_mul(nx, 4), 1.79), __tsover_add(__tsover_mul(ny, 4), 3.51)), .2));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "nx"
		}, {
			"type": "i",
			"name": "ny"
		}],
		"body": [0, [[10, [
			1,
			[
				1,
				[
					1,
					[
						6,
						"valueNoise2D",
						["nx", "ny"]
					],
					"*",
					[5, "0.5"]
				],
				"+",
				[
					1,
					[
						6,
						"valueNoise2D",
						[[
							1,
							[
								1,
								"nx",
								"*",
								[5, "2"]
							],
							"+",
							[5, "5.37"]
						], [
							1,
							[
								1,
								"ny",
								"*",
								[5, "2"]
							],
							"+",
							[5, "9.13"]
						]]
					],
					"*",
					[5, "0.3"]
				]
			],
			"+",
			[
				1,
				[
					6,
					"valueNoise2D",
					[[
						1,
						[
							1,
							"nx",
							"*",
							[5, "4"]
						],
						"+",
						[5, "1.79"]
					], [
						1,
						[
							1,
							"ny",
							"*",
							[5, "4"]
						],
						"+",
						[5, "3.51"]
					]]
				],
				"*",
				[5, "0.2"]
			]
		]]]],
		"externalNames": ["valueNoise2D"]
	},
	externals: () => ({ valueNoise2D })
}) && $.f)({}));
const initKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = fluidLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const jf = d.f32(cx);
	const iF = d.f32(cy);
	const sdx = __tsover_mul(std.fract(__tsover_mul(std.sin(__tsover_mul(p.seed, 37.31)), 43758.5453)), 97);
	const sdy = __tsover_mul(std.fract(__tsover_mul(std.sin(__tsover_mul(p.seed, 83.17)), 43758.5453)), 97);
	const nx = __tsover_add(__tsover_mul(__tsover_div(jf, d.f32(N)), 3), sdx);
	const ny = __tsover_add(__tsover_mul(__tsover_div(iF, d.f32(N)), 3), sdy);
	const fogDens = std.clamp(__tsover_add(__tsover_mul(fbm(__tsover_add(nx, 31.7), __tsover_add(ny, 53.3)), .75), .25), .15, .9);
	const colorVar = std.clamp(fbm(nx, ny), 0, 1);
	fluidLayout.$.densA[idx] = d.vec4f(fogDens, colorVar, 0, 0);
	fluidLayout.$.densB[idx] = d.vec4f(fogDens, colorVar, 0, 0);
	fluidLayout.$.velA[idx] = d.vec4f(0, 0, 0, 0);
	fluidLayout.$.velB[idx] = d.vec4f(0, 0, 0, 0);
	fluidLayout.$.pressure[idx] = 0;
	fluidLayout.$.divergence[idx] = 0;
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"fluidLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"jf",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					["cx"]
				]
			],
			[
				13,
				"iF",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					["cy"]
				]
			],
			[
				13,
				"sdx",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										7,
										"p",
										"seed"
									],
									"*",
									[5, "37.31"]
								]]
							],
							"*",
							[5, "43758.5453"]
						]]
					],
					"*",
					[5, "97"]
				]
			],
			[
				13,
				"sdy",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										7,
										"p",
										"seed"
									],
									"*",
									[5, "83.17"]
								]]
							],
							"*",
							[5, "43758.5453"]
						]]
					],
					"*",
					[5, "97"]
				]
			],
			[
				13,
				"nx",
				[
					1,
					[
						1,
						[
							1,
							"jf",
							"/",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["N"]
							]
						],
						"*",
						[5, "3"]
					],
					"+",
					"sdx"
				]
			],
			[
				13,
				"ny",
				[
					1,
					[
						1,
						[
							1,
							"iF",
							"/",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["N"]
							]
						],
						"*",
						[5, "3"]
					],
					"+",
					"sdy"
				]
			],
			[
				13,
				"fogDens",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									6,
									"fbm",
									[[
										1,
										"nx",
										"+",
										[5, "31.7"]
									], [
										1,
										"ny",
										"+",
										[5, "53.3"]
									]]
								],
								"*",
								[5, "0.75"]
							],
							"+",
							[5, "0.25"]
						],
						[5, "0.15"],
						[5, "0.9"]
					]
				]
			],
			[
				13,
				"colorVar",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							"fbm",
							["nx", "ny"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						"fogDens",
						"colorVar",
						[5, "0"],
						[5, "0"]
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densB"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						"fogDens",
						"colorVar",
						[5, "0"],
						[5, "0"]
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "0"],
						[5, "0"]
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velB"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "0"],
						[5, "0"]
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					"idx"
				],
				"=",
				[5, "0"]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"divergence"
					],
					"idx"
				],
				"=",
				[5, "0"]
			]
		]],
		"externalNames": [
			"fluidLayout",
			"d",
			"N",
			"std",
			"fbm"
		]
	},
	externals: () => ({
		fluidLayout,
		d,
		N,
		std,
		fbm
	})
}) && $.f)({})).$name("fogInit");
const forceKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = fluidLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const jf = __tsover_add(d.f32(cx), .5);
	const iF = __tsover_add(d.f32(cy), .5);
	const x = __tsover_mul(__tsover_div(jf, d.f32(N)), TAU);
	const y = __tsover_mul(__tsover_div(iF, d.f32(N)), TAU);
	const t = p.time;
	const vel = fluidLayout.$.velA[idx];
	const fx = __tsover_mul(__tsover_mul(__tsover_mul(__tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_mul(x, 2), __tsover_mul(t, .31))), std.cos(__tsover_add(__tsover_mul(y, 1.5), __tsover_mul(t, .23)))), __tsover_mul(std.sin(__tsover_add(__tsover_add(__tsover_mul(x, .7), __tsover_mul(y, 1.3)), __tsover_mul(t, .17))), .4)), p.turbulence), __tsover_mul(N, .04)), p.dt);
	const fy = __tsover_mul(__tsover_mul(__tsover_mul(__tsover_add(__tsover_mul(std.cos(__tsover_add(__tsover_mul(x, 1.5), __tsover_mul(t, .27))), std.sin(__tsover_add(__tsover_mul(y, 2), __tsover_mul(t, .19)))), __tsover_mul(std.cos(__tsover_add(__tsover_add(__tsover_mul(x, 1.1), __tsover_mul(y, .9)), __tsover_mul(t, .21))), .4)), p.turbulence), __tsover_mul(N, .04)), p.dt);
	const cdx = __tsover_sub(jf, p.cursorX);
	const cdy = __tsover_sub(iF, p.cursorY);
	const cursorDistSq = __tsover_add(__tsover_mul(cdx, cdx), __tsover_mul(cdy, cdy));
	const cursorInf = __tsover_mul(std.exp(__tsover_mul(__tsover_div(cursorDistSq, std.max(p.mouseRadSq, d.f32(1))), -1)), p.mouseActive);
	fluidLayout.$.velA[idx] = d.vec4f(__tsover_add(__tsover_add(vel.x, fx), __tsover_mul(p.cursorVelX, cursorInf)), __tsover_add(__tsover_add(vel.y, fy), __tsover_mul(p.cursorVelY, cursorInf)), vel.z, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"fluidLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"jf",
				[
					1,
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["cx"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"iF",
				[
					1,
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["cy"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"x",
				[
					1,
					[
						1,
						"jf",
						"/",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["N"]
						]
					],
					"*",
					"TAU"
				]
			],
			[
				13,
				"y",
				[
					1,
					[
						1,
						"iF",
						"/",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["N"]
						]
					],
					"*",
					"TAU"
				]
			],
			[
				13,
				"t",
				[
					7,
					"p",
					"time"
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				13,
				"fx",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									[
										6,
										[
											7,
											"std",
											"sin"
										],
										[[
											1,
											[
												1,
												"x",
												"*",
												[5, "2"]
											],
											"+",
											[
												1,
												"t",
												"*",
												[5, "0.31"]
											]
										]]
									],
									"*",
									[
										6,
										[
											7,
											"std",
											"cos"
										],
										[[
											1,
											[
												1,
												"y",
												"*",
												[5, "1.5"]
											],
											"+",
											[
												1,
												"t",
												"*",
												[5, "0.23"]
											]
										]]
									]
								],
								"+",
								[
									1,
									[
										6,
										[
											7,
											"std",
											"sin"
										],
										[[
											1,
											[
												1,
												[
													1,
													"x",
													"*",
													[5, "0.7"]
												],
												"+",
												[
													1,
													"y",
													"*",
													[5, "1.3"]
												]
											],
											"+",
											[
												1,
												"t",
												"*",
												[5, "0.17"]
											]
										]]
									],
									"*",
									[5, "0.4"]
								]
							],
							"*",
							[
								7,
								"p",
								"turbulence"
							]
						],
						"*",
						[
							1,
							"N",
							"*",
							[5, "0.04"]
						]
					],
					"*",
					[
						7,
						"p",
						"dt"
					]
				]
			],
			[
				13,
				"fy",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									[
										6,
										[
											7,
											"std",
											"cos"
										],
										[[
											1,
											[
												1,
												"x",
												"*",
												[5, "1.5"]
											],
											"+",
											[
												1,
												"t",
												"*",
												[5, "0.27"]
											]
										]]
									],
									"*",
									[
										6,
										[
											7,
											"std",
											"sin"
										],
										[[
											1,
											[
												1,
												"y",
												"*",
												[5, "2"]
											],
											"+",
											[
												1,
												"t",
												"*",
												[5, "0.19"]
											]
										]]
									]
								],
								"+",
								[
									1,
									[
										6,
										[
											7,
											"std",
											"cos"
										],
										[[
											1,
											[
												1,
												[
													1,
													"x",
													"*",
													[5, "1.1"]
												],
												"+",
												[
													1,
													"y",
													"*",
													[5, "0.9"]
												]
											],
											"+",
											[
												1,
												"t",
												"*",
												[5, "0.21"]
											]
										]]
									],
									"*",
									[5, "0.4"]
								]
							],
							"*",
							[
								7,
								"p",
								"turbulence"
							]
						],
						"*",
						[
							1,
							"N",
							"*",
							[5, "0.04"]
						]
					],
					"*",
					[
						7,
						"p",
						"dt"
					]
				]
			],
			[
				13,
				"cdx",
				[
					1,
					"jf",
					"-",
					[
						7,
						"p",
						"cursorX"
					]
				]
			],
			[
				13,
				"cdy",
				[
					1,
					"iF",
					"-",
					[
						7,
						"p",
						"cursorY"
					]
				]
			],
			[
				13,
				"cursorDistSq",
				[
					1,
					[
						1,
						"cdx",
						"*",
						"cdx"
					],
					"+",
					[
						1,
						"cdy",
						"*",
						"cdy"
					]
				]
			],
			[
				13,
				"cursorInf",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"exp"
						],
						[[
							1,
							[
								1,
								"cursorDistSq",
								"/",
								[
									6,
									[
										7,
										"std",
										"max"
									],
									[[
										7,
										"p",
										"mouseRadSq"
									], [
										6,
										[
											7,
											"d",
											"f32"
										],
										[[5, "1"]]
									]]
								]
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						]]
					],
					"*",
					[
						7,
						"p",
						"mouseActive"
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"vel",
									"x"
								],
								"+",
								"fx"
							],
							"+",
							[
								1,
								[
									7,
									"p",
									"cursorVelX"
								],
								"*",
								"cursorInf"
							]
						],
						[
							1,
							[
								1,
								[
									7,
									"vel",
									"y"
								],
								"+",
								"fy"
							],
							"+",
							[
								1,
								[
									7,
									"p",
									"cursorVelY"
								],
								"*",
								"cursorInf"
							]
						],
						[
							7,
							"vel",
							"z"
						],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"fluidLayout",
			"d",
			"N",
			"TAU",
			"std"
		]
	},
	externals: () => ({
		fluidLayout,
		d,
		N,
		TAU,
		std
	})
}) && $.f)({})).$name("fogForce");
const curlKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const ii = d.i32(cy);
	const ji = d.i32(cx);
	const vL = fluidLayout.$.velA[nw(ii, ji, 0, -1)];
	const vR = fluidLayout.$.velA[nw(ii, ji, 0, 1)];
	const vU = fluidLayout.$.velA[nw(ii, ji, -1, 0)];
	const vD = fluidLayout.$.velA[nw(ii, ji, 1, 0)];
	const curl = __tsover_mul(__tsover_sub(__tsover_sub(vR.y, vL.y), __tsover_sub(vD.x, vU.x)), .5);
	const vel = fluidLayout.$.velA[idx];
	fluidLayout.$.velA[idx] = d.vec4f(vel.x, vel.y, curl, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"ii",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cy"]
				]
			],
			[
				13,
				"ji",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cx"]
				]
			],
			[
				13,
				"vL",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[5, "0"],
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]
				]
			],
			[
				13,
				"vR",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[5, "0"],
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"vU",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[
								4,
								"-",
								[5, "1"]
							],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"vD",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[5, "1"],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"curl",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"vR",
								"y"
							],
							"-",
							[
								7,
								"vL",
								"y"
							]
						],
						"-",
						[
							1,
							[
								7,
								"vD",
								"x"
							],
							"-",
							[
								7,
								"vU",
								"x"
							]
						]
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							7,
							"vel",
							"x"
						],
						[
							7,
							"vel",
							"y"
						],
						"curl",
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout",
			"nw"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout,
		nw
	})
}) && $.f)({})).$name("fogCurl");
const vorticityKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = fluidLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const ii = d.i32(cy);
	const ji = d.i32(cx);
	const cL = std.abs(fluidLayout.$.velA[nw(ii, ji, 0, -1)].z);
	const cR = std.abs(fluidLayout.$.velA[nw(ii, ji, 0, 1)].z);
	const cU = std.abs(fluidLayout.$.velA[nw(ii, ji, -1, 0)].z);
	const cD = std.abs(fluidLayout.$.velA[nw(ii, ji, 1, 0)].z);
	const cc = fluidLayout.$.velA[idx].z;
	const gx = __tsover_mul(__tsover_sub(cR, cL), .5);
	const gy = __tsover_mul(__tsover_sub(cD, cU), .5);
	const gl = std.max(std.sqrt(__tsover_add(__tsover_mul(gx, gx), __tsover_mul(gy, gy))), d.f32(1e-5));
	const vel = fluidLayout.$.velA[idx];
	fluidLayout.$.velA[idx] = d.vec4f(__tsover_add(vel.x, __tsover_mul(__tsover_mul(__tsover_mul(__tsover_div(gy, gl), cc), p.curlStrength), p.dt)), __tsover_add(vel.y, __tsover_mul(__tsover_mul(__tsover_mul(__tsover_mul(__tsover_div(gx, gl), -1), cc), p.curlStrength), p.dt)), vel.z, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"fluidLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"ii",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cy"]
				]
			],
			[
				13,
				"ji",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cx"]
				]
			],
			[
				13,
				"cL",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						7,
						[
							8,
							[
								7,
								[
									7,
									"fluidLayout",
									"$"
								],
								"velA"
							],
							[
								6,
								"nw",
								[
									"ii",
									"ji",
									[5, "0"],
									[
										4,
										"-",
										[5, "1"]
									]
								]
							]
						],
						"z"
					]]
				]
			],
			[
				13,
				"cR",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						7,
						[
							8,
							[
								7,
								[
									7,
									"fluidLayout",
									"$"
								],
								"velA"
							],
							[
								6,
								"nw",
								[
									"ii",
									"ji",
									[5, "0"],
									[5, "1"]
								]
							]
						],
						"z"
					]]
				]
			],
			[
				13,
				"cU",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						7,
						[
							8,
							[
								7,
								[
									7,
									"fluidLayout",
									"$"
								],
								"velA"
							],
							[
								6,
								"nw",
								[
									"ii",
									"ji",
									[
										4,
										"-",
										[5, "1"]
									],
									[5, "0"]
								]
							]
						],
						"z"
					]]
				]
			],
			[
				13,
				"cD",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						7,
						[
							8,
							[
								7,
								[
									7,
									"fluidLayout",
									"$"
								],
								"velA"
							],
							[
								6,
								"nw",
								[
									"ii",
									"ji",
									[5, "1"],
									[5, "0"]
								]
							]
						],
						"z"
					]]
				]
			],
			[
				13,
				"cc",
				[
					7,
					[
						8,
						[
							7,
							[
								7,
								"fluidLayout",
								"$"
							],
							"velA"
						],
						"idx"
					],
					"z"
				]
			],
			[
				13,
				"gx",
				[
					1,
					[
						1,
						"cR",
						"-",
						"cL"
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"gy",
				[
					1,
					[
						1,
						"cD",
						"-",
						"cU"
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"gl",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"gx",
								"*",
								"gx"
							],
							"+",
							[
								1,
								"gy",
								"*",
								"gy"
							]
						]]
					], [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0.00001"]]
					]]
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							1,
							[
								7,
								"vel",
								"x"
							],
							"+",
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											"gy",
											"/",
											"gl"
										],
										"*",
										"cc"
									],
									"*",
									[
										7,
										"p",
										"curlStrength"
									]
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						[
							1,
							[
								7,
								"vel",
								"y"
							],
							"+",
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											[
												1,
												"gx",
												"/",
												"gl"
											],
											"*",
											[
												4,
												"-",
												[5, "1"]
											]
										],
										"*",
										"cc"
									],
									"*",
									[
										7,
										"p",
										"curlStrength"
									]
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						[
							7,
							"vel",
							"z"
						],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"fluidLayout",
			"d",
			"N",
			"std",
			"nw"
		]
	},
	externals: () => ({
		fluidLayout,
		d,
		N,
		std,
		nw
	})
}) && $.f)({})).$name("fogVorticity");
const divergenceKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const ii = d.i32(cy);
	const ji = d.i32(cx);
	const vL = fluidLayout.$.velA[nw(ii, ji, 0, -1)].x;
	const vR = fluidLayout.$.velA[nw(ii, ji, 0, 1)].x;
	const vU = fluidLayout.$.velA[nw(ii, ji, -1, 0)].y;
	const vD = fluidLayout.$.velA[nw(ii, ji, 1, 0)].y;
	const div = __tsover_mul(__tsover_add(__tsover_sub(vR, vL), __tsover_sub(vD, vU)), .5);
	fluidLayout.$.pressure[idx] = __tsover_mul(fluidLayout.$.pressure[idx], PRESSURE_DISSIPATION);
	fluidLayout.$.divergence[idx] = div;
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"ii",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cy"]
				]
			],
			[
				13,
				"ji",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cx"]
				]
			],
			[
				13,
				"vL",
				[
					7,
					[
						8,
						[
							7,
							[
								7,
								"fluidLayout",
								"$"
							],
							"velA"
						],
						[
							6,
							"nw",
							[
								"ii",
								"ji",
								[5, "0"],
								[
									4,
									"-",
									[5, "1"]
								]
							]
						]
					],
					"x"
				]
			],
			[
				13,
				"vR",
				[
					7,
					[
						8,
						[
							7,
							[
								7,
								"fluidLayout",
								"$"
							],
							"velA"
						],
						[
							6,
							"nw",
							[
								"ii",
								"ji",
								[5, "0"],
								[5, "1"]
							]
						]
					],
					"x"
				]
			],
			[
				13,
				"vU",
				[
					7,
					[
						8,
						[
							7,
							[
								7,
								"fluidLayout",
								"$"
							],
							"velA"
						],
						[
							6,
							"nw",
							[
								"ii",
								"ji",
								[
									4,
									"-",
									[5, "1"]
								],
								[5, "0"]
							]
						]
					],
					"y"
				]
			],
			[
				13,
				"vD",
				[
					7,
					[
						8,
						[
							7,
							[
								7,
								"fluidLayout",
								"$"
							],
							"velA"
						],
						[
							6,
							"nw",
							[
								"ii",
								"ji",
								[5, "1"],
								[5, "0"]
							]
						]
					],
					"y"
				]
			],
			[
				13,
				"div",
				[
					1,
					[
						1,
						[
							1,
							"vR",
							"-",
							"vL"
						],
						"+",
						[
							1,
							"vD",
							"-",
							"vU"
						]
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					"idx"
				],
				"=",
				[
					1,
					[
						8,
						[
							7,
							[
								7,
								"fluidLayout",
								"$"
							],
							"pressure"
						],
						"idx"
					],
					"*",
					"PRESSURE_DISSIPATION"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"divergence"
					],
					"idx"
				],
				"=",
				"div"
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout",
			"nw",
			"PRESSURE_DISSIPATION"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout,
		nw,
		PRESSURE_DISSIPATION
	})
}) && $.f)({})).$name("fogDivergence");
const jacobiKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const ii = d.i32(cy);
	const ji = d.i32(cx);
	const pL = fluidLayout.$.pressure[nw(ii, ji, 0, -1)];
	const pR = fluidLayout.$.pressure[nw(ii, ji, 0, 1)];
	const pU = fluidLayout.$.pressure[nw(ii, ji, -1, 0)];
	const pD = fluidLayout.$.pressure[nw(ii, ji, 1, 0)];
	const div = fluidLayout.$.divergence[idx];
	fluidLayout.$.pressure[idx] = __tsover_mul(__tsover_sub(__tsover_add(__tsover_add(__tsover_add(pL, pR), pU), pD), div), .25);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"ii",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cy"]
				]
			],
			[
				13,
				"ji",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cx"]
				]
			],
			[
				13,
				"pL",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[5, "0"],
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]
				]
			],
			[
				13,
				"pR",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[5, "0"],
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"pU",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[
								4,
								"-",
								[5, "1"]
							],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"pD",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[5, "1"],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"div",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"divergence"
					],
					"idx"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					"idx"
				],
				"=",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									"pL",
									"+",
									"pR"
								],
								"+",
								"pU"
							],
							"+",
							"pD"
						],
						"-",
						"div"
					],
					"*",
					[5, "0.25"]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout",
			"nw"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout,
		nw
	})
}) && $.f)({})).$name("fogJacobi");
const gradSubtractKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const ii = d.i32(cy);
	const ji = d.i32(cx);
	const pL = fluidLayout.$.pressure[nw(ii, ji, 0, -1)];
	const pR = fluidLayout.$.pressure[nw(ii, ji, 0, 1)];
	const pU = fluidLayout.$.pressure[nw(ii, ji, -1, 0)];
	const pD = fluidLayout.$.pressure[nw(ii, ji, 1, 0)];
	const vel = fluidLayout.$.velA[idx];
	const vx = __tsover_sub(vel.x, __tsover_mul(__tsover_sub(pR, pL), .5));
	const vy = __tsover_sub(vel.y, __tsover_mul(__tsover_sub(pD, pU), .5));
	const maxVel = d.f32(INTERNAL_MAX_VEL);
	const vLen = std.max(std.sqrt(__tsover_add(__tsover_mul(vx, vx), __tsover_mul(vy, vy))), d.f32(1e-6));
	const capScale = __tsover_div(maxVel, std.max(vLen, maxVel));
	fluidLayout.$.velA[idx] = d.vec4f(__tsover_mul(vx, capScale), __tsover_mul(vy, capScale), vel.z, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"ii",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cy"]
				]
			],
			[
				13,
				"ji",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cx"]
				]
			],
			[
				13,
				"pL",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[5, "0"],
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]
				]
			],
			[
				13,
				"pR",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[5, "0"],
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"pU",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[
								4,
								"-",
								[5, "1"]
							],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"pD",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nw",
						[
							"ii",
							"ji",
							[5, "1"],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				13,
				"vx",
				[
					1,
					[
						7,
						"vel",
						"x"
					],
					"-",
					[
						1,
						[
							1,
							"pR",
							"-",
							"pL"
						],
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"vy",
				[
					1,
					[
						7,
						"vel",
						"y"
					],
					"-",
					[
						1,
						[
							1,
							"pD",
							"-",
							"pU"
						],
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"maxVel",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					["INTERNAL_MAX_VEL"]
				]
			],
			[
				13,
				"vLen",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"vx",
								"*",
								"vx"
							],
							"+",
							[
								1,
								"vy",
								"*",
								"vy"
							]
						]]
					], [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0.000001"]]
					]]
				]
			],
			[
				13,
				"capScale",
				[
					1,
					"maxVel",
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						["vLen", "maxVel"]
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							1,
							"vx",
							"*",
							"capScale"
						],
						[
							1,
							"vy",
							"*",
							"capScale"
						],
						[
							7,
							"vel",
							"z"
						],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout",
			"nw",
			"INTERNAL_MAX_VEL",
			"std"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout,
		nw,
		INTERNAL_MAX_VEL,
		std
	})
}) && $.f)({})).$name("fogGradSubtract");
const advectVelKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = fluidLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const vel = fluidLayout.$.velA[idx];
	const srcX = __tsover_mod(__tsover_add(__tsover_sub(d.f32(cx), __tsover_mul(vel.x, p.dt)), d.f32(N)), d.f32(N));
	const srcY = __tsover_mod(__tsover_add(__tsover_sub(d.f32(cy), __tsover_mul(vel.y, p.dt)), d.f32(N)), d.f32(N));
	const x0 = std.floor(srcX);
	const y0 = std.floor(srcY);
	const sx = __tsover_sub(srcX, x0);
	const sy = __tsover_sub(srcY, y0);
	const ix0 = d.i32(x0);
	const iy0 = d.i32(y0);
	const ix1 = __tsover_mod(__tsover_add(ix0, 1), N);
	const iy1 = __tsover_mod(__tsover_add(iy0, 1), N);
	const v00 = fluidLayout.$.velA[d.u32(__tsover_add(__tsover_mul(iy0, N), ix0))];
	const v10 = fluidLayout.$.velA[d.u32(__tsover_add(__tsover_mul(iy0, N), ix1))];
	const v01 = fluidLayout.$.velA[d.u32(__tsover_add(__tsover_mul(iy1, N), ix0))];
	const v11 = fluidLayout.$.velA[d.u32(__tsover_add(__tsover_mul(iy1, N), ix1))];
	const decay = __tsover_add(1, __tsover_mul(VELOCITY_DISSIPATION, p.dt));
	fluidLayout.$.velB[idx] = d.vec4f(__tsover_div(std.mix(std.mix(v00.x, v10.x, sx), std.mix(v01.x, v11.x, sx), sy), decay), __tsover_div(std.mix(std.mix(v00.y, v10.y, sx), std.mix(v01.y, v11.y, sx), sy), decay), 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"fluidLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				13,
				"srcX",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cx"]
							],
							"-",
							[
								1,
								[
									7,
									"vel",
									"x"
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						"+",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["N"]
						]
					],
					"%",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["N"]
					]
				]
			],
			[
				13,
				"srcY",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cy"]
							],
							"-",
							[
								1,
								[
									7,
									"vel",
									"y"
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						"+",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["N"]
						]
					],
					"%",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["N"]
					]
				]
			],
			[
				13,
				"x0",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["srcX"]
				]
			],
			[
				13,
				"y0",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["srcY"]
				]
			],
			[
				13,
				"sx",
				[
					1,
					"srcX",
					"-",
					"x0"
				]
			],
			[
				13,
				"sy",
				[
					1,
					"srcY",
					"-",
					"y0"
				]
			],
			[
				13,
				"ix0",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["x0"]
				]
			],
			[
				13,
				"iy0",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["y0"]
				]
			],
			[
				13,
				"ix1",
				[
					1,
					[
						1,
						"ix0",
						"+",
						[5, "1"]
					],
					"%",
					"N"
				]
			],
			[
				13,
				"iy1",
				[
					1,
					[
						1,
						"iy0",
						"+",
						[5, "1"]
					],
					"%",
					"N"
				]
			],
			[
				13,
				"v00",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								"iy0",
								"*",
								"N"
							],
							"+",
							"ix0"
						]]
					]
				]
			],
			[
				13,
				"v10",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								"iy0",
								"*",
								"N"
							],
							"+",
							"ix1"
						]]
					]
				]
			],
			[
				13,
				"v01",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								"iy1",
								"*",
								"N"
							],
							"+",
							"ix0"
						]]
					]
				]
			],
			[
				13,
				"v11",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								"iy1",
								"*",
								"N"
							],
							"+",
							"ix1"
						]]
					]
				]
			],
			[
				13,
				"decay",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						"VELOCITY_DISSIPATION",
						"*",
						[
							7,
							"p",
							"dt"
						]
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velB"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"mix"
								],
								[
									[
										6,
										[
											7,
											"std",
											"mix"
										],
										[
											[
												7,
												"v00",
												"x"
											],
											[
												7,
												"v10",
												"x"
											],
											"sx"
										]
									],
									[
										6,
										[
											7,
											"std",
											"mix"
										],
										[
											[
												7,
												"v01",
												"x"
											],
											[
												7,
												"v11",
												"x"
											],
											"sx"
										]
									],
									"sy"
								]
							],
							"/",
							"decay"
						],
						[
							1,
							[
								6,
								[
									7,
									"std",
									"mix"
								],
								[
									[
										6,
										[
											7,
											"std",
											"mix"
										],
										[
											[
												7,
												"v00",
												"y"
											],
											[
												7,
												"v10",
												"y"
											],
											"sx"
										]
									],
									[
										6,
										[
											7,
											"std",
											"mix"
										],
										[
											[
												7,
												"v01",
												"y"
											],
											[
												7,
												"v11",
												"y"
											],
											"sx"
										]
									],
									"sy"
								]
							],
							"/",
							"decay"
						],
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"fluidLayout",
			"d",
			"N",
			"std",
			"VELOCITY_DISSIPATION"
		]
	},
	externals: () => ({
		fluidLayout,
		d,
		N,
		std,
		VELOCITY_DISSIPATION
	})
}) && $.f)({})).$name("fogAdvectVel");
const copyVelKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const v = fluidLayout.$.velB[idx];
	fluidLayout.$.velA[idx] = d.vec4f(v.x, v.y, 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"v",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velB"
					],
					"idx"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							7,
							"v",
							"x"
						],
						[
							7,
							"v",
							"y"
						],
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout
	})
}) && $.f)({})).$name("fogCopyVel");
const advectDensKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = fluidLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const vel = fluidLayout.$.velA[idx];
	const srcX = __tsover_mod(__tsover_add(__tsover_sub(d.f32(cx), __tsover_mul(vel.x, p.dt)), d.f32(N)), d.f32(N));
	const srcY = __tsover_mod(__tsover_add(__tsover_sub(d.f32(cy), __tsover_mul(vel.y, p.dt)), d.f32(N)), d.f32(N));
	const x0 = std.floor(srcX);
	const y0 = std.floor(srcY);
	const sx = __tsover_sub(srcX, x0);
	const sy = __tsover_sub(srcY, y0);
	const ix0 = d.i32(x0);
	const iy0 = d.i32(y0);
	const ix1 = __tsover_mod(__tsover_add(ix0, 1), N);
	const iy1 = __tsover_mod(__tsover_add(iy0, 1), N);
	const d00 = fluidLayout.$.densA[d.u32(__tsover_add(__tsover_mul(iy0, N), ix0))];
	const d10 = fluidLayout.$.densA[d.u32(__tsover_add(__tsover_mul(iy0, N), ix1))];
	const d01 = fluidLayout.$.densA[d.u32(__tsover_add(__tsover_mul(iy1, N), ix0))];
	const d11 = fluidLayout.$.densA[d.u32(__tsover_add(__tsover_mul(iy1, N), ix1))];
	fluidLayout.$.densB[idx] = d.vec4f(std.mix(std.mix(d00.x, d10.x, sx), std.mix(d01.x, d11.x, sx), sy), std.mix(std.mix(d00.y, d10.y, sx), std.mix(d01.y, d11.y, sx), sy), 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"fluidLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				13,
				"srcX",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cx"]
							],
							"-",
							[
								1,
								[
									7,
									"vel",
									"x"
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						"+",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["N"]
						]
					],
					"%",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["N"]
					]
				]
			],
			[
				13,
				"srcY",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cy"]
							],
							"-",
							[
								1,
								[
									7,
									"vel",
									"y"
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						"+",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["N"]
						]
					],
					"%",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["N"]
					]
				]
			],
			[
				13,
				"x0",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["srcX"]
				]
			],
			[
				13,
				"y0",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["srcY"]
				]
			],
			[
				13,
				"sx",
				[
					1,
					"srcX",
					"-",
					"x0"
				]
			],
			[
				13,
				"sy",
				[
					1,
					"srcY",
					"-",
					"y0"
				]
			],
			[
				13,
				"ix0",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["x0"]
				]
			],
			[
				13,
				"iy0",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["y0"]
				]
			],
			[
				13,
				"ix1",
				[
					1,
					[
						1,
						"ix0",
						"+",
						[5, "1"]
					],
					"%",
					"N"
				]
			],
			[
				13,
				"iy1",
				[
					1,
					[
						1,
						"iy0",
						"+",
						[5, "1"]
					],
					"%",
					"N"
				]
			],
			[
				13,
				"d00",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								"iy0",
								"*",
								"N"
							],
							"+",
							"ix0"
						]]
					]
				]
			],
			[
				13,
				"d10",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								"iy0",
								"*",
								"N"
							],
							"+",
							"ix1"
						]]
					]
				]
			],
			[
				13,
				"d01",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								"iy1",
								"*",
								"N"
							],
							"+",
							"ix0"
						]]
					]
				]
			],
			[
				13,
				"d11",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								"iy1",
								"*",
								"N"
							],
							"+",
							"ix1"
						]]
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densB"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										[
											7,
											"d00",
											"x"
										],
										[
											7,
											"d10",
											"x"
										],
										"sx"
									]
								],
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										[
											7,
											"d01",
											"x"
										],
										[
											7,
											"d11",
											"x"
										],
										"sx"
									]
								],
								"sy"
							]
						],
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										[
											7,
											"d00",
											"y"
										],
										[
											7,
											"d10",
											"y"
										],
										"sx"
									]
								],
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										[
											7,
											"d01",
											"y"
										],
										[
											7,
											"d11",
											"y"
										],
										"sx"
									]
								],
								"sy"
							]
						],
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"fluidLayout",
			"d",
			"N",
			"std"
		]
	},
	externals: () => ({
		fluidLayout,
		d,
		N,
		std
	})
}) && $.f)({})).$name("fogAdvectDens");
const copyDensKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const dd = fluidLayout.$.densB[idx];
	fluidLayout.$.densA[idx] = d.vec4f(dd.x, dd.y, 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"dd",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densB"
					],
					"idx"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							7,
							"dd",
							"x"
						],
						[
							7,
							"dd",
							"y"
						],
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout
	})
}) && $.f)({})).$name("fogCopyDens");
const colorRestoreKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = fluidLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const jf = d.f32(cx);
	const iF = d.f32(cy);
	const sdx = __tsover_mul(std.fract(__tsover_mul(std.sin(__tsover_mul(p.seed, 37.31)), 43758.5453)), 97);
	const sdy = __tsover_mul(std.fract(__tsover_mul(std.sin(__tsover_mul(p.seed, 83.17)), 43758.5453)), 97);
	const refRaw = fbm(__tsover_add(__tsover_mul(__tsover_div(jf, d.f32(N)), 3), sdx), __tsover_add(__tsover_mul(__tsover_div(iF, d.f32(N)), 3), sdy));
	const invBlend = __tsover_sub(1, p.blending);
	const contrastBoost = __tsover_add(1, __tsover_mul(invBlend, 5));
	const refBoosted = std.clamp(__tsover_add(__tsover_mul(__tsover_sub(refRaw, .5), contrastBoost), .5), 0, 1);
	const cell = fluidLayout.$.densA[idx];
	const rateMult = __tsover_add(1, __tsover_mul(invBlend, 2.5));
	const blendRate = std.min(__tsover_mul(__tsover_mul(p.dt, .5), rateMult), .3);
	fluidLayout.$.densA[idx] = d.vec4f(cell.x, std.mix(cell.y, refBoosted, blendRate), 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"fluidLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"jf",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					["cx"]
				]
			],
			[
				13,
				"iF",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					["cy"]
				]
			],
			[
				13,
				"sdx",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										7,
										"p",
										"seed"
									],
									"*",
									[5, "37.31"]
								]]
							],
							"*",
							[5, "43758.5453"]
						]]
					],
					"*",
					[5, "97"]
				]
			],
			[
				13,
				"sdy",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										7,
										"p",
										"seed"
									],
									"*",
									[5, "83.17"]
								]]
							],
							"*",
							[5, "43758.5453"]
						]]
					],
					"*",
					[5, "97"]
				]
			],
			[
				13,
				"nx",
				[
					1,
					[
						1,
						[
							1,
							"jf",
							"/",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["N"]
							]
						],
						"*",
						[5, "3"]
					],
					"+",
					"sdx"
				]
			],
			[
				13,
				"ny",
				[
					1,
					[
						1,
						[
							1,
							"iF",
							"/",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["N"]
							]
						],
						"*",
						[5, "3"]
					],
					"+",
					"sdy"
				]
			],
			[
				13,
				"refRaw",
				[
					6,
					"fbm",
					["nx", "ny"]
				]
			],
			[
				13,
				"invBlend",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"p",
						"blending"
					]
				]
			],
			[
				13,
				"contrastBoost",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						"invBlend",
						"*",
						[5, "5"]
					]
				]
			],
			[
				13,
				"refBoosted",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									"refRaw",
									"-",
									[5, "0.5"]
								],
								"*",
								"contrastBoost"
							],
							"+",
							[5, "0.5"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"cell",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					"idx"
				]
			],
			[
				13,
				"rateMult",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						"invBlend",
						"*",
						[5, "2.5"]
					]
				]
			],
			[
				13,
				"blendRate",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					[[
						1,
						[
							1,
							[
								7,
								"p",
								"dt"
							],
							"*",
							[5, "0.5"]
						],
						"*",
						"rateMult"
					], [5, "0.3"]]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							7,
							"cell",
							"x"
						],
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									7,
									"cell",
									"y"
								],
								"refBoosted",
								"blendRate"
							]
						],
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"fluidLayout",
			"d",
			"N",
			"std",
			"fbm"
		]
	},
	externals: () => ({
		fluidLayout,
		d,
		N,
		std,
		fbm
	})
}) && $.f)({})).$name("fogColorRestore");
const outputKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const dd = outputLayout.$.densA[idx];
	std.textureStore(outputLayout.$.outTex, d.vec2u(cx, cy), d.vec4f(dd.x, dd.y, 0, 0));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"dd",
				[
					8,
					[
						7,
						[
							7,
							"outputLayout",
							"$"
						],
						"densA"
					],
					"idx"
				]
			],
			[
				6,
				[
					7,
					"std",
					"textureStore"
				],
				[
					[
						7,
						[
							7,
							"outputLayout",
							"$"
						],
						"outTex"
					],
					[
						6,
						[
							7,
							"d",
							"vec2u"
						],
						["cx", "cy"]
					],
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							[
								7,
								"dd",
								"x"
							],
							[
								7,
								"dd",
								"y"
							],
							[5, "0"],
							[5, "0"]
						]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"outputLayout",
			"std"
		]
	},
	externals: () => ({
		d,
		N,
		outputLayout,
		std
	})
}) && $.f)({})).$name("fogOutput");
const componentDefinition = {
	name: "Fog",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Interactive",
	description: "Fog that fills the screen and interacts with the mouse",
	usesPointer: true,
	props: {
		colorA: {
			default: "#e0e0e0",
			transform: transformColor,
			description: "Primary fog color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#888888",
			transform: transformColor,
			description: "Secondary fog color — creates variation across the field",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		seed: {
			default: 0,
			description: "Deterministic starting pattern — different seeds produce different fog configurations",
			ui: {
				type: "range",
				min: 0,
				max: 999,
				step: 1,
				label: "Seed",
				group: "Field"
			}
		},
		speed: {
			default: 1,
			description: "Simulation speed multiplier",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .1,
				label: "Speed",
				group: "Field"
			}
		},
		turbulence: {
			default: 1,
			description: "Ambient motion strength",
			ui: {
				type: "range",
				min: 0,
				max: 3,
				step: .01,
				label: "Turbulence",
				group: "Field"
			}
		},
		detail: {
			default: 15,
			description: "Fine-scale swirling structure — higher values produce more intricate wisps and vortices",
			ui: {
				type: "range",
				min: 0,
				max: 50,
				step: 1,
				label: "Detail",
				group: "Field"
			}
		},
		blending: {
			default: .3,
			description: "How much the two colors blend together — 0 behaves like oil & water (colors stay distinct with sharp boundaries), 1 behaves like food coloring (colors fully mix)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Blending",
				group: "Field"
			}
		},
		mouseInfluence: {
			default: .1,
			description: "Strength of cursor influence — move the cursor to push fog",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Cursor Influence",
				group: "Interaction"
			}
		},
		mouseRadius: {
			default: .1,
			description: "Radius of cursor influence area",
			ui: {
				type: "range",
				min: .02,
				max: .5,
				step: .01,
				label: "Cursor Radius",
				group: "Interaction"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	compute: (params) => {
		const { gpu, getCpuValue, registerComputeTexture, onCleanup } = params;
		const root = gpu?.root;
		if (!root) return null;
		const velA = createStateBuffer(root, d.vec4f, COUNT);
		const velB = createStateBuffer(root, d.vec4f, COUNT);
		const densA = createStateBuffer(root, d.vec4f, COUNT);
		const densB = createStateBuffer(root, d.vec4f, COUNT);
		const pressure = createStateBuffer(root, d.f32, COUNT);
		const divergence = createStateBuffer(root, d.f32, COUNT);
		const outTex = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
			size: [N, N],
			format: STATE_FORMAT
		}).$usage("storage", "sampled"), "outTex");
		onCleanup(() => outTex.destroy());
		const fogTexture = registerComputeTexture(outTex);
		const paramsU = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(FogParams), "paramsU");
		const size = [N, N];
		const fluidBg = root.createBindGroup(fluidLayout, {
			velA,
			velB,
			densA,
			densB,
			pressure,
			divergence,
			params: paramsU.buffer
		});
		const outputBg = root.createBindGroup(outputLayout, {
			densA,
			outTex
		});
		const mk = (kernel) => createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			kernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"kernel",
					["cx", "cy"]
				]]],
				"externalNames": ["kernel"]
			},
			externals: () => ({ kernel })
		}) && $.f)({}), {
			size,
			bindGroup: fluidBg
		});
		const initPass = mk(initKernel);
		const force = mk(forceKernel);
		const curl = mk(curlKernel);
		const vorticity = mk(vorticityKernel);
		const divergencePass = mk(divergenceKernel);
		const jacobi = mk(jacobiKernel);
		const gradSubtract = mk(gradSubtractKernel);
		const advectVel = mk(advectVelKernel);
		const copyVel = mk(copyVelKernel);
		const advectDens = mk(advectDensKernel);
		const copyDens = mk(copyDensKernel);
		const colorRestore = mk(colorRestoreKernel);
		const output = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			outputKernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"outputKernel",
					["cx", "cy"]
				]]],
				"externalNames": ["outputKernel"]
			},
			externals: () => ({ outputKernel })
		}) && $.f)({}), {
			size,
			bindGroup: outputBg
		});
		const solve = (nodes, jacobiIters) => {
			nodes.push(force, curl, vorticity, divergencePass);
			for (let it = 0; it < jacobiIters; it++) nodes.push(jacobi);
			nodes.push(gradSubtract, advectVel, copyVel, advectDens, copyDens, colorRestore);
		};
		let lastTime = Date.now();
		let elapsed = 0;
		let prevMx = .5;
		let prevMy = .5;
		let lastSeed = -1;
		let initialized = false;
		return {
			outputs: { fogTexture },
			getComputeNodes: (frameParams) => {
				const fp = frameParams;
				const now = Date.now();
				const rawDt = Math.min(fp.deltaTime ?? (now - lastTime) / 1e3, .033);
				lastTime = now;
				if (rawDt < .001) return null;
				const num = (key, fallback) => {
					const v = getCpuValue(key);
					return typeof v === "number" ? v : fallback;
				};
				const dt = rawDt * num("speed", 1);
				elapsed += rawDt;
				const turbulence = num("turbulence", 1);
				const curlStrength = num("detail", 15);
				const blending = num("blending", .3);
				const seed = num("seed", 0);
				const nodes = [];
				if (!initialized || seed !== lastSeed) {
					initialized = true;
					lastSeed = seed;
					nodes.push(() => paramsU.write({
						dt: .1,
						time: 0,
						seed,
						turbulence,
						curlStrength,
						blending,
						cursorX: 0,
						cursorY: 0,
						cursorVelX: 0,
						cursorVelY: 0,
						mouseActive: 0,
						mouseRadSq: 1
					}));
					nodes.push(initPass);
					let warmTime = Math.random() * 60;
					for (let w = 0; w < WARM_STEPS; w++) {
						const wt = warmTime;
						warmTime += .1;
						nodes.push(() => paramsU.write({
							dt: .1,
							time: wt,
							seed,
							turbulence,
							curlStrength,
							blending,
							cursorX: 0,
							cursorY: 0,
							cursorVelX: 0,
							cursorVelY: 0,
							mouseActive: 0,
							mouseRadSq: 1
						}));
						solve(nodes, WARM_JACOBI_ITERS);
					}
					elapsed = warmTime;
				}
				const mouseInf = num("mouseInfluence", .1);
				const mouseRadGrid = num("mouseRadius", .1) * N;
				const mx = fp.pointer?.x ?? prevMx;
				const my = fp.pointer?.y ?? prevMy;
				const mdx = mx - prevMx;
				const mdy = my - prevMy;
				prevMx = mx;
				prevMy = my;
				const active = Math.sqrt(mdx * mdx + mdy * mdy) > 5e-4 && mouseInf > 0;
				nodes.push(() => paramsU.write({
					dt,
					time: elapsed,
					seed,
					turbulence,
					curlStrength,
					blending,
					cursorX: active ? mx * N : 0,
					cursorY: active ? my * N : 0,
					cursorVelX: active ? mdx * N * 15 * mouseInf : 0,
					cursorVelY: active ? mdy * N * 15 * mouseInf : 0,
					mouseActive: active ? 1 : 0,
					mouseRadSq: mouseRadGrid * mouseRadGrid
				}));
				solve(nodes, JACOBI_ITERS);
				nodes.push(output);
				return nodes;
			}
		};
	},
	fragment: ({ ctx, computeOutputs, propValues, uniforms }) => {
		const fogField = computeOutputs?.fogTexture;
		if (!fogField) return ZERO;
		const s = fogField.sample(ctx.uv, "linearClamp");
		const fogDensity = s.member("x");
		const colorVar = s.member("y");
		const colorSpaceMode = propValues.colorSpace ?? 0;
		const fogColor = call(cm.mixColorsVariants[colorSpaceMode] ?? cm.mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			colorVar
		]);
		return vec4(fogColor.member("rgb"), fogColor.member("a").mul(fogDensity));
	}
};
var Fog_default = componentDefinition;
export { componentDefinition as a, curlKernel as c, gradSubtractKernel as d, initKernel as f, vorticityKernel as h, colorRestoreKernel as i, divergenceKernel as l, outputKernel as m, advectDensKernel as n, copyDensKernel as o, jacobiKernel as p, advectVelKernel as r, copyVelKernel as s, Fog_default as t, forceKernel as u };
