// shaders/dist/core/fonts-3F3pIODs.js
function googleFontCss2Url(family, weight = 400, italic = false, text) {
	let url = `https://fonts.googleapis.com/css2?family=${family.trim().replace(/ /g, "+")}:ital,wght@${italic ? 1 : 0},${weight}&display=swap`;
	if (text) url += `&text=${encodeURIComponent(text)}`;
	return url;
}
var loadedFonts = /* @__PURE__ */ new Set();
var pendingLoads = /* @__PURE__ */ new Map();
function loadGoogleFont(family, weight = 400, italic = false) {
	if (typeof document === "undefined" || !family) return Promise.resolve();
	const key = `${family}|${weight}|${italic}`;
	if (loadedFonts.has(key)) return Promise.resolve();
	const pending = pendingLoads.get(key);
	if (pending) return pending;
	const promise = loadFontLink(family, weight, italic).catch(() => {
		if (weight !== 400 || italic) return loadGoogleFont(family, 400, false);
	}).then(() => {
		loadedFonts.add(key);
		pendingLoads.delete(key);
	});
	pendingLoads.set(key, promise);
	return promise;
}
function loadFontLink(family, weight, italic) {
	const href = googleFontCss2Url(family, weight, italic);
	return new Promise((resolve, reject) => {
		if (document.querySelector(`link[href="${href}"]`)) {
			resolve();
			return;
		}
		const link = document.createElement("link");
		link.rel = "stylesheet";
		link.href = href;
		link.onload = () => resolve();
		link.onerror = () => {
			link.remove();
			reject(/* @__PURE__ */ new Error(`Failed to load font stylesheet: ${family}`));
		};
		document.head.appendChild(link);
	}).then(async () => {
		if (document.fonts?.load) {
			const spec = `${italic ? "italic " : ""}${weight} 12px "${family}"`;
			try {
				if (!(await document.fonts.load(spec)).length) throw new Error(`No font faces loaded for ${spec}`);
			} catch (e) {
				if (!(await document.fonts.load(`12px "${family}"`)).length) throw e;
			}
			await new Promise((resolve) => setTimeout(resolve, 300));
		} else await new Promise((resolve) => setTimeout(resolve, 1e3));
	});
}
function isFontAvailable(family, weight = 400, italic = false) {
	if (typeof document === "undefined" || !document.fonts?.check) return false;
	try {
		return document.fonts.check(`${italic ? "italic " : ""}${weight} 12px "${family}"`);
	} catch {
		return false;
	}
}
export { loadGoogleFont as n, isFontAvailable as t };
