// shaders/dist/core/Form3D-CrT8E1u5.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, cn as __export, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { c as transformPosition, s as transformEdges } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
var raymarch3d_exports = /* @__PURE__ */ __export({
	Distortion3DTrace: () => Distortion3DTrace$1,
	applyDistortion3D: () => applyDistortion3D,
	applyUVMode3d: () => applyUVMode3d$1,
	buildDistortion3dTrace: () => buildDistortion3dTrace,
	distortion3dLighting: () => distortion3dLighting$1
});
const Distortion3DTrace$1 = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
	hitPos: d.vec3f,
	normal: d.vec3f,
	rd: d.vec3f,
	hit: d.f32
}), "Distortion3DTrace");
const applyUVMode3d$1 = tgpu.fn([d.vec2f, d.f32], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, uvMode) => {
	"use gpu";
	const stretched = std.clamp(uv, d.vec2f(0), d.vec2f(1));
	const mirrored = std.abs(std.fract(uv.mul(.5)).mul(2).sub(1));
	const wrapped = std.fract(uv);
	const tv1 = std.select(d.f32(0), d.f32(1), uv.y <= 1);
	const tv2 = std.select(d.f32(0), tv1, uv.y >= 0);
	const tv3 = std.select(d.f32(0), tv2, uv.x <= 1);
	const transparentVisible = std.select(d.f32(0), tv3, uv.x >= 0);
	const fuv2 = std.select(wrapped, mirrored, uvMode < 2.5);
	const fuv1 = std.select(fuv2, uv, uvMode < 1.5);
	const finalUV = std.select(fuv1, stretched, uvMode < .5);
	const vis1 = std.select(d.f32(1), transparentVisible, uvMode < 1.5);
	const visible = std.select(vis1, d.f32(1), uvMode < .5);
	return d.vec3f(finalUV.x, finalUV.y, visible);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "uv"
		}, {
			"type": "i",
			"name": "uvMode"
		}],
		"body": [0, [
			[
				13,
				"stretched",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"uv",
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "1"]]
						]
					]
				]
			],
			[
				13,
				"mirrored",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									[
										6,
										[
											7,
											"std",
											"fract"
										],
										[[
											6,
											[
												7,
												"uv",
												"mul"
											],
											[[5, "0.5"]]
										]]
									],
									"mul"
								],
								[[5, "2"]]
							],
							"sub"
						],
						[[5, "1"]]
					]]
				]
			],
			[
				13,
				"wrapped",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["uv"]
				]
			],
			[
				13,
				"tv1",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							[
								7,
								"uv",
								"y"
							],
							"<=",
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"tv2",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						"tv1",
						[
							1,
							[
								7,
								"uv",
								"y"
							],
							">=",
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"tv3",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						"tv2",
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"<=",
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"transparentVisible",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						"tv3",
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							">=",
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"fuv2",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"wrapped",
						"mirrored",
						[
							1,
							"uvMode",
							"<",
							[5, "2.5"]
						]
					]
				]
			],
			[
				13,
				"fuv1",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"fuv2",
						"uv",
						[
							1,
							"uvMode",
							"<",
							[5, "1.5"]
						]
					]
				]
			],
			[
				13,
				"finalUV",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"fuv1",
						"stretched",
						[
							1,
							"uvMode",
							"<",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"vis1",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						"transparentVisible",
						[
							1,
							"uvMode",
							"<",
							[5, "1.5"]
						]
					]
				]
			],
			[
				13,
				"visible",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"vis1",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"uvMode",
							"<",
							[5, "0.5"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[
						7,
						"finalUV",
						"x"
					],
					[
						7,
						"finalUV",
						"y"
					],
					"visible"
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const distortion3dLighting$1 = tgpu.fn([
	d.vec4f,
	d.vec3f,
	d.vec3f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((surfaceColor, normal, rd, hit, uvVisible, glossAmt, lightAmt) => {
	"use gpu";
	const lightDir = std.normalize(d.vec3f(.4, .7, -.6));
	const viewDir = rd.mul(-1);
	const diffuse = std.clamp(std.dot(normal, lightDir), 0, 1);
	const halfVec = std.normalize(lightDir.add(viewDir));
	const specPow = __tsover_add(__tsover_mul(glossAmt, 256), 4);
	const specular = __tsover_mul(__tsover_mul(std.pow(std.clamp(std.dot(normal, halfVec), 0, 1), specPow), glossAmt), 1.5);
	const nDotV = std.clamp(std.dot(normal, viewDir), 0, 1);
	const fresnel = __tsover_mul(__tsover_mul(std.pow(__tsover_sub(1, nDotV), 3), .5), glossAmt);
	const lit = std.mix(1, __tsover_add(__tsover_add(__tsover_add(.35, __tsover_mul(diffuse, .65)), specular), fresnel), lightAmt);
	const litClamped = std.clamp(lit, 0, 3);
	const col = surfaceColor.xyz.mul(litClamped);
	const colClamped = std.clamp(col, d.vec3f(0), d.vec3f(1));
	const alpha = __tsover_mul(__tsover_mul(hit, surfaceColor.w), uvVisible);
	return d.vec4f(colClamped.x, colClamped.y, colClamped.z, alpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "surfaceColor"
			},
			{
				"type": "i",
				"name": "normal"
			},
			{
				"type": "i",
				"name": "rd"
			},
			{
				"type": "i",
				"name": "hit"
			},
			{
				"type": "i",
				"name": "uvVisible"
			},
			{
				"type": "i",
				"name": "glossAmt"
			},
			{
				"type": "i",
				"name": "lightAmt"
			}
		],
		"body": [0, [
			[
				13,
				"lightDir",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0.4"],
							[5, "0.7"],
							[
								4,
								"-",
								[5, "0.6"]
							]
						]
					]]
				]
			],
			[
				13,
				"viewDir",
				[
					6,
					[
						7,
						"rd",
						"mul"
					],
					[[
						4,
						"-",
						[5, "1"]
					]]
				]
			],
			[
				13,
				"diffuse",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["normal", "lightDir"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"halfVec",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"lightDir",
							"add"
						],
						["viewDir"]
					]]
				]
			],
			[
				13,
				"specPow",
				[
					1,
					[
						1,
						"glossAmt",
						"*",
						[5, "256"]
					],
					"+",
					[5, "4"]
				]
			],
			[
				13,
				"specular",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							[[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									[
										6,
										[
											7,
											"std",
											"dot"
										],
										["normal", "halfVec"]
									],
									[5, "0"],
									[5, "1"]
								]
							], "specPow"]
						],
						"*",
						"glossAmt"
					],
					"*",
					[5, "1.5"]
				]
			],
			[
				13,
				"nDotV",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["normal", "viewDir"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"fresnel",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							[[
								1,
								[5, "1"],
								"-",
								"nDotV"
							], [5, "3"]]
						],
						"*",
						[5, "0.5"]
					],
					"*",
					"glossAmt"
				]
			],
			[
				13,
				"ambient",
				[5, "0.35"]
			],
			[
				13,
				"lit",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "1"],
						[
							1,
							[
								1,
								[
									1,
									"ambient",
									"+",
									[
										1,
										"diffuse",
										"*",
										[5, "0.65"]
									]
								],
								"+",
								"specular"
							],
							"+",
							"fresnel"
						],
						"lightAmt"
					]
				]
			],
			[
				13,
				"litClamped",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"lit",
						[5, "0"],
						[5, "3"]
					]
				]
			],
			[
				13,
				"col",
				[
					6,
					[
						7,
						[
							7,
							"surfaceColor",
							"xyz"
						],
						"mul"
					],
					["litClamped"]
				]
			],
			[
				13,
				"colClamped",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"col",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[5, "1"]]
						]
					]
				]
			],
			[
				13,
				"alpha",
				[
					1,
					[
						1,
						"hit",
						"*",
						[
							7,
							"surfaceColor",
							"w"
						]
					],
					"*",
					"uvVisible"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"colClamped",
						"x"
					],
					[
						7,
						"colClamped",
						"y"
					],
					[
						7,
						"colClamped",
						"z"
					],
					"alpha"
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
function buildDistortion3dTrace(sdfFn, opts) {
	const STEPS = opts.steps;
	const STEP_CAP = opts.stepCap;
	const HIT_EPS = opts.hitEps;
	const MAX_DIST = opts.maxDistance;
	return tgpu.fn([
		d.vec2f,
		d.f32,
		d.vec2f,
		d.f32
	], Distortion3DTrace$1)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, center, zoom) => {
		"use gpu";
		const cx = center.x;
		const cy = __tsover_sub(1, center.y);
		const camDist = __tsover_sub(8, __tsover_mul(zoom, .06));
		const ro = d.vec3f(0, 0, __tsover_mul(camDist, -1));
		const rd = std.normalize(d.vec3f(__tsover_mul(__tsover_sub(uv.x, cx), aspect), __tsover_sub(uv.y, cy), 1.5));
		let rayT = d.f32(0);
		let hit = d.f32(0);
		const maxDistF = d.f32(MAX_DIST);
		for (let i = 0; i < STEPS; i++) {
			const dd = sdfFn(ro.add(rd.mul(rayT)));
			if (dd < HIT_EPS) {
				hit = d.f32(1);
				break;
			}
			if (rayT > maxDistF) break;
			rayT = __tsover_add(rayT, std.min(dd, STEP_CAP));
		}
		const hitPos = ro.add(rd.mul(rayT));
		const h = .001;
		const hn = __tsover_mul(h, -1);
		const k0 = sdfFn(hitPos.add(d.vec3f(h, hn, hn)));
		const k1 = sdfFn(hitPos.add(d.vec3f(hn, hn, h)));
		const k2 = sdfFn(hitPos.add(d.vec3f(hn, h, hn)));
		const k3 = sdfFn(hitPos.add(d.vec3f(h, h, h)));
		return Distortion3DTrace$1({
			hitPos,
			normal: std.normalize(d.vec3f(1, -1, -1).mul(k0).add(d.vec3f(-1, -1, 1).mul(k1)).add(d.vec3f(-1, 1, -1).mul(k2)).add(d.vec3f(1, 1, 1).mul(k3))),
			rd,
			hit
		});
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [
				{
					"type": "i",
					"name": "uv"
				},
				{
					"type": "i",
					"name": "aspect"
				},
				{
					"type": "i",
					"name": "center"
				},
				{
					"type": "i",
					"name": "zoom"
				}
			],
			"body": [0, [
				[
					13,
					"cx",
					[
						7,
						"center",
						"x"
					]
				],
				[
					13,
					"cy",
					[
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]
				],
				[
					13,
					"camDist",
					[
						1,
						[5, "8"],
						"-",
						[
							1,
							"zoom",
							"*",
							[5, "0.06"]
						]
					]
				],
				[
					13,
					"ro",
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0"],
							[5, "0"],
							[
								1,
								"camDist",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]
						]
					]
				],
				[
					13,
					"focal",
					[5, "1.5"]
				],
				[
					13,
					"rd",
					[
						6,
						[
							7,
							"std",
							"normalize"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									[
										1,
										[
											7,
											"uv",
											"x"
										],
										"-",
										"cx"
									],
									"*",
									"aspect"
								],
								[
									1,
									[
										7,
										"uv",
										"y"
									],
									"-",
									"cy"
								],
								"focal"
							]
						]]
					]
				],
				[
					12,
					"rayT",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					12,
					"hit",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					13,
					"maxDistF",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["MAX_DIST"]
					]
				],
				[
					14,
					[
						12,
						"i",
						[5, "0"]
					],
					[
						1,
						"i",
						"<",
						"STEPS"
					],
					[
						102,
						"++",
						"i"
					],
					[0, [
						[
							13,
							"p",
							[
								6,
								[
									7,
									"ro",
									"add"
								],
								[[
									6,
									[
										7,
										"rd",
										"mul"
									],
									["rayT"]
								]]
							]
						],
						[
							13,
							"dd",
							[
								6,
								"sdfFn",
								["p"]
							]
						],
						[
							11,
							[
								1,
								"dd",
								"<",
								"HIT_EPS"
							],
							[0, [[
								2,
								"hit",
								"=",
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "1"]]
								]
							], [17]]]
						],
						[
							11,
							[
								1,
								"rayT",
								">",
								"maxDistF"
							],
							[0, [[17]]]
						],
						[
							2,
							"rayT",
							"=",
							[
								1,
								"rayT",
								"+",
								[
									6,
									[
										7,
										"std",
										"min"
									],
									["dd", "STEP_CAP"]
								]
							]
						]
					]]
				],
				[
					13,
					"hitPos",
					[
						6,
						[
							7,
							"ro",
							"add"
						],
						[[
							6,
							[
								7,
								"rd",
								"mul"
							],
							["rayT"]
						]]
					]
				],
				[
					13,
					"h",
					[5, "0.001"]
				],
				[
					13,
					"hn",
					[
						1,
						"h",
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					]
				],
				[
					13,
					"k0",
					[
						6,
						"sdfFn",
						[[
							6,
							[
								7,
								"hitPos",
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									"h",
									"hn",
									"hn"
								]
							]]
						]]
					]
				],
				[
					13,
					"k1",
					[
						6,
						"sdfFn",
						[[
							6,
							[
								7,
								"hitPos",
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									"hn",
									"hn",
									"h"
								]
							]]
						]]
					]
				],
				[
					13,
					"k2",
					[
						6,
						"sdfFn",
						[[
							6,
							[
								7,
								"hitPos",
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									"hn",
									"h",
									"hn"
								]
							]]
						]]
					]
				],
				[
					13,
					"k3",
					[
						6,
						"sdfFn",
						[[
							6,
							[
								7,
								"hitPos",
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									"h",
									"h",
									"h"
								]
							]]
						]]
					]
				],
				[
					13,
					"normal",
					[
						6,
						[
							7,
							"std",
							"normalize"
						],
						[[
							6,
							[
								7,
								[
									6,
									[
										7,
										[
											6,
											[
												7,
												[
													6,
													[
														7,
														[
															6,
															[
																7,
																"d",
																"vec3f"
															],
															[
																[5, "1"],
																[
																	4,
																	"-",
																	[5, "1"]
																],
																[
																	4,
																	"-",
																	[5, "1"]
																]
															]
														],
														"mul"
													],
													["k0"]
												],
												"add"
											],
											[[
												6,
												[
													7,
													[
														6,
														[
															7,
															"d",
															"vec3f"
														],
														[
															[
																4,
																"-",
																[5, "1"]
															],
															[
																4,
																"-",
																[5, "1"]
															],
															[5, "1"]
														]
													],
													"mul"
												],
												["k1"]
											]]
										],
										"add"
									],
									[[
										6,
										[
											7,
											[
												6,
												[
													7,
													"d",
													"vec3f"
												],
												[
													[
														4,
														"-",
														[5, "1"]
													],
													[5, "1"],
													[
														4,
														"-",
														[5, "1"]
													]
												]
											],
											"mul"
										],
										["k2"]
									]]
								],
								"add"
							],
							[[
								6,
								[
									7,
									[
										6,
										[
											7,
											"d",
											"vec3f"
										],
										[
											[5, "1"],
											[5, "1"],
											[5, "1"]
										]
									],
									"mul"
								],
								["k3"]
							]]
						]]
					]
				],
				[10, [
					6,
					"Distortion3DTrace",
					[[104, {
						"hitPos": "hitPos",
						"normal": "normal",
						"rd": "rd",
						"hit": "hit"
					}]]
				]]
			]],
			"externalNames": [
				"d",
				"std",
				"MAX_DIST",
				"STEPS",
				"sdfFn",
				"HIT_EPS",
				"STEP_CAP",
				"Distortion3DTrace"
			]
		},
		externals: () => ({
			d,
			std,
			MAX_DIST,
			STEPS,
			sdfFn,
			HIT_EPS,
			STEP_CAP,
			Distortion3DTrace: Distortion3DTrace$1
		})
	}) && $.f)({})).$name("distortion3dTrace");
}
function applyDistortion3D(params, childTexture, shape, march = {}) {
	const { uniforms, ctx } = params;
	const opts = {
		steps: march.steps ?? 96,
		stepCap: march.stepCap ?? .15,
		hitEps: march.hitEps ?? .001,
		maxDistance: march.maxDistance ?? 20
	};
	const trace = buildDistortion3dTrace(shape.sdf, opts);
	const glossAmt = uniforms.glossiness.mul(.005);
	const lightAmt = uniforms.lighting.mul(.005);
	const traceRes = call(trace, "distortion3dTrace", [
		ctx.uv,
		ctx.aspect,
		uniforms.center,
		uniforms.zoom
	]);
	const hitPos = traceRes.member("hitPos");
	const normal = traceRes.member("normal");
	const rd = traceRes.member("rd");
	const hit = traceRes.member("hit");
	const uvm = call(applyUVMode3d$1, "applyUVMode3d", [call(shape.surfaceUV, "surfaceUV", [hitPos, normal]), uniforms.uvMode]);
	const finalUV = uvm.member("xy");
	const uvVisible = uvm.member("z");
	return call(distortion3dLighting$1, "distortion3dLighting", [
		call(unpremultiplyAlpha, "unpremultiplyAlpha", [childTexture.sample(finalUV)]),
		normal,
		rd,
		hit,
		uvVisible,
		glossAmt,
		lightAmt
	]);
}
var { applyUVMode3d, distortion3dLighting, Distortion3DTrace } = raymarch3d_exports;
var DEG = Math.PI / 180;
var TAU_RATE = Math.PI * 2 / 10;
var TWIST_SPEED = .08;
var DEFAULT_RIBBON = JSON.stringify({
	type: "ribbon",
	angle: 0,
	twist: 50,
	width: 40,
	thickness: 20,
	seed: 0
});
const f3dRot = tgpu.fn([
	d.vec3f,
	d.vec4f,
	d.vec4f
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, rp0, rp1) => {
	"use gpu";
	const cx = rp0.x;
	const sx = rp0.y;
	const cy = rp0.z;
	const sy = rp0.w;
	const cz = rp1.x;
	const sz = rp1.y;
	const py = d.vec3f(__tsover_add(__tsover_mul(p.x, cy), __tsover_mul(p.z, sy)), p.y, __tsover_add(__tsover_mul(__tsover_mul(p.x, sy), -1), __tsover_mul(p.z, cy)));
	const px = d.vec3f(py.x, __tsover_sub(__tsover_mul(py.y, cx), __tsover_mul(py.z, sx)), __tsover_add(__tsover_mul(py.y, sx), __tsover_mul(py.z, cx)));
	return d.vec3f(__tsover_sub(__tsover_mul(px.x, cz), __tsover_mul(px.y, sz)), __tsover_add(__tsover_mul(px.x, sz), __tsover_mul(px.y, cz)), px.z);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "p"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			}
		],
		"body": [0, [
			[
				13,
				"cx",
				[
					7,
					"rp0",
					"x"
				]
			],
			[
				13,
				"sx",
				[
					7,
					"rp0",
					"y"
				]
			],
			[
				13,
				"cy",
				[
					7,
					"rp0",
					"z"
				]
			],
			[
				13,
				"sy",
				[
					7,
					"rp0",
					"w"
				]
			],
			[
				13,
				"cz",
				[
					7,
					"rp1",
					"x"
				]
			],
			[
				13,
				"sz",
				[
					7,
					"rp1",
					"y"
				]
			],
			[
				13,
				"py",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"p",
									"x"
								],
								"*",
								"cy"
							],
							"+",
							[
								1,
								[
									7,
									"p",
									"z"
								],
								"*",
								"sy"
							]
						],
						[
							7,
							"p",
							"y"
						],
						[
							1,
							[
								1,
								[
									1,
									[
										7,
										"p",
										"x"
									],
									"*",
									"sy"
								],
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"+",
							[
								1,
								[
									7,
									"p",
									"z"
								],
								"*",
								"cy"
							]
						]
					]
				]
			],
			[
				13,
				"px",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"py",
							"x"
						],
						[
							1,
							[
								1,
								[
									7,
									"py",
									"y"
								],
								"*",
								"cx"
							],
							"-",
							[
								1,
								[
									7,
									"py",
									"z"
								],
								"*",
								"sx"
							]
						],
						[
							1,
							[
								1,
								[
									7,
									"py",
									"y"
								],
								"*",
								"sx"
							],
							"+",
							[
								1,
								[
									7,
									"py",
									"z"
								],
								"*",
								"cx"
							]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[
						1,
						[
							1,
							[
								7,
								"px",
								"x"
							],
							"*",
							"cz"
						],
						"-",
						[
							1,
							[
								7,
								"px",
								"y"
							],
							"*",
							"sz"
						]
					],
					[
						1,
						[
							1,
							[
								7,
								"px",
								"x"
							],
							"*",
							"sz"
						],
						"+",
						[
							1,
							[
								7,
								"px",
								"y"
							],
							"*",
							"cz"
						]
					],
					[
						7,
						"px",
						"z"
					]
				]
			]]
		]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
var sphereSdf = tgpu.fn([
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, rp0, rp1, _rp2, _t) => {
	"use gpu";
	return __tsover_sub(std.length(f3dRot(p, rp0, rp1)), rp1.z);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "p"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "_rp2"
			},
			{
				"type": "i",
				"name": "_t"
			}
		],
		"body": [0, [[10, [
			1,
			[
				6,
				[
					7,
					"std",
					"length"
				],
				[[
					6,
					"f3dRot",
					[
						"p",
						"rp0",
						"rp1"
					]
				]]
			],
			"-",
			[
				7,
				"rp1",
				"z"
			]
		]]]],
		"externalNames": ["std", "f3dRot"]
	},
	externals: () => ({
		std,
		f3dRot
	})
}) && $.f)({}));
var sphereUV = tgpu.fn([
	d.vec3f,
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((h, _n, rp0, rp1, _rp2, _t) => {
	"use gpu";
	const n = std.normalize(f3dRot(h, rp0, rp1));
	return d.vec2f(__tsover_add(__tsover_div(std.atan2(n.z, n.x), 6.283185307179586), .5), __tsover_mul(__tsover_add(n.y, 1), .5));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "h"
			},
			{
				"type": "i",
				"name": "_n"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "_rp2"
			},
			{
				"type": "i",
				"name": "_t"
			}
		],
		"body": [0, [[
			13,
			"n",
			[
				6,
				[
					7,
					"std",
					"normalize"
				],
				[[
					6,
					"f3dRot",
					[
						"h",
						"rp0",
						"rp1"
					]
				]]
			]
		], [10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					1,
					[
						6,
						[
							7,
							"std",
							"atan2"
						],
						[[
							7,
							"n",
							"z"
						], [
							7,
							"n",
							"x"
						]]
					],
					"/",
					[5, "6.283185307179586"]
				],
				"+",
				[5, "0.5"]
			], [
				1,
				[
					1,
					[
						7,
						"n",
						"y"
					],
					"+",
					[5, "1"]
				],
				"*",
				[5, "0.5"]
			]]
		]]]],
		"externalNames": [
			"std",
			"f3dRot",
			"d"
		]
	},
	externals: () => ({
		std,
		f3dRot,
		d
	})
}) && $.f)({}));
var torusSdf = tgpu.fn([
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, rp0, rp1, _rp2, _t) => {
	"use gpu";
	const rp = f3dRot(p, rp0, rp1);
	const q = d.vec2f(__tsover_sub(std.length(d.vec2f(rp.x, rp.z)), rp1.z), rp.y);
	return __tsover_sub(std.length(q), rp1.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "p"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "_rp2"
			},
			{
				"type": "i",
				"name": "_t"
			}
		],
		"body": [0, [
			[
				13,
				"rp",
				[
					6,
					"f3dRot",
					[
						"p",
						"rp0",
						"rp1"
					]
				]
			],
			[
				13,
				"q",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"length"
							],
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									7,
									"rp",
									"x"
								], [
									7,
									"rp",
									"z"
								]]
							]]
						],
						"-",
						[
							7,
							"rp1",
							"z"
						]
					], [
						7,
						"rp",
						"y"
					]]
				]
			],
			[10, [
				1,
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["q"]
				],
				"-",
				[
					7,
					"rp1",
					"w"
				]
			]]
		]],
		"externalNames": [
			"f3dRot",
			"d",
			"std"
		]
	},
	externals: () => ({
		f3dRot,
		d,
		std
	})
}) && $.f)({}));
var torusUV = tgpu.fn([
	d.vec3f,
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((h, _n, rp0, rp1, _rp2, _t) => {
	"use gpu";
	const rp = f3dRot(h, rp0, rp1);
	const u = __tsover_add(__tsover_div(std.atan2(rp.z, rp.x), 6.283185307179586), .5);
	const ringDist = __tsover_sub(std.length(d.vec2f(rp.x, rp.z)), rp1.z);
	return d.vec2f(u, __tsover_add(__tsover_div(std.atan2(rp.y, ringDist), 6.283185307179586), .5));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "h"
			},
			{
				"type": "i",
				"name": "_n"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "_rp2"
			},
			{
				"type": "i",
				"name": "_t"
			}
		],
		"body": [0, [
			[
				13,
				"rp",
				[
					6,
					"f3dRot",
					[
						"h",
						"rp0",
						"rp1"
					]
				]
			],
			[
				13,
				"u",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"atan2"
							],
							[[
								7,
								"rp",
								"z"
							], [
								7,
								"rp",
								"x"
							]]
						],
						"/",
						[5, "6.283185307179586"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"ringDist",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"length"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								7,
								"rp",
								"x"
							], [
								7,
								"rp",
								"z"
							]]
						]]
					],
					"-",
					[
						7,
						"rp1",
						"z"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["u", [
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"atan2"
							],
							[[
								7,
								"rp",
								"y"
							], "ringDist"]
						],
						"/",
						[5, "6.283185307179586"]
					],
					"+",
					[5, "0.5"]
				]]
			]]
		]],
		"externalNames": [
			"f3dRot",
			"std",
			"d"
		]
	},
	externals: () => ({
		f3dRot,
		std,
		d
	})
}) && $.f)({}));
var boxSdf = tgpu.fn([
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, rp0, rp1, rp2, _t) => {
	"use gpu";
	const rp = f3dRot(p, rp0, rp1);
	const rounding = rp2.y;
	const qx = __tsover_add(__tsover_sub(std.abs(rp.x), rp1.z), rounding);
	const qy = __tsover_add(__tsover_sub(std.abs(rp.y), rp1.w), rounding);
	const qz = __tsover_add(__tsover_sub(std.abs(rp.z), rp2.x), rounding);
	const outer = std.length(d.vec3f(std.max(qx, d.f32(0)), std.max(qy, d.f32(0)), std.max(qz, d.f32(0))));
	const inner = std.min(std.max(qx, std.max(qy, qz)), d.f32(0));
	return __tsover_sub(__tsover_add(outer, inner), rounding);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "p"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "rp2"
			},
			{
				"type": "i",
				"name": "_t"
			}
		],
		"body": [0, [
			[
				13,
				"rp",
				[
					6,
					"f3dRot",
					[
						"p",
						"rp0",
						"rp1"
					]
				]
			],
			[
				13,
				"rounding",
				[
					7,
					"rp2",
					"y"
				]
			],
			[
				13,
				"qx",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								7,
								"rp",
								"x"
							]]
						],
						"-",
						[
							7,
							"rp1",
							"z"
						]
					],
					"+",
					"rounding"
				]
			],
			[
				13,
				"qy",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								7,
								"rp",
								"y"
							]]
						],
						"-",
						[
							7,
							"rp1",
							"w"
						]
					],
					"+",
					"rounding"
				]
			],
			[
				13,
				"qz",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								7,
								"rp",
								"z"
							]]
						],
						"-",
						[
							7,
							"rp2",
							"x"
						]
					],
					"+",
					"rounding"
				]
			],
			[
				13,
				"outer",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["qx", [
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "0"]]
								]]
							],
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["qy", [
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "0"]]
								]]
							],
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["qz", [
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "0"]]
								]]
							]
						]
					]]
				]
			],
			[
				13,
				"inner",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					[[
						6,
						[
							7,
							"std",
							"max"
						],
						["qx", [
							6,
							[
								7,
								"std",
								"max"
							],
							["qy", "qz"]
						]]
					], [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]]
				]
			],
			[10, [
				1,
				[
					1,
					"outer",
					"+",
					"inner"
				],
				"-",
				"rounding"
			]]
		]],
		"externalNames": [
			"f3dRot",
			"std",
			"d"
		]
	},
	externals: () => ({
		f3dRot,
		std,
		d
	})
}) && $.f)({}));
var boxUV = tgpu.fn([
	d.vec3f,
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((h, n, rp0, rp1, rp2, _t) => {
	"use gpu";
	const rp = f3dRot(h, rp0, rp1);
	const rn = f3dRot(n, rp0, rp1);
	const an = d.vec3f(std.abs(rn.x), std.abs(rn.y), std.abs(rn.z));
	const hx = rp1.z;
	const hy = rp1.w;
	const hz = rp2.x;
	const scaleX = __tsover_div(.5, std.max(hy, hz));
	const scaleY = __tsover_div(.5, std.max(hx, hz));
	const scaleZ = __tsover_div(.5, std.max(hx, hy));
	const uvX = d.vec2f(__tsover_add(__tsover_mul(rp.z, scaleX), .5), __tsover_add(__tsover_mul(rp.y, scaleX), .5));
	const uvY = d.vec2f(__tsover_add(__tsover_mul(rp.x, scaleY), .5), __tsover_add(__tsover_mul(rp.z, scaleY), .5));
	const uvZ = d.vec2f(__tsover_add(__tsover_mul(rp.x, scaleZ), .5), __tsover_add(__tsover_mul(rp.y, scaleZ), .5));
	const pickXZ = std.select(uvZ, uvX, an.x > an.z);
	const pickYZ = std.select(uvZ, uvY, an.y > an.z);
	return std.select(pickYZ, pickXZ, an.x > an.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "h"
			},
			{
				"type": "i",
				"name": "n"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "rp2"
			},
			{
				"type": "i",
				"name": "_t"
			}
		],
		"body": [0, [
			[
				13,
				"rp",
				[
					6,
					"f3dRot",
					[
						"h",
						"rp0",
						"rp1"
					]
				]
			],
			[
				13,
				"rn",
				[
					6,
					"f3dRot",
					[
						"n",
						"rp0",
						"rp1"
					]
				]
			],
			[
				13,
				"an",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								7,
								"rn",
								"x"
							]]
						],
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								7,
								"rn",
								"y"
							]]
						],
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								7,
								"rn",
								"z"
							]]
						]
					]
				]
			],
			[
				13,
				"hx",
				[
					7,
					"rp1",
					"z"
				]
			],
			[
				13,
				"hy",
				[
					7,
					"rp1",
					"w"
				]
			],
			[
				13,
				"hz",
				[
					7,
					"rp2",
					"x"
				]
			],
			[
				13,
				"scaleX",
				[
					1,
					[5, "0.5"],
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						["hy", "hz"]
					]
				]
			],
			[
				13,
				"scaleY",
				[
					1,
					[5, "0.5"],
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						["hx", "hz"]
					]
				]
			],
			[
				13,
				"scaleZ",
				[
					1,
					[5, "0.5"],
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						["hx", "hy"]
					]
				]
			],
			[
				13,
				"uvX",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								7,
								"rp",
								"z"
							],
							"*",
							"scaleX"
						],
						"+",
						[5, "0.5"]
					], [
						1,
						[
							1,
							[
								7,
								"rp",
								"y"
							],
							"*",
							"scaleX"
						],
						"+",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"uvY",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								7,
								"rp",
								"x"
							],
							"*",
							"scaleY"
						],
						"+",
						[5, "0.5"]
					], [
						1,
						[
							1,
							[
								7,
								"rp",
								"z"
							],
							"*",
							"scaleY"
						],
						"+",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"uvZ",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								7,
								"rp",
								"x"
							],
							"*",
							"scaleZ"
						],
						"+",
						[5, "0.5"]
					], [
						1,
						[
							1,
							[
								7,
								"rp",
								"y"
							],
							"*",
							"scaleZ"
						],
						"+",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"pickXZ",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"uvZ",
						"uvX",
						[
							1,
							[
								7,
								"an",
								"x"
							],
							">",
							[
								7,
								"an",
								"z"
							]
						]
					]
				]
			],
			[
				13,
				"pickYZ",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"uvZ",
						"uvY",
						[
							1,
							[
								7,
								"an",
								"y"
							],
							">",
							[
								7,
								"an",
								"z"
							]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"select"
				],
				[
					"pickYZ",
					"pickXZ",
					[
						1,
						[
							7,
							"an",
							"x"
						],
						">",
						[
							7,
							"an",
							"y"
						]
					]
				]
			]]
		]],
		"externalNames": [
			"f3dRot",
			"d",
			"std"
		]
	},
	externals: () => ({
		f3dRot,
		d,
		std
	})
}) && $.f)({}));
var capsuleSdf = tgpu.fn([
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, rp0, rp1, _rp2, _t) => {
	"use gpu";
	const rp = f3dRot(p, rp0, rp1);
	const radius = rp1.z;
	const halfHeight = rp1.w;
	const clampedY = std.clamp(rp.y, __tsover_mul(halfHeight, -1), halfHeight);
	return __tsover_sub(std.length(d.vec3f(rp.x, __tsover_sub(rp.y, clampedY), rp.z)), radius);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "p"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "_rp2"
			},
			{
				"type": "i",
				"name": "_t"
			}
		],
		"body": [0, [
			[
				13,
				"rp",
				[
					6,
					"f3dRot",
					[
						"p",
						"rp0",
						"rp1"
					]
				]
			],
			[
				13,
				"radius",
				[
					7,
					"rp1",
					"z"
				]
			],
			[
				13,
				"halfHeight",
				[
					7,
					"rp1",
					"w"
				]
			],
			[
				13,
				"clampedY",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							7,
							"rp",
							"y"
						],
						[
							1,
							"halfHeight",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"halfHeight"
					]
				]
			],
			[10, [
				1,
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								7,
								"rp",
								"x"
							],
							[
								1,
								[
									7,
									"rp",
									"y"
								],
								"-",
								"clampedY"
							],
							[
								7,
								"rp",
								"z"
							]
						]
					]]
				],
				"-",
				"radius"
			]]
		]],
		"externalNames": [
			"f3dRot",
			"std",
			"d"
		]
	},
	externals: () => ({
		f3dRot,
		std,
		d
	})
}) && $.f)({}));
var capsuleUV = tgpu.fn([
	d.vec3f,
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((h, _n, rp0, rp1, _rp2, _t) => {
	"use gpu";
	const rp = f3dRot(h, rp0, rp1);
	const u = __tsover_add(__tsover_div(std.atan2(rp.z, rp.x), 6.283185307179586), .5);
	const span = __tsover_add(rp1.w, rp1.z);
	return d.vec2f(u, std.clamp(__tsover_add(__tsover_mul(__tsover_div(rp.y, span), .5), .5), 0, 1));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "h"
			},
			{
				"type": "i",
				"name": "_n"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "_rp2"
			},
			{
				"type": "i",
				"name": "_t"
			}
		],
		"body": [0, [
			[
				13,
				"rp",
				[
					6,
					"f3dRot",
					[
						"h",
						"rp0",
						"rp1"
					]
				]
			],
			[
				13,
				"u",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"atan2"
							],
							[[
								7,
								"rp",
								"z"
							], [
								7,
								"rp",
								"x"
							]]
						],
						"/",
						[5, "6.283185307179586"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"span",
				[
					1,
					[
						7,
						"rp1",
						"w"
					],
					"+",
					[
						7,
						"rp1",
						"z"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["u", [
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									[
										7,
										"rp",
										"y"
									],
									"/",
									"span"
								],
								"*",
								[5, "0.5"]
							],
							"+",
							[5, "0.5"]
						],
						[5, "0"],
						[5, "1"]
					]
				]]
			]]
		]],
		"externalNames": [
			"f3dRot",
			"std",
			"d"
		]
	},
	externals: () => ({
		f3dRot,
		std,
		d
	})
}) && $.f)({}));
var mobiusSdf = tgpu.fn([
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, rp0, rp1, rp2, _t) => {
	"use gpu";
	const rp = f3dRot(p, rp0, rp1);
	const ringRadius = rp1.z;
	const halfWidth = rp1.w;
	const thickness = rp2.x;
	const phi = std.atan2(rp.z, rp.x);
	const spine = d.vec3f(__tsover_mul(std.cos(phi), ringRadius), 0, __tsover_mul(std.sin(phi), ringRadius));
	const dvec = d.vec3f(__tsover_sub(rp.x, spine.x), __tsover_sub(rp.y, spine.y), __tsover_sub(rp.z, spine.z));
	const halfPhi = __tsover_mul(phi, .5);
	const ev = d.vec3f(__tsover_mul(std.cos(phi), std.cos(halfPhi)), std.sin(halfPhi), __tsover_mul(std.sin(phi), std.cos(halfPhi)));
	const vProj = std.dot(dvec, ev);
	const du = std.max(__tsover_sub(std.abs(vProj), halfWidth), d.f32(0));
	const perp = d.vec3f(__tsover_sub(dvec.x, __tsover_mul(ev.x, vProj)), __tsover_sub(dvec.y, __tsover_mul(ev.y, vProj)), __tsover_sub(dvec.z, __tsover_mul(ev.z, vProj)));
	return __tsover_sub(std.length(d.vec2f(du, std.length(perp))), thickness);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "p"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "rp2"
			},
			{
				"type": "i",
				"name": "_t"
			}
		],
		"body": [0, [
			[
				13,
				"rp",
				[
					6,
					"f3dRot",
					[
						"p",
						"rp0",
						"rp1"
					]
				]
			],
			[
				13,
				"ringRadius",
				[
					7,
					"rp1",
					"z"
				]
			],
			[
				13,
				"halfWidth",
				[
					7,
					"rp1",
					"w"
				]
			],
			[
				13,
				"thickness",
				[
					7,
					"rp2",
					"x"
				]
			],
			[
				13,
				"phi",
				[
					6,
					[
						7,
						"std",
						"atan2"
					],
					[[
						7,
						"rp",
						"z"
					], [
						7,
						"rp",
						"x"
					]]
				]
			],
			[
				13,
				"spine",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"cos"
								],
								["phi"]
							],
							"*",
							"ringRadius"
						],
						[5, "0"],
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								["phi"]
							],
							"*",
							"ringRadius"
						]
					]
				]
			],
			[
				13,
				"dvec",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								7,
								"rp",
								"x"
							],
							"-",
							[
								7,
								"spine",
								"x"
							]
						],
						[
							1,
							[
								7,
								"rp",
								"y"
							],
							"-",
							[
								7,
								"spine",
								"y"
							]
						],
						[
							1,
							[
								7,
								"rp",
								"z"
							],
							"-",
							[
								7,
								"spine",
								"z"
							]
						]
					]
				]
			],
			[
				13,
				"halfPhi",
				[
					1,
					"phi",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"ev",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"cos"
								],
								["phi"]
							],
							"*",
							[
								6,
								[
									7,
									"std",
									"cos"
								],
								["halfPhi"]
							]
						],
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							["halfPhi"]
						],
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								["phi"]
							],
							"*",
							[
								6,
								[
									7,
									"std",
									"cos"
								],
								["halfPhi"]
							]
						]
					]
				]
			],
			[
				13,
				"vProj",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["dvec", "ev"]
				]
			],
			[
				13,
				"du",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							["vProj"]
						],
						"-",
						"halfWidth"
					], [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]]
				]
			],
			[
				13,
				"perp",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								7,
								"dvec",
								"x"
							],
							"-",
							[
								1,
								[
									7,
									"ev",
									"x"
								],
								"*",
								"vProj"
							]
						],
						[
							1,
							[
								7,
								"dvec",
								"y"
							],
							"-",
							[
								1,
								[
									7,
									"ev",
									"y"
								],
								"*",
								"vProj"
							]
						],
						[
							1,
							[
								7,
								"dvec",
								"z"
							],
							"-",
							[
								1,
								[
									7,
									"ev",
									"z"
								],
								"*",
								"vProj"
							]
						]
					]
				]
			],
			[10, [
				1,
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						["du", [
							6,
							[
								7,
								"std",
								"length"
							],
							["perp"]
						]]
					]]
				],
				"-",
				"thickness"
			]]
		]],
		"externalNames": [
			"f3dRot",
			"std",
			"d"
		]
	},
	externals: () => ({
		f3dRot,
		std,
		d
	})
}) && $.f)({}));
var mobiusUV = tgpu.fn([
	d.vec3f,
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((h, _n, rp0, rp1, _rp2, _t) => {
	"use gpu";
	const rp = f3dRot(h, rp0, rp1);
	const ringRadius = rp1.z;
	const halfWidth = rp1.w;
	const phi = std.atan2(rp.z, rp.x);
	const spine = d.vec3f(__tsover_mul(std.cos(phi), ringRadius), 0, __tsover_mul(std.sin(phi), ringRadius));
	const dvec = d.vec3f(__tsover_sub(rp.x, spine.x), __tsover_sub(rp.y, spine.y), __tsover_sub(rp.z, spine.z));
	const halfPhi = __tsover_mul(phi, .5);
	const ev = d.vec3f(__tsover_mul(std.cos(phi), std.cos(halfPhi)), std.sin(halfPhi), __tsover_mul(std.sin(phi), std.cos(halfPhi)));
	const vProj = std.dot(dvec, ev);
	return d.vec2f(__tsover_add(__tsover_div(phi, 6.283185307179586), .5), __tsover_add(__tsover_mul(__tsover_div(vProj, halfWidth), .5), .5));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "h"
			},
			{
				"type": "i",
				"name": "_n"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "_rp2"
			},
			{
				"type": "i",
				"name": "_t"
			}
		],
		"body": [0, [
			[
				13,
				"rp",
				[
					6,
					"f3dRot",
					[
						"h",
						"rp0",
						"rp1"
					]
				]
			],
			[
				13,
				"ringRadius",
				[
					7,
					"rp1",
					"z"
				]
			],
			[
				13,
				"halfWidth",
				[
					7,
					"rp1",
					"w"
				]
			],
			[
				13,
				"phi",
				[
					6,
					[
						7,
						"std",
						"atan2"
					],
					[[
						7,
						"rp",
						"z"
					], [
						7,
						"rp",
						"x"
					]]
				]
			],
			[
				13,
				"spine",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"cos"
								],
								["phi"]
							],
							"*",
							"ringRadius"
						],
						[5, "0"],
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								["phi"]
							],
							"*",
							"ringRadius"
						]
					]
				]
			],
			[
				13,
				"dvec",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								7,
								"rp",
								"x"
							],
							"-",
							[
								7,
								"spine",
								"x"
							]
						],
						[
							1,
							[
								7,
								"rp",
								"y"
							],
							"-",
							[
								7,
								"spine",
								"y"
							]
						],
						[
							1,
							[
								7,
								"rp",
								"z"
							],
							"-",
							[
								7,
								"spine",
								"z"
							]
						]
					]
				]
			],
			[
				13,
				"halfPhi",
				[
					1,
					"phi",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"ev",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"cos"
								],
								["phi"]
							],
							"*",
							[
								6,
								[
									7,
									"std",
									"cos"
								],
								["halfPhi"]
							]
						],
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							["halfPhi"]
						],
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								["phi"]
							],
							"*",
							[
								6,
								[
									7,
									"std",
									"cos"
								],
								["halfPhi"]
							]
						]
					]
				]
			],
			[
				13,
				"vProj",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["dvec", "ev"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						1,
						"phi",
						"/",
						[5, "6.283185307179586"]
					],
					"+",
					[5, "0.5"]
				], [
					1,
					[
						1,
						[
							1,
							"vProj",
							"/",
							"halfWidth"
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]]
			]]
		]],
		"externalNames": [
			"f3dRot",
			"std",
			"d"
		]
	},
	externals: () => ({
		f3dRot,
		std,
		d
	})
}) && $.f)({}));
var ribbonSdf = tgpu.fn([
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, rp0, rp1, _rp2, t) => {
	"use gpu";
	const angleRad = rp0.x;
	const twist = rp0.y;
	const halfW = rp0.z;
	const thickness = rp1.x;
	const phase = rp1.y;
	const lipschitz = rp1.z;
	const cosA = std.cos(angleRad);
	const sinA = std.sin(angleRad);
	const rpx = __tsover_add(__tsover_mul(p.x, cosA), __tsover_mul(p.y, sinA));
	const rpy = __tsover_add(__tsover_mul(__tsover_mul(p.x, sinA), -1), __tsover_mul(p.y, cosA));
	const twistAngle = __tsover_add(__tsover_add(__tsover_mul(rpx, twist), phase), __tsover_mul(t, TWIST_SPEED));
	const ca = std.cos(twistAngle);
	const sa = std.sin(twistAngle);
	const localU = __tsover_add(__tsover_mul(rpy, ca), __tsover_mul(p.z, sa));
	const localV = __tsover_add(__tsover_mul(__tsover_mul(rpy, sa), -1), __tsover_mul(p.z, ca));
	const du = std.max(__tsover_sub(std.abs(localU), halfW), d.f32(0));
	return __tsover_div(__tsover_sub(std.length(d.vec2f(du, localV)), thickness), lipschitz);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "p"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "_rp2"
			},
			{
				"type": "i",
				"name": "t"
			}
		],
		"body": [0, [
			[
				13,
				"angleRad",
				[
					7,
					"rp0",
					"x"
				]
			],
			[
				13,
				"twist",
				[
					7,
					"rp0",
					"y"
				]
			],
			[
				13,
				"halfW",
				[
					7,
					"rp0",
					"z"
				]
			],
			[
				13,
				"thickness",
				[
					7,
					"rp1",
					"x"
				]
			],
			[
				13,
				"phase",
				[
					7,
					"rp1",
					"y"
				]
			],
			[
				13,
				"lipschitz",
				[
					7,
					"rp1",
					"z"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"rpx",
				[
					1,
					[
						1,
						[
							7,
							"p",
							"x"
						],
						"*",
						"cosA"
					],
					"+",
					[
						1,
						[
							7,
							"p",
							"y"
						],
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"rpy",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"p",
								"x"
							],
							"*",
							"sinA"
						],
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					],
					"+",
					[
						1,
						[
							7,
							"p",
							"y"
						],
						"*",
						"cosA"
					]
				]
			],
			[
				13,
				"twistAngle",
				[
					1,
					[
						1,
						[
							1,
							"rpx",
							"*",
							"twist"
						],
						"+",
						"phase"
					],
					"+",
					[
						1,
						"t",
						"*",
						"TWIST_SPEED"
					]
				]
			],
			[
				13,
				"ca",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["twistAngle"]
				]
			],
			[
				13,
				"sa",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["twistAngle"]
				]
			],
			[
				13,
				"localU",
				[
					1,
					[
						1,
						"rpy",
						"*",
						"ca"
					],
					"+",
					[
						1,
						[
							7,
							"p",
							"z"
						],
						"*",
						"sa"
					]
				]
			],
			[
				13,
				"localV",
				[
					1,
					[
						1,
						[
							1,
							"rpy",
							"*",
							"sa"
						],
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					],
					"+",
					[
						1,
						[
							7,
							"p",
							"z"
						],
						"*",
						"ca"
					]
				]
			],
			[
				13,
				"du",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							["localU"]
						],
						"-",
						"halfW"
					], [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]]
				]
			],
			[10, [
				1,
				[
					1,
					[
						6,
						[
							7,
							"std",
							"length"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["du", "localV"]
						]]
					],
					"-",
					"thickness"
				],
				"/",
				"lipschitz"
			]]
		]],
		"externalNames": [
			"std",
			"TWIST_SPEED",
			"d"
		]
	},
	externals: () => ({
		std,
		TWIST_SPEED,
		d
	})
}) && $.f)({}));
var ribbonUV = tgpu.fn([
	d.vec3f,
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((h, _n, rp0, rp1, _rp2, t) => {
	"use gpu";
	const angleRad = rp0.x;
	const twist = rp0.y;
	const halfW = rp0.z;
	const width = rp0.w;
	const phase = rp1.y;
	const cosA = std.cos(angleRad);
	const sinA = std.sin(angleRad);
	const rpx = __tsover_add(__tsover_mul(h.x, cosA), __tsover_mul(h.y, sinA));
	const rpy = __tsover_add(__tsover_mul(__tsover_mul(h.x, sinA), -1), __tsover_mul(h.y, cosA));
	const twistAngle = __tsover_add(__tsover_add(__tsover_mul(rpx, twist), phase), __tsover_mul(t, TWIST_SPEED));
	const ca = std.cos(twistAngle);
	const sa = std.sin(twistAngle);
	const localU = __tsover_add(__tsover_mul(rpy, ca), __tsover_mul(h.z, sa));
	const localUClamped = std.max(std.min(localU, halfW), __tsover_mul(halfW, -1));
	return d.vec2f(__tsover_add(__tsover_mul(rpx, .15), .5), __tsover_add(__tsover_div(localUClamped, width), .5));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "h"
			},
			{
				"type": "i",
				"name": "_n"
			},
			{
				"type": "i",
				"name": "rp0"
			},
			{
				"type": "i",
				"name": "rp1"
			},
			{
				"type": "i",
				"name": "_rp2"
			},
			{
				"type": "i",
				"name": "t"
			}
		],
		"body": [0, [
			[
				13,
				"angleRad",
				[
					7,
					"rp0",
					"x"
				]
			],
			[
				13,
				"twist",
				[
					7,
					"rp0",
					"y"
				]
			],
			[
				13,
				"halfW",
				[
					7,
					"rp0",
					"z"
				]
			],
			[
				13,
				"width",
				[
					7,
					"rp0",
					"w"
				]
			],
			[
				13,
				"phase",
				[
					7,
					"rp1",
					"y"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"rpx",
				[
					1,
					[
						1,
						[
							7,
							"h",
							"x"
						],
						"*",
						"cosA"
					],
					"+",
					[
						1,
						[
							7,
							"h",
							"y"
						],
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"rpy",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"h",
								"x"
							],
							"*",
							"sinA"
						],
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					],
					"+",
					[
						1,
						[
							7,
							"h",
							"y"
						],
						"*",
						"cosA"
					]
				]
			],
			[
				13,
				"twistAngle",
				[
					1,
					[
						1,
						[
							1,
							"rpx",
							"*",
							"twist"
						],
						"+",
						"phase"
					],
					"+",
					[
						1,
						"t",
						"*",
						"TWIST_SPEED"
					]
				]
			],
			[
				13,
				"ca",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["twistAngle"]
				]
			],
			[
				13,
				"sa",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["twistAngle"]
				]
			],
			[
				13,
				"localU",
				[
					1,
					[
						1,
						"rpy",
						"*",
						"ca"
					],
					"+",
					[
						1,
						[
							7,
							"h",
							"z"
						],
						"*",
						"sa"
					]
				]
			],
			[
				13,
				"localUClamped",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"min"
						],
						["localU", "halfW"]
					], [
						1,
						"halfW",
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						1,
						"rpx",
						"*",
						[5, "0.15"]
					],
					"+",
					[5, "0.5"]
				], [
					1,
					[
						1,
						"localUClamped",
						"/",
						"width"
					],
					"+",
					[5, "0.5"]
				]]
			]]
		]],
		"externalNames": [
			"std",
			"TWIST_SPEED",
			"d"
		]
	},
	externals: () => ({
		std,
		TWIST_SPEED,
		d
	})
}) && $.f)({}));
var named = (fn, name) => fn.$name(name);
var SHAPES = {
	sphere: {
		sdf: named(sphereSdf, "f3dSphereSdf"),
		surfaceUV: sphereUV,
		uvHint: "f3dSphereUV",
		steps: 64,
		stepCap: 1.5,
		hitEps: .001
	},
	torus: {
		sdf: named(torusSdf, "f3dTorusSdf"),
		surfaceUV: torusUV,
		uvHint: "f3dTorusUV",
		steps: 96,
		stepCap: .5,
		hitEps: .001
	},
	box: {
		sdf: named(boxSdf, "f3dBoxSdf"),
		surfaceUV: boxUV,
		uvHint: "f3dBoxUV",
		steps: 96,
		stepCap: .5,
		hitEps: .001
	},
	capsule: {
		sdf: named(capsuleSdf, "f3dCapsuleSdf"),
		surfaceUV: capsuleUV,
		uvHint: "f3dCapsuleUV",
		steps: 64,
		stepCap: 1.5,
		hitEps: .001
	},
	mobius: {
		sdf: named(mobiusSdf, "f3dMobiusSdf"),
		surfaceUV: mobiusUV,
		uvHint: "f3dMobiusUV",
		steps: 96,
		stepCap: .3,
		hitEps: .001
	},
	ribbon: {
		sdf: named(ribbonSdf, "f3dRibbonSdf"),
		surfaceUV: ribbonUV,
		uvHint: "f3dRibbonUV",
		steps: 128,
		stepCap: .08,
		hitEps: .002
	}
};
function buildForm3dTrace(shape) {
	const STEPS = shape.steps, STEP_CAP = shape.stepCap, HIT_EPS = shape.hitEps, MAX_DIST = 20;
	const sdf = shape.sdf;
	return tgpu.fn([
		d.vec2f,
		d.f32,
		d.vec2f,
		d.f32,
		d.vec4f,
		d.vec4f,
		d.vec4f,
		d.f32
	], Distortion3DTrace)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, center, zoom, rp0, rp1, rp2, t) => {
		"use gpu";
		const cx = center.x;
		const cy = __tsover_sub(1, center.y);
		const camDist = __tsover_sub(8, __tsover_mul(zoom, .06));
		const ro = d.vec3f(0, 0, __tsover_mul(camDist, -1));
		const rd = std.normalize(d.vec3f(__tsover_mul(__tsover_sub(uv.x, cx), aspect), __tsover_sub(uv.y, cy), 1.5));
		let rayT = d.f32(0);
		let hit = d.f32(0);
		const maxDistF = d.f32(MAX_DIST);
		for (let i = 0; i < STEPS; i++) {
			const dd = sdf(ro.add(rd.mul(rayT)), rp0, rp1, rp2, t);
			if (dd < HIT_EPS) {
				hit = d.f32(1);
				break;
			}
			if (rayT > maxDistF) break;
			rayT = __tsover_add(rayT, std.min(dd, STEP_CAP));
		}
		const hitPos = ro.add(rd.mul(rayT));
		const hh = .001;
		const hn = __tsover_mul(hh, -1);
		const k0 = sdf(hitPos.add(d.vec3f(hh, hn, hn)), rp0, rp1, rp2, t);
		const k1 = sdf(hitPos.add(d.vec3f(hn, hn, hh)), rp0, rp1, rp2, t);
		const k2 = sdf(hitPos.add(d.vec3f(hn, hh, hn)), rp0, rp1, rp2, t);
		const k3 = sdf(hitPos.add(d.vec3f(hh, hh, hh)), rp0, rp1, rp2, t);
		return Distortion3DTrace({
			hitPos,
			normal: std.normalize(d.vec3f(1, -1, -1).mul(k0).add(d.vec3f(-1, -1, 1).mul(k1)).add(d.vec3f(-1, 1, -1).mul(k2)).add(d.vec3f(1, 1, 1).mul(k3))),
			rd,
			hit
		});
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [
				{
					"type": "i",
					"name": "uv"
				},
				{
					"type": "i",
					"name": "aspect"
				},
				{
					"type": "i",
					"name": "center"
				},
				{
					"type": "i",
					"name": "zoom"
				},
				{
					"type": "i",
					"name": "rp0"
				},
				{
					"type": "i",
					"name": "rp1"
				},
				{
					"type": "i",
					"name": "rp2"
				},
				{
					"type": "i",
					"name": "t"
				}
			],
			"body": [0, [
				[
					13,
					"cx",
					[
						7,
						"center",
						"x"
					]
				],
				[
					13,
					"cy",
					[
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]
				],
				[
					13,
					"camDist",
					[
						1,
						[5, "8"],
						"-",
						[
							1,
							"zoom",
							"*",
							[5, "0.06"]
						]
					]
				],
				[
					13,
					"ro",
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0"],
							[5, "0"],
							[
								1,
								"camDist",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]
						]
					]
				],
				[
					13,
					"rd",
					[
						6,
						[
							7,
							"std",
							"normalize"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									[
										1,
										[
											7,
											"uv",
											"x"
										],
										"-",
										"cx"
									],
									"*",
									"aspect"
								],
								[
									1,
									[
										7,
										"uv",
										"y"
									],
									"-",
									"cy"
								],
								[5, "1.5"]
							]
						]]
					]
				],
				[
					12,
					"rayT",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					12,
					"hit",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					13,
					"maxDistF",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["MAX_DIST"]
					]
				],
				[
					14,
					[
						12,
						"i",
						[5, "0"]
					],
					[
						1,
						"i",
						"<",
						"STEPS"
					],
					[
						102,
						"++",
						"i"
					],
					[0, [
						[
							13,
							"p",
							[
								6,
								[
									7,
									"ro",
									"add"
								],
								[[
									6,
									[
										7,
										"rd",
										"mul"
									],
									["rayT"]
								]]
							]
						],
						[
							13,
							"dd",
							[
								6,
								"sdf",
								[
									"p",
									"rp0",
									"rp1",
									"rp2",
									"t"
								]
							]
						],
						[
							11,
							[
								1,
								"dd",
								"<",
								"HIT_EPS"
							],
							[0, [[
								2,
								"hit",
								"=",
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "1"]]
								]
							], [17]]]
						],
						[
							11,
							[
								1,
								"rayT",
								">",
								"maxDistF"
							],
							[0, [[17]]]
						],
						[
							2,
							"rayT",
							"=",
							[
								1,
								"rayT",
								"+",
								[
									6,
									[
										7,
										"std",
										"min"
									],
									["dd", "STEP_CAP"]
								]
							]
						]
					]]
				],
				[
					13,
					"hitPos",
					[
						6,
						[
							7,
							"ro",
							"add"
						],
						[[
							6,
							[
								7,
								"rd",
								"mul"
							],
							["rayT"]
						]]
					]
				],
				[
					13,
					"hh",
					[5, "0.001"]
				],
				[
					13,
					"hn",
					[
						1,
						"hh",
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					]
				],
				[
					13,
					"k0",
					[
						6,
						"sdf",
						[
							[
								6,
								[
									7,
									"hitPos",
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec3f"
									],
									[
										"hh",
										"hn",
										"hn"
									]
								]]
							],
							"rp0",
							"rp1",
							"rp2",
							"t"
						]
					]
				],
				[
					13,
					"k1",
					[
						6,
						"sdf",
						[
							[
								6,
								[
									7,
									"hitPos",
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec3f"
									],
									[
										"hn",
										"hn",
										"hh"
									]
								]]
							],
							"rp0",
							"rp1",
							"rp2",
							"t"
						]
					]
				],
				[
					13,
					"k2",
					[
						6,
						"sdf",
						[
							[
								6,
								[
									7,
									"hitPos",
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec3f"
									],
									[
										"hn",
										"hh",
										"hn"
									]
								]]
							],
							"rp0",
							"rp1",
							"rp2",
							"t"
						]
					]
				],
				[
					13,
					"k3",
					[
						6,
						"sdf",
						[
							[
								6,
								[
									7,
									"hitPos",
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec3f"
									],
									[
										"hh",
										"hh",
										"hh"
									]
								]]
							],
							"rp0",
							"rp1",
							"rp2",
							"t"
						]
					]
				],
				[
					13,
					"normal",
					[
						6,
						[
							7,
							"std",
							"normalize"
						],
						[[
							6,
							[
								7,
								[
									6,
									[
										7,
										[
											6,
											[
												7,
												[
													6,
													[
														7,
														[
															6,
															[
																7,
																"d",
																"vec3f"
															],
															[
																[5, "1"],
																[
																	4,
																	"-",
																	[5, "1"]
																],
																[
																	4,
																	"-",
																	[5, "1"]
																]
															]
														],
														"mul"
													],
													["k0"]
												],
												"add"
											],
											[[
												6,
												[
													7,
													[
														6,
														[
															7,
															"d",
															"vec3f"
														],
														[
															[
																4,
																"-",
																[5, "1"]
															],
															[
																4,
																"-",
																[5, "1"]
															],
															[5, "1"]
														]
													],
													"mul"
												],
												["k1"]
											]]
										],
										"add"
									],
									[[
										6,
										[
											7,
											[
												6,
												[
													7,
													"d",
													"vec3f"
												],
												[
													[
														4,
														"-",
														[5, "1"]
													],
													[5, "1"],
													[
														4,
														"-",
														[5, "1"]
													]
												]
											],
											"mul"
										],
										["k2"]
									]]
								],
								"add"
							],
							[[
								6,
								[
									7,
									[
										6,
										[
											7,
											"d",
											"vec3f"
										],
										[
											[5, "1"],
											[5, "1"],
											[5, "1"]
										]
									],
									"mul"
								],
								["k3"]
							]]
						]]
					]
				],
				[10, [
					6,
					"Distortion3DTrace",
					[[104, {
						"hitPos": "hitPos",
						"normal": "normal",
						"rd": "rd",
						"hit": "hit"
					}]]
				]]
			]],
			"externalNames": [
				"d",
				"std",
				"MAX_DIST",
				"STEPS",
				"sdf",
				"HIT_EPS",
				"STEP_CAP",
				"Distortion3DTrace"
			]
		},
		externals: () => ({
			d,
			std,
			MAX_DIST,
			STEPS,
			sdf,
			HIT_EPS,
			STEP_CAP,
			Distortion3DTrace
		})
	}) && $.f)({}));
}
var SPIN_DEFAULTS = {
	torus: {
		rotX: -90,
		spinY: .05
	},
	box: {
		rotX: 15,
		spinY: .1
	},
	sphere: { spinY: .1 },
	capsule: { spinY: .1 },
	mobius: {
		rotX: -30,
		spinY: .05
	}
};
const componentDefinition = {
	name: "Form3D",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Distortions",
	description: "Wraps child content onto a 3D raymarched shape with lighting.",
	requiresRTT: true,
	requiresChild: true,
	animatedTime: { speed: "speed" },
	extraFields: {
		_f3p0: {
			schema: d.vec4f,
			initial: [
				1,
				0,
				1,
				0
			]
		},
		_f3p1: {
			schema: d.vec4f,
			initial: [
				1,
				0,
				.35,
				.1
			]
		},
		_f3p2: {
			schema: d.vec4f,
			initial: [
				.3,
				.02,
				0,
				0
			]
		}
	},
	props: {
		shape3d: {
			default: DEFAULT_RIBBON,
			description: "3D shape and its parameters",
			ui: {
				type: "shape3d",
				label: "Shape",
				group: "Shape"
			}
		},
		shape3dType: {
			default: "ribbon",
			compileTime: true,
			description: "Active shape type — triggers recompile when shape is switched"
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the shape on screen",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		zoom: {
			default: 50,
			description: "Camera zoom level",
			ui: {
				type: ["range", "map"],
				min: 10,
				max: 200,
				step: 1,
				label: "Zoom",
				group: "Scene"
			}
		},
		glossiness: {
			default: 50,
			description: "Specular highlight intensity and sharpness",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 200,
				step: 1,
				label: "Glossiness",
				group: "Lighting"
			}
		},
		lighting: {
			default: 50,
			description: "Overall intensity of lighting effects",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 200,
				step: 1,
				label: "Lighting",
				group: "Lighting"
			}
		},
		uvMode: {
			default: "stretch",
			transform: transformEdges,
			description: "How to handle UV coordinates at shape boundaries",
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "UV Edges",
				group: "Scene"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed — scales all spin rates",
			ui: {
				type: "range",
				min: -10,
				max: 10,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		}
	},
	fragment: (params) => {
		const { childNode, convertToTexture, uniforms, propValues, getCpuValue, onBeforeRender, ctx } = params;
		if (!childNode) return ZERO;
		const childTexture = convertToTexture(childNode);
		const shapeType = propValues.shape3dType || "ribbon";
		const shape = SHAPES[shapeType] ?? SHAPES.ribbon;
		let spinX = 0, spinY = 0, spinZ = 0;
		let lastJson = "";
		let cfg = (() => {
			try {
				return JSON.parse(getCpuValue("shape3d"));
			} catch {
				return {};
			}
		})();
		const num = (k, f) => typeof cfg[k] === "number" ? cfg[k] : f;
		const drive = (deltaTime) => {
			const raw = getCpuValue("shape3d");
			if (typeof raw === "string") {
				if (raw !== lastJson) {
					lastJson = raw;
					try {
						cfg = JSON.parse(raw);
					} catch {}
				}
			} else if (raw && typeof raw === "object") cfg = raw;
			const speed = getCpuValue("speed") ?? 1;
			if (shapeType === "ribbon") {
				const angle = num("angle", 0) * DEG;
				const twist = num("twist", 50) * .03;
				const width = num("width", 40) * .012;
				const halfW = width * .5;
				const thickness = num("thickness", 20) * .001 + .01;
				const seed = num("seed", 0);
				const lipschitz = Math.max(halfW * twist, 1);
				params.setExtraField("_f3p0", [
					angle,
					twist,
					halfW,
					width
				]);
				params.setExtraField("_f3p1", [
					thickness,
					seed,
					lipschitz,
					0
				]);
				params.setExtraField("_f3p2", [
					0,
					0,
					0,
					0
				]);
				return;
			}
			const dfl = SPIN_DEFAULTS[shapeType] ?? {};
			spinX += deltaTime * speed * num("spinX", dfl.spinX ?? 0) * TAU_RATE;
			spinY += deltaTime * speed * num("spinY", dfl.spinY ?? 0) * TAU_RATE;
			spinZ += deltaTime * speed * num("spinZ", dfl.spinZ ?? 0) * TAU_RATE;
			const aX = num("rotX", dfl.rotX ?? 0) * DEG + spinX;
			const aY = num("rotY", dfl.rotY ?? 0) * DEG + spinY;
			const aZ = num("rotZ", dfl.rotZ ?? 0) * DEG + spinZ;
			const rot0 = [
				Math.cos(aX),
				Math.sin(aX),
				Math.cos(aY),
				Math.sin(aY)
			];
			const cz = Math.cos(aZ), sz = Math.sin(aZ);
			if (shapeType === "sphere") {
				params.setExtraField("_f3p0", rot0);
				params.setExtraField("_f3p1", [
					cz,
					sz,
					num("radius", 60) * .008 + .1,
					0
				]);
				params.setExtraField("_f3p2", [
					0,
					0,
					0,
					0
				]);
			} else if (shapeType === "torus") {
				params.setExtraField("_f3p0", rot0);
				params.setExtraField("_f3p1", [
					cz,
					sz,
					num("outerRadius", 60) * .008 + .1,
					num("tubeRadius", 25) * .006 + .05
				]);
				params.setExtraField("_f3p2", [
					0,
					0,
					0,
					0
				]);
			} else if (shapeType === "box") {
				params.setExtraField("_f3p0", rot0);
				params.setExtraField("_f3p1", [
					cz,
					sz,
					num("sizeX", 50) * .006,
					num("sizeY", 50) * .006
				]);
				params.setExtraField("_f3p2", [
					num("sizeZ", 50) * .006,
					num("rounding", 20) * .001,
					0,
					0
				]);
			} else if (shapeType === "capsule") {
				params.setExtraField("_f3p0", rot0);
				params.setExtraField("_f3p1", [
					cz,
					sz,
					num("radius", 25) * .006 + .05,
					num("height", 80) * .005
				]);
				params.setExtraField("_f3p2", [
					0,
					0,
					0,
					0
				]);
			} else {
				params.setExtraField("_f3p0", rot0);
				params.setExtraField("_f3p1", [
					cz,
					sz,
					num("ringRadius", 60) * .008 + .1,
					num("halfWidth", 30) * .005
				]);
				params.setExtraField("_f3p2", [
					num("thickness", 8) * .001 + .005,
					0,
					0,
					0
				]);
			}
		};
		drive(0);
		onBeforeRender((fp) => drive(fp?.deltaTime ?? 0));
		const t = animatedTime(params);
		const traceRes = call(buildForm3dTrace(shape), `form3dTrace_${shapeType}`, [
			ctx.uv,
			ctx.aspect,
			uniforms.center,
			uniforms.zoom,
			uniforms._f3p0,
			uniforms._f3p1,
			uniforms._f3p2,
			t
		]);
		const hitPos = traceRes.member("hitPos");
		const normal = traceRes.member("normal");
		const rd = traceRes.member("rd");
		const hit = traceRes.member("hit");
		const uvm = call(applyUVMode3d, "applyUVMode3d", [call(shape.surfaceUV, shape.uvHint, [
			hitPos,
			normal,
			uniforms._f3p0,
			uniforms._f3p1,
			uniforms._f3p2,
			t
		]), uniforms.uvMode]);
		const finalUV = uvm.member("xy");
		const uvVisible = uvm.member("z");
		return call(distortion3dLighting, "distortion3dLighting", [
			call(unpremultiplyAlpha, "unpremultiplyAlpha", [childTexture.sample(finalUV)]),
			normal,
			rd,
			hit,
			uvVisible,
			uniforms.glossiness.mul(.005),
			uniforms.lighting.mul(.005)
		]);
	}
};
var Form3D_default = componentDefinition;
export { componentDefinition as n, f3dRot as r, Form3D_default as t };
