// shaders/dist/core/FractalNoise-BFs55FIe.js
import { At as call, R as std, U as mixStopsOrColorsExpr, et as mxNoiseFloat2, n as tgpu, on as d, rt as noise_exports } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig, t as animatedTime } from "./porters-Dg80Yg90.js";
var GOLDEN = 2.39996;
const fractalNoiseValue = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, angle, animTime, seed, detail, contrast, octaveCount) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const angleRad = __tsover_mul(angle, .0174533);
	const cosA = std.cos(angleRad);
	const sinA = std.sin(angleRad);
	const au = __tsover_sub(__tsover_mul(uv.x, aspect), __tsover_mul(aspect, .5));
	const av = __tsover_sub(uv.y, .5);
	const ru = __tsover_add(__tsover_mul(au, cosA), __tsover_mul(av, sinA));
	const rv = __tsover_add(__tsover_mul(au, -sinA), __tsover_mul(av, cosA));
	const n = d.i32(octaveCount);
	let octaveScale = d.f32(2);
	let weight = d.f32(1);
	let accumulated = d.f32(0);
	let totalWeight = d.f32(0);
	for (let i = 0; i < 8; i++) {
		const fi = d.f32(i);
		const dirX = std.cos(__tsover_mul(fi, GOLDEN));
		const dirY = std.sin(__tsover_mul(fi, GOLDEN));
		const coord = d.vec2f(__tsover_add(__tsover_add(__tsover_mul(ru, octaveScale), __tsover_mul(animTime, dirX)), seed), __tsover_add(__tsover_add(__tsover_mul(rv, octaveScale), __tsover_mul(animTime, dirY)), __tsover_mul(seed, .7)));
		const active = std.select(d.f32(0), d.f32(1), i < n);
		accumulated = __tsover_add(accumulated, __tsover_mul(__tsover_mul(mxNoiseFloat2(coord), weight), active));
		totalWeight = __tsover_add(totalWeight, __tsover_mul(weight, active));
		octaveScale = __tsover_mul(octaveScale, detail);
		weight = __tsover_mul(weight, contrast);
	}
	return std.clamp(__tsover_add(__tsover_mul(__tsover_div(accumulated, totalWeight), .5), .5), 0, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "detail"
			},
			{
				"type": "i",
				"name": "contrast"
			},
			{
				"type": "i",
				"name": "octaveCount"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					[5, "0.0174533"]
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"au",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					],
					"-",
					[
						1,
						"aspect",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"av",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"ru",
				[
					1,
					[
						1,
						"au",
						"*",
						"cosA"
					],
					"+",
					[
						1,
						"av",
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"rv",
				[
					1,
					[
						1,
						"au",
						"*",
						[
							4,
							"-",
							"sinA"
						]
					],
					"+",
					[
						1,
						"av",
						"*",
						"cosA"
					]
				]
			],
			[
				13,
				"n",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["octaveCount"]
				]
			],
			[
				12,
				"octaveScale",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "2"]]
				]
			],
			[
				12,
				"weight",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "1"]]
				]
			],
			[
				12,
				"accumulated",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0"]]
				]
			],
			[
				12,
				"totalWeight",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0"]]
				]
			],
			[
				14,
				[
					12,
					"i",
					[5, "0"]
				],
				[
					1,
					"i",
					"<",
					[5, "8"]
				],
				[
					102,
					"++",
					"i"
				],
				[0, [
					[
						13,
						"fi",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["i"]
						]
					],
					[
						13,
						"dirX",
						[
							6,
							[
								7,
								"std",
								"cos"
							],
							[[
								1,
								"fi",
								"*",
								"GOLDEN"
							]]
						]
					],
					[
						13,
						"dirY",
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								"fi",
								"*",
								"GOLDEN"
							]]
						]
					],
					[
						13,
						"coord",
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								[
									1,
									[
										1,
										"ru",
										"*",
										"octaveScale"
									],
									"+",
									[
										1,
										"animTime",
										"*",
										"dirX"
									]
								],
								"+",
								"seed"
							], [
								1,
								[
									1,
									[
										1,
										"rv",
										"*",
										"octaveScale"
									],
									"+",
									[
										1,
										"animTime",
										"*",
										"dirY"
									]
								],
								"+",
								[
									1,
									"seed",
									"*",
									[5, "0.7"]
								]
							]]
						]
					],
					[
						13,
						"active",
						[
							6,
							[
								7,
								"std",
								"select"
							],
							[
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "0"]]
								],
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "1"]]
								],
								[
									1,
									"i",
									"<",
									"n"
								]
							]
						]
					],
					[
						2,
						"accumulated",
						"=",
						[
							1,
							"accumulated",
							"+",
							[
								1,
								[
									1,
									[
										6,
										[
											7,
											"noise",
											"mxNoiseFloat2"
										],
										["coord"]
									],
									"*",
									"weight"
								],
								"*",
								"active"
							]
						]
					],
					[
						2,
						"totalWeight",
						"=",
						[
							1,
							"totalWeight",
							"+",
							[
								1,
								"weight",
								"*",
								"active"
							]
						]
					],
					[
						2,
						"octaveScale",
						"=",
						[
							1,
							"octaveScale",
							"*",
							"detail"
						]
					],
					[
						2,
						"weight",
						"=",
						[
							1,
							"weight",
							"*",
							"contrast"
						]
					]
				]]
			],
			[10, [
				6,
				[
					7,
					"std",
					"clamp"
				],
				[
					[
						1,
						[
							1,
							[
								1,
								"accumulated",
								"/",
								"totalWeight"
							],
							"*",
							[5, "0.5"]
						],
						"+",
						[5, "0.5"]
					],
					[5, "0"],
					[5, "1"]
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"GOLDEN",
			"noise"
		]
	},
	externals: () => ({
		std,
		d,
		GOLDEN,
		noise: noise_exports
	})
}) && $.f)({}));
const componentDefinition = {
	name: "FractalNoise",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Multi-octave fractal Brownian motion noise texture with true noise evolution",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "First color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Second color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		stops: colorStopsPropConfig(),
		octaves: {
			default: 4,
			compileTime: true,
			description: "Number of noise octaves (more = more detail)",
			ui: {
				type: "range",
				min: 1,
				max: 8,
				step: 1,
				label: "Octaves",
				group: "Effect"
			}
		},
		detail: {
			default: 2,
			description: "How much finer each successive octave becomes",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 4,
				step: .1,
				label: "Detail",
				group: "Effect"
			}
		},
		contrast: {
			default: .5,
			description: "How strongly finer octaves contribute — higher values create more texture contrast",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 1,
				step: .01,
				label: "Contrast",
				group: "Effect"
			}
		},
		speed: {
			default: .15,
			description: "Speed at which the noise pattern evolves in place",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Rotation angle in degrees",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for pattern variation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		return mixStopsOrColorsExpr(params, call(fractalNoiseValue, "fractalNoiseValue", [
			uv,
			viewport,
			params.uniforms.angle,
			t,
			params.uniforms.seed,
			params.uniforms.detail,
			params.uniforms.contrast,
			params.uniforms.octaves
		]));
	}
};
var FractalNoise_default = componentDefinition;
export { componentDefinition as n, fractalNoiseValue as r, FractalNoise_default as t };
