// shaders/dist/core/Glass-CJFR7oIt.js
import { I as createGaussianBlurCompute, Ot as ZERO, T as sdf_exports, d as sdf3d_exports } from "./coords-BlwgBpJf.js";
import { n as glass_exports } from "./glass-DdgMBWkF.js";
import { a as transformColor, c as transformPosition, i as transformBoolean } from "./transformations-fqzznt4g.js";
import { t as shapeEffectBoundingBoxDeclaration } from "./shapeEffectBounds-BlxxsYMN.js";
var { applyGlassEffect } = glass_exports;
var { createSvgSdfSamplerWithGradients, createAnalyticSdfSampler, driveAnalyticSubProps, ANALYTIC_SDF_EXTRA_FIELDS } = sdf_exports;
var { createVolumetricFieldComputeNode, buildVolumetricFieldSampler, VOLUMETRIC_FIELD_EXTRA_FIELDS, is3dShapeType, isVolumetricShapeType } = sdf3d_exports;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
function resolveShapeType(shapeType, shape) {
	if (shapeType) return shapeType;
	if (shape && typeof shape === "object") return shape.type || "circleSDF";
	if (typeof shape === "string") try {
		return JSON.parse(shape).type || "circleSDF";
	} catch {
		return "circleSDF";
	}
	return "circleSDF";
}
const componentDefinition = {
	name: "Glass",
	boundingBoxDeclaration: shapeEffectBoundingBoxDeclaration,
	category: "Shape Effects",
	description: "Optically realistic glass lens driven in a custom shape",
	requiresRTT: true,
	requiresChild: true,
	extraFields: {
		...VOLUMETRIC_FIELD_EXTRA_FIELDS,
		...ANALYTIC_SDF_EXTRA_FIELDS
	},
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the glass shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		scale: {
			default: 1,
			description: "Scale of the glass shape (1 = default size)",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 3,
				step: .01,
				label: "Scale",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the glass shape in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		},
		cutout: {
			default: false,
			description: "Cut out the alpha outside the glass shape",
			transform: transformBoolean,
			compileTime: true,
			ui: {
				type: "checkbox",
				label: "Cutout",
				group: "Glass"
			}
		},
		refraction: {
			default: 1,
			description: "Lens refraction — how aggressively the edges warp content beneath (0 = none, 1 = max)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .01,
				label: "Refraction",
				group: "Glass"
			}
		},
		edgeSoftness: {
			default: .1,
			description: "Edge softness — higher values give a wider, softer fade at the glass boundary",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Softness",
				group: "Glass"
			}
		},
		blur: {
			default: 0,
			compileTimeWhen: (prev, next) => prev > 0 !== next > 0,
			description: "Frosted blur amount — 0 = clear glass, higher = frosted/diffuse",
			ui: {
				type: "range",
				min: 0,
				max: 20,
				step: .1,
				label: "Blur",
				group: "Glass"
			}
		},
		thickness: {
			default: .2,
			transform: (v) => v * .5,
			description: "Glass depth — how far inward from the edge the refraction extends",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Thickness",
				group: "Glass"
			}
		},
		aberration: {
			default: .5,
			compileTimeWhen: (prev, next) => prev > 0 !== next > 0,
			description: "Chromatic aberration — splits RGB channels along the refraction vector",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Aberration",
				group: "Glass"
			}
		},
		innerZoom: {
			default: 1,
			description: "Inner zoom level — magnifies content seen through the glass",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 3,
				step: .01,
				label: "Inner Zoom",
				group: "Glass"
			}
		},
		lightAngle: {
			default: 300,
			description: "Light angle in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Light Angle",
				group: "Highlight"
			}
		},
		highlight: {
			default: .05,
			description: "Directional edge highlight — bright rim on the light-facing boundary",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .01,
				label: "Highlight",
				group: "Highlight"
			}
		},
		highlightColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color of the directional edge highlight and specular glint",
			ui: {
				type: "color",
				label: "Highlight Color",
				group: "Highlight"
			}
		},
		highlightSoftness: {
			default: .5,
			description: "Specular highlight softness",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Highlight Softness",
				group: "Highlight"
			}
		},
		fresnel: {
			default: .1,
			description: "Fresnel rim glow — a soft luminous halo around the glass boundary",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Fresnel",
				group: "Fresnel"
			}
		},
		fresnelSoftness: {
			default: .1,
			description: "Fresnel rim width — higher values spread the glow further inward",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Fresnel Softness",
				group: "Fresnel"
			}
		},
		fresnelColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color of the fresnel rim glow",
			ui: {
				type: "color",
				label: "Fresnel Color",
				group: "Fresnel"
			}
		},
		tintColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color tint applied to the internal directional gradient",
			ui: {
				type: "color",
				label: "Tint Color",
				group: "Tint"
			}
		},
		tintIntensity: {
			default: 0,
			description: "Intensity of the color tint applied to the glass interior",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Tint Intensity",
				group: "Tint"
			}
		},
		tintPreserveLuminosity: {
			default: true,
			transform: transformBoolean,
			description: "Preserve original brightness when tinting",
			ui: {
				type: "checkbox",
				label: "Preserve Luminosity",
				group: "Tint"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file — when non-empty, activates SVG mode and triggers a shader recompile"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type — triggers recompile when shape is switched. When empty, derived from shape JSON at mount time."
		}
	},
	compute: (params) => {
		const { childNode, gpu, convertToTexture, registerComputeTexture, getCpuValue, onCleanup, onResize, dimensions } = params;
		const root = gpu?.root;
		const fieldResult = createVolumetricFieldComputeNode(params, () => getCpuValue("shape"), "none");
		let blurResult = null;
		const blurValue = getCpuValue("blur");
		if (childNode && root && !(typeof blurValue === "number" && blurValue <= 0)) {
			const childTexture = convertToTexture(childNode);
			let curWidth = Math.max(1, Math.round(dimensions.width));
			let curHeight = Math.max(1, Math.round(dimensions.height));
			const gaussian = createGaussianBlurCompute(root, null, curWidth, curHeight, onCleanup);
			const blurredTexture = registerComputeTexture(gaussian.outputTexture);
			onResize(({ width, height }) => {
				curWidth = Math.max(1, Math.round(width));
				curHeight = Math.max(1, Math.round(height));
				gaussian.setInputDimensions(curWidth, curHeight);
			});
			blurResult = {
				outputs: {
					childTexture,
					blurredTexture
				},
				bindInputs: (resolve) => {
					const src = resolve(childTexture.key);
					if (src) gaussian.setInputTexture(src.texture);
				},
				getComputeNodes: () => {
					const b = getCpuValue("blur");
					if (typeof b === "number") gaussian.updateRadius(b * 2);
					return gaussian.computeSteps;
				}
			};
		}
		if (!fieldResult && !blurResult) return null;
		return {
			outputs: {
				...blurResult?.outputs,
				...fieldResult?.outputs
			},
			bindInputs: blurResult?.bindInputs,
			getComputeNodes: (frameParams) => {
				const nodes = [...blurResult?.getComputeNodes() ?? [], ...fieldResult?.getComputeNodes(frameParams) ?? []];
				return nodes.length ? nodes : null;
			}
		};
	},
	fragment: (params) => {
		const { childNode, computeOutputs, convertToTexture, uniforms, propValues, getCpuValue } = params;
		if (!childNode) return ZERO;
		const childTexture = computeOutputs?.childTexture ?? convertToTexture(childNode);
		const blurredTexture = computeOutputs?.blurredTexture;
		const fieldTex = computeOutputs?.volumetricFieldTexture;
		if (fieldTex) return applyGlassEffect(params, buildVolumetricFieldSampler(fieldTex, {
			originX: uniforms._vfOriginX,
			originY: uniforms._vfOriginY,
			spanX: uniforms._vfSpanX,
			spanY: uniforms._vfSpanY,
			activeResF: uniforms._vfActiveRes,
			rBound: uniforms._vfRBound
		}).volumetricFieldSampler, childTexture, blurredTexture, { volumetric: true });
		const shapeSdfUrl = propValues.shapeSdfUrl || "";
		const shapeType = propValues.shapeType || "";
		if (shapeSdfUrl) return applyGlassEffect(params, createSvgSdfSamplerWithGradients(params, shapeSdfUrl), childTexture, blurredTexture, {
			volumetric: isVolumetricShapeType(shapeType),
			bakedGradients: true
		});
		const st = resolveShapeType(shapeType, getCpuValue("shape"));
		return applyGlassEffect(params, createAnalyticSdfSampler(st, driveAnalyticSubProps(params, () => getCpuValue("shape"))), childTexture, blurredTexture, { volumetric: is3dShapeType(st) });
	}
};
var Glass_default = componentDefinition;
export { componentDefinition as n, Glass_default as t };
