// shaders/dist/core/GlassTiles-BSBCjkhG.js
import { At as call, Ot as ZERO, R as std, St as edgeClampUV, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
var DEG_TO_RAD = Math.PI / 180;
const glassTilesUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, intensity, baseTileCount, rotationDegrees, roundnessAmount) => {
	"use gpu";
	const isWide = aspect > 1;
	const tileCount = d.vec2f(std.select(__tsover_mul(baseTileCount, aspect), baseTileCount, isWide), std.select(baseTileCount, __tsover_div(baseTileCount, aspect), isWide));
	const aspectCorrectedUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const rotationRadians = __tsover_mul(rotationDegrees, DEG_TO_RAD);
	const cosA = std.cos(rotationRadians);
	const sinA = std.sin(rotationRadians);
	const centered = aspectCorrectedUV.sub(d.vec2f(__tsover_mul(.5, aspect), .5));
	const rotatedX = __tsover_sub(__tsover_mul(centered.x, cosA), __tsover_mul(centered.y, sinA));
	const rotatedY = __tsover_add(__tsover_mul(centered.x, sinA), __tsover_mul(centered.y, cosA));
	const rotatedUV = d.vec2f(rotatedX, rotatedY).add(d.vec2f(__tsover_mul(.5, aspect), .5));
	const gridUV = d.vec2f(__tsover_div(rotatedUV.x, aspect), rotatedUV.y);
	const cellCoord = std.floor(gridUV.mul(tileCount)).div(tileCount);
	const localUV = gridUV.sub(cellCoord);
	const oneOverTile = d.vec2f(1, 1).div(tileCount);
	const fromCenter = localUV.div(oneOverTile).sub(d.vec2f(.5, .5));
	const distFromCenter = __tsover_add(__tsover_mul(fromCenter.x, fromCenter.x), __tsover_mul(fromCenter.y, fromCenter.y));
	const roundnessFactor = __tsover_sub(1, __tsover_mul(distFromCenter, __tsover_mul(roundnessAmount, 4)));
	const clampedRoundness = std.max(roundnessFactor, 0);
	const distortionFactor = __tsover_mul(intensity, .025);
	const scaleFactor = __tsover_mul(distortionFactor, clampedRoundness);
	const distortion = d.vec2f(__tsover_div(__tsover_mul(fromCenter.x, scaleFactor), aspect), __tsover_mul(fromCenter.y, scaleFactor));
	return uv.add(distortion);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "baseTileCount"
			},
			{
				"type": "i",
				"name": "rotationDegrees"
			},
			{
				"type": "i",
				"name": "roundnessAmount"
			}
		],
		"body": [0, [
			[
				13,
				"isWide",
				[
					1,
					"aspect",
					">",
					[5, "1"]
				]
			],
			[
				13,
				"tileCount",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								1,
								"baseTileCount",
								"*",
								"aspect"
							],
							"baseTileCount",
							"isWide"
						]
					], [
						6,
						[
							7,
							"std",
							"select"
						],
						[
							"baseTileCount",
							[
								1,
								"baseTileCount",
								"/",
								"aspect"
							],
							"isWide"
						]
					]]
				]
			],
			[
				13,
				"aspectCorrectedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"rotationRadians",
				[
					1,
					"rotationDegrees",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["rotationRadians"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["rotationRadians"]
				]
			],
			[
				13,
				"centered",
				[
					6,
					[
						7,
						"aspectCorrectedUV",
						"sub"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[5, "0.5"],
							"*",
							"aspect"
						], [5, "0.5"]]
					]]
				]
			],
			[
				13,
				"rotatedX",
				[
					1,
					[
						1,
						[
							7,
							"centered",
							"x"
						],
						"*",
						"cosA"
					],
					"-",
					[
						1,
						[
							7,
							"centered",
							"y"
						],
						"*",
						"sinA"
					]
				]
			],
			[
				13,
				"rotatedY",
				[
					1,
					[
						1,
						[
							7,
							"centered",
							"x"
						],
						"*",
						"sinA"
					],
					"+",
					[
						1,
						[
							7,
							"centered",
							"y"
						],
						"*",
						"cosA"
					]
				]
			],
			[
				13,
				"rotatedUV",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["rotatedX", "rotatedY"]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[5, "0.5"],
							"*",
							"aspect"
						], [5, "0.5"]]
					]]
				]
			],
			[
				13,
				"gridUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"rotatedUV",
							"x"
						],
						"/",
						"aspect"
					], [
						7,
						"rotatedUV",
						"y"
					]]
				]
			],
			[
				13,
				"cellCoord",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								6,
								[
									7,
									"gridUV",
									"mul"
								],
								["tileCount"]
							]]
						],
						"div"
					],
					["tileCount"]
				]
			],
			[
				13,
				"localUV",
				[
					6,
					[
						7,
						"gridUV",
						"sub"
					],
					["cellCoord"]
				]
			],
			[
				13,
				"oneOverTile",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "1"], [5, "1"]]
						],
						"div"
					],
					["tileCount"]
				]
			],
			[
				13,
				"normalizedLocalUV",
				[
					6,
					[
						7,
						"localUV",
						"div"
					],
					["oneOverTile"]
				]
			],
			[
				13,
				"fromCenter",
				[
					6,
					[
						7,
						"normalizedLocalUV",
						"sub"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "0.5"], [5, "0.5"]]
					]]
				]
			],
			[
				13,
				"distFromCenter",
				[
					1,
					[
						1,
						[
							7,
							"fromCenter",
							"x"
						],
						"*",
						[
							7,
							"fromCenter",
							"x"
						]
					],
					"+",
					[
						1,
						[
							7,
							"fromCenter",
							"y"
						],
						"*",
						[
							7,
							"fromCenter",
							"y"
						]
					]
				]
			],
			[
				13,
				"roundnessFactor",
				[
					1,
					[5, "1"],
					"-",
					[
						1,
						"distFromCenter",
						"*",
						[
							1,
							"roundnessAmount",
							"*",
							[5, "4"]
						]
					]
				]
			],
			[
				13,
				"clampedRoundness",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["roundnessFactor", [5, "0"]]
				]
			],
			[
				13,
				"distortionFactor",
				[
					1,
					"intensity",
					"*",
					[5, "0.025"]
				]
			],
			[
				13,
				"scaleFactor",
				[
					1,
					"distortionFactor",
					"*",
					"clampedRoundness"
				]
			],
			[
				13,
				"distortion",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								7,
								"fromCenter",
								"x"
							],
							"*",
							"scaleFactor"
						],
						"/",
						"aspect"
					], [
						1,
						[
							7,
							"fromCenter",
							"y"
						],
						"*",
						"scaleFactor"
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"uv",
					"add"
				],
				["distortion"]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"DEG_TO_RAD"
		]
	},
	externals: () => ({
		d,
		std,
		DEG_TO_RAD
	})
}) && $.f)({}));
const componentDefinition = {
	name: "GlassTiles",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Distortions",
	description: "Refraction-like distortion in a tile grid pattern",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 2,
			description: "The intensity of the glass tiles effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 10,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		tileCount: {
			default: 20,
			description: "Number of tiles across the longest dimension",
			ui: {
				type: ["range", "map"],
				min: 5,
				max: 50,
				step: 1,
				label: "Tile Count",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation angle of the tile grid in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		roundness: {
			default: 0,
			description: "Makes tiles more circular instead of square",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Roundness",
				group: "Effect"
			}
		}
	},
	fragment: ({ childNode, ctx, uniforms, convertToTexture }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Glass Tiles shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const distortedUV = call(glassTilesUV, "glassTilesUV", [
			ctx.uv,
			ctx.aspect,
			uniforms.intensity,
			uniforms.tileCount,
			uniforms.rotation,
			uniforms.roundness
		]);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [texture.sample(distortedUV)]);
	},
	uvRemap: ({ uv, mask, uniforms, aspect }) => {
		return {
			uv: call(edgeClampUV, "edgeClampUV", [call(glassTilesUV, "glassTilesUV", [
				uv,
				aspect,
				uniforms.intensity,
				uniforms.tileCount,
				uniforms.rotation,
				uniforms.roundness
			])]),
			mask
		};
	}
};
var GlassTiles_default = componentDefinition;
export { componentDefinition as n, glassTilesUV as r, GlassTiles_default as t };
