// shaders/dist/core/Glitch-B93JzM-d.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
var hash11 = tgpu.fn([d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	return std.fract(__tsover_mul(std.sin(__tsover_mul(p, 78.233)), 43758.5453));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				1,
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						1,
						"p",
						"*",
						[5, "78.233"]
					]]
				],
				"*",
				[5, "43758.5453"]
			]]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const GlitchGeom = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
	redUV: d.vec2f,
	mirroredUV: d.vec2f,
	blueUV: d.vec2f,
	fillColor: d.vec3f,
	colorBarStrength: d.f32,
	scanlineEffect: d.f32
}), "GlitchGeom");
const glitchGeom = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], GlitchGeom)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, time, intensity, speed, rgbShift, blockDensity, colorBarMix, mirrorChance, scanlines) => {
	"use gpu";
	const t = __tsover_mul(time, speed);
	const slowFrame = std.floor(__tsover_mul(t, 2.5));
	const slowHash = hash11(slowFrame);
	const slowPulse = __tsover_mul(__tsover_mul(slowHash, slowHash), slowHash);
	const fastFrame = std.floor(__tsover_mul(t, 10));
	const fastHash = hash11(__tsover_add(fastFrame, 73));
	const fastPulse = __tsover_mul(fastHash, fastHash);
	const threshold = __tsover_sub(1, intensity);
	const majorBurst = std.step(threshold, slowPulse);
	const minorFlicker = std.step(__tsover_add(threshold, .15), fastPulse);
	const glitchStrength = __tsover_mul(std.clamp(__tsover_add(__tsover_mul(majorBurst, .8), __tsover_mul(minorFlicker, .4)), 0, 1), intensity);
	const bandY1 = std.floor(__tsover_mul(uv.y, blockDensity));
	const bandSeed1 = __tsover_add(__tsover_mul(bandY1, .37), __tsover_mul(slowFrame, 7.13));
	const bandHash1 = hash11(bandSeed1);
	const bandActive1 = std.step(__tsover_sub(1, __tsover_mul(glitchStrength, .5)), bandHash1);
	const displace1 = __tsover_mul(__tsover_mul(__tsover_sub(hash11(__tsover_add(bandSeed1, 42)), .5), .25), bandActive1);
	const bandY2 = std.floor(__tsover_mul(uv.y, __tsover_mul(blockDensity, 2)));
	const bandSeed2 = __tsover_add(__tsover_mul(bandY2, .53), __tsover_mul(fastFrame, 3.37));
	const bandHash2 = hash11(bandSeed2);
	const bandActive2 = std.step(__tsover_sub(1, __tsover_mul(glitchStrength, .7)), bandHash2);
	const displace2 = __tsover_mul(__tsover_mul(__tsover_sub(hash11(__tsover_add(bandSeed2, 99)), .5), .12), bandActive2);
	const bandY3 = std.floor(__tsover_mul(uv.y, __tsover_mul(blockDensity, 4)));
	const bandSeed3 = __tsover_add(__tsover_mul(bandY3, .71), __tsover_mul(fastFrame, 11.71));
	const bandHash3 = hash11(bandSeed3);
	const bandActive3 = std.step(__tsover_sub(1, __tsover_mul(glitchStrength, .9)), bandHash3);
	const displace3 = __tsover_mul(__tsover_mul(__tsover_sub(hash11(__tsover_add(bandSeed3, 173)), .5), .05), bandActive3);
	const colX = std.floor(__tsover_mul(uv.x, 4));
	const blockSeed = __tsover_add(__tsover_add(__tsover_mul(colX, .41), __tsover_mul(bandY1, .67)), __tsover_mul(slowFrame, 2.31));
	const blockHash = hash11(blockSeed);
	const blockActive = std.step(__tsover_sub(1, __tsover_mul(glitchStrength, .35)), blockHash);
	const blockDisplace = __tsover_mul(__tsover_mul(__tsover_sub(hash11(__tsover_add(blockSeed, 77)), .5), .18), blockActive);
	const totalDisplaceX = __tsover_add(__tsover_add(__tsover_add(displace1, displace2), displace3), blockDisplace);
	const vertJitter = __tsover_mul(__tsover_mul(bandActive1, __tsover_sub(hash11(__tsover_add(bandSeed1, 200)), .5)), .012);
	const displacedUV = d.vec2f(std.clamp(__tsover_add(uv.x, __tsover_mul(totalDisplaceX, glitchStrength)), 0, 1), std.clamp(__tsover_add(uv.y, __tsover_mul(vertJitter, glitchStrength)), 0, 1));
	const mirrorSeed = __tsover_add(__tsover_add(__tsover_mul(colX, .83), __tsover_mul(bandY1, 1.17)), __tsover_mul(slowFrame, 4.57));
	const hMirrorHash = hash11(mirrorSeed);
	const vMirrorHash = hash11(__tsover_add(mirrorSeed, 500));
	const hMirrorActive = __tsover_mul(std.step(__tsover_sub(1, __tsover_mul(mirrorChance, glitchStrength)), hMirrorHash), bandActive1);
	const vMirrorActive = __tsover_mul(std.step(__tsover_sub(1, __tsover_mul(__tsover_mul(mirrorChance, glitchStrength), .4)), vMirrorHash), bandActive1);
	const mirroredUV = d.vec2f(std.mix(displacedUV.x, __tsover_sub(1, displacedUV.x), hMirrorActive), std.mix(displacedUV.y, __tsover_sub(1, displacedUV.y), vMirrorActive));
	const rgbOffset = __tsover_mul(__tsover_mul(rgbShift, .003), __tsover_add(glitchStrength, __tsover_mul(std.abs(totalDisplaceX), 2)));
	const redUV = d.vec2f(std.clamp(__tsover_add(mirroredUV.x, rgbOffset), 0, 1), mirroredUV.y);
	const blueUV = d.vec2f(std.clamp(__tsover_sub(mirroredUV.x, rgbOffset), 0, 1), mirroredUV.y);
	const fillStyleHash = hash11(__tsover_add(blockSeed, 555));
	const barOffset = hash11(__tsover_add(__tsover_add(bandY1, __tsover_mul(slowFrame, 3.33)), 333));
	const barX = std.fract(__tsover_add(__tsover_mul(uv.x, 3), __tsover_mul(barOffset, 7)));
	const barIndex = std.floor(__tsover_mul(barX, 7));
	const barR = __tsover_add(__tsover_sub(1, std.step(2, barIndex)), __tsover_mul(std.step(4, barIndex), __tsover_sub(1, std.step(6, barIndex))));
	const barG = __tsover_sub(1, std.step(4, barIndex));
	const barB = __tsover_sub(1, std.step(.5, std.fract(__tsover_mul(barIndex, .5))));
	const smpteColor = d.vec3f(barR, barG, barB);
	const solidHue = std.fract(__tsover_add(__tsover_mul(hash11(__tsover_add(bandY1, 13)), .8), __tsover_mul(slowFrame, .07)));
	const sh6 = __tsover_mul(solidHue, 6);
	const solidColor = d.vec3f(std.clamp(__tsover_sub(std.abs(__tsover_sub(sh6, 3)), 1), 0, 1), std.clamp(__tsover_sub(2, std.abs(__tsover_sub(sh6, 2))), 0, 1), std.clamp(__tsover_sub(2, std.abs(__tsover_sub(sh6, 4))), 0, 1)).mul(1.3);
	const fillColor = std.mix(solidColor, smpteColor, d.vec3f(std.step(.5, fillStyleHash)));
	const colorBarStrength = __tsover_mul(__tsover_mul(__tsover_mul(bandActive1, blockActive), colorBarMix), glitchStrength);
	const anyDistortion = std.clamp(__tsover_add(__tsover_add(bandActive1, bandActive2), bandActive3), 0, 1);
	const pixelY = __tsover_mul(uv.y, viewport.y);
	const scanlineWave = __tsover_add(__tsover_mul(std.sin(__tsover_mul(pixelY, 3.14159265)), .5), .5);
	return GlitchGeom({
		redUV,
		mirroredUV,
		blueUV,
		fillColor,
		colorBarStrength,
		scanlineEffect: __tsover_sub(1, __tsover_mul(__tsover_mul(__tsover_mul(scanlines, scanlineWave), .3), anyDistortion))
	});
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "time"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "speed"
			},
			{
				"type": "i",
				"name": "rgbShift"
			},
			{
				"type": "i",
				"name": "blockDensity"
			},
			{
				"type": "i",
				"name": "colorBarMix"
			},
			{
				"type": "i",
				"name": "mirrorChance"
			},
			{
				"type": "i",
				"name": "scanlines"
			}
		],
		"body": [0, [
			[
				13,
				"t",
				[
					1,
					"time",
					"*",
					"speed"
				]
			],
			[
				13,
				"slowFrame",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						1,
						"t",
						"*",
						[5, "2.5"]
					]]
				]
			],
			[
				13,
				"slowHash",
				[
					6,
					"hash11",
					["slowFrame"]
				]
			],
			[
				13,
				"slowPulse",
				[
					1,
					[
						1,
						"slowHash",
						"*",
						"slowHash"
					],
					"*",
					"slowHash"
				]
			],
			[
				13,
				"fastFrame",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						1,
						"t",
						"*",
						[5, "10"]
					]]
				]
			],
			[
				13,
				"fastHash",
				[
					6,
					"hash11",
					[[
						1,
						"fastFrame",
						"+",
						[5, "73"]
					]]
				]
			],
			[
				13,
				"fastPulse",
				[
					1,
					"fastHash",
					"*",
					"fastHash"
				]
			],
			[
				13,
				"threshold",
				[
					1,
					[5, "1"],
					"-",
					"intensity"
				]
			],
			[
				13,
				"majorBurst",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					["threshold", "slowPulse"]
				]
			],
			[
				13,
				"minorFlicker",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					[[
						1,
						"threshold",
						"+",
						[5, "0.15"]
					], "fastPulse"]
				]
			],
			[
				13,
				"glitchStrength",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									"majorBurst",
									"*",
									[5, "0.8"]
								],
								"+",
								[
									1,
									"minorFlicker",
									"*",
									[5, "0.4"]
								]
							],
							[5, "0"],
							[5, "1"]
						]
					],
					"*",
					"intensity"
				]
			],
			[
				13,
				"bandY1",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						1,
						[
							7,
							"uv",
							"y"
						],
						"*",
						"blockDensity"
					]]
				]
			],
			[
				13,
				"bandSeed1",
				[
					1,
					[
						1,
						"bandY1",
						"*",
						[5, "0.37"]
					],
					"+",
					[
						1,
						"slowFrame",
						"*",
						[5, "7.13"]
					]
				]
			],
			[
				13,
				"bandHash1",
				[
					6,
					"hash11",
					["bandSeed1"]
				]
			],
			[
				13,
				"bandActive1",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					[[
						1,
						[5, "1"],
						"-",
						[
							1,
							"glitchStrength",
							"*",
							[5, "0.5"]
						]
					], "bandHash1"]
				]
			],
			[
				13,
				"displace1",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								"hash11",
								[[
									1,
									"bandSeed1",
									"+",
									[5, "42"]
								]]
							],
							"-",
							[5, "0.5"]
						],
						"*",
						[5, "0.25"]
					],
					"*",
					"bandActive1"
				]
			],
			[
				13,
				"bandY2",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						1,
						[
							7,
							"uv",
							"y"
						],
						"*",
						[
							1,
							"blockDensity",
							"*",
							[5, "2"]
						]
					]]
				]
			],
			[
				13,
				"bandSeed2",
				[
					1,
					[
						1,
						"bandY2",
						"*",
						[5, "0.53"]
					],
					"+",
					[
						1,
						"fastFrame",
						"*",
						[5, "3.37"]
					]
				]
			],
			[
				13,
				"bandHash2",
				[
					6,
					"hash11",
					["bandSeed2"]
				]
			],
			[
				13,
				"bandActive2",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					[[
						1,
						[5, "1"],
						"-",
						[
							1,
							"glitchStrength",
							"*",
							[5, "0.7"]
						]
					], "bandHash2"]
				]
			],
			[
				13,
				"displace2",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								"hash11",
								[[
									1,
									"bandSeed2",
									"+",
									[5, "99"]
								]]
							],
							"-",
							[5, "0.5"]
						],
						"*",
						[5, "0.12"]
					],
					"*",
					"bandActive2"
				]
			],
			[
				13,
				"bandY3",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						1,
						[
							7,
							"uv",
							"y"
						],
						"*",
						[
							1,
							"blockDensity",
							"*",
							[5, "4"]
						]
					]]
				]
			],
			[
				13,
				"bandSeed3",
				[
					1,
					[
						1,
						"bandY3",
						"*",
						[5, "0.71"]
					],
					"+",
					[
						1,
						"fastFrame",
						"*",
						[5, "11.71"]
					]
				]
			],
			[
				13,
				"bandHash3",
				[
					6,
					"hash11",
					["bandSeed3"]
				]
			],
			[
				13,
				"bandActive3",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					[[
						1,
						[5, "1"],
						"-",
						[
							1,
							"glitchStrength",
							"*",
							[5, "0.9"]
						]
					], "bandHash3"]
				]
			],
			[
				13,
				"displace3",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								"hash11",
								[[
									1,
									"bandSeed3",
									"+",
									[5, "173"]
								]]
							],
							"-",
							[5, "0.5"]
						],
						"*",
						[5, "0.05"]
					],
					"*",
					"bandActive3"
				]
			],
			[
				13,
				"colX",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						[5, "4"]
					]]
				]
			],
			[
				13,
				"blockSeed",
				[
					1,
					[
						1,
						[
							1,
							"colX",
							"*",
							[5, "0.41"]
						],
						"+",
						[
							1,
							"bandY1",
							"*",
							[5, "0.67"]
						]
					],
					"+",
					[
						1,
						"slowFrame",
						"*",
						[5, "2.31"]
					]
				]
			],
			[
				13,
				"blockHash",
				[
					6,
					"hash11",
					["blockSeed"]
				]
			],
			[
				13,
				"blockActive",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					[[
						1,
						[5, "1"],
						"-",
						[
							1,
							"glitchStrength",
							"*",
							[5, "0.35"]
						]
					], "blockHash"]
				]
			],
			[
				13,
				"blockDisplace",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								"hash11",
								[[
									1,
									"blockSeed",
									"+",
									[5, "77"]
								]]
							],
							"-",
							[5, "0.5"]
						],
						"*",
						[5, "0.18"]
					],
					"*",
					"blockActive"
				]
			],
			[
				13,
				"totalDisplaceX",
				[
					1,
					[
						1,
						[
							1,
							"displace1",
							"+",
							"displace2"
						],
						"+",
						"displace3"
					],
					"+",
					"blockDisplace"
				]
			],
			[
				13,
				"vertJitter",
				[
					1,
					[
						1,
						"bandActive1",
						"*",
						[
							1,
							[
								6,
								"hash11",
								[[
									1,
									"bandSeed1",
									"+",
									[5, "200"]
								]]
							],
							"-",
							[5, "0.5"]
						]
					],
					"*",
					[5, "0.012"]
				]
			],
			[
				13,
				"displacedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									7,
									"uv",
									"x"
								],
								"+",
								[
									1,
									"totalDisplaceX",
									"*",
									"glitchStrength"
								]
							],
							[5, "0"],
							[5, "1"]
						]
					], [
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									7,
									"uv",
									"y"
								],
								"+",
								[
									1,
									"vertJitter",
									"*",
									"glitchStrength"
								]
							],
							[5, "0"],
							[5, "1"]
						]
					]]
				]
			],
			[
				13,
				"mirrorSeed",
				[
					1,
					[
						1,
						[
							1,
							"colX",
							"*",
							[5, "0.83"]
						],
						"+",
						[
							1,
							"bandY1",
							"*",
							[5, "1.17"]
						]
					],
					"+",
					[
						1,
						"slowFrame",
						"*",
						[5, "4.57"]
					]
				]
			],
			[
				13,
				"hMirrorHash",
				[
					6,
					"hash11",
					["mirrorSeed"]
				]
			],
			[
				13,
				"vMirrorHash",
				[
					6,
					"hash11",
					[[
						1,
						"mirrorSeed",
						"+",
						[5, "500"]
					]]
				]
			],
			[
				13,
				"hMirrorActive",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"step"
						],
						[[
							1,
							[5, "1"],
							"-",
							[
								1,
								"mirrorChance",
								"*",
								"glitchStrength"
							]
						], "hMirrorHash"]
					],
					"*",
					"bandActive1"
				]
			],
			[
				13,
				"vMirrorActive",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"step"
						],
						[[
							1,
							[5, "1"],
							"-",
							[
								1,
								[
									1,
									"mirrorChance",
									"*",
									"glitchStrength"
								],
								"*",
								[5, "0.4"]
							]
						], "vMirrorHash"]
					],
					"*",
					"bandActive1"
				]
			],
			[
				13,
				"mirroredUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							[
								7,
								"displacedUV",
								"x"
							],
							[
								1,
								[5, "1"],
								"-",
								[
									7,
									"displacedUV",
									"x"
								]
							],
							"hMirrorActive"
						]
					], [
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							[
								7,
								"displacedUV",
								"y"
							],
							[
								1,
								[5, "1"],
								"-",
								[
									7,
									"displacedUV",
									"y"
								]
							],
							"vMirrorActive"
						]
					]]
				]
			],
			[
				13,
				"rgbOffset",
				[
					1,
					[
						1,
						"rgbShift",
						"*",
						[5, "0.003"]
					],
					"*",
					[
						1,
						"glitchStrength",
						"+",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								["totalDisplaceX"]
							],
							"*",
							[5, "2"]
						]
					]
				]
			],
			[
				13,
				"redUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									7,
									"mirroredUV",
									"x"
								],
								"+",
								"rgbOffset"
							],
							[5, "0"],
							[5, "1"]
						]
					], [
						7,
						"mirroredUV",
						"y"
					]]
				]
			],
			[
				13,
				"blueUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									7,
									"mirroredUV",
									"x"
								],
								"-",
								"rgbOffset"
							],
							[5, "0"],
							[5, "1"]
						]
					], [
						7,
						"mirroredUV",
						"y"
					]]
				]
			],
			[
				13,
				"fillStyleHash",
				[
					6,
					"hash11",
					[[
						1,
						"blockSeed",
						"+",
						[5, "555"]
					]]
				]
			],
			[
				13,
				"barOffset",
				[
					6,
					"hash11",
					[[
						1,
						[
							1,
							"bandY1",
							"+",
							[
								1,
								"slowFrame",
								"*",
								[5, "3.33"]
							]
						],
						"+",
						[5, "333"]
					]]
				]
			],
			[
				13,
				"barX",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"*",
							[5, "3"]
						],
						"+",
						[
							1,
							"barOffset",
							"*",
							[5, "7"]
						]
					]]
				]
			],
			[
				13,
				"barIndex",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						1,
						"barX",
						"*",
						[5, "7"]
					]]
				]
			],
			[
				13,
				"barR",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"step"
							],
							[[5, "2"], "barIndex"]
						]
					],
					"+",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"step"
							],
							[[5, "4"], "barIndex"]
						],
						"*",
						[
							1,
							[5, "1"],
							"-",
							[
								6,
								[
									7,
									"std",
									"step"
								],
								[[5, "6"], "barIndex"]
							]
						]
					]
				]
			],
			[
				13,
				"barG",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"step"
						],
						[[5, "4"], "barIndex"]
					]
				]
			],
			[
				13,
				"barB",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"step"
						],
						[[5, "0.5"], [
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								1,
								"barIndex",
								"*",
								[5, "0.5"]
							]]
						]]
					]
				]
			],
			[
				13,
				"smpteColor",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						"barR",
						"barG",
						"barB"
					]
				]
			],
			[
				13,
				"solidHue",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							1,
							[
								6,
								"hash11",
								[[
									1,
									"bandY1",
									"+",
									[5, "13"]
								]]
							],
							"*",
							[5, "0.8"]
						],
						"+",
						[
							1,
							"slowFrame",
							"*",
							[5, "0.07"]
						]
					]]
				]
			],
			[
				13,
				"sh6",
				[
					1,
					"solidHue",
					"*",
					[5, "6"]
				]
			],
			[
				13,
				"solidColor",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										[
											1,
											[
												6,
												[
													7,
													"std",
													"abs"
												],
												[[
													1,
													"sh6",
													"-",
													[5, "3"]
												]]
											],
											"-",
											[5, "1"]
										],
										[5, "0"],
										[5, "1"]
									]
								],
								[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										[
											1,
											[5, "2"],
											"-",
											[
												6,
												[
													7,
													"std",
													"abs"
												],
												[[
													1,
													"sh6",
													"-",
													[5, "2"]
												]]
											]
										],
										[5, "0"],
										[5, "1"]
									]
								],
								[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										[
											1,
											[5, "2"],
											"-",
											[
												6,
												[
													7,
													"std",
													"abs"
												],
												[[
													1,
													"sh6",
													"-",
													[5, "4"]
												]]
											]
										],
										[5, "0"],
										[5, "1"]
									]
								]
							]
						],
						"mul"
					],
					[[5, "1.3"]]
				]
			],
			[
				13,
				"fillColor",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"solidColor",
						"smpteColor",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[
								6,
								[
									7,
									"std",
									"step"
								],
								[[5, "0.5"], "fillStyleHash"]
							]]
						]
					]
				]
			],
			[
				13,
				"colorBarStrength",
				[
					1,
					[
						1,
						[
							1,
							"bandActive1",
							"*",
							"blockActive"
						],
						"*",
						"colorBarMix"
					],
					"*",
					"glitchStrength"
				]
			],
			[
				13,
				"anyDistortion",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"bandActive1",
								"+",
								"bandActive2"
							],
							"+",
							"bandActive3"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"pixelY",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"*",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"scanlineWave",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								"pixelY",
								"*",
								[5, "3.14159265"]
							]]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"scanlineEffect",
				[
					1,
					[5, "1"],
					"-",
					[
						1,
						[
							1,
							[
								1,
								"scanlines",
								"*",
								"scanlineWave"
							],
							"*",
							[5, "0.3"]
						],
						"*",
						"anyDistortion"
					]
				]
			],
			[10, [
				6,
				"GlitchGeom",
				[[104, {
					"redUV": "redUV",
					"mirroredUV": "mirroredUV",
					"blueUV": "blueUV",
					"fillColor": "fillColor",
					"colorBarStrength": "colorBarStrength",
					"scanlineEffect": "scanlineEffect"
				}]]
			]]
		]],
		"externalNames": [
			"std",
			"hash11",
			"d",
			"GlitchGeom"
		]
	},
	externals: () => ({
		std,
		hash11,
		d,
		GlitchGeom
	})
}) && $.f)({}));
const glitchCombine = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	GlitchGeom
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((r, g, b, geom) => {
	"use gpu";
	const glitchedColor = d.vec3f(r, g, b);
	const finalColor = std.mix(glitchedColor, geom.fillColor, d.vec3f(geom.colorBarStrength)).mul(geom.scanlineEffect);
	return d.vec4f(finalColor.x, finalColor.y, finalColor.z, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "r"
			},
			{
				"type": "i",
				"name": "g"
			},
			{
				"type": "i",
				"name": "b"
			},
			{
				"type": "i",
				"name": "geom"
			}
		],
		"body": [0, [
			[
				13,
				"glitchedColor",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						"r",
						"g",
						"b"
					]
				]
			],
			[
				13,
				"withBars",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"glitchedColor",
						[
							7,
							"geom",
							"fillColor"
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[
								7,
								"geom",
								"colorBarStrength"
							]]
						]
					]
				]
			],
			[
				13,
				"finalColor",
				[
					6,
					[
						7,
						"withBars",
						"mul"
					],
					[[
						7,
						"geom",
						"scanlineEffect"
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"finalColor",
						"x"
					],
					[
						7,
						"finalColor",
						"y"
					],
					[
						7,
						"finalColor",
						"z"
					],
					[5, "1"]
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Glitch",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Digital glitch that melts pixels and distorts colors",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: .5,
			description: "Overall glitch strength and frequency of glitch bursts",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Intensity",
				group: "Effect"
			}
		},
		speed: {
			default: 1,
			description: "How fast the glitch pattern evolves",
			ui: {
				type: "range",
				min: .1,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Effect"
			}
		},
		rgbShift: {
			default: 5,
			description: "Amount of chromatic aberration (RGB channel splitting)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .5,
				label: "RGB Shift",
				group: "Effect"
			}
		},
		blockDensity: {
			default: 10,
			description: "Base number of horizontal glitch bands",
			ui: {
				type: ["range", "map"],
				min: 2,
				max: 50,
				step: 1,
				label: "Block Density",
				group: "Effect"
			}
		},
		colorBarIntensity: {
			default: .2,
			description: "Intensity of vivid neon color bar overlay in glitch regions",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Color Bars",
				group: "Effect"
			}
		},
		mirrorAmount: {
			default: .3,
			description: "Chance of glitch blocks showing mirrored/flipped content",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Mirror",
				group: "Effect"
			}
		},
		scanlineIntensity: {
			default: .2,
			description: "Visibility of CRT-style horizontal scanlines in distorted areas",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Scanlines",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const geom = call(glitchGeom, "glitchGeom", [
			ctx.uv,
			ctx.viewportSize,
			ctx.time,
			uniforms.intensity,
			uniforms.speed,
			uniforms.rgbShift,
			uniforms.blockDensity,
			uniforms.colorBarIntensity,
			uniforms.mirrorAmount,
			uniforms.scanlineIntensity
		]);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [call(glitchCombine, "glitchCombine", [
			texture.sample(geom.member("redUV")).member("r"),
			texture.sample(geom.member("mirroredUV")).member("g"),
			texture.sample(geom.member("blueUV")).member("b"),
			geom
		])]);
	}
};
var Glitch_default = componentDefinition;
export { glitchGeom as a, glitchCombine as i, Glitch_default as n, componentDefinition as r, GlitchGeom as t };
