// shaders/dist/core/Godrays-C0VLmrcc.js
import { At as call, R as std, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
var godraysHash = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	return std.fract(__tsover_mul(std.sin(__tsover_add(__tsover_mul(p.x, 127.1), __tsover_mul(p.y, 311.7))), 43758.5453));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				1,
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						1,
						[
							1,
							[
								7,
								"p",
								"x"
							],
							"*",
							[5, "127.1"]
						],
						"+",
						[
							1,
							[
								7,
								"p",
								"y"
							],
							"*",
							[5, "311.7"]
						]
					]]
				],
				"*",
				[5, "43758.5453"]
			]]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
var godraysValueNoise = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((st) => {
	"use gpu";
	const i = std.floor(st);
	const f = std.fract(st);
	const a = godraysHash(i);
	const b = godraysHash(d.vec2f(__tsover_add(i.x, 1), i.y));
	const c = godraysHash(d.vec2f(i.x, __tsover_add(i.y, 1)));
	const e = godraysHash(d.vec2f(__tsover_add(i.x, 1), __tsover_add(i.y, 1)));
	const u = f.mul(f).mul(f.mul(-2).add(3));
	const x1 = std.mix(a, b, u.x);
	const x2 = std.mix(c, e, u.x);
	return std.mix(x1, x2, u.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "st"
		}],
		"body": [0, [
			[
				13,
				"i",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["st"]
				]
			],
			[
				13,
				"f",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["st"]
				]
			],
			[
				13,
				"a",
				[
					6,
					"godraysHash",
					["i"]
				]
			],
			[
				13,
				"b",
				[
					6,
					"godraysHash",
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								7,
								"i",
								"x"
							],
							"+",
							[5, "1"]
						], [
							7,
							"i",
							"y"
						]]
					]]
				]
			],
			[
				13,
				"c",
				[
					6,
					"godraysHash",
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							7,
							"i",
							"x"
						], [
							1,
							[
								7,
								"i",
								"y"
							],
							"+",
							[5, "1"]
						]]
					]]
				]
			],
			[
				13,
				"e",
				[
					6,
					"godraysHash",
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								7,
								"i",
								"x"
							],
							"+",
							[5, "1"]
						], [
							1,
							[
								7,
								"i",
								"y"
							],
							"+",
							[5, "1"]
						]]
					]]
				]
			],
			[
				13,
				"u",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"f",
								"mul"
							],
							["f"]
						],
						"mul"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"f",
									"mul"
								],
								[[
									4,
									"-",
									[5, "2"]
								]]
							],
							"add"
						],
						[[5, "3"]]
					]]
				]
			],
			[
				13,
				"x1",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"a",
						"b",
						[
							7,
							"u",
							"x"
						]
					]
				]
			],
			[
				13,
				"x2",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"c",
						"e",
						[
							7,
							"u",
							"x"
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"mix"
				],
				[
					"x1",
					"x2",
					[
						7,
						"u",
						"y"
					]
				]
			]]
		]],
		"externalNames": [
			"std",
			"godraysHash",
			"d"
		]
	},
	externals: () => ({
		std,
		godraysHash,
		d
	})
}) && $.f)({}));
var godraysRaysShape = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((a, aWrapped, blend, r, freq, intensity) => {
	"use gpu";
	const nLeft = std.pow(godraysValueNoise(d.vec2f(__tsover_mul(a, freq), r)), intensity);
	const nRight = std.pow(godraysValueNoise(d.vec2f(__tsover_mul(aWrapped, freq), r)), intensity);
	return std.mix(nRight, nLeft, blend);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "a"
			},
			{
				"type": "i",
				"name": "aWrapped"
			},
			{
				"type": "i",
				"name": "blend"
			},
			{
				"type": "i",
				"name": "r"
			},
			{
				"type": "i",
				"name": "freq"
			},
			{
				"type": "i",
				"name": "intensity"
			}
		],
		"body": [0, [
			[
				13,
				"nLeft",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						6,
						"godraysValueNoise",
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								"a",
								"*",
								"freq"
							], "r"]
						]]
					], "intensity"]
				]
			],
			[
				13,
				"nRight",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						6,
						"godraysValueNoise",
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								"aWrapped",
								"*",
								"freq"
							], "r"]
						]]
					], "intensity"]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"mix"
				],
				[
					"nRight",
					"nLeft",
					"blend"
				]
			]]
		]],
		"externalNames": [
			"std",
			"godraysValueNoise",
			"d"
		]
	},
	externals: () => ({
		std,
		godraysValueNoise,
		d
	})
}) && $.f)({}));
var TAU = 6.283185307179586;
const godraysColor = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.vec4f,
	d.vec4f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, animTimeRaw, center, density, intensity, spotty, rayColor, bgColor) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const centerPos = d.vec2f(center.x, __tsover_sub(1, center.y));
	const delta = uv.sub(centerPos);
	const shapeUV = d.vec2f(__tsover_mul(delta.x, aspect), delta.y);
	const animTime = __tsover_mul(animTimeRaw, .2);
	const angle = std.atan2(shapeUV.y, shapeUV.x);
	const angleWrapped = __tsover_sub(angle, __tsover_mul(TAU, std.floor(__tsover_div(angle, TAU))));
	const blendFactor = std.smoothstep(-.15, .15, shapeUV.x);
	const radius = std.length(shapeUV);
	const spots = __tsover_mul(6.5, std.abs(spotty));
	const intensityCalc = __tsover_sub(4, __tsover_mul(3, std.clamp(intensity, 0, 1)));
	const dens = __tsover_mul(6, density);
	let accumRay = d.f32(0);
	const r1 = __tsover_sub(__tsover_mul(radius, 1), __tsover_mul(animTime, 3));
	const r2 = __tsover_sub(__tsover_mul(__tsover_mul(radius, .5), __tsover_add(1, spots)), __tsover_mul(animTime, 2));
	const f1 = __tsover_mul(dens, 5);
	let ray = godraysRaysShape(angle, angleWrapped, blendFactor, r1, f1, intensityCalc);
	ray = __tsover_mul(ray, godraysRaysShape(angle, angleWrapped, blendFactor, r2, __tsover_mul(f1, 4), intensityCalc));
	accumRay = __tsover_add(accumRay, ray);
	const r3 = __tsover_sub(__tsover_mul(radius, 1.4), __tsover_mul(animTime, 2.5));
	const r4 = __tsover_sub(__tsover_mul(__tsover_mul(radius, .7), __tsover_add(1, spots)), __tsover_mul(animTime, 1.8));
	const f2 = __tsover_mul(dens, 4.5);
	let ray2 = godraysRaysShape(angle, angleWrapped, blendFactor, r3, f2, intensityCalc);
	ray2 = __tsover_mul(ray2, godraysRaysShape(angle, angleWrapped, blendFactor, r4, __tsover_mul(f2, 3.5), intensityCalc));
	accumRay = __tsover_add(accumRay, __tsover_mul(ray2, .7));
	const rayEffect = std.clamp(accumRay, 0, 1);
	const rayA = __tsover_mul(rayEffect, rayColor.w);
	const finalAlpha = __tsover_add(rayA, __tsover_mul(bgColor.w, __tsover_sub(1, rayA)));
	const safeAlpha = std.clamp(finalAlpha, .001, 1);
	const rayContribution = rayColor.xyz.mul(rayA);
	const bgContribution = bgColor.xyz.mul(bgColor.w).mul(__tsover_sub(1, rayA));
	const finalColor = rayContribution.add(bgContribution).div(safeAlpha);
	return d.vec4f(finalColor, finalAlpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "animTimeRaw"
			},
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "density"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "spotty"
			},
			{
				"type": "i",
				"name": "rayColor"
			},
			{
				"type": "i",
				"name": "bgColor"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"center",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"delta",
				[
					6,
					[
						7,
						"uv",
						"sub"
					],
					["centerPos"]
				]
			],
			[
				13,
				"shapeUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"delta",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"delta",
						"y"
					]]
				]
			],
			[
				13,
				"animTime",
				[
					1,
					"animTimeRaw",
					"*",
					[5, "0.2"]
				]
			],
			[
				13,
				"angle",
				[
					6,
					[
						7,
						"std",
						"atan2"
					],
					[[
						7,
						"shapeUV",
						"y"
					], [
						7,
						"shapeUV",
						"x"
					]]
				]
			],
			[
				13,
				"angleWrapped",
				[
					1,
					"angle",
					"-",
					[
						1,
						"TAU",
						"*",
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								1,
								"angle",
								"/",
								"TAU"
							]]
						]
					]
				]
			],
			[
				13,
				"blendFactor",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							4,
							"-",
							[5, "0.15"]
						],
						[5, "0.15"],
						[
							7,
							"shapeUV",
							"x"
						]
					]
				]
			],
			[
				13,
				"radius",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["shapeUV"]
				]
			],
			[
				13,
				"spots",
				[
					1,
					[5, "6.5"],
					"*",
					[
						6,
						[
							7,
							"std",
							"abs"
						],
						["spotty"]
					]
				]
			],
			[
				13,
				"intensityCalc",
				[
					1,
					[5, "4"],
					"-",
					[
						1,
						[5, "3"],
						"*",
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								"intensity",
								[5, "0"],
								[5, "1"]
							]
						]
					]
				]
			],
			[
				13,
				"dens",
				[
					1,
					[5, "6"],
					"*",
					"density"
				]
			],
			[
				12,
				"accumRay",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0"]]
				]
			],
			[
				13,
				"r1",
				[
					1,
					[
						1,
						"radius",
						"*",
						[5, "1"]
					],
					"-",
					[
						1,
						"animTime",
						"*",
						[5, "3"]
					]
				]
			],
			[
				13,
				"r2",
				[
					1,
					[
						1,
						[
							1,
							"radius",
							"*",
							[5, "0.5"]
						],
						"*",
						[
							1,
							[5, "1"],
							"+",
							"spots"
						]
					],
					"-",
					[
						1,
						"animTime",
						"*",
						[5, "2"]
					]
				]
			],
			[
				13,
				"f1",
				[
					1,
					"dens",
					"*",
					[5, "5"]
				]
			],
			[
				12,
				"ray",
				[
					6,
					"godraysRaysShape",
					[
						"angle",
						"angleWrapped",
						"blendFactor",
						"r1",
						"f1",
						"intensityCalc"
					]
				]
			],
			[
				2,
				"ray",
				"=",
				[
					1,
					"ray",
					"*",
					[
						6,
						"godraysRaysShape",
						[
							"angle",
							"angleWrapped",
							"blendFactor",
							"r2",
							[
								1,
								"f1",
								"*",
								[5, "4"]
							],
							"intensityCalc"
						]
					]
				]
			],
			[
				2,
				"accumRay",
				"=",
				[
					1,
					"accumRay",
					"+",
					"ray"
				]
			],
			[
				13,
				"r3",
				[
					1,
					[
						1,
						"radius",
						"*",
						[5, "1.4"]
					],
					"-",
					[
						1,
						"animTime",
						"*",
						[5, "2.5"]
					]
				]
			],
			[
				13,
				"r4",
				[
					1,
					[
						1,
						[
							1,
							"radius",
							"*",
							[5, "0.7"]
						],
						"*",
						[
							1,
							[5, "1"],
							"+",
							"spots"
						]
					],
					"-",
					[
						1,
						"animTime",
						"*",
						[5, "1.8"]
					]
				]
			],
			[
				13,
				"f2",
				[
					1,
					"dens",
					"*",
					[5, "4.5"]
				]
			],
			[
				12,
				"ray2",
				[
					6,
					"godraysRaysShape",
					[
						"angle",
						"angleWrapped",
						"blendFactor",
						"r3",
						"f2",
						"intensityCalc"
					]
				]
			],
			[
				2,
				"ray2",
				"=",
				[
					1,
					"ray2",
					"*",
					[
						6,
						"godraysRaysShape",
						[
							"angle",
							"angleWrapped",
							"blendFactor",
							"r4",
							[
								1,
								"f2",
								"*",
								[5, "3.5"]
							],
							"intensityCalc"
						]
					]
				]
			],
			[
				2,
				"accumRay",
				"=",
				[
					1,
					"accumRay",
					"+",
					[
						1,
						"ray2",
						"*",
						[5, "0.7"]
					]
				]
			],
			[
				13,
				"rayEffect",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"accumRay",
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"rayA",
				[
					1,
					"rayEffect",
					"*",
					[
						7,
						"rayColor",
						"w"
					]
				]
			],
			[
				13,
				"finalAlpha",
				[
					1,
					"rayA",
					"+",
					[
						1,
						[
							7,
							"bgColor",
							"w"
						],
						"*",
						[
							1,
							[5, "1"],
							"-",
							"rayA"
						]
					]
				]
			],
			[
				13,
				"safeAlpha",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"finalAlpha",
						[5, "0.001"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"rayContribution",
				[
					6,
					[
						7,
						[
							7,
							"rayColor",
							"xyz"
						],
						"mul"
					],
					["rayA"]
				]
			],
			[
				13,
				"bgContribution",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									7,
									"bgColor",
									"xyz"
								],
								"mul"
							],
							[[
								7,
								"bgColor",
								"w"
							]]
						],
						"mul"
					],
					[[
						1,
						[5, "1"],
						"-",
						"rayA"
					]]
				]
			],
			[
				13,
				"finalColor",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"rayContribution",
								"add"
							],
							["bgContribution"]
						],
						"div"
					],
					["safeAlpha"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				["finalColor", "finalAlpha"]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"TAU",
			"godraysRaysShape"
		]
	},
	externals: () => ({
		d,
		std,
		TAU,
		godraysRaysShape
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Godrays",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Volumetric light rays emanating from a point",
	requiresRTT: false,
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		center: {
			default: {
				x: 0,
				y: 0
			},
			transform: transformPosition,
			description: "The center point of the god rays",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		density: {
			default: .3,
			description: "Frequency of ray sectors",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Density",
				group: "Effect"
			}
		},
		intensity: {
			default: .8,
			description: "Ray visibility within sectors",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		spotty: {
			default: 1,
			description: "Density of spots on rays (higher = more spots)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Spotty",
				group: "Effect"
			}
		},
		speed: {
			default: .5,
			description: "Animation speed of the rays",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		rayColor: {
			default: "#4283fb",
			transform: transformColor,
			description: "Color of the light rays",
			ui: {
				type: "color",
				label: "Ray Color",
				group: "Colors"
			}
		},
		backgroundColor: {
			default: "transparent",
			transform: transformColor,
			description: "Background color",
			ui: {
				type: "color",
				label: "Background Color",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		return call(godraysColor, "godraysColor", [
			params.uvContext ?? params.ctx.uv,
			params.effectiveViewportSize ?? params.ctx.viewportSize,
			animatedTime(params),
			params.uniforms.center,
			params.uniforms.density,
			params.uniforms.intensity,
			params.uniforms.spotty,
			params.uniforms.rayColor,
			params.uniforms.backgroundColor
		]);
	}
};
var Godrays_default = componentDefinition;
export { componentDefinition as n, godraysColor as r, Godrays_default as t };
