// shaders/dist/core/Goo-D-X3mP-n.js
import { At as call, It as vec4, Pt as floatE, R as std, T as sdf_exports, d as sdf3d_exports, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { n as glass_exports } from "./glass-DdgMBWkF.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
import { t as shapeEffectBoundingBoxDeclaration } from "./shapeEffectBounds-BlxxsYMN.js";
var { sdfSpaceUV, offsetUV } = glass_exports;
var { createSvgSdfSampler, createAnalyticSdfSampler, driveAnalyticSubProps, ANALYTIC_SDF_EXTRA_FIELDS } = sdf_exports;
var { createVolumetricFieldComputeNode, buildVolumetricFieldSampler, VOLUMETRIC_FIELD_EXTRA_FIELDS, is3dShapeType, isVolumetricShapeType } = sdf3d_exports;
var DEG_TO_RAD = .017453292519943295;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
var N_CHARGES = 7;
var jhash = (x) => {
	const s = Math.sin(x * 12.9898) * 43758.5453;
	return s - Math.floor(s);
};
var POS_FREQ_INIT = [];
var PHASE_INIT = [];
for (let i = 0; i < N_CHARGES; i++) {
	const theta = i * 2.399963;
	const rr = Math.sqrt((i + .5) / N_CHARGES);
	POS_FREQ_INIT.push(d.vec4f(rr * Math.cos(theta), rr * Math.sin(theta), .5 + jhash(i + 1), .5 + jhash(i + 7)));
	PHASE_INIT.push(d.vec4f(.5 + jhash(i + 13), jhash(i + 2) * 6.2831853, jhash(i + 5) * 6.2831853, jhash(i + 9) * 6.2831853));
}
var POS_FREQ = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.const(d.arrayOf(d.vec4f, N_CHARGES), POS_FREQ_INIT), "POS_FREQ");
var PHASE = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.const(d.arrayOf(d.vec4f, N_CHARGES), PHASE_INIT), "PHASE");
function resolveShapeType(shapeType, shape) {
	if (shapeType) return shapeType;
	if (shape && typeof shape === "object") return shape.type || "circleSDF";
	if (typeof shape === "string") try {
		return JSON.parse(shape).type || "circleSDF";
	} catch {
		return "circleSDF";
	}
	return "circleSDF";
}
const gooComposite = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec2f,
	d.vec4f,
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((surf0, surfX, surfY, sdfUV, gooColor, specColor, p0, p1, p2, p3) => {
	"use gpu";
	const spread = __tsover_mul(p0.x, .45);
	const driftAmp = __tsover_mul(p0.y, .12);
	const breatheAmp = p0.z;
	const merge = p0.w;
	const blobScale = p1.x;
	const contain = p1.y;
	const T = p1.z;
	const edgeSoftness = p1.w;
	const bulge = p2.x;
	const lightAngle = p2.y;
	const wetness = p2.z;
	const fresnelAmt = p2.w;
	const ambientAmt = p3.x;
	const absorbAmt = p3.y;
	const translucency = p3.z;
	const animTime = p3.w;
	const px = __tsover_sub(sdfUV.x, .5);
	const py = __tsover_sub(sdfUV.y, .5);
	const distC = std.sqrt(__tsover_add(__tsover_mul(px, px), __tsover_mul(py, py)));
	const radiusMul = __tsover_add(.65, __tsover_mul(merge, .7));
	const baseR = __tsover_mul(__tsover_mul(blobScale, .38), radiusMul);
	if (distC > __tsover_add(__tsover_add(__tsover_add(spread, __tsover_mul(baseR, __tsover_add(1, breatheAmp))), driftAmp), .2)) return d.vec4f(0, 0, 0, 0);
	const EPS = d.f32(.004);
	let potC = d.f32(0);
	let potX = d.f32(0);
	let potY = d.f32(0);
	for (let i = 0; i < N_CHARGES; i++) {
		const pf = POS_FREQ.$[i];
		const ph = PHASE.$[i];
		const cx = __tsover_add(__tsover_mul(pf.x, spread), __tsover_mul(driftAmp, std.sin(__tsover_add(__tsover_mul(animTime, pf.z), ph.y))));
		const cy = __tsover_add(__tsover_mul(pf.y, spread), __tsover_mul(driftAmp, std.cos(__tsover_add(__tsover_mul(animTime, pf.w), ph.z))));
		const rVal = __tsover_mul(baseR, __tsover_add(1, __tsover_mul(breatheAmp, std.sin(__tsover_add(__tsover_mul(animTime, ph.x), ph.w)))));
		const invR2 = __tsover_div(1, std.max(__tsover_mul(rVal, rVal), d.f32(1e-4)));
		const dcx = __tsover_sub(px, cx);
		const dcy = __tsover_sub(py, cy);
		potC = __tsover_add(potC, std.exp(__tsover_mul(__tsover_mul(__tsover_add(__tsover_mul(dcx, dcx), __tsover_mul(dcy, dcy)), invR2), -1)));
		const dxx = __tsover_sub(__tsover_add(px, EPS), cx);
		potX = __tsover_add(potX, std.exp(__tsover_mul(__tsover_mul(__tsover_add(__tsover_mul(dxx, dxx), __tsover_mul(dcy, dcy)), invR2), -1)));
		const dyy = __tsover_sub(__tsover_add(py, EPS), cy);
		potY = __tsover_add(potY, std.exp(__tsover_mul(__tsover_mul(__tsover_add(__tsover_mul(dcx, dcx), __tsover_mul(dyy, dyy)), invR2), -1)));
	}
	const containerSoft = d.f32(.06);
	const inside0 = __tsover_sub(1, std.smoothstep(__tsover_mul(containerSoft, -1), containerSoft, surf0.x));
	const insideX = __tsover_sub(1, std.smoothstep(__tsover_mul(containerSoft, -1), containerSoft, surfX.x));
	const insideY = __tsover_sub(1, std.smoothstep(__tsover_mul(containerSoft, -1), containerSoft, surfY.x));
	const H0 = __tsover_add(__tsover_mul(potC, std.mix(d.f32(1), inside0, contain)), __tsover_mul(inside0, contain));
	const Hx = __tsover_add(__tsover_mul(potX, std.mix(d.f32(1), insideX, contain)), __tsover_mul(insideX, contain));
	const Hy = __tsover_add(__tsover_mul(potY, std.mix(d.f32(1), insideY, contain)), __tsover_mul(insideY, contain));
	const soft = std.max(edgeSoftness, d.f32(.001));
	const coverage = std.smoothstep(__tsover_sub(T, soft), __tsover_add(T, soft), H0);
	const gradHx = __tsover_div(__tsover_sub(Hx, H0), EPS);
	const gradHy = __tsover_div(__tsover_sub(Hy, H0), EPS);
	const metaSlope = __tsover_add(__tsover_mul(bulge, .12), .03);
	const reliefGate = std.min(std.min(inside0, insideX), insideY);
	const shapeRelief = __tsover_mul(__tsover_mul(contain, 1.6), reliefGate);
	const gradAx = std.clamp(__tsover_div(__tsover_sub(surfX.w, surf0.w), EPS), -6, 6);
	const gradAy = std.clamp(__tsover_div(__tsover_sub(surfY.w, surf0.w), EPS), -6, 6);
	const n = std.normalize(d.vec3f(__tsover_add(__tsover_mul(__tsover_mul(gradHx, -1), metaSlope), __tsover_mul(gradAx, shapeRelief)), __tsover_add(__tsover_mul(__tsover_mul(gradHy, -1), metaSlope), __tsover_mul(gradAy, shapeRelief)), -1));
	const domeZ = __tsover_mul(std.sqrt(std.max(__tsover_sub(H0, T), d.f32(0))), bulge);
	const shapeDepth = std.max(__tsover_mul(surf0.x, -1), d.f32(0));
	const thick = std.clamp(__tsover_add(domeZ, __tsover_mul(__tsover_mul(shapeDepth, contain), 1.2)), 0, 1);
	const lx = std.cos(__tsover_mul(lightAngle, DEG_TO_RAD));
	const ly = std.sin(__tsover_mul(lightAngle, DEG_TO_RAD));
	const L = std.normalize(d.vec3f(lx, ly, -.8));
	const Hh = std.normalize(L.add(d.vec3f(0, 0, -1)));
	const ndl = std.clamp(std.dot(n, L), 0, 1);
	const ndh = std.clamp(std.dot(n, Hh), 0, 1);
	const ndv = std.clamp(__tsover_mul(n.z, -1), 0, 1);
	const specCore = __tsover_mul(std.pow(ndh, d.f32(232)), 1.7);
	const specHalo = __tsover_mul(std.pow(ndh, d.f32(32)), .25);
	const glint = __tsover_mul(std.smoothstep(.992, .999, ndh), 1.9);
	const fres = __tsover_mul(std.pow(__tsover_sub(1, ndv), d.f32(5)), fresnelAmt);
	const specRgb = specColor.mul(__tsover_mul(__tsover_add(__tsover_add(__tsover_add(specCore, specHalo), glint), fres), wetness));
	const transmit = std.exp(__tsover_mul(__tsover_mul(absorbAmt, -1), __tsover_sub(1, thick)));
	const gooRgb = gooColor.xyz;
	const sss = gooRgb.mul(__tsover_mul(translucency, transmit)).mul(__tsover_add(__tsover_mul(ndl, .5), .5));
	const rgb = gooRgb.mul(__tsover_add(ambientAmt, __tsover_mul(ndl, .7))).add(sss).add(specRgb);
	return d.vec4f(rgb.x, rgb.y, rgb.z, __tsover_mul(coverage, gooColor.w));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "surf0"
			},
			{
				"type": "i",
				"name": "surfX"
			},
			{
				"type": "i",
				"name": "surfY"
			},
			{
				"type": "i",
				"name": "sdfUV"
			},
			{
				"type": "i",
				"name": "gooColor"
			},
			{
				"type": "i",
				"name": "specColor"
			},
			{
				"type": "i",
				"name": "p0"
			},
			{
				"type": "i",
				"name": "p1"
			},
			{
				"type": "i",
				"name": "p2"
			},
			{
				"type": "i",
				"name": "p3"
			}
		],
		"body": [0, [
			[
				13,
				"spread",
				[
					1,
					[
						7,
						"p0",
						"x"
					],
					"*",
					[5, "0.45"]
				]
			],
			[
				13,
				"driftAmp",
				[
					1,
					[
						7,
						"p0",
						"y"
					],
					"*",
					[5, "0.12"]
				]
			],
			[
				13,
				"breatheAmp",
				[
					7,
					"p0",
					"z"
				]
			],
			[
				13,
				"merge",
				[
					7,
					"p0",
					"w"
				]
			],
			[
				13,
				"blobScale",
				[
					7,
					"p1",
					"x"
				]
			],
			[
				13,
				"contain",
				[
					7,
					"p1",
					"y"
				]
			],
			[
				13,
				"T",
				[
					7,
					"p1",
					"z"
				]
			],
			[
				13,
				"edgeSoftness",
				[
					7,
					"p1",
					"w"
				]
			],
			[
				13,
				"bulge",
				[
					7,
					"p2",
					"x"
				]
			],
			[
				13,
				"lightAngle",
				[
					7,
					"p2",
					"y"
				]
			],
			[
				13,
				"wetness",
				[
					7,
					"p2",
					"z"
				]
			],
			[
				13,
				"fresnelAmt",
				[
					7,
					"p2",
					"w"
				]
			],
			[
				13,
				"ambientAmt",
				[
					7,
					"p3",
					"x"
				]
			],
			[
				13,
				"absorbAmt",
				[
					7,
					"p3",
					"y"
				]
			],
			[
				13,
				"translucency",
				[
					7,
					"p3",
					"z"
				]
			],
			[
				13,
				"animTime",
				[
					7,
					"p3",
					"w"
				]
			],
			[
				13,
				"px",
				[
					1,
					[
						7,
						"sdfUV",
						"x"
					],
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"py",
				[
					1,
					[
						7,
						"sdfUV",
						"y"
					],
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"distC",
				[
					6,
					[
						7,
						"std",
						"sqrt"
					],
					[[
						1,
						[
							1,
							"px",
							"*",
							"px"
						],
						"+",
						[
							1,
							"py",
							"*",
							"py"
						]
					]]
				]
			],
			[
				13,
				"radiusMul",
				[
					1,
					[5, "0.65"],
					"+",
					[
						1,
						"merge",
						"*",
						[5, "0.7"]
					]
				]
			],
			[
				13,
				"baseR",
				[
					1,
					[
						1,
						"blobScale",
						"*",
						[5, "0.38"]
					],
					"*",
					"radiusMul"
				]
			],
			[
				13,
				"maxReach",
				[
					1,
					[
						1,
						[
							1,
							"spread",
							"+",
							[
								1,
								"baseR",
								"*",
								[
									1,
									[5, "1"],
									"+",
									"breatheAmp"
								]
							]
						],
						"+",
						"driftAmp"
					],
					"+",
					[5, "0.2"]
				]
			],
			[
				11,
				[
					1,
					"distC",
					">",
					"maxReach"
				],
				[10, [
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "0"],
						[5, "0"]
					]
				]]
			],
			[
				13,
				"EPS",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0.004"]]
				]
			],
			[
				12,
				"potC",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0"]]
				]
			],
			[
				12,
				"potX",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0"]]
				]
			],
			[
				12,
				"potY",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0"]]
				]
			],
			[
				14,
				[
					12,
					"i",
					[5, "0"]
				],
				[
					1,
					"i",
					"<",
					"N_CHARGES"
				],
				[
					102,
					"++",
					"i"
				],
				[0, [
					[
						13,
						"pf",
						[
							8,
							[
								7,
								"POS_FREQ",
								"$"
							],
							"i"
						]
					],
					[
						13,
						"ph",
						[
							8,
							[
								7,
								"PHASE",
								"$"
							],
							"i"
						]
					],
					[
						13,
						"cx",
						[
							1,
							[
								1,
								[
									7,
									"pf",
									"x"
								],
								"*",
								"spread"
							],
							"+",
							[
								1,
								"driftAmp",
								"*",
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										[
											1,
											"animTime",
											"*",
											[
												7,
												"pf",
												"z"
											]
										],
										"+",
										[
											7,
											"ph",
											"y"
										]
									]]
								]
							]
						]
					],
					[
						13,
						"cy",
						[
							1,
							[
								1,
								[
									7,
									"pf",
									"y"
								],
								"*",
								"spread"
							],
							"+",
							[
								1,
								"driftAmp",
								"*",
								[
									6,
									[
										7,
										"std",
										"cos"
									],
									[[
										1,
										[
											1,
											"animTime",
											"*",
											[
												7,
												"pf",
												"w"
											]
										],
										"+",
										[
											7,
											"ph",
											"z"
										]
									]]
								]
							]
						]
					],
					[
						13,
						"rVal",
						[
							1,
							"baseR",
							"*",
							[
								1,
								[5, "1"],
								"+",
								[
									1,
									"breatheAmp",
									"*",
									[
										6,
										[
											7,
											"std",
											"sin"
										],
										[[
											1,
											[
												1,
												"animTime",
												"*",
												[
													7,
													"ph",
													"x"
												]
											],
											"+",
											[
												7,
												"ph",
												"w"
											]
										]]
									]
								]
							]
						]
					],
					[
						13,
						"invR2",
						[
							1,
							[5, "1"],
							"/",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								[[
									1,
									"rVal",
									"*",
									"rVal"
								], [
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "0.0001"]]
								]]
							]
						]
					],
					[
						13,
						"dcx",
						[
							1,
							"px",
							"-",
							"cx"
						]
					],
					[
						13,
						"dcy",
						[
							1,
							"py",
							"-",
							"cy"
						]
					],
					[
						2,
						"potC",
						"=",
						[
							1,
							"potC",
							"+",
							[
								6,
								[
									7,
									"std",
									"exp"
								],
								[[
									1,
									[
										1,
										[
											1,
											[
												1,
												"dcx",
												"*",
												"dcx"
											],
											"+",
											[
												1,
												"dcy",
												"*",
												"dcy"
											]
										],
										"*",
										"invR2"
									],
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								]]
							]
						]
					],
					[
						13,
						"dxx",
						[
							1,
							[
								1,
								"px",
								"+",
								"EPS"
							],
							"-",
							"cx"
						]
					],
					[
						2,
						"potX",
						"=",
						[
							1,
							"potX",
							"+",
							[
								6,
								[
									7,
									"std",
									"exp"
								],
								[[
									1,
									[
										1,
										[
											1,
											[
												1,
												"dxx",
												"*",
												"dxx"
											],
											"+",
											[
												1,
												"dcy",
												"*",
												"dcy"
											]
										],
										"*",
										"invR2"
									],
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								]]
							]
						]
					],
					[
						13,
						"dyy",
						[
							1,
							[
								1,
								"py",
								"+",
								"EPS"
							],
							"-",
							"cy"
						]
					],
					[
						2,
						"potY",
						"=",
						[
							1,
							"potY",
							"+",
							[
								6,
								[
									7,
									"std",
									"exp"
								],
								[[
									1,
									[
										1,
										[
											1,
											[
												1,
												"dcx",
												"*",
												"dcx"
											],
											"+",
											[
												1,
												"dyy",
												"*",
												"dyy"
											]
										],
										"*",
										"invR2"
									],
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								]]
							]
						]
					]
				]]
			],
			[
				13,
				"containerSoft",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0.06"]]
				]
			],
			[
				13,
				"inside0",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"containerSoft",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"containerSoft",
							[
								7,
								"surf0",
								"x"
							]
						]
					]
				]
			],
			[
				13,
				"insideX",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"containerSoft",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"containerSoft",
							[
								7,
								"surfX",
								"x"
							]
						]
					]
				]
			],
			[
				13,
				"insideY",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"containerSoft",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"containerSoft",
							[
								7,
								"surfY",
								"x"
							]
						]
					]
				]
			],
			[
				13,
				"H0",
				[
					1,
					[
						1,
						"potC",
						"*",
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "1"]]
								],
								"inside0",
								"contain"
							]
						]
					],
					"+",
					[
						1,
						"inside0",
						"*",
						"contain"
					]
				]
			],
			[
				13,
				"Hx",
				[
					1,
					[
						1,
						"potX",
						"*",
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "1"]]
								],
								"insideX",
								"contain"
							]
						]
					],
					"+",
					[
						1,
						"insideX",
						"*",
						"contain"
					]
				]
			],
			[
				13,
				"Hy",
				[
					1,
					[
						1,
						"potY",
						"*",
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "1"]]
								],
								"insideY",
								"contain"
							]
						]
					],
					"+",
					[
						1,
						"insideY",
						"*",
						"contain"
					]
				]
			],
			[
				13,
				"soft",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["edgeSoftness", [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0.001"]]
					]]
				]
			],
			[
				13,
				"coverage",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"T",
							"-",
							"soft"
						],
						[
							1,
							"T",
							"+",
							"soft"
						],
						"H0"
					]
				]
			],
			[
				13,
				"gradHx",
				[
					1,
					[
						1,
						"Hx",
						"-",
						"H0"
					],
					"/",
					"EPS"
				]
			],
			[
				13,
				"gradHy",
				[
					1,
					[
						1,
						"Hy",
						"-",
						"H0"
					],
					"/",
					"EPS"
				]
			],
			[
				13,
				"metaSlope",
				[
					1,
					[
						1,
						"bulge",
						"*",
						[5, "0.12"]
					],
					"+",
					[5, "0.03"]
				]
			],
			[
				13,
				"reliefGate",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					[[
						6,
						[
							7,
							"std",
							"min"
						],
						["inside0", "insideX"]
					], "insideY"]
				]
			],
			[
				13,
				"shapeRelief",
				[
					1,
					[
						1,
						"contain",
						"*",
						[5, "1.6"]
					],
					"*",
					"reliefGate"
				]
			],
			[
				13,
				"gradAx",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"surfX",
									"w"
								],
								"-",
								[
									7,
									"surf0",
									"w"
								]
							],
							"/",
							"EPS"
						],
						[
							4,
							"-",
							[5, "6"]
						],
						[5, "6"]
					]
				]
			],
			[
				13,
				"gradAy",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"surfY",
									"w"
								],
								"-",
								[
									7,
									"surf0",
									"w"
								]
							],
							"/",
							"EPS"
						],
						[
							4,
							"-",
							[5, "6"]
						],
						[5, "6"]
					]
				]
			],
			[
				13,
				"n",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								[
									1,
									[
										1,
										"gradHx",
										"*",
										[
											4,
											"-",
											[5, "1"]
										]
									],
									"*",
									"metaSlope"
								],
								"+",
								[
									1,
									"gradAx",
									"*",
									"shapeRelief"
								]
							],
							[
								1,
								[
									1,
									[
										1,
										"gradHy",
										"*",
										[
											4,
											"-",
											[5, "1"]
										]
									],
									"*",
									"metaSlope"
								],
								"+",
								[
									1,
									"gradAy",
									"*",
									"shapeRelief"
								]
							],
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]]
				]
			],
			[
				13,
				"domeZ",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								1,
								"H0",
								"-",
								"T"
							], [
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							]]
						]]
					],
					"*",
					"bulge"
				]
			],
			[
				13,
				"shapeDepth",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[
							7,
							"surf0",
							"x"
						],
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					], [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]]
				]
			],
			[
				13,
				"thick",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"domeZ",
							"+",
							[
								1,
								[
									1,
									"shapeDepth",
									"*",
									"contain"
								],
								"*",
								[5, "1.2"]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"lx",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					[[
						1,
						"lightAngle",
						"*",
						"DEG_TO_RAD"
					]]
				]
			],
			[
				13,
				"ly",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						1,
						"lightAngle",
						"*",
						"DEG_TO_RAD"
					]]
				]
			],
			[
				13,
				"L",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"lx",
							"ly",
							[
								4,
								"-",
								[5, "0.8"]
							]
						]
					]]
				]
			],
			[
				13,
				"Hh",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"L",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0"],
								[5, "0"],
								[
									4,
									"-",
									[5, "1"]
								]
							]
						]]
					]]
				]
			],
			[
				13,
				"ndl",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["n", "L"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"ndh",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["n", "Hh"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"ndv",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								7,
								"n",
								"z"
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"specCore",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"pow"
						],
						["ndh", [
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "232"]]
						]]
					],
					"*",
					[5, "1.7"]
				]
			],
			[
				13,
				"specHalo",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"pow"
						],
						["ndh", [
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "32"]]
						]]
					],
					"*",
					[5, "0.25"]
				]
			],
			[
				13,
				"glint",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0.992"],
							[5, "0.999"],
							"ndh"
						]
					],
					"*",
					[5, "1.9"]
				]
			],
			[
				13,
				"fres",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"pow"
						],
						[[
							1,
							[5, "1"],
							"-",
							"ndv"
						], [
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "5"]]
						]]
					],
					"*",
					"fresnelAmt"
				]
			],
			[
				13,
				"specRgb",
				[
					6,
					[
						7,
						"specColor",
						"mul"
					],
					[[
						1,
						[
							1,
							[
								1,
								[
									1,
									"specCore",
									"+",
									"specHalo"
								],
								"+",
								"glint"
							],
							"+",
							"fres"
						],
						"*",
						"wetness"
					]]
				]
			],
			[
				13,
				"transmit",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							"absorbAmt",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"*",
						[
							1,
							[5, "1"],
							"-",
							"thick"
						]
					]]
				]
			],
			[
				13,
				"gooRgb",
				[
					7,
					"gooColor",
					"xyz"
				]
			],
			[
				13,
				"sss",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"gooRgb",
								"mul"
							],
							[[
								1,
								"translucency",
								"*",
								"transmit"
							]]
						],
						"mul"
					],
					[[
						1,
						[
							1,
							"ndl",
							"*",
							[5, "0.5"]
						],
						"+",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"body",
				[
					6,
					[
						7,
						"gooRgb",
						"mul"
					],
					[[
						1,
						"ambientAmt",
						"+",
						[
							1,
							"ndl",
							"*",
							[5, "0.7"]
						]
					]]
				]
			],
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"body",
								"add"
							],
							["sss"]
						],
						"add"
					],
					["specRgb"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"rgb",
						"x"
					],
					[
						7,
						"rgb",
						"y"
					],
					[
						7,
						"rgb",
						"z"
					],
					[
						1,
						"coverage",
						"*",
						[
							7,
							"gooColor",
							"w"
						]
					]
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"N_CHARGES",
			"POS_FREQ",
			"PHASE",
			"DEG_TO_RAD"
		]
	},
	externals: () => ({
		std,
		d,
		N_CHARGES,
		POS_FREQ,
		PHASE,
		DEG_TO_RAD
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Goo",
	boundingBoxDeclaration: shapeEffectBoundingBoxDeclaration,
	category: "Shape Effects",
	description: "Photoreal wet liquid — animated 3D metaball blobs that merge and bulge to loosely form the shape, with sliding wet highlights, a clearcoat rim and a translucent body",
	animatedTime: { speed: "speed" },
	extraFields: {
		...VOLUMETRIC_FIELD_EXTRA_FIELDS,
		...ANALYTIC_SDF_EXTRA_FIELDS
	},
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the goo",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		scale: {
			default: 1,
			description: "Scale of the goo (1 = default size)",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Scale",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the goo in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		},
		gooColor: {
			default: "#37c95a",
			transform: transformColor,
			description: "Base liquid color (a glossy slime green by default)",
			ui: {
				type: "color",
				label: "Goo Color",
				group: "Goo"
			}
		},
		translucency: {
			default: .6,
			description: "Subsurface glow strength — deep lobes transmit more color",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Translucency",
				group: "Goo"
			}
		},
		absorb: {
			default: 2,
			description: "How fast thin areas clear and thick areas saturate (Beer–Lambert)",
			ui: {
				type: "range",
				min: 0,
				max: 6,
				step: .01,
				label: "Absorption",
				group: "Goo"
			}
		},
		containment: {
			default: .9,
			description: "0 = free blobs that ignore the shape, 1 = goo fills the shape and its surface conforms to the 3D form (highlights track the faces as it rotates)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Containment",
				group: "Goo"
			}
		},
		edgeSoftness: {
			default: .06,
			description: "Softness of the blob coverage edge",
			ui: {
				type: "range",
				min: 0,
				max: .3,
				step: .005,
				label: "Edge Softness",
				group: "Goo"
			}
		},
		blobScale: {
			default: .3,
			description: "Base radius of each liquid lobe",
			ui: {
				type: "range",
				min: .1,
				max: 1,
				step: .01,
				label: "Blob Size",
				group: "Blobs"
			}
		},
		spread: {
			default: 1,
			description: "How far the lobes scatter across the shape",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Spread",
				group: "Blobs"
			}
		},
		merge: {
			default: .6,
			description: "How much the lobes fuse together — low = distinct beads, high = one fused mass",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Merge",
				group: "Blobs"
			}
		},
		bulge: {
			default: 1,
			description: "How 3D-rounded each lobe reads — dome height of the wet surface",
			ui: {
				type: "range",
				min: 0,
				max: 1.5,
				step: .01,
				label: "Bulge",
				group: "Blobs"
			}
		},
		threshold: {
			default: .5,
			description: "Iso-surface level — lower makes fatter, fuller blobs",
			ui: {
				type: "range",
				min: .2,
				max: .9,
				step: .01,
				label: "Threshold",
				group: "Blobs"
			}
		},
		lightAngle: {
			default: 300,
			description: "Light direction in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Light Angle",
				group: "Lighting"
			}
		},
		wetness: {
			default: 2,
			description: "Master scale on the sharp specular highlight and clearcoat rim",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .01,
				label: "Highlights",
				group: "Lighting"
			}
		},
		fresnel: {
			default: .25,
			description: "Clearcoat fresnel rim strength wrapping each lobe",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Fresnel",
				group: "Lighting"
			}
		},
		specColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Highlight / rim color (white reads as a clear wet coat)",
			ui: {
				type: "color",
				label: "Highlight Color",
				group: "Lighting"
			}
		},
		ambient: {
			default: .25,
			description: "Base fill so shadowed goo isn't black",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Ambient",
				group: "Lighting"
			}
		},
		speed: {
			default: .5,
			description: "Drift and breathe speed of the blobs. 0 pauses.",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		wobble: {
			default: .5,
			description: "Drift amplitude — how far the lobe centers wander",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Wobble",
				group: "Animation"
			}
		},
		breathe: {
			default: .3,
			description: "Radius pulsing amplitude — lobes swell and shrink",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Breathe",
				group: "Animation"
			}
		},
		seed: {
			default: 1,
			description: "Variation offset — shifts the blob arrangement",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Animation"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type"
		}
	},
	compute: (params) => {
		const vf = createVolumetricFieldComputeNode(params, () => params.getCpuValue("shape"), "none");
		return vf ? {
			outputs: vf.outputs,
			getComputeNodes: (fp) => vf.getComputeNodes(fp)
		} : null;
	},
	fragment: (params) => {
		const { computeOutputs, uniforms, propValues, ctx, getCpuValue } = params;
		let sampler;
		let gradSampler;
		const fieldTex = computeOutputs?.volumetricFieldTexture;
		if (fieldTex) {
			const s = buildVolumetricFieldSampler(fieldTex, {
				originX: uniforms._vfOriginX,
				originY: uniforms._vfOriginY,
				spanX: uniforms._vfSpanX,
				spanY: uniforms._vfSpanY,
				activeResF: uniforms._vfActiveRes,
				rBound: uniforms._vfRBound
			});
			sampler = s.volumetricFieldSampler;
			gradSampler = s.volumetricFieldSamplerFast;
		} else {
			const shapeSdfUrl = propValues.shapeSdfUrl || "";
			const shapeType = propValues.shapeType || "";
			if (shapeSdfUrl) {
				sampler = createSvgSdfSampler(params, shapeSdfUrl);
				gradSampler = sampler;
			} else {
				sampler = createAnalyticSdfSampler(resolveShapeType(shapeType, getCpuValue("shape")), driveAnalyticSubProps(params, () => getCpuValue("shape")));
				gradSampler = sampler;
			}
		}
		const sdfUV = call(sdfSpaceUV, "sdfSpaceUV", [
			uniforms.center,
			uniforms.scale,
			uniforms.rotation,
			ctx.uv,
			ctx.aspect
		]);
		const surf0 = sampler(sdfUV);
		const surfX = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(.004),
			floatE(0)
		]));
		const surfY = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(0),
			floatE(.004)
		]));
		const t = animatedTime(params, "seed");
		const p0 = vec4(uniforms.spread, uniforms.wobble, uniforms.breathe, uniforms.merge);
		const p1 = vec4(uniforms.blobScale, uniforms.containment, uniforms.threshold, uniforms.edgeSoftness);
		const p2 = vec4(uniforms.bulge, uniforms.lightAngle, uniforms.wetness, uniforms.fresnel);
		const p3 = vec4(uniforms.ambient, uniforms.absorb, uniforms.translucency, t);
		return call(gooComposite, "gooComposite", [
			surf0,
			surfX,
			surfY,
			sdfUV,
			uniforms.gooColor,
			uniforms.specColor.member("rgb"),
			p0,
			p1,
			p2,
			p3
		]);
	}
};
var Goo_default = componentDefinition;
export { componentDefinition as n, gooComposite as r, Goo_default as t };
