// shaders/dist/core/GradientMap-uRYa7Yhv.js
import { At as call, Nt as expr, Ot as ZERO, Qt as mixColorsVariants, R as std, Rt as formatFloat, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
var PALETTES = {
	0: {
		a: [
			.5,
			.5,
			.5
		],
		b: [
			.5,
			.5,
			.5
		],
		c: [
			1,
			1,
			1
		],
		d: [
			0,
			.333,
			.667
		]
	},
	1: {
		a: [
			.5,
			.5,
			.5
		],
		b: [
			.5,
			.5,
			.5
		],
		c: [
			1,
			1,
			1
		],
		d: [
			0,
			.1,
			.2
		]
	},
	2: {
		a: [
			.5,
			.5,
			.5
		],
		b: [
			.5,
			.5,
			.5
		],
		c: [
			1,
			1,
			1
		],
		d: [
			.3,
			.2,
			.2
		]
	},
	3: {
		a: [
			.5,
			.5,
			.5
		],
		b: [
			.5,
			.5,
			.5
		],
		c: [
			1,
			1,
			1
		],
		d: [
			.8,
			.9,
			.3
		]
	},
	4: {
		a: [
			.5,
			.5,
			.5
		],
		b: [
			.25,
			.25,
			.25
		],
		c: [
			1,
			1,
			1
		],
		d: [
			0,
			.25,
			.5
		]
	},
	5: {
		a: [
			.5,
			.5,
			.5
		],
		b: [
			.5,
			.5,
			.5
		],
		c: [
			2,
			1,
			1
		],
		d: [
			.5,
			.2,
			.25
		]
	}
};
var transformPalette = (value) => {
	return {
		rainbow: 0,
		sunset: 1,
		ocean: 2,
		fire: 3,
		pastel: 4,
		neon: 5,
		custom: 6
	}[value] ?? 0;
};
var TWO_PI = 6.283185307179586;
var vec3E = (v) => expr(`vec3f(${formatFloat(v[0])}, ${formatFloat(v[1])}, ${formatFloat(v[2])})`);
const gradientMapT = tgpu.fn([
	d.vec4f,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, black, white, contrast) => {
	"use gpu";
	const luma = std.dot(color.xyz, d.vec3f(.2126, .7152, .0722));
	const range = std.max(__tsover_sub(white, black), 1e-4);
	let t = std.clamp(__tsover_div(__tsover_sub(luma, black), range), 0, 1);
	t = std.clamp(__tsover_add(__tsover_mul(__tsover_sub(t, .5), contrast), .5), 0, 1);
	return t;
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "color"
			},
			{
				"type": "i",
				"name": "black"
			},
			{
				"type": "i",
				"name": "white"
			},
			{
				"type": "i",
				"name": "contrast"
			}
		],
		"body": [0, [
			[
				13,
				"luma",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					[[
						7,
						"color",
						"xyz"
					], [
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0.2126"],
							[5, "0.7152"],
							[5, "0.0722"]
						]
					]]
				]
			],
			[
				13,
				"range",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"white",
						"-",
						"black"
					], [5, "0.0001"]]
				]
			],
			[
				12,
				"t",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"luma",
								"-",
								"black"
							],
							"/",
							"range"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				2,
				"t",
				"=",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									"t",
									"-",
									[5, "0.5"]
								],
								"*",
								"contrast"
							],
							"+",
							[5, "0.5"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, "t"]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const gradientMapCosine = tgpu.fn([
	d.f32,
	d.f32,
	d.vec3f,
	d.vec3f,
	d.vec3f,
	d.vec3f
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((t, phase, pa, pb, pc, pd) => {
	"use gpu";
	const arg = pc.mul(t).add(pd).add(d.vec3f(phase, phase, phase)).mul(TWO_PI);
	const mapped = pa.add(pb.mul(std.cos(arg)));
	return std.clamp(mapped, d.vec3f(0, 0, 0), d.vec3f(1, 1, 1));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "t"
			},
			{
				"type": "i",
				"name": "phase"
			},
			{
				"type": "i",
				"name": "pa"
			},
			{
				"type": "i",
				"name": "pb"
			},
			{
				"type": "i",
				"name": "pc"
			},
			{
				"type": "i",
				"name": "pd"
			}
		],
		"body": [0, [
			[
				13,
				"arg",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										[
											6,
											[
												7,
												"pc",
												"mul"
											],
											["t"]
										],
										"add"
									],
									["pd"]
								],
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									"phase",
									"phase",
									"phase"
								]
							]]
						],
						"mul"
					],
					["TWO_PI"]
				]
			],
			[
				13,
				"mapped",
				[
					6,
					[
						7,
						"pa",
						"add"
					],
					[[
						6,
						[
							7,
							"pb",
							"mul"
						],
						[[
							6,
							[
								7,
								"std",
								"cos"
							],
							["arg"]
						]]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"clamp"
				],
				[
					"mapped",
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0"],
							[5, "0"],
							[5, "0"]
						]
					],
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "1"],
							[5, "1"],
							[5, "1"]
						]
					]
				]
			]]
		]],
		"externalNames": [
			"d",
			"TWO_PI",
			"std"
		]
	},
	externals: () => ({
		d,
		TWO_PI,
		std
	})
}) && $.f)({}));
const gradientMapCustomPhase = tgpu.fn([d.f32, d.f32], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((t, phase) => {
	"use gpu";
	const t01 = std.fract(__tsover_add(t, phase));
	const t3 = __tsover_mul(t01, 3);
	return d.vec2f(std.fract(t3), std.floor(t3));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "t"
		}, {
			"type": "i",
			"name": "phase"
		}],
		"body": [0, [
			[
				13,
				"t01",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						"t",
						"+",
						"phase"
					]]
				]
			],
			[
				13,
				"t3",
				[
					1,
					"t01",
					"*",
					[5, "3"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					6,
					[
						7,
						"std",
						"fract"
					],
					["t3"]
				], [
					6,
					[
						7,
						"std",
						"floor"
					],
					["t3"]
				]]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const gradientMapSelect = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((seg0, seg1, seg2, i) => {
	"use gpu";
	const upper = std.select(seg2, seg1, i < 2);
	return std.select(upper, seg0, i < 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "seg0"
			},
			{
				"type": "i",
				"name": "seg1"
			},
			{
				"type": "i",
				"name": "seg2"
			},
			{
				"type": "i",
				"name": "i"
			}
		],
		"body": [0, [[
			13,
			"upper",
			[
				6,
				[
					7,
					"std",
					"select"
				],
				[
					"seg2",
					"seg1",
					[
						1,
						"i",
						"<",
						[5, "2"]
					]
				]
			]
		], [10, [
			6,
			[
				7,
				"std",
				"select"
			],
			[
				"upper",
				"seg0",
				[
					1,
					"i",
					"<",
					[5, "1"]
				]
			]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const gradientMapCompose = tgpu.fn([
	d.vec4f,
	d.vec3f,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, mapped, strength) => {
	"use gpu";
	const outRgb = std.mix(color.xyz, mapped, d.vec3f(strength, strength, strength));
	return d.vec4f(outRgb, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "color"
			},
			{
				"type": "i",
				"name": "mapped"
			},
			{
				"type": "i",
				"name": "strength"
			}
		],
		"body": [0, [[
			13,
			"outRgb",
			[
				6,
				[
					7,
					"std",
					"mix"
				],
				[
					[
						7,
						"color",
						"xyz"
					],
					"mapped",
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"strength",
							"strength",
							"strength"
						]
					]
				]
			]
		], [10, [
			6,
			[
				7,
				"d",
				"vec4f"
			],
			["outRgb", [
				7,
				"color",
				"w"
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "GradientMap",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Maps source luminance through an animated color gradient (Photoshop-style gradient map)",
	requiresChild: true,
	props: {
		palette: {
			default: "rainbow",
			compileTime: true,
			description: "Built-in cosine palette, or Custom Colors to pick your own stops",
			ui: {
				type: "select",
				options: [
					{
						label: "Rainbow",
						value: "rainbow"
					},
					{
						label: "Sunset",
						value: "sunset"
					},
					{
						label: "Ocean",
						value: "ocean"
					},
					{
						label: "Fire",
						value: "fire"
					},
					{
						label: "Pastel",
						value: "pastel"
					},
					{
						label: "Neon",
						value: "neon"
					},
					{
						label: "Custom Colors",
						value: "custom"
					}
				],
				label: "Palette",
				group: "Colors"
			}
		},
		colorLow: {
			default: "#1a0b2e",
			transform: transformColor,
			description: "Color for the darkest tones",
			ui: {
				type: "color",
				label: "Shadows",
				group: "Colors",
				condition: { palette: "custom" }
			}
		},
		colorMid: {
			default: "#e94560",
			transform: transformColor,
			description: "Color for the midtones",
			ui: {
				type: "color",
				label: "Midtones",
				group: "Colors",
				condition: { palette: "custom" }
			}
		},
		colorHigh: {
			default: "#f9ed69",
			transform: transformColor,
			description: "Color for the brightest tones",
			ui: {
				type: "color",
				label: "Highlights",
				group: "Colors",
				condition: { palette: "custom" }
			}
		},
		speed: {
			default: .15,
			description: "Gradient animation speed (0 = static). The gradient cycles seamlessly.",
			ui: {
				type: ["range", "map"],
				min: -2,
				max: 2,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		contrast: {
			default: 1,
			description: "Steepness of the luminance-to-gradient mapping",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 3,
				step: .01,
				label: "Contrast",
				group: "Effect"
			}
		},
		blackPoint: {
			default: 0,
			description: "Input shadow clip — luminance at/below this maps to the start of the gradient",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Black Point",
				group: "Effect"
			}
		},
		whitePoint: {
			default: 1,
			description: "Input highlight clip — luminance at/above this maps to the end of the gradient",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "White Point",
				group: "Effect"
			}
		},
		strength: {
			default: 1,
			description: "Blend between the original image (0) and the gradient-mapped result (1)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Strength",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "oklch",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for interpolating custom colors",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors",
				condition: { palette: "custom" }
			}
		}
	},
	fragment: ({ uniforms, childNode, propValues, ctx }) => {
		if (!childNode) {
			console.error("You must pass a child component into the GradientMap shader.");
			return ZERO;
		}
		const t = call(gradientMapT, "gradientMapT", [
			childNode,
			uniforms.blackPoint,
			uniforms.whitePoint,
			uniforms.contrast
		]);
		const phase = uniforms.speed.mul(ctx.time);
		const paletteIdx = transformPalette(propValues.palette ?? "rainbow");
		if (paletteIdx !== 6) {
			const p = PALETTES[paletteIdx] ?? PALETTES[0];
			return call(gradientMapCompose, "gradientMapCompose", [
				childNode,
				call(gradientMapCosine, "gradientMapCosine", [
					t,
					phase,
					vec3E(p.a),
					vec3E(p.b),
					vec3E(p.c),
					vec3E(p.d)
				]),
				uniforms.strength
			]);
		}
		const variant = mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear;
		const phaseData = call(gradientMapCustomPhase, "gradientMapCustomPhase", [t, phase]);
		const fr = phaseData.member("x");
		const i = phaseData.member("y");
		return call(gradientMapCompose, "gradientMapCompose", [
			childNode,
			call(gradientMapSelect, "gradientMapSelect", [
				call(variant, "mixColors", [
					uniforms.colorLow,
					uniforms.colorMid,
					fr
				]),
				call(variant, "mixColors", [
					uniforms.colorMid,
					uniforms.colorHigh,
					fr
				]),
				call(variant, "mixColors", [
					uniforms.colorHigh,
					uniforms.colorLow,
					fr
				]),
				i
			]).member("rgb"),
			uniforms.strength
		]);
	}
};
var GradientMap_default = componentDefinition;
export { gradientMapCustomPhase as a, gradientMapCosine as i, componentDefinition as n, gradientMapSelect as o, gradientMapCompose as r, gradientMapT as s, GradientMap_default as t };
