// shaders/dist/core/Grayscale-DUGCGltR.js
import { At as call, Ot as ZERO, R as std, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
const grayscaleColor = tgpu.fn([d.vec4f], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color) => {
	"use gpu";
	const luminance = std.dot(color.xyz, d.vec3f(.2126, .7152, .0722));
	return d.vec4f(luminance, luminance, luminance, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}],
		"body": [0, [[
			13,
			"luminance",
			[
				6,
				[
					7,
					"std",
					"dot"
				],
				[[
					7,
					"color",
					"xyz"
				], [
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0.2126"],
						[5, "0.7152"],
						[5, "0.0722"]
					]
				]]
			]
		], [10, [
			6,
			[
				7,
				"d",
				"vec4f"
			],
			[
				"luminance",
				"luminance",
				"luminance",
				[
					7,
					"color",
					"w"
				]
			]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Grayscale",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Convert colors to black and white",
	requiresChild: true,
	props: {},
	fragment: ({ childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Grayscale shader.");
			return ZERO;
		}
		return call(grayscaleColor, "grayscaleColor", [childNode]);
	}
};
var Grayscale_default = componentDefinition;
export { componentDefinition as n, grayscaleColor as r, Grayscale_default as t };
