// shaders/dist/core/GridDistortion-BkkuULR2.js
import { At as call, B as createGuardedCompute, Nt as expr, Ot as ZERO, R as std, V as createStateBuffer, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-CWOT0Wqh.js";
import { s as transformEdges } from "./transformations-fqzznt4g.js";
var DEFAULT_GRID_SIZE = 20;
var MIN_GRID = 8;
var MAX_GRID = 128;
var MAX_CELLS = MAX_GRID * MAX_GRID;
var STATE_FORMAT = "rgba16float";
var clampGrid = (v) => Math.max(MIN_GRID, Math.min(MAX_GRID, Math.floor(v)));
var updateLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	buffer: {
		storage: d.arrayOf(d.vec4f, MAX_CELLS),
		access: "mutable"
	},
	params: { uniform: /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
		cursorX: d.f32,
		cursorY: d.f32,
		mouseVelX: d.f32,
		mouseVelY: d.f32,
		dt: d.f32,
		decay: d.f32,
		intensity: d.f32,
		radius: d.f32,
		aspect: d.f32
	}), "uniform") }
}), "updateLayout");
var outputLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	buffer: {
		storage: d.arrayOf(d.vec4f, MAX_CELLS),
		access: "readonly"
	},
	dispTex: { storageTexture: d.textureStorage2d(STATE_FORMAT, "write-only") }
}), "outputLayout");
function makeUpdateKernel(gridSize) {
	return tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
		"use gpu";
		const p = updateLayout.$.params;
		const idx = __tsover_add(__tsover_mul(cy, d.u32(gridSize)), cx);
		const cell = updateLayout.$.buffer[idx];
		let dx = __tsover_mul(cell.x, __tsover_sub(1, __tsover_mul(p.decay, p.dt)));
		let dy = __tsover_mul(cell.y, __tsover_sub(1, __tsover_mul(p.decay, p.dt)));
		const cellX = __tsover_div(__tsover_add(d.f32(cx), .5), d.f32(gridSize));
		const cellY = __tsover_div(__tsover_add(d.f32(cy), .5), d.f32(gridSize));
		const ddx = std.select(__tsover_sub(cellX, p.cursorX), __tsover_mul(__tsover_sub(cellX, p.cursorX), p.aspect), p.aspect >= 1);
		const ddy = std.select(__tsover_div(__tsover_sub(cellY, p.cursorY), p.aspect), __tsover_sub(cellY, p.cursorY), p.aspect >= 1);
		const distSq = __tsover_add(__tsover_mul(ddx, ddx), __tsover_mul(ddy, ddy));
		const radius2 = __tsover_mul(p.radius, 2);
		if (distSq < __tsover_mul(radius2, radius2) && __tsover_add(std.abs(p.mouseVelX), std.abs(p.mouseVelY)) > .01) {
			const influence = std.exp(__tsover_mul(__tsover_div(distSq, __tsover_mul(p.radius, p.radius)), -1));
			dx = __tsover_add(dx, __tsover_mul(__tsover_mul(__tsover_mul(__tsover_mul(p.mouseVelX, influence), p.intensity), p.dt), .5));
			dy = __tsover_add(dy, __tsover_mul(__tsover_mul(__tsover_mul(__tsover_mul(p.mouseVelY, influence), p.intensity), p.dt), .5));
		}
		dx = std.clamp(dx, -1, 1);
		dy = std.clamp(dy, -1, 1);
		updateLayout.$.buffer[idx] = d.vec4f(dx, dy, 0, 0);
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [{
				"type": "i",
				"name": "cx"
			}, {
				"type": "i",
				"name": "cy"
			}],
			"body": [0, [
				[
					13,
					"p",
					[
						7,
						[
							7,
							"updateLayout",
							"$"
						],
						"params"
					]
				],
				[
					13,
					"idx",
					[
						1,
						[
							1,
							"cy",
							"*",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								["gridSize"]
							]
						],
						"+",
						"cx"
					]
				],
				[
					13,
					"cell",
					[
						8,
						[
							7,
							[
								7,
								"updateLayout",
								"$"
							],
							"buffer"
						],
						"idx"
					]
				],
				[
					12,
					"dx",
					[
						1,
						[
							7,
							"cell",
							"x"
						],
						"*",
						[
							1,
							[5, "1"],
							"-",
							[
								1,
								[
									7,
									"p",
									"decay"
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						]
					]
				],
				[
					12,
					"dy",
					[
						1,
						[
							7,
							"cell",
							"y"
						],
						"*",
						[
							1,
							[5, "1"],
							"-",
							[
								1,
								[
									7,
									"p",
									"decay"
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						]
					]
				],
				[
					13,
					"cellX",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cx"]
							],
							"+",
							[5, "0.5"]
						],
						"/",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["gridSize"]
						]
					]
				],
				[
					13,
					"cellY",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cy"]
							],
							"+",
							[5, "0.5"]
						],
						"/",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["gridSize"]
						]
					]
				],
				[
					13,
					"ddx",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								1,
								"cellX",
								"-",
								[
									7,
									"p",
									"cursorX"
								]
							],
							[
								1,
								[
									1,
									"cellX",
									"-",
									[
										7,
										"p",
										"cursorX"
									]
								],
								"*",
								[
									7,
									"p",
									"aspect"
								]
							],
							[
								1,
								[
									7,
									"p",
									"aspect"
								],
								">=",
								[5, "1"]
							]
						]
					]
				],
				[
					13,
					"ddy",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								1,
								[
									1,
									"cellY",
									"-",
									[
										7,
										"p",
										"cursorY"
									]
								],
								"/",
								[
									7,
									"p",
									"aspect"
								]
							],
							[
								1,
								"cellY",
								"-",
								[
									7,
									"p",
									"cursorY"
								]
							],
							[
								1,
								[
									7,
									"p",
									"aspect"
								],
								">=",
								[5, "1"]
							]
						]
					]
				],
				[
					13,
					"distSq",
					[
						1,
						[
							1,
							"ddx",
							"*",
							"ddx"
						],
						"+",
						[
							1,
							"ddy",
							"*",
							"ddy"
						]
					]
				],
				[
					13,
					"radius2",
					[
						1,
						[
							7,
							"p",
							"radius"
						],
						"*",
						[5, "2"]
					]
				],
				[
					11,
					[
						3,
						[
							1,
							"distSq",
							"<",
							[
								1,
								"radius2",
								"*",
								"radius2"
							]
						],
						"&&",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"abs"
									],
									[[
										7,
										"p",
										"mouseVelX"
									]]
								],
								"+",
								[
									6,
									[
										7,
										"std",
										"abs"
									],
									[[
										7,
										"p",
										"mouseVelY"
									]]
								]
							],
							">",
							[5, "0.01"]
						]
					],
					[0, [
						[
							13,
							"influence",
							[
								6,
								[
									7,
									"std",
									"exp"
								],
								[[
									1,
									[
										1,
										"distSq",
										"/",
										[
											1,
											[
												7,
												"p",
												"radius"
											],
											"*",
											[
												7,
												"p",
												"radius"
											]
										]
									],
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								]]
							]
						],
						[
							2,
							"dx",
							"=",
							[
								1,
								"dx",
								"+",
								[
									1,
									[
										1,
										[
											1,
											[
												1,
												[
													7,
													"p",
													"mouseVelX"
												],
												"*",
												"influence"
											],
											"*",
											[
												7,
												"p",
												"intensity"
											]
										],
										"*",
										[
											7,
											"p",
											"dt"
										]
									],
									"*",
									[5, "0.5"]
								]
							]
						],
						[
							2,
							"dy",
							"=",
							[
								1,
								"dy",
								"+",
								[
									1,
									[
										1,
										[
											1,
											[
												1,
												[
													7,
													"p",
													"mouseVelY"
												],
												"*",
												"influence"
											],
											"*",
											[
												7,
												"p",
												"intensity"
											]
										],
										"*",
										[
											7,
											"p",
											"dt"
										]
									],
									"*",
									[5, "0.5"]
								]
							]
						]
					]]
				],
				[
					2,
					"dx",
					"=",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							"dx",
							[
								4,
								"-",
								[5, "1"]
							],
							[5, "1"]
						]
					]
				],
				[
					2,
					"dy",
					"=",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							"dy",
							[
								4,
								"-",
								[5, "1"]
							],
							[5, "1"]
						]
					]
				],
				[
					2,
					[
						8,
						[
							7,
							[
								7,
								"updateLayout",
								"$"
							],
							"buffer"
						],
						"idx"
					],
					"=",
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							"dx",
							"dy",
							[5, "0"],
							[5, "0"]
						]
					]
				]
			]],
			"externalNames": [
				"updateLayout",
				"d",
				"gridSize",
				"std"
			]
		},
		externals: () => ({
			updateLayout,
			d,
			gridSize,
			std
		})
	}) && $.f)({})).$name("gridDistortionUpdate");
}
function makeOutputKernel(gridSize) {
	return tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
		"use gpu";
		const idx = __tsover_add(__tsover_mul(cy, d.u32(gridSize)), cx);
		const cell = outputLayout.$.buffer[idx];
		std.textureStore(outputLayout.$.dispTex, d.vec2u(cx, cy), d.vec4f(cell.x, cell.y, 0, 1));
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [{
				"type": "i",
				"name": "cx"
			}, {
				"type": "i",
				"name": "cy"
			}],
			"body": [0, [
				[
					13,
					"idx",
					[
						1,
						[
							1,
							"cy",
							"*",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								["gridSize"]
							]
						],
						"+",
						"cx"
					]
				],
				[
					13,
					"cell",
					[
						8,
						[
							7,
							[
								7,
								"outputLayout",
								"$"
							],
							"buffer"
						],
						"idx"
					]
				],
				[
					6,
					[
						7,
						"std",
						"textureStore"
					],
					[
						[
							7,
							[
								7,
								"outputLayout",
								"$"
							],
							"dispTex"
						],
						[
							6,
							[
								7,
								"d",
								"vec2u"
							],
							["cx", "cy"]
						],
						[
							6,
							[
								7,
								"d",
								"vec4f"
							],
							[
								[
									7,
									"cell",
									"x"
								],
								[
									7,
									"cell",
									"y"
								],
								[5, "0"],
								[5, "1"]
							]
						]
					]
				]
			]],
			"externalNames": [
				"d",
				"gridSize",
				"outputLayout",
				"std"
			]
		},
		externals: () => ({
			d,
			gridSize,
			outputLayout,
			std
		})
	}) && $.f)({})).$name("gridDistortionOutput");
}
const gridCellSnap = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, gridSize, aspect) => {
	"use gpu";
	const isWide = aspect > 1;
	const cellsX = std.max(std.select(__tsover_mul(gridSize, aspect), gridSize, isWide), 1);
	const cellsY = std.max(std.select(gridSize, __tsover_div(gridSize, aspect), isWide), 1);
	const cellIndexX = std.floor(__tsover_mul(uv.x, cellsX));
	const cellIndexY = std.floor(__tsover_mul(uv.y, cellsY));
	return d.vec2f(__tsover_div(__tsover_add(cellIndexX, .5), cellsX), __tsover_div(__tsover_add(cellIndexY, .5), cellsY));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "gridSize"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"isWide",
				[
					1,
					"aspect",
					">",
					[5, "1"]
				]
			],
			[
				13,
				"cellsX",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								1,
								"gridSize",
								"*",
								"aspect"
							],
							"gridSize",
							"isWide"
						]
					], [5, "1"]]
				]
			],
			[
				13,
				"cellsY",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							"gridSize",
							[
								1,
								"gridSize",
								"/",
								"aspect"
							],
							"isWide"
						]
					], [5, "1"]]
				]
			],
			[
				13,
				"cellIndexX",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"cellsX"
					]]
				]
			],
			[
				13,
				"cellIndexY",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						1,
						[
							7,
							"uv",
							"y"
						],
						"*",
						"cellsY"
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						1,
						"cellIndexX",
						"+",
						[5, "0.5"]
					],
					"/",
					"cellsX"
				], [
					1,
					[
						1,
						"cellIndexY",
						"+",
						[5, "0.5"]
					],
					"/",
					"cellsY"
				]]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const gridDistortOffsetUV = tgpu.fn([d.vec2f, d.vec2f], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, disp) => {
	"use gpu";
	const maxDisp = .1;
	const clamped = std.clamp(disp, d.vec2f(-maxDisp, -maxDisp), d.vec2f(maxDisp, maxDisp));
	return uv.sub(clamped);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "uv"
		}, {
			"type": "i",
			"name": "disp"
		}],
		"body": [0, [
			[
				13,
				"maxDisp",
				[5, "0.1"]
			],
			[
				13,
				"clamped",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"disp",
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								4,
								"-",
								"maxDisp"
							], [
								4,
								"-",
								"maxDisp"
							]]
						],
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["maxDisp", "maxDisp"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"uv",
					"sub"
				],
				["clamped"]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "GridDistortion",
	category: "Interactive",
	description: "Interactive grid distortion controlled by mouse position",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	usesPointer: true,
	props: {
		intensity: {
			default: 1,
			description: "Strength of the distortion effect",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		decay: {
			default: 3,
			description: "Rate of distortion decay (higher = faster)",
			ui: {
				type: "range",
				min: 0,
				max: 10,
				step: .1,
				label: "Decay",
				group: "Effect"
			}
		},
		radius: {
			default: 1,
			description: "Radius of the distortion effect",
			ui: {
				type: "range",
				min: 0,
				max: 3,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		gridSize: {
			default: 20,
			compileTimeWhen: (prev, next) => clampGrid(prev) !== clampGrid(next),
			description: "Resolution of the distortion grid (higher = more detailed)",
			ui: {
				type: ["range", "map"],
				min: 8,
				max: 128,
				step: 1,
				label: "Grid Size",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	compute: (params) => {
		const { gpu, getCpuValue, registerComputeTexture, onCleanup, onResize, dimensions } = params;
		const root = gpu?.root;
		if (!root) return null;
		const gridSize = clampGrid(getCpuValue("gridSize") ?? DEFAULT_GRID_SIZE);
		const buffer = createStateBuffer(root, d.vec4f, MAX_CELLS);
		const dispTex = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
			size: [gridSize, gridSize],
			format: STATE_FORMAT
		}).$usage("storage", "sampled"), "dispTex");
		onCleanup(() => dispTex.destroy());
		const displacement = registerComputeTexture(dispTex);
		const updateParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(d.struct({
			cursorX: d.f32,
			cursorY: d.f32,
			mouseVelX: d.f32,
			mouseVelY: d.f32,
			dt: d.f32,
			decay: d.f32,
			intensity: d.f32,
			radius: d.f32,
			aspect: d.f32
		})), "updateParams");
		let curW = Math.max(1, Math.round(dimensions.width));
		let curH = Math.max(1, Math.round(dimensions.height));
		onResize(({ width, height }) => {
			curW = Math.max(1, Math.round(width));
			curH = Math.max(1, Math.round(height));
		});
		const size = [gridSize, gridSize];
		const updateKernel = makeUpdateKernel(gridSize);
		const outputKernel = makeOutputKernel(gridSize);
		const updateBg = root.createBindGroup(updateLayout, {
			buffer,
			params: updateParams.buffer
		});
		const outputBg = root.createBindGroup(outputLayout, {
			buffer,
			dispTex
		});
		const update = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			updateKernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"updateKernel",
					["cx", "cy"]
				]]],
				"externalNames": ["updateKernel"]
			},
			externals: () => ({ updateKernel })
		}) && $.f)({}), {
			size,
			bindGroup: updateBg
		});
		const output = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			outputKernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"outputKernel",
					["cx", "cy"]
				]]],
				"externalNames": ["outputKernel"]
			},
			externals: () => ({ outputKernel })
		}) && $.f)({}), {
			size,
			bindGroup: outputBg
		});
		let prevX = .5;
		let prevY = .5;
		let lastActiveTime = Date.now();
		let smoothVelX = 0;
		let smoothVelY = 0;
		return {
			outputs: { displacement },
			getComputeNodes: (frameParams) => {
				const fp = frameParams;
				const currentTime = Date.now();
				const dt = Math.min(fp.deltaTime ?? 0, .016);
				const aspect = Math.max(1, fp.dimensions?.width ?? curW) / Math.max(1, fp.dimensions?.height ?? curH);
				const px = fp.pointer?.x ?? .5;
				const py = fp.pointer?.y ?? .5;
				const velX = dt > 0 ? (px - prevX) / dt : 0;
				const velY = dt > 0 ? (py - prevY) / dt : 0;
				smoothVelX = smoothVelX * .85 + velX * .15;
				smoothVelY = smoothVelY * .85 + velY * .15;
				prevX = px;
				prevY = py;
				if (Math.abs(velX) + Math.abs(velY) > .01) lastActiveTime = currentTime;
				const decayVal = getCpuValue("decay") ?? 3;
				const settleTime = decayVal > 0 ? Math.min(3e4, Math.log(1e-4) / Math.log(Math.max(1e-6, 1 - decayVal * .016)) * 16.67) : 3e4;
				if (currentTime - lastActiveTime > settleTime) return null;
				updateParams.write({
					cursorX: px,
					cursorY: py,
					mouseVelX: smoothVelX,
					mouseVelY: smoothVelY,
					dt,
					decay: decayVal,
					intensity: getCpuValue("intensity") ?? 1,
					radius: (getCpuValue("radius") ?? 1) * .05,
					aspect
				});
				return [update, output];
			}
		};
	},
	fragment: ({ childNode, ctx, computeOutputs, propValues, convertToTexture, uniforms }) => {
		if (!childNode) return ZERO;
		const childTex = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		const dispTex = computeOutputs?.displacement;
		const gridCellUV = call(gridCellSnap, "gridCellSnap", [
			ctx.uv,
			uniforms.gridSize,
			ctx.aspect
		]);
		const disp = dispTex ? dispTex.sample(gridCellUV, "linearClamp").member("xy") : expr("vec2f(0.0, 0.0)");
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(gridDistortOffsetUV, "gridDistortOffsetUV", [ctx.uv, disp]), (uv) => childTex.sample(uv), edgeMode)]);
	},
	uvRemap: ({ uv, mask, uniforms, aspect, propValues, computeOutputs }) => {
		const dispTex = computeOutputs?.displacement;
		if (!dispTex) return {
			uv,
			mask
		};
		const edgeMode = propValues.edges ?? 0;
		const gridCellUV = call(gridCellSnap, "gridCellSnap", [
			uv,
			uniforms.gridSize,
			aspect
		]);
		return composeEdgeRemapExpr(call(gridDistortOffsetUV, "gridDistortOffsetUV", [uv, dispTex.sample(gridCellUV, "linearClamp").member("xy")]), mask, edgeMode);
	}
};
var GridDistortion_default = componentDefinition;
export { makeOutputKernel as a, gridDistortOffsetUV as i, componentDefinition as n, makeUpdateKernel as o, gridCellSnap as r, GridDistortion_default as t };
