// shaders/dist/core/Grid-FtTNW6zy.js
import { At as call, It as vec4, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
var DEG_TO_RAD = Math.PI / 180;
const gridCellCenterUV = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, cells, rotation) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const correctedX = __tsover_mul(uv.x, aspect);
	const correctedY = __tsover_sub(1, uv.y);
	const rotRad = __tsover_mul(rotation, DEG_TO_RAD);
	const cosR = std.cos(rotRad);
	const sinR = std.sin(rotRad);
	const centerX = __tsover_mul(aspect, .5);
	const centerY = .5;
	const cx = __tsover_sub(correctedX, centerX);
	const cy = __tsover_sub(correctedY, centerY);
	const rotX = __tsover_add(__tsover_sub(__tsover_mul(cx, cosR), __tsover_mul(cy, sinR)), centerX);
	const rotY = __tsover_add(__tsover_add(__tsover_mul(cx, sinR), __tsover_mul(cy, cosR)), centerY);
	const cellRX = __tsover_div(__tsover_add(std.floor(__tsover_mul(rotX, cells)), .5), cells);
	const cellRY = __tsover_div(__tsover_add(std.floor(__tsover_mul(rotY, cells)), .5), cells);
	const ccx = __tsover_sub(cellRX, centerX);
	const ccy = __tsover_sub(cellRY, centerY);
	const unrotX = __tsover_add(__tsover_add(__tsover_mul(ccx, cosR), __tsover_mul(ccy, sinR)), centerX);
	const unrotY = __tsover_add(__tsover_add(__tsover_mul(ccx, -sinR), __tsover_mul(ccy, cosR)), centerY);
	return d.vec2f(__tsover_div(unrotX, aspect), __tsover_sub(1, unrotY));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "cells"
			},
			{
				"type": "i",
				"name": "rotation"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"correctedX",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"correctedY",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"uv",
						"y"
					]
				]
			],
			[
				13,
				"rotRad",
				[
					1,
					"rotation",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["rotRad"]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["rotRad"]
				]
			],
			[
				13,
				"centerX",
				[
					1,
					"aspect",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"centerY",
				[5, "0.5"]
			],
			[
				13,
				"cx",
				[
					1,
					"correctedX",
					"-",
					"centerX"
				]
			],
			[
				13,
				"cy",
				[
					1,
					"correctedY",
					"-",
					"centerY"
				]
			],
			[
				13,
				"rotX",
				[
					1,
					[
						1,
						[
							1,
							"cx",
							"*",
							"cosR"
						],
						"-",
						[
							1,
							"cy",
							"*",
							"sinR"
						]
					],
					"+",
					"centerX"
				]
			],
			[
				13,
				"rotY",
				[
					1,
					[
						1,
						[
							1,
							"cx",
							"*",
							"sinR"
						],
						"+",
						[
							1,
							"cy",
							"*",
							"cosR"
						]
					],
					"+",
					"centerY"
				]
			],
			[
				13,
				"cellRX",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								1,
								"rotX",
								"*",
								"cells"
							]]
						],
						"+",
						[5, "0.5"]
					],
					"/",
					"cells"
				]
			],
			[
				13,
				"cellRY",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								1,
								"rotY",
								"*",
								"cells"
							]]
						],
						"+",
						[5, "0.5"]
					],
					"/",
					"cells"
				]
			],
			[
				13,
				"ccx",
				[
					1,
					"cellRX",
					"-",
					"centerX"
				]
			],
			[
				13,
				"ccy",
				[
					1,
					"cellRY",
					"-",
					"centerY"
				]
			],
			[
				13,
				"unrotX",
				[
					1,
					[
						1,
						[
							1,
							"ccx",
							"*",
							"cosR"
						],
						"+",
						[
							1,
							"ccy",
							"*",
							"sinR"
						]
					],
					"+",
					"centerX"
				]
			],
			[
				13,
				"unrotY",
				[
					1,
					[
						1,
						[
							1,
							"ccx",
							"*",
							[
								4,
								"-",
								"sinR"
							]
						],
						"+",
						[
							1,
							"ccy",
							"*",
							"cosR"
						]
					],
					"+",
					"centerY"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					"unrotX",
					"/",
					"aspect"
				], [
					1,
					[5, "1"],
					"-",
					"unrotY"
				]]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"std",
			"d"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		std,
		d
	})
}) && $.f)({}));
const gridFilterAxis = tgpu.fn([
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((px, wx, N) => {
	"use gpu";
	const a = __tsover_add(px, __tsover_mul(wx, .5));
	const b = __tsover_sub(px, __tsover_mul(wx, .5));
	return __tsover_div(__tsover_sub(__tsover_sub(__tsover_add(std.floor(a), std.min(__tsover_mul(std.fract(a), N), 1)), std.floor(b)), std.min(__tsover_mul(std.fract(b), N), 1)), __tsover_mul(N, wx));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "px"
			},
			{
				"type": "i",
				"name": "wx"
			},
			{
				"type": "i",
				"name": "N"
			}
		],
		"body": [0, [
			[
				13,
				"a",
				[
					1,
					"px",
					"+",
					[
						1,
						"wx",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"b",
				[
					1,
					"px",
					"-",
					[
						1,
						"wx",
						"*",
						[5, "0.5"]
					]
				]
			],
			[10, [
				1,
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"floor"
								],
								["a"]
							],
							"+",
							[
								6,
								[
									7,
									"std",
									"min"
								],
								[[
									1,
									[
										6,
										[
											7,
											"std",
											"fract"
										],
										["a"]
									],
									"*",
									"N"
								], [5, "1"]]
							]
						],
						"-",
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							["b"]
						]
					],
					"-",
					[
						6,
						[
							7,
							"std",
							"min"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"fract"
								],
								["b"]
							],
							"*",
							"N"
						], [5, "1"]]
					]
				],
				"/",
				[
					1,
					"N",
					"*",
					"wx"
				]
			]]
		]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const gridField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, cells, thickness, rotation, softness, variation) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const correctedX = __tsover_mul(uv.x, aspect);
	const correctedY = __tsover_sub(1, uv.y);
	const rotRad = __tsover_mul(rotation, DEG_TO_RAD);
	const cosR = std.cos(rotRad);
	const sinR = std.sin(rotRad);
	const centerX = __tsover_mul(aspect, .5);
	const centerY = .5;
	const cx = __tsover_sub(correctedX, centerX);
	const cy = __tsover_sub(correctedY, centerY);
	const rotX = __tsover_add(__tsover_sub(__tsover_mul(cx, cosR), __tsover_mul(cy, sinR)), centerX);
	const rotY = __tsover_add(__tsover_add(__tsover_mul(cx, sinR), __tsover_mul(cy, cosR)), centerY);
	const gridUV = d.vec2f(__tsover_mul(rotX, cells), __tsover_mul(rotY, cells));
	const lineWidth = __tsover_mul(thickness, .01);
	const safeLine = std.max(lineWidth, 1e-4);
	const lineFraction = __tsover_mul(safeLine, 2);
	const N = __tsover_div(1, lineFraction);
	const p = gridUV.add(safeLine);
	const dpx = std.dpdx(p);
	const dpy = std.dpdy(p);
	const wx = __tsover_add(std.max(std.max(std.abs(dpx.x), std.abs(dpy.x)), 1e-5), softness);
	const wy = __tsover_add(std.max(std.max(std.abs(dpx.y), std.abs(dpy.y)), 1e-5), softness);
	const iX = gridFilterAxis(p.x, wx, N);
	const iY = gridFilterAxis(p.y, wy, N);
	const gapBoth = __tsover_mul(__tsover_sub(1, iX), __tsover_sub(1, iY));
	const gridMask = __tsover_mul(__tsover_sub(1, gapBoth), std.step(1e-4, lineWidth));
	const cellId = std.floor(gridUV);
	const q0 = std.fract(cellId.mul(d.vec2f(123.34, 345.45)));
	const q1 = q0.add(std.dot(q0, q0.add(34.345)));
	const rand = std.fract(__tsover_mul(q1.x, q1.y));
	const variedFactor = std.max(__tsover_add(1, __tsover_mul(__tsover_mul(__tsover_sub(rand, .5), 2), variation)), 0);
	return d.vec2f(gridMask, variedFactor);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "cells"
			},
			{
				"type": "i",
				"name": "thickness"
			},
			{
				"type": "i",
				"name": "rotation"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "variation"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"correctedX",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"correctedY",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"uv",
						"y"
					]
				]
			],
			[
				13,
				"rotRad",
				[
					1,
					"rotation",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["rotRad"]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["rotRad"]
				]
			],
			[
				13,
				"centerX",
				[
					1,
					"aspect",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"centerY",
				[5, "0.5"]
			],
			[
				13,
				"cx",
				[
					1,
					"correctedX",
					"-",
					"centerX"
				]
			],
			[
				13,
				"cy",
				[
					1,
					"correctedY",
					"-",
					"centerY"
				]
			],
			[
				13,
				"rotX",
				[
					1,
					[
						1,
						[
							1,
							"cx",
							"*",
							"cosR"
						],
						"-",
						[
							1,
							"cy",
							"*",
							"sinR"
						]
					],
					"+",
					"centerX"
				]
			],
			[
				13,
				"rotY",
				[
					1,
					[
						1,
						[
							1,
							"cx",
							"*",
							"sinR"
						],
						"+",
						[
							1,
							"cy",
							"*",
							"cosR"
						]
					],
					"+",
					"centerY"
				]
			],
			[
				13,
				"gridUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						"rotX",
						"*",
						"cells"
					], [
						1,
						"rotY",
						"*",
						"cells"
					]]
				]
			],
			[
				13,
				"lineWidth",
				[
					1,
					"thickness",
					"*",
					[5, "0.01"]
				]
			],
			[
				13,
				"safeLine",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["lineWidth", [5, "0.0001"]]
				]
			],
			[
				13,
				"lineFraction",
				[
					1,
					"safeLine",
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"N",
				[
					1,
					[5, "1"],
					"/",
					"lineFraction"
				]
			],
			[
				13,
				"p",
				[
					6,
					[
						7,
						"gridUV",
						"add"
					],
					["safeLine"]
				]
			],
			[
				13,
				"dpx",
				[
					6,
					[
						7,
						"std",
						"dpdx"
					],
					["p"]
				]
			],
			[
				13,
				"dpy",
				[
					6,
					[
						7,
						"std",
						"dpdy"
					],
					["p"]
				]
			],
			[
				13,
				"wx",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									7,
									"dpx",
									"x"
								]]
							], [
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									7,
									"dpy",
									"x"
								]]
							]]
						], [5, "0.00001"]]
					],
					"+",
					"softness"
				]
			],
			[
				13,
				"wy",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									7,
									"dpx",
									"y"
								]]
							], [
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									7,
									"dpy",
									"y"
								]]
							]]
						], [5, "0.00001"]]
					],
					"+",
					"softness"
				]
			],
			[
				13,
				"iX",
				[
					6,
					"gridFilterAxis",
					[
						[
							7,
							"p",
							"x"
						],
						"wx",
						"N"
					]
				]
			],
			[
				13,
				"iY",
				[
					6,
					"gridFilterAxis",
					[
						[
							7,
							"p",
							"y"
						],
						"wy",
						"N"
					]
				]
			],
			[
				13,
				"gapBoth",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						"iX"
					],
					"*",
					[
						1,
						[5, "1"],
						"-",
						"iY"
					]
				]
			],
			[
				13,
				"gridMask",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						"gapBoth"
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"step"
						],
						[[5, "0.0001"], "lineWidth"]
					]
				]
			],
			[
				13,
				"cellId",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["gridUV"]
				]
			],
			[
				13,
				"q0",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						6,
						[
							7,
							"cellId",
							"mul"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "123.34"], [5, "345.45"]]
						]]
					]]
				]
			],
			[
				13,
				"q1",
				[
					6,
					[
						7,
						"q0",
						"add"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["q0", [
							6,
							[
								7,
								"q0",
								"add"
							],
							[[5, "34.345"]]
						]]
					]]
				]
			],
			[
				13,
				"rand",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							7,
							"q1",
							"x"
						],
						"*",
						[
							7,
							"q1",
							"y"
						]
					]]
				]
			],
			[
				13,
				"variedFactor",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[5, "1"],
						"+",
						[
							1,
							[
								1,
								[
									1,
									"rand",
									"-",
									[5, "0.5"]
								],
								"*",
								[5, "2"]
							],
							"*",
							"variation"
						]
					], [5, "0"]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["gridMask", "variedFactor"]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"std",
			"d",
			"gridFilterAxis"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		std,
		d,
		gridFilterAxis
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Grid",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Simple grid lines pattern with adjustable thickness and rotation",
	acceptsUVContext: true,
	props: {
		color: {
			default: "#ffffff",
			transform: transformColor,
			description: "The color of the grid lines",
			ui: {
				type: "color",
				label: "Line Color",
				group: "Colors"
			}
		},
		cellColor: {
			default: "transparent",
			transform: transformColor,
			description: "Fill color of the cells (transparent = lines only). Pair with Variation for a tiled look.",
			ui: {
				type: "color",
				label: "Cell Color",
				group: "Colors"
			}
		},
		cells: {
			default: 10,
			description: "Number of cells along the shortest canvas edge (creates square cells)",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 50,
				step: 1,
				label: "Cells",
				group: "Effect"
			}
		},
		thickness: {
			default: 1,
			description: "Thickness of grid lines (normalized, 0.0-1.0)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .1,
				label: "Thickness",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the grid in degrees. At 45° this produces a crosshatch/diamond pattern.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Softness of the line edges (0 = crisp, 1 = very soft)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		variation: {
			default: 0,
			description: "Per-cell random lightening/darkening of the cell fill color",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Variation",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	mapSampleUVs: ({ uniforms, ctx }) => {
		const cellCenter = call(gridCellCenterUV, "gridCellCenterUV", [
			ctx.uv,
			ctx.viewportSize,
			uniforms.cells,
			uniforms.rotation
		]);
		return {
			thickness: cellCenter,
			cells: cellCenter,
			rotation: cellCenter,
			softness: cellCenter
		};
	},
	fragment: (params) => {
		const { uniforms, propValues } = params;
		const field = call(gridField, "gridField", [
			params.uvContext ?? params.ctx.uv,
			params.effectiveViewportSize ?? params.ctx.viewportSize,
			uniforms.cells,
			uniforms.thickness,
			uniforms.rotation,
			uniforms.softness,
			uniforms.variation
		]);
		const gridMask = field.member("x");
		const variedFactor = field.member("y");
		const cellFill = vec4(uniforms.cellColor.member("rgb").mul(variedFactor), uniforms.cellColor.member("a"));
		return call(mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			cellFill,
			uniforms.color,
			gridMask
		]);
	}
};
var Grid_default = componentDefinition;
export { gridFilterAxis as a, gridField as i, componentDefinition as n, gridCellCenterUV as r, Grid_default as t };
