// shaders/dist/core/Halftone-BdmAqFEa.js
import { At as call, It as vec4, Ot as ZERO, Pt as floatE, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, r as transformAngle } from "./transformations-fqzznt4g.js";
var SMOOTHNESS = .1;
var DEG_TO_RAD = Math.PI / 180;
var transformStyle = (value) => value === "cmyk" ? 1 : 0;
var styleModeOf = (raw) => typeof raw === "number" ? raw : raw === "cmyk" ? 1 : 0;
const halftoneClassic = tgpu.fn([
	d.vec4f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((childColor, uv, aspect, angleDeg, frequency) => {
	"use gpu";
	const aspectUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const angleRad = __tsover_mul(angleDeg, DEG_TO_RAD);
	const c = std.cos(angleRad);
	const s = std.sin(angleRad);
	const rotatedUV = d.vec2f(std.dot(aspectUV, d.vec2f(c, __tsover_mul(s, -1))), std.dot(aspectUV, d.vec2f(s, c)));
	const gridUV = std.fract(rotatedUV.mul(frequency)).sub(d.vec2f(.5, .5));
	const brightness = std.dot(d.vec3f(childColor.x, childColor.y, childColor.z), d.vec3f(.299, .587, .114));
	const dotSize = __tsover_add(__tsover_mul(brightness, .7), .15);
	const dist = std.length(gridUV);
	const dotPattern = __tsover_sub(1, std.smoothstep(__tsover_sub(dotSize, SMOOTHNESS), __tsover_add(dotSize, SMOOTHNESS), dist));
	return d.vec4f(__tsover_mul(childColor.x, dotPattern), __tsover_mul(childColor.y, dotPattern), __tsover_mul(childColor.z, dotPattern), __tsover_mul(childColor.w, dotPattern));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "childColor"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "angleDeg"
			},
			{
				"type": "i",
				"name": "frequency"
			}
		],
		"body": [0, [
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angleDeg",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"c",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"s",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"rotatedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["aspectUV", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["c", [
								1,
								"s",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]]
						]]
					], [
						6,
						[
							7,
							"std",
							"dot"
						],
						["aspectUV", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["s", "c"]
						]]
					]]
				]
			],
			[
				13,
				"gridUV",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								6,
								[
									7,
									"rotatedUV",
									"mul"
								],
								["frequency"]
							]]
						],
						"sub"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "0.5"], [5, "0.5"]]
					]]
				]
			],
			[
				13,
				"brightness",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								7,
								"childColor",
								"x"
							],
							[
								7,
								"childColor",
								"y"
							],
							[
								7,
								"childColor",
								"z"
							]
						]
					], [
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0.299"],
							[5, "0.587"],
							[5, "0.114"]
						]
					]]
				]
			],
			[
				13,
				"dotSize",
				[
					1,
					[
						1,
						"brightness",
						"*",
						[5, "0.7"]
					],
					"+",
					[5, "0.15"]
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["gridUV"]
				]
			],
			[
				13,
				"dotPattern",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"dotSize",
								"-",
								"SMOOTHNESS"
							],
							[
								1,
								"dotSize",
								"+",
								"SMOOTHNESS"
							],
							"dist"
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						1,
						[
							7,
							"childColor",
							"x"
						],
						"*",
						"dotPattern"
					],
					[
						1,
						[
							7,
							"childColor",
							"y"
						],
						"*",
						"dotPattern"
					],
					[
						1,
						[
							7,
							"childColor",
							"z"
						],
						"*",
						"dotPattern"
					],
					[
						1,
						[
							7,
							"childColor",
							"w"
						],
						"*",
						"dotPattern"
					]
				]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std",
			"SMOOTHNESS"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std,
		SMOOTHNESS
	})
}) && $.f)({}));
const halftonePlateUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, misprintAngleDeg, plateIndexDeg, misprint, aspect) => {
	"use gpu";
	const angleRad = __tsover_mul(__tsover_add(misprintAngleDeg, plateIndexDeg), DEG_TO_RAD);
	return d.vec2f(__tsover_add(uv.x, __tsover_mul(std.cos(angleRad), misprint)), __tsover_add(uv.y, __tsover_div(__tsover_mul(std.sin(angleRad), misprint), aspect)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "misprintAngleDeg"
			},
			{
				"type": "i",
				"name": "plateIndexDeg"
			},
			{
				"type": "i",
				"name": "misprint"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [[
			13,
			"angleRad",
			[
				1,
				[
					1,
					"misprintAngleDeg",
					"+",
					"plateIndexDeg"
				],
				"*",
				"DEG_TO_RAD"
			]
		], [10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					7,
					"uv",
					"x"
				],
				"+",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						["angleRad"]
					],
					"*",
					"misprint"
				]
			], [
				1,
				[
					7,
					"uv",
					"y"
				],
				"+",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							["angleRad"]
						],
						"*",
						"misprint"
					],
					"/",
					"aspect"
				]
			]]
		]]]],
		"externalNames": [
			"DEG_TO_RAD",
			"d",
			"std"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		d,
		std
	})
}) && $.f)({}));
const halftoneChannelK = tgpu.fn([d.vec4f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((s) => {
	"use gpu";
	return __tsover_sub(1, std.max(std.max(s.x, s.y), s.z));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "s"
		}],
		"body": [0, [[10, [
			1,
			[5, "1"],
			"-",
			[
				6,
				[
					7,
					"std",
					"max"
				],
				[[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						7,
						"s",
						"x"
					], [
						7,
						"s",
						"y"
					]]
				], [
					7,
					"s",
					"z"
				]]
			]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const halftoneChannelC = tgpu.fn([d.vec4f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((s) => {
	"use gpu";
	const k = halftoneChannelK(s);
	const invK = std.max(__tsover_sub(1, k), 1e-4);
	return __tsover_div(__tsover_sub(__tsover_sub(1, s.x), k), invK);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "s"
		}],
		"body": [0, [
			[
				13,
				"k",
				[
					6,
					"halftoneChannelK",
					["s"]
				]
			],
			[
				13,
				"invK",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[5, "1"],
						"-",
						"k"
					], [5, "0.0001"]]
				]
			],
			[10, [
				1,
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						[
							7,
							"s",
							"x"
						]
					],
					"-",
					"k"
				],
				"/",
				"invK"
			]]
		]],
		"externalNames": ["halftoneChannelK", "std"]
	},
	externals: () => ({
		halftoneChannelK,
		std
	})
}) && $.f)({}));
const halftoneChannelM = tgpu.fn([d.vec4f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((s) => {
	"use gpu";
	const k = halftoneChannelK(s);
	const invK = std.max(__tsover_sub(1, k), 1e-4);
	return __tsover_div(__tsover_sub(__tsover_sub(1, s.y), k), invK);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "s"
		}],
		"body": [0, [
			[
				13,
				"k",
				[
					6,
					"halftoneChannelK",
					["s"]
				]
			],
			[
				13,
				"invK",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[5, "1"],
						"-",
						"k"
					], [5, "0.0001"]]
				]
			],
			[10, [
				1,
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						[
							7,
							"s",
							"y"
						]
					],
					"-",
					"k"
				],
				"/",
				"invK"
			]]
		]],
		"externalNames": ["halftoneChannelK", "std"]
	},
	externals: () => ({
		halftoneChannelK,
		std
	})
}) && $.f)({}));
const halftoneChannelY = tgpu.fn([d.vec4f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((s) => {
	"use gpu";
	const k = halftoneChannelK(s);
	const invK = std.max(__tsover_sub(1, k), 1e-4);
	return __tsover_div(__tsover_sub(__tsover_sub(1, s.z), k), invK);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "s"
		}],
		"body": [0, [
			[
				13,
				"k",
				[
					6,
					"halftoneChannelK",
					["s"]
				]
			],
			[
				13,
				"invK",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[5, "1"],
						"-",
						"k"
					], [5, "0.0001"]]
				]
			],
			[10, [
				1,
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						[
							7,
							"s",
							"z"
						]
					],
					"-",
					"k"
				],
				"/",
				"invK"
			]]
		]],
		"externalNames": ["halftoneChannelK", "std"]
	},
	externals: () => ({
		halftoneChannelK,
		std
	})
}) && $.f)({}));
const halftonePlateGrid = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, angleDeg, intensity, frequency) => {
	"use gpu";
	const aspectUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const angleRad = __tsover_mul(angleDeg, DEG_TO_RAD);
	const c = std.cos(angleRad);
	const s = std.sin(angleRad);
	const rotatedUV = d.vec2f(std.dot(aspectUV, d.vec2f(c, __tsover_mul(s, -1))), std.dot(aspectUV, d.vec2f(s, c)));
	const grid = std.fract(rotatedUV.mul(frequency)).sub(d.vec2f(.5, .5));
	const dist = std.length(grid);
	const size = __tsover_sub(__tsover_mul(intensity, .75), SMOOTHNESS);
	return __tsover_sub(1, std.smoothstep(__tsover_sub(size, SMOOTHNESS), __tsover_add(size, SMOOTHNESS), dist));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "angleDeg"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "frequency"
			}
		],
		"body": [0, [
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angleDeg",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"c",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"s",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"rotatedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["aspectUV", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["c", [
								1,
								"s",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]]
						]]
					], [
						6,
						[
							7,
							"std",
							"dot"
						],
						["aspectUV", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["s", "c"]
						]]
					]]
				]
			],
			[
				13,
				"grid",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								6,
								[
									7,
									"rotatedUV",
									"mul"
								],
								["frequency"]
							]]
						],
						"sub"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "0.5"], [5, "0.5"]]
					]]
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["grid"]
				]
			],
			[
				13,
				"size",
				[
					1,
					[
						1,
						"intensity",
						"*",
						[5, "0.75"]
					],
					"-",
					"SMOOTHNESS"
				]
			],
			[10, [
				1,
				[5, "1"],
				"-",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"size",
							"-",
							"SMOOTHNESS"
						],
						[
							1,
							"size",
							"+",
							"SMOOTHNESS"
						],
						"dist"
					]
				]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std",
			"SMOOTHNESS"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std,
		SMOOTHNESS
	})
}) && $.f)({}));
const halftoneTransmission = tgpu.fn([d.vec4f, d.f32], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((inkColor, inkMask) => {
	"use gpu";
	return std.mix(d.vec3f(1, 1, 1), d.vec3f(inkColor.x, inkColor.y, inkColor.z), d.vec3f(__tsover_mul(inkMask, inkColor.w)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "inkColor"
		}, {
			"type": "i",
			"name": "inkMask"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"mix"
			],
			[
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "1"],
						[5, "1"],
						[5, "1"]
					]
				],
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"inkColor",
							"x"
						],
						[
							7,
							"inkColor",
							"y"
						],
						[
							7,
							"inkColor",
							"z"
						]
					]
				],
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[[
						1,
						"inkMask",
						"*",
						[
							7,
							"inkColor",
							"w"
						]
					]]
				]
			]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Halftone",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Halftone dot pattern effect for printing aesthetics",
	requiresRTT: true,
	requiresChild: true,
	blendWithChildren: false,
	props: {
		style: {
			default: "classic",
			transform: transformStyle,
			compileTime: true,
			description: "Halftone rendering style",
			ui: {
				type: "select",
				options: [{
					label: "Simple",
					value: "classic"
				}, {
					label: "CMYK Print",
					value: "cmyk"
				}],
				label: "Style",
				group: "Effect"
			}
		},
		frequency: {
			default: 100,
			description: "Frequency of the halftone dots",
			ui: {
				type: ["range", "map"],
				min: 10,
				max: 300,
				step: 1,
				label: "Frequency",
				group: "Effect"
			}
		},
		angle: {
			default: 45,
			transform: transformAngle,
			description: "Rotation angle of the pattern (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect",
				condition: { style: "classic" }
			}
		},
		cyanAngle: {
			default: 15,
			transform: transformAngle,
			description: "Screen angle for the cyan plate (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Cyan Angle",
				group: "Plate Angles",
				condition: { style: "cmyk" }
			}
		},
		magentaAngle: {
			default: 75,
			transform: transformAngle,
			description: "Screen angle for the magenta plate (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Magenta Angle",
				group: "Plate Angles",
				condition: { style: "cmyk" }
			}
		},
		yellowAngle: {
			default: 0,
			transform: transformAngle,
			description: "Screen angle for the yellow plate (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Yellow Angle",
				group: "Plate Angles",
				condition: { style: "cmyk" }
			}
		},
		blackAngle: {
			default: 45,
			transform: transformAngle,
			description: "Screen angle for the black plate (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Black Angle",
				group: "Plate Angles",
				condition: { style: "cmyk" }
			}
		},
		misprint: {
			default: 0,
			description: "Simulated mis-registration between plates. Plates are offset around the misprint angle, producing colour fringing at the edges of inked regions.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .01,
				step: 5e-4,
				label: "Misprint",
				group: "Effect",
				condition: { style: "cmyk" }
			}
		},
		misprintAngle: {
			default: 0,
			transform: transformAngle,
			description: "Direction the plates drift apart. Rotating this rotates the colour-fringing pattern.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Misprint Angle",
				group: "Effect",
				condition: { style: "cmyk" }
			}
		},
		paperColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Paper/substrate color shown where no ink lands",
			ui: {
				type: "color",
				label: "Paper",
				group: "Inks",
				condition: { style: "cmyk" }
			}
		},
		cyanColor: {
			default: "#00ffff",
			transform: transformColor,
			description: "Cyan ink color",
			ui: {
				type: "color",
				label: "Cyan",
				group: "Inks",
				condition: { style: "cmyk" }
			}
		},
		magentaColor: {
			default: "#ff00ff",
			transform: transformColor,
			description: "Magenta ink color",
			ui: {
				type: "color",
				label: "Magenta",
				group: "Inks",
				condition: { style: "cmyk" }
			}
		},
		yellowColor: {
			default: "#ffff00",
			transform: transformColor,
			description: "Yellow ink color",
			ui: {
				type: "color",
				label: "Yellow",
				group: "Inks",
				condition: { style: "cmyk" }
			}
		},
		blackColor: {
			default: "#000000",
			transform: transformColor,
			description: "Black (key) ink color",
			ui: {
				type: "color",
				label: "Black",
				group: "Inks",
				condition: { style: "cmyk" }
			}
		}
	},
	fragment: ({ childNode, ctx, uniforms, propValues, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const unpremult = (uv) => call(unpremultiplyAlpha, "unpremultiplyAlpha", [texture.sample(uv)]);
		if (styleModeOf(propValues.style) === 0) return call(halftoneClassic, "halftoneClassic", [
			unpremult(ctx.uv),
			ctx.uv,
			ctx.aspect,
			uniforms.angle,
			uniforms.frequency
		]);
		const cSample = unpremult(call(halftonePlateUV, "halftonePlateUV", [
			ctx.uv,
			uniforms.misprintAngle,
			floatE(0),
			uniforms.misprint,
			ctx.aspect
		]));
		const mSample = unpremult(call(halftonePlateUV, "halftonePlateUV", [
			ctx.uv,
			uniforms.misprintAngle,
			floatE(90),
			uniforms.misprint,
			ctx.aspect
		]));
		const ySample = unpremult(call(halftonePlateUV, "halftonePlateUV", [
			ctx.uv,
			uniforms.misprintAngle,
			floatE(180),
			uniforms.misprint,
			ctx.aspect
		]));
		const kSample = unpremult(call(halftonePlateUV, "halftonePlateUV", [
			ctx.uv,
			uniforms.misprintAngle,
			floatE(270),
			uniforms.misprint,
			ctx.aspect
		]));
		const centerSample = unpremult(ctx.uv);
		const c = call(halftoneChannelC, "halftoneChannelC", [cSample]);
		const m = call(halftoneChannelM, "halftoneChannelM", [mSample]);
		const y = call(halftoneChannelY, "halftoneChannelY", [ySample]);
		const k = call(halftoneChannelK, "halftoneChannelK", [kSample]);
		const cInk = call(halftonePlateGrid, "halftonePlateGrid", [
			ctx.uv,
			ctx.aspect,
			uniforms.cyanAngle,
			c,
			uniforms.frequency
		]);
		const mInk = call(halftonePlateGrid, "halftonePlateGrid", [
			ctx.uv,
			ctx.aspect,
			uniforms.magentaAngle,
			m,
			uniforms.frequency
		]);
		const yInk = call(halftonePlateGrid, "halftonePlateGrid", [
			ctx.uv,
			ctx.aspect,
			uniforms.yellowAngle,
			y,
			uniforms.frequency
		]);
		const kInk = call(halftonePlateGrid, "halftonePlateGrid", [
			ctx.uv,
			ctx.aspect,
			uniforms.blackAngle,
			k,
			uniforms.frequency
		]);
		const transC = call(halftoneTransmission, "halftoneTransmission", [uniforms.cyanColor, cInk]);
		const transM = call(halftoneTransmission, "halftoneTransmission", [uniforms.magentaColor, mInk]);
		const transY = call(halftoneTransmission, "halftoneTransmission", [uniforms.yellowColor, yInk]);
		const transK = call(halftoneTransmission, "halftoneTransmission", [uniforms.blackColor, kInk]);
		return vec4(uniforms.paperColor.member("rgb").mul(transC).mul(transM).mul(transY).mul(transK), centerSample.member("a"));
	}
};
var Halftone_default = componentDefinition;
export { halftoneChannelM as a, halftonePlateGrid as c, halftoneChannelK as i, halftonePlateUV as l, componentDefinition as n, halftoneChannelY as o, halftoneChannelC as r, halftoneClassic as s, Halftone_default as t, halftoneTransmission as u };
