// shaders/dist/core/Heatmap-dBfHUyG-.js
import { At as call, It as vec4, Qt as mixColorsVariants, R as std, T as sdf_exports, Z as hash12, Zt as mixColorsLinear, d as sdf3d_exports, gt as mixColorStopsRuntime, lt as value12, n as tgpu, on as d, rt as noise_exports } from "./coords-BlwgBpJf.js";
import { n as glass_exports } from "./glass-DdgMBWkF.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig, t as animatedTime } from "./porters-CoaznI7i.js";
import { t as shapeEffectBoundingBoxDeclaration } from "./shapeEffectBounds-BlxxsYMN.js";
var { sdfSpaceUV } = glass_exports;
var { createSvgSdfSampler, createAnalyticSdfSampler, driveAnalyticSubProps, ANALYTIC_SDF_EXTRA_FIELDS } = sdf_exports;
var { createVolumetricFieldComputeNode, buildVolumetricFieldSampler, VOLUMETRIC_FIELD_EXTRA_FIELDS } = sdf3d_exports;
var DEG_TO_RAD = .017453292519943295;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
var DEFAULT_STOPS = [
	{
		color: "#02010f",
		position: 0
	},
	{
		color: "#2a0a8a",
		position: .2
	},
	{
		color: "#a41c9b",
		position: .45
	},
	{
		color: "#e8632b",
		position: .7
	},
	{
		color: "#f9e25f",
		position: .9
	},
	{
		color: "#ffffff",
		position: 1
	}
];
function resolveShapeType(shapeType, shape) {
	if (shapeType) return shapeType;
	if (shape && typeof shape === "object") return shape.type || "circleSDF";
	if (typeof shape === "string") try {
		return JSON.parse(shape).type || "circleSDF";
	} catch {
		return "circleSDF";
	}
	return "circleSDF";
}
const heatmapShadow = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((a, b, tk, phase) => {
	"use gpu";
	const posS = std.mix(-.6, .6, tk);
	const wob = __tsover_sub(value12(d.vec2f(__tsover_add(__tsover_mul(b, 4), __tsover_mul(phase, 13.7)), __tsover_add(__tsover_mul(tk, 3), __tsover_mul(phase, 7.1)))), .5);
	const aEff = __tsover_add(__tsover_sub(a, posS), __tsover_mul(wob, .12));
	const band = std.exp(__tsover_div(__tsover_mul(__tsover_mul(aEff, aEff), -1), __tsover_mul(.16, .16)));
	const gate = __tsover_mul(std.smoothstep(0, .15, tk), __tsover_sub(1, std.smoothstep(.85, 1, tk)));
	return std.clamp(__tsover_mul(band, gate), 0, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "a"
			},
			{
				"type": "i",
				"name": "b"
			},
			{
				"type": "i",
				"name": "tk"
			},
			{
				"type": "i",
				"name": "phase"
			}
		],
		"body": [0, [
			[
				13,
				"posS",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							4,
							"-",
							[5, "0.6"]
						],
						[5, "0.6"],
						"tk"
					]
				]
			],
			[
				13,
				"wob",
				[
					1,
					[
						6,
						[
							7,
							"noise",
							"value12"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								[
									1,
									"b",
									"*",
									[5, "4"]
								],
								"+",
								[
									1,
									"phase",
									"*",
									[5, "13.7"]
								]
							], [
								1,
								[
									1,
									"tk",
									"*",
									[5, "3"]
								],
								"+",
								[
									1,
									"phase",
									"*",
									[5, "7.1"]
								]
							]]
						]]
					],
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"aEff",
				[
					1,
					[
						1,
						"a",
						"-",
						"posS"
					],
					"+",
					[
						1,
						"wob",
						"*",
						[5, "0.12"]
					]
				]
			],
			[
				13,
				"band",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							[
								1,
								"aEff",
								"*",
								"aEff"
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"/",
						[
							1,
							[5, "0.16"],
							"*",
							[5, "0.16"]
						]
					]]
				]
			],
			[
				13,
				"gate",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							[5, "0.15"],
							"tk"
						]
					],
					"*",
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[5, "0.85"],
								[5, "1"],
								"tk"
							]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"clamp"
				],
				[
					[
						1,
						"band",
						"*",
						"gate"
					],
					[5, "0"],
					[5, "1"]
				]
			]]
		]],
		"externalNames": [
			"std",
			"noise",
			"d"
		]
	},
	externals: () => ({
		std,
		noise: noise_exports,
		d
	})
}) && $.f)({}));
const heatmapHeatField = tgpu.fn([
	d.f32,
	d.vec2f,
	d.f32,
	d.vec4f,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((fieldRaw, sdfUV, animTime, p0, scale) => {
	"use gpu";
	const innerGlow = p0.x;
	const outerGlow = p0.y;
	const contour = p0.z;
	const angleDeg = p0.w;
	const scl = std.max(scale, .001);
	const f = __tsover_div(fieldRaw, scl);
	const inside = __tsover_sub(1, std.smoothstep(-.003, .003, f));
	const angleRad = __tsover_mul(angleDeg, DEG_TO_RAD);
	const dirX = std.cos(angleRad);
	const dirY = std.sin(angleRad);
	const a = __tsover_add(__tsover_mul(dirX, __tsover_sub(sdfUV.x, .5)), __tsover_mul(dirY, __tsover_sub(sdfUV.y, .5)));
	const b = __tsover_sub(__tsover_mul(dirY, __tsover_sub(sdfUV.x, .5)), __tsover_mul(dirX, __tsover_sub(sdfUV.y, .5)));
	const tBase = __tsover_sub(__tsover_mul(.1, animTime), .3);
	const t0 = std.fract(tBase);
	const t1 = std.fract(__tsover_add(tBase, __tsover_div(1, 3)));
	const t2 = std.fract(__tsover_add(tBase, __tsover_div(2, 3)));
	const innerBlur = std.exp(__tsover_div(__tsover_mul(std.max(__tsover_mul(f, -1), 0), -1), .12));
	let inner = __tsover_add(.8, __tsover_mul(.8, innerBlur));
	inner = std.mix(inner, 0, heatmapShadow(a, b, t0, 1));
	inner = std.mix(inner, 0, heatmapShadow(a, b, t1, 2));
	inner = std.mix(inner, 0, heatmapShadow(a, b, t2, 3));
	inner = __tsover_mul(inner, __tsover_mul(2, innerGlow));
	inner = __tsover_add(inner, __tsover_mul(__tsover_mul(contour, 2), std.exp(__tsover_div(__tsover_mul(std.abs(f), -1), .03))));
	inner = __tsover_mul(std.min(inner, 1), inside);
	inner = std.pow(std.max(inner, 0), 1.2);
	const outerBlur = std.exp(__tsover_div(__tsover_mul(std.max(f, 0), -1), __tsover_add(.05, __tsover_mul(.15, outerGlow))));
	const t3 = std.fract(__tsover_sub(__tsover_mul(tBase, 3), .1));
	const yy = std.fract(__tsover_sub(__tsover_mul(a, 1.2), t3));
	const animatedMask = __tsover_add(.5, __tsover_mul(std.smoothstep(.3, .65, yy), __tsover_sub(1, std.smoothstep(.65, 1, yy))));
	const outer = __tsover_mul(__tsover_mul(__tsover_mul(__tsover_mul(__tsover_mul(__tsover_mul(.9, std.pow(outerBlur, .8)), animatedMask), 5), outerGlow), outerGlow), __tsover_sub(1, inside));
	let heat = std.clamp(__tsover_add(inner, outer), 0, 1);
	const frame = __tsover_mul(__tsover_mul(__tsover_mul(std.smoothstep(-.06, .1, sdfUV.x), __tsover_sub(1, std.smoothstep(.9, 1.06, sdfUV.x))), std.smoothstep(-.06, .1, sdfUV.y)), __tsover_sub(1, std.smoothstep(.9, 1.06, sdfUV.y)));
	heat = __tsover_mul(heat, frame);
	heat = __tsover_add(heat, __tsover_mul(.005, __tsover_sub(hash12(d.vec2f(__tsover_mul(sdfUV.x, 991.3), __tsover_mul(sdfUV.y, 787.7))), .5)));
	return std.clamp(heat, 0, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "fieldRaw"
			},
			{
				"type": "i",
				"name": "sdfUV"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "p0"
			},
			{
				"type": "i",
				"name": "scale"
			}
		],
		"body": [0, [
			[
				13,
				"innerGlow",
				[
					7,
					"p0",
					"x"
				]
			],
			[
				13,
				"outerGlow",
				[
					7,
					"p0",
					"y"
				]
			],
			[
				13,
				"contour",
				[
					7,
					"p0",
					"z"
				]
			],
			[
				13,
				"angleDeg",
				[
					7,
					"p0",
					"w"
				]
			],
			[
				13,
				"scl",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["scale", [5, "0.001"]]
				]
			],
			[
				13,
				"f",
				[
					1,
					"fieldRaw",
					"/",
					"scl"
				]
			],
			[
				13,
				"inside",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								4,
								"-",
								[5, "0.003"]
							],
							[5, "0.003"],
							"f"
						]
					]
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angleDeg",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"dirX",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"dirY",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"a",
				[
					1,
					[
						1,
						"dirX",
						"*",
						[
							1,
							[
								7,
								"sdfUV",
								"x"
							],
							"-",
							[5, "0.5"]
						]
					],
					"+",
					[
						1,
						"dirY",
						"*",
						[
							1,
							[
								7,
								"sdfUV",
								"y"
							],
							"-",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"b",
				[
					1,
					[
						1,
						"dirY",
						"*",
						[
							1,
							[
								7,
								"sdfUV",
								"x"
							],
							"-",
							[5, "0.5"]
						]
					],
					"-",
					[
						1,
						"dirX",
						"*",
						[
							1,
							[
								7,
								"sdfUV",
								"y"
							],
							"-",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"tBase",
				[
					1,
					[
						1,
						[5, "0.1"],
						"*",
						"animTime"
					],
					"-",
					[5, "0.3"]
				]
			],
			[
				13,
				"t0",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["tBase"]
				]
			],
			[
				13,
				"t1",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						"tBase",
						"+",
						[
							1,
							[5, "1"],
							"/",
							[5, "3"]
						]
					]]
				]
			],
			[
				13,
				"t2",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						"tBase",
						"+",
						[
							1,
							[5, "2"],
							"/",
							[5, "3"]
						]
					]]
				]
			],
			[
				13,
				"innerBlur",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"max"
								],
								[[
									1,
									"f",
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								], [5, "0"]]
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"/",
						[5, "0.12"]
					]]
				]
			],
			[
				12,
				"inner",
				[
					1,
					[5, "0.8"],
					"+",
					[
						1,
						[5, "0.8"],
						"*",
						"innerBlur"
					]
				]
			],
			[
				2,
				"inner",
				"=",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"inner",
						[5, "0"],
						[
							6,
							"heatmapShadow",
							[
								"a",
								"b",
								"t0",
								[5, "1"]
							]
						]
					]
				]
			],
			[
				2,
				"inner",
				"=",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"inner",
						[5, "0"],
						[
							6,
							"heatmapShadow",
							[
								"a",
								"b",
								"t1",
								[5, "2"]
							]
						]
					]
				]
			],
			[
				2,
				"inner",
				"=",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"inner",
						[5, "0"],
						[
							6,
							"heatmapShadow",
							[
								"a",
								"b",
								"t2",
								[5, "3"]
							]
						]
					]
				]
			],
			[
				2,
				"inner",
				"=",
				[
					1,
					"inner",
					"*",
					[
						1,
						[5, "2"],
						"*",
						"innerGlow"
					]
				]
			],
			[
				2,
				"inner",
				"=",
				[
					1,
					"inner",
					"+",
					[
						1,
						[
							1,
							"contour",
							"*",
							[5, "2"]
						],
						"*",
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									1,
									[
										6,
										[
											7,
											"std",
											"abs"
										],
										["f"]
									],
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								],
								"/",
								[5, "0.03"]
							]]
						]
					]
				]
			],
			[
				2,
				"inner",
				"=",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"min"
						],
						["inner", [5, "1"]]
					],
					"*",
					"inside"
				]
			],
			[
				2,
				"inner",
				"=",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						6,
						[
							7,
							"std",
							"max"
						],
						["inner", [5, "0"]]
					], [5, "1.2"]]
				]
			],
			[
				13,
				"outerBlur",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["f", [5, "0"]]
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"/",
						[
							1,
							[5, "0.05"],
							"+",
							[
								1,
								[5, "0.15"],
								"*",
								"outerGlow"
							]
						]
					]]
				]
			],
			[
				13,
				"t3",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							1,
							"tBase",
							"*",
							[5, "3"]
						],
						"-",
						[5, "0.1"]
					]]
				]
			],
			[
				13,
				"yy",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							1,
							"a",
							"*",
							[5, "1.2"]
						],
						"-",
						"t3"
					]]
				]
			],
			[
				13,
				"animatedMask",
				[
					1,
					[5, "0.5"],
					"+",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[5, "0.3"],
								[5, "0.65"],
								"yy"
							]
						],
						"*",
						[
							1,
							[5, "1"],
							"-",
							[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									[5, "0.65"],
									[5, "1"],
									"yy"
								]
							]
						]
					]
				]
			],
			[
				13,
				"outer",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										[5, "0.9"],
										"*",
										[
											6,
											[
												7,
												"std",
												"pow"
											],
											["outerBlur", [5, "0.8"]]
										]
									],
									"*",
									"animatedMask"
								],
								"*",
								[5, "5"]
							],
							"*",
							"outerGlow"
						],
						"*",
						"outerGlow"
					],
					"*",
					[
						1,
						[5, "1"],
						"-",
						"inside"
					]
				]
			],
			[
				12,
				"heat",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"inner",
							"+",
							"outer"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"frame",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									[
										4,
										"-",
										[5, "0.06"]
									],
									[5, "0.1"],
									[
										7,
										"sdfUV",
										"x"
									]
								]
							],
							"*",
							[
								1,
								[5, "1"],
								"-",
								[
									6,
									[
										7,
										"std",
										"smoothstep"
									],
									[
										[5, "0.9"],
										[5, "1.06"],
										[
											7,
											"sdfUV",
											"x"
										]
									]
								]
							]
						],
						"*",
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[
									4,
									"-",
									[5, "0.06"]
								],
								[5, "0.1"],
								[
									7,
									"sdfUV",
									"y"
								]
							]
						]
					],
					"*",
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[5, "0.9"],
								[5, "1.06"],
								[
									7,
									"sdfUV",
									"y"
								]
							]
						]
					]
				]
			],
			[
				2,
				"heat",
				"=",
				[
					1,
					"heat",
					"*",
					"frame"
				]
			],
			[
				2,
				"heat",
				"=",
				[
					1,
					"heat",
					"+",
					[
						1,
						[5, "0.005"],
						"*",
						[
							1,
							[
								6,
								[
									7,
									"noise",
									"hash12"
								],
								[[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									[[
										1,
										[
											7,
											"sdfUV",
											"x"
										],
										"*",
										[5, "991.3"]
									], [
										1,
										[
											7,
											"sdfUV",
											"y"
										],
										"*",
										[5, "787.7"]
									]]
								]]
							],
							"-",
							[5, "0.5"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"clamp"
				],
				[
					"heat",
					[5, "0"],
					[5, "1"]
				]
			]]
		]],
		"externalNames": [
			"std",
			"DEG_TO_RAD",
			"heatmapShadow",
			"noise",
			"d"
		]
	},
	externals: () => ({
		std,
		DEG_TO_RAD,
		heatmapShadow,
		noise: noise_exports,
		d
	})
}) && $.f)({}));
const heatmapCompose = tgpu.fn([d.f32, d.vec4f], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((heat, ramp) => {
	"use gpu";
	const shapeA = std.smoothstep(0, .15, heat);
	return d.vec4f(std.clamp(ramp.xyz, d.vec3f(0), d.vec3f(1)), __tsover_mul(ramp.w, shapeA));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "heat"
		}, {
			"type": "i",
			"name": "ramp"
		}],
		"body": [0, [[
			13,
			"shapeA",
			[
				6,
				[
					7,
					"std",
					"smoothstep"
				],
				[
					[5, "0"],
					[5, "0.15"],
					"heat"
				]
			]
		], [10, [
			6,
			[
				7,
				"d",
				"vec4f"
			],
			[[
				6,
				[
					7,
					"std",
					"clamp"
				],
				[
					[
						7,
						"ramp",
						"xyz"
					],
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[[5, "0"]]
					],
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[[5, "1"]]
					]
				]
			], [
				1,
				[
					7,
					"ramp",
					"w"
				],
				"*",
				"shapeA"
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Heatmap",
	boundingBoxDeclaration: shapeEffectBoundingBoxDeclaration,
	category: "Shape Effects",
	description: "Thermal-camera heat flowing through any 2D, SVG, or 3D shape. Powered by Paper Shaders.",
	animatedTime: { speed: "speed" },
	extraFields: {
		...VOLUMETRIC_FIELD_EXTRA_FIELDS,
		...ANALYTIC_SDF_EXTRA_FIELDS
	},
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the heatmap shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		scale: {
			default: 1,
			description: "Scale of the heatmap shape (1 = default size)",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Scale",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the heatmap shape in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		},
		colorA: {
			default: "#02010f",
			transform: transformColor,
			description: "Cold color (used when no gradient stops are set)",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#f9e25f",
			transform: transformColor,
			description: "Hot color (used when no gradient stops are set)",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		stops: {
			...colorStopsPropConfig(),
			default: DEFAULT_STOPS
		},
		colorSpace: {
			default: "oklab",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for the heat ramp interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		innerGlow: {
			default: .4,
			description: "Heat filling the inside of the shape",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Inner Glow",
				group: "Heat"
			}
		},
		outerGlow: {
			default: .2,
			description: "Heat radiating beyond the silhouette",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Outer Glow",
				group: "Heat"
			}
		},
		contour: {
			default: .5,
			description: "Heat concentrated along the shape boundary",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Contour",
				group: "Heat"
			}
		},
		angle: {
			default: 90,
			description: "Direction the heat waves travel, in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Heat"
			}
		},
		speed: {
			default: 1,
			description: "Speed of the flowing heat. 0 pauses.",
			ui: {
				type: "range",
				min: 0,
				max: 4,
				step: .01,
				label: "Speed",
				group: "Heat"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type"
		}
	},
	compute: (params) => {
		const vf = createVolumetricFieldComputeNode(params, () => params.getCpuValue("shape"), "none");
		return vf ? {
			outputs: vf.outputs,
			getComputeNodes: (fp) => vf.getComputeNodes(fp)
		} : null;
	},
	fragment: (params) => {
		const { computeOutputs, uniforms, propValues, ctx, getCpuValue } = params;
		let sampler;
		const fieldTex = computeOutputs?.volumetricFieldTexture;
		if (fieldTex) sampler = buildVolumetricFieldSampler(fieldTex, {
			originX: uniforms._vfOriginX,
			originY: uniforms._vfOriginY,
			spanX: uniforms._vfSpanX,
			spanY: uniforms._vfSpanY,
			activeResF: uniforms._vfActiveRes,
			rBound: uniforms._vfRBound
		}).volumetricFieldSampler;
		else {
			const shapeSdfUrl = propValues.shapeSdfUrl || "";
			if (shapeSdfUrl) sampler = createSvgSdfSampler(params, shapeSdfUrl);
			else sampler = createAnalyticSdfSampler(resolveShapeType(propValues.shapeType || "", getCpuValue("shape")), driveAnalyticSubProps(params, () => getCpuValue("shape")));
		}
		const sdfUV = call(sdfSpaceUV, "sdfSpaceUV", [
			uniforms.center,
			uniforms.scale,
			uniforms.rotation,
			ctx.uv,
			ctx.aspect
		]);
		const surf0 = sampler(sdfUV);
		const t = animatedTime(params);
		const p0 = vec4(uniforms.innerGlow, uniforms.outerGlow, uniforms.contour, uniforms.angle);
		const heat = call(heatmapHeatField, "heatmapHeatField", [
			surf0.member("r"),
			sdfUV,
			t,
			p0,
			uniforms.scale
		]);
		const colorSpaceMode = propValues.colorSpace ?? 0;
		const stopCount = propValues.stopCount ?? 0;
		let ramp;
		if (stopCount > 1) ramp = mixColorStopsRuntime(heat, {
			colorsArray: uniforms.colorsArray,
			positionsArray: uniforms.positionsArray,
			convertedColorsArray: uniforms.convertedColorsArray,
			stopCount: uniforms.stopCount
		}, colorSpaceMode);
		else ramp = call(mixColorsVariants[colorSpaceMode] ?? mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			heat
		]);
		return call(heatmapCompose, "heatmapCompose", [heat, ramp]);
	}
};
var Heatmap_default = componentDefinition;
export { heatmapShadow as a, heatmapHeatField as i, componentDefinition as n, heatmapCompose as r, Heatmap_default as t };
