// shaders/dist/core/HexGrid-BTWeNz9Z.js
import { At as call, It as vec4, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
var DEG_TO_RAD = Math.PI / 180;
var SQRT3 = Math.sqrt(3);
var INV_SQRT3_2 = 2 / Math.sqrt(3);
var SQRT3_2 = Math.sqrt(3) / 2;
var flooredMod2 = tgpu.fn([d.vec2f, d.vec2f], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, m) => {
	"use gpu";
	return d.vec2f(__tsover_sub(p.x, __tsover_mul(m.x, std.floor(__tsover_div(p.x, m.x)))), __tsover_sub(p.y, __tsover_mul(m.y, std.floor(__tsover_div(p.y, m.y)))));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}, {
			"type": "i",
			"name": "m"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					7,
					"p",
					"x"
				],
				"-",
				[
					1,
					[
						7,
						"m",
						"x"
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"floor"
						],
						[[
							1,
							[
								7,
								"p",
								"x"
							],
							"/",
							[
								7,
								"m",
								"x"
							]
						]]
					]
				]
			], [
				1,
				[
					7,
					"p",
					"y"
				],
				"-",
				[
					1,
					[
						7,
						"m",
						"y"
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"floor"
						],
						[[
							1,
							[
								7,
								"p",
								"y"
							],
							"/",
							[
								7,
								"m",
								"y"
							]
						]]
					]
				]
			]]
		]]]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const hexGridField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, cells, thickness, rotation, softness, variation) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const correctedUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const ang = -__tsover_mul(rotation, DEG_TO_RAD);
	const cosR = std.cos(ang);
	const sinR = std.sin(ang);
	const centerX = __tsover_mul(aspect, .5);
	const centerY = .5;
	const cx = __tsover_sub(correctedUV.x, centerX);
	const cy = __tsover_sub(correctedUV.y, centerY);
	const scaledUV = d.vec2f(__tsover_add(__tsover_sub(__tsover_mul(cx, cosR), __tsover_mul(cy, sinR)), centerX), __tsover_add(__tsover_add(__tsover_mul(cx, sinR), __tsover_mul(cy, cosR)), centerY)).mul(cells);
	const s = d.vec2f(SQRT3, 1);
	const sh = s.mul(.5);
	const a = flooredMod2(scaledUV, s).sub(sh);
	const b = flooredMod2(scaledUV.sub(sh), s).sub(sh);
	const gv = std.select(b, a, std.dot(a, a) < std.dot(b, b));
	const cellId = scaledUV.sub(gv);
	const ix = std.round(__tsover_mul(cellId.x, INV_SQRT3_2));
	const iy = std.round(__tsover_mul(cellId.y, 2));
	const q0 = std.fract(d.vec2f(ix, iy).mul(d.vec2f(123.34, 345.45)));
	const q1 = q0.add(std.dot(q0, q0.add(34.345)));
	const rand = std.fract(__tsover_mul(q1.x, q1.y));
	const variedFactor = std.max(__tsover_add(1, __tsover_mul(__tsover_mul(__tsover_sub(rand, .5), 2), variation)), 0);
	const pAbs = d.vec2f(std.abs(gv.x), std.abs(gv.y));
	const hexDist = std.max(std.dot(pAbs, d.vec2f(SQRT3_2, .5)), pAbs.y);
	const edgeDist = __tsover_sub(.5, hexDist);
	const lineWidth = __tsover_mul(thickness, .01);
	const pixel = __tsover_mul(std.fwidth(scaledUV.x), .5);
	const soft = __tsover_mul(softness, .5);
	const lo = std.max(__tsover_sub(lineWidth, soft), 0);
	const hi = __tsover_add(__tsover_add(lineWidth, pixel), soft);
	const lineMask = __tsover_mul(__tsover_sub(1, std.smoothstep(lo, hi, edgeDist)), std.step(1e-4, lineWidth));
	return d.vec2f(lineMask, variedFactor);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "cells"
			},
			{
				"type": "i",
				"name": "thickness"
			},
			{
				"type": "i",
				"name": "rotation"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "variation"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"correctedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"ang",
				[
					4,
					"-",
					[
						1,
						"rotation",
						"*",
						"DEG_TO_RAD"
					]
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["ang"]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["ang"]
				]
			],
			[
				13,
				"centerX",
				[
					1,
					"aspect",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"centerY",
				[5, "0.5"]
			],
			[
				13,
				"cx",
				[
					1,
					[
						7,
						"correctedUV",
						"x"
					],
					"-",
					"centerX"
				]
			],
			[
				13,
				"cy",
				[
					1,
					[
						7,
						"correctedUV",
						"y"
					],
					"-",
					"centerY"
				]
			],
			[
				13,
				"rotatedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								1,
								"cx",
								"*",
								"cosR"
							],
							"-",
							[
								1,
								"cy",
								"*",
								"sinR"
							]
						],
						"+",
						"centerX"
					], [
						1,
						[
							1,
							[
								1,
								"cx",
								"*",
								"sinR"
							],
							"+",
							[
								1,
								"cy",
								"*",
								"cosR"
							]
						],
						"+",
						"centerY"
					]]
				]
			],
			[
				13,
				"scaledUV",
				[
					6,
					[
						7,
						"rotatedUV",
						"mul"
					],
					["cells"]
				]
			],
			[
				13,
				"s",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					["SQRT3", [5, "1"]]
				]
			],
			[
				13,
				"sh",
				[
					6,
					[
						7,
						"s",
						"mul"
					],
					[[5, "0.5"]]
				]
			],
			[
				13,
				"a",
				[
					6,
					[
						7,
						[
							6,
							"flooredMod2",
							["scaledUV", "s"]
						],
						"sub"
					],
					["sh"]
				]
			],
			[
				13,
				"b",
				[
					6,
					[
						7,
						[
							6,
							"flooredMod2",
							[[
								6,
								[
									7,
									"scaledUV",
									"sub"
								],
								["sh"]
							], "s"]
						],
						"sub"
					],
					["sh"]
				]
			],
			[
				13,
				"gv",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"b",
						"a",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"dot"
								],
								["a", "a"]
							],
							"<",
							[
								6,
								[
									7,
									"std",
									"dot"
								],
								["b", "b"]
							]
						]
					]
				]
			],
			[
				13,
				"cellId",
				[
					6,
					[
						7,
						"scaledUV",
						"sub"
					],
					["gv"]
				]
			],
			[
				13,
				"ix",
				[
					6,
					[
						7,
						"std",
						"round"
					],
					[[
						1,
						[
							7,
							"cellId",
							"x"
						],
						"*",
						"INV_SQRT3_2"
					]]
				]
			],
			[
				13,
				"iy",
				[
					6,
					[
						7,
						"std",
						"round"
					],
					[[
						1,
						[
							7,
							"cellId",
							"y"
						],
						"*",
						[5, "2"]
					]]
				]
			],
			[
				13,
				"q0",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								["ix", "iy"]
							],
							"mul"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "123.34"], [5, "345.45"]]
						]]
					]]
				]
			],
			[
				13,
				"q1",
				[
					6,
					[
						7,
						"q0",
						"add"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["q0", [
							6,
							[
								7,
								"q0",
								"add"
							],
							[[5, "34.345"]]
						]]
					]]
				]
			],
			[
				13,
				"rand",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							7,
							"q1",
							"x"
						],
						"*",
						[
							7,
							"q1",
							"y"
						]
					]]
				]
			],
			[
				13,
				"variedFactor",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[5, "1"],
						"+",
						[
							1,
							[
								1,
								[
									1,
									"rand",
									"-",
									[5, "0.5"]
								],
								"*",
								[5, "2"]
							],
							"*",
							"variation"
						]
					], [5, "0"]]
				]
			],
			[
				13,
				"pAbs",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							7,
							"gv",
							"x"
						]]
					], [
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							7,
							"gv",
							"y"
						]]
					]]
				]
			],
			[
				13,
				"hexDist",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["pAbs", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["SQRT3_2", [5, "0.5"]]
						]]
					], [
						7,
						"pAbs",
						"y"
					]]
				]
			],
			[
				13,
				"edgeDist",
				[
					1,
					[5, "0.5"],
					"-",
					"hexDist"
				]
			],
			[
				13,
				"lineWidth",
				[
					1,
					"thickness",
					"*",
					[5, "0.01"]
				]
			],
			[
				13,
				"pixel",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fwidth"
						],
						[[
							7,
							"scaledUV",
							"x"
						]]
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"soft",
				[
					1,
					"softness",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"lo",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"lineWidth",
						"-",
						"soft"
					], [5, "0"]]
				]
			],
			[
				13,
				"hi",
				[
					1,
					[
						1,
						"lineWidth",
						"+",
						"pixel"
					],
					"+",
					"soft"
				]
			],
			[
				13,
				"lineMask",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								"lo",
								"hi",
								"edgeDist"
							]
						]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"step"
						],
						[[5, "0.0001"], "lineWidth"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["lineMask", "variedFactor"]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std",
			"SQRT3",
			"flooredMod2",
			"INV_SQRT3_2",
			"SQRT3_2"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std,
		SQRT3,
		flooredMod2,
		INV_SQRT3_2,
		SQRT3_2
	})
}) && $.f)({}));
const componentDefinition = {
	name: "HexGrid",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Honeycomb hexagonal grid pattern",
	acceptsUVContext: true,
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "Cell fill color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Grid line color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		cells: {
			default: 8,
			description: "Number of hexagons across the shortest canvas edge",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 40,
				step: 1,
				label: "Cells",
				group: "Effect"
			}
		},
		thickness: {
			default: 1,
			description: "Thickness of the hex grid lines",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 10,
				step: .1,
				label: "Thickness",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			transform: transformAngle,
			description: "Rotation of the grid in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Softness of the line edges (0 = crisp, 1 = very soft)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		variation: {
			default: 0,
			description: "Per-cell random lightening/darkening of the cell fill color",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Variation",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, propValues } = params;
		const field = call(hexGridField, "hexGridField", [
			params.uvContext ?? params.ctx.uv,
			params.effectiveViewportSize ?? params.ctx.viewportSize,
			uniforms.cells,
			uniforms.thickness,
			uniforms.rotation,
			uniforms.softness,
			uniforms.variation
		]);
		const lineMask = field.member("x");
		const variedFactor = field.member("y");
		const variedColorA = vec4(uniforms.colorA.member("rgb").mul(variedFactor), uniforms.colorA.member("a"));
		return call(mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			variedColorA,
			uniforms.colorB,
			lineMask
		]);
	}
};
var HexGrid_default = componentDefinition;
export { componentDefinition as n, hexGridField as r, HexGrid_default as t };
