// shaders/dist/core/Hologram-BQ1kz8M2.js
import { At as call, It as vec4, Pt as floatE, R as std, T as sdf_exports, d as sdf3d_exports, et as mxNoiseFloat2, n as tgpu, on as d, rt as noise_exports } from "./coords-BlwgBpJf.js";
import { n as glass_exports } from "./glass-DdgMBWkF.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
import { t as shapeEffectBoundingBoxDeclaration } from "./shapeEffectBounds-BlxxsYMN.js";
var { sdfSpaceUV, offsetUV } = glass_exports;
var { createSvgSdfSampler, createAnalyticSdfSampler, driveAnalyticSubProps, ANALYTIC_SDF_EXTRA_FIELDS } = sdf_exports;
var { createVolumetricFieldComputeNode, buildVolumetricFieldSampler, VOLUMETRIC_FIELD_EXTRA_FIELDS, is3dShapeType, isVolumetricShapeType } = sdf3d_exports;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
var TWO_PI = 6.283185307179586;
function resolveShapeType(shapeType, shape) {
	if (shapeType) return shapeType;
	if (shape && typeof shape === "object") return shape.type || "circleSDF";
	if (typeof shape === "string") try {
		return JSON.parse(shape).type || "circleSDF";
	} catch {
		return "circleSDF";
	}
	return "circleSDF";
}
const hologramHash21 = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	return std.fract(__tsover_mul(std.sin(std.dot(p, d.vec2f(12.9898, 78.233))), 43758.5453));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				1,
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["p", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "12.9898"], [5, "78.233"]]
						]]
					]]
				],
				"*",
				[5, "43758.5453"]
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const hologramDisplaceUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, distortion, animTime) => {
	"use gpu";
	const wob = __tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_mul(uv.y, 38), __tsover_mul(animTime, 2.3))), .5), __tsover_mul(std.sin(__tsover_sub(__tsover_mul(uv.y, 11.5), __tsover_mul(animTime, 1.7))), .5));
	const dispX = __tsover_mul(__tsover_mul(wob, distortion), .012);
	return d.vec2f(__tsover_add(uv.x, dispX), uv.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "distortion"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"wob",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									[
										7,
										"uv",
										"y"
									],
									"*",
									[5, "38"]
								],
								"+",
								[
									1,
									"animTime",
									"*",
									[5, "2.3"]
								]
							]]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									[
										7,
										"uv",
										"y"
									],
									"*",
									[5, "11.5"]
								],
								"-",
								[
									1,
									"animTime",
									"*",
									[5, "1.7"]
								]
							]]
						],
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"dispX",
				[
					1,
					[
						1,
						"wob",
						"*",
						"distortion"
					],
					"*",
					[5, "0.012"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						7,
						"uv",
						"x"
					],
					"+",
					"dispX"
				], [
					7,
					"uv",
					"y"
				]]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const hologramComposite = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec2f,
	d.vec2f,
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((sC, sR, sL, sU, sD, uv, viewportSize, color, p0, p1, p2, p3) => {
	"use gpu";
	const scl = p0.x;
	const edgeSoftness = p0.y;
	const brightness = p0.z;
	const fill = p0.w;
	const edgeGlow = p1.x;
	const depthLines = p1.y;
	const depthScale = p1.z;
	const scanlines = p1.w;
	const scanlineScale = p2.x;
	const scanlineSpeed = p2.y;
	const sweep = p2.z;
	const flicker = p2.w;
	const distortion = p3.x;
	const grain = p3.y;
	const animTime = p3.z;
	const volumetric = p3.w;
	const sharp = std.max(__tsover_mul(edgeSoftness, .5), .001);
	const pxH = __tsover_div(1, viewportSize.y);
	const sdf = __tsover_div(sC.x, scl);
	if (sdf > __tsover_mul(pxH, 2)) return d.vec4f(0, 0, 0, 0);
	const EPS = .008;
	const gradX = __tsover_div(__tsover_sub(sR.x, sL.x), __tsover_mul(EPS, 2));
	const thickness = std.max(__tsover_mul(sdf, -1), 0);
	const thin = std.exp(__tsover_mul(thickness, -5.5));
	const nx = std.clamp(__tsover_div(__tsover_sub(sR.w, sL.w), __tsover_mul(EPS, 2)), -2, 2);
	const ny = std.clamp(__tsover_div(__tsover_sub(sU.w, sD.w), __tsover_mul(EPS, 2)), -2, 2);
	const nrm = std.normalize(d.vec3f(nx, ny, -1));
	const surfFresVol = std.pow(std.clamp(__tsover_sub(1, std.abs(nrm.z)), 0, 1), d.f32(2.2));
	const surfFres = std.select(d.f32(0), surfFresVol, volumetric > .5);
	const depthMetric = std.select(thickness, sC.w, volumetric > .5);
	const rimVol = __tsover_add(__tsover_mul(thin, .55), surfFres);
	const rimBase = std.select(thin, rimVol, volumetric > .5);
	const rim = __tsover_mul(rimBase, edgeGlow);
	const bodyFill = __tsover_mul(__tsover_sub(1, std.exp(__tsover_mul(thickness, -5))), fill);
	const phase = __tsover_sub(__tsover_mul(depthMetric, depthScale), __tsover_mul(animTime, 1.2));
	const f = std.fract(phase);
	const distI = std.min(f, __tsover_sub(1, f));
	const contour = __tsover_mul(std.smoothstep(.06, 0, distI), depthLines);
	const scanPhase = __tsover_sub(__tsover_mul(uv.y, scanlineScale), __tsover_mul(animTime, __tsover_mul(scanlineSpeed, 3)));
	const scan = __tsover_add(__tsover_mul(std.sin(__tsover_mul(scanPhase, TWO_PI)), .5), .5);
	const scanMask = std.mix(d.f32(1), scan, scanlines);
	const sweepPos = std.fract(__tsover_mul(animTime, .16));
	const sweepD = __tsover_sub(uv.y, sweepPos);
	const sweepGlow = __tsover_mul(std.exp(__tsover_mul(__tsover_mul(sweepD, sweepD), -650)), sweep);
	const shimmer = __tsover_add(__tsover_mul(std.sin(__tsover_mul(animTime, 40)), .04), __tsover_mul(std.sin(__tsover_mul(animTime, 23.3)), .03));
	const dropNoise = __tsover_add(__tsover_mul(mxNoiseFloat2(d.vec2f(__tsover_mul(animTime, 6), 11)), .5), .5);
	const drop = std.smoothstep(.6, .95, dropNoise);
	const flick = __tsover_mul(__tsover_add(1, shimmer), __tsover_sub(1, __tsover_mul(__tsover_mul(flicker, drop), .7)));
	const gn = hologramHash21(d.vec2f(__tsover_add(__tsover_mul(uv.x, viewportSize.x), __tsover_mul(animTime, 57.3)), __tsover_mul(uv.y, viewportSize.y)));
	const grainMul = __tsover_add(1, __tsover_mul(__tsover_mul(__tsover_sub(gn, .5), grain), .6));
	const glow = __tsover_add(__tsover_add(bodyFill, rim), contour);
	const core = __tsover_mul(__tsover_add(__tsover_mul(__tsover_mul(__tsover_mul(glow, scanMask), flick), grainMul), __tsover_mul(sweepGlow, scanMask)), brightness);
	const ca = __tsover_add(__tsover_mul(distortion, .12), .04);
	const fringe = __tsover_mul(__tsover_mul(__tsover_mul(rim, gradX), ca), .1);
	const rCore = __tsover_add(core, fringe);
	const bCore = __tsover_sub(core, fringe);
	let rgb = d.vec3f(__tsover_mul(color.x, rCore), __tsover_mul(color.y, core), __tsover_mul(color.z, bCore));
	const hot = std.clamp(__tsover_add(__tsover_mul(rim, .5), sweepGlow), 0, .8);
	rgb = std.mix(rgb, d.vec3f(core, core, core), d.vec3f(hot));
	const rb1 = std.clamp(__tsover_mul(__tsover_div(__tsover_mul(sdf, -1), sharp), 32), 0, 1);
	const transition = std.smoothstep(0, 1, rb1);
	const alpha = __tsover_mul(std.clamp(core, 0, 1), transition);
	return d.vec4f(rgb.x, rgb.y, rgb.z, alpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "sC"
			},
			{
				"type": "i",
				"name": "sR"
			},
			{
				"type": "i",
				"name": "sL"
			},
			{
				"type": "i",
				"name": "sU"
			},
			{
				"type": "i",
				"name": "sD"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewportSize"
			},
			{
				"type": "i",
				"name": "color"
			},
			{
				"type": "i",
				"name": "p0"
			},
			{
				"type": "i",
				"name": "p1"
			},
			{
				"type": "i",
				"name": "p2"
			},
			{
				"type": "i",
				"name": "p3"
			}
		],
		"body": [0, [
			[
				13,
				"scl",
				[
					7,
					"p0",
					"x"
				]
			],
			[
				13,
				"edgeSoftness",
				[
					7,
					"p0",
					"y"
				]
			],
			[
				13,
				"brightness",
				[
					7,
					"p0",
					"z"
				]
			],
			[
				13,
				"fill",
				[
					7,
					"p0",
					"w"
				]
			],
			[
				13,
				"edgeGlow",
				[
					7,
					"p1",
					"x"
				]
			],
			[
				13,
				"depthLines",
				[
					7,
					"p1",
					"y"
				]
			],
			[
				13,
				"depthScale",
				[
					7,
					"p1",
					"z"
				]
			],
			[
				13,
				"scanlines",
				[
					7,
					"p1",
					"w"
				]
			],
			[
				13,
				"scanlineScale",
				[
					7,
					"p2",
					"x"
				]
			],
			[
				13,
				"scanlineSpeed",
				[
					7,
					"p2",
					"y"
				]
			],
			[
				13,
				"sweep",
				[
					7,
					"p2",
					"z"
				]
			],
			[
				13,
				"flicker",
				[
					7,
					"p2",
					"w"
				]
			],
			[
				13,
				"distortion",
				[
					7,
					"p3",
					"x"
				]
			],
			[
				13,
				"grain",
				[
					7,
					"p3",
					"y"
				]
			],
			[
				13,
				"animTime",
				[
					7,
					"p3",
					"z"
				]
			],
			[
				13,
				"volumetric",
				[
					7,
					"p3",
					"w"
				]
			],
			[
				13,
				"sharp",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"edgeSoftness",
						"*",
						[5, "0.5"]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"pxH",
				[
					1,
					[5, "1"],
					"/",
					[
						7,
						"viewportSize",
						"y"
					]
				]
			],
			[
				13,
				"sdf",
				[
					1,
					[
						7,
						"sC",
						"x"
					],
					"/",
					"scl"
				]
			],
			[
				11,
				[
					1,
					"sdf",
					">",
					[
						1,
						"pxH",
						"*",
						[5, "2"]
					]
				],
				[10, [
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "0"],
						[5, "0"]
					]
				]]
			],
			[
				13,
				"EPS",
				[5, "0.008"]
			],
			[
				13,
				"gradX",
				[
					1,
					[
						1,
						[
							7,
							"sR",
							"x"
						],
						"-",
						[
							7,
							"sL",
							"x"
						]
					],
					"/",
					[
						1,
						"EPS",
						"*",
						[5, "2"]
					]
				]
			],
			[
				13,
				"thickness",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"sdf",
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					], [5, "0"]]
				]
			],
			[
				13,
				"thin",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						"thickness",
						"*",
						[
							4,
							"-",
							[5, "5.5"]
						]
					]]
				]
			],
			[
				13,
				"nx",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"sR",
									"w"
								],
								"-",
								[
									7,
									"sL",
									"w"
								]
							],
							"/",
							[
								1,
								"EPS",
								"*",
								[5, "2"]
							]
						],
						[
							4,
							"-",
							[5, "2"]
						],
						[5, "2"]
					]
				]
			],
			[
				13,
				"ny",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"sU",
									"w"
								],
								"-",
								[
									7,
									"sD",
									"w"
								]
							],
							"/",
							[
								1,
								"EPS",
								"*",
								[5, "2"]
							]
						],
						[
							4,
							"-",
							[5, "2"]
						],
						[5, "2"]
					]
				]
			],
			[
				13,
				"nrm",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"nx",
							"ny",
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]]
				]
			],
			[
				13,
				"surfFresVol",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[5, "1"],
								"-",
								[
									6,
									[
										7,
										"std",
										"abs"
									],
									[[
										7,
										"nrm",
										"z"
									]]
								]
							],
							[5, "0"],
							[5, "1"]
						]
					], [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "2.2"]]
					]]
				]
			],
			[
				13,
				"surfFres",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						"surfFresVol",
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"depthMetric",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"thickness",
						[
							7,
							"sC",
							"w"
						],
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"rimVol",
				[
					1,
					[
						1,
						"thin",
						"*",
						[5, "0.55"]
					],
					"+",
					"surfFres"
				]
			],
			[
				13,
				"rimBase",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"thin",
						"rimVol",
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"rim",
				[
					1,
					"rimBase",
					"*",
					"edgeGlow"
				]
			],
			[
				13,
				"bodyFill",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								"thickness",
								"*",
								[
									4,
									"-",
									[5, "5"]
								]
							]]
						]
					],
					"*",
					"fill"
				]
			],
			[
				13,
				"phase",
				[
					1,
					[
						1,
						"depthMetric",
						"*",
						"depthScale"
					],
					"-",
					[
						1,
						"animTime",
						"*",
						[5, "1.2"]
					]
				]
			],
			[
				13,
				"f",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["phase"]
				]
			],
			[
				13,
				"distI",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					["f", [
						1,
						[5, "1"],
						"-",
						"f"
					]]
				]
			],
			[
				13,
				"contour",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0.06"],
							[5, "0"],
							"distI"
						]
					],
					"*",
					"depthLines"
				]
			],
			[
				13,
				"scanPhase",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"y"
						],
						"*",
						"scanlineScale"
					],
					"-",
					[
						1,
						"animTime",
						"*",
						[
							1,
							"scanlineSpeed",
							"*",
							[5, "3"]
						]
					]
				]
			],
			[
				13,
				"scan",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								"scanPhase",
								"*",
								"TWO_PI"
							]]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"scanMask",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						"scan",
						"scanlines"
					]
				]
			],
			[
				13,
				"sweepPos",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						"animTime",
						"*",
						[5, "0.16"]
					]]
				]
			],
			[
				13,
				"sweepD",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					"sweepPos"
				]
			],
			[
				13,
				"sweepGlow",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"exp"
						],
						[[
							1,
							[
								1,
								"sweepD",
								"*",
								"sweepD"
							],
							"*",
							[
								4,
								"-",
								[5, "650"]
							]
						]]
					],
					"*",
					"sweep"
				]
			],
			[
				13,
				"shimmer",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								"animTime",
								"*",
								[5, "40"]
							]]
						],
						"*",
						[5, "0.04"]
					],
					"+",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								"animTime",
								"*",
								[5, "23.3"]
							]]
						],
						"*",
						[5, "0.03"]
					]
				]
			],
			[
				13,
				"dropNoise",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"noise",
								"mxNoiseFloat2"
							],
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									1,
									"animTime",
									"*",
									[5, "6"]
								], [5, "11"]]
							]]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"drop",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.6"],
						[5, "0.95"],
						"dropNoise"
					]
				]
			],
			[
				13,
				"flick",
				[
					1,
					[
						1,
						[5, "1"],
						"+",
						"shimmer"
					],
					"*",
					[
						1,
						[5, "1"],
						"-",
						[
							1,
							[
								1,
								"flicker",
								"*",
								"drop"
							],
							"*",
							[5, "0.7"]
						]
					]
				]
			],
			[
				13,
				"gn",
				[
					6,
					"hologramHash21",
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								1,
								[
									7,
									"uv",
									"x"
								],
								"*",
								[
									7,
									"viewportSize",
									"x"
								]
							],
							"+",
							[
								1,
								"animTime",
								"*",
								[5, "57.3"]
							]
						], [
							1,
							[
								7,
								"uv",
								"y"
							],
							"*",
							[
								7,
								"viewportSize",
								"y"
							]
						]]
					]]
				]
			],
			[
				13,
				"grainMul",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						[
							1,
							[
								1,
								"gn",
								"-",
								[5, "0.5"]
							],
							"*",
							"grain"
						],
						"*",
						[5, "0.6"]
					]
				]
			],
			[
				13,
				"glow",
				[
					1,
					[
						1,
						"bodyFill",
						"+",
						"rim"
					],
					"+",
					"contour"
				]
			],
			[
				13,
				"core",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									"glow",
									"*",
									"scanMask"
								],
								"*",
								"flick"
							],
							"*",
							"grainMul"
						],
						"+",
						[
							1,
							"sweepGlow",
							"*",
							"scanMask"
						]
					],
					"*",
					"brightness"
				]
			],
			[
				13,
				"ca",
				[
					1,
					[
						1,
						"distortion",
						"*",
						[5, "0.12"]
					],
					"+",
					[5, "0.04"]
				]
			],
			[
				13,
				"fringe",
				[
					1,
					[
						1,
						[
							1,
							"rim",
							"*",
							"gradX"
						],
						"*",
						"ca"
					],
					"*",
					[5, "0.1"]
				]
			],
			[
				13,
				"rCore",
				[
					1,
					"core",
					"+",
					"fringe"
				]
			],
			[
				13,
				"bCore",
				[
					1,
					"core",
					"-",
					"fringe"
				]
			],
			[
				12,
				"rgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								7,
								"color",
								"x"
							],
							"*",
							"rCore"
						],
						[
							1,
							[
								7,
								"color",
								"y"
							],
							"*",
							"core"
						],
						[
							1,
							[
								7,
								"color",
								"z"
							],
							"*",
							"bCore"
						]
					]
				]
			],
			[
				13,
				"hot",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"rim",
								"*",
								[5, "0.5"]
							],
							"+",
							"sweepGlow"
						],
						[5, "0"],
						[5, "0.8"]
					]
				]
			],
			[
				2,
				"rgb",
				"=",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"rgb",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								"core",
								"core",
								"core"
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["hot"]
						]
					]
				]
			],
			[
				13,
				"rb1",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									"sdf",
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								],
								"/",
								"sharp"
							],
							"*",
							[5, "32"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"transition",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[5, "1"],
						"rb1"
					]
				]
			],
			[
				13,
				"alpha",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							"core",
							[5, "0"],
							[5, "1"]
						]
					],
					"*",
					"transition"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"rgb",
						"x"
					],
					[
						7,
						"rgb",
						"y"
					],
					[
						7,
						"rgb",
						"z"
					],
					"alpha"
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"TWO_PI",
			"noise",
			"hologramHash21"
		]
	},
	externals: () => ({
		std,
		d,
		TWO_PI,
		noise: noise_exports,
		hologramHash21
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Hologram",
	boundingBoxDeclaration: shapeEffectBoundingBoxDeclaration,
	category: "Shape Effects",
	description: "Volumetric sci-fi hologram — a translucent emissive projection of the shape with fresnel-lit edges, depth scan-lines that wrap 3D forms, CRT scanlines, flicker and beam wobble",
	animatedTime: { speed: "speed" },
	extraFields: {
		...VOLUMETRIC_FIELD_EXTRA_FIELDS,
		...ANALYTIC_SDF_EXTRA_FIELDS
	},
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the hologram shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		scale: {
			default: 1,
			description: "Scale of the hologram shape (1 = default size)",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Scale",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the hologram shape in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		},
		color: {
			default: "#5ec8ff",
			transform: transformColor,
			description: "Projection tint — the colour the whole hologram emits (classic sci-fi cyan by default)",
			ui: {
				type: "color",
				label: "Color",
				group: "Style"
			}
		},
		brightness: {
			default: 2.5,
			description: "Overall emission gain. Bright edges blow out toward white for a hot, projected look.",
			ui: {
				type: "range",
				min: .2,
				max: 3,
				step: .01,
				label: "Brightness",
				group: "Style"
			}
		},
		edgeSoftness: {
			default: .04,
			description: "Softness of the shape boundary edge — keep low for crisp projection edges",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Softness",
				group: "Style"
			}
		},
		fill: {
			default: .45,
			description: "Translucent interior glow. Driven by optical thickness, so thicker parts of a 3D form glow more — the core volumetric cue.",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Volume Fill",
				group: "Volume"
			}
		},
		edgeGlow: {
			default: 1,
			description: "Fresnel rim intensity — bright glowing silhouette and, on 3D shapes, the grazing internal walls of extruded logos",
			ui: {
				type: "range",
				min: 0,
				max: 2.5,
				step: .01,
				label: "Edge Glow",
				group: "Volume"
			}
		},
		depthLines: {
			default: .5,
			description: "Iso-depth scan-lines that hug the surface and reveal the 3D form like a volumetric scan. On flat shapes they become concentric contour rings.",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Depth Lines",
				group: "Volume"
			}
		},
		depthScale: {
			default: 26,
			description: "Density of the depth scan-lines — higher packs more contour bands through the volume",
			ui: {
				type: "range",
				min: 4,
				max: 70,
				step: 1,
				label: "Depth Density",
				group: "Volume"
			}
		},
		scanlines: {
			default: .55,
			description: "CRT scanline contrast — how dark the gaps between horizontal projection lines get",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Scanlines",
				group: "Scanlines"
			}
		},
		scanlineScale: {
			default: 130,
			description: "Number of horizontal scanlines across the canvas height",
			ui: {
				type: "range",
				min: 20,
				max: 400,
				step: 1,
				label: "Scanline Count",
				group: "Scanlines"
			}
		},
		scanlineSpeed: {
			default: 1,
			description: "How fast the scanlines drift upward — 0 holds them static",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .01,
				label: "Scanline Drift",
				group: "Scanlines"
			}
		},
		sweep: {
			default: .1,
			description: "Brightness of the refresh bar that sweeps vertically through the projection",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Sweep Bar",
				group: "Scanlines"
			}
		},
		flicker: {
			default: .3,
			description: "Unstable-projector flicker — a fast shimmer plus occasional brightness dropouts",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Flicker",
				group: "Interference"
			}
		},
		distortion: {
			default: .1,
			description: "Continuous horizontal wobble and chromatic edge fringing — the wavering of an unstable beam",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Distortion",
				group: "Interference"
			}
		},
		grain: {
			default: .7,
			description: "Fine projection static — animated per-pixel noise over the emission",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Grain",
				group: "Interference"
			}
		},
		speed: {
			default: 1,
			description: "Master animation speed for every animated layer. 0 freezes the projection.",
			ui: {
				type: "range",
				min: 0,
				max: 3,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file — when non-empty, activates SVG mode and triggers a shader recompile"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type — triggers recompile when shape is switched. When empty, derived from shape JSON at mount time."
		}
	},
	compute: (params) => {
		const vf = createVolumetricFieldComputeNode(params, () => params.getCpuValue("shape"), "none");
		return vf ? {
			outputs: vf.outputs,
			getComputeNodes: (fp) => vf.getComputeNodes(fp)
		} : null;
	},
	fragment: (params) => {
		const { computeOutputs, uniforms, propValues, ctx, getCpuValue } = params;
		let sampler;
		let gradSampler;
		let volumetric = false;
		const fieldTex = computeOutputs?.volumetricFieldTexture;
		if (fieldTex) {
			const s = buildVolumetricFieldSampler(fieldTex, {
				originX: uniforms._vfOriginX,
				originY: uniforms._vfOriginY,
				spanX: uniforms._vfSpanX,
				spanY: uniforms._vfSpanY,
				activeResF: uniforms._vfActiveRes,
				rBound: uniforms._vfRBound
			});
			sampler = s.volumetricFieldSampler;
			gradSampler = s.volumetricFieldSamplerFast;
			volumetric = true;
		} else {
			const shapeSdfUrl = propValues.shapeSdfUrl || "";
			const shapeType = propValues.shapeType || "";
			if (shapeSdfUrl) {
				sampler = createSvgSdfSampler(params, shapeSdfUrl);
				gradSampler = sampler;
				volumetric = isVolumetricShapeType(shapeType);
			} else {
				const st = resolveShapeType(shapeType, getCpuValue("shape"));
				sampler = createAnalyticSdfSampler(st, driveAnalyticSubProps(params, () => getCpuValue("shape")));
				gradSampler = sampler;
				volumetric = is3dShapeType(st);
			}
		}
		const t = animatedTime(params);
		const suv = call(hologramDisplaceUV, "hologramDisplaceUV", [
			ctx.uv,
			uniforms.distortion,
			t
		]);
		const sdfUV = call(sdfSpaceUV, "sdfSpaceUV", [
			uniforms.center,
			uniforms.scale,
			uniforms.rotation,
			suv,
			ctx.aspect
		]);
		const sC = sampler(sdfUV);
		const sR = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(.008),
			floatE(0)
		]));
		const sL = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(-.008),
			floatE(0)
		]));
		const sU = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(0),
			floatE(.008)
		]));
		const sD = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(0),
			floatE(-.008)
		]));
		const volFlag = floatE(volumetric ? 1 : 0);
		const p0 = vec4(uniforms.scale, uniforms.edgeSoftness, uniforms.brightness, uniforms.fill);
		const p1 = vec4(uniforms.edgeGlow, uniforms.depthLines, uniforms.depthScale, uniforms.scanlines);
		const p2 = vec4(uniforms.scanlineScale, uniforms.scanlineSpeed, uniforms.sweep, uniforms.flicker);
		const p3 = vec4(uniforms.distortion, uniforms.grain, t, volFlag);
		return call(hologramComposite, "hologramComposite", [
			sC,
			sR,
			sL,
			sU,
			sD,
			ctx.uv,
			ctx.viewportSize,
			uniforms.color.member("rgb"),
			p0,
			p1,
			p2,
			p3
		]);
	}
};
var Hologram_default = componentDefinition;
export { hologramHash21 as a, hologramDisplaceUV as i, componentDefinition as n, hologramComposite as r, Hologram_default as t };
