// shaders/dist/core/HueShift-BJGcv19-.js
import { At as call, Ot as ZERO, R as std, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
var DEG_TO_RAD = Math.PI / 180;
var SQRT3 = 1.7320508075688772;
const hueShiftColor = tgpu.fn([d.vec4f, d.f32], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, shiftDeg) => {
	"use gpu";
	const angle = __tsover_mul(shiftDeg, DEG_TO_RAD);
	const cosA = std.cos(angle);
	const sinA = std.sin(angle);
	const k = __tsover_div(1, 3);
	const oneMinusCos = __tsover_sub(1, cosA);
	const kOneMinusCos = __tsover_mul(oneMinusCos, k);
	const sinOverSqrt3 = __tsover_div(sinA, SQRT3);
	const m00 = __tsover_add(cosA, kOneMinusCos);
	const m01 = __tsover_sub(kOneMinusCos, sinOverSqrt3);
	const m02 = __tsover_add(kOneMinusCos, sinOverSqrt3);
	const m10 = __tsover_add(kOneMinusCos, sinOverSqrt3);
	const m11 = __tsover_add(cosA, kOneMinusCos);
	const m12 = __tsover_sub(kOneMinusCos, sinOverSqrt3);
	const m20 = __tsover_sub(kOneMinusCos, sinOverSqrt3);
	const m21 = __tsover_add(kOneMinusCos, sinOverSqrt3);
	const m22 = __tsover_add(cosA, kOneMinusCos);
	const r = color.x;
	const g = color.y;
	const b = color.z;
	const newR = __tsover_add(__tsover_add(__tsover_mul(r, m00), __tsover_mul(g, m01)), __tsover_mul(b, m02));
	const newG = __tsover_add(__tsover_add(__tsover_mul(r, m10), __tsover_mul(g, m11)), __tsover_mul(b, m12));
	const newB = __tsover_add(__tsover_add(__tsover_mul(r, m20), __tsover_mul(g, m21)), __tsover_mul(b, m22));
	return d.vec4f(newR, newG, newB, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}, {
			"type": "i",
			"name": "shiftDeg"
		}],
		"body": [0, [
			[
				13,
				"angle",
				[
					1,
					"shiftDeg",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angle"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angle"]
				]
			],
			[
				13,
				"k",
				[
					1,
					[5, "1"],
					"/",
					[5, "3"]
				]
			],
			[
				13,
				"oneMinusCos",
				[
					1,
					[5, "1"],
					"-",
					"cosA"
				]
			],
			[
				13,
				"kOneMinusCos",
				[
					1,
					"oneMinusCos",
					"*",
					"k"
				]
			],
			[
				13,
				"sinOverSqrt3",
				[
					1,
					"sinA",
					"/",
					"SQRT3"
				]
			],
			[
				13,
				"m00",
				[
					1,
					"cosA",
					"+",
					"kOneMinusCos"
				]
			],
			[
				13,
				"m01",
				[
					1,
					"kOneMinusCos",
					"-",
					"sinOverSqrt3"
				]
			],
			[
				13,
				"m02",
				[
					1,
					"kOneMinusCos",
					"+",
					"sinOverSqrt3"
				]
			],
			[
				13,
				"m10",
				[
					1,
					"kOneMinusCos",
					"+",
					"sinOverSqrt3"
				]
			],
			[
				13,
				"m11",
				[
					1,
					"cosA",
					"+",
					"kOneMinusCos"
				]
			],
			[
				13,
				"m12",
				[
					1,
					"kOneMinusCos",
					"-",
					"sinOverSqrt3"
				]
			],
			[
				13,
				"m20",
				[
					1,
					"kOneMinusCos",
					"-",
					"sinOverSqrt3"
				]
			],
			[
				13,
				"m21",
				[
					1,
					"kOneMinusCos",
					"+",
					"sinOverSqrt3"
				]
			],
			[
				13,
				"m22",
				[
					1,
					"cosA",
					"+",
					"kOneMinusCos"
				]
			],
			[
				13,
				"r",
				[
					7,
					"color",
					"x"
				]
			],
			[
				13,
				"g",
				[
					7,
					"color",
					"y"
				]
			],
			[
				13,
				"b",
				[
					7,
					"color",
					"z"
				]
			],
			[
				13,
				"newR",
				[
					1,
					[
						1,
						[
							1,
							"r",
							"*",
							"m00"
						],
						"+",
						[
							1,
							"g",
							"*",
							"m01"
						]
					],
					"+",
					[
						1,
						"b",
						"*",
						"m02"
					]
				]
			],
			[
				13,
				"newG",
				[
					1,
					[
						1,
						[
							1,
							"r",
							"*",
							"m10"
						],
						"+",
						[
							1,
							"g",
							"*",
							"m11"
						]
					],
					"+",
					[
						1,
						"b",
						"*",
						"m12"
					]
				]
			],
			[
				13,
				"newB",
				[
					1,
					[
						1,
						[
							1,
							"r",
							"*",
							"m20"
						],
						"+",
						[
							1,
							"g",
							"*",
							"m21"
						]
					],
					"+",
					[
						1,
						"b",
						"*",
						"m22"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					"newR",
					"newG",
					"newB",
					[
						7,
						"color",
						"w"
					]
				]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"std",
			"SQRT3",
			"d"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		std,
		SQRT3,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "HueShift",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Rotate hue around the color wheel",
	requiresChild: true,
	props: { shift: {
		default: 0,
		description: "The amount to shift the hue by",
		compileTimeWhen: (prev, next) => prev === 0 !== (next === 0),
		ui: {
			type: ["range", "map"],
			min: -180,
			max: 180,
			step: 1,
			label: "Shift",
			group: "Adjustments"
		}
	} },
	fragment: ({ uniforms, childNode, propValues }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Hue Shift shader.");
			return ZERO;
		}
		if ((propValues.shift ?? 0) === 0) return childNode;
		return call(hueShiftColor, "hueShiftColor", [childNode, uniforms.shift]);
	}
};
var HueShift_default = componentDefinition;
export { componentDefinition as n, hueShiftColor as r, HueShift_default as t };
