// shaders/dist/core/ImageTexture-CRueGzAZ.js
import { At as call, a as scaleContain, i as sampleUV, l as scaleScaleDown, o as scaleCover, r as alphaCutLinear, s as scaleFill } from "./coords-CWOT0Wqh.js";
var OBJECT_FIT_MODES = {
	cover: 0,
	contain: 1,
	fill: 2,
	"scale-down": 3,
	none: 2
};
var SVG_RASTER_SIZE = 2048;
function isSvgUrl(url) {
	return /\.svg($|\?|#)/i.test(url);
}
function rasterizeSvg(url) {
	return new Promise((resolve, reject) => {
		const img = new Image();
		img.crossOrigin = "anonymous";
		img.onload = () => {
			const aspect = (img.naturalWidth || SVG_RASTER_SIZE) / (img.naturalHeight || SVG_RASTER_SIZE);
			const canvasW = aspect >= 1 ? SVG_RASTER_SIZE : Math.round(SVG_RASTER_SIZE * aspect);
			const canvasH = aspect >= 1 ? Math.round(SVG_RASTER_SIZE / aspect) : SVG_RASTER_SIZE;
			const canvas = document.createElement("canvas");
			canvas.width = canvasW;
			canvas.height = canvasH;
			const ctx = canvas.getContext("2d");
			if (!ctx) {
				reject(/* @__PURE__ */ new Error("Could not get 2D context"));
				return;
			}
			ctx.drawImage(img, 0, 0, canvasW, canvasH);
			resolve({
				source: canvas,
				width: canvasW,
				height: canvasH
			});
		};
		img.onerror = reject;
		img.src = url;
	});
}
const componentDefinition = {
	name: "ImageTexture",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Display an image with customizable object-fit modes",
	acceptsUVContext: true,
	props: {
		url: {
			default: "https://shaders.com/sample.jpg",
			description: "Upload an image or provide a URL",
			ui: {
				type: "image-upload",
				label: "Image",
				group: "Media"
			}
		},
		objectFit: {
			default: "fill",
			description: "How the image should be sized within the viewport",
			transform: (value) => {
				return {
					"cover": 0,
					"contain": 1,
					"fill": 2,
					"scale-down": 3,
					"none": 2
				}[value] ?? 2;
			},
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Cover",
						value: "cover"
					},
					{
						label: "Contain",
						value: "contain"
					},
					{
						label: "Fill",
						value: "fill"
					},
					{
						label: "Scale Down",
						value: "scale-down"
					}
				],
				label: "Object Fit",
				group: "Sizing"
			}
		}
	},
	fragment: ({ ctx, propValues, uvContext, effectiveViewportSize, getCpuValue, onBeforeRender, onCleanup, createMediaTexture, registerMediaTexture }) => {
		let current = createMediaTexture({
			width: 1,
			height: 1,
			label: "ImageTexture:placeholder"
		});
		let currentUrl = "";
		let isLoading = false;
		let isDisposed = false;
		const applyImage = (source, width, height) => {
			if (isDisposed) return;
			const next = createMediaTexture({
				width: Math.max(1, width),
				height: Math.max(1, height),
				label: "ImageTexture"
			});
			next.write(source);
			const previous = current;
			current = next;
			previous.destroy();
		};
		const loadImage = async (url) => {
			if (!url || url.trim() === "" || isLoading || isDisposed) return;
			isLoading = true;
			try {
				if (isSvgUrl(url)) {
					const { source, width, height } = await rasterizeSvg(url);
					if (isDisposed) return;
					currentUrl = url;
					applyImage(source, width, height);
				} else {
					const response = await fetch(url, { mode: "cors" });
					if (!response.ok) throw new Error(`HTTP ${response.status}`);
					const blob = await response.blob();
					const bitmap = await createImageBitmap(blob);
					if (isDisposed) {
						bitmap.close();
						return;
					}
					currentUrl = url;
					applyImage(bitmap, bitmap.width, bitmap.height);
					bitmap.close();
				}
			} catch (error) {
				const isRelative = !url.startsWith("http://") && !url.startsWith("https://") && !url.startsWith("data:");
				console.error(`[ImageTexture] Failed to load image\n  Requested URL: ${url}\n  Path type: ${isRelative ? "relative" : "absolute"}\n  Error:`, error);
			} finally {
				isLoading = false;
			}
		};
		setTimeout(() => {
			const initialUrl = getCpuValue("url") ?? "";
			if (initialUrl && initialUrl.trim() !== "") loadImage(initialUrl);
		}, 0);
		onBeforeRender(() => {
			const newUrl = getCpuValue("url") ?? "";
			if (newUrl !== currentUrl && !isLoading) loadImage(newUrl);
		});
		onCleanup(() => {
			isDisposed = true;
			current.destroy();
		});
		const image = registerMediaTexture(() => current.texture);
		const baseUV = uvContext ?? ctx.uv;
		const vp = effectiveViewportSize ?? ctx.viewportSize;
		const dims = image.dimensions();
		const rawFit = propValues.objectFit;
		const mode = typeof rawFit === "number" ? rawFit : OBJECT_FIT_MODES[rawFit] ?? 2;
		let uvScale;
		if (mode === 1) uvScale = call(scaleContain, "scaleContain", [dims, vp]);
		else if (mode === 2) uvScale = call(scaleFill, "scaleFill", []);
		else if (mode === 3) uvScale = call(scaleScaleDown, "scaleScaleDown", [dims, vp]);
		else uvScale = call(scaleCover, "scaleCover", [dims, vp]);
		const finalUV = call(sampleUV, "mediaSampleUV", [baseUV, uvScale]);
		return call(alphaCutLinear, "mediaAlphaCut", [image.sample(finalUV), finalUV]);
	}
};
var ImageTexture_default = componentDefinition;
export { componentDefinition as n, ImageTexture_default as t };
