// shaders/dist/core/index.js
import { Bt as needsTransformation, Dt as SHARED_SAMPLER_LAYOUT, Mt as composeNodeTree, Vt as screenUVToBoxLocal$1, _t as packConvertedStops, an as createUniformStore, f as maxPixelRatioForDevice, in as FieldHandle, jt as collectStructuralHashInputs, sn as updateFieldValue, t as SystemUniforms, vt as packStops, yt as getAnimatedTimeState, z as createComputeDispatcher, zt as boundingBoxToUVParams } from "./coords-CWOT0Wqh.js";
import "./glass-hUL2pCYG.js";
import "./Neon-Der_R-Bh.js";
import "./Emboss-D_-1_a3u.js";
import "./Form3D-BmCwmFPZ.js";
import "./ThinFilm-D_y48u9e.js";
import { o as transformColorSpace } from "./transformations-fqzznt4g.js";
import { _ as sortStops, a as buildDimensionalPlan, c as isDimensionalValue, d as colorToRGBA, f as setColorSpaceModeGpu, g as resolveStops, h as colorStopsTransform, i as boxHalfExtentsUV, l as resolveDimensionalProp, n as rootPassthrough, o as buildExtraSizeProps, p as MAX_COLOR_STOPS, r as createGpuUniformsMap, s as isDimensionalProp, u as setShaderCanvasDimensions } from "./porters-Dg80Yg90.js";
import "./AngularBlur-BJKQrsoZ.js";
import "./Arc-BZdfWWlz.js";
import "./Ascii-CR5iFmk8.js";
import "./Aurora-C0KIoJY5.js";
import "./BarShift-_mdK3UWW.js";
import "./Beam-DYvBEj6q.js";
import "./Blob-DgMJ5M6V.js";
import "./noiseColor-WSY8YuLZ.js";
import "./BlockNoise-CsBWCDzA.js";
import "./BlueNoise-D-WjLdgK.js";
import "./Blur-WeJoA31s.js";
import "./BrickPattern-DPez_hvr.js";
import "./BrightnessContrast-XvDV1ixv.js";
import { a as SVG3D_TYPES, c as isVolumetricShapeType, i as SVG3D_DEFAULTS, l as shape3dBoundingRadius, n as SHAPE3D_DEFAULTS, o as is3dShapeType, r as SHAPE3D_TYPES, s as isSvg3dShapeType } from "./shapeEffectBounds-BlxxsYMN.js";
import "./BrushedMetal-DYwNjVpX.js";
import "./Bulge-DcBJaYyW.js";
import "./CarbonFiber-DwYVLmd0.js";
import "./Chalkboard-D_FVdz4V.js";
import "./ChannelBlur-B70l9vql.js";
import "./Checkerboard-B_z4be-A.js";
import "./Chevron-CfGjPdH6.js";
import "./ChromaFlow-B9zZUgzN.js";
import "./ChromaticAberration-BrGXd7kq.js";
import "./Circle-DFx6OVnv.js";
import "./ColorWheel-CI_qR6vL.js";
import "./CompressionArtifacts-CVWkWovO.js";
import "./ConcentricSpin-lAMdNRDl.js";
import "./ConicGradient-C3wD1h2z.js";
import "./ContourLines-C3Hwaw1Y.js";
import "./CornerPin-DxUoEjhR.js";
import "./Crescent-BCf_ZZI3.js";
import "./Cross-CeQhwjVB.js";
import "./CRTScreen-Byau9--w.js";
import "./Crystal-CC7d_6kZ.js";
import "./CurlNoise-KsNJEh6E.js";
import "./CursorRipples-Jgm7iF0Z.js";
import "./CursorTrail-CkNn85vn.js";
import "./DataMosh-Dh7TYpbW.js";
import "./DiamondGradient-HmZwnTdD.js";
import "./DiffuseBlur-uXE0w0DD.js";
import "./Dither-YUNtD1tI.js";
import "./DOMTexture-DFvbXYrj.js";
import "./DotGrid-DABCUp6d.js";
import "./DropShadow-C7KAuipC.js";
import "./Duotone-CuJunWJ0.js";
import "./Ellipse-DrZqqQh2.js";
import "./Engraving-fQ1zZJ89.js";
import "./ErosionNoise-B01HEvZk.js";
import "./Exposure-o_CViEwJ.js";
import "./FallingLines-Bv8X8gRy.js";
import "./FilmGrain-CoW4iHsT.js";
import "./Flip-h2K3wLGX.js";
import "./FloatingParticles-W_QAKOku.js";
import "./Flower-DxOinfag.js";
import "./FlowField-CwU9iWbf.js";
import "./FlowingGradient-DdFrloZC.js";
import "./FlutedGlass-B9652Bh4.js";
import "./Fog-D5sIF3cX.js";
import "./FractalNoise-BFs55FIe.js";
import "./Frost-C6qVP9a2.js";
import "./GaborNoise-C1LVTyG3.js";
import "./Glass-DR9JYQMG.js";
import "./GlassTiles-D5Lrddns.js";
import "./Glitch-B93JzM-d.js";
import "./Glow-DlxNehzC.js";
import "./Godrays-n9Gjl-h6.js";
import "./Goo-BusLBgE3.js";
import "./GradientMap-CapG6NfM.js";
import "./Grayscale-Dx9gKGZb.js";
import "./Grid-gZHTJmwP.js";
import "./GridDistortion-BkkuULR2.js";
import "./Group-pHN72R84.js";
import "./Halftone-dlX8bpPV.js";
import "./Heart-CwELiXMT.js";
import "./Heatmap-unI_z4Uc.js";
import "./HexGrid-BTWeNz9Z.js";
import "./Hologram-3-c4mCzw.js";
import "./Holographic-SchvA6Dy.js";
import "./HueShift-BJGcv19-.js";
import "./ImageTexture-CRueGzAZ.js";
import "./Invert-DLot9mli.js";
import "./IsometricCubes-CFg5QBjO.js";
import "./Kaleidoscope-BQFm59wM.js";
import "./KeyFrames-DiAKRsxm.js";
import "./LensFlare-BpFbnSsM.js";
import "./LightEdge-CZpMFMKO.js";
import "./LightLeak-DGt9quKs.js";
import "./LinearBlur-ClAsNnkC.js";
import "./LinearGradient-B_dBLhFm.js";
import "./LiquidMetal-Bb9b2h9X.js";
import "./Liquify-B9ypclFj.js";
import "./Marble-CZYhaDcz.js";
import "./Mirror--NvGPp-u.js";
import "./MultiPointGradient-IJtw3dEC.js";
import "./fonts-3F3pIODs.js";
import "./ObjectTracker-CfW_aMyp.js";
import "./PagePeel-D_DnKmec.js";
import "./Paper-CZT7vlqM.js";
import "./Parallelogram-BAi9iFi0.js";
import "./Particles-DtxN6LL7.js";
import "./PerlinNoise-BCbLpWKn.js";
import "./Perspective-DXIafIz8.js";
import "./Pixelate-L-pZ8G79.js";
import "./PixelSort-BnZgslHG.js";
import "./PixelThrow-BAVO2zdp.js";
import "./Plasma-BELcebZa.js";
import "./Plastic-e8KLjCgS.js";
import "./PolarCoordinates-CNZ6Df5J.js";
import "./Polygon-BGBTOGLe.js";
import "./Posterize-Cs_kyyjG.js";
import "./Prism-BUalIt7a.js";
import "./ProgressiveBlur-C2wYIOH-.js";
import "./RadialGradient-B3tlBsQZ.js";
import "./RectangularCoordinates-DWWlrqB4.js";
import "./ReflectivePlane-BmM5w76k.js";
import "./Repeater-V37OGlry.js";
import "./Ring-WVyMeyoh.js";
import "./Ripples-Bv__zvme.js";
import "./RoundedRect-B62UFEeb.js";
import "./Saturation-DIRGc5kp.js";
import "./Scratches-ZFLTq06k.js";
import "./Sharpness-WYaGJlSI.js";
import "./Shatter-CPS5Kdei.js";
import "./SimplexNoise-DMWtripU.js";
import "./SineWave-BOoYgny1.js";
import "./Smoke-BGGH2iTS.js";
import "./SmokeFill-ClBFBWz3.js";
import "./SmokeFlow-DkBGrN_d.js";
import "./Solarize-D_ZPY_Cr.js";
import "./SolidColor-rdMtlNyB.js";
import "./Spherize-h1F31u9-.js";
import "./Spiral-BMFhUpkP.js";
import "./Star-D8d3Gk74.js";
import "./noiseStylize-C2jzZv4d.js";
import "./Stone-CfpfDKpn.js";
import "./Strands-256prKRG.js";
import "./Stretch-C7z0wdI7.js";
import "./Stripes-BVvuh6r7.js";
import "./StudioBackground-Cy2qPLYJ.js";
import "./SunBurst-vUglW69s.js";
import "./Surface3D-quYcBdK_.js";
import "./Swirl-BD3Fk-SC.js";
import "./Teardrop-Bxfkjjaz.js";
import "./Text-CrbVixvj.js";
import "./TiltShift-3s5RKGCa.js";
import "./Tint-DZUIRc-Z.js";
import "./Trapezoid-7VOLjftp.js";
import "./TriangularGrid-Qek0gRuV.js";
import "./Tritone-CqYOMKFE.js";
import "./Truchet-D6emdFny.js";
import "./Twirl-Y8OzHc49.js";
import "./Vesica-DFRPMcgz.js";
import "./VHS-B28K9Bfc.js";
import "./Vibrance-BkMNClGf.js";
import "./VideoTexture-DNfGZ4kz.js";
import "./Vignette-DWveIoCk.js";
import "./Voronoi-c1kHPUka.js";
import "./Water-CEmTC2Ii.js";
import "./Watercolor--Bg1ZPIG.js";
import "./WaveDistortion-BSyW8c28.js";
import "./WaveletNoise-cDkFw55m.js";
import "./Weave--Ex2UVBt.js";
import "./WebcamTexture-CPWUtxVm.js";
import "./Wool-BNmVUISf.js";
import "./WorleyNoise-CbhvPZS-.js";
import "./ZoomBlur-BR_do-LV.js";
import { t as getAllShaders } from "./shaderRegistry-DC0r_QlY.js";
import * as d from "typegpu/data";
import tgpu, { common } from "typegpu";
var OPTIONAL_FEATURES = ["float32-filterable", "timestamp-query"];
var rootsByDevice = /* @__PURE__ */ new WeakMap();
var lostCallbacks = /* @__PURE__ */ new WeakMap();
var lostDevices = /* @__PURE__ */ new WeakSet();
var replacementRoots = /* @__PURE__ */ new WeakMap();
async function acquireRoot(options = {}) {
	if (options.device) {
		const device = options.device;
		if (!lostDevices.has(device)) {
			let root = rootsByDevice.get(device);
			if (!root) {
				root = tgpu.initFromDevice({ device });
				rootsByDevice.set(device, root);
			}
			wireDeviceLost(root, options.onDeviceLost);
			return {
				root,
				device,
				adapter: options.adapter,
				createdDevice: false
			};
		}
		let replacement;
		for (;;) {
			const cached = replacementRoots.get(device);
			if (!cached) break;
			const cachedCtx = await cached.catch(() => null);
			if (cachedCtx && !lostDevices.has(cachedCtx.device)) {
				replacement = cached;
				break;
			}
			if (replacementRoots.get(device) === cached) replacementRoots.delete(device);
		}
		if (!replacement) {
			console.warn("[gpu] injected GPU device was lost — creating a replacement device for recovery");
			replacement = createOwnRoot();
			replacementRoots.set(device, replacement);
		}
		const ctx$1 = await replacement;
		wireDeviceLost(ctx$1.root, options.onDeviceLost);
		return {
			root: ctx$1.root,
			device: ctx$1.device,
			adapter: ctx$1.adapter,
			createdDevice: false
		};
	}
	const ctx = await createOwnRoot();
	wireDeviceLost(ctx.root, options.onDeviceLost);
	return {
		...ctx,
		adapter: options.adapter,
		createdDevice: true
	};
}
async function createOwnRoot() {
	if (typeof navigator === "undefined" || !navigator.gpu) throw new Error("[gpu] WebGPU is not available in this environment (navigator.gpu is undefined). This renderer is WebGPU-only.");
	const root = await tgpu.init({
		adapter: { powerPreference: "high-performance" },
		device: { optionalFeatures: OPTIONAL_FEATURES }
	});
	rootsByDevice.set(root.device, root);
	return {
		root,
		device: root.device,
		createdDevice: true
	};
}
function configureCanvasContext(root, canvas, options = {}) {
	return root.configureContext({
		canvas,
		alphaMode: "premultiplied",
		...options
	});
}
function wireDeviceLost(root, onDeviceLost) {
	const device = root.device;
	let callbacks = lostCallbacks.get(device);
	if (!callbacks) {
		callbacks = /* @__PURE__ */ new Set();
		lostCallbacks.set(device, callbacks);
		root.device.lost.then((info) => {
			lostDevices.add(device);
			if (info.reason === "destroyed") return;
			for (const cb of lostCallbacks.get(device) ?? []) cb(info);
		});
	}
	if (onDeviceLost) callbacks.add(onDeviceLost);
}
function structuralHash(inputs) {
	let h = 2166136261;
	const joined = inputs.join("");
	for (let i = 0; i < joined.length; i++) {
		h ^= joined.charCodeAt(i);
		h = Math.imul(h, 16777619);
	}
	return `${(h >>> 0).toString(16).padStart(8, "0")}:${joined.length}`;
}
function createPipelineCache(options = {}) {
	const maxSize = options.maxSize ?? 4;
	const dispose = options.dispose;
	const entries = /* @__PURE__ */ new Map();
	let clock = 0;
	let activeHash = null;
	let pendingHash = null;
	function touch(entry) {
		entry.lastUsed = ++clock;
	}
	function evictIfNeeded() {
		while (entries.size > maxSize) {
			let victim = null;
			for (const e of entries.values()) {
				if (e.hash === activeHash || e.hash === pendingHash) continue;
				if (!victim || e.lastUsed < victim.lastUsed) victim = e;
			}
			if (!victim) break;
			entries.delete(victim.hash);
			dispose?.(victim.value, victim.hash);
		}
	}
	return {
		getOrBuild(hash, buildFn) {
			const existing = entries.get(hash);
			if (existing) {
				touch(existing);
				if (hash === activeHash) pendingHash = null;
				else pendingHash = hash;
				return existing.value;
			}
			const value = buildFn();
			const entry = {
				hash,
				value,
				lastUsed: ++clock
			};
			entries.set(hash, entry);
			if (activeHash === null) {
				activeHash = hash;
				pendingHash = null;
			} else pendingHash = hash;
			evictIfNeeded();
			return value;
		},
		markReady(hash) {
			if (hash !== pendingHash) return;
			const entry = entries.get(hash);
			if (!entry) return;
			activeHash = hash;
			pendingHash = null;
			touch(entry);
			evictIfNeeded();
		},
		get renderValue() {
			const hash = activeHash ?? pendingHash;
			return hash ? entries.get(hash)?.value ?? null : null;
		},
		get activeHash() {
			return activeHash;
		},
		get pendingHash() {
			return pendingHash;
		},
		release(hash) {
			const entry = entries.get(hash);
			if (!entry) return;
			entries.delete(hash);
			dispose?.(entry.value, hash);
			if (activeHash === hash) activeHash = null;
			if (pendingHash === hash) pendingHash = null;
		},
		has(hash) {
			return entries.has(hash);
		},
		get size() {
			return entries.size;
		},
		clear() {
			for (const e of entries.values()) dispose?.(e.value, e.hash);
			entries.clear();
			activeHash = null;
			pendingHash = null;
		}
	};
}
var LABEL = "passManager";
function createPassManager(root, options) {
	const { textureManager, dispatcher } = options;
	const canvasFormat = options.canvasFormat ?? preferredCanvasFormat();
	const rttFormat = options.rttFormat ?? "rgba16float";
	const maxSampled = options.maxSampledTextures ?? 16;
	const vertex = options.fullScreenTriangle ?? common.fullScreenTriangle;
	const rttTextures = /* @__PURE__ */ new Map();
	const liveTextures = /* @__PURE__ */ new Set();
	let current = null;
	function buildSamplerBindGroup() {
		const s = textureManager.samplers;
		try {
			return root.createBindGroup(SHARED_SAMPLER_LAYOUT, {
				linearClamp: s.linearClamp,
				nearestClamp: s.nearestClamp,
				linearRepeat: s.linearRepeat,
				nearestRepeat: s.nearestRepeat
			});
		} catch (e) {
			console.error(`[${LABEL}] failed to build sampler bind group:`, e);
			return;
		}
	}
	function allocateRttTextures(ir, size) {
		const needed = /* @__PURE__ */ new Set();
		for (const t of ir.textures) {
			if (t.kind !== "rtt") continue;
			needed.add(t.key);
			if (!rttTextures.has(t.key)) {
				const rt = textureManager.createRenderTexture({
					width: size.width,
					height: size.height,
					format: rttFormat,
					label: `rtt:${t.key}`
				});
				rttTextures.set(t.key, rt);
				liveTextures.add(rt);
			} else rttTextures.get(t.key).resize(size.width, size.height);
		}
		for (const [key, rt] of [...rttTextures]) if (!needed.has(key)) {
			rt.destroy();
			liveTextures.delete(rt);
			rttTextures.delete(key);
		}
	}
	function buildTextureBindGroup(fragment, mediaResources) {
		if (!fragment.textureLayout || fragment.reads.length === 0) return void 0;
		const entries = {};
		for (const key of fragment.reads) {
			const rt = rttTextures.get(key);
			const media = mediaResources?.[key];
			const resource = rt ?? media;
			if (!resource) {
				console.error(`[${LABEL}] pass "${fragment.entry}" reads unknown texture "${key}"`);
				return;
			}
			entries[key] = resource.texture;
		}
		if (fragment.reads.length > maxSampled) console.warn(`[${LABEL}] pass samples ${fragment.reads.length} textures (> device limit ~${maxSampled}); composition may fail on this device.`);
		try {
			return root.createBindGroup(fragment.textureLayout, entries);
		} catch (e) {
			console.error(`[${LABEL}] failed to build texture bind group:`, e);
			return;
		}
	}
	function buildPipeline(fragment, targetFormat) {
		return root.createRenderPipeline({
			vertex,
			fragment: fragment.entry,
			targets: { format: targetFormat }
		});
	}
	function buildExternalBinding(fragment, externalSources) {
		if (!fragment.externalLayout || fragment.externalReads.length === 0) return void 0;
		const keys = fragment.externalReads;
		if (keys.length === 1) {
			const getSource = externalSources.get(keys[0]);
			if (!getSource) {
				console.error(`[${LABEL}] external read "${keys[0]}" has no registered source`);
				return;
			}
			return textureManager.createExternalTextureBinding({
				layout: fragment.externalLayout,
				entryKey: keys[0],
				getSource
			});
		}
		let current$1 = null;
		return {
			update() {
				const entries = {};
				for (const key of keys) {
					const source = externalSources.get(key)?.();
					if (!source) {
						current$1 = null;
						return null;
					}
					try {
						entries[key] = root.device.importExternalTexture({ source });
					} catch {
						current$1 = null;
						return null;
					}
				}
				current$1 = root.createBindGroup(fragment.externalLayout, entries);
				return current$1;
			},
			get current() {
				return current$1;
			},
			destroy() {
				current$1 = null;
			}
		};
	}
	function buildExecutablePass(fragment, target, mediaResources, externalSources) {
		const targetFormat = target.kind === "canvas" ? canvasFormat : rttFormat;
		const mediaReads = fragment.reads.filter((k) => mediaResources?.[k]);
		const mediaSnapshot = new Map(mediaReads.map((k) => [k, mediaResources[k].texture]));
		return {
			fragment,
			pipeline: buildPipeline(fragment, targetFormat),
			textureBindGroup: buildTextureBindGroup(fragment, mediaResources),
			externalBinding: buildExternalBinding(fragment, externalSources),
			target,
			mediaReads,
			mediaSnapshot
		};
	}
	function setComposition(ir, uniformBindGroup, size, mediaResources) {
		allocateRttTextures(ir, size);
		for (const spec of ir.computeSteps) spec.bindInputs?.((key) => rttTextures.get(key) ?? mediaResources?.[key]);
		const samplerBindGroup = buildSamplerBindGroup();
		const externalSources = /* @__PURE__ */ new Map();
		for (const e of ir.externalTextures) externalSources.set(e.key, e.getSource);
		const rttPasses = ir.rttPasses.map((p) => buildExecutablePass(p.fragment, {
			kind: "rtt",
			key: p.textureKey
		}, mediaResources, externalSources));
		const finalPass = buildExecutablePass(ir.finalPass, { kind: "canvas" }, mediaResources, externalSources);
		current = {
			ir,
			uniformBindGroup,
			samplerBindGroup,
			computeSteps: ir.computeSteps,
			rttPasses,
			finalPass,
			size,
			mediaResources
		};
	}
	function execPass(pass, view) {
		let p = pass.pipeline;
		if (current?.uniformBindGroup) p = p.with(current.uniformBindGroup);
		if (pass.mediaReads && pass.mediaReads.length > 0 && current?.mediaResources) {
			let swapped = false;
			for (const key of pass.mediaReads) if (current.mediaResources[key].texture !== pass.mediaSnapshot.get(key)) swapped = true;
			if (swapped) {
				pass.textureBindGroup = buildTextureBindGroup(pass.fragment, current.mediaResources);
				for (const key of pass.mediaReads) pass.mediaSnapshot.set(key, current.mediaResources[key].texture);
			}
		}
		if (pass.textureBindGroup) p = p.with(pass.textureBindGroup);
		if (pass.fragment.usesSamplers && current?.samplerBindGroup) p = p.with(current.samplerBindGroup);
		if (pass.fragment.externalReads.length > 0) {
			const bg = pass.externalBinding?.update() ?? null;
			if (!bg) {
				if (!pass.externalSkipLogged) {
					console.warn(`[${LABEL}] external texture not ready — skipping pass (${pass.fragment.externalReads.join(", ")})`);
					pass.externalSkipLogged = true;
				}
				return;
			}
			p = p.with(bg);
		}
		p.withColorAttachment({ view }).draw(3);
	}
	function render(canvasView, frameParams) {
		if (!current) return;
		for (const step of current.computeSteps) {
			const steps = step.getComputeNodes(frameParams);
			if (steps && steps.length) dispatcher.dispatch(steps);
		}
		repaintRtt();
		execPass(current.finalPass, canvasView);
	}
	function repaintRtt() {
		if (!current) return;
		for (const pass of current.rttPasses) {
			const rt = rttTextures.get(pass.target.key);
			if (!rt) continue;
			execPass(pass, renderView(rt));
		}
	}
	function renderView(rt) {
		return root.unwrap(rt.texture).createView();
	}
	function rebindPassTextures(pass) {
		if (!pass.fragment.textureLayout || pass.fragment.reads.length === 0) return;
		pass.textureBindGroup = buildTextureBindGroup(pass.fragment, current?.mediaResources);
		if (pass.mediaReads && pass.mediaSnapshot && current?.mediaResources) for (const key of pass.mediaReads) pass.mediaSnapshot.set(key, current.mediaResources[key].texture);
	}
	function resize(width, height) {
		for (const rt of rttTextures.values()) rt.resize(width, height);
		if (!current) return;
		current.size = {
			width,
			height
		};
		for (const spec of current.computeSteps) spec.bindInputs?.((key) => rttTextures.get(key) ?? current.mediaResources?.[key]);
		for (const pass of current.rttPasses) rebindPassTextures(pass);
		rebindPassTextures(current.finalPass);
	}
	return {
		setComposition,
		render,
		repaintRtt,
		resize,
		get textureCount() {
			return liveTextures.size;
		},
		get passCount() {
			return current ? current.rttPasses.length + 1 : 0;
		},
		dispose() {
			for (const rt of [...liveTextures]) rt.destroy();
			liveTextures.clear();
			rttTextures.clear();
			current = null;
		}
	};
}
function preferredCanvasFormat() {
	if (typeof navigator !== "undefined" && navigator.gpu) return navigator.gpu.getPreferredCanvasFormat();
	return "bgra8unorm";
}
function srgbViewFormat(format) {
	return `${format}-srgb`;
}
function createTextureManager(root) {
	const live = /* @__PURE__ */ new Set();
	function createRaw(props, ...usages) {
		return root.createTexture.call(root, props).$usage(...usages);
	}
	let linearClamp;
	let nearestClamp;
	let linearRepeat;
	let nearestRepeat;
	const samplers = {
		get linearClamp() {
			return linearClamp ??= root.createSampler({
				magFilter: "linear",
				minFilter: "linear",
				mipmapFilter: "linear",
				addressModeU: "clamp-to-edge",
				addressModeV: "clamp-to-edge"
			});
		},
		get nearestClamp() {
			return nearestClamp ??= root.createSampler({
				magFilter: "nearest",
				minFilter: "nearest",
				addressModeU: "clamp-to-edge",
				addressModeV: "clamp-to-edge"
			});
		},
		get linearRepeat() {
			return linearRepeat ??= root.createSampler({
				magFilter: "linear",
				minFilter: "linear",
				mipmapFilter: "linear",
				addressModeU: "repeat",
				addressModeV: "repeat"
			});
		},
		get nearestRepeat() {
			return nearestRepeat ??= root.createSampler({
				magFilter: "nearest",
				minFilter: "nearest",
				addressModeU: "repeat",
				addressModeV: "repeat"
			});
		}
	};
	function wrap(texture, width, height, format, srgb) {
		const managed = {
			texture,
			width,
			height,
			format,
			write(source) {
				texture.write(source);
			},
			unwrap() {
				return root.unwrap(texture);
			},
			createSrgbView() {
				if (!srgb) return void 0;
				return root.unwrap(texture).createView({ format: srgbViewFormat(format) });
			},
			generateMipmaps() {
				texture.generateMipmaps?.();
			},
			destroy() {
				if (!live.has(managed)) return;
				texture.destroy();
				live.delete(managed);
			}
		};
		live.add(managed);
		return managed;
	}
	function createMediaTexture(options) {
		const { width, height, srgb = false, mipLevelCount } = options;
		const format = options.format ?? "rgba8unorm";
		return wrap(createRaw({
			size: [width, height],
			format,
			...srgb ? { viewFormats: [srgbViewFormat(format)] } : {},
			...mipLevelCount ? { mipLevelCount } : {}
		}, "sampled", "render"), width, height, format, srgb);
	}
	function createDataTexture(options) {
		const { width, height, format, data } = options;
		const managed = wrap(createRaw({
			size: [width, height],
			format
		}, "sampled"), width, height, format, false);
		if (data) managed.write(data);
		return managed;
	}
	function createRenderTexture(options) {
		const format = options.format ?? "rgba16float";
		const { mipLevelCount } = options;
		let width = options.width;
		let height = options.height;
		function build(w, h) {
			return wrap(createRaw({
				size: [w, h],
				format,
				...mipLevelCount ? { mipLevelCount } : {}
			}, "sampled", "render"), w, h, format, false);
		}
		let inner = build(width, height);
		return {
			get texture() {
				return inner.texture;
			},
			get width() {
				return width;
			},
			get height() {
				return height;
			},
			format,
			write(source) {
				inner.write(source);
			},
			unwrap() {
				return inner.unwrap();
			},
			createSrgbView() {
				return inner.createSrgbView();
			},
			generateMipmaps() {
				inner.generateMipmaps();
			},
			resize(w, h) {
				if (w === width && h === height) return;
				inner.destroy();
				width = w;
				height = h;
				inner = build(w, h);
			},
			destroy() {
				inner.destroy();
			}
		};
	}
	function createExternalTextureBinding(options) {
		const { layout, entryKey, getSource, staticEntries = {} } = options;
		let current = null;
		return {
			update() {
				const source = getSource();
				if (!source) {
					current = null;
					return null;
				}
				let external;
				try {
					external = root.device.importExternalTexture({ source });
				} catch {
					current = null;
					return null;
				}
				current = root.createBindGroup(layout, {
					...staticEntries,
					[entryKey]: external
				});
				return current;
			},
			get current() {
				return current;
			},
			destroy() {
				current = null;
			}
		};
	}
	return {
		samplers,
		createMediaTexture,
		createDataTexture,
		createRenderTexture,
		createExternalTextureBinding,
		get textureCount() {
			return live.size;
		},
		destroy() {
			for (const t of [...live]) t.destroy();
			live.clear();
		}
	};
}
function applySpring(current, velocity, target, smoothing, momentum, dt) {
	if (smoothing === 0 && momentum === 0) return [target, 0];
	const stiffness = 200 * Math.pow(.01, smoothing);
	const damping = 2 * Math.sqrt(stiffness) * (1 - momentum * .85);
	const newVelocity = velocity + (stiffness * (target - current) - damping * velocity) * dt;
	return [current + newVelocity * dt, newVelocity];
}
function applyBounceEase(t) {
	const n1 = 7.5625, d1 = 2.75;
	if (t < 1 / d1) return n1 * t * t;
	if (t < 2 / d1) {
		t -= 1.5 / d1;
		return n1 * t * t + .75;
	}
	if (t < 2.5 / d1) {
		t -= 2.25 / d1;
		return n1 * t * t + .9375;
	}
	t -= 2.625 / d1;
	return n1 * t * t + .984375;
}
function applyEasing(t, easing) {
	switch (easing) {
		case "linear": return t;
		case "quad": return t < .5 ? 2 * t * t : 1 - Math.pow(-2 * t + 2, 2) / 2;
		case "expo":
			if (t === 0) return 0;
			if (t === 1) return 1;
			return t < .5 ? Math.pow(2, 20 * t - 10) / 2 : (2 - Math.pow(2, -20 * t + 10)) / 2;
		case "bounce": return applyBounceEase(t);
		case "sine":
		default: return (1 - Math.cos(Math.PI * t)) / 2;
	}
}
function clampToTextureCap(w, h, env) {
	const pr = env.pixelRatio || 1;
	const gpuCssCap = Math.max(1, Math.floor(env.maxTextureDim / pr) - 1);
	const capW = Math.min(w, env.viewportWidth, gpuCssCap);
	const capH = Math.min(h, env.viewportHeight, gpuCssCap);
	const scale = Math.min(capW / w, capH / h, 1);
	return {
		width: Math.max(1, Math.round(w * scale)),
		height: Math.max(1, Math.round(h * scale))
	};
}
function deriveElapsedTime(sharedTimeOrigin, prevElapsed, deltaTime, now) {
	if (sharedTimeOrigin != null) return (now - sharedTimeOrigin) / 1e3;
	return prevElapsed + deltaTime;
}
function frameGate(now, state) {
	const { lastRenderTime, isVisible, forceFullFrameRate } = state;
	const deltaTime = lastRenderTime > 0 ? (now - lastRenderTime) / 1e3 : .016;
	if (lastRenderTime > 0) {
		const sinceLast = now - lastRenderTime;
		if (!isVisible && !forceFullFrameRate && sinceLast < 1e3) return {
			render: false,
			deltaTime
		};
		if ((isVisible || forceFullFrameRate) && sinceLast < 15.666666666666668) return {
			render: false,
			deltaTime
		};
	}
	return {
		render: true,
		deltaTime: Math.min(deltaTime, .1)
	};
}
async function awaitGpuIdle(root) {
	try {
		const queue = root.device?.queue;
		if (queue && typeof queue.onSubmittedWorkDone === "function") {
			await queue.onSubmittedWorkDone();
			return;
		}
	} catch {}
	await new Promise((resolve) => setTimeout(resolve, 16));
}
function runFrameSequence(seq) {
	seq.updateDrivers();
	if (!seq.ensureComposition()) return;
	seq.beforeRender();
	seq.flush();
	seq.render();
	seq.markReady();
	seq.afterRender();
}
function createFrameLoop(tick) {
	let rafId = null;
	const raf = (cb) => typeof requestAnimationFrame === "function" ? requestAnimationFrame(cb) : setTimeout(() => cb(performance.now()), 16);
	const cancel = (id) => {
		if (typeof cancelAnimationFrame === "function") cancelAnimationFrame(id);
		else clearTimeout(id);
	};
	const step = () => {
		tick();
		if (rafId !== null) rafId = raf(step);
	};
	return {
		start() {
			if (rafId !== null) return;
			rafId = raf(step);
		},
		stop() {
			if (rafId !== null) {
				cancel(rafId);
				rafId = null;
			}
		},
		get running() {
			return rafId !== null;
		}
	};
}
var PerformanceTracker = class {
	frameTimesMs = [];
	frameTimesIndex = 0;
	frameTimesCount = 0;
	maxSamples = 60;
	targetFrameTime = 16.67;
	jankFrameCount = 0;
	totalFrameCount = 0;
	memorySnapshots = [];
	memorySnapshotIndex = 0;
	maxMemorySnapshots = 300;
	nodeCount = 0;
	rttNodeCount = 0;
	lastCpuTime = null;
	lastGpuTime = null;
	isRendering = false;
	lastFrameTimestamp = 0;
	frameIntervals = [];
	frameIntervalsIndex = 0;
	frameIntervalsCount = 0;
	pushToCircularBuffer(buffer, value, index, _count, maxSize) {
		if (buffer.length < maxSize) {
			buffer.push(value);
			return {
				index: buffer.length,
				count: buffer.length
			};
		}
		buffer[index % maxSize] = value;
		return {
			index: index + 1,
			count: maxSize
		};
	}
	getBufferValues(buffer, count) {
		return count <= buffer.length ? buffer.slice(0, count) : buffer;
	}
	recordFrame(frameTimeMs) {
		const result = this.pushToCircularBuffer(this.frameTimesMs, frameTimeMs, this.frameTimesIndex, this.frameTimesCount, this.maxSamples);
		this.frameTimesIndex = result.index;
		this.frameTimesCount = result.count;
		this.totalFrameCount++;
		if (frameTimeMs > this.targetFrameTime) this.jankFrameCount++;
		const now = performance.now();
		if (this.lastFrameTimestamp > 0) {
			const interval = now - this.lastFrameTimestamp;
			const intervalResult = this.pushToCircularBuffer(this.frameIntervals, interval, this.frameIntervalsIndex, this.frameIntervalsCount, this.maxSamples);
			this.frameIntervalsIndex = intervalResult.index;
			this.frameIntervalsCount = intervalResult.count;
		}
		this.lastFrameTimestamp = now;
		this.recordMemorySnapshot();
	}
	recordCpuTime(timeMs) {
		this.lastCpuTime = timeMs;
	}
	recordGpuTime(timeMs) {
		this.lastGpuTime = timeMs;
	}
	updateNodeCounts(nodeCount, rttNodeCount) {
		this.nodeCount = nodeCount;
		this.rttNodeCount = rttNodeCount;
	}
	setRendering(rendering) {
		this.isRendering = rendering;
	}
	recordMemorySnapshot() {
		if (!performance.memory) return;
		const now = performance.now();
		const bytes = performance.memory.usedJSHeapSize;
		if (this.memorySnapshots.length < this.maxMemorySnapshots) this.memorySnapshots.push({
			time: now,
			bytes
		});
		else {
			const entry = this.memorySnapshots[this.memorySnapshotIndex];
			entry.time = now;
			entry.bytes = bytes;
			this.memorySnapshotIndex = (this.memorySnapshotIndex + 1) % this.maxMemorySnapshots;
		}
	}
	calculateMemoryGrowthRate() {
		if (this.memorySnapshots.length < 2) return null;
		const len = this.memorySnapshots.length;
		const oldestIdx = len < this.maxMemorySnapshots ? 0 : this.memorySnapshotIndex;
		const newestIdx = len < this.maxMemorySnapshots ? len - 1 : (this.memorySnapshotIndex - 1 + len) % len;
		const oldest = this.memorySnapshots[oldestIdx];
		const newest = this.memorySnapshots[newestIdx];
		const timeDeltaSeconds = (newest.time - oldest.time) / 1e3;
		if (timeDeltaSeconds <= 0) return null;
		return (newest.bytes - oldest.bytes) / (1024 * 1024) / timeDeltaSeconds;
	}
	calculateStdDev(values) {
		if (values.length === 0) return 0;
		const mean = values.reduce((a, b) => a + b, 0) / values.length;
		let sumSquaredDiff = 0;
		for (let i = 0; i < values.length; i++) {
			const diff = values[i] - mean;
			sumSquaredDiff += diff * diff;
		}
		return Math.sqrt(sumSquaredDiff / values.length);
	}
	sortBuffer = [];
	calculateP99(values) {
		const len = values.length;
		if (len === 0) return 0;
		while (this.sortBuffer.length < len) this.sortBuffer.push(0);
		for (let i = 0; i < len; i++) this.sortBuffer[i] = values[i];
		for (let i = 1; i < len; i++) {
			const current = this.sortBuffer[i];
			let j = i - 1;
			while (j >= 0 && this.sortBuffer[j] > current) {
				this.sortBuffer[j + 1] = this.sortBuffer[j];
				j--;
			}
			this.sortBuffer[j + 1] = current;
		}
		const index = Math.floor(len * .99);
		return this.sortBuffer[Math.min(index, len - 1)];
	}
	calculateComplexityScore() {
		return this.nodeCount + this.rttNodeCount * 10;
	}
	calculateIntensityScore() {
		const frameTimes = this.getBufferValues(this.frameTimesMs, this.frameTimesCount);
		const frameCount = frameTimes.length;
		if (frameCount === 0) return {
			score: 0,
			label: "N/A"
		};
		const avgFrameTime = frameTimes.reduce((a, b) => a + b, 0) / frameCount;
		const frameTimeScore = Math.min(avgFrameTime / 16.67 * 100, 100);
		const complexity = this.calculateComplexityScore();
		const complexityScore = Math.min(complexity / 100 * 100, 100);
		let gpuTimeScore = 0;
		if (this.lastGpuTime !== null) gpuTimeScore = Math.min(this.lastGpuTime / 16.67 * 100, 100);
		const weights = this.lastGpuTime !== null ? {
			frame: .4,
			gpu: .4,
			complexity: .2
		} : {
			frame: .7,
			gpu: 0,
			complexity: .3
		};
		const finalScore = Math.round(frameTimeScore * weights.frame + gpuTimeScore * weights.gpu + complexityScore * weights.complexity);
		let label;
		if (finalScore <= 20) label = "Very Light";
		else if (finalScore <= 40) label = "Light";
		else if (finalScore <= 60) label = "Medium";
		else if (finalScore <= 80) label = "Heavy";
		else label = "Very Heavy";
		return {
			score: Math.min(finalScore, 100),
			label
		};
	}
	getStats(rendererInfo) {
		const frameTimes = this.getBufferValues(this.frameTimesMs, this.frameTimesCount);
		const intervals = this.getBufferValues(this.frameIntervals, this.frameIntervalsCount);
		const frameCount = frameTimes.length;
		const fps = intervals.length > 0 ? 1e3 / (intervals.reduce((a, b) => a + b, 0) / intervals.length) : 0;
		const avgFrameTime = frameCount > 0 ? frameTimes.reduce((a, b) => a + b, 0) / frameCount : 0;
		const minFrameTime = frameCount > 0 ? Math.min(...frameTimes) : 0;
		const maxFrameTime = frameCount > 0 ? Math.max(...frameTimes) : 0;
		const p99FrameTime = this.calculateP99(frameTimes);
		const stdDevFrameTime = this.calculateStdDev(frameTimes);
		const jankPercent = this.totalFrameCount > 0 ? this.jankFrameCount / this.totalFrameCount * 100 : 0;
		const memory = performance.memory;
		const memoryUsedMB = memory ? memory.usedJSHeapSize / (1024 * 1024) : null;
		const memoryGrowthRate = this.calculateMemoryGrowthRate();
		const budgetUsed = avgFrameTime > 0 ? avgFrameTime / this.targetFrameTime * 100 : 0;
		const drawCalls = rendererInfo?.render?.calls ?? null;
		const shaderPrograms = rendererInfo?.programs?.length ?? null;
		const textureCount = rendererInfo?.memory?.textures ?? null;
		const intensity = this.calculateIntensityScore();
		return {
			fps: Math.round(fps),
			avgFrameTime: Math.round(avgFrameTime * 100) / 100,
			minFrameTime: Math.round(minFrameTime * 100) / 100,
			maxFrameTime: Math.round(maxFrameTime * 100) / 100,
			p99FrameTime: Math.round(p99FrameTime * 100) / 100,
			stdDevFrameTime: Math.round(stdDevFrameTime * 100) / 100,
			jankCount: this.jankFrameCount,
			jankPercent: Math.round(jankPercent * 10) / 10,
			nodeCount: this.nodeCount,
			rttNodeCount: this.rttNodeCount,
			complexityScore: this.calculateComplexityScore(),
			memoryUsedMB: memoryUsedMB !== null ? Math.round(memoryUsedMB * 100) / 100 : null,
			memoryGrowthRate: memoryGrowthRate !== null ? Math.round(memoryGrowthRate * 1e3) / 1e3 : null,
			cpuTime: this.lastCpuTime !== null ? Math.round(this.lastCpuTime * 100) / 100 : null,
			gpuTime: this.lastGpuTime !== null ? Math.round(this.lastGpuTime * 100) / 100 : null,
			budgetUsed: Math.round(budgetUsed * 10) / 10,
			drawCalls,
			shaderPrograms,
			textureCount,
			intensityScore: intensity.score,
			intensityLabel: intensity.label,
			isRendering: this.isRendering
		};
	}
	reset() {
		this.frameTimesMs.length = 0;
		this.frameTimesIndex = 0;
		this.frameTimesCount = 0;
		this.frameIntervals.length = 0;
		this.frameIntervalsIndex = 0;
		this.frameIntervalsCount = 0;
		this.lastFrameTimestamp = 0;
		this.jankFrameCount = 0;
		this.totalFrameCount = 0;
		this.memorySnapshots.length = 0;
		this.memorySnapshotIndex = 0;
		this.sortBuffer.length = 0;
		this.lastCpuTime = null;
		this.lastGpuTime = null;
	}
};
var globalEventManager = (() => {
	let instanceCount = 0;
	const callbacks = /* @__PURE__ */ new Set();
	const fan = (key, arg) => {
		callbacks.forEach((cb) => {
			try {
				cb[key](arg);
			} catch (error) {
				console.error("[Shaders] Error in pointer handler:", error);
			}
		});
	};
	const onMove = (e) => fan("onMouseMove", e);
	const onTouch = (e) => fan("onTouchMove", e);
	const onUp = () => fan("onMouseUp");
	const onEnd = () => fan("onTouchEnd");
	return { register(handlers) {
		callbacks.add(handlers);
		instanceCount++;
		if (instanceCount === 1 && typeof window !== "undefined") {
			window.addEventListener("mousemove", onMove);
			window.addEventListener("touchmove", onTouch);
			window.addEventListener("mouseup", onUp);
			window.addEventListener("touchend", onEnd);
		}
		return () => {
			callbacks.delete(handlers);
			instanceCount--;
			if (instanceCount === 0 && typeof window !== "undefined") {
				window.removeEventListener("mousemove", onMove);
				window.removeEventListener("touchmove", onTouch);
				window.removeEventListener("mouseup", onUp);
				window.removeEventListener("touchend", onEnd);
			}
		};
	} };
})();
function shapeConfigUsesMouse(raw) {
	if (!raw) return false;
	let cfg = raw;
	if (typeof raw === "string") try {
		cfg = JSON.parse(raw);
	} catch {
		return false;
	}
	if (!cfg || typeof cfg !== "object") return false;
	for (const key in cfg) {
		const v = cfg[key];
		if (v && typeof v === "object" && v.type === "mouse") return true;
	}
	return false;
}
function shaderRendererGPU() {
	let rootCtx = null;
	let root = null;
	let context = null;
	let textureManager = null;
	let dispatcher = null;
	let canvasElement = null;
	let isInitialized = false;
	let isInitializing = false;
	let isRendererReady = false;
	let initializationAbortController = null;
	const nodes = /* @__PURE__ */ new Map();
	let rootId = null;
	const parentToChildren = /* @__PURE__ */ new Map();
	let pendingRegistrationQueue = [];
	const pipelineCache = createPipelineCache({
		maxSize: 4,
		dispose: (bc) => disposeComposition(bc)
	});
	const liveCompositions = /* @__PURE__ */ new Set();
	let boundComposition = null;
	let desiredHash = null;
	let structuralDirty = true;
	let currentWidth = 0;
	let currentHeight = 0;
	let logicalWidth = 0;
	let logicalHeight = 0;
	let pixelRatio = 1;
	let cachedMaxTextureDim = 8192;
	let lastRequestedWidth = 0;
	let lastRequestedHeight = 0;
	let hasInitialDimensions = false;
	const pxResWidth = () => logicalWidth || currentWidth || 1;
	const pxResHeight = () => logicalHeight || currentHeight || 1;
	let shouldAnimate = true;
	let isVisible = false;
	let forceFullFrameRate = false;
	let lastRenderTime = 0;
	let globalElapsedTime = 0;
	let sharedTimeOrigin = null;
	let pendingRenderRAF = null;
	let pointerX = .5;
	let pointerY = .5;
	let pointerActive = false;
	let cachedCanvasRect = null;
	let mouseListenersNeeded = false;
	let globalEventUnregister = null;
	let resizeObserver = null;
	let intersectionObserver = null;
	let unloadHandler = null;
	let windowResizeHandler = null;
	let pendingResize = null;
	let isResizeScheduled = false;
	let rendererColorSpace = "p3-linear";
	let rendererToneMapping = "linear";
	let enablePerformanceTracking = false;
	let setColorSpaceModeFn = null;
	let onReadyCallback = null;
	let hasEmittedReady = false;
	let firstRenderPending = false;
	let onDeviceLostCallback = null;
	let deviceLostFired = false;
	const performanceTracker = new PerformanceTracker();
	const frameParams = {
		deltaTime: 0,
		pointer: {
			x: .5,
			y: .5
		},
		pointerActive: false,
		dimensions: {
			width: 0,
			height: 0
		}
	};
	const frameLoop = createFrameLoop(() => renderFrame());
	const sanitizeId = (id) => id.replace(/[^a-zA-Z0-9_]/g, "_");
	const findChildNodes = (parentId) => {
		const childIds = parentToChildren.get(parentId);
		if (!childIds) return [];
		const out = [];
		for (const cid of childIds) {
			const info = nodes.get(cid);
			if (info) out.push(info);
		}
		return out;
	};
	const findNodeByCustomId = (customId) => {
		const sanitized = sanitizeId(customId);
		for (const [nodeId, info] of nodes) if (info.metadata.id === sanitized) return nodeId;
		return null;
	};
	const computeDimExtent = (node, origin) => {
		if (origin === "center") return void 0;
		const decl = node.boundingBoxDeclaration;
		if (!decl) return void 0;
		const snap = {};
		for (const [k, u] of Object.entries(node.uniforms)) snap[k] = u._rawDimensional !== void 0 ? u._rawDimensional : u.value;
		return boxHalfExtentsUV(decl, snap, pxResWidth(), pxResHeight());
	};
	const reresolveDimensionalUniforms = (node) => {
		const decl = node.boundingBoxDeclaration;
		const extra = node.extraDimProps;
		const plan = buildDimensionalPlan(decl, extra);
		if (!plan) return;
		const origin = node.uniforms["origin"]?.value ?? "center";
		const he = computeDimExtent(node, origin);
		const names = [...plan.positionProps, ...plan.sizeProps.keys()];
		for (const name of names) {
			const u = node.uniforms[name];
			if (!u || u._rawDimensional === void 0) continue;
			const resolved = resolveDimensionalProp(decl, name, u._rawDimensional, origin, pxResWidth(), pxResHeight(), extra, he);
			u.value = resolved;
			writePropHandle(node, name, resolved);
		}
	};
	const resolveDynamicBound = (node, propName, bound) => {
		if (typeof bound === "number") return bound;
		const origin = node.uniforms["origin"]?.value ?? "center";
		const resolved = resolveDimensionalProp(node.boundingBoxDeclaration, propName, bound, origin, pxResWidth(), pxResHeight(), node.extraDimProps);
		return typeof resolved === "number" ? resolved : bound?.value ?? 0;
	};
	const aspectRatio = () => {
		const { width, height } = bufferSize();
		return width / Math.max(1, height);
	};
	const transformFieldValues = (node) => {
		const t = node.metadata.transform;
		return {
			offsetX: t?.offsetX ?? 0,
			offsetY: t?.offsetY ?? 0,
			rotation: t?.rotation ?? 0,
			scale: t?.scale ?? 1,
			anchorX: t?.anchorX ?? .5,
			anchorY: t?.anchorY ?? .5,
			aspectRatio: aspectRatio()
		};
	};
	const bboxFieldValues = (node) => {
		const bbox = node.metadata.boundingBox;
		if (!bbox) return null;
		const p = boundingBoxToUVParams(bbox, pxResWidth(), pxResHeight());
		return {
			centerX: p.centerX,
			centerY: p.centerY,
			halfWidth: p.halfWidthUV,
			halfHeight: p.halfHeightUV,
			cornerRadius: p.cornerRadiusUV,
			rotation: p.rotation,
			aspectRatio: aspectRatio()
		};
	};
	const bboxActive = (node) => !!node.metadata.boundingBox && !!node.boundingBoxDeclaration;
	const computeNodeBoundsUV = (node) => {
		const decl = node.boundingBoxDeclaration;
		if (!decl || !node.uniforms) return null;
		if (!decl.computeBounds && !decl.propBindings) return null;
		const snap = {};
		for (const [k, u] of Object.entries(node.uniforms)) snap[k] = u._rawDimensional !== void 0 ? u._rawDimensional : u.value;
		const he = boxHalfExtentsUV(decl, snap, pxResWidth(), pxResHeight());
		if (he.hwx <= 0 && he.hhy <= 0) return null;
		let cx = .5;
		let cy = .5;
		const posProp = decl.propBindings?.x?.prop ?? decl.propBindings?.y?.prop;
		const posValue = posProp ? node.uniforms[posProp]?.value : void 0;
		if (posValue && typeof posValue.x === "number" && typeof posValue.y === "number") {
			cx = posValue.x;
			cy = posValue.y;
		}
		return {
			centerX: cx,
			centerY: cy,
			halfWidth: he.hwx,
			halfHeight: he.hhy
		};
	};
	const childBoundsValues = (node) => {
		if (!node.definition.wantsBoundsParams) return null;
		const children = findChildNodes(node.id).sort((a, b) => a.metadata.renderOrder - b.metadata.renderOrder);
		for (const child of children) {
			if (child.metadata.visible === false) continue;
			if (bboxActive(child)) return null;
			const b = computeNodeBoundsUV(child);
			if (b) return b;
		}
		return null;
	};
	const updateChildBounds = () => {
		for (const node of nodes.values()) {
			if (!node.definition.wantsBoundsParams) continue;
			const b = childBoundsValues(node);
			if (!b) continue;
			writeSyntheticScalar(node, "_childBounds_centerX", b.centerX);
			writeSyntheticScalar(node, "_childBounds_centerY", b.centerY);
			writeSyntheticScalar(node, "_childBounds_halfWidth", b.halfWidth);
			writeSyntheticScalar(node, "_childBounds_halfHeight", b.halfHeight);
		}
	};
	const transformActive = (node) => needsTransformation(node.metadata.transform);
	const ensureMapValues = (node, prop, driver) => {
		const mv = node.mapValues[prop] ?? {
			inputMin: 0,
			inputMax: 1,
			outputMin: 0,
			outputMax: 1,
			curve: 0
		};
		if (driver.type === "map") {
			const c = driver;
			mv.inputMin = c.inputMin;
			mv.inputMax = c.inputMax;
			mv.outputMin = resolveDynamicBound(node, prop, c.outputMin);
			mv.outputMax = resolveDynamicBound(node, prop, c.outputMax);
			mv.curve = c.curve ?? 0;
		} else if (driver.type === "mouse") {
			const c = driver;
			mv.outputMin = resolveDynamicBound(node, prop, c.outputMin);
			mv.outputMax = resolveDynamicBound(node, prop, c.outputMax);
			mv.curve = c.curve ?? 0;
		} else if (driver.type === "auto-animate") {
			const c = driver;
			mv.outputMin = resolveDynamicBound(node, prop, c.outputMin);
			mv.outputMax = resolveDynamicBound(node, prop, c.outputMax);
		}
		node.mapValues[prop] = mv;
		return mv;
	};
	const ensureMouseState = (node, prop, driver) => {
		let s = node.mouseDriverState[prop];
		if (s) return s;
		if (driver.type === "mouse-position") {
			const trackX = driver.x === void 0 || driver.x === "mouse";
			const trackY = driver.y === void 0 || driver.y === "mouse";
			s = {
				currentX: trackX ? pointerX : driver.x,
				currentY: trackY ? pointerY : driver.y,
				velocityX: 0,
				velocityY: 0
			};
		} else s = {
			currentX: Math.max(.001, driver.axis === "x" ? pointerX : pointerY),
			currentY: 0,
			velocityX: 0,
			velocityY: 0
		};
		node.mouseDriverState[prop] = s;
		return s;
	};
	const animatedTimeKey = (node) => {
		const at = node.definition.animatedTime;
		return (at ? node.uniforms[at.speed] : void 0) ?? node;
	};
	const extraAnimatedTimeKey = (node, speedProp) => node.uniforms[speedProp] ?? node;
	const buildFieldInits = (node) => {
		const inits = [];
		for (const [propName, u] of Object.entries(node.uniforms)) inits.push({
			name: propName,
			initial: u.value,
			transform: u.transform,
			cpu: u.cpu,
			schema: u.schema
		});
		inits.push({
			name: "_opacity",
			schema: d.f32,
			initial: node.metadata.opacity ?? 1
		});
		if (node.definition.animatedTime) inits.push({
			name: "_animTime",
			schema: d.f32,
			initial: getAnimatedTimeState(animatedTimeKey(node)).value
		});
		if (node.definition.extraAnimatedTimes) for (const [key, speedProp] of Object.entries(node.definition.extraAnimatedTimes)) inits.push({
			name: `_animTime_${key}`,
			schema: d.f32,
			initial: getAnimatedTimeState(extraAnimatedTimeKey(node, speedProp)).value
		});
		const extraFields = node.definition.extraFields;
		if (extraFields) for (const [name, f] of Object.entries(extraFields)) inits.push({
			name,
			schema: f.schema,
			initial: f.initial
		});
		if (transformActive(node)) {
			const tf = transformFieldValues(node);
			for (const [k, v] of Object.entries(tf)) inits.push({
				name: `_xform_${k}`,
				schema: d.f32,
				initial: v
			});
		}
		if (bboxActive(node)) {
			const bb = bboxFieldValues(node);
			if (bb) for (const [k, v] of Object.entries(bb)) inits.push({
				name: `_bbox_${k}`,
				schema: d.f32,
				initial: v
			});
		}
		if (node.definition.wantsBoundsParams) {
			const cb = childBoundsValues(node) ?? {
				centerX: .5,
				centerY: .5,
				halfWidth: .5,
				halfHeight: .5
			};
			for (const [k, v] of Object.entries(cb)) inits.push({
				name: `_childBounds_${k}`,
				schema: d.f32,
				initial: v
			});
		}
		const maps = node.metadata.maps;
		if (maps) for (const [prop, driver] of Object.entries(maps)) {
			if (!(prop in node.uniforms)) continue;
			if (driver.type === "map") {
				const mv = ensureMapValues(node, prop, driver);
				inits.push({
					name: `_map_${prop}_inputMin`,
					schema: d.f32,
					initial: mv.inputMin
				});
				inits.push({
					name: `_map_${prop}_inputMax`,
					schema: d.f32,
					initial: mv.inputMax
				});
				inits.push({
					name: `_map_${prop}_outputMin`,
					schema: d.f32,
					initial: mv.outputMin
				});
				inits.push({
					name: `_map_${prop}_outputMax`,
					schema: d.f32,
					initial: mv.outputMax
				});
				inits.push({
					name: `_map_${prop}_curve`,
					schema: d.f32,
					initial: mv.curve
				});
			} else if (driver.type === "mouse-position") {
				const s = ensureMouseState(node, prop, driver);
				inits.push({
					name: `_smoothed_${prop}`,
					schema: d.vec2f,
					initial: [s.currentX, 1 - s.currentY]
				});
			} else if (driver.type === "mouse") {
				ensureMouseState(node, prop, driver);
				const mv = ensureMapValues(node, prop, driver);
				inits.push({
					name: `_smoothed_${prop}`,
					schema: d.f32,
					initial: mv.outputMin
				});
			} else if (driver.type === "auto-animate") {
				if (!node.autoAnimateState[prop]) node.autoAnimateState[prop] = {};
				const mv = ensureMapValues(node, prop, driver);
				inits.push({
					name: `_smoothed_${prop}`,
					schema: d.f32,
					initial: mv.outputMin
				});
			}
		}
		return inits;
	};
	const writePropHandle = (node, propName, value) => {
		const h = node._liveHandles?.[propName];
		if (!h) return;
		setColorSpaceModeFn?.(rendererColorSpace);
		updateFieldValue(h, value);
	};
	const writeSyntheticScalar = (node, field, value) => {
		const h = node._liveHandles?.[field];
		if (h instanceof FieldHandle) h.value = value;
	};
	const writeExtraFieldValue = (node, field, value) => {
		const h = node._liveHandles?.[field];
		if (h instanceof FieldHandle) h.value = value;
	};
	const nodeColorSpaceMode = (node) => {
		for (const u of Object.values(node.uniforms)) if (u.transform === transformColorSpace) {
			const raw = u.value;
			if (typeof raw === "string") return transformColorSpace(raw);
			if (typeof raw === "number") return raw;
		}
		return 0;
	};
	const repackColorStops = (node) => {
		if (!node.uniforms.convertedColorsArray) return;
		let stops = null;
		for (const u of Object.values(node.uniforms)) if (u.transform === colorStopsTransform) {
			stops = u.value ?? null;
			break;
		}
		const packed = packStops(stops, colorToRGBA);
		const converted = packConvertedStops(packed.colors, packed.stopCount, nodeColorSpaceMode(node));
		const write = (name, value) => {
			const u = node.uniforms[name];
			if (u) u.value = value;
			writePropHandle(node, name, value);
		};
		write("colorsArray", packed.colors);
		write("positionsArray", packed.positions);
		write("convertedColorsArray", converted);
		write("stopCount", packed.stopCount);
	};
	const writeNodeSynthetic = (node) => {
		writeSyntheticScalar(node, "_opacity", node.metadata.opacity ?? 1);
		if (node.definition.animatedTime) writeSyntheticScalar(node, "_animTime", getAnimatedTimeState(animatedTimeKey(node)).value);
		if (node.definition.extraAnimatedTimes) for (const [key, speedProp] of Object.entries(node.definition.extraAnimatedTimes)) writeSyntheticScalar(node, `_animTime_${key}`, getAnimatedTimeState(extraAnimatedTimeKey(node, speedProp)).value);
		if (transformActive(node)) for (const [k, v] of Object.entries(transformFieldValues(node))) writeSyntheticScalar(node, `_xform_${k}`, v);
		if (bboxActive(node)) {
			const bb = bboxFieldValues(node);
			if (bb) for (const [k, v] of Object.entries(bb)) writeSyntheticScalar(node, `_bbox_${k}`, v);
		}
		const maps = node.metadata.maps;
		if (maps) {
			for (const [prop, driver] of Object.entries(maps)) if (driver.type === "map" && node.mapValues[prop]) {
				const mv = node.mapValues[prop];
				writeSyntheticScalar(node, `_map_${prop}_inputMin`, mv.inputMin);
				writeSyntheticScalar(node, `_map_${prop}_inputMax`, mv.inputMax);
				writeSyntheticScalar(node, `_map_${prop}_outputMin`, mv.outputMin);
				writeSyntheticScalar(node, `_map_${prop}_outputMax`, mv.outputMax);
				writeSyntheticScalar(node, `_map_${prop}_curve`, mv.curve);
			}
		}
	};
	const propHandleView = (node, handles) => {
		const out = {};
		for (const propName of Object.keys(node.uniforms)) {
			const h = handles[propName];
			if (h) out[propName] = h;
		}
		return out;
	};
	const sotHandleView = (node) => {
		const out = {};
		for (const [propName, u] of Object.entries(node.uniforms)) out[propName] = {
			accessorPath: "",
			cpu: !!u.cpu,
			value: u.value
		};
		return out;
	};
	const makeRegistryNode = (node, handles) => ({
		id: node.id,
		customId: node.metadata.id,
		componentName: node.componentName,
		parentId: node.parentId,
		definition: node.definition,
		metadata: node.metadata,
		handles,
		domCanvas: node.domCanvas
	});
	const buildRegistryView = (store, handlesById) => {
		const cache = /* @__PURE__ */ new Map();
		const build = (node) => {
			return makeRegistryNode(node, store && handlesById ? propHandleView(node, handlesById.get(node.id) ?? {}) : sotHandleView(node));
		};
		const getNode = (id) => {
			const hit = cache.get(id);
			if (hit) return hit;
			const node = nodes.get(id);
			if (!node) return void 0;
			const rn = build(node);
			cache.set(id, rn);
			return rn;
		};
		return {
			rootId,
			getNode,
			getChildren: (pid) => findChildNodes(pid).map((n) => getNode(n.id)),
			resolveCustomId: (cid) => findNodeByCustomId(cid),
			store: {
				gpuAccessor: (h, layoutVar) => store ? store.gpuAccessor(h, layoutVar) : "",
				layout: store?.layout
			}
		};
	};
	const composeOptions = () => ({
		toneMapping: rendererToneMapping,
		flipY: false,
		premultiplyAlpha: true,
		canvas: canvasElement ?? void 0,
		dimensions: bufferSize(),
		gpu: root ? {
			device: root.device,
			root
		} : void 0,
		writeExtraField: (nodeId, name, value) => {
			const node = nodes.get(nodeId);
			if (node) writeExtraFieldValue(node, name, value);
		},
		createMediaTexture: (opts) => {
			if (!textureManager) throw new Error("[gpu] createMediaTexture before init");
			return textureManager.createMediaTexture(opts);
		},
		createDataTexture: (opts) => {
			if (!textureManager) throw new Error("[gpu] createDataTexture before init");
			return textureManager.createDataTexture(opts);
		},
		resolveMapWindow: (nodeId, prop) => {
			const node = nodes.get(nodeId);
			const driver = node?.metadata.maps?.[prop];
			if (!node || !driver || driver.type !== "map") return void 0;
			const mv = ensureMapValues(node, prop, driver);
			return {
				inputMin: mv.inputMin,
				inputMax: mv.inputMax,
				outputMin: mv.outputMin,
				outputMax: mv.outputMax,
				curve: mv.curve
			};
		},
		resolveDriverCpuValue: (nodeId, prop) => {
			const node = nodes.get(nodeId);
			const driver = node?.metadata.maps?.[prop];
			if (!node || !driver) return void 0;
			if (driver.type === "mouse-position") {
				const s = node.mouseDriverState[prop];
				if (!s || s.liveX === void 0 || s.liveY === void 0) return void 0;
				return {
					x: s.liveX,
					y: 1 - s.liveY
				};
			}
			if (driver.type === "mouse" || driver.type === "auto-animate") {
				const h = node._liveHandles?.[`_smoothed_${prop}`];
				return h instanceof FieldHandle ? h.value : void 0;
			}
		}
	});
	const extraHashInputs = () => {
		const parts = [];
		for (const [id, node] of nodes) {
			const m = node.metadata;
			if (node.hasUvRemap || node.acceptsUVContext) parts.push(`aopac:${id}:${(m.opacity ?? 1) < 1}`);
			const sc = node.uniforms.stopCount?.value;
			if (typeof sc === "number") parts.push(`stops:${id}:${sc > 1 ? sc : 0}`);
			for (const [prop, u] of Object.entries(node.uniforms)) {
				if (!u.compileTimeWhen) continue;
				const v = u._lastCompiledValue;
				if (v === null || typeof v === "object") continue;
				parts.push(`ctw:${id}:${prop}:${String(v)}`);
			}
			if (m.maps) {
				const fp = Object.entries(m.maps).sort(([a], [b]) => a < b ? -1 : a > b ? 1 : 0).map(([p, dr]) => {
					const cfg = dr;
					return `${p}=${cfg.type}:${cfg.source ?? ""}:${cfg.channel ?? ""}:${cfg.axis ?? ""}`;
				}).join(",");
				parts.push(`mapfp:${id}:${fp}`);
			}
		}
		return parts;
	};
	const computeStructuralHash = () => {
		if (!rootId) return null;
		return structuralHash([...collectStructuralHashInputs(buildRegistryView(null, null), composeOptions()), ...extraHashInputs()]);
	};
	const bufferSize = () => ({
		width: Math.max(1, Math.round(currentWidth * pixelRatio)),
		height: Math.max(1, Math.round(currentHeight * pixelRatio))
	});
	const mediaResourcesFromIr = (ir) => {
		const out = {};
		for (const t of ir.textures) if (t.kind !== "rtt" && t.resource) out[t.key] = t.resource;
		return out;
	};
	const buildComposition = (hash) => {
		if (!root || !textureManager || !dispatcher) throw new Error("[gpu] buildComposition before init");
		const store = createUniformStore(root, { systemSchema: SystemUniforms });
		const handlesById = /* @__PURE__ */ new Map();
		for (const [id, node] of nodes) handlesById.set(id, store.defineNode(id, buildFieldInits(node)));
		store.defineSystem();
		store.finalize();
		const ir = composeNodeTree(buildRegistryView(store, handlesById), composeOptions());
		const pm = createPassManager(root, {
			textureManager,
			dispatcher
		});
		pm.setComposition(ir, store.bindGroup, bufferSize(), mediaResourcesFromIr(ir));
		const bc = {
			hash,
			store,
			ir,
			passManager: pm,
			handlesById,
			fresh: true
		};
		liveCompositions.add(bc);
		return bc;
	};
	const disposeComposition = (bc) => {
		try {
			for (const cb of bc.ir.onCleanup) cb();
		} catch (e) {
			console.error("[gpu] composition onCleanup error:", e);
		}
		try {
			bc.passManager.dispose();
		} catch (e) {
			console.error("[gpu] passManager dispose error:", e);
		}
		try {
			bc.store.destroy();
		} catch (e) {
			console.error("[gpu] store destroy error:", e);
		}
		liveCompositions.delete(bc);
		if (boundComposition === bc) boundComposition = null;
	};
	const bindComposition = (bc) => {
		let repointed = false;
		for (const [id, handles] of bc.handlesById) {
			const node = nodes.get(id);
			if (node && node._liveHandles !== handles) {
				node._liveHandles = handles;
				repointed = true;
			}
		}
		if (bc === boundComposition && !repointed) return;
		if (!bc.fresh) {
			for (const node of nodes.values()) {
				for (const [propName, u] of Object.entries(node.uniforms)) writePropHandle(node, propName, u.value);
				writeNodeSynthetic(node);
			}
			bc.store.writeAll();
		}
		bc.fresh = false;
		boundComposition = bc;
		if (!hasEmittedReady && bc.ir.composedNodeIds.size > 0) firstRenderPending = true;
	};
	const markStructuralDirty = () => {
		structuralDirty = true;
		requestRender();
	};
	const requestRender = () => {
		if (frameLoop.running) return;
		if (pendingRenderRAF !== null) return;
		if (typeof requestAnimationFrame !== "function") return;
		pendingRenderRAF = requestAnimationFrame(() => {
			pendingRenderRAF = null;
			renderFrame();
		});
	};
	const registerNode = (id, fragmentNodeFunc, parentId, metadata, uniforms = {}, componentDefinition, domCanvas) => {
		if (fragmentNodeFunc === null) {
			const queueIndex = pendingRegistrationQueue.findIndex((item) => item.id === id);
			if (queueIndex !== -1) {
				pendingRegistrationQueue.splice(queueIndex, 1);
				return;
			}
			removeNode(id);
			return;
		}
		if (!componentDefinition) {
			console.error(`[gpu] registerNode("${id}") requires a componentDefinition`);
			return;
		}
		if (!isRendererReady) {
			const existing = pendingRegistrationQueue.findIndex((item) => item.id === id);
			const entry = {
				id,
				definition: componentDefinition,
				parentId,
				metadata,
				uniforms,
				domCanvas
			};
			if (existing !== -1) pendingRegistrationQueue[existing] = entry;
			else pendingRegistrationQueue.push(entry);
			return;
		}
		let defaultRenderOrder = 0;
		if (parentId !== null) {
			const siblings = parentToChildren.get(parentId);
			if (siblings && siblings.size > 0) {
				let maxOrder = 0;
				for (const siblingId of siblings) {
					const sibling = nodes.get(siblingId);
					if (sibling && sibling.metadata.renderOrder > maxOrder) maxOrder = sibling.metadata.renderOrder;
				}
				defaultRenderOrder = maxOrder + 1;
			}
		}
		const nodeMetadata = {
			blendMode: metadata?.blendMode || "normal",
			opacity: metadata?.opacity,
			visible: metadata?.visible === false ? false : true,
			id: metadata?.id ? sanitizeId(metadata.id) : metadata?.id,
			mask: metadata?.mask,
			maps: metadata?.maps,
			renderOrder: metadata?.renderOrder !== void 0 ? metadata.renderOrder : defaultRenderOrder,
			transform: metadata?.transform,
			boundingBox: metadata?.boundingBox
		};
		const decl = componentDefinition.boundingBoxDeclaration;
		const extraDimProps = componentDefinition.props ? buildExtraSizeProps(componentDefinition.props) : void 0;
		const node = {
			id,
			componentName: componentDefinition.name || "Unknown",
			parentId,
			definition: componentDefinition,
			metadata: nodeMetadata,
			uniforms,
			domCanvas,
			requiresRTT: !!componentDefinition.requiresRTT,
			requiresChild: !!componentDefinition.requiresChild,
			blendWithChildren: !!componentDefinition.blendWithChildren,
			acceptsUVContext: !!componentDefinition.acceptsUVContext,
			providesUVContextViaCompute: !!componentDefinition.providesUVContextViaCompute,
			usesPointer: !!componentDefinition.usesPointer,
			hasUvRemap: !!componentDefinition.uvRemap,
			boundingBoxDeclaration: decl,
			wantsBoundsParams: !!componentDefinition.wantsBoundsParams,
			extraDimProps,
			mouseDriverState: {},
			autoAnimateState: {},
			mapValues: {},
			_liveHandles: null
		};
		if (decl || extraDimProps) {
			const plan = buildDimensionalPlan(decl, extraDimProps);
			if (plan) {
				const origin = uniforms["origin"]?.value ?? "center";
				const he = computeDimExtent(node, origin);
				for (const name of [...plan.positionProps, ...plan.sizeProps.keys()]) {
					const u = uniforms[name];
					if (!u) continue;
					u._rawDimensional = u.value;
					u.value = resolveDimensionalProp(decl, name, u.value, origin, pxResWidth(), pxResHeight(), extraDimProps, he);
				}
			}
		}
		for (const u of Object.values(uniforms)) if (u.compileTimeWhen && u._lastCompiledValue === void 0) u._lastCompiledValue = u.value;
		nodes.set(id, node);
		if (parentId === null) rootId = id;
		const childrenSet = parentToChildren.get(parentId) || /* @__PURE__ */ new Set();
		childrenSet.add(id);
		parentToChildren.set(parentId, childrenSet);
		updateMouseListenerRegistration();
		markStructuralDirty();
	};
	const updateUniformValue = (nodeId, uniformName, value) => {
		const node = nodes.get(nodeId);
		if (!node) return;
		const u = node.uniforms[uniformName];
		if (!u) return;
		const decl = node.boundingBoxDeclaration;
		const extra = node.extraDimProps;
		if ((decl || extra) && isDimensionalProp(decl, uniformName, extra)) {
			u._rawDimensional = value;
			const origin = node.uniforms["origin"]?.value ?? "center";
			const he = computeDimExtent(node, origin);
			value = resolveDimensionalProp(decl, uniformName, value, origin, pxResWidth(), pxResHeight(), extra, he);
		}
		u.value = value;
		writePropHandle(node, uniformName, value);
		if (u.transform === colorStopsTransform || u.transform === transformColorSpace) repackColorStops(node);
		if (uniformName === "origin" && decl) reresolveDimensionalUniforms(node);
		else if (decl) {
			if ((node.uniforms["origin"]?.value ?? "center") !== "center") {
				const plan = buildDimensionalPlan(decl, extra);
				const affectsExtent = !!decl.computeBounds || (plan?.sizeProps.has(uniformName) ?? false);
				const isPositionProp = plan?.positionProps.has(uniformName) ?? false;
				if (affectsExtent && !isPositionProp) reresolveDimensionalUniforms(node);
			}
		}
		if (uniformName === "shape") updateMouseListenerRegistration();
		if (u.compileTime) {
			markStructuralDirty();
			return;
		}
		if (u.compileTimeWhen) {
			const prev = u._lastCompiledValue;
			if (u.compileTimeWhen(prev, value)) {
				u._lastCompiledValue = value;
				markStructuralDirty();
				return;
			}
		}
		requestRender();
	};
	const updateNodeMetadata = (nodeId, metadata) => {
		const node = nodes.get(nodeId);
		if (!node) return;
		let needsRecompose = false;
		if (metadata.blendMode !== void 0 && node.metadata.blendMode !== metadata.blendMode) {
			node.metadata.blendMode = metadata.blendMode;
			needsRecompose = true;
		}
		if (metadata.opacity !== void 0 && node.metadata.opacity !== metadata.opacity) {
			const oldOpacity = node.metadata.opacity ?? 1;
			const newOpacity = metadata.opacity;
			node.metadata.opacity = newOpacity;
			writeSyntheticScalar(node, "_opacity", newOpacity);
			if ((node.hasUvRemap || node.acceptsUVContext) && oldOpacity >= 1 !== newOpacity >= 1) needsRecompose = true;
		}
		if (metadata.renderOrder !== void 0 && node.metadata.renderOrder !== metadata.renderOrder) {
			node.metadata.renderOrder = metadata.renderOrder;
			needsRecompose = true;
		}
		if ("visible" in metadata) {
			const newVisible = metadata.visible === false ? false : true;
			if (node.metadata.visible !== newVisible) {
				node.metadata.visible = newVisible;
				needsRecompose = true;
			}
		}
		if (metadata.id !== void 0) {
			const sanitized = sanitizeId(metadata.id);
			if (node.metadata.id !== sanitized) {
				node.metadata.id = sanitized;
				needsRecompose = true;
			}
		}
		if ("mask" in metadata && (node.metadata.mask?.source !== metadata.mask?.source || node.metadata.mask?.type !== metadata.mask?.type)) {
			node.metadata.mask = metadata.mask;
			needsRecompose = true;
		}
		if ("maps" in metadata && JSON.stringify(node.metadata.maps) !== JSON.stringify(metadata.maps)) {
			const oldMaps = node.metadata.maps;
			const newMaps = metadata.maps;
			node.metadata.maps = newMaps;
			updateMouseListenerRegistration();
			if ((() => {
				if (!oldMaps || !newMaps) return true;
				const oldKeys = Object.keys(oldMaps);
				const newKeys = Object.keys(newMaps);
				if (oldKeys.length !== newKeys.length || !oldKeys.every((k) => k in newMaps)) return true;
				for (const key of newKeys) {
					const oldCfg = oldMaps[key];
					const newCfg = newMaps[key];
					if (oldCfg.type !== newCfg.type) return true;
					if (newCfg.type === "map") {
						const oc = oldCfg;
						const nc = newCfg;
						if (oc.source !== nc.source || oc.channel !== nc.channel) return true;
					} else if (newCfg.type === "mouse") {
						if (oldCfg.axis !== newCfg.axis) return true;
					}
				}
				return false;
			})()) {
				if (oldMaps && newMaps) {
					for (const key of Object.keys(oldMaps)) if (key in newMaps && oldMaps[key].type !== newMaps[key].type) {
						delete node.mouseDriverState[key];
						delete node.autoAnimateState[key];
						delete node.mapValues[key];
					}
				}
				needsRecompose = true;
			} else if (newMaps) for (const [key, driver] of Object.entries(newMaps)) {
				const mv = ensureMapValues(node, key, driver);
				if (driver.type === "map") {
					writeSyntheticScalar(node, `_map_${key}_inputMin`, mv.inputMin);
					writeSyntheticScalar(node, `_map_${key}_inputMax`, mv.inputMax);
					writeSyntheticScalar(node, `_map_${key}_outputMin`, mv.outputMin);
					writeSyntheticScalar(node, `_map_${key}_outputMax`, mv.outputMax);
					writeSyntheticScalar(node, `_map_${key}_curve`, mv.curve);
				}
			}
		}
		if (metadata.transform !== void 0) {
			const prev = node.metadata.transform;
			const merged = {
				offsetX: metadata.transform.offsetX ?? prev?.offsetX ?? 0,
				offsetY: metadata.transform.offsetY ?? prev?.offsetY ?? 0,
				rotation: metadata.transform.rotation ?? prev?.rotation ?? 0,
				scale: metadata.transform.scale ?? prev?.scale ?? 1,
				anchorX: metadata.transform.anchorX ?? prev?.anchorX ?? .5,
				anchorY: metadata.transform.anchorY ?? prev?.anchorY ?? .5,
				edges: metadata.transform.edges ?? prev?.edges ?? "transparent"
			};
			const wasActive = transformActive(node);
			node.metadata.transform = merged;
			const nowActive = needsTransformation(merged);
			if (wasActive !== nowActive || prev && prev.edges !== merged.edges) needsRecompose = true;
			else if (nowActive) for (const [k, v] of Object.entries(transformFieldValues(node))) writeSyntheticScalar(node, `_xform_${k}`, v);
		}
		if (metadata.boundingBox !== void 0 && metadata.boundingBox !== null && node.boundingBoxDeclaration !== void 0) {
			const wasActive = bboxActive(node);
			node.metadata.boundingBox = metadata.boundingBox;
			if (wasActive) {
				const bb = bboxFieldValues(node);
				if (bb) for (const [k, v] of Object.entries(bb)) writeSyntheticScalar(node, `_bbox_${k}`, v);
			} else needsRecompose = true;
		} else if (metadata.boundingBox === null) {
			if (node.metadata.boundingBox !== void 0) {
				node.metadata.boundingBox = void 0;
				needsRecompose = true;
			}
		}
		if (needsRecompose) markStructuralDirty();
		else requestRender();
	};
	const removeNode = (id) => {
		const node = nodes.get(id);
		if (!node) return;
		for (const child of findChildNodes(id)) removeNode(child.id);
		if (rootId === id) rootId = null;
		const siblingSet = parentToChildren.get(node.parentId);
		if (siblingSet) {
			siblingSet.delete(id);
			if (siblingSet.size === 0) parentToChildren.delete(node.parentId);
		}
		nodes.delete(id);
		updateMouseListenerRegistration();
		markStructuralDirty();
	};
	const updateMouseDrivers = () => {
		const deltaTime = frameParams.deltaTime;
		globalElapsedTime = deriveElapsedTime(sharedTimeOrigin, globalElapsedTime, deltaTime, performance.now());
		for (const node of nodes.values()) {
			const maps = node.metadata.maps;
			if (!maps) continue;
			for (const propName in node.mouseDriverState) {
				const state = node.mouseDriverState[propName];
				const driver = maps[propName];
				if (!driver) continue;
				if (driver.type === "mouse-position") {
					const smoothing = driver.smoothing ?? 0;
					const momentum = driver.momentum ?? 0;
					const trackX = driver.x === void 0 || driver.x === "mouse";
					const trackY = driver.y === void 0 || driver.y === "mouse";
					let finalX;
					let finalY;
					if (trackX) {
						const [nx, vx] = applySpring(state.currentX, state.velocityX, pointerX, smoothing, momentum, deltaTime);
						state.currentX = nx;
						state.velocityX = vx;
						finalX = nx;
					} else {
						finalX = driver.x;
						state.velocityX = 0;
					}
					if (trackY) {
						const [ny, vy] = applySpring(state.currentY, state.velocityY, pointerY, smoothing, momentum, deltaTime);
						state.currentY = ny;
						state.velocityY = vy;
						finalY = ny;
					} else {
						finalY = driver.y;
						state.velocityY = 0;
					}
					const reach = driver.reach ?? 1;
					const originX = driver.originX ?? .5;
					const originY = driver.originY ?? .5;
					const xSign = driver.invertX ? -1 : 1;
					const ySign = driver.invertY ? -1 : 1;
					const scaledX = originX + xSign * (finalX - originX) * reach;
					const scaledY = originY + ySign * (finalY - originY) * reach;
					let localX = scaledX;
					let localY = scaledY;
					const bb = node.metadata.boundingBox;
					if (bb && node.boundingBoxDeclaration?.supportsResizeFit && bboxActive(node)) {
						const p = boundingBoxToUVParams(bb, pxResWidth(), pxResHeight());
						const bl = screenUVToBoxLocal$1(scaledX, scaledY, {
							centerX: p.centerX,
							centerY: p.centerY,
							halfWidthUV: p.halfWidthUV,
							halfHeightUV: p.halfHeightUV,
							rotationDeg: p.rotation,
							aspectRatio: aspectRatio()
						});
						localX = bl.x;
						localY = bl.y;
					}
					state.liveX = localX;
					state.liveY = localY;
					const h = node._liveHandles?.[`_smoothed_${propName}`];
					if (h instanceof FieldHandle) h.setComponents(localX, 1 - localY);
				} else if (driver.type === "mouse") {
					const smoothing = driver.smoothing ?? 0;
					const momentum = driver.momentum ?? 0;
					const target = driver.axis === "x" ? pointerX : pointerY;
					const [nx, vx] = applySpring(state.currentX, state.velocityX, target, smoothing, momentum, deltaTime);
					state.currentX = nx;
					state.velocityX = vx;
					const smoothed01 = Math.max(.001, nx);
					const mv = node.mapValues[propName];
					const resolved = mv ? mv.outputMin + (mv.outputMax - mv.outputMin) * Math.pow(smoothed01, Math.pow(2, -mv.curve * 2)) : smoothed01;
					state.liveScalar = resolved;
					writeSyntheticScalar(node, `_smoothed_${propName}`, resolved);
				}
			}
			for (const propName in node.autoAnimateState) {
				const driver = maps[propName];
				if (!driver || driver.type !== "auto-animate") continue;
				const globalT = globalElapsedTime * (driver.speed ?? 1) * .2;
				let phase;
				const easing = driver.easing ?? driver.waveform ?? "sine";
				if (driver.mode === "loop") phase = applyEasing((globalT % 1 + 1) % 1, easing);
				else {
					const t01 = (globalT % 1 + 1) % 1;
					phase = applyEasing(t01 < .5 ? t01 * 2 : (1 - t01) * 2, easing);
				}
				const mv = node.mapValues[propName];
				const outputMin = mv ? mv.outputMin : resolveDynamicBound(node, propName, driver.outputMin);
				const outputMax = mv ? mv.outputMax : resolveDynamicBound(node, propName, driver.outputMax);
				const newValue = outputMin + phase * (outputMax - outputMin);
				const u = node.uniforms[propName];
				if (u?.compileTimeWhen) {
					const prev = u._lastCompiledValue;
					if (u.compileTimeWhen(prev, newValue)) {
						u._lastCompiledValue = newValue;
						markStructuralDirty();
					}
				}
				writeSyntheticScalar(node, `_smoothed_${propName}`, newValue);
			}
		}
	};
	const updateAnimatedTime = () => {
		const dt = frameParams.deltaTime;
		for (const node of nodes.values()) {
			const at = node.definition.animatedTime;
			if (at) {
				const speedU = node.uniforms[at.speed];
				const speed = speedU ? Number(speedU.value) || 0 : 0;
				const state = getAnimatedTimeState(animatedTimeKey(node));
				state.advance(dt, speed);
				writeSyntheticScalar(node, "_animTime", state.value);
			}
			const extras = node.definition.extraAnimatedTimes;
			if (extras) for (const [key, speedProp] of Object.entries(extras)) {
				const su = node.uniforms[speedProp];
				const sp = su ? Number(su.value) || 0 : 0;
				const st = getAnimatedTimeState(extraAnimatedTimeKey(node, speedProp));
				st.advance(dt, sp);
				writeSyntheticScalar(node, `_animTime_${key}`, st.value);
			}
		}
	};
	const clampToTextureCap$1 = (w, h) => clampToTextureCap(w, h, {
		maxTextureDim: cachedMaxTextureDim,
		pixelRatio,
		viewportWidth: typeof window !== "undefined" ? window.innerWidth : w,
		viewportHeight: typeof window !== "undefined" ? window.innerHeight : h
	});
	const basePixelRatio = () => {
		const deviceRatio = (typeof window !== "undefined" ? window.devicePixelRatio : 1) || 1;
		return Math.min(deviceRatio, maxPixelRatioForDevice());
	};
	const applyRendererDimensions = (width, height, rawWidth, rawHeight) => {
		currentWidth = width;
		currentHeight = height;
		logicalWidth = rawWidth;
		logicalHeight = rawHeight;
		setShaderCanvasDimensions(rawWidth, rawHeight);
		if (!hasInitialDimensions) {
			hasInitialDimensions = true;
			if (nodes.size > 0) markStructuralDirty();
		}
		const { width: bufW, height: bufH } = bufferSize();
		if (canvasElement) {
			canvasElement.width = bufW;
			canvasElement.height = bufH;
		}
		for (const c of liveCompositions) c.passManager.resize(bufW, bufH);
		for (const c of liveCompositions) for (const cb of c.ir.onResize) try {
			cb({
				width: bufW,
				height: bufH
			});
		} catch (e) {
			console.error("[gpu] composition onResize error:", e);
		}
		for (const node of nodes.values()) {
			if (transformActive(node)) writeSyntheticScalar(node, "_xform_aspectRatio", aspectRatio());
			if (bboxActive(node)) {
				const bb = bboxFieldValues(node);
				if (bb) for (const [k, v] of Object.entries(bb)) writeSyntheticScalar(node, `_bbox_${k}`, v);
			}
			reresolveDimensionalUniforms(node);
		}
		for (const c of liveCompositions) {
			writeSystemUniforms(c.store);
			c.store.flush();
			c.passManager.repaintRtt();
		}
		lastRenderTime = 0;
		renderFrame();
	};
	const updateRendererDimensions = (width, height) => {
		if (width <= 0 || height <= 0) return;
		if (width === lastRequestedWidth && height === lastRequestedHeight) return;
		lastRequestedWidth = width;
		lastRequestedHeight = height;
		pendingResize = {
			width,
			height
		};
		if (!isResizeScheduled && typeof requestAnimationFrame === "function") {
			isResizeScheduled = true;
			requestAnimationFrame(() => {
				isResizeScheduled = false;
				if (!pendingResize) return;
				const raw = pendingResize;
				pendingResize = null;
				const { width: cw, height: ch } = clampToTextureCap$1(raw.width, raw.height);
				applyRendererDimensions(cw, ch, raw.width, raw.height);
			});
		}
	};
	const handleResize = (entries) => {
		const entry = entries[0];
		if (!entry) return;
		let width, height;
		if (entry.contentBoxSize) {
			const boxSize = Array.isArray(entry.contentBoxSize) ? entry.contentBoxSize[0] : entry.contentBoxSize;
			if (boxSize) {
				width = Math.round(boxSize.inlineSize);
				height = Math.round(boxSize.blockSize);
			} else {
				width = Math.round(entry.contentRect.width);
				height = Math.round(entry.contentRect.height);
			}
		} else {
			width = Math.round(entry.contentRect.width);
			height = Math.round(entry.contentRect.height);
		}
		if (width === 0 || height === 0) return;
		updateRendererDimensions(width, height);
	};
	const handleVisibilityChange = (entries) => {
		const entry = entries[0];
		if (!entry) return;
		isVisible = entry.isIntersecting;
	};
	const canRender = () => isInitialized && !!root && !!context && currentWidth > 0 && currentHeight > 0;
	const writeSystemUniforms = (store) => {
		const sys = store.systemHandles;
		const { width, height } = bufferSize();
		const bpr = basePixelRatio();
		const set = (name, value) => {
			const h = sys[name];
			if (h instanceof FieldHandle) h.value = value;
		};
		set("time", globalElapsedTime);
		set("viewportSize", [width, height]);
		set("logicalViewportSize", [pxResWidth() * bpr, pxResHeight() * bpr]);
		set("aspect", width / Math.max(1, height));
		set("pointer", [pointerX, pointerY]);
		set("pointerActive", pointerActive ? 1 : 0);
	};
	const runCallbacks = (callbacks) => {
		for (const cb of callbacks) try {
			cb(frameParams);
		} catch (e) {
			console.error("[gpu] render callback error:", e);
		}
	};
	const renderFrameInternal = (overrideDeltaTime) => {
		if (!canRender()) return;
		const now = performance.now();
		let delta;
		if (overrideDeltaTime !== void 0) delta = overrideDeltaTime;
		else {
			const gate = frameGate(now, {
				lastRenderTime,
				isVisible,
				forceFullFrameRate
			});
			if (!gate.render) return;
			delta = gate.deltaTime;
		}
		lastRenderTime = now;
		if (enablePerformanceTracking) performance.mark("shader-cpu-start");
		frameParams.deltaTime = delta;
		frameParams.pointer.x = pointerX;
		frameParams.pointer.y = pointerY;
		frameParams.pointerActive = pointerActive;
		frameParams.dimensions.width = currentWidth;
		frameParams.dimensions.height = currentHeight;
		let built = null;
		let renderComp = null;
		runFrameSequence({
			updateDrivers: () => {
				updateMouseDrivers();
				updateAnimatedTime();
				updateChildBounds();
			},
			ensureComposition: () => {
				if (structuralDirty || desiredHash === null) {
					desiredHash = computeStructuralHash();
					structuralDirty = false;
				}
				if (desiredHash === null) return false;
				built = pipelineCache.getOrBuild(desiredHash, () => buildComposition(desiredHash));
				bindComposition(built);
				return !!boundComposition;
			},
			beforeRender: () => runCallbacks(boundComposition.ir.onBeforeRender),
			flush: () => {
				writeSystemUniforms(boundComposition.store);
				boundComposition.store.flush();
			},
			render: () => {
				try {
					if (enablePerformanceTracking) performance.mark("shader-gpu-start");
					renderComp = pipelineCache.renderValue ?? built;
					renderComp?.passManager.render(context, frameParams);
					if (enablePerformanceTracking) {
						performance.mark("shader-gpu-end");
						try {
							performance.measure("shader-gpu-time", "shader-gpu-start", "shader-gpu-end");
							const gpuMeasure = performance.getEntriesByName("shader-gpu-time")[0];
							if (gpuMeasure) performanceTracker.recordGpuTime(gpuMeasure.duration);
							performance.clearMeasures("shader-gpu-time");
							performance.clearMarks("shader-gpu-start");
							performance.clearMarks("shader-gpu-end");
						} catch {}
					}
				} catch (error) {
					console.error("[gpu] render frame error:", error);
				}
			},
			markReady: () => {
				pipelineCache.markReady(desiredHash);
				if (firstRenderPending && !hasEmittedReady) {
					hasEmittedReady = true;
					firstRenderPending = false;
					if (onReadyCallback) queueMicrotask(onReadyCallback);
				}
			},
			afterRender: () => runCallbacks((renderComp ?? boundComposition).ir.onAfterRender)
		});
		if (enablePerformanceTracking) {
			performance.mark("shader-cpu-end");
			try {
				performance.measure("shader-cpu-time", "shader-cpu-start", "shader-cpu-end");
				const cpuMeasure = performance.getEntriesByName("shader-cpu-time")[0];
				if (cpuMeasure) performanceTracker.recordCpuTime(cpuMeasure.duration);
				performance.clearMarks("shader-cpu-start");
				performance.clearMarks("shader-cpu-end");
				performance.clearMeasures("shader-cpu-time");
			} catch {}
		}
		performanceTracker.recordFrame(performance.now() - now);
	};
	const renderFrame = () => renderFrameInternal(void 0);
	const renderAndWait = async () => {
		if (!canRender()) throw new Error("Renderer is not ready");
		const wasAnimating = frameLoop.running;
		if (wasAnimating) stopAnimation();
		lastRenderTime = 0;
		renderFrame();
		if (root) await awaitGpuIdle(root);
		if (wasAnimating) startAnimation();
	};
	const renderSyntheticFrame = async (deltaSeconds) => {
		renderFrameInternal(deltaSeconds);
		if (root) await awaitGpuIdle(root);
	};
	const startAnimation = () => {
		if (frameLoop.running || !shouldAnimate) return;
		performanceTracker.setRendering(true);
		frameLoop.start();
	};
	const stopAnimation = () => {
		if (!frameLoop.running) return;
		frameLoop.stop();
		performanceTracker.setRendering(false);
	};
	const updatePointerCoords = (event) => {
		if (!canvasElement) return;
		cachedCanvasRect = canvasElement.getBoundingClientRect();
		let clientX;
		let clientY;
		if ("touches" in event) {
			if (event.touches.length === 0) return;
			clientX = event.touches[0].clientX;
			clientY = event.touches[0].clientY;
		} else {
			clientX = event.clientX;
			clientY = event.clientY;
		}
		pointerX = (clientX - cachedCanvasRect.left) / cachedCanvasRect.width;
		pointerY = (clientY - cachedCanvasRect.top) / cachedCanvasRect.height;
	};
	const onPointerMove = (event) => {
		if (!isInitialized || !canvasElement) return;
		updatePointerCoords(event);
		requestRender();
	};
	const onTouchMove = (event) => {
		if (!isInitialized || !canvasElement) return;
		updatePointerCoords(event);
		requestRender();
	};
	const onPointerUp = () => {
		if (!isInitialized) return;
		pointerActive = false;
		requestRender();
	};
	const onTouchEnd = () => {
		if (!isInitialized) return;
		pointerActive = false;
		requestRender();
	};
	const onPointerDown = () => {
		pointerActive = true;
		requestRender();
	};
	const onTouchStart = () => {
		pointerActive = true;
		requestRender();
	};
	const updateMouseListenerRegistration = () => {
		let needsMouse = false;
		for (const node of nodes.values()) {
			if (node.usesPointer) {
				needsMouse = true;
				break;
			}
			if (shapeConfigUsesMouse(node.uniforms?.shape?.value)) {
				needsMouse = true;
				break;
			}
			if (!node.metadata.maps) continue;
			for (const driver of Object.values(node.metadata.maps)) if (driver && typeof driver === "object" && "type" in driver) {
				const t = driver.type;
				if (t === "mouse" || t === "mouse-position") {
					needsMouse = true;
					break;
				}
			}
			if (needsMouse) break;
		}
		if (needsMouse && !mouseListenersNeeded) {
			mouseListenersNeeded = true;
			if (!globalEventUnregister && canvasElement) {
				globalEventUnregister = globalEventManager.register({
					onMouseMove: onPointerMove,
					onTouchMove,
					onMouseUp: onPointerUp,
					onTouchEnd
				});
				canvasElement.addEventListener("mousedown", onPointerDown);
				canvasElement.addEventListener("touchstart", onTouchStart);
			}
		} else if (!needsMouse && mouseListenersNeeded) {
			mouseListenersNeeded = false;
			if (globalEventUnregister) {
				globalEventUnregister();
				globalEventUnregister = null;
			}
			if (canvasElement) {
				canvasElement.removeEventListener("mousedown", onPointerDown);
				canvasElement.removeEventListener("touchstart", onTouchStart);
			}
		}
	};
	const processQueuedRegistrations = () => {
		if (pendingRegistrationQueue.length === 0) return;
		const queue = [...pendingRegistrationQueue];
		pendingRegistrationQueue = [];
		for (const q of queue) registerNode(q.id, q.definition.fragment, q.parentId, q.metadata, q.uniforms, q.definition, q.domCanvas);
	};
	const fireDeviceLost = (reason) => {
		if (deviceLostFired) return;
		deviceLostFired = true;
		try {
			stopAnimation();
		} catch {}
		if (onDeviceLostCallback) queueMicrotask(() => {
			try {
				onDeviceLostCallback?.(reason);
			} catch (e) {
				console.error("[gpu] onDeviceLost callback threw:", e);
			}
		});
	};
	const initialize = async (options) => {
		if (options.context !== void 0) throw new Error("[gpu] initialize({ context }) is not supported — this renderer is WebGPU-only. Remove the WebGL `context` option.");
		if (isInitialized || isInitializing) return;
		const { canvas, resizeTarget, enablePerformanceTracking: enableTracking = false, colorSpace = "p3-linear", toneMapping = "linear", gpu, forceFullFrameRate: forceFullFps = false, observeElement: shouldObserveElement = true, timeOrigin: initialTimeOrigin } = options;
		if (initialTimeOrigin != null) sharedTimeOrigin = initialTimeOrigin;
		enablePerformanceTracking = enableTracking;
		forceFullFrameRate = forceFullFps;
		rendererColorSpace = colorSpace;
		rendererToneMapping = toneMapping;
		isInitializing = true;
		initializationAbortController = new AbortController();
		const abort = initializationAbortController;
		try {
			setColorSpaceModeFn = setColorSpaceModeGpu;
			setColorSpaceModeFn(colorSpace);
			if (abort.signal.aborted) return;
			canvasElement = canvas;
			rootCtx = await acquireRoot({
				device: gpu?.device,
				adapter: gpu?.adapter,
				onDeviceLost: (info) => fireDeviceLost(info.message || info.reason || "unknown")
			});
			if (abort.signal.aborted) return;
			root = rootCtx.root;
			pixelRatio = basePixelRatio();
			const lim = root.device.limits?.maxTextureDimension2D;
			if (typeof lim === "number" && lim > 0) cachedMaxTextureDim = lim;
			textureManager = createTextureManager(root);
			dispatcher = createComputeDispatcher(root);
			if (typeof window !== "undefined") {
				unloadHandler = () => cleanup();
				window.addEventListener("beforeunload", unloadHandler);
				windowResizeHandler = () => {
					if (lastRequestedWidth > 0 && lastRequestedHeight > 0) {
						const rawW = lastRequestedWidth;
						const rawH = lastRequestedHeight;
						lastRequestedWidth = 0;
						lastRequestedHeight = 0;
						updateRendererDimensions(rawW, rawH);
					}
				};
				window.addEventListener("resize", windowResizeHandler, { passive: true });
			}
			const rect = canvas.getBoundingClientRect();
			cachedCanvasRect = rect;
			isVisible = !shouldObserveElement || rect.width > 0 && rect.height > 0 && rect.top < window.innerHeight && rect.bottom > 0 && rect.left < window.innerWidth && rect.right > 0;
			shouldAnimate = true;
			const rawWidth = rect.width > 0 ? rect.width : canvas.width;
			const rawHeight = rect.height > 0 ? rect.height : canvas.height;
			const roundedRawWidth = Math.round(rawWidth);
			const roundedRawHeight = Math.round(rawHeight);
			if (roundedRawWidth > 0 && roundedRawHeight > 0) {
				lastRequestedWidth = roundedRawWidth;
				lastRequestedHeight = roundedRawHeight;
				const { width: cw, height: ch } = clampToTextureCap$1(roundedRawWidth, roundedRawHeight);
				currentWidth = cw;
				currentHeight = ch;
				logicalWidth = roundedRawWidth;
				logicalHeight = roundedRawHeight;
				hasInitialDimensions = true;
				setShaderCanvasDimensions(roundedRawWidth, roundedRawHeight);
				const { width: bufW, height: bufH } = bufferSize();
				canvas.width = bufW;
				canvas.height = bufH;
			} else hasInitialDimensions = false;
			context = configureCanvasContext(root, canvas);
			if (shouldObserveElement && typeof ResizeObserver !== "undefined") {
				const elementToObserve = resizeTarget ?? canvas.parentElement;
				if (elementToObserve) {
					resizeObserver = new ResizeObserver(handleResize);
					resizeObserver.observe(elementToObserve);
				} else console.warn("[gpu] no element available for resize observation — use resize() manually");
				if (typeof IntersectionObserver !== "undefined") {
					intersectionObserver = new IntersectionObserver(handleVisibilityChange, {
						threshold: 0,
						rootMargin: "128px"
					});
					intersectionObserver.observe(canvas);
				}
			}
			if (abort.signal.aborted) return;
			isInitialized = true;
			isRendererReady = true;
			processQueuedRegistrations();
			if (currentWidth > 0 && currentHeight > 0) renderFrame();
			if (shouldAnimate) startAnimation();
		} catch (error) {
			console.error("[gpu] initialization error:", error);
		} finally {
			isInitializing = false;
			if (initializationAbortController === abort) initializationAbortController = null;
		}
	};
	const cleanup = () => {
		if (initializationAbortController) initializationAbortController.abort();
		if (resizeObserver) {
			resizeObserver.disconnect();
			resizeObserver = null;
		}
		if (intersectionObserver) {
			intersectionObserver.disconnect();
			intersectionObserver = null;
		}
		if (globalEventUnregister) {
			globalEventUnregister();
			globalEventUnregister = null;
		}
		if (canvasElement) {
			canvasElement.removeEventListener("mousedown", onPointerDown);
			canvasElement.removeEventListener("touchstart", onTouchStart);
		}
		if (unloadHandler && typeof window !== "undefined") {
			window.removeEventListener("beforeunload", unloadHandler);
			unloadHandler = null;
		}
		if (windowResizeHandler && typeof window !== "undefined") {
			window.removeEventListener("resize", windowResizeHandler);
			windowResizeHandler = null;
		}
		if (pendingRenderRAF !== null && typeof cancelAnimationFrame === "function") {
			cancelAnimationFrame(pendingRenderRAF);
			pendingRenderRAF = null;
		}
		stopAnimation();
		pipelineCache.clear();
		for (const c of [...liveCompositions]) disposeComposition(c);
		boundComposition = null;
		try {
			textureManager?.destroy();
		} catch {}
		try {
			if (rootCtx?.createdDevice) root?.destroy();
		} catch {}
		nodes.clear();
		rootId = null;
		parentToChildren.clear();
		pendingRegistrationQueue = [];
		isRendererReady = false;
		pendingResize = null;
		isResizeScheduled = false;
		canvasElement = null;
		context = null;
		textureManager = null;
		dispatcher = null;
		root = null;
		rootCtx = null;
		isInitialized = false;
		isInitializing = false;
		hasInitialDimensions = false;
		initializationAbortController = null;
		currentWidth = 0;
		currentHeight = 0;
		logicalWidth = 0;
		logicalHeight = 0;
		desiredHash = null;
		structuralDirty = true;
		isVisible = false;
		shouldAnimate = true;
		pointerX = .5;
		pointerY = .5;
		pointerActive = false;
		onReadyCallback = null;
		hasEmittedReady = false;
		firstRenderPending = false;
		onDeviceLostCallback = null;
		deviceLostFired = false;
	};
	const getPerformanceStats = () => {
		const nodeCount = Math.max(0, nodes.size - 1);
		const rttNodeCount = boundComposition ? boundComposition.ir.rttPasses.length : 0;
		performanceTracker.updateNodeCounts(nodeCount, rttNodeCount);
		const rendererInfo = {
			render: { calls: boundComposition ? boundComposition.passManager.passCount : 0 },
			programs: { length: pipelineCache.size },
			memory: { textures: textureManager ? textureManager.textureCount : 0 }
		};
		return performanceTracker.getStats(rendererInfo);
	};
	return {
		initialize,
		cleanup,
		registerNode,
		removeNode,
		updateUniformValue,
		updateNodeMetadata,
		isInitialized: () => isInitialized,
		resize: (width, height) => {
			if (!root) return;
			updateRendererDimensions(width, height);
		},
		startAnimation,
		stopAnimation,
		renderAndWait,
		renderSyntheticFrame,
		getPerformanceStats,
		getLiveDriverValue: (nodeId, propName) => {
			const state = nodes.get(nodeId)?.mouseDriverState?.[propName];
			if (!state || state.liveX === void 0 || state.liveY === void 0) return null;
			return {
				x: state.liveX,
				y: state.liveY
			};
		},
		getLiveScalarValue: (nodeId, propName) => {
			const node = nodes.get(nodeId);
			if (!node) return null;
			const driver = node.metadata.maps?.[propName];
			if (driver?.type === "auto-animate") {
				const h = node._liveHandles?.[`_smoothed_${propName}`];
				const v = h instanceof FieldHandle ? h.value : void 0;
				return typeof v === "number" ? v : null;
			}
			if (driver?.type === "mouse") return node.mouseDriverState[propName]?.liveScalar ?? null;
			return null;
		},
		getNodeRegistry: () => ({ nodes: new Map(nodes) }),
		getRendererType: () => "webgpu",
		getInternalRenderer: () => root ? {
			device: root.device,
			adapter: rootCtx?.adapter,
			root
		} : null,
		setForceFullFrameRate: (enabled) => {
			const previous = forceFullFrameRate;
			forceFullFrameRate = enabled;
			return previous;
		},
		beginRecordingResolution: (targetWidth, targetHeight, ratio) => {
			if (!root || targetWidth <= 0 || targetHeight <= 0) return () => {};
			const prevRatio = pixelRatio;
			const prevWidth = currentWidth;
			const prevHeight = currentHeight;
			const prevLogicalWidth = logicalWidth;
			const prevLogicalHeight = logicalHeight;
			const prevLastRequestedWidth = lastRequestedWidth;
			const prevLastRequestedHeight = lastRequestedHeight;
			pixelRatio = ratio;
			applyRendererDimensions(targetWidth, targetHeight, targetWidth, targetHeight);
			return () => {
				if (!root) return;
				pixelRatio = prevRatio;
				applyRendererDimensions(prevWidth, prevHeight, prevLogicalWidth, prevLogicalHeight);
				lastRequestedWidth = prevLastRequestedWidth;
				lastRequestedHeight = prevLastRequestedHeight;
			};
		},
		setResolutionScale: (scale) => {
			if (!root) return;
			const clampedScale = Math.max(.05, Math.min(scale, 1));
			const targetRatio = Math.max(.01, basePixelRatio() * clampedScale);
			if (Math.abs(pixelRatio - targetRatio) < .001) return;
			pixelRatio = targetRatio;
			const logicalW = logicalWidth || currentWidth;
			const logicalH = logicalHeight || currentHeight;
			if (logicalW <= 0 || logicalH <= 0) return;
			const { width, height } = clampToTextureCap$1(logicalW, logicalH);
			applyRendererDimensions(width, height, logicalW, logicalH);
		},
		setTimeOrigin: (origin) => {
			sharedTimeOrigin = origin;
		},
		setOnReady: (callback) => {
			onReadyCallback = callback;
			if (callback && hasEmittedReady) queueMicrotask(callback);
		},
		setOnDeviceLost: (callback) => {
			onDeviceLostCallback = callback;
		},
		__testing: {
			needsTransformation,
			findChildNodes,
			buildFieldInits: (id) => {
				const node = nodes.get(id);
				return node ? buildFieldInits(node) : null;
			},
			stepAnimatedTime: (deltaTime) => {
				frameParams.deltaTime = deltaTime;
				updateAnimatedTime();
			},
			getAnimatedTimeValue: (id) => {
				const node = nodes.get(id);
				if (!node?.definition.animatedTime) return null;
				return getAnimatedTimeState(animatedTimeKey(node)).value;
			},
			writeExtraField: (id, name, value) => {
				const node = nodes.get(id);
				if (node) writeExtraFieldValue(node, name, value);
			},
			readExtraField: (id, name) => {
				const h = nodes.get(id)?._liveHandles?.[name];
				return h instanceof FieldHandle ? h.value : void 0;
			},
			getNodeRegistry: () => ({
				nodes,
				rootId,
				parentToChildren
			}),
			getParentToChildren: () => parentToChildren,
			getPendingRegistrations: () => [...pendingRegistrationQueue],
			getFrameDiagnostics: () => ({
				deltaTime: frameParams.deltaTime,
				pointerX,
				pointerY,
				pointerActive,
				globalElapsedTime,
				lastRenderTime,
				isVisible,
				forceFullFrameRate,
				sharedTimeOrigin,
				animating: frameLoop.running,
				mouseListenersNeeded
			}),
			computeStructuralHash,
			isStructuralDirty: () => structuralDirty,
			clearStructuralDirty: () => {
				structuralDirty = false;
			},
			setTestReady: (dims) => {
				isInitialized = true;
				isRendererReady = true;
				if (dims) {
					currentWidth = dims.width;
					currentHeight = dims.height;
					logicalWidth = dims.width;
					logicalHeight = dims.height;
					hasInitialDimensions = true;
				}
				processQueuedRegistrations();
			}
		}
	};
}
function createRendererFromJSON(preset, options) {
	const coreRenderer = shaderRendererGPU();
	const componentRegistry = /* @__PURE__ */ new Map();
	const nodeIdMap = /* @__PURE__ */ new Map();
	let presetCanvas = null;
	getAllShaders().forEach((shader) => {
		componentRegistry.set(shader.definition.name, shader.definition);
	});
	async function initialize(canvas) {
		presetCanvas = canvas;
		await coreRenderer.initialize({
			canvas,
			enablePerformanceTracking: options?.enablePerformanceTracking || false,
			gpu: options?.gpu,
			forceFullFrameRate: options?.forceFullFrameRate || false
		});
		registerPresetNodes();
		coreRenderer.stopAnimation();
	}
	async function renderFrame() {
		if (!coreRenderer.isInitialized()) {
			console.warn("[presetRenderer] Renderer not yet initialized, skipping frame");
			return;
		}
		await coreRenderer.renderAndWait();
	}
	function updatePreset(newPreset) {
		nodeIdMap.forEach((_, nodeId) => {
			coreRenderer.removeNode(nodeId);
		});
		nodeIdMap.clear();
		preset = newPreset;
		registerPresetNodes();
	}
	function getGPUContext() {
		const internal = coreRenderer.getInternalRenderer();
		return {
			type: "webgpu",
			device: internal?.device,
			adapter: internal?.adapter,
			canvas: presetCanvas
		};
	}
	function dispose() {
		coreRenderer.cleanup();
	}
	function registerPresetNodes() {
		const rootId = "preset-root";
		coreRenderer.registerNode(rootId, rootPassthrough.fragment, null, null, {}, rootPassthrough);
		preset.components.forEach((component, index) => {
			registerComponent(component, rootId, index);
		});
	}
	function registerComponent(component, parentId, renderOrder) {
		const componentDef = componentRegistry.get(component.type);
		if (!componentDef) {
			console.warn(`[createRendererFromJSON] Unknown component type: ${component.type}`);
			return;
		}
		const nodeId = component.id || `${component.type}_${Date.now()}_${Math.random().toString(36).slice(2, 9)}`;
		nodeIdMap.set(nodeId, component.type);
		const uniforms = createGpuUniformsMap(componentDef, Object.fromEntries(Object.entries(componentDef.props).map(([key, propConfig]) => [key, component.props?.[key] !== void 0 ? component.props[key] : propConfig.default])), nodeId);
		const metadata = {
			blendMode: component.props?.blendMode || "normal",
			opacity: component.props?.opacity,
			visible: component.props?.visible,
			renderOrder,
			id: component.id,
			mask: component.props?.maskSource ? {
				source: component.props.maskSource,
				type: component.props.maskType || "alpha"
			} : void 0,
			transform: component.props?.transform ? {
				offsetX: 0,
				offsetY: 0,
				rotation: 0,
				scale: 1,
				anchorX: .5,
				anchorY: .5,
				edges: "transparent",
				...component.props.transform
			} : void 0
		};
		coreRenderer.registerNode(nodeId, componentDef.fragment, parentId, metadata, uniforms, componentDef);
		component.children?.forEach((child, index) => {
			registerComponent(child, nodeId, index);
		});
	}
	return {
		initialize,
		renderFrame,
		updatePreset,
		getGPUContext,
		dispose
	};
}
function resolveMaskDependencies(nodes) {
	const elementIdToNodeId = /* @__PURE__ */ new Map();
	const maskDependencies = /* @__PURE__ */ new Map();
	const result = {
		processingOrder: [],
		maskSources: /* @__PURE__ */ new Map(),
		circularDependencies: []
	};
	nodes.forEach((node) => {
		if (node.elementId) elementIdToNodeId.set(node.elementId, node.id);
	});
	nodes.forEach((node) => {
		if (node.mask?.source) {
			const sourceElementId = node.mask.source.startsWith("#") ? node.mask.source.substring(1) : node.mask.source;
			const sourceNodeId = elementIdToNodeId.get(sourceElementId);
			if (sourceNodeId) {
				if (!maskDependencies.has(node.id)) maskDependencies.set(node.id, []);
				maskDependencies.get(node.id)?.push(sourceNodeId);
				result.maskSources.set(node.id, sourceNodeId);
			} else console.warn(`Mask source "${sourceElementId}" not found for node "${node.id}"`);
		}
	});
	result.processingOrder = topologicalSort(maskDependencies, result.circularDependencies);
	return result;
}
function topologicalSort(dependencies, circularDependencies) {
	const result = [];
	const visited = /* @__PURE__ */ new Set();
	const tempVisited = /* @__PURE__ */ new Set();
	function visit(nodeId, path = []) {
		if (visited.has(nodeId)) return;
		if (tempVisited.has(nodeId)) {
			const cycle = [...path.slice(path.indexOf(nodeId)), nodeId];
			circularDependencies.push(cycle);
			return;
		}
		tempVisited.add(nodeId);
		path.push(nodeId);
		const deps = dependencies.get(nodeId) || [];
		for (const depId of deps) visit(depId, [...path]);
		tempVisited.delete(nodeId);
		visited.add(nodeId);
		result.push(nodeId);
	}
	for (const nodeId of dependencies.keys()) if (!visited.has(nodeId)) visit(nodeId);
	const allSourceNodeIds = new Set([...dependencies.values()].flat());
	for (const sourceId of allSourceNodeIds) if (!visited.has(sourceId) && !dependencies.has(sourceId)) result.push(sourceId);
	return result.reverse();
}
const screenUVToBoxLocal = (sx, sy, g) => {
	const th = g.rotationDeg * Math.PI / 180;
	const c = Math.cos(th);
	const s = Math.sin(th);
	const sqx = (sx - g.centerX) * g.aspectRatio;
	const sqy = sy - g.centerY;
	const rx = sqx * c + sqy * s;
	const ry = sqx * -s + sqy * c;
	const hwSq = Math.max(g.halfWidthUV * g.aspectRatio, 1e-6);
	const hhSq = Math.max(g.halfHeightUV, 1e-6);
	return {
		x: rx / (2 * hwSq) + .5,
		y: ry / (2 * hhSq) + .5
	};
};
const boxLocalToScreenUV = (bx, by, g) => {
	const th = g.rotationDeg * Math.PI / 180;
	const c = Math.cos(th);
	const s = Math.sin(th);
	const lx = (bx - .5) * 2 * (g.halfWidthUV * g.aspectRatio);
	const ly = (by - .5) * 2 * g.halfHeightUV;
	const sqx = lx * c - ly * s;
	const sqy = lx * s + ly * c;
	return {
		x: sqx / g.aspectRatio + g.centerX,
		y: sqy + g.centerY
	};
};
const DEFAULT_BOUNDING_BOX = {
	x: {
		value: 0,
		unit: "uv"
	},
	y: {
		value: 0,
		unit: "uv"
	},
	width: {
		value: 1,
		unit: "uv"
	},
	height: {
		value: 1,
		unit: "uv"
	},
	origin: "top-left",
	rotation: 0
};
function resolveBoundingBox(bbox) {
	if (!bbox || typeof bbox !== "object") return void 0;
	const dim = (d$1, fallback) => d$1 && typeof d$1 === "object" ? {
		value: d$1.value ?? fallback.value,
		unit: d$1.unit ?? fallback.unit
	} : { ...fallback };
	return {
		x: dim(bbox.x, DEFAULT_BOUNDING_BOX.x),
		y: dim(bbox.y, DEFAULT_BOUNDING_BOX.y),
		width: dim(bbox.width, DEFAULT_BOUNDING_BOX.width),
		height: dim(bbox.height, DEFAULT_BOUNDING_BOX.height),
		origin: bbox.origin ?? "top-left",
		rotation: bbox.rotation ?? 0,
		...bbox.cornerRadius ? { cornerRadius: bbox.cornerRadius } : {},
		...bbox.lockAspect !== void 0 ? { lockAspect: bbox.lockAspect } : {}
	};
}
export { DEFAULT_BOUNDING_BOX, MAX_COLOR_STOPS, SHAPE3D_DEFAULTS, SHAPE3D_TYPES, SVG3D_DEFAULTS, SVG3D_TYPES, boxLocalToScreenUV, createGpuUniformsMap, createRendererFromJSON, is3dShapeType, isDimensionalValue, isSvg3dShapeType, isVolumetricShapeType, resolveBoundingBox, resolveMaskDependencies, resolveStops, rootPassthrough, screenUVToBoxLocal, shaderRendererGPU, shape3dBoundingRadius, sortStops };
