// shaders/dist/core/Invert-DLot9mli.js
import { At as call, Ot as ZERO, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
const invertColor = tgpu.fn([d.vec4f], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color) => {
	"use gpu";
	return d.vec4f(__tsover_sub(1, color.x), __tsover_sub(1, color.y), __tsover_sub(1, color.z), color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec4f"
			],
			[
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"color",
						"x"
					]
				],
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"color",
						"y"
					]
				],
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"color",
						"z"
					]
				],
				[
					7,
					"color",
					"w"
				]
			]
		]]]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const componentDefinition = {
	name: "Invert",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Invert RGB colors while preserving alpha",
	requiresChild: true,
	props: {},
	fragment: ({ childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Invert shader.");
			return ZERO;
		}
		return call(invertColor, "invertColor", [childNode]);
	}
};
var Invert_default = componentDefinition;
export { componentDefinition as n, invertColor as r, Invert_default as t };
