// shaders/dist/core/IsometricCubes-BVYfIhbb.js
import { At as call, Ft as mixExpr, It as vec4, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
var DEG_TO_RAD = Math.PI / 180;
var SQRT3 = Math.sqrt(3);
var SQRT3_2 = Math.sqrt(3) / 2;
var INV_2_SQRT3 = 2 / Math.sqrt(3);
var flooredMod2 = tgpu.fn([d.vec2f, d.vec2f], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, m) => {
	"use gpu";
	return d.vec2f(__tsover_sub(p.x, __tsover_mul(m.x, std.floor(__tsover_div(p.x, m.x)))), __tsover_sub(p.y, __tsover_mul(m.y, std.floor(__tsover_div(p.y, m.y)))));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}, {
			"type": "i",
			"name": "m"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					7,
					"p",
					"x"
				],
				"-",
				[
					1,
					[
						7,
						"m",
						"x"
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"floor"
						],
						[[
							1,
							[
								7,
								"p",
								"x"
							],
							"/",
							[
								7,
								"m",
								"x"
							]
						]]
					]
				]
			], [
				1,
				[
					7,
					"p",
					"y"
				],
				"-",
				[
					1,
					[
						7,
						"m",
						"y"
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"floor"
						],
						[[
							1,
							[
								7,
								"p",
								"y"
							],
							"/",
							[
								7,
								"m",
								"y"
							]
						]]
					]
				]
			]]
		]]]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const isoCubeField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, cells, thickness, rotation, softness) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const correctedUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const ang = -__tsover_mul(rotation, DEG_TO_RAD);
	const cosR = std.cos(ang);
	const sinR = std.sin(ang);
	const centerX = __tsover_mul(aspect, .5);
	const centerY = .5;
	const cx = __tsover_sub(correctedUV.x, centerX);
	const cy = __tsover_sub(correctedUV.y, centerY);
	const scaledUV = d.vec2f(__tsover_add(__tsover_sub(__tsover_mul(cx, cosR), __tsover_mul(cy, sinR)), centerX), __tsover_add(__tsover_add(__tsover_mul(cx, sinR), __tsover_mul(cy, cosR)), centerY)).mul(cells);
	const s = d.vec2f(1, SQRT3);
	const sh = s.mul(.5);
	const a = flooredMod2(scaledUV, s).sub(sh);
	const b = flooredMod2(scaledUV.sub(sh), s).sub(sh);
	const gv = std.select(b, a, std.dot(a, a) < std.dot(b, b));
	const id = scaledUV.sub(gv);
	const s0 = std.dot(gv, d.vec2f(0, 1));
	const s1 = std.dot(gv, d.vec2f(-SQRT3_2, -.5));
	const s2 = std.dot(gv, d.vec2f(SQRT3_2, -.5));
	const isTop = s0 >= std.max(s1, s2);
	const isLeft = s1 >= s2;
	const tone = std.select(std.select(.74, .5, isLeft), 1, isTop);
	const ix = std.round(__tsover_mul(id.x, 2));
	const iy = std.round(__tsover_mul(id.y, INV_2_SQRT3));
	const q0 = std.fract(d.vec2f(ix, iy).mul(d.vec2f(123.34, 345.45)));
	const q1 = q0.add(std.dot(q0, q0.add(34.345)));
	const rand = std.fract(__tsover_mul(q1.x, q1.y));
	const mx = std.max(s0, std.max(s1, s2));
	const mn = std.min(s0, std.min(s1, s2));
	const midScore = __tsover_sub(__tsover_sub(__tsover_add(__tsover_add(s0, s1), s2), mx), mn);
	const internal = __tsover_sub(mx, midScore);
	const pAbs = d.vec2f(std.abs(gv.x), std.abs(gv.y));
	const hexDist = std.max(std.dot(pAbs, d.vec2f(.5, SQRT3_2)), pAbs.x);
	const rimDist = __tsover_sub(.5, hexDist);
	const edgeField = std.min(internal, rimDist);
	const lineWidth = __tsover_mul(thickness, .02);
	const aa = __tsover_add(__tsover_mul(std.fwidth(scaledUV.x), .5), __tsover_mul(softness, .5));
	const wire = __tsover_mul(__tsover_sub(1, std.smoothstep(__tsover_sub(lineWidth, aa), __tsover_add(lineWidth, aa), edgeField)), std.step(1e-4, lineWidth));
	return d.vec3f(rand, tone, wire);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "cells"
			},
			{
				"type": "i",
				"name": "thickness"
			},
			{
				"type": "i",
				"name": "rotation"
			},
			{
				"type": "i",
				"name": "softness"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"correctedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"ang",
				[
					4,
					"-",
					[
						1,
						"rotation",
						"*",
						"DEG_TO_RAD"
					]
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["ang"]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["ang"]
				]
			],
			[
				13,
				"centerX",
				[
					1,
					"aspect",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"centerY",
				[5, "0.5"]
			],
			[
				13,
				"cx",
				[
					1,
					[
						7,
						"correctedUV",
						"x"
					],
					"-",
					"centerX"
				]
			],
			[
				13,
				"cy",
				[
					1,
					[
						7,
						"correctedUV",
						"y"
					],
					"-",
					"centerY"
				]
			],
			[
				13,
				"rotatedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								1,
								"cx",
								"*",
								"cosR"
							],
							"-",
							[
								1,
								"cy",
								"*",
								"sinR"
							]
						],
						"+",
						"centerX"
					], [
						1,
						[
							1,
							[
								1,
								"cx",
								"*",
								"sinR"
							],
							"+",
							[
								1,
								"cy",
								"*",
								"cosR"
							]
						],
						"+",
						"centerY"
					]]
				]
			],
			[
				13,
				"scaledUV",
				[
					6,
					[
						7,
						"rotatedUV",
						"mul"
					],
					["cells"]
				]
			],
			[
				13,
				"s",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[5, "1"], "SQRT3"]
				]
			],
			[
				13,
				"sh",
				[
					6,
					[
						7,
						"s",
						"mul"
					],
					[[5, "0.5"]]
				]
			],
			[
				13,
				"a",
				[
					6,
					[
						7,
						[
							6,
							"flooredMod2",
							["scaledUV", "s"]
						],
						"sub"
					],
					["sh"]
				]
			],
			[
				13,
				"b",
				[
					6,
					[
						7,
						[
							6,
							"flooredMod2",
							[[
								6,
								[
									7,
									"scaledUV",
									"sub"
								],
								["sh"]
							], "s"]
						],
						"sub"
					],
					["sh"]
				]
			],
			[
				13,
				"gv",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"b",
						"a",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"dot"
								],
								["a", "a"]
							],
							"<",
							[
								6,
								[
									7,
									"std",
									"dot"
								],
								["b", "b"]
							]
						]
					]
				]
			],
			[
				13,
				"id",
				[
					6,
					[
						7,
						"scaledUV",
						"sub"
					],
					["gv"]
				]
			],
			[
				13,
				"s0",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["gv", [
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "0"], [5, "1"]]
					]]
				]
			],
			[
				13,
				"s1",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["gv", [
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							4,
							"-",
							"SQRT3_2"
						], [
							4,
							"-",
							[5, "0.5"]
						]]
					]]
				]
			],
			[
				13,
				"s2",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["gv", [
						6,
						[
							7,
							"d",
							"vec2f"
						],
						["SQRT3_2", [
							4,
							"-",
							[5, "0.5"]
						]]
					]]
				]
			],
			[
				13,
				"isTop",
				[
					1,
					"s0",
					">=",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						["s1", "s2"]
					]
				]
			],
			[
				13,
				"isLeft",
				[
					1,
					"s1",
					">=",
					"s2"
				]
			],
			[
				13,
				"tone",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"std",
								"select"
							],
							[
								[5, "0.74"],
								[5, "0.5"],
								"isLeft"
							]
						],
						[5, "1"],
						"isTop"
					]
				]
			],
			[
				13,
				"ix",
				[
					6,
					[
						7,
						"std",
						"round"
					],
					[[
						1,
						[
							7,
							"id",
							"x"
						],
						"*",
						[5, "2"]
					]]
				]
			],
			[
				13,
				"iy",
				[
					6,
					[
						7,
						"std",
						"round"
					],
					[[
						1,
						[
							7,
							"id",
							"y"
						],
						"*",
						"INV_2_SQRT3"
					]]
				]
			],
			[
				13,
				"q0",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								["ix", "iy"]
							],
							"mul"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "123.34"], [5, "345.45"]]
						]]
					]]
				]
			],
			[
				13,
				"q1",
				[
					6,
					[
						7,
						"q0",
						"add"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["q0", [
							6,
							[
								7,
								"q0",
								"add"
							],
							[[5, "34.345"]]
						]]
					]]
				]
			],
			[
				13,
				"rand",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							7,
							"q1",
							"x"
						],
						"*",
						[
							7,
							"q1",
							"y"
						]
					]]
				]
			],
			[
				13,
				"mx",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["s0", [
						6,
						[
							7,
							"std",
							"max"
						],
						["s1", "s2"]
					]]
				]
			],
			[
				13,
				"mn",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					["s0", [
						6,
						[
							7,
							"std",
							"min"
						],
						["s1", "s2"]
					]]
				]
			],
			[
				13,
				"midScore",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								"s0",
								"+",
								"s1"
							],
							"+",
							"s2"
						],
						"-",
						"mx"
					],
					"-",
					"mn"
				]
			],
			[
				13,
				"internal",
				[
					1,
					"mx",
					"-",
					"midScore"
				]
			],
			[
				13,
				"pAbs",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							7,
							"gv",
							"x"
						]]
					], [
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							7,
							"gv",
							"y"
						]]
					]]
				]
			],
			[
				13,
				"hexDist",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["pAbs", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "0.5"], "SQRT3_2"]
						]]
					], [
						7,
						"pAbs",
						"x"
					]]
				]
			],
			[
				13,
				"rimDist",
				[
					1,
					[5, "0.5"],
					"-",
					"hexDist"
				]
			],
			[
				13,
				"edgeField",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					["internal", "rimDist"]
				]
			],
			[
				13,
				"lineWidth",
				[
					1,
					"thickness",
					"*",
					[5, "0.02"]
				]
			],
			[
				13,
				"aa",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"fwidth"
							],
							[[
								7,
								"scaledUV",
								"x"
							]]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[
						1,
						"softness",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"wire",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[
									1,
									"lineWidth",
									"-",
									"aa"
								],
								[
									1,
									"lineWidth",
									"+",
									"aa"
								],
								"edgeField"
							]
						]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"step"
						],
						[[5, "0.0001"], "lineWidth"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					"rand",
					"tone",
					"wire"
				]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std",
			"SQRT3",
			"flooredMod2",
			"SQRT3_2",
			"INV_2_SQRT3"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std,
		SQRT3,
		flooredMod2,
		SQRT3_2,
		INV_2_SQRT3
	})
}) && $.f)({}));
const componentDefinition = {
	name: "IsometricCubes",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Isometric tumbling-blocks tiling — a 3D cube illusion (rhombille pattern)",
	acceptsUVContext: true,
	props: {
		colorA: {
			default: "#7c5cff",
			transform: transformColor,
			description: "Base cube color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ff5c9d",
			transform: transformColor,
			description: "Second cube color, mixed in per-cube for variation",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		lineColor: {
			default: "#000000",
			transform: transformColor,
			description: "Color of the cube edge lines",
			ui: {
				type: "color",
				label: "Edges",
				group: "Colors"
			}
		},
		cells: {
			default: 6,
			description: "Number of cubes across the shortest canvas edge",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Cells",
				group: "Effect"
			}
		},
		thickness: {
			default: 1,
			description: "Thickness of the cube edge lines (0 = no edges)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 10,
				step: .1,
				label: "Thickness",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			transform: transformAngle,
			description: "Rotation of the tiling in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Softness of the edge lines (0 = crisp, 1 = very soft)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		colorVariation: {
			default: 1,
			description: "Per-cube random blend between the two colors (0 = uniform cubes, the clean 3D look)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Variation",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, propValues } = params;
		const field = call(isoCubeField, "isoCubeField", [
			params.uvContext ?? params.ctx.uv,
			params.effectiveViewportSize ?? params.ctx.viewportSize,
			uniforms.cells,
			uniforms.thickness,
			uniforms.rotation,
			uniforms.softness
		]);
		const rand = field.member("x");
		const tone = field.member("y");
		const wire = field.member("z");
		return vec4(mixExpr(call(mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			rand.mul(uniforms.colorVariation)
		]).member("rgb").mul(tone), uniforms.lineColor.member("rgb"), wire), 1);
	}
};
var IsometricCubes_default = componentDefinition;
export { componentDefinition as n, isoCubeField as r, IsometricCubes_default as t };
