// shaders/dist/core/Kaleidoscope-VxQtO7zA.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-BlwgBpJf.js";
import { c as transformPosition, s as transformEdges } from "./transformations-fqzznt4g.js";
var DEG_TO_RAD = Math.PI / 180;
var TAU = Math.PI * 2;
const kaleidoscopeUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, segments, angle, uv, aspect) => {
	"use gpu";
	const aspectCorrectedUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y));
	const delta = aspectCorrectedUV.sub(centerPos);
	const radius = std.length(delta);
	const currentAngle = std.atan2(delta.y, delta.x);
	const rotatedAngle = __tsover_add(currentAngle, __tsover_mul(angle, DEG_TO_RAD));
	const segmentAngle = __tsover_div(TAU, segments);
	const segTwo = __tsover_mul(segmentAngle, 2);
	const wrappedAngle = __tsover_sub(rotatedAngle, __tsover_mul(segTwo, std.floor(__tsover_div(rotatedAngle, segTwo))));
	const finalAngle = std.select(wrappedAngle, __tsover_sub(segTwo, wrappedAngle), wrappedAngle > segmentAngle);
	const kaleidoscopeX = __tsover_mul(std.cos(finalAngle), radius);
	const kaleidoscopeY = __tsover_mul(std.sin(finalAngle), radius);
	const kaleidoscopePos = centerPos.add(d.vec2f(kaleidoscopeX, kaleidoscopeY));
	return d.vec2f(__tsover_div(kaleidoscopePos.x, aspect), kaleidoscopePos.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "segments"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"aspectCorrectedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"delta",
				[
					6,
					[
						7,
						"aspectCorrectedUV",
						"sub"
					],
					["centerPos"]
				]
			],
			[
				13,
				"radius",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["delta"]
				]
			],
			[
				13,
				"currentAngle",
				[
					6,
					[
						7,
						"std",
						"atan2"
					],
					[[
						7,
						"delta",
						"y"
					], [
						7,
						"delta",
						"x"
					]]
				]
			],
			[
				13,
				"rotatedAngle",
				[
					1,
					"currentAngle",
					"+",
					[
						1,
						"angle",
						"*",
						"DEG_TO_RAD"
					]
				]
			],
			[
				13,
				"segmentAngle",
				[
					1,
					"TAU",
					"/",
					"segments"
				]
			],
			[
				13,
				"segTwo",
				[
					1,
					"segmentAngle",
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"wrappedAngle",
				[
					1,
					"rotatedAngle",
					"-",
					[
						1,
						"segTwo",
						"*",
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								1,
								"rotatedAngle",
								"/",
								"segTwo"
							]]
						]
					]
				]
			],
			[
				13,
				"finalAngle",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"wrappedAngle",
						[
							1,
							"segTwo",
							"-",
							"wrappedAngle"
						],
						[
							1,
							"wrappedAngle",
							">",
							"segmentAngle"
						]
					]
				]
			],
			[
				13,
				"kaleidoscopeX",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						["finalAngle"]
					],
					"*",
					"radius"
				]
			],
			[
				13,
				"kaleidoscopeY",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						["finalAngle"]
					],
					"*",
					"radius"
				]
			],
			[
				13,
				"kaleidoscopePos",
				[
					6,
					[
						7,
						"centerPos",
						"add"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						["kaleidoscopeX", "kaleidoscopeY"]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						7,
						"kaleidoscopePos",
						"x"
					],
					"/",
					"aspect"
				], [
					7,
					"kaleidoscopePos",
					"y"
				]]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"DEG_TO_RAD",
			"TAU"
		]
	},
	externals: () => ({
		d,
		std,
		DEG_TO_RAD,
		TAU
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Kaleidoscope",
	category: "Distortions",
	description: "Create a kaleidoscope effect with radial mirrored segments",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the kaleidoscope effect",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		segments: {
			default: 6,
			description: "Number of radial segments in the kaleidoscope",
			ui: {
				type: ["range", "map"],
				min: 2,
				max: 24,
				step: 1,
				label: "Segments",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			description: "Rotation offset for the entire kaleidoscope pattern",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, propValues, convertToTexture }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Kaleidoscope shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(kaleidoscopeUV, "kaleidoscopeUV", [
			uniforms.center,
			uniforms.segments,
			uniforms.angle,
			ctx.uv,
			ctx.aspect
		]), (uv) => texture.sample(uv), edgeMode)]);
	},
	uvRemap: ({ uv, mask, uniforms, aspect, propValues }) => {
		const edgeMode = propValues.edges ?? 0;
		return composeEdgeRemapExpr(call(kaleidoscopeUV, "kaleidoscopeUV", [
			uniforms.center,
			uniforms.segments,
			uniforms.angle,
			uv,
			aspect
		]), mask, edgeMode);
	}
};
var Kaleidoscope_default = componentDefinition;
export { componentDefinition as n, kaleidoscopeUV as r, Kaleidoscope_default as t };
