// shaders/dist/core/KeyFrames-DiAKRsxm.js
import { B as createGuardedCompute, Lt as Expr, Ot as ZERO, R as std, Rt as formatFloat, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor } from "./transformations-fqzznt4g.js";
var MAX_TRACKERS = 64;
var TRAIL_LEN = 12;
var TRAIL_EVERY = 8;
var AABB_COL = 2 + TRAIL_LEN;
var STATE_W = AABB_COL + 1;
var STATE_FORMAT = "rgba32float";
var FEAT_GRID = 96;
var REFERENCE_HEIGHT = 1080;
var TWO_PI = 6.283185307179586;
var TrackParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
	dt: d.f32,
	trackersF: d.f32,
	agility: d.f32,
	threshold: d.f32,
	variance: d.f32,
	lifespan: d.f32
}), "TrackParams");
var featureLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	src: { texture: d.texture2d(d.f32) },
	grid: { storageTexture: d.textureStorage2d(STATE_FORMAT, "write-only") },
	params: { uniform: TrackParams }
}), "featureLayout");
var trackLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	src: { texture: d.texture2d(d.f32) },
	featGrid: {
		texture: d.texture2d(d.f32),
		sampleType: "unfilterable-float"
	},
	prev: {
		texture: d.texture2d(d.f32),
		sampleType: "unfilterable-float"
	},
	next: { storageTexture: d.textureStorage2d(STATE_FORMAT, "write-only") },
	present: { storageTexture: d.textureStorage2d(STATE_FORMAT, "write-only") },
	params: { uniform: TrackParams }
}), "trackLayout");
const trackHash = tgpu.fn([d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((x) => {
	"use gpu";
	return std.fract(__tsover_mul(std.sin(x), 43758.5453123));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "x"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				1,
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["x"]
				],
				"*",
				[5, "43758.5453123"]
			]]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
function buildScoreFn(mode) {
	return tgpu.fn([d.vec4f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((c) => {
		"use gpu";
		const a = std.max(c.w, 1e-4);
		const lumaU = std.clamp(__tsover_div(__tsover_add(__tsover_add(__tsover_mul(c.x, .299), __tsover_mul(c.y, .587)), __tsover_mul(c.z, .114)), a), 0, 1);
		let score = d.f32(0);
		if (mode === "alpha") score = c.w;
		else if (mode === "dark") score = __tsover_mul(__tsover_sub(1, lumaU), std.select(d.f32(0), d.f32(1), c.w > .5));
		else if (mode === "red") score = __tsover_mul(__tsover_mul(__tsover_mul(std.clamp(__tsover_div(c.x, a), 0, 1), std.select(d.f32(0), d.f32(1), c.x >= c.y)), std.select(d.f32(0), d.f32(1), c.x >= c.z)), std.select(d.f32(0), d.f32(1), c.w > .02));
		else if (mode === "green") score = __tsover_mul(__tsover_mul(__tsover_mul(std.clamp(__tsover_div(c.y, a), 0, 1), std.select(d.f32(0), d.f32(1), c.y >= c.x)), std.select(d.f32(0), d.f32(1), c.y >= c.z)), std.select(d.f32(0), d.f32(1), c.w > .02));
		else if (mode === "blue") score = __tsover_mul(__tsover_mul(__tsover_mul(std.clamp(__tsover_div(c.z, a), 0, 1), std.select(d.f32(0), d.f32(1), c.z >= c.x)), std.select(d.f32(0), d.f32(1), c.z >= c.y)), std.select(d.f32(0), d.f32(1), c.w > .02));
		else score = __tsover_mul(lumaU, std.select(d.f32(0), d.f32(1), c.w > .02));
		return score;
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [{
				"type": "i",
				"name": "c"
			}],
			"body": [0, [
				[
					13,
					"a",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							7,
							"c",
							"w"
						], [5, "0.0001"]]
					]
				],
				[
					13,
					"lumaU",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"c",
												"x"
											],
											"*",
											[5, "0.299"]
										],
										"+",
										[
											1,
											[
												7,
												"c",
												"y"
											],
											"*",
											[5, "0.587"]
										]
									],
									"+",
									[
										1,
										[
											7,
											"c",
											"z"
										],
										"*",
										[5, "0.114"]
									]
								],
								"/",
								"a"
							],
							[5, "0"],
							[5, "1"]
						]
					]
				],
				[
					12,
					"score",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					11,
					[
						1,
						"mode",
						"===",
						[103, "alpha"]
					],
					[
						2,
						"score",
						"=",
						[
							7,
							"c",
							"w"
						]
					],
					[
						11,
						[
							1,
							"mode",
							"===",
							[103, "dark"]
						],
						[
							2,
							"score",
							"=",
							[
								1,
								[
									1,
									[5, "1"],
									"-",
									"lumaU"
								],
								"*",
								[
									6,
									[
										7,
										"std",
										"select"
									],
									[
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											[[5, "0"]]
										],
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											[[5, "1"]]
										],
										[
											1,
											[
												7,
												"c",
												"w"
											],
											">",
											[5, "0.5"]
										]
									]
								]
							]
						],
						[
							11,
							[
								1,
								"mode",
								"===",
								[103, "red"]
							],
							[
								2,
								"score",
								"=",
								[
									1,
									[
										1,
										[
											1,
											[
												6,
												[
													7,
													"std",
													"clamp"
												],
												[
													[
														1,
														[
															7,
															"c",
															"x"
														],
														"/",
														"a"
													],
													[5, "0"],
													[5, "1"]
												]
											],
											"*",
											[
												6,
												[
													7,
													"std",
													"select"
												],
												[
													[
														6,
														[
															7,
															"d",
															"f32"
														],
														[[5, "0"]]
													],
													[
														6,
														[
															7,
															"d",
															"f32"
														],
														[[5, "1"]]
													],
													[
														1,
														[
															7,
															"c",
															"x"
														],
														">=",
														[
															7,
															"c",
															"y"
														]
													]
												]
											]
										],
										"*",
										[
											6,
											[
												7,
												"std",
												"select"
											],
											[
												[
													6,
													[
														7,
														"d",
														"f32"
													],
													[[5, "0"]]
												],
												[
													6,
													[
														7,
														"d",
														"f32"
													],
													[[5, "1"]]
												],
												[
													1,
													[
														7,
														"c",
														"x"
													],
													">=",
													[
														7,
														"c",
														"z"
													]
												]
											]
										]
									],
									"*",
									[
										6,
										[
											7,
											"std",
											"select"
										],
										[
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												[[5, "0"]]
											],
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												[[5, "1"]]
											],
											[
												1,
												[
													7,
													"c",
													"w"
												],
												">",
												[5, "0.02"]
											]
										]
									]
								]
							],
							[
								11,
								[
									1,
									"mode",
									"===",
									[103, "green"]
								],
								[
									2,
									"score",
									"=",
									[
										1,
										[
											1,
											[
												1,
												[
													6,
													[
														7,
														"std",
														"clamp"
													],
													[
														[
															1,
															[
																7,
																"c",
																"y"
															],
															"/",
															"a"
														],
														[5, "0"],
														[5, "1"]
													]
												],
												"*",
												[
													6,
													[
														7,
														"std",
														"select"
													],
													[
														[
															6,
															[
																7,
																"d",
																"f32"
															],
															[[5, "0"]]
														],
														[
															6,
															[
																7,
																"d",
																"f32"
															],
															[[5, "1"]]
														],
														[
															1,
															[
																7,
																"c",
																"y"
															],
															">=",
															[
																7,
																"c",
																"x"
															]
														]
													]
												]
											],
											"*",
											[
												6,
												[
													7,
													"std",
													"select"
												],
												[
													[
														6,
														[
															7,
															"d",
															"f32"
														],
														[[5, "0"]]
													],
													[
														6,
														[
															7,
															"d",
															"f32"
														],
														[[5, "1"]]
													],
													[
														1,
														[
															7,
															"c",
															"y"
														],
														">=",
														[
															7,
															"c",
															"z"
														]
													]
												]
											]
										],
										"*",
										[
											6,
											[
												7,
												"std",
												"select"
											],
											[
												[
													6,
													[
														7,
														"d",
														"f32"
													],
													[[5, "0"]]
												],
												[
													6,
													[
														7,
														"d",
														"f32"
													],
													[[5, "1"]]
												],
												[
													1,
													[
														7,
														"c",
														"w"
													],
													">",
													[5, "0.02"]
												]
											]
										]
									]
								],
								[
									11,
									[
										1,
										"mode",
										"===",
										[103, "blue"]
									],
									[
										2,
										"score",
										"=",
										[
											1,
											[
												1,
												[
													1,
													[
														6,
														[
															7,
															"std",
															"clamp"
														],
														[
															[
																1,
																[
																	7,
																	"c",
																	"z"
																],
																"/",
																"a"
															],
															[5, "0"],
															[5, "1"]
														]
													],
													"*",
													[
														6,
														[
															7,
															"std",
															"select"
														],
														[
															[
																6,
																[
																	7,
																	"d",
																	"f32"
																],
																[[5, "0"]]
															],
															[
																6,
																[
																	7,
																	"d",
																	"f32"
																],
																[[5, "1"]]
															],
															[
																1,
																[
																	7,
																	"c",
																	"z"
																],
																">=",
																[
																	7,
																	"c",
																	"x"
																]
															]
														]
													]
												],
												"*",
												[
													6,
													[
														7,
														"std",
														"select"
													],
													[
														[
															6,
															[
																7,
																"d",
																"f32"
															],
															[[5, "0"]]
														],
														[
															6,
															[
																7,
																"d",
																"f32"
															],
															[[5, "1"]]
														],
														[
															1,
															[
																7,
																"c",
																"z"
															],
															">=",
															[
																7,
																"c",
																"y"
															]
														]
													]
												]
											],
											"*",
											[
												6,
												[
													7,
													"std",
													"select"
												],
												[
													[
														6,
														[
															7,
															"d",
															"f32"
														],
														[[5, "0"]]
													],
													[
														6,
														[
															7,
															"d",
															"f32"
														],
														[[5, "1"]]
													],
													[
														1,
														[
															7,
															"c",
															"w"
														],
														">",
														[5, "0.02"]
													]
												]
											]
										]
									],
									[
										2,
										"score",
										"=",
										[
											1,
											"lumaU",
											"*",
											[
												6,
												[
													7,
													"std",
													"select"
												],
												[
													[
														6,
														[
															7,
															"d",
															"f32"
														],
														[[5, "0"]]
													],
													[
														6,
														[
															7,
															"d",
															"f32"
														],
														[[5, "1"]]
													],
													[
														1,
														[
															7,
															"c",
															"w"
														],
														">",
														[5, "0.02"]
													]
												]
											]
										]
									]
								]
							]
						]
					]
				],
				[10, "score"]
			]],
			"externalNames": [
				"std",
				"d",
				"mode"
			]
		},
		externals: () => ({
			std,
			d,
			mode
		})
	}) && $.f)({})).$name(`keyframesScore_${mode}`);
}
function buildFeatureKernel(scoreFn) {
	return tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((gx, gy) => {
		"use gpu";
		const srcDims = std.textureDimensions(featureLayout.$.src);
		const srcW = d.f32(srcDims.x);
		const srcH = d.f32(srcDims.y);
		const g = d.f32(FEAT_GRID);
		const cu = __tsover_div(__tsover_add(d.f32(gx), .5), g);
		const cv = __tsover_div(__tsover_add(d.f32(gy), .5), g);
		const q = __tsover_div(.25, g);
		let score = d.f32(0);
		for (let sy = 0; sy < 2; sy++) for (let sx = 0; sx < 2; sx++) {
			const u = __tsover_add(cu, __tsover_mul(__tsover_mul(__tsover_sub(d.f32(sx), .5), 2), q));
			const v = __tsover_add(cv, __tsover_mul(__tsover_mul(__tsover_sub(d.f32(sy), .5), 2), q));
			const tx = d.u32(std.clamp(__tsover_mul(u, srcW), d.f32(0), __tsover_sub(srcW, 1)));
			const ty = d.u32(std.clamp(__tsover_mul(v, srcH), d.f32(0), __tsover_sub(srcH, 1)));
			score = std.max(score, scoreFn(std.textureLoad(featureLayout.$.src, d.vec2u(tx, ty), 0)));
		}
		std.textureStore(featureLayout.$.grid, d.vec2u(gx, gy), d.vec4f(score, 0, 0, 1));
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [{
				"type": "i",
				"name": "gx"
			}, {
				"type": "i",
				"name": "gy"
			}],
			"body": [0, [
				[
					13,
					"srcDims",
					[
						6,
						[
							7,
							"std",
							"textureDimensions"
						],
						[[
							7,
							[
								7,
								"featureLayout",
								"$"
							],
							"src"
						]]
					]
				],
				[
					13,
					"srcW",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[
							7,
							"srcDims",
							"x"
						]]
					]
				],
				[
					13,
					"srcH",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[
							7,
							"srcDims",
							"y"
						]]
					]
				],
				[
					13,
					"g",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["FEAT_GRID"]
					]
				],
				[
					13,
					"cu",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["gx"]
							],
							"+",
							[5, "0.5"]
						],
						"/",
						"g"
					]
				],
				[
					13,
					"cv",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["gy"]
							],
							"+",
							[5, "0.5"]
						],
						"/",
						"g"
					]
				],
				[
					13,
					"q",
					[
						1,
						[5, "0.25"],
						"/",
						"g"
					]
				],
				[
					12,
					"score",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					14,
					[
						12,
						"sy",
						[5, "0"]
					],
					[
						1,
						"sy",
						"<",
						[5, "2"]
					],
					[
						102,
						"++",
						"sy"
					],
					[0, [[
						14,
						[
							12,
							"sx",
							[5, "0"]
						],
						[
							1,
							"sx",
							"<",
							[5, "2"]
						],
						[
							102,
							"++",
							"sx"
						],
						[0, [
							[
								13,
								"u",
								[
									1,
									"cu",
									"+",
									[
										1,
										[
											1,
											[
												1,
												[
													6,
													[
														7,
														"d",
														"f32"
													],
													["sx"]
												],
												"-",
												[5, "0.5"]
											],
											"*",
											[5, "2"]
										],
										"*",
										"q"
									]
								]
							],
							[
								13,
								"v",
								[
									1,
									"cv",
									"+",
									[
										1,
										[
											1,
											[
												1,
												[
													6,
													[
														7,
														"d",
														"f32"
													],
													["sy"]
												],
												"-",
												[5, "0.5"]
											],
											"*",
											[5, "2"]
										],
										"*",
										"q"
									]
								]
							],
							[
								13,
								"tx",
								[
									6,
									[
										7,
										"d",
										"u32"
									],
									[[
										6,
										[
											7,
											"std",
											"clamp"
										],
										[
											[
												1,
												"u",
												"*",
												"srcW"
											],
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												[[5, "0"]]
											],
											[
												1,
												"srcW",
												"-",
												[5, "1"]
											]
										]
									]]
								]
							],
							[
								13,
								"ty",
								[
									6,
									[
										7,
										"d",
										"u32"
									],
									[[
										6,
										[
											7,
											"std",
											"clamp"
										],
										[
											[
												1,
												"v",
												"*",
												"srcH"
											],
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												[[5, "0"]]
											],
											[
												1,
												"srcH",
												"-",
												[5, "1"]
											]
										]
									]]
								]
							],
							[
								2,
								"score",
								"=",
								[
									6,
									[
										7,
										"std",
										"max"
									],
									["score", [
										6,
										"scoreFn",
										[[
											6,
											[
												7,
												"std",
												"textureLoad"
											],
											[
												[
													7,
													[
														7,
														"featureLayout",
														"$"
													],
													"src"
												],
												[
													6,
													[
														7,
														"d",
														"vec2u"
													],
													["tx", "ty"]
												],
												[5, "0"]
											]
										]]
									]]
								]
							]
						]]
					]]]
				],
				[
					6,
					[
						7,
						"std",
						"textureStore"
					],
					[
						[
							7,
							[
								7,
								"featureLayout",
								"$"
							],
							"grid"
						],
						[
							6,
							[
								7,
								"d",
								"vec2u"
							],
							["gx", "gy"]
						],
						[
							6,
							[
								7,
								"d",
								"vec4f"
							],
							[
								"score",
								[5, "0"],
								[5, "0"],
								[5, "1"]
							]
						]
					]
				]
			]],
			"externalNames": [
				"std",
				"featureLayout",
				"d",
				"FEAT_GRID",
				"scoreFn"
			]
		},
		externals: () => ({
			std,
			featureLayout,
			d,
			FEAT_GRID,
			scoreFn
		})
	}) && $.f)({})).$name("keyframesFeatureGrid");
}
function buildTrackKernel(mode, scoreFn) {
	const score = scoreFn ?? buildScoreFn(mode);
	return tgpu.fn([d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((ti) => {
		"use gpu";
		const P = trackLayout.$.params;
		const iF = d.f32(ti);
		const head = std.textureLoad(trackLayout.$.prev, d.vec2u(d.u32(0), ti), 0);
		const meta = std.textureLoad(trackLayout.$.prev, d.vec2u(d.u32(1), ti), 0);
		let posX = head.x;
		let posY = head.y;
		let frame = head.z;
		let genStart = __tsover_sub(head.w, 1);
		let missStreak = meta.x;
		let gen = meta.z;
		let crowdStreak = meta.w;
		if (head.w < .5) {
			posX = __tsover_add(.15, __tsover_mul(.7, trackHash(__tsover_add(__tsover_mul(iF, 7.31), 1.3))));
			posY = __tsover_add(.15, __tsover_mul(.7, trackHash(__tsover_add(__tsover_mul(iF, 11.93), 4.7))));
			frame = d.f32(0);
			genStart = d.f32(0);
			missStreak = d.f32(0);
			gen = d.f32(1);
		}
		frame = __tsover_add(frame, 1);
		const minFeat = std.max(P.threshold, .02);
		const Lpref = std.max(__tsover_sub(1, __tsover_mul(P.variance, trackHash(__tsover_add(__tsover_add(__tsover_mul(iF, 6.13), __tsover_mul(gen, .37)), 2.9)))), minFeat);
		const srcDims = std.textureDimensions(trackLayout.$.src);
		const srcW = d.f32(srcDims.x);
		const srcH = d.f32(srcDims.y);
		const radius = __tsover_add(.035, __tsover_mul(.11, P.agility));
		let sumW = d.f32(0);
		let sumX = d.f32(0);
		let sumY = d.f32(0);
		for (let dy = -3; dy <= 3; dy++) for (let dx = -3; dx <= 3; dx++) {
			const ox = __tsover_mul(__tsover_div(d.f32(dx), 3), radius);
			const oy = __tsover_mul(__tsover_div(d.f32(dy), 3), radius);
			const cx01 = std.clamp(__tsover_add(posX, ox), .02, .98);
			const cy01 = std.clamp(__tsover_add(posY, oy), .02, .98);
			const tx = d.u32(std.clamp(__tsover_mul(cx01, srcW), d.f32(0), __tsover_sub(srcW, 1)));
			const ty = d.u32(std.clamp(__tsover_mul(cy01, srcH), d.f32(0), __tsover_sub(srcH, 1)));
			const s = score(std.textureLoad(trackLayout.$.src, d.vec2u(tx, ty), 0));
			const gate = std.select(d.f32(0), d.f32(1), s > minFeat);
			const dBand = __tsover_sub(s, Lpref);
			const aff = __tsover_mul(std.exp(__tsover_mul(__tsover_mul(dBand, dBand), -24)), gate);
			const off = __tsover_div(std.sqrt(__tsover_add(__tsover_mul(ox, ox), __tsover_mul(oy, oy))), std.max(radius, .001));
			const w = __tsover_mul(aff, __tsover_sub(1, __tsover_mul(.3, off)));
			sumW = __tsover_add(sumW, w);
			sumX = __tsover_add(sumX, __tsover_mul(cx01, w));
			sumY = __tsover_add(sumY, __tsover_mul(cy01, w));
		}
		const locked = sumW > .8;
		let bestX = posX;
		let bestY = posY;
		if (locked) {
			bestX = __tsover_div(sumX, sumW);
			bestY = __tsover_div(sumY, sumW);
		}
		missStreak = std.select(__tsover_add(missStreak, 1), d.f32(0), locked);
		if (!locked) {
			const g = d.f32(FEAT_GRID);
			let bestW = d.f32(0);
			let tgtX = posX;
			let tgtY = posY;
			for (let gy = 0; gy < FEAT_GRID; gy++) for (let gx = 0; gx < FEAT_GRID; gx++) {
				const cs = std.textureLoad(trackLayout.$.featGrid, d.vec2u(d.u32(gx), d.u32(gy)), 0).x;
				const cGate = std.select(d.f32(0), d.f32(1), cs > minFeat);
				const dB = __tsover_sub(cs, Lpref);
				const aff = __tsover_mul(std.exp(__tsover_mul(__tsover_mul(dB, dB), -14)), cGate);
				const cx = __tsover_div(__tsover_add(d.f32(gx), .5), g);
				const cy = __tsover_div(__tsover_add(d.f32(gy), .5), g);
				const ddx = __tsover_sub(cx, posX);
				const ddy = __tsover_sub(cy, posY);
				const pref = __tsover_add(.6, __tsover_mul(.8, trackHash(__tsover_add(__tsover_add(__tsover_mul(iF, 3.17), __tsover_mul(d.f32(gx), 13.71)), __tsover_mul(d.f32(gy), 29.37)))));
				const wgt = __tsover_div(__tsover_mul(aff, pref), __tsover_add(1, __tsover_mul(5, __tsover_add(__tsover_mul(ddx, ddx), __tsover_mul(ddy, ddy)))));
				if (wgt > bestW) {
					bestW = wgt;
					tgtX = cx;
					tgtY = cy;
				}
			}
			if (bestW > .02) {
				bestX = std.clamp(__tsover_add(tgtX, __tsover_div(__tsover_sub(trackHash(__tsover_add(__tsover_mul(iF, 5.77), __tsover_mul(gen, 2.13))), .5), g)), .02, .98);
				bestY = std.clamp(__tsover_add(tgtY, __tsover_div(__tsover_sub(trackHash(__tsover_add(__tsover_mul(iF, 8.31), __tsover_mul(gen, 4.97))), .5), g)), .02, .98);
			} else {
				const ang = __tsover_add(__tsover_mul(trackHash(__tsover_add(__tsover_mul(iF, 13.77), __tsover_mul(gen, 3.19))), TWO_PI), __tsover_mul(std.sin(__tsover_add(__tsover_mul(frame, .02), iF)), 1.4));
				bestX = std.clamp(__tsover_add(posX, __tsover_mul(std.cos(ang), .08)), .03, .97);
				bestY = std.clamp(__tsover_add(posY, __tsover_mul(std.sin(ang), .08)), .03, .97);
			}
		}
		const maxStep = __tsover_mul(P.dt, __tsover_add(.15, __tsover_mul(1.4, P.agility)));
		const dX = __tsover_sub(bestX, posX);
		const dY = __tsover_sub(bestY, posY);
		const dist = std.sqrt(__tsover_add(__tsover_mul(dX, dX), __tsover_mul(dY, dY)));
		const stepK = std.min(d.f32(1), __tsover_div(maxStep, std.max(dist, 1e-5)));
		posX = __tsover_add(posX, __tsover_mul(dX, stepK));
		posY = __tsover_add(posY, __tsover_mul(dY, stepK));
		let crowded = d.f32(0);
		const iN = d.i32(ti);
		for (let k = 0; k < iN; k++) {
			const kF = d.f32(k);
			const other = std.textureLoad(trackLayout.$.prev, d.vec2u(d.u32(0), d.u32(k)), 0);
			const odx = __tsover_sub(other.x, posX);
			const ody = __tsover_sub(other.y, posY);
			const near = std.select(d.f32(0), d.f32(1), std.sqrt(__tsover_add(__tsover_mul(odx, odx), __tsover_mul(ody, ody))) < .022);
			crowded = std.max(crowded, __tsover_mul(__tsover_mul(near, other.w), std.select(d.f32(0), d.f32(1), kF < P.trackersF)));
		}
		crowdStreak = std.select(d.f32(0), __tsover_add(crowdStreak, 1), crowded > .5);
		if (crowdStreak > 30) {
			const hopAng = __tsover_mul(trackHash(__tsover_add(__tsover_add(__tsover_mul(iF, 9.13), __tsover_mul(gen, 5.77)), 2.3)), TWO_PI);
			const hopDist = __tsover_add(.05, __tsover_mul(.08, trackHash(__tsover_add(__tsover_mul(iF, 4.71), __tsover_mul(gen, 8.11)))));
			posX = std.clamp(__tsover_add(posX, __tsover_mul(std.cos(hopAng), hopDist)), .03, .97);
			posY = std.clamp(__tsover_add(posY, __tsover_mul(std.sin(hopAng), hopDist)), .03, .97);
			gen = __tsover_add(gen, 1);
			genStart = frame;
			missStreak = d.f32(0);
			crowdStreak = d.f32(0);
		}
		const lifeFrames = __tsover_mul(__tsover_mul(P.lifespan, 60), __tsover_add(.7, __tsover_mul(.6, trackHash(__tsover_add(__tsover_mul(iF, 4.77), __tsover_mul(gen, 1.31))))));
		const expired = __tsover_mul(std.select(d.f32(0), d.f32(1), __tsover_sub(frame, genStart) > lifeFrames), std.select(d.f32(0), d.f32(1), P.lifespan > .01));
		const lost = std.select(d.f32(0), d.f32(1), missStreak > 90);
		if (__tsover_add(expired, lost) > .5) {
			posX = __tsover_add(.1, __tsover_mul(.8, trackHash(__tsover_add(__tsover_add(__tsover_mul(iF, 7.31), __tsover_mul(gen, 13.7)), __tsover_mul(frame, .013)))));
			posY = __tsover_add(.1, __tsover_mul(.8, trackHash(__tsover_add(__tsover_add(__tsover_mul(iF, 11.93), __tsover_mul(gen, 17.3)), __tsover_mul(frame, .019)))));
			gen = __tsover_add(gen, 1);
			genStart = frame;
			missStreak = d.f32(0);
			crowdStreak = d.f32(0);
		}
		const writeIdx = __tsover_mod(std.floor(__tsover_div(frame, d.f32(TRAIL_EVERY))), d.f32(TRAIL_LEN));
		const shouldWrite = std.select(d.f32(0), d.f32(1), __tsover_mod(frame, d.f32(TRAIL_EVERY)) < .5);
		let bbMinX = posX;
		let bbMinY = posY;
		let bbMaxX = posX;
		let bbMaxY = posY;
		for (let j = 0; j < TRAIL_LEN; j++) {
			const jF = d.f32(j);
			let e = std.textureLoad(trackLayout.$.prev, d.vec2u(d.u32(__tsover_add(2, j)), ti), 0);
			const isSlot = std.select(d.f32(0), d.f32(1), std.abs(__tsover_sub(jF, writeIdx)) < .5);
			const doWrite = __tsover_mul(isSlot, shouldWrite);
			e = std.mix(e, d.vec4f(posX, posY, frame, gen), doWrite);
			std.textureStore(trackLayout.$.next, d.vec2u(d.u32(__tsover_add(2, j)), ti), e);
			std.textureStore(trackLayout.$.present, d.vec2u(d.u32(__tsover_add(2, j)), ti), e);
			const live = __tsover_mul(std.select(d.f32(0), d.f32(1), std.abs(__tsover_sub(e.w, gen)) < .5), std.select(d.f32(0), d.f32(1), e.z > .5));
			bbMinX = std.min(bbMinX, std.mix(posX, e.x, live));
			bbMinY = std.min(bbMinY, std.mix(posY, e.y, live));
			bbMaxX = std.max(bbMaxX, std.mix(posX, e.x, live));
			bbMaxY = std.max(bbMaxY, std.mix(posY, e.y, live));
		}
		const bbOut = d.vec4f(bbMinX, bbMinY, bbMaxX, bbMaxY);
		std.textureStore(trackLayout.$.next, d.vec2u(d.u32(AABB_COL), ti), bbOut);
		std.textureStore(trackLayout.$.present, d.vec2u(d.u32(AABB_COL), ti), bbOut);
		const headOut = d.vec4f(posX, posY, frame, __tsover_add(genStart, 1));
		const metaOut = d.vec4f(missStreak, sumW, gen, crowdStreak);
		std.textureStore(trackLayout.$.next, d.vec2u(d.u32(0), ti), headOut);
		std.textureStore(trackLayout.$.present, d.vec2u(d.u32(0), ti), headOut);
		std.textureStore(trackLayout.$.next, d.vec2u(d.u32(1), ti), metaOut);
		std.textureStore(trackLayout.$.present, d.vec2u(d.u32(1), ti), metaOut);
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [{
				"type": "i",
				"name": "ti"
			}],
			"body": [0, [
				[
					13,
					"P",
					[
						7,
						[
							7,
							"trackLayout",
							"$"
						],
						"params"
					]
				],
				[
					13,
					"iF",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["ti"]
					]
				],
				[
					13,
					"head",
					[
						6,
						[
							7,
							"std",
							"textureLoad"
						],
						[
							[
								7,
								[
									7,
									"trackLayout",
									"$"
								],
								"prev"
							],
							[
								6,
								[
									7,
									"d",
									"vec2u"
								],
								[[
									6,
									[
										7,
										"d",
										"u32"
									],
									[[5, "0"]]
								], "ti"]
							],
							[5, "0"]
						]
					]
				],
				[
					13,
					"meta",
					[
						6,
						[
							7,
							"std",
							"textureLoad"
						],
						[
							[
								7,
								[
									7,
									"trackLayout",
									"$"
								],
								"prev"
							],
							[
								6,
								[
									7,
									"d",
									"vec2u"
								],
								[[
									6,
									[
										7,
										"d",
										"u32"
									],
									[[5, "1"]]
								], "ti"]
							],
							[5, "0"]
						]
					]
				],
				[
					12,
					"posX",
					[
						7,
						"head",
						"x"
					]
				],
				[
					12,
					"posY",
					[
						7,
						"head",
						"y"
					]
				],
				[
					12,
					"frame",
					[
						7,
						"head",
						"z"
					]
				],
				[
					12,
					"genStart",
					[
						1,
						[
							7,
							"head",
							"w"
						],
						"-",
						[5, "1"]
					]
				],
				[
					12,
					"missStreak",
					[
						7,
						"meta",
						"x"
					]
				],
				[
					12,
					"gen",
					[
						7,
						"meta",
						"z"
					]
				],
				[
					12,
					"crowdStreak",
					[
						7,
						"meta",
						"w"
					]
				],
				[
					11,
					[
						1,
						[
							7,
							"head",
							"w"
						],
						"<",
						[5, "0.5"]
					],
					[0, [
						[
							2,
							"posX",
							"=",
							[
								1,
								[5, "0.15"],
								"+",
								[
									1,
									[5, "0.7"],
									"*",
									[
										6,
										"trackHash",
										[[
											1,
											[
												1,
												"iF",
												"*",
												[5, "7.31"]
											],
											"+",
											[5, "1.3"]
										]]
									]
								]
							]
						],
						[
							2,
							"posY",
							"=",
							[
								1,
								[5, "0.15"],
								"+",
								[
									1,
									[5, "0.7"],
									"*",
									[
										6,
										"trackHash",
										[[
											1,
											[
												1,
												"iF",
												"*",
												[5, "11.93"]
											],
											"+",
											[5, "4.7"]
										]]
									]
								]
							]
						],
						[
							2,
							"frame",
							"=",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							]
						],
						[
							2,
							"genStart",
							"=",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							]
						],
						[
							2,
							"missStreak",
							"=",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							]
						],
						[
							2,
							"gen",
							"=",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "1"]]
							]
						]
					]]
				],
				[
					2,
					"frame",
					"=",
					[
						1,
						"frame",
						"+",
						[5, "1"]
					]
				],
				[
					13,
					"minFeat",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							7,
							"P",
							"threshold"
						], [5, "0.02"]]
					]
				],
				[
					13,
					"Lpref",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							1,
							[5, "1"],
							"-",
							[
								1,
								[
									7,
									"P",
									"variance"
								],
								"*",
								[
									6,
									"trackHash",
									[[
										1,
										[
											1,
											[
												1,
												"iF",
												"*",
												[5, "6.13"]
											],
											"+",
											[
												1,
												"gen",
												"*",
												[5, "0.37"]
											]
										],
										"+",
										[5, "2.9"]
									]]
								]
							]
						], "minFeat"]
					]
				],
				[
					13,
					"srcDims",
					[
						6,
						[
							7,
							"std",
							"textureDimensions"
						],
						[[
							7,
							[
								7,
								"trackLayout",
								"$"
							],
							"src"
						]]
					]
				],
				[
					13,
					"srcW",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[
							7,
							"srcDims",
							"x"
						]]
					]
				],
				[
					13,
					"srcH",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[
							7,
							"srcDims",
							"y"
						]]
					]
				],
				[
					13,
					"radius",
					[
						1,
						[5, "0.035"],
						"+",
						[
							1,
							[5, "0.11"],
							"*",
							[
								7,
								"P",
								"agility"
							]
						]
					]
				],
				[
					12,
					"sumW",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					12,
					"sumX",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					12,
					"sumY",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					14,
					[
						12,
						"dy",
						[
							4,
							"-",
							[5, "3"]
						]
					],
					[
						1,
						"dy",
						"<=",
						[5, "3"]
					],
					[
						102,
						"++",
						"dy"
					],
					[0, [[
						14,
						[
							12,
							"dx",
							[
								4,
								"-",
								[5, "3"]
							]
						],
						[
							1,
							"dx",
							"<=",
							[5, "3"]
						],
						[
							102,
							"++",
							"dx"
						],
						[0, [
							[
								13,
								"ox",
								[
									1,
									[
										1,
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											["dx"]
										],
										"/",
										[5, "3"]
									],
									"*",
									"radius"
								]
							],
							[
								13,
								"oy",
								[
									1,
									[
										1,
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											["dy"]
										],
										"/",
										[5, "3"]
									],
									"*",
									"radius"
								]
							],
							[
								13,
								"cx01",
								[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										[
											1,
											"posX",
											"+",
											"ox"
										],
										[5, "0.02"],
										[5, "0.98"]
									]
								]
							],
							[
								13,
								"cy01",
								[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										[
											1,
											"posY",
											"+",
											"oy"
										],
										[5, "0.02"],
										[5, "0.98"]
									]
								]
							],
							[
								13,
								"tx",
								[
									6,
									[
										7,
										"d",
										"u32"
									],
									[[
										6,
										[
											7,
											"std",
											"clamp"
										],
										[
											[
												1,
												"cx01",
												"*",
												"srcW"
											],
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												[[5, "0"]]
											],
											[
												1,
												"srcW",
												"-",
												[5, "1"]
											]
										]
									]]
								]
							],
							[
								13,
								"ty",
								[
									6,
									[
										7,
										"d",
										"u32"
									],
									[[
										6,
										[
											7,
											"std",
											"clamp"
										],
										[
											[
												1,
												"cy01",
												"*",
												"srcH"
											],
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												[[5, "0"]]
											],
											[
												1,
												"srcH",
												"-",
												[5, "1"]
											]
										]
									]]
								]
							],
							[
								13,
								"s",
								[
									6,
									"score",
									[[
										6,
										[
											7,
											"std",
											"textureLoad"
										],
										[
											[
												7,
												[
													7,
													"trackLayout",
													"$"
												],
												"src"
											],
											[
												6,
												[
													7,
													"d",
													"vec2u"
												],
												["tx", "ty"]
											],
											[5, "0"]
										]
									]]
								]
							],
							[
								13,
								"gate",
								[
									6,
									[
										7,
										"std",
										"select"
									],
									[
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											[[5, "0"]]
										],
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											[[5, "1"]]
										],
										[
											1,
											"s",
											">",
											"minFeat"
										]
									]
								]
							],
							[
								13,
								"dBand",
								[
									1,
									"s",
									"-",
									"Lpref"
								]
							],
							[
								13,
								"aff",
								[
									1,
									[
										6,
										[
											7,
											"std",
											"exp"
										],
										[[
											1,
											[
												1,
												"dBand",
												"*",
												"dBand"
											],
											"*",
											[
												4,
												"-",
												[5, "24"]
											]
										]]
									],
									"*",
									"gate"
								]
							],
							[
								13,
								"off",
								[
									1,
									[
										6,
										[
											7,
											"std",
											"sqrt"
										],
										[[
											1,
											[
												1,
												"ox",
												"*",
												"ox"
											],
											"+",
											[
												1,
												"oy",
												"*",
												"oy"
											]
										]]
									],
									"/",
									[
										6,
										[
											7,
											"std",
											"max"
										],
										["radius", [5, "0.001"]]
									]
								]
							],
							[
								13,
								"w",
								[
									1,
									"aff",
									"*",
									[
										1,
										[5, "1"],
										"-",
										[
											1,
											[5, "0.3"],
											"*",
											"off"
										]
									]
								]
							],
							[
								2,
								"sumW",
								"=",
								[
									1,
									"sumW",
									"+",
									"w"
								]
							],
							[
								2,
								"sumX",
								"=",
								[
									1,
									"sumX",
									"+",
									[
										1,
										"cx01",
										"*",
										"w"
									]
								]
							],
							[
								2,
								"sumY",
								"=",
								[
									1,
									"sumY",
									"+",
									[
										1,
										"cy01",
										"*",
										"w"
									]
								]
							]
						]]
					]]]
				],
				[
					13,
					"locked",
					[
						1,
						"sumW",
						">",
						[5, "0.8"]
					]
				],
				[
					12,
					"bestX",
					"posX"
				],
				[
					12,
					"bestY",
					"posY"
				],
				[
					11,
					"locked",
					[0, [[
						2,
						"bestX",
						"=",
						[
							1,
							"sumX",
							"/",
							"sumW"
						]
					], [
						2,
						"bestY",
						"=",
						[
							1,
							"sumY",
							"/",
							"sumW"
						]
					]]]
				],
				[
					2,
					"missStreak",
					"=",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								1,
								"missStreak",
								"+",
								[5, "1"]
							],
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							],
							"locked"
						]
					]
				],
				[
					11,
					[
						4,
						"!",
						"locked"
					],
					[0, [
						[
							13,
							"g",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["FEAT_GRID"]
							]
						],
						[
							12,
							"bestW",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							]
						],
						[
							12,
							"tgtX",
							"posX"
						],
						[
							12,
							"tgtY",
							"posY"
						],
						[
							14,
							[
								12,
								"gy",
								[5, "0"]
							],
							[
								1,
								"gy",
								"<",
								"FEAT_GRID"
							],
							[
								102,
								"++",
								"gy"
							],
							[0, [[
								14,
								[
									12,
									"gx",
									[5, "0"]
								],
								[
									1,
									"gx",
									"<",
									"FEAT_GRID"
								],
								[
									102,
									"++",
									"gx"
								],
								[0, [
									[
										13,
										"cs",
										[
											7,
											[
												6,
												[
													7,
													"std",
													"textureLoad"
												],
												[
													[
														7,
														[
															7,
															"trackLayout",
															"$"
														],
														"featGrid"
													],
													[
														6,
														[
															7,
															"d",
															"vec2u"
														],
														[[
															6,
															[
																7,
																"d",
																"u32"
															],
															["gx"]
														], [
															6,
															[
																7,
																"d",
																"u32"
															],
															["gy"]
														]]
													],
													[5, "0"]
												]
											],
											"x"
										]
									],
									[
										13,
										"cGate",
										[
											6,
											[
												7,
												"std",
												"select"
											],
											[
												[
													6,
													[
														7,
														"d",
														"f32"
													],
													[[5, "0"]]
												],
												[
													6,
													[
														7,
														"d",
														"f32"
													],
													[[5, "1"]]
												],
												[
													1,
													"cs",
													">",
													"minFeat"
												]
											]
										]
									],
									[
										13,
										"dB",
										[
											1,
											"cs",
											"-",
											"Lpref"
										]
									],
									[
										13,
										"aff",
										[
											1,
											[
												6,
												[
													7,
													"std",
													"exp"
												],
												[[
													1,
													[
														1,
														"dB",
														"*",
														"dB"
													],
													"*",
													[
														4,
														"-",
														[5, "14"]
													]
												]]
											],
											"*",
											"cGate"
										]
									],
									[
										13,
										"cx",
										[
											1,
											[
												1,
												[
													6,
													[
														7,
														"d",
														"f32"
													],
													["gx"]
												],
												"+",
												[5, "0.5"]
											],
											"/",
											"g"
										]
									],
									[
										13,
										"cy",
										[
											1,
											[
												1,
												[
													6,
													[
														7,
														"d",
														"f32"
													],
													["gy"]
												],
												"+",
												[5, "0.5"]
											],
											"/",
											"g"
										]
									],
									[
										13,
										"ddx",
										[
											1,
											"cx",
											"-",
											"posX"
										]
									],
									[
										13,
										"ddy",
										[
											1,
											"cy",
											"-",
											"posY"
										]
									],
									[
										13,
										"pref",
										[
											1,
											[5, "0.6"],
											"+",
											[
												1,
												[5, "0.8"],
												"*",
												[
													6,
													"trackHash",
													[[
														1,
														[
															1,
															[
																1,
																"iF",
																"*",
																[5, "3.17"]
															],
															"+",
															[
																1,
																[
																	6,
																	[
																		7,
																		"d",
																		"f32"
																	],
																	["gx"]
																],
																"*",
																[5, "13.71"]
															]
														],
														"+",
														[
															1,
															[
																6,
																[
																	7,
																	"d",
																	"f32"
																],
																["gy"]
															],
															"*",
															[5, "29.37"]
														]
													]]
												]
											]
										]
									],
									[
										13,
										"wgt",
										[
											1,
											[
												1,
												"aff",
												"*",
												"pref"
											],
											"/",
											[
												1,
												[5, "1"],
												"+",
												[
													1,
													[5, "5"],
													"*",
													[
														1,
														[
															1,
															"ddx",
															"*",
															"ddx"
														],
														"+",
														[
															1,
															"ddy",
															"*",
															"ddy"
														]
													]
												]
											]
										]
									],
									[
										11,
										[
											1,
											"wgt",
											">",
											"bestW"
										],
										[0, [
											[
												2,
												"bestW",
												"=",
												"wgt"
											],
											[
												2,
												"tgtX",
												"=",
												"cx"
											],
											[
												2,
												"tgtY",
												"=",
												"cy"
											]
										]]
									]
								]]
							]]]
						],
						[
							11,
							[
								1,
								"bestW",
								">",
								[5, "0.02"]
							],
							[0, [[
								2,
								"bestX",
								"=",
								[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										[
											1,
											"tgtX",
											"+",
											[
												1,
												[
													1,
													[
														6,
														"trackHash",
														[[
															1,
															[
																1,
																"iF",
																"*",
																[5, "5.77"]
															],
															"+",
															[
																1,
																"gen",
																"*",
																[5, "2.13"]
															]
														]]
													],
													"-",
													[5, "0.5"]
												],
												"/",
												"g"
											]
										],
										[5, "0.02"],
										[5, "0.98"]
									]
								]
							], [
								2,
								"bestY",
								"=",
								[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										[
											1,
											"tgtY",
											"+",
											[
												1,
												[
													1,
													[
														6,
														"trackHash",
														[[
															1,
															[
																1,
																"iF",
																"*",
																[5, "8.31"]
															],
															"+",
															[
																1,
																"gen",
																"*",
																[5, "4.97"]
															]
														]]
													],
													"-",
													[5, "0.5"]
												],
												"/",
												"g"
											]
										],
										[5, "0.02"],
										[5, "0.98"]
									]
								]
							]]],
							[0, [
								[
									13,
									"ang",
									[
										1,
										[
											1,
											[
												6,
												"trackHash",
												[[
													1,
													[
														1,
														"iF",
														"*",
														[5, "13.77"]
													],
													"+",
													[
														1,
														"gen",
														"*",
														[5, "3.19"]
													]
												]]
											],
											"*",
											"TWO_PI"
										],
										"+",
										[
											1,
											[
												6,
												[
													7,
													"std",
													"sin"
												],
												[[
													1,
													[
														1,
														"frame",
														"*",
														[5, "0.02"]
													],
													"+",
													"iF"
												]]
											],
											"*",
											[5, "1.4"]
										]
									]
								],
								[
									2,
									"bestX",
									"=",
									[
										6,
										[
											7,
											"std",
											"clamp"
										],
										[
											[
												1,
												"posX",
												"+",
												[
													1,
													[
														6,
														[
															7,
															"std",
															"cos"
														],
														["ang"]
													],
													"*",
													[5, "0.08"]
												]
											],
											[5, "0.03"],
											[5, "0.97"]
										]
									]
								],
								[
									2,
									"bestY",
									"=",
									[
										6,
										[
											7,
											"std",
											"clamp"
										],
										[
											[
												1,
												"posY",
												"+",
												[
													1,
													[
														6,
														[
															7,
															"std",
															"sin"
														],
														["ang"]
													],
													"*",
													[5, "0.08"]
												]
											],
											[5, "0.03"],
											[5, "0.97"]
										]
									]
								]
							]]
						]
					]]
				],
				[
					13,
					"maxStep",
					[
						1,
						[
							7,
							"P",
							"dt"
						],
						"*",
						[
							1,
							[5, "0.15"],
							"+",
							[
								1,
								[5, "1.4"],
								"*",
								[
									7,
									"P",
									"agility"
								]
							]
						]
					]
				],
				[
					13,
					"dX",
					[
						1,
						"bestX",
						"-",
						"posX"
					]
				],
				[
					13,
					"dY",
					[
						1,
						"bestY",
						"-",
						"posY"
					]
				],
				[
					13,
					"dist",
					[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"dX",
								"*",
								"dX"
							],
							"+",
							[
								1,
								"dY",
								"*",
								"dY"
							]
						]]
					]
				],
				[
					13,
					"stepK",
					[
						6,
						[
							7,
							"std",
							"min"
						],
						[[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						], [
							1,
							"maxStep",
							"/",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["dist", [5, "0.00001"]]
							]
						]]
					]
				],
				[
					2,
					"posX",
					"=",
					[
						1,
						"posX",
						"+",
						[
							1,
							"dX",
							"*",
							"stepK"
						]
					]
				],
				[
					2,
					"posY",
					"=",
					[
						1,
						"posY",
						"+",
						[
							1,
							"dY",
							"*",
							"stepK"
						]
					]
				],
				[
					12,
					"crowded",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					]
				],
				[
					13,
					"iN",
					[
						6,
						[
							7,
							"d",
							"i32"
						],
						["ti"]
					]
				],
				[
					14,
					[
						12,
						"k",
						[5, "0"]
					],
					[
						1,
						"k",
						"<",
						"iN"
					],
					[
						102,
						"++",
						"k"
					],
					[0, [
						[
							13,
							"kF",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["k"]
							]
						],
						[
							13,
							"other",
							[
								6,
								[
									7,
									"std",
									"textureLoad"
								],
								[
									[
										7,
										[
											7,
											"trackLayout",
											"$"
										],
										"prev"
									],
									[
										6,
										[
											7,
											"d",
											"vec2u"
										],
										[[
											6,
											[
												7,
												"d",
												"u32"
											],
											[[5, "0"]]
										], [
											6,
											[
												7,
												"d",
												"u32"
											],
											["k"]
										]]
									],
									[5, "0"]
								]
							]
						],
						[
							13,
							"odx",
							[
								1,
								[
									7,
									"other",
									"x"
								],
								"-",
								"posX"
							]
						],
						[
							13,
							"ody",
							[
								1,
								[
									7,
									"other",
									"y"
								],
								"-",
								"posY"
							]
						],
						[
							13,
							"near",
							[
								6,
								[
									7,
									"std",
									"select"
								],
								[
									[
										6,
										[
											7,
											"d",
											"f32"
										],
										[[5, "0"]]
									],
									[
										6,
										[
											7,
											"d",
											"f32"
										],
										[[5, "1"]]
									],
									[
										1,
										[
											6,
											[
												7,
												"std",
												"sqrt"
											],
											[[
												1,
												[
													1,
													"odx",
													"*",
													"odx"
												],
												"+",
												[
													1,
													"ody",
													"*",
													"ody"
												]
											]]
										],
										"<",
										[5, "0.022"]
									]
								]
							]
						],
						[
							2,
							"crowded",
							"=",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["crowded", [
									1,
									[
										1,
										"near",
										"*",
										[
											7,
											"other",
											"w"
										]
									],
									"*",
									[
										6,
										[
											7,
											"std",
											"select"
										],
										[
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												[[5, "0"]]
											],
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												[[5, "1"]]
											],
											[
												1,
												"kF",
												"<",
												[
													7,
													"P",
													"trackersF"
												]
											]
										]
									]
								]]
							]
						]
					]]
				],
				[
					2,
					"crowdStreak",
					"=",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							],
							[
								1,
								"crowdStreak",
								"+",
								[5, "1"]
							],
							[
								1,
								"crowded",
								">",
								[5, "0.5"]
							]
						]
					]
				],
				[
					11,
					[
						1,
						"crowdStreak",
						">",
						[5, "30"]
					],
					[0, [
						[
							13,
							"hopAng",
							[
								1,
								[
									6,
									"trackHash",
									[[
										1,
										[
											1,
											[
												1,
												"iF",
												"*",
												[5, "9.13"]
											],
											"+",
											[
												1,
												"gen",
												"*",
												[5, "5.77"]
											]
										],
										"+",
										[5, "2.3"]
									]]
								],
								"*",
								"TWO_PI"
							]
						],
						[
							13,
							"hopDist",
							[
								1,
								[5, "0.05"],
								"+",
								[
									1,
									[5, "0.08"],
									"*",
									[
										6,
										"trackHash",
										[[
											1,
											[
												1,
												"iF",
												"*",
												[5, "4.71"]
											],
											"+",
											[
												1,
												"gen",
												"*",
												[5, "8.11"]
											]
										]]
									]
								]
							]
						],
						[
							2,
							"posX",
							"=",
							[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									[
										1,
										"posX",
										"+",
										[
											1,
											[
												6,
												[
													7,
													"std",
													"cos"
												],
												["hopAng"]
											],
											"*",
											"hopDist"
										]
									],
									[5, "0.03"],
									[5, "0.97"]
								]
							]
						],
						[
							2,
							"posY",
							"=",
							[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									[
										1,
										"posY",
										"+",
										[
											1,
											[
												6,
												[
													7,
													"std",
													"sin"
												],
												["hopAng"]
											],
											"*",
											"hopDist"
										]
									],
									[5, "0.03"],
									[5, "0.97"]
								]
							]
						],
						[
							2,
							"gen",
							"=",
							[
								1,
								"gen",
								"+",
								[5, "1"]
							]
						],
						[
							2,
							"genStart",
							"=",
							"frame"
						],
						[
							2,
							"missStreak",
							"=",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							]
						],
						[
							2,
							"crowdStreak",
							"=",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							]
						]
					]]
				],
				[
					13,
					"lifeFrames",
					[
						1,
						[
							1,
							[
								7,
								"P",
								"lifespan"
							],
							"*",
							[5, "60"]
						],
						"*",
						[
							1,
							[5, "0.7"],
							"+",
							[
								1,
								[5, "0.6"],
								"*",
								[
									6,
									"trackHash",
									[[
										1,
										[
											1,
											"iF",
											"*",
											[5, "4.77"]
										],
										"+",
										[
											1,
											"gen",
											"*",
											[5, "1.31"]
										]
									]]
								]
							]
						]
					]
				],
				[
					13,
					"expired",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"select"
							],
							[
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "0"]]
								],
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "1"]]
								],
								[
									1,
									[
										1,
										"frame",
										"-",
										"genStart"
									],
									">",
									"lifeFrames"
								]
							]
						],
						"*",
						[
							6,
							[
								7,
								"std",
								"select"
							],
							[
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "0"]]
								],
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "1"]]
								],
								[
									1,
									[
										7,
										"P",
										"lifespan"
									],
									">",
									[5, "0.01"]
								]
							]
						]
					]
				],
				[
					13,
					"lost",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							],
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "1"]]
							],
							[
								1,
								"missStreak",
								">",
								[5, "90"]
							]
						]
					]
				],
				[
					11,
					[
						1,
						[
							1,
							"expired",
							"+",
							"lost"
						],
						">",
						[5, "0.5"]
					],
					[0, [
						[
							2,
							"posX",
							"=",
							[
								1,
								[5, "0.1"],
								"+",
								[
									1,
									[5, "0.8"],
									"*",
									[
										6,
										"trackHash",
										[[
											1,
											[
												1,
												[
													1,
													"iF",
													"*",
													[5, "7.31"]
												],
												"+",
												[
													1,
													"gen",
													"*",
													[5, "13.7"]
												]
											],
											"+",
											[
												1,
												"frame",
												"*",
												[5, "0.013"]
											]
										]]
									]
								]
							]
						],
						[
							2,
							"posY",
							"=",
							[
								1,
								[5, "0.1"],
								"+",
								[
									1,
									[5, "0.8"],
									"*",
									[
										6,
										"trackHash",
										[[
											1,
											[
												1,
												[
													1,
													"iF",
													"*",
													[5, "11.93"]
												],
												"+",
												[
													1,
													"gen",
													"*",
													[5, "17.3"]
												]
											],
											"+",
											[
												1,
												"frame",
												"*",
												[5, "0.019"]
											]
										]]
									]
								]
							]
						],
						[
							2,
							"gen",
							"=",
							[
								1,
								"gen",
								"+",
								[5, "1"]
							]
						],
						[
							2,
							"genStart",
							"=",
							"frame"
						],
						[
							2,
							"missStreak",
							"=",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							]
						],
						[
							2,
							"crowdStreak",
							"=",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							]
						]
					]]
				],
				[
					13,
					"writeIdx",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								1,
								"frame",
								"/",
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									["TRAIL_EVERY"]
								]
							]]
						],
						"%",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["TRAIL_LEN"]
						]
					]
				],
				[
					13,
					"shouldWrite",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							],
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "1"]]
							],
							[
								1,
								[
									1,
									"frame",
									"%",
									[
										6,
										[
											7,
											"d",
											"f32"
										],
										["TRAIL_EVERY"]
									]
								],
								"<",
								[5, "0.5"]
							]
						]
					]
				],
				[
					12,
					"bbMinX",
					"posX"
				],
				[
					12,
					"bbMinY",
					"posY"
				],
				[
					12,
					"bbMaxX",
					"posX"
				],
				[
					12,
					"bbMaxY",
					"posY"
				],
				[
					14,
					[
						12,
						"j",
						[5, "0"]
					],
					[
						1,
						"j",
						"<",
						"TRAIL_LEN"
					],
					[
						102,
						"++",
						"j"
					],
					[0, [
						[
							13,
							"jF",
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["j"]
							]
						],
						[
							12,
							"e",
							[
								6,
								[
									7,
									"std",
									"textureLoad"
								],
								[
									[
										7,
										[
											7,
											"trackLayout",
											"$"
										],
										"prev"
									],
									[
										6,
										[
											7,
											"d",
											"vec2u"
										],
										[[
											6,
											[
												7,
												"d",
												"u32"
											],
											[[
												1,
												[5, "2"],
												"+",
												"j"
											]]
										], "ti"]
									],
									[5, "0"]
								]
							]
						],
						[
							13,
							"isSlot",
							[
								6,
								[
									7,
									"std",
									"select"
								],
								[
									[
										6,
										[
											7,
											"d",
											"f32"
										],
										[[5, "0"]]
									],
									[
										6,
										[
											7,
											"d",
											"f32"
										],
										[[5, "1"]]
									],
									[
										1,
										[
											6,
											[
												7,
												"std",
												"abs"
											],
											[[
												1,
												"jF",
												"-",
												"writeIdx"
											]]
										],
										"<",
										[5, "0.5"]
									]
								]
							]
						],
						[
							13,
							"doWrite",
							[
								1,
								"isSlot",
								"*",
								"shouldWrite"
							]
						],
						[
							2,
							"e",
							"=",
							[
								6,
								[
									7,
									"std",
									"mix"
								],
								[
									"e",
									[
										6,
										[
											7,
											"d",
											"vec4f"
										],
										[
											"posX",
											"posY",
											"frame",
											"gen"
										]
									],
									"doWrite"
								]
							]
						],
						[
							6,
							[
								7,
								"std",
								"textureStore"
							],
							[
								[
									7,
									[
										7,
										"trackLayout",
										"$"
									],
									"next"
								],
								[
									6,
									[
										7,
										"d",
										"vec2u"
									],
									[[
										6,
										[
											7,
											"d",
											"u32"
										],
										[[
											1,
											[5, "2"],
											"+",
											"j"
										]]
									], "ti"]
								],
								"e"
							]
						],
						[
							6,
							[
								7,
								"std",
								"textureStore"
							],
							[
								[
									7,
									[
										7,
										"trackLayout",
										"$"
									],
									"present"
								],
								[
									6,
									[
										7,
										"d",
										"vec2u"
									],
									[[
										6,
										[
											7,
											"d",
											"u32"
										],
										[[
											1,
											[5, "2"],
											"+",
											"j"
										]]
									], "ti"]
								],
								"e"
							]
						],
						[
							13,
							"live",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"select"
									],
									[
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											[[5, "0"]]
										],
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											[[5, "1"]]
										],
										[
											1,
											[
												6,
												[
													7,
													"std",
													"abs"
												],
												[[
													1,
													[
														7,
														"e",
														"w"
													],
													"-",
													"gen"
												]]
											],
											"<",
											[5, "0.5"]
										]
									]
								],
								"*",
								[
									6,
									[
										7,
										"std",
										"select"
									],
									[
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											[[5, "0"]]
										],
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											[[5, "1"]]
										],
										[
											1,
											[
												7,
												"e",
												"z"
											],
											">",
											[5, "0.5"]
										]
									]
								]
							]
						],
						[
							2,
							"bbMinX",
							"=",
							[
								6,
								[
									7,
									"std",
									"min"
								],
								["bbMinX", [
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										"posX",
										[
											7,
											"e",
											"x"
										],
										"live"
									]
								]]
							]
						],
						[
							2,
							"bbMinY",
							"=",
							[
								6,
								[
									7,
									"std",
									"min"
								],
								["bbMinY", [
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										"posY",
										[
											7,
											"e",
											"y"
										],
										"live"
									]
								]]
							]
						],
						[
							2,
							"bbMaxX",
							"=",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["bbMaxX", [
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										"posX",
										[
											7,
											"e",
											"x"
										],
										"live"
									]
								]]
							]
						],
						[
							2,
							"bbMaxY",
							"=",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["bbMaxY", [
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										"posY",
										[
											7,
											"e",
											"y"
										],
										"live"
									]
								]]
							]
						]
					]]
				],
				[
					13,
					"bbOut",
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							"bbMinX",
							"bbMinY",
							"bbMaxX",
							"bbMaxY"
						]
					]
				],
				[
					6,
					[
						7,
						"std",
						"textureStore"
					],
					[
						[
							7,
							[
								7,
								"trackLayout",
								"$"
							],
							"next"
						],
						[
							6,
							[
								7,
								"d",
								"vec2u"
							],
							[[
								6,
								[
									7,
									"d",
									"u32"
								],
								["AABB_COL"]
							], "ti"]
						],
						"bbOut"
					]
				],
				[
					6,
					[
						7,
						"std",
						"textureStore"
					],
					[
						[
							7,
							[
								7,
								"trackLayout",
								"$"
							],
							"present"
						],
						[
							6,
							[
								7,
								"d",
								"vec2u"
							],
							[[
								6,
								[
									7,
									"d",
									"u32"
								],
								["AABB_COL"]
							], "ti"]
						],
						"bbOut"
					]
				],
				[
					13,
					"headOut",
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							"posX",
							"posY",
							"frame",
							[
								1,
								"genStart",
								"+",
								[5, "1"]
							]
						]
					]
				],
				[
					13,
					"metaOut",
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							"missStreak",
							"sumW",
							"gen",
							"crowdStreak"
						]
					]
				],
				[
					6,
					[
						7,
						"std",
						"textureStore"
					],
					[
						[
							7,
							[
								7,
								"trackLayout",
								"$"
							],
							"next"
						],
						[
							6,
							[
								7,
								"d",
								"vec2u"
							],
							[[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "0"]]
							], "ti"]
						],
						"headOut"
					]
				],
				[
					6,
					[
						7,
						"std",
						"textureStore"
					],
					[
						[
							7,
							[
								7,
								"trackLayout",
								"$"
							],
							"present"
						],
						[
							6,
							[
								7,
								"d",
								"vec2u"
							],
							[[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "0"]]
							], "ti"]
						],
						"headOut"
					]
				],
				[
					6,
					[
						7,
						"std",
						"textureStore"
					],
					[
						[
							7,
							[
								7,
								"trackLayout",
								"$"
							],
							"next"
						],
						[
							6,
							[
								7,
								"d",
								"vec2u"
							],
							[[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							], "ti"]
						],
						"metaOut"
					]
				],
				[
					6,
					[
						7,
						"std",
						"textureStore"
					],
					[
						[
							7,
							[
								7,
								"trackLayout",
								"$"
							],
							"present"
						],
						[
							6,
							[
								7,
								"d",
								"vec2u"
							],
							[[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							], "ti"]
						],
						"metaOut"
					]
				]
			]],
			"externalNames": [
				"trackLayout",
				"d",
				"std",
				"trackHash",
				"score",
				"FEAT_GRID",
				"TWO_PI",
				"TRAIL_EVERY",
				"TRAIL_LEN",
				"AABB_COL"
			]
		},
		externals: () => ({
			trackLayout,
			d,
			std,
			trackHash,
			score,
			FEAT_GRID,
			TWO_PI,
			TRAIL_EVERY,
			TRAIL_LEN,
			AABB_COL
		})
	}) && $.f)({})).$name("keyframesTrack");
}
const keyframesFeatureKernelBright = buildFeatureKernel(buildScoreFn("bright"));
const componentDefinition = {
	name: "KeyFrames",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Motion-tracking overlay — trackers hunt and follow features in the content, dropping keyframe diamonds behind them along their motion paths, like a compositor mid-track",
	requiresRTT: true,
	requiresChild: true,
	props: {
		trackers: {
			default: 12,
			description: "Number of simultaneous trackers.",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 64,
				step: 1,
				label: "Trackers",
				group: "Tracking"
			}
		},
		detect: {
			default: "bright",
			compileTime: true,
			description: "What the trackers hunt for.",
			ui: {
				type: "select",
				options: [
					{
						label: "Brightness",
						value: "bright"
					},
					{
						label: "Darkness",
						value: "dark"
					},
					{
						label: "Transparency",
						value: "alpha"
					},
					{
						label: "Red",
						value: "red"
					},
					{
						label: "Green",
						value: "green"
					},
					{
						label: "Blue",
						value: "blue"
					}
				],
				label: "Detect",
				group: "Tracking"
			}
		},
		threshold: {
			default: .15,
			description: "Feature floor — content scoring below this doesn't count as trackable at all.",
			ui: {
				type: "range",
				min: 0,
				max: .9,
				step: .01,
				label: "Threshold",
				group: "Tracking"
			}
		},
		variance: {
			default: .8,
			description: "Spread of tracker preferences — 0 hunts only the strongest features, 1 distributes trackers across the whole brightness range.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Variance",
				group: "Tracking"
			}
		},
		agility: {
			default: .5,
			description: "How fast trackers travel and how wide they search.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Agility",
				group: "Tracking"
			}
		},
		lifespan: {
			default: 6,
			description: "Seconds a tracker follows its target before retiring and reacquiring elsewhere. 0 = track forever.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .5,
				label: "Lifespan",
				group: "Tracking"
			}
		},
		trail: {
			default: 1,
			description: "Visibility of the keyframe trail each tracker leaves behind.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Trail",
				group: "Style"
			}
		},
		markerSize: {
			default: 28,
			description: "Tracker gizmo size in pixels.",
			ui: {
				type: "range",
				min: 8,
				max: 80,
				step: 1,
				label: "Marker Size",
				group: "Style"
			}
		},
		lineWidth: {
			default: 1.5,
			description: "Stroke width in pixels.",
			ui: {
				type: "range",
				min: .5,
				max: 6,
				step: .25,
				label: "Line Width",
				group: "Style"
			}
		},
		markerColor: {
			default: "#54d66b",
			transform: transformColor,
			description: "Tracker gizmo color.",
			ui: {
				type: "color",
				label: "Tracker",
				group: "Colors"
			}
		},
		keyframeColor: {
			default: "#ffcf33",
			transform: transformColor,
			description: "Keyframe diamond color.",
			ui: {
				type: "color",
				label: "Keyframes",
				group: "Colors"
			}
		},
		pathColor: {
			default: "#ffffffb8",
			transform: transformColor,
			description: "Motion path color.",
			ui: {
				type: "color",
				label: "Path",
				group: "Colors"
			}
		}
	},
	compute: (params) => {
		const { childNode, gpu, registerComputeTexture, convertToTexture, getCpuValue, onCleanup, propValues } = params;
		if (!childNode) return null;
		const root = gpu?.root;
		if (!root) return null;
		const childTexture = convertToTexture(childNode);
		const mode = typeof propValues.detect === "string" ? propValues.detect : "bright";
		const scoreFn = buildScoreFn(mode);
		const featureKernel = buildFeatureKernel(scoreFn);
		const trackKernel = buildTrackKernel(mode, scoreFn);
		const stateA = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
			size: [STATE_W, MAX_TRACKERS],
			format: STATE_FORMAT
		}).$usage("storage", "sampled"), "stateA");
		const stateB = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
			size: [STATE_W, MAX_TRACKERS],
			format: STATE_FORMAT
		}).$usage("storage", "sampled"), "stateB");
		const presentTex = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
			size: [STATE_W, MAX_TRACKERS],
			format: STATE_FORMAT
		}).$usage("storage", "sampled"), "presentTex");
		const gridTex = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
			size: [FEAT_GRID, FEAT_GRID],
			format: STATE_FORMAT
		}).$usage("storage", "sampled"), "gridTex");
		onCleanup(() => {
			stateA.destroy();
			stateB.destroy();
			presentTex.destroy();
			gridTex.destroy();
		});
		const trackState = registerComputeTexture(presentTex);
		const paramsBuf = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(TrackParams), "paramsBuf");
		const featurePipeline = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((gx, gy) => {
			"use gpu";
			featureKernel(gx, gy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "gx"
				}, {
					"type": "i",
					"name": "gy"
				}],
				"body": [0, [[
					6,
					"featureKernel",
					["gx", "gy"]
				]]],
				"externalNames": ["featureKernel"]
			},
			externals: () => ({ featureKernel })
		}) && $.f)({}), { size: [FEAT_GRID, FEAT_GRID] });
		const simPipeline = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((ti) => {
			"use gpu";
			trackKernel(ti);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "ti"
				}],
				"body": [0, [[
					6,
					"trackKernel",
					["ti"]
				]]],
				"externalNames": ["trackKernel"]
			},
			externals: () => ({ trackKernel })
		}) && $.f)({}), { size: [MAX_TRACKERS] });
		let featBg = null;
		let bgAB = null;
		let bgBA = null;
		let direction = 0;
		return {
			outputs: {
				childTexture,
				trackState
			},
			bindInputs: (resolve) => {
				const src = resolve(childTexture.key);
				if (!src) return;
				featBg = root.createBindGroup(featureLayout, {
					src: src.texture,
					grid: gridTex,
					params: paramsBuf.buffer
				});
				bgAB = root.createBindGroup(trackLayout, {
					src: src.texture,
					featGrid: gridTex,
					prev: stateA,
					next: stateB,
					present: presentTex,
					params: paramsBuf.buffer
				});
				bgBA = root.createBindGroup(trackLayout, {
					src: src.texture,
					featGrid: gridTex,
					prev: stateB,
					next: stateA,
					present: presentTex,
					params: paramsBuf.buffer
				});
			},
			getComputeNodes: (frameParams) => {
				if (!featBg || !bgAB || !bgBA) return null;
				const fp = frameParams;
				paramsBuf.write({
					dt: Math.min(fp.deltaTime ?? .016, .05),
					trackersF: getCpuValue("trackers") ?? 4,
					agility: getCpuValue("agility") ?? .5,
					threshold: getCpuValue("threshold") ?? .15,
					variance: getCpuValue("variance") ?? .5,
					lifespan: getCpuValue("lifespan") ?? 6
				});
				const nodes = [featurePipeline.with(featBg), simPipeline.with(direction === 0 ? bgAB : bgBA)];
				direction = 1 - direction;
				return nodes;
			}
		};
	},
	fragment: (params) => {
		const { uniforms, ctx, childNode, convertToTexture, computeOutputs } = params;
		if (!childNode) return ZERO;
		const texKey = (computeOutputs?.childTexture ?? convertToTexture(childNode)).key;
		const stateKit = computeOutputs?.trackState;
		return new Expr((ec) => {
			const U = (name) => uniforms[name]._emit(ec);
			const L = (n) => formatFloat(n);
			const p = ec.freshLocal("kft");
			const s = [];
			const uvE = ctx.uv._emit(ec);
			s.push(`let ${p}_res = ${ctx.logicalViewportSize._emit(ec)};`);
			s.push(`let ${p}_sf = ${p}_res.y / ${L(REFERENCE_HEIGHT)};`);
			s.push(`let ${p}_px = vec2f((${uvE}).x * ${p}_res.x, (${uvE}).y * ${p}_res.y);`);
			s.push(`let ${p}_aa = max(max(fwidth(${p}_px.x), fwidth(${p}_px.y)), ${L(1e-4)});`);
			s.push(`let ${p}_lw = max(${U("lineWidth")} * ${p}_sf, ${L(.5)});`);
			s.push(`let ${p}_ms = max(${U("markerSize")} * ${p}_sf, ${L(4)});`);
			s.push(`var ${p}_dst = textureSampleLevel(tex.$.${texKey}, samp.$.linearClamp, (${uvE}), 0.0);`);
			const overP = (rgb, a) => {
				s.push(`{ let av = clamp(${a}, 0.0, 1.0); ${p}_dst = vec4f((${rgb}) * av + ${p}_dst.rgb * (1.0 - av), av + ${p}_dst.a * (1.0 - av)); }`);
			};
			if (stateKit) {
				const stateKey = stateKit.key;
				const load = (col) => `textureLoad(tex.$.${stateKey}, vec2u(${col}, ti), 0)`;
				const maxAge = TRAIL_LEN * TRAIL_EVERY;
				s.push(`let ${p}_n = u32(clamp(${U("trackers")}, ${L(1)}, ${L(MAX_TRACKERS)}));`);
				s.push(`for (var ti = 0u; ti < ${p}_n; ti = ti + 1u) {`);
				s.push(`let head = ${load("0u")};`);
				s.push(`if (head.w < 0.5) { continue; }`);
				s.push(`let bb = ${load(`${AABB_COL}u`)};`);
				s.push(`let pad = ${p}_ms + ${p}_lw * ${L(4)} + ${p}_aa * ${L(2)};`);
				s.push(`let bbMin = vec2f(bb.x * ${p}_res.x, bb.y * ${p}_res.y) - vec2f(pad);`);
				s.push(`let bbMax = vec2f(bb.z * ${p}_res.x, bb.w * ${p}_res.y) + vec2f(pad);`);
				s.push(`if (all(${p}_px >= bbMin) && all(${p}_px <= bbMax)) {`);
				s.push(`let mta = ${load("1u")};`);
				s.push(`let tp = vec2f(head.x * ${p}_res.x, head.y * ${p}_res.y);`);
				s.push(`let searching = select(${L(1)}, ${L(.45)}, mta.x > ${L(4)});`);
				s.push(`if (${U("trail")} > ${L(.001)}) {`);
				s.push(`var ePrev = ${load(`${2 + TRAIL_LEN - 1}u`)};`);
				s.push(`for (var j = 0u; j < ${TRAIL_LEN}u; j = j + 1u) {`);
				s.push(`let e = ${load("2u + j")};`);
				s.push(`let valid = select(${L(0)}, ${L(1)}, abs(e.w - mta.z) < ${L(.5)}) * select(${L(0)}, ${L(1)}, e.z > ${L(.5)});`);
				s.push(`let age = head.z - e.z;`);
				s.push(`let fade = clamp(${L(1)} - age / ${L(maxAge)}, ${L(0)}, ${L(1)}) * valid;`);
				s.push(`let ep = vec2f(e.x * ${p}_res.x, e.y * ${p}_res.y);`);
				s.push(`let dp = ${p}_px - ep;`);
				s.push(`let dsz = ${p}_ms * ${L(.16)};`);
				s.push(`let dSd = (abs(dp.x) + abs(dp.y)) * ${L(.7071)} - dsz;`);
				overP(`${U("keyframeColor")}.rgb`, `(${L(1)} - smoothstep(${L(0)}, ${p}_aa, dSd)) * ${U("keyframeColor")}.a * fade * ${U("trail")}`);
				s.push(`let validP = select(${L(0)}, ${L(1)}, abs(ePrev.w - mta.z) < ${L(.5)}) * select(${L(0)}, ${L(1)}, ePrev.z > ${L(.5)});`);
				s.push(`let stampGap = abs(e.z - ePrev.z);`);
				s.push(`let contiguous = select(${L(0)}, ${L(1)}, stampGap < ${L(TRAIL_EVERY * 1.5)});`);
				s.push(`let epp = vec2f(ePrev.x * ${p}_res.x, ePrev.y * ${p}_res.y);`);
				s.push(`let pa = ${p}_px - epp; let ba = ep - epp;`);
				s.push(`let h = clamp(dot(pa, ba) / max(dot(ba, ba), ${L(1e-4)}), ${L(0)}, ${L(1)});`);
				s.push(`let segD = length(pa - ba * h);`);
				overP(`${U("pathColor")}.rgb`, `(${L(1)} - smoothstep(${p}_lw * ${L(.4)}, ${p}_lw * ${L(.4)} + ${p}_aa, segD)) * ${U("pathColor")}.a * fade * validP * contiguous * ${U("trail")} * ${L(.8)}`);
				s.push(`ePrev = e;`);
				s.push(`}`);
				s.push(`}`);
				s.push(`let gp = ${p}_px - tp;`);
				s.push(`let hlf = ${p}_ms * ${L(.5)};`);
				s.push(`let q = abs(gp) - vec2f(hlf);`);
				s.push(`let boxSd = length(max(q, vec2f(0.0))) + min(max(q.x, q.y), ${L(0)});`);
				s.push(`var strokeA = ${L(1)} - smoothstep(${L(0)}, ${p}_aa, abs(boxSd) - ${p}_lw * ${L(.5)});`);
				s.push(`let bl = hlf * ${L(.45)};`);
				s.push(`let inCorner = abs(gp.x) > hlf - bl && abs(gp.y) > hlf - bl;`);
				s.push(`strokeA = strokeA * select(${L(0)}, ${L(1)}, inCorner);`);
				overP(`${U("markerColor")}.rgb`, `strokeA * ${U("markerColor")}.a * searching`);
				s.push(`let tick = hlf * ${L(.55)};`);
				s.push(`let chH = ${L(1)} - smoothstep(${L(0)}, ${p}_aa, max(abs(gp.y) - ${p}_lw * ${L(.5)}, abs(gp.x) - tick));`);
				s.push(`let chV = ${L(1)} - smoothstep(${L(0)}, ${p}_aa, max(abs(gp.x) - ${p}_lw * ${L(.5)}, abs(gp.y) - tick));`);
				overP(`${U("markerColor")}.rgb`, `max(chH, chV) * ${L(.55)} * ${U("markerColor")}.a * searching`);
				overP(`vec3f(1.0, 1.0, 1.0)`, `(${L(1)} - smoothstep(${p}_lw, ${p}_lw + ${p}_aa, length(gp))) * searching`);
				s.push(`}`);
				s.push(`}`);
			}
			for (const line of s) ec.statement(line);
			return `vec4f(${p}_dst.rgb / max(${p}_dst.a, ${L(1e-4)}), ${p}_dst.a)`;
		});
	}
};
var KeyFrames_default = componentDefinition;
export { trackHash as a, keyframesFeatureKernelBright as i, buildTrackKernel as n, componentDefinition as r, KeyFrames_default as t };
