// shaders/dist/core/LensFlare-BpFbnSsM.js
import { At as call, It as vec4, R as std, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { c as transformPosition } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
var TAU = 6.2831853;
var lensFlareSpectral = tgpu.fn([d.f32], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((t) => {
	"use gpu";
	return d.vec3f(__tsover_add(__tsover_mul(std.cos(__tsover_mul(TAU, t)), .5), .5), __tsover_add(__tsover_mul(std.cos(__tsover_mul(TAU, __tsover_sub(t, .333))), .5), .5), __tsover_add(__tsover_mul(std.cos(__tsover_mul(TAU, __tsover_sub(t, .666))), .5), .5));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "t"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec3f"
			],
			[
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"cos"
							],
							[[
								1,
								"TAU",
								"*",
								"t"
							]]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				],
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"cos"
							],
							[[
								1,
								"TAU",
								"*",
								[
									1,
									"t",
									"-",
									[5, "0.333"]
								]
							]]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				],
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"cos"
							],
							[[
								1,
								"TAU",
								"*",
								[
									1,
									"t",
									"-",
									[5, "0.666"]
								]
							]]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			]
		]]]],
		"externalNames": [
			"d",
			"std",
			"TAU"
		]
	},
	externals: () => ({
		d,
		std,
		TAU
	})
}) && $.f)({}));
var lensFlareGhost = tgpu.fn([
	d.vec2f,
	d.f32,
	d.vec2f,
	d.vec2f,
	d.f32,
	d.vec3f,
	d.vec4f,
	d.vec4f
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, lightPos, flareAxis, t, shared, disc, tint) => {
	"use gpu";
	const ghostSpread = shared.x;
	const ghostChroma = shared.y;
	const ghostInt = shared.z;
	const offset = disc.x;
	const size = disc.y;
	const bright = disc.z;
	const hollow = disc.w;
	const r = tint.x;
	const g = tint.y;
	const b = tint.z;
	const gPhase = tint.w;
	const ghostPos = lightPos.add(flareAxis.mul(__tsover_mul(__tsover_mul(offset, ghostSpread), 2)));
	const toGhost = d.vec2f(__tsover_mul(__tsover_sub(uv.x, ghostPos.x), aspect), __tsover_sub(uv.y, ghostPos.y));
	const dist = std.length(toGhost);
	const sz = size;
	const discShape = __tsover_sub(1, std.smoothstep(__tsover_mul(sz, .12), sz, dist));
	const innerFade = __tsover_sub(1, std.smoothstep(__tsover_mul(sz, .05), __tsover_mul(sz, .45), dist));
	const shape = __tsover_mul(discShape, __tsover_sub(1, __tsover_mul(innerFade, hollow)));
	const gShimmer = __tsover_add(__tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_mul(t, 4.3), gPhase)), .08), __tsover_mul(std.sin(__tsover_add(__tsover_mul(t, 9.1), __tsover_mul(gPhase, 1.7))), .05)), .95);
	const ringCenter = __tsover_mul(sz, .78);
	const chromaSpread = __tsover_mul(__tsover_mul(sz, ghostChroma), .14);
	const ringWidth = __tsover_mul(sz, .1);
	const rRing = __tsover_sub(1, std.smoothstep(0, ringWidth, std.abs(__tsover_sub(dist, __tsover_add(ringCenter, chromaSpread)))));
	const gRing = __tsover_sub(1, std.smoothstep(0, ringWidth, std.abs(__tsover_sub(dist, ringCenter))));
	const bRing = __tsover_sub(1, std.smoothstep(0, ringWidth, std.abs(__tsover_sub(dist, __tsover_sub(ringCenter, chromaSpread)))));
	const brightness = __tsover_mul(__tsover_mul(bright, ghostInt), gShimmer);
	return d.vec3f(__tsover_mul(__tsover_add(__tsover_mul(shape, r), __tsover_mul(rRing, ghostChroma)), brightness), __tsover_mul(__tsover_add(__tsover_mul(shape, g), __tsover_mul(gRing, ghostChroma)), brightness), __tsover_mul(__tsover_add(__tsover_mul(shape, b), __tsover_mul(bRing, ghostChroma)), brightness));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "lightPos"
			},
			{
				"type": "i",
				"name": "flareAxis"
			},
			{
				"type": "i",
				"name": "t"
			},
			{
				"type": "i",
				"name": "shared"
			},
			{
				"type": "i",
				"name": "disc"
			},
			{
				"type": "i",
				"name": "tint"
			}
		],
		"body": [0, [
			[
				13,
				"ghostSpread",
				[
					7,
					"shared",
					"x"
				]
			],
			[
				13,
				"ghostChroma",
				[
					7,
					"shared",
					"y"
				]
			],
			[
				13,
				"ghostInt",
				[
					7,
					"shared",
					"z"
				]
			],
			[
				13,
				"offset",
				[
					7,
					"disc",
					"x"
				]
			],
			[
				13,
				"size",
				[
					7,
					"disc",
					"y"
				]
			],
			[
				13,
				"bright",
				[
					7,
					"disc",
					"z"
				]
			],
			[
				13,
				"hollow",
				[
					7,
					"disc",
					"w"
				]
			],
			[
				13,
				"r",
				[
					7,
					"tint",
					"x"
				]
			],
			[
				13,
				"g",
				[
					7,
					"tint",
					"y"
				]
			],
			[
				13,
				"b",
				[
					7,
					"tint",
					"z"
				]
			],
			[
				13,
				"gPhase",
				[
					7,
					"tint",
					"w"
				]
			],
			[
				13,
				"ghostPos",
				[
					6,
					[
						7,
						"lightPos",
						"add"
					],
					[[
						6,
						[
							7,
							"flareAxis",
							"mul"
						],
						[[
							1,
							[
								1,
								"offset",
								"*",
								"ghostSpread"
							],
							"*",
							[5, "2"]
						]]
					]]
				]
			],
			[
				13,
				"toGhost",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"-",
							[
								7,
								"ghostPos",
								"x"
							]
						],
						"*",
						"aspect"
					], [
						1,
						[
							7,
							"uv",
							"y"
						],
						"-",
						[
							7,
							"ghostPos",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["toGhost"]
				]
			],
			[
				13,
				"sz",
				"size"
			],
			[
				13,
				"discShape",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"sz",
								"*",
								[5, "0.12"]
							],
							"sz",
							"dist"
						]
					]
				]
			],
			[
				13,
				"innerFade",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"sz",
								"*",
								[5, "0.05"]
							],
							[
								1,
								"sz",
								"*",
								[5, "0.45"]
							],
							"dist"
						]
					]
				]
			],
			[
				13,
				"shape",
				[
					1,
					"discShape",
					"*",
					[
						1,
						[5, "1"],
						"-",
						[
							1,
							"innerFade",
							"*",
							"hollow"
						]
					]
				]
			],
			[
				13,
				"gShimmer",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										1,
										"t",
										"*",
										[5, "4.3"]
									],
									"+",
									"gPhase"
								]]
							],
							"*",
							[5, "0.08"]
						],
						"+",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										1,
										"t",
										"*",
										[5, "9.1"]
									],
									"+",
									[
										1,
										"gPhase",
										"*",
										[5, "1.7"]
									]
								]]
							],
							"*",
							[5, "0.05"]
						]
					],
					"+",
					[5, "0.95"]
				]
			],
			[
				13,
				"ringCenter",
				[
					1,
					"sz",
					"*",
					[5, "0.78"]
				]
			],
			[
				13,
				"chromaSpread",
				[
					1,
					[
						1,
						"sz",
						"*",
						"ghostChroma"
					],
					"*",
					[5, "0.14"]
				]
			],
			[
				13,
				"ringWidth",
				[
					1,
					"sz",
					"*",
					[5, "0.1"]
				]
			],
			[
				13,
				"rRing",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							"ringWidth",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"dist",
									"-",
									[
										1,
										"ringCenter",
										"+",
										"chromaSpread"
									]
								]]
							]
						]
					]
				]
			],
			[
				13,
				"gRing",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							"ringWidth",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"dist",
									"-",
									"ringCenter"
								]]
							]
						]
					]
				]
			],
			[
				13,
				"bRing",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							"ringWidth",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"dist",
									"-",
									[
										1,
										"ringCenter",
										"-",
										"chromaSpread"
									]
								]]
							]
						]
					]
				]
			],
			[
				13,
				"brightness",
				[
					1,
					[
						1,
						"bright",
						"*",
						"ghostInt"
					],
					"*",
					"gShimmer"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[
						1,
						[
							1,
							[
								1,
								"shape",
								"*",
								"r"
							],
							"+",
							[
								1,
								"rRing",
								"*",
								"ghostChroma"
							]
						],
						"*",
						"brightness"
					],
					[
						1,
						[
							1,
							[
								1,
								"shape",
								"*",
								"g"
							],
							"+",
							[
								1,
								"gRing",
								"*",
								"ghostChroma"
							]
						],
						"*",
						"brightness"
					],
					[
						1,
						[
							1,
							[
								1,
								"shape",
								"*",
								"b"
							],
							"+",
							[
								1,
								"bRing",
								"*",
								"ghostChroma"
							]
						],
						"*",
						"brightness"
					]
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const lensFlareColor = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.vec2f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, animTime, lightPosRaw, p1, p2, p3, p4) => {
	"use gpu";
	const intensity = p1.x;
	const ghostInt = p1.y;
	const ghostSpread = p1.z;
	const ghostChroma = p1.w;
	const haloInt = p2.x;
	const haloRad = p2.y;
	const haloChroma = p2.z;
	const haloSoft = p2.w;
	const starburstInt = p3.x;
	const starburstPts = p3.y;
	const streakInt = p3.z;
	const streakLen = p3.w;
	const glareInt = p4.x;
	const glareSz = p4.y;
	const edgeFade = p4.z;
	const aspect = __tsover_div(viewport.x, viewport.y);
	const lightPos = d.vec2f(lightPosRaw.x, __tsover_sub(1, lightPosRaw.y));
	const flareAxis = d.vec2f(__tsover_sub(.5, lightPos.x), __tsover_sub(.5, lightPos.y));
	const t = animTime;
	const shimmer = __tsover_add(__tsover_add(__tsover_add(__tsover_mul(std.sin(__tsover_mul(t, 3.7)), .04), __tsover_mul(std.sin(__tsover_mul(t, 7.3)), .03)), __tsover_mul(std.sin(__tsover_mul(t, 13.1)), .02)), .97);
	const edgeDist = std.max(__tsover_mul(std.abs(__tsover_sub(lightPos.x, .5)), 2), __tsover_mul(std.abs(__tsover_sub(lightPos.y, .5)), 2));
	const fadeStart = __tsover_sub(1, __tsover_mul(edgeFade, .5));
	const edgeFadeV = std.mix(1, __tsover_sub(1, std.smoothstep(fadeStart, 1, edgeDist)), edgeFade);
	const masterFade = __tsover_mul(__tsover_mul(intensity, edgeFadeV), shimmer);
	const toLightAR = d.vec2f(__tsover_mul(__tsover_sub(uv.x, lightPos.x), aspect), __tsover_sub(uv.y, lightPos.y));
	const lightDist = std.length(toLightAR);
	const gShared = d.vec3f(ghostSpread, ghostChroma, ghostInt);
	let ghostAccum = d.vec3f(0, 0, 0);
	ghostAccum = ghostAccum.add(lensFlareGhost(uv, aspect, lightPos, flareAxis, t, gShared, d.vec4f(.15, .2, .3, .1), d.vec4f(1, .95, .9, 0)));
	ghostAccum = ghostAccum.add(lensFlareGhost(uv, aspect, lightPos, flareAxis, t, gShared, d.vec4f(.3, .12, .25, .65), d.vec4f(1, .85, .55, 2.17)));
	ghostAccum = ghostAccum.add(lensFlareGhost(uv, aspect, lightPos, flareAxis, t, gShared, d.vec4f(.5, .08, .35, .15), d.vec4f(.62, .82, 1, 4.34)));
	ghostAccum = ghostAccum.add(lensFlareGhost(uv, aspect, lightPos, flareAxis, t, gShared, d.vec4f(.72, .14, .18, .75), d.vec4f(.75, 1, .7, 6.51)));
	ghostAccum = ghostAccum.add(lensFlareGhost(uv, aspect, lightPos, flareAxis, t, gShared, d.vec4f(.95, .06, .3, 0), d.vec4f(1, .6, .85, 8.68)));
	ghostAccum = ghostAccum.add(lensFlareGhost(uv, aspect, lightPos, flareAxis, t, gShared, d.vec4f(1.25, .1, .15, .8), d.vec4f(.55, .88, 1, 10.85)));
	ghostAccum = ghostAccum.add(lensFlareGhost(uv, aspect, lightPos, flareAxis, t, gShared, d.vec4f(1.55, .07, .2, .25), d.vec4f(1, .92, .6, 13.02)));
	const antiLight = d.vec2f(__tsover_sub(1, lightPos.x), __tsover_sub(1, lightPos.y));
	const toAntiLight = d.vec2f(__tsover_mul(__tsover_sub(uv.x, antiLight.x), aspect), __tsover_sub(uv.y, antiLight.y));
	const antiLightDist = std.length(toAntiLight);
	const haloChromaSpread = __tsover_mul(__tsover_mul(haloRad, haloChroma), .06);
	const haloWidth = __tsover_add(__tsover_mul(haloSoft, .12), .005);
	const rHalo = __tsover_sub(1, std.smoothstep(0, haloWidth, std.abs(__tsover_sub(antiLightDist, __tsover_add(haloRad, haloChromaSpread)))));
	const gHalo = __tsover_sub(1, std.smoothstep(0, haloWidth, std.abs(__tsover_sub(antiLightDist, haloRad))));
	const bHalo = __tsover_sub(1, std.smoothstep(0, haloWidth, std.abs(__tsover_sub(antiLightDist, __tsover_sub(haloRad, haloChromaSpread)))));
	const haloAngle = std.atan2(toAntiLight.y, toAntiLight.x);
	const haloAngVar = __tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_mul(haloAngle, 3), __tsover_mul(t, .5))), .12), .88);
	const haloColor = d.vec3f(rHalo, gHalo, bHalo).mul(haloInt).mul(haloAngVar);
	const lightAngle = std.atan2(toLightAR.y, toLightAR.x);
	const axisAng = std.atan2(flareAxis.y, __tsover_mul(flareAxis.x, aspect));
	const rotation = __tsover_add(__tsover_mul(axisAng, .5), __tsover_mul(t, .06));
	const spikeAngle = __tsover_add(lightAngle, rotation);
	const primarySpikes = std.pow(std.abs(std.cos(__tsover_mul(spikeAngle, __tsover_div(starburstPts, 2)))), 3.5);
	const secondarySpikes = __tsover_mul(std.pow(std.abs(std.cos(__tsover_mul(__tsover_add(spikeAngle, .25), __tsover_div(starburstPts, 2)))), 7), .4);
	const radialFall = __tsover_div(1, __tsover_add(1, __tsover_mul(__tsover_mul(lightDist, lightDist), 60)));
	const starSpectral = lensFlareSpectral(__tsover_add(__tsover_div(spikeAngle, TAU), .5));
	const starburstColor = std.mix(d.vec3f(1, .97, .92), starSpectral, .12).mul(__tsover_mul(__tsover_mul(__tsover_add(primarySpikes, secondarySpikes), radialFall), starburstInt));
	const dy = std.abs(__tsover_sub(uv.y, lightPos.y));
	const dx = std.abs(__tsover_sub(uv.x, lightPos.x));
	const vertFall = std.exp(__tsover_mul(__tsover_mul(-800, dy), dy));
	const lenSq = __tsover_add(__tsover_mul(__tsover_mul(streakLen, streakLen), .25), .001);
	const rStreak = std.exp(__tsover_div(__tsover_mul(__tsover_mul(-1, dx), dx), __tsover_mul(lenSq, .7)));
	const gStreak = std.exp(__tsover_div(__tsover_mul(__tsover_mul(-1, dx), dx), lenSq));
	const bStreak = std.exp(__tsover_div(__tsover_mul(__tsover_mul(-1, dx), dx), __tsover_mul(lenSq, 1.4)));
	const streakColor = d.vec3f(__tsover_mul(__tsover_mul(vertFall, rStreak), .5), __tsover_mul(__tsover_mul(vertFall, gStreak), .65), __tsover_mul(__tsover_mul(vertFall, bStreak), .9)).mul(streakInt);
	const glareSharpness = __tsover_div(1, __tsover_add(__tsover_mul(glareSz, glareSz), .01));
	const glareFalloff = std.exp(__tsover_mul(__tsover_mul(__tsover_mul(-1, lightDist), lightDist), glareSharpness));
	const glareColor = d.vec3f(1, .98, .95).mul(__tsover_mul(glareFalloff, glareInt));
	const coreSoft = std.exp(__tsover_mul(__tsover_mul(-500, lightDist), lightDist));
	const coreSharp = __tsover_sub(1, std.smoothstep(0, .025, lightDist));
	const coreColor = d.vec3f(1, .98, .96).mul(__tsover_mul(__tsover_add(coreSoft, __tsover_mul(coreSharp, 1.5)), intensity));
	const finalColor = ghostAccum.add(haloColor).add(starburstColor).add(streakColor).add(glareColor).add(coreColor).mul(masterFade);
	const flareAlpha = std.clamp(__tsover_mul(std.dot(finalColor, d.vec3f(.299, .587, .114)), 2), 0, 1);
	return d.vec4f(finalColor, flareAlpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "lightPosRaw"
			},
			{
				"type": "i",
				"name": "p1"
			},
			{
				"type": "i",
				"name": "p2"
			},
			{
				"type": "i",
				"name": "p3"
			},
			{
				"type": "i",
				"name": "p4"
			}
		],
		"body": [0, [
			[
				13,
				"intensity",
				[
					7,
					"p1",
					"x"
				]
			],
			[
				13,
				"ghostInt",
				[
					7,
					"p1",
					"y"
				]
			],
			[
				13,
				"ghostSpread",
				[
					7,
					"p1",
					"z"
				]
			],
			[
				13,
				"ghostChroma",
				[
					7,
					"p1",
					"w"
				]
			],
			[
				13,
				"haloInt",
				[
					7,
					"p2",
					"x"
				]
			],
			[
				13,
				"haloRad",
				[
					7,
					"p2",
					"y"
				]
			],
			[
				13,
				"haloChroma",
				[
					7,
					"p2",
					"z"
				]
			],
			[
				13,
				"haloSoft",
				[
					7,
					"p2",
					"w"
				]
			],
			[
				13,
				"starburstInt",
				[
					7,
					"p3",
					"x"
				]
			],
			[
				13,
				"starburstPts",
				[
					7,
					"p3",
					"y"
				]
			],
			[
				13,
				"streakInt",
				[
					7,
					"p3",
					"z"
				]
			],
			[
				13,
				"streakLen",
				[
					7,
					"p3",
					"w"
				]
			],
			[
				13,
				"glareInt",
				[
					7,
					"p4",
					"x"
				]
			],
			[
				13,
				"glareSz",
				[
					7,
					"p4",
					"y"
				]
			],
			[
				13,
				"edgeFade",
				[
					7,
					"p4",
					"z"
				]
			],
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"lightPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"lightPosRaw",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"lightPosRaw",
							"y"
						]
					]]
				]
			],
			[
				13,
				"flareAxis",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[5, "0.5"],
						"-",
						[
							7,
							"lightPos",
							"x"
						]
					], [
						1,
						[5, "0.5"],
						"-",
						[
							7,
							"lightPos",
							"y"
						]
					]]
				]
			],
			[
				13,
				"t",
				"animTime"
			],
			[
				13,
				"shimmer",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										"t",
										"*",
										[5, "3.7"]
									]]
								],
								"*",
								[5, "0.04"]
							],
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										"t",
										"*",
										[5, "7.3"]
									]]
								],
								"*",
								[5, "0.03"]
							]
						],
						"+",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									"t",
									"*",
									[5, "13.1"]
								]]
							],
							"*",
							[5, "0.02"]
						]
					],
					"+",
					[5, "0.97"]
				]
			],
			[
				13,
				"edgeDist",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								1,
								[
									7,
									"lightPos",
									"x"
								],
								"-",
								[5, "0.5"]
							]]
						],
						"*",
						[5, "2"]
					], [
						1,
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								1,
								[
									7,
									"lightPos",
									"y"
								],
								"-",
								[5, "0.5"]
							]]
						],
						"*",
						[5, "2"]
					]]
				]
			],
			[
				13,
				"fadeStart",
				[
					1,
					[5, "1"],
					"-",
					[
						1,
						"edgeFade",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"edgeFadeV",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "1"],
						[
							1,
							[5, "1"],
							"-",
							[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									"fadeStart",
									[5, "1"],
									"edgeDist"
								]
							]
						],
						"edgeFade"
					]
				]
			],
			[
				13,
				"masterFade",
				[
					1,
					[
						1,
						"intensity",
						"*",
						"edgeFadeV"
					],
					"*",
					"shimmer"
				]
			],
			[
				13,
				"toLightAR",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"-",
							[
								7,
								"lightPos",
								"x"
							]
						],
						"*",
						"aspect"
					], [
						1,
						[
							7,
							"uv",
							"y"
						],
						"-",
						[
							7,
							"lightPos",
							"y"
						]
					]]
				]
			],
			[
				13,
				"lightDist",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["toLightAR"]
				]
			],
			[
				13,
				"gShared",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						"ghostSpread",
						"ghostChroma",
						"ghostInt"
					]
				]
			],
			[
				12,
				"ghostAccum",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "0"]
					]
				]
			],
			[
				2,
				"ghostAccum",
				"=",
				[
					6,
					[
						7,
						"ghostAccum",
						"add"
					],
					[[
						6,
						"lensFlareGhost",
						[
							"uv",
							"aspect",
							"lightPos",
							"flareAxis",
							"t",
							"gShared",
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "0.15"],
									[5, "0.2"],
									[5, "0.3"],
									[5, "0.1"]
								]
							],
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "1"],
									[5, "0.95"],
									[5, "0.9"],
									[5, "0"]
								]
							]
						]
					]]
				]
			],
			[
				2,
				"ghostAccum",
				"=",
				[
					6,
					[
						7,
						"ghostAccum",
						"add"
					],
					[[
						6,
						"lensFlareGhost",
						[
							"uv",
							"aspect",
							"lightPos",
							"flareAxis",
							"t",
							"gShared",
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "0.3"],
									[5, "0.12"],
									[5, "0.25"],
									[5, "0.65"]
								]
							],
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "1"],
									[5, "0.85"],
									[5, "0.55"],
									[5, "2.17"]
								]
							]
						]
					]]
				]
			],
			[
				2,
				"ghostAccum",
				"=",
				[
					6,
					[
						7,
						"ghostAccum",
						"add"
					],
					[[
						6,
						"lensFlareGhost",
						[
							"uv",
							"aspect",
							"lightPos",
							"flareAxis",
							"t",
							"gShared",
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "0.5"],
									[5, "0.08"],
									[5, "0.35"],
									[5, "0.15"]
								]
							],
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "0.62"],
									[5, "0.82"],
									[5, "1"],
									[5, "4.34"]
								]
							]
						]
					]]
				]
			],
			[
				2,
				"ghostAccum",
				"=",
				[
					6,
					[
						7,
						"ghostAccum",
						"add"
					],
					[[
						6,
						"lensFlareGhost",
						[
							"uv",
							"aspect",
							"lightPos",
							"flareAxis",
							"t",
							"gShared",
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "0.72"],
									[5, "0.14"],
									[5, "0.18"],
									[5, "0.75"]
								]
							],
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "0.75"],
									[5, "1"],
									[5, "0.7"],
									[5, "6.51"]
								]
							]
						]
					]]
				]
			],
			[
				2,
				"ghostAccum",
				"=",
				[
					6,
					[
						7,
						"ghostAccum",
						"add"
					],
					[[
						6,
						"lensFlareGhost",
						[
							"uv",
							"aspect",
							"lightPos",
							"flareAxis",
							"t",
							"gShared",
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "0.95"],
									[5, "0.06"],
									[5, "0.3"],
									[5, "0"]
								]
							],
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "1"],
									[5, "0.6"],
									[5, "0.85"],
									[5, "8.68"]
								]
							]
						]
					]]
				]
			],
			[
				2,
				"ghostAccum",
				"=",
				[
					6,
					[
						7,
						"ghostAccum",
						"add"
					],
					[[
						6,
						"lensFlareGhost",
						[
							"uv",
							"aspect",
							"lightPos",
							"flareAxis",
							"t",
							"gShared",
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "1.25"],
									[5, "0.1"],
									[5, "0.15"],
									[5, "0.8"]
								]
							],
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "0.55"],
									[5, "0.88"],
									[5, "1"],
									[5, "10.85"]
								]
							]
						]
					]]
				]
			],
			[
				2,
				"ghostAccum",
				"=",
				[
					6,
					[
						7,
						"ghostAccum",
						"add"
					],
					[[
						6,
						"lensFlareGhost",
						[
							"uv",
							"aspect",
							"lightPos",
							"flareAxis",
							"t",
							"gShared",
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "1.55"],
									[5, "0.07"],
									[5, "0.2"],
									[5, "0.25"]
								]
							],
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									[5, "1"],
									[5, "0.92"],
									[5, "0.6"],
									[5, "13.02"]
								]
							]
						]
					]]
				]
			],
			[
				13,
				"antiLight",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[5, "1"],
						"-",
						[
							7,
							"lightPos",
							"x"
						]
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"lightPos",
							"y"
						]
					]]
				]
			],
			[
				13,
				"toAntiLight",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"-",
							[
								7,
								"antiLight",
								"x"
							]
						],
						"*",
						"aspect"
					], [
						1,
						[
							7,
							"uv",
							"y"
						],
						"-",
						[
							7,
							"antiLight",
							"y"
						]
					]]
				]
			],
			[
				13,
				"antiLightDist",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["toAntiLight"]
				]
			],
			[
				13,
				"haloChromaSpread",
				[
					1,
					[
						1,
						"haloRad",
						"*",
						"haloChroma"
					],
					"*",
					[5, "0.06"]
				]
			],
			[
				13,
				"haloWidth",
				[
					1,
					[
						1,
						"haloSoft",
						"*",
						[5, "0.12"]
					],
					"+",
					[5, "0.005"]
				]
			],
			[
				13,
				"rHalo",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							"haloWidth",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"antiLightDist",
									"-",
									[
										1,
										"haloRad",
										"+",
										"haloChromaSpread"
									]
								]]
							]
						]
					]
				]
			],
			[
				13,
				"gHalo",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							"haloWidth",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"antiLightDist",
									"-",
									"haloRad"
								]]
							]
						]
					]
				]
			],
			[
				13,
				"bHalo",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							"haloWidth",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"antiLightDist",
									"-",
									[
										1,
										"haloRad",
										"-",
										"haloChromaSpread"
									]
								]]
							]
						]
					]
				]
			],
			[
				13,
				"haloAngle",
				[
					6,
					[
						7,
						"std",
						"atan2"
					],
					[[
						7,
						"toAntiLight",
						"y"
					], [
						7,
						"toAntiLight",
						"x"
					]]
				]
			],
			[
				13,
				"haloAngVar",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									"haloAngle",
									"*",
									[5, "3"]
								],
								"+",
								[
									1,
									"t",
									"*",
									[5, "0.5"]
								]
							]]
						],
						"*",
						[5, "0.12"]
					],
					"+",
					[5, "0.88"]
				]
			],
			[
				13,
				"haloColor",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										"d",
										"vec3f"
									],
									[
										"rHalo",
										"gHalo",
										"bHalo"
									]
								],
								"mul"
							],
							["haloInt"]
						],
						"mul"
					],
					["haloAngVar"]
				]
			],
			[
				13,
				"lightAngle",
				[
					6,
					[
						7,
						"std",
						"atan2"
					],
					[[
						7,
						"toLightAR",
						"y"
					], [
						7,
						"toLightAR",
						"x"
					]]
				]
			],
			[
				13,
				"axisAng",
				[
					6,
					[
						7,
						"std",
						"atan2"
					],
					[[
						7,
						"flareAxis",
						"y"
					], [
						1,
						[
							7,
							"flareAxis",
							"x"
						],
						"*",
						"aspect"
					]]
				]
			],
			[
				13,
				"rotation",
				[
					1,
					[
						1,
						"axisAng",
						"*",
						[5, "0.5"]
					],
					"+",
					[
						1,
						"t",
						"*",
						[5, "0.06"]
					]
				]
			],
			[
				13,
				"spikeAngle",
				[
					1,
					"lightAngle",
					"+",
					"rotation"
				]
			],
			[
				13,
				"primarySpikes",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							6,
							[
								7,
								"std",
								"cos"
							],
							[[
								1,
								"spikeAngle",
								"*",
								[
									1,
									"starburstPts",
									"/",
									[5, "2"]
								]
							]]
						]]
					], [5, "3.5"]]
				]
			],
			[
				13,
				"secondarySpikes",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"pow"
						],
						[[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								6,
								[
									7,
									"std",
									"cos"
								],
								[[
									1,
									[
										1,
										"spikeAngle",
										"+",
										[5, "0.25"]
									],
									"*",
									[
										1,
										"starburstPts",
										"/",
										[5, "2"]
									]
								]]
							]]
						], [5, "7"]]
					],
					"*",
					[5, "0.4"]
				]
			],
			[
				13,
				"radialFall",
				[
					1,
					[5, "1"],
					"/",
					[
						1,
						[5, "1"],
						"+",
						[
							1,
							[
								1,
								"lightDist",
								"*",
								"lightDist"
							],
							"*",
							[5, "60"]
						]
					]
				]
			],
			[
				13,
				"starSpectral",
				[
					6,
					"lensFlareSpectral",
					[[
						1,
						[
							1,
							"spikeAngle",
							"/",
							"TAU"
						],
						"+",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"starTint",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1"],
								[5, "0.97"],
								[5, "0.92"]
							]
						],
						"starSpectral",
						[5, "0.12"]
					]
				]
			],
			[
				13,
				"starburstColor",
				[
					6,
					[
						7,
						"starTint",
						"mul"
					],
					[[
						1,
						[
							1,
							[
								1,
								"primarySpikes",
								"+",
								"secondarySpikes"
							],
							"*",
							"radialFall"
						],
						"*",
						"starburstInt"
					]]
				]
			],
			[
				13,
				"dy",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						1,
						[
							7,
							"uv",
							"y"
						],
						"-",
						[
							7,
							"lightPos",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dx",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"-",
						[
							7,
							"lightPos",
							"x"
						]
					]]
				]
			],
			[
				13,
				"vertFall",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							[
								4,
								"-",
								[5, "800"]
							],
							"*",
							"dy"
						],
						"*",
						"dy"
					]]
				]
			],
			[
				13,
				"lenSq",
				[
					1,
					[
						1,
						[
							1,
							"streakLen",
							"*",
							"streakLen"
						],
						"*",
						[5, "0.25"]
					],
					"+",
					[5, "0.001"]
				]
			],
			[
				13,
				"rStreak",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							[
								1,
								[
									4,
									"-",
									[5, "1"]
								],
								"*",
								"dx"
							],
							"*",
							"dx"
						],
						"/",
						[
							1,
							"lenSq",
							"*",
							[5, "0.7"]
						]
					]]
				]
			],
			[
				13,
				"gStreak",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							[
								1,
								[
									4,
									"-",
									[5, "1"]
								],
								"*",
								"dx"
							],
							"*",
							"dx"
						],
						"/",
						"lenSq"
					]]
				]
			],
			[
				13,
				"bStreak",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							[
								1,
								[
									4,
									"-",
									[5, "1"]
								],
								"*",
								"dx"
							],
							"*",
							"dx"
						],
						"/",
						[
							1,
							"lenSq",
							"*",
							[5, "1.4"]
						]
					]]
				]
			],
			[
				13,
				"streakColor",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									[
										1,
										"vertFall",
										"*",
										"rStreak"
									],
									"*",
									[5, "0.5"]
								],
								[
									1,
									[
										1,
										"vertFall",
										"*",
										"gStreak"
									],
									"*",
									[5, "0.65"]
								],
								[
									1,
									[
										1,
										"vertFall",
										"*",
										"bStreak"
									],
									"*",
									[5, "0.9"]
								]
							]
						],
						"mul"
					],
					["streakInt"]
				]
			],
			[
				13,
				"glareSharpness",
				[
					1,
					[5, "1"],
					"/",
					[
						1,
						[
							1,
							"glareSz",
							"*",
							"glareSz"
						],
						"+",
						[5, "0.01"]
					]
				]
			],
			[
				13,
				"glareFalloff",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							[
								1,
								[
									4,
									"-",
									[5, "1"]
								],
								"*",
								"lightDist"
							],
							"*",
							"lightDist"
						],
						"*",
						"glareSharpness"
					]]
				]
			],
			[
				13,
				"glareColor",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1"],
								[5, "0.98"],
								[5, "0.95"]
							]
						],
						"mul"
					],
					[[
						1,
						"glareFalloff",
						"*",
						"glareInt"
					]]
				]
			],
			[
				13,
				"coreSoft",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							[
								4,
								"-",
								[5, "500"]
							],
							"*",
							"lightDist"
						],
						"*",
						"lightDist"
					]]
				]
			],
			[
				13,
				"coreSharp",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							[5, "0.025"],
							"lightDist"
						]
					]
				]
			],
			[
				13,
				"coreColor",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1"],
								[5, "0.98"],
								[5, "0.96"]
							]
						],
						"mul"
					],
					[[
						1,
						[
							1,
							"coreSoft",
							"+",
							[
								1,
								"coreSharp",
								"*",
								[5, "1.5"]
							]
						],
						"*",
						"intensity"
					]]
				]
			],
			[
				13,
				"flareTotal",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										[
											6,
											[
												7,
												[
													6,
													[
														7,
														"ghostAccum",
														"add"
													],
													["haloColor"]
												],
												"add"
											],
											["starburstColor"]
										],
										"add"
									],
									["streakColor"]
								],
								"add"
							],
							["glareColor"]
						],
						"add"
					],
					["coreColor"]
				]
			],
			[
				13,
				"finalColor",
				[
					6,
					[
						7,
						"flareTotal",
						"mul"
					],
					["masterFade"]
				]
			],
			[
				13,
				"flareAlpha",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"dot"
								],
								["finalColor", [
									6,
									[
										7,
										"d",
										"vec3f"
									],
									[
										[5, "0.299"],
										[5, "0.587"],
										[5, "0.114"]
									]
								]]
							],
							"*",
							[5, "2"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				["finalColor", "flareAlpha"]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"lensFlareGhost",
			"lensFlareSpectral",
			"TAU"
		]
	},
	externals: () => ({
		d,
		std,
		lensFlareGhost,
		lensFlareSpectral,
		TAU
	})
}) && $.f)({}));
const componentDefinition = {
	name: "LensFlare",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Stylize",
	description: "Realistic camera lens flare with artifacts.",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		lightPosition: {
			default: {
				x: .3,
				y: .3
			},
			transform: transformPosition,
			description: "Position of the light source",
			ui: {
				type: "position",
				label: "Light Position",
				group: "Light"
			}
		},
		intensity: {
			default: .5,
			description: "Master brightness of the entire lens flare effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .01,
				label: "Intensity",
				group: "Light"
			}
		},
		ghostIntensity: {
			default: .4,
			description: "Brightness of internal reflection ghost discs along the flare axis",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Ghosts",
				group: "Reflections"
			}
		},
		ghostSpread: {
			default: .7,
			description: "Spacing between ghost reflections along the flare axis",
			ui: {
				type: "range",
				min: .1,
				max: 2,
				step: .01,
				label: "Ghost Spread",
				group: "Reflections"
			}
		},
		ghostChroma: {
			default: .3,
			description: "Rainbow chromatic fringing around ghost element edges",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Ghost Chroma",
				group: "Reflections"
			}
		},
		haloIntensity: {
			default: .4,
			description: "Brightness of the circular halo ring from internal reflection",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Halo",
				group: "Reflections"
			}
		},
		haloRadius: {
			default: .6,
			description: "Radius of the halo ring",
			ui: {
				type: "range",
				min: .1,
				max: 1,
				step: .01,
				label: "Halo Size",
				group: "Reflections"
			}
		},
		haloChroma: {
			default: .6,
			description: "Spectral dispersion on the halo creating rainbow color separation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Halo Chroma",
				group: "Reflections"
			}
		},
		haloSoftness: {
			default: .8,
			description: "Thickness and softness of the halo ring",
			ui: {
				type: ["range", "map"],
				min: .01,
				max: 3,
				step: .01,
				label: "Halo Softness",
				group: "Reflections"
			}
		},
		starburstIntensity: {
			default: .3,
			description: "Brightness of diffraction spikes radiating from the light source",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Starburst",
				group: "Effects"
			}
		},
		starburstPoints: {
			default: 6,
			description: "Number of starburst spikes (simulates aperture blade count)",
			ui: {
				type: ["range", "map"],
				min: 4,
				max: 16,
				step: 1,
				label: "Blades",
				group: "Effects"
			}
		},
		streakIntensity: {
			default: .15,
			description: "Brightness of horizontal anamorphic light streak",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Streak",
				group: "Effects"
			}
		},
		streakLength: {
			default: .5,
			description: "Horizontal extent of the anamorphic streak",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 1,
				step: .01,
				label: "Streak Length",
				group: "Effects"
			}
		},
		glareIntensity: {
			default: .2,
			description: "Soft veiling glare that washes out contrast around the light",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Glare",
				group: "Effects"
			}
		},
		glareSize: {
			default: .5,
			description: "Size of the soft glare glow",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 1,
				step: .01,
				label: "Glare Size",
				group: "Effects"
			}
		},
		edgeFade: {
			default: .2,
			description: "How much the flare fades when the light source is near the screen edge (0 = no fade, 1 = heavy fade)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Fade",
				group: "Light"
			}
		},
		speed: {
			default: .5,
			description: "Speed of subtle flare shimmer and starburst rotation",
			ui: {
				type: "range",
				min: 0,
				max: 3,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		}
	},
	fragment: (params) => {
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		const u = params.uniforms;
		return call(lensFlareColor, "lensFlareColor", [
			uv,
			viewport,
			t,
			u.lightPosition,
			vec4(u.intensity, u.ghostIntensity, u.ghostSpread, u.ghostChroma),
			vec4(u.haloIntensity, u.haloRadius, u.haloChroma, u.haloSoftness),
			vec4(u.starburstIntensity, u.starburstPoints, u.streakIntensity, u.streakLength),
			vec4(u.glareIntensity, u.glareSize, u.edgeFade, 0)
		]);
	}
};
var LensFlare_default = componentDefinition;
export { componentDefinition as n, lensFlareColor as r, LensFlare_default as t };
