// shaders/dist/core/LightLeak-DGt9quKs.js
import { At as call, Ot as ZERO, R as std, Z as hash12, lt as value12, n as tgpu, on as d, rt as noise_exports } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
const lightLeakRamp = tgpu.fn([
	d.f32,
	d.vec3f,
	d.vec3f,
	d.vec3f
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((heat, hot, mid, fringe) => {
	"use gpu";
	const toMid = std.smoothstep(0, .5, heat);
	const toHot = std.smoothstep(.55, 1.15, heat);
	return std.mix(std.mix(fringe, mid, d.vec3f(toMid)), hot, d.vec3f(toHot)).mul(std.clamp(heat, 0, 1.6));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "heat"
			},
			{
				"type": "i",
				"name": "hot"
			},
			{
				"type": "i",
				"name": "mid"
			},
			{
				"type": "i",
				"name": "fringe"
			}
		],
		"body": [0, [
			[
				13,
				"toMid",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[5, "0.5"],
						"heat"
					]
				]
			],
			[
				13,
				"toHot",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.55"],
						[5, "1.15"],
						"heat"
					]
				]
			],
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								"fringe",
								"mid",
								[
									6,
									[
										7,
										"d",
										"vec3f"
									],
									["toMid"]
								]
							]
						],
						"hot",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["toHot"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"rgb",
					"mul"
				],
				[[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"heat",
						[5, "0"],
						[5, "1.6"]
					]
				]]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const lightLeakComposite = tgpu.fn([
	d.vec4f,
	d.vec2f,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec3f,
	d.vec3f,
	d.vec3f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, uv, aspect, animTime, position, spread, intensity, streaks, flicker, seed, hot, mid, fringe) => {
	"use gpu";
	const t = __tsover_add(__tsover_mul(animTime, .6), __tsover_mul(seed, 7.9));
	const pA = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const anchor = d.vec2f(__tsover_mul(position.x, aspect), __tsover_sub(1, position.y));
	const toCenter = d.vec2f(__tsover_mul(aspect, .5), .5).sub(anchor);
	const cLen = std.length(toCenter);
	const dirX = std.select(__tsover_div(toCenter.x, std.max(cLen, .001)), d.f32(1), cLen < .02);
	const dirY = std.select(__tsover_div(toCenter.y, std.max(cLen, .001)), d.f32(0), cLen < .02);
	const relX = __tsover_sub(pA.x, anchor.x);
	const relY = __tsover_sub(pA.y, anchor.y);
	const u = __tsover_add(__tsover_mul(relX, dirX), __tsover_mul(relY, dirY));
	const v = __tsover_sub(__tsover_mul(relY, dirX), __tsover_mul(relX, dirY));
	const nA = value12(d.vec2f(__tsover_mul(t, .9), __tsover_add(__tsover_mul(seed, .53), 11.7)));
	const nB = value12(d.vec2f(__tsover_add(__tsover_mul(t, 2.7), 41.3), __tsover_add(__tsover_mul(seed, .29), 5.1)));
	const breathe = __tsover_add(__tsover_mul(nA, .65), __tsover_mul(nB, .35));
	const flickerMod = std.mix(1, __tsover_add(.25, __tsover_mul(1.5, breathe)), flicker);
	const evo1 = value12(d.vec2f(__tsover_add(__tsover_mul(t, .13), 3.1), __tsover_add(__tsover_mul(seed, .41), 27)));
	const evo2 = value12(d.vec2f(__tsover_add(__tsover_mul(t, .09), 17.7), __tsover_add(__tsover_mul(seed, .61), 8.4)));
	const evo3 = value12(d.vec2f(__tsover_add(__tsover_mul(t, .17), 51.2), __tsover_add(__tsover_mul(seed, .23), 2.9)));
	const sp = std.max(spread, .05);
	const elong = __tsover_mul(sp, __tsover_add(1.5, __tsover_mul(.7, evo2)));
	const du = __tsover_div(u, sp);
	const dv = __tsover_div(v, elong);
	const main = std.exp(__tsover_mul(__tsover_add(__tsover_mul(du, du), __tsover_mul(dv, dv)), -1.9));
	const alongFade = std.exp(__tsover_mul(__tsover_mul(dv, dv), -.55));
	const s1c = __tsover_mul(sp, __tsover_add(.42, __tsover_mul(.22, evo1)));
	const s1w = std.max(__tsover_mul(sp, __tsover_add(.07, __tsover_mul(.07, evo3))), .008);
	const streak1 = __tsover_mul(std.exp(__tsover_div(__tsover_mul(__tsover_mul(__tsover_sub(u, s1c), __tsover_sub(u, s1c)), -1), __tsover_mul(s1w, s1w))), alongFade);
	const s2c = __tsover_mul(sp, __tsover_add(.8, __tsover_mul(.3, evo1)));
	const s2w = std.max(__tsover_mul(sp, __tsover_add(.16, __tsover_mul(.12, evo3))), .015);
	const streak2 = __tsover_mul(__tsover_mul(std.exp(__tsover_div(__tsover_mul(__tsover_mul(__tsover_sub(u, s2c), __tsover_sub(u, s2c)), -1), __tsover_mul(s2w, s2w))), alongFade), .45);
	const streakSum = __tsover_mul(__tsover_add(streak1, streak2), streaks);
	const shoulder = __tsover_mul(std.exp(__tsover_mul(__tsover_add(__tsover_mul(du, du), __tsover_mul(dv, dv)), -.45)), .22);
	let heat = __tsover_mul(__tsover_mul(__tsover_mul(__tsover_add(__tsover_add(__tsover_mul(main, 1.25), shoulder), __tsover_mul(streakSum, .85)), intensity), 2.2), flickerMod);
	heat = __tsover_add(heat, __tsover_mul(__tsover_sub(hash12(uv.mul(1024).add(d.vec2f(t, __tsover_mul(t, 1.7)))), .5), .012));
	heat = std.max(heat, 0);
	const rampR = lightLeakRamp(__tsover_mul(heat, 1.08), hot, mid, fringe);
	const rampG = lightLeakRamp(heat, hot, mid, fringe);
	const rampB = lightLeakRamp(__tsover_mul(heat, .88), hot, mid, fringe);
	const leakRgb = d.vec3f(rampR.x, rampG.y, rampB.z);
	const base = color.xyz;
	const lifted = d.vec3f(1, 1, 1).sub(d.vec3f(1, 1, 1).sub(std.clamp(base, d.vec3f(0), d.vec3f(1))).mul(d.vec3f(1, 1, 1).sub(std.clamp(leakRgb, d.vec3f(0), d.vec3f(1))))).add(leakRgb.mul(.16));
	const outRgb = std.clamp(lifted, d.vec3f(0), d.vec3f(1));
	const leakLuma = std.clamp(std.dot(leakRgb, d.vec3f(.299, .587, .114)), 0, 1);
	const outA = std.clamp(__tsover_add(color.w, __tsover_mul(leakLuma, .85)), 0, 1);
	return d.vec4f(outRgb, outA);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "color"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "position"
			},
			{
				"type": "i",
				"name": "spread"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "streaks"
			},
			{
				"type": "i",
				"name": "flicker"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "hot"
			},
			{
				"type": "i",
				"name": "mid"
			},
			{
				"type": "i",
				"name": "fringe"
			}
		],
		"body": [0, [
			[
				13,
				"t",
				[
					1,
					[
						1,
						"animTime",
						"*",
						[5, "0.6"]
					],
					"+",
					[
						1,
						"seed",
						"*",
						[5, "7.9"]
					]
				]
			],
			[
				13,
				"pA",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"anchor",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"position",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"position",
							"y"
						]
					]]
				]
			],
			[
				13,
				"toCenter",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								"aspect",
								"*",
								[5, "0.5"]
							], [5, "0.5"]]
						],
						"sub"
					],
					["anchor"]
				]
			],
			[
				13,
				"cLen",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["toCenter"]
				]
			],
			[
				13,
				"dirX",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							1,
							[
								7,
								"toCenter",
								"x"
							],
							"/",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["cLen", [5, "0.001"]]
							]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"cLen",
							"<",
							[5, "0.02"]
						]
					]
				]
			],
			[
				13,
				"dirY",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							1,
							[
								7,
								"toCenter",
								"y"
							],
							"/",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["cLen", [5, "0.001"]]
							]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							1,
							"cLen",
							"<",
							[5, "0.02"]
						]
					]
				]
			],
			[
				13,
				"relX",
				[
					1,
					[
						7,
						"pA",
						"x"
					],
					"-",
					[
						7,
						"anchor",
						"x"
					]
				]
			],
			[
				13,
				"relY",
				[
					1,
					[
						7,
						"pA",
						"y"
					],
					"-",
					[
						7,
						"anchor",
						"y"
					]
				]
			],
			[
				13,
				"u",
				[
					1,
					[
						1,
						"relX",
						"*",
						"dirX"
					],
					"+",
					[
						1,
						"relY",
						"*",
						"dirY"
					]
				]
			],
			[
				13,
				"v",
				[
					1,
					[
						1,
						"relY",
						"*",
						"dirX"
					],
					"-",
					[
						1,
						"relX",
						"*",
						"dirY"
					]
				]
			],
			[
				13,
				"nA",
				[
					6,
					[
						7,
						"noise",
						"value12"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							"t",
							"*",
							[5, "0.9"]
						], [
							1,
							[
								1,
								"seed",
								"*",
								[5, "0.53"]
							],
							"+",
							[5, "11.7"]
						]]
					]]
				]
			],
			[
				13,
				"nB",
				[
					6,
					[
						7,
						"noise",
						"value12"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								1,
								"t",
								"*",
								[5, "2.7"]
							],
							"+",
							[5, "41.3"]
						], [
							1,
							[
								1,
								"seed",
								"*",
								[5, "0.29"]
							],
							"+",
							[5, "5.1"]
						]]
					]]
				]
			],
			[
				13,
				"breathe",
				[
					1,
					[
						1,
						"nA",
						"*",
						[5, "0.65"]
					],
					"+",
					[
						1,
						"nB",
						"*",
						[5, "0.35"]
					]
				]
			],
			[
				13,
				"flickerMod",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "1"],
						[
							1,
							[5, "0.25"],
							"+",
							[
								1,
								[5, "1.5"],
								"*",
								"breathe"
							]
						],
						"flicker"
					]
				]
			],
			[
				13,
				"evo1",
				[
					6,
					[
						7,
						"noise",
						"value12"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								1,
								"t",
								"*",
								[5, "0.13"]
							],
							"+",
							[5, "3.1"]
						], [
							1,
							[
								1,
								"seed",
								"*",
								[5, "0.41"]
							],
							"+",
							[5, "27"]
						]]
					]]
				]
			],
			[
				13,
				"evo2",
				[
					6,
					[
						7,
						"noise",
						"value12"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								1,
								"t",
								"*",
								[5, "0.09"]
							],
							"+",
							[5, "17.7"]
						], [
							1,
							[
								1,
								"seed",
								"*",
								[5, "0.61"]
							],
							"+",
							[5, "8.4"]
						]]
					]]
				]
			],
			[
				13,
				"evo3",
				[
					6,
					[
						7,
						"noise",
						"value12"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								1,
								"t",
								"*",
								[5, "0.17"]
							],
							"+",
							[5, "51.2"]
						], [
							1,
							[
								1,
								"seed",
								"*",
								[5, "0.23"]
							],
							"+",
							[5, "2.9"]
						]]
					]]
				]
			],
			[
				13,
				"sp",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["spread", [5, "0.05"]]
				]
			],
			[
				13,
				"elong",
				[
					1,
					"sp",
					"*",
					[
						1,
						[5, "1.5"],
						"+",
						[
							1,
							[5, "0.7"],
							"*",
							"evo2"
						]
					]
				]
			],
			[
				13,
				"du",
				[
					1,
					"u",
					"/",
					"sp"
				]
			],
			[
				13,
				"dv",
				[
					1,
					"v",
					"/",
					"elong"
				]
			],
			[
				13,
				"main",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							[
								1,
								"du",
								"*",
								"du"
							],
							"+",
							[
								1,
								"dv",
								"*",
								"dv"
							]
						],
						"*",
						[
							4,
							"-",
							[5, "1.9"]
						]
					]]
				]
			],
			[
				13,
				"alongFade",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							"dv",
							"*",
							"dv"
						],
						"*",
						[
							4,
							"-",
							[5, "0.55"]
						]
					]]
				]
			],
			[
				13,
				"s1c",
				[
					1,
					"sp",
					"*",
					[
						1,
						[5, "0.42"],
						"+",
						[
							1,
							[5, "0.22"],
							"*",
							"evo1"
						]
					]
				]
			],
			[
				13,
				"s1w",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"sp",
						"*",
						[
							1,
							[5, "0.07"],
							"+",
							[
								1,
								[5, "0.07"],
								"*",
								"evo3"
							]
						]
					], [5, "0.008"]]
				]
			],
			[
				13,
				"streak1",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"exp"
						],
						[[
							1,
							[
								1,
								[
									1,
									[
										1,
										"u",
										"-",
										"s1c"
									],
									"*",
									[
										1,
										"u",
										"-",
										"s1c"
									]
								],
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"/",
							[
								1,
								"s1w",
								"*",
								"s1w"
							]
						]]
					],
					"*",
					"alongFade"
				]
			],
			[
				13,
				"s2c",
				[
					1,
					"sp",
					"*",
					[
						1,
						[5, "0.8"],
						"+",
						[
							1,
							[5, "0.3"],
							"*",
							"evo1"
						]
					]
				]
			],
			[
				13,
				"s2w",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"sp",
						"*",
						[
							1,
							[5, "0.16"],
							"+",
							[
								1,
								[5, "0.12"],
								"*",
								"evo3"
							]
						]
					], [5, "0.015"]]
				]
			],
			[
				13,
				"streak2",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									1,
									[
										1,
										[
											1,
											"u",
											"-",
											"s2c"
										],
										"*",
										[
											1,
											"u",
											"-",
											"s2c"
										]
									],
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								],
								"/",
								[
									1,
									"s2w",
									"*",
									"s2w"
								]
							]]
						],
						"*",
						"alongFade"
					],
					"*",
					[5, "0.45"]
				]
			],
			[
				13,
				"streakSum",
				[
					1,
					[
						1,
						"streak1",
						"+",
						"streak2"
					],
					"*",
					"streaks"
				]
			],
			[
				13,
				"shoulder",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"exp"
						],
						[[
							1,
							[
								1,
								[
									1,
									"du",
									"*",
									"du"
								],
								"+",
								[
									1,
									"dv",
									"*",
									"dv"
								]
							],
							"*",
							[
								4,
								"-",
								[5, "0.45"]
							]
						]]
					],
					"*",
					[5, "0.22"]
				]
			],
			[
				12,
				"heat",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										"main",
										"*",
										[5, "1.25"]
									],
									"+",
									"shoulder"
								],
								"+",
								[
									1,
									"streakSum",
									"*",
									[5, "0.85"]
								]
							],
							"*",
							"intensity"
						],
						"*",
						[5, "2.2"]
					],
					"*",
					"flickerMod"
				]
			],
			[
				2,
				"heat",
				"=",
				[
					1,
					"heat",
					"+",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"noise",
									"hash12"
								],
								[[
									6,
									[
										7,
										[
											6,
											[
												7,
												"uv",
												"mul"
											],
											[[5, "1024"]]
										],
										"add"
									],
									[[
										6,
										[
											7,
											"d",
											"vec2f"
										],
										["t", [
											1,
											"t",
											"*",
											[5, "1.7"]
										]]
									]]
								]]
							],
							"-",
							[5, "0.5"]
						],
						"*",
						[5, "0.012"]
					]
				]
			],
			[
				2,
				"heat",
				"=",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["heat", [5, "0"]]
				]
			],
			[
				13,
				"rampR",
				[
					6,
					"lightLeakRamp",
					[
						[
							1,
							"heat",
							"*",
							[5, "1.08"]
						],
						"hot",
						"mid",
						"fringe"
					]
				]
			],
			[
				13,
				"rampG",
				[
					6,
					"lightLeakRamp",
					[
						"heat",
						"hot",
						"mid",
						"fringe"
					]
				]
			],
			[
				13,
				"rampB",
				[
					6,
					"lightLeakRamp",
					[
						[
							1,
							"heat",
							"*",
							[5, "0.88"]
						],
						"hot",
						"mid",
						"fringe"
					]
				]
			],
			[
				13,
				"leakRgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"rampR",
							"x"
						],
						[
							7,
							"rampG",
							"y"
						],
						[
							7,
							"rampB",
							"z"
						]
					]
				]
			],
			[
				13,
				"base",
				[
					7,
					"color",
					"xyz"
				]
			],
			[
				13,
				"screened",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1"],
								[5, "1"],
								[5, "1"]
							]
						],
						"sub"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									[
										6,
										[
											7,
											"d",
											"vec3f"
										],
										[
											[5, "1"],
											[5, "1"],
											[5, "1"]
										]
									],
									"sub"
								],
								[[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										"base",
										[
											6,
											[
												7,
												"d",
												"vec3f"
											],
											[[5, "0"]]
										],
										[
											6,
											[
												7,
												"d",
												"vec3f"
											],
											[[5, "1"]]
										]
									]
								]]
							],
							"mul"
						],
						[[
							6,
							[
								7,
								[
									6,
									[
										7,
										"d",
										"vec3f"
									],
									[
										[5, "1"],
										[5, "1"],
										[5, "1"]
									]
								],
								"sub"
							],
							[[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									"leakRgb",
									[
										6,
										[
											7,
											"d",
											"vec3f"
										],
										[[5, "0"]]
									],
									[
										6,
										[
											7,
											"d",
											"vec3f"
										],
										[[5, "1"]]
									]
								]
							]]
						]]
					]]
				]
			],
			[
				13,
				"lifted",
				[
					6,
					[
						7,
						"screened",
						"add"
					],
					[[
						6,
						[
							7,
							"leakRgb",
							"mul"
						],
						[[5, "0.16"]]
					]]
				]
			],
			[
				13,
				"outRgb",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"lifted",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[5, "1"]]
						]
					]
				]
			],
			[
				13,
				"leakLuma",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["leakRgb", [
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[5, "0.299"],
									[5, "0.587"],
									[5, "0.114"]
								]
							]]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"outA",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								7,
								"color",
								"w"
							],
							"+",
							[
								1,
								"leakLuma",
								"*",
								[5, "0.85"]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				["outRgb", "outA"]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"noise",
			"lightLeakRamp"
		]
	},
	externals: () => ({
		d,
		std,
		noise: noise_exports,
		lightLeakRamp
	})
}) && $.f)({}));
const componentDefinition = {
	name: "LightLeak",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Photorealistic film light leak — warm overexposed light bleeding in from a draggable anchor point, with streak bands, chromatic fringing, and slow breathing that evolves over time",
	requiresChild: true,
	usesPointer: false,
	animatedTime: { speed: "speed" },
	props: {
		position: {
			default: {
				x: .95,
				y: .4
			},
			transform: transformPosition,
			description: "Anchor point of the leak — the light beams from here toward the canvas centre",
			ui: {
				type: "position",
				label: "Position",
				group: "Leak",
				units: ["%", "px"]
			}
		},
		spread: {
			default: .55,
			description: "How far the light bleeds into the frame",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 1.5,
				step: .01,
				label: "Spread",
				group: "Leak"
			}
		},
		intensity: {
			default: .15,
			description: "Exposure strength of the leak",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Intensity",
				group: "Leak"
			}
		},
		streaks: {
			default: .6,
			description: "Strength of the parallel streak bands further into the frame",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Streaks",
				group: "Leak"
			}
		},
		colorHot: {
			default: "#fff3c4",
			transform: transformColor,
			description: "Core color where the leak is fully overexposed",
			ui: {
				type: "color",
				label: "Hot Core",
				group: "Colors"
			}
		},
		colorMid: {
			default: "#ff7a2f",
			transform: transformColor,
			description: "Main body color of the leak",
			ui: {
				type: "color",
				label: "Midtone",
				group: "Colors"
			}
		},
		colorFringe: {
			default: "#a63d8f",
			transform: transformColor,
			description: "Outer fringe color where the light fades out",
			ui: {
				type: "color",
				label: "Fringe",
				group: "Colors"
			}
		},
		flicker: {
			default: .35,
			description: "Breathing of the leak intensity",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Flicker",
				group: "Animation"
			}
		},
		speed: {
			default: 1,
			description: "Speed of the breathing, drift and evolution. 0 freezes the leak.",
			ui: {
				type: "range",
				min: 0,
				max: 4,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		seed: {
			default: 0,
			description: "Random seed — re-rolls the drift and breathing pattern",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Animation"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, childNode, ctx } = params;
		if (!childNode) {
			console.error("You must pass a child component into the Light Leak shader.");
			return ZERO;
		}
		const t = animatedTime(params);
		return call(lightLeakComposite, "lightLeakComposite", [
			childNode,
			ctx.uv,
			ctx.aspect,
			t,
			uniforms.position,
			uniforms.spread,
			uniforms.intensity,
			uniforms.streaks,
			uniforms.flicker,
			uniforms.seed,
			uniforms.colorHot.member("rgb"),
			uniforms.colorMid.member("rgb"),
			uniforms.colorFringe.member("rgb")
		]);
	}
};
var LightLeak_default = componentDefinition;
export { lightLeakRamp as i, componentDefinition as n, lightLeakComposite as r, LightLeak_default as t };
