// shaders/dist/core/LinearBlur-ClAsNnkC.js
import { At as call, Ot as ZERO, Pt as floatE, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { r as transformAngle } from "./transformations-fqzznt4g.js";
var rawWeights = Array.from({ length: 32 }, (_, i) => {
	const t = (i / 31 - .5) * 2;
	return Math.exp(-.5 * t * t / .64);
});
var wSum = rawWeights.reduce((a, b) => a + b);
const BLUR_WEIGHTS = rawWeights.map((w) => w / wSum);
const TAP_COUNT = 32;
var DEG_TO_RAD = Math.PI / 180;
const linearBlurTapCoord = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, angleDeg, intensity, viewport, tapT) => {
	"use gpu";
	const angleRad = __tsover_mul(angleDeg, DEG_TO_RAD);
	const aspect = __tsover_div(viewport.x, viewport.y);
	const dirX = __tsover_div(std.cos(angleRad), aspect);
	const dirY = std.sin(angleRad);
	const bvX = __tsover_mul(__tsover_div(__tsover_mul(dirX, intensity), viewport.x), 2);
	const bvY = __tsover_mul(__tsover_div(__tsover_mul(dirY, intensity), viewport.y), 2);
	return d.vec2f(__tsover_add(uv.x, __tsover_mul(bvX, tapT)), __tsover_add(uv.y, __tsover_mul(bvY, tapT)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "angleDeg"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "tapT"
			}
		],
		"body": [0, [
			[
				13,
				"angleRad",
				[
					1,
					"angleDeg",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"dirX",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						["angleRad"]
					],
					"/",
					"aspect"
				]
			],
			[
				13,
				"dirY",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"bvX",
				[
					1,
					[
						1,
						[
							1,
							"dirX",
							"*",
							"intensity"
						],
						"/",
						[
							7,
							"viewport",
							"x"
						]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"bvY",
				[
					1,
					[
						1,
						[
							1,
							"dirY",
							"*",
							"intensity"
						],
						"/",
						[
							7,
							"viewport",
							"y"
						]
					],
					"*",
					[5, "2"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						7,
						"uv",
						"x"
					],
					"+",
					[
						1,
						"bvX",
						"*",
						"tapT"
					]
				], [
					1,
					[
						7,
						"uv",
						"y"
					],
					"+",
					[
						1,
						"bvY",
						"*",
						"tapT"
					]
				]]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"std",
			"d"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "LinearBlur",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Blurs",
	description: "Directional motion blur in a specific angle",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 30,
			description: "Intensity of the linear blur effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Direction of the linear blur (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		let total;
		for (let i = 0; i < 32; i++) {
			const tapT = i / 31 - .5;
			const coord = call(linearBlurTapCoord, "linearBlurTapCoord", [
				ctx.uv,
				uniforms.angle,
				uniforms.intensity,
				ctx.viewportSize,
				floatE(tapT)
			]);
			const weighted = texture.sample(coord).mul(BLUR_WEIGHTS[i]);
			total = total ? total.add(weighted) : weighted;
		}
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [total]);
	}
};
var LinearBlur_default = componentDefinition;
export { linearBlurTapCoord as a, componentDefinition as i, LinearBlur_default as n, TAP_COUNT as r, BLUR_WEIGHTS as t };
