// shaders/dist/core/LinearGradient-RDNQes5r.js
import { At as call, Gt as colorMixing_exports, It as vec4, R as std, gt as mixColorStopsRuntime, ht as gradientStopsInSpace$1, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, r as transformAngle, s as transformEdges, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig } from "./porters-CoaznI7i.js";
const gradientStopsInSpace = gradientStopsInSpace$1;
var DEG_TO_RAD = .017453292519943295;
var TRANSPARENT = 1;
var MIRROR = 2;
var WRAP = 3;
const gradientRawT = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.vec2f,
	d.vec2f
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((start, end, angle, uv, viewport) => {
	"use gpu";
	const startPos = d.vec2f(start.x, __tsover_sub(1, start.y));
	const endPos = d.vec2f(end.x, __tsover_sub(1, end.y));
	const gradientVector = endPos.sub(startPos);
	const gradientLength = std.length(gradientVector);
	const gradientDir = std.normalize(gradientVector);
	const safeViewportY = std.max(viewport.y, 1e-6);
	const aspect = __tsover_div(viewport.x, safeViewportY);
	const angleRad = __tsover_mul(__tsover_mul(angle, DEG_TO_RAD), -1);
	const cosAngle = std.cos(angleRad);
	const sinAngle = std.sin(angleRad);
	const midpoint = startPos.add(endPos).mul(.5);
	const centeredUV = uv.sub(midpoint);
	const acX = __tsover_mul(centeredUV.x, aspect);
	const acY = centeredUV.y;
	const rotX = __tsover_sub(__tsover_mul(acX, cosAngle), __tsover_mul(acY, sinAngle));
	const rotY = __tsover_add(__tsover_mul(acX, sinAngle), __tsover_mul(acY, cosAngle));
	const relativePos = d.vec2f(__tsover_div(rotX, aspect), rotY).add(midpoint).sub(startPos);
	const projection = std.dot(relativePos, gradientDir);
	return __tsover_div(projection, std.max(gradientLength, 1e-6));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "start"
			},
			{
				"type": "i",
				"name": "end"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			}
		],
		"body": [0, [
			[
				13,
				"startPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"start",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"start",
							"y"
						]
					]]
				]
			],
			[
				13,
				"endPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"end",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"end",
							"y"
						]
					]]
				]
			],
			[
				13,
				"gradientVector",
				[
					6,
					[
						7,
						"endPos",
						"sub"
					],
					["startPos"]
				]
			],
			[
				13,
				"gradientLength",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["gradientVector"]
				]
			],
			[
				13,
				"gradientDir",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					["gradientVector"]
				]
			],
			[
				13,
				"safeViewportY",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						7,
						"viewport",
						"y"
					], [5, "0.000001"]]
				]
			],
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					"safeViewportY"
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					[
						1,
						"angle",
						"*",
						"DEG_TO_RAD"
					],
					"*",
					[
						4,
						"-",
						[5, "1"]
					]
				]
			],
			[
				13,
				"cosAngle",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinAngle",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"midpoint",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"startPos",
								"add"
							],
							["endPos"]
						],
						"mul"
					],
					[[5, "0.5"]]
				]
			],
			[
				13,
				"centeredUV",
				[
					6,
					[
						7,
						"uv",
						"sub"
					],
					["midpoint"]
				]
			],
			[
				13,
				"acX",
				[
					1,
					[
						7,
						"centeredUV",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"acY",
				[
					7,
					"centeredUV",
					"y"
				]
			],
			[
				13,
				"rotX",
				[
					1,
					[
						1,
						"acX",
						"*",
						"cosAngle"
					],
					"-",
					[
						1,
						"acY",
						"*",
						"sinAngle"
					]
				]
			],
			[
				13,
				"rotY",
				[
					1,
					[
						1,
						"acX",
						"*",
						"sinAngle"
					],
					"+",
					[
						1,
						"acY",
						"*",
						"cosAngle"
					]
				]
			],
			[
				13,
				"rotatedUV",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								"rotX",
								"/",
								"aspect"
							], "rotY"]
						],
						"add"
					],
					["midpoint"]
				]
			],
			[
				13,
				"relativePos",
				[
					6,
					[
						7,
						"rotatedUV",
						"sub"
					],
					["startPos"]
				]
			],
			[
				13,
				"projection",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["relativePos", "gradientDir"]
				]
			],
			[10, [
				1,
				"projection",
				"/",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["gradientLength", [5, "0.000001"]]
				]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"DEG_TO_RAD"
		]
	},
	externals: () => ({
		d,
		std,
		DEG_TO_RAD
	})
}) && $.f)({}));
const gradientEdgeMirror = tgpu.fn([d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((t) => {
	"use gpu";
	const m = std.mod(std.abs(t), 2);
	return std.select(m, __tsover_sub(2, m), m > 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "t"
		}],
		"body": [0, [[
			13,
			"m",
			[
				6,
				[
					7,
					"std",
					"mod"
				],
				[[
					6,
					[
						7,
						"std",
						"abs"
					],
					["t"]
				], [5, "2"]]
			]
		], [10, [
			6,
			[
				7,
				"std",
				"select"
			],
			[
				"m",
				[
					1,
					[5, "2"],
					"-",
					"m"
				],
				[
					1,
					"m",
					">",
					[5, "1"]
				]
			]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const gradientEdgeWrap = tgpu.fn([d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((t) => {
	"use gpu";
	return std.fract(t);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "t"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			["t"]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const gradientClamp01 = tgpu.fn([d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((t) => {
	"use gpu";
	return std.clamp(t, 0, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "t"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"clamp"
			],
			[
				"t",
				[5, "0"],
				[5, "1"]
			]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const gradientTransparentAlpha = tgpu.fn([d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((t) => {
	"use gpu";
	const inLow = std.select(d.f32(0), d.f32(1), t >= 0);
	const inHigh = std.select(d.f32(0), d.f32(1), t <= 1);
	return __tsover_mul(inLow, inHigh);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "t"
		}],
		"body": [0, [
			[
				13,
				"inLow",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"t",
							">=",
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"inHigh",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"t",
							"<=",
							[5, "1"]
						]
					]
				]
			],
			[10, [
				1,
				"inLow",
				"*",
				"inHigh"
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
var cm = colorMixing_exports;
function gradientColor(params, uv, viewport) {
	const { uniforms, propValues } = params;
	const edgeMode = propValues.edges ?? 0;
	const colorSpaceMode = propValues.colorSpace ?? 0;
	const stopCount = propValues.stopCount ?? 0;
	const t = call(gradientRawT, "gradientRawT", [
		uniforms.start,
		uniforms.end,
		uniforms.angle,
		uv,
		viewport
	]);
	let finalT;
	if (edgeMode === MIRROR) finalT = call(gradientEdgeMirror, "gradientEdgeMirror", [t]);
	else if (edgeMode === WRAP) finalT = call(gradientEdgeWrap, "gradientEdgeWrap", [t]);
	else if (edgeMode === TRANSPARENT) finalT = t;
	else finalT = call(gradientClamp01, "gradientClamp01", [t]);
	const clampedT = call(gradientClamp01, "gradientClamp01", [finalT]);
	let color;
	if (stopCount > 1) color = mixColorStopsRuntime(clampedT, {
		colorsArray: uniforms.colorsArray,
		positionsArray: uniforms.positionsArray,
		convertedColorsArray: uniforms.convertedColorsArray,
		stopCount: uniforms.stopCount
	}, colorSpaceMode);
	else color = call(cm.mixColorsVariants[colorSpaceMode] ?? cm.mixColorsLinear, "mixColors", [
		uniforms.colorA,
		uniforms.colorB,
		clampedT
	]);
	if (edgeMode === TRANSPARENT) {
		const cut = call(gradientTransparentAlpha, "gradientTransparentAlpha", [t]);
		return vec4(color.member("rgb"), color.member("a").mul(cut));
	}
	return color;
}
const componentDefinition = {
	name: "LinearGradient",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Create smooth linear color gradients",
	acceptsUVContext: true,
	props: {
		colorA: {
			default: "#1aff00",
			transform: transformColor,
			description: "The starting color of the gradient",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#0000ff",
			transform: transformColor,
			description: "The ending color of the gradient",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		stops: colorStopsPropConfig(),
		start: {
			default: {
				x: 0,
				y: .5
			},
			transform: transformPosition,
			description: "The starting point of the gradient",
			ui: {
				type: "position",
				label: "Start",
				group: "Position"
			}
		},
		end: {
			default: {
				x: 1,
				y: .5
			},
			transform: transformPosition,
			description: "The ending point of the gradient",
			ui: {
				type: "position",
				label: "End",
				group: "Position"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Additional rotation angle of the gradient (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			transform: transformEdges,
			compileTime: true,
			description: "How to handle areas beyond the gradient endpoints",
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		return gradientColor(params, params.uvContext ?? params.ctx.uv, params.effectiveViewportSize ?? params.ctx.viewportSize);
	}
};
var LinearGradient_default = componentDefinition;
export { gradientEdgeWrap as a, gradientTransparentAlpha as c, gradientEdgeMirror as i, componentDefinition as n, gradientRawT as o, gradientClamp01 as r, gradientStopsInSpace as s, LinearGradient_default as t };
