// shaders/dist/core/Liquify-CmkepjAk.js
import { At as call, B as createGuardedCompute, Nt as expr, Ot as ZERO, R as std, V as createStateBuffer, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-BlwgBpJf.js";
import { s as transformEdges } from "./transformations-fqzznt4g.js";
var GRID_SIZE = 64;
var CELL_COUNT = GRID_SIZE * GRID_SIZE;
var GRID_MAX = GRID_SIZE - 1;
var STATE_FORMAT = "rgba16float";
var springLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	readBuf: {
		storage: d.arrayOf(d.vec4f, CELL_COUNT),
		access: "readonly"
	},
	writeBuf: {
		storage: d.arrayOf(d.vec4f, CELL_COUNT),
		access: "mutable"
	},
	params: { uniform: /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
		cursorX: d.f32,
		cursorY: d.f32,
		dirX: d.f32,
		dirY: d.f32,
		clampedSpeed: d.f32,
		dt: d.f32,
		subDt: d.f32,
		stiffness: d.f32,
		dampFactor: d.f32,
		radius: d.f32,
		aspect: d.f32
	}), "uniform") }
}), "springLayout");
var outputLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	srcBuf: {
		storage: d.arrayOf(d.vec4f, CELL_COUNT),
		access: "readonly"
	},
	dispTex: { storageTexture: d.textureStorage2d(STATE_FORMAT, "write-only") }
}), "outputLayout");
const springKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	if (cx > d.u32(0) && cx < d.u32(GRID_MAX) && cy > d.u32(0) && cy < d.u32(GRID_MAX)) {
		const p = springLayout.$.params;
		const gs = d.u32(GRID_SIZE);
		const idx = __tsover_add(__tsover_mul(cy, gs), cx);
		const current = springLayout.$.readBuf[idx];
		const dispX = current.x;
		const dispY = current.y;
		const velX = current.z;
		const velY = current.w;
		const nL = springLayout.$.readBuf[__tsover_sub(idx, d.u32(1))];
		const nR = springLayout.$.readBuf[__tsover_add(idx, d.u32(1))];
		const nU = springLayout.$.readBuf[__tsover_sub(idx, gs)];
		const nD = springLayout.$.readBuf[__tsover_add(idx, gs)];
		const nUL = springLayout.$.readBuf[__tsover_sub(__tsover_sub(idx, gs), d.u32(1))];
		const nUR = springLayout.$.readBuf[__tsover_add(__tsover_sub(idx, gs), d.u32(1))];
		const nDL = springLayout.$.readBuf[__tsover_sub(__tsover_add(idx, gs), d.u32(1))];
		const nDR = springLayout.$.readBuf[__tsover_add(__tsover_add(idx, gs), d.u32(1))];
		let fx = __tsover_mul(p.stiffness, __tsover_add(__tsover_add(__tsover_add(__tsover_sub(nL.x, dispX), __tsover_sub(nR.x, dispX)), __tsover_sub(nU.x, dispX)), __tsover_sub(nD.x, dispX)));
		let fy = __tsover_mul(p.stiffness, __tsover_add(__tsover_add(__tsover_add(__tsover_sub(nL.y, dispY), __tsover_sub(nR.y, dispY)), __tsover_sub(nU.y, dispY)), __tsover_sub(nD.y, dispY)));
		const shearK = __tsover_mul(p.stiffness, .35);
		fx = __tsover_add(fx, __tsover_mul(shearK, __tsover_add(__tsover_add(__tsover_add(__tsover_sub(nUL.x, dispX), __tsover_sub(nUR.x, dispX)), __tsover_sub(nDL.x, dispX)), __tsover_sub(nDR.x, dispX))));
		fy = __tsover_add(fy, __tsover_mul(shearK, __tsover_add(__tsover_add(__tsover_add(__tsover_sub(nUL.y, dispY), __tsover_sub(nUR.y, dispY)), __tsover_sub(nDL.y, dispY)), __tsover_sub(nDR.y, dispY))));
		fx = __tsover_sub(fx, __tsover_mul(__tsover_mul(p.stiffness, .1), dispX));
		fy = __tsover_sub(fy, __tsover_mul(__tsover_mul(p.stiffness, .1), dispY));
		let newVelX = __tsover_add(__tsover_mul(velX, p.dampFactor), __tsover_mul(fx, p.subDt));
		let newVelY = __tsover_add(__tsover_mul(velY, p.dampFactor), __tsover_mul(fy, p.subDt));
		if (p.clampedSpeed > .01) {
			const cellX = __tsover_div(__tsover_add(d.f32(cx), .5), d.f32(GRID_SIZE));
			const cellY = __tsover_div(__tsover_add(d.f32(cy), .5), d.f32(GRID_SIZE));
			const ddx = std.select(__tsover_sub(cellX, p.cursorX), __tsover_mul(__tsover_sub(cellX, p.cursorX), p.aspect), p.aspect >= 1);
			const ddy = std.select(__tsover_div(__tsover_sub(cellY, p.cursorY), p.aspect), __tsover_sub(cellY, p.cursorY), p.aspect >= 1);
			const distSq = __tsover_add(__tsover_mul(ddx, ddx), __tsover_mul(ddy, ddy));
			const influenceRadius = __tsover_mul(p.radius, 3);
			if (distSq < __tsover_mul(influenceRadius, influenceRadius)) {
				const influence = std.exp(__tsover_mul(__tsover_div(distSq, __tsover_mul(p.radius, p.radius)), -1));
				const forceMag = __tsover_mul(__tsover_mul(__tsover_mul(influence, p.clampedSpeed), p.dt), 2);
				newVelX = __tsover_add(newVelX, __tsover_mul(p.dirX, forceMag));
				newVelY = __tsover_add(newVelY, __tsover_mul(p.dirY, forceMag));
			}
		}
		const newDispX = std.clamp(__tsover_add(dispX, __tsover_mul(newVelX, p.subDt)), -.5, .5);
		const newDispY = std.clamp(__tsover_add(dispY, __tsover_mul(newVelY, p.subDt)), -.5, .5);
		springLayout.$.writeBuf[idx] = d.vec4f(newDispX, newDispY, newVelX, newVelY);
	}
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [[
			11,
			[
				3,
				[
					3,
					[
						3,
						[
							1,
							"cx",
							">",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "0"]]
							]
						],
						"&&",
						[
							1,
							"cx",
							"<",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								["GRID_MAX"]
							]
						]
					],
					"&&",
					[
						1,
						"cy",
						">",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							[[5, "0"]]
						]
					]
				],
				"&&",
				[
					1,
					"cy",
					"<",
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						["GRID_MAX"]
					]
				]
			],
			[0, [
				[
					13,
					"p",
					[
						7,
						[
							7,
							"springLayout",
							"$"
						],
						"params"
					]
				],
				[
					13,
					"gs",
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						["GRID_SIZE"]
					]
				],
				[
					13,
					"idx",
					[
						1,
						[
							1,
							"cy",
							"*",
							"gs"
						],
						"+",
						"cx"
					]
				],
				[
					13,
					"current",
					[
						8,
						[
							7,
							[
								7,
								"springLayout",
								"$"
							],
							"readBuf"
						],
						"idx"
					]
				],
				[
					13,
					"dispX",
					[
						7,
						"current",
						"x"
					]
				],
				[
					13,
					"dispY",
					[
						7,
						"current",
						"y"
					]
				],
				[
					13,
					"velX",
					[
						7,
						"current",
						"z"
					]
				],
				[
					13,
					"velY",
					[
						7,
						"current",
						"w"
					]
				],
				[
					13,
					"nL",
					[
						8,
						[
							7,
							[
								7,
								"springLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							"idx",
							"-",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							]
						]
					]
				],
				[
					13,
					"nR",
					[
						8,
						[
							7,
							[
								7,
								"springLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							"idx",
							"+",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							]
						]
					]
				],
				[
					13,
					"nU",
					[
						8,
						[
							7,
							[
								7,
								"springLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							"idx",
							"-",
							"gs"
						]
					]
				],
				[
					13,
					"nD",
					[
						8,
						[
							7,
							[
								7,
								"springLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							"idx",
							"+",
							"gs"
						]
					]
				],
				[
					13,
					"nUL",
					[
						8,
						[
							7,
							[
								7,
								"springLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							[
								1,
								"idx",
								"-",
								"gs"
							],
							"-",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							]
						]
					]
				],
				[
					13,
					"nUR",
					[
						8,
						[
							7,
							[
								7,
								"springLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							[
								1,
								"idx",
								"-",
								"gs"
							],
							"+",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							]
						]
					]
				],
				[
					13,
					"nDL",
					[
						8,
						[
							7,
							[
								7,
								"springLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							[
								1,
								"idx",
								"+",
								"gs"
							],
							"-",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							]
						]
					]
				],
				[
					13,
					"nDR",
					[
						8,
						[
							7,
							[
								7,
								"springLayout",
								"$"
							],
							"readBuf"
						],
						[
							1,
							[
								1,
								"idx",
								"+",
								"gs"
							],
							"+",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								[[5, "1"]]
							]
						]
					]
				],
				[
					12,
					"fx",
					[
						1,
						[
							7,
							"p",
							"stiffness"
						],
						"*",
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										[
											7,
											"nL",
											"x"
										],
										"-",
										"dispX"
									],
									"+",
									[
										1,
										[
											7,
											"nR",
											"x"
										],
										"-",
										"dispX"
									]
								],
								"+",
								[
									1,
									[
										7,
										"nU",
										"x"
									],
									"-",
									"dispX"
								]
							],
							"+",
							[
								1,
								[
									7,
									"nD",
									"x"
								],
								"-",
								"dispX"
							]
						]
					]
				],
				[
					12,
					"fy",
					[
						1,
						[
							7,
							"p",
							"stiffness"
						],
						"*",
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										[
											7,
											"nL",
											"y"
										],
										"-",
										"dispY"
									],
									"+",
									[
										1,
										[
											7,
											"nR",
											"y"
										],
										"-",
										"dispY"
									]
								],
								"+",
								[
									1,
									[
										7,
										"nU",
										"y"
									],
									"-",
									"dispY"
								]
							],
							"+",
							[
								1,
								[
									7,
									"nD",
									"y"
								],
								"-",
								"dispY"
							]
						]
					]
				],
				[
					13,
					"shearK",
					[
						1,
						[
							7,
							"p",
							"stiffness"
						],
						"*",
						[5, "0.35"]
					]
				],
				[
					2,
					"fx",
					"=",
					[
						1,
						"fx",
						"+",
						[
							1,
							"shearK",
							"*",
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"nUL",
												"x"
											],
											"-",
											"dispX"
										],
										"+",
										[
											1,
											[
												7,
												"nUR",
												"x"
											],
											"-",
											"dispX"
										]
									],
									"+",
									[
										1,
										[
											7,
											"nDL",
											"x"
										],
										"-",
										"dispX"
									]
								],
								"+",
								[
									1,
									[
										7,
										"nDR",
										"x"
									],
									"-",
									"dispX"
								]
							]
						]
					]
				],
				[
					2,
					"fy",
					"=",
					[
						1,
						"fy",
						"+",
						[
							1,
							"shearK",
							"*",
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"nUL",
												"y"
											],
											"-",
											"dispY"
										],
										"+",
										[
											1,
											[
												7,
												"nUR",
												"y"
											],
											"-",
											"dispY"
										]
									],
									"+",
									[
										1,
										[
											7,
											"nDL",
											"y"
										],
										"-",
										"dispY"
									]
								],
								"+",
								[
									1,
									[
										7,
										"nDR",
										"y"
									],
									"-",
									"dispY"
								]
							]
						]
					]
				],
				[
					2,
					"fx",
					"=",
					[
						1,
						"fx",
						"-",
						[
							1,
							[
								1,
								[
									7,
									"p",
									"stiffness"
								],
								"*",
								[5, "0.1"]
							],
							"*",
							"dispX"
						]
					]
				],
				[
					2,
					"fy",
					"=",
					[
						1,
						"fy",
						"-",
						[
							1,
							[
								1,
								[
									7,
									"p",
									"stiffness"
								],
								"*",
								[5, "0.1"]
							],
							"*",
							"dispY"
						]
					]
				],
				[
					12,
					"newVelX",
					[
						1,
						[
							1,
							"velX",
							"*",
							[
								7,
								"p",
								"dampFactor"
							]
						],
						"+",
						[
							1,
							"fx",
							"*",
							[
								7,
								"p",
								"subDt"
							]
						]
					]
				],
				[
					12,
					"newVelY",
					[
						1,
						[
							1,
							"velY",
							"*",
							[
								7,
								"p",
								"dampFactor"
							]
						],
						"+",
						[
							1,
							"fy",
							"*",
							[
								7,
								"p",
								"subDt"
							]
						]
					]
				],
				[
					11,
					[
						1,
						[
							7,
							"p",
							"clampedSpeed"
						],
						">",
						[5, "0.01"]
					],
					[0, [
						[
							13,
							"cellX",
							[
								1,
								[
									1,
									[
										6,
										[
											7,
											"d",
											"f32"
										],
										["cx"]
									],
									"+",
									[5, "0.5"]
								],
								"/",
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									["GRID_SIZE"]
								]
							]
						],
						[
							13,
							"cellY",
							[
								1,
								[
									1,
									[
										6,
										[
											7,
											"d",
											"f32"
										],
										["cy"]
									],
									"+",
									[5, "0.5"]
								],
								"/",
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									["GRID_SIZE"]
								]
							]
						],
						[
							13,
							"ddx",
							[
								6,
								[
									7,
									"std",
									"select"
								],
								[
									[
										1,
										"cellX",
										"-",
										[
											7,
											"p",
											"cursorX"
										]
									],
									[
										1,
										[
											1,
											"cellX",
											"-",
											[
												7,
												"p",
												"cursorX"
											]
										],
										"*",
										[
											7,
											"p",
											"aspect"
										]
									],
									[
										1,
										[
											7,
											"p",
											"aspect"
										],
										">=",
										[5, "1"]
									]
								]
							]
						],
						[
							13,
							"ddy",
							[
								6,
								[
									7,
									"std",
									"select"
								],
								[
									[
										1,
										[
											1,
											"cellY",
											"-",
											[
												7,
												"p",
												"cursorY"
											]
										],
										"/",
										[
											7,
											"p",
											"aspect"
										]
									],
									[
										1,
										"cellY",
										"-",
										[
											7,
											"p",
											"cursorY"
										]
									],
									[
										1,
										[
											7,
											"p",
											"aspect"
										],
										">=",
										[5, "1"]
									]
								]
							]
						],
						[
							13,
							"distSq",
							[
								1,
								[
									1,
									"ddx",
									"*",
									"ddx"
								],
								"+",
								[
									1,
									"ddy",
									"*",
									"ddy"
								]
							]
						],
						[
							13,
							"influenceRadius",
							[
								1,
								[
									7,
									"p",
									"radius"
								],
								"*",
								[5, "3"]
							]
						],
						[
							11,
							[
								1,
								"distSq",
								"<",
								[
									1,
									"influenceRadius",
									"*",
									"influenceRadius"
								]
							],
							[0, [
								[
									13,
									"influence",
									[
										6,
										[
											7,
											"std",
											"exp"
										],
										[[
											1,
											[
												1,
												"distSq",
												"/",
												[
													1,
													[
														7,
														"p",
														"radius"
													],
													"*",
													[
														7,
														"p",
														"radius"
													]
												]
											],
											"*",
											[
												4,
												"-",
												[5, "1"]
											]
										]]
									]
								],
								[
									13,
									"forceMag",
									[
										1,
										[
											1,
											[
												1,
												"influence",
												"*",
												[
													7,
													"p",
													"clampedSpeed"
												]
											],
											"*",
											[
												7,
												"p",
												"dt"
											]
										],
										"*",
										[5, "2"]
									]
								],
								[
									2,
									"newVelX",
									"=",
									[
										1,
										"newVelX",
										"+",
										[
											1,
											[
												7,
												"p",
												"dirX"
											],
											"*",
											"forceMag"
										]
									]
								],
								[
									2,
									"newVelY",
									"=",
									[
										1,
										"newVelY",
										"+",
										[
											1,
											[
												7,
												"p",
												"dirY"
											],
											"*",
											"forceMag"
										]
									]
								]
							]]
						]
					]]
				],
				[
					13,
					"newDispX",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								"dispX",
								"+",
								[
									1,
									"newVelX",
									"*",
									[
										7,
										"p",
										"subDt"
									]
								]
							],
							[
								4,
								"-",
								[5, "0.5"]
							],
							[5, "0.5"]
						]
					]
				],
				[
					13,
					"newDispY",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								"dispY",
								"+",
								[
									1,
									"newVelY",
									"*",
									[
										7,
										"p",
										"subDt"
									]
								]
							],
							[
								4,
								"-",
								[5, "0.5"]
							],
							[5, "0.5"]
						]
					]
				],
				[
					2,
					[
						8,
						[
							7,
							[
								7,
								"springLayout",
								"$"
							],
							"writeBuf"
						],
						"idx"
					],
					"=",
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							"newDispX",
							"newDispY",
							"newVelX",
							"newVelY"
						]
					]
				]
			]]
		]]],
		"externalNames": [
			"d",
			"GRID_MAX",
			"springLayout",
			"GRID_SIZE",
			"std"
		]
	},
	externals: () => ({
		d,
		GRID_MAX,
		springLayout,
		GRID_SIZE,
		std
	})
}) && $.f)({})).$name("liquifySpring");
const outputKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(GRID_SIZE)), cx);
	const cell = outputLayout.$.srcBuf[idx];
	std.textureStore(outputLayout.$.dispTex, d.vec2u(cx, cy), d.vec4f(cell.x, cell.y, 0, 1));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["GRID_SIZE"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"cell",
				[
					8,
					[
						7,
						[
							7,
							"outputLayout",
							"$"
						],
						"srcBuf"
					],
					"idx"
				]
			],
			[
				6,
				[
					7,
					"std",
					"textureStore"
				],
				[
					[
						7,
						[
							7,
							"outputLayout",
							"$"
						],
						"dispTex"
					],
					[
						6,
						[
							7,
							"d",
							"vec2u"
						],
						["cx", "cy"]
					],
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							[
								7,
								"cell",
								"x"
							],
							[
								7,
								"cell",
								"y"
							],
							[5, "0"],
							[5, "1"]
						]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"GRID_SIZE",
			"outputLayout",
			"std"
		]
	},
	externals: () => ({
		d,
		GRID_SIZE,
		outputLayout,
		std
	})
}) && $.f)({})).$name("liquifyOutput");
const liquifyOffsetUV = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, disp, intensity) => {
	"use gpu";
	const scaled = disp.mul(__tsover_mul(intensity, .1));
	const maxDisp = .15;
	const clamped = std.clamp(scaled, d.vec2f(-maxDisp, -maxDisp), d.vec2f(maxDisp, maxDisp));
	return uv.sub(clamped);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "disp"
			},
			{
				"type": "i",
				"name": "intensity"
			}
		],
		"body": [0, [
			[
				13,
				"scaled",
				[
					6,
					[
						7,
						"disp",
						"mul"
					],
					[[
						1,
						"intensity",
						"*",
						[5, "0.1"]
					]]
				]
			],
			[
				13,
				"maxDisp",
				[5, "0.15"]
			],
			[
				13,
				"clamped",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"scaled",
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								4,
								"-",
								"maxDisp"
							], [
								4,
								"-",
								"maxDisp"
							]]
						],
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["maxDisp", "maxDisp"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"uv",
					"sub"
				],
				["clamped"]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Liquify",
	category: "Interactive",
	description: "Liquid-like interactive deformation effect",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	usesPointer: true,
	props: {
		intensity: {
			default: 10,
			description: "Scale of the fabric displacement effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		stiffness: {
			default: 3,
			description: "Fabric rigidity (higher = stiffer canvas, lower = stretchy silk)",
			ui: {
				type: "range",
				min: 1,
				max: 30,
				step: .5,
				label: "Stiffness",
				group: "Effect"
			}
		},
		damping: {
			default: 3,
			description: "How quickly fabric motion settles",
			ui: {
				type: "range",
				min: 0,
				max: 10,
				step: .1,
				label: "Damping",
				group: "Effect"
			}
		},
		radius: {
			default: 1,
			description: "Cursor influence area",
			ui: {
				type: "range",
				min: .1,
				max: 1.5,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	compute: (params) => {
		const { gpu, registerComputeTexture, getCpuValue, onCleanup, onResize, dimensions } = params;
		const root = gpu?.root;
		if (!root) return null;
		const bufferA = createStateBuffer(root, d.vec4f, CELL_COUNT);
		const bufferB = createStateBuffer(root, d.vec4f, CELL_COUNT);
		const dispTex = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
			size: [GRID_SIZE, GRID_SIZE],
			format: STATE_FORMAT
		}).$usage("storage", "sampled"), "dispTex");
		onCleanup(() => dispTex.destroy());
		const displacement = registerComputeTexture(dispTex);
		const springParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(d.struct({
			cursorX: d.f32,
			cursorY: d.f32,
			dirX: d.f32,
			dirY: d.f32,
			clampedSpeed: d.f32,
			dt: d.f32,
			subDt: d.f32,
			stiffness: d.f32,
			dampFactor: d.f32,
			radius: d.f32,
			aspect: d.f32
		})), "springParams");
		let curW = Math.max(1, Math.round(dimensions.width));
		let curH = Math.max(1, Math.round(dimensions.height));
		onResize(({ width, height }) => {
			curW = Math.max(1, Math.round(width));
			curH = Math.max(1, Math.round(height));
		});
		const size = [GRID_SIZE, GRID_SIZE];
		const springBgAB = root.createBindGroup(springLayout, {
			readBuf: bufferA,
			writeBuf: bufferB,
			params: springParams.buffer
		});
		const springBgBA = root.createBindGroup(springLayout, {
			readBuf: bufferB,
			writeBuf: bufferA,
			params: springParams.buffer
		});
		const spring = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			springKernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"springKernel",
					["cx", "cy"]
				]]],
				"externalNames": ["springKernel"]
			},
			externals: () => ({ springKernel })
		}) && $.f)({}), { size });
		const outputBgA = root.createBindGroup(outputLayout, {
			srcBuf: bufferA,
			dispTex
		});
		const output = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			outputKernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"outputKernel",
					["cx", "cy"]
				]]],
				"externalNames": ["outputKernel"]
			},
			externals: () => ({ outputKernel })
		}) && $.f)({}), {
			size,
			bindGroup: outputBgA
		});
		let prevX = .5;
		let prevY = .5;
		let lastActiveTime = Date.now();
		return {
			outputs: { displacement },
			getComputeNodes: (frameParams) => {
				const fp = frameParams;
				const currentTime = Date.now();
				const dt = Math.min(fp.deltaTime ?? 0, .016);
				if (dt <= 0) return null;
				const aspect = Math.max(1, fp.dimensions?.width ?? curW) / Math.max(1, fp.dimensions?.height ?? curH);
				const stiffness = getCpuValue("stiffness") ?? 3;
				const damping = getCpuValue("damping") ?? 3;
				const radius = (getCpuValue("radius") ?? 1) * .08;
				const px = fp.pointer?.x ?? .5;
				const py = fp.pointer?.y ?? .5;
				const rawVelX = (px - prevX) / dt;
				const rawVelY = (py - prevY) / dt;
				const speed = Math.sqrt(rawVelX * rawVelX + rawVelY * rawVelY);
				const clampedSpeed = Math.min(speed, 3);
				prevX = px;
				prevY = py;
				if (clampedSpeed > .01) lastActiveTime = currentTime;
				const settleMs = damping > 0 ? Math.min(3e4, 5e3 / damping) : 3e4;
				if (currentTime - lastActiveTime > settleMs) return null;
				const subDt = dt / 2;
				springParams.write({
					cursorX: px,
					cursorY: py,
					dirX: speed > .01 ? rawVelX / speed : 0,
					dirY: speed > .01 ? rawVelY / speed : 0,
					clampedSpeed,
					dt,
					subDt,
					stiffness,
					dampFactor: Math.max(0, Math.min(1, 1 - damping * subDt)),
					radius,
					aspect
				});
				return [
					spring.with(springBgAB),
					spring.with(springBgBA),
					output
				];
			}
		};
	},
	fragment: ({ childNode, ctx, computeOutputs, propValues, convertToTexture, uniforms }) => {
		if (!childNode) return ZERO;
		const childTex = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		const dispTex = computeOutputs?.displacement;
		const disp = dispTex ? dispTex.sample(ctx.uv, "linearClamp").member("xy") : expr("vec2f(0.0, 0.0)");
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(liquifyOffsetUV, "liquifyOffsetUV", [
			ctx.uv,
			disp,
			uniforms.intensity
		]), (uv) => childTex.sample(uv), edgeMode)]);
	},
	uvRemap: ({ uv, mask, uniforms, propValues, computeOutputs }) => {
		const dispTex = computeOutputs?.displacement;
		if (!dispTex) return {
			uv,
			mask
		};
		const edgeMode = propValues.edges ?? 0;
		return composeEdgeRemapExpr(call(liquifyOffsetUV, "liquifyOffsetUV", [
			uv,
			dispTex.sample(uv, "linearClamp").member("xy"),
			uniforms.intensity
		]), mask, edgeMode);
	}
};
var Liquify_default = componentDefinition;
export { springKernel as a, outputKernel as i, componentDefinition as n, liquifyOffsetUV as r, Liquify_default as t };
