// shaders/dist/core/Marble-D_a2ctfR.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, et as mxNoiseFloat2, n as tgpu, on as d, rt as noise_exports } from "./coords-BlwgBpJf.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
var PI = Math.PI;
const marbleFactors = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, scale, turbulence, seed, animTime) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const au = __tsover_mul(uv.x, aspect);
	const av = uv.y;
	const n1 = mxNoiseFloat2(d.vec2f(__tsover_add(__tsover_mul(au, scale), seed), __tsover_add(__tsover_mul(av, scale), animTime)));
	const n2 = __tsover_mul(mxNoiseFloat2(d.vec2f(__tsover_add(__tsover_mul(__tsover_mul(au, scale), 2), __tsover_mul(seed, 1.7)), __tsover_add(__tsover_mul(__tsover_mul(av, scale), 2), __tsover_mul(animTime, .7)))), .5);
	const n3 = __tsover_mul(mxNoiseFloat2(d.vec2f(__tsover_add(__tsover_mul(__tsover_mul(au, scale), 4), __tsover_mul(seed, 3.1)), __tsover_add(__tsover_mul(__tsover_mul(av, scale), 4), __tsover_mul(animTime, 1.3)))), .25);
	const totalNoise = __tsover_add(__tsover_add(n1, n2), n3);
	const marbleCoord = __tsover_add(__tsover_mul(__tsover_mul(au, scale), PI), __tsover_mul(totalNoise, turbulence));
	const marbleT = __tsover_add(__tsover_mul(std.sin(marbleCoord), .5), .5);
	const n4 = __tsover_add(__tsover_mul(mxNoiseFloat2(d.vec2f(__tsover_add(__tsover_mul(__tsover_mul(av, scale), 1.5), __tsover_mul(seed, 2)), __tsover_mul(__tsover_mul(au, scale), 1.2))), .3), .5);
	const t1 = std.smoothstep(.3, .6, marbleT);
	const t2 = std.smoothstep(.6, .9, marbleT);
	return d.vec2f(t1, __tsover_mul(t2, n4));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "turbulence"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"au",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"av",
				[
					7,
					"uv",
					"y"
				]
			],
			[
				13,
				"n1",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat2"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								1,
								"au",
								"*",
								"scale"
							],
							"+",
							"seed"
						], [
							1,
							[
								1,
								"av",
								"*",
								"scale"
							],
							"+",
							"animTime"
						]]
					]]
				]
			],
			[
				13,
				"n2",
				[
					1,
					[
						6,
						[
							7,
							"noise",
							"mxNoiseFloat2"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								[
									1,
									[
										1,
										"au",
										"*",
										"scale"
									],
									"*",
									[5, "2"]
								],
								"+",
								[
									1,
									"seed",
									"*",
									[5, "1.7"]
								]
							], [
								1,
								[
									1,
									[
										1,
										"av",
										"*",
										"scale"
									],
									"*",
									[5, "2"]
								],
								"+",
								[
									1,
									"animTime",
									"*",
									[5, "0.7"]
								]
							]]
						]]
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"n3",
				[
					1,
					[
						6,
						[
							7,
							"noise",
							"mxNoiseFloat2"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								[
									1,
									[
										1,
										"au",
										"*",
										"scale"
									],
									"*",
									[5, "4"]
								],
								"+",
								[
									1,
									"seed",
									"*",
									[5, "3.1"]
								]
							], [
								1,
								[
									1,
									[
										1,
										"av",
										"*",
										"scale"
									],
									"*",
									[5, "4"]
								],
								"+",
								[
									1,
									"animTime",
									"*",
									[5, "1.3"]
								]
							]]
						]]
					],
					"*",
					[5, "0.25"]
				]
			],
			[
				13,
				"totalNoise",
				[
					1,
					[
						1,
						"n1",
						"+",
						"n2"
					],
					"+",
					"n3"
				]
			],
			[
				13,
				"marbleCoord",
				[
					1,
					[
						1,
						[
							1,
							"au",
							"*",
							"scale"
						],
						"*",
						"PI"
					],
					"+",
					[
						1,
						"totalNoise",
						"*",
						"turbulence"
					]
				]
			],
			[
				13,
				"marbleT",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							["marbleCoord"]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"n4",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"noise",
								"mxNoiseFloat2"
							],
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									1,
									[
										1,
										[
											1,
											"av",
											"*",
											"scale"
										],
										"*",
										[5, "1.5"]
									],
									"+",
									[
										1,
										"seed",
										"*",
										[5, "2"]
									]
								], [
									1,
									[
										1,
										"au",
										"*",
										"scale"
									],
									"*",
									[5, "1.2"]
								]]
							]]
						],
						"*",
						[5, "0.3"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"t1",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.3"],
						[5, "0.6"],
						"marbleT"
					]
				]
			],
			[
				13,
				"t2",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.6"],
						[5, "0.9"],
						"marbleT"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["t1", [
					1,
					"t2",
					"*",
					"n4"
				]]
			]]
		]],
		"externalNames": [
			"noise",
			"d",
			"PI",
			"std"
		]
	},
	externals: () => ({
		noise: noise_exports,
		d,
		PI,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Marble",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Classic marble swirl and vein texture using noise-warped sine waves",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#ffffff",
			transform: transformColor,
			description: "Base background color of the marble",
			ui: {
				type: "color",
				label: "Base",
				group: "Colors"
			}
		},
		colorB: {
			default: "#3a2d54",
			transform: transformColor,
			description: "Secondary marble tone",
			ui: {
				type: "color",
				label: "Tone",
				group: "Colors"
			}
		},
		colorC: {
			default: "#0f0f0f",
			transform: transformColor,
			description: "Deepest marble color",
			ui: {
				type: "color",
				label: "Depth",
				group: "Colors"
			}
		},
		scale: {
			default: 2,
			description: "Scale and density of the marble vein pattern",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 10,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		turbulence: {
			default: 10,
			description: "Amount of noise-driven distortion applied to the veins",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 50,
				step: .5,
				label: "Turbulence",
				group: "Effect"
			}
		},
		speed: {
			default: .05,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: .25,
				step: .005,
				label: "Speed",
				group: "Animation"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for pattern variation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const uv = params.uvContext ?? ctx.uv;
		const viewport = params.effectiveViewportSize ?? ctx.viewportSize;
		const animTime = animatedTime(params);
		const factors = call(marbleFactors, "marbleFactors", [
			uv,
			viewport,
			uniforms.scale,
			uniforms.turbulence,
			uniforms.seed,
			animTime
		]);
		const t1 = factors.member("x");
		const t2n4 = factors.member("y");
		const variant = mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear;
		return call(variant, "mixColors", [
			call(variant, "mixColors", [
				uniforms.colorA,
				uniforms.colorB,
				t1
			]),
			uniforms.colorC,
			t2n4
		]);
	}
};
var Marble_default = componentDefinition;
export { componentDefinition as n, marbleFactors as r, Marble_default as t };
