// shaders/dist/core/Mirror-B98LBGCI.js
import { At as call, Ft as mixExpr, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-BlwgBpJf.js";
import { c as transformPosition, s as transformEdges } from "./transformations-fqzznt4g.js";
var DEG_TO_RAD = Math.PI / 180;
const mirrorReflect = tgpu.fn([
	d.vec2f,
	d.f32,
	d.vec2f,
	d.f32
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, angle, uv, aspect) => {
	"use gpu";
	const aspectCorrectedUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y));
	const delta = aspectCorrectedUV.sub(centerPos);
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const lineDirection = d.vec2f(std.cos(angleRad), std.sin(angleRad));
	const lineNormal = d.vec2f(__tsover_mul(lineDirection.y, -1), lineDirection.x);
	const signedDistance = std.dot(delta, lineNormal);
	const reflectedDelta = delta.sub(lineNormal.mul(__tsover_mul(signedDistance, 2)));
	const mirroredUV = centerPos.add(reflectedDelta);
	const finalMirroredUV = d.vec2f(__tsover_div(mirroredUV.x, aspect), mirroredUV.y);
	const shouldMirror = std.step(0, signedDistance);
	return d.vec3f(finalMirroredUV.x, finalMirroredUV.y, shouldMirror);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"aspectCorrectedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"delta",
				[
					6,
					[
						7,
						"aspectCorrectedUV",
						"sub"
					],
					["centerPos"]
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"lineDirection",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"cos"
						],
						["angleRad"]
					], [
						6,
						[
							7,
							"std",
							"sin"
						],
						["angleRad"]
					]]
				]
			],
			[
				13,
				"lineNormal",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"lineDirection",
							"y"
						],
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					], [
						7,
						"lineDirection",
						"x"
					]]
				]
			],
			[
				13,
				"signedDistance",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["delta", "lineNormal"]
				]
			],
			[
				13,
				"reflectedDelta",
				[
					6,
					[
						7,
						"delta",
						"sub"
					],
					[[
						6,
						[
							7,
							"lineNormal",
							"mul"
						],
						[[
							1,
							"signedDistance",
							"*",
							[5, "2"]
						]]
					]]
				]
			],
			[
				13,
				"mirroredUV",
				[
					6,
					[
						7,
						"centerPos",
						"add"
					],
					["reflectedDelta"]
				]
			],
			[
				13,
				"finalMirroredUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"mirroredUV",
							"x"
						],
						"/",
						"aspect"
					], [
						7,
						"mirroredUV",
						"y"
					]]
				]
			],
			[
				13,
				"shouldMirror",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					[[5, "0"], "signedDistance"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[
						7,
						"finalMirroredUV",
						"x"
					],
					[
						7,
						"finalMirroredUV",
						"y"
					],
					"shouldMirror"
				]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Mirror",
	category: "Distortions",
	description: "Mirror content across a line defined by center point and angle",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The point the mirror line passes through",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		angle: {
			default: 0,
			description: "The angle of the mirror line in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, propValues, convertToTexture }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Mirror shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		const m = call(mirrorReflect, "mirrorReflect", [
			uniforms.center,
			uniforms.angle,
			ctx.uv,
			ctx.aspect
		]);
		const mirroredUV = m.member("xy");
		const shouldMirror = m.member("z");
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [mixExpr(texture.sample(ctx.uv), applyEdgeHandlingExpr(mirroredUV, (uv) => texture.sample(uv), edgeMode), shouldMirror)]);
	},
	uvRemap: ({ uv, mask, uniforms, aspect, propValues }) => {
		const edgeMode = propValues.edges ?? 0;
		const m = call(mirrorReflect, "mirrorReflect", [
			uniforms.center,
			uniforms.angle,
			uv,
			aspect
		]);
		return composeEdgeRemapExpr(mixExpr(uv, m.member("xy"), m.member("z")), mask, edgeMode);
	}
};
var Mirror_default = componentDefinition;
export { componentDefinition as n, mirrorReflect as r, Mirror_default as t };
