// shaders/dist/core/MultiPointGradient-IJtw3dEC.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
var mpgWeight = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uvAspect, pos, effPower) => {
	"use gpu";
	const dx = __tsover_sub(uvAspect.x, pos.x);
	const dy = __tsover_sub(uvAspect.y, pos.y);
	const dist = __tsover_add(std.sqrt(__tsover_add(__tsover_mul(dx, dx), __tsover_mul(dy, dy))), .001);
	return __tsover_div(1, std.pow(dist, effPower));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uvAspect"
			},
			{
				"type": "i",
				"name": "pos"
			},
			{
				"type": "i",
				"name": "effPower"
			}
		],
		"body": [0, [
			[
				13,
				"dx",
				[
					1,
					[
						7,
						"uvAspect",
						"x"
					],
					"-",
					[
						7,
						"pos",
						"x"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"uvAspect",
						"y"
					],
					"-",
					[
						7,
						"pos",
						"y"
					]
				]
			],
			[
				13,
				"dist",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"dx",
								"*",
								"dx"
							],
							"+",
							[
								1,
								"dy",
								"*",
								"dy"
							]
						]]
					],
					"+",
					[5, "0.001"]
				]
			],
			[10, [
				1,
				[5, "1"],
				"/",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					["dist", "effPower"]
				]
			]]
		]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const mpgFactors = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, vp, posA, posB, posC, posD, posE, smoothness) => {
	"use gpu";
	const aspect = __tsover_div(vp.x, vp.y);
	const uvAspect = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const pA = d.vec2f(__tsover_mul(posA.x, aspect), __tsover_sub(1, posA.y));
	const pB = d.vec2f(__tsover_mul(posB.x, aspect), __tsover_sub(1, posB.y));
	const pC = d.vec2f(__tsover_mul(posC.x, aspect), __tsover_sub(1, posC.y));
	const pD = d.vec2f(__tsover_mul(posD.x, aspect), __tsover_sub(1, posD.y));
	const pE = d.vec2f(__tsover_mul(posE.x, aspect), __tsover_sub(1, posE.y));
	const effPower = std.min(__tsover_div(8, __tsover_add(smoothness, .5)), 8);
	const wA = mpgWeight(uvAspect, pA, effPower);
	const wB = mpgWeight(uvAspect, pB, effPower);
	const wC = mpgWeight(uvAspect, pC, effPower);
	const wD = mpgWeight(uvAspect, pD, effPower);
	const wE = mpgWeight(uvAspect, pE, effPower);
	const accW1 = wA;
	const t1 = __tsover_div(wB, __tsover_add(accW1, wB));
	const accW2 = __tsover_add(accW1, wB);
	const t2 = __tsover_div(wC, __tsover_add(accW2, wC));
	const accW3 = __tsover_add(accW2, wC);
	const t3 = __tsover_div(wD, __tsover_add(accW3, wD));
	const accW4 = __tsover_add(accW3, wD);
	const t4 = __tsover_div(wE, __tsover_add(accW4, wE));
	return d.vec4f(t1, t2, t3, t4);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "vp"
			},
			{
				"type": "i",
				"name": "posA"
			},
			{
				"type": "i",
				"name": "posB"
			},
			{
				"type": "i",
				"name": "posC"
			},
			{
				"type": "i",
				"name": "posD"
			},
			{
				"type": "i",
				"name": "posE"
			},
			{
				"type": "i",
				"name": "smoothness"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"vp",
						"x"
					],
					"/",
					[
						7,
						"vp",
						"y"
					]
				]
			],
			[
				13,
				"uvAspect",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"pA",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"posA",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"posA",
							"y"
						]
					]]
				]
			],
			[
				13,
				"pB",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"posB",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"posB",
							"y"
						]
					]]
				]
			],
			[
				13,
				"pC",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"posC",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"posC",
							"y"
						]
					]]
				]
			],
			[
				13,
				"pD",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"posD",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"posD",
							"y"
						]
					]]
				]
			],
			[
				13,
				"pE",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"posE",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"posE",
							"y"
						]
					]]
				]
			],
			[
				13,
				"effPower",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					[[
						1,
						[5, "8"],
						"/",
						[
							1,
							"smoothness",
							"+",
							[5, "0.5"]
						]
					], [5, "8"]]
				]
			],
			[
				13,
				"wA",
				[
					6,
					"mpgWeight",
					[
						"uvAspect",
						"pA",
						"effPower"
					]
				]
			],
			[
				13,
				"wB",
				[
					6,
					"mpgWeight",
					[
						"uvAspect",
						"pB",
						"effPower"
					]
				]
			],
			[
				13,
				"wC",
				[
					6,
					"mpgWeight",
					[
						"uvAspect",
						"pC",
						"effPower"
					]
				]
			],
			[
				13,
				"wD",
				[
					6,
					"mpgWeight",
					[
						"uvAspect",
						"pD",
						"effPower"
					]
				]
			],
			[
				13,
				"wE",
				[
					6,
					"mpgWeight",
					[
						"uvAspect",
						"pE",
						"effPower"
					]
				]
			],
			[
				13,
				"accW1",
				"wA"
			],
			[
				13,
				"t1",
				[
					1,
					"wB",
					"/",
					[
						1,
						"accW1",
						"+",
						"wB"
					]
				]
			],
			[
				13,
				"accW2",
				[
					1,
					"accW1",
					"+",
					"wB"
				]
			],
			[
				13,
				"t2",
				[
					1,
					"wC",
					"/",
					[
						1,
						"accW2",
						"+",
						"wC"
					]
				]
			],
			[
				13,
				"accW3",
				[
					1,
					"accW2",
					"+",
					"wC"
				]
			],
			[
				13,
				"t3",
				[
					1,
					"wD",
					"/",
					[
						1,
						"accW3",
						"+",
						"wD"
					]
				]
			],
			[
				13,
				"accW4",
				[
					1,
					"accW3",
					"+",
					"wD"
				]
			],
			[
				13,
				"t4",
				[
					1,
					"wE",
					"/",
					[
						1,
						"accW4",
						"+",
						"wE"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					"t1",
					"t2",
					"t3",
					"t4"
				]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"mpgWeight"
		]
	},
	externals: () => ({
		d,
		std,
		mpgWeight
	})
}) && $.f)({}));
const componentDefinition = {
	name: "MultiPointGradient",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Five individually placed color points blended together by proximity — drag each point to shape the gradient",
	acceptsUVContext: true,
	props: {
		colorA: {
			default: "#4776E6",
			transform: transformColor,
			description: "Color of control point A",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		positionA: {
			default: {
				x: .2,
				y: .2
			},
			transform: transformPosition,
			description: "Position of control point A",
			ui: {
				type: "position",
				label: "Position A",
				group: "Position"
			}
		},
		colorB: {
			default: "#C44DFF",
			transform: transformColor,
			description: "Color of control point B",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		positionB: {
			default: {
				x: .8,
				y: .2
			},
			transform: transformPosition,
			description: "Position of control point B",
			ui: {
				type: "position",
				label: "Position B",
				group: "Position"
			}
		},
		colorC: {
			default: "#1ABC9C",
			transform: transformColor,
			description: "Color of control point C",
			ui: {
				type: "color",
				label: "Color C",
				group: "Colors"
			}
		},
		positionC: {
			default: {
				x: .2,
				y: .8
			},
			transform: transformPosition,
			description: "Position of control point C",
			ui: {
				type: "position",
				label: "Position C",
				group: "Position"
			}
		},
		colorD: {
			default: "#F8BBD9",
			transform: transformColor,
			description: "Color of control point D",
			ui: {
				type: "color",
				label: "Color D",
				group: "Colors"
			}
		},
		positionD: {
			default: {
				x: .8,
				y: .8
			},
			transform: transformPosition,
			description: "Position of control point D",
			ui: {
				type: "position",
				label: "Position D",
				group: "Position"
			}
		},
		colorE: {
			default: "#FF8C42",
			transform: transformColor,
			description: "Color of control point E",
			ui: {
				type: "color",
				label: "Color E",
				group: "Colors"
			}
		},
		positionE: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Position of control point E",
			ui: {
				type: "position",
				label: "Position E",
				group: "Position"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		smoothness: {
			default: 2,
			description: "Controls how smoothly colors blend.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .01,
				label: "Smoothness",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const uv = params.uvContext ?? ctx.uv;
		const vp = params.effectiveViewportSize ?? ctx.viewportSize;
		const mode = propValues.colorSpace ?? 0;
		const factors = call(mpgFactors, "mpgFactors", [
			uv,
			vp,
			uniforms.positionA,
			uniforms.positionB,
			uniforms.positionC,
			uniforms.positionD,
			uniforms.positionE,
			uniforms.smoothness
		]);
		const variant = mixColorsVariants[mode] ?? mixColorsLinear;
		let acc = uniforms.colorA;
		acc = call(variant, "mixColors", [
			acc,
			uniforms.colorB,
			factors.member("x")
		]);
		acc = call(variant, "mixColors", [
			acc,
			uniforms.colorC,
			factors.member("y")
		]);
		acc = call(variant, "mixColors", [
			acc,
			uniforms.colorD,
			factors.member("z")
		]);
		acc = call(variant, "mixColors", [
			acc,
			uniforms.colorE,
			factors.member("w")
		]);
		return acc;
	}
};
var MultiPointGradient_default = componentDefinition;
export { componentDefinition as n, mpgFactors as r, MultiPointGradient_default as t };
