// shaders/dist/core/noiseStylize-C2jzZv4d.js
function reliefStylizeProps(defaults) {
	return {
		intensity: {
			default: defaults.intensity,
			description: "Strength of the relief shading applied to the child content",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Intensity",
				group: "Texture"
			}
		},
		scale: {
			default: defaults.scale,
			description: "Scale of the pattern — lower = larger features, higher = finer grain",
			ui: {
				type: "range",
				min: .1,
				max: 8,
				step: .05,
				label: "Scale",
				group: "Texture"
			}
		},
		contrast: {
			default: defaults.contrast,
			description: "Contrast of the texture — negative flattens it into a subtle overlay",
			ui: {
				type: "range",
				min: -1,
				max: 2,
				step: .05,
				label: "Contrast",
				group: "Texture"
			}
		},
		distortion: {
			default: defaults.distortion,
			description: "Surface distortion — warps the child along the texture like carved relief",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Distortion",
				group: "Texture"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for pattern variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Texture"
			}
		}
	};
}
export { reliefStylizeProps as t };
