// shaders/dist/core/PagePeel-CTy_-lpL.js
import { At as call, Ot as ZERO, Pt as floatE, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
var transformCorner = (value) => {
	switch (value) {
		case "top-left": return 0;
		case "top-right": return 1;
		case "bottom-left": return 2;
		case "bottom-right": return 3;
		default: return 3;
	}
};
var CORNERS = {
	0: {
		corner: [0, 0],
		opposite: [1, 1]
	},
	1: {
		corner: [1, 0],
		opposite: [0, 1]
	},
	2: {
		corner: [0, 1],
		opposite: [1, 0]
	},
	3: {
		corner: [1, 1],
		opposite: [0, 0]
	}
};
var HALF_PI = Math.PI / 2;
const PagePeelGeom = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
	curlUV: d.vec2f,
	curlBright: d.f32,
	spec: d.f32,
	flatShade: d.f32,
	peelAlpha: d.f32,
	foldBlend: d.f32,
	showCurl: d.f32
}), "PagePeelGeom");
const pagePeelGeom = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], PagePeelGeom)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, cornerX, cornerY, oppX, oppY, amount, radius, shading, highlight, highlightSoftness, shadow) => {
	"use gpu";
	const pA = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const cornerA = d.vec2f(__tsover_mul(cornerX, aspect), cornerY);
	const axisVec = d.vec2f(__tsover_mul(oppX, aspect), oppY).sub(cornerA);
	const L = std.length(axisVec);
	const axis = axisVec.mul(__tsover_div(1, L));
	const t = std.dot(pA.sub(cornerA), axis);
	const peel = __tsover_mul(amount, L);
	const cFold = peel;
	const R = std.max(__tsover_mul(radius, L), 1e-4);
	const thetaMax = __tsover_div(peel, R);
	const X = __tsover_div(__tsover_sub(t, cFold), R);
	const Xc = std.clamp(X, 0, 1);
	const theta = std.asin(Xc);
	const inBand = __tsover_mul(std.step(0, X), std.step(X, 1));
	const paper = std.step(theta, thetaMax);
	const cCurl = __tsover_sub(peel, __tsover_mul(R, theta));
	const sourceA = pA.add(axis.mul(__tsover_sub(cCurl, t)));
	const curlUV = d.vec2f(__tsover_div(sourceA.x, aspect), sourceA.y);
	const curlInBounds = __tsover_mul(__tsover_mul(__tsover_mul(std.step(0, curlUV.x), std.step(curlUV.x, 1)), std.step(0, curlUV.y)), std.step(curlUV.y, 1));
	const facing = std.cos(theta);
	const curlBright = __tsover_sub(1, __tsover_mul(shading, facing));
	const thetaN = __tsover_div(theta, HALF_PI);
	const specSigma = std.mix(.05, .4, highlightSoftness);
	const specDelta = __tsover_div(__tsover_sub(thetaN, .65), specSigma);
	const spec = __tsover_mul(std.exp(__tsover_mul(__tsover_mul(specDelta, specDelta), -.5)), highlight);
	const liftFrac = __tsover_sub(1, std.cos(std.clamp(thetaMax, 0, HALF_PI)));
	const lipHeight = __tsover_mul(R, liftFrac);
	const shadowStrength = __tsover_mul(std.smoothstep(0, .04, amount), std.pow(__tsover_sub(1, amount), .7));
	const flatReach = std.max(__tsover_add(R, __tsover_mul(lipHeight, .5)), 1e-4);
	const distPastCrease = __tsover_sub(t, cFold);
	const flatFall = __tsover_mul(__tsover_mul(std.pow(__tsover_sub(1, std.smoothstep(0, flatReach, distPastCrease)), 2), std.step(0, distPastCrease)), shadowStrength);
	const flatShade = __tsover_sub(1, __tsover_mul(__tsover_mul(shadow, flatFall), .5));
	const peelReach = std.max(__tsover_mul(lipHeight, 2.5), 1e-4);
	const distIntoPeel = __tsover_sub(cFold, t);
	const peelFall = __tsover_mul(__tsover_mul(__tsover_mul(std.pow(__tsover_sub(1, std.smoothstep(0, peelReach, distIntoPeel)), 1.5), std.step(0, distIntoPeel)), shadowStrength), shadow);
	return PagePeelGeom({
		curlUV,
		curlBright,
		spec,
		flatShade,
		peelAlpha: __tsover_mul(peelFall, .6),
		foldBlend: std.smoothstep(__tsover_sub(cFold, .0015), __tsover_add(cFold, .0015), t),
		showCurl: __tsover_mul(__tsover_mul(inBand, paper), curlInBounds)
	});
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "cornerX"
			},
			{
				"type": "i",
				"name": "cornerY"
			},
			{
				"type": "i",
				"name": "oppX"
			},
			{
				"type": "i",
				"name": "oppY"
			},
			{
				"type": "i",
				"name": "amount"
			},
			{
				"type": "i",
				"name": "radius"
			},
			{
				"type": "i",
				"name": "shading"
			},
			{
				"type": "i",
				"name": "highlight"
			},
			{
				"type": "i",
				"name": "highlightSoftness"
			},
			{
				"type": "i",
				"name": "shadow"
			}
		],
		"body": [0, [
			[
				13,
				"pA",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"cornerA",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						"cornerX",
						"*",
						"aspect"
					], "cornerY"]
				]
			],
			[
				13,
				"oppA",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						"oppX",
						"*",
						"aspect"
					], "oppY"]
				]
			],
			[
				13,
				"axisVec",
				[
					6,
					[
						7,
						"oppA",
						"sub"
					],
					["cornerA"]
				]
			],
			[
				13,
				"L",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["axisVec"]
				]
			],
			[
				13,
				"axis",
				[
					6,
					[
						7,
						"axisVec",
						"mul"
					],
					[[
						1,
						[5, "1"],
						"/",
						"L"
					]]
				]
			],
			[
				13,
				"t",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					[[
						6,
						[
							7,
							"pA",
							"sub"
						],
						["cornerA"]
					], "axis"]
				]
			],
			[
				13,
				"peel",
				[
					1,
					"amount",
					"*",
					"L"
				]
			],
			[
				13,
				"cFold",
				"peel"
			],
			[
				13,
				"R",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"radius",
						"*",
						"L"
					], [5, "0.0001"]]
				]
			],
			[
				13,
				"thetaMax",
				[
					1,
					"peel",
					"/",
					"R"
				]
			],
			[
				13,
				"X",
				[
					1,
					[
						1,
						"t",
						"-",
						"cFold"
					],
					"/",
					"R"
				]
			],
			[
				13,
				"Xc",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"X",
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"theta",
				[
					6,
					[
						7,
						"std",
						"asin"
					],
					["Xc"]
				]
			],
			[
				13,
				"inBand",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"step"
						],
						[[5, "0"], "X"]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"step"
						],
						["X", [5, "1"]]
					]
				]
			],
			[
				13,
				"paper",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					["theta", "thetaMax"]
				]
			],
			[
				13,
				"cCurl",
				[
					1,
					"peel",
					"-",
					[
						1,
						"R",
						"*",
						"theta"
					]
				]
			],
			[
				13,
				"sourceA",
				[
					6,
					[
						7,
						"pA",
						"add"
					],
					[[
						6,
						[
							7,
							"axis",
							"mul"
						],
						[[
							1,
							"cCurl",
							"-",
							"t"
						]]
					]]
				]
			],
			[
				13,
				"curlUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"sourceA",
							"x"
						],
						"/",
						"aspect"
					], [
						7,
						"sourceA",
						"y"
					]]
				]
			],
			[
				13,
				"curlInBounds",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"step"
								],
								[[5, "0"], [
									7,
									"curlUV",
									"x"
								]]
							],
							"*",
							[
								6,
								[
									7,
									"std",
									"step"
								],
								[[
									7,
									"curlUV",
									"x"
								], [5, "1"]]
							]
						],
						"*",
						[
							6,
							[
								7,
								"std",
								"step"
							],
							[[5, "0"], [
								7,
								"curlUV",
								"y"
							]]
						]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"step"
						],
						[[
							7,
							"curlUV",
							"y"
						], [5, "1"]]
					]
				]
			],
			[
				13,
				"facing",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["theta"]
				]
			],
			[
				13,
				"curlBright",
				[
					1,
					[5, "1"],
					"-",
					[
						1,
						"shading",
						"*",
						"facing"
					]
				]
			],
			[
				13,
				"thetaN",
				[
					1,
					"theta",
					"/",
					"HALF_PI"
				]
			],
			[
				13,
				"specSigma",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "0.05"],
						[5, "0.4"],
						"highlightSoftness"
					]
				]
			],
			[
				13,
				"specDelta",
				[
					1,
					[
						1,
						"thetaN",
						"-",
						[5, "0.65"]
					],
					"/",
					"specSigma"
				]
			],
			[
				13,
				"spec",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"exp"
						],
						[[
							1,
							[
								1,
								"specDelta",
								"*",
								"specDelta"
							],
							"*",
							[
								4,
								"-",
								[5, "0.5"]
							]
						]]
					],
					"*",
					"highlight"
				]
			],
			[
				13,
				"liftFrac",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						[[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								"thetaMax",
								[5, "0"],
								"HALF_PI"
							]
						]]
					]
				]
			],
			[
				13,
				"lipHeight",
				[
					1,
					"R",
					"*",
					"liftFrac"
				]
			],
			[
				13,
				"shadowStrength",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							[5, "0.04"],
							"amount"
						]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"pow"
						],
						[[
							1,
							[5, "1"],
							"-",
							"amount"
						], [5, "0.7"]]
					]
				]
			],
			[
				13,
				"flatReach",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"R",
						"+",
						[
							1,
							"lipHeight",
							"*",
							[5, "0.5"]
						]
					], [5, "0.0001"]]
				]
			],
			[
				13,
				"distPastCrease",
				[
					1,
					"t",
					"-",
					"cFold"
				]
			],
			[
				13,
				"flatFall",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							[[
								1,
								[5, "1"],
								"-",
								[
									6,
									[
										7,
										"std",
										"smoothstep"
									],
									[
										[5, "0"],
										"flatReach",
										"distPastCrease"
									]
								]
							], [5, "2"]]
						],
						"*",
						[
							6,
							[
								7,
								"std",
								"step"
							],
							[[5, "0"], "distPastCrease"]
						]
					],
					"*",
					"shadowStrength"
				]
			],
			[
				13,
				"flatShade",
				[
					1,
					[5, "1"],
					"-",
					[
						1,
						[
							1,
							"shadow",
							"*",
							"flatFall"
						],
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"peelReach",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"lipHeight",
						"*",
						[5, "2.5"]
					], [5, "0.0001"]]
				]
			],
			[
				13,
				"distIntoPeel",
				[
					1,
					"cFold",
					"-",
					"t"
				]
			],
			[
				13,
				"peelFall",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"pow"
								],
								[[
									1,
									[5, "1"],
									"-",
									[
										6,
										[
											7,
											"std",
											"smoothstep"
										],
										[
											[5, "0"],
											"peelReach",
											"distIntoPeel"
										]
									]
								], [5, "1.5"]]
							],
							"*",
							[
								6,
								[
									7,
									"std",
									"step"
								],
								[[5, "0"], "distIntoPeel"]
							]
						],
						"*",
						"shadowStrength"
					],
					"*",
					"shadow"
				]
			],
			[
				13,
				"peelAlpha",
				[
					1,
					"peelFall",
					"*",
					[5, "0.6"]
				]
			],
			[
				13,
				"foldBlend",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"cFold",
							"-",
							[5, "0.0015"]
						],
						[
							1,
							"cFold",
							"+",
							[5, "0.0015"]
						],
						"t"
					]
				]
			],
			[
				13,
				"showCurl",
				[
					1,
					[
						1,
						"inBand",
						"*",
						"paper"
					],
					"*",
					"curlInBounds"
				]
			],
			[10, [
				6,
				"PagePeelGeom",
				[[104, {
					"curlUV": "curlUV",
					"curlBright": "curlBright",
					"spec": "spec",
					"flatShade": "flatShade",
					"peelAlpha": "peelAlpha",
					"foldBlend": "foldBlend",
					"showCurl": "showCurl"
				}]]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"HALF_PI",
			"PagePeelGeom"
		]
	},
	externals: () => ({
		d,
		std,
		HALF_PI,
		PagePeelGeom
	})
}) && $.f)({}));
const pagePeelCompose = tgpu.fn([
	d.vec4f,
	d.vec4f,
	PagePeelGeom
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((curlSample, flatSample, g) => {
	"use gpu";
	const litRgb = curlSample.xyz.mul(g.curlBright).add(__tsover_mul(g.spec, curlSample.w));
	const curlColor = d.vec4f(litRgb, curlSample.w);
	const flatColor = d.vec4f(flatSample.xyz.mul(g.flatShade), flatSample.w);
	const peeledColor = d.vec4f(0, 0, 0, g.peelAlpha);
	const base = std.mix(peeledColor, flatColor, d.vec4f(g.foldBlend));
	return std.select(base, curlColor, g.showCurl > .5);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "curlSample"
			},
			{
				"type": "i",
				"name": "flatSample"
			},
			{
				"type": "i",
				"name": "g"
			}
		],
		"body": [0, [
			[
				13,
				"litRgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									7,
									"curlSample",
									"xyz"
								],
								"mul"
							],
							[[
								7,
								"g",
								"curlBright"
							]]
						],
						"add"
					],
					[[
						1,
						[
							7,
							"g",
							"spec"
						],
						"*",
						[
							7,
							"curlSample",
							"w"
						]
					]]
				]
			],
			[
				13,
				"curlColor",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					["litRgb", [
						7,
						"curlSample",
						"w"
					]]
				]
			],
			[
				13,
				"flatColor",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[[
						6,
						[
							7,
							[
								7,
								"flatSample",
								"xyz"
							],
							"mul"
						],
						[[
							7,
							"g",
							"flatShade"
						]]
					], [
						7,
						"flatSample",
						"w"
					]]
				]
			],
			[
				13,
				"peeledColor",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "0"],
						[
							7,
							"g",
							"peelAlpha"
						]
					]
				]
			],
			[
				13,
				"base",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"peeledColor",
						"flatColor",
						[
							6,
							[
								7,
								"d",
								"vec4f"
							],
							[[
								7,
								"g",
								"foldBlend"
							]]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"select"
				],
				[
					"base",
					"curlColor",
					[
						1,
						[
							7,
							"g",
							"showCurl"
						],
						">",
						[5, "0.5"]
					]
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "PagePeel",
	category: "Distortions",
	description: "Curl the content up from a corner like a peeling page",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		corner: {
			default: "bottom-right",
			description: "Which corner the page peels from",
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Top Left",
						value: "top-left"
					},
					{
						label: "Top Right",
						value: "top-right"
					},
					{
						label: "Bottom Left",
						value: "bottom-left"
					},
					{
						label: "Bottom Right",
						value: "bottom-right"
					}
				],
				label: "Corner",
				group: "Effect"
			}
		},
		amount: {
			default: .2,
			description: "How far the peel has progressed across the page (0 = flat, 1 = fully peeled)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Amount",
				group: "Effect"
			}
		},
		radius: {
			default: .2,
			description: "Tightness of the curl (fraction of the page diagonal)",
			ui: {
				type: ["range", "map"],
				min: .02,
				max: .4,
				step: .01,
				label: "Curl Radius",
				group: "Effect"
			}
		},
		shading: {
			default: .55,
			description: "How much the underside of the curl is shaded for depth",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Shading",
				group: "Effect"
			}
		},
		highlight: {
			default: .4,
			description: "Strength of the specular sheen running along the curl",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Highlight",
				group: "Effect"
			}
		},
		highlightSoftness: {
			default: .2,
			description: "Width of the specular highlight band (low = tight gloss, high = broad satin)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Highlight Softness",
				group: "Effect"
			}
		},
		shadow: {
			default: 1,
			description: "Strength of the contact shadow the curl casts on the page",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Shadow",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, propValues, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const geom = CORNERS[transformCorner(propValues.corner ?? "bottom-right")] ?? CORNERS[3];
		const g = call(pagePeelGeom, "pagePeelGeom", [
			ctx.uv,
			ctx.aspect,
			floatE(geom.corner[0]),
			floatE(geom.corner[1]),
			floatE(geom.opposite[0]),
			floatE(geom.opposite[1]),
			uniforms.amount,
			uniforms.radius,
			uniforms.shading,
			uniforms.highlight,
			uniforms.highlightSoftness,
			uniforms.shadow
		]);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [call(pagePeelCompose, "pagePeelCompose", [
			texture.sample(g.member("curlUV")),
			texture.sample(ctx.uv),
			g
		])]);
	}
};
var PagePeel_default = componentDefinition;
export { pagePeelGeom as a, pagePeelCompose as i, PagePeel_default as n, componentDefinition as r, PagePeelGeom as t };
