// shaders/dist/core/Paper-CZT7vlqM.js
import { At as call, It as vec4, K as curl22, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, it as paper12, n as tgpu, on as d, rt as noise_exports } from "./coords-CWOT0Wqh.js";
const paperSurface = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, grainScale, seed, displacement, roughness) => {
	"use gpu";
	const aspectUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const grainFreq = __tsover_mul(grainScale, 6);
	const grainUV = aspectUV.mul(grainFreq).add(seed);
	const disp = curl22(grainUV).mul(displacement).mul(.0025);
	const displacedUV = uv.add(disp);
	const grain = paper12(grainUV, d.i32(6));
	const brightness = std.clamp(__tsover_add(1, __tsover_mul(__tsover_sub(grain, .7), __tsover_mul(roughness, 1.2))), 0, 1.5);
	return d.vec3f(displacedUV.x, displacedUV.y, brightness);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "grainScale"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "displacement"
			},
			{
				"type": "i",
				"name": "roughness"
			}
		],
		"body": [0, [
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"grainFreq",
				[
					1,
					"grainScale",
					"*",
					[5, "6"]
				]
			],
			[
				13,
				"grainUV",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"aspectUV",
								"mul"
							],
							["grainFreq"]
						],
						"add"
					],
					["seed"]
				]
			],
			[
				13,
				"disp",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										"noise",
										"curl22"
									],
									["grainUV"]
								],
								"mul"
							],
							["displacement"]
						],
						"mul"
					],
					[[5, "0.0025"]]
				]
			],
			[
				13,
				"displacedUV",
				[
					6,
					[
						7,
						"uv",
						"add"
					],
					["disp"]
				]
			],
			[
				13,
				"grain",
				[
					6,
					[
						7,
						"noise",
						"paper12"
					],
					["grainUV", [
						6,
						[
							7,
							"d",
							"i32"
						],
						[[5, "6"]]
					]]
				]
			],
			[
				13,
				"brightness",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[5, "1"],
							"+",
							[
								1,
								[
									1,
									"grain",
									"-",
									[5, "0.7"]
								],
								"*",
								[
									1,
									"roughness",
									"*",
									[5, "1.2"]
								]
							]
						],
						[5, "0"],
						[5, "1.5"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[
						7,
						"displacedUV",
						"x"
					],
					[
						7,
						"displacedUV",
						"y"
					],
					"brightness"
				]
			]]
		]],
		"externalNames": [
			"d",
			"noise",
			"std"
		]
	},
	externals: () => ({
		d,
		noise: noise_exports,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Paper",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Applies realistic paper grain and surface roughness to child content",
	requiresRTT: true,
	requiresChild: true,
	props: {
		roughness: {
			default: .3,
			description: "Surface roughness — higher values create more pronounced brightness variation",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Roughness",
				group: "Texture"
			}
		},
		grainScale: {
			default: 1,
			description: "Scale of the paper grain — lower = coarser, higher = finer",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Grain Scale",
				group: "Texture"
			}
		},
		displacement: {
			default: .15,
			description: "Surface micro-roughness — shifts pixels at grain scale like real paper fiber bumps",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Displacement",
				group: "Texture"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for pattern variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Texture"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, convertToTexture }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Paper shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const surface = call(paperSurface, "paperSurface", [
			ctx.uv,
			ctx.aspect,
			uniforms.grainScale,
			uniforms.seed,
			uniforms.displacement,
			uniforms.roughness
		]);
		const straight = call(unpremultiplyAlpha, "unpremultiplyAlpha", [texture.sample(surface.member("xy"))]);
		const brightness = surface.member("z");
		return vec4(straight.member("rgb").mul(brightness), straight.member("a"));
	}
};
var Paper_default = componentDefinition;
export { componentDefinition as n, paperSurface as r, Paper_default as t };
