// shaders/dist/core/Particles-DtxN6LL7.js
import { $ as hash33, B as createGuardedCompute, Ot as ZERO, Qt as mixColorsVariants, R as std, T as sdf_exports, V as createStateBuffer, X as hash11, Zt as mixColorsLinear, d as sdf3d_exports, n as tgpu, on as d, rt as noise_exports } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as shapeEffectBoundingBoxDeclaration } from "./shapeEffectBounds-BlxxsYMN.js";
var { buildAnalyticSdfFn, createSdfDataTexture, SVG_SDF_SIZE } = sdf_exports;
var { createAnalytic3dSdfSetup, createSvg3dSdfSetup, rotateVec3, rotateVecCpu, is3dShapeType, isSvg3dShapeType, isMobileGpuViewport, resolveShapeSubProp, SHAPE3D_DEFAULTS, SVG3D_DEFAULTS, VOLUMETRIC_FIELD_FORMAT } = sdf3d_exports;
var DEG_TO_RAD = .017453292519943295;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
function resolveShapeType(shapeType, shape) {
	if (shapeType) return shapeType;
	if (shape && typeof shape === "object") return shape.type || "circleSDF";
	if (typeof shape === "string") try {
		return JSON.parse(shape).type || "circleSDF";
	} catch {
		return "circleSDF";
	}
	return "circleSDF";
}
function parseShapeConfig(raw) {
	if (raw && typeof raw === "object") return raw;
	if (typeof raw === "string") try {
		return JSON.parse(raw);
	} catch {
		return {};
	}
	return {};
}
var MAX_PARTICLES = 16e3;
var DGRID = 32;
var DCELLS = DGRID * DGRID * DGRID;
var DOMAIN = 1;
var OUT_RES = isMobileGpuViewport() ? 768 : 1024;
var SPLAT_R = 6;
var SPLAT_WIN = 13;
var MAX_SPEED = 3;
var STATE_FORMAT = "rgba16float";
var SimParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
	colA: d.vec4f,
	colB: d.vec4f,
	dt: d.f32,
	time: d.f32,
	spread: d.f32,
	agitation: d.f32,
	dragMul: d.f32,
	gravX: d.f32,
	gravY: d.f32,
	halfDepth: d.f32,
	mouseX: d.f32,
	mouseY: d.f32,
	mouseForce: d.f32,
	mouseRadSq: d.f32,
	saRadius: d.f32,
	saSides: d.f32,
	saRounding: d.f32,
	saInnerRatio: d.f32,
	saRotation: d.f32,
	saHeight: d.f32,
	saOffset: d.f32,
	saAperture: d.f32,
	centerX: d.f32,
	centerYv: d.f32,
	scale: d.f32,
	rotC: d.f32,
	rotS: d.f32,
	aspect: d.f32,
	size: d.f32,
	exposure: d.f32,
	softness: d.f32,
	speedColorK: d.f32,
	omegaX: d.f32,
	omegaY: d.f32,
	omegaZ: d.f32,
	entrain: d.f32,
	gridOffX: d.f32,
	gridOffY: d.f32,
	gridOffZ: d.f32
}), "SimParams");
var simLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	pos: {
		storage: d.arrayOf(d.vec4f, MAX_PARTICLES),
		access: "mutable"
	},
	vel: {
		storage: d.arrayOf(d.vec4f, MAX_PARTICLES),
		access: "mutable"
	},
	dens: {
		storage: d.arrayOf(d.atomic(d.u32), DCELLS),
		access: "mutable"
	},
	accumE: {
		storage: d.arrayOf(d.atomic(d.u32), OUT_RES * OUT_RES),
		access: "mutable"
	},
	accumS: {
		storage: d.arrayOf(d.atomic(d.u32), OUT_RES * OUT_RES),
		access: "mutable"
	},
	params: { uniform: SimParams },
	outTex: { storageTexture: d.textureStorage2d(STATE_FORMAT, "write-only") }
}), "simLayout");
var svgSdfLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({ sdfSource: { texture: d.texture2d(d.f32) } }), "svgSdfLayout");
var densAt = tgpu.fn([
	d.i32,
	d.i32,
	d.i32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy, cz) => {
	"use gpu";
	const g = d.i32(DGRID);
	const x = std.clamp(cx, 0, __tsover_sub(g, 1));
	const y = std.clamp(cy, 0, __tsover_sub(g, 1));
	const z = std.clamp(cz, 0, __tsover_sub(g, 1));
	const idx = d.u32(__tsover_add(__tsover_mul(__tsover_add(__tsover_mul(z, g), y), g), x));
	return __tsover_mul(d.f32(std.atomicLoad(simLayout.$.dens[idx])), __tsover_div(1, 64));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "cx"
			},
			{
				"type": "i",
				"name": "cy"
			},
			{
				"type": "i",
				"name": "cz"
			}
		],
		"body": [0, [
			[
				13,
				"g",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["DGRID"]
				]
			],
			[
				13,
				"x",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"cx",
						[5, "0"],
						[
							1,
							"g",
							"-",
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"y",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"cy",
						[5, "0"],
						[
							1,
							"g",
							"-",
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"z",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"cz",
						[5, "0"],
						[
							1,
							"g",
							"-",
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"idx",
				[
					6,
					[
						7,
						"d",
						"u32"
					],
					[[
						1,
						[
							1,
							[
								1,
								[
									1,
									"z",
									"*",
									"g"
								],
								"+",
								"y"
							],
							"*",
							"g"
						],
						"+",
						"x"
					]]
				]
			],
			[10, [
				1,
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[
						6,
						[
							7,
							"std",
							"atomicLoad"
						],
						[[
							8,
							[
								7,
								[
									7,
									"simLayout",
									"$"
								],
								"dens"
							],
							"idx"
						]]
					]]
				],
				"*",
				[
					1,
					[5, "1"],
					"/",
					[5, "64"]
				]
			]]
		]],
		"externalNames": [
			"d",
			"DGRID",
			"std",
			"simLayout"
		]
	},
	externals: () => ({
		d,
		DGRID,
		std,
		simLayout
	})
}) && $.f)({}));
var densTrilinear = tgpu.fn([d.vec3f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((pos) => {
	"use gpu";
	const prm = simLayout.$.params;
	const g = d.f32(DGRID);
	const c = d.vec3f(__tsover_sub(__tsover_mul(__tsover_div(__tsover_add(__tsover_add(pos.x, prm.gridOffX), DOMAIN), __tsover_mul(2, DOMAIN)), g), .5), __tsover_sub(__tsover_mul(__tsover_div(__tsover_add(__tsover_add(pos.y, prm.gridOffY), DOMAIN), __tsover_mul(2, DOMAIN)), g), .5), __tsover_sub(__tsover_mul(__tsover_div(__tsover_add(__tsover_add(pos.z, prm.gridOffZ), DOMAIN), __tsover_mul(2, DOMAIN)), g), .5));
	const base = std.clamp(std.floor(c), d.vec3f(0, 0, 0), d.vec3f(__tsover_sub(g, 2), __tsover_sub(g, 2), __tsover_sub(g, 2)));
	const f = std.clamp(c.sub(base), d.vec3f(0, 0, 0), d.vec3f(1, 1, 1));
	const bi = d.i32(base.x);
	const bj = d.i32(base.y);
	const bk = d.i32(base.z);
	let acc = d.f32(0);
	for (let dz = 0; dz < 2; dz++) for (let dy = 0; dy < 2; dy++) for (let dx = 0; dx < 2; dx++) {
		const w = __tsover_mul(__tsover_mul(std.mix(__tsover_sub(1, f.x), f.x, d.f32(dx)), std.mix(__tsover_sub(1, f.y), f.y, d.f32(dy))), std.mix(__tsover_sub(1, f.z), f.z, d.f32(dz)));
		acc = __tsover_add(acc, __tsover_mul(densAt(__tsover_add(bi, dx), __tsover_add(bj, dy), __tsover_add(bk, dz)), w));
	}
	return acc;
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "pos"
		}],
		"body": [0, [
			[
				13,
				"prm",
				[
					7,
					[
						7,
						"simLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"g",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					["DGRID"]
				]
			],
			[
				13,
				"c",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"pos",
												"x"
											],
											"+",
											[
												7,
												"prm",
												"gridOffX"
											]
										],
										"+",
										"DOMAIN"
									],
									"/",
									[
										1,
										[5, "2"],
										"*",
										"DOMAIN"
									]
								],
								"*",
								"g"
							],
							"-",
							[5, "0.5"]
						],
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"pos",
												"y"
											],
											"+",
											[
												7,
												"prm",
												"gridOffY"
											]
										],
										"+",
										"DOMAIN"
									],
									"/",
									[
										1,
										[5, "2"],
										"*",
										"DOMAIN"
									]
								],
								"*",
								"g"
							],
							"-",
							[5, "0.5"]
						],
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"pos",
												"z"
											],
											"+",
											[
												7,
												"prm",
												"gridOffZ"
											]
										],
										"+",
										"DOMAIN"
									],
									"/",
									[
										1,
										[5, "2"],
										"*",
										"DOMAIN"
									]
								],
								"*",
								"g"
							],
							"-",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"base",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							["c"]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0"],
								[5, "0"],
								[5, "0"]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									"g",
									"-",
									[5, "2"]
								],
								[
									1,
									"g",
									"-",
									[5, "2"]
								],
								[
									1,
									"g",
									"-",
									[5, "2"]
								]
							]
						]
					]
				]
			],
			[
				13,
				"f",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"c",
								"sub"
							],
							["base"]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0"],
								[5, "0"],
								[5, "0"]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1"],
								[5, "1"],
								[5, "1"]
							]
						]
					]
				]
			],
			[
				13,
				"bi",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					[[
						7,
						"base",
						"x"
					]]
				]
			],
			[
				13,
				"bj",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					[[
						7,
						"base",
						"y"
					]]
				]
			],
			[
				13,
				"bk",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					[[
						7,
						"base",
						"z"
					]]
				]
			],
			[
				12,
				"acc",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0"]]
				]
			],
			[
				14,
				[
					12,
					"dz",
					[5, "0"]
				],
				[
					1,
					"dz",
					"<",
					[5, "2"]
				],
				[
					102,
					"++",
					"dz"
				],
				[0, [[
					14,
					[
						12,
						"dy",
						[5, "0"]
					],
					[
						1,
						"dy",
						"<",
						[5, "2"]
					],
					[
						102,
						"++",
						"dy"
					],
					[0, [[
						14,
						[
							12,
							"dx",
							[5, "0"]
						],
						[
							1,
							"dx",
							"<",
							[5, "2"]
						],
						[
							102,
							"++",
							"dx"
						],
						[0, [[
							13,
							"w",
							[
								1,
								[
									1,
									[
										6,
										[
											7,
											"std",
											"mix"
										],
										[
											[
												1,
												[5, "1"],
												"-",
												[
													7,
													"f",
													"x"
												]
											],
											[
												7,
												"f",
												"x"
											],
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												["dx"]
											]
										]
									],
									"*",
									[
										6,
										[
											7,
											"std",
											"mix"
										],
										[
											[
												1,
												[5, "1"],
												"-",
												[
													7,
													"f",
													"y"
												]
											],
											[
												7,
												"f",
												"y"
											],
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												["dy"]
											]
										]
									]
								],
								"*",
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										[
											1,
											[5, "1"],
											"-",
											[
												7,
												"f",
												"z"
											]
										],
										[
											7,
											"f",
											"z"
										],
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											["dz"]
										]
									]
								]
							]
						], [
							2,
							"acc",
							"=",
							[
								1,
								"acc",
								"+",
								[
									1,
									[
										6,
										"densAt",
										[
											[
												1,
												"bi",
												"+",
												"dx"
											],
											[
												1,
												"bj",
												"+",
												"dy"
											],
											[
												1,
												"bk",
												"+",
												"dz"
											]
										]
									],
									"*",
									"w"
								]
							]
						]]]
					]]]
				]]]
			],
			[10, "acc"]
		]],
		"externalNames": [
			"simLayout",
			"d",
			"DGRID",
			"DOMAIN",
			"std",
			"densAt"
		]
	},
	externals: () => ({
		simLayout,
		d,
		DGRID,
		DOMAIN,
		std,
		densAt
	})
}) && $.f)({}));
var extrudeSdf = tgpu.fn([d.f32, d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((d2, wz) => {
	"use gpu";
	const ox = std.max(d2, 0);
	const oz = std.max(wz, 0);
	return __tsover_add(std.min(std.max(d2, wz), 0), std.sqrt(__tsover_add(__tsover_mul(ox, ox), __tsover_mul(oz, oz))));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "d2"
		}, {
			"type": "i",
			"name": "wz"
		}],
		"body": [0, [
			[
				13,
				"ox",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["d2", [5, "0"]]
				]
			],
			[
				13,
				"oz",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["wz", [5, "0"]]
				]
			],
			[10, [
				1,
				[
					6,
					[
						7,
						"std",
						"min"
					],
					[[
						6,
						[
							7,
							"std",
							"max"
						],
						["d2", "wz"]
					], [5, "0"]]
				],
				"+",
				[
					6,
					[
						7,
						"std",
						"sqrt"
					],
					[[
						1,
						[
							1,
							"ox",
							"*",
							"ox"
						],
						"+",
						[
							1,
							"oz",
							"*",
							"oz"
						]
					]]
				]
			]]
		]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const particleInitKernel = tgpu.fn([d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((i) => {
	"use gpu";
	const fi = d.f32(i);
	const h3 = hash33(d.vec3f(__tsover_add(__tsover_mul(fi, 1.6180339), .71), __tsover_add(__tsover_mul(fi, 2.7182818), 4.17), __tsover_add(__tsover_mul(fi, 3.1415926), 9.03)));
	const dir = std.normalize(h3.sub(d.vec3f(.5, .5, .5)).add(d.vec3f(1e-4, 2e-4, 3e-4)));
	const rad = __tsover_mul(std.pow(hash11(__tsover_add(__tsover_mul(fi, 2.2360679), 1.7)), d.f32(.3333)), .55);
	const seed = hash11(__tsover_add(__tsover_mul(fi, 1.7320508), .37));
	simLayout.$.pos[i] = d.vec4f(__tsover_mul(dir.x, rad), __tsover_mul(dir.y, rad), __tsover_mul(__tsover_mul(dir.z, rad), .8), seed);
	simLayout.$.vel[i] = d.vec4f(0, 0, 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "i"
		}],
		"body": [0, [
			[
				13,
				"fi",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					["i"]
				]
			],
			[
				13,
				"h3",
				[
					6,
					[
						7,
						"noise",
						"hash33"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								[
									1,
									"fi",
									"*",
									[5, "1.6180339"]
								],
								"+",
								[5, "0.71"]
							],
							[
								1,
								[
									1,
									"fi",
									"*",
									[5, "2.7182818"]
								],
								"+",
								[5, "4.17"]
							],
							[
								1,
								[
									1,
									"fi",
									"*",
									[5, "3.1415926"]
								],
								"+",
								[5, "9.03"]
							]
						]
					]]
				]
			],
			[
				13,
				"dir",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"h3",
									"sub"
								],
								[[
									6,
									[
										7,
										"d",
										"vec3f"
									],
									[
										[5, "0.5"],
										[5, "0.5"],
										[5, "0.5"]
									]
								]]
							],
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0.0001"],
								[5, "0.0002"],
								[5, "0.0003"]
							]
						]]
					]]
				]
			],
			[
				13,
				"rad",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"pow"
						],
						[[
							6,
							[
								7,
								"noise",
								"hash11"
							],
							[[
								1,
								[
									1,
									"fi",
									"*",
									[5, "2.2360679"]
								],
								"+",
								[5, "1.7"]
							]]
						], [
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0.3333"]]
						]]
					],
					"*",
					[5, "0.55"]
				]
			],
			[
				13,
				"seed",
				[
					6,
					[
						7,
						"noise",
						"hash11"
					],
					[[
						1,
						[
							1,
							"fi",
							"*",
							[5, "1.7320508"]
						],
						"+",
						[5, "0.37"]
					]]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"simLayout",
							"$"
						],
						"pos"
					],
					"i"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							1,
							[
								7,
								"dir",
								"x"
							],
							"*",
							"rad"
						],
						[
							1,
							[
								7,
								"dir",
								"y"
							],
							"*",
							"rad"
						],
						[
							1,
							[
								1,
								[
									7,
									"dir",
									"z"
								],
								"*",
								"rad"
							],
							"*",
							[5, "0.8"]
						],
						"seed"
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"simLayout",
							"$"
						],
						"vel"
					],
					"i"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"noise",
			"std",
			"simLayout"
		]
	},
	externals: () => ({
		d,
		noise: noise_exports,
		std,
		simLayout
	})
}) && $.f)({})).$name("particlesInit");
const particleClearDensityKernel = tgpu.fn([d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((i) => {
	"use gpu";
	std.atomicStore(simLayout.$.dens[i], d.u32(0));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "i"
		}],
		"body": [0, [[
			6,
			[
				7,
				"std",
				"atomicStore"
			],
			[[
				8,
				[
					7,
					[
						7,
						"simLayout",
						"$"
					],
					"dens"
				],
				"i"
			], [
				6,
				[
					7,
					"d",
					"u32"
				],
				[[5, "0"]]
			]]
		]]],
		"externalNames": [
			"std",
			"simLayout",
			"d"
		]
	},
	externals: () => ({
		std,
		simLayout,
		d
	})
}) && $.f)({})).$name("particlesClearDensity");
const particleDensityKernel = tgpu.fn([d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((i) => {
	"use gpu";
	const prm = simLayout.$.params;
	const pos = simLayout.$.pos[i];
	const g = d.f32(DGRID);
	const c = d.vec3f(__tsover_sub(__tsover_mul(__tsover_div(__tsover_add(__tsover_add(pos.x, prm.gridOffX), DOMAIN), __tsover_mul(2, DOMAIN)), g), .5), __tsover_sub(__tsover_mul(__tsover_div(__tsover_add(__tsover_add(pos.y, prm.gridOffY), DOMAIN), __tsover_mul(2, DOMAIN)), g), .5), __tsover_sub(__tsover_mul(__tsover_div(__tsover_add(__tsover_add(pos.z, prm.gridOffZ), DOMAIN), __tsover_mul(2, DOMAIN)), g), .5));
	const base = std.clamp(std.floor(c), d.vec3f(0, 0, 0), d.vec3f(__tsover_sub(g, 2), __tsover_sub(g, 2), __tsover_sub(g, 2)));
	const f = std.clamp(c.sub(base), d.vec3f(0, 0, 0), d.vec3f(1, 1, 1));
	const bi = d.i32(base.x);
	const bj = d.i32(base.y);
	const bk = d.i32(base.z);
	const gi = d.i32(DGRID);
	for (let dz = 0; dz < 2; dz++) for (let dy = 0; dy < 2; dy++) for (let dx = 0; dx < 2; dx++) {
		const wx = std.mix(__tsover_sub(1, f.x), f.x, d.f32(dx));
		const wy = std.mix(__tsover_sub(1, f.y), f.y, d.f32(dy));
		const wz = std.mix(__tsover_sub(1, f.z), f.z, d.f32(dz));
		const w = __tsover_mul(__tsover_mul(wx, wy), wz);
		const idx = d.u32(__tsover_add(__tsover_mul(__tsover_add(__tsover_mul(__tsover_add(bk, dz), gi), __tsover_add(bj, dy)), gi), __tsover_add(bi, dx)));
		std.atomicAdd(simLayout.$.dens[idx], d.u32(__tsover_add(__tsover_mul(w, 64), .5)));
	}
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "i"
		}],
		"body": [0, [
			[
				13,
				"prm",
				[
					7,
					[
						7,
						"simLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"pos",
				[
					8,
					[
						7,
						[
							7,
							"simLayout",
							"$"
						],
						"pos"
					],
					"i"
				]
			],
			[
				13,
				"g",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					["DGRID"]
				]
			],
			[
				13,
				"c",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"pos",
												"x"
											],
											"+",
											[
												7,
												"prm",
												"gridOffX"
											]
										],
										"+",
										"DOMAIN"
									],
									"/",
									[
										1,
										[5, "2"],
										"*",
										"DOMAIN"
									]
								],
								"*",
								"g"
							],
							"-",
							[5, "0.5"]
						],
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"pos",
												"y"
											],
											"+",
											[
												7,
												"prm",
												"gridOffY"
											]
										],
										"+",
										"DOMAIN"
									],
									"/",
									[
										1,
										[5, "2"],
										"*",
										"DOMAIN"
									]
								],
								"*",
								"g"
							],
							"-",
							[5, "0.5"]
						],
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											[
												7,
												"pos",
												"z"
											],
											"+",
											[
												7,
												"prm",
												"gridOffZ"
											]
										],
										"+",
										"DOMAIN"
									],
									"/",
									[
										1,
										[5, "2"],
										"*",
										"DOMAIN"
									]
								],
								"*",
								"g"
							],
							"-",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"base",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							["c"]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0"],
								[5, "0"],
								[5, "0"]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									"g",
									"-",
									[5, "2"]
								],
								[
									1,
									"g",
									"-",
									[5, "2"]
								],
								[
									1,
									"g",
									"-",
									[5, "2"]
								]
							]
						]
					]
				]
			],
			[
				13,
				"f",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"c",
								"sub"
							],
							["base"]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0"],
								[5, "0"],
								[5, "0"]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1"],
								[5, "1"],
								[5, "1"]
							]
						]
					]
				]
			],
			[
				13,
				"bi",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					[[
						7,
						"base",
						"x"
					]]
				]
			],
			[
				13,
				"bj",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					[[
						7,
						"base",
						"y"
					]]
				]
			],
			[
				13,
				"bk",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					[[
						7,
						"base",
						"z"
					]]
				]
			],
			[
				13,
				"gi",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["DGRID"]
				]
			],
			[
				14,
				[
					12,
					"dz",
					[5, "0"]
				],
				[
					1,
					"dz",
					"<",
					[5, "2"]
				],
				[
					102,
					"++",
					"dz"
				],
				[0, [[
					14,
					[
						12,
						"dy",
						[5, "0"]
					],
					[
						1,
						"dy",
						"<",
						[5, "2"]
					],
					[
						102,
						"++",
						"dy"
					],
					[0, [[
						14,
						[
							12,
							"dx",
							[5, "0"]
						],
						[
							1,
							"dx",
							"<",
							[5, "2"]
						],
						[
							102,
							"++",
							"dx"
						],
						[0, [
							[
								13,
								"wx",
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										[
											1,
											[5, "1"],
											"-",
											[
												7,
												"f",
												"x"
											]
										],
										[
											7,
											"f",
											"x"
										],
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											["dx"]
										]
									]
								]
							],
							[
								13,
								"wy",
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										[
											1,
											[5, "1"],
											"-",
											[
												7,
												"f",
												"y"
											]
										],
										[
											7,
											"f",
											"y"
										],
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											["dy"]
										]
									]
								]
							],
							[
								13,
								"wz",
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										[
											1,
											[5, "1"],
											"-",
											[
												7,
												"f",
												"z"
											]
										],
										[
											7,
											"f",
											"z"
										],
										[
											6,
											[
												7,
												"d",
												"f32"
											],
											["dz"]
										]
									]
								]
							],
							[
								13,
								"w",
								[
									1,
									[
										1,
										"wx",
										"*",
										"wy"
									],
									"*",
									"wz"
								]
							],
							[
								13,
								"idx",
								[
									6,
									[
										7,
										"d",
										"u32"
									],
									[[
										1,
										[
											1,
											[
												1,
												[
													1,
													[
														1,
														"bk",
														"+",
														"dz"
													],
													"*",
													"gi"
												],
												"+",
												[
													1,
													"bj",
													"+",
													"dy"
												]
											],
											"*",
											"gi"
										],
										"+",
										[
											1,
											"bi",
											"+",
											"dx"
										]
									]]
								]
							],
							[
								6,
								[
									7,
									"std",
									"atomicAdd"
								],
								[[
									8,
									[
										7,
										[
											7,
											"simLayout",
											"$"
										],
										"dens"
									],
									"idx"
								], [
									6,
									[
										7,
										"d",
										"u32"
									],
									[[
										1,
										[
											1,
											"w",
											"*",
											[5, "64"]
										],
										"+",
										[5, "0.5"]
									]]
								]]
							]
						]]
					]]]
				]]]
			]
		]],
		"externalNames": [
			"simLayout",
			"d",
			"DGRID",
			"DOMAIN",
			"std"
		]
	},
	externals: () => ({
		simLayout,
		d,
		DGRID,
		DOMAIN,
		std
	})
}) && $.f)({})).$name("particlesDensity");
const particleClearAccumKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((x, y) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(y, d.u32(OUT_RES)), x);
	std.atomicStore(simLayout.$.accumE[idx], d.u32(0));
	std.atomicStore(simLayout.$.accumS[idx], d.u32(0));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "x"
		}, {
			"type": "i",
			"name": "y"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"y",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["OUT_RES"]
						]
					],
					"+",
					"x"
				]
			],
			[
				6,
				[
					7,
					"std",
					"atomicStore"
				],
				[[
					8,
					[
						7,
						[
							7,
							"simLayout",
							"$"
						],
						"accumE"
					],
					"idx"
				], [
					6,
					[
						7,
						"d",
						"u32"
					],
					[[5, "0"]]
				]]
			],
			[
				6,
				[
					7,
					"std",
					"atomicStore"
				],
				[[
					8,
					[
						7,
						[
							7,
							"simLayout",
							"$"
						],
						"accumS"
					],
					"idx"
				], [
					6,
					[
						7,
						"d",
						"u32"
					],
					[[5, "0"]]
				]]
			]
		]],
		"externalNames": [
			"d",
			"OUT_RES",
			"std",
			"simLayout"
		]
	},
	externals: () => ({
		d,
		OUT_RES,
		std,
		simLayout
	})
}) && $.f)({})).$name("particlesClearAccum");
const particleSplatKernel = tgpu.fn([d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((i) => {
	"use gpu";
	const prm = simLayout.$.params;
	const pos = simLayout.$.pos[i];
	const vel = simLayout.$.vel[i];
	const persp = __tsover_div(1, std.clamp(__tsover_sub(1, __tsover_mul(pos.z, .35)), .45, 2));
	const xs = __tsover_mul(pos.x, persp);
	const ys = __tsover_mul(__tsover_mul(pos.y, persp), -1);
	const rx = __tsover_sub(__tsover_mul(xs, prm.rotC), __tsover_mul(ys, prm.rotS));
	const ry = __tsover_add(__tsover_mul(ys, prm.rotC), __tsover_mul(xs, prm.rotS));
	const u = __tsover_add(prm.centerX, __tsover_div(__tsover_mul(rx, prm.scale), prm.aspect));
	const v = __tsover_add(prm.centerYv, __tsover_mul(ry, prm.scale));
	const outF = d.f32(OUT_RES);
	const tx = __tsover_mul(u, outF);
	const ty = __tsover_mul(v, outF);
	const size = std.clamp(__tsover_mul(__tsover_mul(__tsover_mul(prm.size, __tsover_add(.65, __tsover_mul(.7, pos.w))), persp), prm.scale), .6, d.f32(SPLAT_R));
	const bright = __tsover_mul(__tsover_mul(__tsover_add(.7, __tsover_mul(.6, pos.w)), persp), persp);
	const speedNorm = std.clamp(__tsover_mul(vel.w, prm.speedColorK), 0, 1);
	const cxi = d.i32(std.floor(tx));
	const cyi = d.i32(std.floor(ty));
	for (let ky = 0; ky < SPLAT_WIN; ky++) for (let kx = 0; kx < SPLAT_WIN; kx++) {
		const px = __tsover_sub(__tsover_add(cxi, kx), SPLAT_R);
		const py = __tsover_sub(__tsover_add(cyi, ky), SPLAT_R);
		if (px >= 0 && px < OUT_RES && py >= 0 && py < OUT_RES) {
			const ddx = __tsover_sub(__tsover_add(d.f32(px), .5), tx);
			const ddy = __tsover_sub(__tsover_add(d.f32(py), .5), ty);
			const q = __tsover_div(__tsover_add(__tsover_mul(ddx, ddx), __tsover_mul(ddy, ddy)), std.max(__tsover_mul(size, size), .01));
			if (q < 2.6) {
				const nd = std.sqrt(q);
				const wHard = __tsover_sub(1, std.smoothstep(.72, 1, nd));
				const wSoft = std.exp(__tsover_mul(q, -2.5));
				const w = std.mix(wHard, wSoft, prm.softness);
				const e = d.u32(__tsover_mul(__tsover_mul(w, bright), 256));
				if (e > d.u32(0)) {
					const idx = d.u32(__tsover_add(__tsover_mul(py, OUT_RES), px));
					std.atomicAdd(simLayout.$.accumE[idx], e);
					std.atomicAdd(simLayout.$.accumS[idx], d.u32(__tsover_mul(__tsover_mul(__tsover_mul(w, bright), speedNorm), 256)));
				}
			}
		}
	}
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "i"
		}],
		"body": [0, [
			[
				13,
				"prm",
				[
					7,
					[
						7,
						"simLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"pos",
				[
					8,
					[
						7,
						[
							7,
							"simLayout",
							"$"
						],
						"pos"
					],
					"i"
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"simLayout",
							"$"
						],
						"vel"
					],
					"i"
				]
			],
			[
				13,
				"persp",
				[
					1,
					[5, "1"],
					"/",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[5, "1"],
								"-",
								[
									1,
									[
										7,
										"pos",
										"z"
									],
									"*",
									[5, "0.35"]
								]
							],
							[5, "0.45"],
							[5, "2"]
						]
					]
				]
			],
			[
				13,
				"xs",
				[
					1,
					[
						7,
						"pos",
						"x"
					],
					"*",
					"persp"
				]
			],
			[
				13,
				"ys",
				[
					1,
					[
						1,
						[
							7,
							"pos",
							"y"
						],
						"*",
						"persp"
					],
					"*",
					[
						4,
						"-",
						[5, "1"]
					]
				]
			],
			[
				13,
				"rx",
				[
					1,
					[
						1,
						"xs",
						"*",
						[
							7,
							"prm",
							"rotC"
						]
					],
					"-",
					[
						1,
						"ys",
						"*",
						[
							7,
							"prm",
							"rotS"
						]
					]
				]
			],
			[
				13,
				"ry",
				[
					1,
					[
						1,
						"ys",
						"*",
						[
							7,
							"prm",
							"rotC"
						]
					],
					"+",
					[
						1,
						"xs",
						"*",
						[
							7,
							"prm",
							"rotS"
						]
					]
				]
			],
			[
				13,
				"u",
				[
					1,
					[
						7,
						"prm",
						"centerX"
					],
					"+",
					[
						1,
						[
							1,
							"rx",
							"*",
							[
								7,
								"prm",
								"scale"
							]
						],
						"/",
						[
							7,
							"prm",
							"aspect"
						]
					]
				]
			],
			[
				13,
				"v",
				[
					1,
					[
						7,
						"prm",
						"centerYv"
					],
					"+",
					[
						1,
						"ry",
						"*",
						[
							7,
							"prm",
							"scale"
						]
					]
				]
			],
			[
				13,
				"outF",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					["OUT_RES"]
				]
			],
			[
				13,
				"tx",
				[
					1,
					"u",
					"*",
					"outF"
				]
			],
			[
				13,
				"ty",
				[
					1,
					"v",
					"*",
					"outF"
				]
			],
			[
				13,
				"size",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									[
										7,
										"prm",
										"size"
									],
									"*",
									[
										1,
										[5, "0.65"],
										"+",
										[
											1,
											[5, "0.7"],
											"*",
											[
												7,
												"pos",
												"w"
											]
										]
									]
								],
								"*",
								"persp"
							],
							"*",
							[
								7,
								"prm",
								"scale"
							]
						],
						[5, "0.6"],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["SPLAT_R"]
						]
					]
				]
			],
			[
				13,
				"bright",
				[
					1,
					[
						1,
						[
							1,
							[5, "0.7"],
							"+",
							[
								1,
								[5, "0.6"],
								"*",
								[
									7,
									"pos",
									"w"
								]
							]
						],
						"*",
						"persp"
					],
					"*",
					"persp"
				]
			],
			[
				13,
				"speedNorm",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								7,
								"vel",
								"w"
							],
							"*",
							[
								7,
								"prm",
								"speedColorK"
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"cxi",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					[[
						6,
						[
							7,
							"std",
							"floor"
						],
						["tx"]
					]]
				]
			],
			[
				13,
				"cyi",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					[[
						6,
						[
							7,
							"std",
							"floor"
						],
						["ty"]
					]]
				]
			],
			[
				14,
				[
					12,
					"ky",
					[5, "0"]
				],
				[
					1,
					"ky",
					"<",
					"SPLAT_WIN"
				],
				[
					102,
					"++",
					"ky"
				],
				[0, [[
					14,
					[
						12,
						"kx",
						[5, "0"]
					],
					[
						1,
						"kx",
						"<",
						"SPLAT_WIN"
					],
					[
						102,
						"++",
						"kx"
					],
					[0, [
						[
							13,
							"px",
							[
								1,
								[
									1,
									"cxi",
									"+",
									"kx"
								],
								"-",
								"SPLAT_R"
							]
						],
						[
							13,
							"py",
							[
								1,
								[
									1,
									"cyi",
									"+",
									"ky"
								],
								"-",
								"SPLAT_R"
							]
						],
						[
							11,
							[
								3,
								[
									3,
									[
										3,
										[
											1,
											"px",
											">=",
											[5, "0"]
										],
										"&&",
										[
											1,
											"px",
											"<",
											"OUT_RES"
										]
									],
									"&&",
									[
										1,
										"py",
										">=",
										[5, "0"]
									]
								],
								"&&",
								[
									1,
									"py",
									"<",
									"OUT_RES"
								]
							],
							[0, [
								[
									13,
									"ddx",
									[
										1,
										[
											1,
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												["px"]
											],
											"+",
											[5, "0.5"]
										],
										"-",
										"tx"
									]
								],
								[
									13,
									"ddy",
									[
										1,
										[
											1,
											[
												6,
												[
													7,
													"d",
													"f32"
												],
												["py"]
											],
											"+",
											[5, "0.5"]
										],
										"-",
										"ty"
									]
								],
								[
									13,
									"q",
									[
										1,
										[
											1,
											[
												1,
												"ddx",
												"*",
												"ddx"
											],
											"+",
											[
												1,
												"ddy",
												"*",
												"ddy"
											]
										],
										"/",
										[
											6,
											[
												7,
												"std",
												"max"
											],
											[[
												1,
												"size",
												"*",
												"size"
											], [5, "0.01"]]
										]
									]
								],
								[
									11,
									[
										1,
										"q",
										"<",
										[5, "2.6"]
									],
									[0, [
										[
											13,
											"nd",
											[
												6,
												[
													7,
													"std",
													"sqrt"
												],
												["q"]
											]
										],
										[
											13,
											"wHard",
											[
												1,
												[5, "1"],
												"-",
												[
													6,
													[
														7,
														"std",
														"smoothstep"
													],
													[
														[5, "0.72"],
														[5, "1"],
														"nd"
													]
												]
											]
										],
										[
											13,
											"wSoft",
											[
												6,
												[
													7,
													"std",
													"exp"
												],
												[[
													1,
													"q",
													"*",
													[
														4,
														"-",
														[5, "2.5"]
													]
												]]
											]
										],
										[
											13,
											"w",
											[
												6,
												[
													7,
													"std",
													"mix"
												],
												[
													"wHard",
													"wSoft",
													[
														7,
														"prm",
														"softness"
													]
												]
											]
										],
										[
											13,
											"e",
											[
												6,
												[
													7,
													"d",
													"u32"
												],
												[[
													1,
													[
														1,
														"w",
														"*",
														"bright"
													],
													"*",
													[5, "256"]
												]]
											]
										],
										[
											11,
											[
												1,
												"e",
												">",
												[
													6,
													[
														7,
														"d",
														"u32"
													],
													[[5, "0"]]
												]
											],
											[0, [
												[
													13,
													"idx",
													[
														6,
														[
															7,
															"d",
															"u32"
														],
														[[
															1,
															[
																1,
																"py",
																"*",
																"OUT_RES"
															],
															"+",
															"px"
														]]
													]
												],
												[
													6,
													[
														7,
														"std",
														"atomicAdd"
													],
													[[
														8,
														[
															7,
															[
																7,
																"simLayout",
																"$"
															],
															"accumE"
														],
														"idx"
													], "e"]
												],
												[
													6,
													[
														7,
														"std",
														"atomicAdd"
													],
													[[
														8,
														[
															7,
															[
																7,
																"simLayout",
																"$"
															],
															"accumS"
														],
														"idx"
													], [
														6,
														[
															7,
															"d",
															"u32"
														],
														[[
															1,
															[
																1,
																[
																	1,
																	"w",
																	"*",
																	"bright"
																],
																"*",
																"speedNorm"
															],
															"*",
															[5, "256"]
														]]
													]]
												]
											]]
										]
									]]
								]
							]]
						]
					]]
				]]]
			]
		]],
		"externalNames": [
			"simLayout",
			"std",
			"d",
			"OUT_RES",
			"SPLAT_R",
			"SPLAT_WIN"
		]
	},
	externals: () => ({
		simLayout,
		std,
		d,
		OUT_RES,
		SPLAT_R,
		SPLAT_WIN
	})
}) && $.f)({})).$name("particlesSplat");
function makeParticleResolveKernel(colorSpaceMode) {
	const mixFn = mixColorsVariants[colorSpaceMode] ?? mixColorsLinear;
	return tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((x, y) => {
		"use gpu";
		const prm = simLayout.$.params;
		const idx = __tsover_add(__tsover_mul(y, d.u32(OUT_RES)), x);
		const eRaw = d.f32(std.atomicLoad(simLayout.$.accumE[idx]));
		const sRaw = d.f32(std.atomicLoad(simLayout.$.accumS[idx]));
		const energy = __tsover_mul(__tsover_mul(eRaw, __tsover_div(1, 256)), prm.exposure);
		const speedAvg = __tsover_div(sRaw, std.max(eRaw, 1));
		const col = mixFn(prm.colA, prm.colB, std.clamp(speedAvg, 0, 1));
		const alpha = __tsover_sub(1, std.exp(__tsover_mul(energy, -1.6)));
		std.textureStore(simLayout.$.outTex, d.vec2u(x, y), d.vec4f(col.x, col.y, col.z, __tsover_mul(alpha, col.w)));
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [{
				"type": "i",
				"name": "x"
			}, {
				"type": "i",
				"name": "y"
			}],
			"body": [0, [
				[
					13,
					"prm",
					[
						7,
						[
							7,
							"simLayout",
							"$"
						],
						"params"
					]
				],
				[
					13,
					"idx",
					[
						1,
						[
							1,
							"y",
							"*",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								["OUT_RES"]
							]
						],
						"+",
						"x"
					]
				],
				[
					13,
					"eRaw",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[
							6,
							[
								7,
								"std",
								"atomicLoad"
							],
							[[
								8,
								[
									7,
									[
										7,
										"simLayout",
										"$"
									],
									"accumE"
								],
								"idx"
							]]
						]]
					]
				],
				[
					13,
					"sRaw",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[
							6,
							[
								7,
								"std",
								"atomicLoad"
							],
							[[
								8,
								[
									7,
									[
										7,
										"simLayout",
										"$"
									],
									"accumS"
								],
								"idx"
							]]
						]]
					]
				],
				[
					13,
					"energy",
					[
						1,
						[
							1,
							"eRaw",
							"*",
							[
								1,
								[5, "1"],
								"/",
								[5, "256"]
							]
						],
						"*",
						[
							7,
							"prm",
							"exposure"
						]
					]
				],
				[
					13,
					"speedAvg",
					[
						1,
						"sRaw",
						"/",
						[
							6,
							[
								7,
								"std",
								"max"
							],
							["eRaw", [5, "1"]]
						]
					]
				],
				[
					13,
					"col",
					[
						6,
						"mixFn",
						[
							[
								7,
								"prm",
								"colA"
							],
							[
								7,
								"prm",
								"colB"
							],
							[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									"speedAvg",
									[5, "0"],
									[5, "1"]
								]
							]
						]
					]
				],
				[
					13,
					"alpha",
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								"energy",
								"*",
								[
									4,
									"-",
									[5, "1.6"]
								]
							]]
						]
					]
				],
				[
					6,
					[
						7,
						"std",
						"textureStore"
					],
					[
						[
							7,
							[
								7,
								"simLayout",
								"$"
							],
							"outTex"
						],
						[
							6,
							[
								7,
								"d",
								"vec2u"
							],
							["x", "y"]
						],
						[
							6,
							[
								7,
								"d",
								"vec4f"
							],
							[
								[
									7,
									"col",
									"x"
								],
								[
									7,
									"col",
									"y"
								],
								[
									7,
									"col",
									"z"
								],
								[
									1,
									"alpha",
									"*",
									[
										7,
										"col",
										"w"
									]
								]
							]
						]
					]
				]
			]],
			"externalNames": [
				"simLayout",
				"d",
				"OUT_RES",
				"std",
				"mixFn"
			]
		},
		externals: () => ({
			simLayout,
			d,
			OUT_RES,
			std,
			mixFn
		})
	}) && $.f)({})).$name("particlesResolve");
}
function makeParticleUpdateKernel(evalSdf) {
	return tgpu.fn([d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((i) => {
		"use gpu";
		const prm = simLayout.$.params;
		const p4 = simLayout.$.pos[i];
		const v4 = simLayout.$.vel[i];
		let pos = d.vec3f(p4.x, p4.y, p4.z);
		const dt = prm.dt;
		const sd = evalSdf(pos);
		const e = d.f32(.02);
		const gx = __tsover_sub(evalSdf(d.vec3f(__tsover_add(pos.x, e), pos.y, pos.z)), evalSdf(d.vec3f(__tsover_sub(pos.x, e), pos.y, pos.z)));
		const gy = __tsover_sub(evalSdf(d.vec3f(pos.x, __tsover_add(pos.y, e), pos.z)), evalSdf(d.vec3f(pos.x, __tsover_sub(pos.y, e), pos.z)));
		const gz = __tsover_sub(evalSdf(d.vec3f(pos.x, pos.y, __tsover_add(pos.z, e))), evalSdf(d.vec3f(pos.x, pos.y, __tsover_sub(pos.z, e))));
		const gv = d.vec3f(gx, gy, gz);
		const gn = gv.div(std.max(std.length(gv), 1e-4));
		const he = d.f32(__tsover_div(DOMAIN, DGRID));
		const c0 = densTrilinear(pos);
		const dgx = __tsover_sub(densTrilinear(d.vec3f(__tsover_add(pos.x, he), pos.y, pos.z)), densTrilinear(d.vec3f(__tsover_sub(pos.x, he), pos.y, pos.z)));
		const dgy = __tsover_sub(densTrilinear(d.vec3f(pos.x, __tsover_add(pos.y, he), pos.z)), densTrilinear(d.vec3f(pos.x, __tsover_sub(pos.y, he), pos.z)));
		const dgz = __tsover_sub(densTrilinear(d.vec3f(pos.x, pos.y, __tsover_add(pos.z, he))), densTrilinear(d.vec3f(pos.x, pos.y, __tsover_sub(pos.z, he))));
		const pressNorm = __tsover_div(__tsover_mul(prm.spread, 4.4), std.max(__tsover_add(__tsover_mul(c0, 2), 6), 6));
		let F = d.vec3f(__tsover_mul(__tsover_mul(dgx, -1), pressNorm), __tsover_mul(__tsover_mul(dgy, -1), pressNorm), __tsover_mul(__tsover_mul(dgz, -1), pressNorm));
		const out = std.max(sd, 0);
		const wall = __tsover_add(__tsover_mul(std.smoothstep(-.1, .01, sd), 1.1), __tsover_mul(out, 22));
		F = F.sub(gn.mul(wall));
		const rl = std.length(pos);
		const recall = __tsover_mul(std.smoothstep(__tsover_mul(DOMAIN, .85), __tsover_mul(DOMAIN, 1.1), rl), 9);
		F = F.sub(pos.mul(__tsover_div(recall, std.max(rl, 1e-4))));
		const homing = __tsover_mul(std.smoothstep(.12, .45, sd), 5);
		F = F.sub(pos.mul(__tsover_div(homing, std.max(rl, 1e-4))));
		const vPrev = d.vec3f(v4.x, v4.y, v4.z);
		const featureVel = std.cross(d.vec3f(prm.omegaX, prm.omegaY, prm.omegaZ), pos);
		const insideM = __tsover_sub(1, std.smoothstep(-.05, .05, sd));
		F = F.add(featureVel.sub(vPrev).mul(__tsover_mul(prm.entrain, insideM)));
		const dmx = __tsover_sub(pos.x, prm.mouseX);
		const dmy = __tsover_sub(pos.y, prm.mouseY);
		const dl = std.sqrt(__tsover_add(__tsover_mul(dmx, dmx), __tsover_mul(dmy, dmy)));
		const mg = __tsover_mul(std.exp(__tsover_mul(__tsover_div(__tsover_mul(dl, dl), std.max(prm.mouseRadSq, 1e-5)), -1)), prm.mouseForce);
		F = F.add(d.vec3f(__tsover_mul(__tsover_div(dmx, std.max(dl, 1e-4)), mg), __tsover_mul(__tsover_div(dmy, std.max(dl, 1e-4)), mg), 0));
		F = F.add(d.vec3f(prm.gravX, prm.gravY, 0));
		const rnd = hash33(pos.mul(193).add(d.vec3f(__tsover_mul(prm.time, 31.7), __tsover_mul(prm.time, 27.3), __tsover_mul(p4.w, 997)))).sub(d.vec3f(.5, .5, .5));
		F = F.add(rnd.mul(__tsover_mul(prm.agitation, 5)));
		let v3 = vPrev.add(F.mul(dt)).mul(prm.dragMul);
		const spd = std.length(v3);
		v3 = v3.mul(__tsover_div(std.min(spd, d.f32(MAX_SPEED)), std.max(spd, 1e-5)));
		pos = std.clamp(pos.add(v3.mul(dt)), d.vec3f(-1.9, -1.9, -1.9), d.vec3f(1.9, 1.9, 1.9));
		simLayout.$.pos[i] = d.vec4f(pos.x, pos.y, pos.z, p4.w);
		simLayout.$.vel[i] = d.vec4f(v3.x, v3.y, v3.z, std.length(v3));
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [{
				"type": "i",
				"name": "i"
			}],
			"body": [0, [
				[
					13,
					"prm",
					[
						7,
						[
							7,
							"simLayout",
							"$"
						],
						"params"
					]
				],
				[
					13,
					"p4",
					[
						8,
						[
							7,
							[
								7,
								"simLayout",
								"$"
							],
							"pos"
						],
						"i"
					]
				],
				[
					13,
					"v4",
					[
						8,
						[
							7,
							[
								7,
								"simLayout",
								"$"
							],
							"vel"
						],
						"i"
					]
				],
				[
					12,
					"pos",
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								7,
								"p4",
								"x"
							],
							[
								7,
								"p4",
								"y"
							],
							[
								7,
								"p4",
								"z"
							]
						]
					]
				],
				[
					13,
					"dt",
					[
						7,
						"prm",
						"dt"
					]
				],
				[
					13,
					"sd",
					[
						6,
						"evalSdf",
						["pos"]
					]
				],
				[
					13,
					"e",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0.02"]]
					]
				],
				[
					13,
					"gx",
					[
						1,
						[
							6,
							"evalSdf",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										1,
										[
											7,
											"pos",
											"x"
										],
										"+",
										"e"
									],
									[
										7,
										"pos",
										"y"
									],
									[
										7,
										"pos",
										"z"
									]
								]
							]]
						],
						"-",
						[
							6,
							"evalSdf",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										1,
										[
											7,
											"pos",
											"x"
										],
										"-",
										"e"
									],
									[
										7,
										"pos",
										"y"
									],
									[
										7,
										"pos",
										"z"
									]
								]
							]]
						]
					]
				],
				[
					13,
					"gy",
					[
						1,
						[
							6,
							"evalSdf",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										7,
										"pos",
										"x"
									],
									[
										1,
										[
											7,
											"pos",
											"y"
										],
										"+",
										"e"
									],
									[
										7,
										"pos",
										"z"
									]
								]
							]]
						],
						"-",
						[
							6,
							"evalSdf",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										7,
										"pos",
										"x"
									],
									[
										1,
										[
											7,
											"pos",
											"y"
										],
										"-",
										"e"
									],
									[
										7,
										"pos",
										"z"
									]
								]
							]]
						]
					]
				],
				[
					13,
					"gz",
					[
						1,
						[
							6,
							"evalSdf",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										7,
										"pos",
										"x"
									],
									[
										7,
										"pos",
										"y"
									],
									[
										1,
										[
											7,
											"pos",
											"z"
										],
										"+",
										"e"
									]
								]
							]]
						],
						"-",
						[
							6,
							"evalSdf",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										7,
										"pos",
										"x"
									],
									[
										7,
										"pos",
										"y"
									],
									[
										1,
										[
											7,
											"pos",
											"z"
										],
										"-",
										"e"
									]
								]
							]]
						]
					]
				],
				[
					13,
					"gv",
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"gx",
							"gy",
							"gz"
						]
					]
				],
				[
					13,
					"gn",
					[
						6,
						[
							7,
							"gv",
							"div"
						],
						[[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								6,
								[
									7,
									"std",
									"length"
								],
								["gv"]
							], [5, "0.0001"]]
						]]
					]
				],
				[
					13,
					"he",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[
							1,
							"DOMAIN",
							"/",
							"DGRID"
						]]
					]
				],
				[
					13,
					"c0",
					[
						6,
						"densTrilinear",
						["pos"]
					]
				],
				[
					13,
					"dgx",
					[
						1,
						[
							6,
							"densTrilinear",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										1,
										[
											7,
											"pos",
											"x"
										],
										"+",
										"he"
									],
									[
										7,
										"pos",
										"y"
									],
									[
										7,
										"pos",
										"z"
									]
								]
							]]
						],
						"-",
						[
							6,
							"densTrilinear",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										1,
										[
											7,
											"pos",
											"x"
										],
										"-",
										"he"
									],
									[
										7,
										"pos",
										"y"
									],
									[
										7,
										"pos",
										"z"
									]
								]
							]]
						]
					]
				],
				[
					13,
					"dgy",
					[
						1,
						[
							6,
							"densTrilinear",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										7,
										"pos",
										"x"
									],
									[
										1,
										[
											7,
											"pos",
											"y"
										],
										"+",
										"he"
									],
									[
										7,
										"pos",
										"z"
									]
								]
							]]
						],
						"-",
						[
							6,
							"densTrilinear",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										7,
										"pos",
										"x"
									],
									[
										1,
										[
											7,
											"pos",
											"y"
										],
										"-",
										"he"
									],
									[
										7,
										"pos",
										"z"
									]
								]
							]]
						]
					]
				],
				[
					13,
					"dgz",
					[
						1,
						[
							6,
							"densTrilinear",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										7,
										"pos",
										"x"
									],
									[
										7,
										"pos",
										"y"
									],
									[
										1,
										[
											7,
											"pos",
											"z"
										],
										"+",
										"he"
									]
								]
							]]
						],
						"-",
						[
							6,
							"densTrilinear",
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										7,
										"pos",
										"x"
									],
									[
										7,
										"pos",
										"y"
									],
									[
										1,
										[
											7,
											"pos",
											"z"
										],
										"-",
										"he"
									]
								]
							]]
						]
					]
				],
				[
					13,
					"pressNorm",
					[
						1,
						[
							1,
							[
								7,
								"prm",
								"spread"
							],
							"*",
							[5, "4.4"]
						],
						"/",
						[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								1,
								[
									1,
									"c0",
									"*",
									[5, "2"]
								],
								"+",
								[5, "6"]
							], [5, "6"]]
						]
					]
				],
				[
					12,
					"F",
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								[
									1,
									"dgx",
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								],
								"*",
								"pressNorm"
							],
							[
								1,
								[
									1,
									"dgy",
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								],
								"*",
								"pressNorm"
							],
							[
								1,
								[
									1,
									"dgz",
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								],
								"*",
								"pressNorm"
							]
						]
					]
				],
				[
					13,
					"out",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						["sd", [5, "0"]]
					]
				],
				[
					13,
					"wall",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									[
										4,
										"-",
										[5, "0.1"]
									],
									[5, "0.01"],
									"sd"
								]
							],
							"*",
							[5, "1.1"]
						],
						"+",
						[
							1,
							"out",
							"*",
							[5, "22"]
						]
					]
				],
				[
					2,
					"F",
					"=",
					[
						6,
						[
							7,
							"F",
							"sub"
						],
						[[
							6,
							[
								7,
								"gn",
								"mul"
							],
							["wall"]
						]]
					]
				],
				[
					13,
					"rl",
					[
						6,
						[
							7,
							"std",
							"length"
						],
						["pos"]
					]
				],
				[
					13,
					"recall",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[
									1,
									"DOMAIN",
									"*",
									[5, "0.85"]
								],
								[
									1,
									"DOMAIN",
									"*",
									[5, "1.1"]
								],
								"rl"
							]
						],
						"*",
						[5, "9"]
					]
				],
				[
					2,
					"F",
					"=",
					[
						6,
						[
							7,
							"F",
							"sub"
						],
						[[
							6,
							[
								7,
								"pos",
								"mul"
							],
							[[
								1,
								"recall",
								"/",
								[
									6,
									[
										7,
										"std",
										"max"
									],
									["rl", [5, "0.0001"]]
								]
							]]
						]]
					]
				],
				[
					13,
					"homing",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[5, "0.12"],
								[5, "0.45"],
								"sd"
							]
						],
						"*",
						[5, "5"]
					]
				],
				[
					2,
					"F",
					"=",
					[
						6,
						[
							7,
							"F",
							"sub"
						],
						[[
							6,
							[
								7,
								"pos",
								"mul"
							],
							[[
								1,
								"homing",
								"/",
								[
									6,
									[
										7,
										"std",
										"max"
									],
									["rl", [5, "0.0001"]]
								]
							]]
						]]
					]
				],
				[
					13,
					"vPrev",
					[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								7,
								"v4",
								"x"
							],
							[
								7,
								"v4",
								"y"
							],
							[
								7,
								"v4",
								"z"
							]
						]
					]
				],
				[
					13,
					"featureVel",
					[
						6,
						[
							7,
							"std",
							"cross"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									7,
									"prm",
									"omegaX"
								],
								[
									7,
									"prm",
									"omegaY"
								],
								[
									7,
									"prm",
									"omegaZ"
								]
							]
						], "pos"]
					]
				],
				[
					13,
					"insideM",
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[
									4,
									"-",
									[5, "0.05"]
								],
								[5, "0.05"],
								"sd"
							]
						]
					]
				],
				[
					2,
					"F",
					"=",
					[
						6,
						[
							7,
							"F",
							"add"
						],
						[[
							6,
							[
								7,
								[
									6,
									[
										7,
										"featureVel",
										"sub"
									],
									["vPrev"]
								],
								"mul"
							],
							[[
								1,
								[
									7,
									"prm",
									"entrain"
								],
								"*",
								"insideM"
							]]
						]]
					]
				],
				[
					13,
					"dmx",
					[
						1,
						[
							7,
							"pos",
							"x"
						],
						"-",
						[
							7,
							"prm",
							"mouseX"
						]
					]
				],
				[
					13,
					"dmy",
					[
						1,
						[
							7,
							"pos",
							"y"
						],
						"-",
						[
							7,
							"prm",
							"mouseY"
						]
					]
				],
				[
					13,
					"dl",
					[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"dmx",
								"*",
								"dmx"
							],
							"+",
							[
								1,
								"dmy",
								"*",
								"dmy"
							]
						]]
					]
				],
				[
					13,
					"mg",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									1,
									[
										1,
										"dl",
										"*",
										"dl"
									],
									"/",
									[
										6,
										[
											7,
											"std",
											"max"
										],
										[[
											7,
											"prm",
											"mouseRadSq"
										], [5, "0.00001"]]
									]
								],
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]]
						],
						"*",
						[
							7,
							"prm",
							"mouseForce"
						]
					]
				],
				[
					2,
					"F",
					"=",
					[
						6,
						[
							7,
							"F",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									[
										1,
										"dmx",
										"/",
										[
											6,
											[
												7,
												"std",
												"max"
											],
											["dl", [5, "0.0001"]]
										]
									],
									"*",
									"mg"
								],
								[
									1,
									[
										1,
										"dmy",
										"/",
										[
											6,
											[
												7,
												"std",
												"max"
											],
											["dl", [5, "0.0001"]]
										]
									],
									"*",
									"mg"
								],
								[5, "0"]
							]
						]]
					]
				],
				[
					2,
					"F",
					"=",
					[
						6,
						[
							7,
							"F",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									7,
									"prm",
									"gravX"
								],
								[
									7,
									"prm",
									"gravY"
								],
								[5, "0"]
							]
						]]
					]
				],
				[
					13,
					"rnd",
					[
						6,
						[
							7,
							[
								6,
								[
									7,
									"noise",
									"hash33"
								],
								[[
									6,
									[
										7,
										[
											6,
											[
												7,
												"pos",
												"mul"
											],
											[[5, "193"]]
										],
										"add"
									],
									[[
										6,
										[
											7,
											"d",
											"vec3f"
										],
										[
											[
												1,
												[
													7,
													"prm",
													"time"
												],
												"*",
												[5, "31.7"]
											],
											[
												1,
												[
													7,
													"prm",
													"time"
												],
												"*",
												[5, "27.3"]
											],
											[
												1,
												[
													7,
													"p4",
													"w"
												],
												"*",
												[5, "997"]
											]
										]
									]]
								]]
							],
							"sub"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0.5"],
								[5, "0.5"],
								[5, "0.5"]
							]
						]]
					]
				],
				[
					2,
					"F",
					"=",
					[
						6,
						[
							7,
							"F",
							"add"
						],
						[[
							6,
							[
								7,
								"rnd",
								"mul"
							],
							[[
								1,
								[
									7,
									"prm",
									"agitation"
								],
								"*",
								[5, "5"]
							]]
						]]
					]
				],
				[
					12,
					"v3",
					[
						6,
						[
							7,
							[
								6,
								[
									7,
									"vPrev",
									"add"
								],
								[[
									6,
									[
										7,
										"F",
										"mul"
									],
									["dt"]
								]]
							],
							"mul"
						],
						[[
							7,
							"prm",
							"dragMul"
						]]
					]
				],
				[
					13,
					"spd",
					[
						6,
						[
							7,
							"std",
							"length"
						],
						["v3"]
					]
				],
				[
					2,
					"v3",
					"=",
					[
						6,
						[
							7,
							"v3",
							"mul"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"min"
								],
								["spd", [
									6,
									[
										7,
										"d",
										"f32"
									],
									["MAX_SPEED"]
								]]
							],
							"/",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["spd", [5, "0.00001"]]
							]
						]]
					]
				],
				[
					2,
					"pos",
					"=",
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								6,
								[
									7,
									"pos",
									"add"
								],
								[[
									6,
									[
										7,
										"v3",
										"mul"
									],
									["dt"]
								]]
							],
							[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[
										4,
										"-",
										[5, "1.9"]
									],
									[
										4,
										"-",
										[5, "1.9"]
									],
									[
										4,
										"-",
										[5, "1.9"]
									]
								]
							],
							[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[5, "1.9"],
									[5, "1.9"],
									[5, "1.9"]
								]
							]
						]
					]
				],
				[
					2,
					[
						8,
						[
							7,
							[
								7,
								"simLayout",
								"$"
							],
							"pos"
						],
						"i"
					],
					"=",
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							[
								7,
								"pos",
								"x"
							],
							[
								7,
								"pos",
								"y"
							],
							[
								7,
								"pos",
								"z"
							],
							[
								7,
								"p4",
								"w"
							]
						]
					]
				],
				[
					2,
					[
						8,
						[
							7,
							[
								7,
								"simLayout",
								"$"
							],
							"vel"
						],
						"i"
					],
					"=",
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							[
								7,
								"v3",
								"x"
							],
							[
								7,
								"v3",
								"y"
							],
							[
								7,
								"v3",
								"z"
							],
							[
								6,
								[
									7,
									"std",
									"length"
								],
								["v3"]
							]
						]
					]
				]
			]],
			"externalNames": [
				"simLayout",
				"d",
				"evalSdf",
				"std",
				"DOMAIN",
				"DGRID",
				"densTrilinear",
				"noise",
				"MAX_SPEED"
			]
		},
		externals: () => ({
			simLayout,
			d,
			evalSdf,
			std,
			DOMAIN,
			DGRID,
			densTrilinear,
			noise: noise_exports,
			MAX_SPEED
		})
	}) && $.f)({})).$name("particlesUpdate");
}
const componentDefinition = {
	name: "Particles",
	boundingBoxDeclaration: shapeEffectBoundingBoxDeclaration,
	category: "Shape Effects",
	description: "A swarm of simulated particles that settles into the shape, filling it evenly — flat, SVG or true 3D volumes — scattering from the cursor (or chasing it), tumbling when the shape moves, and always drifting back home",
	requiresRTT: false,
	requiresChild: false,
	usesPointer: true,
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		scale: {
			default: 1,
			description: "Scale of the shape (1 = default size)",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Scale",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the shape in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		},
		colorA: {
			default: "#8ec5ff",
			transform: transformColor,
			description: "Color of particles at rest",
			ui: {
				type: "color",
				label: "Rest Color",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ff7ad9",
			transform: transformColor,
			description: "Color particles shift toward when agitated — scattering from the cursor or catching up to a moving shape",
			ui: {
				type: "color",
				label: "Excited Color",
				group: "Colors"
			}
		},
		count: {
			default: 4e3,
			description: "Number of simulated particles",
			ui: {
				type: "range",
				min: 1e3,
				max: MAX_PARTICLES,
				step: 1e3,
				label: "Count",
				group: "Particles"
			}
		},
		size: {
			default: 1.5,
			description: "Base particle size (each particle also varies slightly, and grows as it drifts toward the viewer)",
			ui: {
				type: "range",
				min: .6,
				max: 6,
				step: .1,
				label: "Size",
				group: "Particles"
			}
		},
		spread: {
			default: 1,
			description: "The pressure pushing particles apart — how strongly they insist on even spacing inside the shape",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Spread",
				group: "Physics"
			}
		},
		agitation: {
			default: .12,
			description: "Idle thermal motion of the swarm — 0 freezes it crystal-still once settled",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Agitation",
				group: "Physics"
			}
		},
		damping: {
			default: .4,
			description: "How quickly disturbed particles calm back down and settle into place",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Damping",
				group: "Physics"
			}
		},
		gravity: {
			default: 0,
			description: "Downward pull on the swarm — negative floats the particles upward",
			ui: {
				type: "range",
				min: -2,
				max: 2,
				step: .01,
				label: "Gravity",
				group: "Physics"
			}
		},
		mouseInfluence: {
			default: 2,
			description: "Strength of the cursor field. Positive scatters particles away; negative pulls the swarm toward the cursor",
			ui: {
				type: "range",
				min: -5,
				max: 5,
				step: .01,
				label: "Cursor Force",
				group: "Interaction"
			}
		},
		mouseRadius: {
			default: .2,
			description: "Reach of the cursor field, in shape-local units",
			ui: {
				type: "range",
				min: .05,
				max: .8,
				step: .01,
				label: "Cursor Radius",
				group: "Interaction"
			}
		},
		exposure: {
			default: 1,
			description: "Brightness of the additive particle glow",
			ui: {
				type: "range",
				min: .2,
				max: 3,
				step: .01,
				label: "Exposure",
				group: "Particles"
			}
		},
		softness: {
			default: .1,
			description: "Edge softness of each particle — 0 is a crisp disc, 1 a soft glow puff",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Particles"
			}
		},
		depth: {
			default: .18,
			description: "Thickness of the particle volume behind flat shapes (3D shapes use their own true depth)",
			ui: {
				type: "range",
				min: .02,
				max: .6,
				step: .01,
				label: "Depth",
				group: "Particles"
			}
		},
		colorSpace: {
			default: "oklab",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for the rest→excited color ramp",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type"
		}
	},
	compute: (params) => {
		const { gpu, getCpuValue, registerComputeTexture, onCleanup } = params;
		const root = gpu?.root;
		if (!root) return null;
		const shapeSdfUrl = getCpuValue("shapeSdfUrl") || "";
		const shapeType = resolveShapeType(getCpuValue("shapeType") || "", getCpuValue("shape"));
		const posBuf = createStateBuffer(root, d.vec4f, MAX_PARTICLES);
		const velBuf = createStateBuffer(root, d.vec4f, MAX_PARTICLES);
		const densBuf = createStateBuffer(root, d.atomic(d.u32), DCELLS);
		const accumE = createStateBuffer(root, d.atomic(d.u32), OUT_RES * OUT_RES);
		const accumS = createStateBuffer(root, d.atomic(d.u32), OUT_RES * OUT_RES);
		const outTex = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
			size: [OUT_RES, OUT_RES],
			format: STATE_FORMAT
		}).$usage("storage", "sampled"), "outTex");
		onCleanup(() => outTex.destroy());
		const particleTexture = registerComputeTexture(outTex);
		const paramsU = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(SimParams), "paramsU");
		const simBg = root.createBindGroup(simLayout, {
			pos: posBuf,
			vel: velBuf,
			dens: densBuf,
			accumE,
			accumS,
			params: paramsU.buffer,
			outTex
		});
		let extraBg = null;
		let setup3d = null;
		let evalSdf;
		if (shapeSdfUrl && isSvg3dShapeType(shapeType)) {
			setup3d = createSvg3dSdfSetup(params, shapeSdfUrl, () => getCpuValue("shape"));
			const dummyField = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
				size: [1, 1],
				format: VOLUMETRIC_FIELD_FORMAT
			}).$usage("storage"), "dummyField");
			onCleanup(() => dummyField.destroy());
			extraBg = root.createBindGroup(setup3d.layout, {
				field: dummyField,
				params: setup3d.marchParamsBuffer.buffer,
				sdfSource: setup3d.sdfSourceTexture
			});
			const baked = setup3d.sdfFn;
			const layoutSvg = setup3d.layout;
			evalSdf = tgpu.fn([d.vec3f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p3) => {
				"use gpu";
				const ps = rotateVec3(p3, layoutSvg.$.params.rot);
				const exX = std.max(__tsover_sub(std.abs(ps.x), .5), 0);
				const exY = std.max(__tsover_sub(std.abs(ps.y), .5), 0);
				return __tsover_add(baked(ps), std.sqrt(__tsover_add(__tsover_mul(exX, exX), __tsover_mul(exY, exY))));
			}), {
				v: 1,
				name: void 0,
				ast: {
					"params": [{
						"type": "i",
						"name": "p3"
					}],
					"body": [0, [
						[
							13,
							"ps",
							[
								6,
								"rotateVec3",
								["p3", [
									7,
									[
										7,
										[
											7,
											"layoutSvg",
											"$"
										],
										"params"
									],
									"rot"
								]]
							]
						],
						[
							13,
							"exX",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								[[
									1,
									[
										6,
										[
											7,
											"std",
											"abs"
										],
										[[
											7,
											"ps",
											"x"
										]]
									],
									"-",
									[5, "0.5"]
								], [5, "0"]]
							]
						],
						[
							13,
							"exY",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								[[
									1,
									[
										6,
										[
											7,
											"std",
											"abs"
										],
										[[
											7,
											"ps",
											"y"
										]]
									],
									"-",
									[5, "0.5"]
								], [5, "0"]]
							]
						],
						[10, [
							1,
							[
								6,
								"baked",
								["ps"]
							],
							"+",
							[
								6,
								[
									7,
									"std",
									"sqrt"
								],
								[[
									1,
									[
										1,
										"exX",
										"*",
										"exX"
									],
									"+",
									[
										1,
										"exY",
										"*",
										"exY"
									]
								]]
							]
						]]
					]],
					"externalNames": [
						"rotateVec3",
						"layoutSvg",
						"std",
						"baked"
					]
				},
				externals: () => ({
					rotateVec3,
					layoutSvg,
					std,
					baked
				})
			}) && $.f)({}));
		} else if (shapeSdfUrl) {
			const sdfTex = createSdfDataTexture(params, shapeSdfUrl);
			extraBg = root.createBindGroup(svgSdfLayout, { sdfSource: sdfTex.texture.texture });
			evalSdf = tgpu.fn([d.vec3f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p3) => {
				"use gpu";
				const prm = simLayout.$.params;
				const sizeM1 = __tsover_sub(SVG_SDF_SIZE, 1);
				const tx = std.clamp(d.i32(__tsover_mul(__tsover_add(p3.x, .5), SVG_SDF_SIZE)), 0, sizeM1);
				const ty = std.clamp(d.i32(__tsover_mul(__tsover_sub(.5, p3.y), SVG_SDF_SIZE)), 0, sizeM1);
				const d2 = std.textureLoad(svgSdfLayout.$.sdfSource, d.vec2u(d.u32(tx), d.u32(ty)), 0).x;
				const exX = std.max(__tsover_sub(std.abs(p3.x), .5), 0);
				const exY = std.max(__tsover_sub(std.abs(p3.y), .5), 0);
				return extrudeSdf(__tsover_add(d2, std.sqrt(__tsover_add(__tsover_mul(exX, exX), __tsover_mul(exY, exY)))), __tsover_sub(std.abs(p3.z), prm.halfDepth));
			}), {
				v: 1,
				name: void 0,
				ast: {
					"params": [{
						"type": "i",
						"name": "p3"
					}],
					"body": [0, [
						[
							13,
							"prm",
							[
								7,
								[
									7,
									"simLayout",
									"$"
								],
								"params"
							]
						],
						[
							13,
							"sizeM1",
							[
								1,
								"SVG_SDF_SIZE",
								"-",
								[5, "1"]
							]
						],
						[
							13,
							"tx",
							[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									[
										6,
										[
											7,
											"d",
											"i32"
										],
										[[
											1,
											[
												1,
												[
													7,
													"p3",
													"x"
												],
												"+",
												[5, "0.5"]
											],
											"*",
											"SVG_SDF_SIZE"
										]]
									],
									[5, "0"],
									"sizeM1"
								]
							]
						],
						[
							13,
							"ty",
							[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									[
										6,
										[
											7,
											"d",
											"i32"
										],
										[[
											1,
											[
												1,
												[5, "0.5"],
												"-",
												[
													7,
													"p3",
													"y"
												]
											],
											"*",
											"SVG_SDF_SIZE"
										]]
									],
									[5, "0"],
									"sizeM1"
								]
							]
						],
						[
							13,
							"d2",
							[
								7,
								[
									6,
									[
										7,
										"std",
										"textureLoad"
									],
									[
										[
											7,
											[
												7,
												"svgSdfLayout",
												"$"
											],
											"sdfSource"
										],
										[
											6,
											[
												7,
												"d",
												"vec2u"
											],
											[[
												6,
												[
													7,
													"d",
													"u32"
												],
												["tx"]
											], [
												6,
												[
													7,
													"d",
													"u32"
												],
												["ty"]
											]]
										],
										[5, "0"]
									]
								],
								"x"
							]
						],
						[
							13,
							"exX",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								[[
									1,
									[
										6,
										[
											7,
											"std",
											"abs"
										],
										[[
											7,
											"p3",
											"x"
										]]
									],
									"-",
									[5, "0.5"]
								], [5, "0"]]
							]
						],
						[
							13,
							"exY",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								[[
									1,
									[
										6,
										[
											7,
											"std",
											"abs"
										],
										[[
											7,
											"p3",
											"y"
										]]
									],
									"-",
									[5, "0.5"]
								], [5, "0"]]
							]
						],
						[
							13,
							"d2e",
							[
								1,
								"d2",
								"+",
								[
									6,
									[
										7,
										"std",
										"sqrt"
									],
									[[
										1,
										[
											1,
											"exX",
											"*",
											"exX"
										],
										"+",
										[
											1,
											"exY",
											"*",
											"exY"
										]
									]]
								]
							]
						],
						[10, [
							6,
							"extrudeSdf",
							["d2e", [
								1,
								[
									6,
									[
										7,
										"std",
										"abs"
									],
									[[
										7,
										"p3",
										"z"
									]]
								],
								"-",
								[
									7,
									"prm",
									"halfDepth"
								]
							]]
						]]
					]],
					"externalNames": [
						"simLayout",
						"SVG_SDF_SIZE",
						"std",
						"d",
						"svgSdfLayout",
						"extrudeSdf"
					]
				},
				externals: () => ({
					simLayout,
					SVG_SDF_SIZE,
					std,
					d,
					svgSdfLayout,
					extrudeSdf
				})
			}) && $.f)({}));
		} else if (is3dShapeType(shapeType)) {
			setup3d = createAnalytic3dSdfSetup(root, shapeType, parseShapeConfig(getCpuValue("shape")), () => getCpuValue("shape"));
			const dummyField = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
				size: [1, 1],
				format: VOLUMETRIC_FIELD_FORMAT
			}).$usage("storage"), "dummyField");
			onCleanup(() => dummyField.destroy());
			extraBg = root.createBindGroup(setup3d.layout, {
				field: dummyField,
				params: setup3d.marchParamsBuffer.buffer
			});
			const baked = setup3d.sdfFn;
			const layout3d = setup3d.layout;
			evalSdf = tgpu.fn([d.vec3f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p3) => {
				"use gpu";
				return baked(rotateVec3(p3, layout3d.$.params.rot));
			}), {
				v: 1,
				name: void 0,
				ast: {
					"params": [{
						"type": "i",
						"name": "p3"
					}],
					"body": [0, [[10, [
						6,
						"baked",
						[[
							6,
							"rotateVec3",
							["p3", [
								7,
								[
									7,
									[
										7,
										"layout3d",
										"$"
									],
									"params"
								],
								"rot"
							]]
						]]
					]]]],
					"externalNames": [
						"baked",
						"rotateVec3",
						"layout3d"
					]
				},
				externals: () => ({
					baked,
					rotateVec3,
					layout3d
				})
			}) && $.f)({}));
		} else {
			const sdf2d = buildAnalyticSdfFn(shapeType);
			evalSdf = tgpu.fn([d.vec3f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p3) => {
				"use gpu";
				const prm = simLayout.$.params;
				const d2 = sdf2d(d.vec2f(__tsover_add(p3.x, .5), __tsover_sub(.5, p3.y)), prm.saRadius, prm.saSides, prm.saRounding, prm.saInnerRatio, prm.saRotation, prm.saHeight, prm.saOffset, prm.saAperture).x;
				return extrudeSdf(d2, __tsover_sub(std.abs(p3.z), prm.halfDepth));
			}), {
				v: 1,
				name: void 0,
				ast: {
					"params": [{
						"type": "i",
						"name": "p3"
					}],
					"body": [0, [
						[
							13,
							"prm",
							[
								7,
								[
									7,
									"simLayout",
									"$"
								],
								"params"
							]
						],
						[
							13,
							"uv2",
							[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									1,
									[
										7,
										"p3",
										"x"
									],
									"+",
									[5, "0.5"]
								], [
									1,
									[5, "0.5"],
									"-",
									[
										7,
										"p3",
										"y"
									]
								]]
							]
						],
						[
							13,
							"d2",
							[
								7,
								[
									6,
									"sdf2d",
									[
										"uv2",
										[
											7,
											"prm",
											"saRadius"
										],
										[
											7,
											"prm",
											"saSides"
										],
										[
											7,
											"prm",
											"saRounding"
										],
										[
											7,
											"prm",
											"saInnerRatio"
										],
										[
											7,
											"prm",
											"saRotation"
										],
										[
											7,
											"prm",
											"saHeight"
										],
										[
											7,
											"prm",
											"saOffset"
										],
										[
											7,
											"prm",
											"saAperture"
										]
									]
								],
								"x"
							]
						],
						[10, [
							6,
							"extrudeSdf",
							["d2", [
								1,
								[
									6,
									[
										7,
										"std",
										"abs"
									],
									[[
										7,
										"p3",
										"z"
									]]
								],
								"-",
								[
									7,
									"prm",
									"halfDepth"
								]
							]]
						]]
					]],
					"externalNames": [
						"simLayout",
						"d",
						"sdf2d",
						"extrudeSdf",
						"std"
					]
				},
				externals: () => ({
					simLayout,
					d,
					sdf2d,
					extrudeSdf,
					std
				})
			}) && $.f)({}));
		}
		const updateKernel = makeParticleUpdateKernel(evalSdf);
		const resolveKernel = makeParticleResolveKernel(getCpuValue("colorSpace") ?? 0);
		const mk1 = (kernel) => {
			let pipe = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((i) => {
				"use gpu";
				kernel(i);
			}), {
				v: 1,
				name: void 0,
				ast: {
					"params": [{
						"type": "i",
						"name": "i"
					}],
					"body": [0, [[
						6,
						"kernel",
						["i"]
					]]],
					"externalNames": ["kernel"]
				},
				externals: () => ({ kernel })
			}) && $.f)({})).with(simBg);
			if (extraBg) pipe = pipe.with(extraBg);
			return pipe;
		};
		const init = mk1(particleInitKernel);
		const clearDensity = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((i) => {
			"use gpu";
			particleClearDensityKernel(i);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "i"
				}],
				"body": [0, [[
					6,
					"particleClearDensityKernel",
					["i"]
				]]],
				"externalNames": ["particleClearDensityKernel"]
			},
			externals: () => ({ particleClearDensityKernel })
		}) && $.f)({}), { size: [DCELLS] }).with(simBg);
		const density = mk1(particleDensityKernel);
		const update = mk1(updateKernel);
		const clearAccum = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((x, y) => {
			"use gpu";
			particleClearAccumKernel(x, y);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "x"
				}, {
					"type": "i",
					"name": "y"
				}],
				"body": [0, [[
					6,
					"particleClearAccumKernel",
					["x", "y"]
				]]],
				"externalNames": ["particleClearAccumKernel"]
			},
			externals: () => ({ particleClearAccumKernel })
		}) && $.f)({}), { size: [OUT_RES, OUT_RES] }).with(simBg);
		const splat = mk1(particleSplatKernel);
		const resolve = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((x, y) => {
			"use gpu";
			resolveKernel(x, y);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "x"
				}, {
					"type": "i",
					"name": "y"
				}],
				"body": [0, [[
					6,
					"resolveKernel",
					["x", "y"]
				]]],
				"externalNames": ["resolveKernel"]
			},
			externals: () => ({ resolveKernel })
		}) && $.f)({}), { size: [OUT_RES, OUT_RES] }).with(simBg);
		let elapsed = 0;
		const springs = /* @__PURE__ */ new Map();
		const num = (v, f, frame, k) => resolveShapeSubProp(v, f, frame, k);
		let time = 0;
		let initialized = false;
		let frameIdx = 0;
		let prevRot = null;
		const CELL = 2 * DOMAIN / DGRID;
		const R3A = .8191725133961645;
		const R3B = .6710436067037893;
		const R3C = .5497004779019703;
		return {
			outputs: { particleTexture },
			getComputeNodes: (frameParams) => {
				const fp = frameParams;
				const dt = Math.min(Math.max(fp.deltaTime ?? .016, .001), .033);
				elapsed += dt;
				time += dt;
				const frame = {
					elapsed,
					deltaTime: dt,
					pointerX: fp.pointer?.x ?? .5,
					pointerY: fp.pointer?.y ?? .5,
					springs
				};
				const g = (key, fallback) => {
					const v = getCpuValue(key);
					return typeof v === "number" ? v : fallback;
				};
				const h = fp.dimensions?.height ?? 0;
				const aspect = h > 0 ? (fp.dimensions?.width ?? 0) / h : 1;
				const centerView = getCpuValue("center");
				const centerX = centerView?.x ?? .5;
				const centerYv = 1 - (centerView?.y ?? .5);
				const scale = Math.max(g("scale", 1), .01);
				const rotRad = g("rotation", 0) * DEG_TO_RAD;
				const rotC = Math.cos(rotRad);
				const rotS = Math.sin(rotRad);
				const mu = fp.pointer?.x ?? .5;
				const mv = fp.pointer?.y ?? .5;
				const dxAc = (mu - centerX) * aspect;
				const dyv = mv - centerYv;
				const mouseX = (dxAc * rotC + dyv * rotS) / scale;
				const mouseY = -((dyv * rotC - dxAc * rotS) / scale);
				const gravMag = g("gravity", 0) * 1.2;
				const gravX = rotS * gravMag;
				const gravY = -(rotC * gravMag);
				const cfg = parseShapeConfig(getCpuValue("shape"));
				const isSvg3d = isSvg3dShapeType(shapeType);
				const halfDepth = shapeSdfUrl && isSvg3d ? num(cfg.depth, SVG3D_DEFAULTS.depth ?? .25, frame, "depth") * .5 : Math.max(g("depth", .18), .01) * .5;
				const saRotationV = num(cfg.rotation, 0, frame, "shapeRotation");
				let omX = 0;
				let omY = 0;
				let omZ = 0;
				if (setup3d) {
					const d3 = isSvg3dShapeType(shapeType) ? SVG3D_DEFAULTS : SHAPE3D_DEFAULTS[shapeType] ?? {};
					const rx = num(cfg.rotX, d3.rotX ?? 0, frame, "rotX") * DEG_TO_RAD;
					const ry = num(cfg.rotY, d3.rotY ?? 0, frame, "rotY") * DEG_TO_RAD;
					const rz = num(cfg.rotZ, d3.rotZ ?? 0, frame, "rotZ") * DEG_TO_RAD;
					if (prevRot) {
						const wx = (rx - prevRot.x) / dt;
						const wy = (ry - prevRot.y) / dt;
						const wz = (rz - prevRot.z) / dt;
						const s = {
							cx: Math.cos(rx),
							sx: Math.sin(rx),
							cy: Math.cos(ry),
							sy: Math.sin(ry),
							cz: Math.cos(rz),
							sz: Math.sin(rz)
						};
						const ex = rotateVecCpu(1, 0, 0, s);
						const ey = rotateVecCpu(0, 1, 0, s);
						const ez = rotateVecCpu(0, 0, 1, s);
						omX = -(ex.x * wx + ex.y * wy + ex.z * wz);
						omY = -(ey.x * wx + ey.y * wy + ey.z * wz);
						omZ = -(ez.x * wx + ez.y * wy + ez.z * wz);
					}
					prevRot = {
						x: rx,
						y: ry,
						z: rz
					};
				} else if (!shapeSdfUrl) {
					const r2 = saRotationV * DEG_TO_RAD;
					if (prevRot) omZ = -((r2 - prevRot.x) / dt);
					prevRot = {
						x: r2,
						y: 0,
						z: 0
					};
				}
				const omLen = Math.hypot(omX, omY, omZ);
				const omCap = omLen > 10 ? 10 / omLen : 1;
				const entrain = Math.min(1, omLen * 2.5) * 5;
				const damping = g("damping", .4);
				const mouseRadius = g("mouseRadius", .22);
				const colA = getCpuValue("colorA");
				const colB = getCpuValue("colorB");
				paramsU.write({
					colA: d.vec4f(colA?.x ?? 1, colA?.y ?? 1, colA?.z ?? 1, colA?.w ?? 1),
					colB: d.vec4f(colB?.x ?? 1, colB?.y ?? 1, colB?.z ?? 1, colB?.w ?? 1),
					dt,
					time,
					spread: g("spread", 1),
					agitation: g("agitation", .12),
					dragMul: Math.exp(-(2.2 + damping * 6) * dt),
					gravX,
					gravY,
					halfDepth,
					mouseX,
					mouseY,
					mouseForce: g("mouseInfluence", 1.2) * 3,
					mouseRadSq: mouseRadius * mouseRadius,
					saRadius: num(cfg.radius ?? cfg.width ?? cfg.bottomWidth, .35, frame, "radius"),
					saSides: num(cfg.sides, 6, frame, "sides"),
					saRounding: num(cfg.rounding, 0, frame, "rounding"),
					saInnerRatio: num(cfg.innerRatio ?? cfg.thickness ?? cfg.spread ?? cfg.topWidth ?? cfg.topRatio, .4, frame, "innerRatio"),
					saRotation: saRotationV,
					saHeight: num(cfg.height, .25, frame, "height"),
					saOffset: num(cfg.offset ?? cfg.skew, .2, frame, "offset"),
					saAperture: num(cfg.aperture, 270, frame, "aperture"),
					centerX,
					centerYv,
					scale,
					rotC,
					rotS,
					aspect,
					size: g("size", 2.2),
					exposure: g("exposure", 1),
					softness: Math.min(Math.max(g("softness", .5), 0), 1),
					speedColorK: 2 / MAX_SPEED,
					omegaX: omX * omCap,
					omegaY: omY * omCap,
					omegaZ: omZ * omCap,
					entrain,
					gridOffX: frameIdx * R3A % 1 * CELL,
					gridOffY: frameIdx * R3B % 1 * CELL,
					gridOffZ: frameIdx * R3C % 1 * CELL
				});
				frameIdx++;
				setup3d?.update({
					deltaTime: dt,
					pointer: fp.pointer
				});
				const count = Math.min(Math.max(Math.round(g("count", 4e3)), 16), MAX_PARTICLES);
				const nodes = [];
				if (!initialized) {
					initialized = true;
					nodes.push(() => init.dispatchThreads(MAX_PARTICLES));
				}
				nodes.push(clearDensity, () => density.dispatchThreads(count), () => update.dispatchThreads(count), clearAccum, () => splat.dispatchThreads(count), resolve);
				return nodes;
			}
		};
	},
	fragment: ({ ctx, computeOutputs }) => {
		const particleField = computeOutputs?.particleTexture;
		if (!particleField) return ZERO;
		return particleField.sample(ctx.uv, "linearClamp");
	}
};
var Particles_default = componentDefinition;
export { particleClearAccumKernel as a, particleInitKernel as c, makeParticleUpdateKernel as i, particleSplatKernel as l, componentDefinition as n, particleClearDensityKernel as o, makeParticleResolveKernel as r, particleDensityKernel as s, Particles_default as t };
