// shaders/dist/core/Perspective-CoYBx1wH.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-BlwgBpJf.js";
import { c as transformPosition, s as transformEdges } from "./transformations-fqzznt4g.js";
var DEG_TO_RAD = Math.PI / 180;
const perspectiveUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.vec2f
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, pan, tilt, fov, zoom, offset, uv) => {
	"use gpu";
	const centerPos = d.vec2f(center.x, __tsover_sub(1, center.y));
	const x = __tsover_div(__tsover_sub(uv.x, centerPos.x), zoom);
	const y = __tsover_div(__tsover_sub(uv.y, centerPos.y), zoom);
	const panRad = __tsover_mul(pan, DEG_TO_RAD);
	const tiltRad = __tsover_mul(tilt, DEG_TO_RAD);
	const fovRad = __tsover_mul(fov, DEG_TO_RAD);
	const cosPan = std.cos(panRad);
	const sinPan = std.sin(panRad);
	const cosTilt = std.cos(tiltRad);
	const sinTilt = std.sin(tiltRad);
	const perspectiveFactor = __tsover_mul(std.tan(__tsover_mul(fovRad, .5)), 2);
	const panDivisor = __tsover_add(cosPan, __tsover_mul(__tsover_mul(sinPan, x), perspectiveFactor));
	const safePanDivisor = std.max(panDivisor, .001);
	const afterPanX = __tsover_div(x, safePanDivisor);
	const afterPanY = __tsover_div(__tsover_mul(y, cosPan), safePanDivisor);
	const tiltDivisor = __tsover_add(cosTilt, __tsover_mul(__tsover_mul(sinTilt, afterPanY), perspectiveFactor));
	const safeTiltDivisor = std.max(tiltDivisor, .001);
	const finalX = __tsover_div(__tsover_mul(afterPanX, cosTilt), safeTiltDivisor);
	const finalY = __tsover_div(afterPanY, safeTiltDivisor);
	const offsetPos = d.vec2f(offset.x, __tsover_sub(1, offset.y));
	return d.vec2f(finalX, finalY).add(centerPos).sub(offsetPos).add(d.vec2f(.5, .5));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "pan"
			},
			{
				"type": "i",
				"name": "tilt"
			},
			{
				"type": "i",
				"name": "fov"
			},
			{
				"type": "i",
				"name": "zoom"
			},
			{
				"type": "i",
				"name": "offset"
			},
			{
				"type": "i",
				"name": "uv"
			}
		],
		"body": [0, [
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"center",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"x",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"-",
						[
							7,
							"centerPos",
							"x"
						]
					],
					"/",
					"zoom"
				]
			],
			[
				13,
				"y",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"y"
						],
						"-",
						[
							7,
							"centerPos",
							"y"
						]
					],
					"/",
					"zoom"
				]
			],
			[
				13,
				"panRad",
				[
					1,
					"pan",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"tiltRad",
				[
					1,
					"tilt",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"fovRad",
				[
					1,
					"fov",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosPan",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["panRad"]
				]
			],
			[
				13,
				"sinPan",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["panRad"]
				]
			],
			[
				13,
				"cosTilt",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["tiltRad"]
				]
			],
			[
				13,
				"sinTilt",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["tiltRad"]
				]
			],
			[
				13,
				"perspectiveFactor",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"tan"
						],
						[[
							1,
							"fovRad",
							"*",
							[5, "0.5"]
						]]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"panDivisor",
				[
					1,
					"cosPan",
					"+",
					[
						1,
						[
							1,
							"sinPan",
							"*",
							"x"
						],
						"*",
						"perspectiveFactor"
					]
				]
			],
			[
				13,
				"safePanDivisor",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["panDivisor", [5, "0.001"]]
				]
			],
			[
				13,
				"afterPanX",
				[
					1,
					"x",
					"/",
					"safePanDivisor"
				]
			],
			[
				13,
				"afterPanY",
				[
					1,
					[
						1,
						"y",
						"*",
						"cosPan"
					],
					"/",
					"safePanDivisor"
				]
			],
			[
				13,
				"tiltDivisor",
				[
					1,
					"cosTilt",
					"+",
					[
						1,
						[
							1,
							"sinTilt",
							"*",
							"afterPanY"
						],
						"*",
						"perspectiveFactor"
					]
				]
			],
			[
				13,
				"safeTiltDivisor",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["tiltDivisor", [5, "0.001"]]
				]
			],
			[
				13,
				"finalX",
				[
					1,
					[
						1,
						"afterPanX",
						"*",
						"cosTilt"
					],
					"/",
					"safeTiltDivisor"
				]
			],
			[
				13,
				"finalY",
				[
					1,
					"afterPanY",
					"/",
					"safeTiltDivisor"
				]
			],
			[
				13,
				"offsetPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"offset",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"offset",
							"y"
						]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					[
						6,
						[
							7,
							[
								6,
								[
									7,
									[
										6,
										[
											7,
											"d",
											"vec2f"
										],
										["finalX", "finalY"]
									],
									"add"
								],
								["centerPos"]
							],
							"sub"
						],
						["offsetPos"]
					],
					"add"
				],
				[[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[5, "0.5"], [5, "0.5"]]
				]]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Perspective",
	category: "Distortions",
	description: "Rotate the plane in 3D space with pan and tilt",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of rotation",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		pan: {
			default: 0,
			description: "Horizontal rotation (left/right)",
			ui: {
				type: ["range", "map"],
				min: -90,
				max: 90,
				step: .1,
				label: "Pan",
				group: "Effect"
			}
		},
		tilt: {
			default: 0,
			description: "Vertical rotation (up/down)",
			ui: {
				type: ["range", "map"],
				min: -90,
				max: 90,
				step: .1,
				label: "Tilt",
				group: "Effect"
			}
		},
		fov: {
			default: 60,
			description: "Field of view - controls perspective intensity",
			ui: {
				type: ["range", "map"],
				min: 30,
				max: 120,
				step: 1,
				label: "FOV",
				group: "Effect"
			}
		},
		zoom: {
			default: 1,
			description: "Zoom in to fill the frame after rotation",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 3,
				step: .1,
				label: "Zoom",
				group: "Effect"
			}
		},
		offset: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Shift the result in X/Y",
			ui: {
				type: "position",
				label: "Offset",
				group: "Position"
			}
		},
		edges: {
			default: "transparent",
			description: "How to handle edges",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: ({ childNode, ctx, uniforms, propValues, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(perspectiveUV, "perspectiveUV", [
			uniforms.center,
			uniforms.pan,
			uniforms.tilt,
			uniforms.fov,
			uniforms.zoom,
			uniforms.offset,
			ctx.uv
		]), (uv) => texture.sample(uv), edgeMode)]);
	},
	uvRemap: ({ uv, mask, uniforms, propValues }) => {
		const edgeMode = propValues.edges ?? 0;
		return composeEdgeRemapExpr(call(perspectiveUV, "perspectiveUV", [
			uniforms.center,
			uniforms.pan,
			uniforms.tilt,
			uniforms.fov,
			uniforms.zoom,
			uniforms.offset,
			uv
		]), mask, edgeMode);
	}
};
var Perspective_default = componentDefinition;
export { componentDefinition as n, perspectiveUV as r, Perspective_default as t };
