// shaders/dist/core/Pixelate-L-pZ8G79.js
import { At as call, It as vec4, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
const pixelateSample = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, scale, gap, roundness) => {
	"use gpu";
	const isWide = aspect > 1;
	const pixelCountX = std.max(std.select(__tsover_mul(scale, aspect), scale, isWide), 1);
	const pixelCountY = std.max(std.select(scale, __tsover_div(scale, aspect), isWide), 1);
	const sampleUV = d.vec2f(__tsover_div(std.floor(__tsover_mul(uv.x, pixelCountX)), pixelCountX), __tsover_div(std.floor(__tsover_mul(uv.y, pixelCountY)), pixelCountY));
	const localUV = d.vec2f(__tsover_sub(std.fract(__tsover_mul(uv.x, pixelCountX)), .5), __tsover_sub(std.fract(__tsover_mul(uv.y, pixelCountY)), .5));
	const halfSize = __tsover_sub(.5, __tsover_mul(gap, .5));
	const cornerRadius = __tsover_mul(roundness, halfSize);
	const qx = std.max(__tsover_add(__tsover_sub(std.abs(localUV.x), halfSize), cornerRadius), 0);
	const qy = std.max(__tsover_add(__tsover_sub(std.abs(localUV.y), halfSize), cornerRadius), 0);
	const sdf = __tsover_sub(std.length(d.vec2f(qx, qy)), cornerRadius);
	const mask = std.select(d.f32(0), d.f32(1), sdf <= 0);
	return d.vec3f(sampleUV.x, sampleUV.y, mask);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "gap"
			},
			{
				"type": "i",
				"name": "roundness"
			}
		],
		"body": [0, [
			[
				13,
				"isWide",
				[
					1,
					"aspect",
					">",
					[5, "1"]
				]
			],
			[
				13,
				"pixelCountX",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								1,
								"scale",
								"*",
								"aspect"
							],
							"scale",
							"isWide"
						]
					], [5, "1"]]
				]
			],
			[
				13,
				"pixelCountY",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							"scale",
							[
								1,
								"scale",
								"/",
								"aspect"
							],
							"isWide"
						]
					], [5, "1"]]
				]
			],
			[
				13,
				"sampleUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								1,
								[
									7,
									"uv",
									"x"
								],
								"*",
								"pixelCountX"
							]]
						],
						"/",
						"pixelCountX"
					], [
						1,
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								1,
								[
									7,
									"uv",
									"y"
								],
								"*",
								"pixelCountY"
							]]
						],
						"/",
						"pixelCountY"
					]]
				]
			],
			[
				13,
				"localUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								1,
								[
									7,
									"uv",
									"x"
								],
								"*",
								"pixelCountX"
							]]
						],
						"-",
						[5, "0.5"]
					], [
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								1,
								[
									7,
									"uv",
									"y"
								],
								"*",
								"pixelCountY"
							]]
						],
						"-",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"halfSize",
				[
					1,
					[5, "0.5"],
					"-",
					[
						1,
						"gap",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"cornerRadius",
				[
					1,
					"roundness",
					"*",
					"halfSize"
				]
			],
			[
				13,
				"qx",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									7,
									"localUV",
									"x"
								]]
							],
							"-",
							"halfSize"
						],
						"+",
						"cornerRadius"
					], [5, "0"]]
				]
			],
			[
				13,
				"qy",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									7,
									"localUV",
									"y"
								]]
							],
							"-",
							"halfSize"
						],
						"+",
						"cornerRadius"
					], [5, "0"]]
				]
			],
			[
				13,
				"sdf",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"length"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["qx", "qy"]
						]]
					],
					"-",
					"cornerRadius"
				]
			],
			[
				13,
				"mask",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"sdf",
							"<=",
							[5, "0"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[
						7,
						"sampleUV",
						"x"
					],
					[
						7,
						"sampleUV",
						"y"
					],
					"mask"
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Pixelate",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Pixelation effect with adjustable cell size",
	requiresRTT: true,
	requiresChild: true,
	props: {
		scale: {
			default: 50,
			description: "Number of pixels along the longest edge (higher = smaller pixels)",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 200,
				step: 1,
				label: "Scale",
				group: "Effect"
			}
		},
		gap: {
			default: 0,
			description: "Space between pixels as a fraction of cell size (0 = no gap, 1 = fully invisible)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .95,
				step: .01,
				label: "Gap",
				group: "Effect"
			}
		},
		roundness: {
			default: 0,
			description: "Roundness of each pixel's corners (0 = square, 1 = circle)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Roundness",
				group: "Effect"
			}
		}
	},
	fragment: ({ childNode, ctx, uniforms, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const packed = call(pixelateSample, "pixelateSample", [
			ctx.uv,
			ctx.aspect,
			uniforms.scale,
			uniforms.gap,
			uniforms.roundness
		]);
		const sampled = call(unpremultiplyAlpha, "unpremultiplyAlpha", [texture.sample(packed.member("xy"))]);
		return vec4(sampled.member("rgb"), sampled.member("a").mul(packed.member("z")));
	}
};
var Pixelate_default = componentDefinition;
export { componentDefinition as n, pixelateSample as r, Pixelate_default as t };
