// shaders/dist/core/Plastic-CpmtjSSV.js
import { At as call, It as vec4, Pt as floatE, Qt as mixColorsVariants, R as std, T as sdf_exports, Zt as mixColorsLinear, d as sdf3d_exports, et as mxNoiseFloat2, gt as mixColorStopsRuntime, n as tgpu, on as d, rt as noise_exports } from "./coords-BlwgBpJf.js";
import { n as glass_exports } from "./glass-DdgMBWkF.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig, t as animatedTime } from "./porters-CoaznI7i.js";
import { t as shapeEffectBoundingBoxDeclaration } from "./shapeEffectBounds-BlxxsYMN.js";
var { sdfSpaceUV, offsetUV } = glass_exports;
var { createSvgSdfSampler, createAnalyticSdfSampler, driveAnalyticSubProps, ANALYTIC_SDF_EXTRA_FIELDS } = sdf_exports;
var { createVolumetricFieldComputeNode, buildVolumetricFieldSampler, VOLUMETRIC_FIELD_EXTRA_FIELDS, is3dShapeType, isVolumetricShapeType } = sdf3d_exports;
var DEG_TO_RAD = .017453292519943295;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
function resolveShapeType(shapeType, shape) {
	if (shapeType) return shapeType;
	if (shape && typeof shape === "object") return shape.type || "circleSDF";
	if (typeof shape === "string") try {
		return JSON.parse(shape).type || "circleSDF";
	} catch {
		return "circleSDF";
	}
	return "circleSDF";
}
const plasticEnvBox = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((x, lo, hi, soft) => {
	"use gpu";
	const a = std.smoothstep(__tsover_sub(lo, soft), __tsover_add(lo, soft), x);
	const b = __tsover_sub(1, std.smoothstep(__tsover_sub(hi, soft), __tsover_add(hi, soft), x));
	return __tsover_mul(a, b);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "x"
			},
			{
				"type": "i",
				"name": "lo"
			},
			{
				"type": "i",
				"name": "hi"
			},
			{
				"type": "i",
				"name": "soft"
			}
		],
		"body": [0, [
			[
				13,
				"a",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"lo",
							"-",
							"soft"
						],
						[
							1,
							"lo",
							"+",
							"soft"
						],
						"x"
					]
				]
			],
			[
				13,
				"b",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[
								1,
								"hi",
								"-",
								"soft"
							],
							[
								1,
								"hi",
								"+",
								"soft"
							],
							"x"
						]
					]
				]
			],
			[10, [
				1,
				"a",
				"*",
				"b"
			]]
		]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const plasticCrumple = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((px, py, cScale, seedOff) => {
	"use gpu";
	const p1 = d.vec2f(__tsover_add(__tsover_mul(px, cScale), seedOff), __tsover_add(__tsover_mul(py, cScale), __tsover_mul(seedOff, .617)));
	const p2 = d.vec2f(__tsover_add(__tsover_add(__tsover_mul(__tsover_mul(px, cScale), 2.3), 13.1), seedOff), __tsover_add(__tsover_add(__tsover_mul(__tsover_mul(py, cScale), 2.3), 7.7), __tsover_mul(seedOff, .617)));
	return __tsover_add(std.abs(mxNoiseFloat2(p1)), __tsover_mul(std.abs(mxNoiseFloat2(p2)), .45));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "px"
			},
			{
				"type": "i",
				"name": "py"
			},
			{
				"type": "i",
				"name": "cScale"
			},
			{
				"type": "i",
				"name": "seedOff"
			}
		],
		"body": [0, [
			[
				13,
				"p1",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							"px",
							"*",
							"cScale"
						],
						"+",
						"seedOff"
					], [
						1,
						[
							1,
							"py",
							"*",
							"cScale"
						],
						"+",
						[
							1,
							"seedOff",
							"*",
							[5, "0.617"]
						]
					]]
				]
			],
			[
				13,
				"p2",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								1,
								[
									1,
									"px",
									"*",
									"cScale"
								],
								"*",
								[5, "2.3"]
							],
							"+",
							[5, "13.1"]
						],
						"+",
						"seedOff"
					], [
						1,
						[
							1,
							[
								1,
								[
									1,
									"py",
									"*",
									"cScale"
								],
								"*",
								[5, "2.3"]
							],
							"+",
							[5, "7.7"]
						],
						"+",
						[
							1,
							"seedOff",
							"*",
							[5, "0.617"]
						]
					]]
				]
			],
			[10, [
				1,
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						6,
						[
							7,
							"noise",
							"mxNoiseFloat2"
						],
						["p1"]
					]]
				],
				"+",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							6,
							[
								7,
								"noise",
								"mxNoiseFloat2"
							],
							["p2"]
						]]
					],
					"*",
					[5, "0.45"]
				]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"noise"
		]
	},
	externals: () => ({
		d,
		std,
		noise: noise_exports
	})
}) && $.f)({}));
const plasticGradientT = tgpu.fn([d.vec2f, d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((sdfUV, gradientAngle) => {
	"use gpu";
	const gr = __tsover_mul(gradientAngle, DEG_TO_RAD);
	const gx = std.cos(gr);
	const gy = std.sin(gr);
	const gdx = __tsover_sub(sdfUV.x, .5);
	const gdy = __tsover_sub(sdfUV.y, .5);
	return std.smoothstep(0, 1, std.clamp(__tsover_add(__tsover_div(__tsover_add(__tsover_mul(gdx, gx), __tsover_mul(gdy, gy)), .7), .5), 0, 1));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "sdfUV"
		}, {
			"type": "i",
			"name": "gradientAngle"
		}],
		"body": [0, [
			[
				13,
				"gr",
				[
					1,
					"gradientAngle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"gx",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["gr"]
				]
			],
			[
				13,
				"gy",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["gr"]
				]
			],
			[
				13,
				"gdx",
				[
					1,
					[
						7,
						"sdfUV",
						"x"
					],
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"gdy",
				[
					1,
					[
						7,
						"sdfUV",
						"y"
					],
					"-",
					[5, "0.5"]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"smoothstep"
				],
				[
					[5, "0"],
					[5, "1"],
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											"gdx",
											"*",
											"gx"
										],
										"+",
										[
											1,
											"gdy",
											"*",
											"gy"
										]
									],
									"/",
									[5, "0.7"]
								],
								"+",
								[5, "0.5"]
							],
							[5, "0"],
							[5, "1"]
						]
					]
				]
			]]
		]],
		"externalNames": ["DEG_TO_RAD", "std"]
	},
	externals: () => ({
		DEG_TO_RAD,
		std
	})
}) && $.f)({}));
const plasticComposite = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec2f,
	d.vec2f,
	d.vec3f,
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((surf0, surfX, surfY, sdfUV, viewportSize, baseRgb, rimColor, p0, p1, p2, p3) => {
	"use gpu";
	const scale = p0.x;
	const edgeSoftness = p0.y;
	const thickness = p0.z;
	const roughness = p0.w;
	const reflectivity = p1.x;
	const crumple = p1.y;
	const crumpleScale = p1.z;
	const crumpleCoverage = p1.w;
	const seed = p2.x;
	const lightAngle = p2.y;
	const shading = p2.z;
	const rimP = p2.w;
	const animTime = p3.x;
	const volumetric = p3.y;
	const sharp = std.max(__tsover_mul(edgeSoftness, .5), .001);
	const pxH = __tsover_div(1, viewportSize.y);
	const sdfRaw = surf0.x;
	const sdf = __tsover_div(sdfRaw, scale);
	if (sdf > __tsover_mul(pxH, 2)) return d.vec4f(0, 0, 0, 0);
	const lr = __tsover_mul(__tsover_add(lightAngle, __tsover_mul(animTime, 6)), DEG_TO_RAD);
	const lx = std.cos(lr);
	const ly = std.sin(lr);
	const EPS = .01;
	const gradX = __tsover_div(__tsover_sub(surfX.x, sdfRaw), EPS);
	const gradY = __tsover_div(__tsover_sub(surfY.x, sdfRaw), EPS);
	const rb1 = std.clamp(__tsover_mul(__tsover_div(__tsover_mul(sdf, -1), sharp), 32), 0, 1);
	const pdx = std.select(__tsover_sub(sdfUV.x, .5), surf0.y, volumetric > .5);
	const pdy = std.select(__tsover_sub(sdfUV.y, .5), surf0.z, volumetric > .5);
	const thicknessRange = std.max(__tsover_mul(thickness, .3), .01);
	const depthNorm = std.clamp(__tsover_div(__tsover_mul(sdf, -1), thicknessRange), 0, 1);
	const ddx = std.clamp(__tsover_div(__tsover_sub(surfX.w, surf0.w), EPS), -4, 4);
	const ddy = std.clamp(__tsover_div(__tsover_sub(surfY.w, surf0.w), EPS), -4, 4);
	const nVol = std.normalize(d.vec3f(ddx, ddy, -1));
	const bevelSlope = __tsover_mul(std.sqrt(std.max(__tsover_sub(1, __tsover_mul(depthNorm, depthNorm)), d.f32(0))), 2.4);
	const nFlat = std.normalize(d.vec3f(__tsover_mul(gradX, bevelSlope), __tsover_mul(gradY, bevelSlope), -1));
	const nBase = std.select(nFlat, nVol, volumetric > .5);
	const cScale = __tsover_mul(crumpleScale, 4);
	const cAmp = __tsover_mul(crumple, .9);
	const seedOff = __tsover_mul(seed, 7.31);
	const neps = .012;
	const cr0 = plasticCrumple(pdx, pdy, cScale, seedOff);
	const crx = __tsover_div(__tsover_sub(plasticCrumple(__tsover_add(pdx, neps), pdy, cScale, seedOff), cr0), neps);
	const cry = __tsover_div(__tsover_sub(plasticCrumple(pdx, __tsover_add(pdy, neps), cScale, seedOff), cr0), neps);
	const coverage = std.clamp(crumpleCoverage, 0, 1);
	const covNoise = __tsover_add(__tsover_mul(mxNoiseFloat2(d.vec2f(__tsover_add(__tsover_add(__tsover_mul(pdx, 2.6), seedOff), 31.7), __tsover_add(__tsover_add(__tsover_mul(pdy, 2.6), __tsover_mul(seedOff, .617)), 17.3))), .5), .5);
	const covThreshold = __tsover_sub(1.25, __tsover_mul(coverage, 1.75));
	const covMask = std.smoothstep(__tsover_sub(covThreshold, .25), __tsover_add(covThreshold, .25), covNoise);
	const crumpleW = __tsover_mul(__tsover_mul(cAmp, covMask), __tsover_add(__tsover_mul(std.smoothstep(0, .35, depthNorm), .85), .15));
	const n = std.normalize(nBase.add(d.vec3f(__tsover_mul(__tsover_mul(crx, crumpleW), .25), __tsover_mul(__tsover_mul(cry, crumpleW), .25), 0)));
	const cosView = std.clamp(__tsover_mul(n.z, -1), 0, 1);
	const grazing = __tsover_sub(1, cosView);
	const R = std.reflect(d.vec3f(0, 0, 1), n);
	const rough = std.clamp(roughness, 0, 1);
	const grain = mxNoiseFloat2(d.vec2f(__tsover_mul(pdx, 140), __tsover_mul(pdy, 140)));
	const rJit = __tsover_mul(__tsover_mul(rough, grain), .18);
	const eu = __tsover_add(__tsover_add(__tsover_mul(R.x, lx), __tsover_mul(R.y, ly)), rJit);
	const ev = __tsover_add(__tsover_add(__tsover_mul(R.x, __tsover_mul(ly, -1)), __tsover_mul(R.y, lx)), __tsover_mul(rJit, .6));
	const soft = std.mix(.02, .45, rough);
	const windowBox = __tsover_mul(plasticEnvBox(eu, .18, .95, soft), plasticEnvBox(ev, -.42, .42, soft));
	const mullion = __tsover_mul(__tsover_mul(plasticEnvBox(ev, -.045, .045, __tsover_mul(soft, .6)), .8), __tsover_sub(1, rough));
	const keyLight = __tsover_mul(windowBox, __tsover_sub(1, mullion));
	const strip = __tsover_mul(__tsover_mul(plasticEnvBox(eu, -.16, -.05, __tsover_mul(soft, .7)), plasticEnvBox(ev, -.85, .85, soft)), .85);
	const fill = __tsover_mul(__tsover_mul(plasticEnvBox(eu, -.95, -.45, __tsover_mul(soft, 1.6)), plasticEnvBox(ev, -.6, .6, __tsover_mul(soft, 1.6))), .3);
	const dome = __tsover_mul(std.clamp(__tsover_add(__tsover_mul(eu, .5), .5), 0, 1), .12);
	const env = __tsover_mul(__tsover_add(__tsover_add(__tsover_add(keyLight, strip), fill), dome), __tsover_sub(1, __tsover_mul(rough, .55)));
	const fresnel = __tsover_add(.3, __tsover_mul(std.pow(grazing, d.f32(3)), .7));
	const specW = __tsover_mul(reflectivity, fresnel);
	const L = std.normalize(d.vec3f(lx, ly, -.8));
	const H = std.normalize(L.add(d.vec3f(0, 0, -1)));
	const ndh = std.clamp(std.dot(n, H), 0, 1);
	const shininess = std.exp2(std.mix(d.f32(9), d.f32(4), rough));
	const glint = __tsover_mul(__tsover_mul(std.pow(ndh, shininess), 1.6), __tsover_sub(1, __tsover_mul(rough, .7)));
	const specular = __tsover_mul(__tsover_add(env, glint), specW);
	const ndl = std.clamp(std.dot(n, L), -1, 1);
	const lit = __tsover_add(__tsover_mul(ndl, .5), .5);
	const diffuse = std.mix(__tsover_sub(1, __tsover_mul(shading, .75)), __tsover_add(1, __tsover_mul(shading, .45)), lit);
	const ao = std.mix(.6, 1, std.smoothstep(0, 1, std.clamp(__tsover_div(__tsover_mul(sdf, -1), .025), 0, 1)));
	const rimGlow = __tsover_mul(__tsover_mul(__tsover_mul(std.pow(grazing, d.f32(3.5)), rimP), __tsover_add(__tsover_mul(lit, .6), .4)), 1.6);
	const plasticRgb = baseRgb.mul(diffuse).mul(ao).add(d.vec3f(specular, specular, specular)).add(rimColor.mul(rimGlow));
	const transition = std.smoothstep(0, 1, rb1);
	return d.vec4f(plasticRgb.x, plasticRgb.y, plasticRgb.z, transition);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "surf0"
			},
			{
				"type": "i",
				"name": "surfX"
			},
			{
				"type": "i",
				"name": "surfY"
			},
			{
				"type": "i",
				"name": "sdfUV"
			},
			{
				"type": "i",
				"name": "viewportSize"
			},
			{
				"type": "i",
				"name": "baseRgb"
			},
			{
				"type": "i",
				"name": "rimColor"
			},
			{
				"type": "i",
				"name": "p0"
			},
			{
				"type": "i",
				"name": "p1"
			},
			{
				"type": "i",
				"name": "p2"
			},
			{
				"type": "i",
				"name": "p3"
			}
		],
		"body": [0, [
			[
				13,
				"scale",
				[
					7,
					"p0",
					"x"
				]
			],
			[
				13,
				"edgeSoftness",
				[
					7,
					"p0",
					"y"
				]
			],
			[
				13,
				"thickness",
				[
					7,
					"p0",
					"z"
				]
			],
			[
				13,
				"roughness",
				[
					7,
					"p0",
					"w"
				]
			],
			[
				13,
				"reflectivity",
				[
					7,
					"p1",
					"x"
				]
			],
			[
				13,
				"crumple",
				[
					7,
					"p1",
					"y"
				]
			],
			[
				13,
				"crumpleScale",
				[
					7,
					"p1",
					"z"
				]
			],
			[
				13,
				"crumpleCoverage",
				[
					7,
					"p1",
					"w"
				]
			],
			[
				13,
				"seed",
				[
					7,
					"p2",
					"x"
				]
			],
			[
				13,
				"lightAngle",
				[
					7,
					"p2",
					"y"
				]
			],
			[
				13,
				"shading",
				[
					7,
					"p2",
					"z"
				]
			],
			[
				13,
				"rimP",
				[
					7,
					"p2",
					"w"
				]
			],
			[
				13,
				"animTime",
				[
					7,
					"p3",
					"x"
				]
			],
			[
				13,
				"volumetric",
				[
					7,
					"p3",
					"y"
				]
			],
			[
				13,
				"sharp",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"edgeSoftness",
						"*",
						[5, "0.5"]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"pxH",
				[
					1,
					[5, "1"],
					"/",
					[
						7,
						"viewportSize",
						"y"
					]
				]
			],
			[
				13,
				"sdfRaw",
				[
					7,
					"surf0",
					"x"
				]
			],
			[
				13,
				"sdf",
				[
					1,
					"sdfRaw",
					"/",
					"scale"
				]
			],
			[
				11,
				[
					1,
					"sdf",
					">",
					[
						1,
						"pxH",
						"*",
						[5, "2"]
					]
				],
				[10, [
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "0"],
						[5, "0"]
					]
				]]
			],
			[
				13,
				"lr",
				[
					1,
					[
						1,
						"lightAngle",
						"+",
						[
							1,
							"animTime",
							"*",
							[5, "6"]
						]
					],
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"lx",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["lr"]
				]
			],
			[
				13,
				"ly",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["lr"]
				]
			],
			[
				13,
				"EPS",
				[5, "0.01"]
			],
			[
				13,
				"gradX",
				[
					1,
					[
						1,
						[
							7,
							"surfX",
							"x"
						],
						"-",
						"sdfRaw"
					],
					"/",
					"EPS"
				]
			],
			[
				13,
				"gradY",
				[
					1,
					[
						1,
						[
							7,
							"surfY",
							"x"
						],
						"-",
						"sdfRaw"
					],
					"/",
					"EPS"
				]
			],
			[
				13,
				"rb1",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									"sdf",
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								],
								"/",
								"sharp"
							],
							"*",
							[5, "32"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"pdx",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							1,
							[
								7,
								"sdfUV",
								"x"
							],
							"-",
							[5, "0.5"]
						],
						[
							7,
							"surf0",
							"y"
						],
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"pdy",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							1,
							[
								7,
								"sdfUV",
								"y"
							],
							"-",
							[5, "0.5"]
						],
						[
							7,
							"surf0",
							"z"
						],
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"thicknessRange",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"thickness",
						"*",
						[5, "0.3"]
					], [5, "0.01"]]
				]
			],
			[
				13,
				"depthNorm",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"sdf",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"/",
							"thicknessRange"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"ddx",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"surfX",
									"w"
								],
								"-",
								[
									7,
									"surf0",
									"w"
								]
							],
							"/",
							"EPS"
						],
						[
							4,
							"-",
							[5, "4"]
						],
						[5, "4"]
					]
				]
			],
			[
				13,
				"ddy",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"surfY",
									"w"
								],
								"-",
								[
									7,
									"surf0",
									"w"
								]
							],
							"/",
							"EPS"
						],
						[
							4,
							"-",
							[5, "4"]
						],
						[5, "4"]
					]
				]
			],
			[
				13,
				"nVol",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"ddx",
							"ddy",
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]]
				]
			],
			[
				13,
				"bevelSlope",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								1,
								[5, "1"],
								"-",
								[
									1,
									"depthNorm",
									"*",
									"depthNorm"
								]
							], [
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							]]
						]]
					],
					"*",
					[5, "2.4"]
				]
			],
			[
				13,
				"nFlat",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								"gradX",
								"*",
								"bevelSlope"
							],
							[
								1,
								"gradY",
								"*",
								"bevelSlope"
							],
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]]
				]
			],
			[
				13,
				"nBase",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"nFlat",
						"nVol",
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"cScale",
				[
					1,
					"crumpleScale",
					"*",
					[5, "4"]
				]
			],
			[
				13,
				"cAmp",
				[
					1,
					"crumple",
					"*",
					[5, "0.9"]
				]
			],
			[
				13,
				"seedOff",
				[
					1,
					"seed",
					"*",
					[5, "7.31"]
				]
			],
			[
				13,
				"neps",
				[5, "0.012"]
			],
			[
				13,
				"cr0",
				[
					6,
					"plasticCrumple",
					[
						"pdx",
						"pdy",
						"cScale",
						"seedOff"
					]
				]
			],
			[
				13,
				"crx",
				[
					1,
					[
						1,
						[
							6,
							"plasticCrumple",
							[
								[
									1,
									"pdx",
									"+",
									"neps"
								],
								"pdy",
								"cScale",
								"seedOff"
							]
						],
						"-",
						"cr0"
					],
					"/",
					"neps"
				]
			],
			[
				13,
				"cry",
				[
					1,
					[
						1,
						[
							6,
							"plasticCrumple",
							[
								"pdx",
								[
									1,
									"pdy",
									"+",
									"neps"
								],
								"cScale",
								"seedOff"
							]
						],
						"-",
						"cr0"
					],
					"/",
					"neps"
				]
			],
			[
				13,
				"coverage",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"crumpleCoverage",
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"covNoise",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"noise",
								"mxNoiseFloat2"
							],
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									1,
									[
										1,
										[
											1,
											"pdx",
											"*",
											[5, "2.6"]
										],
										"+",
										"seedOff"
									],
									"+",
									[5, "31.7"]
								], [
									1,
									[
										1,
										[
											1,
											"pdy",
											"*",
											[5, "2.6"]
										],
										"+",
										[
											1,
											"seedOff",
											"*",
											[5, "0.617"]
										]
									],
									"+",
									[5, "17.3"]
								]]
							]]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"covThreshold",
				[
					1,
					[5, "1.25"],
					"-",
					[
						1,
						"coverage",
						"*",
						[5, "1.75"]
					]
				]
			],
			[
				13,
				"covMask",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"covThreshold",
							"-",
							[5, "0.25"]
						],
						[
							1,
							"covThreshold",
							"+",
							[5, "0.25"]
						],
						"covNoise"
					]
				]
			],
			[
				13,
				"crumpleW",
				[
					1,
					[
						1,
						"cAmp",
						"*",
						"covMask"
					],
					"*",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									[5, "0"],
									[5, "0.35"],
									"depthNorm"
								]
							],
							"*",
							[5, "0.85"]
						],
						"+",
						[5, "0.15"]
					]
				]
			],
			[
				13,
				"n",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"nBase",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									[
										1,
										"crx",
										"*",
										"crumpleW"
									],
									"*",
									[5, "0.25"]
								],
								[
									1,
									[
										1,
										"cry",
										"*",
										"crumpleW"
									],
									"*",
									[5, "0.25"]
								],
								[5, "0"]
							]
						]]
					]]
				]
			],
			[
				13,
				"cosView",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								7,
								"n",
								"z"
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"grazing",
				[
					1,
					[5, "1"],
					"-",
					"cosView"
				]
			],
			[
				13,
				"R",
				[
					6,
					[
						7,
						"std",
						"reflect"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0"],
							[5, "0"],
							[5, "1"]
						]
					], "n"]
				]
			],
			[
				13,
				"rough",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"roughness",
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"grain",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat2"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							"pdx",
							"*",
							[5, "140"]
						], [
							1,
							"pdy",
							"*",
							[5, "140"]
						]]
					]]
				]
			],
			[
				13,
				"rJit",
				[
					1,
					[
						1,
						"rough",
						"*",
						"grain"
					],
					"*",
					[5, "0.18"]
				]
			],
			[
				13,
				"eu",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"R",
								"x"
							],
							"*",
							"lx"
						],
						"+",
						[
							1,
							[
								7,
								"R",
								"y"
							],
							"*",
							"ly"
						]
					],
					"+",
					"rJit"
				]
			],
			[
				13,
				"ev",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"R",
								"x"
							],
							"*",
							[
								1,
								"ly",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]
						],
						"+",
						[
							1,
							[
								7,
								"R",
								"y"
							],
							"*",
							"lx"
						]
					],
					"+",
					[
						1,
						"rJit",
						"*",
						[5, "0.6"]
					]
				]
			],
			[
				13,
				"soft",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "0.02"],
						[5, "0.45"],
						"rough"
					]
				]
			],
			[
				13,
				"windowBox",
				[
					1,
					[
						6,
						"plasticEnvBox",
						[
							"eu",
							[5, "0.18"],
							[5, "0.95"],
							"soft"
						]
					],
					"*",
					[
						6,
						"plasticEnvBox",
						[
							"ev",
							[
								4,
								"-",
								[5, "0.42"]
							],
							[5, "0.42"],
							"soft"
						]
					]
				]
			],
			[
				13,
				"mullion",
				[
					1,
					[
						1,
						[
							6,
							"plasticEnvBox",
							[
								"ev",
								[
									4,
									"-",
									[5, "0.045"]
								],
								[5, "0.045"],
								[
									1,
									"soft",
									"*",
									[5, "0.6"]
								]
							]
						],
						"*",
						[5, "0.8"]
					],
					"*",
					[
						1,
						[5, "1"],
						"-",
						"rough"
					]
				]
			],
			[
				13,
				"keyLight",
				[
					1,
					"windowBox",
					"*",
					[
						1,
						[5, "1"],
						"-",
						"mullion"
					]
				]
			],
			[
				13,
				"strip",
				[
					1,
					[
						1,
						[
							6,
							"plasticEnvBox",
							[
								"eu",
								[
									4,
									"-",
									[5, "0.16"]
								],
								[
									4,
									"-",
									[5, "0.05"]
								],
								[
									1,
									"soft",
									"*",
									[5, "0.7"]
								]
							]
						],
						"*",
						[
							6,
							"plasticEnvBox",
							[
								"ev",
								[
									4,
									"-",
									[5, "0.85"]
								],
								[5, "0.85"],
								"soft"
							]
						]
					],
					"*",
					[5, "0.85"]
				]
			],
			[
				13,
				"fill",
				[
					1,
					[
						1,
						[
							6,
							"plasticEnvBox",
							[
								"eu",
								[
									4,
									"-",
									[5, "0.95"]
								],
								[
									4,
									"-",
									[5, "0.45"]
								],
								[
									1,
									"soft",
									"*",
									[5, "1.6"]
								]
							]
						],
						"*",
						[
							6,
							"plasticEnvBox",
							[
								"ev",
								[
									4,
									"-",
									[5, "0.6"]
								],
								[5, "0.6"],
								[
									1,
									"soft",
									"*",
									[5, "1.6"]
								]
							]
						]
					],
					"*",
					[5, "0.3"]
				]
			],
			[
				13,
				"dome",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									"eu",
									"*",
									[5, "0.5"]
								],
								"+",
								[5, "0.5"]
							],
							[5, "0"],
							[5, "1"]
						]
					],
					"*",
					[5, "0.12"]
				]
			],
			[
				13,
				"env",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								"keyLight",
								"+",
								"strip"
							],
							"+",
							"fill"
						],
						"+",
						"dome"
					],
					"*",
					[
						1,
						[5, "1"],
						"-",
						[
							1,
							"rough",
							"*",
							[5, "0.55"]
						]
					]
				]
			],
			[
				13,
				"fresnel",
				[
					1,
					[5, "0.3"],
					"+",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							["grazing", [
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "3"]]
							]]
						],
						"*",
						[5, "0.7"]
					]
				]
			],
			[
				13,
				"specW",
				[
					1,
					"reflectivity",
					"*",
					"fresnel"
				]
			],
			[
				13,
				"L",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"lx",
							"ly",
							[
								4,
								"-",
								[5, "0.8"]
							]
						]
					]]
				]
			],
			[
				13,
				"H",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"L",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0"],
								[5, "0"],
								[
									4,
									"-",
									[5, "1"]
								]
							]
						]]
					]]
				]
			],
			[
				13,
				"ndh",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["n", "H"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"shininess",
				[
					6,
					[
						7,
						"std",
						"exp2"
					],
					[[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "9"]]
							],
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "4"]]
							],
							"rough"
						]
					]]
				]
			],
			[
				13,
				"glint",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							["ndh", "shininess"]
						],
						"*",
						[5, "1.6"]
					],
					"*",
					[
						1,
						[5, "1"],
						"-",
						[
							1,
							"rough",
							"*",
							[5, "0.7"]
						]
					]
				]
			],
			[
				13,
				"specular",
				[
					1,
					[
						1,
						"env",
						"+",
						"glint"
					],
					"*",
					"specW"
				]
			],
			[
				13,
				"ndl",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["n", "L"]
						],
						[
							4,
							"-",
							[5, "1"]
						],
						[5, "1"]
					]
				]
			],
			[
				13,
				"lit",
				[
					1,
					[
						1,
						"ndl",
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"diffuse",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							1,
							[5, "1"],
							"-",
							[
								1,
								"shading",
								"*",
								[5, "0.75"]
							]
						],
						[
							1,
							[5, "1"],
							"+",
							[
								1,
								"shading",
								"*",
								[5, "0.45"]
							]
						],
						"lit"
					]
				]
			],
			[
				13,
				"ao",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "0.6"],
						[5, "1"],
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[5, "0"],
								[5, "1"],
								[
									6,
									[
										7,
										"std",
										"clamp"
									],
									[
										[
											1,
											[
												1,
												"sdf",
												"*",
												[
													4,
													"-",
													[5, "1"]
												]
											],
											"/",
											[5, "0.025"]
										],
										[5, "0"],
										[5, "1"]
									]
								]
							]
						]
					]
				]
			],
			[
				13,
				"rimGlow",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"pow"
								],
								["grazing", [
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "3.5"]]
								]]
							],
							"*",
							"rimP"
						],
						"*",
						[
							1,
							[
								1,
								"lit",
								"*",
								[5, "0.6"]
							],
							"+",
							[5, "0.4"]
						]
					],
					"*",
					[5, "1.6"]
				]
			],
			[
				13,
				"plasticRgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										[
											6,
											[
												7,
												"baseRgb",
												"mul"
											],
											["diffuse"]
										],
										"mul"
									],
									["ao"]
								],
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									"specular",
									"specular",
									"specular"
								]
							]]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"rimColor",
							"mul"
						],
						["rimGlow"]
					]]
				]
			],
			[
				13,
				"transition",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[5, "1"],
						"rb1"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"plasticRgb",
						"x"
					],
					[
						7,
						"plasticRgb",
						"y"
					],
					[
						7,
						"plasticRgb",
						"z"
					],
					"transition"
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"DEG_TO_RAD",
			"plasticCrumple",
			"noise",
			"plasticEnvBox"
		]
	},
	externals: () => ({
		std,
		d,
		DEG_TO_RAD,
		plasticCrumple,
		noise: noise_exports,
		plasticEnvBox
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Plastic",
	boundingBoxDeclaration: shapeEffectBoundingBoxDeclaration,
	category: "Shape Effects",
	description: "Glossy molded plastic with photorealistic studio reflections, driven in a custom shape",
	animatedTime: { speed: "speed" },
	extraFields: {
		...VOLUMETRIC_FIELD_EXTRA_FIELDS,
		...ANALYTIC_SDF_EXTRA_FIELDS
	},
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the plastic shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		scale: {
			default: 1,
			description: "Scale of the plastic shape (1 = default size)",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Scale",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the plastic shape in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		},
		colorA: {
			default: "#f5f5f7",
			transform: transformColor,
			description: "Body color at the start of the gradient",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#d9dade",
			transform: transformColor,
			description: "Body color at the end of the gradient",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		stops: colorStopsPropConfig(),
		gradientAngle: {
			default: 90,
			description: "Direction of the body color gradient in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Gradient Angle",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for gradient color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		thickness: {
			default: .5,
			description: "Puffiness — how far the rounded bevel reaches inward from the edge before the surface flattens",
			ui: {
				type: "range",
				min: .05,
				max: 1,
				step: .01,
				label: "Puffiness",
				group: "Plastic"
			}
		},
		roughness: {
			default: .12,
			description: "Surface roughness — 0 = polished gloss with crisp mirror reflections, 1 = matte satin",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Roughness",
				group: "Plastic"
			}
		},
		reflectivity: {
			default: 1,
			description: "Strength of the environment reflections and specular highlights",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Reflectivity",
				group: "Plastic"
			}
		},
		crumple: {
			default: .25,
			description: "Surface crumple — irregular dents that break reflections into hard organic shapes, like blown or vacuum-formed plastic",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Crumple",
				group: "Plastic"
			}
		},
		crumpleScale: {
			default: 1,
			description: "Size of the crumple dents — higher = smaller, denser dents",
			ui: {
				type: "range",
				min: .5,
				max: 10,
				step: .1,
				label: "Crumple Scale",
				group: "Plastic"
			}
		},
		crumpleCoverage: {
			default: 1,
			description: "How much of the surface is crumpled — 0 = pristine, 0.5 = patches of crumple, 1 = fully crumpled",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Crumple Coverage",
				group: "Plastic"
			}
		},
		seed: {
			default: 0,
			description: "Random seed — re-rolls the crumple pattern for reflection variety",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Plastic"
			}
		},
		edgeSoftness: {
			default: .05,
			description: "Softness of the shape boundary edge",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Softness",
				group: "Plastic"
			}
		},
		lightAngle: {
			default: 315,
			description: "Direction the studio key light comes from, in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Light Angle",
				group: "Lighting"
			}
		},
		speed: {
			default: 1,
			description: "Speed of the environment drift — the studio reflections slowly orbit the surface. 0 pauses.",
			ui: {
				type: "range",
				min: 0,
				max: 4,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		shading: {
			default: .35,
			description: "Diffuse shading contrast — how strongly the body darkens away from the light",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Shading",
				group: "Lighting"
			}
		},
		rim: {
			default: .25,
			description: "Fresnel rim light intensity on grazing surfaces",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Rim Light",
				group: "Lighting"
			}
		},
		rimColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color of the fresnel rim light",
			ui: {
				type: "color",
				label: "Rim Color",
				group: "Lighting"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type"
		}
	},
	compute: (params) => {
		const vf = createVolumetricFieldComputeNode(params, () => params.getCpuValue("shape"), "raw");
		return vf ? {
			outputs: vf.outputs,
			getComputeNodes: (fp) => vf.getComputeNodes(fp)
		} : null;
	},
	fragment: (params) => {
		const { computeOutputs, uniforms, propValues, ctx, getCpuValue } = params;
		let sampler;
		let gradSampler;
		let volumetric = false;
		const fieldTex = computeOutputs?.volumetricFieldTexture;
		if (fieldTex) {
			const s = buildVolumetricFieldSampler(fieldTex, {
				originX: uniforms._vfOriginX,
				originY: uniforms._vfOriginY,
				spanX: uniforms._vfSpanX,
				spanY: uniforms._vfSpanY,
				activeResF: uniforms._vfActiveRes,
				rBound: uniforms._vfRBound
			});
			sampler = s.volumetricFieldSampler;
			gradSampler = s.volumetricFieldSamplerFast;
			volumetric = true;
		} else {
			const shapeSdfUrl = propValues.shapeSdfUrl || "";
			const shapeType = propValues.shapeType || "";
			if (shapeSdfUrl) {
				sampler = createSvgSdfSampler(params, shapeSdfUrl);
				gradSampler = sampler;
				volumetric = isVolumetricShapeType(shapeType);
			} else {
				const st = resolveShapeType(shapeType, getCpuValue("shape"));
				sampler = createAnalyticSdfSampler(st, driveAnalyticSubProps(params, () => getCpuValue("shape")), "raw");
				gradSampler = sampler;
				volumetric = is3dShapeType(st);
			}
		}
		const sdfUV = call(sdfSpaceUV, "sdfSpaceUV", [
			uniforms.center,
			uniforms.scale,
			uniforms.rotation,
			ctx.uv,
			ctx.aspect
		]);
		const surf0 = sampler(sdfUV);
		const surfX = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(.01),
			floatE(0)
		]));
		const surfY = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(0),
			floatE(.01)
		]));
		const gT = call(plasticGradientT, "plasticGradientT", [sdfUV, uniforms.gradientAngle]);
		const colorSpaceMode = propValues.colorSpace ?? 0;
		const stopCount = propValues.stopCount ?? 0;
		let baseColor;
		if (stopCount > 1) baseColor = mixColorStopsRuntime(gT, {
			colorsArray: uniforms.colorsArray,
			positionsArray: uniforms.positionsArray,
			convertedColorsArray: uniforms.convertedColorsArray,
			stopCount: uniforms.stopCount
		}, colorSpaceMode);
		else baseColor = call(mixColorsVariants[colorSpaceMode] ?? mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			gT
		]);
		const t = animatedTime(params);
		const volFlag = floatE(volumetric ? 1 : 0);
		const p0 = vec4(uniforms.scale, uniforms.edgeSoftness, uniforms.thickness, uniforms.roughness);
		const p1 = vec4(uniforms.reflectivity, uniforms.crumple, uniforms.crumpleScale, uniforms.crumpleCoverage);
		const p2 = vec4(uniforms.seed, uniforms.lightAngle, uniforms.shading, uniforms.rim);
		const p3 = vec4(t, volFlag, floatE(0), floatE(0));
		return call(plasticComposite, "plasticComposite", [
			surf0,
			surfX,
			surfY,
			sdfUV,
			ctx.viewportSize,
			baseColor.member("rgb"),
			uniforms.rimColor.member("rgb"),
			p0,
			p1,
			p2,
			p3
		]);
	}
};
var Plastic_default = componentDefinition;
export { plasticEnvBox as a, plasticCrumple as i, componentDefinition as n, plasticGradientT as o, plasticComposite as r, Plastic_default as t };
