// shaders/dist/core/Plasma-BELcebZa.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, gt as mixColorStopsRuntime, n as tgpu, on as d, rt as noise_exports, tt as mxNoiseFloat3 } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig, t as animatedTime } from "./porters-Dg80Yg90.js";
const plasmaTone = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((fluid, intensity, contrast, balance) => {
	"use gpu";
	const plasma = std.pow(fluid, __tsover_div(5, intensity));
	const contrasted = std.clamp(__tsover_add(__tsover_mul(__tsover_sub(plasma, .5), contrast), .5), 0, 1);
	const balanceShift = __tsover_sub(__tsover_div(balance, 100), .5);
	const balanced = std.clamp(__tsover_add(contrasted, balanceShift), 0, 1);
	return __tsover_sub(1, balanced);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "fluid"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "contrast"
			},
			{
				"type": "i",
				"name": "balance"
			}
		],
		"body": [0, [
			[
				13,
				"plasma",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					["fluid", [
						1,
						[5, "5"],
						"/",
						"intensity"
					]]
				]
			],
			[
				13,
				"contrasted",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									"plasma",
									"-",
									[5, "0.5"]
								],
								"*",
								"contrast"
							],
							"+",
							[5, "0.5"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"balanceShift",
				[
					1,
					[
						1,
						"balance",
						"/",
						[5, "100"]
					],
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"balanced",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"contrasted",
							"+",
							"balanceShift"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, [
				1,
				[5, "1"],
				"-",
				"balanced"
			]]
		]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const plasmaT = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, density, warp, intensity, contrast, balance, animTime) => {
	"use gpu";
	const t = __tsover_mul(animTime, .125);
	const aspect = __tsover_div(viewport.x, viewport.y);
	const pos = d.vec3f(__tsover_mul(uv.x, aspect), uv.y, 0).mul(std.exp(__tsover_sub(density, 1)));
	const warpAmount = __tsover_mul(warp, 4);
	const qx = mxNoiseFloat3(pos.add(d.vec3f(__tsover_mul(t, .3), __tsover_mul(t, -.2), 0)));
	const qy = mxNoiseFloat3(pos.add(d.vec3f(5.2, 1.3, __tsover_mul(t, .25))));
	const warped1 = pos.add(d.vec3f(qx, qy, 0).mul(warpAmount));
	const rx = mxNoiseFloat3(warped1.add(d.vec3f(1.7, 9.2, __tsover_mul(t, .15))));
	const ry = mxNoiseFloat3(warped1.add(d.vec3f(8.3, 2.8, __tsover_mul(t, .13))));
	const warped2 = pos.add(d.vec3f(rx, ry, 0).mul(warpAmount));
	return plasmaTone(__tsover_add(__tsover_mul(mxNoiseFloat3(warped2), .5), .5), intensity, contrast, balance);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "density"
			},
			{
				"type": "i",
				"name": "warp"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "contrast"
			},
			{
				"type": "i",
				"name": "balance"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"t",
				[
					1,
					"animTime",
					"*",
					[5, "0.125"]
				]
			],
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"base",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"*",
							"aspect"
						],
						[
							7,
							"uv",
							"y"
						],
						[5, "0"]
					]
				]
			],
			[
				13,
				"pos",
				[
					6,
					[
						7,
						"base",
						"mul"
					],
					[[
						6,
						[
							7,
							"std",
							"exp"
						],
						[[
							1,
							"density",
							"-",
							[5, "1"]
						]]
					]]
				]
			],
			[
				13,
				"warpAmount",
				[
					1,
					"warp",
					"*",
					[5, "4"]
				]
			],
			[
				13,
				"qx",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					[[
						6,
						[
							7,
							"pos",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									1,
									"t",
									"*",
									[5, "0.3"]
								],
								[
									1,
									"t",
									"*",
									[
										4,
										"-",
										[5, "0.2"]
									]
								],
								[5, "0"]
							]
						]]
					]]
				]
			],
			[
				13,
				"qy",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					[[
						6,
						[
							7,
							"pos",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "5.2"],
								[5, "1.3"],
								[
									1,
									"t",
									"*",
									[5, "0.25"]
								]
							]
						]]
					]]
				]
			],
			[
				13,
				"warped1",
				[
					6,
					[
						7,
						"pos",
						"add"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									"qx",
									"qy",
									[5, "0"]
								]
							],
							"mul"
						],
						["warpAmount"]
					]]
				]
			],
			[
				13,
				"rx",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					[[
						6,
						[
							7,
							"warped1",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1.7"],
								[5, "9.2"],
								[
									1,
									"t",
									"*",
									[5, "0.15"]
								]
							]
						]]
					]]
				]
			],
			[
				13,
				"ry",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					[[
						6,
						[
							7,
							"warped1",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "8.3"],
								[5, "2.8"],
								[
									1,
									"t",
									"*",
									[5, "0.13"]
								]
							]
						]]
					]]
				]
			],
			[
				13,
				"warped2",
				[
					6,
					[
						7,
						"pos",
						"add"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									"rx",
									"ry",
									[5, "0"]
								]
							],
							"mul"
						],
						["warpAmount"]
					]]
				]
			],
			[
				13,
				"fluid",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"noise",
								"mxNoiseFloat3"
							],
							["warped2"]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[10, [
				6,
				"plasmaTone",
				[
					"fluid",
					"intensity",
					"contrast",
					"balance"
				]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"noise",
			"plasmaTone"
		]
	},
	externals: () => ({
		d,
		std,
		noise: noise_exports,
		plasmaTone
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Plasma",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Animated effect of glowing plasma",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		density: {
			default: 2,
			description: "Density of the plasma pattern",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 4,
				step: .1,
				label: "Density",
				group: "Effect"
			}
		},
		speed: {
			default: 2,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Effect"
			}
		},
		intensity: {
			default: 1.5,
			description: "Brightness and spread of the plasma glow",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 3,
				step: .1,
				label: "Intensity",
				group: "Colors"
			}
		},
		warp: {
			default: .4,
			description: "How much the flow distorts and swirls",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Warp",
				group: "Effect"
			}
		},
		contrast: {
			default: 1,
			description: "Push darks darker and lights lighter",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 3,
				step: .1,
				label: "Contrast",
				group: "Colors"
			}
		},
		balance: {
			default: 50,
			description: "Skew color balance toward A (higher) or B (lower)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Balance",
				group: "Colors"
			}
		},
		colorA: {
			default: "#7018be",
			description: "Primary color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#000000",
			description: "Secondary color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		stops: colorStopsPropConfig(),
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const uv = params.uvContext ?? ctx.uv;
		const viewport = params.effectiveViewportSize ?? ctx.viewportSize;
		const t = animatedTime(params);
		const tColor = call(plasmaT, "plasmaT", [
			uv,
			viewport,
			uniforms.density,
			uniforms.warp,
			uniforms.intensity,
			uniforms.contrast,
			uniforms.balance,
			t
		]);
		const colorSpaceMode = propValues.colorSpace ?? 0;
		if ((propValues.stopCount ?? 0) > 1) return mixColorStopsRuntime(tColor, {
			colorsArray: uniforms.colorsArray,
			positionsArray: uniforms.positionsArray,
			convertedColorsArray: uniforms.convertedColorsArray,
			stopCount: uniforms.stopCount
		}, colorSpaceMode);
		return call(mixColorsVariants[colorSpaceMode] ?? mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			tColor
		]);
	}
};
var Plasma_default = componentDefinition;
export { plasmaTone as i, componentDefinition as n, plasmaT as r, Plasma_default as t };
