// shaders/dist/core/PolarCoordinates-CNZ6Df5J.js
import { At as call, Ft as mixExpr, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-CWOT0Wqh.js";
import { c as transformPosition, s as transformEdges } from "./transformations-fqzznt4g.js";
var PI = Math.PI;
var TAU = Math.PI * 2;
const polarCoordsUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, wrap, radius, uv, aspect) => {
	"use gpu";
	const aspectCorrectedUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y));
	const delta = aspectCorrectedUV.sub(centerPos);
	const theta = std.atan2(delta.y, delta.x);
	const normalizedAngle = __tsover_div(__tsover_add(theta, PI), TAU);
	const r = std.length(delta);
	const u = __tsover_mul(normalizedAngle, wrap);
	const v = __tsover_mul(r, radius);
	return d.vec2f(u, v);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "wrap"
			},
			{
				"type": "i",
				"name": "radius"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"aspectCorrectedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"delta",
				[
					6,
					[
						7,
						"aspectCorrectedUV",
						"sub"
					],
					["centerPos"]
				]
			],
			[
				13,
				"theta",
				[
					6,
					[
						7,
						"std",
						"atan2"
					],
					[[
						7,
						"delta",
						"y"
					], [
						7,
						"delta",
						"x"
					]]
				]
			],
			[
				13,
				"normalizedAngle",
				[
					1,
					[
						1,
						"theta",
						"+",
						"PI"
					],
					"/",
					"TAU"
				]
			],
			[
				13,
				"r",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["delta"]
				]
			],
			[
				13,
				"u",
				[
					1,
					"normalizedAngle",
					"*",
					"wrap"
				]
			],
			[
				13,
				"v",
				[
					1,
					"r",
					"*",
					"radius"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["u", "v"]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"PI",
			"TAU"
		]
	},
	externals: () => ({
		d,
		std,
		PI,
		TAU
	})
}) && $.f)({}));
const componentDefinition = {
	name: "PolarCoordinates",
	category: "Distortions",
	description: "Convert rectangular coordinates to polar space",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point for polar coordinate conversion",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		wrap: {
			default: 1,
			description: "Controls how much of the angular range to use (1 = full 360°, 0.5 = 180°)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "Wrap",
				group: "Effect"
			}
		},
		radius: {
			default: 1,
			description: "Controls how much of the radius range to use (affects the radial mapping)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		intensity: {
			default: 1,
			description: "Blends between original UVs (0) and polar coordinates (1)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		edges: {
			default: "transparent",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, propValues, convertToTexture }) => {
		if (!childNode) {
			console.error("You must pass a child component into the PolarCoordinates shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		const polar = call(polarCoordsUV, "polarCoordsUV", [
			uniforms.center,
			uniforms.wrap,
			uniforms.radius,
			ctx.uv,
			ctx.aspect
		]);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(mixExpr(ctx.uv, polar, uniforms.intensity), (uv) => texture.sample(uv), edgeMode)]);
	},
	uvRemap: ({ uv, mask, uniforms, aspect, propValues }) => {
		const edgeMode = propValues.edges ?? 0;
		return composeEdgeRemapExpr(mixExpr(uv, call(polarCoordsUV, "polarCoordsUV", [
			uniforms.center,
			uniforms.wrap,
			uniforms.radius,
			uv,
			aspect
		]), uniforms.intensity), mask, edgeMode);
	}
};
var PolarCoordinates_default = componentDefinition;
export { componentDefinition as n, polarCoordsUV as r, PolarCoordinates_default as t };
