// shaders/dist/core/porters-Dg80Yg90.js
import { Nt as expr, _t as packConvertedStops, pt as MAX_COLOR_STOPS, vt as packStops } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition, n as setColorSpaceMode, o as transformColorSpace, u as Color } from "./transformations-fqzznt4g.js";
import * as d from "typegpu/data";
const MAX_COLOR_STOPS$1 = 8;
const colorStopsTransform = (value) => value;
function sortStops(stops) {
	return stops.map((stop, index) => ({
		stop,
		index
	})).sort((a, b) => a.stop.position - b.stop.position || a.index - b.index).map(({ stop }) => stop);
}
function resolveStops(props) {
	if (Array.isArray(props.stops) && props.stops.length > 0) return sortStops(props.stops);
	return [{
		color: props.colorA ?? "#1aff00",
		position: 0
	}, {
		color: props.colorB ?? "#0000ff",
		position: 1
	}];
}
function colorStopsPropConfig() {
	return {
		default: null,
		transform: colorStopsTransform,
		compileTimeWhen: (previousValue, newValue) => {
			const effective = (v) => {
				const n = Array.isArray(v) ? v.length : 0;
				return n > 1 ? n : 0;
			};
			return effective(previousValue) !== effective(newValue);
		},
		description: "Multi-stop gradient colors (overrides Color A / Color B when set)",
		ui: {
			type: "gradient-stops",
			label: "Colors",
			group: "Colors"
		}
	};
}
var gpuColorSpaceMode = "p3-linear";
function setColorSpaceModeGpu(mode) {
	gpuColorSpaceMode = mode;
	try {
		setColorSpaceMode(mode);
	} catch {}
}
function parseColorChannelsGpu(value) {
	try {
		const parsed = new Color(value);
		const targetSpace = gpuColorSpaceMode === "srgb" ? "srgb-linear" : "p3-linear";
		const coords = parsed.to(targetSpace).coords;
		return {
			r: Number.isFinite(coords?.[0]) ? coords[0] : 0,
			g: Number.isFinite(coords?.[1]) ? coords[1] : 0,
			b: Number.isFinite(coords?.[2]) ? coords[2] : 0,
			a: Number.isFinite(parsed.alpha) ? parsed.alpha : 0
		};
	} catch {
		return {
			r: 0,
			g: 0,
			b: 0,
			a: 0
		};
	}
}
function transformColorGpu(value) {
	const { r, g, b, a } = parseColorChannelsGpu(value);
	return d.vec4f(r, g, b, a);
}
function colorToRGBA(value) {
	const c = transformColorGpu(value);
	return [
		c.x,
		c.y,
		c.z,
		c.w
	];
}
function parsePositionAxis(val, isVertical) {
	if (typeof val === "number") return val;
	if (typeof val === "object" && val !== null) return val.value;
	const str = val.toLowerCase().trim();
	if (isVertical) {
		if (str === "top") return 0;
		if (str === "bottom") return 1;
		if (str === "center") return .5;
	} else {
		if (str === "left") return 0;
		if (str === "right") return 1;
		if (str === "center") return .5;
	}
	console.warn(`Invalid position value: ${val}. Defaulting to center.`);
	return .5;
}
function transformPositionGpu(value) {
	let x = .5;
	let y = .5;
	if (typeof value === "string") {
		const parts = value.toLowerCase().trim().split(/\s+/);
		if (parts.includes("left")) x = 0;
		else if (parts.includes("right")) x = 1;
		else if (parts.includes("center") || parts.length === 1 && (parts[0] === "top" || parts[0] === "bottom")) x = .5;
		if (parts.includes("top")) y = 0;
		else if (parts.includes("bottom")) y = 1;
		else if (parts.includes("center") || parts.length === 1 && (parts[0] === "left" || parts[0] === "right")) y = .5;
		if (parts.length === 1 && parts[0] === "center") {
			x = .5;
			y = .5;
		}
	} else if (typeof value === "object" && value !== null) {
		x = parsePositionAxis(value.x, false);
		y = parsePositionAxis(value.y, true);
	} else console.warn(`Invalid position value provided: ${value}. Defaulting to center.`);
	return d.vec2f(x, 1 - y);
}
function gpuTransformFor(propTransform) {
	if (!propTransform) return void 0;
	if (propTransform === transformColor) return transformColorGpu;
	if (propTransform === transformPosition) return transformPositionGpu;
	if (propTransform === colorStopsTransform) return colorStopsTransform;
	return propTransform;
}
function setShaderCanvasDimensions(width, height) {
	if (width > 0);
	if (height > 0);
}
function isDimensionalValue(v) {
	return v != null && typeof v === "object" && typeof v.value === "number" && (v.unit === "px" || v.unit === "uv");
}
function buildExtraSizeProps(props) {
	if (!props) return void 0;
	let out;
	for (const [name, cfg] of Object.entries(props)) {
		const d$1 = cfg?.ui?.dimensional;
		if (d$1 === "canvas-height" || d$1 === "half-canvas-height" || d$1 === "canvas-width" || d$1 === "count-canvas-height") (out ??= /* @__PURE__ */ new Map()).set(name, d$1);
	}
	return out;
}
function buildDimensionalPlan(decl, extraSizeProps) {
	const pb = decl?.propBindings;
	const positionProps = /* @__PURE__ */ new Set();
	const sizeProps = /* @__PURE__ */ new Map();
	if (pb) for (const axis of [
		"x",
		"y",
		"width",
		"height"
	]) {
		const b = pb[axis];
		if (!b) continue;
		if (b.as === "position-x" || b.as === "position-y") positionProps.add(b.prop);
		else if (b.as === "canvas-height" || b.as === "half-canvas-height") sizeProps.set(b.prop, b.as);
	}
	if (extraSizeProps) for (const [prop, conv] of extraSizeProps) sizeProps.set(prop, conv);
	return positionProps.size > 0 || sizeProps.size > 0 ? {
		positionProps,
		sizeProps
	} : null;
}
var resolveOriginEdges = (origin) => ({
	ox: origin.includes("left") ? "start" : origin.includes("right") ? "end" : "center",
	oy: origin.includes("top") ? "start" : origin.includes("bottom") ? "end" : "center"
});
var resolveOriginCenter = (anchorUV, edge, he) => edge === "start" ? anchorUV + he : edge === "end" ? anchorUV - he : anchorUV;
function bindingSizePx(binding, v, width, height) {
	const toPx = (val, dim) => isDimensionalValue(val) ? val.unit === "px" ? val.value : val.value * dim : typeof val === "number" ? val * dim : 0;
	if (binding.as === "canvas-height") return toPx(v, height);
	if (binding.as === "half-canvas-height") return toPx(v, 2 * height);
	if (binding.as === "canvas-width") return toPx(v, width);
	return 0;
}
function boxHalfExtentsUV(decl, props, width, height) {
	if (decl?.computeBounds) {
		const b = decl.computeBounds(props, width, height);
		return {
			hwx: width > 0 ? b.widthPx / width / 2 : 0,
			hhy: height > 0 ? b.heightPx / height / 2 : 0
		};
	}
	const pb = decl?.propBindings;
	let hwx = 0, hhy = 0;
	if (pb?.width && width > 0) hwx = bindingSizePx(pb.width, props[pb.width.prop], width, height) / (2 * width);
	if (pb?.height && height > 0) hhy = bindingSizePx(pb.height, props[pb.height.prop], width, height) / (2 * height);
	return {
		hwx,
		hhy
	};
}
function resolvePositionAxis(v, dim, edge, he) {
	let anchorUV;
	if (isDimensionalValue(v)) if (v.unit === "px") {
		const gap = dim > 0 ? v.value / dim : 0;
		anchorUV = edge === "end" ? 1 - gap : edge === "center" ? .5 + gap : gap;
	} else anchorUV = v.value;
	else if (typeof v === "number") anchorUV = v;
	else return v;
	return resolveOriginCenter(anchorUV, edge, he);
}
function resolveSize(v, conversion, width, height) {
	if (v.unit !== "px") return v.value;
	if (conversion === "canvas-width") return width > 0 ? v.value / width : v.value;
	if (height <= 0) return v.value;
	if (conversion === "count-canvas-height") return v.value > 0 ? height / v.value : v.value;
	return conversion === "canvas-height" ? v.value / height : v.value / (2 * height);
}
function resolveDimensionalProp(decl, propName, rawValue, origin, width, height, extraSizeProps, he) {
	const plan = buildDimensionalPlan(decl, extraSizeProps);
	if (!plan) return rawValue;
	const { ox, oy } = resolveOriginEdges(origin);
	if (plan.positionProps.has(propName)) {
		if (!rawValue || typeof rawValue !== "object") return rawValue;
		return {
			...rawValue,
			x: resolvePositionAxis(rawValue.x, width, ox, he?.hwx ?? 0),
			y: resolvePositionAxis(rawValue.y, height, oy, he?.hhy ?? 0)
		};
	}
	const conversion = plan.sizeProps.get(propName);
	if (conversion && isDimensionalValue(rawValue)) return resolveSize(rawValue, conversion, width, height);
	return rawValue;
}
function isDimensionalProp(decl, propName, extraSizeProps) {
	const plan = buildDimensionalPlan(decl, extraSizeProps);
	if (!plan) return false;
	return plan.positionProps.has(propName) || plan.sizeProps.has(propName);
}
function createGpuUniformsMap(definition, reactiveProps, _instanceId) {
	const uniformsMap = {};
	const decl = definition.boundingBoxDeclaration;
	const extraSizeProps = buildExtraSizeProps(definition.props);
	const activeColorSpaceMode = resolveColorSpaceMode(definition.props, reactiveProps);
	for (const [propName, propConfig] of Object.entries(definition.props)) {
		const transform = propConfig.transform;
		if (transform === colorStopsTransform) {
			expandColorStops(uniformsMap, propName, propConfig, reactiveProps[propName], activeColorSpaceMode);
			continue;
		}
		const rawValue = reactiveProps[propName];
		let value = rawValue;
		if (value === null || value === void 0) value = propConfig.default;
		if (value === null || value === void 0) {
			console.error(`[Shaders] Uniform "${propName}" is null/undefined after fallback. PropConfig:`, {
				propName,
				initialValue: rawValue,
				default: propConfig.default,
				hasTransform: !!transform
			});
			value = 0;
		}
		const gpuTransform = gpuTransformFor(transform);
		const isDim = isDimensionalProp(decl, propName, extraSizeProps);
		const cpu = !gpuTransform && !isDim && (typeof value === "string" || typeof value === "object" && value !== null && !Array.isArray(value));
		uniformsMap[propName] = {
			value,
			transform: gpuTransform,
			compileTime: propConfig.compileTime || false,
			...propConfig.compileTimeWhen ? {
				compileTimeWhen: propConfig.compileTimeWhen,
				_lastCompiledValue: rawValue ?? propConfig.default
			} : {},
			...isDim ? { _rawDimensional: rawValue } : {},
			...cpu ? { cpu: true } : {}
		};
	}
	return uniformsMap;
}
function resolveColorSpaceMode(props, reactiveProps) {
	for (const [name, cfg] of Object.entries(props)) if (cfg.transform === transformColorSpace) {
		const raw = reactiveProps[name] ?? cfg.default;
		return typeof raw === "string" ? transformColorSpace(raw) : 0;
	}
	return 0;
}
function expandColorStops(uniformsMap, propName, propConfig, rawValue, colorSpaceMode) {
	const stops = rawValue ?? null;
	const packed = packStops(stops, colorToRGBA);
	const converted = packConvertedStops(packed.colors, packed.stopCount, colorSpaceMode);
	uniformsMap[propName] = {
		value: stops,
		transform: colorStopsTransform,
		cpu: true,
		compileTime: propConfig.compileTime || false,
		...propConfig.compileTimeWhen ? {
			compileTimeWhen: propConfig.compileTimeWhen,
			_lastCompiledValue: stops ?? propConfig.default
		} : {}
	};
	uniformsMap.colorsArray = {
		value: packed.colors,
		schema: d.arrayOf(d.vec4f, 8)
	};
	uniformsMap.positionsArray = {
		value: packed.positions,
		schema: d.arrayOf(d.vec4f, 8 / 4)
	};
	uniformsMap.convertedColorsArray = {
		value: converted,
		schema: d.arrayOf(d.vec3f, 8)
	};
	uniformsMap.stopCount = {
		value: packed.stopCount,
		schema: d.f32
	};
}
const rootPassthrough = {
	name: "Root",
	props: {},
	fragment: ({ childNode }) => childNode ?? expr("vec4f(0.0, 0.0, 0.0, 0.0)")
};
function animatedTime(params, seedName, fieldName = "_animTime") {
	const base = params.props.member(fieldName);
	if (seedName) {
		const seed = params.uniforms[seedName];
		if (seed) return base.add(seed);
	}
	return base;
}
export { sortStops as _, buildDimensionalPlan as a, isDimensionalValue as c, colorToRGBA as d, setColorSpaceModeGpu as f, resolveStops as g, colorStopsTransform as h, boxHalfExtentsUV as i, resolveDimensionalProp as l, colorStopsPropConfig as m, rootPassthrough as n, buildExtraSizeProps as o, MAX_COLOR_STOPS$1 as p, createGpuUniformsMap as r, isDimensionalProp as s, animatedTime as t, setShaderCanvasDimensions as u };
