// shaders/dist/core/Posterize-Cs_kyyjG.js
import { At as call, Ot as ZERO, R as std, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
const posterizeColor = tgpu.fn([d.vec4f, d.f32], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, steps) => {
	"use gpu";
	const r = __tsover_div(std.floor(__tsover_mul(color.x, steps)), steps);
	const g = __tsover_div(std.floor(__tsover_mul(color.y, steps)), steps);
	const b = __tsover_div(std.floor(__tsover_mul(color.z, steps)), steps);
	return d.vec4f(r, g, b, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}, {
			"type": "i",
			"name": "steps"
		}],
		"body": [0, [
			[
				13,
				"r",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"floor"
						],
						[[
							1,
							[
								7,
								"color",
								"x"
							],
							"*",
							"steps"
						]]
					],
					"/",
					"steps"
				]
			],
			[
				13,
				"g",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"floor"
						],
						[[
							1,
							[
								7,
								"color",
								"y"
							],
							"*",
							"steps"
						]]
					],
					"/",
					"steps"
				]
			],
			[
				13,
				"b",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"floor"
						],
						[[
							1,
							[
								7,
								"color",
								"z"
							],
							"*",
							"steps"
						]]
					],
					"/",
					"steps"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					"r",
					"g",
					"b",
					[
						7,
						"color",
						"w"
					]
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Posterize",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Reduce color depth to create a poster effect",
	requiresChild: true,
	props: { intensity: {
		default: 5,
		description: "The intensity of the posterization effect (lower is more posterized)",
		ui: {
			type: ["range", "map"],
			min: 2,
			max: 20,
			step: 1,
			label: "Intensity",
			group: "Effect"
		}
	} },
	fragment: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Posterize shader.");
			return ZERO;
		}
		return call(posterizeColor, "posterizeColor", [childNode, uniforms.intensity]);
	}
};
var Posterize_default = componentDefinition;
export { componentDefinition as n, posterizeColor as r, Posterize_default as t };
