// shaders/dist/core/Prism-BUalIt7a.js
import { At as call, R as std, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
const prismHue2Rgb = tgpu.fn([d.f32], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((h) => {
	"use gpu";
	const x = __tsover_mul(std.fract(h), 6);
	const r = std.clamp(__tsover_sub(std.abs(__tsover_sub(x, 3)), 1), 0, 1);
	const g = std.clamp(__tsover_sub(2, std.abs(__tsover_sub(x, 2))), 0, 1);
	const b = std.clamp(__tsover_sub(2, std.abs(__tsover_sub(x, 4))), 0, 1);
	return d.vec3f(r, g, b);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "h"
		}],
		"body": [0, [
			[
				13,
				"x",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						["h"]
					],
					"*",
					[5, "6"]
				]
			],
			[
				13,
				"r",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"x",
									"-",
									[5, "3"]
								]]
							],
							"-",
							[5, "1"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"g",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[5, "2"],
							"-",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"x",
									"-",
									[5, "2"]
								]]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"b",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[5, "2"],
							"-",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									"x",
									"-",
									[5, "4"]
								]]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					"r",
					"g",
					"b"
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const prismBeam = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.vec2f,
	d.vec2f,
	d.vec3f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, vp, animTime, position, splitPosition, beamColor, beamWidth, intensity, startFalloff, endFalloff, spread, softness, saturation) => {
	"use gpu";
	const aspect = __tsover_div(vp.x, vp.y);
	const cx = __tsover_mul(position.x, aspect);
	const cy = __tsover_sub(1, position.y);
	const relX = __tsover_sub(__tsover_mul(uv.x, aspect), cx);
	const relY = __tsover_sub(uv.y, cy);
	const splitRelX = __tsover_sub(__tsover_mul(splitPosition.x, aspect), cx);
	const splitRelY = __tsover_sub(__tsover_sub(1, splitPosition.y), cy);
	const splitAlong = std.max(std.length(d.vec2f(splitRelX, splitRelY)), 1e-4);
	const dirX = __tsover_div(splitRelX, splitAlong);
	const dirY = __tsover_div(splitRelY, splitAlong);
	const perpX = __tsover_mul(dirY, -1);
	const perpY = dirX;
	const along = __tsover_add(__tsover_mul(relX, dirX), __tsover_mul(relY, dirY));
	const across = __tsover_add(__tsover_mul(relX, perpX), __tsover_mul(relY, perpY));
	const soft = std.max(softness, .001);
	const raw = __tsover_sub(along, splitAlong);
	const k = __tsover_mul(soft, .6);
	const past = __tsover_mul(__tsover_add(raw, std.sqrt(__tsover_add(__tsover_mul(raw, raw), __tsover_mul(k, k)))), .5);
	const fadeIn = std.smoothstep(0, std.max(startFalloff, .001), along);
	const fadeOut = std.exp(__tsover_mul(__tsover_div(past, std.max(endFalloff, .001)), -1));
	const longEnv = __tsover_mul(fadeIn, fadeOut);
	const fanWidth = std.max(__tsover_add(beamWidth, __tsover_mul(past, spread)), .001);
	const an = __tsover_div(across, fanWidth);
	const profile = std.exp(__tsover_mul(__tsover_mul(an, an), -2.5));
	const splitSat = std.smoothstep(splitAlong, __tsover_add(splitAlong, soft), along);
	const hue = __tsover_add(__tsover_mul(an, .42), animTime);
	const col = std.mix(beamColor, prismHue2Rgb(hue), d.vec3f(__tsover_mul(splitSat, saturation)));
	const amt = __tsover_mul(profile, longEnv);
	const rgb = col.mul(amt).mul(intensity);
	const alpha = std.clamp(amt, 0, 1);
	return d.vec4f(rgb.x, rgb.y, rgb.z, alpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "vp"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "position"
			},
			{
				"type": "i",
				"name": "splitPosition"
			},
			{
				"type": "i",
				"name": "beamColor"
			},
			{
				"type": "i",
				"name": "beamWidth"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "startFalloff"
			},
			{
				"type": "i",
				"name": "endFalloff"
			},
			{
				"type": "i",
				"name": "spread"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "saturation"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"vp",
						"x"
					],
					"/",
					[
						7,
						"vp",
						"y"
					]
				]
			],
			[
				13,
				"cx",
				[
					1,
					[
						7,
						"position",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"cy",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"position",
						"y"
					]
				]
			],
			[
				13,
				"relX",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					],
					"-",
					"cx"
				]
			],
			[
				13,
				"relY",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					"cy"
				]
			],
			[
				13,
				"splitRelX",
				[
					1,
					[
						1,
						[
							7,
							"splitPosition",
							"x"
						],
						"*",
						"aspect"
					],
					"-",
					"cx"
				]
			],
			[
				13,
				"splitRelY",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						[
							7,
							"splitPosition",
							"y"
						]
					],
					"-",
					"cy"
				]
			],
			[
				13,
				"splitAlong",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"length"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							["splitRelX", "splitRelY"]
						]]
					], [5, "0.0001"]]
				]
			],
			[
				13,
				"dirX",
				[
					1,
					"splitRelX",
					"/",
					"splitAlong"
				]
			],
			[
				13,
				"dirY",
				[
					1,
					"splitRelY",
					"/",
					"splitAlong"
				]
			],
			[
				13,
				"perpX",
				[
					1,
					"dirY",
					"*",
					[
						4,
						"-",
						[5, "1"]
					]
				]
			],
			[
				13,
				"perpY",
				"dirX"
			],
			[
				13,
				"along",
				[
					1,
					[
						1,
						"relX",
						"*",
						"dirX"
					],
					"+",
					[
						1,
						"relY",
						"*",
						"dirY"
					]
				]
			],
			[
				13,
				"across",
				[
					1,
					[
						1,
						"relX",
						"*",
						"perpX"
					],
					"+",
					[
						1,
						"relY",
						"*",
						"perpY"
					]
				]
			],
			[
				13,
				"soft",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["softness", [5, "0.001"]]
				]
			],
			[
				13,
				"raw",
				[
					1,
					"along",
					"-",
					"splitAlong"
				]
			],
			[
				13,
				"k",
				[
					1,
					"soft",
					"*",
					[5, "0.6"]
				]
			],
			[
				13,
				"past",
				[
					1,
					[
						1,
						"raw",
						"+",
						[
							6,
							[
								7,
								"std",
								"sqrt"
							],
							[[
								1,
								[
									1,
									"raw",
									"*",
									"raw"
								],
								"+",
								[
									1,
									"k",
									"*",
									"k"
								]
							]]
						]
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"fadeIn",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[
							6,
							[
								7,
								"std",
								"max"
							],
							["startFalloff", [5, "0.001"]]
						],
						"along"
					]
				]
			],
			[
				13,
				"fadeOut",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							"past",
							"/",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								["endFalloff", [5, "0.001"]]
							]
						],
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					]]
				]
			],
			[
				13,
				"longEnv",
				[
					1,
					"fadeIn",
					"*",
					"fadeOut"
				]
			],
			[
				13,
				"fanWidth",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"beamWidth",
						"+",
						[
							1,
							"past",
							"*",
							"spread"
						]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"an",
				[
					1,
					"across",
					"/",
					"fanWidth"
				]
			],
			[
				13,
				"profile",
				[
					6,
					[
						7,
						"std",
						"exp"
					],
					[[
						1,
						[
							1,
							"an",
							"*",
							"an"
						],
						"*",
						[
							4,
							"-",
							[5, "2.5"]
						]
					]]
				]
			],
			[
				13,
				"splitSat",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						"splitAlong",
						[
							1,
							"splitAlong",
							"+",
							"soft"
						],
						"along"
					]
				]
			],
			[
				13,
				"hue",
				[
					1,
					[
						1,
						"an",
						"*",
						[5, "0.42"]
					],
					"+",
					"animTime"
				]
			],
			[
				13,
				"col",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"beamColor",
						[
							6,
							"prismHue2Rgb",
							["hue"]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[
								1,
								"splitSat",
								"*",
								"saturation"
							]]
						]
					]
				]
			],
			[
				13,
				"amt",
				[
					1,
					"profile",
					"*",
					"longEnv"
				]
			],
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"col",
								"mul"
							],
							["amt"]
						],
						"mul"
					],
					["intensity"]
				]
			],
			[
				13,
				"alpha",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"amt",
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"rgb",
						"x"
					],
					[
						7,
						"rgb",
						"y"
					],
					[
						7,
						"rgb",
						"z"
					],
					"alpha"
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"prismHue2Rgb"
		]
	},
	externals: () => ({
		std,
		d,
		prismHue2Rgb
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Prism",
	category: "Textures",
	description: "A beam of light that fans out and splits into a slowly-rotating rainbow past a controllable point. Transparent background — composite over a dark layer.",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		position: {
			default: {
				x: .18,
				y: .18
			},
			transform: transformPosition,
			description: "Where the beam originates",
			ui: {
				type: "position",
				label: "Source",
				group: "Beam"
			}
		},
		beamWidth: {
			default: .04,
			description: "Thickness of the white beam",
			ui: {
				type: ["range", "map"],
				min: .002,
				max: .3,
				step: .002,
				label: "Beam Width",
				group: "Beam"
			}
		},
		intensity: {
			default: 1.6,
			description: "Brightness (above 1 blows the core to white)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 4,
				step: .01,
				label: "Intensity",
				group: "Beam"
			}
		},
		beamColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Colour of the beam before it splits",
			ui: {
				type: "color",
				label: "Beam Color",
				group: "Beam"
			}
		},
		startFalloff: {
			default: .15,
			description: "How softly the beam fades in at its source",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Start Falloff",
				group: "Beam"
			}
		},
		endFalloff: {
			default: .6,
			description: "How quickly the rainbow fades out with distance",
			ui: {
				type: ["range", "map"],
				min: .02,
				max: 2,
				step: .01,
				label: "End Falloff",
				group: "Beam"
			}
		},
		splitPosition: {
			default: {
				x: .5,
				y: .42
			},
			transform: transformPosition,
			description: "Where the beam starts to fan out and split (projected onto the beam axis)",
			ui: {
				type: "position",
				label: "Split Point",
				group: "Spectrum"
			}
		},
		spread: {
			default: .7,
			description: "How widely the rainbow diffuses as it travels past the split point",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 3,
				step: .01,
				label: "Spread",
				group: "Spectrum"
			}
		},
		softness: {
			default: .12,
			description: "How quickly the white beam transitions into the rainbow",
			ui: {
				type: ["range", "map"],
				min: .001,
				max: 1,
				step: .005,
				label: "Softness",
				group: "Spectrum"
			}
		},
		saturation: {
			default: .95,
			description: "Rainbow saturation — 0 = white, 1 = full colour",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Saturation",
				group: "Spectrum"
			}
		},
		speed: {
			default: .1,
			description: "Speed the hue slowly rotates through the rainbow (0 = static)",
			ui: {
				type: "range",
				min: -1,
				max: 1,
				step: .01,
				label: "Hue Speed",
				group: "Spectrum"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, uvContext, effectiveViewportSize } = params;
		return call(prismBeam, "prismBeam", [
			uvContext ?? ctx.uv,
			effectiveViewportSize ?? ctx.viewportSize,
			animatedTime(params),
			uniforms.position,
			uniforms.splitPosition,
			uniforms.beamColor.member("rgb"),
			uniforms.beamWidth,
			uniforms.intensity,
			uniforms.startFalloff,
			uniforms.endFalloff,
			uniforms.spread,
			uniforms.softness,
			uniforms.saturation
		]);
	}
};
var Prism_default = componentDefinition;
export { prismHue2Rgb as i, componentDefinition as n, prismBeam as r, Prism_default as t };
