// shaders/dist/core/ProgressiveBlur-C2wYIOH-.js
import { At as call, B as createGuardedCompute, It as vec4, L as createVariableGaussianBlurCompute, N as BLUR_MAP_FORMAT, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { c as transformPosition, r as transformAngle } from "./transformations-fqzznt4g.js";
var DEG_TO_RAD = Math.PI / 180;
const INTENSITY_TO_RADIUS = .36;
const intensityToRadius = (intensity) => intensity * INTENSITY_TO_RADIUS;
const progressiveBlurAmount = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((angleDeg, centerX, centerY, falloff, uv, aspect) => {
	"use gpu";
	const angleRad = __tsover_mul(angleDeg, DEG_TO_RAD);
	const direction = d.vec2f(__tsover_div(std.cos(angleRad), aspect), std.sin(angleRad));
	const centerPos = d.vec2f(centerX, __tsover_sub(1, centerY));
	const centeredUV = uv.sub(centerPos);
	const directionalDist = std.max(0, std.dot(centeredUV, direction));
	return std.smoothstep(0, falloff, directionalDist);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "angleDeg"
			},
			{
				"type": "i",
				"name": "centerX"
			},
			{
				"type": "i",
				"name": "centerY"
			},
			{
				"type": "i",
				"name": "falloff"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"angleRad",
				[
					1,
					"angleDeg",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"direction",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"cos"
							],
							["angleRad"]
						],
						"/",
						"aspect"
					], [
						6,
						[
							7,
							"std",
							"sin"
						],
						["angleRad"]
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					["centerX", [
						1,
						[5, "1"],
						"-",
						"centerY"
					]]
				]
			],
			[
				13,
				"centeredUV",
				[
					6,
					[
						7,
						"uv",
						"sub"
					],
					["centerPos"]
				]
			],
			[
				13,
				"directionalDist",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[5, "0"], [
						6,
						[
							7,
							"std",
							"dot"
						],
						["centeredUV", "direction"]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"smoothstep"
				],
				[
					[5, "0"],
					"falloff",
					"directionalDist"
				]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"d",
			"std"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		d,
		std
	})
}) && $.f)({})).$name("progressiveBlurAmount");
function progressiveFillParams() {
	return d.struct({
		angle: d.f32,
		centerX: d.f32,
		centerY: d.f32,
		falloff: d.f32,
		aspect: d.f32,
		maxRadius: d.f32
	});
}
function progressiveFillMapParams() {
	return d.struct({
		angle: d.f32,
		centerX: d.f32,
		centerY: d.f32,
		falloff: d.f32,
		aspect: d.f32,
		inputWidth: d.f32,
		inputHeight: d.f32,
		inputMin: d.f32,
		inputMax: d.f32,
		outputMin: d.f32,
		outputMax: d.f32,
		curve: d.f32
	});
}
function buildProgressiveBlurFillGraph(computeWidth, computeHeight) {
	const Params = progressiveFillParams();
	const layout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
		blurMap: { storageTexture: d.textureStorage2d(BLUR_MAP_FORMAT, "write-only") },
		params: { uniform: Params }
	}), "layout");
	return {
		layout,
		kernel: tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			const p = layout.$.params;
			const u = __tsover_div(__tsover_add(d.f32(cx), .5), computeWidth);
			const v = __tsover_div(__tsover_add(d.f32(cy), .5), computeHeight);
			const blurAmount = progressiveBlurAmount(p.angle, p.centerX, p.centerY, p.falloff, d.vec2f(u, v), p.aspect);
			const radius = __tsover_mul(blurAmount, p.maxRadius);
			std.textureStore(layout.$.blurMap, d.vec2u(cx, cy), d.vec4f(radius, 0, 0, 1));
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [
					[
						13,
						"p",
						[
							7,
							[
								7,
								"layout",
								"$"
							],
							"params"
						]
					],
					[
						13,
						"u",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									["cx"]
								],
								"+",
								[5, "0.5"]
							],
							"/",
							"computeWidth"
						]
					],
					[
						13,
						"v",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									["cy"]
								],
								"+",
								[5, "0.5"]
							],
							"/",
							"computeHeight"
						]
					],
					[
						13,
						"blurAmount",
						[
							6,
							"progressiveBlurAmount",
							[
								[
									7,
									"p",
									"angle"
								],
								[
									7,
									"p",
									"centerX"
								],
								[
									7,
									"p",
									"centerY"
								],
								[
									7,
									"p",
									"falloff"
								],
								[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									["u", "v"]
								],
								[
									7,
									"p",
									"aspect"
								]
							]
						]
					],
					[
						13,
						"radius",
						[
							1,
							"blurAmount",
							"*",
							[
								7,
								"p",
								"maxRadius"
							]
						]
					],
					[
						6,
						[
							7,
							"std",
							"textureStore"
						],
						[
							[
								7,
								[
									7,
									"layout",
									"$"
								],
								"blurMap"
							],
							[
								6,
								[
									7,
									"d",
									"vec2u"
								],
								["cx", "cy"]
							],
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									"radius",
									[5, "0"],
									[5, "0"],
									[5, "1"]
								]
							]
						]
					]
				]],
				"externalNames": [
					"layout",
					"d",
					"computeWidth",
					"computeHeight",
					"progressiveBlurAmount",
					"std"
				]
			},
			externals: () => ({
				layout,
				d,
				computeWidth,
				computeHeight,
				progressiveBlurAmount,
				std
			})
		}) && $.f)({})).$name("progressiveBlurFillBlurMap"),
		Params
	};
}
function buildProgressiveBlurFillMapGraph(computeWidth, computeHeight, channel) {
	const Params = progressiveFillMapParams();
	const layout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
		source: { texture: d.texture2d(d.f32) },
		blurMap: { storageTexture: d.textureStorage2d(BLUR_MAP_FORMAT, "write-only") },
		params: { uniform: Params }
	}), "layout");
	return {
		layout,
		kernel: tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			const p = layout.$.params;
			const u = __tsover_div(__tsover_add(d.f32(cx), .5), computeWidth);
			const v = __tsover_div(__tsover_add(d.f32(cy), .5), computeHeight);
			const blurAmount = progressiveBlurAmount(p.angle, p.centerX, p.centerY, p.falloff, d.vec2f(u, v), p.aspect);
			const srcX = std.clamp(d.i32(std.round(__tsover_mul(u, p.inputWidth))), 0, d.i32(__tsover_sub(p.inputWidth, 1)));
			const srcY = std.clamp(d.i32(std.round(__tsover_mul(v, p.inputHeight))), 0, d.i32(__tsover_sub(p.inputHeight, 1)));
			const sample = std.textureLoad(layout.$.source, d.vec2u(d.u32(srcX), d.u32(srcY)), 0);
			const raw = channel === "alpha" ? sample.w : channel === "alphaInverted" ? __tsover_sub(1, sample.w) : channel === "luminanceInverted" ? __tsover_sub(1, std.dot(sample.xyz, d.vec3f(.2126, .7152, .0722))) : std.dot(sample.xyz, d.vec3f(.2126, .7152, .0722));
			const inputRange = std.max(__tsover_sub(p.inputMax, p.inputMin), 1e-4);
			const normalised = std.clamp(__tsover_div(__tsover_sub(raw, p.inputMin), inputRange), 0, 1);
			const exponent = std.pow(2, __tsover_mul(-p.curve, 2));
			const eased = std.pow(normalised, exponent);
			const maxRadius = __tsover_mul(std.mix(p.outputMin, p.outputMax, eased), .36);
			const radius = __tsover_mul(blurAmount, maxRadius);
			std.textureStore(layout.$.blurMap, d.vec2u(cx, cy), d.vec4f(radius, 0, 0, 1));
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [
					[
						13,
						"p",
						[
							7,
							[
								7,
								"layout",
								"$"
							],
							"params"
						]
					],
					[
						13,
						"u",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									["cx"]
								],
								"+",
								[5, "0.5"]
							],
							"/",
							"computeWidth"
						]
					],
					[
						13,
						"v",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									["cy"]
								],
								"+",
								[5, "0.5"]
							],
							"/",
							"computeHeight"
						]
					],
					[
						13,
						"blurAmount",
						[
							6,
							"progressiveBlurAmount",
							[
								[
									7,
									"p",
									"angle"
								],
								[
									7,
									"p",
									"centerX"
								],
								[
									7,
									"p",
									"centerY"
								],
								[
									7,
									"p",
									"falloff"
								],
								[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									["u", "v"]
								],
								[
									7,
									"p",
									"aspect"
								]
							]
						]
					],
					[
						13,
						"srcX",
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								[
									6,
									[
										7,
										"d",
										"i32"
									],
									[[
										6,
										[
											7,
											"std",
											"round"
										],
										[[
											1,
											"u",
											"*",
											[
												7,
												"p",
												"inputWidth"
											]
										]]
									]]
								],
								[5, "0"],
								[
									6,
									[
										7,
										"d",
										"i32"
									],
									[[
										1,
										[
											7,
											"p",
											"inputWidth"
										],
										"-",
										[5, "1"]
									]]
								]
							]
						]
					],
					[
						13,
						"srcY",
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								[
									6,
									[
										7,
										"d",
										"i32"
									],
									[[
										6,
										[
											7,
											"std",
											"round"
										],
										[[
											1,
											"v",
											"*",
											[
												7,
												"p",
												"inputHeight"
											]
										]]
									]]
								],
								[5, "0"],
								[
									6,
									[
										7,
										"d",
										"i32"
									],
									[[
										1,
										[
											7,
											"p",
											"inputHeight"
										],
										"-",
										[5, "1"]
									]]
								]
							]
						]
					],
					[
						13,
						"sample",
						[
							6,
							[
								7,
								"std",
								"textureLoad"
							],
							[
								[
									7,
									[
										7,
										"layout",
										"$"
									],
									"source"
								],
								[
									6,
									[
										7,
										"d",
										"vec2u"
									],
									[[
										6,
										[
											7,
											"d",
											"u32"
										],
										["srcX"]
									], [
										6,
										[
											7,
											"d",
											"u32"
										],
										["srcY"]
									]]
								],
								[5, "0"]
							]
						]
					],
					[
						13,
						"raw",
						[
							105,
							[
								1,
								"channel",
								"===",
								[103, "alpha"]
							],
							[
								7,
								"sample",
								"w"
							],
							[
								105,
								[
									1,
									"channel",
									"===",
									[103, "alphaInverted"]
								],
								[
									1,
									[5, "1"],
									"-",
									[
										7,
										"sample",
										"w"
									]
								],
								[
									105,
									[
										1,
										"channel",
										"===",
										[103, "luminanceInverted"]
									],
									[
										1,
										[5, "1"],
										"-",
										[
											6,
											[
												7,
												"std",
												"dot"
											],
											[[
												7,
												"sample",
												"xyz"
											], [
												6,
												[
													7,
													"d",
													"vec3f"
												],
												[
													[5, "0.2126"],
													[5, "0.7152"],
													[5, "0.0722"]
												]
											]]
										]
									],
									[
										6,
										[
											7,
											"std",
											"dot"
										],
										[[
											7,
											"sample",
											"xyz"
										], [
											6,
											[
												7,
												"d",
												"vec3f"
											],
											[
												[5, "0.2126"],
												[5, "0.7152"],
												[5, "0.0722"]
											]
										]]
									]
								]
							]
						]
					],
					[
						13,
						"inputRange",
						[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								1,
								[
									7,
									"p",
									"inputMax"
								],
								"-",
								[
									7,
									"p",
									"inputMin"
								]
							], [5, "0.0001"]]
						]
					],
					[
						13,
						"normalised",
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								[
									1,
									[
										1,
										"raw",
										"-",
										[
											7,
											"p",
											"inputMin"
										]
									],
									"/",
									"inputRange"
								],
								[5, "0"],
								[5, "1"]
							]
						]
					],
					[
						13,
						"exponent",
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							[[5, "2"], [
								1,
								[
									4,
									"-",
									[
										7,
										"p",
										"curve"
									]
								],
								"*",
								[5, "2"]
							]]
						]
					],
					[
						13,
						"eased",
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							["normalised", "exponent"]
						]
					],
					[
						13,
						"maxRadius",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"mix"
								],
								[
									[
										7,
										"p",
										"outputMin"
									],
									[
										7,
										"p",
										"outputMax"
									],
									"eased"
								]
							],
							"*",
							"INTENSITY_TO_RADIUS"
						]
					],
					[
						13,
						"radius",
						[
							1,
							"blurAmount",
							"*",
							"maxRadius"
						]
					],
					[
						6,
						[
							7,
							"std",
							"textureStore"
						],
						[
							[
								7,
								[
									7,
									"layout",
									"$"
								],
								"blurMap"
							],
							[
								6,
								[
									7,
									"d",
									"vec2u"
								],
								["cx", "cy"]
							],
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									"radius",
									[5, "0"],
									[5, "0"],
									[5, "1"]
								]
							]
						]
					]
				]],
				"externalNames": [
					"layout",
					"d",
					"computeWidth",
					"computeHeight",
					"progressiveBlurAmount",
					"std",
					"channel",
					"INTENSITY_TO_RADIUS"
				]
			},
			externals: () => ({
				layout,
				d,
				computeWidth,
				computeHeight,
				progressiveBlurAmount,
				std,
				channel,
				INTENSITY_TO_RADIUS: .36
			})
		}) && $.f)({})).$name("progressiveBlurFillBlurMapVariable"),
		Params
	};
}
const componentDefinition = {
	name: "ProgressiveBlur",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Blurs",
	description: "Blur that increases progressively in one direction",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 50,
			description: "Maximum intensity of the blur effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Direction of the blur gradient (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		center: {
			default: {
				x: 0,
				y: .5
			},
			transform: transformPosition,
			description: "Center point where blur begins",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		falloff: {
			default: 1,
			description: "Distance over which blur transitions to full strength",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Falloff",
				group: "Effect"
			}
		}
	},
	compute: (params) => {
		const { childNode, gpu, convertToTexture, registerComputeTexture, getCpuValue, getMapInfo, onCleanup, onResize, dimensions } = params;
		if (!childNode) return null;
		const root = gpu?.root;
		if (!root) return null;
		const childTexture = convertToTexture(childNode);
		let curWidth = Math.max(1, Math.round(dimensions.width));
		let curHeight = Math.max(1, Math.round(dimensions.height));
		const variable = createVariableGaussianBlurCompute(root, null, curWidth, curHeight, onCleanup);
		const blurredTexture = registerComputeTexture(variable.outputTexture);
		const readGeometry = () => {
			const center = getCpuValue("center");
			return {
				angle: getCpuValue("angle") ?? 0,
				centerX: typeof center?.x === "number" ? center.x : 0,
				centerY: typeof center?.y === "number" ? center.y : .5,
				falloff: getCpuValue("falloff") ?? 1,
				aspect: curWidth / curHeight
			};
		};
		const mapInfo = getMapInfo("intensity");
		if (mapInfo) {
			const graph$1 = buildProgressiveBlurFillMapGraph(variable.computeWidth, variable.computeHeight, mapInfo.channel);
			const fillParams$1 = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(graph$1.Params), "fillParams");
			let fillPass$1 = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
				"use gpu";
				graph$1.kernel(cx, cy);
			}), {
				v: 1,
				name: void 0,
				ast: {
					"params": [{
						"type": "i",
						"name": "cx"
					}, {
						"type": "i",
						"name": "cy"
					}],
					"body": [0, [[
						6,
						[
							7,
							"graph",
							"kernel"
						],
						["cx", "cy"]
					]]],
					"externalNames": ["graph"]
				},
				externals: () => ({ graph: graph$1 })
			}) && $.f)({}), { size: [variable.computeWidth, variable.computeHeight] });
			onResize(({ width, height }) => {
				curWidth = Math.max(1, Math.round(width));
				curHeight = Math.max(1, Math.round(height));
				variable.setInputDimensions(curWidth, curHeight);
			});
			return {
				outputs: {
					childTexture,
					blurredTexture
				},
				bindInputs: (resolve) => {
					const src = resolve(childTexture.key);
					if (src) variable.setInputTexture(src.texture);
					const mapSrc = resolve(mapInfo.sourceTexture.key);
					if (mapSrc) {
						const bg = root.createBindGroup(graph$1.layout, {
							source: mapSrc.texture,
							blurMap: variable.blurMapTexture,
							params: fillParams$1.buffer
						});
						fillPass$1 = fillPass$1.with(bg);
					}
				},
				getComputeNodes: () => {
					const g = readGeometry();
					const w = mapInfo.window();
					fillParams$1.write({
						...g,
						inputWidth: curWidth,
						inputHeight: curHeight,
						inputMin: w.inputMin,
						inputMax: w.inputMax,
						outputMin: w.outputMin,
						outputMax: w.outputMax,
						curve: w.curve
					});
					return [fillPass$1, ...variable.computeSteps];
				}
			};
		}
		const graph = buildProgressiveBlurFillGraph(variable.computeWidth, variable.computeHeight);
		const fillParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(graph.Params), "fillParams");
		const fillBindGroup = root.createBindGroup(graph.layout, {
			blurMap: variable.blurMapTexture,
			params: fillParams.buffer
		});
		const fillPass = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			graph.kernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					[
						7,
						"graph",
						"kernel"
					],
					["cx", "cy"]
				]]],
				"externalNames": ["graph"]
			},
			externals: () => ({ graph })
		}) && $.f)({}), {
			size: [variable.computeWidth, variable.computeHeight],
			bindGroup: fillBindGroup
		});
		onResize(({ width, height }) => {
			curWidth = Math.max(1, Math.round(width));
			curHeight = Math.max(1, Math.round(height));
			variable.setInputDimensions(curWidth, curHeight);
		});
		return {
			outputs: {
				childTexture,
				blurredTexture
			},
			bindInputs: (resolve) => {
				const src = resolve(childTexture.key);
				if (src) variable.setInputTexture(src.texture);
			},
			getComputeNodes: () => {
				const g = readGeometry();
				const intensity = getCpuValue("intensity") ?? 0;
				fillParams.write({
					...g,
					maxRadius: intensityToRadius(intensity)
				});
				return [fillPass, ...variable.computeSteps];
			}
		};
	},
	fragment: ({ childNode, computeOutputs, ctx, convertToTexture }) => {
		if (!childNode) return ZERO;
		const blurred = computeOutputs?.blurredTexture;
		const sharp = computeOutputs?.childTexture;
		if (!blurred || !sharp) return call(unpremultiplyAlpha, "unpremultiplyAlpha", [convertToTexture(childNode).sample(ctx.uv)]);
		const sharpSample = sharp.sample(ctx.uv);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [vec4(blurred.sample(ctx.uv).member("rgb"), sharpSample.member("a"))]);
	}
};
var ProgressiveBlur_default = componentDefinition;
export { componentDefinition as a, buildProgressiveBlurFillMapGraph as i, ProgressiveBlur_default as n, intensityToRadius as o, buildProgressiveBlurFillGraph as r, progressiveBlurAmount as s, INTENSITY_TO_RADIUS as t };
