// shaders/dist/core/RadialGradient-B3tlBsQZ.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, gt as mixColorStopsRuntime, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig } from "./porters-Dg80Yg90.js";
var DEG_TO_RAD = .017453292519943295;
const radialFinalT = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.vec2f
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, skewAngle, ellipseAspect, radius, repeat, uv, viewport) => {
	"use gpu";
	const safeViewportY = std.max(viewport.y, 1e-6);
	const aspectRatio = __tsover_div(viewport.x, safeViewportY);
	const acUV = d.vec2f(__tsover_mul(uv.x, aspectRatio), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspectRatio), __tsover_sub(1, center.y));
	const dx = __tsover_sub(acUV.x, centerPos.x);
	const dy = __tsover_sub(acUV.y, centerPos.y);
	const skewRad = __tsover_mul(skewAngle, DEG_TO_RAD);
	const cosS = std.cos(skewRad);
	const sinS = std.sin(skewRad);
	const rdx = __tsover_add(__tsover_mul(dx, cosS), __tsover_mul(dy, sinS));
	const rdy = __tsover_sub(__tsover_mul(dy, cosS), __tsover_mul(dx, sinS));
	const stretched = d.vec2f(__tsover_mul(rdx, ellipseAspect), rdy);
	const dist = std.length(stretched);
	const safeRadius = std.max(radius, 1e-6);
	const t = __tsover_mul(__tsover_div(dist, safeRadius), repeat);
	return std.select(std.clamp(t, 0, 1), std.fract(t), repeat > 1.0001);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "skewAngle"
			},
			{
				"type": "i",
				"name": "ellipseAspect"
			},
			{
				"type": "i",
				"name": "radius"
			},
			{
				"type": "i",
				"name": "repeat"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			}
		],
		"body": [0, [
			[
				13,
				"safeViewportY",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						7,
						"viewport",
						"y"
					], [5, "0.000001"]]
				]
			],
			[
				13,
				"aspectRatio",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					"safeViewportY"
				]
			],
			[
				13,
				"acUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspectRatio"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspectRatio"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						7,
						"acUV",
						"x"
					],
					"-",
					[
						7,
						"centerPos",
						"x"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"acUV",
						"y"
					],
					"-",
					[
						7,
						"centerPos",
						"y"
					]
				]
			],
			[
				13,
				"skewRad",
				[
					1,
					"skewAngle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosS",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["skewRad"]
				]
			],
			[
				13,
				"sinS",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["skewRad"]
				]
			],
			[
				13,
				"rdx",
				[
					1,
					[
						1,
						"dx",
						"*",
						"cosS"
					],
					"+",
					[
						1,
						"dy",
						"*",
						"sinS"
					]
				]
			],
			[
				13,
				"rdy",
				[
					1,
					[
						1,
						"dy",
						"*",
						"cosS"
					],
					"-",
					[
						1,
						"dx",
						"*",
						"sinS"
					]
				]
			],
			[
				13,
				"stretched",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						"rdx",
						"*",
						"ellipseAspect"
					], "rdy"]
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					["stretched"]
				]
			],
			[
				13,
				"safeRadius",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["radius", [5, "0.000001"]]
				]
			],
			[
				13,
				"t",
				[
					1,
					[
						1,
						"dist",
						"/",
						"safeRadius"
					],
					"*",
					"repeat"
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"select"
				],
				[
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							"t",
							[5, "0"],
							[5, "1"]
						]
					],
					[
						6,
						[
							7,
							"std",
							"fract"
						],
						["t"]
					],
					[
						1,
						"repeat",
						">",
						[5, "1.0001"]
					]
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"DEG_TO_RAD"
		]
	},
	externals: () => ({
		std,
		d,
		DEG_TO_RAD
	})
}) && $.f)({}));
function gradientColor(params, finalT) {
	const { uniforms, propValues } = params;
	const colorSpaceMode = propValues.colorSpace ?? 0;
	if ((propValues.stopCount ?? 0) > 1) return mixColorStopsRuntime(finalT, {
		colorsArray: uniforms.colorsArray,
		positionsArray: uniforms.positionsArray,
		convertedColorsArray: uniforms.convertedColorsArray,
		stopCount: uniforms.stopCount
	}, colorSpaceMode);
	return call(mixColorsVariants[colorSpaceMode] ?? mixColorsLinear, "mixColors", [
		uniforms.colorA,
		uniforms.colorB,
		finalT
	]);
}
const componentDefinition = {
	name: "RadialGradient",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Radial gradient radiating from a center point",
	acceptsUVContext: true,
	props: {
		colorA: {
			default: "#ff0000",
			transform: transformColor,
			description: "The starting color at the center of the gradient",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#0000ff",
			transform: transformColor,
			description: "The ending color at the edge of the gradient",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		stops: colorStopsPropConfig(),
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the radial gradient",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		radius: {
			default: 1,
			description: "The radius of the gradient (normalized, 0.0-1.0)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		repeat: {
			default: 1,
			description: "Number of times the gradient repeats. Values above 1 create concentric rings.",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 20,
				step: .5,
				label: "Repeat",
				group: "Effect"
			}
		},
		aspect: {
			default: 1,
			description: "Stretches the gradient into an ellipse. Values below 1 compress vertically, above 1 compress horizontally.",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 4,
				step: .01,
				label: "Aspect",
				group: "Effect"
			}
		},
		skewAngle: {
			default: 0,
			description: "Rotates the ellipse axis in degrees. Only visible when Aspect is not 1.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Skew Angle",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		return gradientColor(params, call(radialFinalT, "radialFinalT", [
			params.uniforms.center,
			params.uniforms.skewAngle,
			params.uniforms.aspect,
			params.uniforms.radius,
			params.uniforms.repeat,
			uv,
			viewport
		]));
	}
};
var RadialGradient_default = componentDefinition;
export { componentDefinition as n, radialFinalT as r, RadialGradient_default as t };
