// shaders/dist/core/RectangularCoordinates-DJGtXzgn.js
import { At as call, Ft as mixExpr, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-BlwgBpJf.js";
import { c as transformPosition, s as transformEdges } from "./transformations-fqzznt4g.js";
var PI = Math.PI;
var TAU = Math.PI * 2;
const rectCoordsUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, scale, uv, aspect) => {
	"use gpu";
	const theta = __tsover_sub(__tsover_mul(uv.x, TAU), PI);
	const r = __tsover_mul(uv.y, scale);
	const rectX = __tsover_mul(r, std.cos(theta));
	const rectY = __tsover_mul(r, std.sin(theta));
	const centerPos = d.vec2f(center.x, __tsover_sub(1, center.y));
	return d.vec2f(__tsover_add(__tsover_div(rectX, aspect), centerPos.x), __tsover_add(rectY, centerPos.y));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"theta",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"TAU"
					],
					"-",
					"PI"
				]
			],
			[
				13,
				"r",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"*",
					"scale"
				]
			],
			[
				13,
				"rectX",
				[
					1,
					"r",
					"*",
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						["theta"]
					]
				]
			],
			[
				13,
				"rectY",
				[
					1,
					"r",
					"*",
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						["theta"]
					]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"center",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						1,
						"rectX",
						"/",
						"aspect"
					],
					"+",
					[
						7,
						"centerPos",
						"x"
					]
				], [
					1,
					"rectY",
					"+",
					[
						7,
						"centerPos",
						"y"
					]
				]]
			]]
		]],
		"externalNames": [
			"TAU",
			"PI",
			"std",
			"d"
		]
	},
	externals: () => ({
		TAU,
		PI,
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "RectangularCoordinates",
	category: "Distortions",
	description: "Convert polar coordinates back to rectangular space",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point for rectangular coordinate conversion",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		scale: {
			default: 1,
			description: "Scale factor for the rectangular output",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 3,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		intensity: {
			default: 1,
			description: "Blends between original UVs (0) and rectangular coordinates (1)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		edges: {
			default: "transparent",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, propValues, convertToTexture }) => {
		if (!childNode) {
			console.error("You must pass a child component into the RectangularCoordinates shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		const rect = call(rectCoordsUV, "rectCoordsUV", [
			uniforms.center,
			uniforms.scale,
			ctx.uv,
			ctx.aspect
		]);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(mixExpr(ctx.uv, rect, uniforms.intensity), (uv) => texture.sample(uv), edgeMode)]);
	},
	uvRemap: ({ uv, mask, uniforms, aspect, propValues }) => {
		const edgeMode = propValues.edges ?? 0;
		return composeEdgeRemapExpr(mixExpr(uv, call(rectCoordsUV, "rectCoordsUV", [
			uniforms.center,
			uniforms.scale,
			uv,
			aspect
		]), uniforms.intensity), mask, edgeMode);
	}
};
var RectangularCoordinates_default = componentDefinition;
export { componentDefinition as n, rectCoordsUV as r, RectangularCoordinates_default as t };
