// shaders/dist/core/ReflectivePlane-C4JxzSK_.js
import { At as call, B as createGuardedCompute, L as createVariableGaussianBlurCompute, N as BLUR_MAP_FORMAT, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { s as transformEdges } from "./transformations-fqzznt4g.js";
var HALF_KERNEL = 30;
var BLUR_TO_RADIUS_PX = 12;
const reflectedPlaneUV = tgpu.fn([d.vec2f, d.f32], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, floorY) => {
	"use gpu";
	return d.vec2f(uv.x, __tsover_sub(__tsover_mul(floorY, 2), uv.y));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "uv"
		}, {
			"type": "i",
			"name": "floorY"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				7,
				"uv",
				"x"
			], [
				1,
				[
					1,
					"floorY",
					"*",
					[5, "2"]
				],
				"-",
				[
					7,
					"uv",
					"y"
				]
			]]
		]]]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const reflectivePlaneCompose = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((original, refl, uvY, floorY, distance, falloff) => {
	"use gpu";
	const distBelow = std.max(__tsover_sub(uvY, floorY), 0);
	const distanceMax = std.max(distance, .001);
	const fadeWidth = std.max(__tsover_mul(distanceMax, falloff), 1e-4);
	const fadeStart = __tsover_sub(distanceMax, fadeWidth);
	const visibility = __tsover_sub(1, std.smoothstep(fadeStart, distanceMax, distBelow));
	const effReflA = __tsover_mul(refl.w, visibility);
	const oneMinusReflA = __tsover_sub(1, effReflA);
	const belowRgb = refl.xyz.mul(visibility).add(original.xyz.mul(oneMinusReflA));
	const belowA = __tsover_add(effReflA, __tsover_mul(original.w, oneMinusReflA));
	const belowFloor = d.vec4f(belowRgb, belowA);
	const isBelowFloor = std.step(floorY, uvY);
	return std.mix(original, belowFloor, d.vec4f(isBelowFloor));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "original"
			},
			{
				"type": "i",
				"name": "refl"
			},
			{
				"type": "i",
				"name": "uvY"
			},
			{
				"type": "i",
				"name": "floorY"
			},
			{
				"type": "i",
				"name": "distance"
			},
			{
				"type": "i",
				"name": "falloff"
			}
		],
		"body": [0, [
			[
				13,
				"distBelow",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"uvY",
						"-",
						"floorY"
					], [5, "0"]]
				]
			],
			[
				13,
				"distanceMax",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["distance", [5, "0.001"]]
				]
			],
			[
				13,
				"fadeWidth",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"distanceMax",
						"*",
						"falloff"
					], [5, "0.0001"]]
				]
			],
			[
				13,
				"fadeStart",
				[
					1,
					"distanceMax",
					"-",
					"fadeWidth"
				]
			],
			[
				13,
				"visibility",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							"fadeStart",
							"distanceMax",
							"distBelow"
						]
					]
				]
			],
			[
				13,
				"effReflA",
				[
					1,
					[
						7,
						"refl",
						"w"
					],
					"*",
					"visibility"
				]
			],
			[
				13,
				"oneMinusReflA",
				[
					1,
					[5, "1"],
					"-",
					"effReflA"
				]
			],
			[
				13,
				"belowRgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									7,
									"refl",
									"xyz"
								],
								"mul"
							],
							["visibility"]
						],
						"add"
					],
					[[
						6,
						[
							7,
							[
								7,
								"original",
								"xyz"
							],
							"mul"
						],
						["oneMinusReflA"]
					]]
				]
			],
			[
				13,
				"belowA",
				[
					1,
					"effReflA",
					"+",
					[
						1,
						[
							7,
							"original",
							"w"
						],
						"*",
						"oneMinusReflA"
					]
				]
			],
			[
				13,
				"belowFloor",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					["belowRgb", "belowA"]
				]
			],
			[
				13,
				"isBelowFloor",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					["floorY", "uvY"]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"mix"
				],
				[
					"original",
					"belowFloor",
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						["isBelowFloor"]
					]
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
function reflectiveFillParams() {
	return d.struct({
		height: d.f32,
		blurAmount: d.f32,
		blurDistance: d.f32
	});
}
function buildReflectivePlaneFillGraph(computeWidth, computeHeight) {
	const Params = reflectiveFillParams();
	const layout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
		blurMap: { storageTexture: d.textureStorage2d(BLUR_MAP_FORMAT, "write-only") },
		params: { uniform: Params }
	}), "layout");
	return {
		layout,
		kernel: tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			const p = layout.$.params;
			const v = __tsover_div(__tsover_add(d.f32(cy), .5), computeHeight);
			const distBelowAtOutput = std.max(__tsover_sub(p.height, v), 0);
			const ramp = std.smoothstep(0, p.blurDistance, distBelowAtOutput);
			const radius = __tsover_mul(__tsover_mul(p.blurAmount, ramp), BLUR_TO_RADIUS_PX);
			std.textureStore(layout.$.blurMap, d.vec2u(cx, cy), d.vec4f(radius, 0, 0, 1));
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [
					[
						13,
						"p",
						[
							7,
							[
								7,
								"layout",
								"$"
							],
							"params"
						]
					],
					[
						13,
						"v",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									["cy"]
								],
								"+",
								[5, "0.5"]
							],
							"/",
							"computeHeight"
						]
					],
					[
						13,
						"distBelowAtOutput",
						[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								1,
								[
									7,
									"p",
									"height"
								],
								"-",
								"v"
							], [5, "0"]]
						]
					],
					[
						13,
						"ramp",
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[5, "0"],
								[
									7,
									"p",
									"blurDistance"
								],
								"distBelowAtOutput"
							]
						]
					],
					[
						13,
						"radius",
						[
							1,
							[
								1,
								[
									7,
									"p",
									"blurAmount"
								],
								"*",
								"ramp"
							],
							"*",
							"BLUR_TO_RADIUS_PX"
						]
					],
					[
						6,
						[
							7,
							"std",
							"textureStore"
						],
						[
							[
								7,
								[
									7,
									"layout",
									"$"
								],
								"blurMap"
							],
							[
								6,
								[
									7,
									"d",
									"vec2u"
								],
								["cx", "cy"]
							],
							[
								6,
								[
									7,
									"d",
									"vec4f"
								],
								[
									"radius",
									[5, "0"],
									[5, "0"],
									[5, "1"]
								]
							]
						]
					]
				]],
				"externalNames": [
					"layout",
					"d",
					"computeHeight",
					"std",
					"BLUR_TO_RADIUS_PX"
				]
			},
			externals: () => ({
				layout,
				d,
				computeHeight,
				std,
				BLUR_TO_RADIUS_PX
			})
		}) && $.f)({})).$name("reflectivePlaneFillBlurMap"),
		Params
	};
}
const componentDefinition = {
	name: "ReflectivePlane",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Reflective floor that mirrors the content above it",
	requiresRTT: true,
	requiresChild: true,
	props: {
		height: {
			default: .7,
			description: "Vertical position of the reflective surface",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Floor Height",
				group: "Plane"
			}
		},
		distance: {
			default: .5,
			description: "How far below the floor the reflection remains visible before fully fading to transparent.",
			ui: {
				type: ["range", "map"],
				min: .01,
				max: 1,
				step: .01,
				label: "Reflection Distance",
				group: "Plane"
			}
		},
		falloff: {
			default: .5,
			description: "Width of the fade zone, as a fraction of reflection distance.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 3,
				step: .01,
				label: "Reflection Falloff",
				group: "Plane"
			}
		},
		blur: {
			default: 3,
			description: "Maximum blur applied to the reflection far from the surface.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .01,
				label: "Blur",
				group: "Surface"
			}
		},
		blurDistance: {
			default: .3,
			description: "How far below the surface the blur takes to ramp from sharp to maximum.",
			ui: {
				type: "range",
				min: .01,
				max: 1,
				step: .01,
				label: "Blur Distance",
				group: "Surface"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle reflected samples that fall outside the source content.",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Surface"
			}
		}
	},
	compute: (params) => {
		const { childNode, gpu, convertToTexture, registerComputeTexture, getCpuValue, onCleanup, onResize, dimensions } = params;
		if (!childNode) return null;
		const root = gpu?.root;
		if (!root) return null;
		const childTexture = convertToTexture(childNode);
		let curWidth = Math.max(1, Math.round(dimensions.width));
		let curHeight = Math.max(1, Math.round(dimensions.height));
		const variable = createVariableGaussianBlurCompute(root, null, curWidth, curHeight, onCleanup, HALF_KERNEL);
		const blurredTexture = registerComputeTexture(variable.outputTexture);
		const graph = buildReflectivePlaneFillGraph(variable.computeWidth, variable.computeHeight);
		const fillParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(graph.Params), "fillParams");
		const fillBindGroup = root.createBindGroup(graph.layout, {
			blurMap: variable.blurMapTexture,
			params: fillParams.buffer
		});
		const fillPass = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			graph.kernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					[
						7,
						"graph",
						"kernel"
					],
					["cx", "cy"]
				]]],
				"externalNames": ["graph"]
			},
			externals: () => ({ graph })
		}) && $.f)({}), {
			size: [variable.computeWidth, variable.computeHeight],
			bindGroup: fillBindGroup
		});
		onResize(({ width, height }) => {
			curWidth = Math.max(1, Math.round(width));
			curHeight = Math.max(1, Math.round(height));
			variable.setInputDimensions(curWidth, curHeight);
		});
		return {
			outputs: {
				childTexture,
				blurredTexture
			},
			bindInputs: (resolve) => {
				const src = resolve(childTexture.key);
				if (src) variable.setInputTexture(src.texture);
			},
			getComputeNodes: () => {
				const height = getCpuValue("height") ?? .5;
				const blurAmount = getCpuValue("blur") ?? 1;
				const blurDistance = Math.max(.01, getCpuValue("blurDistance") ?? .5);
				fillParams.write({
					height,
					blurAmount,
					blurDistance
				});
				return [fillPass, ...variable.computeSteps];
			}
		};
	},
	fragment: ({ uniforms, childNode, ctx, propValues, computeOutputs, convertToTexture }) => {
		if (!childNode) return ZERO;
		const edgeMode = propValues.edges ?? 0;
		const floorY = uniforms.height;
		const reflectedUV = call(reflectedPlaneUV, "reflectedPlaneUV", [ctx.uv, floorY]);
		const blurred = computeOutputs?.blurredTexture;
		const source = computeOutputs?.childTexture ?? convertToTexture(childNode);
		const reflectionTex = blurred ?? source;
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [call(reflectivePlaneCompose, "reflectivePlaneCompose", [
			source.sample(ctx.uv),
			applyEdgeHandlingExpr(reflectedUV, (uv) => reflectionTex.sample(uv), edgeMode),
			ctx.uv.member("y"),
			floorY,
			uniforms.distance,
			uniforms.falloff
		])]);
	}
};
var ReflectivePlane_default = componentDefinition;
export { reflectivePlaneCompose as a, reflectedPlaneUV as i, buildReflectivePlaneFillGraph as n, componentDefinition as r, ReflectivePlane_default as t };
