// shaders/dist/core/Repeater-V37OGlry.js
import { At as call, Lt as Expr, Nt as expr, Ot as ZERO, Rt as formatFloat, Wt as unpremultiplyAlpha } from "./coords-CWOT0Wqh.js";
import { i as transformBoolean } from "./transformations-fqzznt4g.js";
const componentDefinition = {
	name: "Repeater",
	category: "Distortions",
	description: "Repeat the child content in grid, radial or linear layouts with per-instance variation",
	requiresRTT: true,
	requiresChild: true,
	wantsBoundsParams: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		mode: {
			default: "grid",
			transform: (value) => ({
				grid: 0,
				radial: 1,
				linear: 2
			})[value] ?? 0,
			compileTime: true,
			description: "Layout used to place the repeated instances",
			ui: {
				type: "select",
				options: [
					{
						label: "Grid",
						value: "grid"
					},
					{
						label: "Radial",
						value: "radial"
					},
					{
						label: "Linear",
						value: "linear"
					}
				],
				label: "Mode",
				group: "Layout"
			}
		},
		cropLeft: {
			default: 0,
			description: "Crop inset from the left edge of the source element (fraction of its width)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .49,
				step: .005,
				label: "Crop Left",
				group: "Source"
			}
		},
		cropRight: {
			default: 0,
			description: "Crop inset from the right edge of the source element (fraction of its width)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .49,
				step: .005,
				label: "Crop Right",
				group: "Source"
			}
		},
		cropTop: {
			default: 0,
			description: "Crop inset from the top edge of the source element (fraction of its height)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .49,
				step: .005,
				label: "Crop Top",
				group: "Source"
			}
		},
		cropBottom: {
			default: 0,
			description: "Crop inset from the bottom edge of the source element (fraction of its height)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .49,
				step: .005,
				label: "Crop Bottom",
				group: "Source"
			}
		},
		columns: {
			default: 3,
			description: "Number of grid columns",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 12,
				step: 1,
				label: "Columns",
				group: "Layout",
				condition: { mode: "grid" }
			}
		},
		rows: {
			default: 3,
			description: "Number of grid rows",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 12,
				step: 1,
				label: "Rows",
				group: "Layout",
				condition: { mode: "grid" }
			}
		},
		gapX: {
			default: .05,
			description: "Horizontal gap between grid cells",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .5,
				step: .005,
				label: "Gap X",
				group: "Layout",
				condition: { mode: "grid" },
				dimensional: "canvas-width"
			}
		},
		gapY: {
			default: .05,
			description: "Vertical gap between grid cells",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .5,
				step: .005,
				label: "Gap Y",
				group: "Layout",
				condition: { mode: "grid" },
				dimensional: "canvas-height"
			}
		},
		stagger: {
			default: 0,
			description: "Horizontal offset applied to every other row (brick layout)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Stagger",
				group: "Layout",
				condition: { mode: "grid" }
			}
		},
		flip: {
			default: "none",
			transform: (value) => ({
				"none": 0,
				"alternate-flip-x": 1,
				"alternate-flip-y": 2,
				"both": 3
			})[value] ?? 0,
			compileTime: true,
			description: "Mirror alternating columns/rows for seamless tiling patterns",
			ui: {
				type: "select",
				options: [
					{
						label: "None",
						value: "none"
					},
					{
						label: "Alternate Flip X",
						value: "alternate-flip-x"
					},
					{
						label: "Alternate Flip Y",
						value: "alternate-flip-y"
					},
					{
						label: "Both",
						value: "both"
					}
				],
				label: "Flip",
				group: "Layout",
				condition: { mode: "grid" }
			}
		},
		count: {
			default: 8,
			description: "Number of repeated instances",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 64,
				step: 1,
				label: "Count",
				group: "Layout",
				condition: { mode: ["radial", "linear"] }
			}
		},
		radius: {
			default: .3,
			description: "Orbit radius of the radial layout",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .005,
				label: "Radius",
				group: "Layout",
				condition: { mode: "radial" },
				dimensional: "canvas-height"
			}
		},
		startAngle: {
			default: 0,
			description: "Angle of the first instance on the orbit",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Start Angle",
				group: "Layout",
				condition: { mode: "radial" }
			}
		},
		sweep: {
			default: 360,
			description: "Angular span the instances are distributed across",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Sweep",
				group: "Layout",
				condition: { mode: "radial" }
			}
		},
		faceCenter: {
			default: false,
			transform: transformBoolean,
			description: "Rotate each instance to face the orbit center",
			ui: {
				type: "checkbox",
				label: "Face Center",
				group: "Layout",
				condition: { mode: "radial" }
			}
		},
		direction: {
			default: 0,
			description: "Direction of the linear layout in degrees (0 = rightward)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Direction",
				group: "Layout",
				condition: { mode: "linear" }
			}
		},
		spacing: {
			default: .2,
			description: "Distance between consecutive instances",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .005,
				label: "Spacing",
				group: "Layout",
				condition: { mode: "linear" },
				dimensional: "canvas-height"
			}
		},
		instanceScale: {
			default: 1,
			description: "Progressive scale multiplier applied per instance (compounds with each copy)",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 2,
				step: .005,
				label: "Scale",
				group: "Instances"
			}
		},
		instanceRotation: {
			default: 0,
			transform: (value) => value * Math.PI / 180,
			description: "Additional rotation in degrees applied per instance (accumulates with each copy)",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Rotation",
				group: "Instances"
			}
		},
		instanceOpacity: {
			default: 1,
			description: "Progressive opacity falloff applied per instance (compounds with each copy)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .005,
				label: "Opacity",
				group: "Instances"
			}
		},
		hueShift: {
			default: 0,
			transform: (value) => value * Math.PI / 180,
			compileTimeWhen: (prev, next) => prev === 0 !== (next === 0),
			description: "Hue rotation in degrees added per instance",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Hue Shift",
				group: "Instances"
			}
		},
		phase: {
			default: 0,
			description: "Seamless layout animation phase (0-1 wraps to the identical layout)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .001,
				label: "Phase",
				group: "Instances"
			}
		},
		zOrder: {
			default: "forward",
			transform: (value) => value === "backward" ? 1 : 0,
			compileTime: true,
			description: "Stacking order of overlapping instances",
			ui: {
				type: "select",
				options: [{
					label: "Forward",
					value: "forward"
				}, {
					label: "Backward",
					value: "backward"
				}],
				label: "Z-Order",
				group: "Instances"
			}
		},
		jitterPosition: {
			default: 0,
			description: "Random per-instance position offset",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .005,
				label: "Position",
				group: "Jitter"
			}
		},
		jitterRotation: {
			default: 0,
			description: "Random per-instance rotation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .005,
				label: "Rotation",
				group: "Jitter"
			}
		},
		jitterScale: {
			default: 0,
			description: "Random per-instance scale variation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .005,
				label: "Scale",
				group: "Jitter"
			}
		},
		jitterOpacity: {
			default: 0,
			description: "Random per-instance opacity variation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .005,
				label: "Opacity",
				group: "Jitter"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for the jitter values",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Jitter"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, childNode, ctx, propValues, childBoundsParams, ownBoundsParams, convertToTexture } = params;
		if (!childNode) return ZERO;
		const texKey = convertToTexture(childNode).key;
		const modeMap = {
			grid: 0,
			radial: 1,
			linear: 2
		};
		const flipMap = {
			"none": 0,
			"alternate-flip-x": 1,
			"alternate-flip-y": 2,
			"both": 3
		};
		const mode = typeof propValues.mode === "number" ? propValues.mode : modeMap[propValues.mode] ?? 0;
		const flipMode = typeof propValues.flip === "number" ? propValues.flip : flipMap[propValues.flip] ?? 0;
		const zBack = propValues.zOrder === "backward" || propValues.zOrder === 1;
		const hueShiftActive = (propValues.hueShift ?? 0) !== 0;
		const PI = Math.PI;
		return new Expr((ec) => {
			const U = (name) => uniforms[name]._emit(ec);
			const L = (n) => formatFloat(n);
			const p = ec.freshLocal("rep");
			const stmts = [];
			const uv = `${p}_uv`;
			stmts.push(`let ${uv} = ${ctx.uv._emit(ec)};`);
			stmts.push(`let ${p}_aspect = ${ctx.aspect._emit(ec)};`);
			stmts.push(`let ${p}_pix = vec2f(${uv}.x * ${p}_aspect, ${uv}.y);`);
			const srcCx0 = childBoundsParams ? childBoundsParams.centerX._emit(ec) : L(.5);
			const srcCy0 = childBoundsParams ? childBoundsParams.centerY._emit(ec) : L(.5);
			const srcHw0 = childBoundsParams ? childBoundsParams.halfWidth._emit(ec) : L(.5);
			const srcHh0 = childBoundsParams ? childBoundsParams.halfHeight._emit(ec) : L(.5);
			stmts.push(`let ${p}_cropL = ${U("cropLeft")};`);
			stmts.push(`let ${p}_cropR = ${U("cropRight")};`);
			stmts.push(`let ${p}_cropT = ${U("cropTop")};`);
			stmts.push(`let ${p}_cropB = ${U("cropBottom")};`);
			stmts.push(`let ${p}_srcCx = (${srcCx0}) + (${p}_cropL - ${p}_cropR) * (${srcHw0});`);
			stmts.push(`let ${p}_srcCy = (${srcCy0}) + (${p}_cropT - ${p}_cropB) * (${srcHh0});`);
			stmts.push(`let ${p}_srcHw = max((${srcHw0}) * (1.0 - ${p}_cropL - ${p}_cropR), ${L(1e-6)});`);
			stmts.push(`let ${p}_srcHh = max((${srcHh0}) * (1.0 - ${p}_cropT - ${p}_cropB), ${L(1e-6)});`);
			stmts.push(`let ${p}_srcHwSq = ${p}_srcHw * ${p}_aspect;`);
			stmts.push(`let ${p}_srcHhSq = ${p}_srcHh;`);
			const pCx = ownBoundsParams ? ownBoundsParams.centerX._emit(ec) : L(.5);
			const pCy = ownBoundsParams ? ownBoundsParams.centerY._emit(ec) : L(.5);
			const pHw = ownBoundsParams ? ownBoundsParams.halfWidth._emit(ec) : L(.5);
			const pHh = ownBoundsParams ? ownBoundsParams.halfHeight._emit(ec) : L(.5);
			stmts.push(`let ${p}_fieldCenterSq = vec2f((${pCx}) * ${p}_aspect, ${pCy});`);
			stmts.push(`let ${p}_fieldHwSq = max((${pHw}) * ${p}_aspect, ${L(1e-6)});`);
			stmts.push(`let ${p}_fieldHh = max(${pHh}, ${L(1e-6)});`);
			stmts.push(`let ${p}_phase = ${U("phase")};`);
			stmts.push(`let ${p}_seed = ${U("seed")};`);
			const requested = mode === 0 ? `(${U("columns")} * ${U("rows")})` : U("count");
			stmts.push(`let ${p}_totalN = min(max(${requested}, 1.0), 64.0);`);
			stmts.push(`var ${p}_accum = vec4f(0.0, 0.0, 0.0, 0.0);`);
			stmts.push(`var ${p}_cellFitInv = 1.0;`);
			const body = [];
			body.push(`let ${p}_fi = f32(${p}_i);`);
			body.push(zBack ? `let ${p}_j = ${p}_totalN - 1.0 - ${p}_fi;` : `let ${p}_j = ${p}_fi;`);
			body.push(`var ${p}_pos = vec2f(0.0, 0.0);`);
			body.push(`var ${p}_layoutRot = 0.0;`);
			body.push(`var ${p}_flipX = 1.0;`);
			body.push(`var ${p}_flipY = 1.0;`);
			if (mode === 0) {
				body.push(`let ${p}_cols = ${U("columns")};`);
				body.push(`let ${p}_rows = ${U("rows")};`);
				body.push(`let ${p}_col = ${p}_j - ${p}_cols * floor(${p}_j / ${p}_cols);`);
				body.push(`let ${p}_row = floor(${p}_j / ${p}_cols);`);
				body.push(`let ${p}_gapXSq = ${U("gapX")} * ${p}_aspect;`);
				body.push(`let ${p}_gapY = ${U("gapY")};`);
				body.push(`let ${p}_fieldW = ${p}_fieldHwSq * 2.0;`);
				body.push(`let ${p}_fieldH = ${p}_fieldHh * 2.0;`);
				body.push(`let ${p}_pitchX = (${p}_fieldW - ${p}_gapXSq * (${p}_cols - 1.0)) / ${p}_cols;`);
				body.push(`let ${p}_pitchY = (${p}_fieldH - ${p}_gapY * (${p}_rows - 1.0)) / ${p}_rows;`);
				body.push(`let ${p}_periodX = ${p}_pitchX + ${p}_gapXSq;`);
				body.push(`let ${p}_periodY = ${p}_pitchY + ${p}_gapY;`);
				body.push(`${p}_cellFitInv = max(${p}_srcHwSq * 2.0 / max(${p}_pitchX, ${L(1e-5)}), ${p}_srcHhSq * 2.0 / max(${p}_pitchY, ${L(1e-5)}));`);
				body.push(`let ${p}_rowMod2 = ${p}_row - 2.0 * floor(${p}_row / 2.0);`);
				body.push(`let ${p}_staggerOff = ${p}_rowMod2 * ${U("stagger")} * ${p}_periodX;`);
				body.push(`let ${p}_cxArg = ${p}_col * ${p}_periodX + ${p}_staggerOff + ${p}_phase * ${p}_periodX;`);
				body.push(`let ${p}_cxMod = ${p}_periodX * ${p}_cols;`);
				body.push(`let ${p}_cellX = ${p}_cxArg - ${p}_cxMod * floor(${p}_cxArg / ${p}_cxMod);`);
				body.push(`let ${p}_cyArg = ${p}_row * ${p}_periodY + ${p}_phase * ${p}_periodY;`);
				body.push(`let ${p}_cyMod = ${p}_periodY * ${p}_rows;`);
				body.push(`let ${p}_cellY = ${p}_cyArg - ${p}_cyMod * floor(${p}_cyArg / ${p}_cyMod);`);
				body.push(`${p}_pos = vec2f(${p}_fieldCenterSq.x - ${p}_fieldW * 0.5 + ${p}_pitchX * 0.5 + ${p}_cellX, ${p}_fieldCenterSq.y - ${p}_fieldH * 0.5 + ${p}_pitchY * 0.5 + ${p}_cellY);`);
				if (flipMode === 1 || flipMode === 3) body.push(`${p}_flipX = 1.0 - (${p}_col - 2.0 * floor(${p}_col / 2.0)) * 2.0;`);
				if (flipMode === 2 || flipMode === 3) body.push(`${p}_flipY = 1.0 - (${p}_row - 2.0 * floor(${p}_row / 2.0)) * 2.0;`);
			} else {
				const cSq = childBoundsParams ? `vec2f((${childBoundsParams.centerX._emit(ec)}) * ${p}_aspect, ${childBoundsParams.centerY._emit(ec)})` : `${p}_fieldCenterSq`;
				body.push(`let ${p}_cSq = ${cSq};`);
				if (mode === 1) {
					body.push(`let ${p}_sweep = ${U("sweep")} * ${L(PI / 180)};`);
					body.push(`let ${p}_start = ${U("startAngle")} * ${L(PI / 180)};`);
					body.push(`let ${p}_ang = ${p}_start + ${p}_sweep * (${p}_j / ${p}_totalN) + ${p}_phase * ${p}_sweep;`);
					body.push(`${p}_pos = ${p}_cSq + vec2f(cos(${p}_ang), sin(${p}_ang)) * ${U("radius")};`);
					body.push(`${p}_layoutRot = select(0.0, ${p}_ang + ${L(PI / 2)}, ${U("faceCenter")} > 0.0);`);
				} else {
					body.push(`let ${p}_dirRad = ${U("direction")} * ${L(PI / 180)};`);
					body.push(`let ${p}_dir = vec2f(cos(${p}_dirRad), sin(${p}_dirRad));`);
					body.push(`let ${p}_tArg = ${p}_j + ${p}_phase;`);
					body.push(`let ${p}_t = ${p}_tArg - ${p}_totalN * floor(${p}_tArg / ${p}_totalN);`);
					body.push(`${p}_pos = ${p}_cSq + ${p}_dir * ${U("spacing")} * ${p}_t;`);
				}
			}
			body.push(`let ${p}_hashBase = ${p}_j + ${p}_seed * 7.0;`);
			const hash = (off) => `fract(sin((${p}_hashBase + ${L(off)}) * 78.233) * 43758.5453)`;
			body.push(`let ${p}_h1 = ${hash(.123)};`);
			body.push(`let ${p}_h2 = ${hash(5.456)};`);
			body.push(`let ${p}_h3 = ${hash(9.789)};`);
			body.push(`let ${p}_h4 = ${hash(13.21)};`);
			body.push(`let ${p}_h5 = ${hash(17.65)};`);
			body.push(`let ${p}_scale = max(pow(${U("instanceScale")}, ${p}_j) * (1.0 + ${U("jitterScale")} * (${p}_h3 - 0.5)), ${L(1e-4)});`);
			body.push(`let ${p}_rot = ${p}_layoutRot + ${U("instanceRotation")} * ${p}_j + ${U("jitterRotation")} * (${p}_h4 - 0.5) * ${L(PI * 2)};`);
			body.push(`let ${p}_instOpacity = pow(${U("instanceOpacity")}, ${p}_j) * (1.0 - ${U("jitterOpacity")} * ${p}_h5);`);
			body.push(`${p}_pos = ${p}_pos + vec2f((${p}_h1 - 0.5) * ${p}_fieldHwSq, (${p}_h2 - 0.5) * ${p}_fieldHh) * ${U("jitterPosition")};`);
			body.push(`let ${p}_negRot = -${p}_rot;`);
			body.push(`let ${p}_cosA = cos(${p}_negRot);`);
			body.push(`let ${p}_sinA = sin(${p}_negRot);`);
			body.push(`let ${p}_d0 = ${p}_pix - ${p}_pos;`);
			body.push(`let ${p}_dr = vec2f(${p}_d0.x * ${p}_cosA - ${p}_d0.y * ${p}_sinA, ${p}_d0.x * ${p}_sinA + ${p}_d0.y * ${p}_cosA);`);
			body.push(`let ${p}_dl = ${p}_dr / ${p}_scale;`);
			body.push(`let ${p}_localU = ${p}_dl.x * ${p}_flipX * ${p}_cellFitInv / (${p}_srcHwSq * 2.0) + 0.5;`);
			body.push(`let ${p}_localV = ${p}_dl.y * ${p}_flipY * ${p}_cellFitInv / (${p}_srcHhSq * 2.0) + 0.5;`);
			const sample = [];
			sample.push(`let ${p}_sampleUV = vec2f(${p}_srcCx - ${p}_srcHw + ${p}_localU * ${p}_srcHw * 2.0, ${p}_srcCy - ${p}_srcHh + ${p}_localV * ${p}_srcHh * 2.0);`);
			sample.push(`let ${p}_src = textureSampleLevel(tex.$.${texKey}, samp.$.linearClamp, ${p}_sampleUV, 0.0);`);
			sample.push(`var ${p}_rgb = ${p}_src.rgb;`);
			if (hueShiftActive) {
				sample.push(`let ${p}_hueAngle = ${U("hueShift")} * ${p}_j;`);
				sample.push(`let ${p}_cosH = cos(${p}_hueAngle);`);
				sample.push(`let ${p}_sinH = sin(${p}_hueAngle);`);
				sample.push(`let ${p}_kOMC = (1.0 - ${p}_cosH) * ${L(1 / 3)};`);
				sample.push(`let ${p}_sOS3 = ${p}_sinH / 1.7320508075688772;`);
				sample.push(`let ${p}_mA = ${p}_cosH + ${p}_kOMC;`);
				sample.push(`let ${p}_mB = ${p}_kOMC - ${p}_sOS3;`);
				sample.push(`let ${p}_mC = ${p}_kOMC + ${p}_sOS3;`);
				sample.push(`${p}_rgb = vec3f(${p}_src.r * ${p}_mA + ${p}_src.g * ${p}_mB + ${p}_src.b * ${p}_mC, ${p}_src.r * ${p}_mC + ${p}_src.g * ${p}_mA + ${p}_src.b * ${p}_mB, ${p}_src.r * ${p}_mB + ${p}_src.g * ${p}_mC + ${p}_src.b * ${p}_mA);`);
			}
			sample.push(`let ${p}_a = ${p}_src.a * ${p}_instOpacity;`);
			sample.push(`let ${p}_premul = ${p}_rgb * ${p}_a;`);
			sample.push(`${p}_accum = vec4f(${p}_premul + ${p}_accum.rgb * (1.0 - ${p}_a), ${p}_a + ${p}_accum.a * (1.0 - ${p}_a));`);
			body.push(`if (${p}_localU >= 0.0 && ${p}_localU <= 1.0 && ${p}_localV >= 0.0 && ${p}_localV <= 1.0) {`);
			for (const s of sample) body.push(`  ${s}`);
			body.push(`}`);
			stmts.push(`for (var ${p}_i = 0u; ${p}_i < u32(${p}_totalN); ${p}_i++) {`);
			for (const b of body) stmts.push(`  ${b}`);
			stmts.push(`}`);
			for (const s of stmts) ec.statement(s);
			return call(unpremultiplyAlpha, "unpremultiplyAlpha", [expr(`${p}_accum`)])._emit(ec);
		});
	}
};
var Repeater_default = componentDefinition;
export { componentDefinition as n, Repeater_default as t };
