// shaders/dist/core/Ring-W9djENkw.js
import { At as call, C as ringSdf, It as vec4, O as strokeMaskFromSdf, Qt as mixColorsVariants, T as sdf_exports, Zt as mixColorsLinear, n as tgpu, on as d, p as ShapeMask } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition, l as transformStrokePosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
const ringShape = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32
], ShapeMask)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, radius, thickness, softness, strokeThickness, strokePosition, uv, aspect) => {
	"use gpu";
	const aspectUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y));
	return strokeMaskFromSdf(ringSdf(__tsover_sub(aspectUV.x, centerPos.x), __tsover_sub(aspectUV.y, centerPos.y), radius, thickness), softness, strokeThickness, strokePosition);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "radius"
			},
			{
				"type": "i",
				"name": "thickness"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "strokeThickness"
			},
			{
				"type": "i",
				"name": "strokePosition"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						7,
						"aspectUV",
						"x"
					],
					"-",
					[
						7,
						"centerPos",
						"x"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"aspectUV",
						"y"
					],
					"-",
					[
						7,
						"centerPos",
						"y"
					]
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"sdf",
						"ringSdf"
					],
					[
						"dx",
						"dy",
						"radius",
						"thickness"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"sdf",
					"strokeMaskFromSdf"
				],
				[
					"dist",
					"softness",
					"strokeThickness",
					"strokePosition"
				]
			]]
		]],
		"externalNames": ["d", "sdf"]
	},
	externals: () => ({
		d,
		sdf: sdf_exports
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Ring",
	boundingBoxDeclaration: {
		propBindings: {
			x: {
				prop: "center",
				as: "position-x"
			},
			y: {
				prop: "center",
				as: "position-y"
			}
		},
		computeBounds(props, cw, ch) {
			const radius = props.radius ?? .5;
			const thickness = props.thickness ?? 0;
			const cx = props.center?.x ?? .5;
			const cy = props.center?.y ?? .5;
			const diameter = (radius + thickness) * 2 * ch;
			return {
				centerXPx: cx * cw,
				centerYPx: cy * ch,
				widthPx: diameter,
				heightPx: diameter,
				rotationDeg: 0
			};
		},
		writeBounds(bounds, currentProps, cw, ch) {
			const thickness = currentProps.thickness ?? 0;
			const newOuter = Math.max(0, bounds.widthPx / (2 * ch));
			return {
				center: {
					...currentProps.center ?? {
						x: .5,
						y: .5
					},
					x: bounds.centerXPx / cw,
					y: bounds.centerYPx / ch
				},
				radius: Math.max(0, newOuter - thickness)
			};
		}
	},
	category: "Shapes",
	description: "Annular ring (donut) with adjustable radius and band thickness",
	acceptsUVContext: true,
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		color: {
			default: "#ffffff",
			transform: transformColor,
			description: "Fill color of the ring",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the ring",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		radius: {
			default: .3,
			description: "Distance from center to the ring's midline in UV space",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Radius",
				group: "Shape"
			}
		},
		thickness: {
			default: .07,
			description: "Half-width of the ring band — total ring width is twice this value",
			ui: {
				type: "range",
				min: .005,
				max: .3,
				step: .005,
				label: "Thickness",
				group: "Shape"
			}
		},
		softness: {
			default: 0,
			description: "Edge softness for antialiasing (applied to both inner and outer ring edges)",
			ui: {
				type: "range",
				min: 0,
				max: .1,
				step: .001,
				label: "Softness",
				group: "Effect"
			}
		},
		strokeThickness: {
			default: 0,
			description: "Stroke thickness. Zero means no stroke.",
			ui: {
				type: "range",
				min: 0,
				max: .1,
				step: .005,
				label: "Stroke Thickness",
				group: "Stroke"
			}
		},
		strokeColor: {
			default: "#000000",
			transform: transformColor,
			description: "Color of the stroke outline",
			ui: {
				type: "color",
				label: "Stroke Color",
				group: "Colors"
			}
		},
		strokePosition: {
			default: "center",
			transform: transformStrokePosition,
			description: "Position of the stroke relative to the ring edge",
			ui: {
				type: "select",
				options: [
					{
						label: "Outside",
						value: "outside"
					},
					{
						label: "Center",
						value: "center"
					},
					{
						label: "Inside",
						value: "inside"
					}
				],
				label: "Stroke Position",
				group: "Stroke"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for blending fill and stroke colors",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Blending",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const uv = params.uvContext ?? ctx.uv;
		const colorSpaceMode = propValues.colorSpace ?? 0;
		const mask = call(ringShape, "ringShape", [
			uniforms.center,
			uniforms.radius,
			uniforms.thickness,
			uniforms.softness,
			uniforms.strokeThickness,
			uniforms.strokePosition,
			uv,
			ctx.aspect
		]);
		const blended = call(mixColorsVariants[colorSpaceMode] ?? mixColorsLinear, "mixColors", [
			uniforms.color,
			uniforms.strokeColor,
			mask.member("strokeBlend")
		]);
		return vec4(blended.member("rgb"), blended.member("a").mul(mask.member("overallMask")));
	}
};
var Ring_default = componentDefinition;
export { componentDefinition as n, ringShape as r, Ring_default as t };
