// shaders/dist/core/Ripples-Bv__zvme.js
import { At as call, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
const ripplesMask = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.vec2f
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, frequency, softness, thickness, phase, animTime, uv, viewport) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const dx = __tsover_sub(__tsover_mul(uv.x, aspect), __tsover_mul(center.x, aspect));
	const dy = __tsover_sub(uv.y, __tsover_sub(1, center.y));
	const dist = std.length(d.vec2f(dx, dy));
	const waveInput = __tsover_add(__tsover_sub(__tsover_mul(dist, frequency), animTime), phase);
	const baseWave = std.sin(waveInput);
	const thicknessThreshold = __tsover_sub(__tsover_mul(thickness, 2), 1);
	const banded = std.smoothstep(__tsover_sub(thicknessThreshold, softness), __tsover_add(thicknessThreshold, softness), baseWave);
	return __tsover_sub(1, banded);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "frequency"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "thickness"
			},
			{
				"type": "i",
				"name": "phase"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					],
					"-",
					[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					[
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						["dx", "dy"]
					]]
				]
			],
			[
				13,
				"waveInput",
				[
					1,
					[
						1,
						[
							1,
							"dist",
							"*",
							"frequency"
						],
						"-",
						"animTime"
					],
					"+",
					"phase"
				]
			],
			[
				13,
				"baseWave",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["waveInput"]
				]
			],
			[
				13,
				"thicknessThreshold",
				[
					1,
					[
						1,
						"thickness",
						"*",
						[5, "2"]
					],
					"-",
					[5, "1"]
				]
			],
			[
				13,
				"banded",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"thicknessThreshold",
							"-",
							"softness"
						],
						[
							1,
							"thicknessThreshold",
							"+",
							"softness"
						],
						"baseWave"
					]
				]
			],
			[10, [
				1,
				[5, "1"],
				"-",
				"banded"
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Ripples",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Concentric animated ripples emanating from a point",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point where ripples emanate from",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		colorA: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color of the ripple waves",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#000000",
			transform: transformColor,
			description: "Background color between ripples",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		speed: {
			default: 1,
			description: "Speed of ripple animation",
			ui: {
				type: "range",
				min: -5,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		frequency: {
			default: 20,
			description: "Number of ripples/spacing between them",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 80,
				step: .1,
				label: "Frequency",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Softness of ripple edges",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 3,
				step: .1,
				label: "Softness",
				group: "Effect"
			}
		},
		thickness: {
			default: .5,
			description: "Thickness of each ripple band",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Thickness",
				group: "Effect"
			}
		},
		phase: {
			default: 0,
			description: "Phase offset for ripple animation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 6.28,
				step: .1,
				label: "Phase",
				group: "Animation"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx } = params;
		const uv = params.uvContext ?? ctx.uv;
		const viewport = params.effectiveViewportSize ?? ctx.viewportSize;
		const t = animatedTime(params);
		const mask = call(ripplesMask, "ripplesMask", [
			uniforms.center,
			uniforms.frequency,
			uniforms.softness,
			uniforms.thickness,
			uniforms.phase,
			t,
			uv,
			viewport
		]);
		return call(mixColorsLinear, "mixColors", [
			uniforms.colorB,
			uniforms.colorA,
			mask
		]);
	}
};
var Ripples_default = componentDefinition;
export { componentDefinition as n, ripplesMask as r, Ripples_default as t };
