// shaders/dist/core/Saturation-Wct3aUol.js
import { At as call, Ot as ZERO, R as std, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
const saturationColor = tgpu.fn([d.vec4f, d.f32], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, intensity) => {
	"use gpu";
	const rgb = d.vec3f(color.x, color.y, color.z);
	const luminance = std.dot(rgb, d.vec3f(.2126, .7152, .0722));
	const gray = d.vec3f(luminance, luminance, luminance);
	const saturated = std.mix(gray, rgb, d.vec3f(intensity));
	return d.vec4f(saturated.x, saturated.y, saturated.z, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}, {
			"type": "i",
			"name": "intensity"
		}],
		"body": [0, [
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"color",
							"x"
						],
						[
							7,
							"color",
							"y"
						],
						[
							7,
							"color",
							"z"
						]
					]
				]
			],
			[
				13,
				"luminance",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["rgb", [
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0.2126"],
							[5, "0.7152"],
							[5, "0.0722"]
						]
					]]
				]
			],
			[
				13,
				"gray",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						"luminance",
						"luminance",
						"luminance"
					]
				]
			],
			[
				13,
				"saturated",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"gray",
						"rgb",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["intensity"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"saturated",
						"x"
					],
					[
						7,
						"saturated",
						"y"
					],
					[
						7,
						"saturated",
						"z"
					],
					[
						7,
						"color",
						"w"
					]
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Saturation",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Adjust color saturation intensity",
	requiresChild: true,
	props: { intensity: {
		default: 1,
		description: "The intensity of the saturation effect (1 being no change)",
		compileTimeWhen: (prev, next) => prev === 1 !== (next === 1),
		ui: {
			type: ["range", "map"],
			min: 0,
			max: 3,
			step: .1,
			label: "Intensity",
			group: "Adjustments"
		}
	} },
	fragment: ({ uniforms, childNode, propValues }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Saturation shader.");
			return ZERO;
		}
		if ((propValues.intensity ?? 1) === 1) return childNode;
		return call(saturationColor, "saturationColor", [childNode, uniforms.intensity]);
	}
};
var Saturation_default = componentDefinition;
export { componentDefinition as n, saturationColor as r, Saturation_default as t };
