// shaders/dist/core/Scratches-CpfAuRQw.js
import { At as call, R as std, W as toneAndColorExpr, n as tgpu, on as d, rt as noise_exports, st as scratches12 } from "./coords-BlwgBpJf.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
import { n as noiseColorPropsAB, r as noiseSpeedProp } from "./noiseColor-Cr76BYLN.js";
const scratchesField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, scale, seed, animTime, thickness) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	return scratches12(d.vec2f(__tsover_mul(uv.x, aspect), uv.y).mul(std.exp(scale)).add(seed), animTime, thickness);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "thickness"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"pos",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"aspectUV",
								"mul"
							],
							[[
								6,
								[
									7,
									"std",
									"exp"
								],
								["scale"]
							]]
						],
						"add"
					],
					["seed"]
				]
			],
			[10, [
				6,
				[
					7,
					"noise",
					"scratches12"
				],
				[
					"pos",
					"animTime",
					"thickness"
				]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"noise"
		]
	},
	externals: () => ({
		d,
		std,
		noise: noise_exports
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Scratches",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Fine hairline scratches, like a worn film or scratched surface",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		...noiseColorPropsAB(),
		scale: {
			default: 2,
			description: "Pattern scale (higher = more, finer scratches)",
			ui: {
				type: "range",
				min: -2,
				max: 5,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		thickness: {
			default: 1,
			description: "Thickness of the scratches (higher = bolder streaks)",
			ui: {
				type: "range",
				min: .2,
				max: 5,
				step: .01,
				label: "Thickness",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for pattern variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		...noiseSpeedProp()
	},
	fragment: (params) => {
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		return toneAndColorExpr(params, call(scratchesField, "scratchesField", [
			uv,
			viewport,
			params.uniforms.scale,
			params.uniforms.seed,
			t,
			params.uniforms.thickness
		]));
	}
};
var Scratches_default = componentDefinition;
export { componentDefinition as n, scratchesField as r, Scratches_default as t };
