// shaders/dist/core/shaderRegistry-DC0r_QlY.js
import { n as componentDefinition$94 } from "./Neon-Der_R-Bh.js";
import { n as componentDefinition$44 } from "./Emboss-D_-1_a3u.js";
import { n as componentDefinition$57 } from "./Form3D-BmCwmFPZ.js";
import { n as componentDefinition$143 } from "./ThinFilm-D_y48u9e.js";
import { a as componentDefinition } from "./AngularBlur-BJKQrsoZ.js";
import { r as componentDefinition$1 } from "./Arc-BZdfWWlz.js";
import { a as componentDefinition$2 } from "./Ascii-CR5iFmk8.js";
import { a as componentDefinition$3 } from "./Aurora-C0KIoJY5.js";
import { a as componentDefinition$4 } from "./BarShift-_mdK3UWW.js";
import { r as componentDefinition$5 } from "./Beam-DYvBEj6q.js";
import { a as componentDefinition$6 } from "./Blob-DgMJ5M6V.js";
import { r as componentDefinition$7 } from "./BlockNoise-CsBWCDzA.js";
import { r as componentDefinition$8 } from "./BlueNoise-D-WjLdgK.js";
import { i as componentDefinition$9 } from "./Blur-WeJoA31s.js";
import { r as componentDefinition$10 } from "./BrickPattern-DPez_hvr.js";
import { r as componentDefinition$11 } from "./BrightnessContrast-XvDV1ixv.js";
import { i as componentDefinition$12 } from "./BrushedMetal-DYwNjVpX.js";
import { r as componentDefinition$13 } from "./Bulge-DcBJaYyW.js";
import { r as componentDefinition$14 } from "./CarbonFiber-DwYVLmd0.js";
import { i as componentDefinition$15 } from "./Chalkboard-D_FVdz4V.js";
import { a as componentDefinition$16 } from "./ChannelBlur-B70l9vql.js";
import { a as componentDefinition$17 } from "./Checkerboard-B_z4be-A.js";
import { r as componentDefinition$18 } from "./Chevron-CfGjPdH6.js";
import { i as componentDefinition$19 } from "./ChromaFlow-B9zZUgzN.js";
import { r as componentDefinition$20 } from "./ChromaticAberration-BrGXd7kq.js";
import { r as componentDefinition$21 } from "./Circle-DFx6OVnv.js";
import { o as componentDefinition$22 } from "./ColorWheel-CI_qR6vL.js";
import { r as componentDefinition$23 } from "./CompressionArtifacts-CVWkWovO.js";
import { n as componentDefinition$24 } from "./ConcentricSpin-lAMdNRDl.js";
import { n as componentDefinition$25 } from "./ConicGradient-C3wD1h2z.js";
import { n as componentDefinition$26 } from "./ContourLines-C3Hwaw1Y.js";
import { n as componentDefinition$27 } from "./CornerPin-DxUoEjhR.js";
import { n as componentDefinition$28 } from "./Crescent-BCf_ZZI3.js";
import { n as componentDefinition$29 } from "./Cross-CeQhwjVB.js";
import { n as componentDefinition$30 } from "./CRTScreen-Byau9--w.js";
import { r as componentDefinition$31 } from "./Crystal-CC7d_6kZ.js";
import { n as componentDefinition$32 } from "./CurlNoise-KsNJEh6E.js";
import { n as componentDefinition$33 } from "./CursorRipples-Jgm7iF0Z.js";
import { n as componentDefinition$34 } from "./CursorTrail-CkNn85vn.js";
import { n as componentDefinition$35 } from "./DataMosh-Dh7TYpbW.js";
import { n as componentDefinition$36 } from "./DiamondGradient-HmZwnTdD.js";
import { n as componentDefinition$37 } from "./DiffuseBlur-uXE0w0DD.js";
import { n as componentDefinition$38 } from "./Dither-YUNtD1tI.js";
import { n as componentDefinition$39 } from "./DOMTexture-DFvbXYrj.js";
import { n as componentDefinition$40 } from "./DotGrid-DABCUp6d.js";
import { n as componentDefinition$41 } from "./DropShadow-C7KAuipC.js";
import { n as componentDefinition$42 } from "./Duotone-CuJunWJ0.js";
import { n as componentDefinition$43 } from "./Ellipse-DrZqqQh2.js";
import { n as componentDefinition$45 } from "./Engraving-fQ1zZJ89.js";
import { n as componentDefinition$46 } from "./ErosionNoise-B01HEvZk.js";
import { n as componentDefinition$47 } from "./Exposure-o_CViEwJ.js";
import { n as componentDefinition$48 } from "./FallingLines-Bv8X8gRy.js";
import { n as componentDefinition$49 } from "./FilmGrain-CoW4iHsT.js";
import { n as componentDefinition$50 } from "./Flip-h2K3wLGX.js";
import { n as componentDefinition$51 } from "./FloatingParticles-W_QAKOku.js";
import { n as componentDefinition$52 } from "./Flower-DxOinfag.js";
import { n as componentDefinition$53 } from "./FlowField-CwU9iWbf.js";
import { n as componentDefinition$54 } from "./FlowingGradient-DdFrloZC.js";
import { r as componentDefinition$55 } from "./FlutedGlass-B9652Bh4.js";
import { a as componentDefinition$56 } from "./Fog-D5sIF3cX.js";
import { n as componentDefinition$58 } from "./FractalNoise-BFs55FIe.js";
import { n as componentDefinition$59 } from "./Frost-C6qVP9a2.js";
import { n as componentDefinition$60 } from "./GaborNoise-C1LVTyG3.js";
import { n as componentDefinition$61 } from "./Glass-DR9JYQMG.js";
import { n as componentDefinition$62 } from "./GlassTiles-D5Lrddns.js";
import { r as componentDefinition$63 } from "./Glitch-B93JzM-d.js";
import { i as componentDefinition$64 } from "./Glow-DlxNehzC.js";
import { n as componentDefinition$65 } from "./Godrays-n9Gjl-h6.js";
import { n as componentDefinition$66 } from "./Goo-BusLBgE3.js";
import { n as componentDefinition$67 } from "./GradientMap-CapG6NfM.js";
import { n as componentDefinition$68 } from "./Grayscale-Dx9gKGZb.js";
import { n as componentDefinition$69 } from "./Grid-gZHTJmwP.js";
import { n as componentDefinition$70 } from "./GridDistortion-BkkuULR2.js";
import { n as componentDefinition$71 } from "./Group-pHN72R84.js";
import { n as componentDefinition$72 } from "./Halftone-dlX8bpPV.js";
import { n as componentDefinition$73 } from "./Heart-CwELiXMT.js";
import { n as componentDefinition$74 } from "./Heatmap-unI_z4Uc.js";
import { n as componentDefinition$75 } from "./HexGrid-BTWeNz9Z.js";
import { n as componentDefinition$76 } from "./Hologram-3-c4mCzw.js";
import { n as componentDefinition$77 } from "./Holographic-SchvA6Dy.js";
import { n as componentDefinition$78 } from "./HueShift-BJGcv19-.js";
import { n as componentDefinition$79 } from "./ImageTexture-CRueGzAZ.js";
import { n as componentDefinition$80 } from "./Invert-DLot9mli.js";
import { n as componentDefinition$81 } from "./IsometricCubes-CFg5QBjO.js";
import { n as componentDefinition$82 } from "./Kaleidoscope-BQFm59wM.js";
import { r as componentDefinition$83 } from "./KeyFrames-DiAKRsxm.js";
import { n as componentDefinition$84 } from "./LensFlare-BpFbnSsM.js";
import { r as componentDefinition$85 } from "./LightEdge-CZpMFMKO.js";
import { n as componentDefinition$86 } from "./LightLeak-DGt9quKs.js";
import { i as componentDefinition$87 } from "./LinearBlur-ClAsNnkC.js";
import { n as componentDefinition$88 } from "./LinearGradient-B_dBLhFm.js";
import { n as componentDefinition$89 } from "./LiquidMetal-Bb9b2h9X.js";
import { n as componentDefinition$90 } from "./Liquify-B9ypclFj.js";
import { n as componentDefinition$91 } from "./Marble-CZYhaDcz.js";
import { n as componentDefinition$92 } from "./Mirror--NvGPp-u.js";
import { n as componentDefinition$93 } from "./MultiPointGradient-IJtw3dEC.js";
import { n as componentDefinition$95 } from "./ObjectTracker-CfW_aMyp.js";
import { r as componentDefinition$96 } from "./PagePeel-D_DnKmec.js";
import { n as componentDefinition$97 } from "./Paper-CZT7vlqM.js";
import { n as componentDefinition$98 } from "./Parallelogram-BAi9iFi0.js";
import { n as componentDefinition$99 } from "./Particles-DtxN6LL7.js";
import { n as componentDefinition$100 } from "./PerlinNoise-BCbLpWKn.js";
import { n as componentDefinition$101 } from "./Perspective-DXIafIz8.js";
import { n as componentDefinition$102 } from "./Pixelate-L-pZ8G79.js";
import { i as componentDefinition$103 } from "./PixelSort-BnZgslHG.js";
import { i as componentDefinition$104 } from "./PixelThrow-BAVO2zdp.js";
import { n as componentDefinition$105 } from "./Plasma-BELcebZa.js";
import { n as componentDefinition$106 } from "./Plastic-e8KLjCgS.js";
import { n as componentDefinition$107 } from "./PolarCoordinates-CNZ6Df5J.js";
import { n as componentDefinition$108 } from "./Polygon-BGBTOGLe.js";
import { n as componentDefinition$109 } from "./Posterize-Cs_kyyjG.js";
import { n as componentDefinition$110 } from "./Prism-BUalIt7a.js";
import { a as componentDefinition$111 } from "./ProgressiveBlur-C2wYIOH-.js";
import { n as componentDefinition$112 } from "./RadialGradient-B3tlBsQZ.js";
import { n as componentDefinition$113 } from "./RectangularCoordinates-DWWlrqB4.js";
import { r as componentDefinition$114 } from "./ReflectivePlane-BmM5w76k.js";
import { n as componentDefinition$115 } from "./Repeater-V37OGlry.js";
import { n as componentDefinition$116 } from "./Ring-WVyMeyoh.js";
import { n as componentDefinition$117 } from "./Ripples-Bv__zvme.js";
import { n as componentDefinition$118 } from "./RoundedRect-B62UFEeb.js";
import { n as componentDefinition$119 } from "./Saturation-DIRGc5kp.js";
import { n as componentDefinition$120 } from "./Scratches-ZFLTq06k.js";
import { n as componentDefinition$121 } from "./Sharpness-WYaGJlSI.js";
import { i as componentDefinition$122 } from "./Shatter-CPS5Kdei.js";
import { n as componentDefinition$123 } from "./SimplexNoise-DMWtripU.js";
import { n as componentDefinition$124 } from "./SineWave-BOoYgny1.js";
import { i as componentDefinition$125 } from "./Smoke-BGGH2iTS.js";
import { i as componentDefinition$126 } from "./SmokeFill-ClBFBWz3.js";
import { i as componentDefinition$127 } from "./SmokeFlow-DkBGrN_d.js";
import { n as componentDefinition$128 } from "./Solarize-D_ZPY_Cr.js";
import { n as componentDefinition$129 } from "./SolidColor-rdMtlNyB.js";
import { n as componentDefinition$130 } from "./Spherize-h1F31u9-.js";
import { n as componentDefinition$131 } from "./Spiral-BMFhUpkP.js";
import { n as componentDefinition$132 } from "./Star-D8d3Gk74.js";
import { n as componentDefinition$133 } from "./Stone-CfpfDKpn.js";
import { n as componentDefinition$134 } from "./Strands-256prKRG.js";
import { n as componentDefinition$135 } from "./Stretch-C7z0wdI7.js";
import { n as componentDefinition$136 } from "./Stripes-BVvuh6r7.js";
import { n as componentDefinition$137 } from "./StudioBackground-Cy2qPLYJ.js";
import { n as componentDefinition$138 } from "./SunBurst-vUglW69s.js";
import { n as componentDefinition$139 } from "./Surface3D-quYcBdK_.js";
import { n as componentDefinition$140 } from "./Swirl-BD3Fk-SC.js";
import { n as componentDefinition$141 } from "./Teardrop-Bxfkjjaz.js";
import { r as componentDefinition$142 } from "./Text-CrbVixvj.js";
import { a as componentDefinition$144 } from "./TiltShift-3s5RKGCa.js";
import { n as componentDefinition$145 } from "./Tint-DZUIRc-Z.js";
import { n as componentDefinition$146 } from "./Trapezoid-7VOLjftp.js";
import { n as componentDefinition$147 } from "./TriangularGrid-Qek0gRuV.js";
import { n as componentDefinition$148 } from "./Tritone-CqYOMKFE.js";
import { n as componentDefinition$149 } from "./Truchet-D6emdFny.js";
import { n as componentDefinition$150 } from "./Twirl-Y8OzHc49.js";
import { n as componentDefinition$151 } from "./Vesica-DFRPMcgz.js";
import { n as componentDefinition$152 } from "./VHS-B28K9Bfc.js";
import { n as componentDefinition$153 } from "./Vibrance-BkMNClGf.js";
import { n as componentDefinition$154 } from "./VideoTexture-DNfGZ4kz.js";
import { n as componentDefinition$155 } from "./Vignette-DWveIoCk.js";
import { n as componentDefinition$156 } from "./Voronoi-c1kHPUka.js";
import { n as componentDefinition$157 } from "./Water-CEmTC2Ii.js";
import { n as componentDefinition$158 } from "./Watercolor--Bg1ZPIG.js";
import { n as componentDefinition$159 } from "./WaveDistortion-BSyW8c28.js";
import { n as componentDefinition$160 } from "./WaveletNoise-cDkFw55m.js";
import { n as componentDefinition$161 } from "./Weave--Ex2UVBt.js";
import { r as componentDefinition$162 } from "./WebcamTexture-CPWUtxVm.js";
import { n as componentDefinition$163 } from "./Wool-BNmVUISf.js";
import { n as componentDefinition$164 } from "./WorleyNoise-CbhvPZS-.js";
import { i as componentDefinition$165 } from "./ZoomBlur-BR_do-LV.js";
var createShaderRegistry = () => {
	const registry = {};
	const shaderDefinitions = {
		AngularBlur: componentDefinition,
		Arc: componentDefinition$1,
		Ascii: componentDefinition$2,
		Aurora: componentDefinition$3,
		BarShift: componentDefinition$4,
		Beam: componentDefinition$5,
		Blob: componentDefinition$6,
		BlockNoise: componentDefinition$7,
		BlueNoise: componentDefinition$8,
		Blur: componentDefinition$9,
		BrickPattern: componentDefinition$10,
		BrightnessContrast: componentDefinition$11,
		BrushedMetal: componentDefinition$12,
		Bulge: componentDefinition$13,
		CarbonFiber: componentDefinition$14,
		Chalkboard: componentDefinition$15,
		ChannelBlur: componentDefinition$16,
		Checkerboard: componentDefinition$17,
		Chevron: componentDefinition$18,
		ChromaFlow: componentDefinition$19,
		ChromaticAberration: componentDefinition$20,
		Circle: componentDefinition$21,
		ColorWheel: componentDefinition$22,
		CompressionArtifacts: componentDefinition$23,
		ConcentricSpin: componentDefinition$24,
		ConicGradient: componentDefinition$25,
		ContourLines: componentDefinition$26,
		CornerPin: componentDefinition$27,
		Crescent: componentDefinition$28,
		Cross: componentDefinition$29,
		CRTScreen: componentDefinition$30,
		Crystal: componentDefinition$31,
		CurlNoise: componentDefinition$32,
		CursorRipples: componentDefinition$33,
		CursorTrail: componentDefinition$34,
		DataMosh: componentDefinition$35,
		DiamondGradient: componentDefinition$36,
		DiffuseBlur: componentDefinition$37,
		Dither: componentDefinition$38,
		DOMTexture: componentDefinition$39,
		DotGrid: componentDefinition$40,
		DropShadow: componentDefinition$41,
		Duotone: componentDefinition$42,
		Ellipse: componentDefinition$43,
		Emboss: componentDefinition$44,
		Engraving: componentDefinition$45,
		ErosionNoise: componentDefinition$46,
		Exposure: componentDefinition$47,
		FallingLines: componentDefinition$48,
		FilmGrain: componentDefinition$49,
		Flip: componentDefinition$50,
		FloatingParticles: componentDefinition$51,
		Flower: componentDefinition$52,
		FlowField: componentDefinition$53,
		FlowingGradient: componentDefinition$54,
		FlutedGlass: componentDefinition$55,
		Fog: componentDefinition$56,
		Form3D: componentDefinition$57,
		FractalNoise: componentDefinition$58,
		Frost: componentDefinition$59,
		GaborNoise: componentDefinition$60,
		Glass: componentDefinition$61,
		GlassTiles: componentDefinition$62,
		Glitch: componentDefinition$63,
		Glow: componentDefinition$64,
		Godrays: componentDefinition$65,
		Goo: componentDefinition$66,
		GradientMap: componentDefinition$67,
		Grayscale: componentDefinition$68,
		Grid: componentDefinition$69,
		GridDistortion: componentDefinition$70,
		Group: componentDefinition$71,
		Halftone: componentDefinition$72,
		Heart: componentDefinition$73,
		Heatmap: componentDefinition$74,
		HexGrid: componentDefinition$75,
		Hologram: componentDefinition$76,
		Holographic: componentDefinition$77,
		HueShift: componentDefinition$78,
		ImageTexture: componentDefinition$79,
		Invert: componentDefinition$80,
		IsometricCubes: componentDefinition$81,
		Kaleidoscope: componentDefinition$82,
		KeyFrames: componentDefinition$83,
		LensFlare: componentDefinition$84,
		LightEdge: componentDefinition$85,
		LightLeak: componentDefinition$86,
		LinearBlur: componentDefinition$87,
		LinearGradient: componentDefinition$88,
		LiquidMetal: componentDefinition$89,
		Liquify: componentDefinition$90,
		Marble: componentDefinition$91,
		Mirror: componentDefinition$92,
		MultiPointGradient: componentDefinition$93,
		Neon: componentDefinition$94,
		ObjectTracker: componentDefinition$95,
		PagePeel: componentDefinition$96,
		Paper: componentDefinition$97,
		Parallelogram: componentDefinition$98,
		Particles: componentDefinition$99,
		PerlinNoise: componentDefinition$100,
		Perspective: componentDefinition$101,
		Pixelate: componentDefinition$102,
		PixelSort: componentDefinition$103,
		PixelThrow: componentDefinition$104,
		Plasma: componentDefinition$105,
		Plastic: componentDefinition$106,
		PolarCoordinates: componentDefinition$107,
		Polygon: componentDefinition$108,
		Posterize: componentDefinition$109,
		Prism: componentDefinition$110,
		ProgressiveBlur: componentDefinition$111,
		RadialGradient: componentDefinition$112,
		RectangularCoordinates: componentDefinition$113,
		ReflectivePlane: componentDefinition$114,
		Repeater: componentDefinition$115,
		Ring: componentDefinition$116,
		Ripples: componentDefinition$117,
		RoundedRect: componentDefinition$118,
		Saturation: componentDefinition$119,
		Scratches: componentDefinition$120,
		Sharpness: componentDefinition$121,
		Shatter: componentDefinition$122,
		SimplexNoise: componentDefinition$123,
		SineWave: componentDefinition$124,
		Smoke: componentDefinition$125,
		SmokeFill: componentDefinition$126,
		SmokeFlow: componentDefinition$127,
		Solarize: componentDefinition$128,
		SolidColor: componentDefinition$129,
		Spherize: componentDefinition$130,
		Spiral: componentDefinition$131,
		Star: componentDefinition$132,
		Stone: componentDefinition$133,
		Strands: componentDefinition$134,
		Stretch: componentDefinition$135,
		Stripes: componentDefinition$136,
		StudioBackground: componentDefinition$137,
		SunBurst: componentDefinition$138,
		Surface3D: componentDefinition$139,
		Swirl: componentDefinition$140,
		Teardrop: componentDefinition$141,
		Text: componentDefinition$142,
		ThinFilm: componentDefinition$143,
		TiltShift: componentDefinition$144,
		Tint: componentDefinition$145,
		Trapezoid: componentDefinition$146,
		TriangularGrid: componentDefinition$147,
		Tritone: componentDefinition$148,
		Truchet: componentDefinition$149,
		Twirl: componentDefinition$150,
		Vesica: componentDefinition$151,
		VHS: componentDefinition$152,
		Vibrance: componentDefinition$153,
		VideoTexture: componentDefinition$154,
		Vignette: componentDefinition$155,
		Voronoi: componentDefinition$156,
		Water: componentDefinition$157,
		Watercolor: componentDefinition$158,
		WaveDistortion: componentDefinition$159,
		WaveletNoise: componentDefinition$160,
		Weave: componentDefinition$161,
		WebcamTexture: componentDefinition$162,
		Wool: componentDefinition$163,
		WorleyNoise: componentDefinition$164,
		ZoomBlur: componentDefinition$165
	};
	Object.entries(shaderDefinitions).forEach(([fileName, def]) => {
		const propsMetadata = {};
		if (def.props) Object.entries(def.props).forEach(([propName, propConfig]) => {
			if (propConfig.ui) propsMetadata[propName] = {
				ui: propConfig.ui,
				default: propConfig.default,
				description: propConfig.description
			};
		});
		registry[def.name] = {
			name: def.name,
			fileName,
			category: def.category || "Uncategorized",
			definition: def,
			propsMetadata
		};
	});
	return registry;
};
const shaderRegistry = createShaderRegistry();
function getAllShaders() {
	return Object.values(shaderRegistry);
}
function getShaderByName(name) {
	return shaderRegistry[name];
}
function getShadersByCategory(category) {
	return Object.values(shaderRegistry).filter((shader) => shader.category === category);
}
function getShaderCategories() {
	const categories = new Set(Object.values(shaderRegistry).map((shader) => shader.category));
	return Array.from(categories).sort();
}
export { shaderRegistry as a, getShadersByCategory as i, getShaderByName as n, getShaderCategories as r, getAllShaders as t };
