// shaders/dist/core/shaderRegistry-BIiRyksy.js
import { n as componentDefinition$94 } from "./Neon-DFuPjkcF.js";
import { n as componentDefinition$44 } from "./Emboss-D-qOxM2u.js";
import { n as componentDefinition$57 } from "./Form3D-CrT8E1u5.js";
import { n as componentDefinition$143 } from "./ThinFilm-9l3KGv4Q.js";
import { a as componentDefinition } from "./AngularBlur-CFsh0PlD.js";
import { r as componentDefinition$1 } from "./Arc-CUnOmySr.js";
import { a as componentDefinition$2 } from "./Ascii-BLLIv-3N.js";
import { a as componentDefinition$3 } from "./Aurora-B_5FkTB2.js";
import { a as componentDefinition$4 } from "./BarShift-Ca-1XYfx.js";
import { r as componentDefinition$5 } from "./Beam-DRdKJKir.js";
import { a as componentDefinition$6 } from "./Blob-BRTcopJC.js";
import { r as componentDefinition$7 } from "./BlockNoise-CA933kae.js";
import { r as componentDefinition$8 } from "./BlueNoise-BeFfRr-o.js";
import { i as componentDefinition$9 } from "./Blur-l7HgDuUW.js";
import { r as componentDefinition$10 } from "./BrickPattern-BWK7-KwE.js";
import { r as componentDefinition$11 } from "./BrightnessContrast-Dh8qSRLx.js";
import { i as componentDefinition$12 } from "./BrushedMetal-CvIXubZz.js";
import { r as componentDefinition$13 } from "./Bulge-d93jKEwv.js";
import { r as componentDefinition$14 } from "./CarbonFiber-G7vIOz5W.js";
import { i as componentDefinition$15 } from "./Chalkboard-C9PtcnYN.js";
import { a as componentDefinition$16 } from "./ChannelBlur-B1bY63Fb.js";
import { a as componentDefinition$17 } from "./Checkerboard-DbjFO-fF.js";
import { r as componentDefinition$18 } from "./Chevron-D5-x4Wwd.js";
import { i as componentDefinition$19 } from "./ChromaFlow-CuLYMObY.js";
import { r as componentDefinition$20 } from "./ChromaticAberration-Dwvzvb8a.js";
import { r as componentDefinition$21 } from "./Circle-BLNjdX3R.js";
import { o as componentDefinition$22 } from "./ColorWheel-CA_34-9I.js";
import { r as componentDefinition$23 } from "./CompressionArtifacts-3lIojU64.js";
import { n as componentDefinition$24 } from "./ConcentricSpin-FJuBPPHi.js";
import { n as componentDefinition$25 } from "./ConicGradient-f9jCXFuK.js";
import { n as componentDefinition$26 } from "./ContourLines-Clnm2pQU.js";
import { n as componentDefinition$27 } from "./CornerPin-ecpjb9do.js";
import { n as componentDefinition$28 } from "./Crescent-CO5EZzD1.js";
import { n as componentDefinition$29 } from "./Cross-CofGHkIn.js";
import { n as componentDefinition$30 } from "./CRTScreen-CSBjpbn4.js";
import { r as componentDefinition$31 } from "./Crystal-C_aHlQ_3.js";
import { n as componentDefinition$32 } from "./CurlNoise-Br5XEf1K.js";
import { n as componentDefinition$33 } from "./CursorRipples-CLYoq-BF.js";
import { n as componentDefinition$34 } from "./CursorTrail-c-t96YgV.js";
import { n as componentDefinition$35 } from "./DataMosh-BSYAol1u.js";
import { n as componentDefinition$36 } from "./DiamondGradient-uyUPtPc3.js";
import { n as componentDefinition$37 } from "./DiffuseBlur-Cw9duoJ7.js";
import { n as componentDefinition$38 } from "./Dither-B3ulhQOh.js";
import { n as componentDefinition$39 } from "./DOMTexture-Zxx_WTGZ.js";
import { n as componentDefinition$40 } from "./DotGrid-ioUj-Lim.js";
import { n as componentDefinition$41 } from "./DropShadow-BRL7yRmy.js";
import { n as componentDefinition$42 } from "./Duotone-Df34HelN.js";
import { n as componentDefinition$43 } from "./Ellipse-BHxcxCpl.js";
import { n as componentDefinition$45 } from "./Engraving-B1ebFc1E.js";
import { n as componentDefinition$46 } from "./ErosionNoise-ZrJulbtK.js";
import { n as componentDefinition$47 } from "./Exposure-D8apJapP.js";
import { n as componentDefinition$48 } from "./FallingLines-DsMBKjEZ.js";
import { n as componentDefinition$49 } from "./FilmGrain-DeW-yw_v.js";
import { n as componentDefinition$50 } from "./Flip-Dw194RY6.js";
import { n as componentDefinition$51 } from "./FloatingParticles--Y81hnnp.js";
import { n as componentDefinition$52 } from "./Flower-Cdg8fOzZ.js";
import { n as componentDefinition$53 } from "./FlowField-CiK49M8B.js";
import { n as componentDefinition$54 } from "./FlowingGradient-B1Iywvbd.js";
import { r as componentDefinition$55 } from "./FlutedGlass-Crp432sd.js";
import { a as componentDefinition$56 } from "./Fog-ClrsfQfj.js";
import { n as componentDefinition$58 } from "./FractalNoise-DeDp7KQE.js";
import { n as componentDefinition$59 } from "./Frost-IvGmzNqf.js";
import { n as componentDefinition$60 } from "./GaborNoise-Bmhmuq_Z.js";
import { n as componentDefinition$61 } from "./Glass-CJFR7oIt.js";
import { n as componentDefinition$62 } from "./GlassTiles-BSBCjkhG.js";
import { r as componentDefinition$63 } from "./Glitch-Bp0LbC1u.js";
import { i as componentDefinition$64 } from "./Glow-CVBYSMxr.js";
import { n as componentDefinition$65 } from "./Godrays-C0VLmrcc.js";
import { n as componentDefinition$66 } from "./Goo-D-X3mP-n.js";
import { n as componentDefinition$67 } from "./GradientMap-uRYa7Yhv.js";
import { n as componentDefinition$68 } from "./Grayscale-DUGCGltR.js";
import { n as componentDefinition$69 } from "./Grid-FtTNW6zy.js";
import { n as componentDefinition$70 } from "./GridDistortion-lKZ4wpxx.js";
import { n as componentDefinition$71 } from "./Group-BNE6w-Ls.js";
import { n as componentDefinition$72 } from "./Halftone-BdmAqFEa.js";
import { n as componentDefinition$73 } from "./Heart-iGATzBCZ.js";
import { n as componentDefinition$74 } from "./Heatmap-dBfHUyG-.js";
import { n as componentDefinition$75 } from "./HexGrid-BUe3dyym.js";
import { n as componentDefinition$76 } from "./Hologram-BQ1kz8M2.js";
import { n as componentDefinition$77 } from "./Holographic-B39ULAli.js";
import { n as componentDefinition$78 } from "./HueShift-DeP_0kOt.js";
import { n as componentDefinition$79 } from "./ImageTexture-Bzh0A17b.js";
import { n as componentDefinition$80 } from "./Invert-CqrJ-5l4.js";
import { n as componentDefinition$81 } from "./IsometricCubes-BVYfIhbb.js";
import { n as componentDefinition$82 } from "./Kaleidoscope-VxQtO7zA.js";
import { r as componentDefinition$83 } from "./KeyFrames-WFOyBjoF.js";
import { n as componentDefinition$84 } from "./LensFlare-2M6sCguV.js";
import { r as componentDefinition$85 } from "./LightEdge-CvV9PYbe.js";
import { n as componentDefinition$86 } from "./LightLeak-BQxRr11E.js";
import { i as componentDefinition$87 } from "./LinearBlur-CyNRaqj8.js";
import { n as componentDefinition$88 } from "./LinearGradient-RDNQes5r.js";
import { n as componentDefinition$89 } from "./LiquidMetal-PGUaQGV-.js";
import { n as componentDefinition$90 } from "./Liquify-CmkepjAk.js";
import { n as componentDefinition$91 } from "./Marble-D_a2ctfR.js";
import { n as componentDefinition$92 } from "./Mirror-B98LBGCI.js";
import { n as componentDefinition$93 } from "./MultiPointGradient-DBWhUtj8.js";
import { n as componentDefinition$95 } from "./ObjectTracker-Bolz91qz.js";
import { r as componentDefinition$96 } from "./PagePeel-CTy_-lpL.js";
import { n as componentDefinition$97 } from "./Paper-BE3SU_ac.js";
import { n as componentDefinition$98 } from "./Parallelogram-C9qwJkYa.js";
import { n as componentDefinition$99 } from "./Particles-x_HCygMr.js";
import { n as componentDefinition$100 } from "./PerlinNoise-Di7dE-hc.js";
import { n as componentDefinition$101 } from "./Perspective-CoYBx1wH.js";
import { n as componentDefinition$102 } from "./Pixelate-Dg41Ap0g.js";
import { i as componentDefinition$103 } from "./PixelSort-DMWTXIcO.js";
import { r as componentDefinition$104 } from "./PixelThrow-CiZj7iGN.js";
import { n as componentDefinition$105 } from "./Plasma-BpNK-Jnv.js";
import { n as componentDefinition$106 } from "./Plastic-CpmtjSSV.js";
import { n as componentDefinition$107 } from "./PolarCoordinates-CBh8MX8L.js";
import { n as componentDefinition$108 } from "./Polygon-KpW71r_g.js";
import { n as componentDefinition$109 } from "./Posterize-DPwFDwMx.js";
import { n as componentDefinition$110 } from "./Prism-BZcRvoMJ.js";
import { a as componentDefinition$111 } from "./ProgressiveBlur-BAYvLhUa.js";
import { n as componentDefinition$112 } from "./RadialGradient-DxIBw-EZ.js";
import { n as componentDefinition$113 } from "./RectangularCoordinates-DJGtXzgn.js";
import { r as componentDefinition$114 } from "./ReflectivePlane-C4JxzSK_.js";
import { n as componentDefinition$115 } from "./Repeater-1TW-fdff.js";
import { n as componentDefinition$116 } from "./Ring-W9djENkw.js";
import { n as componentDefinition$117 } from "./Ripples-Dbvh__ZY.js";
import { n as componentDefinition$118 } from "./RoundedRect-DXjLPZee.js";
import { n as componentDefinition$119 } from "./Saturation-Wct3aUol.js";
import { n as componentDefinition$120 } from "./Scratches-CpfAuRQw.js";
import { n as componentDefinition$121 } from "./Sharpness-BgWKE-3E.js";
import { i as componentDefinition$122 } from "./Shatter-DiBeVj4M.js";
import { n as componentDefinition$123 } from "./SimplexNoise-Bgiw5XJP.js";
import { n as componentDefinition$124 } from "./SineWave-CHygTRV-.js";
import { i as componentDefinition$125 } from "./Smoke-CbAIiwKG.js";
import { i as componentDefinition$126 } from "./SmokeFill-BnJoSbE3.js";
import { i as componentDefinition$127 } from "./SmokeFlow-H01ajJ5Q.js";
import { n as componentDefinition$128 } from "./Solarize-CUpThxJh.js";
import { n as componentDefinition$129 } from "./SolidColor-rdMtlNyB.js";
import { n as componentDefinition$130 } from "./Spherize-bBe44dU7.js";
import { n as componentDefinition$131 } from "./Spiral-sBR7Fn0z.js";
import { n as componentDefinition$132 } from "./Star-DQUfQjbf.js";
import { n as componentDefinition$133 } from "./Stone-DnY9YuMc.js";
import { n as componentDefinition$134 } from "./Strands-e7O0GIHI.js";
import { n as componentDefinition$135 } from "./Stretch-CXFrCyH_.js";
import { n as componentDefinition$136 } from "./Stripes-BS5Y7yeT.js";
import { n as componentDefinition$137 } from "./StudioBackground-D7eIlsQA.js";
import { n as componentDefinition$138 } from "./SunBurst-CuawxiUI.js";
import { n as componentDefinition$139 } from "./Surface3D-ZtGSv9V6.js";
import { n as componentDefinition$140 } from "./Swirl-CxRi1nI8.js";
import { n as componentDefinition$141 } from "./Teardrop-D_ePwja5.js";
import { r as componentDefinition$142 } from "./Text-C1Wjl8fJ.js";
import { a as componentDefinition$144 } from "./TiltShift-CBCuQwu9.js";
import { n as componentDefinition$145 } from "./Tint-HLecBeer.js";
import { n as componentDefinition$146 } from "./Trapezoid-CztpQJcb.js";
import { n as componentDefinition$147 } from "./TriangularGrid-BWzP25ws.js";
import { n as componentDefinition$148 } from "./Tritone-DczuOC88.js";
import { n as componentDefinition$149 } from "./Truchet-Bj0dM4pc.js";
import { n as componentDefinition$150 } from "./Twirl-CgT4-JNU.js";
import { n as componentDefinition$151 } from "./Vesica-hhR3uVm2.js";
import { n as componentDefinition$152 } from "./VHS-BO2ImGGx.js";
import { n as componentDefinition$153 } from "./Vibrance-Bxbm2XXk.js";
import { n as componentDefinition$154 } from "./VideoTexture-Zj56qULM.js";
import { n as componentDefinition$155 } from "./Vignette-DhIYUi7B.js";
import { n as componentDefinition$156 } from "./Voronoi-I5jNa5lH.js";
import { n as componentDefinition$157 } from "./Water-DcoqDm5J.js";
import { n as componentDefinition$158 } from "./Watercolor-DhrRCOJ_.js";
import { n as componentDefinition$159 } from "./WaveDistortion-DJ1itI6L.js";
import { n as componentDefinition$160 } from "./WaveletNoise-Cy1pQ99p.js";
import { n as componentDefinition$161 } from "./Weave-CPIeOoWv.js";
import { r as componentDefinition$162 } from "./WebcamTexture-DNCb4NqI.js";
import { n as componentDefinition$163 } from "./Wool-aEiSh89V.js";
import { n as componentDefinition$164 } from "./WorleyNoise-C9l_C4G0.js";
import { i as componentDefinition$165 } from "./ZoomBlur-CpM-IOSu.js";
var createShaderRegistry = () => {
	const registry = {};
	const shaderDefinitions = {
		AngularBlur: componentDefinition,
		Arc: componentDefinition$1,
		Ascii: componentDefinition$2,
		Aurora: componentDefinition$3,
		BarShift: componentDefinition$4,
		Beam: componentDefinition$5,
		Blob: componentDefinition$6,
		BlockNoise: componentDefinition$7,
		BlueNoise: componentDefinition$8,
		Blur: componentDefinition$9,
		BrickPattern: componentDefinition$10,
		BrightnessContrast: componentDefinition$11,
		BrushedMetal: componentDefinition$12,
		Bulge: componentDefinition$13,
		CarbonFiber: componentDefinition$14,
		Chalkboard: componentDefinition$15,
		ChannelBlur: componentDefinition$16,
		Checkerboard: componentDefinition$17,
		Chevron: componentDefinition$18,
		ChromaFlow: componentDefinition$19,
		ChromaticAberration: componentDefinition$20,
		Circle: componentDefinition$21,
		ColorWheel: componentDefinition$22,
		CompressionArtifacts: componentDefinition$23,
		ConcentricSpin: componentDefinition$24,
		ConicGradient: componentDefinition$25,
		ContourLines: componentDefinition$26,
		CornerPin: componentDefinition$27,
		Crescent: componentDefinition$28,
		Cross: componentDefinition$29,
		CRTScreen: componentDefinition$30,
		Crystal: componentDefinition$31,
		CurlNoise: componentDefinition$32,
		CursorRipples: componentDefinition$33,
		CursorTrail: componentDefinition$34,
		DataMosh: componentDefinition$35,
		DiamondGradient: componentDefinition$36,
		DiffuseBlur: componentDefinition$37,
		Dither: componentDefinition$38,
		DOMTexture: componentDefinition$39,
		DotGrid: componentDefinition$40,
		DropShadow: componentDefinition$41,
		Duotone: componentDefinition$42,
		Ellipse: componentDefinition$43,
		Emboss: componentDefinition$44,
		Engraving: componentDefinition$45,
		ErosionNoise: componentDefinition$46,
		Exposure: componentDefinition$47,
		FallingLines: componentDefinition$48,
		FilmGrain: componentDefinition$49,
		Flip: componentDefinition$50,
		FloatingParticles: componentDefinition$51,
		Flower: componentDefinition$52,
		FlowField: componentDefinition$53,
		FlowingGradient: componentDefinition$54,
		FlutedGlass: componentDefinition$55,
		Fog: componentDefinition$56,
		Form3D: componentDefinition$57,
		FractalNoise: componentDefinition$58,
		Frost: componentDefinition$59,
		GaborNoise: componentDefinition$60,
		Glass: componentDefinition$61,
		GlassTiles: componentDefinition$62,
		Glitch: componentDefinition$63,
		Glow: componentDefinition$64,
		Godrays: componentDefinition$65,
		Goo: componentDefinition$66,
		GradientMap: componentDefinition$67,
		Grayscale: componentDefinition$68,
		Grid: componentDefinition$69,
		GridDistortion: componentDefinition$70,
		Group: componentDefinition$71,
		Halftone: componentDefinition$72,
		Heart: componentDefinition$73,
		Heatmap: componentDefinition$74,
		HexGrid: componentDefinition$75,
		Hologram: componentDefinition$76,
		Holographic: componentDefinition$77,
		HueShift: componentDefinition$78,
		ImageTexture: componentDefinition$79,
		Invert: componentDefinition$80,
		IsometricCubes: componentDefinition$81,
		Kaleidoscope: componentDefinition$82,
		KeyFrames: componentDefinition$83,
		LensFlare: componentDefinition$84,
		LightEdge: componentDefinition$85,
		LightLeak: componentDefinition$86,
		LinearBlur: componentDefinition$87,
		LinearGradient: componentDefinition$88,
		LiquidMetal: componentDefinition$89,
		Liquify: componentDefinition$90,
		Marble: componentDefinition$91,
		Mirror: componentDefinition$92,
		MultiPointGradient: componentDefinition$93,
		Neon: componentDefinition$94,
		ObjectTracker: componentDefinition$95,
		PagePeel: componentDefinition$96,
		Paper: componentDefinition$97,
		Parallelogram: componentDefinition$98,
		Particles: componentDefinition$99,
		PerlinNoise: componentDefinition$100,
		Perspective: componentDefinition$101,
		Pixelate: componentDefinition$102,
		PixelSort: componentDefinition$103,
		PixelThrow: componentDefinition$104,
		Plasma: componentDefinition$105,
		Plastic: componentDefinition$106,
		PolarCoordinates: componentDefinition$107,
		Polygon: componentDefinition$108,
		Posterize: componentDefinition$109,
		Prism: componentDefinition$110,
		ProgressiveBlur: componentDefinition$111,
		RadialGradient: componentDefinition$112,
		RectangularCoordinates: componentDefinition$113,
		ReflectivePlane: componentDefinition$114,
		Repeater: componentDefinition$115,
		Ring: componentDefinition$116,
		Ripples: componentDefinition$117,
		RoundedRect: componentDefinition$118,
		Saturation: componentDefinition$119,
		Scratches: componentDefinition$120,
		Sharpness: componentDefinition$121,
		Shatter: componentDefinition$122,
		SimplexNoise: componentDefinition$123,
		SineWave: componentDefinition$124,
		Smoke: componentDefinition$125,
		SmokeFill: componentDefinition$126,
		SmokeFlow: componentDefinition$127,
		Solarize: componentDefinition$128,
		SolidColor: componentDefinition$129,
		Spherize: componentDefinition$130,
		Spiral: componentDefinition$131,
		Star: componentDefinition$132,
		Stone: componentDefinition$133,
		Strands: componentDefinition$134,
		Stretch: componentDefinition$135,
		Stripes: componentDefinition$136,
		StudioBackground: componentDefinition$137,
		SunBurst: componentDefinition$138,
		Surface3D: componentDefinition$139,
		Swirl: componentDefinition$140,
		Teardrop: componentDefinition$141,
		Text: componentDefinition$142,
		ThinFilm: componentDefinition$143,
		TiltShift: componentDefinition$144,
		Tint: componentDefinition$145,
		Trapezoid: componentDefinition$146,
		TriangularGrid: componentDefinition$147,
		Tritone: componentDefinition$148,
		Truchet: componentDefinition$149,
		Twirl: componentDefinition$150,
		Vesica: componentDefinition$151,
		VHS: componentDefinition$152,
		Vibrance: componentDefinition$153,
		VideoTexture: componentDefinition$154,
		Vignette: componentDefinition$155,
		Voronoi: componentDefinition$156,
		Water: componentDefinition$157,
		Watercolor: componentDefinition$158,
		WaveDistortion: componentDefinition$159,
		WaveletNoise: componentDefinition$160,
		Weave: componentDefinition$161,
		WebcamTexture: componentDefinition$162,
		Wool: componentDefinition$163,
		WorleyNoise: componentDefinition$164,
		ZoomBlur: componentDefinition$165
	};
	Object.entries(shaderDefinitions).forEach(([fileName, def]) => {
		const propsMetadata = {};
		if (def.props) Object.entries(def.props).forEach(([propName, propConfig]) => {
			if (propConfig.ui) propsMetadata[propName] = {
				ui: propConfig.ui,
				default: propConfig.default,
				description: propConfig.description
			};
		});
		registry[def.name] = {
			name: def.name,
			fileName,
			category: def.category || "Uncategorized",
			definition: def,
			propsMetadata
		};
	});
	return registry;
};
const shaderRegistry = createShaderRegistry();
function getAllShaders() {
	return Object.values(shaderRegistry);
}
function getShaderByName(name) {
	return shaderRegistry[name];
}
function getShadersByCategory(category) {
	return Object.values(shaderRegistry).filter((shader) => shader.category === category);
}
function getShaderCategories() {
	const categories = new Set(Object.values(shaderRegistry).map((shader) => shader.category));
	return Array.from(categories).sort();
}
export { shaderRegistry as a, getShadersByCategory as i, getShaderByName as n, getShaderCategories as r, getAllShaders as t };
