// shaders/dist/core/shapeEffectBounds-BlxxsYMN.js
var SVG_SDF_SIZE = 512;
var cache = {};
var pending = /* @__PURE__ */ new Set();
var waiters = {};
function getSdfContentBounds(url, onReady) {
	if (!url) return null;
	if (url in cache) return cache[url];
	if (onReady) (waiters[url] ??= []).push(onReady);
	if (!pending.has(url)) {
		pending.add(url);
		scan(url);
	}
	return null;
}
async function scan(url) {
	try {
		const res = await fetch(url, { mode: "cors" });
		if (!res.ok) {
			cache[url] = null;
			return;
		}
		const buf = await res.arrayBuffer();
		const pixelCount = SVG_SDF_SIZE * SVG_SDF_SIZE;
		if (buf.byteLength !== pixelCount * 2 && buf.byteLength !== pixelCount * 4) {
			cache[url] = null;
			return;
		}
		const valueAt = buf.byteLength === pixelCount * 2 ? (() => {
			const u = new Uint16Array(buf);
			return (i) => u[i] / 32767.5 - 1;
		})() : (() => {
			const f = new Float32Array(buf);
			return (i) => f[i];
		})();
		let minX = SVG_SDF_SIZE, minY = SVG_SDF_SIZE, maxX = -1, maxY = -1;
		for (let y = 0; y < SVG_SDF_SIZE; y++) {
			const row = y * SVG_SDF_SIZE;
			for (let x = 0; x < SVG_SDF_SIZE; x++) if (valueAt(row + x) < 0) {
				if (x < minX) minX = x;
				if (x > maxX) maxX = x;
				if (y < minY) minY = y;
				if (y > maxY) maxY = y;
			}
		}
		cache[url] = maxX < 0 ? null : {
			hwf: (maxX - minX + 1) / SVG_SDF_SIZE / 2,
			hhf: (maxY - minY + 1) / SVG_SDF_SIZE / 2
		};
	} catch {
		cache[url] = null;
	} finally {
		pending.delete(url);
		const fns = waiters[url];
		if (fns) {
			delete waiters[url];
			for (const fn of fns) fn();
		}
	}
}
const SHAPE3D_TYPES = [
	"sphere3D",
	"cube3D",
	"torus3D",
	"octahedron3D",
	"cylinder3D",
	"capsule3D",
	"cone3D",
	"pyramid3D",
	"prism3D",
	"ellipsoid3D",
	"diamond3D",
	"link3D",
	"gem3D",
	"helix3D",
	"metaballs3D",
	"dodecahedron3D",
	"hemisphere3D",
	"ribbon3D",
	"blob3D",
	"gyroscope3D"
];
function is3dShapeType(shapeType) {
	return SHAPE3D_TYPES.includes(shapeType);
}
const SVG3D_TYPES = ["svgExtrude3D"];
function isSvg3dShapeType(shapeType) {
	return shapeType === "svgExtrude3D";
}
function isVolumetricShapeType(shapeType) {
	return is3dShapeType(shapeType) || isSvg3dShapeType(shapeType);
}
const SVG3D_DEFAULTS = {
	depth: .12,
	bevel: .02,
	rotX: 15,
	rotY: 30,
	rotZ: 0
};
function shapeSubPropMax(value, fallback) {
	if (typeof value === "number") return value;
	if (value && typeof value === "object" && (value.type === "auto-animate" || value.type === "mouse")) {
		const oMin = typeof value.outputMin === "number" ? value.outputMin : fallback;
		const oMax = typeof value.outputMax === "number" ? value.outputMax : fallback;
		return Math.max(oMin, oMax);
	}
	return fallback;
}
const SHAPE3D_DEFAULTS = {
	sphere3D: {
		radius: .35,
		rotX: 0,
		rotY: 0,
		rotZ: 0
	},
	cube3D: {
		sizeX: .27,
		sizeY: .27,
		sizeZ: .27,
		rounding: .02,
		rotX: 25,
		rotY: 35,
		rotZ: 0
	},
	torus3D: {
		radius: .3,
		tube: .12,
		rotX: 55,
		rotY: 0,
		rotZ: 0
	},
	octahedron3D: {
		radius: .42,
		rotX: 10,
		rotY: 25,
		rotZ: 0
	},
	cylinder3D: {
		radius: .22,
		height: .26,
		rounding: .02,
		rotX: 60,
		rotY: 0,
		rotZ: 30
	},
	capsule3D: {
		radius: .16,
		height: .2,
		rotX: 0,
		rotY: 0,
		rotZ: 35
	},
	cone3D: {
		radius: .3,
		topRadius: .02,
		height: .3,
		rotX: 25,
		rotY: 0,
		rotZ: 15
	},
	pyramid3D: {
		size: .3,
		height: .45,
		rotX: -10,
		rotY: 30,
		rotZ: 0
	},
	prism3D: {
		radius: .28,
		height: .14,
		rotX: 25,
		rotY: 30,
		rotZ: 0
	},
	ellipsoid3D: {
		radiusX: .38,
		radiusY: .22,
		radiusZ: .3,
		rotX: 0,
		rotY: 0,
		rotZ: 20
	},
	diamond3D: {
		radius: .3,
		height: .42,
		rotX: 15,
		rotY: 0,
		rotZ: 0
	},
	link3D: {
		radius: .18,
		length: .15,
		tube: .08,
		rotX: 20,
		rotY: 30,
		rotZ: 0
	},
	gem3D: {
		radius: .3,
		height: .32,
		facets: 8,
		rotX: 15,
		rotY: 0,
		rotZ: 0
	},
	helix3D: {
		radius: .22,
		tube: .06,
		pitch: .16,
		rotX: 20,
		rotY: 0,
		rotZ: 0
	},
	metaballs3D: {
		ballRadius: .16,
		spread: .2,
		blend: .12,
		speed: 1,
		rotX: 0,
		rotY: 0,
		rotZ: 0
	},
	dodecahedron3D: {
		radius: .34,
		rotX: 20,
		rotY: 10,
		rotZ: 0
	},
	hemisphere3D: {
		radius: .4,
		cut: 0,
		rotX: 30,
		rotY: 0,
		rotZ: 0
	},
	ribbon3D: {
		width: .15,
		thickness: .03,
		length: .38,
		wave: .1,
		waveFrequency: 7,
		twist: 140,
		speed: 1,
		rotX: 12,
		rotY: 18,
		rotZ: 0
	},
	blob3D: {
		radius: .3,
		wobble: .07,
		wobbliness: 5,
		speed: 1,
		rotX: 0,
		rotY: 0,
		rotZ: 0
	},
	gyroscope3D: {
		radius: .34,
		width: .1,
		thickness: .028,
		rounding: .006,
		core: .14,
		speed: 1,
		rotX: 15,
		rotY: 0,
		rotZ: 0
	}
};
function shape3dBoundingRadius(cfg) {
	const d = SHAPE3D_DEFAULTS[cfg.type] ?? {};
	const n = (k, f = .3) => shapeSubPropMax(cfg[k], d[k] ?? f);
	switch (cfg.type) {
		case "cube3D": return Math.hypot(n("sizeX"), n("sizeY"), n("sizeZ")) + n("rounding", 0);
		case "torus3D": return n("radius") + n("tube");
		case "cylinder3D": return Math.hypot(n("radius"), n("height")) + n("rounding", 0);
		case "capsule3D": return n("height") + n("radius");
		case "cone3D": return Math.hypot(Math.max(n("radius"), n("topRadius")), n("height"));
		case "pyramid3D": return Math.hypot(n("size") * Math.SQRT2, n("height") * .5);
		case "prism3D": return Math.hypot(n("radius") * 1.1548, n("height"));
		case "ellipsoid3D": return Math.max(n("radiusX"), n("radiusY"), n("radiusZ"));
		case "diamond3D": return Math.max(n("radius"), n("height"));
		case "link3D": return n("length") + n("radius") + n("tube");
		case "gem3D": return Math.max(n("radius"), n("height"));
		case "helix3D": return Math.hypot(n("radius") + n("tube"), n("pitch") * 3 * .5 + n("tube"));
		case "metaballs3D": return n("spread") + n("ballRadius", .16) + n("blend", .12);
		case "dodecahedron3D": return n("radius") * 1.32;
		case "hemisphere3D": return n("radius");
		case "ribbon3D": return Math.hypot(n("length", .38), n("wave", .1) * 1.2 + Math.hypot(n("width", .15), n("thickness", .03))) + .01;
		case "blob3D": return n("radius", .3) + n("wobble", .07) + .02;
		case "gyroscope3D": return n("radius", .34) + n("thickness", .028) * .5 + n("width", .1) * .5 + n("rounding", .006);
		case "sphere3D":
		case "octahedron3D":
		default: return n("radius", .35);
	}
}
const shapeEffectBoundingBoxDeclaration = {
	propBindings: {
		x: {
			prop: "center",
			as: "position-x"
		},
		y: {
			prop: "center",
			as: "position-y"
		},
		rotation: {
			prop: "rotation",
			as: "degrees"
		}
	},
	computeBounds(props, cw, ch) {
		const scale = props.scale ?? 1;
		const cx = props.center?.x ?? .5;
		const cy = props.center?.y ?? .5;
		const { hwf, hhf } = shapeContentExtent(props);
		const fieldPx = scale * ch;
		return {
			centerXPx: cx * cw,
			centerYPx: cy * ch,
			widthPx: 2 * hwf * fieldPx,
			heightPx: 2 * hhf * fieldPx,
			rotationDeg: props.rotation ?? 0
		};
	},
	writeBounds(bounds, currentProps, cw, ch) {
		const center = {
			...currentProps.center ?? {
				x: .5,
				y: .5
			},
			x: bounds.centerXPx / cw,
			y: bounds.centerYPx / ch
		};
		const scale = currentProps.scale ?? 1;
		const cfg = isNonUniformShapeEffect(currentProps) ? parseShapeConfig(currentProps.shape) : null;
		if (cfg && scale > 0) return {
			center,
			rotation: bounds.rotationDeg,
			shape: JSON.stringify({
				...cfg,
				radius: Math.max(.001, bounds.widthPx / (2 * scale * ch)),
				height: Math.max(.001, bounds.heightPx / (2 * scale * ch))
			})
		};
		const { hwf } = shapeContentExtent(currentProps);
		return {
			center,
			scale: hwf > 0 ? Math.max(.01, bounds.widthPx / (2 * hwf * ch)) : .01,
			rotation: bounds.rotationDeg
		};
	}
};
function parseShapeConfig(shape) {
	try {
		return typeof shape === "string" ? JSON.parse(shape) : shape;
	} catch {
		return null;
	}
}
function isNonUniformShapeEffect(props) {
	if (props.shapeSdfUrl) return false;
	const t = parseShapeConfig(props.shape)?.type;
	return t === "roundedRectSDF" || t === "ellipseSDF";
}
function shapeContentExtent(props) {
	if (props.shapeSdfUrl) {
		const sb = props._shapeBounds ?? getSdfContentBounds(props.shapeSdfUrl);
		const base = sb && typeof sb.hwf === "number" && typeof sb.hhf === "number" ? {
			hwf: sb.hwf,
			hhf: sb.hhf
		} : {
			hwf: .5,
			hhf: .5
		};
		const t = props.shapeType;
		if (typeof t === "string" && isSvg3dShapeType(t)) {
			const cfg$1 = parseShapeConfig(props.shape) ?? {};
			const sub = (k) => shapeSubPropMax(cfg$1[k], SVG3D_DEFAULTS[k] ?? 0);
			const br = Math.hypot(base.hwf, base.hhf, sub("depth") * .5) + sub("bevel") + .02;
			return {
				hwf: br,
				hhf: br
			};
		}
		return base;
	}
	let cfg = null;
	try {
		cfg = typeof props.shape === "string" ? JSON.parse(props.shape) : props.shape;
	} catch {}
	if (!cfg || typeof cfg !== "object") return {
		hwf: .5,
		hhf: .5
	};
	if (typeof cfg.type === "string" && is3dShapeType(cfg.type)) {
		const br = shape3dBoundingRadius(cfg);
		return {
			hwf: br,
			hhf: br
		};
	}
	const r = typeof cfg.radius === "number" ? cfg.radius : .35;
	switch (cfg.type) {
		case "roundedRectSDF":
		case "ellipseSDF": return {
			hwf: r,
			hhf: typeof cfg.height === "number" ? cfg.height : r
		};
		case "ringSDF": {
			const t = typeof cfg.thickness === "number" ? cfg.thickness : 0;
			return {
				hwf: r + t,
				hhf: r + t
			};
		}
		case "trapezoidSDF": {
			const bw = typeof cfg.bottomWidth === "number" ? cfg.bottomWidth : .3;
			const tw = typeof cfg.topWidth === "number" ? cfg.topWidth : .2;
			const h = typeof cfg.height === "number" ? cfg.height : .25;
			return {
				hwf: Math.max(bw, tw),
				hhf: h
			};
		}
		case "vesicaSDF": {
			const spread = typeof cfg.spread === "number" ? Math.min(1, Math.max(0, cfg.spread)) : .4;
			return {
				hwf: r * (1 - spread),
				hhf: r * Math.sqrt(Math.max(0, 1 - spread * spread))
			};
		}
		case "teardropSDF": return {
			hwf: r,
			hhf: ((typeof cfg.height === "number" ? cfg.height : .4) + r) / 2
		};
		case "parallelogramSDF": {
			const w = typeof cfg.width === "number" ? cfg.width : .32;
			const h = typeof cfg.height === "number" ? cfg.height : .22;
			const sk = typeof cfg.skew === "number" ? cfg.skew : 0;
			return {
				hwf: w + Math.abs(sk),
				hhf: h
			};
		}
		default: return {
			hwf: r,
			hhf: r
		};
	}
}
export { SVG3D_TYPES as a, isVolumetricShapeType as c, SVG3D_DEFAULTS as i, shape3dBoundingRadius as l, SHAPE3D_DEFAULTS as n, is3dShapeType as o, SHAPE3D_TYPES as r, isSvg3dShapeType as s, shapeEffectBoundingBoxDeclaration as t };
