// shaders/dist/core/Sharpness-BgWKE-3E.js
import { At as call, Ot as ZERO, Pt as floatE, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
const sharpnessTapUV = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, dirX, dirY) => {
	"use gpu";
	return d.vec2f(__tsover_add(uv.x, __tsover_div(dirX, viewport.x)), __tsover_add(uv.y, __tsover_div(dirY, viewport.y)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "dirX"
			},
			{
				"type": "i",
				"name": "dirY"
			}
		],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					7,
					"uv",
					"x"
				],
				"+",
				[
					1,
					"dirX",
					"/",
					[
						7,
						"viewport",
						"x"
					]
				]
			], [
				1,
				[
					7,
					"uv",
					"y"
				],
				"+",
				[
					1,
					"dirY",
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			]]
		]]]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const sharpnessCompose = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, top, bottom, left, right, amount) => {
	"use gpu";
	const centerWeight = __tsover_add(1, __tsover_mul(amount, 4));
	const neighborWeight = __tsover_mul(amount, -1);
	const rgb = center.xyz.mul(centerWeight).add(top.xyz.mul(neighborWeight)).add(bottom.xyz.mul(neighborWeight)).add(left.xyz.mul(neighborWeight)).add(right.xyz.mul(neighborWeight));
	const clamped = std.clamp(rgb, d.vec3f(0, 0, 0), d.vec3f(1, 1, 1));
	return d.vec4f(clamped.x, clamped.y, clamped.z, center.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "top"
			},
			{
				"type": "i",
				"name": "bottom"
			},
			{
				"type": "i",
				"name": "left"
			},
			{
				"type": "i",
				"name": "right"
			},
			{
				"type": "i",
				"name": "amount"
			}
		],
		"body": [0, [
			[
				13,
				"centerWeight",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						"amount",
						"*",
						[5, "4"]
					]
				]
			],
			[
				13,
				"neighborWeight",
				[
					1,
					"amount",
					"*",
					[
						4,
						"-",
						[5, "1"]
					]
				]
			],
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										[
											6,
											[
												7,
												[
													6,
													[
														7,
														[
															7,
															"center",
															"xyz"
														],
														"mul"
													],
													["centerWeight"]
												],
												"add"
											],
											[[
												6,
												[
													7,
													[
														7,
														"top",
														"xyz"
													],
													"mul"
												],
												["neighborWeight"]
											]]
										],
										"add"
									],
									[[
										6,
										[
											7,
											[
												7,
												"bottom",
												"xyz"
											],
											"mul"
										],
										["neighborWeight"]
									]]
								],
								"add"
							],
							[[
								6,
								[
									7,
									[
										7,
										"left",
										"xyz"
									],
									"mul"
								],
								["neighborWeight"]
							]]
						],
						"add"
					],
					[[
						6,
						[
							7,
							[
								7,
								"right",
								"xyz"
							],
							"mul"
						],
						["neighborWeight"]
					]]
				]
			],
			[
				13,
				"clamped",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"rgb",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0"],
								[5, "0"],
								[5, "0"]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1"],
								[5, "1"],
								[5, "1"]
							]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"clamped",
						"x"
					],
					[
						7,
						"clamped",
						"y"
					],
					[
						7,
						"clamped",
						"z"
					],
					[
						7,
						"center",
						"w"
					]
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Sharpness",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Adjust image sharpness using a convolution kernel",
	requiresRTT: true,
	requiresChild: true,
	props: { sharpness: {
		default: 0,
		description: "How sharp to make the underlying image",
		compileTimeWhen: (prev, next) => prev === 0 !== (next === 0),
		ui: {
			type: ["range", "map"],
			min: 0,
			max: 5,
			step: .1,
			label: "Sharpness",
			group: "Adjustments"
		}
	} },
	fragment: ({ uniforms, childNode, ctx, propValues, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		if ((propValues.sharpness ?? 0) === 0) return call(unpremultiplyAlpha, "unpremultiplyAlpha", [texture.sample(ctx.uv)]);
		const vp = ctx.viewportSize;
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [call(sharpnessCompose, "sharpnessCompose", [
			texture.sample(ctx.uv),
			texture.sample(call(sharpnessTapUV, "sharpnessTapUV", [
				ctx.uv,
				vp,
				floatE(0),
				floatE(1)
			])),
			texture.sample(call(sharpnessTapUV, "sharpnessTapUV", [
				ctx.uv,
				vp,
				floatE(0),
				floatE(-1)
			])),
			texture.sample(call(sharpnessTapUV, "sharpnessTapUV", [
				ctx.uv,
				vp,
				floatE(-1),
				floatE(0)
			])),
			texture.sample(call(sharpnessTapUV, "sharpnessTapUV", [
				ctx.uv,
				vp,
				floatE(1),
				floatE(0)
			])),
			uniforms.sharpness
		])]);
	}
};
var Sharpness_default = componentDefinition;
export { sharpnessTapUV as i, componentDefinition as n, sharpnessCompose as r, Sharpness_default as t };
