// shaders/dist/core/SimplexNoise-DMWtripU.js
import { At as call, R as std, U as mixStopsOrColorsExpr, n as tgpu, on as d, rt as noise_exports, tt as mxNoiseFloat3 } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig, t as animatedTime } from "./porters-Dg80Yg90.js";
const simplexNoiseField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, scale, seed, contrast, balance, animTime) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const pos = d.vec2f(__tsover_mul(uv.x, aspect), uv.y).mul(std.exp(scale)).add(seed);
	const n = mxNoiseFloat3(d.vec3f(pos.x, pos.y, __tsover_mul(animTime, .5)));
	const adjusted = __tsover_add(__tsover_mul(n, __tsover_add(contrast, 1)), balance);
	const k = std.clamp(__tsover_add(__tsover_mul(adjusted, .5), .5), 0, 1);
	return __tsover_sub(1, k);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "contrast"
			},
			{
				"type": "i",
				"name": "balance"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"pos",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"aspectUV",
								"mul"
							],
							[[
								6,
								[
									7,
									"std",
									"exp"
								],
								["scale"]
							]]
						],
						"add"
					],
					["seed"]
				]
			],
			[
				13,
				"n",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat3"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								7,
								"pos",
								"x"
							],
							[
								7,
								"pos",
								"y"
							],
							[
								1,
								"animTime",
								"*",
								[5, "0.5"]
							]
						]
					]]
				]
			],
			[
				13,
				"adjusted",
				[
					1,
					[
						1,
						"n",
						"*",
						[
							1,
							"contrast",
							"+",
							[5, "1"]
						]
					],
					"+",
					"balance"
				]
			],
			[
				13,
				"k",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"adjusted",
								"*",
								[5, "0.5"]
							],
							"+",
							[5, "0.5"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, [
				1,
				[5, "1"],
				"-",
				"k"
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"noise"
		]
	},
	externals: () => ({
		d,
		std,
		noise: noise_exports
	})
}) && $.f)({}));
const componentDefinition = {
	name: "SimplexNoise",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Organic noise with animated movement",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#ffffff",
			description: "First color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#000000",
			description: "Second color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		stops: colorStopsPropConfig(),
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		scale: {
			default: 2,
			description: "Pattern scale (higher = larger patterns)",
			ui: {
				type: "range",
				min: -2,
				max: 5,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		balance: {
			default: 0,
			description: "Balance between colors (negative = more colorB, positive = more colorA)",
			ui: {
				type: ["range", "map"],
				min: -1,
				max: 1,
				step: .1,
				label: "Balance",
				group: "Effect"
			}
		},
		contrast: {
			default: 0,
			description: "Pattern contrast (higher = sharper transitions)",
			ui: {
				type: ["range", "map"],
				min: -2,
				max: 5,
				step: .1,
				label: "Contrast",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for pattern variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		}
	},
	fragment: (params) => {
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		return mixStopsOrColorsExpr(params, call(simplexNoiseField, "simplexNoiseField", [
			uv,
			viewport,
			params.uniforms.scale,
			params.uniforms.seed,
			params.uniforms.contrast,
			params.uniforms.balance,
			t
		]));
	}
};
var SimplexNoise_default = componentDefinition;
export { componentDefinition as n, simplexNoiseField as r, SimplexNoise_default as t };
