// shaders/dist/core/SineWave-CHygTRV-.js
import { At as call, It as vec4, R as std, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition, r as transformAngle } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
var DEG_TO_RAD = Math.PI / 180;
var TAU = Math.PI * 2;
const sineWaveMask = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, angle, position, frequency, amplitude, thickness, softness, animTime) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const aspectCorrectedX = __tsover_mul(uv.x, aspect);
	const centerX = __tsover_mul(position.x, aspect);
	const centerY = __tsover_sub(1, position.y);
	const centeredX = __tsover_sub(aspectCorrectedX, centerX);
	const centeredY = __tsover_sub(uv.y, centerY);
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const cosAngle = std.cos(angleRad);
	const sinAngle = std.sin(angleRad);
	const rotatedX = __tsover_sub(__tsover_mul(centeredX, cosAngle), __tsover_mul(centeredY, sinAngle));
	const rotatedY = __tsover_add(__tsover_mul(centeredX, sinAngle), __tsover_mul(centeredY, cosAngle));
	const waveInput = __tsover_add(__tsover_mul(__tsover_mul(rotatedX, frequency), TAU), animTime);
	const sineWave = __tsover_mul(std.sin(waveInput), amplitude);
	const distanceFromWave = std.abs(__tsover_sub(rotatedY, sineWave));
	const halfThickness = __tsover_mul(thickness, .5);
	const halfSoftness = __tsover_mul(softness, .5);
	return __tsover_sub(1, std.smoothstep(__tsover_sub(halfThickness, halfSoftness), __tsover_add(halfThickness, halfSoftness), distanceFromWave));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "position"
			},
			{
				"type": "i",
				"name": "frequency"
			},
			{
				"type": "i",
				"name": "amplitude"
			},
			{
				"type": "i",
				"name": "thickness"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"aspectCorrectedX",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"centerX",
				[
					1,
					[
						7,
						"position",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"centerY",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"position",
						"y"
					]
				]
			],
			[
				13,
				"centeredX",
				[
					1,
					"aspectCorrectedX",
					"-",
					"centerX"
				]
			],
			[
				13,
				"centeredY",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					"centerY"
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosAngle",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinAngle",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"rotatedX",
				[
					1,
					[
						1,
						"centeredX",
						"*",
						"cosAngle"
					],
					"-",
					[
						1,
						"centeredY",
						"*",
						"sinAngle"
					]
				]
			],
			[
				13,
				"rotatedY",
				[
					1,
					[
						1,
						"centeredX",
						"*",
						"sinAngle"
					],
					"+",
					[
						1,
						"centeredY",
						"*",
						"cosAngle"
					]
				]
			],
			[
				13,
				"waveInput",
				[
					1,
					[
						1,
						[
							1,
							"rotatedX",
							"*",
							"frequency"
						],
						"*",
						"TAU"
					],
					"+",
					"animTime"
				]
			],
			[
				13,
				"sineWave",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						["waveInput"]
					],
					"*",
					"amplitude"
				]
			],
			[
				13,
				"distanceFromWave",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						1,
						"rotatedY",
						"-",
						"sineWave"
					]]
				]
			],
			[
				13,
				"halfThickness",
				[
					1,
					"thickness",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"halfSoftness",
				[
					1,
					"softness",
					"*",
					[5, "0.5"]
				]
			],
			[10, [
				1,
				[5, "1"],
				"-",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"halfThickness",
							"-",
							"halfSoftness"
						],
						[
							1,
							"halfThickness",
							"+",
							"halfSoftness"
						],
						"distanceFromWave"
					]
				]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"std",
			"TAU"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		std,
		TAU
	})
}) && $.f)({}));
const componentDefinition = {
	name: "SineWave",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Animated wave with thickness and softness",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		color: {
			default: "#ffffff",
			transform: transformColor,
			description: "The color of the sine wave",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		amplitude: {
			default: .15,
			description: "The height/amplitude of the sine wave",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Amplitude",
				group: "Effect"
			}
		},
		frequency: {
			default: 1,
			description: "The frequency/number of wave cycles",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 20,
				step: .1,
				label: "Frequency",
				group: "Effect"
			}
		},
		speed: {
			default: 1,
			description: "The animation speed of the wave",
			ui: {
				type: "range",
				min: -5,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "The rotation angle of the wave (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		position: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center position of the wave",
			ui: {
				type: "position",
				label: "Position",
				group: "Position"
			}
		},
		thickness: {
			default: .2,
			description: "The thickness of the wave line",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "Thickness",
				group: "Effect"
			}
		},
		softness: {
			default: .4,
			description: "Edge softness of the wave line",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Softness",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { uniforms } = params;
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		const mask = call(sineWaveMask, "sineWaveMask", [
			uv,
			viewport,
			uniforms.angle,
			uniforms.position,
			uniforms.frequency,
			uniforms.amplitude,
			uniforms.thickness,
			uniforms.softness,
			t
		]);
		return vec4(uniforms.color.member("rgb"), uniforms.color.member("a").mul(mask));
	}
};
var SineWave_default = componentDefinition;
export { componentDefinition as n, sineWaveMask as r, SineWave_default as t };
