// shaders/dist/core/SmokeFill-ClBFBWz3.js
import { At as call, B as createGuardedCompute, Gt as colorMixing_exports, It as vec4, Ot as ZERO, R as std, T as sdf_exports, V as createStateBuffer, d as sdf3d_exports, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { n as glass_exports } from "./glass-hUL2pCYG.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as shapeEffectBoundingBoxDeclaration } from "./shapeEffectBounds-BlxxsYMN.js";
var { sdfSpaceUV } = glass_exports;
var { createSvgSdfSampler, createSdfDataTexture, SVG_SDF_SIZE, createAnalyticSdfSampler, driveAnalyticSubProps, ANALYTIC_SDF_EXTRA_FIELDS, buildAnalyticSdfFn } = sdf_exports;
var { createVolumetricFieldComputeNode, buildVolumetricFieldSampler, VOLUMETRIC_FIELD_EXTRA_FIELDS } = sdf3d_exports;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
var N = 256;
var COUNT = N * N;
var JACOBI_ITERS = 10;
var PRESSURE_DISSIPATION = .8;
var VELOCITY_DISSIPATION = .2;
var STATE_FORMAT = "rgba16float";
var cm = colorMixing_exports;
function resolveShapeType(shapeType, shape) {
	if (shapeType) return shapeType;
	if (shape && typeof shape === "object") return shape.type || "circleSDF";
	if (typeof shape === "string") try {
		return JSON.parse(shape).type || "circleSDF";
	} catch {
		return "circleSDF";
	}
	return "circleSDF";
}
var FillParams = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(d.struct({
	dt: d.f32,
	emitX: d.f32,
	emitY: d.f32,
	emitVelX: d.f32,
	emitVelY: d.f32,
	perpDirX: d.f32,
	perpDirY: d.f32,
	spreadFactor: d.f32,
	emitRad: d.f32,
	emitIntensity: d.f32,
	densDissip: d.f32,
	curlStrength: d.f32,
	gravity: d.f32,
	colorDecay: d.f32,
	centerX: d.f32,
	centerY: d.f32,
	scale: d.f32,
	rotation: d.f32,
	aspect: d.f32,
	cursorX: d.f32,
	cursorY: d.f32,
	cursorVelX: d.f32,
	cursorVelY: d.f32,
	mouseActive: d.f32,
	mouseRadSq: d.f32,
	saRadius: d.f32,
	saSides: d.f32,
	saRounding: d.f32,
	saInnerRatio: d.f32,
	saRotation: d.f32,
	saHeight: d.f32,
	saOffset: d.f32,
	saAperture: d.f32,
	vfOriginX: d.f32,
	vfOriginY: d.f32,
	vfSpanX: d.f32,
	vfSpanY: d.f32,
	vfActiveRes: d.f32
}), "FillParams");
var fluidLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	velA: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "mutable"
	},
	velB: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "mutable"
	},
	densA: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "mutable"
	},
	densB: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "mutable"
	},
	pressure: {
		storage: d.arrayOf(d.f32, COUNT),
		access: "mutable"
	},
	divergence: {
		storage: d.arrayOf(d.f32, COUNT),
		access: "mutable"
	},
	maskBuf: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "readonly"
	},
	params: { uniform: FillParams }
}), "fluidLayout");
var maskLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	maskBuf: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "mutable"
	},
	params: { uniform: FillParams }
}), "maskLayout");
var maskSvgLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	maskBuf: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "mutable"
	},
	params: { uniform: FillParams },
	sdfSource: { texture: d.texture2d(d.f32) }
}), "maskSvgLayout");
var maskVfLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	maskBuf: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "mutable"
	},
	params: { uniform: FillParams },
	fieldTex: {
		texture: d.texture2d(d.f32),
		sampleType: "unfilterable-float"
	}
}), "maskVfLayout");
var outputLayout = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(tgpu.bindGroupLayout({
	densA: {
		storage: d.arrayOf(d.vec4f, COUNT),
		access: "readonly"
	},
	outTex: { storageTexture: d.textureStorage2d(STATE_FORMAT, "write-only") }
}), "outputLayout");
var nidx = tgpu.fn([
	d.i32,
	d.i32,
	d.i32,
	d.i32
], d.u32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((ii, ji, di, dj) => {
	"use gpu";
	const row = std.clamp(__tsover_add(ii, di), 0, __tsover_sub(N, 1));
	const col = std.clamp(__tsover_add(ji, dj), 0, __tsover_sub(N, 1));
	return d.u32(__tsover_add(__tsover_mul(row, N), col));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "ii"
			},
			{
				"type": "i",
				"name": "ji"
			},
			{
				"type": "i",
				"name": "di"
			},
			{
				"type": "i",
				"name": "dj"
			}
		],
		"body": [0, [
			[
				13,
				"row",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"ii",
							"+",
							"di"
						],
						[5, "0"],
						[
							1,
							"N",
							"-",
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"col",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"ji",
							"+",
							"dj"
						],
						[5, "0"],
						[
							1,
							"N",
							"-",
							[5, "1"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"u32"
				],
				[[
					1,
					[
						1,
						"row",
						"*",
						"N"
					],
					"+",
					"col"
				]]
			]]
		]],
		"externalNames": [
			"std",
			"N",
			"d"
		]
	},
	externals: () => ({
		std,
		N,
		d
	})
}) && $.f)({}));
function makeMaskKernel(shapeType) {
	const sdfFn = buildAnalyticSdfFn(shapeType);
	return tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
		"use gpu";
		const p = maskLayout.$.params;
		const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
		const u = __tsover_div(__tsover_add(d.f32(cx), .5), d.f32(N));
		const v = __tsover_div(__tsover_add(d.f32(cy), .5), d.f32(N));
		const dxAc = __tsover_mul(__tsover_sub(u, p.centerX), p.aspect);
		const dy = __tsover_sub(v, p.centerY);
		const cosR = std.cos(p.rotation);
		const sinR = std.sin(p.rotation);
		const rdx = __tsover_add(__tsover_mul(dxAc, cosR), __tsover_mul(dy, sinR));
		const rdy = __tsover_sub(__tsover_mul(dy, cosR), __tsover_mul(dxAc, sinR));
		const res = sdfFn(d.vec2f(__tsover_add(__tsover_div(rdx, p.scale), .5), __tsover_add(__tsover_div(rdy, p.scale), .5)), p.saRadius, p.saSides, p.saRounding, p.saInnerRatio, p.saRotation, p.saHeight, p.saOffset, p.saAperture);
		const inside = std.select(d.f32(0), d.f32(1), res.x < 0);
		maskLayout.$.maskBuf[idx] = d.vec4f(inside, res.x, 0, 0);
	}), {
		v: 1,
		name: void 0,
		ast: {
			"params": [{
				"type": "i",
				"name": "cx"
			}, {
				"type": "i",
				"name": "cy"
			}],
			"body": [0, [
				[
					13,
					"p",
					[
						7,
						[
							7,
							"maskLayout",
							"$"
						],
						"params"
					]
				],
				[
					13,
					"idx",
					[
						1,
						[
							1,
							"cy",
							"*",
							[
								6,
								[
									7,
									"d",
									"u32"
								],
								["N"]
							]
						],
						"+",
						"cx"
					]
				],
				[
					13,
					"u",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cx"]
							],
							"+",
							[5, "0.5"]
						],
						"/",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["N"]
						]
					]
				],
				[
					13,
					"v",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cy"]
							],
							"+",
							[5, "0.5"]
						],
						"/",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["N"]
						]
					]
				],
				[
					13,
					"dxAc",
					[
						1,
						[
							1,
							"u",
							"-",
							[
								7,
								"p",
								"centerX"
							]
						],
						"*",
						[
							7,
							"p",
							"aspect"
						]
					]
				],
				[
					13,
					"dy",
					[
						1,
						"v",
						"-",
						[
							7,
							"p",
							"centerY"
						]
					]
				],
				[
					13,
					"cosR",
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						[[
							7,
							"p",
							"rotation"
						]]
					]
				],
				[
					13,
					"sinR",
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						[[
							7,
							"p",
							"rotation"
						]]
					]
				],
				[
					13,
					"rdx",
					[
						1,
						[
							1,
							"dxAc",
							"*",
							"cosR"
						],
						"+",
						[
							1,
							"dy",
							"*",
							"sinR"
						]
					]
				],
				[
					13,
					"rdy",
					[
						1,
						[
							1,
							"dy",
							"*",
							"cosR"
						],
						"-",
						[
							1,
							"dxAc",
							"*",
							"sinR"
						]
					]
				],
				[
					13,
					"sdfUV",
					[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								1,
								"rdx",
								"/",
								[
									7,
									"p",
									"scale"
								]
							],
							"+",
							[5, "0.5"]
						], [
							1,
							[
								1,
								"rdy",
								"/",
								[
									7,
									"p",
									"scale"
								]
							],
							"+",
							[5, "0.5"]
						]]
					]
				],
				[
					13,
					"res",
					[
						6,
						"sdfFn",
						[
							"sdfUV",
							[
								7,
								"p",
								"saRadius"
							],
							[
								7,
								"p",
								"saSides"
							],
							[
								7,
								"p",
								"saRounding"
							],
							[
								7,
								"p",
								"saInnerRatio"
							],
							[
								7,
								"p",
								"saRotation"
							],
							[
								7,
								"p",
								"saHeight"
							],
							[
								7,
								"p",
								"saOffset"
							],
							[
								7,
								"p",
								"saAperture"
							]
						]
					]
				],
				[
					13,
					"inside",
					[
						6,
						[
							7,
							"std",
							"select"
						],
						[
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							],
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "1"]]
							],
							[
								1,
								[
									7,
									"res",
									"x"
								],
								"<",
								[5, "0"]
							]
						]
					]
				],
				[
					2,
					[
						8,
						[
							7,
							[
								7,
								"maskLayout",
								"$"
							],
							"maskBuf"
						],
						"idx"
					],
					"=",
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							"inside",
							[
								7,
								"res",
								"x"
							],
							[5, "0"],
							[5, "0"]
						]
					]
				]
			]],
			"externalNames": [
				"maskLayout",
				"d",
				"N",
				"std",
				"sdfFn"
			]
		},
		externals: () => ({
			maskLayout,
			d,
			N,
			std,
			sdfFn
		})
	}) && $.f)({})).$name("smokeFillMask");
}
const svgMaskKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = maskSvgLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const u = __tsover_div(__tsover_add(d.f32(cx), .5), d.f32(N));
	const v = __tsover_div(__tsover_add(d.f32(cy), .5), d.f32(N));
	const dxAc = __tsover_mul(__tsover_sub(u, p.centerX), p.aspect);
	const dy = __tsover_sub(v, p.centerY);
	const cosR = std.cos(p.rotation);
	const sinR = std.sin(p.rotation);
	const rdx = __tsover_add(__tsover_mul(dxAc, cosR), __tsover_mul(dy, sinR));
	const rdy = __tsover_sub(__tsover_mul(dy, cosR), __tsover_mul(dxAc, sinR));
	const sdfUV = d.vec2f(__tsover_add(__tsover_div(rdx, p.scale), .5), __tsover_add(__tsover_div(rdy, p.scale), .5));
	const sizeM1 = __tsover_sub(SVG_SDF_SIZE, 1);
	const tx = std.clamp(d.i32(__tsover_mul(sdfUV.x, SVG_SDF_SIZE)), 0, sizeM1);
	const ty = std.clamp(d.i32(__tsover_mul(sdfUV.y, SVG_SDF_SIZE)), 0, sizeM1);
	const dist = std.textureLoad(maskSvgLayout.$.sdfSource, d.vec2u(d.u32(tx), d.u32(ty)), 0).x;
	const inside = std.select(d.f32(0), d.f32(1), dist < 0);
	maskSvgLayout.$.maskBuf[idx] = d.vec4f(inside, dist, 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"maskSvgLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"u",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["cx"]
						],
						"+",
						[5, "0.5"]
					],
					"/",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["N"]
					]
				]
			],
			[
				13,
				"v",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["cy"]
						],
						"+",
						[5, "0.5"]
					],
					"/",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["N"]
					]
				]
			],
			[
				13,
				"dxAc",
				[
					1,
					[
						1,
						"u",
						"-",
						[
							7,
							"p",
							"centerX"
						]
					],
					"*",
					[
						7,
						"p",
						"aspect"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					"v",
					"-",
					[
						7,
						"p",
						"centerY"
					]
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					[[
						7,
						"p",
						"rotation"
					]]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						7,
						"p",
						"rotation"
					]]
				]
			],
			[
				13,
				"rdx",
				[
					1,
					[
						1,
						"dxAc",
						"*",
						"cosR"
					],
					"+",
					[
						1,
						"dy",
						"*",
						"sinR"
					]
				]
			],
			[
				13,
				"rdy",
				[
					1,
					[
						1,
						"dy",
						"*",
						"cosR"
					],
					"-",
					[
						1,
						"dxAc",
						"*",
						"sinR"
					]
				]
			],
			[
				13,
				"sdfUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							"rdx",
							"/",
							[
								7,
								"p",
								"scale"
							]
						],
						"+",
						[5, "0.5"]
					], [
						1,
						[
							1,
							"rdy",
							"/",
							[
								7,
								"p",
								"scale"
							]
						],
						"+",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"sizeM1",
				[
					1,
					"SVG_SDF_SIZE",
					"-",
					[5, "1"]
				]
			],
			[
				13,
				"tx",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"d",
								"i32"
							],
							[[
								1,
								[
									7,
									"sdfUV",
									"x"
								],
								"*",
								"SVG_SDF_SIZE"
							]]
						],
						[5, "0"],
						"sizeM1"
					]
				]
			],
			[
				13,
				"ty",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"d",
								"i32"
							],
							[[
								1,
								[
									7,
									"sdfUV",
									"y"
								],
								"*",
								"SVG_SDF_SIZE"
							]]
						],
						[5, "0"],
						"sizeM1"
					]
				]
			],
			[
				13,
				"dist",
				[
					7,
					[
						6,
						[
							7,
							"std",
							"textureLoad"
						],
						[
							[
								7,
								[
									7,
									"maskSvgLayout",
									"$"
								],
								"sdfSource"
							],
							[
								6,
								[
									7,
									"d",
									"vec2u"
								],
								[[
									6,
									[
										7,
										"d",
										"u32"
									],
									["tx"]
								], [
									6,
									[
										7,
										"d",
										"u32"
									],
									["ty"]
								]]
							],
							[5, "0"]
						]
					],
					"x"
				]
			],
			[
				13,
				"inside",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"dist",
							"<",
							[5, "0"]
						]
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"maskSvgLayout",
							"$"
						],
						"maskBuf"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						"inside",
						"dist",
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"maskSvgLayout",
			"d",
			"N",
			"std",
			"SVG_SDF_SIZE"
		]
	},
	externals: () => ({
		maskSvgLayout,
		d,
		N,
		std,
		SVG_SDF_SIZE
	})
}) && $.f)({})).$name("smokeFillSvgMask");
const vfMaskKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = maskVfLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const u = __tsover_div(__tsover_add(d.f32(cx), .5), d.f32(N));
	const v = __tsover_div(__tsover_add(d.f32(cy), .5), d.f32(N));
	const dxAc = __tsover_mul(__tsover_sub(u, p.centerX), p.aspect);
	const dy = __tsover_sub(v, p.centerY);
	const cosR = std.cos(p.rotation);
	const sinR = std.sin(p.rotation);
	const rdx = __tsover_add(__tsover_mul(dxAc, cosR), __tsover_mul(dy, sinR));
	const rdy = __tsover_sub(__tsover_mul(dy, cosR), __tsover_mul(dxAc, sinR));
	const sdfUV = d.vec2f(__tsover_add(__tsover_div(rdx, p.scale), .5), __tsover_add(__tsover_div(rdy, p.scale), .5));
	const uxc = std.clamp(sdfUV.x, p.vfOriginX, __tsover_add(p.vfOriginX, p.vfSpanX));
	const uyc = std.clamp(sdfUV.y, p.vfOriginY, __tsover_add(p.vfOriginY, p.vfSpanY));
	const outsideDist = std.length(d.vec2f(__tsover_sub(sdfUV.x, uxc), __tsover_sub(sdfUV.y, uyc)));
	const maxT = __tsover_sub(p.vfActiveRes, 1);
	const tx = std.clamp(__tsover_sub(__tsover_mul(__tsover_div(__tsover_sub(uxc, p.vfOriginX), p.vfSpanX), p.vfActiveRes), .5), d.f32(0), maxT);
	const ty = std.clamp(__tsover_sub(__tsover_mul(__tsover_div(__tsover_sub(uyc, p.vfOriginY), p.vfSpanY), p.vfActiveRes), .5), d.f32(0), maxT);
	const dist = __tsover_add(std.textureLoad(maskVfLayout.$.fieldTex, d.vec2u(d.u32(tx), d.u32(ty)), 0).x, outsideDist);
	const inside = std.select(d.f32(0), d.f32(1), dist < 0);
	maskVfLayout.$.maskBuf[idx] = d.vec4f(inside, dist, 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"maskVfLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"u",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["cx"]
						],
						"+",
						[5, "0.5"]
					],
					"/",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["N"]
					]
				]
			],
			[
				13,
				"v",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["cy"]
						],
						"+",
						[5, "0.5"]
					],
					"/",
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["N"]
					]
				]
			],
			[
				13,
				"dxAc",
				[
					1,
					[
						1,
						"u",
						"-",
						[
							7,
							"p",
							"centerX"
						]
					],
					"*",
					[
						7,
						"p",
						"aspect"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					"v",
					"-",
					[
						7,
						"p",
						"centerY"
					]
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					[[
						7,
						"p",
						"rotation"
					]]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						7,
						"p",
						"rotation"
					]]
				]
			],
			[
				13,
				"rdx",
				[
					1,
					[
						1,
						"dxAc",
						"*",
						"cosR"
					],
					"+",
					[
						1,
						"dy",
						"*",
						"sinR"
					]
				]
			],
			[
				13,
				"rdy",
				[
					1,
					[
						1,
						"dy",
						"*",
						"cosR"
					],
					"-",
					[
						1,
						"dxAc",
						"*",
						"sinR"
					]
				]
			],
			[
				13,
				"sdfUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							"rdx",
							"/",
							[
								7,
								"p",
								"scale"
							]
						],
						"+",
						[5, "0.5"]
					], [
						1,
						[
							1,
							"rdy",
							"/",
							[
								7,
								"p",
								"scale"
							]
						],
						"+",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"uxc",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							7,
							"sdfUV",
							"x"
						],
						[
							7,
							"p",
							"vfOriginX"
						],
						[
							1,
							[
								7,
								"p",
								"vfOriginX"
							],
							"+",
							[
								7,
								"p",
								"vfSpanX"
							]
						]
					]
				]
			],
			[
				13,
				"uyc",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							7,
							"sdfUV",
							"y"
						],
						[
							7,
							"p",
							"vfOriginY"
						],
						[
							1,
							[
								7,
								"p",
								"vfOriginY"
							],
							"+",
							[
								7,
								"p",
								"vfSpanY"
							]
						]
					]
				]
			],
			[
				13,
				"outsideDist",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								7,
								"sdfUV",
								"x"
							],
							"-",
							"uxc"
						], [
							1,
							[
								7,
								"sdfUV",
								"y"
							],
							"-",
							"uyc"
						]]
					]]
				]
			],
			[
				13,
				"maxT",
				[
					1,
					[
						7,
						"p",
						"vfActiveRes"
					],
					"-",
					[5, "1"]
				]
			],
			[
				13,
				"tx",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										"uxc",
										"-",
										[
											7,
											"p",
											"vfOriginX"
										]
									],
									"/",
									[
										7,
										"p",
										"vfSpanX"
									]
								],
								"*",
								[
									7,
									"p",
									"vfActiveRes"
								]
							],
							"-",
							[5, "0.5"]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						"maxT"
					]
				]
			],
			[
				13,
				"ty",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									[
										1,
										"uyc",
										"-",
										[
											7,
											"p",
											"vfOriginY"
										]
									],
									"/",
									[
										7,
										"p",
										"vfSpanY"
									]
								],
								"*",
								[
									7,
									"p",
									"vfActiveRes"
								]
							],
							"-",
							[5, "0.5"]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						"maxT"
					]
				]
			],
			[
				13,
				"dist",
				[
					1,
					[
						7,
						[
							6,
							[
								7,
								"std",
								"textureLoad"
							],
							[
								[
									7,
									[
										7,
										"maskVfLayout",
										"$"
									],
									"fieldTex"
								],
								[
									6,
									[
										7,
										"d",
										"vec2u"
									],
									[[
										6,
										[
											7,
											"d",
											"u32"
										],
										["tx"]
									], [
										6,
										[
											7,
											"d",
											"u32"
										],
										["ty"]
									]]
								],
								[5, "0"]
							]
						],
						"x"
					],
					"+",
					"outsideDist"
				]
			],
			[
				13,
				"inside",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"dist",
							"<",
							[5, "0"]
						]
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"maskVfLayout",
							"$"
						],
						"maskBuf"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						"inside",
						"dist",
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"maskVfLayout",
			"d",
			"N",
			"std"
		]
	},
	externals: () => ({
		maskVfLayout,
		d,
		N,
		std
	})
}) && $.f)({})).$name("smokeFillVfMask");
const splatKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = fluidLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const mask = fluidLayout.$.maskBuf[idx].x;
	const cellX = __tsover_add(d.f32(cx), .5);
	const cellY = __tsover_add(d.f32(cy), .5);
	const dx = __tsover_sub(cellX, p.emitX);
	const dy = __tsover_sub(cellY, p.emitY);
	const distSq = __tsover_add(__tsover_mul(dx, dx), __tsover_mul(dy, dy));
	const radSq = std.max(__tsover_mul(p.emitRad, p.emitRad), d.f32(1));
	const influence = __tsover_mul(std.exp(__tsover_mul(__tsover_div(distSq, radSq), -1)), mask);
	const cell = fluidLayout.$.densA[idx];
	const oldDens = cell.x;
	const addDens = __tsover_mul(__tsover_mul(__tsover_mul(p.emitIntensity, influence), p.dt), 8);
	const newDens = std.min(__tsover_add(oldDens, addDens), 1);
	const rejuvenate = __tsover_div(addDens, std.max(newDens, d.f32(.001)));
	fluidLayout.$.densA[idx] = d.vec4f(newDens, __tsover_mul(cell.y, __tsover_sub(1, std.min(rejuvenate, 1))), 0, 0);
	const vel = fluidLayout.$.velA[idx];
	const perpOff = __tsover_div(__tsover_add(__tsover_mul(dx, p.perpDirX), __tsover_mul(dy, p.perpDirY)), std.max(p.emitRad, d.f32(1)));
	const vmSq = std.max(__tsover_add(__tsover_mul(p.emitVelX, p.emitVelX), __tsover_mul(p.emitVelY, p.emitVelY)), d.f32(1));
	const tvx = __tsover_add(p.emitVelX, __tsover_mul(__tsover_mul(__tsover_mul(p.perpDirX, perpOff), p.spreadFactor), std.sqrt(vmSq)));
	const tvy = __tsover_add(p.emitVelY, __tsover_mul(__tsover_mul(__tsover_mul(p.perpDirY, perpOff), p.spreadFactor), std.sqrt(vmSq)));
	const br = std.min(__tsover_mul(__tsover_mul(influence, p.dt), 3), 1);
	const gravForce = __tsover_mul(__tsover_mul(__tsover_mul(__tsover_mul(p.gravity, oldDens), p.dt), __tsover_mul(N, .08)), mask);
	const cursorDX = __tsover_sub(cellX, p.cursorX);
	const cursorDY = __tsover_sub(cellY, p.cursorY);
	const cursorDistSq = __tsover_add(__tsover_mul(cursorDX, cursorDX), __tsover_mul(cursorDY, cursorDY));
	const cursorInf = __tsover_mul(__tsover_mul(std.exp(__tsover_mul(__tsover_div(cursorDistSq, std.max(p.mouseRadSq, d.f32(1))), -1)), p.mouseActive), mask);
	fluidLayout.$.velA[idx] = d.vec4f(__tsover_add(std.mix(vel.x, tvx, br), __tsover_mul(p.cursorVelX, cursorInf)), __tsover_add(__tsover_add(std.mix(vel.y, tvy, br), gravForce), __tsover_mul(p.cursorVelY, cursorInf)), vel.z, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"fluidLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"mask",
				[
					7,
					[
						8,
						[
							7,
							[
								7,
								"fluidLayout",
								"$"
							],
							"maskBuf"
						],
						"idx"
					],
					"x"
				]
			],
			[
				13,
				"cellX",
				[
					1,
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["cx"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"cellY",
				[
					1,
					[
						6,
						[
							7,
							"d",
							"f32"
						],
						["cy"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"dx",
				[
					1,
					"cellX",
					"-",
					[
						7,
						"p",
						"emitX"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					"cellY",
					"-",
					[
						7,
						"p",
						"emitY"
					]
				]
			],
			[
				13,
				"distSq",
				[
					1,
					[
						1,
						"dx",
						"*",
						"dx"
					],
					"+",
					[
						1,
						"dy",
						"*",
						"dy"
					]
				]
			],
			[
				13,
				"radSq",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[
							7,
							"p",
							"emitRad"
						],
						"*",
						[
							7,
							"p",
							"emitRad"
						]
					], [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "1"]]
					]]
				]
			],
			[
				13,
				"influence",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"exp"
						],
						[[
							1,
							[
								1,
								"distSq",
								"/",
								"radSq"
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						]]
					],
					"*",
					"mask"
				]
			],
			[
				13,
				"cell",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					"idx"
				]
			],
			[
				13,
				"oldDens",
				[
					7,
					"cell",
					"x"
				]
			],
			[
				13,
				"addDens",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"p",
								"emitIntensity"
							],
							"*",
							"influence"
						],
						"*",
						[
							7,
							"p",
							"dt"
						]
					],
					"*",
					[5, "8"]
				]
			],
			[
				13,
				"newDens",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					[[
						1,
						"oldDens",
						"+",
						"addDens"
					], [5, "1"]]
				]
			],
			[
				13,
				"rejuvenate",
				[
					1,
					"addDens",
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						["newDens", [
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0.001"]]
						]]
					]
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						"newDens",
						[
							1,
							[
								7,
								"cell",
								"y"
							],
							"*",
							[
								1,
								[5, "1"],
								"-",
								[
									6,
									[
										7,
										"std",
										"min"
									],
									["rejuvenate", [5, "1"]]
								]
							]
						],
						[5, "0"],
						[5, "0"]
					]
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				13,
				"perpOff",
				[
					1,
					[
						1,
						[
							1,
							"dx",
							"*",
							[
								7,
								"p",
								"perpDirX"
							]
						],
						"+",
						[
							1,
							"dy",
							"*",
							[
								7,
								"p",
								"perpDirY"
							]
						]
					],
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							7,
							"p",
							"emitRad"
						], [
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						]]
					]
				]
			],
			[
				13,
				"vmSq",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[
							1,
							[
								7,
								"p",
								"emitVelX"
							],
							"*",
							[
								7,
								"p",
								"emitVelX"
							]
						],
						"+",
						[
							1,
							[
								7,
								"p",
								"emitVelY"
							],
							"*",
							[
								7,
								"p",
								"emitVelY"
							]
						]
					], [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "1"]]
					]]
				]
			],
			[
				13,
				"tvx",
				[
					1,
					[
						7,
						"p",
						"emitVelX"
					],
					"+",
					[
						1,
						[
							1,
							[
								1,
								[
									7,
									"p",
									"perpDirX"
								],
								"*",
								"perpOff"
							],
							"*",
							[
								7,
								"p",
								"spreadFactor"
							]
						],
						"*",
						[
							6,
							[
								7,
								"std",
								"sqrt"
							],
							["vmSq"]
						]
					]
				]
			],
			[
				13,
				"tvy",
				[
					1,
					[
						7,
						"p",
						"emitVelY"
					],
					"+",
					[
						1,
						[
							1,
							[
								1,
								[
									7,
									"p",
									"perpDirY"
								],
								"*",
								"perpOff"
							],
							"*",
							[
								7,
								"p",
								"spreadFactor"
							]
						],
						"*",
						[
							6,
							[
								7,
								"std",
								"sqrt"
							],
							["vmSq"]
						]
					]
				]
			],
			[
				13,
				"br",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					[[
						1,
						[
							1,
							"influence",
							"*",
							[
								7,
								"p",
								"dt"
							]
						],
						"*",
						[5, "3"]
					], [5, "1"]]
				]
			],
			[
				13,
				"gravForce",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									7,
									"p",
									"gravity"
								],
								"*",
								"oldDens"
							],
							"*",
							[
								7,
								"p",
								"dt"
							]
						],
						"*",
						[
							1,
							"N",
							"*",
							[5, "0.08"]
						]
					],
					"*",
					"mask"
				]
			],
			[
				13,
				"cursorDX",
				[
					1,
					"cellX",
					"-",
					[
						7,
						"p",
						"cursorX"
					]
				]
			],
			[
				13,
				"cursorDY",
				[
					1,
					"cellY",
					"-",
					[
						7,
						"p",
						"cursorY"
					]
				]
			],
			[
				13,
				"cursorDistSq",
				[
					1,
					[
						1,
						"cursorDX",
						"*",
						"cursorDX"
					],
					"+",
					[
						1,
						"cursorDY",
						"*",
						"cursorDY"
					]
				]
			],
			[
				13,
				"cursorInf",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									1,
									"cursorDistSq",
									"/",
									[
										6,
										[
											7,
											"std",
											"max"
										],
										[[
											7,
											"p",
											"mouseRadSq"
										], [
											6,
											[
												7,
												"d",
												"f32"
											],
											[[5, "1"]]
										]]
									]
								],
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]]
						],
						"*",
						[
							7,
							"p",
							"mouseActive"
						]
					],
					"*",
					"mask"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"mix"
								],
								[
									[
										7,
										"vel",
										"x"
									],
									"tvx",
									"br"
								]
							],
							"+",
							[
								1,
								[
									7,
									"p",
									"cursorVelX"
								],
								"*",
								"cursorInf"
							]
						],
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										[
											7,
											"vel",
											"y"
										],
										"tvy",
										"br"
									]
								],
								"+",
								"gravForce"
							],
							"+",
							[
								1,
								[
									7,
									"p",
									"cursorVelY"
								],
								"*",
								"cursorInf"
							]
						],
						[
							7,
							"vel",
							"z"
						],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"fluidLayout",
			"d",
			"N",
			"std"
		]
	},
	externals: () => ({
		fluidLayout,
		d,
		N,
		std
	})
}) && $.f)({})).$name("smokeFillSplat");
const curlKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const ii = d.i32(cy);
	const ji = d.i32(cx);
	const vL = fluidLayout.$.velA[nidx(ii, ji, 0, -1)];
	const vR = fluidLayout.$.velA[nidx(ii, ji, 0, 1)];
	const vU = fluidLayout.$.velA[nidx(ii, ji, -1, 0)];
	const vD = fluidLayout.$.velA[nidx(ii, ji, 1, 0)];
	const curl = __tsover_mul(__tsover_sub(__tsover_sub(vR.y, vL.y), __tsover_sub(vD.x, vU.x)), .5);
	const vel = fluidLayout.$.velA[idx];
	fluidLayout.$.velA[idx] = d.vec4f(vel.x, vel.y, curl, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"ii",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cy"]
				]
			],
			[
				13,
				"ji",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cx"]
				]
			],
			[
				13,
				"vL",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[5, "0"],
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]
				]
			],
			[
				13,
				"vR",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[5, "0"],
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"vU",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[
								4,
								"-",
								[5, "1"]
							],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"vD",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[5, "1"],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"curl",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"vR",
								"y"
							],
							"-",
							[
								7,
								"vL",
								"y"
							]
						],
						"-",
						[
							1,
							[
								7,
								"vD",
								"x"
							],
							"-",
							[
								7,
								"vU",
								"x"
							]
						]
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							7,
							"vel",
							"x"
						],
						[
							7,
							"vel",
							"y"
						],
						"curl",
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout",
			"nidx"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout,
		nidx
	})
}) && $.f)({})).$name("smokeFillCurl");
const vorticityKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = fluidLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const ii = d.i32(cy);
	const ji = d.i32(cx);
	const cL = std.abs(fluidLayout.$.velA[nidx(ii, ji, 0, -1)].z);
	const cR = std.abs(fluidLayout.$.velA[nidx(ii, ji, 0, 1)].z);
	const cU = std.abs(fluidLayout.$.velA[nidx(ii, ji, -1, 0)].z);
	const cD = std.abs(fluidLayout.$.velA[nidx(ii, ji, 1, 0)].z);
	const cc = fluidLayout.$.velA[idx].z;
	const gx = __tsover_mul(__tsover_sub(cR, cL), .5);
	const gy = __tsover_mul(__tsover_sub(cD, cU), .5);
	const gl = std.max(std.sqrt(__tsover_add(__tsover_mul(gx, gx), __tsover_mul(gy, gy))), d.f32(1e-5));
	const vel = fluidLayout.$.velA[idx];
	fluidLayout.$.velA[idx] = d.vec4f(__tsover_add(vel.x, __tsover_mul(__tsover_mul(__tsover_mul(__tsover_div(gy, gl), cc), p.curlStrength), p.dt)), __tsover_add(vel.y, __tsover_mul(__tsover_mul(__tsover_mul(__tsover_mul(__tsover_div(gx, gl), -1), cc), p.curlStrength), p.dt)), vel.z, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"fluidLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"ii",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cy"]
				]
			],
			[
				13,
				"ji",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cx"]
				]
			],
			[
				13,
				"cL",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						7,
						[
							8,
							[
								7,
								[
									7,
									"fluidLayout",
									"$"
								],
								"velA"
							],
							[
								6,
								"nidx",
								[
									"ii",
									"ji",
									[5, "0"],
									[
										4,
										"-",
										[5, "1"]
									]
								]
							]
						],
						"z"
					]]
				]
			],
			[
				13,
				"cR",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						7,
						[
							8,
							[
								7,
								[
									7,
									"fluidLayout",
									"$"
								],
								"velA"
							],
							[
								6,
								"nidx",
								[
									"ii",
									"ji",
									[5, "0"],
									[5, "1"]
								]
							]
						],
						"z"
					]]
				]
			],
			[
				13,
				"cU",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						7,
						[
							8,
							[
								7,
								[
									7,
									"fluidLayout",
									"$"
								],
								"velA"
							],
							[
								6,
								"nidx",
								[
									"ii",
									"ji",
									[
										4,
										"-",
										[5, "1"]
									],
									[5, "0"]
								]
							]
						],
						"z"
					]]
				]
			],
			[
				13,
				"cD",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						7,
						[
							8,
							[
								7,
								[
									7,
									"fluidLayout",
									"$"
								],
								"velA"
							],
							[
								6,
								"nidx",
								[
									"ii",
									"ji",
									[5, "1"],
									[5, "0"]
								]
							]
						],
						"z"
					]]
				]
			],
			[
				13,
				"cc",
				[
					7,
					[
						8,
						[
							7,
							[
								7,
								"fluidLayout",
								"$"
							],
							"velA"
						],
						"idx"
					],
					"z"
				]
			],
			[
				13,
				"gx",
				[
					1,
					[
						1,
						"cR",
						"-",
						"cL"
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"gy",
				[
					1,
					[
						1,
						"cD",
						"-",
						"cU"
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"gl",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"gx",
								"*",
								"gx"
							],
							"+",
							[
								1,
								"gy",
								"*",
								"gy"
							]
						]]
					], [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0.00001"]]
					]]
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							1,
							[
								7,
								"vel",
								"x"
							],
							"+",
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											"gy",
											"/",
											"gl"
										],
										"*",
										"cc"
									],
									"*",
									[
										7,
										"p",
										"curlStrength"
									]
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						[
							1,
							[
								7,
								"vel",
								"y"
							],
							"+",
							[
								1,
								[
									1,
									[
										1,
										[
											1,
											[
												1,
												"gx",
												"/",
												"gl"
											],
											"*",
											[
												4,
												"-",
												[5, "1"]
											]
										],
										"*",
										"cc"
									],
									"*",
									[
										7,
										"p",
										"curlStrength"
									]
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						[
							7,
							"vel",
							"z"
						],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"fluidLayout",
			"d",
			"N",
			"std",
			"nidx"
		]
	},
	externals: () => ({
		fluidLayout,
		d,
		N,
		std,
		nidx
	})
}) && $.f)({})).$name("smokeFillVorticity");
const divergenceKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const ii = d.i32(cy);
	const ji = d.i32(cx);
	const jf = d.f32(cx);
	const iF = d.f32(cy);
	const selfMask = fluidLayout.$.maskBuf[idx].x;
	const selfVel = fluidLayout.$.velA[idx];
	const isWallL = jf < .5 || fluidLayout.$.maskBuf[nidx(ii, ji, 0, -1)].x < .5;
	const isWallR = jf > d.f32(__tsover_sub(N, 1.5)) || fluidLayout.$.maskBuf[nidx(ii, ji, 0, 1)].x < .5;
	const isWallT = iF < .5 || fluidLayout.$.maskBuf[nidx(ii, ji, -1, 0)].x < .5;
	const isWallB = iF > d.f32(__tsover_sub(N, 1.5)) || fluidLayout.$.maskBuf[nidx(ii, ji, 1, 0)].x < .5;
	const L = std.select(fluidLayout.$.velA[nidx(ii, ji, 0, -1)].x, __tsover_mul(selfVel.x, -1), isWallL);
	const R = std.select(fluidLayout.$.velA[nidx(ii, ji, 0, 1)].x, __tsover_mul(selfVel.x, -1), isWallR);
	const T = std.select(fluidLayout.$.velA[nidx(ii, ji, -1, 0)].y, __tsover_mul(selfVel.y, -1), isWallT);
	const B = std.select(fluidLayout.$.velA[nidx(ii, ji, 1, 0)].y, __tsover_mul(selfVel.y, -1), isWallB);
	const div = __tsover_mul(__tsover_mul(__tsover_add(__tsover_sub(R, L), __tsover_sub(B, T)), .5), selfMask);
	fluidLayout.$.pressure[idx] = __tsover_mul(__tsover_mul(fluidLayout.$.pressure[idx], PRESSURE_DISSIPATION), selfMask);
	fluidLayout.$.divergence[idx] = div;
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"ii",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cy"]
				]
			],
			[
				13,
				"ji",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cx"]
				]
			],
			[
				13,
				"jf",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					["cx"]
				]
			],
			[
				13,
				"iF",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					["cy"]
				]
			],
			[
				13,
				"selfMask",
				[
					7,
					[
						8,
						[
							7,
							[
								7,
								"fluidLayout",
								"$"
							],
							"maskBuf"
						],
						"idx"
					],
					"x"
				]
			],
			[
				13,
				"selfVel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				13,
				"isWallL",
				[
					3,
					[
						1,
						"jf",
						"<",
						[5, "0.5"]
					],
					"||",
					[
						1,
						[
							7,
							[
								8,
								[
									7,
									[
										7,
										"fluidLayout",
										"$"
									],
									"maskBuf"
								],
								[
									6,
									"nidx",
									[
										"ii",
										"ji",
										[5, "0"],
										[
											4,
											"-",
											[5, "1"]
										]
									]
								]
							],
							"x"
						],
						"<",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"isWallR",
				[
					3,
					[
						1,
						"jf",
						">",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[
								1,
								"N",
								"-",
								[5, "1.5"]
							]]
						]
					],
					"||",
					[
						1,
						[
							7,
							[
								8,
								[
									7,
									[
										7,
										"fluidLayout",
										"$"
									],
									"maskBuf"
								],
								[
									6,
									"nidx",
									[
										"ii",
										"ji",
										[5, "0"],
										[5, "1"]
									]
								]
							],
							"x"
						],
						"<",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"isWallT",
				[
					3,
					[
						1,
						"iF",
						"<",
						[5, "0.5"]
					],
					"||",
					[
						1,
						[
							7,
							[
								8,
								[
									7,
									[
										7,
										"fluidLayout",
										"$"
									],
									"maskBuf"
								],
								[
									6,
									"nidx",
									[
										"ii",
										"ji",
										[
											4,
											"-",
											[5, "1"]
										],
										[5, "0"]
									]
								]
							],
							"x"
						],
						"<",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"isWallB",
				[
					3,
					[
						1,
						"iF",
						">",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[
								1,
								"N",
								"-",
								[5, "1.5"]
							]]
						]
					],
					"||",
					[
						1,
						[
							7,
							[
								8,
								[
									7,
									[
										7,
										"fluidLayout",
										"$"
									],
									"maskBuf"
								],
								[
									6,
									"nidx",
									[
										"ii",
										"ji",
										[5, "1"],
										[5, "0"]
									]
								]
							],
							"x"
						],
						"<",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"L",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							7,
							[
								8,
								[
									7,
									[
										7,
										"fluidLayout",
										"$"
									],
									"velA"
								],
								[
									6,
									"nidx",
									[
										"ii",
										"ji",
										[5, "0"],
										[
											4,
											"-",
											[5, "1"]
										]
									]
								]
							],
							"x"
						],
						[
							1,
							[
								7,
								"selfVel",
								"x"
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"isWallL"
					]
				]
			],
			[
				13,
				"R",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							7,
							[
								8,
								[
									7,
									[
										7,
										"fluidLayout",
										"$"
									],
									"velA"
								],
								[
									6,
									"nidx",
									[
										"ii",
										"ji",
										[5, "0"],
										[5, "1"]
									]
								]
							],
							"x"
						],
						[
							1,
							[
								7,
								"selfVel",
								"x"
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"isWallR"
					]
				]
			],
			[
				13,
				"T",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							7,
							[
								8,
								[
									7,
									[
										7,
										"fluidLayout",
										"$"
									],
									"velA"
								],
								[
									6,
									"nidx",
									[
										"ii",
										"ji",
										[
											4,
											"-",
											[5, "1"]
										],
										[5, "0"]
									]
								]
							],
							"y"
						],
						[
							1,
							[
								7,
								"selfVel",
								"y"
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"isWallT"
					]
				]
			],
			[
				13,
				"B",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							7,
							[
								8,
								[
									7,
									[
										7,
										"fluidLayout",
										"$"
									],
									"velA"
								],
								[
									6,
									"nidx",
									[
										"ii",
										"ji",
										[5, "1"],
										[5, "0"]
									]
								]
							],
							"y"
						],
						[
							1,
							[
								7,
								"selfVel",
								"y"
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"isWallB"
					]
				]
			],
			[
				13,
				"div",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								"R",
								"-",
								"L"
							],
							"+",
							[
								1,
								"B",
								"-",
								"T"
							]
						],
						"*",
						[5, "0.5"]
					],
					"*",
					"selfMask"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					"idx"
				],
				"=",
				[
					1,
					[
						1,
						[
							8,
							[
								7,
								[
									7,
									"fluidLayout",
									"$"
								],
								"pressure"
							],
							"idx"
						],
						"*",
						"PRESSURE_DISSIPATION"
					],
					"*",
					"selfMask"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"divergence"
					],
					"idx"
				],
				"=",
				"div"
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout",
			"nidx",
			"std",
			"PRESSURE_DISSIPATION"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout,
		nidx,
		std,
		PRESSURE_DISSIPATION
	})
}) && $.f)({})).$name("smokeFillDivergence");
const jacobiKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const ii = d.i32(cy);
	const ji = d.i32(cx);
	const pL = fluidLayout.$.pressure[nidx(ii, ji, 0, -1)];
	const pR = fluidLayout.$.pressure[nidx(ii, ji, 0, 1)];
	const pU = fluidLayout.$.pressure[nidx(ii, ji, -1, 0)];
	const pD = fluidLayout.$.pressure[nidx(ii, ji, 1, 0)];
	const div = fluidLayout.$.divergence[idx];
	fluidLayout.$.pressure[idx] = __tsover_mul(__tsover_sub(__tsover_add(__tsover_add(__tsover_add(pL, pR), pU), pD), div), .25);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"ii",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cy"]
				]
			],
			[
				13,
				"ji",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cx"]
				]
			],
			[
				13,
				"pL",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[5, "0"],
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]
				]
			],
			[
				13,
				"pR",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[5, "0"],
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"pU",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[
								4,
								"-",
								[5, "1"]
							],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"pD",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[5, "1"],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"div",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"divergence"
					],
					"idx"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					"idx"
				],
				"=",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									"pL",
									"+",
									"pR"
								],
								"+",
								"pU"
							],
							"+",
							"pD"
						],
						"-",
						"div"
					],
					"*",
					[5, "0.25"]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout",
			"nidx"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout,
		nidx
	})
}) && $.f)({})).$name("smokeFillJacobi");
const gradSubtractKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const ii = d.i32(cy);
	const ji = d.i32(cx);
	const pL = fluidLayout.$.pressure[nidx(ii, ji, 0, -1)];
	const pR = fluidLayout.$.pressure[nidx(ii, ji, 0, 1)];
	const pU = fluidLayout.$.pressure[nidx(ii, ji, -1, 0)];
	const pD = fluidLayout.$.pressure[nidx(ii, ji, 1, 0)];
	const vel = fluidLayout.$.velA[idx];
	fluidLayout.$.velA[idx] = d.vec4f(__tsover_sub(vel.x, __tsover_mul(__tsover_sub(pR, pL), .5)), __tsover_sub(vel.y, __tsover_mul(__tsover_sub(pD, pU), .5)), vel.z, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"ii",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cy"]
				]
			],
			[
				13,
				"ji",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["cx"]
				]
			],
			[
				13,
				"pL",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[5, "0"],
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]
				]
			],
			[
				13,
				"pR",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[5, "0"],
							[5, "1"]
						]
					]
				]
			],
			[
				13,
				"pU",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[
								4,
								"-",
								[5, "1"]
							],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"pD",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"pressure"
					],
					[
						6,
						"nidx",
						[
							"ii",
							"ji",
							[5, "1"],
							[5, "0"]
						]
					]
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							1,
							[
								7,
								"vel",
								"x"
							],
							"-",
							[
								1,
								[
									1,
									"pR",
									"-",
									"pL"
								],
								"*",
								[5, "0.5"]
							]
						],
						[
							1,
							[
								7,
								"vel",
								"y"
							],
							"-",
							[
								1,
								[
									1,
									"pD",
									"-",
									"pU"
								],
								"*",
								[5, "0.5"]
							]
						],
						[
							7,
							"vel",
							"z"
						],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout",
			"nidx"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout,
		nidx
	})
}) && $.f)({})).$name("smokeFillGradSubtract");
const advectVelKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = fluidLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const vel = fluidLayout.$.velA[idx];
	const srcX = std.clamp(__tsover_sub(d.f32(cx), __tsover_mul(vel.x, p.dt)), 0, d.f32(__tsover_sub(N, 1)));
	const srcY = std.clamp(__tsover_sub(d.f32(cy), __tsover_mul(vel.y, p.dt)), 0, d.f32(__tsover_sub(N, 1)));
	const x0 = std.clamp(std.floor(srcX), 0, d.f32(__tsover_sub(N, 2)));
	const y0 = std.clamp(std.floor(srcY), 0, d.f32(__tsover_sub(N, 2)));
	const sx = __tsover_sub(srcX, x0);
	const sy = __tsover_sub(srcY, y0);
	const ix0 = d.i32(x0);
	const iy0 = d.i32(y0);
	const v00 = fluidLayout.$.velA[d.u32(__tsover_add(__tsover_mul(iy0, N), ix0))];
	const v10 = fluidLayout.$.velA[d.u32(__tsover_add(__tsover_add(__tsover_mul(iy0, N), ix0), 1))];
	const v01 = fluidLayout.$.velA[d.u32(__tsover_add(__tsover_mul(__tsover_add(iy0, 1), N), ix0))];
	const v11 = fluidLayout.$.velA[d.u32(__tsover_add(__tsover_add(__tsover_mul(__tsover_add(iy0, 1), N), ix0), 1))];
	const decay = __tsover_add(1, __tsover_mul(VELOCITY_DISSIPATION, p.dt));
	const mask = fluidLayout.$.maskBuf[idx].x;
	fluidLayout.$.velB[idx] = d.vec4f(__tsover_mul(__tsover_div(std.mix(std.mix(v00.x, v10.x, sx), std.mix(v01.x, v11.x, sx), sy), decay), mask), __tsover_mul(__tsover_div(std.mix(std.mix(v00.y, v10.y, sx), std.mix(v01.y, v11.y, sx), sy), decay), mask), 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"fluidLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				13,
				"srcX",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cx"]
							],
							"-",
							[
								1,
								[
									7,
									"vel",
									"x"
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						[5, "0"],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[
								1,
								"N",
								"-",
								[5, "1"]
							]]
						]
					]
				]
			],
			[
				13,
				"srcY",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cy"]
							],
							"-",
							[
								1,
								[
									7,
									"vel",
									"y"
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						[5, "0"],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[
								1,
								"N",
								"-",
								[5, "1"]
							]]
						]
					]
				]
			],
			[
				13,
				"x0",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							["srcX"]
						],
						[5, "0"],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[
								1,
								"N",
								"-",
								[5, "2"]
							]]
						]
					]
				]
			],
			[
				13,
				"y0",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							["srcY"]
						],
						[5, "0"],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[
								1,
								"N",
								"-",
								[5, "2"]
							]]
						]
					]
				]
			],
			[
				13,
				"sx",
				[
					1,
					"srcX",
					"-",
					"x0"
				]
			],
			[
				13,
				"sy",
				[
					1,
					"srcY",
					"-",
					"y0"
				]
			],
			[
				13,
				"ix0",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["x0"]
				]
			],
			[
				13,
				"iy0",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["y0"]
				]
			],
			[
				13,
				"v00",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								"iy0",
								"*",
								"N"
							],
							"+",
							"ix0"
						]]
					]
				]
			],
			[
				13,
				"v10",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								[
									1,
									"iy0",
									"*",
									"N"
								],
								"+",
								"ix0"
							],
							"+",
							[5, "1"]
						]]
					]
				]
			],
			[
				13,
				"v01",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								[
									1,
									"iy0",
									"+",
									[5, "1"]
								],
								"*",
								"N"
							],
							"+",
							"ix0"
						]]
					]
				]
			],
			[
				13,
				"v11",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								[
									1,
									[
										1,
										"iy0",
										"+",
										[5, "1"]
									],
									"*",
									"N"
								],
								"+",
								"ix0"
							],
							"+",
							[5, "1"]
						]]
					]
				]
			],
			[
				13,
				"decay",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						"VELOCITY_DISSIPATION",
						"*",
						[
							7,
							"p",
							"dt"
						]
					]
				]
			],
			[
				13,
				"mask",
				[
					7,
					[
						8,
						[
							7,
							[
								7,
								"fluidLayout",
								"$"
							],
							"maskBuf"
						],
						"idx"
					],
					"x"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velB"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										[
											6,
											[
												7,
												"std",
												"mix"
											],
											[
												[
													7,
													"v00",
													"x"
												],
												[
													7,
													"v10",
													"x"
												],
												"sx"
											]
										],
										[
											6,
											[
												7,
												"std",
												"mix"
											],
											[
												[
													7,
													"v01",
													"x"
												],
												[
													7,
													"v11",
													"x"
												],
												"sx"
											]
										],
										"sy"
									]
								],
								"/",
								"decay"
							],
							"*",
							"mask"
						],
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										[
											6,
											[
												7,
												"std",
												"mix"
											],
											[
												[
													7,
													"v00",
													"y"
												],
												[
													7,
													"v10",
													"y"
												],
												"sx"
											]
										],
										[
											6,
											[
												7,
												"std",
												"mix"
											],
											[
												[
													7,
													"v01",
													"y"
												],
												[
													7,
													"v11",
													"y"
												],
												"sx"
											]
										],
										"sy"
									]
								],
								"/",
								"decay"
							],
							"*",
							"mask"
						],
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"fluidLayout",
			"d",
			"N",
			"std",
			"VELOCITY_DISSIPATION"
		]
	},
	externals: () => ({
		fluidLayout,
		d,
		N,
		std,
		VELOCITY_DISSIPATION
	})
}) && $.f)({})).$name("smokeFillAdvectVel");
const copyVelKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const v = fluidLayout.$.velB[idx];
	fluidLayout.$.velA[idx] = d.vec4f(v.x, v.y, 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"v",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velB"
					],
					"idx"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							7,
							"v",
							"x"
						],
						[
							7,
							"v",
							"y"
						],
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout
	})
}) && $.f)({})).$name("smokeFillCopyVel");
const advectDensKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const p = fluidLayout.$.params;
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const vel = fluidLayout.$.velA[idx];
	const srcX = std.clamp(__tsover_sub(d.f32(cx), __tsover_mul(vel.x, p.dt)), 0, d.f32(__tsover_sub(N, 1)));
	const srcY = std.clamp(__tsover_sub(d.f32(cy), __tsover_mul(vel.y, p.dt)), 0, d.f32(__tsover_sub(N, 1)));
	const x0 = std.clamp(std.floor(srcX), 0, d.f32(__tsover_sub(N, 2)));
	const y0 = std.clamp(std.floor(srcY), 0, d.f32(__tsover_sub(N, 2)));
	const sx = __tsover_sub(srcX, x0);
	const sy = __tsover_sub(srcY, y0);
	const ix0 = d.i32(x0);
	const iy0 = d.i32(y0);
	const d00 = fluidLayout.$.densA[d.u32(__tsover_add(__tsover_mul(iy0, N), ix0))];
	const d10 = fluidLayout.$.densA[d.u32(__tsover_add(__tsover_add(__tsover_mul(iy0, N), ix0), 1))];
	const d01 = fluidLayout.$.densA[d.u32(__tsover_add(__tsover_mul(__tsover_add(iy0, 1), N), ix0))];
	const d11 = fluidLayout.$.densA[d.u32(__tsover_add(__tsover_add(__tsover_mul(__tsover_add(iy0, 1), N), ix0), 1))];
	const dens = std.mix(std.mix(d00.x, d10.x, sx), std.mix(d01.x, d11.x, sx), sy);
	const age = std.mix(std.mix(d00.y, d10.y, sx), std.mix(d01.y, d11.y, sx), sy);
	const decay = __tsover_add(1, __tsover_mul(p.densDissip, p.dt));
	const mask = fluidLayout.$.maskBuf[idx].x;
	fluidLayout.$.densB[idx] = d.vec4f(__tsover_mul(std.max(__tsover_div(dens, decay), d.f32(0)), mask), std.min(__tsover_add(age, __tsover_mul(__tsover_mul(p.dt, .4), p.colorDecay)), 1), 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"p",
				[
					7,
					[
						7,
						"fluidLayout",
						"$"
					],
					"params"
				]
			],
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"vel",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"velA"
					],
					"idx"
				]
			],
			[
				13,
				"srcX",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cx"]
							],
							"-",
							[
								1,
								[
									7,
									"vel",
									"x"
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						[5, "0"],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[
								1,
								"N",
								"-",
								[5, "1"]
							]]
						]
					]
				]
			],
			[
				13,
				"srcY",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"d",
									"f32"
								],
								["cy"]
							],
							"-",
							[
								1,
								[
									7,
									"vel",
									"y"
								],
								"*",
								[
									7,
									"p",
									"dt"
								]
							]
						],
						[5, "0"],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[
								1,
								"N",
								"-",
								[5, "1"]
							]]
						]
					]
				]
			],
			[
				13,
				"x0",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							["srcX"]
						],
						[5, "0"],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[
								1,
								"N",
								"-",
								[5, "2"]
							]]
						]
					]
				]
			],
			[
				13,
				"y0",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							["srcY"]
						],
						[5, "0"],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[
								1,
								"N",
								"-",
								[5, "2"]
							]]
						]
					]
				]
			],
			[
				13,
				"sx",
				[
					1,
					"srcX",
					"-",
					"x0"
				]
			],
			[
				13,
				"sy",
				[
					1,
					"srcY",
					"-",
					"y0"
				]
			],
			[
				13,
				"ix0",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["x0"]
				]
			],
			[
				13,
				"iy0",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					["y0"]
				]
			],
			[
				13,
				"d00",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								"iy0",
								"*",
								"N"
							],
							"+",
							"ix0"
						]]
					]
				]
			],
			[
				13,
				"d10",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								[
									1,
									"iy0",
									"*",
									"N"
								],
								"+",
								"ix0"
							],
							"+",
							[5, "1"]
						]]
					]
				]
			],
			[
				13,
				"d01",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								[
									1,
									"iy0",
									"+",
									[5, "1"]
								],
								"*",
								"N"
							],
							"+",
							"ix0"
						]]
					]
				]
			],
			[
				13,
				"d11",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					[
						6,
						[
							7,
							"d",
							"u32"
						],
						[[
							1,
							[
								1,
								[
									1,
									[
										1,
										"iy0",
										"+",
										[5, "1"]
									],
									"*",
									"N"
								],
								"+",
								"ix0"
							],
							"+",
							[5, "1"]
						]]
					]
				]
			],
			[
				13,
				"dens",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									7,
									"d00",
									"x"
								],
								[
									7,
									"d10",
									"x"
								],
								"sx"
							]
						],
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									7,
									"d01",
									"x"
								],
								[
									7,
									"d11",
									"x"
								],
								"sx"
							]
						],
						"sy"
					]
				]
			],
			[
				13,
				"age",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									7,
									"d00",
									"y"
								],
								[
									7,
									"d10",
									"y"
								],
								"sx"
							]
						],
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									7,
									"d01",
									"y"
								],
								[
									7,
									"d11",
									"y"
								],
								"sx"
							]
						],
						"sy"
					]
				]
			],
			[
				13,
				"decay",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						[
							7,
							"p",
							"densDissip"
						],
						"*",
						[
							7,
							"p",
							"dt"
						]
					]
				]
			],
			[
				13,
				"mask",
				[
					7,
					[
						8,
						[
							7,
							[
								7,
								"fluidLayout",
								"$"
							],
							"maskBuf"
						],
						"idx"
					],
					"x"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densB"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							1,
							[
								6,
								[
									7,
									"std",
									"max"
								],
								[[
									1,
									"dens",
									"/",
									"decay"
								], [
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "0"]]
								]]
							],
							"*",
							"mask"
						],
						[
							6,
							[
								7,
								"std",
								"min"
							],
							[[
								1,
								"age",
								"+",
								[
									1,
									[
										1,
										[
											7,
											"p",
											"dt"
										],
										"*",
										[5, "0.4"]
									],
									"*",
									[
										7,
										"p",
										"colorDecay"
									]
								]
							], [5, "1"]]
						],
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"fluidLayout",
			"d",
			"N",
			"std"
		]
	},
	externals: () => ({
		fluidLayout,
		d,
		N,
		std
	})
}) && $.f)({})).$name("smokeFillAdvectDens");
const copyDensKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const dd = fluidLayout.$.densB[idx];
	fluidLayout.$.densA[idx] = d.vec4f(dd.x, dd.y, 0, 0);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"dd",
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densB"
					],
					"idx"
				]
			],
			[
				2,
				[
					8,
					[
						7,
						[
							7,
							"fluidLayout",
							"$"
						],
						"densA"
					],
					"idx"
				],
				"=",
				[
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[
							7,
							"dd",
							"x"
						],
						[
							7,
							"dd",
							"y"
						],
						[5, "0"],
						[5, "0"]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"fluidLayout"
		]
	},
	externals: () => ({
		d,
		N,
		fluidLayout
	})
}) && $.f)({})).$name("smokeFillCopyDens");
const outputKernel = tgpu.fn([d.u32, d.u32])(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
	"use gpu";
	const idx = __tsover_add(__tsover_mul(cy, d.u32(N)), cx);
	const dd = outputLayout.$.densA[idx];
	std.textureStore(outputLayout.$.outTex, d.vec2u(cx, cy), d.vec4f(dd.x, dd.y, 0, 0));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "cx"
		}, {
			"type": "i",
			"name": "cy"
		}],
		"body": [0, [
			[
				13,
				"idx",
				[
					1,
					[
						1,
						"cy",
						"*",
						[
							6,
							[
								7,
								"d",
								"u32"
							],
							["N"]
						]
					],
					"+",
					"cx"
				]
			],
			[
				13,
				"dd",
				[
					8,
					[
						7,
						[
							7,
							"outputLayout",
							"$"
						],
						"densA"
					],
					"idx"
				]
			],
			[
				6,
				[
					7,
					"std",
					"textureStore"
				],
				[
					[
						7,
						[
							7,
							"outputLayout",
							"$"
						],
						"outTex"
					],
					[
						6,
						[
							7,
							"d",
							"vec2u"
						],
						["cx", "cy"]
					],
					[
						6,
						[
							7,
							"d",
							"vec4f"
						],
						[
							[
								7,
								"dd",
								"x"
							],
							[
								7,
								"dd",
								"y"
							],
							[5, "0"],
							[5, "0"]
						]
					]
				]
			]
		]],
		"externalNames": [
			"d",
			"N",
			"outputLayout",
			"std"
		]
	},
	externals: () => ({
		d,
		N,
		outputLayout,
		std
	})
}) && $.f)({})).$name("smokeFillOutput");
const smokeFillShapeMask = tgpu.fn([d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((dist) => {
	"use gpu";
	return std.clamp(__tsover_mul(__tsover_mul(dist, -1), __tsover_mul(N, 4)), 0, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "dist"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"clamp"
			],
			[
				[
					1,
					[
						1,
						"dist",
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					],
					"*",
					[
						1,
						"N",
						"*",
						[5, "4"]
					]
				],
				[5, "0"],
				[5, "1"]
			]
		]]]],
		"externalNames": ["std", "N"]
	},
	externals: () => ({
		std,
		N
	})
}) && $.f)({}));
const componentDefinition = {
	name: "SmokeFill",
	boundingBoxDeclaration: shapeEffectBoundingBoxDeclaration,
	category: "Shape Effects",
	description: "Fill a shape with swirling fluid smoke that interacts with the shape boundary",
	requiresRTT: false,
	requiresChild: false,
	usesPointer: true,
	extraFields: {
		...VOLUMETRIC_FIELD_EXTRA_FIELDS,
		...ANALYTIC_SDF_EXTRA_FIELDS
	},
	props: {
		colorA: {
			default: "#8cf3ff",
			transform: transformColor,
			description: "Color of fresh smoke",
			ui: {
				type: "color",
				label: "Start Color",
				group: "Colors"
			}
		},
		colorB: {
			default: "#04a0d6",
			transform: transformColor,
			description: "Color smoke transitions to as it ages",
			ui: {
				type: "color",
				label: "End Color",
				group: "Colors"
			}
		},
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		scale: {
			default: 1,
			description: "Scale of the shape (1 = default size)",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Scale",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the shape in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		},
		emitFrom: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Emission source point within the shape",
			ui: {
				type: "position",
				label: "Emit From",
				group: "Emission"
			}
		},
		direction: {
			default: 0,
			description: "Emission direction (0 = up, 90 = right, 180 = down, 270 = left)",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Direction",
				group: "Emission"
			}
		},
		speed: {
			default: 10,
			description: "Emission velocity strength",
			ui: {
				type: "range",
				min: .1,
				max: 30,
				step: .1,
				label: "Speed",
				group: "Emission"
			}
		},
		spread: {
			default: 60,
			description: "Emission cone angle in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 180,
				step: 1,
				label: "Spread",
				group: "Emission"
			}
		},
		emitRadius: {
			default: .03,
			description: "Size of the emission area",
			ui: {
				type: "range",
				min: .01,
				max: .3,
				step: .01,
				label: "Emit Size",
				group: "Emission"
			}
		},
		intensity: {
			default: 1,
			description: "Smoke emission density",
			ui: {
				type: "range",
				min: .1,
				max: 1,
				step: .01,
				label: "Intensity",
				group: "Effect"
			}
		},
		dissipation: {
			default: .3,
			description: "How fast smoke fades over time",
			ui: {
				type: "range",
				min: .1,
				max: 5,
				step: .1,
				label: "Dissipation",
				group: "Effect"
			}
		},
		detail: {
			default: 25,
			description: "Fine-scale swirling detail",
			ui: {
				type: "range",
				min: 0,
				max: 50,
				step: 1,
				label: "Detail",
				group: "Effect"
			}
		},
		gravity: {
			default: .5,
			description: "Downward gravitational pull on smoke — 0 = weightless, negative values = smoke rises",
			ui: {
				type: "range",
				min: -2,
				max: 2,
				step: .1,
				label: "Gravity",
				group: "Effect"
			}
		},
		colorDecay: {
			default: .4,
			description: "How quickly smoke shifts from Color A to Color B",
			ui: {
				type: "range",
				min: 0,
				max: 3,
				step: .1,
				label: "Color Decay",
				group: "Colors"
			}
		},
		mouseInfluence: {
			default: .1,
			description: "Strength of cursor influence",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Cursor Influence",
				group: "Interaction"
			}
		},
		mouseRadius: {
			default: .1,
			description: "Radius of cursor influence area",
			ui: {
				type: "range",
				min: .02,
				max: .5,
				step: .01,
				label: "Cursor Radius",
				group: "Interaction"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type"
		}
	},
	compute: (params) => {
		const { gpu, getCpuValue, registerComputeTexture, onCleanup } = params;
		const root = gpu?.root;
		if (!root) return null;
		const shapeSdfUrl = getCpuValue("shapeSdfUrl") || "";
		const shapeType = resolveShapeType(getCpuValue("shapeType") || "", getCpuValue("shape"));
		const vf = createVolumetricFieldComputeNode(params, () => getCpuValue("shape"), "none");
		const velA = createStateBuffer(root, d.vec4f, COUNT);
		const velB = createStateBuffer(root, d.vec4f, COUNT);
		const densA = createStateBuffer(root, d.vec4f, COUNT);
		const densB = createStateBuffer(root, d.vec4f, COUNT);
		const pressure = createStateBuffer(root, d.f32, COUNT);
		const divergence = createStateBuffer(root, d.f32, COUNT);
		const maskBuf = createStateBuffer(root, d.vec4f, COUNT);
		const outTex = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createTexture({
			size: [N, N],
			format: STATE_FORMAT
		}).$usage("storage", "sampled"), "outTex");
		onCleanup(() => outTex.destroy());
		const smokeTexture = registerComputeTexture(outTex);
		const paramsU = /* @__PURE__ */ (globalThis.__TYPEGPU_AUTONAME__ ?? ((a) => a))(root.createUniform(FillParams), "paramsU");
		const size = [N, N];
		const fluidBg = root.createBindGroup(fluidLayout, {
			velA,
			velB,
			densA,
			densB,
			pressure,
			divergence,
			maskBuf,
			params: paramsU.buffer
		});
		const outputBg = root.createBindGroup(outputLayout, {
			densA,
			outTex
		});
		const mk = (kernel, bg = fluidBg) => createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			kernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"kernel",
					["cx", "cy"]
				]]],
				"externalNames": ["kernel"]
			},
			externals: () => ({ kernel })
		}) && $.f)({}), {
			size,
			bindGroup: bg
		});
		let maskPass;
		if (vf) maskPass = mk(vfMaskKernel, root.createBindGroup(maskVfLayout, {
			maskBuf,
			params: paramsU.buffer,
			fieldTex: vf.fieldTexture
		}));
		else if (shapeSdfUrl) {
			const sdfTex = createSdfDataTexture(params, shapeSdfUrl);
			maskPass = mk(svgMaskKernel, root.createBindGroup(maskSvgLayout, {
				maskBuf,
				params: paramsU.buffer,
				sdfSource: sdfTex.texture.texture
			}));
		} else {
			const maskBg = root.createBindGroup(maskLayout, {
				maskBuf,
				params: paramsU.buffer
			});
			maskPass = mk(makeMaskKernel(shapeType), maskBg);
		}
		const splat = mk(splatKernel);
		const curl = mk(curlKernel);
		const vorticity = mk(vorticityKernel);
		const divergencePass = mk(divergenceKernel);
		const jacobi = mk(jacobiKernel);
		const gradSubtract = mk(gradSubtractKernel);
		const advectVel = mk(advectVelKernel);
		const copyVel = mk(copyVelKernel);
		const advectDens = mk(advectDensKernel);
		const copyDens = mk(copyDensKernel);
		const output = createGuardedCompute(root, /* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((cx, cy) => {
			"use gpu";
			outputKernel(cx, cy);
		}), {
			v: 1,
			name: void 0,
			ast: {
				"params": [{
					"type": "i",
					"name": "cx"
				}, {
					"type": "i",
					"name": "cy"
				}],
				"body": [0, [[
					6,
					"outputKernel",
					["cx", "cy"]
				]]],
				"externalNames": ["outputKernel"]
			},
			externals: () => ({ outputKernel })
		}) && $.f)({}), {
			size,
			bindGroup: outputBg
		});
		const num = (v, f) => typeof v === "number" ? v : f;
		const resolveSub = () => {
			const raw = getCpuValue("shape");
			let c = {};
			if (raw && typeof raw === "object") c = raw;
			else if (typeof raw === "string") try {
				c = JSON.parse(raw);
			} catch {
				c = {};
			}
			return {
				saRadius: num(c.radius, num(c.width, num(c.bottomWidth, .35))),
				saSides: num(c.sides, 6),
				saRounding: num(c.rounding, 0),
				saInnerRatio: num(c.innerRatio, num(c.thickness, num(c.spread, num(c.topWidth, num(c.topRatio, .4))))),
				saRotation: num(c.rotation, 0),
				saHeight: num(c.height, .25),
				saOffset: num(c.offset, num(c.skew, .2)),
				saAperture: num(c.aperture, 270)
			};
		};
		let lastTime = Date.now();
		let prevMx = .5;
		let prevMy = .5;
		return {
			outputs: {
				smokeTexture,
				...vf ? vf.outputs : {}
			},
			getComputeNodes: (frameParams) => {
				const fp = frameParams;
				const vfNodes = vf?.getComputeNodes(fp) ?? null;
				const now = Date.now();
				const dt = Math.min(fp.deltaTime ?? (now - lastTime) / 1e3, .033);
				lastTime = now;
				if (dt < .001) return vfNodes;
				const g = (key, fallback) => {
					const v = getCpuValue(key);
					return typeof v === "number" ? v : fallback;
				};
				const speed = g("speed", 2);
				const dirRad = g("direction", 0) * Math.PI / 180;
				const spreadHalf = Math.min(89 * Math.PI / 180, g("spread", 60) * Math.PI / 360);
				const centerView = getCpuValue("center");
				const emitView = getCpuValue("emitFrom");
				const h = fp.dimensions?.height ?? 0;
				const aspect = h > 0 ? (fp.dimensions?.width ?? 0) / h : 1;
				const velMag = speed * N * .15;
				const sub = resolveSub();
				const mouseInf = g("mouseInfluence", .5);
				const mouseRad = g("mouseRadius", .12);
				const mx = fp.pointer?.x ?? prevMx;
				const my = fp.pointer?.y ?? prevMy;
				const mdx = mx - prevMx;
				const mdy = my - prevMy;
				prevMx = mx;
				prevMy = my;
				const active = Math.sqrt(mdx * mdx + mdy * mdy) > 5e-4 && mouseInf > 0;
				const mouseRadGrid = mouseRad * N;
				const dom = vf?.getFieldDomain();
				paramsU.write({
					dt,
					emitX: (emitView?.x ?? .5) * N,
					emitY: (1 - (emitView?.y ?? .5)) * N,
					emitVelX: Math.sin(dirRad) * velMag,
					emitVelY: -Math.cos(dirRad) * velMag,
					perpDirX: Math.cos(dirRad),
					perpDirY: Math.sin(dirRad),
					spreadFactor: Math.tan(spreadHalf),
					emitRad: g("emitRadius", .03) * N,
					emitIntensity: g("intensity", .8),
					densDissip: g("dissipation", .5),
					curlStrength: g("detail", 30),
					gravity: g("gravity", .5),
					colorDecay: g("colorDecay", 1),
					centerX: centerView?.x ?? .5,
					centerY: 1 - (centerView?.y ?? .5),
					scale: g("scale", 1),
					rotation: g("rotation", 0) * Math.PI / 180,
					aspect,
					cursorX: active ? mx * N : 0,
					cursorY: active ? my * N : 0,
					cursorVelX: active ? mdx * N * 15 * mouseInf : 0,
					cursorVelY: active ? mdy * N * 15 * mouseInf : 0,
					mouseActive: active ? 1 : 0,
					mouseRadSq: mouseRadGrid * mouseRadGrid,
					...sub,
					vfOriginX: dom?.originX ?? 0,
					vfOriginY: dom?.originY ?? 0,
					vfSpanX: dom?.spanX ?? 1,
					vfSpanY: dom?.spanY ?? 1,
					vfActiveRes: dom?.activeRes ?? 1
				});
				const nodes = [
					...vfNodes ?? [],
					maskPass,
					splat,
					curl,
					vorticity,
					divergencePass
				];
				for (let iter = 0; iter < JACOBI_ITERS; iter++) nodes.push(jacobi);
				nodes.push(gradSubtract, advectVel, copyVel, advectDens, copyDens, output);
				return nodes;
			}
		};
	},
	fragment: (params) => {
		const { ctx, computeOutputs, propValues, uniforms, getCpuValue } = params;
		const smokeField = computeOutputs?.smokeTexture;
		if (!smokeField) return ZERO;
		let sampler;
		const fieldTex = computeOutputs?.volumetricFieldTexture;
		if (fieldTex) sampler = buildVolumetricFieldSampler(fieldTex, {
			originX: uniforms._vfOriginX,
			originY: uniforms._vfOriginY,
			spanX: uniforms._vfSpanX,
			spanY: uniforms._vfSpanY,
			activeResF: uniforms._vfActiveRes,
			rBound: uniforms._vfRBound
		}).volumetricFieldSampler;
		else {
			const shapeSdfUrl = propValues.shapeSdfUrl || "";
			if (shapeSdfUrl) sampler = createSvgSdfSampler(params, shapeSdfUrl);
			else sampler = createAnalyticSdfSampler(resolveShapeType(propValues.shapeType || "", getCpuValue("shape")), driveAnalyticSubProps(params, () => getCpuValue("shape")));
		}
		const sdfUV = call(sdfSpaceUV, "sdfSpaceUV", [
			uniforms.center,
			uniforms.scale,
			uniforms.rotation,
			ctx.uv,
			ctx.aspect
		]);
		const shapeMask = call(smokeFillShapeMask, "smokeFillShapeMask", [sampler(sdfUV).member("x")]);
		const s = smokeField.sample(ctx.uv, "linearClamp");
		const smokeDensity = s.member("x");
		const smokeAge = s.member("y");
		const colorSpaceMode = propValues.colorSpace ?? 0;
		const smokeColor = call(cm.mixColorsVariants[colorSpaceMode] ?? cm.mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			smokeAge
		]);
		return vec4(smokeColor.member("rgb"), smokeColor.member("a").mul(smokeDensity).mul(shapeMask));
	}
};
var SmokeFill_default = componentDefinition;
export { vorticityKernel as _, copyDensKernel as a, divergenceKernel as c, makeMaskKernel as d, outputKernel as f, vfMaskKernel as g, svgMaskKernel as h, componentDefinition as i, gradSubtractKernel as l, splatKernel as m, advectDensKernel as n, copyVelKernel as o, smokeFillShapeMask as p, advectVelKernel as r, curlKernel as s, SmokeFill_default as t, jacobiKernel as u };
