// shaders/dist/core/Solarize-D_ZPY_Cr.js
import { At as call, Ot as ZERO, R as std, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
const solarizeColor = tgpu.fn([
	d.vec4f,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, threshold, strength) => {
	"use gpu";
	const rgb = d.vec3f(color.x, color.y, color.z);
	const luminance = std.dot(rgb, d.vec3f(.299, .587, .114));
	const inverted = d.vec3f(__tsover_sub(1, color.x), __tsover_sub(1, color.y), __tsover_sub(1, color.z));
	const solarized = std.select(rgb, inverted, luminance > threshold);
	const result = std.mix(rgb, solarized, d.vec3f(strength));
	return d.vec4f(result.x, result.y, result.z, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "color"
			},
			{
				"type": "i",
				"name": "threshold"
			},
			{
				"type": "i",
				"name": "strength"
			}
		],
		"body": [0, [
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"color",
							"x"
						],
						[
							7,
							"color",
							"y"
						],
						[
							7,
							"color",
							"z"
						]
					]
				]
			],
			[
				13,
				"luminance",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["rgb", [
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0.299"],
							[5, "0.587"],
							[5, "0.114"]
						]
					]]
				]
			],
			[
				13,
				"inverted",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[5, "1"],
							"-",
							[
								7,
								"color",
								"x"
							]
						],
						[
							1,
							[5, "1"],
							"-",
							[
								7,
								"color",
								"y"
							]
						],
						[
							1,
							[5, "1"],
							"-",
							[
								7,
								"color",
								"z"
							]
						]
					]
				]
			],
			[
				13,
				"solarized",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"rgb",
						"inverted",
						[
							1,
							"luminance",
							">",
							"threshold"
						]
					]
				]
			],
			[
				13,
				"result",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"rgb",
						"solarized",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["strength"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"result",
						"x"
					],
					[
						7,
						"result",
						"y"
					],
					[
						7,
						"result",
						"z"
					],
					[
						7,
						"color",
						"w"
					]
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Solarize",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Inverts tones above a luminance threshold — a classic darkroom and photo effect",
	requiresChild: true,
	props: {
		threshold: {
			default: .5,
			description: "Luminance level above which colors are inverted. Pixels brighter than this threshold get flipped.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Threshold",
				group: "Effect"
			}
		},
		strength: {
			default: 1,
			description: "Blend between original (0) and fully solarized (1)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Strength",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Solarize shader.");
			return ZERO;
		}
		return call(solarizeColor, "solarizeColor", [
			childNode,
			uniforms.threshold,
			uniforms.strength
		]);
	}
};
var Solarize_default = componentDefinition;
export { componentDefinition as n, solarizeColor as r, Solarize_default as t };
