// shaders/dist/core/Spherize-bBe44dU7.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
const spherizeGeom = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, center, radius, depth, lightPos, lightIntensity, lightSoftness) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const centerX = center.x;
	const centerY = __tsover_sub(1, center.y);
	const cuX = __tsover_mul(__tsover_sub(uv.x, centerX), aspect);
	const cuY = __tsover_sub(uv.y, centerY);
	const sphereX = __tsover_div(__tsover_mul(cuX, 2), radius);
	const sphereY = __tsover_div(__tsover_mul(cuY, 2), radius);
	const radiusSq = __tsover_add(__tsover_mul(sphereX, sphereX), __tsover_mul(sphereY, sphereY));
	const sphereAlpha = __tsover_sub(1, std.smoothstep(.98, 1, radiusSq));
	const z = std.sqrt(std.max(d.f32(0), __tsover_sub(1, radiusSq)));
	const depthFactor = __tsover_add(1, __tsover_mul(z, depth));
	const finalX = __tsover_div(__tsover_mul(__tsover_div(sphereX, depthFactor), radius), 2);
	const finalY = __tsover_div(__tsover_mul(__tsover_div(sphereY, depthFactor), radius), 2);
	const transformedX = __tsover_add(__tsover_div(finalX, aspect), centerX);
	const transformedY = __tsover_add(finalY, centerY);
	const normal = std.normalize(d.vec3f(sphereX, sphereY, z));
	const viewDir = d.vec3f(0, 0, 1);
	const fresnelPower = __tsover_add(1, __tsover_mul(4, __tsover_sub(1, lightSoftness)));
	const fresnel = std.pow(__tsover_sub(1, std.max(std.dot(normal, viewDir), d.f32(0))), fresnelPower);
	const lightDir2Dx = __tsover_mul(__tsover_sub(lightPos.x, .5), 2);
	const lightDir2Dy = __tsover_mul(__tsover_sub(__tsover_sub(1, lightPos.y), .5), 2);
	const normalDir = std.normalize(d.vec2f(__tsover_add(normal.x, 1e-4), __tsover_add(normal.y, 1e-4)));
	const lightDirNorm = std.normalize(d.vec2f(__tsover_add(lightDir2Dx, 1e-4), __tsover_add(lightDir2Dy, 1e-4)));
	const directionality = std.max(d.f32(0), std.dot(normalDir, lightDirNorm));
	const directionalBias = std.pow(directionality, d.f32(2));
	const rimLight = __tsover_mul(__tsover_mul(__tsover_mul(fresnel, directionalBias), lightIntensity), 2);
	return d.vec4f(transformedX, transformedY, sphereAlpha, rimLight);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "radius"
			},
			{
				"type": "i",
				"name": "depth"
			},
			{
				"type": "i",
				"name": "lightPos"
			},
			{
				"type": "i",
				"name": "lightIntensity"
			},
			{
				"type": "i",
				"name": "lightSoftness"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"centerX",
				[
					7,
					"center",
					"x"
				]
			],
			[
				13,
				"centerY",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"center",
						"y"
					]
				]
			],
			[
				13,
				"cuX",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"-",
						"centerX"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"cuY",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					"centerY"
				]
			],
			[
				13,
				"sphereX",
				[
					1,
					[
						1,
						"cuX",
						"*",
						[5, "2"]
					],
					"/",
					"radius"
				]
			],
			[
				13,
				"sphereY",
				[
					1,
					[
						1,
						"cuY",
						"*",
						[5, "2"]
					],
					"/",
					"radius"
				]
			],
			[
				13,
				"radiusSq",
				[
					1,
					[
						1,
						"sphereX",
						"*",
						"sphereX"
					],
					"+",
					[
						1,
						"sphereY",
						"*",
						"sphereY"
					]
				]
			],
			[
				13,
				"sphereAlpha",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0.98"],
							[5, "1"],
							"radiusSq"
						]
					]
				]
			],
			[
				13,
				"z",
				[
					6,
					[
						7,
						"std",
						"sqrt"
					],
					[[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						], [
							1,
							[5, "1"],
							"-",
							"radiusSq"
						]]
					]]
				]
			],
			[
				13,
				"depthFactor",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						"z",
						"*",
						"depth"
					]
				]
			],
			[
				13,
				"finalX",
				[
					1,
					[
						1,
						[
							1,
							"sphereX",
							"/",
							"depthFactor"
						],
						"*",
						"radius"
					],
					"/",
					[5, "2"]
				]
			],
			[
				13,
				"finalY",
				[
					1,
					[
						1,
						[
							1,
							"sphereY",
							"/",
							"depthFactor"
						],
						"*",
						"radius"
					],
					"/",
					[5, "2"]
				]
			],
			[
				13,
				"transformedX",
				[
					1,
					[
						1,
						"finalX",
						"/",
						"aspect"
					],
					"+",
					"centerX"
				]
			],
			[
				13,
				"transformedY",
				[
					1,
					"finalY",
					"+",
					"centerY"
				]
			],
			[
				13,
				"normal",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"sphereX",
							"sphereY",
							"z"
						]
					]]
				]
			],
			[
				13,
				"viewDir",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"fresnelPower",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						[5, "4"],
						"*",
						[
							1,
							[5, "1"],
							"-",
							"lightSoftness"
						]
					]
				]
			],
			[
				13,
				"fresnel",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								6,
								[
									7,
									"std",
									"dot"
								],
								["normal", "viewDir"]
							], [
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0"]]
							]]
						]
					], "fresnelPower"]
				]
			],
			[
				13,
				"lightDir2Dx",
				[
					1,
					[
						1,
						[
							7,
							"lightPos",
							"x"
						],
						"-",
						[5, "0.5"]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"lightDir2Dy",
				[
					1,
					[
						1,
						[
							1,
							[5, "1"],
							"-",
							[
								7,
								"lightPos",
								"y"
							]
						],
						"-",
						[5, "0.5"]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"normalDir",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								7,
								"normal",
								"x"
							],
							"+",
							[5, "0.0001"]
						], [
							1,
							[
								7,
								"normal",
								"y"
							],
							"+",
							[5, "0.0001"]
						]]
					]]
				]
			],
			[
				13,
				"lightDirNorm",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							"lightDir2Dx",
							"+",
							[5, "0.0001"]
						], [
							1,
							"lightDir2Dy",
							"+",
							[5, "0.0001"]
						]]
					]]
				]
			],
			[
				13,
				"directionality",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "0"]]
					], [
						6,
						[
							7,
							"std",
							"dot"
						],
						["normalDir", "lightDirNorm"]
					]]
				]
			],
			[
				13,
				"directionalBias",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					["directionality", [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "2"]]
					]]
				]
			],
			[
				13,
				"rimLight",
				[
					1,
					[
						1,
						[
							1,
							"fresnel",
							"*",
							"directionalBias"
						],
						"*",
						"lightIntensity"
					],
					"*",
					[5, "2"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					"transformedX",
					"transformedY",
					"sphereAlpha",
					"rimLight"
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const spherizeCompose = tgpu.fn([
	d.vec4f,
	d.vec3f,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((straight, lightColor, rimLight, sphereAlpha) => {
	"use gpu";
	const r = __tsover_add(straight.x, __tsover_mul(lightColor.x, rimLight));
	const g = __tsover_add(straight.y, __tsover_mul(lightColor.y, rimLight));
	const b = __tsover_add(straight.z, __tsover_mul(lightColor.z, rimLight));
	const a = __tsover_mul(straight.w, sphereAlpha);
	return d.vec4f(r, g, b, a);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "straight"
			},
			{
				"type": "i",
				"name": "lightColor"
			},
			{
				"type": "i",
				"name": "rimLight"
			},
			{
				"type": "i",
				"name": "sphereAlpha"
			}
		],
		"body": [0, [
			[
				13,
				"r",
				[
					1,
					[
						7,
						"straight",
						"x"
					],
					"+",
					[
						1,
						[
							7,
							"lightColor",
							"x"
						],
						"*",
						"rimLight"
					]
				]
			],
			[
				13,
				"g",
				[
					1,
					[
						7,
						"straight",
						"y"
					],
					"+",
					[
						1,
						[
							7,
							"lightColor",
							"y"
						],
						"*",
						"rimLight"
					]
				]
			],
			[
				13,
				"b",
				[
					1,
					[
						7,
						"straight",
						"z"
					],
					"+",
					[
						1,
						[
							7,
							"lightColor",
							"z"
						],
						"*",
						"rimLight"
					]
				]
			],
			[
				13,
				"a",
				[
					1,
					[
						7,
						"straight",
						"w"
					],
					"*",
					"sphereAlpha"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					"r",
					"g",
					"b",
					"a"
				]
			]]
		]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const componentDefinition = {
	name: "Spherize",
	category: "Distortions",
	description: "Map content onto a 3D sphere surface with depth distortion",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		radius: {
			default: 1,
			description: "Radius of the sphere (1 = half viewport height)",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 3,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		depth: {
			default: 1,
			description: "How much the sphere bulges toward viewer (0 = flat, higher = more bulge)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 3,
				step: .1,
				label: "Depth",
				group: "Effect"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the sphere",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		lightPosition: {
			default: {
				x: .3,
				y: .3
			},
			transform: transformPosition,
			description: "Position of the specular light source",
			ui: {
				type: "position",
				label: "Light Position",
				group: "Position"
			}
		},
		lightIntensity: {
			default: .5,
			description: "Intensity of the rim light (0 = off)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Light Intensity",
				group: "Effect"
			}
		},
		lightSoftness: {
			default: .5,
			description: "Softness of the rim light falloff (0 = hard edge, 1 = soft glow)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Light Softness",
				group: "Effect"
			}
		},
		lightColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color of the specular highlight",
			ui: {
				type: "color",
				label: "Light Color",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { childNode, convertToTexture, uniforms, ctx } = params;
		if (!childNode) return ZERO;
		const childTexture = convertToTexture(childNode);
		const geom = call(spherizeGeom, "spherizeGeom", [
			ctx.uv,
			ctx.viewportSize,
			uniforms.center,
			uniforms.radius,
			uniforms.depth,
			uniforms.lightPosition,
			uniforms.lightIntensity,
			uniforms.lightSoftness
		]);
		return call(spherizeCompose, "spherizeCompose", [
			call(unpremultiplyAlpha, "unpremultiplyAlpha", [childTexture.sample(geom.member("xy"))]),
			uniforms.lightColor.member("rgb"),
			geom.member("w"),
			geom.member("z")
		]);
	}
};
var Spherize_default = componentDefinition;
export { spherizeGeom as i, componentDefinition as n, spherizeCompose as r, Spherize_default as t };
