// shaders/dist/core/Spiral-sBR7Fn0z.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
var TWO_PI = 6.283185307;
const spiralOffset = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.vec2f,
	d.vec2f
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, scale, animTime, uv, viewport) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const dx = __tsover_sub(__tsover_mul(uv.x, aspect), __tsover_mul(center.x, aspect));
	const dy = __tsover_sub(uv.y, __tsover_sub(1, center.y));
	const l = std.length(d.vec2f(dx, dy));
	const angle = __tsover_sub(std.atan2(dy, dx), animTime);
	return __tsover_add(__tsover_mul(l, scale), __tsover_div(angle, TWO_PI));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					],
					"-",
					[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					[
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]
				]
			],
			[
				13,
				"l",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						["dx", "dy"]
					]]
				]
			],
			[
				13,
				"angle",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"atan2"
						],
						["dy", "dx"]
					],
					"-",
					"animTime"
				]
			],
			[10, [
				1,
				[
					1,
					"l",
					"*",
					"scale"
				],
				"+",
				[
					1,
					"angle",
					"/",
					"TWO_PI"
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"TWO_PI"
		]
	},
	externals: () => ({
		std,
		d,
		TWO_PI
	})
}) && $.f)({}));
const spiralMask = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.vec2f
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, scale, strokeWidth, strokeFalloff, softness, animTime, uv, viewport) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const dx = __tsover_sub(__tsover_mul(uv.x, aspect), __tsover_mul(center.x, aspect));
	const dy = __tsover_sub(uv.y, __tsover_sub(1, center.y));
	const l = std.length(d.vec2f(dx, dy));
	const angle = __tsover_sub(std.atan2(dy, dx), animTime);
	const offset = __tsover_add(__tsover_mul(l, scale), __tsover_div(angle, TWO_PI));
	const stripe = std.fract(offset);
	const shape = __tsover_mul(std.abs(__tsover_sub(stripe, .5)), 2);
	const baseWidth = std.clamp(strokeWidth, __tsover_mul(strokeFalloff, .005), 1);
	const falloffAmount = __tsover_sub(1, __tsover_mul(std.clamp(strokeFalloff, 0, 1), l));
	const width = __tsover_mul(baseWidth, falloffAmount);
	const fw = std.fwidth(offset);
	const fwMult = __tsover_sub(4, __tsover_mul(__tsover_mul(std.smoothstep(.05, .4, __tsover_mul(strokeWidth, 2)), std.smoothstep(.05, .4, __tsover_mul(__tsover_sub(1, strokeWidth), 2))), 3));
	const pixelSize = std.mix(__tsover_mul(fwMult, fw), std.fwidth(shape), std.clamp(fw, 0, 1));
	return std.smoothstep(__tsover_sub(__tsover_sub(width, pixelSize), softness), __tsover_add(__tsover_add(width, pixelSize), softness), shape);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "strokeWidth"
			},
			{
				"type": "i",
				"name": "strokeFalloff"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					],
					"-",
					[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					[
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]
				]
			],
			[
				13,
				"l",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						["dx", "dy"]
					]]
				]
			],
			[
				13,
				"angle",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"atan2"
						],
						["dy", "dx"]
					],
					"-",
					"animTime"
				]
			],
			[
				13,
				"offset",
				[
					1,
					[
						1,
						"l",
						"*",
						"scale"
					],
					"+",
					[
						1,
						"angle",
						"/",
						"TWO_PI"
					]
				]
			],
			[
				13,
				"stripe",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["offset"]
				]
			],
			[
				13,
				"shape",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							1,
							"stripe",
							"-",
							[5, "0.5"]
						]]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"baseWidth",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"strokeWidth",
						[
							1,
							"strokeFalloff",
							"*",
							[5, "0.005"]
						],
						[5, "1"]
					]
				]
			],
			[
				13,
				"falloffAmount",
				[
					1,
					[5, "1"],
					"-",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								"strokeFalloff",
								[5, "0"],
								[5, "1"]
							]
						],
						"*",
						"l"
					]
				]
			],
			[
				13,
				"width",
				[
					1,
					"baseWidth",
					"*",
					"falloffAmount"
				]
			],
			[
				13,
				"fw",
				[
					6,
					[
						7,
						"std",
						"fwidth"
					],
					["offset"]
				]
			],
			[
				13,
				"fwMult",
				[
					1,
					[5, "4"],
					"-",
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									[5, "0.05"],
									[5, "0.4"],
									[
										1,
										"strokeWidth",
										"*",
										[5, "2"]
									]
								]
							],
							"*",
							[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									[5, "0.05"],
									[5, "0.4"],
									[
										1,
										[
											1,
											[5, "1"],
											"-",
											"strokeWidth"
										],
										"*",
										[5, "2"]
									]
								]
							]
						],
						"*",
						[5, "3"]
					]
				]
			],
			[
				13,
				"pixelSize",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							1,
							"fwMult",
							"*",
							"fw"
						],
						[
							6,
							[
								7,
								"std",
								"fwidth"
							],
							["shape"]
						],
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								"fw",
								[5, "0"],
								[5, "1"]
							]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"smoothstep"
				],
				[
					[
						1,
						[
							1,
							"width",
							"-",
							"pixelSize"
						],
						"-",
						"softness"
					],
					[
						1,
						[
							1,
							"width",
							"+",
							"pixelSize"
						],
						"+",
						"softness"
					],
					"shape"
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"TWO_PI"
		]
	},
	externals: () => ({
		std,
		d,
		TWO_PI
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Spiral",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Rotating spiral pattern with animated movement",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "Background color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Spiral stroke color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		strokeWidth: {
			default: .5,
			description: "Thickness of spiral stroke",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "Stroke Width",
				group: "Effect"
			}
		},
		strokeFalloff: {
			default: 0,
			description: "Stroke losing width further from center",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Stroke Falloff",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Color transition sharpness (0 = hard edge, 1 = smooth fade)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Softness",
				group: "Effect"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed (negative values reverse direction)",
			ui: {
				type: "range",
				min: -3,
				max: 3,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the spiral",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		scale: {
			default: 1,
			description: "Scale factor for spiral bands (higher = more bands, lower = fewer bands)",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 5,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const uv = params.uvContext ?? ctx.uv;
		const viewport = params.effectiveViewportSize ?? ctx.viewportSize;
		const t = animatedTime(params);
		const mask = call(spiralMask, "spiralMask", [
			uniforms.center,
			uniforms.scale,
			uniforms.strokeWidth,
			uniforms.strokeFalloff,
			uniforms.softness,
			t,
			uv,
			viewport
		]);
		return call(mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			mask
		]);
	}
};
var Spiral_default = componentDefinition;
export { spiralOffset as i, componentDefinition as n, spiralMask as r, Spiral_default as t };
