// shaders/dist/core/Strands-e7O0GIHI.js
import { At as call, It as vec4, Jt as hsvToRgb, R as std, Xt as lchToLab, Yt as labToRgb, ht as gradientStopsInSpace, mt as colorStops_exports, n as tgpu, nn as oklchToOklab, on as d, pt as MAX_COLOR_STOPS, qt as hslToRgb, rn as sRGBToP3, tn as oklabToRgb } from "./coords-BlwgBpJf.js";
import { c as transformPosition, i as transformBoolean, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig, t as animatedTime } from "./porters-CoaznI7i.js";
var TAU = Math.PI * 2;
var PI = Math.PI;
var MAX = 8;
var STRAND_MAX = 16;
var strandsTonePow = tgpu.fn([d.vec3f], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((rgb) => {
	"use gpu";
	return std.pow(rgb, d.vec3f(.85, .85, .85));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "rgb"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"pow"
			],
			["rgb", [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[5, "0.85"],
					[5, "0.85"],
					[5, "0.85"]
				]
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const strandsField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.arrayOf(d.vec4f, MAX),
	d.arrayOf(d.vec4f, MAX / 4),
	d.arrayOf(d.vec3f, MAX)
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, start, end, packed1, packed2, packed3, colors, positions, converted) => {
	"use gpu";
	const amplitude = packed1.x;
	const frequency = packed1.y;
	const lineCount = packed1.z;
	const lineWidth = packed1.w;
	const softnessP = packed2.x;
	const spread = packed2.y;
	const colorScale = packed2.z;
	const colorVariance = packed2.w;
	const pinOn = packed3.x > 0;
	const waveT = packed3.y;
	const colorT = packed3.z;
	const stopCount = packed3.w;
	const aspectRatio = __tsover_div(viewport.x, std.max(viewport.y, 1e-6));
	const startPos = d.vec2f(__tsover_mul(start.x, aspectRatio), __tsover_sub(1, start.y));
	const endPos = d.vec2f(__tsover_mul(end.x, aspectRatio), __tsover_sub(1, end.y));
	const aspectUV = d.vec2f(__tsover_mul(uv.x, aspectRatio), uv.y);
	const direction = d.vec2f(__tsover_sub(endPos.x, startPos.x), __tsover_sub(endPos.y, startPos.y));
	const len = std.max(std.length(direction), 1e-6);
	const normalizedDir = d.vec2f(__tsover_div(direction.x, len), __tsover_div(direction.y, len));
	const perpDir = d.vec2f(-normalizedDir.y, normalizedDir.x);
	const relativePos = d.vec2f(__tsover_sub(aspectUV.x, startPos.x), __tsover_sub(aspectUV.y, startPos.y));
	const progress = __tsover_div(std.dot(relativePos, normalizedDir), len);
	const perpDist = std.dot(relativePos, perpDir);
	const clampedProg = std.clamp(progress, 0, 1);
	const pinch = std.select(d.f32(1), std.sin(__tsover_mul(clampedProg, PI)), pinOn);
	const fade = __tsover_mul(std.smoothstep(0, .06, progress), __tsover_sub(1, std.smoothstep(.94, 1, progress)));
	const edgeFade = std.select(d.f32(1), fade, pinOn);
	const thickness = __tsover_mul(lineWidth, .07);
	const softness = __tsover_add(__tsover_mul(softnessP, .08), 8e-4);
	const denom = std.max(__tsover_sub(lineCount, 1), 1);
	let accumColor = d.vec3f(0, 0, 0);
	let accumWeight = d.f32(0);
	const waveBase = __tsover_mul(__tsover_mul(progress, TAU), frequency);
	const shimmerBase = __tsover_add(__tsover_mul(progress, 8), __tsover_mul(waveT, 2));
	const colorBase = __tsover_add(__tsover_mul(progress, colorScale), colorT);
	const n = d.i32(std.round(lineCount));
	for (let idx = 0; idx < n; idx++) {
		const fi = d.f32(idx);
		const norm = __tsover_div(fi, denom);
		const phase = __tsover_mul(fi, .9);
		const strandT = __tsover_mul(waveT, __tsover_add(__tsover_mul(norm, .6), .7));
		const wave = __tsover_add(std.sin(__tsover_add(__tsover_add(waveBase, strandT), phase)), __tsover_mul(std.sin(__tsover_add(__tsover_sub(__tsover_mul(waveBase, 2.1), __tsover_mul(strandT, .7)), phase)), .5));
		const displacement = __tsover_mul(__tsover_mul(wave, amplitude), .06);
		const baseline = __tsover_mul(__tsover_sub(norm, .5), spread);
		const centre = __tsover_mul(__tsover_add(baseline, displacement), pinch);
		const dd = std.abs(__tsover_sub(perpDist, centre));
		const line = __tsover_sub(1, std.smoothstep(thickness, __tsover_add(thickness, softness), dd));
		const shimmer = __tsover_add(__tsover_mul(std.sin(__tsover_add(shimmerBase, fi)), .15), .85);
		const cc = __tsover_add(colorBase, __tsover_mul(norm, colorVariance));
		const strandColor = gradientStopsInSpace(__tsover_sub(1, std.abs(__tsover_sub(__tsover_mul(std.fract(__tsover_mul(cc, .5)), 2), 1))), colors, positions, converted, stopCount).conv;
		const weight = __tsover_mul(line, shimmer);
		accumColor = accumColor.add(strandColor.mul(weight));
		accumWeight = __tsover_add(accumWeight, weight);
	}
	const workingColor = accumColor.div(std.max(accumWeight, 1e-4));
	const alpha = std.clamp(__tsover_mul(accumWeight, edgeFade), 0, 1);
	return d.vec4f(workingColor, alpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "start"
			},
			{
				"type": "i",
				"name": "end"
			},
			{
				"type": "i",
				"name": "packed1"
			},
			{
				"type": "i",
				"name": "packed2"
			},
			{
				"type": "i",
				"name": "packed3"
			},
			{
				"type": "i",
				"name": "colors"
			},
			{
				"type": "i",
				"name": "positions"
			},
			{
				"type": "i",
				"name": "converted"
			}
		],
		"body": [0, [
			[
				13,
				"amplitude",
				[
					7,
					"packed1",
					"x"
				]
			],
			[
				13,
				"frequency",
				[
					7,
					"packed1",
					"y"
				]
			],
			[
				13,
				"lineCount",
				[
					7,
					"packed1",
					"z"
				]
			],
			[
				13,
				"lineWidth",
				[
					7,
					"packed1",
					"w"
				]
			],
			[
				13,
				"softnessP",
				[
					7,
					"packed2",
					"x"
				]
			],
			[
				13,
				"spread",
				[
					7,
					"packed2",
					"y"
				]
			],
			[
				13,
				"colorScale",
				[
					7,
					"packed2",
					"z"
				]
			],
			[
				13,
				"colorVariance",
				[
					7,
					"packed2",
					"w"
				]
			],
			[
				13,
				"pinOn",
				[
					1,
					[
						7,
						"packed3",
						"x"
					],
					">",
					[5, "0"]
				]
			],
			[
				13,
				"waveT",
				[
					7,
					"packed3",
					"y"
				]
			],
			[
				13,
				"colorT",
				[
					7,
					"packed3",
					"z"
				]
			],
			[
				13,
				"stopCount",
				[
					7,
					"packed3",
					"w"
				]
			],
			[
				13,
				"aspectRatio",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							7,
							"viewport",
							"y"
						], [5, "0.000001"]]
					]
				]
			],
			[
				13,
				"startPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"start",
							"x"
						],
						"*",
						"aspectRatio"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"start",
							"y"
						]
					]]
				]
			],
			[
				13,
				"endPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"end",
							"x"
						],
						"*",
						"aspectRatio"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"end",
							"y"
						]
					]]
				]
			],
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspectRatio"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"direction",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"endPos",
							"x"
						],
						"-",
						[
							7,
							"startPos",
							"x"
						]
					], [
						1,
						[
							7,
							"endPos",
							"y"
						],
						"-",
						[
							7,
							"startPos",
							"y"
						]
					]]
				]
			],
			[
				13,
				"len",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"length"
						],
						["direction"]
					], [5, "0.000001"]]
				]
			],
			[
				13,
				"normalizedDir",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"direction",
							"x"
						],
						"/",
						"len"
					], [
						1,
						[
							7,
							"direction",
							"y"
						],
						"/",
						"len"
					]]
				]
			],
			[
				13,
				"perpDir",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						4,
						"-",
						[
							7,
							"normalizedDir",
							"y"
						]
					], [
						7,
						"normalizedDir",
						"x"
					]]
				]
			],
			[
				13,
				"relativePos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"aspectUV",
							"x"
						],
						"-",
						[
							7,
							"startPos",
							"x"
						]
					], [
						1,
						[
							7,
							"aspectUV",
							"y"
						],
						"-",
						[
							7,
							"startPos",
							"y"
						]
					]]
				]
			],
			[
				13,
				"progress",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"dot"
						],
						["relativePos", "normalizedDir"]
					],
					"/",
					"len"
				]
			],
			[
				13,
				"perpDist",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["relativePos", "perpDir"]
				]
			],
			[
				13,
				"clampedProg",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"progress",
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"pinch",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								"clampedProg",
								"*",
								"PI"
							]]
						],
						"pinOn"
					]
				]
			],
			[
				13,
				"fade",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							[5, "0.06"],
							"progress"
						]
					],
					"*",
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[5, "0.94"],
								[5, "1"],
								"progress"
							]
						]
					]
				]
			],
			[
				13,
				"edgeFade",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						"fade",
						"pinOn"
					]
				]
			],
			[
				13,
				"thickness",
				[
					1,
					"lineWidth",
					"*",
					[5, "0.07"]
				]
			],
			[
				13,
				"softness",
				[
					1,
					[
						1,
						"softnessP",
						"*",
						[5, "0.08"]
					],
					"+",
					[5, "0.0008"]
				]
			],
			[
				13,
				"denom",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"lineCount",
						"-",
						[5, "1"]
					], [5, "1"]]
				]
			],
			[
				12,
				"accumColor",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "0"]
					]
				]
			],
			[
				12,
				"accumWeight",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0"]]
				]
			],
			[
				13,
				"waveBase",
				[
					1,
					[
						1,
						"progress",
						"*",
						"TAU"
					],
					"*",
					"frequency"
				]
			],
			[
				13,
				"shimmerBase",
				[
					1,
					[
						1,
						"progress",
						"*",
						[5, "8"]
					],
					"+",
					[
						1,
						"waveT",
						"*",
						[5, "2"]
					]
				]
			],
			[
				13,
				"colorBase",
				[
					1,
					[
						1,
						"progress",
						"*",
						"colorScale"
					],
					"+",
					"colorT"
				]
			],
			[
				13,
				"n",
				[
					6,
					[
						7,
						"d",
						"i32"
					],
					[[
						6,
						[
							7,
							"std",
							"round"
						],
						["lineCount"]
					]]
				]
			],
			[
				14,
				[
					12,
					"idx",
					[5, "0"]
				],
				[
					1,
					"idx",
					"<",
					"n"
				],
				[
					102,
					"++",
					"idx"
				],
				[0, [
					[
						13,
						"fi",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["idx"]
						]
					],
					[
						13,
						"norm",
						[
							1,
							"fi",
							"/",
							"denom"
						]
					],
					[
						13,
						"phase",
						[
							1,
							"fi",
							"*",
							[5, "0.9"]
						]
					],
					[
						13,
						"strandT",
						[
							1,
							"waveT",
							"*",
							[
								1,
								[
									1,
									"norm",
									"*",
									[5, "0.6"]
								],
								"+",
								[5, "0.7"]
							]
						]
					],
					[
						13,
						"wave",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									[
										1,
										"waveBase",
										"+",
										"strandT"
									],
									"+",
									"phase"
								]]
							],
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										[
											1,
											[
												1,
												"waveBase",
												"*",
												[5, "2.1"]
											],
											"-",
											[
												1,
												"strandT",
												"*",
												[5, "0.7"]
											]
										],
										"+",
										"phase"
									]]
								],
								"*",
								[5, "0.5"]
							]
						]
					],
					[
						13,
						"displacement",
						[
							1,
							[
								1,
								"wave",
								"*",
								"amplitude"
							],
							"*",
							[5, "0.06"]
						]
					],
					[
						13,
						"baseline",
						[
							1,
							[
								1,
								"norm",
								"-",
								[5, "0.5"]
							],
							"*",
							"spread"
						]
					],
					[
						13,
						"centre",
						[
							1,
							[
								1,
								"baseline",
								"+",
								"displacement"
							],
							"*",
							"pinch"
						]
					],
					[
						13,
						"dd",
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								1,
								"perpDist",
								"-",
								"centre"
							]]
						]
					],
					[
						13,
						"line",
						[
							1,
							[5, "1"],
							"-",
							[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									"thickness",
									[
										1,
										"thickness",
										"+",
										"softness"
									],
									"dd"
								]
							]
						]
					],
					[
						13,
						"shimmer",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										"shimmerBase",
										"+",
										"fi"
									]]
								],
								"*",
								[5, "0.15"]
							],
							"+",
							[5, "0.85"]
						]
					],
					[
						13,
						"cc",
						[
							1,
							"colorBase",
							"+",
							[
								1,
								"norm",
								"*",
								"colorVariance"
							]
						]
					],
					[
						13,
						"colorCoord",
						[
							1,
							[5, "1"],
							"-",
							[
								6,
								[
									7,
									"std",
									"abs"
								],
								[[
									1,
									[
										1,
										[
											6,
											[
												7,
												"std",
												"fract"
											],
											[[
												1,
												"cc",
												"*",
												[5, "0.5"]
											]]
										],
										"*",
										[5, "2"]
									],
									"-",
									[5, "1"]
								]]
							]
						]
					],
					[
						13,
						"stepRes",
						[
							6,
							[
								7,
								"kitColorStops",
								"gradientStopsInSpace"
							],
							[
								"colorCoord",
								"colors",
								"positions",
								"converted",
								"stopCount"
							]
						]
					],
					[
						13,
						"strandColor",
						[
							7,
							"stepRes",
							"conv"
						]
					],
					[
						13,
						"weight",
						[
							1,
							"line",
							"*",
							"shimmer"
						]
					],
					[
						2,
						"accumColor",
						"=",
						[
							6,
							[
								7,
								"accumColor",
								"add"
							],
							[[
								6,
								[
									7,
									"strandColor",
									"mul"
								],
								["weight"]
							]]
						]
					],
					[
						2,
						"accumWeight",
						"=",
						[
							1,
							"accumWeight",
							"+",
							"weight"
						]
					]
				]]
			],
			[
				13,
				"workingColor",
				[
					6,
					[
						7,
						"accumColor",
						"div"
					],
					[[
						6,
						[
							7,
							"std",
							"max"
						],
						["accumWeight", [5, "0.0001"]]
					]]
				]
			],
			[
				13,
				"alpha",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"accumWeight",
							"*",
							"edgeFade"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				["workingColor", "alpha"]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"PI",
			"TAU",
			"kitColorStops"
		]
	},
	externals: () => ({
		std,
		d,
		PI,
		TAU,
		kitColorStops: colorStops_exports
	})
}) && $.f)({}));
function backConvert(conv, mode) {
	switch (mode) {
		case 1: return call(sRGBToP3, "sRGBToP3", [call(oklabToRgb, "oklabToRgb", [call(oklchToOklab, "oklchToOklab", [conv])])]);
		case 2: return call(sRGBToP3, "sRGBToP3", [call(oklabToRgb, "oklabToRgb", [conv])]);
		case 3: return call(sRGBToP3, "sRGBToP3", [call(hslToRgb, "hslToRgb", [conv])]);
		case 4: return call(sRGBToP3, "sRGBToP3", [call(hsvToRgb, "hsvToRgb", [conv])]);
		case 5: return call(sRGBToP3, "sRGBToP3", [call(labToRgb, "labToRgb", [call(lchToLab, "lchToLab", [conv])])]);
		default: return conv;
	}
}
const componentDefinition = {
	name: "Strands",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Flowing ribbons of light with a multi-color gradient",
	requiresRTT: false,
	requiresChild: false,
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	extraAnimatedTimes: { color: "colorSpeed" },
	props: {
		speed: {
			default: .5,
			description: "Overall animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		amplitude: {
			default: 2,
			description: "How far the strands wave away from their resting line",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .1,
				label: "Amplitude",
				group: "Effect"
			}
		},
		frequency: {
			default: .3,
			description: "How many waves run along each strand",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .1,
				label: "Frequency",
				group: "Effect"
			}
		},
		lineCount: {
			default: 8,
			description: "Number of strands",
			ui: {
				type: "range",
				min: 1,
				max: STRAND_MAX,
				step: 1,
				label: "Strand Count",
				group: "Effect"
			}
		},
		lineWidth: {
			default: .05,
			description: "Thickness of each strand’s solid core",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Strand Width",
				group: "Effect"
			}
		},
		softness: {
			default: .05,
			description: "Softness of each strand’s edge (0 is crisp, higher is feathered)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Strand Softness",
				group: "Effect"
			}
		},
		spread: {
			default: .2,
			description: "How far the strands fan out across the field",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Spread",
				group: "Effect"
			}
		},
		pinEdges: {
			default: false,
			transform: transformBoolean,
			description: "Pin the strands so they converge to the start and end points",
			ui: {
				type: "checkbox",
				label: "Pin Edges",
				group: "Effect"
			}
		},
		stops: {
			...colorStopsPropConfig(),
			default: [
				{
					color: "#00e5ff",
					position: 0
				},
				{
					color: "#5b8cff",
					position: .34
				},
				{
					color: "#b06bff",
					position: .67
				},
				{
					color: "#ff5fa2",
					position: 1
				}
			]
		},
		colorSpace: {
			default: "oklab",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space used to blend the gradient",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		colorScale: {
			default: 1,
			description: "How many times the gradient repeats along the strands",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 6,
				step: 1,
				label: "Color Scale",
				group: "Colors"
			}
		},
		colorVariance: {
			default: 1,
			description: "At 0 every strand shares the same color; at 1 the strands spread across the full gradient",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Color Variance",
				group: "Colors"
			}
		},
		colorSpeed: {
			default: 1,
			description: "Speed at which the colors flow along the strands (loops)",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .05,
				label: "Color Speed",
				group: "Colors"
			}
		},
		start: {
			default: {
				x: 0,
				y: .5
			},
			transform: transformPosition,
			description: "Starting point of the strands",
			ui: {
				type: "position",
				label: "Start",
				group: "Position"
			}
		},
		end: {
			default: {
				x: 1,
				y: .5
			},
			transform: transformPosition,
			description: "Ending point of the strands",
			ui: {
				type: "position",
				label: "End",
				group: "Position"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, propValues } = params;
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const waveT = animatedTime(params);
		const colorT = animatedTime(params, void 0, "_animTime_color");
		const packed1 = vec4(uniforms.amplitude, uniforms.frequency, uniforms.lineCount, uniforms.lineWidth);
		const packed2 = vec4(uniforms.softness, uniforms.spread, uniforms.colorScale, uniforms.colorVariance);
		const packed3 = vec4(uniforms.pinEdges, waveT, colorT, uniforms.stopCount);
		const field = call(strandsField, "strandsField", [
			uv,
			viewport,
			uniforms.start,
			uniforms.end,
			packed1,
			packed2,
			packed3,
			uniforms.colorsArray,
			uniforms.positionsArray,
			uniforms.convertedColorsArray
		]);
		const colorSpaceMode = propValues.colorSpace ?? 0;
		return vec4(call(strandsTonePow, "strandsTonePow", [backConvert(field.member("rgb"), colorSpaceMode)]), field.member("a"));
	}
};
var Strands_default = componentDefinition;
export { componentDefinition as n, strandsField as r, Strands_default as t };
