// shaders/dist/core/Stretch-CXFrCyH_.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-BlwgBpJf.js";
import { c as transformPosition, s as transformEdges } from "./transformations-fqzznt4g.js";
var DEG_TO_RAD = Math.PI / 180;
const stretchUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, strength, angle, falloff, uv, aspect) => {
	"use gpu";
	const aspectCorrectedUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y));
	const delta = aspectCorrectedUV.sub(centerPos);
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const directionVector = d.vec2f(std.cos(angleRad), std.sin(angleRad));
	const projection = std.dot(delta, directionVector);
	const parallel = directionVector.mul(projection);
	const perpendicular = delta.sub(parallel);
	const transitionWidth = std.mix(.001, 75, falloff);
	const effectMask = std.clamp(__tsover_div(projection, transitionWidth), 0, 1);
	const scaledStrength = __tsover_mul(strength, 100);
	const stretchFactor = __tsover_add(1, __tsover_mul(scaledStrength, effectMask));
	const finalProjection = __tsover_div(projection, stretchFactor);
	const stretchedDelta = directionVector.mul(finalProjection).add(perpendicular);
	const stretchedUV = centerPos.add(stretchedDelta);
	return d.vec2f(__tsover_div(stretchedUV.x, aspect), stretchedUV.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "falloff"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"aspectCorrectedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"delta",
				[
					6,
					[
						7,
						"aspectCorrectedUV",
						"sub"
					],
					["centerPos"]
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"directionVector",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						6,
						[
							7,
							"std",
							"cos"
						],
						["angleRad"]
					], [
						6,
						[
							7,
							"std",
							"sin"
						],
						["angleRad"]
					]]
				]
			],
			[
				13,
				"projection",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["delta", "directionVector"]
				]
			],
			[
				13,
				"parallel",
				[
					6,
					[
						7,
						"directionVector",
						"mul"
					],
					["projection"]
				]
			],
			[
				13,
				"perpendicular",
				[
					6,
					[
						7,
						"delta",
						"sub"
					],
					["parallel"]
				]
			],
			[
				13,
				"transitionWidth",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "0.001"],
						[5, "75"],
						"falloff"
					]
				]
			],
			[
				13,
				"effectMask",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"projection",
							"/",
							"transitionWidth"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"scaledStrength",
				[
					1,
					"strength",
					"*",
					[5, "100"]
				]
			],
			[
				13,
				"stretchFactor",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						"scaledStrength",
						"*",
						"effectMask"
					]
				]
			],
			[
				13,
				"finalProjection",
				[
					1,
					"projection",
					"/",
					"stretchFactor"
				]
			],
			[
				13,
				"finalParallel",
				[
					6,
					[
						7,
						"directionVector",
						"mul"
					],
					["finalProjection"]
				]
			],
			[
				13,
				"stretchedDelta",
				[
					6,
					[
						7,
						"finalParallel",
						"add"
					],
					["perpendicular"]
				]
			],
			[
				13,
				"stretchedUV",
				[
					6,
					[
						7,
						"centerPos",
						"add"
					],
					["stretchedDelta"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						7,
						"stretchedUV",
						"x"
					],
					"/",
					"aspect"
				], [
					7,
					"stretchedUV",
					"y"
				]]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Stretch",
	category: "Distortions",
	description: "Stretch content towards a direction from a center point",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the stretch effect",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		strength: {
			default: 1,
			description: "The intensity of the stretch effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Strength",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			description: "The direction of the stretch in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		falloff: {
			default: 0,
			description: "Controls the sharpness of the transition (0 = sharp edge, 1 = gradual transition)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Falloff",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, propValues, convertToTexture }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Stretch shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(stretchUV, "stretchUV", [
			uniforms.center,
			uniforms.strength,
			uniforms.angle,
			uniforms.falloff,
			ctx.uv,
			ctx.aspect
		]), (uv) => texture.sample(uv), edgeMode)]);
	},
	uvRemap: ({ uv, mask, uniforms, aspect, propValues }) => {
		const edgeMode = propValues.edges ?? 0;
		return composeEdgeRemapExpr(call(stretchUV, "stretchUV", [
			uniforms.center,
			uniforms.strength,
			uniforms.angle,
			uniforms.falloff,
			uv,
			aspect
		]), mask, edgeMode);
	}
};
var Stretch_default = componentDefinition;
export { componentDefinition as n, stretchUV as r, Stretch_default as t };
