// shaders/dist/core/Stripes-BVvuh6r7.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-Dg80Yg90.js";
var DEG_TO_RAD = .0174533;
const stripeCoordP = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, angle, density, animTime, offset) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, std.max(viewport.y, 1e-6));
	const acx = __tsover_mul(uv.x, aspect);
	const acy = uv.y;
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const cosA = std.cos(angleRad);
	const sinA = std.sin(angleRad);
	const rotatedCoord = __tsover_add(__tsover_mul(acx, cosA), __tsover_mul(acy, sinA));
	return __tsover_add(__tsover_add(__tsover_mul(rotatedCoord, density), animTime), offset);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "density"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "offset"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							7,
							"viewport",
							"y"
						], [5, "0.000001"]]
					]
				]
			],
			[
				13,
				"acx",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"acy",
				[
					7,
					"uv",
					"y"
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"rotatedCoord",
				[
					1,
					[
						1,
						"acx",
						"*",
						"cosA"
					],
					"+",
					[
						1,
						"acy",
						"*",
						"sinA"
					]
				]
			],
			[10, [
				1,
				[
					1,
					[
						1,
						"rotatedCoord",
						"*",
						"density"
					],
					"+",
					"animTime"
				],
				"+",
				"offset"
			]]
		]],
		"externalNames": ["std", "DEG_TO_RAD"]
	},
	externals: () => ({
		std,
		DEG_TO_RAD
	})
}) && $.f)({}));
const stripeMaskValue = tgpu.fn([
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p, w, balance) => {
	"use gpu";
	const a = __tsover_add(p, __tsover_mul(w, .5));
	const b = __tsover_sub(p, __tsover_mul(w, .5));
	const oneMinusBalance = __tsover_sub(1, balance);
	const Fa = __tsover_add(__tsover_mul(std.floor(a), oneMinusBalance), std.max(__tsover_sub(std.fract(a), balance), 0));
	const Fb = __tsover_add(__tsover_mul(std.floor(b), oneMinusBalance), std.max(__tsover_sub(std.fract(b), balance), 0));
	return std.clamp(__tsover_div(__tsover_sub(Fa, Fb), w), 0, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "p"
			},
			{
				"type": "i",
				"name": "w"
			},
			{
				"type": "i",
				"name": "balance"
			}
		],
		"body": [0, [
			[
				13,
				"a",
				[
					1,
					"p",
					"+",
					[
						1,
						"w",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"b",
				[
					1,
					"p",
					"-",
					[
						1,
						"w",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"oneMinusBalance",
				[
					1,
					[5, "1"],
					"-",
					"balance"
				]
			],
			[
				13,
				"Fa",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							["a"]
						],
						"*",
						"oneMinusBalance"
					],
					"+",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"fract"
								],
								["a"]
							],
							"-",
							"balance"
						], [5, "0"]]
					]
				]
			],
			[
				13,
				"Fb",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							["b"]
						],
						"*",
						"oneMinusBalance"
					],
					"+",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"fract"
								],
								["b"]
							],
							"-",
							"balance"
						], [5, "0"]]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"clamp"
				],
				[
					[
						1,
						[
							1,
							"Fa",
							"-",
							"Fb"
						],
						"/",
						"w"
					],
					[5, "0"],
					[5, "1"]
				]
			]]
		]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const stripesMask = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, angle, density, animTime, offset, balance, softness) => {
	"use gpu";
	const p = stripeCoordP(uv, viewport, angle, density, animTime, offset);
	const dpx = std.dpdx(p);
	const dpy = std.dpdy(p);
	const baseW = std.max(std.abs(dpx), std.abs(dpy));
	return stripeMaskValue(p, std.max(__tsover_add(baseW, softness), 1e-5), balance);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "density"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "offset"
			},
			{
				"type": "i",
				"name": "balance"
			},
			{
				"type": "i",
				"name": "softness"
			}
		],
		"body": [0, [
			[
				13,
				"p",
				[
					6,
					"stripeCoordP",
					[
						"uv",
						"viewport",
						"angle",
						"density",
						"animTime",
						"offset"
					]
				]
			],
			[
				13,
				"dpx",
				[
					6,
					[
						7,
						"std",
						"dpdx"
					],
					["p"]
				]
			],
			[
				13,
				"dpy",
				[
					6,
					[
						7,
						"std",
						"dpdy"
					],
					["p"]
				]
			],
			[
				13,
				"baseW",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"abs"
						],
						["dpx"]
					], [
						6,
						[
							7,
							"std",
							"abs"
						],
						["dpy"]
					]]
				]
			],
			[
				13,
				"w",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"baseW",
						"+",
						"softness"
					], [5, "0.00001"]]
				]
			],
			[10, [
				6,
				"stripeMaskValue",
				[
					"p",
					"w",
					"balance"
				]
			]]
		]],
		"externalNames": [
			"stripeCoordP",
			"std",
			"stripeMaskValue"
		]
	},
	externals: () => ({
		stripeCoordP,
		std,
		stripeMaskValue
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Stripes",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Alternating colored stripes with animation",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "First stripe color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Second stripe color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		angle: {
			default: 45,
			transform: transformAngle,
			description: "Angle of stripes in degrees",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		density: {
			default: 5,
			description: "Number of stripe pairs visible",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Density",
				group: "Effect"
			}
		},
		balance: {
			default: .5,
			description: "Ratio of the two colors",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Balance",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Edge softness",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Softness",
				group: "Effect"
			}
		},
		speed: {
			default: .2,
			description: "Animation speed",
			ui: {
				type: "range",
				min: -1,
				max: 1,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		offset: {
			default: 0,
			description: "Phase offset for pattern positioning",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Offset",
				group: "Animation"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		const mask = call(stripesMask, "stripesMask", [
			uv,
			viewport,
			params.uniforms.angle,
			params.uniforms.density,
			t,
			params.uniforms.offset,
			params.uniforms.balance,
			params.uniforms.softness
		]);
		return call(mixColorsVariants[params.propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			params.uniforms.colorA,
			params.uniforms.colorB,
			mask
		]);
	}
};
var Stripes_default = componentDefinition;
export { stripesMask as a, stripeMaskValue as i, componentDefinition as n, stripeCoordP as r, Stripes_default as t };
