// shaders/dist/core/StudioBackground-D7eIlsQA.js
import { At as call, It as vec4, R as std, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
var TAU = Math.PI * 2;
var hash01 = tgpu.fn([d.f32], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((x) => {
	"use gpu";
	return std.fract(__tsover_mul(std.sin(__tsover_mul(x, 12.9898)), 43758.5453));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "x"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				1,
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						1,
						"x",
						"*",
						[5, "12.9898"]
					]]
				],
				"*",
				[5, "43758.5453"]
			]]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
var studioDownSpot = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((sourceX, decay, uvx, aspect, centerX, wallNormY2, coneWidth, notFloor, wallFallY2, wallBias, floorDy2, isFloor, floorBias) => {
	"use gpu";
	const beamX = std.mix(sourceX, centerX, wallNormY2);
	const wDx = __tsover_div(__tsover_mul(__tsover_sub(uvx, beamX), aspect), coneWidth);
	const wD2 = __tsover_add(__tsover_mul(wDx, wDx), wallFallY2);
	const wall = __tsover_mul(__tsover_mul(std.exp(__tsover_mul(-wD2, decay)), notFloor), wallBias);
	const poolX = std.mix(centerX, sourceX, .25);
	const fDx = __tsover_mul(__tsover_mul(__tsover_sub(uvx, poolX), aspect), 1.4);
	const fD2 = __tsover_add(__tsover_mul(fDx, fDx), floorDy2);
	const floorV = __tsover_mul(__tsover_mul(std.exp(__tsover_mul(-fD2, __tsover_mul(decay, 1.5))), isFloor), floorBias);
	return __tsover_add(wall, floorV);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "sourceX"
			},
			{
				"type": "i",
				"name": "decay"
			},
			{
				"type": "i",
				"name": "uvx"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "centerX"
			},
			{
				"type": "i",
				"name": "wallNormY2"
			},
			{
				"type": "i",
				"name": "coneWidth"
			},
			{
				"type": "i",
				"name": "notFloor"
			},
			{
				"type": "i",
				"name": "wallFallY2"
			},
			{
				"type": "i",
				"name": "wallBias"
			},
			{
				"type": "i",
				"name": "floorDy2"
			},
			{
				"type": "i",
				"name": "isFloor"
			},
			{
				"type": "i",
				"name": "floorBias"
			}
		],
		"body": [0, [
			[
				13,
				"beamX",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"sourceX",
						"centerX",
						"wallNormY2"
					]
				]
			],
			[
				13,
				"wDx",
				[
					1,
					[
						1,
						[
							1,
							"uvx",
							"-",
							"beamX"
						],
						"*",
						"aspect"
					],
					"/",
					"coneWidth"
				]
			],
			[
				13,
				"wD2",
				[
					1,
					[
						1,
						"wDx",
						"*",
						"wDx"
					],
					"+",
					"wallFallY2"
				]
			],
			[
				13,
				"wall",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									4,
									"-",
									"wD2"
								],
								"*",
								"decay"
							]]
						],
						"*",
						"notFloor"
					],
					"*",
					"wallBias"
				]
			],
			[
				13,
				"poolX",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"centerX",
						"sourceX",
						[5, "0.25"]
					]
				]
			],
			[
				13,
				"fDx",
				[
					1,
					[
						1,
						[
							1,
							"uvx",
							"-",
							"poolX"
						],
						"*",
						"aspect"
					],
					"*",
					[5, "1.4"]
				]
			],
			[
				13,
				"fD2",
				[
					1,
					[
						1,
						"fDx",
						"*",
						"fDx"
					],
					"+",
					"floorDy2"
				]
			],
			[
				13,
				"floorV",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									4,
									"-",
									"fD2"
								],
								"*",
								[
									1,
									"decay",
									"*",
									[5, "1.5"]
								]
							]]
						],
						"*",
						"isFloor"
					],
					"*",
					"floorBias"
				]
			],
			[10, [
				1,
				"wall",
				"+",
				"floorV"
			]]
		]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const studioBackgroundColor = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, center, surfaceColor, keyColor, fillColor, backColor, lights1, lights2, scene, misc) => {
	"use gpu";
	const brightness = __tsover_mul(lights1.x, .02);
	const keyInt = __tsover_mul(lights1.y, .015);
	const keySoft01 = __tsover_mul(lights1.z, .01);
	const fillInt = __tsover_mul(lights1.w, .015);
	const fillSoft01 = __tsover_mul(lights2.x, .01);
	const fillOffset = __tsover_mul(lights2.y, .005);
	const backInt = __tsover_mul(lights2.z, .015);
	const backSoft01 = __tsover_mul(lights2.w, .01);
	const lightTarget = __tsover_mul(scene.x, .01);
	const curvature = __tsover_mul(scene.y, .01);
	const vignetteStrength = __tsover_mul(scene.z, .025);
	const ambIntensity = __tsover_mul(scene.w, .012);
	const seed = misc.x;
	const animTime = misc.y;
	const aspect = __tsover_div(viewport.x, viewport.y);
	const centerX = center.x;
	const centerY = __tsover_sub(1, center.y);
	const keyDecay = std.mix(d.f32(5), .8, keySoft01);
	const fillDecay = std.mix(d.f32(5), .8, fillSoft01);
	const backDecay = std.mix(d.f32(5), .8, backSoft01);
	const transitionWidth = std.mix(.015, .15, curvature);
	const isFloor = std.smoothstep(__tsover_sub(centerY, transitionWidth), __tsover_add(centerY, transitionWidth), uv.y);
	const notFloor = __tsover_sub(1, isFloor);
	const wallNormY = std.clamp(__tsover_div(uv.y, centerY), 0, 1);
	const wallNormY2 = __tsover_mul(wallNormY, wallNormY);
	const floorNormY = std.clamp(__tsover_div(__tsover_sub(uv.y, centerY), __tsover_sub(1, centerY)), 0, 1);
	const uvXCentered = __tsover_mul(__tsover_sub(uv.x, centerX), aspect);
	const coneWidth = std.mix(.05, .45, wallNormY2);
	const wallFallY = __tsover_mul(__tsover_sub(1, wallNormY), 1.2);
	const wallFallY2 = __tsover_mul(wallFallY, wallFallY);
	const floorPoolY = __tsover_add(centerY, __tsover_mul(__tsover_sub(1, centerY), .3));
	const floorDy = __tsover_mul(__tsover_sub(uv.y, floorPoolY), 4);
	const floorDy2 = __tsover_mul(floorDy, floorDy);
	const wallAmbient = std.mix(.2, .4, wallNormY);
	const floorAmbient = std.mix(.38, .18, floorNormY);
	const ambient = __tsover_mul(std.mix(wallAmbient, floorAmbient, isFloor), brightness);
	const curveDist = __tsover_div(__tsover_sub(uv.y, centerY), __tsover_add(transitionWidth, .001));
	const curveGlow = __tsover_mul(__tsover_mul(__tsover_mul(std.exp(__tsover_mul(__tsover_mul(-.5, curveDist), curveDist)), curvature), brightness), .1);
	const neutralBase = __tsover_add(ambient, curveGlow);
	let lightAccum = d.vec3f(neutralBase, neutralBase, neutralBase);
	const wallBias = __tsover_add(__tsover_mul(__tsover_sub(1, lightTarget), 1.4), .3);
	const floorBias = __tsover_add(__tsover_mul(lightTarget, 1.4), .3);
	const keyPat = __tsover_mul(studioDownSpot(centerX, keyDecay, uv.x, aspect, centerX, wallNormY2, coneWidth, notFloor, wallFallY2, wallBias, floorDy2, isFloor, floorBias), keyInt);
	lightAccum = lightAccum.add(keyColor.xyz.mul(keyPat));
	const leftPat = __tsover_mul(studioDownSpot(__tsover_sub(centerX, fillOffset), fillDecay, uv.x, aspect, centerX, wallNormY2, coneWidth, notFloor, wallFallY2, wallBias, floorDy2, isFloor, floorBias), fillInt);
	const rightPat = __tsover_mul(studioDownSpot(__tsover_add(centerX, fillOffset), fillDecay, uv.x, aspect, centerX, wallNormY2, coneWidth, notFloor, wallFallY2, wallBias, floorDy2, isFloor, floorBias), fillInt);
	lightAccum = lightAccum.add(fillColor.xyz.mul(__tsover_add(leftPat, rightPat)));
	const backBottomY = std.mix(d.f32(1), centerY, .15);
	const bfDx = __tsover_mul(uvXCentered, .5);
	const bfDy = __tsover_mul(__tsover_sub(uv.y, backBottomY), 4);
	const bfD2 = __tsover_add(__tsover_mul(bfDx, bfDx), __tsover_mul(bfDy, bfDy));
	const backFloor = __tsover_mul(std.exp(__tsover_mul(-backDecay, bfD2)), isFloor);
	const bwDx = __tsover_mul(uvXCentered, .6);
	const bwDy = __tsover_mul(__tsover_sub(centerY, uv.y), 3);
	const bwD2 = __tsover_add(__tsover_mul(bwDx, bwDx), __tsover_mul(bwDy, bwDy));
	const backWall = __tsover_mul(__tsover_mul(std.exp(__tsover_mul(-__tsover_mul(backDecay, .8), bwD2)), notFloor), wallNormY2);
	const backPat = __tsover_mul(__tsover_add(backFloor, backWall), backInt);
	lightAccum = lightAccum.add(backColor.xyz.mul(backPat));
	const bnDx = __tsover_mul(uvXCentered, .7);
	const bnDy = __tsover_mul(__tsover_add(__tsover_sub(uv.y, centerY), .02), 10);
	const bnD2 = __tsover_add(__tsover_mul(bnDx, bnDx), __tsover_mul(bnDy, bnDy));
	const bounce = __tsover_mul(__tsover_mul(std.exp(__tsover_mul(-2.5, bnD2)), __tsover_mul(keyInt, .12)), notFloor);
	lightAccum = lightAccum.add(d.vec3f(bounce, bounce, bounce));
	const ambGlowScale = __tsover_mul(ambIntensity, .25);
	for (let i = 0; i < 4; i++) {
		const s = __tsover_add(seed, __tsover_mul(d.f32(i), 73.1));
		const sizeVal = __tsover_add(__tsover_mul(hash01(__tsover_add(s, 311)), .12), .1);
		const radiusX = __tsover_add(__tsover_mul(hash01(s), .5), .2);
		const radiusY = __tsover_add(__tsover_mul(hash01(__tsover_add(s, 37)), .25), .1);
		const freqX = __tsover_add(__tsover_mul(hash01(__tsover_add(s, 91)), .3), .1);
		const freqY = __tsover_add(__tsover_mul(hash01(__tsover_add(s, 143)), .25), .08);
		const phaseX = __tsover_mul(hash01(__tsover_add(s, 200)), TAU);
		const phaseY = __tsover_mul(hash01(__tsover_add(s, 257)), TAU);
		const sizeSq = __tsover_mul(sizeVal, sizeVal);
		const lx = __tsover_add(centerX, __tsover_mul(std.sin(__tsover_add(__tsover_mul(animTime, freqX), phaseX)), radiusX));
		const ly = __tsover_add(__tsover_sub(centerY, .2), __tsover_mul(std.cos(__tsover_add(__tsover_mul(animTime, freqY), phaseY)), radiusY));
		const dx = __tsover_mul(__tsover_sub(uv.x, lx), aspect);
		const dy = __tsover_sub(uv.y, ly);
		const dd2 = __tsover_add(__tsover_mul(dx, dx), __tsover_mul(dy, dy));
		const glow = __tsover_mul(std.exp(__tsover_div(__tsover_mul(-.5, dd2), sizeSq)), ambGlowScale);
		lightAccum = lightAccum.add(d.vec3f(glow, glow, glow));
	}
	const vigX = __tsover_mul(__tsover_sub(uv.x, .5), aspect);
	const vigY = __tsover_sub(uv.y, .5);
	const vigD2 = __tsover_add(__tsover_mul(vigX, vigX), __tsover_mul(vigY, vigY));
	const vig = std.clamp(__tsover_sub(1, __tsover_mul(vigD2, vignetteStrength)), 0, 1);
	const clamped = std.clamp(lightAccum, d.vec3f(0), d.vec3f(3));
	const lit = std.clamp(surfaceColor.xyz.mul(clamped).mul(vig), d.vec3f(0), d.vec3f(1));
	return d.vec4f(lit, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "surfaceColor"
			},
			{
				"type": "i",
				"name": "keyColor"
			},
			{
				"type": "i",
				"name": "fillColor"
			},
			{
				"type": "i",
				"name": "backColor"
			},
			{
				"type": "i",
				"name": "lights1"
			},
			{
				"type": "i",
				"name": "lights2"
			},
			{
				"type": "i",
				"name": "scene"
			},
			{
				"type": "i",
				"name": "misc"
			}
		],
		"body": [0, [
			[
				13,
				"brightness",
				[
					1,
					[
						7,
						"lights1",
						"x"
					],
					"*",
					[5, "0.02"]
				]
			],
			[
				13,
				"keyInt",
				[
					1,
					[
						7,
						"lights1",
						"y"
					],
					"*",
					[5, "0.015"]
				]
			],
			[
				13,
				"keySoft01",
				[
					1,
					[
						7,
						"lights1",
						"z"
					],
					"*",
					[5, "0.01"]
				]
			],
			[
				13,
				"fillInt",
				[
					1,
					[
						7,
						"lights1",
						"w"
					],
					"*",
					[5, "0.015"]
				]
			],
			[
				13,
				"fillSoft01",
				[
					1,
					[
						7,
						"lights2",
						"x"
					],
					"*",
					[5, "0.01"]
				]
			],
			[
				13,
				"fillOffset",
				[
					1,
					[
						7,
						"lights2",
						"y"
					],
					"*",
					[5, "0.005"]
				]
			],
			[
				13,
				"backInt",
				[
					1,
					[
						7,
						"lights2",
						"z"
					],
					"*",
					[5, "0.015"]
				]
			],
			[
				13,
				"backSoft01",
				[
					1,
					[
						7,
						"lights2",
						"w"
					],
					"*",
					[5, "0.01"]
				]
			],
			[
				13,
				"lightTarget",
				[
					1,
					[
						7,
						"scene",
						"x"
					],
					"*",
					[5, "0.01"]
				]
			],
			[
				13,
				"curvature",
				[
					1,
					[
						7,
						"scene",
						"y"
					],
					"*",
					[5, "0.01"]
				]
			],
			[
				13,
				"vignetteStrength",
				[
					1,
					[
						7,
						"scene",
						"z"
					],
					"*",
					[5, "0.025"]
				]
			],
			[
				13,
				"ambIntensity",
				[
					1,
					[
						7,
						"scene",
						"w"
					],
					"*",
					[5, "0.012"]
				]
			],
			[
				13,
				"seed",
				[
					7,
					"misc",
					"x"
				]
			],
			[
				13,
				"animTime",
				[
					7,
					"misc",
					"y"
				]
			],
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"centerX",
				[
					7,
					"center",
					"x"
				]
			],
			[
				13,
				"centerY",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"center",
						"y"
					]
				]
			],
			[
				13,
				"keyDecay",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "5"]]
						],
						[5, "0.8"],
						"keySoft01"
					]
				]
			],
			[
				13,
				"fillDecay",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "5"]]
						],
						[5, "0.8"],
						"fillSoft01"
					]
				]
			],
			[
				13,
				"backDecay",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "5"]]
						],
						[5, "0.8"],
						"backSoft01"
					]
				]
			],
			[
				13,
				"transitionWidth",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "0.015"],
						[5, "0.15"],
						"curvature"
					]
				]
			],
			[
				13,
				"isFloor",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"centerY",
							"-",
							"transitionWidth"
						],
						[
							1,
							"centerY",
							"+",
							"transitionWidth"
						],
						[
							7,
							"uv",
							"y"
						]
					]
				]
			],
			[
				13,
				"notFloor",
				[
					1,
					[5, "1"],
					"-",
					"isFloor"
				]
			],
			[
				13,
				"wallNormY",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								7,
								"uv",
								"y"
							],
							"/",
							"centerY"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"wallNormY2",
				[
					1,
					"wallNormY",
					"*",
					"wallNormY"
				]
			],
			[
				13,
				"floorNormY",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"uv",
									"y"
								],
								"-",
								"centerY"
							],
							"/",
							[
								1,
								[5, "1"],
								"-",
								"centerY"
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"uvXCentered",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"-",
						"centerX"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"coneWidth",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "0.05"],
						[5, "0.45"],
						"wallNormY2"
					]
				]
			],
			[
				13,
				"wallFallY",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						"wallNormY"
					],
					"*",
					[5, "1.2"]
				]
			],
			[
				13,
				"wallFallY2",
				[
					1,
					"wallFallY",
					"*",
					"wallFallY"
				]
			],
			[
				13,
				"floorPoolY",
				[
					1,
					"centerY",
					"+",
					[
						1,
						[
							1,
							[5, "1"],
							"-",
							"centerY"
						],
						"*",
						[5, "0.3"]
					]
				]
			],
			[
				13,
				"floorDy",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"y"
						],
						"-",
						"floorPoolY"
					],
					"*",
					[5, "4"]
				]
			],
			[
				13,
				"floorDy2",
				[
					1,
					"floorDy",
					"*",
					"floorDy"
				]
			],
			[
				13,
				"wallAmbient",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "0.2"],
						[5, "0.4"],
						"wallNormY"
					]
				]
			],
			[
				13,
				"floorAmbient",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "0.38"],
						[5, "0.18"],
						"floorNormY"
					]
				]
			],
			[
				13,
				"ambient",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"wallAmbient",
							"floorAmbient",
							"isFloor"
						]
					],
					"*",
					"brightness"
				]
			],
			[
				13,
				"curveDist",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"y"
						],
						"-",
						"centerY"
					],
					"/",
					[
						1,
						"transitionWidth",
						"+",
						[5, "0.001"]
					]
				]
			],
			[
				13,
				"curveGlow",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"exp"
								],
								[[
									1,
									[
										1,
										[
											4,
											"-",
											[5, "0.5"]
										],
										"*",
										"curveDist"
									],
									"*",
									"curveDist"
								]]
							],
							"*",
							"curvature"
						],
						"*",
						"brightness"
					],
					"*",
					[5, "0.1"]
				]
			],
			[
				13,
				"neutralBase",
				[
					1,
					"ambient",
					"+",
					"curveGlow"
				]
			],
			[
				12,
				"lightAccum",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						"neutralBase",
						"neutralBase",
						"neutralBase"
					]
				]
			],
			[
				13,
				"wallBias",
				[
					1,
					[
						1,
						[
							1,
							[5, "1"],
							"-",
							"lightTarget"
						],
						"*",
						[5, "1.4"]
					],
					"+",
					[5, "0.3"]
				]
			],
			[
				13,
				"floorBias",
				[
					1,
					[
						1,
						"lightTarget",
						"*",
						[5, "1.4"]
					],
					"+",
					[5, "0.3"]
				]
			],
			[
				13,
				"keyPat",
				[
					1,
					[
						6,
						"studioDownSpot",
						[
							"centerX",
							"keyDecay",
							[
								7,
								"uv",
								"x"
							],
							"aspect",
							"centerX",
							"wallNormY2",
							"coneWidth",
							"notFloor",
							"wallFallY2",
							"wallBias",
							"floorDy2",
							"isFloor",
							"floorBias"
						]
					],
					"*",
					"keyInt"
				]
			],
			[
				2,
				"lightAccum",
				"=",
				[
					6,
					[
						7,
						"lightAccum",
						"add"
					],
					[[
						6,
						[
							7,
							[
								7,
								"keyColor",
								"xyz"
							],
							"mul"
						],
						["keyPat"]
					]]
				]
			],
			[
				13,
				"leftPat",
				[
					1,
					[
						6,
						"studioDownSpot",
						[
							[
								1,
								"centerX",
								"-",
								"fillOffset"
							],
							"fillDecay",
							[
								7,
								"uv",
								"x"
							],
							"aspect",
							"centerX",
							"wallNormY2",
							"coneWidth",
							"notFloor",
							"wallFallY2",
							"wallBias",
							"floorDy2",
							"isFloor",
							"floorBias"
						]
					],
					"*",
					"fillInt"
				]
			],
			[
				13,
				"rightPat",
				[
					1,
					[
						6,
						"studioDownSpot",
						[
							[
								1,
								"centerX",
								"+",
								"fillOffset"
							],
							"fillDecay",
							[
								7,
								"uv",
								"x"
							],
							"aspect",
							"centerX",
							"wallNormY2",
							"coneWidth",
							"notFloor",
							"wallFallY2",
							"wallBias",
							"floorDy2",
							"isFloor",
							"floorBias"
						]
					],
					"*",
					"fillInt"
				]
			],
			[
				2,
				"lightAccum",
				"=",
				[
					6,
					[
						7,
						"lightAccum",
						"add"
					],
					[[
						6,
						[
							7,
							[
								7,
								"fillColor",
								"xyz"
							],
							"mul"
						],
						[[
							1,
							"leftPat",
							"+",
							"rightPat"
						]]
					]]
				]
			],
			[
				13,
				"backBottomY",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						"centerY",
						[5, "0.15"]
					]
				]
			],
			[
				13,
				"bfDx",
				[
					1,
					"uvXCentered",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"bfDy",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"y"
						],
						"-",
						"backBottomY"
					],
					"*",
					[5, "4"]
				]
			],
			[
				13,
				"bfD2",
				[
					1,
					[
						1,
						"bfDx",
						"*",
						"bfDx"
					],
					"+",
					[
						1,
						"bfDy",
						"*",
						"bfDy"
					]
				]
			],
			[
				13,
				"backFloor",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"exp"
						],
						[[
							1,
							[
								4,
								"-",
								"backDecay"
							],
							"*",
							"bfD2"
						]]
					],
					"*",
					"isFloor"
				]
			],
			[
				13,
				"bwDx",
				[
					1,
					"uvXCentered",
					"*",
					[5, "0.6"]
				]
			],
			[
				13,
				"bwDy",
				[
					1,
					[
						1,
						"centerY",
						"-",
						[
							7,
							"uv",
							"y"
						]
					],
					"*",
					[5, "3"]
				]
			],
			[
				13,
				"bwD2",
				[
					1,
					[
						1,
						"bwDx",
						"*",
						"bwDx"
					],
					"+",
					[
						1,
						"bwDy",
						"*",
						"bwDy"
					]
				]
			],
			[
				13,
				"backWall",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									4,
									"-",
									[
										1,
										"backDecay",
										"*",
										[5, "0.8"]
									]
								],
								"*",
								"bwD2"
							]]
						],
						"*",
						"notFloor"
					],
					"*",
					"wallNormY2"
				]
			],
			[
				13,
				"backPat",
				[
					1,
					[
						1,
						"backFloor",
						"+",
						"backWall"
					],
					"*",
					"backInt"
				]
			],
			[
				2,
				"lightAccum",
				"=",
				[
					6,
					[
						7,
						"lightAccum",
						"add"
					],
					[[
						6,
						[
							7,
							[
								7,
								"backColor",
								"xyz"
							],
							"mul"
						],
						["backPat"]
					]]
				]
			],
			[
				13,
				"bnDx",
				[
					1,
					"uvXCentered",
					"*",
					[5, "0.7"]
				]
			],
			[
				13,
				"bnDy",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"uv",
								"y"
							],
							"-",
							"centerY"
						],
						"+",
						[5, "0.02"]
					],
					"*",
					[5, "10"]
				]
			],
			[
				13,
				"bnD2",
				[
					1,
					[
						1,
						"bnDx",
						"*",
						"bnDx"
					],
					"+",
					[
						1,
						"bnDy",
						"*",
						"bnDy"
					]
				]
			],
			[
				13,
				"bounce",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									4,
									"-",
									[5, "2.5"]
								],
								"*",
								"bnD2"
							]]
						],
						"*",
						[
							1,
							"keyInt",
							"*",
							[5, "0.12"]
						]
					],
					"*",
					"notFloor"
				]
			],
			[
				2,
				"lightAccum",
				"=",
				[
					6,
					[
						7,
						"lightAccum",
						"add"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"bounce",
							"bounce",
							"bounce"
						]
					]]
				]
			],
			[
				13,
				"ambGlowScale",
				[
					1,
					"ambIntensity",
					"*",
					[5, "0.25"]
				]
			],
			[
				14,
				[
					12,
					"i",
					[5, "0"]
				],
				[
					1,
					"i",
					"<",
					[5, "4"]
				],
				[
					102,
					"++",
					"i"
				],
				[0, [
					[
						13,
						"s",
						[
							1,
							"seed",
							"+",
							[
								1,
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									["i"]
								],
								"*",
								[5, "73.1"]
							]
						]
					],
					[
						13,
						"sizeVal",
						[
							1,
							[
								1,
								[
									6,
									"hash01",
									[[
										1,
										"s",
										"+",
										[5, "311"]
									]]
								],
								"*",
								[5, "0.12"]
							],
							"+",
							[5, "0.1"]
						]
					],
					[
						13,
						"radiusX",
						[
							1,
							[
								1,
								[
									6,
									"hash01",
									["s"]
								],
								"*",
								[5, "0.5"]
							],
							"+",
							[5, "0.2"]
						]
					],
					[
						13,
						"radiusY",
						[
							1,
							[
								1,
								[
									6,
									"hash01",
									[[
										1,
										"s",
										"+",
										[5, "37"]
									]]
								],
								"*",
								[5, "0.25"]
							],
							"+",
							[5, "0.1"]
						]
					],
					[
						13,
						"freqX",
						[
							1,
							[
								1,
								[
									6,
									"hash01",
									[[
										1,
										"s",
										"+",
										[5, "91"]
									]]
								],
								"*",
								[5, "0.3"]
							],
							"+",
							[5, "0.1"]
						]
					],
					[
						13,
						"freqY",
						[
							1,
							[
								1,
								[
									6,
									"hash01",
									[[
										1,
										"s",
										"+",
										[5, "143"]
									]]
								],
								"*",
								[5, "0.25"]
							],
							"+",
							[5, "0.08"]
						]
					],
					[
						13,
						"phaseX",
						[
							1,
							[
								6,
								"hash01",
								[[
									1,
									"s",
									"+",
									[5, "200"]
								]]
							],
							"*",
							"TAU"
						]
					],
					[
						13,
						"phaseY",
						[
							1,
							[
								6,
								"hash01",
								[[
									1,
									"s",
									"+",
									[5, "257"]
								]]
							],
							"*",
							"TAU"
						]
					],
					[
						13,
						"sizeSq",
						[
							1,
							"sizeVal",
							"*",
							"sizeVal"
						]
					],
					[
						13,
						"lx",
						[
							1,
							"centerX",
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										[
											1,
											"animTime",
											"*",
											"freqX"
										],
										"+",
										"phaseX"
									]]
								],
								"*",
								"radiusX"
							]
						]
					],
					[
						13,
						"ly",
						[
							1,
							[
								1,
								"centerY",
								"-",
								[5, "0.2"]
							],
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"cos"
									],
									[[
										1,
										[
											1,
											"animTime",
											"*",
											"freqY"
										],
										"+",
										"phaseY"
									]]
								],
								"*",
								"radiusY"
							]
						]
					],
					[
						13,
						"dx",
						[
							1,
							[
								1,
								[
									7,
									"uv",
									"x"
								],
								"-",
								"lx"
							],
							"*",
							"aspect"
						]
					],
					[
						13,
						"dy",
						[
							1,
							[
								7,
								"uv",
								"y"
							],
							"-",
							"ly"
						]
					],
					[
						13,
						"dd2",
						[
							1,
							[
								1,
								"dx",
								"*",
								"dx"
							],
							"+",
							[
								1,
								"dy",
								"*",
								"dy"
							]
						]
					],
					[
						13,
						"glow",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"exp"
								],
								[[
									1,
									[
										1,
										[
											4,
											"-",
											[5, "0.5"]
										],
										"*",
										"dd2"
									],
									"/",
									"sizeSq"
								]]
							],
							"*",
							"ambGlowScale"
						]
					],
					[
						2,
						"lightAccum",
						"=",
						[
							6,
							[
								7,
								"lightAccum",
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									"glow",
									"glow",
									"glow"
								]
							]]
						]
					]
				]]
			],
			[
				13,
				"vigX",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"-",
						[5, "0.5"]
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"vigY",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"vigD2",
				[
					1,
					[
						1,
						"vigX",
						"*",
						"vigX"
					],
					"+",
					[
						1,
						"vigY",
						"*",
						"vigY"
					]
				]
			],
			[
				13,
				"vig",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[5, "1"],
							"-",
							[
								1,
								"vigD2",
								"*",
								"vignetteStrength"
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"clamped",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"lightAccum",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[5, "3"]]
						]
					]
				]
			],
			[
				13,
				"lit",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										[
											7,
											"surfaceColor",
											"xyz"
										],
										"mul"
									],
									["clamped"]
								],
								"mul"
							],
							["vig"]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[5, "1"]]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				["lit", [5, "1"]]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"studioDownSpot",
			"hash01",
			"TAU"
		]
	},
	externals: () => ({
		std,
		d,
		studioDownSpot,
		hash01,
		TAU
	})
}) && $.f)({}));
const componentDefinition = {
	name: "StudioBackground",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Multi-light studio background with ambient motion.",
	animatedTime: { speed: "ambientSpeed" },
	props: {
		color: {
			default: "#d8dbec",
			transform: transformColor,
			description: "Base studio surface color",
			ui: {
				type: "color",
				label: "Surface Color",
				group: "Colors"
			}
		},
		keyColor: {
			default: "#d5e4ea",
			transform: transformColor,
			description: "Color of the overhead key light",
			ui: {
				type: "color",
				label: "Key Color",
				group: "Key Light"
			}
		},
		keyIntensity: {
			default: 40,
			description: "Intensity of the key light",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Key Light"
			}
		},
		keySoftness: {
			default: 50,
			description: "How diffuse the key light is",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Softness",
				group: "Key Light"
			}
		},
		fillColor: {
			default: "#d5e4ea",
			transform: transformColor,
			description: "Color of the side fill lights",
			ui: {
				type: "color",
				label: "Fill Color",
				group: "Fill Lights"
			}
		},
		fillIntensity: {
			default: 10,
			description: "Intensity of the fill lights",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Fill Lights"
			}
		},
		fillSoftness: {
			default: 70,
			description: "How diffuse the fill lights are",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Softness",
				group: "Fill Lights"
			}
		},
		fillAngle: {
			default: 70,
			description: "How far apart the fill lights are from center",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Spread",
				group: "Fill Lights"
			}
		},
		backColor: {
			default: "#c8d4e8",
			transform: transformColor,
			description: "Color of the upward back wash",
			ui: {
				type: "color",
				label: "Back Color",
				group: "Back Light"
			}
		},
		backIntensity: {
			default: 20,
			description: "Intensity of the back wash",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Back Light"
			}
		},
		backSoftness: {
			default: 80,
			description: "How diffuse the back wash is",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Softness",
				group: "Back Light"
			}
		},
		brightness: {
			default: 20,
			description: "Overall ambient light level",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Ambient",
				group: "Lighting"
			}
		},
		vignette: {
			default: 0,
			description: "Edge darkening",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Vignette",
				group: "Lighting"
			}
		},
		center: {
			default: {
				x: .5,
				y: .8
			},
			transform: transformPosition,
			description: "Where the spotlight meets the floor",
			ui: {
				type: "position",
				label: "Stage Center",
				group: "Scene"
			}
		},
		lightTarget: {
			default: 100,
			description: "How far toward the floor vs wall the spotlights aim",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Light Depth",
				group: "Scene"
			}
		},
		wallCurvature: {
			default: 10,
			description: "How rounded the cove is",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Wall Curvature",
				group: "Scene"
			}
		},
		ambientIntensity: {
			default: 50,
			description: "Intensity of drifting ambient lights",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Ambient Detail",
				group: "Environment"
			}
		},
		ambientSpeed: {
			default: 2,
			description: "Drift speed",
			ui: {
				type: "range",
				min: -5,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Environment"
			}
		},
		seed: {
			default: 0,
			description: "Seed for ambient pattern",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Environment"
			}
		}
	},
	fragment: (params) => {
		const { uniforms } = params;
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		const lights1 = vec4(uniforms.brightness, uniforms.keyIntensity, uniforms.keySoftness, uniforms.fillIntensity);
		const lights2 = vec4(uniforms.fillSoftness, uniforms.fillAngle, uniforms.backIntensity, uniforms.backSoftness);
		const scene = vec4(uniforms.lightTarget, uniforms.wallCurvature, uniforms.vignette, uniforms.ambientIntensity);
		const misc = vec4(uniforms.seed, t, 0, 0);
		return call(studioBackgroundColor, "studioBackgroundColor", [
			uv,
			viewport,
			uniforms.center,
			uniforms.color,
			uniforms.keyColor,
			uniforms.fillColor,
			uniforms.backColor,
			lights1,
			lights2,
			scene,
			misc
		]);
	}
};
var StudioBackground_default = componentDefinition;
export { componentDefinition as n, studioBackgroundColor as r, StudioBackground_default as t };
