// shaders/dist/core/SunBurst-CuawxiUI.js
import { At as call, Ft as mixExpr, It as vec4, R as std, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
const sunBurstAlpha = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, center, rayCount, softness, radius, feather, animTime) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const cx = center.x;
	const cy = __tsover_sub(1, center.y);
	const dx = __tsover_sub(__tsover_mul(uv.x, aspect), __tsover_mul(cx, aspect));
	const dy = __tsover_sub(uv.y, cy);
	const dist = std.sqrt(__tsover_add(__tsover_mul(dx, dx), __tsover_mul(dy, dy)));
	const angle = std.atan2(dy, dx);
	const t = __tsover_add(__tsover_mul(std.sin(__tsover_sub(__tsover_mul(angle, rayCount), animTime)), .5), .5);
	const sharpness = __tsover_mul(__tsover_div(1, __tsover_add(softness, .05)), .3);
	const rayMask = std.pow(t, sharpness);
	const falloffStart = __tsover_mul(radius, __tsover_sub(1, feather));
	const outerMask = __tsover_sub(1, std.smoothstep(falloffStart, radius, dist));
	return __tsover_mul(rayMask, outerMask);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "rayCount"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "radius"
			},
			{
				"type": "i",
				"name": "feather"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"cx",
				[
					7,
					"center",
					"x"
				]
			],
			[
				13,
				"cy",
				[
					1,
					[5, "1"],
					"-",
					[
						7,
						"center",
						"y"
					]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					],
					"-",
					[
						1,
						"cx",
						"*",
						"aspect"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					"cy"
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"std",
						"sqrt"
					],
					[[
						1,
						[
							1,
							"dx",
							"*",
							"dx"
						],
						"+",
						[
							1,
							"dy",
							"*",
							"dy"
						]
					]]
				]
			],
			[
				13,
				"angle",
				[
					6,
					[
						7,
						"std",
						"atan2"
					],
					["dy", "dx"]
				]
			],
			[
				13,
				"t",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									"angle",
									"*",
									"rayCount"
								],
								"-",
								"animTime"
							]]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"sharpness",
				[
					1,
					[
						1,
						[5, "1"],
						"/",
						[
							1,
							"softness",
							"+",
							[5, "0.05"]
						]
					],
					"*",
					[5, "0.3"]
				]
			],
			[
				13,
				"rayMask",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					["t", "sharpness"]
				]
			],
			[
				13,
				"falloffStart",
				[
					1,
					"radius",
					"*",
					[
						1,
						[5, "1"],
						"-",
						"feather"
					]
				]
			],
			[
				13,
				"outerMask",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							"falloffStart",
							"radius",
							"dist"
						]
					]
				]
			],
			[10, [
				1,
				"rayMask",
				"*",
				"outerMask"
			]]
		]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const componentDefinition = {
	name: "SunBurst",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Radial sunburst rays emanating from a center point",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		color: {
			default: "#ffdd88",
			transform: transformColor,
			description: "Ray color",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		background: {
			default: "#000000",
			transform: transformColor,
			description: "Background color",
			ui: {
				type: "color",
				label: "Background",
				group: "Colors"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the sunburst",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		rayCount: {
			default: 12,
			description: "Number of rays",
			ui: {
				type: "range",
				min: 3,
				max: 64,
				step: 1,
				label: "Ray Count",
				group: "Effect"
			}
		},
		softness: {
			default: .3,
			description: "Softness of ray edges",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		radius: {
			default: .8,
			description: "How far the rays extend from the center",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1.2,
				step: .01,
				label: "Radius",
				group: "Effect"
			}
		},
		feather: {
			default: .5,
			description: "How gradually the rays fade at their outer edge",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .01,
				label: "Feather",
				group: "Effect"
			}
		},
		speed: {
			default: .2,
			description: "Rotation speed — positive values rotate clockwise",
			ui: {
				type: "range",
				min: -2,
				max: 2,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		}
	},
	fragment: (params) => {
		const { uniforms } = params;
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		const alpha = call(sunBurstAlpha, "sunBurstAlpha", [
			uv,
			viewport,
			uniforms.center,
			uniforms.rayCount,
			uniforms.softness,
			uniforms.radius,
			uniforms.feather,
			t
		]);
		const col = uniforms.color;
		const bg = uniforms.background;
		return vec4(mixExpr(bg.member("rgb"), col.member("rgb"), alpha), mixExpr(bg.member("a"), col.member("a"), alpha));
	}
};
var SunBurst_default = componentDefinition;
export { componentDefinition as n, sunBurstAlpha as r, SunBurst_default as t };
