// shaders/dist/core/Swirl-BD3Fk-SC.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, gt as mixColorStopsRuntime, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig, t as animatedTime } from "./porters-Dg80Yg90.js";
const swirlField = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, detail, blend, animTime) => {
	"use gpu";
	const t = animTime;
	const freq1 = detail;
	const d1 = d.vec2f(__tsover_add(uv.x, __tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_mul(uv.y, __tsover_mul(freq1, 1.7)), __tsover_mul(t, .8))), .12), __tsover_mul(std.cos(__tsover_sub(__tsover_mul(uv.x, __tsover_mul(freq1, .9)), __tsover_mul(t, .5))), .05))), __tsover_add(uv.y, __tsover_add(__tsover_mul(std.cos(__tsover_sub(__tsover_mul(uv.x, __tsover_mul(freq1, 1.3)), __tsover_mul(t, .6))), .12), __tsover_mul(std.sin(__tsover_add(__tsover_mul(uv.y, __tsover_mul(freq1, 1.1)), __tsover_mul(t, .7))), .05))));
	const pattern1 = std.sin(__tsover_add(__tsover_add(__tsover_mul(d1.x, __tsover_mul(freq1, 2.1)), __tsover_mul(d1.y, __tsover_mul(freq1, 1.8))), __tsover_mul(t, .4)));
	const freq2 = __tsover_mul(detail, 2.1);
	const d2 = d.vec2f(__tsover_add(d1.x, __tsover_add(__tsover_mul(std.cos(__tsover_sub(__tsover_mul(d1.y, __tsover_mul(freq2, 2.7)), __tsover_mul(t, .45))), .07), __tsover_mul(std.sin(__tsover_add(__tsover_mul(d1.x, __tsover_mul(freq2, 1.9)), __tsover_mul(t, .6))), .04))), __tsover_add(d1.y, __tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_mul(d1.x, __tsover_mul(freq2, 2.3)), __tsover_mul(t, .65))), .07), __tsover_mul(std.cos(__tsover_sub(__tsover_mul(d1.y, __tsover_mul(freq2, 1.6)), __tsover_mul(t, .4))), .04))));
	const pattern2 = std.cos(__tsover_add(__tsover_sub(__tsover_mul(d2.x, __tsover_mul(freq2, 1.4)), __tsover_mul(d2.y, __tsover_mul(freq2, 1.9))), __tsover_mul(t, .35)));
	const freq3 = __tsover_mul(detail, 3.7);
	const d3 = d.vec2f(__tsover_add(d2.x, __tsover_add(__tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_mul(d2.y, __tsover_mul(freq3, 1.8)), __tsover_mul(t, .85))), .04), __tsover_mul(std.cos(__tsover_sub(__tsover_mul(d2.x, __tsover_mul(freq3, 1.3)), __tsover_mul(t, .55))), .025)), __tsover_mul(std.sin(__tsover_add(__tsover_mul(__tsover_add(d2.x, d2.y), __tsover_mul(freq3, .7)), __tsover_mul(t, .9))), .02))), __tsover_add(d2.y, __tsover_add(__tsover_add(__tsover_mul(std.cos(__tsover_sub(__tsover_mul(d2.x, __tsover_mul(freq3, 1.6)), __tsover_mul(t, .75))), .04), __tsover_mul(std.sin(__tsover_add(__tsover_mul(d2.y, __tsover_mul(freq3, 1.1)), __tsover_mul(t, .5))), .025)), __tsover_mul(std.cos(__tsover_sub(__tsover_mul(__tsover_add(d2.x, d2.y), __tsover_mul(freq3, .8)), __tsover_mul(t, .95))), .02))));
	const pattern3 = std.sin(__tsover_sub(__tsover_add(__tsover_mul(d3.x, __tsover_mul(freq3, 1.1)), __tsover_mul(d3.y, __tsover_mul(freq3, 1.5))), __tsover_mul(t, .55)));
	const combinedPattern = __tsover_add(__tsover_add(__tsover_mul(pattern1, .45), __tsover_mul(pattern2, .35)), __tsover_mul(pattern3, .2));
	const blendBias = __tsover_mul(__tsover_sub(blend, 50), .006);
	const normalizedPattern = __tsover_add(__tsover_add(__tsover_mul(combinedPattern, .5), .5), blendBias);
	const blendFactor = std.smoothstep(.3, .7, normalizedPattern);
	const shimmer = __tsover_add(__tsover_mul(std.sin(__tsover_add(__tsover_mul(t, 2.5), __tsover_mul(combinedPattern, 8))), .015), 1);
	return d.vec2f(blendFactor, shimmer);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "detail"
			},
			{
				"type": "i",
				"name": "blend"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"t",
				"animTime"
			],
			[
				13,
				"freq1",
				"detail"
			],
			[
				13,
				"d1",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"+",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										[
											1,
											[
												7,
												"uv",
												"y"
											],
											"*",
											[
												1,
												"freq1",
												"*",
												[5, "1.7"]
											]
										],
										"+",
										[
											1,
											"t",
											"*",
											[5, "0.8"]
										]
									]]
								],
								"*",
								[5, "0.12"]
							],
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"cos"
									],
									[[
										1,
										[
											1,
											[
												7,
												"uv",
												"x"
											],
											"*",
											[
												1,
												"freq1",
												"*",
												[5, "0.9"]
											]
										],
										"-",
										[
											1,
											"t",
											"*",
											[5, "0.5"]
										]
									]]
								],
								"*",
								[5, "0.05"]
							]
						]
					], [
						1,
						[
							7,
							"uv",
							"y"
						],
						"+",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"cos"
									],
									[[
										1,
										[
											1,
											[
												7,
												"uv",
												"x"
											],
											"*",
											[
												1,
												"freq1",
												"*",
												[5, "1.3"]
											]
										],
										"-",
										[
											1,
											"t",
											"*",
											[5, "0.6"]
										]
									]]
								],
								"*",
								[5, "0.12"]
							],
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										[
											1,
											[
												7,
												"uv",
												"y"
											],
											"*",
											[
												1,
												"freq1",
												"*",
												[5, "1.1"]
											]
										],
										"+",
										[
											1,
											"t",
											"*",
											[5, "0.7"]
										]
									]]
								],
								"*",
								[5, "0.05"]
							]
						]
					]]
				]
			],
			[
				13,
				"pattern1",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						1,
						[
							1,
							[
								1,
								[
									7,
									"d1",
									"x"
								],
								"*",
								[
									1,
									"freq1",
									"*",
									[5, "2.1"]
								]
							],
							"+",
							[
								1,
								[
									7,
									"d1",
									"y"
								],
								"*",
								[
									1,
									"freq1",
									"*",
									[5, "1.8"]
								]
							]
						],
						"+",
						[
							1,
							"t",
							"*",
							[5, "0.4"]
						]
					]]
				]
			],
			[
				13,
				"freq2",
				[
					1,
					"detail",
					"*",
					[5, "2.1"]
				]
			],
			[
				13,
				"d2",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"d1",
							"x"
						],
						"+",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"cos"
									],
									[[
										1,
										[
											1,
											[
												7,
												"d1",
												"y"
											],
											"*",
											[
												1,
												"freq2",
												"*",
												[5, "2.7"]
											]
										],
										"-",
										[
											1,
											"t",
											"*",
											[5, "0.45"]
										]
									]]
								],
								"*",
								[5, "0.07"]
							],
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										[
											1,
											[
												7,
												"d1",
												"x"
											],
											"*",
											[
												1,
												"freq2",
												"*",
												[5, "1.9"]
											]
										],
										"+",
										[
											1,
											"t",
											"*",
											[5, "0.6"]
										]
									]]
								],
								"*",
								[5, "0.04"]
							]
						]
					], [
						1,
						[
							7,
							"d1",
							"y"
						],
						"+",
						[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										[
											1,
											[
												7,
												"d1",
												"x"
											],
											"*",
											[
												1,
												"freq2",
												"*",
												[5, "2.3"]
											]
										],
										"+",
										[
											1,
											"t",
											"*",
											[5, "0.65"]
										]
									]]
								],
								"*",
								[5, "0.07"]
							],
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"cos"
									],
									[[
										1,
										[
											1,
											[
												7,
												"d1",
												"y"
											],
											"*",
											[
												1,
												"freq2",
												"*",
												[5, "1.6"]
											]
										],
										"-",
										[
											1,
											"t",
											"*",
											[5, "0.4"]
										]
									]]
								],
								"*",
								[5, "0.04"]
							]
						]
					]]
				]
			],
			[
				13,
				"pattern2",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					[[
						1,
						[
							1,
							[
								1,
								[
									7,
									"d2",
									"x"
								],
								"*",
								[
									1,
									"freq2",
									"*",
									[5, "1.4"]
								]
							],
							"-",
							[
								1,
								[
									7,
									"d2",
									"y"
								],
								"*",
								[
									1,
									"freq2",
									"*",
									[5, "1.9"]
								]
							]
						],
						"+",
						[
							1,
							"t",
							"*",
							[5, "0.35"]
						]
					]]
				]
			],
			[
				13,
				"freq3",
				[
					1,
					"detail",
					"*",
					[5, "3.7"]
				]
			],
			[
				13,
				"d3",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"d2",
							"x"
						],
						"+",
						[
							1,
							[
								1,
								[
									1,
									[
										6,
										[
											7,
											"std",
											"sin"
										],
										[[
											1,
											[
												1,
												[
													7,
													"d2",
													"y"
												],
												"*",
												[
													1,
													"freq3",
													"*",
													[5, "1.8"]
												]
											],
											"+",
											[
												1,
												"t",
												"*",
												[5, "0.85"]
											]
										]]
									],
									"*",
									[5, "0.04"]
								],
								"+",
								[
									1,
									[
										6,
										[
											7,
											"std",
											"cos"
										],
										[[
											1,
											[
												1,
												[
													7,
													"d2",
													"x"
												],
												"*",
												[
													1,
													"freq3",
													"*",
													[5, "1.3"]
												]
											],
											"-",
											[
												1,
												"t",
												"*",
												[5, "0.55"]
											]
										]]
									],
									"*",
									[5, "0.025"]
								]
							],
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										[
											1,
											[
												1,
												[
													7,
													"d2",
													"x"
												],
												"+",
												[
													7,
													"d2",
													"y"
												]
											],
											"*",
											[
												1,
												"freq3",
												"*",
												[5, "0.7"]
											]
										],
										"+",
										[
											1,
											"t",
											"*",
											[5, "0.9"]
										]
									]]
								],
								"*",
								[5, "0.02"]
							]
						]
					], [
						1,
						[
							7,
							"d2",
							"y"
						],
						"+",
						[
							1,
							[
								1,
								[
									1,
									[
										6,
										[
											7,
											"std",
											"cos"
										],
										[[
											1,
											[
												1,
												[
													7,
													"d2",
													"x"
												],
												"*",
												[
													1,
													"freq3",
													"*",
													[5, "1.6"]
												]
											],
											"-",
											[
												1,
												"t",
												"*",
												[5, "0.75"]
											]
										]]
									],
									"*",
									[5, "0.04"]
								],
								"+",
								[
									1,
									[
										6,
										[
											7,
											"std",
											"sin"
										],
										[[
											1,
											[
												1,
												[
													7,
													"d2",
													"y"
												],
												"*",
												[
													1,
													"freq3",
													"*",
													[5, "1.1"]
												]
											],
											"+",
											[
												1,
												"t",
												"*",
												[5, "0.5"]
											]
										]]
									],
									"*",
									[5, "0.025"]
								]
							],
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"cos"
									],
									[[
										1,
										[
											1,
											[
												1,
												[
													7,
													"d2",
													"x"
												],
												"+",
												[
													7,
													"d2",
													"y"
												]
											],
											"*",
											[
												1,
												"freq3",
												"*",
												[5, "0.8"]
											]
										],
										"-",
										[
											1,
											"t",
											"*",
											[5, "0.95"]
										]
									]]
								],
								"*",
								[5, "0.02"]
							]
						]
					]]
				]
			],
			[
				13,
				"pattern3",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						1,
						[
							1,
							[
								1,
								[
									7,
									"d3",
									"x"
								],
								"*",
								[
									1,
									"freq3",
									"*",
									[5, "1.1"]
								]
							],
							"+",
							[
								1,
								[
									7,
									"d3",
									"y"
								],
								"*",
								[
									1,
									"freq3",
									"*",
									[5, "1.5"]
								]
							]
						],
						"-",
						[
							1,
							"t",
							"*",
							[5, "0.55"]
						]
					]]
				]
			],
			[
				13,
				"combinedPattern",
				[
					1,
					[
						1,
						[
							1,
							"pattern1",
							"*",
							[5, "0.45"]
						],
						"+",
						[
							1,
							"pattern2",
							"*",
							[5, "0.35"]
						]
					],
					"+",
					[
						1,
						"pattern3",
						"*",
						[5, "0.2"]
					]
				]
			],
			[
				13,
				"blendBias",
				[
					1,
					[
						1,
						"blend",
						"-",
						[5, "50"]
					],
					"*",
					[5, "0.006"]
				]
			],
			[
				13,
				"normalizedPattern",
				[
					1,
					[
						1,
						[
							1,
							"combinedPattern",
							"*",
							[5, "0.5"]
						],
						"+",
						[5, "0.5"]
					],
					"+",
					"blendBias"
				]
			],
			[
				13,
				"blendFactor",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.3"],
						[5, "0.7"],
						"normalizedPattern"
					]
				]
			],
			[
				13,
				"shimmer",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									"t",
									"*",
									[5, "2.5"]
								],
								"+",
								[
									1,
									"combinedPattern",
									"*",
									[5, "8"]
								]
							]]
						],
						"*",
						[5, "0.015"]
					],
					"+",
					[5, "1"]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["blendFactor", "shimmer"]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Swirl",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Textures",
	description: "Flowing swirl pattern with multi-layered noise",
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#1275d8",
			description: "Primary gradient color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#e19136",
			description: "Secondary gradient color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		stops: colorStopsPropConfig(),
		speed: {
			default: 1,
			description: "Flow animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		detail: {
			default: 1,
			description: "Level of detail and intricacy in the swirl patterns",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .1,
				label: "Detail",
				group: "Effect"
			}
		},
		blend: {
			default: 50,
			description: "Skew color balance toward A (lower values) or B (higher values)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Blend",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, ctx, propValues } = params;
		const t = animatedTime(params);
		const field = call(swirlField, "swirlField", [
			ctx.uv,
			uniforms.detail,
			uniforms.blend,
			t
		]);
		const blendFactor = field.member("x");
		const shimmer = field.member("y");
		const colorSpaceMode = propValues.colorSpace ?? 0;
		const stopCount = propValues.stopCount ?? 0;
		let colorMix;
		if (stopCount > 1) colorMix = mixColorStopsRuntime(blendFactor, {
			colorsArray: uniforms.colorsArray,
			positionsArray: uniforms.positionsArray,
			convertedColorsArray: uniforms.convertedColorsArray,
			stopCount: uniforms.stopCount
		}, colorSpaceMode);
		else colorMix = call(mixColorsVariants[colorSpaceMode] ?? mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			blendFactor
		]);
		return colorMix.mul(shimmer);
	}
};
var Swirl_default = componentDefinition;
export { componentDefinition as n, swirlField as r, Swirl_default as t };
