// shaders/dist/core/Text-CrbVixvj.js
import { At as call, It as vec4, R as std, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
import { n as loadGoogleFont, t as isFontAvailable } from "./fonts-3F3pIODs.js";
function applyTextTransform(text, mode) {
	switch (mode) {
		case "uppercase": return text.toUpperCase();
		case "lowercase": return text.toLowerCase();
		case "capitalize": return text.replace(/(^|\s)(\S)/g, (_, ws, ch) => ws + ch.toUpperCase());
		default: return text;
	}
}
var generation = 0;
function bumpMeasureGeneration() {
	generation++;
}
var measureCtx = null;
var letterSpacingSupported = false;
function getMeasureContext() {
	if (measureCtx) return measureCtx;
	if (typeof document === "undefined") return null;
	const canvas = document.createElement("canvas");
	canvas.width = 1;
	canvas.height = 1;
	measureCtx = canvas.getContext("2d", { willReadFrequently: false });
	if (measureCtx) letterSpacingSupported = "letterSpacing" in measureCtx;
	return measureCtx;
}
function supportsCanvasLetterSpacing() {
	getMeasureContext();
	return letterSpacingSupported;
}
var MAX_CACHE_ENTRIES = 500;
var cache = /* @__PURE__ */ new Map();
function measureText(spec) {
	const text = applyTextTransform(spec.text, spec.textTransform);
	const ctx = getMeasureContext();
	if (!ctx) return {
		widthPer100: text.length * 55,
		ascentPer100: 80,
		descentPer100: 20
	};
	const fontLoaded = isFontAvailable(spec.fontFamily, spec.fontWeight, spec.italic);
	const key = `${generation}|${fontLoaded ? 1 : 0}|${spec.fontFamily}|${spec.fontWeight}|${spec.italic ? 1 : 0}|${spec.letterSpacingEm}|${text}`;
	const cached = cache.get(key);
	if (cached) return cached;
	ctx.font = `${spec.italic ? "italic " : ""}${spec.fontWeight} 100px "${spec.fontFamily}", sans-serif`;
	if (letterSpacingSupported) ctx.letterSpacing = `${spec.letterSpacingEm * 100}px`;
	const metrics = ctx.measureText(text);
	let widthPer100 = metrics.width;
	if (letterSpacingSupported && text.length > 0) widthPer100 = Math.max(0, widthPer100 - spec.letterSpacingEm * 100);
	const ascentPer100 = metrics.fontBoundingBoxAscent ?? metrics.actualBoundingBoxAscent ?? 80;
	const descentPer100 = metrics.fontBoundingBoxDescent ?? metrics.actualBoundingBoxDescent ?? 20;
	const result = {
		widthPer100,
		ascentPer100,
		descentPer100
	};
	if (cache.size >= MAX_CACHE_ENTRIES) cache.clear();
	cache.set(key, result);
	return result;
}
var MAX_TEXTURE_AXIS = 4096;
var DEG_TO_RAD = Math.PI / 180;
const TEXT_LINE_BOX_RATIO = 1.2;
var fontSizePxOf = (v, canvasHeight) => {
	if (v && typeof v === "object" && typeof v.value === "number") return v.unit === "px" ? v.value : v.value * canvasHeight;
	return (typeof v === "number" ? v : .1) * canvasHeight;
};
const textField = tgpu.fn([
	d.vec2f,
	d.f32,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, center, rotationDeg, halfW, halfH) => {
	"use gpu";
	const aspectUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y));
	const dx = __tsover_sub(aspectUV.x, centerPos.x);
	const dy = __tsover_sub(aspectUV.y, centerPos.y);
	const rotRad = __tsover_mul(rotationDeg, DEG_TO_RAD);
	const cosR = std.cos(rotRad);
	const sinR = std.sin(rotRad);
	const rdx = __tsover_add(__tsover_mul(dx, cosR), __tsover_mul(dy, sinR));
	const rdy = __tsover_sub(__tsover_mul(dy, cosR), __tsover_mul(dx, sinR));
	const localU = __tsover_add(__tsover_div(rdx, __tsover_mul(halfW, 2)), .5);
	const localV = __tsover_add(__tsover_div(rdy, __tsover_mul(halfH, 2)), .5);
	const insideMask = __tsover_mul(__tsover_mul(__tsover_mul(std.step(0, localU), std.step(localU, 1)), std.step(0, localV)), std.step(localV, 1));
	return d.vec3f(localU, localV, insideMask);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "rotationDeg"
			},
			{
				"type": "i",
				"name": "halfW"
			},
			{
				"type": "i",
				"name": "halfH"
			}
		],
		"body": [0, [
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dx",
				[
					1,
					[
						7,
						"aspectUV",
						"x"
					],
					"-",
					[
						7,
						"centerPos",
						"x"
					]
				]
			],
			[
				13,
				"dy",
				[
					1,
					[
						7,
						"aspectUV",
						"y"
					],
					"-",
					[
						7,
						"centerPos",
						"y"
					]
				]
			],
			[
				13,
				"rotRad",
				[
					1,
					"rotationDeg",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["rotRad"]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["rotRad"]
				]
			],
			[
				13,
				"rdx",
				[
					1,
					[
						1,
						"dx",
						"*",
						"cosR"
					],
					"+",
					[
						1,
						"dy",
						"*",
						"sinR"
					]
				]
			],
			[
				13,
				"rdy",
				[
					1,
					[
						1,
						"dy",
						"*",
						"cosR"
					],
					"-",
					[
						1,
						"dx",
						"*",
						"sinR"
					]
				]
			],
			[
				13,
				"localU",
				[
					1,
					[
						1,
						"rdx",
						"/",
						[
							1,
							"halfW",
							"*",
							[5, "2"]
						]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"localV",
				[
					1,
					[
						1,
						"rdy",
						"/",
						[
							1,
							"halfH",
							"*",
							[5, "2"]
						]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"insideMask",
				[
					1,
					[
						1,
						[
							1,
							[
								6,
								[
									7,
									"std",
									"step"
								],
								[[5, "0"], "localU"]
							],
							"*",
							[
								6,
								[
									7,
									"std",
									"step"
								],
								["localU", [5, "1"]]
							]
						],
						"*",
						[
							6,
							[
								7,
								"std",
								"step"
							],
							[[5, "0"], "localV"]
						]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"step"
						],
						["localV", [5, "1"]]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					"localU",
					"localV",
					"insideMask"
				]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Text",
	category: "Textures",
	description: "Single-line text with any Google font, rendered crisp via a glyph texture",
	acceptsUVContext: true,
	extraFields: {
		halfW: {
			schema: d.f32,
			initial: .001
		},
		halfH: {
			schema: d.f32,
			initial: .001
		}
	},
	boundingBoxDeclaration: {
		propBindings: {
			x: {
				prop: "center",
				as: "position-x"
			},
			y: {
				prop: "center",
				as: "position-y"
			},
			rotation: {
				prop: "rotation",
				as: "degrees"
			}
		},
		computeBounds(props, cw, ch) {
			const fontSizePx = fontSizePxOf(props.fontSize, ch);
			const m = measureText({
				text: String(props.text ?? "").replace(/[\r\n]+/g, " "),
				fontFamily: String(props.fontFamily ?? "Inter"),
				fontWeight: Number(props.fontWeight ?? 400),
				italic: !!props.italic,
				letterSpacingEm: Number(props.letterSpacing ?? 0),
				textTransform: props.textTransform ?? "none"
			});
			const cx = typeof props.center?.x === "number" ? props.center.x : .5;
			const cy = typeof props.center?.y === "number" ? props.center.y : .5;
			return {
				centerXPx: cx * cw,
				centerYPx: cy * ch,
				widthPx: Math.max(0, m.widthPer100 * fontSizePx / 100),
				heightPx: Math.max(0, fontSizePx * TEXT_LINE_BOX_RATIO),
				rotationDeg: props.rotation ?? 0
			};
		},
		writeBounds(bounds, currentProps, cw, ch) {
			const newFontPx = bounds.heightPx / TEXT_LINE_BOX_RATIO;
			const raw = currentProps.fontSize;
			const fontSize = raw && typeof raw === "object" && raw.unit === "px" ? {
				value: Math.max(1, Math.round(newFontPx)),
				unit: "px"
			} : Math.max(.001, ch > 0 ? newFontPx / ch : .001);
			return {
				center: {
					...currentProps.center ?? {
						x: .5,
						y: .5
					},
					x: bounds.centerXPx / cw,
					y: bounds.centerYPx / ch
				},
				fontSize,
				rotation: bounds.rotationDeg
			};
		}
	},
	props: {
		text: {
			default: "Hello World",
			description: "The text to display (single line)",
			ui: {
				type: "text",
				label: "Text",
				group: "Text",
				hidden: true
			}
		},
		fontFamily: {
			default: "Inter",
			description: "Google Fonts family name",
			ui: {
				type: "font-family",
				label: "Font",
				group: "Text"
			}
		},
		fontWeight: {
			default: 400,
			description: "Font weight (only weights published for the family render true; others are synthesized)",
			ui: {
				type: "font-weight",
				label: "Weight",
				group: "Text"
			}
		},
		italic: {
			default: false,
			description: "Italic style",
			ui: {
				type: "checkbox",
				label: "Italic",
				group: "Text"
			},
			transform: (value) => value ? 1 : 0
		},
		textTransform: {
			default: "none",
			description: "Case transformation applied to the text",
			ui: {
				type: "select",
				options: [
					{
						label: "None",
						value: "none"
					},
					{
						label: "Uppercase",
						value: "uppercase"
					},
					{
						label: "Lowercase",
						value: "lowercase"
					},
					{
						label: "Capitalize",
						value: "capitalize"
					}
				],
				label: "Transform",
				group: "Text"
			}
		},
		fontSize: {
			default: .07,
			description: "Font size as a fraction of canvas height",
			ui: {
				type: "range",
				min: .01,
				max: .5,
				step: .005,
				label: "Size",
				group: "Typography",
				dimensional: "canvas-height"
			}
		},
		letterSpacing: {
			default: 0,
			description: "Letter spacing in em units",
			ui: {
				type: "range",
				min: -.1,
				max: .5,
				step: .005,
				label: "Letter Spacing",
				group: "Typography"
			}
		},
		color: {
			default: "#ffffff",
			transform: transformColor,
			description: "Text color",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the text relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the text",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		rotation: {
			default: 0,
			description: "Rotation in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		}
	},
	fragment: ({ uniforms, ctx, uvContext, onCleanup, onBeforeRender, canvas, dimensions, getCpuValue, setExtraField, createMediaTexture, registerMediaTexture }) => {
		const rasterCanvas = document.createElement("canvas");
		rasterCanvas.width = 2;
		rasterCanvas.height = 2;
		const rasterCtx = rasterCanvas.getContext("2d");
		let glyphTex = createMediaTexture({
			width: 2,
			height: 2,
			label: "Text:glyph"
		});
		let isDisposed = false;
		const ensureTexSize = (w, h) => {
			if (isDisposed || glyphTex.width === w && glyphTex.height === h) return;
			const next = createMediaTexture({
				width: w,
				height: h,
				label: "Text:glyph"
			});
			const old = glyphTex;
			glyphTex = next;
			old.destroy();
		};
		const defaults = componentDefinition.props;
		const readProps = () => ({
			text: String(getCpuValue("text") ?? defaults.text.default),
			fontFamily: String(getCpuValue("fontFamily") ?? defaults.fontFamily.default),
			fontWeight: Number(getCpuValue("fontWeight") ?? defaults.fontWeight.default),
			italic: Number(getCpuValue("italic") ?? 0) > .5,
			fontSizeUV: Number(getCpuValue("fontSize") ?? defaults.fontSize.default),
			letterSpacing: Number(getCpuValue("letterSpacing") ?? defaults.letterSpacing.default),
			textTransform: String(getCpuValue("textTransform") ?? defaults.textTransform.default)
		});
		const raster = () => {
			if (!rasterCtx || isDisposed) return;
			const p = readProps();
			const dimW = dimensions.width || 1;
			const dimH = dimensions.height || 1;
			const dpr = Math.max(.1, (canvas.width || dimW) / dimW);
			const singleLine = p.text.replace(/[\r\n]+/g, " ");
			const drawText = applyTextTransform(singleLine, p.textTransform);
			const fontSizePx = p.fontSizeUV * dimH;
			if (!drawText.trim() || fontSizePx <= 0) {
				rasterCanvas.width = 2;
				rasterCanvas.height = 2;
				rasterCtx.clearRect(0, 0, 2, 2);
				ensureTexSize(2, 2);
				glyphTex.write(rasterCanvas);
				setExtraField("halfW", .001);
				setExtraField("halfH", .001);
				return;
			}
			const m = measureText({
				text: singleLine,
				fontFamily: p.fontFamily,
				fontWeight: p.fontWeight,
				italic: p.italic,
				letterSpacingEm: p.letterSpacing,
				textTransform: p.textTransform
			});
			const textWPx = Math.max(1, m.widthPer100 * fontSizePx / 100);
			const lineBoxPx = Math.max(1, fontSizePx * TEXT_LINE_BOX_RATIO);
			const ascentPx = m.ascentPer100 * fontSizePx / 100;
			const descentPx = m.descentPer100 * fontSizePx / 100;
			const padPx = Math.max(.25 * fontSizePx, (ascentPx + descentPx - lineBoxPx) / 2 + .25 * fontSizePx);
			const cssW = textWPx + padPx * 2;
			const cssH = lineBoxPx + padPx * 2;
			let scale = dpr * 2;
			scale = Math.min(scale, MAX_TEXTURE_AXIS / cssW, MAX_TEXTURE_AXIS / cssH);
			const texW = Math.max(2, Math.round(cssW * scale));
			const texH = Math.max(2, Math.round(cssH * scale));
			rasterCanvas.width = texW;
			rasterCanvas.height = texH;
			rasterCtx.clearRect(0, 0, texW, texH);
			rasterCtx.fillStyle = "#ffffff";
			rasterCtx.font = `${p.italic ? "italic " : ""}${p.fontWeight} ${fontSizePx * scale}px "${p.fontFamily}", sans-serif`;
			if (supportsCanvasLetterSpacing()) rasterCtx.letterSpacing = `${p.letterSpacing * fontSizePx * scale}px`;
			rasterCtx.textAlign = "left";
			rasterCtx.textBaseline = "alphabetic";
			const drawX = (texW - textWPx * scale) / 2;
			const baselineY = texH / 2 + (ascentPx - descentPx) / 2 * scale;
			rasterCtx.fillText(drawText, drawX, baselineY);
			ensureTexSize(texW, texH);
			glyphTex.write(rasterCanvas);
			setExtraField("halfW", Math.max(1e-4, cssW / 2 / dimH));
			setExtraField("halfH", Math.max(1e-4, cssH / 2 / dimH));
		};
		let loadedFontKey = "";
		const ensureFont = () => {
			const p = readProps();
			const key = `${p.fontFamily}|${p.fontWeight}|${p.italic}`;
			if (key === loadedFontKey) return;
			loadedFontKey = key;
			loadGoogleFont(p.fontFamily, p.fontWeight, p.italic).then(() => {
				bumpMeasureGeneration();
				raster();
			}).catch(() => {});
		};
		setTimeout(() => {
			raster();
			ensureFont();
		}, 0);
		let lastRasterKey = "";
		let lastUpdateTime = 0;
		onBeforeRender(() => {
			const now = Date.now();
			if (now - lastUpdateTime < 16) return;
			lastUpdateTime = now;
			const p = readProps();
			const key = `${p.text}|${p.fontFamily}|${p.fontWeight}|${p.italic}|${p.fontSizeUV}|${p.letterSpacing}|${p.textTransform}|${dimensions.width}x${dimensions.height}|${canvas.width}`;
			if (key === lastRasterKey) return;
			lastRasterKey = key;
			raster();
			ensureFont();
		});
		onCleanup(() => {
			isDisposed = true;
			glyphTex.destroy();
			rasterCanvas.width = 0;
			rasterCanvas.height = 0;
		});
		const glyphField = registerMediaTexture(() => glyphTex.texture);
		const field = call(textField, "textField", [
			uvContext ?? ctx.uv,
			ctx.aspect,
			uniforms.center,
			uniforms.rotation,
			uniforms.halfW,
			uniforms.halfH
		]);
		const sampled = glyphField.sample(field.member("xy"));
		const color = uniforms.color;
		const alpha = sampled.member("a").mul(color.member("a")).mul(field.member("z"));
		return vec4(color.member("rgb"), alpha);
	}
};
var Text_default = componentDefinition;
export { textField as i, Text_default as n, componentDefinition as r, TEXT_LINE_BOX_RATIO as t };
