// shaders/dist/core/ThinFilm-D_y48u9e.js
import { At as call, Ft as mixExpr, Pt as floatE, Qt as mixColorsVariants, R as std, T as sdf_exports, Zt as mixColorsLinear, cn as __export, d as sdf3d_exports, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { i as sdfSpaceUV, r as offsetUV, t as glassGradient } from "./glass-hUL2pCYG.js";
import { a as transformColor, c as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as shapeEffectBoundingBoxDeclaration } from "./shapeEffectBounds-BlxxsYMN.js";
var thinFilm_exports = /* @__PURE__ */ __export({
	applyThinFilmEffect: () => applyThinFilmEffect$1,
	thinFilmCompose: () => thinFilmCompose,
	thinFilmCustomParams: () => thinFilmCustomParams,
	thinFilmRainbow: () => thinFilmRainbow,
	thinFilmShade: () => thinFilmShade
});
var DEG_TO_RAD = .017453292519943295;
var TWO_PI = 6.283185307179586;
const thinFilmShade = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((sdfRaw, scale, gradX, gradY, edgeSoftness, rimWidthUniform, thickness, dispersion, hueShift, lightAngle, animTime, volumetric) => {
	"use gpu";
	const sdf = __tsover_div(sdfRaw, scale);
	const sharp = std.max(__tsover_mul(edgeSoftness, .5), .001);
	const gLen = std.max(std.sqrt(__tsover_add(__tsover_mul(gradX, gradX), __tsover_mul(gradY, gradY))), 1e-4);
	const nx = __tsover_div(gradX, gLen);
	const ny = __tsover_div(gradY, gLen);
	const rb1 = std.clamp(__tsover_mul(__tsover_div(__tsover_mul(sdf, -1), sharp), 32), 0, 1);
	const rimWidth = std.max(__tsover_mul(rimWidthUniform, .1), .001);
	const depth = std.clamp(__tsover_div(__tsover_mul(sdf, -1), rimWidth), 0, 1);
	const cosView = __tsover_div(1, std.sqrt(__tsover_add(__tsover_mul(gLen, gLen), 1)));
	const exponent = std.mix(7, 1.5, std.clamp(rimWidthUniform, 0, 1));
	const volRim = __tsover_mul(std.pow(__tsover_sub(1, cosView), exponent), rb1);
	const flatRim = __tsover_mul(__tsover_mul(__tsover_sub(1, depth), __tsover_sub(1, depth)), rb1);
	const rim = std.select(flatRim, volRim, volumetric > .5);
	const lx = std.cos(__tsover_mul(lightAngle, DEG_TO_RAD));
	const ly = std.sin(__tsover_mul(lightAngle, DEG_TO_RAD));
	const baseCos = __tsover_add(__tsover_mul(nx, lx), __tsover_mul(ny, ly));
	const baseSin = __tsover_sub(__tsover_mul(ny, lx), __tsover_mul(nx, ly));
	const theta = __tsover_mul(animTime, TWO_PI);
	const rotatedDot = __tsover_add(__tsover_mul(baseCos, std.cos(theta)), __tsover_mul(baseSin, std.sin(theta)));
	const thickVol = __tsover_mul(__tsover_mul(thickness, std.max(__tsover_mul(sdf, -1), 0)), 10);
	const thickFlat = __tsover_mul(__tsover_mul(thickness, depth), 2);
	const thicknessTerm = std.select(thickFlat, thickVol, volumetric > .5);
	const phase = __tsover_add(__tsover_add(hueShift, __tsover_mul(dispersion, rotatedDot)), thicknessTerm);
	return d.vec2f(phase, rim);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "sdfRaw"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "gradX"
			},
			{
				"type": "i",
				"name": "gradY"
			},
			{
				"type": "i",
				"name": "edgeSoftness"
			},
			{
				"type": "i",
				"name": "rimWidthUniform"
			},
			{
				"type": "i",
				"name": "thickness"
			},
			{
				"type": "i",
				"name": "dispersion"
			},
			{
				"type": "i",
				"name": "hueShift"
			},
			{
				"type": "i",
				"name": "lightAngle"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "volumetric"
			}
		],
		"body": [0, [
			[
				13,
				"sdf",
				[
					1,
					"sdfRaw",
					"/",
					"scale"
				]
			],
			[
				13,
				"sharp",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"edgeSoftness",
						"*",
						[5, "0.5"]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"gLen",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"gradX",
								"*",
								"gradX"
							],
							"+",
							[
								1,
								"gradY",
								"*",
								"gradY"
							]
						]]
					], [5, "0.0001"]]
				]
			],
			[
				13,
				"nx",
				[
					1,
					"gradX",
					"/",
					"gLen"
				]
			],
			[
				13,
				"ny",
				[
					1,
					"gradY",
					"/",
					"gLen"
				]
			],
			[
				13,
				"rb1",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									1,
									"sdf",
									"*",
									[
										4,
										"-",
										[5, "1"]
									]
								],
								"/",
								"sharp"
							],
							"*",
							[5, "32"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"rimWidth",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"rimWidthUniform",
						"*",
						[5, "0.1"]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"depth",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"sdf",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"/",
							"rimWidth"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"cosView",
				[
					1,
					[5, "1"],
					"/",
					[
						6,
						[
							7,
							"std",
							"sqrt"
						],
						[[
							1,
							[
								1,
								"gLen",
								"*",
								"gLen"
							],
							"+",
							[5, "1"]
						]]
					]
				]
			],
			[
				13,
				"exponent",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "7"],
						[5, "1.5"],
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								"rimWidthUniform",
								[5, "0"],
								[5, "1"]
							]
						]
					]
				]
			],
			[
				13,
				"volRim",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"pow"
						],
						[[
							1,
							[5, "1"],
							"-",
							"cosView"
						], "exponent"]
					],
					"*",
					"rb1"
				]
			],
			[
				13,
				"flatRim",
				[
					1,
					[
						1,
						[
							1,
							[5, "1"],
							"-",
							"depth"
						],
						"*",
						[
							1,
							[5, "1"],
							"-",
							"depth"
						]
					],
					"*",
					"rb1"
				]
			],
			[
				13,
				"rim",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"flatRim",
						"volRim",
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"lx",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					[[
						1,
						"lightAngle",
						"*",
						"DEG_TO_RAD"
					]]
				]
			],
			[
				13,
				"ly",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						1,
						"lightAngle",
						"*",
						"DEG_TO_RAD"
					]]
				]
			],
			[
				13,
				"baseCos",
				[
					1,
					[
						1,
						"nx",
						"*",
						"lx"
					],
					"+",
					[
						1,
						"ny",
						"*",
						"ly"
					]
				]
			],
			[
				13,
				"baseSin",
				[
					1,
					[
						1,
						"ny",
						"*",
						"lx"
					],
					"-",
					[
						1,
						"nx",
						"*",
						"ly"
					]
				]
			],
			[
				13,
				"theta",
				[
					1,
					"animTime",
					"*",
					"TWO_PI"
				]
			],
			[
				13,
				"rotatedDot",
				[
					1,
					[
						1,
						"baseCos",
						"*",
						[
							6,
							[
								7,
								"std",
								"cos"
							],
							["theta"]
						]
					],
					"+",
					[
						1,
						"baseSin",
						"*",
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							["theta"]
						]
					]
				]
			],
			[
				13,
				"thickVol",
				[
					1,
					[
						1,
						"thickness",
						"*",
						[
							6,
							[
								7,
								"std",
								"max"
							],
							[[
								1,
								"sdf",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							], [5, "0"]]
						]
					],
					"*",
					[5, "10"]
				]
			],
			[
				13,
				"thickFlat",
				[
					1,
					[
						1,
						"thickness",
						"*",
						"depth"
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"thicknessTerm",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"thickFlat",
						"thickVol",
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"phase",
				[
					1,
					[
						1,
						"hueShift",
						"+",
						[
							1,
							"dispersion",
							"*",
							"rotatedDot"
						]
					],
					"+",
					"thicknessTerm"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["phase", "rim"]
			]]
		]],
		"externalNames": [
			"std",
			"DEG_TO_RAD",
			"TWO_PI",
			"d"
		]
	},
	externals: () => ({
		std,
		DEG_TO_RAD,
		TWO_PI,
		d
	})
}) && $.f)({}));
const thinFilmRainbow = tgpu.fn([d.f32], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((phase) => {
	"use gpu";
	const r = std.cos(__tsover_mul(__tsover_add(phase, 0), TWO_PI));
	const g = std.cos(__tsover_mul(__tsover_add(phase, .3333), TWO_PI));
	const b = std.cos(__tsover_mul(__tsover_add(phase, .6667), TWO_PI));
	const half = d.vec3f(r, g, b).mul(.5);
	return d.vec3f(__tsover_add(half.x, .5), __tsover_add(half.y, .5), __tsover_add(half.z, .5));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "phase"
		}],
		"body": [0, [
			[
				13,
				"r",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					[[
						1,
						[
							1,
							"phase",
							"+",
							[5, "0"]
						],
						"*",
						"TWO_PI"
					]]
				]
			],
			[
				13,
				"g",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					[[
						1,
						[
							1,
							"phase",
							"+",
							[5, "0.3333"]
						],
						"*",
						"TWO_PI"
					]]
				]
			],
			[
				13,
				"b",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					[[
						1,
						[
							1,
							"phase",
							"+",
							[5, "0.6667"]
						],
						"*",
						"TWO_PI"
					]]
				]
			],
			[
				13,
				"half",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								"r",
								"g",
								"b"
							]
						],
						"mul"
					],
					[[5, "0.5"]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					[
						1,
						[
							7,
							"half",
							"x"
						],
						"+",
						[5, "0.5"]
					],
					[
						1,
						[
							7,
							"half",
							"y"
						],
						"+",
						[5, "0.5"]
					],
					[
						1,
						[
							7,
							"half",
							"z"
						],
						"+",
						[5, "0.5"]
					]
				]
			]]
		]],
		"externalNames": [
			"std",
			"TWO_PI",
			"d"
		]
	},
	externals: () => ({
		std,
		TWO_PI,
		d
	})
}) && $.f)({}));
const thinFilmCustomParams = tgpu.fn([d.f32], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((phase) => {
	"use gpu";
	const t = std.fract(phase);
	const t3 = __tsover_mul(t, 3);
	const si = std.fract(t3);
	const floorT3 = std.floor(t3);
	const gate12 = std.select(d.f32(0), d.f32(1), floorT3 <= 1.5);
	const gate01 = std.select(d.f32(0), d.f32(1), floorT3 <= .5);
	return d.vec3f(si, gate12, gate01);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "phase"
		}],
		"body": [0, [
			[
				13,
				"t",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["phase"]
				]
			],
			[
				13,
				"t3",
				[
					1,
					"t",
					"*",
					[5, "3"]
				]
			],
			[
				13,
				"si",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["t3"]
				]
			],
			[
				13,
				"floorT3",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["t3"]
				]
			],
			[
				13,
				"gate12",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"floorT3",
							"<=",
							[5, "1.5"]
						]
					]
				]
			],
			[
				13,
				"gate01",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							1,
							"floorT3",
							"<=",
							[5, "0.5"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					"si",
					"gate12",
					"gate01"
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const thinFilmCompose = tgpu.fn([
	d.vec3f,
	d.f32,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((baseColor, rim, intensity, saturation) => {
	"use gpu";
	const white = d.vec3f(1, 1, 1);
	const rgb = std.mix(white, baseColor, saturation).mul(rim).mul(intensity);
	const alpha = std.clamp(rim, 0, 1);
	return d.vec4f(rgb.x, rgb.y, rgb.z, alpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "baseColor"
			},
			{
				"type": "i",
				"name": "rim"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "saturation"
			}
		],
		"body": [0, [
			[
				13,
				"white",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "1"],
						[5, "1"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"col",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"white",
						"baseColor",
						"saturation"
					]
				]
			],
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"col",
								"mul"
							],
							["rim"]
						],
						"mul"
					],
					["intensity"]
				]
			],
			[
				13,
				"alpha",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"rim",
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"rgb",
						"x"
					],
					[
						7,
						"rgb",
						"y"
					],
					[
						7,
						"rgb",
						"z"
					],
					"alpha"
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
function applyThinFilmEffect$1(params, sdfSampler, options) {
	const { uniforms, propValues, ctx } = params;
	const volumetric = options?.volumetric ?? false;
	const gradSampler = options?.gradSampler ?? sdfSampler;
	const volFlag = floatE(volumetric ? 1 : 0);
	const animTime = params.props.member("_animTime");
	const sdfUV = call(sdfSpaceUV, "sdfSpaceUV", [
		uniforms.center,
		uniforms.scale,
		uniforms.rotation,
		ctx.uv,
		ctx.aspect
	]);
	const surf0 = sdfSampler(sdfUV);
	const sdfRaw = surf0.member("r");
	let gradX;
	let gradY;
	if (options?.bakedGradients) {
		gradX = surf0.member("g");
		gradY = surf0.member("b");
	} else {
		const surfX = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(.01),
			floatE(0)
		]));
		const surfY = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(0),
			floatE(.01)
		]));
		const grad = call(glassGradient, "glassGradient", [
			sdfRaw,
			surfX.member("r"),
			surfY.member("r")
		]);
		gradX = grad.member("x");
		gradY = grad.member("y");
	}
	const shade = call(thinFilmShade, "thinFilmShade", [
		sdfRaw,
		uniforms.scale,
		gradX,
		gradY,
		uniforms.edgeSoftness,
		uniforms.rimWidth,
		uniforms.thickness,
		uniforms.dispersion,
		uniforms.hueShift,
		uniforms.lightAngle,
		animTime,
		volFlag
	]);
	const phase = shade.member("x");
	const rim = shade.member("y");
	let baseColor;
	if (propValues.mode === "custom") {
		const mixFn = mixColorsVariants[typeof propValues.colorSpace === "number" ? propValues.colorSpace : 0] ?? mixColorsLinear;
		const cp = call(thinFilmCustomParams, "thinFilmCustomParams", [phase]);
		const si = cp.member("x");
		const gate12 = cp.member("y");
		const gate01 = cp.member("z");
		const c01 = call(mixFn, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			si
		]);
		const c12 = call(mixFn, "mixColors", [
			uniforms.colorB,
			uniforms.colorC,
			si
		]);
		baseColor = mixExpr(mixExpr(call(mixFn, "mixColors", [
			uniforms.colorC,
			uniforms.colorA,
			si
		]), c12, gate12), c01, gate01).member("rgb");
	} else baseColor = call(thinFilmRainbow, "thinFilmRainbow", [phase]);
	return call(thinFilmCompose, "thinFilmCompose", [
		baseColor,
		rim,
		uniforms.intensity,
		uniforms.saturation
	]);
}
var { applyThinFilmEffect } = thinFilm_exports;
var { createSvgSdfSampler, createAnalyticSdfSampler, driveAnalyticSubProps, ANALYTIC_SDF_EXTRA_FIELDS } = sdf_exports;
var { createVolumetricFieldComputeNode, buildVolumetricFieldSampler, VOLUMETRIC_FIELD_EXTRA_FIELDS, is3dShapeType, isVolumetricShapeType } = sdf3d_exports;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
function resolveShapeType(shapeType, shape) {
	if (shapeType) return shapeType;
	if (shape && typeof shape === "object") return shape.type || "circleSDF";
	if (typeof shape === "string") try {
		return JSON.parse(shape).type || "circleSDF";
	} catch {
		return "circleSDF";
	}
	return "circleSDF";
}
const componentDefinition = {
	name: "ThinFilm",
	boundingBoxDeclaration: shapeEffectBoundingBoxDeclaration,
	category: "Shape Effects",
	description: "Iridescent thin-film edge",
	animatedTime: { speed: "speed" },
	extraFields: {
		...VOLUMETRIC_FIELD_EXTRA_FIELDS,
		...ANALYTIC_SDF_EXTRA_FIELDS
	},
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		scale: {
			default: 1,
			description: "Scale of the shape (1 = default size)",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 3,
				step: .01,
				label: "Scale",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the shape in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		},
		intensity: {
			default: 1,
			description: "Rim brightness — values above 1 push the edge past white for HDR bloom",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 4,
				step: .01,
				label: "Intensity",
				group: "Film"
			}
		},
		rimWidth: {
			default: 1,
			description: "How far inward from the edge the iridescence reaches",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 3,
				step: .01,
				label: "Rim Width",
				group: "Film"
			}
		},
		edgeSoftness: {
			default: .3,
			description: "Softness of the shape boundary",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Softness",
				group: "Film"
			}
		},
		thickness: {
			default: .5,
			description: "Film thickness — higher values pack more spectral bands across the rim",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Thickness",
				group: "Film"
			}
		},
		dispersion: {
			default: .5,
			description: "How strongly colour separates around the perimeter (warm vs cool sides)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Dispersion",
				group: "Film"
			}
		},
		saturation: {
			default: 1,
			description: "Spectral saturation — 0 = white rim, 1 = full iridescence",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Saturation",
				group: "Film"
			}
		},
		hueShift: {
			default: 0,
			description: "Rotate the spectrum to reposition the colours",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Hue Shift",
				group: "Film"
			}
		},
		lightAngle: {
			default: 300,
			description: "Light angle in degrees — sets which side of the rim runs warm vs cool",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Light Angle",
				group: "Film"
			}
		},
		mode: {
			default: "rainbow",
			compileTime: true,
			description: "Rainbow uses the full iridescent spectrum; Custom cycles through your three chosen colours",
			ui: {
				type: "select",
				options: [{
					label: "Rainbow",
					value: "rainbow"
				}, {
					label: "Custom",
					value: "custom"
				}],
				label: "Mode",
				group: "Colors"
			}
		},
		colorA: {
			default: "#2b6fff",
			transform: transformColor,
			description: "First colour in the rim cycle",
			ui: {
				type: "color",
				label: "Color 1",
				group: "Colors",
				condition: { mode: "custom" }
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Second colour in the rim cycle",
			ui: {
				type: "color",
				label: "Color 2",
				group: "Colors",
				condition: { mode: "custom" }
			}
		},
		colorC: {
			default: "#ff7a21",
			transform: transformColor,
			description: "Third colour in the rim cycle",
			ui: {
				type: "color",
				label: "Color 3",
				group: "Colors",
				condition: { mode: "custom" }
			}
		},
		colorSpace: {
			default: "oklch",
			transform: transformColorSpace,
			compileTime: true,
			description: "Colour space used to blend between custom colours",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors",
				condition: { mode: "custom" }
			}
		},
		speed: {
			default: .1,
			description: "Speed at which the colours rotate around the rim (0 = static)",
			ui: {
				type: "range",
				min: -1,
				max: 1,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file — when non-empty, activates SVG mode and triggers a shader recompile"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type — triggers recompile when shape is switched. When empty, derived from shape JSON at mount time."
		}
	},
	compute: (params) => {
		const vf = createVolumetricFieldComputeNode(params, () => params.getCpuValue("shape"), "none", "firstLobe");
		return vf ? {
			outputs: vf.outputs,
			getComputeNodes: (fp) => vf.getComputeNodes(fp)
		} : null;
	},
	fragment: (params) => {
		const { computeOutputs, uniforms, propValues, getCpuValue } = params;
		const fieldTex = computeOutputs?.volumetricFieldTexture;
		if (fieldTex) return applyThinFilmEffect(params, buildVolumetricFieldSampler(fieldTex, {
			originX: uniforms._vfOriginX,
			originY: uniforms._vfOriginY,
			spanX: uniforms._vfSpanX,
			spanY: uniforms._vfSpanY,
			activeResF: uniforms._vfActiveRes,
			rBound: uniforms._vfRBound
		}).volumetricFieldSamplerGrad, {
			volumetric: true,
			bakedGradients: true
		});
		const shapeSdfUrl = propValues.shapeSdfUrl || "";
		const shapeType = propValues.shapeType || "";
		if (shapeSdfUrl) return applyThinFilmEffect(params, createSvgSdfSampler(params, shapeSdfUrl), { volumetric: isVolumetricShapeType(shapeType) });
		const st = resolveShapeType(shapeType, getCpuValue("shape"));
		return applyThinFilmEffect(params, createAnalyticSdfSampler(st, driveAnalyticSubProps(params, () => getCpuValue("shape"))), { volumetric: is3dShapeType(st) });
	}
};
var ThinFilm_default = componentDefinition;
export { componentDefinition as n, ThinFilm_default as t };
