// shaders/dist/core/Tint-HLecBeer.js
import { At as call, Ot as ZERO, R as std, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor } from "./transformations-fqzznt4g.js";
const tintColorPlain = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, tintColor, amount) => {
	"use gpu";
	const rgb = d.vec3f(color.x, color.y, color.z);
	const tinted = std.mix(rgb, d.vec3f(tintColor.x, tintColor.y, tintColor.z), d.vec3f(amount));
	return d.vec4f(tinted.x, tinted.y, tinted.z, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "color"
			},
			{
				"type": "i",
				"name": "tintColor"
			},
			{
				"type": "i",
				"name": "amount"
			}
		],
		"body": [0, [
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"color",
							"x"
						],
						[
							7,
							"color",
							"y"
						],
						[
							7,
							"color",
							"z"
						]
					]
				]
			],
			[
				13,
				"tinted",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"rgb",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									7,
									"tintColor",
									"x"
								],
								[
									7,
									"tintColor",
									"y"
								],
								[
									7,
									"tintColor",
									"z"
								]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["amount"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"tinted",
						"x"
					],
					[
						7,
						"tinted",
						"y"
					],
					[
						7,
						"tinted",
						"z"
					],
					[
						7,
						"color",
						"w"
					]
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const tintColorPreserve = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, tintColor, amount) => {
	"use gpu";
	const rgb = d.vec3f(color.x, color.y, color.z);
	const tinted = std.mix(rgb, d.vec3f(tintColor.x, tintColor.y, tintColor.z), d.vec3f(amount));
	const luminanceWeights = d.vec3f(.299, .587, .114);
	const originalLum = std.dot(rgb, luminanceWeights);
	const tintedLum = std.dot(tinted, luminanceWeights);
	const result = tinted.mul(__tsover_div(originalLum, std.max(tintedLum, 1e-4)));
	return d.vec4f(result.x, result.y, result.z, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "color"
			},
			{
				"type": "i",
				"name": "tintColor"
			},
			{
				"type": "i",
				"name": "amount"
			}
		],
		"body": [0, [
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"color",
							"x"
						],
						[
							7,
							"color",
							"y"
						],
						[
							7,
							"color",
							"z"
						]
					]
				]
			],
			[
				13,
				"tinted",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"rgb",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[
									7,
									"tintColor",
									"x"
								],
								[
									7,
									"tintColor",
									"y"
								],
								[
									7,
									"tintColor",
									"z"
								]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["amount"]
						]
					]
				]
			],
			[
				13,
				"luminanceWeights",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0.299"],
						[5, "0.587"],
						[5, "0.114"]
					]
				]
			],
			[
				13,
				"originalLum",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["rgb", "luminanceWeights"]
				]
			],
			[
				13,
				"tintedLum",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["tinted", "luminanceWeights"]
				]
			],
			[
				13,
				"result",
				[
					6,
					[
						7,
						"tinted",
						"mul"
					],
					[[
						1,
						"originalLum",
						"/",
						[
							6,
							[
								7,
								"std",
								"max"
							],
							["tintedLum", [5, "0.0001"]]
						]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"result",
						"x"
					],
					[
						7,
						"result",
						"y"
					],
					[
						7,
						"result",
						"z"
					],
					[
						7,
						"color",
						"w"
					]
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Tint",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Apply a color tint to the image",
	requiresChild: true,
	props: {
		color: {
			default: "#ff8800",
			transform: transformColor,
			description: "Tint color",
			ui: {
				type: "color",
				label: "Tint Color",
				group: "Colors"
			}
		},
		amount: {
			default: .5,
			description: "Tint amount (0 = no tint, 1 = full tint)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Amount",
				group: "Effect"
			}
		},
		preserveLuminosity: {
			default: true,
			compileTime: true,
			description: "Preserve original brightness",
			transform: (v) => v ? 1 : 0,
			ui: {
				type: "checkbox",
				label: "Preserve Luminosity",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, propValues }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Tint shader.");
			return ZERO;
		}
		if (Number(propValues.preserveLuminosity) > 0) return call(tintColorPreserve, "tintColorPreserve", [
			childNode,
			uniforms.color,
			uniforms.amount
		]);
		return call(tintColorPlain, "tintColorPlain", [
			childNode,
			uniforms.color,
			uniforms.amount
		]);
	}
};
var Tint_default = componentDefinition;
export { tintColorPreserve as i, componentDefinition as n, tintColorPlain as r, Tint_default as t };
