// shaders/dist/core/TriangularGrid-BWzP25ws.js
import { At as call, It as vec4, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
var DEG_TO_RAD = Math.PI / 180;
var SQRT3 = Math.sqrt(3);
var TWO_OVER_SQRT3 = 2 / Math.sqrt(3);
const triangularGridField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, cells, thickness, rotation, softness, variation, speedVariance, animTime) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const correctedUV = d.vec2f(__tsover_mul(uv.x, aspect), __tsover_sub(1, uv.y));
	const ang = __tsover_mul(rotation, DEG_TO_RAD);
	const cosR = std.cos(ang);
	const sinR = std.sin(ang);
	const centerX = __tsover_mul(aspect, .5);
	const centerY = .5;
	const cx = __tsover_sub(correctedUV.x, centerX);
	const cy = __tsover_sub(correctedUV.y, centerY);
	const p = d.vec2f(__tsover_add(__tsover_sub(__tsover_mul(cx, cosR), __tsover_mul(cy, sinR)), centerX), __tsover_add(__tsover_add(__tsover_mul(cx, sinR), __tsover_mul(cy, cosR)), centerY)).mul(cells);
	const ps = d.vec2f(__tsover_sub(p.x, __tsover_div(p.y, SQRT3)), __tsover_mul(p.y, TWO_OVER_SQRT3));
	const rowIndex = std.floor(ps.y);
	const rowHash = std.fract(__tsover_mul(std.sin(__tsover_mul(rowIndex, 127.1)), 43758.5453));
	const speedMod = __tsover_add(1, __tsover_mul(__tsover_mul(__tsover_sub(rowHash, .5), 4), speedVariance));
	const drift = __tsover_mul(animTime, speedMod);
	const psShifted = d.vec2f(__tsover_sub(ps.x, drift), ps.y);
	const cell = std.floor(psShifted);
	const f = std.fract(psShifted);
	const upper = std.step(1, __tsover_add(f.x, f.y));
	const triId = d.vec2f(__tsover_add(__tsover_mul(cell.x, 2), upper), cell.y);
	const q0 = std.fract(triId.mul(d.vec2f(123.34, 345.45)));
	const q1 = q0.add(std.dot(q0, q0.add(34.345)));
	const rand = std.fract(__tsover_mul(q1.x, q1.y));
	const variedFactor = std.max(__tsover_add(1, __tsover_mul(__tsover_mul(__tsover_sub(rand, .5), 2), variation)), 0);
	const dEdge = std.min(std.min(std.min(f.x, __tsover_sub(1, f.x)), std.min(f.y, __tsover_sub(1, f.y))), std.abs(__tsover_sub(__tsover_add(f.x, f.y), 1)));
	const lineWidth = __tsover_mul(thickness, .02);
	const aa = __tsover_add(std.max(std.fwidth(ps.x), std.fwidth(ps.y)), __tsover_mul(softness, .5));
	const lineMask = __tsover_mul(__tsover_sub(1, std.smoothstep(__tsover_sub(lineWidth, aa), __tsover_add(lineWidth, aa), dEdge)), std.step(1e-4, lineWidth));
	return d.vec2f(lineMask, variedFactor);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "cells"
			},
			{
				"type": "i",
				"name": "thickness"
			},
			{
				"type": "i",
				"name": "rotation"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "variation"
			},
			{
				"type": "i",
				"name": "speedVariance"
			},
			{
				"type": "i",
				"name": "animTime"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"correctedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"uv",
							"y"
						]
					]]
				]
			],
			[
				13,
				"ang",
				[
					1,
					"rotation",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["ang"]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["ang"]
				]
			],
			[
				13,
				"centerX",
				[
					1,
					"aspect",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"centerY",
				[5, "0.5"]
			],
			[
				13,
				"cx",
				[
					1,
					[
						7,
						"correctedUV",
						"x"
					],
					"-",
					"centerX"
				]
			],
			[
				13,
				"cy",
				[
					1,
					[
						7,
						"correctedUV",
						"y"
					],
					"-",
					"centerY"
				]
			],
			[
				13,
				"rotatedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								1,
								"cx",
								"*",
								"cosR"
							],
							"-",
							[
								1,
								"cy",
								"*",
								"sinR"
							]
						],
						"+",
						"centerX"
					], [
						1,
						[
							1,
							[
								1,
								"cx",
								"*",
								"sinR"
							],
							"+",
							[
								1,
								"cy",
								"*",
								"cosR"
							]
						],
						"+",
						"centerY"
					]]
				]
			],
			[
				13,
				"p",
				[
					6,
					[
						7,
						"rotatedUV",
						"mul"
					],
					["cells"]
				]
			],
			[
				13,
				"ps",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"p",
							"x"
						],
						"-",
						[
							1,
							[
								7,
								"p",
								"y"
							],
							"/",
							"SQRT3"
						]
					], [
						1,
						[
							7,
							"p",
							"y"
						],
						"*",
						"TWO_OVER_SQRT3"
					]]
				]
			],
			[
				13,
				"rowIndex",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						7,
						"ps",
						"y"
					]]
				]
			],
			[
				13,
				"rowHash",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								"rowIndex",
								"*",
								[5, "127.1"]
							]]
						],
						"*",
						[5, "43758.5453"]
					]]
				]
			],
			[
				13,
				"speedMod",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						[
							1,
							[
								1,
								"rowHash",
								"-",
								[5, "0.5"]
							],
							"*",
							[5, "4"]
						],
						"*",
						"speedVariance"
					]
				]
			],
			[
				13,
				"drift",
				[
					1,
					"animTime",
					"*",
					"speedMod"
				]
			],
			[
				13,
				"psShifted",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"ps",
							"x"
						],
						"-",
						"drift"
					], [
						7,
						"ps",
						"y"
					]]
				]
			],
			[
				13,
				"cell",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["psShifted"]
				]
			],
			[
				13,
				"f",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["psShifted"]
				]
			],
			[
				13,
				"upper",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					[[5, "1"], [
						1,
						[
							7,
							"f",
							"x"
						],
						"+",
						[
							7,
							"f",
							"y"
						]
					]]
				]
			],
			[
				13,
				"triId",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								7,
								"cell",
								"x"
							],
							"*",
							[5, "2"]
						],
						"+",
						"upper"
					], [
						7,
						"cell",
						"y"
					]]
				]
			],
			[
				13,
				"q0",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						6,
						[
							7,
							"triId",
							"mul"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "123.34"], [5, "345.45"]]
						]]
					]]
				]
			],
			[
				13,
				"q1",
				[
					6,
					[
						7,
						"q0",
						"add"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["q0", [
							6,
							[
								7,
								"q0",
								"add"
							],
							[[5, "34.345"]]
						]]
					]]
				]
			],
			[
				13,
				"rand",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						[
							7,
							"q1",
							"x"
						],
						"*",
						[
							7,
							"q1",
							"y"
						]
					]]
				]
			],
			[
				13,
				"variedFactor",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						[5, "1"],
						"+",
						[
							1,
							[
								1,
								[
									1,
									"rand",
									"-",
									[5, "0.5"]
								],
								"*",
								[5, "2"]
							],
							"*",
							"variation"
						]
					], [5, "0"]]
				]
			],
			[
				13,
				"dEdge",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					[[
						6,
						[
							7,
							"std",
							"min"
						],
						[[
							6,
							[
								7,
								"std",
								"min"
							],
							[[
								7,
								"f",
								"x"
							], [
								1,
								[5, "1"],
								"-",
								[
									7,
									"f",
									"x"
								]
							]]
						], [
							6,
							[
								7,
								"std",
								"min"
							],
							[[
								7,
								"f",
								"y"
							], [
								1,
								[5, "1"],
								"-",
								[
									7,
									"f",
									"y"
								]
							]]
						]]
					], [
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							1,
							[
								1,
								[
									7,
									"f",
									"x"
								],
								"+",
								[
									7,
									"f",
									"y"
								]
							],
							"-",
							[5, "1"]
						]]
					]]
				]
			],
			[
				13,
				"lineWidth",
				[
					1,
					"thickness",
					"*",
					[5, "0.02"]
				]
			],
			[
				13,
				"aa",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"max"
						],
						[[
							6,
							[
								7,
								"std",
								"fwidth"
							],
							[[
								7,
								"ps",
								"x"
							]]
						], [
							6,
							[
								7,
								"std",
								"fwidth"
							],
							[[
								7,
								"ps",
								"y"
							]]
						]]
					],
					"+",
					[
						1,
						"softness",
						"*",
						[5, "0.5"]
					]
				]
			],
			[
				13,
				"lineMask",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						[
							6,
							[
								7,
								"std",
								"smoothstep"
							],
							[
								[
									1,
									"lineWidth",
									"-",
									"aa"
								],
								[
									1,
									"lineWidth",
									"+",
									"aa"
								],
								"dEdge"
							]
						]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"step"
						],
						[[5, "0.0001"], "lineWidth"]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["lineMask", "variedFactor"]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std",
			"SQRT3",
			"TWO_OVER_SQRT3"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std,
		SQRT3,
		TWO_OVER_SQRT3
	})
}) && $.f)({}));
const componentDefinition = {
	name: "TriangularGrid",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Tiling grid of equilateral triangles with optional animated row offsets",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#1a1a1a",
			transform: transformColor,
			description: "Triangle fill color",
			ui: {
				type: "color",
				label: "Fill",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Grid line color",
			ui: {
				type: "color",
				label: "Lines",
				group: "Colors"
			}
		},
		cells: {
			default: 8,
			description: "Number of triangle rows down the shortest canvas edge",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 40,
				step: 1,
				label: "Cells",
				group: "Effect"
			}
		},
		thickness: {
			default: 1,
			description: "Thickness of the grid lines (0 = no lines)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 10,
				step: .1,
				label: "Thickness",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			transform: transformAngle,
			description: "Rotation of the grid in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Softness of the line edges (0 = crisp, 1 = very soft)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		variation: {
			default: 0,
			description: "Per-triangle random lightening/darkening of the fill color",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Variation",
				group: "Effect"
			}
		},
		speed: {
			default: 0,
			description: "Animates the triangle rows drifting horizontally (0 = static)",
			ui: {
				type: "range",
				min: -3,
				max: 3,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		speedVariance: {
			default: .3,
			description: "Per-row random speed variance for irregular drifting motion",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Speed Variance",
				group: "Animation"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, propValues } = params;
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		const field = call(triangularGridField, "triangularGridField", [
			uv,
			viewport,
			uniforms.cells,
			uniforms.thickness,
			uniforms.rotation,
			uniforms.softness,
			uniforms.variation,
			uniforms.speedVariance,
			t
		]);
		const lineMask = field.member("x");
		const variedFactor = field.member("y");
		const variedFill = vec4(uniforms.colorA.member("rgb").mul(variedFactor), uniforms.colorA.member("a"));
		return call(mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			variedFill,
			uniforms.colorB,
			lineMask
		]);
	}
};
var TriangularGrid_default = componentDefinition;
export { componentDefinition as n, triangularGridField as r, TriangularGrid_default as t };
