// shaders/dist/core/Tritone-CqYOMKFE.js
import { At as call, It as vec4, Ot as ZERO, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
const tritoneFactors = tgpu.fn([d.vec4f, d.f32], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, blendMid) => {
	"use gpu";
	const luminance = std.dot(color.xyz, d.vec3f(.299, .587, .114));
	const shadowToMid = std.smoothstep(__tsover_sub(blendMid, .25), blendMid, luminance);
	const midToHighlight = std.smoothstep(blendMid, __tsover_add(blendMid, .25), luminance);
	const finalT = std.smoothstep(__tsover_sub(blendMid, .1), __tsover_add(blendMid, .1), luminance);
	return d.vec3f(shadowToMid, midToHighlight, finalT);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}, {
			"type": "i",
			"name": "blendMid"
		}],
		"body": [0, [
			[
				13,
				"luminance",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					[[
						7,
						"color",
						"xyz"
					], [
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0.299"],
							[5, "0.587"],
							[5, "0.114"]
						]
					]]
				]
			],
			[
				13,
				"shadowToMid",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"blendMid",
							"-",
							[5, "0.25"]
						],
						"blendMid",
						"luminance"
					]
				]
			],
			[
				13,
				"midToHighlight",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						"blendMid",
						[
							1,
							"blendMid",
							"+",
							[5, "0.25"]
						],
						"luminance"
					]
				]
			],
			[
				13,
				"finalT",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[
							1,
							"blendMid",
							"-",
							[5, "0.1"]
						],
						[
							1,
							"blendMid",
							"+",
							[5, "0.1"]
						],
						"luminance"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec3f"
				],
				[
					"shadowToMid",
					"midToHighlight",
					"finalT"
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Tritone",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Map colors to three tones: shadows, midtones, highlights",
	requiresChild: true,
	props: {
		colorA: {
			default: "#ce1bea",
			transform: transformColor,
			description: "First color (used for shadows/darkest areas)",
			ui: {
				type: "color",
				label: "Color A (Shadows)",
				group: "Colors"
			}
		},
		colorB: {
			default: "#2fff00",
			transform: transformColor,
			description: "Second color (used for midtones)",
			ui: {
				type: "color",
				label: "Color B (Midtones)",
				group: "Colors"
			}
		},
		colorC: {
			default: "#ffff00",
			transform: transformColor,
			description: "Third color (used for highlights/brightest areas)",
			ui: {
				type: "color",
				label: "Color C (Highlights)",
				group: "Colors"
			}
		},
		blendMid: {
			default: .5,
			description: "Midpoint position between the three colors",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Midpoint",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: ({ uniforms, childNode, propValues }) => {
		if (!childNode) return ZERO;
		const factors = call(tritoneFactors, "tritoneFactors", [childNode, uniforms.blendMid]);
		const shadowToMid = factors.member("x");
		const midToHighlight = factors.member("y");
		const finalT = factors.member("z");
		const variant = mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear;
		return vec4(call(variant, "mixColors", [
			call(variant, "mixColors", [
				uniforms.colorA,
				uniforms.colorB,
				shadowToMid
			]),
			call(variant, "mixColors", [
				uniforms.colorB,
				uniforms.colorC,
				midToHighlight
			]),
			finalT
		]).member("rgb"), childNode.member("a"));
	}
};
var Tritone_default = componentDefinition;
export { componentDefinition as n, tritoneFactors as r, Tritone_default as t };
