// shaders/dist/core/Truchet-Bj0dM4pc.js
import { At as call, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
var DEG_TO_RAD = Math.PI / 180;
var truchetHash = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	const h = std.dot(p, d.vec2f(127.1, 311.7));
	return std.fract(__tsover_mul(std.sin(h), 43758.5453));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [[
			13,
			"h",
			[
				6,
				[
					7,
					"std",
					"dot"
				],
				["p", [
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[5, "127.1"], [5, "311.7"]]
				]]
			]
		], [10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				1,
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["h"]
				],
				"*",
				[5, "43758.5453"]
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const truchetField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, cells, thickness, rotation, softness, seed) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const correctedUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const ang = -__tsover_mul(rotation, DEG_TO_RAD);
	const cosR = std.cos(ang);
	const sinR = std.sin(ang);
	const centerX = __tsover_mul(aspect, .5);
	const centerY = .5;
	const cx = __tsover_sub(correctedUV.x, centerX);
	const cy = __tsover_sub(correctedUV.y, centerY);
	const scaledUV = d.vec2f(__tsover_add(__tsover_sub(__tsover_mul(cx, cosR), __tsover_mul(cy, sinR)), centerX), __tsover_add(__tsover_add(__tsover_mul(cx, sinR), __tsover_mul(cy, cosR)), centerY)).mul(cells);
	const cell = std.floor(scaledUV);
	const localUV = std.fract(scaledUV);
	const flip = truchetHash(d.vec2f(__tsover_add(cell.x, seed), __tsover_add(cell.y, seed))) > .5;
	const c1x = std.select(d.f32(0), d.f32(1), flip);
	const c2x = std.select(d.f32(1), d.f32(0), flip);
	const d1 = std.abs(__tsover_sub(std.length(d.vec2f(__tsover_sub(localUV.x, c1x), localUV.y)), .5));
	const d2 = std.abs(__tsover_sub(std.length(d.vec2f(__tsover_sub(localUV.x, c2x), __tsover_sub(localUV.y, 1))), .5));
	const arcDist = std.min(d1, d2);
	const lineWidth = __tsover_mul(thickness, .005);
	const pixel = __tsover_mul(std.fwidth(scaledUV.x), .5);
	const soft = __tsover_mul(softness, .5);
	const lo = std.max(__tsover_sub(lineWidth, soft), 0);
	const hi = __tsover_add(__tsover_add(lineWidth, pixel), soft);
	return __tsover_mul(__tsover_sub(1, std.smoothstep(lo, hi, arcDist)), std.step(1e-4, lineWidth));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "cells"
			},
			{
				"type": "i",
				"name": "thickness"
			},
			{
				"type": "i",
				"name": "rotation"
			},
			{
				"type": "i",
				"name": "softness"
			},
			{
				"type": "i",
				"name": "seed"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"correctedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"ang",
				[
					4,
					"-",
					[
						1,
						"rotation",
						"*",
						"DEG_TO_RAD"
					]
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["ang"]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["ang"]
				]
			],
			[
				13,
				"centerX",
				[
					1,
					"aspect",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"centerY",
				[5, "0.5"]
			],
			[
				13,
				"cx",
				[
					1,
					[
						7,
						"correctedUV",
						"x"
					],
					"-",
					"centerX"
				]
			],
			[
				13,
				"cy",
				[
					1,
					[
						7,
						"correctedUV",
						"y"
					],
					"-",
					"centerY"
				]
			],
			[
				13,
				"rotatedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								1,
								"cx",
								"*",
								"cosR"
							],
							"-",
							[
								1,
								"cy",
								"*",
								"sinR"
							]
						],
						"+",
						"centerX"
					], [
						1,
						[
							1,
							[
								1,
								"cx",
								"*",
								"sinR"
							],
							"+",
							[
								1,
								"cy",
								"*",
								"cosR"
							]
						],
						"+",
						"centerY"
					]]
				]
			],
			[
				13,
				"scaledUV",
				[
					6,
					[
						7,
						"rotatedUV",
						"mul"
					],
					["cells"]
				]
			],
			[
				13,
				"cell",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["scaledUV"]
				]
			],
			[
				13,
				"localUV",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["scaledUV"]
				]
			],
			[
				13,
				"h",
				[
					6,
					"truchetHash",
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								7,
								"cell",
								"x"
							],
							"+",
							"seed"
						], [
							1,
							[
								7,
								"cell",
								"y"
							],
							"+",
							"seed"
						]]
					]]
				]
			],
			[
				13,
				"flip",
				[
					1,
					"h",
					">",
					[5, "0.5"]
				]
			],
			[
				13,
				"c1x",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						"flip"
					]
				]
			],
			[
				13,
				"c2x",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "1"]]
						],
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							[[5, "0"]]
						],
						"flip"
					]
				]
			],
			[
				13,
				"d1",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"length"
							],
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									1,
									[
										7,
										"localUV",
										"x"
									],
									"-",
									"c1x"
								], [
									7,
									"localUV",
									"y"
								]]
							]]
						],
						"-",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"d2",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						1,
						[
							6,
							[
								7,
								"std",
								"length"
							],
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									1,
									[
										7,
										"localUV",
										"x"
									],
									"-",
									"c2x"
								], [
									1,
									[
										7,
										"localUV",
										"y"
									],
									"-",
									[5, "1"]
								]]
							]]
						],
						"-",
						[5, "0.5"]
					]]
				]
			],
			[
				13,
				"arcDist",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					["d1", "d2"]
				]
			],
			[
				13,
				"lineWidth",
				[
					1,
					"thickness",
					"*",
					[5, "0.005"]
				]
			],
			[
				13,
				"pixel",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fwidth"
						],
						[[
							7,
							"scaledUV",
							"x"
						]]
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"soft",
				[
					1,
					"softness",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"lo",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"lineWidth",
						"-",
						"soft"
					], [5, "0"]]
				]
			],
			[
				13,
				"hi",
				[
					1,
					[
						1,
						"lineWidth",
						"+",
						"pixel"
					],
					"+",
					"soft"
				]
			],
			[10, [
				1,
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							"lo",
							"hi",
							"arcDist"
						]
					]
				],
				"*",
				[
					6,
					[
						7,
						"std",
						"step"
					],
					[[5, "0.0001"], "lineWidth"]
				]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std",
			"truchetHash"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std,
		truchetHash
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Truchet",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Quarter-circle arc tiles that connect to form organic, maze-like flowing curves",
	acceptsUVContext: true,
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "Background color between the arcs",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Arc line color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		cells: {
			default: 10,
			description: "Number of tiles across the shortest canvas edge",
			ui: {
				type: ["range", "map"],
				min: 2,
				max: 40,
				step: 1,
				label: "Cells",
				group: "Effect"
			}
		},
		thickness: {
			default: 2,
			description: "Thickness of the arc lines",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .1,
				label: "Thickness",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			transform: transformAngle,
			description: "Rotation of the tiling in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Softness of the arc edges (0 = crisp, 1 = very soft)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Random seed — changes which tiles flip, producing a different maze pattern",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, propValues } = params;
		const lineMask = call(truchetField, "truchetField", [
			params.uvContext ?? params.ctx.uv,
			params.effectiveViewportSize ?? params.ctx.viewportSize,
			uniforms.cells,
			uniforms.thickness,
			uniforms.rotation,
			uniforms.softness,
			uniforms.seed
		]);
		return call(mixColorsVariants[propValues.colorSpace ?? 0] ?? mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			lineMask
		]);
	}
};
var Truchet_default = componentDefinition;
export { componentDefinition as n, truchetField as r, Truchet_default as t };
