// shaders/dist/core/Twirl-Y8OzHc49.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-CWOT0Wqh.js";
import { c as transformPosition, s as transformEdges } from "./transformations-fqzznt4g.js";
const twirlUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.vec2f,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, intensity, uv, aspect) => {
	"use gpu";
	const centerPos = d.vec2f(center.x, __tsover_sub(1, center.y));
	const delta = uv.sub(centerPos);
	const acd = d.vec2f(__tsover_mul(delta.x, aspect), delta.y);
	const angle = __tsover_mul(intensity, std.length(acd));
	const cosA = std.cos(angle);
	const sinA = std.sin(angle);
	const rotatedX = __tsover_sub(__tsover_mul(cosA, acd.x), __tsover_mul(sinA, acd.y));
	const rotatedY = __tsover_add(__tsover_mul(sinA, acd.x), __tsover_mul(cosA, acd.y));
	return d.vec2f(__tsover_add(__tsover_div(rotatedX, aspect), centerPos.x), __tsover_add(rotatedY, centerPos.y));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			}
		],
		"body": [0, [
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"center",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"delta",
				[
					6,
					[
						7,
						"uv",
						"sub"
					],
					["centerPos"]
				]
			],
			[
				13,
				"acd",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"delta",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"delta",
						"y"
					]]
				]
			],
			[
				13,
				"angle",
				[
					1,
					"intensity",
					"*",
					[
						6,
						[
							7,
							"std",
							"length"
						],
						["acd"]
					]
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angle"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angle"]
				]
			],
			[
				13,
				"rotatedX",
				[
					1,
					[
						1,
						"cosA",
						"*",
						[
							7,
							"acd",
							"x"
						]
					],
					"-",
					[
						1,
						"sinA",
						"*",
						[
							7,
							"acd",
							"y"
						]
					]
				]
			],
			[
				13,
				"rotatedY",
				[
					1,
					[
						1,
						"sinA",
						"*",
						[
							7,
							"acd",
							"x"
						]
					],
					"+",
					[
						1,
						"cosA",
						"*",
						[
							7,
							"acd",
							"y"
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						1,
						"rotatedX",
						"/",
						"aspect"
					],
					"+",
					[
						7,
						"centerPos",
						"x"
					]
				], [
					1,
					"rotatedY",
					"+",
					[
						7,
						"centerPos",
						"y"
					]
				]]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Twirl",
	category: "Distortions",
	description: "Rotate and twist content around a center point",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the twirl effect",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		intensity: {
			default: 1,
			description: "The strength of the twirl effect",
			ui: {
				type: ["range", "map"],
				min: -5,
				max: 5,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, propValues, convertToTexture }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Twirl shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(twirlUV, "twirlUV", [
			uniforms.center,
			uniforms.intensity,
			ctx.uv,
			ctx.aspect
		]), (uv) => texture.sample(uv), edgeMode)]);
	},
	uvRemap: ({ uv, mask, uniforms, aspect, propValues }) => {
		const edgeMode = propValues.edges ?? 0;
		return composeEdgeRemapExpr(call(twirlUV, "twirlUV", [
			uniforms.center,
			uniforms.intensity,
			uv,
			aspect
		]), mask, edgeMode);
	}
};
var Twirl_default = componentDefinition;
export { componentDefinition as n, twirlUV as r, Twirl_default as t };
