// shaders/dist/core/VHS-B28K9Bfc.js
import { At as call, Ot as ZERO, Pt as floatE, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
var FIELD_LINES = 487;
var hash2D = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	return std.fract(__tsover_mul(std.sin(std.dot(p, d.vec2f(12.9898, 78.233))), 43758.5453));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				1,
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["p", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "12.9898"], [5, "78.233"]]
						]]
					]]
				],
				"*",
				[5, "43758.5453"]
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
var smoothNoise2D = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	const i = std.floor(p);
	const f = std.fract(p);
	const u = f.mul(f).mul(d.vec2f(3, 3).sub(f.mul(2)));
	const a = hash2D(i);
	const b = hash2D(i.add(d.vec2f(1, 0)));
	const c = hash2D(i.add(d.vec2f(0, 1)));
	const dd = hash2D(i.add(d.vec2f(1, 1)));
	return std.mix(std.mix(a, b, u.x), std.mix(c, dd, u.x), u.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [
			[
				13,
				"i",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["p"]
				]
			],
			[
				13,
				"f",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["p"]
				]
			],
			[
				13,
				"u",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"f",
								"mul"
							],
							["f"]
						],
						"mul"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[5, "3"], [5, "3"]]
							],
							"sub"
						],
						[[
							6,
							[
								7,
								"f",
								"mul"
							],
							[[5, "2"]]
						]]
					]]
				]
			],
			[
				13,
				"a",
				[
					6,
					"hash2D",
					["i"]
				]
			],
			[
				13,
				"b",
				[
					6,
					"hash2D",
					[[
						6,
						[
							7,
							"i",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "1"], [5, "0"]]
						]]
					]]
				]
			],
			[
				13,
				"c",
				[
					6,
					"hash2D",
					[[
						6,
						[
							7,
							"i",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "0"], [5, "1"]]
						]]
					]]
				]
			],
			[
				13,
				"dd",
				[
					6,
					"hash2D",
					[[
						6,
						[
							7,
							"i",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "1"], [5, "1"]]
						]]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"mix"
				],
				[
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"a",
							"b",
							[
								7,
								"u",
								"x"
							]
						]
					],
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"c",
							"dd",
							[
								7,
								"u",
								"x"
							]
						]
					],
					[
						7,
						"u",
						"y"
					]
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"hash2D"
		]
	},
	externals: () => ({
		std,
		d,
		hash2D
	})
}) && $.f)({}));
const vhsSampleUVs = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, time, speed, wobble, scanlineNoise) => {
	"use gpu";
	const t = __tsover_mul(time, speed);
	const burst = std.smoothstep(.45, .8, smoothNoise2D(d.vec2f(__tsover_mul(t, .4), 0)));
	const rowInt = std.floor(__tsover_mul(uv.y, FIELD_LINES));
	const h0 = __tsover_mul(__tsover_sub(std.fract(__tsover_mul(std.sin(__tsover_add(__tsover_mul(rowInt, 12.9898), __tsover_mul(t, 78.233))), 43758.5453)), .5), 2);
	const h1 = __tsover_mul(__tsover_sub(std.fract(__tsover_mul(std.sin(__tsover_add(__tsover_mul(rowInt, 12.9898), __tsover_mul(__tsover_add(t, 69.42), 78.233))), 43758.5453)), .5), 2);
	const fineGate = __tsover_add(.25, __tsover_mul(burst, .75));
	const chromaRowOffset = __tsover_mul(__tsover_mul(__tsover_mul(h0, scanlineNoise), .008), fineGate);
	const lumaRowOffset = __tsover_mul(__tsover_mul(__tsover_mul(h1, scanlineNoise), .004), fineGate);
	const wave1 = __tsover_sub(smoothNoise2D(d.vec2f(__tsover_mul(uv.y, 3), __tsover_mul(t, .8))), .5);
	const wave2 = __tsover_sub(smoothNoise2D(d.vec2f(__tsover_mul(uv.y, 30), __tsover_mul(t, 6))), .5);
	const tapeWave = __tsover_mul(__tsover_add(__tsover_mul(wave1, .008), __tsover_mul(wave2, .002)), wobble);
	const creasePhase = std.smoothstep(.92, .99, std.sin(__tsover_sub(__tsover_mul(uv.y, 8), __tsover_mul(t, 3.77))));
	const creaseNoise = std.smoothstep(.3, 1, smoothNoise2D(d.vec2f(__tsover_mul(uv.y, 4.77), t)));
	const creaseShift = __tsover_mul(__tsover_mul(__tsover_mul(creasePhase, creaseNoise), wobble), -.018);
	const switchPhase = __tsover_sub(1, std.smoothstep(0, .06, uv.y));
	const switchNoise = __tsover_sub(smoothNoise2D(d.vec2f(__tsover_mul(uv.y, 60), __tsover_mul(t, 14))), .5);
	const switchX = __tsover_mul(__tsover_mul(__tsover_mul(switchPhase, switchNoise), wobble), .09);
	const switchY = __tsover_mul(__tsover_mul(__tsover_mul(switchPhase, wobble), burst), .02);
	const globalX = __tsover_add(__tsover_add(tapeWave, creaseShift), switchX);
	const sy = __tsover_add(uv.y, switchY);
	return d.vec4f(__tsover_add(__tsover_add(uv.x, globalX), lumaRowOffset), sy, __tsover_add(__tsover_add(uv.x, globalX), chromaRowOffset), sy);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "time"
			},
			{
				"type": "i",
				"name": "speed"
			},
			{
				"type": "i",
				"name": "wobble"
			},
			{
				"type": "i",
				"name": "scanlineNoise"
			}
		],
		"body": [0, [
			[
				13,
				"t",
				[
					1,
					"time",
					"*",
					"speed"
				]
			],
			[
				13,
				"burst",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.45"],
						[5, "0.8"],
						[
							6,
							"smoothNoise2D",
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									1,
									"t",
									"*",
									[5, "0.4"]
								], [5, "0"]]
							]]
						]
					]
				]
			],
			[
				13,
				"rowInt",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					[[
						1,
						[
							7,
							"uv",
							"y"
						],
						"*",
						"FIELD_LINES"
					]]
				]
			],
			[
				13,
				"h0",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										[
											1,
											"rowInt",
											"*",
											[5, "12.9898"]
										],
										"+",
										[
											1,
											"t",
											"*",
											[5, "78.233"]
										]
									]]
								],
								"*",
								[5, "43758.5453"]
							]]
						],
						"-",
						[5, "0.5"]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"h1",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"fract"
							],
							[[
								1,
								[
									6,
									[
										7,
										"std",
										"sin"
									],
									[[
										1,
										[
											1,
											"rowInt",
											"*",
											[5, "12.9898"]
										],
										"+",
										[
											1,
											[
												1,
												"t",
												"+",
												[5, "69.42"]
											],
											"*",
											[5, "78.233"]
										]
									]]
								],
								"*",
								[5, "43758.5453"]
							]]
						],
						"-",
						[5, "0.5"]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"fineGate",
				[
					1,
					[5, "0.25"],
					"+",
					[
						1,
						"burst",
						"*",
						[5, "0.75"]
					]
				]
			],
			[
				13,
				"chromaRowOffset",
				[
					1,
					[
						1,
						[
							1,
							"h0",
							"*",
							"scanlineNoise"
						],
						"*",
						[5, "0.008"]
					],
					"*",
					"fineGate"
				]
			],
			[
				13,
				"lumaRowOffset",
				[
					1,
					[
						1,
						[
							1,
							"h1",
							"*",
							"scanlineNoise"
						],
						"*",
						[5, "0.004"]
					],
					"*",
					"fineGate"
				]
			],
			[
				13,
				"wave1",
				[
					1,
					[
						6,
						"smoothNoise2D",
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								[
									7,
									"uv",
									"y"
								],
								"*",
								[5, "3"]
							], [
								1,
								"t",
								"*",
								[5, "0.8"]
							]]
						]]
					],
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"wave2",
				[
					1,
					[
						6,
						"smoothNoise2D",
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								[
									7,
									"uv",
									"y"
								],
								"*",
								[5, "30"]
							], [
								1,
								"t",
								"*",
								[5, "6"]
							]]
						]]
					],
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"tapeWave",
				[
					1,
					[
						1,
						[
							1,
							"wave1",
							"*",
							[5, "0.008"]
						],
						"+",
						[
							1,
							"wave2",
							"*",
							[5, "0.002"]
						]
					],
					"*",
					"wobble"
				]
			],
			[
				13,
				"creasePhase",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.92"],
						[5, "0.99"],
						[
							6,
							[
								7,
								"std",
								"sin"
							],
							[[
								1,
								[
									1,
									[
										7,
										"uv",
										"y"
									],
									"*",
									[5, "8"]
								],
								"-",
								[
									1,
									"t",
									"*",
									[5, "3.77"]
								]
							]]
						]
					]
				]
			],
			[
				13,
				"creaseNoise",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.3"],
						[5, "1"],
						[
							6,
							"smoothNoise2D",
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									1,
									[
										7,
										"uv",
										"y"
									],
									"*",
									[5, "4.77"]
								], "t"]
							]]
						]
					]
				]
			],
			[
				13,
				"creaseShift",
				[
					1,
					[
						1,
						[
							1,
							"creasePhase",
							"*",
							"creaseNoise"
						],
						"*",
						"wobble"
					],
					"*",
					[
						4,
						"-",
						[5, "0.018"]
					]
				]
			],
			[
				13,
				"switchPhase",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							[5, "0.06"],
							[
								7,
								"uv",
								"y"
							]
						]
					]
				]
			],
			[
				13,
				"switchNoise",
				[
					1,
					[
						6,
						"smoothNoise2D",
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								[
									7,
									"uv",
									"y"
								],
								"*",
								[5, "60"]
							], [
								1,
								"t",
								"*",
								[5, "14"]
							]]
						]]
					],
					"-",
					[5, "0.5"]
				]
			],
			[
				13,
				"switchX",
				[
					1,
					[
						1,
						[
							1,
							"switchPhase",
							"*",
							"switchNoise"
						],
						"*",
						"wobble"
					],
					"*",
					[5, "0.09"]
				]
			],
			[
				13,
				"switchY",
				[
					1,
					[
						1,
						[
							1,
							"switchPhase",
							"*",
							"wobble"
						],
						"*",
						"burst"
					],
					"*",
					[5, "0.02"]
				]
			],
			[
				13,
				"globalX",
				[
					1,
					[
						1,
						"tapeWave",
						"+",
						"creaseShift"
					],
					"+",
					"switchX"
				]
			],
			[
				13,
				"sy",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"+",
					"switchY"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						1,
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"+",
							"globalX"
						],
						"+",
						"lumaRowOffset"
					],
					"sy",
					[
						1,
						[
							1,
							[
								7,
								"uv",
								"x"
							],
							"+",
							"globalX"
						],
						"+",
						"chromaRowOffset"
					],
					"sy"
				]
			]]
		]],
		"externalNames": [
			"std",
			"smoothNoise2D",
			"d",
			"FIELD_LINES"
		]
	},
	externals: () => ({
		std,
		smoothNoise2D,
		d,
		FIELD_LINES
	})
}) && $.f)({}));
const vhsAcBeat = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, time, speed, wobble) => {
	"use gpu";
	const t = __tsover_mul(time, speed);
	return __tsover_add(1, __tsover_mul(__tsover_mul(std.cos(__tsover_add(__tsover_mul(std.mod(t, 6.2831853), 2), __tsover_mul(uv.y, .5))), .015), wobble));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "time"
			},
			{
				"type": "i",
				"name": "speed"
			},
			{
				"type": "i",
				"name": "wobble"
			}
		],
		"body": [0, [[
			13,
			"t",
			[
				1,
				"time",
				"*",
				"speed"
			]
		], [10, [
			1,
			[5, "1"],
			"+",
			[
				1,
				[
					1,
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						[[
							1,
							[
								1,
								[
									6,
									[
										7,
										"std",
										"mod"
									],
									["t", [5, "6.2831853"]]
								],
								"*",
								[5, "2"]
							],
							"+",
							[
								1,
								[
									7,
									"uv",
									"y"
								],
								"*",
								[5, "0.5"]
							]
						]]
					],
					"*",
					[5, "0.015"]
				],
				"*",
				"wobble"
			]
		]]]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const vhsChromaTapUV = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((chromaUV, smearScale, i) => {
	"use gpu";
	return d.vec2f(__tsover_add(chromaUV.x, __tsover_mul(__tsover_mul(i, -1), smearScale)), chromaUV.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "chromaUV"
			},
			{
				"type": "i",
				"name": "smearScale"
			},
			{
				"type": "i",
				"name": "i"
			}
		],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					7,
					"chromaUV",
					"x"
				],
				"+",
				[
					1,
					[
						1,
						"i",
						"*",
						[
							4,
							"-",
							[5, "1"]
						]
					],
					"*",
					"smearScale"
				]
			], [
				7,
				"chromaUV",
				"y"
			]]
		]]]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const vhsCompose = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((luma, c1, c2, c3, c4, c5, acBeat) => {
	"use gpu";
	const sharpY = std.dot(luma.xyz, d.vec3f(.299, .587, .114));
	const iAxis = d.vec3f(.596, -.274, -.322);
	const qAxis = d.vec3f(.211, -.523, .312);
	const accI = __tsover_add(__tsover_add(__tsover_add(__tsover_add(__tsover_mul(std.dot(c1.xyz, iAxis), .06666666666666667), __tsover_mul(std.dot(c2.xyz, iAxis), .13333333333333333)), __tsover_mul(std.dot(c3.xyz, iAxis), .2)), __tsover_mul(std.dot(c4.xyz, iAxis), .26666666666666666)), __tsover_mul(std.dot(c5.xyz, iAxis), .3333333333333333));
	const accQ = __tsover_add(__tsover_add(__tsover_add(__tsover_add(__tsover_mul(std.dot(c1.xyz, qAxis), .06666666666666667), __tsover_mul(std.dot(c2.xyz, qAxis), .13333333333333333)), __tsover_mul(std.dot(c3.xyz, qAxis), .2)), __tsover_mul(std.dot(c4.xyz, qAxis), .26666666666666666)), __tsover_mul(std.dot(c5.xyz, qAxis), .3333333333333333));
	const finalRgb = d.vec3f(__tsover_add(__tsover_add(sharpY, __tsover_mul(accI, .956)), __tsover_mul(accQ, .621)), __tsover_sub(__tsover_sub(sharpY, __tsover_mul(accI, .272)), __tsover_mul(accQ, .647)), __tsover_add(__tsover_sub(sharpY, __tsover_mul(accI, 1.106)), __tsover_mul(accQ, 1.703)));
	const clamped = std.clamp(finalRgb.mul(acBeat), d.vec3f(0, 0, 0), d.vec3f(1, 1, 1));
	return d.vec4f(clamped.x, clamped.y, clamped.z, luma.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "luma"
			},
			{
				"type": "i",
				"name": "c1"
			},
			{
				"type": "i",
				"name": "c2"
			},
			{
				"type": "i",
				"name": "c3"
			},
			{
				"type": "i",
				"name": "c4"
			},
			{
				"type": "i",
				"name": "c5"
			},
			{
				"type": "i",
				"name": "acBeat"
			}
		],
		"body": [0, [
			[
				13,
				"sharpY",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					[[
						7,
						"luma",
						"xyz"
					], [
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0.299"],
							[5, "0.587"],
							[5, "0.114"]
						]
					]]
				]
			],
			[
				13,
				"iAxis",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0.596"],
						[
							4,
							"-",
							[5, "0.274"]
						],
						[
							4,
							"-",
							[5, "0.322"]
						]
					]
				]
			],
			[
				13,
				"qAxis",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0.211"],
						[
							4,
							"-",
							[5, "0.523"]
						],
						[5, "0.312"]
					]
				]
			],
			[
				13,
				"accI",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									[
										6,
										[
											7,
											"std",
											"dot"
										],
										[[
											7,
											"c1",
											"xyz"
										], "iAxis"]
									],
									"*",
									[5, "0.06666666666666667"]
								],
								"+",
								[
									1,
									[
										6,
										[
											7,
											"std",
											"dot"
										],
										[[
											7,
											"c2",
											"xyz"
										], "iAxis"]
									],
									"*",
									[5, "0.13333333333333333"]
								]
							],
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"dot"
									],
									[[
										7,
										"c3",
										"xyz"
									], "iAxis"]
								],
								"*",
								[5, "0.2"]
							]
						],
						"+",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"dot"
								],
								[[
									7,
									"c4",
									"xyz"
								], "iAxis"]
							],
							"*",
							[5, "0.26666666666666666"]
						]
					],
					"+",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							[[
								7,
								"c5",
								"xyz"
							], "iAxis"]
						],
						"*",
						[5, "0.3333333333333333"]
					]
				]
			],
			[
				13,
				"accQ",
				[
					1,
					[
						1,
						[
							1,
							[
								1,
								[
									1,
									[
										6,
										[
											7,
											"std",
											"dot"
										],
										[[
											7,
											"c1",
											"xyz"
										], "qAxis"]
									],
									"*",
									[5, "0.06666666666666667"]
								],
								"+",
								[
									1,
									[
										6,
										[
											7,
											"std",
											"dot"
										],
										[[
											7,
											"c2",
											"xyz"
										], "qAxis"]
									],
									"*",
									[5, "0.13333333333333333"]
								]
							],
							"+",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"dot"
									],
									[[
										7,
										"c3",
										"xyz"
									], "qAxis"]
								],
								"*",
								[5, "0.2"]
							]
						],
						"+",
						[
							1,
							[
								6,
								[
									7,
									"std",
									"dot"
								],
								[[
									7,
									"c4",
									"xyz"
								], "qAxis"]
							],
							"*",
							[5, "0.26666666666666666"]
						]
					],
					"+",
					[
						1,
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							[[
								7,
								"c5",
								"xyz"
							], "qAxis"]
						],
						"*",
						[5, "0.3333333333333333"]
					]
				]
			],
			[
				13,
				"finalRgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								1,
								"sharpY",
								"+",
								[
									1,
									"accI",
									"*",
									[5, "0.956"]
								]
							],
							"+",
							[
								1,
								"accQ",
								"*",
								[5, "0.621"]
							]
						],
						[
							1,
							[
								1,
								"sharpY",
								"-",
								[
									1,
									"accI",
									"*",
									[5, "0.272"]
								]
							],
							"-",
							[
								1,
								"accQ",
								"*",
								[5, "0.647"]
							]
						],
						[
							1,
							[
								1,
								"sharpY",
								"-",
								[
									1,
									"accI",
									"*",
									[5, "1.106"]
								]
							],
							"+",
							[
								1,
								"accQ",
								"*",
								[5, "1.703"]
							]
						]
					]
				]
			],
			[
				13,
				"clamped",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"finalRgb",
								"mul"
							],
							["acBeat"]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0"],
								[5, "0"],
								[5, "0"]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1"],
								[5, "1"],
								[5, "1"]
							]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"clamped",
						"x"
					],
					[
						7,
						"clamped",
						"y"
					],
					[
						7,
						"clamped",
						"z"
					],
					[
						7,
						"luma",
						"w"
					]
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "VHS",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Analog VHS tape with intermittent tape damage, chroma bleed, and per-scanline noise",
	requiresRTT: true,
	requiresChild: true,
	props: {
		wobble: {
			default: 1,
			description: "Overall amount of tape damage — waves, creases, and head-switching noise. Bursts on and off organically over time.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .01,
				label: "Wobble",
				group: "Effect"
			}
		},
		scanlineNoise: {
			default: .6,
			description: "Per-scanline fine chroma/luma jitter. Adds the classic horizontal-streak detail.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Scanline Noise",
				group: "Effect"
			}
		},
		smear: {
			default: .2,
			description: "Horizontal chroma smear (color bleed) amount. Positive trails colour to the right (classic VHS), negative trails it to the left.",
			ui: {
				type: ["range", "map"],
				min: -2,
				max: 2,
				step: .01,
				label: "Smear",
				group: "Effect"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed of the tape effects.",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .1,
				label: "Speed",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const uvs = call(vhsSampleUVs, "vhsSampleUVs", [
			ctx.uv,
			ctx.time,
			uniforms.speed,
			uniforms.wobble,
			uniforms.scanlineNoise
		]);
		const lumaUV = uvs.member("xy");
		const chromaUV = uvs.member("zw");
		const lumaSample = texture.sample(lumaUV);
		const smearScale = uniforms.smear.mul(.0075);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [call(vhsCompose, "vhsCompose", [
			lumaSample,
			texture.sample(call(vhsChromaTapUV, "vhsChromaTapUV", [
				chromaUV,
				smearScale,
				floatE(1)
			])),
			texture.sample(call(vhsChromaTapUV, "vhsChromaTapUV", [
				chromaUV,
				smearScale,
				floatE(2)
			])),
			texture.sample(call(vhsChromaTapUV, "vhsChromaTapUV", [
				chromaUV,
				smearScale,
				floatE(3)
			])),
			texture.sample(call(vhsChromaTapUV, "vhsChromaTapUV", [
				chromaUV,
				smearScale,
				floatE(4)
			])),
			texture.sample(call(vhsChromaTapUV, "vhsChromaTapUV", [
				chromaUV,
				smearScale,
				floatE(5)
			])),
			call(vhsAcBeat, "vhsAcBeat", [
				ctx.uv,
				ctx.time,
				uniforms.speed,
				uniforms.wobble
			])
		])]);
	}
};
var VHS_default = componentDefinition;
export { vhsCompose as a, vhsChromaTapUV as i, componentDefinition as n, vhsSampleUVs as o, vhsAcBeat as r, VHS_default as t };
