// shaders/dist/core/Vibrance-Bxbm2XXk.js
import { At as call, Ot as ZERO, R as std, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
const vibranceColor = tgpu.fn([d.vec4f, d.f32], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, adjustment) => {
	"use gpu";
	const r = color.x;
	const g = color.y;
	const b = color.z;
	const average = __tsover_div(__tsover_add(__tsover_add(r, g), b), 3);
	const mx = std.max(r, std.max(g, b));
	const amt = __tsover_mul(__tsover_mul(__tsover_sub(mx, average), adjustment), -3);
	const rgb = d.vec3f(r, g, b);
	const vibrant = std.mix(rgb, d.vec3f(mx, mx, mx), d.vec3f(amt));
	return d.vec4f(vibrant.x, vibrant.y, vibrant.z, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "color"
		}, {
			"type": "i",
			"name": "adjustment"
		}],
		"body": [0, [
			[
				13,
				"r",
				[
					7,
					"color",
					"x"
				]
			],
			[
				13,
				"g",
				[
					7,
					"color",
					"y"
				]
			],
			[
				13,
				"b",
				[
					7,
					"color",
					"z"
				]
			],
			[
				13,
				"average",
				[
					1,
					[
						1,
						[
							1,
							"r",
							"+",
							"g"
						],
						"+",
						"b"
					],
					"/",
					[5, "3"]
				]
			],
			[
				13,
				"mx",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					["r", [
						6,
						[
							7,
							"std",
							"max"
						],
						["g", "b"]
					]]
				]
			],
			[
				13,
				"amt",
				[
					1,
					[
						1,
						[
							1,
							"mx",
							"-",
							"average"
						],
						"*",
						"adjustment"
					],
					"*",
					[
						4,
						"-",
						[5, "3"]
					]
				]
			],
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						"r",
						"g",
						"b"
					]
				]
			],
			[
				13,
				"vibrant",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"rgb",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								"mx",
								"mx",
								"mx"
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["amt"]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"vibrant",
						"x"
					],
					[
						7,
						"vibrant",
						"y"
					],
					[
						7,
						"vibrant",
						"z"
					],
					[
						7,
						"color",
						"w"
					]
				]
			]]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Vibrance",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Adjustments",
	description: "Selective saturation adjustment protecting skin tones",
	requiresChild: true,
	props: { intensity: {
		default: 0,
		description: "The intensity of the vibrance effect",
		ui: {
			type: ["range", "map"],
			min: -2,
			max: 2,
			step: .1,
			label: "Intensity",
			group: "Adjustments"
		}
	} },
	fragment: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Vibrance shader.");
			return ZERO;
		}
		return call(vibranceColor, "vibranceColor", [childNode, uniforms.intensity]);
	}
};
var Vibrance_default = componentDefinition;
export { componentDefinition as n, vibranceColor as r, Vibrance_default as t };
