// shaders/dist/core/Vignette-DhIYUi7B.js
import { At as call, Ot as ZERO, R as std, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
const vignetteColor = tgpu.fn([
	d.vec4f,
	d.vec2f,
	d.f32,
	d.vec2f,
	d.vec4f,
	d.f32,
	d.f32,
	d.f32
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((color, uv, aspect, center, vigColor, radius, falloff, intensity) => {
	"use gpu";
	const aspectUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const centerPos = d.vec2f(__tsover_mul(center.x, aspect), __tsover_sub(1, center.y));
	const dist = std.length(aspectUV.sub(centerPos));
	const vignetteMask = std.smoothstep(radius, __tsover_add(radius, falloff), dist);
	const rgb = d.vec3f(color.x, color.y, color.z);
	const vigRgb = d.vec3f(vigColor.x, vigColor.y, vigColor.z);
	const result = std.mix(rgb, vigRgb, d.vec3f(__tsover_mul(vignetteMask, intensity)));
	return d.vec4f(result.x, result.y, result.z, color.w);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "color"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "vigColor"
			},
			{
				"type": "i",
				"name": "radius"
			},
			{
				"type": "i",
				"name": "falloff"
			},
			{
				"type": "i",
				"name": "intensity"
			}
		],
		"body": [0, [
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"center",
							"x"
						],
						"*",
						"aspect"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"dist",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							"aspectUV",
							"sub"
						],
						["centerPos"]
					]]
				]
			],
			[
				13,
				"vignetteMask",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						"radius",
						[
							1,
							"radius",
							"+",
							"falloff"
						],
						"dist"
					]
				]
			],
			[
				13,
				"rgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"color",
							"x"
						],
						[
							7,
							"color",
							"y"
						],
						[
							7,
							"color",
							"z"
						]
					]
				]
			],
			[
				13,
				"vigRgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"vigColor",
							"x"
						],
						[
							7,
							"vigColor",
							"y"
						],
						[
							7,
							"vigColor",
							"z"
						]
					]
				]
			],
			[
				13,
				"result",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"rgb",
						"vigRgb",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[
								1,
								"vignetteMask",
								"*",
								"intensity"
							]]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"result",
						"x"
					],
					[
						7,
						"result",
						"y"
					],
					[
						7,
						"result",
						"z"
					],
					[
						7,
						"color",
						"w"
					]
				]
			]]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Vignette",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Darkens or tints the edges of the frame, drawing attention toward the center",
	requiresChild: true,
	props: {
		color: {
			default: "#000000",
			transform: transformColor,
			description: "Color of the vignette at the edges",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center of the clear area where the vignette begins",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		radius: {
			default: .5,
			description: "Distance from center where the vignette begins to fade in",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1.5,
				step: .01,
				label: "Radius",
				group: "Effect"
			}
		},
		falloff: {
			default: .5,
			description: "Width of the transition zone from clear to full vignette",
			ui: {
				type: ["range", "map"],
				min: .01,
				max: 1.5,
				step: .01,
				label: "Falloff",
				group: "Effect"
			}
		},
		intensity: {
			default: 1,
			description: "Strength of the vignette effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Intensity",
				group: "Effect"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Vignette shader.");
			return ZERO;
		}
		return call(vignetteColor, "vignetteColor", [
			childNode,
			ctx.uv,
			ctx.aspect,
			uniforms.center,
			uniforms.color,
			uniforms.radius,
			uniforms.falloff,
			uniforms.intensity
		]);
	}
};
var Vignette_default = componentDefinition;
export { componentDefinition as n, vignetteColor as r, Vignette_default as t };
