// shaders/dist/core/Voronoi-c1kHPUka.js
import { At as call, Ft as mixExpr, It as vec4, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, gt as mixColorStopsRuntime, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { m as colorStopsPropConfig, t as animatedTime } from "./porters-Dg80Yg90.js";
var voronoiHash = tgpu.fn([d.vec2f], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	const h = d.vec2f(std.dot(p, d.vec2f(127.1, 311.7)), std.dot(p, d.vec2f(269.5, 183.3)));
	return std.fract(std.sin(h).mul(43758.5453));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [[
			13,
			"h",
			[
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					6,
					[
						7,
						"std",
						"dot"
					],
					["p", [
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "127.1"], [5, "311.7"]]
					]]
				], [
					6,
					[
						7,
						"std",
						"dot"
					],
					["p", [
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "269.5"], [5, "183.3"]]
					]]
				]]
			]
		], [10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				6,
				[
					7,
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						["h"]
					],
					"mul"
				],
				[[5, "43758.5453"]]
			]]
		]]]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const voronoiField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, scale, animT, seed, edgeIntensity, edgeSoftness) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const scaledUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y).mul(scale);
	const cell = std.floor(scaledUV);
	const localUV = std.fract(scaledUV);
	let d1 = d.f32(10);
	let d2 = d.f32(10);
	for (let ny = -1; ny <= 1; ny++) for (let nx = -1; nx <= 1; nx++) {
		const offset = d.vec2f(d.f32(nx), d.f32(ny));
		const h = voronoiHash(d.vec2f(__tsover_add(__tsover_add(cell.x, offset.x), seed), __tsover_add(__tsover_add(cell.y, offset.y), seed)));
		const pointX = std.clamp(__tsover_add(h.x, __tsover_mul(std.sin(__tsover_add(animT, __tsover_mul(h.x, 6.28))), .15)), .05, .95);
		const pointY = std.clamp(__tsover_add(h.y, __tsover_mul(std.cos(__tsover_add(__tsover_mul(animT, .7), __tsover_mul(h.y, 6.28))), .15)), .05, .95);
		const dist = std.length(d.vec2f(__tsover_sub(localUV.x, __tsover_add(offset.x, pointX)), __tsover_sub(localUV.y, __tsover_add(offset.y, pointY))));
		const isCloser = dist < d1;
		d2 = std.select(std.select(d2, dist, dist < d2), d1, isCloser);
		d1 = std.select(d1, dist, isCloser);
	}
	const safeSum = std.max(__tsover_add(d1, d2), 1e-4);
	const normalizedDist = std.clamp(__tsover_div(__tsover_mul(d1, 2), safeSum), 0, 1);
	const gradientPower = __tsover_sub(4, __tsover_mul(edgeIntensity, 3));
	const tColor = std.clamp(std.pow(normalizedDist, gradientPower), 0, 1);
	const scaledEdgeSoftness = __tsover_div(__tsover_mul(edgeSoftness, scale), 6);
	const edgeMetric = __tsover_div(__tsover_sub(d2, d1), safeSum);
	const edgeMask = std.smoothstep(0, __tsover_add(scaledEdgeSoftness, .001), edgeMetric);
	return d.vec2f(tColor, edgeMask);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "animT"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "edgeIntensity"
			},
			{
				"type": "i",
				"name": "edgeSoftness"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"scaledUV",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								[
									7,
									"uv",
									"x"
								],
								"*",
								"aspect"
							], [
								7,
								"uv",
								"y"
							]]
						],
						"mul"
					],
					["scale"]
				]
			],
			[
				13,
				"cell",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["scaledUV"]
				]
			],
			[
				13,
				"localUV",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["scaledUV"]
				]
			],
			[
				12,
				"d1",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "10"]]
				]
			],
			[
				12,
				"d2",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "10"]]
				]
			],
			[
				14,
				[
					12,
					"ny",
					[
						4,
						"-",
						[5, "1"]
					]
				],
				[
					1,
					"ny",
					"<=",
					[5, "1"]
				],
				[
					102,
					"++",
					"ny"
				],
				[0, [[
					14,
					[
						12,
						"nx",
						[
							4,
							"-",
							[5, "1"]
						]
					],
					[
						1,
						"nx",
						"<=",
						[5, "1"]
					],
					[
						102,
						"++",
						"nx"
					],
					[0, [
						[
							13,
							"offset",
							[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									6,
									[
										7,
										"d",
										"f32"
									],
									["nx"]
								], [
									6,
									[
										7,
										"d",
										"f32"
									],
									["ny"]
								]]
							]
						],
						[
							13,
							"h",
							[
								6,
								"voronoiHash",
								[[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									[[
										1,
										[
											1,
											[
												7,
												"cell",
												"x"
											],
											"+",
											[
												7,
												"offset",
												"x"
											]
										],
										"+",
										"seed"
									], [
										1,
										[
											1,
											[
												7,
												"cell",
												"y"
											],
											"+",
											[
												7,
												"offset",
												"y"
											]
										],
										"+",
										"seed"
									]]
								]]
							]
						],
						[
							13,
							"pointX",
							[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									[
										1,
										[
											7,
											"h",
											"x"
										],
										"+",
										[
											1,
											[
												6,
												[
													7,
													"std",
													"sin"
												],
												[[
													1,
													"animT",
													"+",
													[
														1,
														[
															7,
															"h",
															"x"
														],
														"*",
														[5, "6.28"]
													]
												]]
											],
											"*",
											[5, "0.15"]
										]
									],
									[5, "0.05"],
									[5, "0.95"]
								]
							]
						],
						[
							13,
							"pointY",
							[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									[
										1,
										[
											7,
											"h",
											"y"
										],
										"+",
										[
											1,
											[
												6,
												[
													7,
													"std",
													"cos"
												],
												[[
													1,
													[
														1,
														"animT",
														"*",
														[5, "0.7"]
													],
													"+",
													[
														1,
														[
															7,
															"h",
															"y"
														],
														"*",
														[5, "6.28"]
													]
												]]
											],
											"*",
											[5, "0.15"]
										]
									],
									[5, "0.05"],
									[5, "0.95"]
								]
							]
						],
						[
							13,
							"dist",
							[
								6,
								[
									7,
									"std",
									"length"
								],
								[[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									[[
										1,
										[
											7,
											"localUV",
											"x"
										],
										"-",
										[
											1,
											[
												7,
												"offset",
												"x"
											],
											"+",
											"pointX"
										]
									], [
										1,
										[
											7,
											"localUV",
											"y"
										],
										"-",
										[
											1,
											[
												7,
												"offset",
												"y"
											],
											"+",
											"pointY"
										]
									]]
								]]
							]
						],
						[
							13,
							"isCloser",
							[
								1,
								"dist",
								"<",
								"d1"
							]
						],
						[
							13,
							"nd2",
							[
								6,
								[
									7,
									"std",
									"select"
								],
								[
									[
										6,
										[
											7,
											"std",
											"select"
										],
										[
											"d2",
											"dist",
											[
												1,
												"dist",
												"<",
												"d2"
											]
										]
									],
									"d1",
									"isCloser"
								]
							]
						],
						[
							2,
							"d2",
							"=",
							"nd2"
						],
						[
							2,
							"d1",
							"=",
							[
								6,
								[
									7,
									"std",
									"select"
								],
								[
									"d1",
									"dist",
									"isCloser"
								]
							]
						]
					]]
				]]]
			],
			[
				13,
				"safeSum",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"d1",
						"+",
						"d2"
					], [5, "0.0001"]]
				]
			],
			[
				13,
				"normalizedDist",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"d1",
								"*",
								[5, "2"]
							],
							"/",
							"safeSum"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"gradientPower",
				[
					1,
					[5, "4"],
					"-",
					[
						1,
						"edgeIntensity",
						"*",
						[5, "3"]
					]
				]
			],
			[
				13,
				"tColor",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							["normalizedDist", "gradientPower"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"scaledEdgeSoftness",
				[
					1,
					[
						1,
						"edgeSoftness",
						"*",
						"scale"
					],
					"/",
					[5, "6"]
				]
			],
			[
				13,
				"edgeMetric",
				[
					1,
					[
						1,
						"d2",
						"-",
						"d1"
					],
					"/",
					"safeSum"
				]
			],
			[
				13,
				"edgeMask",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[
							1,
							"scaledEdgeSoftness",
							"+",
							[5, "0.001"]
						],
						"edgeMetric"
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				["tColor", "edgeMask"]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"voronoiHash"
		]
	},
	externals: () => ({
		d,
		std,
		voronoiHash
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Voronoi",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Cellular pattern where each pixel is colored by its distance to the nearest of many scattered points",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#3186cf",
			transform: transformColor,
			description: "Color near each cell's center point",
			ui: {
				type: "color",
				label: "Color A (Center)",
				group: "Colors"
			}
		},
		colorB: {
			default: "#fc02dd",
			transform: transformColor,
			description: "Color at cell boundaries, far from any center point",
			ui: {
				type: "color",
				label: "Color B (Edge)",
				group: "Colors"
			}
		},
		stops: colorStopsPropConfig(),
		colorBorder: {
			default: "#000000",
			transform: transformColor,
			description: "Color of the cell boundary lines",
			ui: {
				type: "color",
				label: "Border Color",
				group: "Colors"
			}
		},
		scale: {
			default: 6,
			description: "Number of cells across the canvas",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 20,
				step: .5,
				label: "Scale",
				group: "Effect"
			}
		},
		speed: {
			default: .5,
			description: "Animation speed — how fast the cell points drift",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		seed: {
			default: 0,
			description: "Random seed — shifts the cell pattern without changing the overall structure",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		edgeIntensity: {
			default: .5,
			description: "Controls how much of the cell interior is filled by the edge color. Low = center color dominates with a sharp boundary. High = edge color spreads further into the cell.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Intensity",
				group: "Effect"
			}
		},
		edgeSoftness: {
			default: .05,
			description: "Width of the cell boundary lines.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .4,
				step: .005,
				label: "Edge Softness",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, propValues } = params;
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		const field = call(voronoiField, "voronoiField", [
			uv,
			viewport,
			uniforms.scale,
			t,
			uniforms.seed,
			uniforms.edgeIntensity,
			uniforms.edgeSoftness
		]);
		const tColor = field.member("x");
		const edgeMask = field.member("y");
		const colorSpaceMode = propValues.colorSpace ?? 0;
		const stopCount = propValues.stopCount ?? 0;
		let cellColor;
		if (stopCount > 1) cellColor = mixColorStopsRuntime(tColor, {
			colorsArray: uniforms.colorsArray,
			positionsArray: uniforms.positionsArray,
			convertedColorsArray: uniforms.convertedColorsArray,
			stopCount: uniforms.stopCount
		}, colorSpaceMode);
		else cellColor = call(mixColorsVariants[colorSpaceMode] ?? mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			tColor
		]);
		return vec4(mixExpr(uniforms.colorBorder.member("rgb"), cellColor.member("rgb"), edgeMask), cellColor.member("a"));
	}
};
var Voronoi_default = componentDefinition;
export { componentDefinition as n, voronoiField as r, Voronoi_default as t };
