// shaders/dist/core/Water-DcoqDm5J.js
import { At as call, It as vec4, Pt as floatE, R as std, T as sdf_exports, d as sdf3d_exports, et as mxNoiseFloat2, n as tgpu, on as d, rt as noise_exports } from "./coords-BlwgBpJf.js";
import { n as glass_exports } from "./glass-DdgMBWkF.js";
import { a as transformColor, c as transformPosition } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
import { t as shapeEffectBoundingBoxDeclaration } from "./shapeEffectBounds-BlxxsYMN.js";
var { sdfSpaceUV, offsetUV } = glass_exports;
var { createSvgSdfSampler, createAnalyticSdfSampler, driveAnalyticSubProps, ANALYTIC_SDF_EXTRA_FIELDS } = sdf_exports;
var { createVolumetricFieldComputeNode, buildVolumetricFieldSampler, VOLUMETRIC_FIELD_EXTRA_FIELDS, is3dShapeType, isVolumetricShapeType } = sdf3d_exports;
var DEG_TO_RAD = .017453292519943295;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
function resolveShapeType(shapeType, shape) {
	if (shapeType) return shapeType;
	if (shape && typeof shape === "object") return shape.type || "circleSDF";
	if (typeof shape === "string") try {
		return JSON.parse(shape).type || "circleSDF";
	} catch {
		return "circleSDF";
	}
	return "circleSDF";
}
const waterFbm = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv) => {
	"use gpu";
	const n0 = mxNoiseFloat2(uv);
	const n1 = __tsover_mul(mxNoiseFloat2(uv.mul(2.07).add(d.vec2f(4.3, 9.1))), .5);
	const n2 = __tsover_mul(mxNoiseFloat2(uv.mul(4.13).add(d.vec2f(1.7, 8.2))), .25);
	return __tsover_mul(__tsover_add(__tsover_add(n0, n1), n2), .52);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "uv"
		}],
		"body": [0, [
			[
				13,
				"n0",
				[
					6,
					[
						7,
						"noise",
						"mxNoiseFloat2"
					],
					["uv"]
				]
			],
			[
				13,
				"n1",
				[
					1,
					[
						6,
						[
							7,
							"noise",
							"mxNoiseFloat2"
						],
						[[
							6,
							[
								7,
								[
									6,
									[
										7,
										"uv",
										"mul"
									],
									[[5, "2.07"]]
								],
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[5, "4.3"], [5, "9.1"]]
							]]
						]]
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"n2",
				[
					1,
					[
						6,
						[
							7,
							"noise",
							"mxNoiseFloat2"
						],
						[[
							6,
							[
								7,
								[
									6,
									[
										7,
										"uv",
										"mul"
									],
									[[5, "4.13"]]
								],
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[5, "1.7"], [5, "8.2"]]
							]]
						]]
					],
					"*",
					[5, "0.25"]
				]
			],
			[10, [
				1,
				[
					1,
					[
						1,
						"n0",
						"+",
						"n1"
					],
					"+",
					"n2"
				],
				"*",
				[5, "0.52"]
			]]
		]],
		"externalNames": ["noise", "d"]
	},
	externals: () => ({
		noise: noise_exports,
		d
	})
}) && $.f)({}));
const waterNoise1 = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv) => {
	"use gpu";
	return __tsover_mul(mxNoiseFloat2(uv), .52);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "uv"
		}],
		"body": [0, [[10, [
			1,
			[
				6,
				[
					7,
					"noise",
					"mxNoiseFloat2"
				],
				["uv"]
			],
			"*",
			[5, "0.52"]
		]]]],
		"externalNames": ["noise"]
	},
	externals: () => ({ noise: noise_exports })
}) && $.f)({}));
const waterSky = tgpu.fn([
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((rx, ry, lx, ly, envC, envS, sharp) => {
	"use gpu";
	const ex = __tsover_sub(__tsover_mul(rx, envC), __tsover_mul(ry, envS));
	const ey = __tsover_add(__tsover_mul(rx, envS), __tsover_mul(ry, envC));
	const up = std.clamp(__tsover_add(__tsover_mul(ey, .5), .5), 0, 1);
	const horizon = d.vec3f(.9, .96, 1);
	const zenith = d.vec3f(.2, .46, .85);
	const below = d.vec3f(.04, .1, .16);
	const aboveCol = std.mix(horizon, zenith, d.vec3f(std.smoothstep(.5, 1, up)));
	let base = std.mix(below, aboveCol, d.vec3f(std.smoothstep(.4, .55, up)));
	const sunCenter = d.vec2f(__tsover_mul(lx, .8), __tsover_add(__tsover_mul(ly, .8), .12));
	const sunRad = std.mix(.5, .13, sharp);
	const sd = std.length(d.vec2f(ex, ey).sub(sunCenter));
	const sun = std.smoothstep(sunRad, __tsover_mul(sunRad, .15), sd);
	base = base.add(d.vec3f(1, .97, .88).mul(__tsover_mul(sun, 1.5)));
	return base;
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "rx"
			},
			{
				"type": "i",
				"name": "ry"
			},
			{
				"type": "i",
				"name": "lx"
			},
			{
				"type": "i",
				"name": "ly"
			},
			{
				"type": "i",
				"name": "envC"
			},
			{
				"type": "i",
				"name": "envS"
			},
			{
				"type": "i",
				"name": "sharp"
			}
		],
		"body": [0, [
			[
				13,
				"ex",
				[
					1,
					[
						1,
						"rx",
						"*",
						"envC"
					],
					"-",
					[
						1,
						"ry",
						"*",
						"envS"
					]
				]
			],
			[
				13,
				"ey",
				[
					1,
					[
						1,
						"rx",
						"*",
						"envS"
					],
					"+",
					[
						1,
						"ry",
						"*",
						"envC"
					]
				]
			],
			[
				13,
				"up",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"ey",
								"*",
								[5, "0.5"]
							],
							"+",
							[5, "0.5"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"horizon",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0.9"],
						[5, "0.96"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"zenith",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0.2"],
						[5, "0.46"],
						[5, "0.85"]
					]
				]
			],
			[
				13,
				"below",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[5, "0.04"],
						[5, "0.1"],
						[5, "0.16"]
					]
				]
			],
			[
				13,
				"aboveCol",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"horizon",
						"zenith",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									[5, "0.5"],
									[5, "1"],
									"up"
								]
							]]
						]
					]
				]
			],
			[
				12,
				"base",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"below",
						"aboveCol",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[
								6,
								[
									7,
									"std",
									"smoothstep"
								],
								[
									[5, "0.4"],
									[5, "0.55"],
									"up"
								]
							]]
						]
					]
				]
			],
			[
				13,
				"sunCenter",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						"lx",
						"*",
						[5, "0.8"]
					], [
						1,
						[
							1,
							"ly",
							"*",
							[5, "0.8"]
						],
						"+",
						[5, "0.12"]
					]]
				]
			],
			[
				13,
				"sunRad",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "0.5"],
						[5, "0.13"],
						"sharp"
					]
				]
			],
			[
				13,
				"sd",
				[
					6,
					[
						7,
						"std",
						"length"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								["ex", "ey"]
							],
							"sub"
						],
						["sunCenter"]
					]]
				]
			],
			[
				13,
				"sun",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						"sunRad",
						[
							1,
							"sunRad",
							"*",
							[5, "0.15"]
						],
						"sd"
					]
				]
			],
			[
				2,
				"base",
				"=",
				[
					6,
					[
						7,
						"base",
						"add"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[5, "1"],
									[5, "0.97"],
									[5, "0.88"]
								]
							],
							"mul"
						],
						[[
							1,
							"sun",
							"*",
							[5, "1.5"]
						]]
					]]
				]
			],
			[10, "base"]
		]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const waterComposite = tgpu.fn([
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec2f,
	d.vec2f,
	d.vec3f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f,
	d.vec4f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((surf0, surfX, surfY, sdfUV, viewportSize, waterColor, p0, p1, p2, p3, p4) => {
	"use gpu";
	const scl = p0.x;
	const edgeSoftness = p0.y;
	const waveScale = p0.z;
	const swirl = p0.w;
	const choppiness = p1.x;
	const depth = p1.y;
	const shallows = p1.z;
	const clarity = p1.w;
	const caustics = p2.x;
	const causticScale = p2.y;
	const reflection = p2.z;
	const sharpness = p2.w;
	const foam = p3.x;
	const lightAngle = p3.y;
	const envRotation = p3.z;
	const animTime = p3.w;
	const volumetric = p4.x;
	const sharpEdge = std.max(__tsover_mul(edgeSoftness, .5), .001);
	const pxH = __tsover_div(1, viewportSize.y);
	const sdfRaw = surf0.x;
	const sdf = __tsover_div(sdfRaw, scl);
	if (sdf > __tsover_mul(pxH, 2)) return d.vec4f(0, 0, 0, 0);
	const lx = std.cos(__tsover_mul(lightAngle, DEG_TO_RAD));
	const ly = std.sin(__tsover_mul(lightAngle, DEG_TO_RAD));
	const envC = std.cos(__tsover_mul(envRotation, DEG_TO_RAD));
	const envS = std.sin(__tsover_mul(envRotation, DEG_TO_RAD));
	const EPS = .01;
	const gradX = __tsover_div(__tsover_sub(surfX.x, sdfRaw), EPS);
	const gradY = __tsover_div(__tsover_sub(surfY.x, sdfRaw), EPS);
	const ddx = std.clamp(__tsover_div(__tsover_sub(surfX.w, surf0.w), EPS), -5, 5);
	const ddy = std.clamp(__tsover_div(__tsover_sub(surfY.w, surf0.w), EPS), -5, 5);
	const nVol = std.normalize(d.vec3f(ddx, ddy, -1));
	const rimT = std.clamp(__tsover_div(__tsover_mul(sdf, -1), .05), 0, 1);
	const rimBevel = __tsover_mul(__tsover_sub(1, rimT), 1.6);
	const nFlat = std.normalize(d.vec3f(__tsover_mul(gradX, rimBevel), __tsover_mul(gradY, rimBevel), -1));
	const nGeo = std.select(nFlat, nVol, volumetric > .5);
	const fx = std.select(__tsover_sub(sdfUV.x, .5), surf0.y, volumetric > .5);
	const fy = std.select(__tsover_sub(sdfUV.y, .5), surf0.z, volumetric > .5);
	const freq = __tsover_mul(waveScale, 2.4);
	const flowT = __tsover_mul(animTime, .2);
	const q = d.vec2f(__tsover_mul(__tsover_mul(fx, freq), .8), __tsover_mul(__tsover_mul(fy, freq), 1.25));
	const wAmt = __tsover_mul(swirl, .9);
	const warpInput = q.mul(.6);
	const wx = __tsover_mul(std.sin(__tsover_add(__tsover_add(__tsover_mul(warpInput.x, 1.7), __tsover_mul(flowT, .7)), 1.7)), std.cos(__tsover_add(__tsover_add(__tsover_mul(warpInput.y, 1.3), __tsover_mul(flowT, .5)), 2.3)));
	const wy = __tsover_mul(std.cos(__tsover_add(__tsover_add(__tsover_mul(warpInput.x, 1.4), __tsover_mul(flowT, .9)), 8.7)), std.sin(__tsover_add(__tsover_add(__tsover_mul(warpInput.y, 1.6), __tsover_mul(flowT, .6)), 5.1)));
	const qw = q.add(d.vec2f(wx, wy).mul(wAmt)).add(d.vec2f(0, __tsover_mul(flowT, -1)));
	const re = .06;
	const gMax = d.f32(4);
	const h0 = waterFbm(qw);
	const h0n1 = waterNoise1(qw);
	const hx = std.clamp(__tsover_div(__tsover_sub(waterNoise1(qw.add(d.vec2f(re, 0))), h0n1), re), __tsover_mul(gMax, -1), gMax);
	const hy = std.clamp(__tsover_div(__tsover_sub(waterNoise1(qw.add(d.vec2f(0, re))), h0n1), re), __tsover_mul(gMax, -1), gMax);
	const fineRe = .025;
	const qf = qw.mul(3.3).add(d.vec2f(__tsover_mul(flowT, 1.7), __tsover_mul(flowT, -1.15))).add(d.vec2f(11.3, 4.7));
	const f0n1 = waterNoise1(qf);
	const fxg = std.clamp(__tsover_div(__tsover_sub(waterNoise1(qf.add(d.vec2f(fineRe, 0))), f0n1), fineRe), __tsover_mul(gMax, -1), gMax);
	const fyg = std.clamp(__tsover_div(__tsover_sub(waterNoise1(qf.add(d.vec2f(0, fineRe))), f0n1), fineRe), __tsover_mul(gMax, -1), gMax);
	const relief = __tsover_mul(choppiness, .5);
	const reliefFine = __tsover_mul(choppiness, .22);
	const n = std.normalize(d.vec3f(__tsover_sub(__tsover_sub(nGeo.x, __tsover_mul(hx, relief)), __tsover_mul(fxg, reliefFine)), __tsover_sub(__tsover_sub(nGeo.y, __tsover_mul(hy, relief)), __tsover_mul(fyg, reliefFine)), nGeo.z));
	const sharp = std.clamp(sharpness, 0, 1);
	const R = std.reflect(d.vec3f(0, 0, 1), n);
	const reflCol = waterSky(R.x, R.y, lx, ly, envC, envS, sharp).mul(reflection);
	const thickVol = std.clamp(__tsover_mul(__tsover_mul(sdf, -1), 2), 0, 1.2);
	const thickFlat = __tsover_mul(std.clamp(__tsover_mul(sdf, -1), 0, .5), 2);
	const thicknessRaw = std.select(thickFlat, thickVol, volumetric > .5);
	const thickness = __tsover_mul(thicknessRaw, depth);
	const depthT = std.clamp(__tsover_mul(thickness, 1.3), 0, 1);
	const wc = waterColor;
	const absorb = d.vec3f(__tsover_mul(__tsover_sub(1, wc.x), 2.2), __tsover_mul(__tsover_sub(1, wc.y), 2.2), __tsover_mul(__tsover_sub(1, wc.z), 2.2));
	const transmit = d.vec3f(std.exp(__tsover_mul(__tsover_mul(thickness, absorb.x), -1)), std.exp(__tsover_mul(__tsover_mul(thickness, absorb.y), -1)), std.exp(__tsover_mul(__tsover_mul(thickness, absorb.z), -1)));
	const shallowsC = std.clamp(shallows, 0, 1);
	const shallowCol = std.mix(wc, d.vec3f(1, 1, 1), d.vec3f(shallowsC));
	const bodyTint = std.mix(shallowCol, wc, d.vec3f(depthT)).mul(transmit).add(wc.mul(__tsover_mul(depthT, .18)));
	const L = std.normalize(d.vec3f(lx, ly, -.7));
	const ndl = std.dot(n, L);
	const diffuse = __tsover_add(__tsover_mul(std.pow(std.clamp(__tsover_add(__tsover_mul(ndl, .5), .5), 0, 1), d.f32(.8)), .5), .5);
	const eta = __tsover_div(1, 1.333);
	const tDir = std.refract(d.vec3f(0, 0, 1), n, eta);
	const transSky = waterSky(tDir.x, tDir.y, lx, ly, envC, envS, sharp).mul(transmit);
	const clarityT = __tsover_mul(std.clamp(clarity, 0, 1), __tsover_sub(1, __tsover_mul(depthT, .8)));
	let body = std.mix(bodyTint, transSky.mul(.6).add(bodyTint.mul(.4)), d.vec3f(clarityT)).mul(diffuse);
	const causFreq = causticScale;
	const causT = __tsover_mul(animTime, 3);
	const causP = d.vec2f(fx, fy).mul(causFreq).add(d.vec2f(__tsover_mul(std.sin(__tsover_mul(causT, .15)), .4), __tsover_mul(__tsover_mul(causT, .3), -1)));
	const c1 = __tsover_sub(1, std.abs(waterFbm(causP)));
	const c2 = __tsover_sub(1, std.abs(waterFbm(causP.mul(1.9).add(d.vec2f(3.7, __tsover_mul(__tsover_mul(causT, .18), -1))))));
	const causV = std.pow(std.clamp(__tsover_add(__tsover_mul(c1, .6), __tsover_mul(c2, .4)), 0, 1), d.f32(3));
	const causMask = __tsover_mul(__tsover_add(__tsover_mul(depthT, .8), .2), caustics);
	body = body.mul(__tsover_add(1, __tsover_mul(__tsover_mul(causV, causMask), .7)));
	body = body.add(d.vec3f(.55, .82, 1).mul(__tsover_mul(__tsover_mul(causV, causMask), .45)));
	const cosView = std.clamp(__tsover_mul(n.z, -1), 0, 1);
	const fresMix = std.clamp(__tsover_add(.04, __tsover_mul(.96, std.pow(__tsover_sub(1, cosView), d.f32(5)))), 0, 1);
	let surface = std.mix(body, reflCol, d.vec3f(fresMix));
	const H = std.normalize(L.add(d.vec3f(0, 0, -1)));
	const ndh = std.clamp(std.dot(n, H), 0, 1);
	const spec = __tsover_mul(std.pow(ndh, std.mix(d.f32(40), d.f32(600), sharp)), std.mix(.3, 1.3, sharp));
	surface = surface.add(d.vec3f(1, .98, .92).mul(spec));
	const crestH = std.clamp(__tsover_add(__tsover_mul(h0, .5), .5), 0, 1);
	const crest = std.smoothstep(.62, .92, crestH);
	const shore = __tsover_sub(1, std.smoothstep(0, .045, __tsover_mul(sdf, -1)));
	const foamMask = __tsover_mul(std.clamp(__tsover_add(crest, __tsover_mul(shore, .7)), 0, 1), foam);
	surface = std.mix(surface, d.vec3f(1, 1, 1), d.vec3f(__tsover_mul(foamMask, .85)));
	const transition = std.smoothstep(0, 1, std.clamp(__tsover_mul(__tsover_div(__tsover_mul(sdf, -1), sharpEdge), 32), 0, 1));
	return d.vec4f(surface.x, surface.y, surface.z, transition);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "surf0"
			},
			{
				"type": "i",
				"name": "surfX"
			},
			{
				"type": "i",
				"name": "surfY"
			},
			{
				"type": "i",
				"name": "sdfUV"
			},
			{
				"type": "i",
				"name": "viewportSize"
			},
			{
				"type": "i",
				"name": "waterColor"
			},
			{
				"type": "i",
				"name": "p0"
			},
			{
				"type": "i",
				"name": "p1"
			},
			{
				"type": "i",
				"name": "p2"
			},
			{
				"type": "i",
				"name": "p3"
			},
			{
				"type": "i",
				"name": "p4"
			}
		],
		"body": [0, [
			[
				13,
				"scl",
				[
					7,
					"p0",
					"x"
				]
			],
			[
				13,
				"edgeSoftness",
				[
					7,
					"p0",
					"y"
				]
			],
			[
				13,
				"waveScale",
				[
					7,
					"p0",
					"z"
				]
			],
			[
				13,
				"swirl",
				[
					7,
					"p0",
					"w"
				]
			],
			[
				13,
				"choppiness",
				[
					7,
					"p1",
					"x"
				]
			],
			[
				13,
				"depth",
				[
					7,
					"p1",
					"y"
				]
			],
			[
				13,
				"shallows",
				[
					7,
					"p1",
					"z"
				]
			],
			[
				13,
				"clarity",
				[
					7,
					"p1",
					"w"
				]
			],
			[
				13,
				"caustics",
				[
					7,
					"p2",
					"x"
				]
			],
			[
				13,
				"causticScale",
				[
					7,
					"p2",
					"y"
				]
			],
			[
				13,
				"reflection",
				[
					7,
					"p2",
					"z"
				]
			],
			[
				13,
				"sharpness",
				[
					7,
					"p2",
					"w"
				]
			],
			[
				13,
				"foam",
				[
					7,
					"p3",
					"x"
				]
			],
			[
				13,
				"lightAngle",
				[
					7,
					"p3",
					"y"
				]
			],
			[
				13,
				"envRotation",
				[
					7,
					"p3",
					"z"
				]
			],
			[
				13,
				"animTime",
				[
					7,
					"p3",
					"w"
				]
			],
			[
				13,
				"volumetric",
				[
					7,
					"p4",
					"x"
				]
			],
			[
				13,
				"sharpEdge",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"edgeSoftness",
						"*",
						[5, "0.5"]
					], [5, "0.001"]]
				]
			],
			[
				13,
				"pxH",
				[
					1,
					[5, "1"],
					"/",
					[
						7,
						"viewportSize",
						"y"
					]
				]
			],
			[
				13,
				"sdfRaw",
				[
					7,
					"surf0",
					"x"
				]
			],
			[
				13,
				"sdf",
				[
					1,
					"sdfRaw",
					"/",
					"scl"
				]
			],
			[
				11,
				[
					1,
					"sdf",
					">",
					[
						1,
						"pxH",
						"*",
						[5, "2"]
					]
				],
				[10, [
					6,
					[
						7,
						"d",
						"vec4f"
					],
					[
						[5, "0"],
						[5, "0"],
						[5, "0"],
						[5, "0"]
					]
				]]
			],
			[
				13,
				"lx",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					[[
						1,
						"lightAngle",
						"*",
						"DEG_TO_RAD"
					]]
				]
			],
			[
				13,
				"ly",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						1,
						"lightAngle",
						"*",
						"DEG_TO_RAD"
					]]
				]
			],
			[
				13,
				"envC",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					[[
						1,
						"envRotation",
						"*",
						"DEG_TO_RAD"
					]]
				]
			],
			[
				13,
				"envS",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						1,
						"envRotation",
						"*",
						"DEG_TO_RAD"
					]]
				]
			],
			[
				13,
				"EPS",
				[5, "0.01"]
			],
			[
				13,
				"gradX",
				[
					1,
					[
						1,
						[
							7,
							"surfX",
							"x"
						],
						"-",
						"sdfRaw"
					],
					"/",
					"EPS"
				]
			],
			[
				13,
				"gradY",
				[
					1,
					[
						1,
						[
							7,
							"surfY",
							"x"
						],
						"-",
						"sdfRaw"
					],
					"/",
					"EPS"
				]
			],
			[
				13,
				"ddx",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"surfX",
									"w"
								],
								"-",
								[
									7,
									"surf0",
									"w"
								]
							],
							"/",
							"EPS"
						],
						[
							4,
							"-",
							[5, "5"]
						],
						[5, "5"]
					]
				]
			],
			[
				13,
				"ddy",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									7,
									"surfY",
									"w"
								],
								"-",
								[
									7,
									"surf0",
									"w"
								]
							],
							"/",
							"EPS"
						],
						[
							4,
							"-",
							[5, "5"]
						],
						[5, "5"]
					]
				]
			],
			[
				13,
				"nVol",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"ddx",
							"ddy",
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]]
				]
			],
			[
				13,
				"rimT",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"sdf",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"/",
							[5, "0.05"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"rimBevel",
				[
					1,
					[
						1,
						[5, "1"],
						"-",
						"rimT"
					],
					"*",
					[5, "1.6"]
				]
			],
			[
				13,
				"nFlat",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								"gradX",
								"*",
								"rimBevel"
							],
							[
								1,
								"gradY",
								"*",
								"rimBevel"
							],
							[
								4,
								"-",
								[5, "1"]
							]
						]
					]]
				]
			],
			[
				13,
				"nGeo",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"nFlat",
						"nVol",
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"fx",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							1,
							[
								7,
								"sdfUV",
								"x"
							],
							"-",
							[5, "0.5"]
						],
						[
							7,
							"surf0",
							"y"
						],
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"fy",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						[
							1,
							[
								7,
								"sdfUV",
								"y"
							],
							"-",
							[5, "0.5"]
						],
						[
							7,
							"surf0",
							"z"
						],
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"freq",
				[
					1,
					"waveScale",
					"*",
					[5, "2.4"]
				]
			],
			[
				13,
				"flowT",
				[
					1,
					"animTime",
					"*",
					[5, "0.2"]
				]
			],
			[
				13,
				"q",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							"fx",
							"*",
							"freq"
						],
						"*",
						[5, "0.8"]
					], [
						1,
						[
							1,
							"fy",
							"*",
							"freq"
						],
						"*",
						[5, "1.25"]
					]]
				]
			],
			[
				13,
				"wAmt",
				[
					1,
					"swirl",
					"*",
					[5, "0.9"]
				]
			],
			[
				13,
				"warpInput",
				[
					6,
					[
						7,
						"q",
						"mul"
					],
					[[5, "0.6"]]
				]
			],
			[
				13,
				"wx",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						[[
							1,
							[
								1,
								[
									1,
									[
										7,
										"warpInput",
										"x"
									],
									"*",
									[5, "1.7"]
								],
								"+",
								[
									1,
									"flowT",
									"*",
									[5, "0.7"]
								]
							],
							"+",
							[5, "1.7"]
						]]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						[[
							1,
							[
								1,
								[
									1,
									[
										7,
										"warpInput",
										"y"
									],
									"*",
									[5, "1.3"]
								],
								"+",
								[
									1,
									"flowT",
									"*",
									[5, "0.5"]
								]
							],
							"+",
							[5, "2.3"]
						]]
					]
				]
			],
			[
				13,
				"wy",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"cos"
						],
						[[
							1,
							[
								1,
								[
									1,
									[
										7,
										"warpInput",
										"x"
									],
									"*",
									[5, "1.4"]
								],
								"+",
								[
									1,
									"flowT",
									"*",
									[5, "0.9"]
								]
							],
							"+",
							[5, "8.7"]
						]]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						[[
							1,
							[
								1,
								[
									1,
									[
										7,
										"warpInput",
										"y"
									],
									"*",
									[5, "1.6"]
								],
								"+",
								[
									1,
									"flowT",
									"*",
									[5, "0.6"]
								]
							],
							"+",
							[5, "5.1"]
						]]
					]
				]
			],
			[
				13,
				"qw",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"q",
								"add"
							],
							[[
								6,
								[
									7,
									[
										6,
										[
											7,
											"d",
											"vec2f"
										],
										["wx", "wy"]
									],
									"mul"
								],
								["wAmt"]
							]]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "0"], [
							1,
							"flowT",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						]]
					]]
				]
			],
			[
				13,
				"re",
				[5, "0.06"]
			],
			[
				13,
				"gMax",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "4"]]
				]
			],
			[
				13,
				"h0",
				[
					6,
					"waterFbm",
					["qw"]
				]
			],
			[
				13,
				"h0n1",
				[
					6,
					"waterNoise1",
					["qw"]
				]
			],
			[
				13,
				"hx",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									6,
									"waterNoise1",
									[[
										6,
										[
											7,
											"qw",
											"add"
										],
										[[
											6,
											[
												7,
												"d",
												"vec2f"
											],
											["re", [5, "0"]]
										]]
									]]
								],
								"-",
								"h0n1"
							],
							"/",
							"re"
						],
						[
							1,
							"gMax",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"gMax"
					]
				]
			],
			[
				13,
				"hy",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									6,
									"waterNoise1",
									[[
										6,
										[
											7,
											"qw",
											"add"
										],
										[[
											6,
											[
												7,
												"d",
												"vec2f"
											],
											[[5, "0"], "re"]
										]]
									]]
								],
								"-",
								"h0n1"
							],
							"/",
							"re"
						],
						[
							1,
							"gMax",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"gMax"
					]
				]
			],
			[
				13,
				"fineRe",
				[5, "0.025"]
			],
			[
				13,
				"qf",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										"qw",
										"mul"
									],
									[[5, "3.3"]]
								],
								"add"
							],
							[[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									1,
									"flowT",
									"*",
									[5, "1.7"]
								], [
									1,
									"flowT",
									"*",
									[
										4,
										"-",
										[5, "1.15"]
									]
								]]
							]]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "11.3"], [5, "4.7"]]
					]]
				]
			],
			[
				13,
				"f0n1",
				[
					6,
					"waterNoise1",
					["qf"]
				]
			],
			[
				13,
				"fxg",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									6,
									"waterNoise1",
									[[
										6,
										[
											7,
											"qf",
											"add"
										],
										[[
											6,
											[
												7,
												"d",
												"vec2f"
											],
											["fineRe", [5, "0"]]
										]]
									]]
								],
								"-",
								"f0n1"
							],
							"/",
							"fineRe"
						],
						[
							1,
							"gMax",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"gMax"
					]
				]
			],
			[
				13,
				"fyg",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								[
									6,
									"waterNoise1",
									[[
										6,
										[
											7,
											"qf",
											"add"
										],
										[[
											6,
											[
												7,
												"d",
												"vec2f"
											],
											[[5, "0"], "fineRe"]
										]]
									]]
								],
								"-",
								"f0n1"
							],
							"/",
							"fineRe"
						],
						[
							1,
							"gMax",
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						"gMax"
					]
				]
			],
			[
				13,
				"relief",
				[
					1,
					"choppiness",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"reliefFine",
				[
					1,
					"choppiness",
					"*",
					[5, "0.22"]
				]
			],
			[
				13,
				"n",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[
								1,
								[
									1,
									[
										7,
										"nGeo",
										"x"
									],
									"-",
									[
										1,
										"hx",
										"*",
										"relief"
									]
								],
								"-",
								[
									1,
									"fxg",
									"*",
									"reliefFine"
								]
							],
							[
								1,
								[
									1,
									[
										7,
										"nGeo",
										"y"
									],
									"-",
									[
										1,
										"hy",
										"*",
										"relief"
									]
								],
								"-",
								[
									1,
									"fyg",
									"*",
									"reliefFine"
								]
							],
							[
								7,
								"nGeo",
								"z"
							]
						]
					]]
				]
			],
			[
				13,
				"sharp",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"sharpness",
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"R",
				[
					6,
					[
						7,
						"std",
						"reflect"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							[5, "0"],
							[5, "0"],
							[5, "1"]
						]
					], "n"]
				]
			],
			[
				13,
				"reflCol",
				[
					6,
					[
						7,
						[
							6,
							"waterSky",
							[
								[
									7,
									"R",
									"x"
								],
								[
									7,
									"R",
									"y"
								],
								"lx",
								"ly",
								"envC",
								"envS",
								"sharp"
							]
						],
						"mul"
					],
					["reflection"]
				]
			],
			[
				13,
				"thickVol",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"sdf",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							"*",
							[5, "2"]
						],
						[5, "0"],
						[5, "1.2"]
					]
				]
			],
			[
				13,
				"thickFlat",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								"sdf",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							],
							[5, "0"],
							[5, "0.5"]
						]
					],
					"*",
					[5, "2"]
				]
			],
			[
				13,
				"thicknessRaw",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"thickFlat",
						"thickVol",
						[
							1,
							"volumetric",
							">",
							[5, "0.5"]
						]
					]
				]
			],
			[
				13,
				"thickness",
				[
					1,
					"thicknessRaw",
					"*",
					"depth"
				]
			],
			[
				13,
				"depthT",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"thickness",
							"*",
							[5, "1.3"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"wc",
				"waterColor"
			],
			[
				13,
				"absorb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							1,
							[
								1,
								[5, "1"],
								"-",
								[
									7,
									"wc",
									"x"
								]
							],
							"*",
							[5, "2.2"]
						],
						[
							1,
							[
								1,
								[5, "1"],
								"-",
								[
									7,
									"wc",
									"y"
								]
							],
							"*",
							[5, "2.2"]
						],
						[
							1,
							[
								1,
								[5, "1"],
								"-",
								[
									7,
									"wc",
									"z"
								]
							],
							"*",
							[5, "2.2"]
						]
					]
				]
			],
			[
				13,
				"transmit",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									1,
									"thickness",
									"*",
									[
										7,
										"absorb",
										"x"
									]
								],
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]]
						],
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									1,
									"thickness",
									"*",
									[
										7,
										"absorb",
										"y"
									]
								],
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]]
						],
						[
							6,
							[
								7,
								"std",
								"exp"
							],
							[[
								1,
								[
									1,
									"thickness",
									"*",
									[
										7,
										"absorb",
										"z"
									]
								],
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]]
						]
					]
				]
			],
			[
				13,
				"shallowsC",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						"shallows",
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"shallowCol",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"wc",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1"],
								[5, "1"],
								[5, "1"]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["shallowsC"]
						]
					]
				]
			],
			[
				13,
				"bodyTint",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										"std",
										"mix"
									],
									[
										"shallowCol",
										"wc",
										[
											6,
											[
												7,
												"d",
												"vec3f"
											],
											["depthT"]
										]
									]
								],
								"mul"
							],
							["transmit"]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"wc",
							"mul"
						],
						[[
							1,
							"depthT",
							"*",
							[5, "0.18"]
						]]
					]]
				]
			],
			[
				13,
				"L",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"d",
							"vec3f"
						],
						[
							"lx",
							"ly",
							[
								4,
								"-",
								[5, "0.7"]
							]
						]
					]]
				]
			],
			[
				13,
				"ndl",
				[
					6,
					[
						7,
						"std",
						"dot"
					],
					["n", "L"]
				]
			],
			[
				13,
				"diffuse",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"pow"
							],
							[[
								6,
								[
									7,
									"std",
									"clamp"
								],
								[
									[
										1,
										[
											1,
											"ndl",
											"*",
											[5, "0.5"]
										],
										"+",
										[5, "0.5"]
									],
									[5, "0"],
									[5, "1"]
								]
							], [
								6,
								[
									7,
									"d",
									"f32"
								],
								[[5, "0.8"]]
							]]
						],
						"*",
						[5, "0.5"]
					],
					"+",
					[5, "0.5"]
				]
			],
			[
				13,
				"eta",
				[
					1,
					[5, "1"],
					"/",
					[5, "1.333"]
				]
			],
			[
				13,
				"tDir",
				[
					6,
					[
						7,
						"std",
						"refract"
					],
					[
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0"],
								[5, "0"],
								[5, "1"]
							]
						],
						"n",
						"eta"
					]
				]
			],
			[
				13,
				"transSky",
				[
					6,
					[
						7,
						[
							6,
							"waterSky",
							[
								[
									7,
									"tDir",
									"x"
								],
								[
									7,
									"tDir",
									"y"
								],
								"lx",
								"ly",
								"envC",
								"envS",
								"sharp"
							]
						],
						"mul"
					],
					["transmit"]
				]
			],
			[
				13,
				"clarityT",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							"clarity",
							[5, "0"],
							[5, "1"]
						]
					],
					"*",
					[
						1,
						[5, "1"],
						"-",
						[
							1,
							"depthT",
							"*",
							[5, "0.8"]
						]
					]
				]
			],
			[
				12,
				"body",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								"bodyTint",
								[
									6,
									[
										7,
										[
											6,
											[
												7,
												"transSky",
												"mul"
											],
											[[5, "0.6"]]
										],
										"add"
									],
									[[
										6,
										[
											7,
											"bodyTint",
											"mul"
										],
										[[5, "0.4"]]
									]]
								],
								[
									6,
									[
										7,
										"d",
										"vec3f"
									],
									["clarityT"]
								]
							]
						],
						"mul"
					],
					["diffuse"]
				]
			],
			[
				13,
				"causFreq",
				"causticScale"
			],
			[
				13,
				"causT",
				[
					1,
					"animTime",
					"*",
					[5, "3"]
				]
			],
			[
				13,
				"causP",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									["fx", "fy"]
								],
								"mul"
							],
							["causFreq"]
						],
						"add"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[
							1,
							[
								6,
								[
									7,
									"std",
									"sin"
								],
								[[
									1,
									"causT",
									"*",
									[5, "0.15"]
								]]
							],
							"*",
							[5, "0.4"]
						], [
							1,
							[
								1,
								"causT",
								"*",
								[5, "0.3"]
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						]]
					]]
				]
			],
			[
				13,
				"c1",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							6,
							"waterFbm",
							["causP"]
						]]
					]
				]
			],
			[
				13,
				"c2",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"abs"
						],
						[[
							6,
							"waterFbm",
							[[
								6,
								[
									7,
									[
										6,
										[
											7,
											"causP",
											"mul"
										],
										[[5, "1.9"]]
									],
									"add"
								],
								[[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									[[5, "3.7"], [
										1,
										[
											1,
											"causT",
											"*",
											[5, "0.18"]
										],
										"*",
										[
											4,
											"-",
											[5, "1"]
										]
									]]
								]]
							]]
						]]
					]
				]
			],
			[
				13,
				"causV",
				[
					6,
					[
						7,
						"std",
						"pow"
					],
					[[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								[
									1,
									"c1",
									"*",
									[5, "0.6"]
								],
								"+",
								[
									1,
									"c2",
									"*",
									[5, "0.4"]
								]
							],
							[5, "0"],
							[5, "1"]
						]
					], [
						6,
						[
							7,
							"d",
							"f32"
						],
						[[5, "3"]]
					]]
				]
			],
			[
				13,
				"causMask",
				[
					1,
					[
						1,
						[
							1,
							"depthT",
							"*",
							[5, "0.8"]
						],
						"+",
						[5, "0.2"]
					],
					"*",
					"caustics"
				]
			],
			[
				2,
				"body",
				"=",
				[
					6,
					[
						7,
						"body",
						"mul"
					],
					[[
						1,
						[5, "1"],
						"+",
						[
							1,
							[
								1,
								"causV",
								"*",
								"causMask"
							],
							"*",
							[5, "0.7"]
						]
					]]
				]
			],
			[
				2,
				"body",
				"=",
				[
					6,
					[
						7,
						"body",
						"add"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[5, "0.55"],
									[5, "0.82"],
									[5, "1"]
								]
							],
							"mul"
						],
						[[
							1,
							[
								1,
								"causV",
								"*",
								"causMask"
							],
							"*",
							[5, "0.45"]
						]]
					]]
				]
			],
			[
				13,
				"cosView",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								7,
								"n",
								"z"
							],
							"*",
							[
								4,
								"-",
								[5, "1"]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"fresMix",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[5, "0.04"],
							"+",
							[
								1,
								[5, "0.96"],
								"*",
								[
									6,
									[
										7,
										"std",
										"pow"
									],
									[[
										1,
										[5, "1"],
										"-",
										"cosView"
									], [
										6,
										[
											7,
											"d",
											"f32"
										],
										[[5, "5"]]
									]]
								]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				12,
				"surface",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"body",
						"reflCol",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["fresMix"]
						]
					]
				]
			],
			[
				13,
				"H",
				[
					6,
					[
						7,
						"std",
						"normalize"
					],
					[[
						6,
						[
							7,
							"L",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "0"],
								[5, "0"],
								[
									4,
									"-",
									[5, "1"]
								]
							]
						]]
					]]
				]
			],
			[
				13,
				"ndh",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							6,
							[
								7,
								"std",
								"dot"
							],
							["n", "H"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"spec",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"pow"
						],
						["ndh", [
							6,
							[
								7,
								"std",
								"mix"
							],
							[
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "40"]]
								],
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "600"]]
								],
								"sharp"
							]
						]]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							[5, "0.3"],
							[5, "1.3"],
							"sharp"
						]
					]
				]
			],
			[
				2,
				"surface",
				"=",
				[
					6,
					[
						7,
						"surface",
						"add"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"d",
									"vec3f"
								],
								[
									[5, "1"],
									[5, "0.98"],
									[5, "0.92"]
								]
							],
							"mul"
						],
						["spec"]
					]]
				]
			],
			[
				13,
				"crestH",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							[
								1,
								"h0",
								"*",
								[5, "0.5"]
							],
							"+",
							[5, "0.5"]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"crest",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0.62"],
						[5, "0.92"],
						"crestH"
					]
				]
			],
			[
				13,
				"shore",
				[
					1,
					[5, "1"],
					"-",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							[5, "0"],
							[5, "0.045"],
							[
								1,
								"sdf",
								"*",
								[
									4,
									"-",
									[5, "1"]
								]
							]
						]
					]
				]
			],
			[
				13,
				"foamMask",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"clamp"
						],
						[
							[
								1,
								"crest",
								"+",
								[
									1,
									"shore",
									"*",
									[5, "0.7"]
								]
							],
							[5, "0"],
							[5, "1"]
						]
					],
					"*",
					"foam"
				]
			],
			[
				2,
				"surface",
				"=",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"surface",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[
								[5, "1"],
								[5, "1"],
								[5, "1"]
							]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[
								1,
								"foamMask",
								"*",
								[5, "0.85"]
							]]
						]
					]
				]
			],
			[
				13,
				"transition",
				[
					6,
					[
						7,
						"std",
						"smoothstep"
					],
					[
						[5, "0"],
						[5, "1"],
						[
							6,
							[
								7,
								"std",
								"clamp"
							],
							[
								[
									1,
									[
										1,
										[
											1,
											"sdf",
											"*",
											[
												4,
												"-",
												[5, "1"]
											]
										],
										"/",
										"sharpEdge"
									],
									"*",
									[5, "32"]
								],
								[5, "0"],
								[5, "1"]
							]
						]
					]
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					[
						7,
						"surface",
						"x"
					],
					[
						7,
						"surface",
						"y"
					],
					[
						7,
						"surface",
						"z"
					],
					"transition"
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"DEG_TO_RAD",
			"waterFbm",
			"waterNoise1",
			"waterSky"
		]
	},
	externals: () => ({
		std,
		d,
		DEG_TO_RAD,
		waterFbm,
		waterNoise1,
		waterSky
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Water",
	boundingBoxDeclaration: shapeEffectBoundingBoxDeclaration,
	category: "Shape Effects",
	description: "Translucent water — a refractive, light-absorbing body that deepens to its colour with thickness, wrapped in a wind-driven wave surface that reflects a procedural sky, with Fresnel-bright grazing edges and foam breaking on the crests and shoreline",
	animatedTime: { speed: "speed" },
	extraFields: {
		...VOLUMETRIC_FIELD_EXTRA_FIELDS,
		...ANALYTIC_SDF_EXTRA_FIELDS
	},
	props: {
		origin: {
			default: "center",
			description: "Reference edge the center position is measured from (center default). Lets you pin the shape relative to a corner or the canvas centre.",
			ui: {
				type: "origin",
				label: "Origin",
				group: "Position"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the water shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position",
				units: ["%", "px"]
			}
		},
		scale: {
			default: 1,
			description: "Scale of the water shape (1 = default size)",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Scale",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the water shape in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Position"
			}
		},
		waterColor: {
			default: "#00bfeb",
			transform: transformColor,
			description: "The colour the water saturates toward with depth",
			ui: {
				type: "color",
				label: "Water Color",
				group: "Water"
			}
		},
		clarity: {
			default: .4,
			description: "How much of the reflected sky transmits through the body",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Clarity",
				group: "Water"
			}
		},
		depth: {
			default: 5,
			description: "How quickly the water darkens and saturates toward its colour with thickness",
			ui: {
				type: "range",
				min: 0,
				max: 8,
				step: .01,
				label: "Depth",
				group: "Water"
			}
		},
		shallows: {
			default: .5,
			description: "Brightness of the thin shallow water at the edges — high fades the rim to light translucent water, low keeps the edges close to the deep water colour",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Shallows",
				group: "Water"
			}
		},
		caustics: {
			default: .15,
			description: "Intensity of the light caustics bubbling up through the body — bright veins of focused light churning inside the water",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Caustics",
				group: "Water"
			}
		},
		causticScale: {
			default: 6,
			description: "Size of the caustic veins — higher = finer, busier light patterns inside the body",
			ui: {
				type: "range",
				min: 1,
				max: 20,
				step: .1,
				label: "Caustic Scale",
				group: "Water"
			}
		},
		choppiness: {
			default: .8,
			description: "Height of the wind-driven waves — the depth of the folds that bend the reflection",
			ui: {
				type: "range",
				min: 0,
				max: 1.5,
				step: .01,
				label: "Choppiness",
				group: "Waves"
			}
		},
		waveScale: {
			default: 3.5,
			description: "Scale of the wave fronts — higher = smaller, busier ripples",
			ui: {
				type: "range",
				min: .5,
				max: 8,
				step: .01,
				label: "Wave Scale",
				group: "Waves"
			}
		},
		swirl: {
			default: .6,
			description: "Swirl of the flow — domain-warps the wave fronts into curling eddies",
			ui: {
				type: "range",
				min: 0,
				max: 1.5,
				step: .01,
				label: "Swirl",
				group: "Waves"
			}
		},
		speed: {
			default: .5,
			description: "Speed of the rolling waves. 0 pauses the surface.",
			ui: {
				type: "range",
				min: 0,
				max: 4,
				step: .01,
				label: "Speed",
				group: "Waves"
			}
		},
		reflection: {
			default: 2,
			description: "Strength of the reflected sky across the surface",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Reflection",
				group: "Surface"
			}
		},
		envRotation: {
			default: 0,
			description: "Rotates the reflected sky — spins where the bright sky and sun fall",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Env Rotation",
				group: "Surface"
			}
		},
		sharpness: {
			default: .1,
			description: "Tightness of the sun glint and reflected sun",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Sharpness",
				group: "Surface"
			}
		},
		lightAngle: {
			default: 30,
			description: "Direction of the sun, in degrees — drives the glint and where the sun reflects",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Light Angle",
				group: "Surface"
			}
		},
		foam: {
			default: .3,
			description: "White foam breaking on the wave crests and along the shoreline edge",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Foam",
				group: "Surface"
			}
		},
		edgeSoftness: {
			default: .05,
			description: "Softness of the shape boundary edge",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Softness",
				group: "Surface"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type"
		}
	},
	compute: (params) => {
		const vf = createVolumetricFieldComputeNode(params, () => params.getCpuValue("shape"), "raw");
		return vf ? {
			outputs: vf.outputs,
			getComputeNodes: (fp) => vf.getComputeNodes(fp)
		} : null;
	},
	fragment: (params) => {
		const { computeOutputs, uniforms, propValues, ctx, getCpuValue } = params;
		let sampler;
		let gradSampler;
		let volumetric = false;
		const fieldTex = computeOutputs?.volumetricFieldTexture;
		if (fieldTex) {
			const s = buildVolumetricFieldSampler(fieldTex, {
				originX: uniforms._vfOriginX,
				originY: uniforms._vfOriginY,
				spanX: uniforms._vfSpanX,
				spanY: uniforms._vfSpanY,
				activeResF: uniforms._vfActiveRes,
				rBound: uniforms._vfRBound
			});
			sampler = s.volumetricFieldSampler;
			gradSampler = s.volumetricFieldSamplerFast;
			volumetric = true;
		} else {
			const shapeSdfUrl = propValues.shapeSdfUrl || "";
			const shapeType = propValues.shapeType || "";
			if (shapeSdfUrl) {
				sampler = createSvgSdfSampler(params, shapeSdfUrl);
				gradSampler = sampler;
				volumetric = isVolumetricShapeType(shapeType);
			} else {
				const st = resolveShapeType(shapeType, getCpuValue("shape"));
				sampler = createAnalyticSdfSampler(st, driveAnalyticSubProps(params, () => getCpuValue("shape")), "raw");
				gradSampler = sampler;
				volumetric = is3dShapeType(st);
			}
		}
		const sdfUV = call(sdfSpaceUV, "sdfSpaceUV", [
			uniforms.center,
			uniforms.scale,
			uniforms.rotation,
			ctx.uv,
			ctx.aspect
		]);
		const surf0 = sampler(sdfUV);
		const surfX = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(.01),
			floatE(0)
		]));
		const surfY = gradSampler(call(offsetUV, "offsetUV", [
			sdfUV,
			floatE(0),
			floatE(.01)
		]));
		const t = animatedTime(params);
		const volFlag = floatE(volumetric ? 1 : 0);
		const p0 = vec4(uniforms.scale, uniforms.edgeSoftness, uniforms.waveScale, uniforms.swirl);
		const p1 = vec4(uniforms.choppiness, uniforms.depth, uniforms.shallows, uniforms.clarity);
		const p2 = vec4(uniforms.caustics, uniforms.causticScale, uniforms.reflection, uniforms.sharpness);
		const p3 = vec4(uniforms.foam, uniforms.lightAngle, uniforms.envRotation, t);
		const p4 = vec4(volFlag, floatE(0), floatE(0), floatE(0));
		return call(waterComposite, "waterComposite", [
			surf0,
			surfX,
			surfY,
			sdfUV,
			ctx.viewportSize,
			uniforms.waterColor.member("rgb"),
			p0,
			p1,
			p2,
			p3,
			p4
		]);
	}
};
var Water_default = componentDefinition;
export { waterNoise1 as a, waterFbm as i, componentDefinition as n, waterSky as o, waterComposite as r, Water_default as t };
