// shaders/dist/core/Watercolor--Bg1ZPIG.js
import { At as call, It as vec4, Nt as expr, Ot as ZERO, Pt as floatE, R as std, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor } from "./transformations-fqzznt4g.js";
var hash21 = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	return std.fract(__tsover_mul(std.sin(std.dot(p, d.vec2f(12.9898, 78.233))), 43758.5453));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [[10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				1,
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					[[
						6,
						[
							7,
							"std",
							"dot"
						],
						["p", [
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "12.9898"], [5, "78.233"]]
						]]
					]]
				],
				"*",
				[5, "43758.5453"]
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
var vnoise = tgpu.fn([d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	const i = std.floor(p);
	const f = std.fract(p);
	const u = f.mul(f).mul(d.vec2f(3, 3).sub(f.mul(2)));
	const a = hash21(i);
	const b = hash21(i.add(d.vec2f(1, 0)));
	const c = hash21(i.add(d.vec2f(0, 1)));
	const dd = hash21(i.add(d.vec2f(1, 1)));
	return std.mix(std.mix(a, b, u.x), std.mix(c, dd, u.x), u.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [
			[
				13,
				"i",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["p"]
				]
			],
			[
				13,
				"f",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["p"]
				]
			],
			[
				13,
				"u",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"f",
								"mul"
							],
							["f"]
						],
						"mul"
					],
					[[
						6,
						[
							7,
							[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[5, "3"], [5, "3"]]
							],
							"sub"
						],
						[[
							6,
							[
								7,
								"f",
								"mul"
							],
							[[5, "2"]]
						]]
					]]
				]
			],
			[
				13,
				"a",
				[
					6,
					"hash21",
					["i"]
				]
			],
			[
				13,
				"b",
				[
					6,
					"hash21",
					[[
						6,
						[
							7,
							"i",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "1"], [5, "0"]]
						]]
					]]
				]
			],
			[
				13,
				"c",
				[
					6,
					"hash21",
					[[
						6,
						[
							7,
							"i",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "0"], [5, "1"]]
						]]
					]]
				]
			],
			[
				13,
				"dd",
				[
					6,
					"hash21",
					[[
						6,
						[
							7,
							"i",
							"add"
						],
						[[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[5, "1"], [5, "1"]]
						]]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"mix"
				],
				[
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"a",
							"b",
							[
								7,
								"u",
								"x"
							]
						]
					],
					[
						6,
						[
							7,
							"std",
							"mix"
						],
						[
							"c",
							"dd",
							[
								7,
								"u",
								"x"
							]
						]
					],
					[
						7,
						"u",
						"y"
					]
				]
			]]
		]],
		"externalNames": [
			"std",
			"d",
			"hash21"
		]
	},
	externals: () => ({
		std,
		d,
		hash21
	})
}) && $.f)({}));
const watercolorUvBase = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, bleed, time) => {
	"use gpu";
	const t = __tsover_mul(time, .03);
	const wob = d.vec2f(vnoise(uv.mul(5).add(d.vec2f(t, t))), vnoise(uv.mul(5).add(d.vec2f(31.4, 31.4)).add(d.vec2f(t, t)))).sub(d.vec2f(.5, .5));
	const texel = d.vec2f(__tsover_div(1, viewport.x), __tsover_div(1, viewport.y));
	return uv.add(wob.mul(bleed).mul(texel).mul(8));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "bleed"
			},
			{
				"type": "i",
				"name": "time"
			}
		],
		"body": [0, [
			[
				13,
				"t",
				[
					1,
					"time",
					"*",
					[5, "0.03"]
				]
			],
			[
				13,
				"wob",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								6,
								"vnoise",
								[[
									6,
									[
										7,
										[
											6,
											[
												7,
												"uv",
												"mul"
											],
											[[5, "5"]]
										],
										"add"
									],
									[[
										6,
										[
											7,
											"d",
											"vec2f"
										],
										["t", "t"]
									]]
								]]
							], [
								6,
								"vnoise",
								[[
									6,
									[
										7,
										[
											6,
											[
												7,
												[
													6,
													[
														7,
														"uv",
														"mul"
													],
													[[5, "5"]]
												],
												"add"
											],
											[[
												6,
												[
													7,
													"d",
													"vec2f"
												],
												[[5, "31.4"], [5, "31.4"]]
											]]
										],
										"add"
									],
									[[
										6,
										[
											7,
											"d",
											"vec2f"
										],
										["t", "t"]
									]]
								]]
							]]
						],
						"sub"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "0.5"], [5, "0.5"]]
					]]
				]
			],
			[
				13,
				"texel",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[5, "1"],
						"/",
						[
							7,
							"viewport",
							"x"
						]
					], [
						1,
						[5, "1"],
						"/",
						[
							7,
							"viewport",
							"y"
						]
					]]
				]
			],
			[10, [
				6,
				[
					7,
					"uv",
					"add"
				],
				[[
					6,
					[
						7,
						[
							6,
							[
								7,
								[
									6,
									[
										7,
										"wob",
										"mul"
									],
									["bleed"]
								],
								"mul"
							],
							["texel"]
						],
						"mul"
					],
					[[5, "8"]]
				]]
			]]
		]],
		"externalNames": ["d", "vnoise"]
	},
	externals: () => ({
		d,
		vnoise
	})
}) && $.f)({}));
const watercolorTapUV = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uvBase, viewport, ox, oy) => {
	"use gpu";
	return d.vec2f(__tsover_add(uvBase.x, __tsover_div(ox, viewport.x)), __tsover_add(uvBase.y, __tsover_div(oy, viewport.y)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uvBase"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "ox"
			},
			{
				"type": "i",
				"name": "oy"
			}
		],
		"body": [0, [[10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			[[
				1,
				[
					7,
					"uvBase",
					"x"
				],
				"+",
				[
					1,
					"ox",
					"/",
					[
						7,
						"viewport",
						"x"
					]
				]
			], [
				1,
				[
					7,
					"uvBase",
					"y"
				],
				"+",
				[
					1,
					"oy",
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			]]
		]]]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const watercolorGrain = tgpu.fn([d.vec2f, d.vec2f], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport) => {
	"use gpu";
	return vnoise(uv.mul(viewport).mul(.5));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "uv"
		}, {
			"type": "i",
			"name": "viewport"
		}],
		"body": [0, [[10, [
			6,
			"vnoise",
			[[
				6,
				[
					7,
					[
						6,
						[
							7,
							"uv",
							"mul"
						],
						["viewport"]
					],
					"mul"
				],
				[[5, "0.5"]]
			]]
		]]]],
		"externalNames": ["vnoise"]
	},
	externals: () => ({ vnoise })
}) && $.f)({}));
const watercolorCompose = tgpu.fn([
	d.vec3f,
	d.vec3f,
	d.vec3f,
	d.vec3f,
	d.vec3f,
	d.vec3f,
	d.vec3f,
	d.vec3f,
	d.f32,
	d.vec4f,
	d.vec4f,
	d.f32,
	d.f32,
	d.f32
], d.vec3f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((sum0, sum1, sum2, sum3, sq0, sq1, sq2, sq3, invN, originalBleed, paperColor, paper, strength, grain) => {
	"use gpu";
	const mean0 = sum0.mul(invN);
	const mean1 = sum1.mul(invN);
	const mean2 = sum2.mul(invN);
	const mean3 = sum3.mul(invN);
	const var0 = sq0.mul(invN).sub(mean0.mul(mean0));
	const var1 = sq1.mul(invN).sub(mean1.mul(mean1));
	const var2 = sq2.mul(invN).sub(mean2.mul(mean2));
	const var3 = sq3.mul(invN).sub(mean3.mul(mean3));
	const sigma0 = __tsover_add(__tsover_add(var0.x, var0.y), var0.z);
	const sigma1 = __tsover_add(__tsover_add(var1.x, var1.y), var1.z);
	const sigma2 = __tsover_add(__tsover_add(var2.x, var2.y), var2.z);
	const sigma3 = __tsover_add(__tsover_add(var3.x, var3.y), var3.z);
	let result = d.vec3f(mean0.x, mean0.y, mean0.z);
	let best = sigma0;
	result = std.select(result, mean1, sigma1 < best);
	best = std.min(best, sigma1);
	result = std.select(result, mean2, sigma2 < best);
	best = std.min(best, sigma2);
	result = std.select(result, mean3, sigma3 < best);
	best = std.min(best, sigma3);
	const paperMul = std.mix(1, grain, paper);
	const paperTint = d.vec3f(paperColor.x, paperColor.y, paperColor.z);
	result = std.mix(result.mul(paperMul), result.mul(paperTint), d.vec3f(__tsover_mul(paper, .4)));
	const origRgb = d.vec3f(originalBleed.x, originalBleed.y, originalBleed.z);
	return std.mix(origRgb, result, d.vec3f(strength));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "sum0"
			},
			{
				"type": "i",
				"name": "sum1"
			},
			{
				"type": "i",
				"name": "sum2"
			},
			{
				"type": "i",
				"name": "sum3"
			},
			{
				"type": "i",
				"name": "sq0"
			},
			{
				"type": "i",
				"name": "sq1"
			},
			{
				"type": "i",
				"name": "sq2"
			},
			{
				"type": "i",
				"name": "sq3"
			},
			{
				"type": "i",
				"name": "invN"
			},
			{
				"type": "i",
				"name": "originalBleed"
			},
			{
				"type": "i",
				"name": "paperColor"
			},
			{
				"type": "i",
				"name": "paper"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "grain"
			}
		],
		"body": [0, [
			[
				13,
				"mean0",
				[
					6,
					[
						7,
						"sum0",
						"mul"
					],
					["invN"]
				]
			],
			[
				13,
				"mean1",
				[
					6,
					[
						7,
						"sum1",
						"mul"
					],
					["invN"]
				]
			],
			[
				13,
				"mean2",
				[
					6,
					[
						7,
						"sum2",
						"mul"
					],
					["invN"]
				]
			],
			[
				13,
				"mean3",
				[
					6,
					[
						7,
						"sum3",
						"mul"
					],
					["invN"]
				]
			],
			[
				13,
				"var0",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"sq0",
								"mul"
							],
							["invN"]
						],
						"sub"
					],
					[[
						6,
						[
							7,
							"mean0",
							"mul"
						],
						["mean0"]
					]]
				]
			],
			[
				13,
				"var1",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"sq1",
								"mul"
							],
							["invN"]
						],
						"sub"
					],
					[[
						6,
						[
							7,
							"mean1",
							"mul"
						],
						["mean1"]
					]]
				]
			],
			[
				13,
				"var2",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"sq2",
								"mul"
							],
							["invN"]
						],
						"sub"
					],
					[[
						6,
						[
							7,
							"mean2",
							"mul"
						],
						["mean2"]
					]]
				]
			],
			[
				13,
				"var3",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"sq3",
								"mul"
							],
							["invN"]
						],
						"sub"
					],
					[[
						6,
						[
							7,
							"mean3",
							"mul"
						],
						["mean3"]
					]]
				]
			],
			[
				13,
				"sigma0",
				[
					1,
					[
						1,
						[
							7,
							"var0",
							"x"
						],
						"+",
						[
							7,
							"var0",
							"y"
						]
					],
					"+",
					[
						7,
						"var0",
						"z"
					]
				]
			],
			[
				13,
				"sigma1",
				[
					1,
					[
						1,
						[
							7,
							"var1",
							"x"
						],
						"+",
						[
							7,
							"var1",
							"y"
						]
					],
					"+",
					[
						7,
						"var1",
						"z"
					]
				]
			],
			[
				13,
				"sigma2",
				[
					1,
					[
						1,
						[
							7,
							"var2",
							"x"
						],
						"+",
						[
							7,
							"var2",
							"y"
						]
					],
					"+",
					[
						7,
						"var2",
						"z"
					]
				]
			],
			[
				13,
				"sigma3",
				[
					1,
					[
						1,
						[
							7,
							"var3",
							"x"
						],
						"+",
						[
							7,
							"var3",
							"y"
						]
					],
					"+",
					[
						7,
						"var3",
						"z"
					]
				]
			],
			[
				12,
				"result",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"mean0",
							"x"
						],
						[
							7,
							"mean0",
							"y"
						],
						[
							7,
							"mean0",
							"z"
						]
					]
				]
			],
			[
				12,
				"best",
				"sigma0"
			],
			[
				2,
				"result",
				"=",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"result",
						"mean1",
						[
							1,
							"sigma1",
							"<",
							"best"
						]
					]
				]
			],
			[
				2,
				"best",
				"=",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					["best", "sigma1"]
				]
			],
			[
				2,
				"result",
				"=",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"result",
						"mean2",
						[
							1,
							"sigma2",
							"<",
							"best"
						]
					]
				]
			],
			[
				2,
				"best",
				"=",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					["best", "sigma2"]
				]
			],
			[
				2,
				"result",
				"=",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"result",
						"mean3",
						[
							1,
							"sigma3",
							"<",
							"best"
						]
					]
				]
			],
			[
				2,
				"best",
				"=",
				[
					6,
					[
						7,
						"std",
						"min"
					],
					["best", "sigma3"]
				]
			],
			[
				13,
				"paperMul",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[5, "1"],
						"grain",
						"paper"
					]
				]
			],
			[
				13,
				"paperTint",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"paperColor",
							"x"
						],
						[
							7,
							"paperColor",
							"y"
						],
						[
							7,
							"paperColor",
							"z"
						]
					]
				]
			],
			[
				2,
				"result",
				"=",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						[
							6,
							[
								7,
								"result",
								"mul"
							],
							["paperMul"]
						],
						[
							6,
							[
								7,
								"result",
								"mul"
							],
							["paperTint"]
						],
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							[[
								1,
								"paper",
								"*",
								[5, "0.4"]
							]]
						]
					]
				]
			],
			[
				13,
				"origRgb",
				[
					6,
					[
						7,
						"d",
						"vec3f"
					],
					[
						[
							7,
							"originalBleed",
							"x"
						],
						[
							7,
							"originalBleed",
							"y"
						],
						[
							7,
							"originalBleed",
							"z"
						]
					]
				]
			],
			[
				13,
				"finalRGB",
				[
					6,
					[
						7,
						"std",
						"mix"
					],
					[
						"origRgb",
						"result",
						[
							6,
							[
								7,
								"d",
								"vec3f"
							],
							["strength"]
						]
					]
				]
			],
			[10, "finalRGB"]
		]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Watercolor",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Stylize",
	description: "Painterly watercolor look — Kuwahara flattening, pigment edge darkening, paper grain and bleeding",
	requiresRTT: true,
	requiresChild: true,
	props: {
		radius: {
			default: 3,
			compileTime: true,
			description: "Brush size — radius of the painterly flattening",
			ui: {
				type: "range",
				min: 1,
				max: 8,
				step: 1,
				label: "Brush",
				group: "Effect",
				hidden: true
			}
		},
		bleed: {
			default: 1,
			description: "Amount the pigment bleeds and wobbles past hard edges",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 3,
				step: .01,
				label: "Bleed",
				group: "Effect"
			}
		},
		strength: {
			default: 1,
			description: "Blend between the original image (0) and the full watercolor effect (1)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Strength",
				group: "Effect"
			}
		},
		paper: {
			default: .35,
			description: "Strength of the paper grain the pigment settles into",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Paper",
				group: "Effect"
			}
		},
		paperColor: {
			default: "#fbf7ec",
			transform: transformColor,
			description: "Tint of the paper grain",
			ui: {
				type: "color",
				label: "Paper Color",
				group: "Colors"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, propValues, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		const vp = ctx.viewportSize;
		const R = Math.max(1, Math.min(8, Math.round(propValues.radius ?? 4)));
		const invN = 1 / ((R + 1) * (R + 1));
		const uvBase = call(watercolorUvBase, "watercolorUvBase", [
			ctx.uv,
			vp,
			uniforms.bleed,
			ctx.time
		]);
		const unpremult = (uv) => call(unpremultiplyAlpha, "unpremultiplyAlpha", [texture.sample(uv)]).member("rgb");
		const quadrant = (dx, dy) => {
			let sum = expr("vec3f(0.0, 0.0, 0.0)");
			let sq = expr("vec3f(0.0, 0.0, 0.0)");
			for (let j = 0; j <= R; j++) for (let i = 0; i <= R; i++) {
				const c = unpremult(call(watercolorTapUV, "watercolorTapUV", [
					uvBase,
					vp,
					floatE(i * dx),
					floatE(j * dy)
				]));
				sum = sum.add(c);
				sq = sq.add(c.mul(c));
			}
			return {
				sum,
				sq
			};
		};
		const q0 = quadrant(-1, -1);
		const q1 = quadrant(1, -1);
		const q2 = quadrant(-1, 1);
		const q3 = quadrant(1, 1);
		const grain = call(watercolorGrain, "watercolorGrain", [ctx.uv, vp]);
		const originalBleedSample = call(unpremultiplyAlpha, "unpremultiplyAlpha", [texture.sample(uvBase)]);
		return vec4(call(watercolorCompose, "watercolorCompose", [
			q0.sum,
			q1.sum,
			q2.sum,
			q3.sum,
			q0.sq,
			q1.sq,
			q2.sq,
			q3.sq,
			floatE(invN),
			originalBleedSample,
			uniforms.paperColor,
			uniforms.paper,
			uniforms.strength,
			grain
		]), originalBleedSample.member("a"));
	}
};
var Watercolor_default = componentDefinition;
export { watercolorTapUV as a, watercolorGrain as i, componentDefinition as n, watercolorUvBase as o, watercolorCompose as r, Watercolor_default as t };
