// shaders/dist/core/WaveDistortion-DJ1itI6L.js
import { At as call, Ot as ZERO, R as std, Wt as unpremultiplyAlpha, bt as applyEdgeHandlingExpr, n as tgpu, on as d, xt as composeEdgeRemapExpr } from "./coords-BlwgBpJf.js";
import { s as transformEdges } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
var DEG_TO_RAD = Math.PI / 180;
var TAU = Math.PI * 2;
const wavePhase = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, angle, frequency, t) => {
	"use gpu";
	const centeredUV = uv.sub(d.vec2f(.5, .5));
	const aspectCorrectedUV = d.vec2f(__tsover_mul(centeredUV.x, aspect), centeredUV.y);
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const cosA = std.cos(angleRad);
	const sinA = std.sin(angleRad);
	const rotatedY = __tsover_add(__tsover_mul(aspectCorrectedUV.x, sinA), __tsover_mul(aspectCorrectedUV.y, cosA));
	return __tsover_add(__tsover_mul(__tsover_mul(__tsover_add(rotatedY, .5), frequency), TAU), t);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "frequency"
			},
			{
				"type": "i",
				"name": "t"
			}
		],
		"body": [0, [
			[
				13,
				"centeredUV",
				[
					6,
					[
						7,
						"uv",
						"sub"
					],
					[[
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "0.5"], [5, "0.5"]]
					]]
				]
			],
			[
				13,
				"aspectCorrectedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"centeredUV",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"centeredUV",
						"y"
					]]
				]
			],
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"rotatedY",
				[
					1,
					[
						1,
						[
							7,
							"aspectCorrectedUV",
							"x"
						],
						"*",
						"sinA"
					],
					"+",
					[
						1,
						[
							7,
							"aspectCorrectedUV",
							"y"
						],
						"*",
						"cosA"
					]
				]
			],
			[10, [
				1,
				[
					1,
					[
						1,
						[
							1,
							"rotatedY",
							"+",
							[5, "0.5"]
						],
						"*",
						"frequency"
					],
					"*",
					"TAU"
				],
				"+",
				"t"
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std",
			"TAU"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std,
		TAU
	})
}) && $.f)({}));
const waveApply = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, angle, strength, wave) => {
	"use gpu";
	const angleRad = __tsover_mul(angle, DEG_TO_RAD);
	const cosA = std.cos(angleRad);
	const sinA = std.sin(angleRad);
	const displacement = __tsover_mul(__tsover_mul(wave, strength), .5);
	const displacementX = __tsover_div(__tsover_mul(displacement, cosA), aspect);
	const displacementY = __tsover_mul(displacement, sinA);
	return d.vec2f(__tsover_add(uv.x, displacementX), __tsover_add(uv.y, displacementY));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "wave"
			}
		],
		"body": [0, [
			[
				13,
				"angleRad",
				[
					1,
					"angle",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosA",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["angleRad"]
				]
			],
			[
				13,
				"sinA",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["angleRad"]
				]
			],
			[
				13,
				"displacement",
				[
					1,
					[
						1,
						"wave",
						"*",
						"strength"
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"displacementX",
				[
					1,
					[
						1,
						"displacement",
						"*",
						"cosA"
					],
					"/",
					"aspect"
				]
			],
			[
				13,
				"displacementY",
				[
					1,
					"displacement",
					"*",
					"sinA"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						7,
						"uv",
						"x"
					],
					"+",
					"displacementX"
				], [
					1,
					[
						7,
						"uv",
						"y"
					],
					"+",
					"displacementY"
				]]
			]]
		]],
		"externalNames": [
			"DEG_TO_RAD",
			"std",
			"d"
		]
	},
	externals: () => ({
		DEG_TO_RAD,
		std,
		d
	})
}) && $.f)({}));
const waveDistortSine = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, angle, frequency, strength, t) => {
	"use gpu";
	const phase = wavePhase(uv, aspect, angle, frequency, t);
	return waveApply(uv, aspect, angle, strength, std.sin(phase));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "frequency"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "t"
			}
		],
		"body": [0, [
			[
				13,
				"phase",
				[
					6,
					"wavePhase",
					[
						"uv",
						"aspect",
						"angle",
						"frequency",
						"t"
					]
				]
			],
			[
				13,
				"wave",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["phase"]
				]
			],
			[10, [
				6,
				"waveApply",
				[
					"uv",
					"aspect",
					"angle",
					"strength",
					"wave"
				]
			]]
		]],
		"externalNames": [
			"wavePhase",
			"std",
			"waveApply"
		]
	},
	externals: () => ({
		wavePhase,
		std,
		waveApply
	})
}) && $.f)({}));
const waveDistortTriangle = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, angle, frequency, strength, t) => {
	"use gpu";
	const phase = wavePhase(uv, aspect, angle, frequency, t);
	const normalizedPhase = std.fract(__tsover_div(phase, TAU));
	return waveApply(uv, aspect, angle, strength, __tsover_sub(__tsover_mul(std.abs(__tsover_sub(__tsover_mul(normalizedPhase, 2), 1)), 2), 1));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "frequency"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "t"
			}
		],
		"body": [0, [
			[
				13,
				"phase",
				[
					6,
					"wavePhase",
					[
						"uv",
						"aspect",
						"angle",
						"frequency",
						"t"
					]
				]
			],
			[
				13,
				"normalizedPhase",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						"phase",
						"/",
						"TAU"
					]]
				]
			],
			[
				13,
				"wave",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"abs"
							],
							[[
								1,
								[
									1,
									"normalizedPhase",
									"*",
									[5, "2"]
								],
								"-",
								[5, "1"]
							]]
						],
						"*",
						[5, "2"]
					],
					"-",
					[5, "1"]
				]
			],
			[10, [
				6,
				"waveApply",
				[
					"uv",
					"aspect",
					"angle",
					"strength",
					"wave"
				]
			]]
		]],
		"externalNames": [
			"wavePhase",
			"std",
			"TAU",
			"waveApply"
		]
	},
	externals: () => ({
		wavePhase,
		std,
		TAU,
		waveApply
	})
}) && $.f)({}));
const waveDistortSquare = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, angle, frequency, strength, t) => {
	"use gpu";
	const phase = wavePhase(uv, aspect, angle, frequency, t);
	const normalizedPhase = std.fract(__tsover_div(phase, TAU));
	return waveApply(uv, aspect, angle, strength, __tsover_sub(__tsover_mul(std.step(.5, normalizedPhase), 2), 1));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "frequency"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "t"
			}
		],
		"body": [0, [
			[
				13,
				"phase",
				[
					6,
					"wavePhase",
					[
						"uv",
						"aspect",
						"angle",
						"frequency",
						"t"
					]
				]
			],
			[
				13,
				"normalizedPhase",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						"phase",
						"/",
						"TAU"
					]]
				]
			],
			[
				13,
				"wave",
				[
					1,
					[
						1,
						[
							6,
							[
								7,
								"std",
								"step"
							],
							[[5, "0.5"], "normalizedPhase"]
						],
						"*",
						[5, "2"]
					],
					"-",
					[5, "1"]
				]
			],
			[10, [
				6,
				"waveApply",
				[
					"uv",
					"aspect",
					"angle",
					"strength",
					"wave"
				]
			]]
		]],
		"externalNames": [
			"wavePhase",
			"std",
			"TAU",
			"waveApply"
		]
	},
	externals: () => ({
		wavePhase,
		std,
		TAU,
		waveApply
	})
}) && $.f)({}));
const waveDistortSawtooth = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, angle, frequency, strength, t) => {
	"use gpu";
	const phase = wavePhase(uv, aspect, angle, frequency, t);
	const normalizedPhase = std.fract(__tsover_div(phase, TAU));
	return waveApply(uv, aspect, angle, strength, __tsover_sub(__tsover_mul(normalizedPhase, 2), 1));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "frequency"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "t"
			}
		],
		"body": [0, [
			[
				13,
				"phase",
				[
					6,
					"wavePhase",
					[
						"uv",
						"aspect",
						"angle",
						"frequency",
						"t"
					]
				]
			],
			[
				13,
				"normalizedPhase",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					[[
						1,
						"phase",
						"/",
						"TAU"
					]]
				]
			],
			[
				13,
				"wave",
				[
					1,
					[
						1,
						"normalizedPhase",
						"*",
						[5, "2"]
					],
					"-",
					[5, "1"]
				]
			],
			[10, [
				6,
				"waveApply",
				[
					"uv",
					"aspect",
					"angle",
					"strength",
					"wave"
				]
			]]
		]],
		"externalNames": [
			"wavePhase",
			"std",
			"TAU",
			"waveApply"
		]
	},
	externals: () => ({
		wavePhase,
		std,
		TAU,
		waveApply
	})
}) && $.f)({}));
const waveDistortBounce = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, aspect, angle, frequency, strength, t) => {
	"use gpu";
	const phase = wavePhase(uv, aspect, angle, frequency, t);
	return waveApply(uv, aspect, angle, strength, std.abs(std.sin(phase)));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "angle"
			},
			{
				"type": "i",
				"name": "frequency"
			},
			{
				"type": "i",
				"name": "strength"
			},
			{
				"type": "i",
				"name": "t"
			}
		],
		"body": [0, [
			[
				13,
				"phase",
				[
					6,
					"wavePhase",
					[
						"uv",
						"aspect",
						"angle",
						"frequency",
						"t"
					]
				]
			],
			[
				13,
				"wave",
				[
					6,
					[
						7,
						"std",
						"abs"
					],
					[[
						6,
						[
							7,
							"std",
							"sin"
						],
						["phase"]
					]]
				]
			],
			[10, [
				6,
				"waveApply",
				[
					"uv",
					"aspect",
					"angle",
					"strength",
					"wave"
				]
			]]
		]],
		"externalNames": [
			"wavePhase",
			"std",
			"waveApply"
		]
	},
	externals: () => ({
		wavePhase,
		std,
		waveApply
	})
}) && $.f)({}));
var WAVE_MODES = {
	sine: 0,
	triangle: 1,
	square: 2,
	sawtooth: 3,
	bounce: 4
};
function waveTypeMode(raw) {
	if (typeof raw === "number") return raw;
	return WAVE_MODES[raw] ?? 0;
}
function waveBodyFor(waveType) {
	switch (waveType) {
		case 1: return {
			fn: waveDistortTriangle,
			hint: "waveDistortTriangle"
		};
		case 2: return {
			fn: waveDistortSquare,
			hint: "waveDistortSquare"
		};
		case 3: return {
			fn: waveDistortSawtooth,
			hint: "waveDistortSawtooth"
		};
		case 4: return {
			fn: waveDistortBounce,
			hint: "waveDistortBounce"
		};
		default: return {
			fn: waveDistortSine,
			hint: "waveDistortSine"
		};
	}
}
const componentDefinition = {
	name: "WaveDistortion",
	category: "Distortions",
	description: "Wave-based distortion with multiple waveform types",
	requiresRTT: true,
	requiresChild: true,
	boundingBoxDeclaration: { aspectRatio: null },
	animatedTime: { speed: "speed" },
	props: {
		strength: {
			default: .3,
			description: "Distortion intensity",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Strength",
				group: "Effect"
			}
		},
		frequency: {
			default: 1,
			description: "Number of bends/waves",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 10,
				step: .1,
				label: "Frequency",
				group: "Effect"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		angle: {
			default: 0,
			description: "Direction of wave distortion in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		waveType: {
			default: "sine",
			description: "Shape of the distortion wave",
			compileTime: true,
			transform: (value) => {
				return {
					"sine": 0,
					"triangle": 1,
					"square": 2,
					"sawtooth": 3,
					"bounce": 4
				}[value] ?? 0;
			},
			ui: {
				type: "select",
				options: [
					{
						label: "Sine",
						value: "sine"
					},
					{
						label: "Triangle",
						value: "triangle"
					},
					{
						label: "Square",
						value: "square"
					},
					{
						label: "Sawtooth",
						value: "sawtooth"
					},
					{
						label: "Bounce",
						value: "bounce"
					}
				],
				label: "Wave Type",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, childNode, ctx, propValues, convertToTexture } = params;
		if (!childNode) {
			console.error("You must pass a child component into the WaveDistortion shader.");
			return ZERO;
		}
		const texture = convertToTexture(childNode);
		const edgeMode = propValues.edges ?? 0;
		const waveType = waveTypeMode(propValues.waveType);
		const t = animatedTime(params).mul(.5);
		const { fn, hint } = waveBodyFor(waveType);
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [applyEdgeHandlingExpr(call(fn, hint, [
			ctx.uv,
			ctx.aspect,
			uniforms.angle,
			uniforms.frequency,
			uniforms.strength,
			t
		]), (uv) => texture.sample(uv), edgeMode)]);
	},
	uvRemap: (ctx) => {
		const edgeMode = ctx.propValues.edges ?? 0;
		const waveType = waveTypeMode(ctx.propValues.waveType);
		const t = animatedTime(ctx).mul(.5);
		const { fn, hint } = waveBodyFor(waveType);
		return composeEdgeRemapExpr(call(fn, hint, [
			ctx.uv,
			ctx.aspect,
			ctx.uniforms.angle,
			ctx.uniforms.frequency,
			ctx.uniforms.strength,
			t
		]), ctx.mask, edgeMode);
	}
};
var WaveDistortion_default = componentDefinition;
export { waveDistortSawtooth as a, waveDistortTriangle as c, waveDistortBounce as i, wavePhase as l, componentDefinition as n, waveDistortSine as o, waveApply as r, waveDistortSquare as s, WaveDistortion_default as t };
