// shaders/dist/core/WaveletNoise-Cy1pQ99p.js
import { At as call, R as std, W as toneAndColorExpr, dt as wavelet12, n as tgpu, on as d, rt as noise_exports } from "./coords-BlwgBpJf.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
import { i as noiseToneProps, r as noiseSpeedProp, t as noiseColorProps } from "./noiseColor-Cr76BYLN.js";
const waveletNoiseField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, scale, seed, animTime, detail) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const pos = d.vec2f(__tsover_mul(uv.x, aspect), uv.y).mul(std.exp(scale)).add(seed);
	return __tsover_add(__tsover_mul(wavelet12(pos, animTime, detail), .5), .5);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "detail"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"aspectUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"pos",
				[
					6,
					[
						7,
						[
							6,
							[
								7,
								"aspectUV",
								"mul"
							],
							[[
								6,
								[
									7,
									"std",
									"exp"
								],
								["scale"]
							]]
						],
						"add"
					],
					["seed"]
				]
			],
			[10, [
				1,
				[
					1,
					[
						6,
						[
							7,
							"noise",
							"wavelet12"
						],
						[
							"pos",
							"animTime",
							"detail"
						]
					],
					"*",
					[5, "0.5"]
				],
				"+",
				[5, "0.5"]
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"noise"
		]
	},
	externals: () => ({
		d,
		std,
		noise: noise_exports
	})
}) && $.f)({}));
const componentDefinition = {
	name: "WaveletNoise",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Rotating banded wavelets that ripple as they animate",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		...noiseColorProps(),
		scale: {
			default: 1.5,
			description: "Pattern scale (higher = larger patterns)",
			ui: {
				type: "range",
				min: -2,
				max: 5,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		detail: {
			default: 1.24,
			description: "Per-octave frequency ratio — higher adds finer wavelet detail",
			ui: {
				type: "range",
				min: 1.05,
				max: 2,
				step: .01,
				label: "Detail",
				group: "Effect"
			}
		},
		...noiseToneProps(),
		...noiseSpeedProp()
	},
	fragment: (params) => {
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		return toneAndColorExpr(params, call(waveletNoiseField, "waveletNoiseField", [
			uv,
			viewport,
			params.uniforms.scale,
			params.uniforms.seed,
			t,
			params.uniforms.detail
		]));
	}
};
var WaveletNoise_default = componentDefinition;
export { componentDefinition as n, waveletNoiseField as r, WaveletNoise_default as t };
