// shaders/dist/core/Weave--Ex2UVBt.js
import { At as call, R as std, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { a as transformColor } from "./transformations-fqzznt4g.js";
var DEG_TO_RAD = Math.PI / 180;
const weaveColor = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.vec4f,
	d.vec4f
], d.vec4f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, cells, gap, rotation, colorA, colorB) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const correctedUV = d.vec2f(__tsover_mul(uv.x, aspect), uv.y);
	const rotRad = __tsover_mul(rotation, DEG_TO_RAD);
	const cosR = std.cos(rotRad);
	const sinR = std.sin(rotRad);
	const centerX = __tsover_mul(aspect, .5);
	const centerY = .5;
	const cx = __tsover_sub(correctedUV.x, centerX);
	const cy = __tsover_sub(correctedUV.y, centerY);
	const scaledUV = d.vec2f(__tsover_add(__tsover_sub(__tsover_mul(cx, cosR), __tsover_mul(cy, sinR)), centerX), __tsover_add(__tsover_add(__tsover_mul(cx, sinR), __tsover_mul(cy, cosR)), centerY)).mul(cells);
	const cell = std.floor(scaledUV);
	const localUV = std.fract(scaledUV);
	const pixel = __tsover_mul(std.fwidth(scaledUV.x), .5);
	const hBand = __tsover_mul(std.smoothstep(gap, __tsover_add(gap, pixel), localUV.y), std.smoothstep(gap, __tsover_add(gap, pixel), __tsover_sub(1, localUV.y)));
	const vBand = __tsover_mul(std.smoothstep(gap, __tsover_add(gap, pixel), localUV.x), std.smoothstep(gap, __tsover_add(gap, pixel), __tsover_sub(1, localUV.x)));
	const s = __tsover_add(cell.x, cell.y);
	const isHTopFloat = __tsover_sub(s, __tsover_mul(2, std.floor(__tsover_div(s, 2))));
	const intersection = __tsover_mul(hBand, vBand);
	const hVis = std.clamp(__tsover_sub(hBand, __tsover_mul(intersection, isHTopFloat)), 0, 1);
	const vVis = std.clamp(__tsover_sub(vBand, __tsover_mul(intersection, __tsover_sub(1, isHTopFloat))), 0, 1);
	const weightA = __tsover_mul(colorA.w, hVis);
	const weightB = __tsover_mul(colorB.w, vVis);
	const totalAlpha = std.clamp(__tsover_add(weightA, weightB), 0, 1);
	const safeWeight = std.max(__tsover_add(weightA, weightB), .001);
	const r = __tsover_div(__tsover_add(__tsover_mul(colorA.x, weightA), __tsover_mul(colorB.x, weightB)), safeWeight);
	const g = __tsover_div(__tsover_add(__tsover_mul(colorA.y, weightA), __tsover_mul(colorB.y, weightB)), safeWeight);
	const b = __tsover_div(__tsover_add(__tsover_mul(colorA.z, weightA), __tsover_mul(colorB.z, weightB)), safeWeight);
	return d.vec4f(r, g, b, totalAlpha);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "cells"
			},
			{
				"type": "i",
				"name": "gap"
			},
			{
				"type": "i",
				"name": "rotation"
			},
			{
				"type": "i",
				"name": "colorA"
			},
			{
				"type": "i",
				"name": "colorB"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"correctedUV",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							7,
							"uv",
							"x"
						],
						"*",
						"aspect"
					], [
						7,
						"uv",
						"y"
					]]
				]
			],
			[
				13,
				"rotRad",
				[
					1,
					"rotation",
					"*",
					"DEG_TO_RAD"
				]
			],
			[
				13,
				"cosR",
				[
					6,
					[
						7,
						"std",
						"cos"
					],
					["rotRad"]
				]
			],
			[
				13,
				"sinR",
				[
					6,
					[
						7,
						"std",
						"sin"
					],
					["rotRad"]
				]
			],
			[
				13,
				"centerX",
				[
					1,
					"aspect",
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"centerY",
				[5, "0.5"]
			],
			[
				13,
				"cx",
				[
					1,
					[
						7,
						"correctedUV",
						"x"
					],
					"-",
					"centerX"
				]
			],
			[
				13,
				"cy",
				[
					1,
					[
						7,
						"correctedUV",
						"y"
					],
					"-",
					"centerY"
				]
			],
			[
				13,
				"rotated",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						1,
						[
							1,
							[
								1,
								"cx",
								"*",
								"cosR"
							],
							"-",
							[
								1,
								"cy",
								"*",
								"sinR"
							]
						],
						"+",
						"centerX"
					], [
						1,
						[
							1,
							[
								1,
								"cx",
								"*",
								"sinR"
							],
							"+",
							[
								1,
								"cy",
								"*",
								"cosR"
							]
						],
						"+",
						"centerY"
					]]
				]
			],
			[
				13,
				"scaledUV",
				[
					6,
					[
						7,
						"rotated",
						"mul"
					],
					["cells"]
				]
			],
			[
				13,
				"cell",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["scaledUV"]
				]
			],
			[
				13,
				"localUV",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["scaledUV"]
				]
			],
			[
				13,
				"pixel",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"fwidth"
						],
						[[
							7,
							"scaledUV",
							"x"
						]]
					],
					"*",
					[5, "0.5"]
				]
			],
			[
				13,
				"hBand",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							"gap",
							[
								1,
								"gap",
								"+",
								"pixel"
							],
							[
								7,
								"localUV",
								"y"
							]
						]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							"gap",
							[
								1,
								"gap",
								"+",
								"pixel"
							],
							[
								1,
								[5, "1"],
								"-",
								[
									7,
									"localUV",
									"y"
								]
							]
						]
					]
				]
			],
			[
				13,
				"vBand",
				[
					1,
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							"gap",
							[
								1,
								"gap",
								"+",
								"pixel"
							],
							[
								7,
								"localUV",
								"x"
							]
						]
					],
					"*",
					[
						6,
						[
							7,
							"std",
							"smoothstep"
						],
						[
							"gap",
							[
								1,
								"gap",
								"+",
								"pixel"
							],
							[
								1,
								[5, "1"],
								"-",
								[
									7,
									"localUV",
									"x"
								]
							]
						]
					]
				]
			],
			[
				13,
				"s",
				[
					1,
					[
						7,
						"cell",
						"x"
					],
					"+",
					[
						7,
						"cell",
						"y"
					]
				]
			],
			[
				13,
				"isHTopFloat",
				[
					1,
					"s",
					"-",
					[
						1,
						[5, "2"],
						"*",
						[
							6,
							[
								7,
								"std",
								"floor"
							],
							[[
								1,
								"s",
								"/",
								[5, "2"]
							]]
						]
					]
				]
			],
			[
				13,
				"intersection",
				[
					1,
					"hBand",
					"*",
					"vBand"
				]
			],
			[
				13,
				"hVis",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"hBand",
							"-",
							[
								1,
								"intersection",
								"*",
								"isHTopFloat"
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"vVis",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"vBand",
							"-",
							[
								1,
								"intersection",
								"*",
								[
									1,
									[5, "1"],
									"-",
									"isHTopFloat"
								]
							]
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"weightA",
				[
					1,
					[
						7,
						"colorA",
						"w"
					],
					"*",
					"hVis"
				]
			],
			[
				13,
				"weightB",
				[
					1,
					[
						7,
						"colorB",
						"w"
					],
					"*",
					"vVis"
				]
			],
			[
				13,
				"totalAlpha",
				[
					6,
					[
						7,
						"std",
						"clamp"
					],
					[
						[
							1,
							"weightA",
							"+",
							"weightB"
						],
						[5, "0"],
						[5, "1"]
					]
				]
			],
			[
				13,
				"safeWeight",
				[
					6,
					[
						7,
						"std",
						"max"
					],
					[[
						1,
						"weightA",
						"+",
						"weightB"
					], [5, "0.001"]]
				]
			],
			[
				13,
				"r",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"colorA",
								"x"
							],
							"*",
							"weightA"
						],
						"+",
						[
							1,
							[
								7,
								"colorB",
								"x"
							],
							"*",
							"weightB"
						]
					],
					"/",
					"safeWeight"
				]
			],
			[
				13,
				"g",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"colorA",
								"y"
							],
							"*",
							"weightA"
						],
						"+",
						[
							1,
							[
								7,
								"colorB",
								"y"
							],
							"*",
							"weightB"
						]
					],
					"/",
					"safeWeight"
				]
			],
			[
				13,
				"b",
				[
					1,
					[
						1,
						[
							1,
							[
								7,
								"colorA",
								"z"
							],
							"*",
							"weightA"
						],
						"+",
						[
							1,
							[
								7,
								"colorB",
								"z"
							],
							"*",
							"weightB"
						]
					],
					"/",
					"safeWeight"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec4f"
				],
				[
					"r",
					"g",
					"b",
					"totalAlpha"
				]
			]]
		]],
		"externalNames": [
			"d",
			"DEG_TO_RAD",
			"std"
		]
	},
	externals: () => ({
		d,
		DEG_TO_RAD,
		std
	})
}) && $.f)({}));
const componentDefinition = {
	name: "Weave",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Interlaced textile weave pattern with two thread colors going over and under each other",
	acceptsUVContext: true,
	props: {
		colorA: {
			default: "#c4c4c4",
			transform: transformColor,
			description: "Horizontal thread color",
			ui: {
				type: "color",
				label: "X Color",
				group: "Colors"
			}
		},
		colorB: {
			default: "#4d4d4d",
			transform: transformColor,
			description: "Vertical thread color",
			ui: {
				type: "color",
				label: "Y Color",
				group: "Colors"
			}
		},
		cells: {
			default: 10,
			description: "Number of threads across the shortest canvas edge",
			ui: {
				type: ["range", "map"],
				min: 2,
				max: 40,
				step: 1,
				label: "Cells",
				group: "Effect"
			}
		},
		gap: {
			default: .25,
			description: "Gap between threads (0 = no gap, 0.5 = maximum gap)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .45,
				step: .01,
				label: "Gap",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the weave pattern in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		}
	},
	fragment: (params) => {
		const { uniforms } = params;
		return call(weaveColor, "weaveColor", [
			params.uvContext ?? params.ctx.uv,
			params.effectiveViewportSize ?? params.ctx.viewportSize,
			uniforms.cells,
			uniforms.gap,
			uniforms.rotation,
			uniforms.colorA,
			uniforms.colorB
		]);
	}
};
var Weave_default = componentDefinition;
export { componentDefinition as n, weaveColor as r, Weave_default as t };
