// shaders/dist/core/WebcamTexture-DNCb4NqI.js
import { At as call, R as std, a as scaleContain, c as scaleNone, i as sampleUV, l as scaleScaleDown, n as tgpu, o as scaleCover, on as d, r as alphaCutLinear, s as scaleFill } from "./coords-BlwgBpJf.js";
var OBJECT_FIT_MODES = {
	cover: 0,
	contain: 1,
	fill: 2,
	"scale-down": 3,
	none: 4
};
const applyMirror = tgpu.fn([d.vec2f, d.f32], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, mirror) => {
	"use gpu";
	const mirroredX = std.select(uv.x, __tsover_sub(1, uv.x), mirror > 0);
	return d.vec2f(mirroredX, uv.y);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "uv"
		}, {
			"type": "i",
			"name": "mirror"
		}],
		"body": [0, [[
			13,
			"mirroredX",
			[
				6,
				[
					7,
					"std",
					"select"
				],
				[
					[
						7,
						"uv",
						"x"
					],
					[
						1,
						[5, "1"],
						"-",
						[
							7,
							"uv",
							"x"
						]
					],
					[
						1,
						"mirror",
						">",
						[5, "0"]
					]
				]
			]
		], [10, [
			6,
			[
				7,
				"d",
				"vec2f"
			],
			["mirroredX", [
				7,
				"uv",
				"y"
			]]
		]]]],
		"externalNames": ["std", "d"]
	},
	externals: () => ({
		std,
		d
	})
}) && $.f)({}));
const componentDefinition = {
	name: "WebcamTexture",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Display a live webcam feed with customizable object-fit modes",
	acceptsUVContext: true,
	props: {
		objectFit: {
			default: "cover",
			description: "How the webcam feed should be sized within the viewport",
			transform: (value) => {
				return {
					"cover": 0,
					"contain": 1,
					"fill": 2,
					"scale-down": 3,
					"none": 4
				}[value] ?? 0;
			},
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Cover",
						value: "cover"
					},
					{
						label: "Contain",
						value: "contain"
					},
					{
						label: "Fill",
						value: "fill"
					},
					{
						label: "Scale Down",
						value: "scale-down"
					},
					{
						label: "None",
						value: "none"
					}
				],
				label: "Object Fit",
				group: "Sizing"
			}
		},
		mirror: {
			default: true,
			description: "Mirror the webcam feed horizontally (selfie mode)",
			ui: {
				type: "checkbox",
				label: "Mirror",
				group: "Effect"
			}
		}
	},
	fragment: ({ ctx, propValues, uniforms, uvContext, effectiveViewportSize, onBeforeRender, onCleanup, registerExternalTexture }) => {
		let mediaStream = null;
		let videoElement = null;
		let isInitialized = false;
		let isDisposed = false;
		const startWebcam = async () => {
			if (isInitialized || isDisposed) return;
			try {
				mediaStream = await navigator.mediaDevices.getUserMedia({
					video: {
						width: { ideal: 1280 },
						height: { ideal: 720 },
						facingMode: "user"
					},
					audio: false
				});
				if (isDisposed) {
					mediaStream.getTracks().forEach((track) => track.stop());
					mediaStream = null;
					return;
				}
				const video = document.createElement("video");
				video.srcObject = mediaStream;
				video.playsInline = true;
				video.muted = true;
				await new Promise((resolve, reject) => {
					video.onloadedmetadata = () => resolve();
					video.onerror = () => reject(/* @__PURE__ */ new Error("Failed to load video metadata"));
				});
				if (isDisposed) return;
				await video.play();
				if (isDisposed) return;
				videoElement = video;
				isInitialized = true;
				console.log(`[WebcamTexture] Webcam started: ${video.videoWidth}x${video.videoHeight}`);
			} catch (error) {
				if (error instanceof DOMException) if (error.name === "NotAllowedError") console.error("[WebcamTexture] Camera permission denied by user");
				else if (error.name === "NotFoundError") console.error("[WebcamTexture] No camera found on this device");
				else if (error.name === "NotReadableError") console.error("[WebcamTexture] Camera is already in use by another application");
				else console.error("[WebcamTexture] Camera error:", error.message);
				else console.error("[WebcamTexture] Failed to start webcam:", error);
			}
		};
		setTimeout(() => startWebcam(), 0);
		onBeforeRender(() => {});
		onCleanup(() => {
			isDisposed = true;
			if (mediaStream) {
				mediaStream.getTracks().forEach((track) => track.stop());
				mediaStream = null;
			}
			if (videoElement) {
				videoElement.srcObject = null;
				videoElement = null;
			}
			isInitialized = false;
		});
		const external = registerExternalTexture(() => {
			if (!videoElement || isDisposed) return null;
			if (videoElement.readyState < 2 || videoElement.videoWidth === 0) return null;
			return videoElement;
		});
		const baseUV = uvContext ?? ctx.uv;
		const vp = effectiveViewportSize ?? ctx.viewportSize;
		const videoDims = external.dimensions();
		const rawFit = propValues.objectFit;
		const mode = typeof rawFit === "number" ? rawFit : OBJECT_FIT_MODES[rawFit] ?? 0;
		let uvScale;
		if (mode === 1) uvScale = call(scaleContain, "scaleContain", [videoDims, vp]);
		else if (mode === 2) uvScale = call(scaleFill, "scaleFill", []);
		else if (mode === 3) uvScale = call(scaleScaleDown, "scaleScaleDown", [videoDims, vp]);
		else if (mode === 4) uvScale = call(scaleNone, "scaleNone", [videoDims, vp]);
		else uvScale = call(scaleCover, "scaleCover", [videoDims, vp]);
		const finalUV = call(applyMirror, "applyMirror", [call(sampleUV, "mediaSampleUV", [baseUV, uvScale]), uniforms.mirror]);
		return call(alphaCutLinear, "mediaAlphaCut", [external.sample(finalUV), finalUV]);
	}
};
var WebcamTexture_default = componentDefinition;
export { applyMirror as n, componentDefinition as r, WebcamTexture_default as t };
