// shaders/dist/core/WorleyNoise-C9l_C4G0.js
import { At as call, Pt as floatE, Qt as mixColorsVariants, R as std, Zt as mixColorsLinear, n as tgpu, on as d } from "./coords-BlwgBpJf.js";
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-fqzznt4g.js";
import { t as animatedTime } from "./porters-CoaznI7i.js";
var TAU = Math.PI * 2;
const worleyHash = tgpu.fn([d.vec2f], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((p) => {
	"use gpu";
	const h = d.vec2f(std.dot(p, d.vec2f(127.1, 311.7)), std.dot(p, d.vec2f(269.5, 183.3)));
	return std.fract(std.sin(h).mul(43758.5453));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [{
			"type": "i",
			"name": "p"
		}],
		"body": [0, [[
			13,
			"h",
			[
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					6,
					[
						7,
						"std",
						"dot"
					],
					["p", [
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "127.1"], [5, "311.7"]]
					]]
				], [
					6,
					[
						7,
						"std",
						"dot"
					],
					["p", [
						6,
						[
							7,
							"d",
							"vec2f"
						],
						[[5, "269.5"], [5, "183.3"]]
					]]
				]]
			]
		], [10, [
			6,
			[
				7,
				"std",
				"fract"
			],
			[[
				6,
				[
					7,
					[
						6,
						[
							7,
							"std",
							"sin"
						],
						["h"]
					],
					"mul"
				],
				[[5, "43758.5453"]]
			]]
		]]]],
		"externalNames": ["d", "std"]
	},
	externals: () => ({
		d,
		std
	})
}) && $.f)({}));
var transformMode = (value) => {
	return {
		f1: 0,
		f2: 1,
		f2MinusF1: 2,
		f1PlusF2: 3,
		f1TimesF2: 4
	}[value] ?? 0;
};
var transformDistance = (value) => {
	return {
		euclidean: 0,
		manhattan: 1,
		chebyshev: 2
	}[value] ?? 0;
};
function worleyModeNum(raw) {
	return typeof raw === "number" ? raw : transformMode(raw);
}
function worleyDistNum(raw) {
	return typeof raw === "number" ? raw : transformDistance(raw);
}
const worleyEvalOctave = tgpu.fn([
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((oUV, animT, seedOff, jitter, distNum, modeNum) => {
	"use gpu";
	const animTSlow = __tsover_mul(animT, .7);
	const cell = std.floor(oUV);
	const localUV = std.fract(oUV);
	let d1 = d.f32(10);
	let d2 = d.f32(10);
	for (let ny = -1; ny <= 1; ny++) for (let nx = -1; nx <= 1; nx++) {
		const offset = d.vec2f(d.f32(nx), d.f32(ny));
		const h = worleyHash(d.vec2f(__tsover_add(__tsover_add(cell.x, offset.x), seedOff), __tsover_add(__tsover_add(cell.y, offset.y), seedOff)));
		const px = __tsover_add(h.x, __tsover_mul(std.sin(__tsover_add(animT, __tsover_mul(h.x, TAU))), .15));
		const py = __tsover_add(h.y, __tsover_mul(std.cos(__tsover_add(animTSlow, __tsover_mul(h.y, TAU))), .15));
		const jx = std.mix(.5, std.clamp(px, .05, .95), jitter);
		const jy = std.mix(.5, std.clamp(py, .05, .95), jitter);
		const deltaX = __tsover_sub(localUV.x, __tsover_add(offset.x, jx));
		const deltaY = __tsover_sub(localUV.y, __tsover_add(offset.y, jy));
		const dEuclid = __tsover_add(__tsover_mul(deltaX, deltaX), __tsover_mul(deltaY, deltaY));
		const dManhattan = __tsover_add(std.abs(deltaX), std.abs(deltaY));
		const dChebyshev = std.max(std.abs(deltaX), std.abs(deltaY));
		let dd = dEuclid;
		dd = std.select(dd, dManhattan, distNum === 1);
		dd = std.select(dd, dChebyshev, distNum === 2);
		d2 = std.min(d2, std.max(dd, d1));
		d1 = std.min(d1, dd);
	}
	const isEuclid = distNum === 0;
	const sd1 = std.select(d1, std.sqrt(d1), isEuclid);
	const sd2 = std.select(d2, std.sqrt(d2), isEuclid);
	let raw = sd1;
	raw = std.select(raw, sd2, modeNum === 1);
	raw = std.select(raw, __tsover_sub(sd2, sd1), modeNum === 2);
	raw = std.select(raw, __tsover_add(sd1, sd2), modeNum === 3);
	raw = std.select(raw, __tsover_mul(sd1, sd2), modeNum === 4);
	return raw;
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "oUV"
			},
			{
				"type": "i",
				"name": "animT"
			},
			{
				"type": "i",
				"name": "seedOff"
			},
			{
				"type": "i",
				"name": "jitter"
			},
			{
				"type": "i",
				"name": "distNum"
			},
			{
				"type": "i",
				"name": "modeNum"
			}
		],
		"body": [0, [
			[
				13,
				"animTSlow",
				[
					1,
					"animT",
					"*",
					[5, "0.7"]
				]
			],
			[
				13,
				"cell",
				[
					6,
					[
						7,
						"std",
						"floor"
					],
					["oUV"]
				]
			],
			[
				13,
				"localUV",
				[
					6,
					[
						7,
						"std",
						"fract"
					],
					["oUV"]
				]
			],
			[
				12,
				"d1",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "10"]]
				]
			],
			[
				12,
				"d2",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "10"]]
				]
			],
			[
				14,
				[
					12,
					"ny",
					[
						4,
						"-",
						[5, "1"]
					]
				],
				[
					1,
					"ny",
					"<=",
					[5, "1"]
				],
				[
					102,
					"++",
					"ny"
				],
				[0, [[
					14,
					[
						12,
						"nx",
						[
							4,
							"-",
							[5, "1"]
						]
					],
					[
						1,
						"nx",
						"<=",
						[5, "1"]
					],
					[
						102,
						"++",
						"nx"
					],
					[0, [
						[
							13,
							"offset",
							[
								6,
								[
									7,
									"d",
									"vec2f"
								],
								[[
									6,
									[
										7,
										"d",
										"f32"
									],
									["nx"]
								], [
									6,
									[
										7,
										"d",
										"f32"
									],
									["ny"]
								]]
							]
						],
						[
							13,
							"h",
							[
								6,
								"worleyHash",
								[[
									6,
									[
										7,
										"d",
										"vec2f"
									],
									[[
										1,
										[
											1,
											[
												7,
												"cell",
												"x"
											],
											"+",
											[
												7,
												"offset",
												"x"
											]
										],
										"+",
										"seedOff"
									], [
										1,
										[
											1,
											[
												7,
												"cell",
												"y"
											],
											"+",
											[
												7,
												"offset",
												"y"
											]
										],
										"+",
										"seedOff"
									]]
								]]
							]
						],
						[
							13,
							"px",
							[
								1,
								[
									7,
									"h",
									"x"
								],
								"+",
								[
									1,
									[
										6,
										[
											7,
											"std",
											"sin"
										],
										[[
											1,
											"animT",
											"+",
											[
												1,
												[
													7,
													"h",
													"x"
												],
												"*",
												"TAU"
											]
										]]
									],
									"*",
									[5, "0.15"]
								]
							]
						],
						[
							13,
							"py",
							[
								1,
								[
									7,
									"h",
									"y"
								],
								"+",
								[
									1,
									[
										6,
										[
											7,
											"std",
											"cos"
										],
										[[
											1,
											"animTSlow",
											"+",
											[
												1,
												[
													7,
													"h",
													"y"
												],
												"*",
												"TAU"
											]
										]]
									],
									"*",
									[5, "0.15"]
								]
							]
						],
						[
							13,
							"jx",
							[
								6,
								[
									7,
									"std",
									"mix"
								],
								[
									[5, "0.5"],
									[
										6,
										[
											7,
											"std",
											"clamp"
										],
										[
											"px",
											[5, "0.05"],
											[5, "0.95"]
										]
									],
									"jitter"
								]
							]
						],
						[
							13,
							"jy",
							[
								6,
								[
									7,
									"std",
									"mix"
								],
								[
									[5, "0.5"],
									[
										6,
										[
											7,
											"std",
											"clamp"
										],
										[
											"py",
											[5, "0.05"],
											[5, "0.95"]
										]
									],
									"jitter"
								]
							]
						],
						[
							13,
							"deltaX",
							[
								1,
								[
									7,
									"localUV",
									"x"
								],
								"-",
								[
									1,
									[
										7,
										"offset",
										"x"
									],
									"+",
									"jx"
								]
							]
						],
						[
							13,
							"deltaY",
							[
								1,
								[
									7,
									"localUV",
									"y"
								],
								"-",
								[
									1,
									[
										7,
										"offset",
										"y"
									],
									"+",
									"jy"
								]
							]
						],
						[
							13,
							"dEuclid",
							[
								1,
								[
									1,
									"deltaX",
									"*",
									"deltaX"
								],
								"+",
								[
									1,
									"deltaY",
									"*",
									"deltaY"
								]
							]
						],
						[
							13,
							"dManhattan",
							[
								1,
								[
									6,
									[
										7,
										"std",
										"abs"
									],
									["deltaX"]
								],
								"+",
								[
									6,
									[
										7,
										"std",
										"abs"
									],
									["deltaY"]
								]
							]
						],
						[
							13,
							"dChebyshev",
							[
								6,
								[
									7,
									"std",
									"max"
								],
								[[
									6,
									[
										7,
										"std",
										"abs"
									],
									["deltaX"]
								], [
									6,
									[
										7,
										"std",
										"abs"
									],
									["deltaY"]
								]]
							]
						],
						[
							12,
							"dd",
							"dEuclid"
						],
						[
							2,
							"dd",
							"=",
							[
								6,
								[
									7,
									"std",
									"select"
								],
								[
									"dd",
									"dManhattan",
									[
										1,
										"distNum",
										"===",
										[5, "1"]
									]
								]
							]
						],
						[
							2,
							"dd",
							"=",
							[
								6,
								[
									7,
									"std",
									"select"
								],
								[
									"dd",
									"dChebyshev",
									[
										1,
										"distNum",
										"===",
										[5, "2"]
									]
								]
							]
						],
						[
							2,
							"d2",
							"=",
							[
								6,
								[
									7,
									"std",
									"min"
								],
								["d2", [
									6,
									[
										7,
										"std",
										"max"
									],
									["dd", "d1"]
								]]
							]
						],
						[
							2,
							"d1",
							"=",
							[
								6,
								[
									7,
									"std",
									"min"
								],
								["d1", "dd"]
							]
						]
					]]
				]]]
			],
			[
				13,
				"isEuclid",
				[
					1,
					"distNum",
					"===",
					[5, "0"]
				]
			],
			[
				13,
				"sd1",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"d1",
						[
							6,
							[
								7,
								"std",
								"sqrt"
							],
							["d1"]
						],
						"isEuclid"
					]
				]
			],
			[
				13,
				"sd2",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"d2",
						[
							6,
							[
								7,
								"std",
								"sqrt"
							],
							["d2"]
						],
						"isEuclid"
					]
				]
			],
			[
				12,
				"raw",
				"sd1"
			],
			[
				2,
				"raw",
				"=",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"raw",
						"sd2",
						[
							1,
							"modeNum",
							"===",
							[5, "1"]
						]
					]
				]
			],
			[
				2,
				"raw",
				"=",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"raw",
						[
							1,
							"sd2",
							"-",
							"sd1"
						],
						[
							1,
							"modeNum",
							"===",
							[5, "2"]
						]
					]
				]
			],
			[
				2,
				"raw",
				"=",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"raw",
						[
							1,
							"sd1",
							"+",
							"sd2"
						],
						[
							1,
							"modeNum",
							"===",
							[5, "3"]
						]
					]
				]
			],
			[
				2,
				"raw",
				"=",
				[
					6,
					[
						7,
						"std",
						"select"
					],
					[
						"raw",
						[
							1,
							"sd1",
							"*",
							"sd2"
						],
						[
							1,
							"modeNum",
							"===",
							[5, "4"]
						]
					]
				]
			],
			[10, "raw"]
		]],
		"externalNames": [
			"std",
			"d",
			"worleyHash",
			"TAU"
		]
	},
	externals: () => ({
		std,
		d,
		worleyHash,
		TAU
	})
}) && $.f)({}));
const worleyNoiseField = tgpu.fn([
	d.vec2f,
	d.vec2f,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((uv, viewport, scale, seed, jitter, lacunarity, persistence, animTime, octaves, distNum, modeNum, modeScale) => {
	"use gpu";
	const aspect = __tsover_div(viewport.x, viewport.y);
	const baseUVx = __tsover_mul(uv.x, aspect);
	const baseUVy = uv.y;
	let acc = d.f32(0);
	let totalAmp = d.f32(0);
	let amp = d.f32(1);
	let freq = scale;
	for (let o = 0; o < 4; o++) {
		const fo = d.f32(o);
		const gate = std.select(d.f32(0), d.f32(1), fo < octaves);
		const raw = worleyEvalOctave(d.vec2f(__tsover_mul(baseUVx, freq), __tsover_mul(baseUVy, freq)), __tsover_add(animTime, __tsover_mul(fo, 17)), __tsover_add(seed, __tsover_mul(fo, 31)), jitter, distNum, modeNum);
		acc = __tsover_add(acc, __tsover_mul(__tsover_mul(raw, amp), gate));
		totalAmp = __tsover_add(totalAmp, __tsover_mul(amp, gate));
		amp = __tsover_mul(amp, persistence);
		freq = __tsover_mul(freq, lacunarity);
	}
	return __tsover_mul(__tsover_div(acc, std.max(totalAmp, 1e-4)), modeScale);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "viewport"
			},
			{
				"type": "i",
				"name": "scale"
			},
			{
				"type": "i",
				"name": "seed"
			},
			{
				"type": "i",
				"name": "jitter"
			},
			{
				"type": "i",
				"name": "lacunarity"
			},
			{
				"type": "i",
				"name": "persistence"
			},
			{
				"type": "i",
				"name": "animTime"
			},
			{
				"type": "i",
				"name": "octaves"
			},
			{
				"type": "i",
				"name": "distNum"
			},
			{
				"type": "i",
				"name": "modeNum"
			},
			{
				"type": "i",
				"name": "modeScale"
			}
		],
		"body": [0, [
			[
				13,
				"aspect",
				[
					1,
					[
						7,
						"viewport",
						"x"
					],
					"/",
					[
						7,
						"viewport",
						"y"
					]
				]
			],
			[
				13,
				"baseUVx",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"*",
					"aspect"
				]
			],
			[
				13,
				"baseUVy",
				[
					7,
					"uv",
					"y"
				]
			],
			[
				12,
				"acc",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0"]]
				]
			],
			[
				12,
				"totalAmp",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "0"]]
				]
			],
			[
				12,
				"amp",
				[
					6,
					[
						7,
						"d",
						"f32"
					],
					[[5, "1"]]
				]
			],
			[
				12,
				"freq",
				"scale"
			],
			[
				14,
				[
					12,
					"o",
					[5, "0"]
				],
				[
					1,
					"o",
					"<",
					[5, "4"]
				],
				[
					102,
					"++",
					"o"
				],
				[0, [
					[
						13,
						"fo",
						[
							6,
							[
								7,
								"d",
								"f32"
							],
							["o"]
						]
					],
					[
						13,
						"gate",
						[
							6,
							[
								7,
								"std",
								"select"
							],
							[
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "0"]]
								],
								[
									6,
									[
										7,
										"d",
										"f32"
									],
									[[5, "1"]]
								],
								[
									1,
									"fo",
									"<",
									"octaves"
								]
							]
						]
					],
					[
						13,
						"oUV",
						[
							6,
							[
								7,
								"d",
								"vec2f"
							],
							[[
								1,
								"baseUVx",
								"*",
								"freq"
							], [
								1,
								"baseUVy",
								"*",
								"freq"
							]]
						]
					],
					[
						13,
						"animT",
						[
							1,
							"animTime",
							"+",
							[
								1,
								"fo",
								"*",
								[5, "17"]
							]
						]
					],
					[
						13,
						"seedOff",
						[
							1,
							"seed",
							"+",
							[
								1,
								"fo",
								"*",
								[5, "31"]
							]
						]
					],
					[
						13,
						"raw",
						[
							6,
							"worleyEvalOctave",
							[
								"oUV",
								"animT",
								"seedOff",
								"jitter",
								"distNum",
								"modeNum"
							]
						]
					],
					[
						2,
						"acc",
						"=",
						[
							1,
							"acc",
							"+",
							[
								1,
								[
									1,
									"raw",
									"*",
									"amp"
								],
								"*",
								"gate"
							]
						]
					],
					[
						2,
						"totalAmp",
						"=",
						[
							1,
							"totalAmp",
							"+",
							[
								1,
								"amp",
								"*",
								"gate"
							]
						]
					],
					[
						2,
						"amp",
						"=",
						[
							1,
							"amp",
							"*",
							"persistence"
						]
					],
					[
						2,
						"freq",
						"=",
						[
							1,
							"freq",
							"*",
							"lacunarity"
						]
					]
				]]
			],
			[10, [
				1,
				[
					1,
					"acc",
					"/",
					[
						6,
						[
							7,
							"std",
							"max"
						],
						["totalAmp", [5, "0.0001"]]
					]
				],
				"*",
				"modeScale"
			]]
		]],
		"externalNames": [
			"d",
			"std",
			"worleyEvalOctave"
		]
	},
	externals: () => ({
		d,
		std,
		worleyEvalOctave
	})
}) && $.f)({}));
const worleyTone = tgpu.fn([
	d.f32,
	d.f32,
	d.f32
], d.f32)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((normalized, contrast, balance) => {
	"use gpu";
	const signed = __tsover_sub(__tsover_mul(normalized, 2), 1);
	const adjusted = __tsover_add(__tsover_mul(signed, contrast), balance);
	return std.clamp(__tsover_add(__tsover_mul(adjusted, .5), .5), 0, 1);
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "normalized"
			},
			{
				"type": "i",
				"name": "contrast"
			},
			{
				"type": "i",
				"name": "balance"
			}
		],
		"body": [0, [
			[
				13,
				"signed",
				[
					1,
					[
						1,
						"normalized",
						"*",
						[5, "2"]
					],
					"-",
					[5, "1"]
				]
			],
			[
				13,
				"adjusted",
				[
					1,
					[
						1,
						"signed",
						"*",
						"contrast"
					],
					"+",
					"balance"
				]
			],
			[10, [
				6,
				[
					7,
					"std",
					"clamp"
				],
				[
					[
						1,
						[
							1,
							"adjusted",
							"*",
							[5, "0.5"]
						],
						"+",
						[5, "0.5"]
					],
					[5, "0"],
					[5, "1"]
				]
			]]
		]],
		"externalNames": ["std"]
	},
	externals: () => ({ std })
}) && $.f)({}));
const componentDefinition = {
	name: "WorleyNoise",
	boundingBoxDeclaration: {
		aspectRatio: null,
		supportsResizeFit: true
	},
	category: "Textures",
	description: "Cellular noise field — distance-based, with selectable feature combinations and fractal octaves",
	acceptsUVContext: true,
	animatedTime: { speed: "speed" },
	props: {
		colorA: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color where the noise field is low (typically near cell centers)",
			ui: {
				type: "color",
				label: "Color A (Low)",
				group: "Colors"
			}
		},
		colorB: {
			default: "#000000",
			transform: transformColor,
			description: "Color where the noise field is high (typically near cell boundaries)",
			ui: {
				type: "color",
				label: "Color B (High)",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		scale: {
			default: 6,
			description: "Number of cells across the canvas at the base octave",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: .5,
				label: "Scale",
				group: "Effect"
			}
		},
		mode: {
			default: "f1",
			transform: transformMode,
			compileTime: true,
			description: "Field type. F1 = distance to nearest point. F2 = distance to second-nearest. F2 − F1 emphasises cell boundaries.",
			ui: {
				type: "select",
				options: [
					{
						label: "F1 (Cells)",
						value: "f1"
					},
					{
						label: "F2 (Wider Cells)",
						value: "f2"
					},
					{
						label: "F2 − F1 (Edges)",
						value: "f2MinusF1"
					},
					{
						label: "F1 + F2",
						value: "f1PlusF2"
					},
					{
						label: "F1 × F2",
						value: "f1TimesF2"
					}
				],
				label: "Mode",
				group: "Effect"
			}
		},
		distance: {
			default: "euclidean",
			transform: transformDistance,
			compileTime: true,
			description: "Distance metric. Euclidean = round cells. Manhattan = diamond. Chebyshev = square.",
			ui: {
				type: "select",
				options: [
					{
						label: "Euclidean",
						value: "euclidean"
					},
					{
						label: "Manhattan",
						value: "manhattan"
					},
					{
						label: "Chebyshev",
						value: "chebyshev"
					}
				],
				label: "Distance",
				group: "Effect"
			}
		},
		octaves: {
			default: 1,
			compileTime: true,
			description: "Number of fractal layers stacked at progressively finer scales",
			ui: {
				type: "range",
				min: 1,
				max: 4,
				step: 1,
				label: "Octaves",
				group: "Effect"
			}
		},
		lacunarity: {
			default: 2,
			description: "Scale multiplier between octaves (only active when Octaves > 1)",
			ui: {
				type: ["range", "map"],
				min: 1.5,
				max: 4,
				step: .1,
				label: "Lacunarity",
				group: "Effect",
				condition: { octaves: [
					2,
					3,
					4
				] }
			}
		},
		persistence: {
			default: .5,
			description: "Amplitude multiplier between octaves (only active when Octaves > 1)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Persistence",
				group: "Effect",
				condition: { octaves: [
					2,
					3,
					4
				] }
			}
		},
		jitter: {
			default: 1,
			description: "How much each cell's point drifts inside its cell. 0 = rigid grid (banded look), 1 = fully random.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Jitter",
				group: "Effect"
			}
		},
		contrast: {
			default: 1,
			description: "Steepness of the gradient between low and high regions",
			ui: {
				type: ["range", "map"],
				min: .25,
				max: 4,
				step: .01,
				label: "Contrast",
				group: "Effect"
			}
		},
		balance: {
			default: 0,
			description: "Shifts the gradient midpoint. Negative pulls the field toward Color A, positive toward Color B.",
			ui: {
				type: ["range", "map"],
				min: -1,
				max: 1,
				step: .01,
				label: "Balance",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Random seed — shifts the cell pattern without changing its overall structure",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		speed: {
			default: .5,
			description: "Animation speed — how fast each cell's point drifts",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		}
	},
	fragment: (params) => {
		const { uniforms, propValues } = params;
		const uv = params.uvContext ?? params.ctx.uv;
		const viewport = params.effectiveViewportSize ?? params.ctx.viewportSize;
		const t = animatedTime(params);
		const csMode = propValues.colorSpace ?? 0;
		const modeNum = worleyModeNum(propValues.mode);
		const distNum = worleyDistNum(propValues.distance);
		const octavesNum = Math.max(1, Math.min(4, Math.round(propValues.octaves ?? 1)));
		let modeScale = 1.5;
		if (modeNum === 1) modeScale = 1;
		else if (modeNum === 2) modeScale = 2.5;
		else if (modeNum === 3) modeScale = .6;
		else if (modeNum === 4) modeScale = 2;
		if (distNum === 1) modeScale *= .7;
		else if (distNum === 2) modeScale *= 1.4;
		const k = call(worleyTone, "worleyTone", [
			call(worleyNoiseField, "worleyNoiseField", [
				uv,
				viewport,
				uniforms.scale,
				uniforms.seed,
				uniforms.jitter,
				uniforms.lacunarity,
				uniforms.persistence,
				t,
				floatE(octavesNum),
				floatE(distNum),
				floatE(modeNum),
				floatE(modeScale)
			]),
			uniforms.contrast,
			uniforms.balance
		]);
		return call(mixColorsVariants[csMode] ?? mixColorsLinear, "mixColors", [
			uniforms.colorA,
			uniforms.colorB,
			k
		]);
	}
};
var WorleyNoise_default = componentDefinition;
export { worleyNoiseField as a, worleyHash as i, componentDefinition as n, worleyTone as o, worleyEvalOctave as r, WorleyNoise_default as t };
