// shaders/dist/core/ZoomBlur-BR_do-LV.js
import { At as call, Ot as ZERO, Pt as floatE, Wt as unpremultiplyAlpha, n as tgpu, on as d } from "./coords-CWOT0Wqh.js";
import { c as transformPosition } from "./transformations-fqzznt4g.js";
var rawWeights = Array.from({ length: 32 }, (_, i) => {
	const t = (i / 31 - .5) * 2;
	return Math.exp(-.5 * t * t / .64);
});
var wSum = rawWeights.reduce((a, b) => a + b);
const BLUR_WEIGHTS = rawWeights.map((w) => w / wSum);
const TAP_COUNT = 32;
const zoomBlurTapCoord = tgpu.fn([
	d.vec2f,
	d.f32,
	d.vec2f,
	d.f32,
	d.f32
], d.vec2f)(/* @__PURE__ */ (($) => (globalThis.__TYPEGPU_META__ ??= /* @__PURE__ */ new WeakMap()).set($.f = ((center, intensity, uv, aspect, tapIndex) => {
	"use gpu";
	const centerPos = d.vec2f(center.x, __tsover_sub(1, center.y));
	const radius = __tsover_mul(intensity, .01);
	const scale = __tsover_add(1, __tsover_mul(radius, __tsover_div(tapIndex, 31)));
	const deltaX = __tsover_sub(uv.x, centerPos.x);
	const deltaY = __tsover_sub(uv.y, centerPos.y);
	const acdX = __tsover_mul(deltaX, aspect);
	const acdY = deltaY;
	const scaledDeltaX = __tsover_div(acdX, scale);
	const scaledDeltaY = __tsover_div(acdY, scale);
	return d.vec2f(__tsover_add(__tsover_div(scaledDeltaX, aspect), centerPos.x), __tsover_add(scaledDeltaY, centerPos.y));
}), {
	v: 1,
	name: void 0,
	ast: {
		"params": [
			{
				"type": "i",
				"name": "center"
			},
			{
				"type": "i",
				"name": "intensity"
			},
			{
				"type": "i",
				"name": "uv"
			},
			{
				"type": "i",
				"name": "aspect"
			},
			{
				"type": "i",
				"name": "tapIndex"
			}
		],
		"body": [0, [
			[
				13,
				"centerPos",
				[
					6,
					[
						7,
						"d",
						"vec2f"
					],
					[[
						7,
						"center",
						"x"
					], [
						1,
						[5, "1"],
						"-",
						[
							7,
							"center",
							"y"
						]
					]]
				]
			],
			[
				13,
				"radius",
				[
					1,
					"intensity",
					"*",
					[5, "0.01"]
				]
			],
			[
				13,
				"scale",
				[
					1,
					[5, "1"],
					"+",
					[
						1,
						"radius",
						"*",
						[
							1,
							"tapIndex",
							"/",
							[5, "31"]
						]
					]
				]
			],
			[
				13,
				"deltaX",
				[
					1,
					[
						7,
						"uv",
						"x"
					],
					"-",
					[
						7,
						"centerPos",
						"x"
					]
				]
			],
			[
				13,
				"deltaY",
				[
					1,
					[
						7,
						"uv",
						"y"
					],
					"-",
					[
						7,
						"centerPos",
						"y"
					]
				]
			],
			[
				13,
				"acdX",
				[
					1,
					"deltaX",
					"*",
					"aspect"
				]
			],
			[
				13,
				"acdY",
				"deltaY"
			],
			[
				13,
				"scaledDeltaX",
				[
					1,
					"acdX",
					"/",
					"scale"
				]
			],
			[
				13,
				"scaledDeltaY",
				[
					1,
					"acdY",
					"/",
					"scale"
				]
			],
			[10, [
				6,
				[
					7,
					"d",
					"vec2f"
				],
				[[
					1,
					[
						1,
						"scaledDeltaX",
						"/",
						"aspect"
					],
					"+",
					[
						7,
						"centerPos",
						"x"
					]
				], [
					1,
					"scaledDeltaY",
					"+",
					[
						7,
						"centerPos",
						"y"
					]
				]]
			]]
		]],
		"externalNames": ["d"]
	},
	externals: () => ({ d })
}) && $.f)({}));
const componentDefinition = {
	name: "ZoomBlur",
	boundingBoxDeclaration: { aspectRatio: null },
	category: "Blurs",
	description: "Radial zoom blur expanding from a center point",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 30,
			description: "Intensity of the zoom blur effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Effect"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the zoom blur",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		}
	},
	fragment: ({ uniforms, childNode, ctx, convertToTexture }) => {
		if (!childNode) return ZERO;
		const texture = convertToTexture(childNode);
		let total;
		for (let i = 0; i < 32; i++) {
			const coord = call(zoomBlurTapCoord, "zoomBlurTapCoord", [
				uniforms.center,
				uniforms.intensity,
				ctx.uv,
				ctx.aspect,
				floatE(i)
			]);
			const weighted = texture.sample(coord).mul(BLUR_WEIGHTS[i]);
			total = total ? total.add(weighted) : weighted;
		}
		return call(unpremultiplyAlpha, "unpremultiplyAlpha", [total]);
	}
};
var ZoomBlur_default = componentDefinition;
export { zoomBlurTapCoord as a, componentDefinition as i, TAP_COUNT as n, ZoomBlur_default as r, BLUR_WEIGHTS as t };
