// typegpu/core/querySet/querySet.js
import { $internal } from "../../shared/symbols.js";
import { setName } from "../../shared/meta.js";

//#region src/core/querySet/querySet.ts
function INTERNAL_createQuerySet(group, type, count, rawQuerySet) {
	return new TgpuQuerySetImpl(group, type, count, rawQuerySet);
}
function isQuerySet(value) {
	const maybe = value;
	return maybe?.resourceType === "query-set" && !!maybe[$internal];
}
var TgpuQuerySetImpl = class {
	resourceType = "query-set";
	type;
	count;
	#rawQuerySet;
	#device;
	#querySet;
	#ownQuerySet;
	#destroyed = false;
	#available = true;
	#readBuffer = void 0;
	#resolveBuffer = void 0;
	constructor(root, type, count, rawQuerySet) {
		this.#device = root.device;
		this.type = type;
		this.count = count;
		this.#rawQuerySet = rawQuerySet;
		this.#ownQuerySet = !rawQuerySet;
		this.#querySet = rawQuerySet;
	}
	get querySet() {
		if (this.#destroyed) throw new Error("This QuerySet has been destroyed.");
		if (this.#rawQuerySet) return this.#rawQuerySet;
		if (this.#querySet) return this.#querySet;
		this.#querySet = this.#device.createQuerySet({
			type: this.type,
			count: this.count
		});
		return this.#querySet;
	}
	get destroyed() {
		return this.#destroyed;
	}
	get available() {
		return this.#available;
	}
	get [$internal]() {
		const self = this;
		return {
			get readBuffer() {
				if (!self.#readBuffer) self.#readBuffer = self.#device.createBuffer({
					size: self.count * BigUint64Array.BYTES_PER_ELEMENT,
					usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ
				});
				return self.#readBuffer;
			},
			get resolveBuffer() {
				if (!self.#resolveBuffer) self.#resolveBuffer = self.#device.createBuffer({
					size: self.count * BigUint64Array.BYTES_PER_ELEMENT,
					usage: GPUBufferUsage.QUERY_RESOLVE | GPUBufferUsage.COPY_SRC
				});
				return self.#resolveBuffer;
			}
		};
	}
	$name(label) {
		setName(this, label);
		if (this.#querySet) this.#querySet.label = label;
		return this;
	}
	resolve() {
		if (this.#destroyed) throw new Error("This QuerySet has been destroyed.");
		if (!this.#available) throw new Error("This QuerySet is busy resolving or reading.");
		const commandEncoder = this.#device.createCommandEncoder();
		commandEncoder.resolveQuerySet(this.querySet, 0, this.count, this[$internal].resolveBuffer, 0);
		this.#device.queue.submit([commandEncoder.finish()]);
	}
	async read() {
		if (!this.#resolveBuffer) throw new Error("QuerySet must be resolved before reading.");
		this.#available = false;
		try {
			const commandEncoder = this.#device.createCommandEncoder();
			commandEncoder.copyBufferToBuffer(this[$internal].resolveBuffer, 0, this[$internal].readBuffer, 0, this.count * BigUint64Array.BYTES_PER_ELEMENT);
			this.#device.queue.submit([commandEncoder.finish()]);
			const readBuffer = this[$internal].readBuffer;
			await readBuffer.mapAsync(GPUMapMode.READ);
			const data = new BigUint64Array(readBuffer.getMappedRange().slice());
			readBuffer.unmap();
			return Array.from(data);
		} finally {
			this.#available = true;
		}
	}
	destroy() {
		if (this.#destroyed) return;
		this.#destroyed = true;
		if (this.#querySet && this.#ownQuerySet) this.#querySet.destroy();
		this.#readBuffer?.destroy();
		this.#resolveBuffer?.destroy();
		this.#readBuffer = this.#resolveBuffer = void 0;
	}
};

//#endregion
export { INTERNAL_createQuerySet, isQuerySet };
