// typegpu/shared/meta.js
import { version } from "../package.js";
import { DEV, TEST } from "./env.js";
import { $getNameForward, isMarkedInternal } from "./symbols.js";
import { normalizeMetadata } from "./normalizeMetadata.js";

//#region src/shared/meta.ts
const globalExt = globalThis;
if (globalExt.__TYPEGPU_VERSION__ !== void 0) console.warn(`Found duplicate TypeGPU version. First was ${globalExt.__TYPEGPU_VERSION__}, this one is ${version}. This may cause unexpected behavior.`);
globalExt.__TYPEGPU_VERSION__ = version;
globalExt.__TYPEGPU_AUTONAME__ = (exp, label) => isNamable(exp) && isMarkedInternal(exp) && !getName(exp) ? exp.$name(label) : exp;
const nameMap = /* @__PURE__ */ new WeakMap();
function isNamable(value) {
	return !!value?.$name;
}
function isForwarded(value) {
	return !!value?.[$getNameForward];
}
function getName(definition) {
	if (isForwarded(definition)) return getName(definition[$getNameForward]);
	return nameMap.get(definition) ?? globalExt.__TYPEGPU_META__?.get(definition)?.name;
}
function setName(definition, name) {
	if (isForwarded(definition)) {
		setName(definition[$getNameForward], name);
		return;
	}
	nameMap.set(definition, name);
}
const metadataMap = /* @__PURE__ */ new WeakMap();
/**
* Retrieves normalized (non-raw) function metadata.
* If `globalExt.__TYPEGPU_META__` contains raw metadata for the function,
* it is normalized, and then deleted to avoid unnecessary re-normalization.
*/
function getFunctionMetadata(definition) {
	const maybeRawMeta = globalExt.__TYPEGPU_META__?.get(definition);
	if (maybeRawMeta) {
		globalExt.__TYPEGPU_META__?.delete(definition);
		const normalized = normalizeMetadata(maybeRawMeta);
		metadataMap.set(definition, normalized);
		if (maybeRawMeta.name && nameMap.get(definition) === void 0) nameMap.set(definition, maybeRawMeta.name);
	}
	return metadataMap.get(definition);
}
/**
* AST's are given to functions with a 'use gpu' directive, which this function checks for.
*/
function hasTinyestMetadata(value) {
	return typeof value === "function" && !!getFunctionMetadata(value);
}
/**
* Performance measurements are only enabled in dev & test environments for now
*/
const PERF = (DEV || TEST) && {
	get enabled() {
		return !!globalExt.__TYPEGPU_MEASURE_PERF__;
	},
	record(name, data) {
		const records = globalExt.__TYPEGPU_PERF_RECORDS__ ??= /* @__PURE__ */ new Map();
		let entries = records.get(name);
		if (!entries) {
			entries = [];
			records.set(name, entries);
		}
		entries.push(data);
	}
} || void 0;

//#endregion
export { PERF, getFunctionMetadata, getName, hasTinyestMetadata, isNamable, setName };
