// typegpu/shared/symbols.js
import { version } from "../package.js";

//#region src/shared/symbols.ts
const $internal = Symbol(`typegpu:${version}:$internal`);
/**
* The getter to the value of this resource, accessible on the GPU
*/
const $gpuValueOf = Symbol(`typegpu:${version}:$gpuValueOf`);
/**
* If this symbol is present, this means that getName and setName
* will refer to object behind this property instead.
*/
const $getNameForward = Symbol(`typegpu:${version}:$getNameForward`);
/**
* Marks an object with slot-value bindings
*/
const $providing = Symbol(`typegpu:${version}:$providing`);
/**
* Objects can provide the snippet that represents them.
*/
const $ownSnippet = Symbol(`typegpu:${version}:$ownSnippet`);
const $resolve = Symbol(`typegpu:${version}:$resolve`);
/**
* A way for a schema to provide casting behavior, without the need to be explicitly
* callable by the end-user (e.g. vertex formats)
*/
const $cast = Symbol(`typegpu:${version}:$cast`);
/**
* Can be called on the GPU
*/
const $gpuCallable = Symbol(`typegpu:${version}:$gpuCallable`);
/**
* Type token for the inferred (CPU & GPU) representation of a resource
*/
const $repr = Symbol(`typegpu:${version}:$repr`);
/**
* Type token for the inferred (GPU-side) representation of a resource
* If present, it shadows the value of `$repr` for GPU-side inference.
*/
const $gpuRepr = Symbol(`typegpu:${version}:$gpuRepr`);
/**
* Type token for the inferred partial representation of a resource.
* If present, it shadows the value of `$repr` for use in partial IO.
*/
const $reprPartial = Symbol(`typegpu:${version}:$reprPartial`);
/**
* Type token for the inferred patch representation of a resource.
* Used by the `buffer.patch` API with `Record<number, T>` sparse arrays.
*/
const $reprPatch = Symbol(`typegpu:${version}:$reprPatch`);
/**
* Type token for the write-side (input) representation of a resource.
*/
const $inRepr = Symbol(`typegpu:${version}:$inRepr`);
/**
* Type token holding schemas that are identical in memory layout.
*/
const $memIdent = Symbol(`typegpu:${version}:$memIdent`);
/**
* Type token, signaling that a schema can be used in a storage buffer.
*/
const $validStorageSchema = Symbol(`typegpu:${version}:$invalidStorageSchema`);
/**
* Type token, signaling that a schema can be used in a uniform buffer.
*/
const $validUniformSchema = Symbol(`typegpu:${version}:$validUniformSchema`);
/**
* Type token, signaling that a schema can be used in a vertex buffer.
*/
const $validVertexSchema = Symbol(`typegpu:${version}:$validVertexSchema`);
/**
* Type token, containing a reason for why the schema is invalid (if it is).
*/
const $invalidSchemaReason = Symbol(`typegpu:${version}:$invalidSchemaReason`);
function isMarkedInternal(value) {
	return !!value?.[$internal];
}

//#endregion
export { $cast, $getNameForward, $gpuCallable, $gpuValueOf, $internal, $ownSnippet, $providing, $repr, $resolve, isMarkedInternal };
