// typegpu/errors.js
import { DEV, TEST } from "./shared/env.js";
import { getName, hasTinyestMetadata } from "./shared/meta.js";

//#region src/errors.ts
const prefix = "Invariant failed";
/**
* Inspired by: https://github.com/alexreardon/tiny-invariant/blob/master/src/tiny-invariant.ts
*/
function invariant(condition, message) {
	if (condition) return;
	if (!DEV && !TEST) throw new Error(prefix);
	const provided = typeof message === "function" ? message() : message;
	const value = provided ? `${prefix}: ${provided}` : prefix;
	throw new Error(value);
}
/**
* An error that happens during resolution of WGSL code.
* Contains a trace of all ancestor resolvables in
* which this error originated.
*
* @category Errors
*/
var ResolutionError = class ResolutionError extends Error {
	cause;
	trace;
	constructor(cause, trace) {
		let entries = trace.map((ancestor) => `- ${hasTinyestMetadata(ancestor) ? `fn*:${getName(ancestor)}` : ancestor}`);
		if (entries.length > 20) entries = [
			...entries.slice(0, 11),
			"...",
			...entries.slice(-10)
		];
		super(`Resolution of the following tree failed:\n${entries.join("\n")}: ${cause && typeof cause === "object" && "message" in cause ? cause.message : cause}`);
		this.cause = cause;
		this.trace = trace;
		Object.setPrototypeOf(this, ResolutionError.prototype);
	}
	appendToTrace(ancestor) {
		const newTrace = [ancestor, ...this.trace];
		return new ResolutionError(this.cause, newTrace);
	}
};
/**
* An error that happens during execution of TypeGPU functions.
* Contains a trace of all TypeGPU functions called along the way.
*
* @category Errors
*/
var ExecutionError = class ExecutionError extends Error {
	cause;
	trace;
	constructor(cause, trace) {
		let entries = trace.map((ancestor) => `- ${ancestor}`);
		if (entries.length > 20) entries = [
			...entries.slice(0, 11),
			"...",
			...entries.slice(-10)
		];
		super(`Execution of the following tree failed:\n${entries.join("\n")}: ${cause && typeof cause === "object" && "message" in cause ? cause.message : cause}`);
		this.cause = cause;
		this.trace = trace;
		Object.setPrototypeOf(this, ExecutionError.prototype);
	}
	appendToTrace(ancestor) {
		const newTrace = [ancestor, ...this.trace];
		return new ExecutionError(this.cause, newTrace);
	}
};
/**
* @category Errors
*/
var MissingSlotValueError = class MissingSlotValueError extends Error {
	slot;
	constructor(slot) {
		super(`Missing value for '${slot}'`);
		this.slot = slot;
		Object.setPrototypeOf(this, MissingSlotValueError.prototype);
	}
};
/**
* @category Errors
*/
var NotUniformError = class NotUniformError extends Error {
	constructor(value) {
		super(`Buffer '${getName(value) ?? "<unnamed>"}' is not bindable as a uniform. Use .$usage('uniform') to allow it.`);
		Object.setPrototypeOf(this, NotUniformError.prototype);
	}
};
var MissingBindGroupsError = class MissingBindGroupsError extends Error {
	constructor(layouts) {
		super(`Missing bind groups for layouts: '${[...layouts].map((layout) => getName(layout) ?? "<unnamed>").join(", ")}'. Please provide it using pipeline.with(bindGroup).(...)`);
		Object.setPrototypeOf(this, MissingBindGroupsError.prototype);
	}
};
var MissingVertexBuffersError = class MissingVertexBuffersError extends Error {
	constructor(layouts) {
		super(`Missing vertex buffers for layouts: '${[...layouts].map((layout) => getName(layout) ?? "<unnamed>").join(", ")}'. Please provide it using pipeline.with(layout, buffer).(...)`);
		Object.setPrototypeOf(this, MissingVertexBuffersError.prototype);
	}
};
var IllegalVarAccessError = class IllegalVarAccessError extends Error {
	constructor(msg) {
		super(msg);
		Object.setPrototypeOf(this, IllegalVarAccessError.prototype);
	}
};
var IllegalBufferAccessError = class IllegalBufferAccessError extends Error {
	constructor(msg) {
		super(msg);
		Object.setPrototypeOf(this, IllegalBufferAccessError.prototype);
	}
};
var WgslTypeError = class WgslTypeError extends Error {
	constructor(msg) {
		super(msg);
		Object.setPrototypeOf(this, WgslTypeError.prototype);
	}
};
var SignatureNotSupportedError = class SignatureNotSupportedError extends Error {
	constructor(actual, candidates) {
		super(`Unsupported data types: ${actual.map((a) => a.type).join(", ")}. Supported types are: ${candidates.map((r) => r.type).join(", ")}.`);
		Object.setPrototypeOf(this, SignatureNotSupportedError.prototype);
	}
};
var FiniteMathAssumptionError = class FiniteMathAssumptionError extends Error {
	constructor(value, type) {
		super(`Cannot convert value '${value}' to type ${type.type} because of the Finite Math Assumption (see: https://www.w3.org/TR/WGSL/#finite-math-assumption)`);
		Object.setPrototypeOf(this, FiniteMathAssumptionError.prototype);
	}
};

//#endregion
export { ExecutionError, FiniteMathAssumptionError, IllegalBufferAccessError, IllegalVarAccessError, MissingBindGroupsError, MissingSlotValueError, MissingVertexBuffersError, NotUniformError, ResolutionError, SignatureNotSupportedError, WgslTypeError, invariant };
