// typegpu/execMode.js
import { invariant } from "./errors.js";
import { NormalState } from "./types.js";

//#region src/execMode.ts
/**
* Used to track if the code we're currently
* executing is inside an executing TypeGPU function.
*
* Helpful for providing better error messages.
*/
let insideTgpuFn = false;
function provideInsideTgpuFn(callback) {
	if (insideTgpuFn) return callback();
	try {
		insideTgpuFn = true;
		return callback();
	} finally {
		insideTgpuFn = false;
	}
}
function isInsideTgpuFn() {
	return insideTgpuFn;
}
let resolutionCtx;
function provideCtx(ctx, callback) {
	invariant(resolutionCtx === void 0 || resolutionCtx === ctx, "Cannot nest context providers");
	if (resolutionCtx === ctx) return callback();
	resolutionCtx = ctx;
	try {
		return callback();
	} finally {
		resolutionCtx = void 0;
	}
}
function getResolutionCtx() {
	return resolutionCtx;
}
/**
* Applicable when code is being executed outside of any
* execution-altering APIs.
*/
const topLevelState = new NormalState();
function getExecMode() {
	return resolutionCtx?.mode ?? topLevelState;
}
function inCodegenMode() {
	return resolutionCtx?.mode.type === "codegen";
}

//#endregion
export { getExecMode, getResolutionCtx, inCodegenMode, isInsideTgpuFn, provideCtx, provideInsideTgpuFn, topLevelState };
