// typegpu/index.js
import { MissingBindGroupsError, MissingSlotValueError, MissingVertexBuffersError, NotUniformError, ResolutionError } from "./errors.js";
import { isAccessor, isLazy, isMutableAccessor, isSlot } from "./core/slot/slotTypes.js";
import { shaderGenerator_members_exports } from "./tgsl/shaderGenerator_members.js";
import { isTgpuFn } from "./core/function/tgpuFn.js";
import { readFromArrayBuffer, writeToArrayBuffer } from "./data/dataIO.js";
import { patchArrayBuffer } from "./data/partialIO.js";
import { isUsableAsStorage } from "./extension.js";
import { isUsableAsUniform } from "./core/buffer/bufferUsage.js";
import { isBuffer, isUsableAsVertex } from "./core/buffer/buffer.js";
import { isComparisonSampler, isSampler } from "./core/sampler/sampler.js";
import { isTexture } from "./core/texture/texture.js";
import { isUsableAsRender, isUsableAsSampled } from "./core/texture/usageExtension.js";
import { isVariable } from "./core/variable/tgpuVariable.js";
import { WgslGenerator } from "./tgsl/wgslGenerator.js";
import { isBufferShorthand } from "./core/buffer/bufferShorthand.js";
import { isTgpuComputeFn } from "./core/function/tgpuComputeFn.js";
import { isTgpuFragmentFn } from "./core/function/tgpuFragmentFn.js";
import { isTgpuVertexFn } from "./core/function/tgpuVertexFn.js";
import { tgpu_exports } from "./tgpu.js";
import { data_exports } from "./data/index.js";
import { std_exports } from "./std/index.js";
import { common_exports } from "./common/index.js";

//#region src/index.js
var src_default = tgpu_exports;

//#endregion
export { MissingBindGroupsError, MissingSlotValueError, MissingVertexBuffersError, NotUniformError, ResolutionError, shaderGenerator_members_exports as ShaderGenerator, WgslGenerator, common_exports as common, data_exports as d, src_default as default, isAccessor, isBuffer, isBufferShorthand, isComparisonSampler, isLazy, isMutableAccessor, isSampler, isSlot, isTexture, isTgpuComputeFn, isTgpuFn, isTgpuFragmentFn, isTgpuVertexFn, isUsableAsRender, isUsableAsSampled, isUsableAsStorage, isUsableAsUniform, isUsableAsVertex, isVariable, patchArrayBuffer, readFromArrayBuffer, std_exports as std, tgpu_exports as tgpu, writeToArrayBuffer };
