// typegpu/tgpuUnstable.js
import { __export } from "./_virtual/rolldown_runtime.js";
import { comptime } from "./core/function/comptime.js";
import { constant } from "./core/constant/tgpuConstant.js";
import { fn } from "./core/function/tgpuFn.js";
import { slot } from "./core/slot/slot.js";
import { privateVar, workgroupVar } from "./core/variable/tgpuVariable.js";
import { namespace } from "./core/resolve/namespace.js";
import { computeFn } from "./core/function/tgpuComputeFn.js";
import { vertexLayout } from "./core/vertexLayout/vertexLayout.js";
import { lazy } from "./core/slot/lazy.js";
import { accessor, mutableAccessor } from "./core/slot/accessor.js";
import { fragmentFn } from "./core/function/tgpuFragmentFn.js";
import { vertexFn } from "./core/function/tgpuVertexFn.js";
import { declare } from "./core/declare/tgpuDeclare.js";
import { rawCodeSnippet } from "./core/rawCodeSnippet/tgpuRawCodeSnippet.js";
import { simulate } from "./core/simulate/tgpuSimulate.js";

//#region src/tgpuUnstable.ts
var tgpuUnstable_exports = /* @__PURE__ */ __export({
	accessor: () => _accessor,
	comptime: () => _comptime,
	computeFn: () => _computeFn,
	const: () => _constant,
	declare: () => declare,
	derived: () => _lazy,
	fn: () => _fn,
	fragmentFn: () => _fragmentFn,
	mutableAccessor: () => _mutableAccessor,
	namespace: () => namespace,
	privateVar: () => _privateVar,
	rawCodeSnippet: () => rawCodeSnippet,
	simulate: () => simulate,
	slot: () => _slot,
	vertexFn: () => _vertexFn,
	vertexLayout: () => _vertexLayout,
	workgroupVar: () => _workgroupVar
});
/** @deprecated This feature is now stable, use tgpu.const. */
const _constant = constant;
/** @deprecated This feature is now stable, use tgpu.comptime. */
const _comptime = comptime;
/** @deprecated This feature is now stable, use tgpu.computeFn. */
const _computeFn = computeFn;
/** @deprecated This feature is now stable, use tgpu.fn. */
const _fn = fn;
/** @deprecated This feature is now stable, use tgpu.fragmentFn. */
const _fragmentFn = fragmentFn;
/** @deprecated This feature is now stable, use tgpu.vertexFn. */
const _vertexFn = vertexFn;
/** @deprecated This feature is now stable, use tgpu.accessor. */
const _accessor = accessor;
/** @deprecated This feature is now stable, use tgpu.mutableAccessor. */
const _mutableAccessor = mutableAccessor;
/** @deprecated This feature is now stable, use tgpu.lazy. */
const _lazy = lazy;
/** @deprecated This feature is now stable, use tgpu.slot. */
const _slot = slot;
/** @deprecated This feature is now stable, use tgpu.privateVar. */
const _privateVar = privateVar;
/** @deprecated This feature is now stable, use tgpu.workgroupVar. */
const _workgroupVar = workgroupVar;
/** @deprecated This feature is now stable, use tgpu.vertexLayout. */
const _vertexLayout = vertexLayout;

//#endregion
export { tgpuUnstable_exports };
