// typegpu/tgpu.js
import { __export } from "./_virtual/rolldown_runtime.js";
import { comptime } from "./core/function/comptime.js";
import { constant } from "./core/constant/tgpuConstant.js";
import { fn } from "./core/function/tgpuFn.js";
import { bindGroupLayout } from "./tgpuBindGroupLayout.js";
import { slot } from "./core/slot/slot.js";
import { privateVar, workgroupVar } from "./core/variable/tgpuVariable.js";
import { unroll } from "./core/unroll/tgpuUnroll.js";
import { resolve, resolveWithContext } from "./core/resolve/tgpuResolve.js";
import { computeFn } from "./core/function/tgpuComputeFn.js";
import { vertexLayout } from "./core/vertexLayout/vertexLayout.js";
import { init, initFromDevice } from "./core/root/init.js";
import { lazy } from "./core/slot/lazy.js";
import { accessor, mutableAccessor } from "./core/slot/accessor.js";
import { fragmentFn } from "./core/function/tgpuFragmentFn.js";
import { vertexFn } from "./core/function/tgpuVertexFn.js";
import { tgpuUnstable_exports } from "./tgpuUnstable.js";

//#region src/tgpu.ts
var tgpu_exports = /* @__PURE__ */ __export({
	accessor: () => accessor,
	bindGroupLayout: () => bindGroupLayout,
	comptime: () => comptime,
	computeFn: () => computeFn,
	const: () => constant,
	fn: () => fn,
	fragmentFn: () => fragmentFn,
	init: () => init,
	initFromDevice: () => initFromDevice,
	lazy: () => lazy,
	mutableAccessor: () => mutableAccessor,
	privateVar: () => privateVar,
	resolve: () => resolve,
	resolveWithContext: () => resolveWithContext,
	slot: () => slot,
	unroll: () => unroll,
	vertexFn: () => vertexFn,
	vertexLayout: () => vertexLayout,
	workgroupVar: () => workgroupVar,
	"~unstable": () => tgpuUnstable_exports
});

//#endregion
export { tgpu_exports };
