// typegpu/types.js
import { $cast, $gpuCallable, $ownSnippet, $resolve } from "./shared/symbols.js";
import { isWgslData } from "./data/wgslTypes.js";
import { UnknownData } from "./data/dataTypes.js";
import { isLazy, isProviding, isSlot } from "./core/slot/slotTypes.js";

//#region src/types.ts
var NormalState = class {
	type = "normal";
};
var CodegenState = class {
	type = "codegen";
};
var SimulationState = class {
	type = "simulate";
	buffers;
	vars;
	constructor(buffers, vars) {
		this.buffers = buffers;
		this.vars = vars;
	}
};
function isSelfResolvable(value) {
	return !!value?.[$resolve];
}
function getOwnSnippet(value) {
	return value?.[$ownSnippet];
}
function isGPUCallable(value) {
	return !!value?.[$gpuCallable];
}
function hasCast(value) {
	return !!value?.[$cast];
}
function isKnownAtComptime(snippet) {
	return (typeof snippet.value !== "string" || snippet.dataType === UnknownData) && getOwnSnippet(snippet.value) === void 0;
}
function isWgsl(value) {
	return typeof value === "number" || typeof value === "boolean" || typeof value === "string" || isSelfResolvable(value) || isWgslData(value) || isSlot(value) || isLazy(value) || isProviding(value);
}
function isGPUBuffer(value) {
	return !!value && typeof value === "object" && "getMappedRange" in value && "mapAsync" in value;
}

//#endregion
export { CodegenState, NormalState, SimulationState, getOwnSnippet, hasCast, isGPUBuffer, isGPUCallable, isKnownAtComptime, isSelfResolvable, isWgsl };
