// typegpu/core/variable/tgpuVariable.js
import { $gpuValueOf, $internal, $ownSnippet, $resolve } from "../../shared/symbols.js";
import { getName, setName } from "../../shared/meta.js";
import { snip } from "../../data/snippet.js";
import { IllegalVarAccessError } from "../../errors.js";
import { getExecMode, isInsideTgpuFn } from "../../execMode.js";
import { assertExhaustive } from "../../shared/utilityTypes.js";
import { valueProxyHandler } from "../valueProxyUtils.js";

//#region src/core/variable/tgpuVariable.ts
/**
* Defines a variable scoped to each entry function (private).
*
* @param dataType The schema of the held data's type
* @param initialValue If not provided, the variable will be initialized to the dataType's "zero-value".
*/
function privateVar(dataType, initialValue) {
	return new TgpuVarImpl("private", dataType, initialValue);
}
/**
* Defines a variable scoped to the whole workgroup, shared between entry functions
* of the same invocation.
*
* @param dataType The schema of the held data's type
*/
function workgroupVar(dataType) {
	return new TgpuVarImpl("workgroup", dataType);
}
function isVariable(value) {
	return value instanceof TgpuVarImpl;
}
var TgpuVarImpl = class {
	[$internal] = {};
	resourceType;
	#scope;
	#dataType;
	#initialValue;
	constructor(scope, dataType, initialValue) {
		this.resourceType = "var";
		this.#scope = scope;
		this.#dataType = dataType;
		this.#initialValue = initialValue;
	}
	[$resolve](ctx) {
		const id = ctx.makeUniqueIdentifier(getName(this), "global");
		const init = this.#initialValue ? snip(this.#initialValue, this.#dataType, "constant") : void 0;
		return ctx.gen.declareGlobalVar({
			scope: this.#scope,
			id,
			dataType: this.#dataType,
			init
		});
	}
	$name(label) {
		setName(this, label);
		return this;
	}
	toString() {
		return `var:${getName(this) ?? "<unnamed>"}`;
	}
	get [$gpuValueOf]() {
		const dataType = this.#dataType;
		const origin = this.#scope;
		return new Proxy({
			[$internal]: true,
			get [$ownSnippet]() {
				return snip(this, dataType, origin);
			},
			[$resolve]: (ctx) => ctx.resolve(this),
			toString: () => `var:${getName(this) ?? "<unnamed>"}.$`
		}, valueProxyHandler);
	}
	get $() {
		const mode = getExecMode();
		const insideTgpuFn = isInsideTgpuFn();
		if (mode.type === "normal") throw new IllegalVarAccessError(insideTgpuFn ? `Cannot access variable '${getName(this) ?? "<unnamed>"}'. TypeGPU functions that depends on GPU resources need to be part of a compute dispatch, draw call or simulation` : "TypeGPU variables are inaccessible during normal JS execution. If you wanted to simulate GPU behavior, try `tgpu.simulate()`");
		if (mode.type === "codegen") return this[$gpuValueOf];
		if (mode.type === "simulate") {
			if (!mode.vars[this.#scope].has(this)) mode.vars[this.#scope].set(this, this.#initialValue);
			return mode.vars[this.#scope].get(this);
		}
		return assertExhaustive(mode, "tgpuVariable.ts#TgpuVarImpl/$");
	}
	set $(value) {
		const mode = getExecMode();
		const insideTgpuFn = isInsideTgpuFn();
		if (mode.type === "normal") throw new IllegalVarAccessError(insideTgpuFn ? `Cannot access ${String(this)}. TypeGPU functions that depends on GPU resources need to be part of a compute dispatch, draw call or simulation` : "TypeGPU variables are inaccessible during normal JS execution. If you wanted to simulate GPU behavior, try `tgpu.simulate()`");
		if (mode.type === "codegen") throw new Error("Unreachable tgpuVariable.ts#TgpuVarImpl/$");
		if (mode.type === "simulate") {
			mode.vars[this.#scope].set(this, value);
			return;
		}
		assertExhaustive(mode, "tgpuVariable.ts#TgpuVarImpl/$");
	}
	get value() {
		return this.$;
	}
	set value(v) {
		this.$ = v;
	}
};

//#endregion
export { isVariable, privateVar, workgroupVar };
