// typegpu/core/resolve/externals.js
import { isWgslStruct } from "../../data/wgslTypes.js";
import { getName, hasTinyestMetadata, isNamable, setName } from "../../shared/meta.js";
import { isLooseData } from "../../data/dataTypes.js";
import { isWgsl } from "../../types.js";

//#region src/core/resolve/externals.ts
function isResolvable(value) {
	return isWgsl(value) || isLooseData(value) || hasTinyestMetadata(value);
}
/**
* Merges function externals into one map.
*/
function mergeFunctionExternals(fnExternals) {
	const base = fnExternals.pluginProvided ?? fnExternals.userProvided ?? {};
	const result = Object.defineProperties({}, Object.getOwnPropertyDescriptors(base));
	for (const flatExternal of [fnExternals.args, fnExternals.out].filter((e) => e !== void 0)) for (const [key, value] of Object.entries(flatExternal)) {
		if (key in result && result[key] !== value) throw new Error(`Key '${key}' appears in externals despite already being used for argument/return type. Please rename this external.`);
		result[key] = value;
	}
	return result;
}
function addArgTypesToExternals(implementation, argTypes, core) {
	const argTypeNames = [...implementation.matchAll(/:\s*(?<arg>.*?)\s*[,)]/g)].map((found) => found?.[1]);
	const args = Object.fromEntries(argTypes.flatMap((argType, i) => {
		const argTypeName = argTypeNames?.[i];
		return isWgslStruct(argType) && argTypeName !== void 0 ? [[argTypeName, argType]] : [];
	}));
	core.setExternals("args", args);
}
function addReturnTypeToExternals(implementation, returnType, core) {
	const matched = implementation.match(/->\s(?<output>[\w\d_]+)\s{/);
	const outputName = matched ? matched[1]?.trim() : void 0;
	if (isWgslStruct(returnType) && outputName && !/\s/g.test(outputName)) core.setExternals("out", { [outputName]: returnType });
}
function identifierRegex(name) {
	return new RegExp(`(?<![\\w\\$_.])${name.replaceAll(".", "\\.").replaceAll("$", "\\$")}(?![\\w\\$_])`, "g");
}
/**
* Replaces all occurrences of external names in WGSL code with their resolved values.
* It adds all necessary definitions to the resolution context.
* @param ctx - The resolution context.
* @param externalMap - The external map.
* @param wgsl - The WGSL code.
*
* @returns The WGSL code with all external names replaced with their resolved values.
*/
function replaceExternalsInWgsl(ctx, externalMap, wgsl) {
	return Object.entries(externalMap).reduce((acc, [externalName, external]) => {
		const externalRegex = identifierRegex(externalName);
		if (wgsl && externalName !== "Out" && externalName !== "in" && !externalRegex.test(wgsl)) console.warn(`The external '${externalName}' wasn't used in the resolved template.`);
		if (isResolvable(external)) {
			if (isNamable(external) && getName(external) === void 0) setName(external, externalName.split(".").at(-1));
			return acc.replaceAll(externalRegex, ctx.resolve(external).value);
		}
		if (external !== null && typeof external === "object") {
			const foundProperties = [...wgsl.matchAll(new RegExp(`${externalName.replaceAll(".", "\\.").replaceAll("$", "\\$")}\\.(?<prop>.*?)(?![\\w\\$_])`, "g"))].map((found) => found[1]);
			return [...new Set(foundProperties)].reduce((innerAcc, prop) => prop && prop in external ? replaceExternalsInWgsl(ctx, { [`${externalName}.${prop}`]: external[prop] }, innerAcc) : innerAcc, acc);
		}
		console.warn(`During resolution, the external '${externalName}' has been omitted. Only TGPU resources, 'use gpu' functions, primitives, and plain JS objects can be used as externals.`);
		return acc;
	}, wgsl);
}

//#endregion
export { addArgTypesToExternals, addReturnTypeToExternals, mergeFunctionExternals, replaceExternalsInWgsl };
