// typegpu/core/resolve/tgpuResolve.js
import { $internal, $resolve } from "../../shared/symbols.js";
import { Void } from "../../data/wgslTypes.js";
import { snip } from "../../data/snippet.js";
import { replaceExternalsInWgsl } from "./externals.js";
import { isBindGroupLayout } from "../../tgpuBindGroupLayout.js";
import { resolve } from "../../resolutionCtx.js";
import { isPipeline } from "../pipeline/typeGuards.js";
import { namespace } from "./namespace.js";

//#region src/core/resolve/tgpuResolve.ts
function resolveWithContext(arg0, options) {
	if (Array.isArray(arg0)) return resolveFromArray(arg0, options);
	return resolveFromTemplate(arg0);
}
function resolve$1(arg, options) {
	if (Array.isArray(arg)) return resolveWithContext(arg, options).code;
	return resolveWithContext(arg).code;
}
function resolveFromTemplate(options) {
	const { template, externals, unstable_shaderGenerator: shaderGenerator, names = "strict", config, enableExtensions } = options;
	if (!template) console.warn("Calling resolve with an empty template is deprecated and will soon return an empty string. Consider using the 'tgpu.resolve(resolvableArray, options)' API instead.");
	return resolve({
		[$internal]: true,
		[$resolve](ctx) {
			return snip(replaceExternalsInWgsl(ctx, externals, template ?? ""), Void, "runtime");
		},
		toString: () => "<root>"
	}, {
		namespace: typeof names === "string" ? namespace({ names }) : names,
		enableExtensions,
		shaderGenerator,
		config,
		root: tryFindRoot(Object.values(externals))
	});
}
function resolveFromArray(items, options) {
	const { unstable_shaderGenerator: shaderGenerator, names = "strict", config, enableExtensions } = options ?? {};
	return resolve({
		[$internal]: true,
		[$resolve](ctx) {
			for (const item of items) if (isBindGroupLayout(item)) for (const binding of Object.values(item[$internal].bound)) ctx.resolve(binding);
			else ctx.resolve(item);
			return snip("", Void, "runtime");
		},
		toString: () => "<root>"
	}, {
		namespace: typeof names === "string" ? namespace({ names }) : names,
		enableExtensions,
		shaderGenerator,
		config,
		root: tryFindRoot(items)
	});
}
/**
* Attempts to locate a pipeline in a list of items and returns the root.
* Does not check recursively.
* Throws an error if multiple pipelines are found.
*/
function tryFindRoot(items) {
	const pipelines = items.filter(isPipeline);
	if (pipelines.length > 1) throw new Error(`Found ${pipelines.length} pipelines but can only resolve one at a time.`);
	return pipelines[0]?.[$internal].root;
}

//#endregion
export { resolve$1 as resolve, resolveWithContext };
