// typegpu/core/vertexLayout/connectAttributesToShader.js
import { getCustomLocation, isData } from "../../data/dataTypes.js";
import { isBuiltin } from "../../data/attributes.js";

//#region src/core/vertexLayout/connectAttributesToShader.ts
function isAttribute(value) {
	return typeof value?.format === "string";
}
function connectAttributesToShader(shaderInputLayout, attributes) {
	const usedVertexLayouts = [];
	if (isData(shaderInputLayout)) {
		if (!isAttribute(attributes)) throw new Error("Shader expected a single attribute, not a record of attributes to be passed in.");
		usedVertexLayouts.push(attributes._layout);
		return {
			usedVertexLayouts,
			bufferDefinitions: [{
				arrayStride: attributes._layout.stride,
				stepMode: attributes._layout.stepMode,
				attributes: [{
					format: attributes.format,
					offset: attributes.offset,
					shaderLocation: getCustomLocation(shaderInputLayout) ?? 0
				}]
			}]
		};
	}
	const bufferDefinitions = [];
	const layoutToAttribListMap = /* @__PURE__ */ new WeakMap();
	let nextShaderLocation = 0;
	for (const [key, member] of Object.entries(shaderInputLayout)) {
		if (isBuiltin(member)) continue;
		const matchingAttribute = attributes[key];
		if (!matchingAttribute) throw new Error(`An attribute by the name of '${key}' was not provided to the shader.`);
		const layout = matchingAttribute._layout;
		let attribList = layoutToAttribListMap.get(layout);
		if (!attribList) {
			usedVertexLayouts.push(layout);
			attribList = [];
			bufferDefinitions.push({
				arrayStride: layout.stride,
				stepMode: layout.stepMode,
				attributes: attribList
			});
			layoutToAttribListMap.set(layout, attribList);
		}
		nextShaderLocation = getCustomLocation(member) ?? nextShaderLocation;
		attribList.push({
			format: matchingAttribute.format,
			offset: matchingAttribute.offset,
			shaderLocation: nextShaderLocation++
		});
	}
	return {
		usedVertexLayouts,
		bufferDefinitions
	};
}

//#endregion
export { connectAttributesToShader, isAttribute };
