// typegpu/core/vertexLayout/vertexLayout.js
import { $internal } from "../../shared/symbols.js";
import { isDecorated, isWgslStruct } from "../../data/wgslTypes.js";
import { setName } from "../../shared/meta.js";
import { kindToDefaultFormatMap, vertexFormats } from "../../shared/vertexFormat.js";
import { getCustomLocation, isLooseDecorated, isUnstruct } from "../../data/dataTypes.js";
import { alignmentOf, customAlignmentOf } from "../../data/alignmentOf.js";
import { roundUp } from "../../mathUtils.js";
import { sizeOf } from "../../data/sizeOf.js";

//#region src/core/vertexLayout/vertexLayout.ts
function vertexLayout(schemaForCount, stepMode = "vertex") {
	return new TgpuVertexLayoutImpl(schemaForCount, stepMode);
}
function isVertexLayout(value) {
	return value?.resourceType === "vertex-layout";
}
const defaultAttribEntry = Symbol("defaultAttribEntry");
function dataToContainedAttribs(layout, data, offset, customLocationMap, key) {
	if (isDecorated(data) || isLooseDecorated(data)) {
		const customLocation = getCustomLocation(data);
		if (customLocation !== void 0) customLocationMap[key ?? defaultAttribEntry] = customLocation;
		return dataToContainedAttribs(layout, data.inner, roundUp(offset, customAlignmentOf(data)), customLocationMap);
	}
	if (isWgslStruct(data)) {
		let memberOffset = offset;
		const propTypes = data.propTypes;
		return Object.fromEntries(Object.entries(propTypes).map(([key$1, value]) => {
			memberOffset = roundUp(memberOffset, alignmentOf(value));
			const attrib = [key$1, dataToContainedAttribs(layout, value, memberOffset, customLocationMap, key$1)];
			memberOffset += sizeOf(value);
			return attrib;
		}));
	}
	if (isUnstruct(data)) {
		let memberOffset = offset;
		const propTypes = data.propTypes;
		return Object.fromEntries(Object.entries(propTypes).map(([key$1, value]) => {
			memberOffset = roundUp(memberOffset, customAlignmentOf(value));
			const attrib = [key$1, dataToContainedAttribs(layout, value, memberOffset, customLocationMap, key$1)];
			memberOffset += sizeOf(value);
			return attrib;
		}));
	}
	if ("type" in data && typeof data.type === "string") {
		if (vertexFormats.includes(data.type)) return {
			_layout: layout,
			format: data.type,
			offset
		};
		const format = kindToDefaultFormatMap[data.type];
		if (format) return {
			_layout: layout,
			format,
			offset
		};
	}
	throw new Error(`Unsupported data used in vertex layout: ${String(data)}`);
}
var TgpuVertexLayoutImpl = class {
	[$internal] = true;
	resourceType = "vertex-layout";
	stride;
	attrib;
	schemaForCount;
	stepMode;
	#customLocationMap = {};
	constructor(schemaForCount, stepMode) {
		this.schemaForCount = schemaForCount;
		this.stepMode = stepMode;
		const arraySchema = schemaForCount(0);
		this.stride = roundUp(sizeOf(arraySchema.elementType), alignmentOf(arraySchema));
		this.attrib = dataToContainedAttribs(this, arraySchema.elementType, 0, this.#customLocationMap);
	}
	get vertexLayout() {
		if (this.#customLocationMap[defaultAttribEntry] !== void 0) {
			if (typeof this.attrib.format !== "string" || typeof this.attrib.offset !== "number") throw new Error("Single attribute vertex layouts must have a format and offset.");
			return {
				arrayStride: this.stride,
				stepMode: this.stepMode,
				attributes: [{
					format: this.attrib.format,
					offset: this.attrib.offset,
					shaderLocation: this.#customLocationMap[defaultAttribEntry]
				}]
			};
		}
		if (!Object.keys(this.attrib).every((key) => this.#customLocationMap[key] !== void 0)) throw new Error("All attributes must have custom locations in order to unwrap a vertex layout.");
		return {
			arrayStride: this.stride,
			stepMode: this.stepMode,
			attributes: Object.entries(this.attrib).map(([key, attrib]) => ({
				format: attrib.format,
				offset: attrib.offset,
				shaderLocation: this.#customLocationMap[key]
			}))
		};
	}
	$name(label) {
		setName(this, label);
		return this;
	}
};

//#endregion
export { isVertexLayout, vertexLayout };
