// typegpu/core/pipeline/computePipeline.js
import { $getNameForward, $internal, $resolve } from "../../shared/symbols.js";
import { Void } from "../../data/wgslTypes.js";
import { PERF, getName, setName } from "../../shared/meta.js";
import { snip } from "../../data/snippet.js";
import { isGPUBuffer } from "../../types.js";
import { isBindGroup } from "../../tgpuBindGroupLayout.js";
import { resolve } from "../../resolutionCtx.js";
import { isGPUCommandEncoder, isGPUComputePassEncoder } from "./typeGuards.js";
import { namespace } from "../resolve/namespace.js";
import { applyBindGroups } from "./applyPipelineState.js";
import { logDataFromGPU } from "../../tgsl/consoleLog/deserializers.js";
import { wgslEnableExtensionToFeatureName, wgslEnableExtensions } from "../../wgslExtensions.js";
import { resolveIndirectOffset } from "./pipelineUtils.js";
import { createWithPerformanceCallback, createWithTimestampWrites, setupTimestampWrites, triggerPerformanceCallback } from "./timeable.js";

//#region src/core/pipeline/computePipeline.ts
function INTERNAL_createComputePipeline(branch, slotBindings, descriptor) {
	return new TgpuComputePipelineImpl(new ComputePipelineCore(branch, slotBindings, descriptor), {});
}
const _lastAppliedCompute = /* @__PURE__ */ new WeakMap();
var TgpuComputePipelineImpl = class TgpuComputePipelineImpl {
	[$internal];
	resourceType = "compute-pipeline";
	[$getNameForward];
	#core;
	#priors;
	constructor(core, priors) {
		this.#core = core;
		this.#priors = priors;
		this[$internal] = {
			get rawPipeline() {
				return core.unwrap().pipeline;
			},
			get priors() {
				return priors;
			},
			get root() {
				return core.root;
			}
		};
		this[$getNameForward] = core;
	}
	[$resolve](ctx) {
		return ctx.resolve(this.#core);
	}
	toString() {
		return `computePipeline:${getName(this) ?? "<unnamed>"}`;
	}
	get rawPipeline() {
		return this.#core.unwrap().pipeline;
	}
	with(first, bindGroup) {
		if (isGPUComputePassEncoder(first)) return new TgpuComputePipelineImpl(this.#core, {
			...this.#priors,
			externalPass: first,
			externalEncoder: void 0
		});
		if (isGPUCommandEncoder(first)) return new TgpuComputePipelineImpl(this.#core, {
			...this.#priors,
			externalEncoder: first,
			externalPass: void 0
		});
		if (isBindGroup(first)) return new TgpuComputePipelineImpl(this.#core, {
			...this.#priors,
			bindGroupLayoutMap: new Map([...this.#priors.bindGroupLayoutMap ?? [], [first.layout, first]])
		});
		return new TgpuComputePipelineImpl(this.#core, {
			...this.#priors,
			bindGroupLayoutMap: new Map([...this.#priors.bindGroupLayoutMap ?? [], [first, bindGroup]])
		});
	}
	withPerformanceCallback(callback) {
		if (this.#priors.timestampWrites) return new TgpuComputePipelineImpl(this.#core, {
			...this.#priors,
			performanceCallback: callback
		});
		const querySet = this.#core.performanceCallbackQuerySet;
		if (!querySet) {
			console.warn("Performance callback cannot be used because the timestamp-query feature is not enabled on the root.");
			return this;
		}
		const newPriors = createWithPerformanceCallback(this.#priors, callback, querySet);
		return new TgpuComputePipelineImpl(this.#core, newPriors);
	}
	withTimestampWrites(options) {
		const newPriors = createWithTimestampWrites(this.#priors, options, this.#core.root);
		return new TgpuComputePipelineImpl(this.#core, newPriors);
	}
	dispatchWorkgroups(x, y, z) {
		this._executeComputePass((pass) => pass.dispatchWorkgroups(x, y, z));
	}
	dispatchWorkgroupsIndirect(indirectBuffer, start) {
		const DISPATCH_SIZE = 12;
		const rawBuffer = isGPUBuffer(indirectBuffer) ? indirectBuffer : indirectBuffer.buffer;
		const offset = resolveIndirectOffset(indirectBuffer, start, DISPATCH_SIZE, "dispatchWorkgroupsIndirect");
		this._executeComputePass((pass) => pass.dispatchWorkgroupsIndirect(rawBuffer, offset));
	}
	_applyComputeState(pass) {
		const memo = this.#core.unwrap();
		const { root } = this.#core;
		pass.setPipeline(memo.pipeline);
		applyBindGroups(pass, root, memo.usedBindGroupLayouts, memo.catchall, (layout) => this.#priors.bindGroupLayoutMap?.get(layout));
	}
	_executeComputePass(dispatch) {
		const { root } = this.#core;
		if (this.#priors.externalPass) {
			if (_lastAppliedCompute.get(this.#priors.externalPass) !== this) {
				this._applyComputeState(this.#priors.externalPass);
				_lastAppliedCompute.set(this.#priors.externalPass, this);
			}
			dispatch(this.#priors.externalPass);
			return;
		}
		if (this.#priors.externalEncoder) {
			const passDescriptor$1 = {
				label: getName(this.#core) ?? "<unnamed>",
				...setupTimestampWrites(this.#priors, root)
			};
			const pass$1 = this.#priors.externalEncoder.beginComputePass(passDescriptor$1);
			this._applyComputeState(pass$1);
			dispatch(pass$1);
			pass$1.end();
			return;
		}
		const memo = this.#core.unwrap();
		const passDescriptor = {
			label: getName(this.#core) ?? "<unnamed>",
			...setupTimestampWrites(this.#priors, root)
		};
		const commandEncoder = root.device.createCommandEncoder();
		const pass = commandEncoder.beginComputePass(passDescriptor);
		this._applyComputeState(pass);
		dispatch(pass);
		pass.end();
		root.device.queue.submit([commandEncoder.finish()]);
		if (memo.logResources) logDataFromGPU(memo.logResources);
		if (this.#priors.performanceCallback) triggerPerformanceCallback({
			root,
			priors: this.#priors
		});
	}
	$name(label) {
		setName(this, label);
		return this;
	}
};
var ComputePipelineCore = class {
	[$internal] = true;
	root;
	_memo;
	#slotBindings;
	#descriptor;
	#performanceCallbackQuerySet;
	constructor(root, slotBindings, descriptor) {
		this.root = root;
		this.#slotBindings = slotBindings;
		this.#descriptor = descriptor;
	}
	[$resolve](ctx) {
		return ctx.withSlots(this.#slotBindings, () => {
			ctx.resolve(this.#descriptor.compute);
			return snip("", Void, "runtime");
		});
	}
	toString() {
		return "computePipelineCore";
	}
	get performanceCallbackQuerySet() {
		if (!this.root.enabledFeatures.has("timestamp-query")) return;
		return this.#performanceCallbackQuerySet ??= this.root.createQuerySet("timestamp", 2);
	}
	unwrap() {
		if (this._memo === void 0) {
			const device = this.root.device;
			const enableExtensions = wgslEnableExtensions.filter((extension) => this.root.enabledFeatures.has(wgslEnableExtensionToFeatureName[extension]));
			let resolutionResult;
			let resolveMeasure;
			const ns = namespace({ names: this.root.nameRegistrySetting });
			if (PERF?.enabled) {
				const resolveStart = performance.mark("typegpu:resolution:start");
				resolutionResult = resolve(this, {
					namespace: ns,
					enableExtensions,
					shaderGenerator: this.root.shaderGenerator,
					root: this.root
				});
				resolveMeasure = performance.measure("typegpu:resolution", { start: resolveStart.name });
			} else resolutionResult = resolve(this, {
				namespace: ns,
				enableExtensions,
				shaderGenerator: this.root.shaderGenerator,
				root: this.root
			});
			const { code, usedBindGroupLayouts, catchall, logResources } = resolutionResult;
			if (catchall !== void 0) usedBindGroupLayouts[catchall[0]]?.$name(`${getName(this) ?? "<unnamed>"} - Automatic Bind Group & Layout`);
			const module = device.createShaderModule({
				label: `${getName(this) ?? "<unnamed>"} - Shader`,
				code
			});
			this._memo = {
				pipeline: device.createComputePipeline({
					label: getName(this) ?? "<unnamed>",
					layout: device.createPipelineLayout({
						label: `${getName(this) ?? "<unnamed>"} - Pipeline Layout`,
						bindGroupLayouts: usedBindGroupLayouts.map((l) => this.root.unwrap(l))
					}),
					compute: { module }
				}),
				usedBindGroupLayouts,
				catchall,
				logResources
			};
			if (PERF?.enabled) (async () => {
				const start = performance.mark("typegpu:compile-start");
				await device.queue.onSubmittedWorkDone();
				const compileMeasure = performance.measure("typegpu:compiled", { start: start.name });
				PERF?.record("resolution", {
					resolveDuration: resolveMeasure?.duration,
					compileDuration: compileMeasure.duration,
					wgslSize: code.length
				});
			})();
		}
		return this._memo;
	}
};

//#endregion
export { INTERNAL_createComputePipeline };
