// typegpu/core/pipeline/pipelineUtils.js
import { isGPUBuffer } from "../../types.js";
import { sizeOf } from "../../data/sizeOf.js";
import { memoryLayoutOf } from "../../data/offsetUtils.js";

//#region src/core/pipeline/pipelineUtils.ts
const IndirectOperationToRequiredData = {
	dispatchWorkgroupsIndirect: "3 x u32",
	drawIndirect: "4 x u32",
	drawIndexedIndirect: "3 x u32, i32, u32"
};
function validateIndirectBufferSize(bufferSize, offset, requiredBytes, operation) {
	if (offset + requiredBytes > bufferSize) throw new Error(`Buffer too small for ${operation}. Required: ${requiredBytes} bytes at offset ${offset}, but buffer is only ${bufferSize} bytes.`);
	if (offset % 4 !== 0) throw new Error(`Indirect buffer offset must be a multiple of 4. Got: ${offset}`);
}
function resolveIndirectOffset(indirectBuffer, start, requiredSize, operation) {
	if (isGPUBuffer(indirectBuffer)) {
		const offset$1 = typeof start === "number" ? start : start?.offset ?? 0;
		validateIndirectBufferSize(indirectBuffer.size, offset$1, requiredSize, operation);
		return offset$1;
	}
	const { offset, contiguous } = start ? typeof start === "number" ? {
		offset: start,
		contiguous: requiredSize
	} : start : memoryLayoutOf(indirectBuffer.dataType);
	validateIndirectBufferSize(sizeOf(indirectBuffer.dataType), offset, requiredSize, operation);
	if (contiguous < requiredSize) console.warn(`${operation}: Starting at offset ${offset}, only ${contiguous} contiguous bytes are available before padding. '${operation}' requires ${requiredSize} bytes (${IndirectOperationToRequiredData[operation]}). Reading across padding may result in undefined behavior.`);
	return offset;
}

//#endregion
export { resolveIndirectOffset };
