// typegpu/core/pipeline/timeable.js
import { $internal } from "../../shared/symbols.js";
import { isQuerySet } from "../querySet/querySet.js";

//#region src/core/pipeline/timeable.ts
function createWithPerformanceCallback(currentPriors, callback, querySet) {
	if (!currentPriors.timestampWrites) return {
		...currentPriors,
		performanceCallback: callback,
		timestampWrites: {
			querySet,
			beginningOfPassWriteIndex: 0,
			endOfPassWriteIndex: 1
		}
	};
	return {
		...currentPriors,
		performanceCallback: callback
	};
}
function createWithTimestampWrites(currentPriors, options, root) {
	if (!root.enabledFeatures.has("timestamp-query")) throw new Error("Timestamp writes require the \"timestamp-query\" feature to be enabled on GPU device.");
	const timestampWrites = { querySet: options.querySet };
	if (options.beginningOfPassWriteIndex !== void 0) timestampWrites.beginningOfPassWriteIndex = options.beginningOfPassWriteIndex;
	if (options.endOfPassWriteIndex !== void 0) timestampWrites.endOfPassWriteIndex = options.endOfPassWriteIndex;
	return {
		...currentPriors,
		timestampWrites
	};
}
function setupTimestampWrites(priors, root) {
	if (!priors.timestampWrites) return {};
	const { querySet, beginningOfPassWriteIndex, endOfPassWriteIndex } = priors.timestampWrites;
	const timestampWrites = { querySet: isQuerySet(querySet) ? root.unwrap(querySet) : querySet };
	if (beginningOfPassWriteIndex !== void 0) timestampWrites.beginningOfPassWriteIndex = beginningOfPassWriteIndex;
	if (endOfPassWriteIndex !== void 0) timestampWrites.endOfPassWriteIndex = endOfPassWriteIndex;
	return { timestampWrites };
}
function triggerPerformanceCallback({ root, priors }) {
	const querySet = priors.timestampWrites?.querySet;
	const callback = priors.performanceCallback;
	if (!querySet) throw new Error("Cannot dispatch workgroups with performance callback without a query set.");
	if (!isQuerySet(querySet)) throw new Error("Performance callback with raw GPUQuerySet is not supported. Use TgpuQuerySet instead.");
	const commandEncoder = root.device.createCommandEncoder();
	commandEncoder.resolveQuerySet(root.unwrap(querySet), 0, querySet.count, querySet[$internal].resolveBuffer, 0);
	root.device.queue.submit([commandEncoder.finish()]);
	root.device.queue.onSubmittedWorkDone().then(async () => {
		if (!querySet.available) return;
		const result = await querySet.read();
		const start = result[priors.timestampWrites?.beginningOfPassWriteIndex ?? 0];
		const end = result[priors.timestampWrites?.endOfPassWriteIndex ?? 1];
		if (start === void 0 || end === void 0) throw new Error("QuerySet did not return valid timestamps.");
		await callback(start, end);
	});
}

//#endregion
export { createWithPerformanceCallback, createWithTimestampWrites, setupTimestampWrites, triggerPerformanceCallback };
