// typegpu/core/pipeline/typeGuards.js
import { $internal } from "../../shared/symbols.js";

//#region src/core/pipeline/typeGuards.ts
function isComputePipeline(value) {
	const maybe = value;
	return maybe?.resourceType === "compute-pipeline" && !!maybe[$internal];
}
function isRenderPipeline(value) {
	const maybe = value;
	return maybe?.resourceType === "render-pipeline" && !!maybe[$internal];
}
function isPipeline(value) {
	return isRenderPipeline(value) || isComputePipeline(value);
}
function isGPUCommandEncoder(value) {
	return !!value && typeof value === "object" && "beginRenderPass" in value && "beginComputePass" in value;
}
function isGPUComputePassEncoder(value) {
	return !!value && typeof value === "object" && "dispatchWorkgroups" in value && !("beginRenderPass" in value);
}
function isGPURenderPassEncoder(value) {
	return !!value && typeof value === "object" && "executeBundles" in value && "draw" in value;
}
function isGPURenderBundleEncoder(value) {
	return !!value && typeof value === "object" && "draw" in value && "finish" in value && !("executeBundles" in value) && !("beginRenderPass" in value) && !("dispatchWorkgroups" in value);
}

//#endregion
export { isComputePipeline, isGPUCommandEncoder, isGPUComputePassEncoder, isGPURenderBundleEncoder, isGPURenderPassEncoder, isPipeline, isRenderPipeline };
