// typegpu/core/rawCodeSnippet/tgpuRawCodeSnippet.js
import { $gpuValueOf, $internal, $ownSnippet, $resolve } from "../../shared/symbols.js";
import { snip } from "../../data/snippet.js";
import { inCodegenMode } from "../../execMode.js";
import { valueProxyHandler } from "../valueProxyUtils.js";
import { replaceExternalsInWgsl } from "../resolve/externals.js";

//#region src/core/rawCodeSnippet/tgpuRawCodeSnippet.ts
/**
* An advanced API that creates a typed shader expression which
* can be injected into the final shader bundle upon use.
*
* @param expression The code snippet that will be injected in place of `foo.$`
* @param type The type of the expression
* @param [origin='runtime'] Where the value originates from.
*
* **-- Which origin to choose?**
*
* Usually 'runtime' (the default) is a safe bet, but if you're sure that the expression or
* computation is constant (either a reference to a constant, a numeric literal,
* or an operation on constants), then pass 'constant' as it might lead to better
* optimizations.
*
* If what the expression is a direct reference to an existing value (e.g. a uniform, a
* storage binding, ...), then choose from 'uniform', 'mutable', 'readonly', 'workgroup',
* 'private' or 'handle' depending on the address space of the referred value.
*
* @example
* ```ts
* // An identifier that we know will be in the
* // final shader bundle, but we cannot
* // refer to it in any other way.
* const existingGlobal = tgpu['~unstable']
*   .rawCodeSnippet('EXISTING_GLOBAL', d.f32, 'constant');
*
* const foo = () => {
*   'use gpu';
*   return existingGlobal.$ * 2;
* };
*
* const wgsl = tgpu.resolve([foo]);
* // fn foo() -> f32 {
* //   return EXISTING_GLOBAL * 2;
* // }
* ```
*/
function rawCodeSnippet(expression, type, origin = "runtime") {
	return new TgpuRawCodeSnippetImpl(expression, type, origin);
}
var TgpuRawCodeSnippetImpl = class {
	[$internal];
	dataType;
	origin;
	#expression;
	#externals;
	constructor(expression, type, origin) {
		this[$internal] = true;
		this.dataType = type;
		this.origin = origin;
		this.#expression = expression;
	}
	$uses(dependencyMap) {
		if (this.#externals !== void 0) throw new Error("Cannot call '$uses' multiple times. If you wish to override dependencies, use slots or accessors instead.");
		this.#externals = dependencyMap;
		return this;
	}
	[$resolve](ctx) {
		return snip(replaceExternalsInWgsl(ctx, this.#externals ?? {}, this.#expression), this.dataType, this.origin);
	}
	toString() {
		return `raw(${String(this.dataType)}): "${this.#expression}"`;
	}
	get [$gpuValueOf]() {
		const dataType = this.dataType;
		const origin = this.origin;
		return new Proxy({
			[$internal]: true,
			get [$ownSnippet]() {
				return snip(this, dataType, origin);
			},
			[$resolve]: (ctx) => ctx.resolve(this),
			toString: () => `raw(${String(this.dataType)}): "${this.#expression}".$`
		}, valueProxyHandler);
	}
	get $() {
		if (!inCodegenMode()) throw new Error("Raw code snippets can only be used on the GPU.");
		return this[$gpuValueOf];
	}
	get value() {
		return this.$;
	}
};

//#endregion
export { rawCodeSnippet };
