// typegpu/core/texture/texture.js
import { $gpuValueOf, $internal, $ownSnippet, $resolve } from "../../shared/symbols.js";
import { getName, setName } from "../../shared/meta.js";
import { snip } from "../../data/snippet.js";
import { inCodegenMode } from "../../execMode.js";
import { valueProxyHandler } from "../valueProxyUtils.js";
import { isWgslStorageTexture, textureDescriptorToSchema } from "../../data/texture.js";
import { getTextureFormatInfo } from "./textureFormats.js";
import { generateTextureMipmaps, getImageSourceDimensions, resampleImage } from "./textureUtils.js";

//#region src/core/texture/texture.ts
function getDescriptorForProps(props) {
	return {
		dimension: props.dimension ?? "2d",
		sampleType: getTextureFormatInfo(props.format).channelType,
		multisampled: !((props.sampleCount ?? 1) === 1)
	};
}
function INTERNAL_createTexture(props, branch) {
	return new TgpuTextureImpl(props, branch);
}
function isTexture(value) {
	return value?.resourceType === "texture" && !!value[$internal];
}
function isTextureView(value) {
	return value?.resourceType === "texture-view" && !!value[$internal];
}
var TgpuTextureImpl = class {
	[$internal];
	resourceType = "texture";
	props;
	usableAsSampled = false;
	usableAsStorage = false;
	usableAsRender = false;
	#formatInfo;
	#destroyed = false;
	#flags = GPUTextureUsage.COPY_DST | GPUTextureUsage.COPY_SRC;
	#texture = null;
	#branch;
	constructor(props, branch) {
		this.props = props;
		const format = props.format;
		this.#branch = branch;
		this.#formatInfo = getTextureFormatInfo(format);
		this[$internal] = { unwrap: () => {
			if (this.#destroyed) throw new Error("This texture has been destroyed");
			if (!this.#texture) this.#texture = branch.device.createTexture({
				label: getName(this) ?? "<unnamed>",
				format: props.format,
				size: props.size,
				usage: this.#flags,
				dimension: props.dimension ?? "2d",
				viewFormats: props.viewFormats ?? [],
				mipLevelCount: props.mipLevelCount ?? 1,
				sampleCount: props.sampleCount ?? 1
			});
			return this.#texture;
		} };
	}
	$name(label) {
		setName(this, label);
		return this;
	}
	$usage(...usages) {
		const hasStorage = usages.includes("storage");
		const hasSampled = usages.includes("sampled");
		const hasRender = usages.includes("render");
		this.#flags |= hasSampled ? GPUTextureUsage.TEXTURE_BINDING : 0;
		this.#flags |= hasStorage ? GPUTextureUsage.STORAGE_BINDING : 0;
		this.#flags |= hasRender ? GPUTextureUsage.RENDER_ATTACHMENT : 0;
		this.usableAsStorage ||= hasStorage;
		this.usableAsSampled ||= hasSampled;
		this.usableAsRender ||= hasRender;
		return this;
	}
	createView(schema, viewDescriptor) {
		if (schema === "render") return new TgpuTextureRenderViewImpl(this, viewDescriptor);
		return new TgpuFixedTextureViewImpl(schema ?? textureDescriptorToSchema(getDescriptorForProps(this.props)), this, viewDescriptor);
	}
	#clearMipLevel(mip = 0) {
		const scale = 2 ** mip;
		const [width, height, depth] = [
			Math.max(1, Math.floor((this.props.size[0] ?? 1) / scale)),
			Math.max(1, Math.floor((this.props.size[1] ?? 1) / scale)),
			Math.max(1, Math.floor((this.props.size[2] ?? 1) / scale))
		];
		const texelSize = this.#formatInfo.texelSize;
		if (texelSize === "non-copyable") throw new Error(`Cannot clear texture with format '${this.props.format}': this format does not support copy operations.`);
		this.#branch.device.queue.writeTexture({
			texture: this[$internal].unwrap(),
			mipLevel: mip
		}, new Uint8Array(width * height * depth * texelSize), {
			bytesPerRow: texelSize * width,
			rowsPerImage: height
		}, [
			width,
			height,
			depth
		]);
	}
	clear(mipLevel = "all") {
		if (mipLevel === "all") {
			const mipLevels = this.props.mipLevelCount ?? 1;
			for (let i = 0; i < mipLevels; i++) this.#clearMipLevel(i);
		} else this.#clearMipLevel(mipLevel);
	}
	generateMipmaps(baseMipLevel = 0, mipLevels) {
		if (!this.usableAsRender) throw new Error("generateMipmaps called without specifying 'render' usage. Add it via the $usage('render') method.");
		const actualMipLevels = mipLevels ?? (this.props.mipLevelCount ?? 1) - baseMipLevel;
		if (actualMipLevels <= 1) {
			console.warn(`generateMipmaps is a no-op: would generate ${actualMipLevels} mip levels (base: ${baseMipLevel}, total: ${this.props.mipLevelCount ?? 1})`);
			return;
		}
		if (baseMipLevel >= (this.props.mipLevelCount ?? 1)) throw new Error(`Base mip level ${baseMipLevel} is out of range. Texture has ${this.props.mipLevelCount ?? 1} mip levels.`);
		generateTextureMipmaps(this.#branch.device, this[$internal].unwrap(), baseMipLevel, actualMipLevels);
	}
	write(source, mipLevel = 0) {
		if (source instanceof ArrayBuffer || ArrayBuffer.isView(source)) {
			this.#writeBufferData(source, mipLevel);
			return;
		}
		const dimension = this.props.dimension ?? "2d";
		if (!Array.isArray(source)) {
			this.#writeSingleLayer(source, dimension === "3d" ? 0 : void 0);
			return;
		}
		const layerCount = this.props.size[2] ?? 1;
		if (source.length > layerCount) console.warn(`Too many image sources provided. Expected ${layerCount} layers, got ${source.length}. Extra sources will be ignored.`);
		for (let layer = 0; layer < Math.min(source.length, layerCount); layer++) {
			const bitmap = source[layer];
			if (bitmap) this.#writeSingleLayer(bitmap, layer);
		}
	}
	#writeBufferData(source, mipLevel) {
		const mipWidth = Math.max(1, this.props.size[0] >> mipLevel);
		const mipHeight = Math.max(1, (this.props.size[1] ?? 1) >> mipLevel);
		const mipDepth = Math.max(1, (this.props.size[2] ?? 1) >> mipLevel);
		const texelSize = this.#formatInfo.texelSize;
		if (texelSize === "non-copyable") throw new Error(`Cannot write to texture with format '${this.props.format}': this format does not support copy operations.`);
		const expectedSize = mipWidth * mipHeight * mipDepth * texelSize;
		const actualSize = source.byteLength ?? source.byteLength;
		if (actualSize !== expectedSize) throw new Error(`Buffer size mismatch. Expected ${expectedSize} bytes for mip level ${mipLevel}, got ${actualSize} bytes.`);
		this.#branch.device.queue.writeTexture({
			texture: this[$internal].unwrap(),
			mipLevel
		}, "buffer" in source ? source.buffer : source, {
			bytesPerRow: texelSize * mipWidth,
			rowsPerImage: mipHeight
		}, [
			mipWidth,
			mipHeight,
			mipDepth
		]);
	}
	#writeSingleLayer(source, layer) {
		const targetWidth = this.props.size[0];
		const targetHeight = this.props.size[1] ?? 1;
		const { width: sourceWidth, height: sourceHeight } = getImageSourceDimensions(source);
		if (sourceWidth !== targetWidth || sourceHeight !== targetHeight) {
			resampleImage(this.#branch.device, this[$internal].unwrap(), source, layer);
			return;
		}
		this.#branch.device.queue.copyExternalImageToTexture({ source }, {
			texture: this[$internal].unwrap(),
			...layer !== void 0 && { origin: {
				x: 0,
				y: 0,
				z: layer
			} }
		}, layer !== void 0 ? [
			targetWidth,
			targetHeight,
			1
		] : this.props.size);
	}
	copyFrom(source) {
		if (source.props.format !== this.props.format) throw new Error(`Texture format mismatch. Source texture has format ${source.props.format}, target texture has format ${this.props.format}`);
		if (source.props.size[0] !== this.props.size[0] || (source.props.size[1] ?? 1) !== (this.props.size[1] ?? 1) || (source.props.size[2] ?? 1) !== (this.props.size[2] ?? 1)) throw new Error(`Texture size mismatch. Source texture has size ${source.props.size.join("x")}, target texture has size ${this.props.size.join("x")}`);
		const commandEncoder = this.#branch.device.createCommandEncoder();
		commandEncoder.copyTextureToTexture({ texture: source[$internal].unwrap() }, { texture: this[$internal].unwrap() }, source.props.size);
		this.#branch.device.queue.submit([commandEncoder.finish()]);
	}
	toString() {
		return `${this.resourceType}:${getName(this) ?? "<unnamed>"}`;
	}
	get destroyed() {
		return this.#destroyed;
	}
	destroy() {
		if (this.#destroyed) return;
		this.#destroyed = true;
		this.#texture?.destroy();
	}
};
var TgpuFixedTextureViewImpl = class {
	[$internal];
	resourceType = "texture-view";
	schema;
	#baseTexture;
	#view;
	#descriptor;
	constructor(schema, baseTexture, descriptor) {
		this.schema = schema;
		this.#baseTexture = baseTexture;
		this.#descriptor = descriptor;
		this[$internal] = { unwrap: () => {
			if (!this.#view) {
				const schema$1 = this.schema;
				const format = isWgslStorageTexture(schema$1) ? schema$1.format : this.#baseTexture.props.format;
				this.#view = this.#baseTexture[$internal].unwrap().createView({
					...this.#descriptor,
					label: getName(this) ?? "<unnamed>",
					format: this.#descriptor?.format ?? format,
					dimension: schema$1.dimension
				});
			}
			return this.#view;
		} };
	}
	$name(label) {
		setName(this, label);
		if (this.#view) this.#view.label = label;
		return this;
	}
	get [$gpuValueOf]() {
		const schema = this.schema;
		return new Proxy({
			[$internal]: true,
			get [$ownSnippet]() {
				return snip(this, schema, "handle");
			},
			[$resolve]: (ctx) => ctx.resolve(this),
			toString: () => `${this.toString()}.$`
		}, valueProxyHandler);
	}
	get $() {
		if (inCodegenMode()) return this[$gpuValueOf];
		throw new Error("Direct access to texture view values is possible only as part of a compute dispatch or draw call. Try .read() or .write() instead");
	}
	get value() {
		return this.$;
	}
	get size() {
		return this.#baseTexture.props.size;
	}
	toString() {
		return `textureView:${getName(this) ?? "<unnamed>"}`;
	}
	[$resolve](ctx) {
		const id = ctx.makeUniqueIdentifier(getName(this), "global");
		const { group, binding } = ctx.allocateFixedEntry(isWgslStorageTexture(this.schema) ? { storageTexture: this.schema } : {
			texture: this.schema,
			sampleType: this.#descriptor?.sampleType ?? this.schema.bindingSampleType[0]
		}, this);
		return ctx.gen.declareGlobalVar({
			group,
			binding,
			id,
			dataType: this.schema,
			scope: "handle",
			init: void 0
		});
	}
};
var TgpuLaidOutTextureViewImpl = class {
	[$internal] = { unwrap: void 0 };
	resourceType = "texture-view";
	#membership;
	schema;
	constructor(schema, membership) {
		this.schema = schema;
		this.#membership = membership;
		setName(this, membership.key);
	}
	toString() {
		return `textureView:${getName(this) ?? "<unnamed>"}`;
	}
	[$resolve](ctx) {
		const id = ctx.makeUniqueIdentifier(getName(this), "global");
		const group = ctx.allocateLayoutEntry(this.#membership.layout);
		ctx.addDeclaration(`@group(${group}) @binding(${this.#membership.idx}) var ${id}: ${ctx.resolve(this.schema).value};`);
		return snip(id, this.schema, "handle");
	}
	get [$gpuValueOf]() {
		const schema = this.schema;
		return new Proxy({
			[$internal]: true,
			get [$ownSnippet]() {
				return snip(this, schema, "handle");
			},
			[$resolve]: (ctx) => ctx.resolve(this),
			toString: () => `${this.toString()}.$`
		}, valueProxyHandler);
	}
	get $() {
		if (inCodegenMode()) return this[$gpuValueOf];
		throw new Error(`Accessed view '${getName(this) ?? "<unnamed>"}' outside of codegen mode. Direct access to texture views values is possible only as part of a compute dispatch or draw call. Try .read() or .write() instead`);
	}
	get value() {
		return this.$;
	}
	$name(label) {
		setName(this, label);
		return this;
	}
};
var TgpuTextureRenderViewImpl = class {
	[$internal];
	resourceType = "texture-view";
	descriptor;
	constructor(baseTexture, descriptor = {}) {
		this.descriptor = descriptor;
		this[$internal] = { unwrap: () => {
			return baseTexture[$internal].unwrap().createView({
				label: getName(this) ?? "<unnamed>",
				...this.descriptor
			});
		} };
	}
};

//#endregion
export { INTERNAL_createTexture, TgpuLaidOutTextureViewImpl, isTexture, isTextureView };
