// typegpu/core/texture/textureFormats.js
import { f32, i32, u32 } from "../../data/numeric.js";
import { vec4f, vec4i, vec4u } from "../../data/vector.js";

//#region src/core/texture/textureFormats.ts
const DEPTH_ASPECT_NON_COPYABLE = {
	channelType: f32,
	vectorType: vec4f,
	sampleTypes: ["depth", "unfilterable-float"],
	texelSize: "non-copyable"
};
const DEPTH_ASPECT_16 = {
	channelType: f32,
	vectorType: vec4f,
	sampleTypes: ["depth", "unfilterable-float"],
	texelSize: 2
};
const DEPTH_ASPECT_32 = {
	channelType: f32,
	vectorType: vec4f,
	sampleTypes: ["depth", "unfilterable-float"],
	texelSize: 4
};
const STENCIL_ASPECT = {
	channelType: u32,
	vectorType: vec4u,
	sampleTypes: ["uint"],
	texelSize: 1
};
const formatInfoCache = /* @__PURE__ */ new Map();
function getTextureFormatInfo(format) {
	let info = formatInfoCache.get(format);
	if (info === void 0) {
		info = createFormatInfo(format);
		formatInfoCache.set(format, info);
	}
	return info;
}
function createFormatInfo(format) {
	const channelType = parseChannelType(format);
	const depthAspect = getDepthAspect(format);
	const hasStencil = format.includes("stencil");
	return {
		channelType,
		vectorType: channelType === u32 ? vec4u : channelType === i32 ? vec4i : vec4f,
		texelSize: parseTexelSize(format),
		sampleTypes: parseSampleTypes(format),
		canRenderAttachment: canRenderAttachment(format),
		...depthAspect && { depthAspect },
		...hasStencil && { stencilAspect: STENCIL_ASPECT }
	};
}
function getDepthAspect(format) {
	if (format === "depth16unorm") return DEPTH_ASPECT_16;
	if (format === "depth32float" || format === "depth32float-stencil8") return DEPTH_ASPECT_32;
	if (format === "depth24plus" || format === "depth24plus-stencil8") return DEPTH_ASPECT_NON_COPYABLE;
}
function canRenderAttachment(format) {
	if (format.startsWith("bc") || format.startsWith("etc2") || format.startsWith("eac") || format.startsWith("astc")) return false;
	if (format === "rgb9e5ufloat") return false;
	return true;
}
function parseChannelType(format) {
	if (format === "stencil8") return u32;
	if (format.includes("uint")) return u32;
	if (format.includes("sint")) return i32;
	return f32;
}
function parseTexelSize(format) {
	const [, channels, bits] = format.match(/^(rgba|bgra|rg|r)(8|16|32)/) ?? [];
	if (channels && bits) return channels.length * Number(bits) / 8;
	if (format === "stencil8") return 1;
	if (format === "depth16unorm") return 2;
	if (format === "depth32float") return 4;
	if (format === "depth32float-stencil8") return 5;
	if (format === "depth24plus" || format === "depth24plus-stencil8") return "non-copyable";
	if (/^(bc[14]-|etc2-rgb8|eac-r11)/.test(format)) return 8;
	if (/^(bc|astc-|etc2-rgba|eac-rg)/.test(format)) return 16;
	return 4;
}
function parseSampleTypes(format) {
	if (format === "stencil8") return ["uint"];
	if (format.includes("uint")) return ["uint"];
	if (format.includes("sint")) return ["sint"];
	if (format.includes("depth")) return ["depth", "unfilterable-float"];
	if (/^(r|rg|rgba)16(u|s)norm$/.test(format)) return ["unfilterable-float"];
	return ["float", "unfilterable-float"];
}
const FLOAT32_FORMATS = new Set([
	"r32float",
	"rg32float",
	"rgba32float"
]);
function getEffectiveSampleTypes(device, format) {
	if (FLOAT32_FORMATS.has(format) && !device.features.has("float32-filterable")) return ["unfilterable-float"];
	return getTextureFormatInfo(format).sampleTypes;
}

//#endregion
export { getEffectiveSampleTypes, getTextureFormatInfo };
