// typegpu/core/function/autoIO.js
import { $internal, $resolve } from "../../shared/symbols.js";
import { getName, setName } from "../../shared/meta.js";
import { vec4f } from "../../data/vector.js";
import { AutoStruct } from "../../data/autoStruct.js";
import { createFnCore } from "./fnCore.js";
import { shaderStageSlot } from "../slot/internalSlots.js";
import { builtin } from "../../builtin.js";

//#region src/core/function/autoIO.ts
const builtinVertexIn = {
	$vertexIndex: builtin.vertexIndex,
	$instanceIndex: builtin.instanceIndex
};
const builtinVertexOut = {
	$clipDistances: builtin.clipDistances,
	$position: builtin.position
};
const builtinFragmentIn = {
	$position: builtin.position,
	$frontFacing: builtin.frontFacing,
	$primitiveIndex: builtin.primitiveIndex,
	$sampleIndex: builtin.sampleIndex,
	$sampleMask: builtin.sampleMask,
	$subgroupInvocationId: builtin.subgroupInvocationId,
	$subgroupSize: builtin.subgroupSize
};
const builtinFragmentOut = {
	$fragDepth: builtin.fragDepth,
	$sampleMask: builtin.sampleMask
};
/**
* Only used internally. Meant to be recreated on every resolution.
*/
var AutoFragmentFn = class {
	#core;
	autoIn;
	autoOut;
	constructor(impl, varyings, locations) {
		if (!getName(impl)) setName(impl, "fragmentFn");
		this.#core = createFnCore(impl, "fragment");
		this.autoIn = new AutoStruct({
			...builtinFragmentIn,
			...varyings
		}, void 0, locations);
		setName(this.autoIn, "FragmentIn");
		this.autoOut = new AutoStruct(builtinFragmentOut, vec4f);
		setName(this.autoOut, "FragmentOut");
	}
	toString() {
		return "autoFragmentFn";
	}
	[$resolve](ctx) {
		return ctx.withSlots([[shaderStageSlot, "fragment"]], () => this.#core.resolve(ctx, [this.autoIn], this.autoOut));
	}
};
AutoFragmentFn.prototype[$internal] = true;
AutoFragmentFn.prototype.resourceType = "auto-fragment-fn";
/**
* Only used internally. Meant to be recreated on every resolution.
*/
var AutoVertexFn = class {
	#core;
	autoIn;
	autoOut;
	constructor(impl, attribs, locations) {
		if (!getName(impl)) setName(impl, "vertexFn");
		this.#core = createFnCore(impl, "vertex");
		this.autoIn = new AutoStruct({
			...builtinVertexIn,
			...attribs
		}, void 0, locations);
		setName(this.autoIn, "VertexIn");
		this.autoOut = new AutoStruct(builtinVertexOut, void 0);
		setName(this.autoOut, "VertexOut");
	}
	toString() {
		return "autoVertexFn";
	}
	[$resolve](ctx) {
		return ctx.withSlots([[shaderStageSlot, "vertex"]], () => this.#core.resolve(ctx, [this.autoIn], this.autoOut));
	}
};
AutoVertexFn.prototype[$internal] = true;
AutoVertexFn.prototype.resourceType = "auto-vertex-fn";

//#endregion
export { AutoFragmentFn, AutoVertexFn };
