// typegpu/core/function/dualImpl.js
import { $gpuCallable } from "../../shared/symbols.js";
import { isPtr } from "../../data/wgslTypes.js";
import { setName } from "../../shared/meta.js";
import { snip } from "../../data/snippet.js";
import { NormalState, isKnownAtComptime } from "../../types.js";
import { tryConvertSnippet } from "../../tgsl/conversion.js";
import { concretize } from "../../tgsl/generationHelpers.js";

//#region src/core/function/dualImpl.ts
var MissingCpuImplError = class extends Error {
	constructor(message) {
		super(message);
		this.name = this.constructor.name;
	}
};
function dualImpl(options) {
	const impl = ((...args) => {
		if (typeof options.normalImpl === "string") throw new MissingCpuImplError(options.normalImpl);
		return options.normalImpl(...args);
	});
	if (options.name) setName(impl, options.name);
	impl.toString = () => options.name ?? "<unknown>";
	impl[$gpuCallable] = {
		get strictSignature() {
			return typeof options.signature !== "function" ? options.signature : void 0;
		},
		call(ctx, args) {
			const { argTypes, returnType } = typeof options.signature === "function" ? options.signature(...args.map((s) => {
				if (isPtr(s.dataType) && s.dataType.implicit) return s.dataType.inner;
				return s.dataType;
			})) : options.signature;
			const converted = args.map((s, idx) => {
				const argType = argTypes[idx];
				if (!argType) throw new Error("Function called with invalid arguments");
				return tryConvertSnippet(ctx, s, argType, !options.ignoreImplicitCastWarning);
			});
			if (!options.noComptime && converted.every((s) => isKnownAtComptime(s)) && typeof options.normalImpl === "function") {
				ctx.pushMode(new NormalState());
				try {
					return snip(options.normalImpl(...converted.map((s) => s.value)), returnType, "constant");
				} catch (e) {
					if (!(e instanceof MissingCpuImplError)) throw e;
				} finally {
					ctx.popMode("normal");
				}
			}
			return snip(options.codegenImpl(ctx, converted), concretize(returnType), "runtime");
		}
	};
	return impl;
}

//#endregion
export { MissingCpuImplError, dualImpl };
