// typegpu/core/function/entryInputRouter.js
import { $internal } from "../../shared/symbols.js";
import { isWgslStruct } from "../../data/wgslTypes.js";

//#region src/core/function/entryInputRouter.ts
/**
* Routes `(input) => { input.x }` style property access to the correct WGSL
* expression: positional args get a direct snippet, struct fields get
* `structArgName.fieldName`.
*/
var EntryInputRouter = class {
	[$internal] = {};
	type = "entry-input-router";
	structArg;
	/** Maps schemaKey → { WGSL arg name, type } */
	positionalArgsMap;
	constructor(structArg, positionalArgs) {
		this.structArg = structArg;
		this.positionalArgsMap = new Map(positionalArgs.map((a) => [a.schemaKey, a.arg]));
	}
	toString() {
		return "entry-input-router";
	}
	accessProp(propName) {
		const positionalEntry = this.positionalArgsMap.get(propName);
		if (positionalEntry) return positionalEntry();
		const structSnippet = this.structArg?.();
		if (structSnippet && isWgslStruct(structSnippet.dataType)) return {
			target: structSnippet,
			prop: propName
		};
	}
};

//#endregion
export { EntryInputRouter };
