// typegpu/core/function/fnCore.js
import { $getNameForward } from "../../shared/symbols.js";
import { Void, isWgslData, isWgslStruct } from "../../data/wgslTypes.js";
import { getFunctionMetadata, getName } from "../../shared/meta.js";
import { undecorate } from "../../data/dataTypes.js";
import { snip } from "../../data/snippet.js";
import { getAttributesString } from "../../data/attributes.js";
import { validateIdentifier } from "../../nameUtils.js";
import { mergeFunctionExternals, replaceExternalsInWgsl } from "../resolve/externals.js";
import { extractArgs } from "./extractArgs.js";

//#region src/core/function/fnCore.ts
function createFnCore(implementation, functionType, workgroupSize) {
	/**
	* External application has to be deferred until resolution because
	* some externals can reference the owner function which has not been
	* initialized yet (like when accessing the Output struct of a vertex
	* entry fn).
	*/
	const externals = {};
	return {
		[$getNameForward]: typeof implementation === "function" ? implementation : void 0,
		setExternals(key, newExternal) {
			if (key === "userProvided" && "userProvided" in externals) throw new Error("Cannot call '$uses' multiple times. If you wish to override dependencies, use slots or accessors instead.");
			externals[key] = newExternal;
			if (externals.userProvided && externals.pluginProvided) throw new Error("Cannot call '$uses' on functions whose metadata was provided by unplugin-typegpu.");
		},
		resolve(ctx, argTypes, returnType, entryInput) {
			let attributes = "";
			if (functionType === "compute") attributes = `@compute @workgroup_size(${workgroupSize?.join(", ")}) `;
			else if (functionType === "vertex") attributes = `@vertex `;
			else if (functionType === "fragment") attributes = `@fragment `;
			const id = ctx.makeUniqueIdentifier(getName(this), "global");
			if (typeof implementation === "string") {
				if (!returnType) throw new Error("Explicit return type is required for string implementation");
				if (entryInput) {
					for (const arg of entryInput.positionalArgs) {
						const result = /* @__PURE__ */ validateIdentifier(arg.schemaKey);
						if (!result.success) throw new Error(`Invalid argument name "${arg.schemaKey}"${result.error ? `: ${result.error}` : ""}`);
						if (ctx.isIdentifierBanned(arg.schemaKey)) throw new Error(`Invalid argument name "${arg.schemaKey}", the identifier is a reserved keyword.`);
					}
					this.setExternals("args", { in: Object.fromEntries(entryInput.positionalArgs.map((a) => [a.schemaKey, a.schemaKey])) });
				}
				const replacedImpl = replaceExternalsInWgsl(ctx, mergeFunctionExternals(externals), implementation);
				let header = "";
				let body = "";
				if (functionType !== "normal" && entryInput) {
					const { dataSchema, positionalArgs } = entryInput;
					const parts = [];
					if (dataSchema && isArgUsedInBody("in", replacedImpl)) parts.push(`in: ${ctx.resolve(dataSchema).value}`);
					for (const a of positionalArgs) {
						const argName = a.schemaKey;
						if (isArgUsedInBody(argName, replacedImpl)) parts.push(`${getAttributesString(a.type)}${argName}: ${ctx.resolve(a.type).value}`);
					}
					const input = `(${parts.join(", ")})`;
					const attributes$1 = isWgslData(returnType) ? getAttributesString(returnType) : "";
					header = `${input}${returnType !== Void ? isWgslStruct(returnType) ? ` -> ${ctx.resolve(returnType).value} ` : ` -> ${attributes$1 !== "" ? attributes$1 : "@location(0)"} ${ctx.resolve(returnType).value} ` : " "}`;
					body = replacedImpl;
				} else {
					const providedArgs = extractArgs(replacedImpl);
					if (providedArgs.args.length !== argTypes.length) throw new Error(`WGSL implementation has ${providedArgs.args.length} arguments, while the shell has ${argTypes.length} arguments.`);
					header = `(${providedArgs.args.map((argInfo, i) => `${argInfo.identifier}: ${checkAndReturnType(ctx, `parameter ${argInfo.identifier}`, argInfo.type, argTypes[i])}`).join(", ")})${returnType === Void ? " " : ` -> ${checkAndReturnType(ctx, "return type", providedArgs.ret?.type, returnType)} `}`;
					body = replacedImpl.slice(providedArgs.range.end);
				}
				ctx.addDeclaration(`${attributes}fn ${id}${header}${body}`);
				return snip(id, returnType, "runtime");
			}
			const pluginData = getFunctionMetadata(implementation);
			const pluginExternals = pluginData?.externals();
			if (pluginExternals) this.setExternals("pluginProvided", pluginExternals);
			const ast = pluginData?.ast;
			if (!ast) throw new Error("Missing metadata for tgpu.fn function body (either missing 'use gpu' directive, or misconfigured `unplugin-typegpu`)");
			const maybeSecondArg = ast.params[1];
			if (maybeSecondArg && maybeSecondArg.type === "i" && functionType !== "normal") this.setExternals("out", { [maybeSecondArg.name]: undecorate(returnType) });
			const { code, returnType: actualReturnType } = ctx.resolveFunction({
				functionType,
				name: id,
				workgroupSize,
				argTypes,
				entryInput,
				params: ast.params,
				returnType,
				body: ast.body,
				externalMap: mergeFunctionExternals(externals)
			});
			ctx.addDeclaration(code);
			return snip(id, actualReturnType, "runtime");
		}
	};
}
function isArgUsedInBody(argName, body) {
	return (/* @__PURE__ */ new RegExp(`\\b${argName}\\b`)).test(body);
}
function checkAndReturnType(ctx, name, wgslType, jsType) {
	const resolvedJsType = ctx.resolve(jsType).value.replace(/\s/g, "");
	if (!wgslType) return resolvedJsType;
	const resolvedWgslType = wgslType.replace(/\s/g, "");
	if (resolvedWgslType !== resolvedJsType) throw new Error(`Type mismatch between TGPU shell and WGSL code string: ${name}, JS type "${resolvedJsType}", WGSL type "${resolvedWgslType}".`);
	return wgslType;
}

//#endregion
export { createFnCore };
