// typegpu/core/function/ioSchema.js
import { isVoid } from "../../data/wgslTypes.js";
import { getCustomLocation, isData } from "../../data/dataTypes.js";
import { isBuiltin, location } from "../../data/attributes.js";
import { INTERNAL_createStruct } from "../../data/struct.js";

//#region src/core/function/ioSchema.ts
function withLocations(members, locations = {}) {
	let nextLocation = 0;
	const usedCustomLocations = /* @__PURE__ */ new Set();
	return Object.fromEntries(Object.entries(members ?? {}).map(([key, member]) => {
		const customLocation = getCustomLocation(member);
		if (customLocation !== void 0) {
			if (usedCustomLocations.has(customLocation)) throw new Error("Duplicate custom location attributes found");
			usedCustomLocations.add(customLocation);
		}
		return [key, member];
	}).map(([key, member]) => {
		if (isBuiltin(member)) return [key, member];
		if (getCustomLocation(member) !== void 0) return [key, member];
		if (locations[key]) return [key, location(locations[key], member)];
		while (usedCustomLocations.has(nextLocation)) nextLocation++;
		return [key, location(nextLocation++, member)];
	}));
}
function separateBuiltins(schema, locations = {}) {
	const positionalArgs = [];
	const dataFields = {};
	for (const [key, type] of Object.entries(schema)) if (isBuiltin(type)) positionalArgs.push({
		schemaKey: key,
		type
	});
	else dataFields[key] = type;
	return {
		dataSchema: Object.keys(dataFields).length > 0 ? INTERNAL_createStruct(withLocations(dataFields, locations), false) : void 0,
		positionalArgs
	};
}
function separateAllAsPositional(schema) {
	const withLocs = withLocations(schema);
	return {
		dataSchema: void 0,
		positionalArgs: Object.entries(withLocs).map(([key, type]) => ({
			schemaKey: key,
			type
		}))
	};
}
function createIoSchema(layout, locations = {}) {
	return isData(layout) ? isVoid(layout) ? layout : isBuiltin(layout) ? layout : getCustomLocation(layout) !== void 0 ? layout : location(0, layout) : INTERNAL_createStruct(withLocations(layout, locations), false);
}

//#endregion
export { createIoSchema, separateAllAsPositional, separateBuiltins };
