// typegpu/core/function/tgpuFn.js
import { $getNameForward, $internal, $providing, $resolve, isMarkedInternal } from "../../shared/symbols.js";
import { Void } from "../../data/wgslTypes.js";
import { getName, setName } from "../../shared/meta.js";
import { ExecutionError } from "../../errors.js";
import { isAccessor, isMutableAccessor } from "../slot/slotTypes.js";
import { provideInsideTgpuFn } from "../../execMode.js";
import { stitch } from "../resolve/stitch.js";
import { schemaCallWrapper } from "../../data/schemaCallWrapper.js";
import { dualImpl } from "./dualImpl.js";
import { comptime } from "./comptime.js";
import { addArgTypesToExternals, addReturnTypeToExternals } from "../resolve/externals.js";
import { createFnCore } from "./fnCore.js";
import { stripTemplate } from "./templateUtils.js";

//#region src/core/function/tgpuFn.ts
function fn(argTypesOrCallback, returnType) {
	if (typeof argTypesOrCallback === "function") return createGenericFn(argTypesOrCallback, []);
	const argTypes = argTypesOrCallback;
	const shell = {
		[$internal]: true,
		argTypes,
		returnType: returnType ?? Void
	};
	const call = (arg, ...values) => createFn(shell, stripTemplate(arg, ...values));
	return Object.assign(call, shell);
}
function isTgpuFn(value) {
	return isMarkedInternal(value) && value?.resourceType === "function";
}
function isGenericFn(value) {
	return isMarkedInternal(value) && value?.resourceType === "generic-function";
}
function stringifyPair([slot, value]) {
	return `${getName(slot) ?? "<unnamed>"}=${value}`;
}
function createFn(shell, _implementation) {
	let pairs = [];
	let implementation;
	if (isGenericFn(_implementation)) {
		pairs = _implementation[$providing]?.pairs ?? [];
		implementation = _implementation[$internal].inner;
	} else implementation = _implementation;
	const core = createFnCore(implementation, "normal");
	const fnBase = {
		shell,
		resourceType: "function",
		[$internal]: { implementation },
		$uses(newExternals) {
			core.setExternals("userProvided", newExternals);
			return this;
		},
		[$getNameForward]: core,
		$name(label) {
			setName(this, label);
			return this;
		},
		with: comptime((slot, value) => {
			return createBoundFunction(fn$1, [[isAccessor(slot) || isMutableAccessor(slot) ? slot.slot : slot, value]]);
		}),
		[$resolve](ctx) {
			if (typeof implementation === "string") {
				addArgTypesToExternals(implementation, shell.argTypes, core);
				addReturnTypeToExternals(implementation, shell.returnType, core);
			}
			return core.resolve(ctx, shell.argTypes, shell.returnType);
		}
	};
	const call = dualImpl({
		name: void 0,
		noComptime: true,
		signature: {
			argTypes: shell.argTypes,
			returnType: shell.returnType
		},
		normalImpl: (...args) => provideInsideTgpuFn(() => {
			try {
				if (typeof implementation === "string") throw new Error("Cannot execute on the CPU functions constructed with raw WGSL");
				const castAndCopiedArgs = args.map((arg, index) => schemaCallWrapper(shell.argTypes[index], arg));
				const result = implementation(...castAndCopiedArgs);
				return schemaCallWrapper(shell.returnType, result);
			} catch (err) {
				if (err instanceof ExecutionError) throw err.appendToTrace(fn$1);
				throw new ExecutionError(err, [fn$1]);
			}
		}),
		codegenImpl: (ctx, args) => ctx.withResetIndentLevel(() => stitch`${ctx.resolve(fn$1).value}(${args})`)
	});
	const fn$1 = Object.assign(call, fnBase);
	Object.defineProperty(fn$1, "toString", { value() {
		return `fn:${getName(core) ?? "<unnamed>"}`;
	} });
	if (pairs.length > 0) return createBoundFunction(fn$1, pairs);
	return fn$1;
}
function createBoundFunction(innerFn, pairs) {
	const fnBase = {
		resourceType: "function",
		shell: innerFn.shell,
		[$internal]: { implementation: innerFn[$internal].implementation },
		[$providing]: {
			inner: innerFn,
			pairs
		},
		$uses(newExternals) {
			innerFn.$uses(newExternals);
			return this;
		},
		$name(label) {
			setName(this, label);
			return this;
		},
		with: comptime((slot, value) => {
			const s = isAccessor(slot) || isMutableAccessor(slot) ? slot.slot : slot;
			return createBoundFunction(innerFn, [...pairs, [s, value]]);
		})
	};
	const call = dualImpl({
		name: void 0,
		noComptime: true,
		signature: {
			argTypes: innerFn.shell.argTypes,
			returnType: innerFn.shell.returnType
		},
		normalImpl: innerFn,
		codegenImpl: (ctx, args) => ctx.withResetIndentLevel(() => stitch`${ctx.resolve(fn$1).value}(${args})`)
	});
	const fn$1 = Object.assign(call, fnBase);
	Object.defineProperty(fn$1, "toString", { value() {
		return `fn:${getName(this) ?? "<unnamed>"}[${pairs.map(stringifyPair).join(", ")}]`;
	} });
	const innerName = getName(innerFn);
	if (innerName) setName(fn$1, innerName);
	return fn$1;
}
function createGenericFn(inner, pairs) {
	const fnBase = {
		[$internal]: { inner },
		resourceType: "generic-function",
		[$providing]: pairs.length > 0 ? {
			inner,
			pairs
		} : void 0,
		$name(label) {
			setName(this, label);
			if (!getName(inner)) setName(inner, label);
			return this;
		},
		[$resolve](ctx) {
			return ctx.resolve(inner);
		},
		with(slot, value) {
			const s = isAccessor(slot) || isMutableAccessor(slot) ? slot.slot : slot;
			return createGenericFn(inner, [...pairs, [s, value]]);
		}
	};
	const call = (...args) => {
		return inner(...args);
	};
	const genericFn = Object.assign(call, fnBase);
	const innerName = getName(inner);
	if (innerName) setName(genericFn, innerName);
	Object.defineProperty(genericFn, "toString", { value() {
		const fnLabel = getName(genericFn) ?? "<unnamed>";
		if (pairs.length > 0) return `fn*:${fnLabel}[${pairs.map(stringifyPair).join(", ")}]`;
		return `fn*:${fnLabel}`;
	} });
	return genericFn;
}

//#endregion
export { fn, isGenericFn, isTgpuFn };
