// typegpu/core/slot/accessor.js
import { $getNameForward, $gpuValueOf, $internal, $ownSnippet, $resolve } from "../../shared/symbols.js";
import { getName, hasTinyestMetadata, setName } from "../../shared/meta.js";
import { isData } from "../../data/dataTypes.js";
import { isSnippet, snip } from "../../data/snippet.js";
import { NormalState, getOwnSnippet } from "../../types.js";
import { getResolutionCtx, inCodegenMode } from "../../execMode.js";
import { schemaCallWrapper } from "../../data/schemaCallWrapper.js";
import { getGpuValueRecursively, valueProxyHandler } from "../valueProxyUtils.js";
import { isTgpuFn } from "../function/tgpuFn.js";
import { slot } from "./slot.js";

//#region src/core/slot/accessor.ts
function accessor(schemaOrConstructor, defaultValue) {
	return new TgpuAccessorImpl(schemaOrConstructor, defaultValue);
}
function mutableAccessor(schemaOrConstructor, defaultValue) {
	return new TgpuMutableAccessorImpl(schemaOrConstructor, defaultValue);
}
var AccessorBase = class {
	[$internal] = true;
	[$getNameForward];
	slot;
	schema;
	defaultValue;
	constructor(schemaOrConstructor, defaultValue = void 0) {
		this.schema = isData(schemaOrConstructor) ? schemaOrConstructor : schemaOrConstructor(0);
		this.defaultValue = defaultValue;
		this.slot = (() => slot(defaultValue))();
		this[$getNameForward] = this.slot;
	}
	get [$gpuValueOf]() {
		return new Proxy({
			[$internal]: true,
			[$ownSnippet]: this.#createSnippet(),
			[$resolve]: (ctx) => ctx.resolve(this),
			toString: () => `${this.resourceType}:${getName(this) ?? "<unnamed>"}.$`
		}, valueProxyHandler);
	}
	/**
	* @returns A snippet representing the accessor.
	*/
	#createSnippet() {
		const ctx = getResolutionCtx();
		let value = getGpuValueRecursively(ctx.unwrap(this.slot));
		while (typeof value === "function" && !isTgpuFn(value) && !hasTinyestMetadata(value)) {
			value = value();
			if (isSnippet(value)) value = value.value;
		}
		const ownSnippet = getOwnSnippet(value);
		if (ownSnippet) return ownSnippet;
		if (isTgpuFn(value) || hasTinyestMetadata(value)) return ctx.withResetIndentLevel(() => snip(`${ctx.resolve(value).value}()`, this.schema, "runtime"));
		ctx.pushMode(new NormalState());
		try {
			return snip(schemaCallWrapper(this.schema, value), this.schema, "constant", false);
		} finally {
			ctx.popMode("normal");
		}
	}
	$name(label) {
		setName(this, label);
		if (this.defaultValue && typeof this.defaultValue === "function" && !getName(this.defaultValue)) setName(this.defaultValue, label);
		return this;
	}
	toString() {
		return `${this.resourceType}:${getName(this) ?? "<unnamed>"}`;
	}
	get value() {
		return this.$;
	}
	[$resolve](ctx) {
		const snippet = this.#createSnippet();
		return snip(ctx.resolve(snippet.value, snippet.dataType).value, snippet.dataType, snippet.origin);
	}
};
var TgpuAccessorImpl = class extends AccessorBase {
	resourceType = "accessor";
	constructor(schemaOrConstructor, defaultValue = void 0) {
		super(schemaOrConstructor, defaultValue);
	}
	get $() {
		if (inCodegenMode()) return this[$gpuValueOf];
		throw new Error("`tgpu.accessor` relies on GPU resources and cannot be accessed outside of a compute dispatch or draw call. Use `tgpu.slot` for non-WGSL values instead.");
	}
};
var TgpuMutableAccessorImpl = class extends AccessorBase {
	resourceType = "mutable-accessor";
	constructor(schemaOrConstructor, defaultValue = void 0) {
		super(schemaOrConstructor, defaultValue);
	}
	get $() {
		if (inCodegenMode()) return this[$gpuValueOf];
		throw new Error("`tgpu.mutableAccessor` relies on GPU resources and cannot be accessed outside of a compute dispatch or draw call. Use `tgpu.slot` for non-WGSL values instead.");
	}
};

//#endregion
export { accessor, mutableAccessor };
