// typegpu/core/slot/lazy.js
import { $gpuValueOf, $internal, $providing } from "../../shared/symbols.js";
import { isAccessor, isMutableAccessor } from "./slotTypes.js";
import { getResolutionCtx } from "../../execMode.js";
import { getGpuValueRecursively } from "../valueProxyUtils.js";

//#region src/core/slot/lazy.ts
function lazy(compute) {
	if (getResolutionCtx()) throw new Error("Cannot create tgpu.lazy objects during shader resolution.");
	return new TgpuLazyImpl(compute, void 0);
}
var TgpuLazyImpl = class TgpuLazyImpl {
	[$internal];
	[$providing];
	constructor(compute, providing) {
		this[$internal] = { compute };
		this[$providing] = providing;
	}
	get [$gpuValueOf]() {
		const ctx = getResolutionCtx();
		if (!ctx) throw new Error(`Cannot access tgpu.lazy's value outside of resolution.`);
		return getGpuValueRecursively(ctx.unwrap(this));
	}
	get value() {
		return this[$gpuValueOf];
	}
	get $() {
		return this[$gpuValueOf];
	}
	toString() {
		return "lazy";
	}
	with(slot, value) {
		return new TgpuLazyImpl(this[$internal].compute.bind(this), {
			inner: this[$providing]?.inner ?? this,
			pairs: [...this[$providing]?.pairs ?? [], [isAccessor(slot) || isMutableAccessor(slot) ? slot.slot : slot, value]]
		});
	}
};
TgpuLazyImpl.prototype.resourceType = "lazy";

//#endregion
export { lazy };
