// typegpu/std/atomic.js
import { Void, isAtomic } from "../data/wgslTypes.js";
import { safeStringify } from "../shared/stringify.js";
import { stitch } from "../core/resolve/stitch.js";
import { i32, u32 } from "../data/numeric.js";
import { dualImpl } from "../core/function/dualImpl.js";

//#region src/std/atomic.ts
const workgroupBarrier = dualImpl({
	name: "workgroupBarrier",
	normalImpl: "workgroupBarrier is a no-op outside of CODEGEN mode.",
	signature: {
		argTypes: [],
		returnType: Void
	},
	codegenImpl: () => "workgroupBarrier()"
});
const storageBarrier = dualImpl({
	name: "storageBarrier",
	normalImpl: "storageBarrier is a no-op outside of CODEGEN mode.",
	signature: {
		argTypes: [],
		returnType: Void
	},
	codegenImpl: () => "storageBarrier()"
});
const textureBarrier = dualImpl({
	name: "textureBarrier",
	normalImpl: "textureBarrier is a no-op outside of CODEGEN mode.",
	signature: {
		argTypes: [],
		returnType: Void
	},
	codegenImpl: () => "textureBarrier()"
});
const atomicNormalError = "Atomic operations are not supported outside of CODEGEN mode.";
const atomicLoad = dualImpl({
	name: "atomicLoad",
	normalImpl: atomicNormalError,
	signature: (a) => {
		if (!isAtomic(a)) throw new Error(`Invalid atomic type: ${safeStringify(a)}`);
		return {
			argTypes: [a],
			returnType: a.inner
		};
	},
	codegenImpl: (_ctx, [a]) => stitch`atomicLoad(&${a})`
});
const atomicActionSignature = (a) => {
	if (!isAtomic(a)) throw new Error(`Invalid atomic type: ${safeStringify(a)}`);
	return {
		argTypes: [a, a.inner.type === "u32" ? u32 : i32],
		returnType: Void
	};
};
const atomicOpSignature = (a) => {
	if (!isAtomic(a)) throw new Error(`Invalid atomic type: ${safeStringify(a)}`);
	const paramType = a.inner.type === "u32" ? u32 : i32;
	return {
		argTypes: [a, paramType],
		returnType: paramType
	};
};
const atomicStore = dualImpl({
	name: "atomicStore",
	normalImpl: atomicNormalError,
	signature: atomicActionSignature,
	codegenImpl: (_ctx, [a, value]) => stitch`atomicStore(&${a}, ${value})`
});
const atomicAdd = dualImpl({
	name: "atomicAdd",
	normalImpl: atomicNormalError,
	signature: atomicOpSignature,
	codegenImpl: (_ctx, [a, value]) => stitch`atomicAdd(&${a}, ${value})`
});
const atomicSub = dualImpl({
	name: "atomicSub",
	normalImpl: atomicNormalError,
	signature: atomicOpSignature,
	codegenImpl: (_ctx, [a, value]) => stitch`atomicSub(&${a}, ${value})`
});
const atomicMax = dualImpl({
	name: "atomicMax",
	normalImpl: atomicNormalError,
	signature: atomicOpSignature,
	codegenImpl: (_ctx, [a, value]) => stitch`atomicMax(&${a}, ${value})`
});
const atomicMin = dualImpl({
	name: "atomicMin",
	normalImpl: atomicNormalError,
	signature: atomicOpSignature,
	codegenImpl: (_ctx, [a, value]) => stitch`atomicMin(&${a}, ${value})`
});
const atomicAnd = dualImpl({
	name: "atomicAnd",
	normalImpl: atomicNormalError,
	signature: atomicOpSignature,
	codegenImpl: (_ctx, [a, value]) => stitch`atomicAnd(&${a}, ${value})`
});
const atomicOr = dualImpl({
	name: "atomicOr",
	normalImpl: atomicNormalError,
	signature: atomicOpSignature,
	codegenImpl: (_ctx, [a, value]) => stitch`atomicOr(&${a}, ${value})`
});
const atomicXor = dualImpl({
	name: "atomicXor",
	normalImpl: atomicNormalError,
	signature: atomicOpSignature,
	codegenImpl: (_ctx, [a, value]) => stitch`atomicXor(&${a}, ${value})`
});

//#endregion
export { atomicAdd, atomicAnd, atomicLoad, atomicMax, atomicMin, atomicOr, atomicStore, atomicSub, atomicXor, storageBarrier, textureBarrier, workgroupBarrier };
