// typegpu/std/bitcast.js
import { isVec } from "../data/wgslTypes.js";
import { stitch } from "../core/resolve/stitch.js";
import { unify } from "../tgsl/conversion.js";
import { f32, i32, u32 } from "../data/numeric.js";
import { vec2f, vec2i, vec3f, vec3i, vec4f, vec4i } from "../data/vector.js";
import { dualImpl } from "../core/function/dualImpl.js";
import { bitcastU32toF32Impl, bitcastU32toI32Impl } from "../data/numberOps.js";
import { VectorOps } from "../data/vectorOps.js";

//#region src/std/bitcast.ts
const bitcastU32toF32 = dualImpl({
	name: "bitcastU32toF32",
	normalImpl: ((value) => {
		if (typeof value === "number") return bitcastU32toF32Impl(value);
		return VectorOps.bitcastU32toF32[value.kind](value);
	}),
	codegenImpl: (_ctx, [n]) => stitch`bitcast<f32>(${n})`,
	signature: (...arg) => {
		const uargs = unify(arg, [u32]) ?? arg;
		return {
			argTypes: uargs,
			returnType: isVec(uargs[0]) ? uargs[0].type === "vec2u" ? vec2f : uargs[0].type === "vec3u" ? vec3f : vec4f : f32
		};
	}
});
const bitcastU32toI32 = dualImpl({
	name: "bitcastU32toI32",
	normalImpl: ((value) => {
		if (typeof value === "number") return bitcastU32toI32Impl(value);
		return VectorOps.bitcastU32toI32[value.kind](value);
	}),
	codegenImpl: (_ctx, [n]) => stitch`bitcast<i32>(${n})`,
	signature: (...arg) => {
		const uargs = unify(arg, [u32]) ?? arg;
		return {
			argTypes: uargs,
			returnType: isVec(uargs[0]) ? uargs[0].type === "vec2u" ? vec2i : uargs[0].type === "vec3u" ? vec3i : vec4i : i32
		};
	}
});

//#endregion
export { bitcastU32toF32, bitcastU32toI32 };
