// typegpu/std/derivative.js
import { stitch } from "../core/resolve/stitch.js";
import { dualImpl } from "../core/function/dualImpl.js";

//#region src/std/derivative.ts
const derivativeNormalError = "Derivative builtins are not allowed on the CPU";
const dpdx = dualImpl({
	name: "dpdx",
	normalImpl: derivativeNormalError,
	signature: (value) => ({
		argTypes: [value],
		returnType: value
	}),
	codegenImpl: (_ctx, [value]) => stitch`dpdx(${value})`
});
const dpdxCoarse = dualImpl({
	name: "dpdxCoarse",
	normalImpl: derivativeNormalError,
	signature: (value) => ({
		argTypes: [value],
		returnType: value
	}),
	codegenImpl: (_ctx, [value]) => stitch`dpdxCoarse(${value})`
});
const dpdxFine = dualImpl({
	name: "dpdxFine",
	normalImpl: derivativeNormalError,
	signature: (value) => ({
		argTypes: [value],
		returnType: value
	}),
	codegenImpl: (_ctx, [value]) => stitch`dpdxFine(${value})`
});
const dpdy = dualImpl({
	name: "dpdy",
	normalImpl: derivativeNormalError,
	signature: (value) => ({
		argTypes: [value],
		returnType: value
	}),
	codegenImpl: (_ctx, [value]) => stitch`dpdy(${value})`
});
const dpdyCoarse = dualImpl({
	name: "dpdyCoarse",
	normalImpl: derivativeNormalError,
	signature: (value) => ({
		argTypes: [value],
		returnType: value
	}),
	codegenImpl: (_ctx, [value]) => stitch`dpdyCoarse(${value})`
});
const dpdyFine = dualImpl({
	name: "dpdyFine",
	normalImpl: derivativeNormalError,
	signature: (value) => ({
		argTypes: [value],
		returnType: value
	}),
	codegenImpl: (_ctx, [value]) => stitch`dpdyFine(${value})`
});
const fwidth = dualImpl({
	name: "fwidth",
	normalImpl: derivativeNormalError,
	signature: (value) => ({
		argTypes: [value],
		returnType: value
	}),
	codegenImpl: (_ctx, [value]) => stitch`fwidth(${value})`
});
const fwidthCoarse = dualImpl({
	name: "fwidthCoarse",
	normalImpl: derivativeNormalError,
	signature: (value) => ({
		argTypes: [value],
		returnType: value
	}),
	codegenImpl: (_ctx, [value]) => stitch`fwidthCoarse(${value})`
});
const fwidthFine = dualImpl({
	name: "fwidthFine",
	normalImpl: derivativeNormalError,
	signature: (value) => ({
		argTypes: [value],
		returnType: value
	}),
	codegenImpl: (_ctx, [value]) => stitch`fwidthFine(${value})`
});

//#endregion
export { dpdx, dpdxCoarse, dpdxFine, dpdy, dpdyCoarse, dpdyFine, fwidth, fwidthCoarse, fwidthFine };
