// typegpu/std/index.js
import { __export } from "../_virtual/rolldown_runtime.js";
import { identity2, identity3, identity4, rotationX4, rotationY4, rotationZ4, scaling4, translation4 } from "../data/matrix.js";
import { add, bitShiftLeft, bitShiftRight, div, mod, mul, neg, sub } from "./operators.js";
import { abs, acos, acosh, asin, asinh, atan, atan2, atanh, ceil, clamp, cos, cosh, countLeadingZeros, countOneBits, countTrailingZeros, cross, degrees, determinant, distance, dot, dot4I8Packed, dot4U8Packed, exp, exp2, extractBits, faceForward, firstLeadingBit, firstTrailingBit, floor, fma, fract, frexp, insertBits, inverseSqrt, ldexp, length, log, log2, max, min, mix, modf, normalize, pow, quantizeToF16, radians, reflect, refract, reverseBits, round, saturate, sign, sin, sinh, smoothstep, sqrt, step, tan, tanh, transpose, trunc } from "./numeric.js";
import { arrayLength } from "./array.js";
import { range } from "./range.js";
import { all, allEq, and, any, eq, ge, gt, isCloseTo, le, lt, ne, not, or, select } from "./boolean.js";
import { bitcastU32toF32, bitcastU32toI32 } from "./bitcast.js";
import { pack2x16float, pack4x8unorm, unpack2x16float, unpack4x8unorm } from "./packing.js";
import { copy } from "./copy.js";
import { discard } from "./discard.js";
import { rotateX4, rotateY4, rotateZ4, scale4, translate4 } from "./matrix.js";
import { atomicAdd, atomicAnd, atomicLoad, atomicMax, atomicMin, atomicOr, atomicStore, atomicSub, atomicXor, storageBarrier, textureBarrier, workgroupBarrier } from "./atomic.js";
import { dpdx, dpdxCoarse, dpdxFine, dpdy, dpdyCoarse, dpdyFine, fwidth, fwidthCoarse, fwidthFine } from "./derivative.js";
import { textureDimensions, textureGather, textureLoad, textureSample, textureSampleBaseClampToEdge, textureSampleBias, textureSampleCompare, textureSampleCompareLevel, textureSampleGrad, textureSampleLevel, textureStore } from "./texture.js";
import { subgroupAdd, subgroupAll, subgroupAnd, subgroupAny, subgroupBallot, subgroupBroadcast, subgroupBroadcastFirst, subgroupElect, subgroupExclusiveAdd, subgroupExclusiveMul, subgroupInclusiveAdd, subgroupInclusiveMul, subgroupMax, subgroupMin, subgroupMul, subgroupOr, subgroupShuffle, subgroupShuffleDown, subgroupShuffleUp, subgroupShuffleXor, subgroupXor } from "./subgroup.js";
import { extensionEnabled } from "./extensions.js";
import { getTargetShaderLanguage, isBeingTranspiled } from "./environment.js";

//#region src/std/index.ts
var std_exports = /* @__PURE__ */ __export({
	abs: () => abs,
	acos: () => acos,
	acosh: () => acosh,
	add: () => add,
	all: () => all,
	allEq: () => allEq,
	and: () => and,
	any: () => any,
	arrayLength: () => arrayLength,
	asin: () => asin,
	asinh: () => asinh,
	atan: () => atan,
	atan2: () => atan2,
	atanh: () => atanh,
	atomicAdd: () => atomicAdd,
	atomicAnd: () => atomicAnd,
	atomicLoad: () => atomicLoad,
	atomicMax: () => atomicMax,
	atomicMin: () => atomicMin,
	atomicOr: () => atomicOr,
	atomicStore: () => atomicStore,
	atomicSub: () => atomicSub,
	atomicXor: () => atomicXor,
	bitShiftLeft: () => bitShiftLeft,
	bitShiftRight: () => bitShiftRight,
	bitcastU32toF32: () => bitcastU32toF32,
	bitcastU32toI32: () => bitcastU32toI32,
	ceil: () => ceil,
	clamp: () => clamp,
	copy: () => copy,
	cos: () => cos,
	cosh: () => cosh,
	countLeadingZeros: () => countLeadingZeros,
	countOneBits: () => countOneBits,
	countTrailingZeros: () => countTrailingZeros,
	cross: () => cross,
	degrees: () => degrees,
	determinant: () => determinant,
	discard: () => discard,
	distance: () => distance,
	div: () => div,
	dot: () => dot,
	dot4I8Packed: () => dot4I8Packed,
	dot4U8Packed: () => dot4U8Packed,
	dpdx: () => dpdx,
	dpdxCoarse: () => dpdxCoarse,
	dpdxFine: () => dpdxFine,
	dpdy: () => dpdy,
	dpdyCoarse: () => dpdyCoarse,
	dpdyFine: () => dpdyFine,
	eq: () => eq,
	exp: () => exp,
	exp2: () => exp2,
	extensionEnabled: () => extensionEnabled,
	extractBits: () => extractBits,
	faceForward: () => faceForward,
	firstLeadingBit: () => firstLeadingBit,
	firstTrailingBit: () => firstTrailingBit,
	floor: () => floor,
	fma: () => fma,
	fract: () => fract,
	frexp: () => frexp,
	fwidth: () => fwidth,
	fwidthCoarse: () => fwidthCoarse,
	fwidthFine: () => fwidthFine,
	ge: () => ge,
	getTargetShaderLanguage: () => getTargetShaderLanguage,
	gt: () => gt,
	identity2: () => identity2,
	identity3: () => identity3,
	identity4: () => identity4,
	insertBits: () => insertBits,
	inverseSqrt: () => inverseSqrt,
	isBeingTranspiled: () => isBeingTranspiled,
	isCloseTo: () => isCloseTo,
	ldexp: () => ldexp,
	le: () => le,
	length: () => length,
	log: () => log,
	log2: () => log2,
	lt: () => lt,
	max: () => max,
	min: () => min,
	mix: () => mix,
	mod: () => mod,
	modf: () => modf,
	mul: () => mul,
	ne: () => ne,
	neg: () => neg,
	normalize: () => normalize,
	not: () => not,
	or: () => or,
	pack2x16float: () => pack2x16float,
	pack4x8unorm: () => pack4x8unorm,
	pow: () => pow,
	quantizeToF16: () => quantizeToF16,
	radians: () => radians,
	range: () => range,
	reflect: () => reflect,
	refract: () => refract,
	reverseBits: () => reverseBits,
	rotateX4: () => rotateX4,
	rotateY4: () => rotateY4,
	rotateZ4: () => rotateZ4,
	rotationX4: () => rotationX4,
	rotationY4: () => rotationY4,
	rotationZ4: () => rotationZ4,
	round: () => round,
	saturate: () => saturate,
	scale4: () => scale4,
	scaling4: () => scaling4,
	select: () => select,
	sign: () => sign,
	sin: () => sin,
	sinh: () => sinh,
	smoothstep: () => smoothstep,
	sqrt: () => sqrt,
	step: () => step,
	storageBarrier: () => storageBarrier,
	sub: () => sub,
	subgroupAdd: () => subgroupAdd,
	subgroupAll: () => subgroupAll,
	subgroupAnd: () => subgroupAnd,
	subgroupAny: () => subgroupAny,
	subgroupBallot: () => subgroupBallot,
	subgroupBroadcast: () => subgroupBroadcast,
	subgroupBroadcastFirst: () => subgroupBroadcastFirst,
	subgroupElect: () => subgroupElect,
	subgroupExclusiveAdd: () => subgroupExclusiveAdd,
	subgroupExclusiveMul: () => subgroupExclusiveMul,
	subgroupInclusiveAdd: () => subgroupInclusiveAdd,
	subgroupInclusiveMul: () => subgroupInclusiveMul,
	subgroupMax: () => subgroupMax,
	subgroupMin: () => subgroupMin,
	subgroupMul: () => subgroupMul,
	subgroupOr: () => subgroupOr,
	subgroupShuffle: () => subgroupShuffle,
	subgroupShuffleDown: () => subgroupShuffleDown,
	subgroupShuffleUp: () => subgroupShuffleUp,
	subgroupShuffleXor: () => subgroupShuffleXor,
	subgroupXor: () => subgroupXor,
	tan: () => tan,
	tanh: () => tanh,
	textureBarrier: () => textureBarrier,
	textureDimensions: () => textureDimensions,
	textureGather: () => textureGather,
	textureLoad: () => textureLoad,
	textureSample: () => textureSample,
	textureSampleBaseClampToEdge: () => textureSampleBaseClampToEdge,
	textureSampleBias: () => textureSampleBias,
	textureSampleCompare: () => textureSampleCompare,
	textureSampleCompareLevel: () => textureSampleCompareLevel,
	textureSampleGrad: () => textureSampleGrad,
	textureSampleLevel: () => textureSampleLevel,
	textureStore: () => textureStore,
	translate4: () => translate4,
	translation4: () => translation4,
	transpose: () => transpose,
	trunc: () => trunc,
	unpack2x16float: () => unpack2x16float,
	unpack4x8unorm: () => unpack4x8unorm,
	workgroupBarrier: () => workgroupBarrier
});

//#endregion
export { abs, acos, acosh, add, all, allEq, and, any, arrayLength, asin, asinh, atan, atan2, atanh, atomicAdd, atomicAnd, atomicLoad, atomicMax, atomicMin, atomicOr, atomicStore, atomicSub, atomicXor, bitShiftLeft, bitShiftRight, bitcastU32toF32, bitcastU32toI32, ceil, clamp, copy, cos, cosh, countLeadingZeros, countOneBits, countTrailingZeros, cross, degrees, determinant, discard, distance, div, dot, dot4I8Packed, dot4U8Packed, dpdx, dpdxCoarse, dpdxFine, dpdy, dpdyCoarse, dpdyFine, eq, exp, exp2, extensionEnabled, extractBits, faceForward, firstLeadingBit, firstTrailingBit, floor, fma, fract, frexp, fwidth, fwidthCoarse, fwidthFine, ge, getTargetShaderLanguage, gt, identity2, identity3, identity4, insertBits, inverseSqrt, isBeingTranspiled, isCloseTo, ldexp, le, length, log, log2, lt, max, min, mix, mod, modf, mul, ne, neg, normalize, not, or, pack2x16float, pack4x8unorm, pow, quantizeToF16, radians, range, reflect, refract, reverseBits, rotateX4, rotateY4, rotateZ4, rotationX4, rotationY4, rotationZ4, round, saturate, scale4, scaling4, select, sign, sin, sinh, smoothstep, sqrt, std_exports, step, storageBarrier, sub, subgroupAdd, subgroupAll, subgroupAnd, subgroupAny, subgroupBallot, subgroupBroadcast, subgroupBroadcastFirst, subgroupElect, subgroupExclusiveAdd, subgroupExclusiveMul, subgroupInclusiveAdd, subgroupInclusiveMul, subgroupMax, subgroupMin, subgroupMul, subgroupOr, subgroupShuffle, subgroupShuffleDown, subgroupShuffleUp, subgroupShuffleXor, subgroupXor, tan, tanh, textureBarrier, textureDimensions, textureGather, textureLoad, textureSample, textureSampleBaseClampToEdge, textureSampleBias, textureSampleCompare, textureSampleCompareLevel, textureSampleGrad, textureSampleLevel, textureStore, translate4, translation4, transpose, trunc, unpack2x16float, unpack4x8unorm, workgroupBarrier };
