// typegpu/std/operators.js
import { isFloat32VecInstance, isInteger32VecInstance, isMat, isMatInstance, isUint32VecInstance, isVec, isVecInstance } from "../data/wgslTypes.js";
import { SignatureNotSupportedError } from "../errors.js";
import { stitch } from "../core/resolve/stitch.js";
import { unify } from "../tgsl/conversion.js";
import { abstractFloat, f16, f32, i32, u32 } from "../data/numeric.js";
import { vec2i, vec2u, vec3i, vec3u, vec4i, vec4u, vecTypeToConstructor } from "../data/vector.js";
import { dualImpl } from "../core/function/dualImpl.js";
import { VectorOps } from "../data/vectorOps.js";

//#region src/std/operators.ts
const getPrimitive = (t) => "primitive" in t ? t.primitive : t;
const makeBinarySignature = (opts) => (lhs, rhs) => {
	const { restrict } = opts ?? {};
	const fail = (msg) => {
		if (restrict) throw new SignatureNotSupportedError([lhs, rhs], restrict);
		throw new Error(`Cannot apply operator to ${lhs.type} and ${rhs.type}: ${msg}`);
	};
	if (opts?.noMat && (isMat(lhs) || isMat(rhs))) return fail("matrices not supported");
	const lhsC = isVec(lhs) || isMat(lhs);
	const rhsC = isVec(rhs) || isMat(rhs);
	if (!lhsC && !rhsC) {
		const unified$1 = unify([lhs, rhs], restrict);
		if (!unified$1) return fail("incompatible scalar types");
		return {
			argTypes: unified$1,
			returnType: unified$1[0]
		};
	}
	if (lhsC && rhsC) {
		if (opts?.matVecProduct && isVec(lhs) !== isVec(rhs)) return {
			argTypes: [lhs, rhs],
			returnType: isVec(lhs) ? lhs : rhs
		};
		if (lhs.type !== rhs.type) return fail("operands must have the same type");
		return {
			argTypes: [lhs, rhs],
			returnType: lhs
		};
	}
	const [scalar, composite] = lhsC ? [rhs, lhs] : [lhs, rhs];
	const unified = unify([scalar], [getPrimitive(composite)]);
	if (!unified) return fail(`scalar not convertible to ${getPrimitive(composite).type}`);
	return {
		argTypes: lhsC ? [lhs, unified[0]] : [unified[0], rhs],
		returnType: composite
	};
};
const binaryArithmeticSignature = makeBinarySignature();
const binaryMulSignature = makeBinarySignature({ matVecProduct: true });
const binaryDivSignature = makeBinarySignature({
	noMat: true,
	restrict: [
		f32,
		f16,
		abstractFloat
	]
});
function cpuAdd(lhs, rhs) {
	if (typeof lhs === "number" && typeof rhs === "number") return lhs + rhs;
	if (typeof lhs === "number" && isVecInstance(rhs)) return VectorOps.addMixed[rhs.kind](rhs, lhs);
	if (isVecInstance(lhs) && typeof rhs === "number") return VectorOps.addMixed[lhs.kind](lhs, rhs);
	if (isVecInstance(lhs) && isVecInstance(rhs) || isMatInstance(lhs) && isMatInstance(rhs)) return VectorOps.add[lhs.kind](lhs, rhs);
	throw new Error("Add/Sub called with invalid arguments.");
}
const add = dualImpl({
	name: "add",
	signature: binaryArithmeticSignature,
	normalImpl: cpuAdd,
	codegenImpl: (_ctx, [lhs, rhs]) => stitch`(${lhs} + ${rhs})`
});
function cpuSub(lhs, rhs) {
	return cpuAdd(lhs, cpuMul(-1, rhs));
}
const sub = dualImpl({
	name: "sub",
	signature: binaryArithmeticSignature,
	normalImpl: cpuSub,
	codegenImpl: (_ctx, [lhs, rhs]) => stitch`(${lhs} - ${rhs})`
});
function cpuMul(lhs, rhs) {
	if (typeof lhs === "number" && typeof rhs === "number") return lhs * rhs;
	if (typeof lhs === "number" && (isVecInstance(rhs) || isMatInstance(rhs))) return VectorOps.mulSxV[rhs.kind](lhs, rhs);
	if ((isVecInstance(lhs) || isMatInstance(lhs)) && typeof rhs === "number") return VectorOps.mulSxV[lhs.kind](rhs, lhs);
	if (isVecInstance(lhs) && isVecInstance(rhs)) return VectorOps.mulVxV[lhs.kind](lhs, rhs);
	if (isFloat32VecInstance(lhs) && isMatInstance(rhs)) return VectorOps.mulVxM[rhs.kind](lhs, rhs);
	if (isMatInstance(lhs) && isFloat32VecInstance(rhs)) return VectorOps.mulMxV[lhs.kind](lhs, rhs);
	if (isMatInstance(lhs) && isMatInstance(rhs)) return VectorOps.mulVxV[lhs.kind](lhs, rhs);
	throw new Error("Mul called with invalid arguments.");
}
const mul = dualImpl({
	name: "mul",
	signature: binaryMulSignature,
	normalImpl: cpuMul,
	codegenImpl: (_ctx, [lhs, rhs]) => stitch`(${lhs} * ${rhs})`
});
function cpuDiv(lhs, rhs) {
	if (typeof lhs === "number" && typeof rhs === "number") return lhs / rhs;
	if (typeof lhs === "number" && isVecInstance(rhs)) {
		const schema = vecTypeToConstructor[rhs.kind];
		return VectorOps.div[rhs.kind](schema(lhs), rhs);
	}
	if (isVecInstance(lhs) && typeof rhs === "number") {
		const schema = vecTypeToConstructor[lhs.kind];
		return VectorOps.div[lhs.kind](lhs, schema(rhs));
	}
	if (isVecInstance(lhs) && isVecInstance(rhs)) return VectorOps.div[lhs.kind](lhs, rhs);
	throw new Error("Div called with invalid arguments.");
}
const div = dualImpl({
	name: "div",
	signature: binaryDivSignature,
	normalImpl: cpuDiv,
	codegenImpl: (_ctx, [lhs, rhs]) => stitch`(${lhs} / ${rhs})`,
	ignoreImplicitCastWarning: true
});
/**
* @privateRemarks
* Both JS and WGSL implementations use truncated definition of modulo
*/
const mod = dualImpl({
	name: "mod",
	signature: binaryDivSignature,
	normalImpl: ((a, b) => {
		if (typeof a === "number" && typeof b === "number") return a % b;
		if (typeof a === "number" && isVecInstance(b)) {
			const schema = vecTypeToConstructor[b.kind];
			return VectorOps.mod[b.kind](schema(a), b);
		}
		if (isVecInstance(a) && typeof b === "number") {
			const schema = vecTypeToConstructor[a.kind];
			return VectorOps.mod[a.kind](a, schema(b));
		}
		if (isVecInstance(a) && isVecInstance(b)) return VectorOps.mod[a.kind](a, b);
		throw new Error("Mod called with invalid arguments, expected types: number or vector.");
	}),
	codegenImpl: (_ctx, [lhs, rhs]) => stitch`(${lhs} % ${rhs})`
});
function cpuNeg(value) {
	if (typeof value === "number") return -value;
	return VectorOps.neg[value.kind](value);
}
const neg = dualImpl({
	name: "neg",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: cpuNeg,
	codegenImpl: (_ctx, [arg]) => stitch`-(${arg})`
});
const anyConcreteInteger = [
	i32,
	u32,
	vec2i,
	vec3i,
	vec4i,
	vec2u,
	vec3u,
	vec4u
];
const intVecToUnsignedVec = {
	vec2i: vec2u,
	vec2u,
	vec3i: vec3u,
	vec3u,
	vec4i: vec4u,
	vec4u
};
const bitShiftSignature = (lhs, rhs) => {
	const lhsUnified = unify([lhs], anyConcreteInteger)?.[0];
	if (!lhsUnified) throw new SignatureNotSupportedError([lhs], anyConcreteInteger);
	let rhsType;
	if (isVec(lhsUnified)) {
		const cc = lhsUnified.componentCount;
		const vecU = cc === 2 ? vec2u : cc === 3 ? vec3u : vec4u;
		const rhsUnified = unify([rhs], [u32, vecU])?.[0];
		if (!rhsUnified) throw new SignatureNotSupportedError([rhs], [u32, vecU]);
		rhsType = rhsUnified;
	} else rhsType = u32;
	return {
		argTypes: [lhsUnified, rhsType],
		returnType: lhsUnified
	};
};
function cpuBitShiftLeft(lhs, rhs) {
	if (typeof lhs === "number" && typeof rhs === "number") return lhs << rhs;
	if (isInteger32VecInstance(lhs) && isUint32VecInstance(rhs) && lhs.length == rhs.length) return VectorOps.bitShiftLeft[lhs.kind](lhs, rhs);
	if (isInteger32VecInstance(lhs) && typeof rhs === "number") {
		const rhsVec = intVecToUnsignedVec[lhs.kind](rhs);
		return VectorOps.bitShiftLeft[lhs.kind](lhs, rhsVec);
	}
	throw new Error("bitShiftLeft called with invalid arguments, expected types: number or integer vector (rhs must be the same arity as lhs).");
}
const bitShiftLeft = dualImpl({
	name: "bitShiftLeft",
	signature: bitShiftSignature,
	normalImpl: cpuBitShiftLeft,
	codegenImpl: (_ctx, [lhs, rhs]) => {
		if (isVec(lhs.dataType) && !isVec(rhs.dataType)) {
			const cc = lhs.dataType.componentCount;
			return stitch`(${lhs} << ${cc === 2 ? "vec2u" : cc === 3 ? "vec3u" : "vec4u"}(${rhs}))`;
		}
		return stitch`(${lhs} << ${rhs})`;
	}
});
function cpuBitShiftRight(lhs, rhs) {
	if (typeof lhs === "number" && typeof rhs === "number") return lhs >> rhs;
	if (isInteger32VecInstance(lhs) && isUint32VecInstance(rhs) && lhs.length == rhs.length) return VectorOps.bitShiftRight[lhs.kind](lhs, rhs);
	if (isInteger32VecInstance(lhs) && typeof rhs === "number") {
		const rhsVec = intVecToUnsignedVec[lhs.kind](rhs);
		return VectorOps.bitShiftRight[lhs.kind](lhs, rhsVec);
	}
	throw new Error("bitShiftRight called with invalid arguments, expected types: number or integer vector (rhs must be the same arity as lhs).");
}
const bitShiftRight = dualImpl({
	name: "bitShiftRight",
	signature: bitShiftSignature,
	normalImpl: cpuBitShiftRight,
	codegenImpl: (_ctx, [lhs, rhs]) => {
		if (isVec(lhs.dataType) && !isVec(rhs.dataType)) {
			const cc = lhs.dataType.componentCount;
			return stitch`(${lhs} >> ${cc === 2 ? "vec2u" : cc === 3 ? "vec3u" : "vec4u"}(${rhs}))`;
		}
		return stitch`(${lhs} >> ${rhs})`;
	}
});

//#endregion
export { add, bitShiftLeft, bitShiftRight, div, mod, mul, neg, sub };
