// typegpu/std/packing.js
import { stitch } from "../core/resolve/stitch.js";
import { u32 } from "../data/numeric.js";
import { vec2f, vec4f } from "../data/vector.js";
import { dualImpl } from "../core/function/dualImpl.js";
import * as TB from "typed-binary";

//#region src/std/packing.ts
/**
* @privateRemarks
* https://gpuweb.github.io/gpuweb/wgsl/#unpack2x16float-builtin
*/
const unpack2x16float = dualImpl({
	name: "unpack2x16float",
	normalImpl: (e) => {
		const buffer = /* @__PURE__ */ new ArrayBuffer(4);
		new TB.BufferWriter(buffer).writeUint32(e);
		const reader = new TB.BufferReader(buffer);
		return vec2f(reader.readFloat16(), reader.readFloat16());
	},
	signature: {
		argTypes: [u32],
		returnType: vec2f
	},
	codegenImpl: (_ctx, [e]) => stitch`unpack2x16float(${e})`
});
/**
* @privateRemarks
* https://gpuweb.github.io/gpuweb/wgsl/#pack2x16float-builtin
*/
const pack2x16float = dualImpl({
	name: "pack2x16float",
	normalImpl: (e) => {
		const buffer = /* @__PURE__ */ new ArrayBuffer(4);
		const writer = new TB.BufferWriter(buffer);
		writer.writeFloat16(e.x);
		writer.writeFloat16(e.y);
		return u32(new TB.BufferReader(buffer).readUint32());
	},
	signature: {
		argTypes: [vec2f],
		returnType: u32
	},
	codegenImpl: (_ctx, [e]) => stitch`pack2x16float(${e})`
});
/**
* @privateRemarks
* https://gpuweb.github.io/gpuweb/wgsl/#unpack4x8unorm-builtin
*/
const unpack4x8unorm = dualImpl({
	name: "unpack4x8unorm",
	normalImpl: (e) => {
		const buffer = /* @__PURE__ */ new ArrayBuffer(4);
		new TB.BufferWriter(buffer).writeUint32(e);
		const reader = new TB.BufferReader(buffer);
		return vec4f(reader.readUint8() / 255, reader.readUint8() / 255, reader.readUint8() / 255, reader.readUint8() / 255);
	},
	signature: {
		argTypes: [u32],
		returnType: vec4f
	},
	codegenImpl: (_ctx, [e]) => stitch`unpack4x8unorm(${e})`
});
/**
* @privateRemarks
* https://gpuweb.github.io/gpuweb/wgsl/#pack4x8unorm-builtin
*/
const pack4x8unorm = dualImpl({
	name: "pack4x8unorm",
	normalImpl: (e) => {
		const buffer = /* @__PURE__ */ new ArrayBuffer(4);
		const writer = new TB.BufferWriter(buffer);
		writer.writeUint8(e.x * 255);
		writer.writeUint8(e.y * 255);
		writer.writeUint8(e.z * 255);
		writer.writeUint8(e.w * 255);
		return u32(new TB.BufferReader(buffer).readUint32());
	},
	signature: {
		argTypes: [vec4f],
		returnType: u32
	},
	codegenImpl: (_ctx, [e]) => stitch`pack4x8unorm(${e})`
});

//#endregion
export { pack2x16float, pack4x8unorm, unpack2x16float, unpack4x8unorm };
