// typegpu/std/subgroup.js
import { stitch } from "../core/resolve/stitch.js";
import { unify } from "../tgsl/conversion.js";
import { bool, i32, u32 } from "../data/numeric.js";
import { vec4u } from "../data/vector.js";
import { dualImpl } from "../core/function/dualImpl.js";

//#region src/std/subgroup.ts
const errorMessage = "Subgroup operations can only be used in the GPU context.";
const subgroupAdd = dualImpl({
	name: "subgroupAdd",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [arg]) => stitch`subgroupAdd(${arg})`
});
const subgroupExclusiveAdd = dualImpl({
	name: "subgroupExclusiveAdd",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [arg]) => stitch`subgroupExclusiveAdd(${arg})`
});
const subgroupInclusiveAdd = dualImpl({
	name: "subgroupInclusiveAdd",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [arg]) => stitch`subgroupInclusiveAdd(${arg})`
});
const subgroupAll = dualImpl({
	name: "subgroupAll",
	signature: {
		argTypes: [bool],
		returnType: bool
	},
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e]) => stitch`subgroupAll(${e})`
});
const subgroupAnd = dualImpl({
	name: "subgroupAnd",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e]) => stitch`subgroupAnd(${e})`
});
const subgroupAny = dualImpl({
	name: "subgroupAny",
	signature: {
		argTypes: [bool],
		returnType: bool
	},
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e]) => stitch`subgroupAny(${e})`
});
const subgroupBallot = dualImpl({
	name: "subgroupBallot",
	signature: {
		argTypes: [bool],
		returnType: vec4u
	},
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e]) => stitch`subgroupBallot(${e})`
});
const subgroupBroadcast = dualImpl({
	name: "subgroupBroadcast",
	signature: (...args) => {
		const id = unify([args[1]], [i32, u32]);
		if (!id) throw new Error(`subgroupBroadcast's second argument has to be compatible with i32 or u32. Got: ${args[1].type}`);
		return {
			argTypes: [args[0], id[0]],
			returnType: args[0]
		};
	},
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e, index]) => stitch`subgroupBroadcast(${e}, ${index})`
});
const subgroupBroadcastFirst = dualImpl({
	name: "subgroupBroadcastFirst",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e]) => stitch`subgroupBroadcastFirst(${e})`
});
const subgroupElect = dualImpl({
	name: "subgroupElect",
	signature: {
		argTypes: [],
		returnType: bool
	},
	normalImpl: errorMessage,
	codegenImpl: () => stitch`subgroupElect()`
});
const subgroupMax = dualImpl({
	name: "subgroupMax",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [arg]) => stitch`subgroupMax(${arg})`
});
const subgroupMin = dualImpl({
	name: "subgroupMin",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [arg]) => stitch`subgroupMin(${arg})`
});
const subgroupMul = dualImpl({
	name: "subgroupMul",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [arg]) => stitch`subgroupMul(${arg})`
});
const subgroupExclusiveMul = dualImpl({
	name: "subgroupExclusiveMul",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [arg]) => stitch`subgroupExclusiveMul(${arg})`
});
const subgroupInclusiveMul = dualImpl({
	name: "subgroupInclusiveMul",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [arg]) => stitch`subgroupInclusiveMul(${arg})`
});
const subgroupOr = dualImpl({
	name: "subgroupOr",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e]) => stitch`subgroupOr(${e})`
});
const subgroupShuffle = dualImpl({
	name: "subgroupShuffle",
	signature: (...args) => {
		const id = unify([args[1]], [i32, u32]);
		if (!id) throw new Error(`subgroupShuffle's second argument has to be compatible with i32 or u32. Got: ${args[1].type}`);
		return {
			argTypes: [args[0], id[0]],
			returnType: args[0]
		};
	},
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e, index]) => stitch`subgroupShuffle(${e}, ${index})`
});
const subgroupShuffleDown = dualImpl({
	name: "subgroupShuffleDown",
	signature: (...args) => {
		const delta = unify([args[1]], [u32]);
		if (!delta) throw new Error(`subgroupShuffleDown's second argument has to be compatible with u32. Got: ${args[1].type}`);
		return {
			argTypes: [args[0], delta[0]],
			returnType: args[0]
		};
	},
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e, delta]) => stitch`subgroupShuffleDown(${e}, ${delta})`
});
const subgroupShuffleUp = dualImpl({
	name: "subgroupShuffleUp",
	signature: (...args) => {
		const delta = unify([args[1]], [u32]);
		if (!delta) throw new Error(`subgroupShuffleUp's second argument has to be compatible with u32. Got: ${args[1].type}`);
		return {
			argTypes: [args[0], delta[0]],
			returnType: args[0]
		};
	},
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e, delta]) => stitch`subgroupShuffleUp(${e}, ${delta})`
});
const subgroupShuffleXor = dualImpl({
	name: "subgroupShuffleXor",
	signature: (...args) => {
		const mask = unify([args[1]], [u32]);
		if (!mask) throw new Error(`subgroupShuffleXor's second argument has to be compatible with u32. Got: ${args[1].type}`);
		return {
			argTypes: [args[0], mask[0]],
			returnType: args[0]
		};
	},
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e, mask]) => stitch`subgroupShuffleXor(${e}, ${mask})`
});
const subgroupXor = dualImpl({
	name: "subgroupXor",
	signature: (arg) => ({
		argTypes: [arg],
		returnType: arg
	}),
	normalImpl: errorMessage,
	codegenImpl: (_ctx, [e]) => stitch`subgroupXor(${e})`
});

//#endregion
export { subgroupAdd, subgroupAll, subgroupAnd, subgroupAny, subgroupBallot, subgroupBroadcast, subgroupBroadcastFirst, subgroupElect, subgroupExclusiveAdd, subgroupExclusiveMul, subgroupInclusiveAdd, subgroupInclusiveMul, subgroupMax, subgroupMin, subgroupMul, subgroupOr, subgroupShuffle, subgroupShuffleDown, subgroupShuffleUp, subgroupShuffleXor, subgroupXor };
