// typegpu/std/texture.js
import { Void } from "../data/wgslTypes.js";
import { stitch } from "../core/resolve/stitch.js";
import { f32, i32, u32 } from "../data/numeric.js";
import { vec2u, vec3u, vec4f, vec4i, vec4u } from "../data/vector.js";
import { MissingCpuImplError, dualImpl } from "../core/function/dualImpl.js";
import { isWgslExternalTexture, isWgslTexture } from "../data/texture.js";
import { getTextureFormatInfo } from "../core/texture/textureFormats.js";

//#region src/std/texture.ts
function sampleCpu(_texture, _sampler, _coords, _offsetOrArrayIndex, _maybeOffset) {
	throw new MissingCpuImplError("Texture sampling relies on GPU resources and cannot be executed outside of a draw call");
}
const textureSample = dualImpl({
	name: "textureSample",
	normalImpl: sampleCpu,
	codegenImpl: (_ctx, args) => stitch`textureSample(${args})`,
	signature: (...args) => {
		return {
			argTypes: args,
			returnType: args[0].type.startsWith("texture_depth") ? f32 : vec4f
		};
	}
});
function sampleBiasCpu(_texture, _sampler, _coords, _biasOrArrayIndex, _biasOrOffset, _maybeOffset) {
	throw new MissingCpuImplError("Texture sampling with bias relies on GPU resources and cannot be executed outside of a draw call");
}
const textureSampleBias = dualImpl({
	name: "textureSampleBias",
	normalImpl: sampleBiasCpu,
	codegenImpl: (_ctx, args) => stitch`textureSampleBias(${args})`,
	signature: (...args) => ({
		argTypes: args,
		returnType: vec4f
	})
});
function sampleLevelCpu(_texture, _sampler, _coords, _level, _offsetOrArrayIndex, _maybeOffset) {
	throw new MissingCpuImplError("Texture sampling relies on GPU resources and cannot be executed outside of a draw call");
}
const textureSampleLevel = dualImpl({
	name: "textureSampleLevel",
	normalImpl: sampleLevelCpu,
	codegenImpl: (_ctx, args) => stitch`textureSampleLevel(${args})`,
	signature: (...args) => {
		return {
			argTypes: args,
			returnType: args[0].type.startsWith("texture_depth") ? f32 : vec4f
		};
	}
});
function textureLoadCpu(_texture, _coords, _levelOrArrayIndex) {
	throw new MissingCpuImplError("`textureLoad` relies on GPU resources and cannot be executed outside of a draw call");
}
const textureLoad = dualImpl({
	name: "textureLoad",
	normalImpl: textureLoadCpu,
	codegenImpl: (_ctx, args) => stitch`textureLoad(${args})`,
	signature: (...args) => {
		const texture = args[0];
		if (isWgslTexture(texture)) {
			const isDepth = texture.type.startsWith("texture_depth");
			const sampleType = texture.sampleType;
			return {
				argTypes: args,
				returnType: isDepth ? f32 : sampleType.type === "f32" ? vec4f : sampleType.type === "u32" ? vec4u : vec4i
			};
		}
		if (isWgslExternalTexture(texture)) return {
			argTypes: args,
			returnType: vec4f
		};
		const format = texture.format;
		return {
			argTypes: args,
			returnType: getTextureFormatInfo(format).vectorType
		};
	}
});
function textureStoreCpu(_texture, _coords, _arrayIndexOrValue, _maybeValue) {
	throw new MissingCpuImplError("`textureStore` relies on GPU resources and cannot be executed outside of a draw call");
}
const textureStore = dualImpl({
	name: "textureStore",
	normalImpl: textureStoreCpu,
	codegenImpl: (_ctx, args) => stitch`textureStore(${args})`,
	signature: (...args) => ({
		argTypes: args,
		returnType: Void
	})
});
function textureDimensionsCpu(_texture, _level) {
	throw new MissingCpuImplError("`textureDimensions` relies on GPU resources and cannot be executed outside of a draw call");
}
const textureDimensions = dualImpl({
	name: "textureDimensions",
	normalImpl: textureDimensionsCpu,
	codegenImpl: (_ctx, args) => stitch`textureDimensions(${args})`,
	signature: (...args) => {
		const dim = args[0].dimension;
		if (dim === "1d") return {
			argTypes: args,
			returnType: u32
		};
		if (dim === "3d") return {
			argTypes: args,
			returnType: vec3u
		};
		return {
			argTypes: args,
			returnType: vec2u
		};
	}
});
const textureGatherCpu = (..._args) => {
	throw new Error("Texture gather relies on GPU resources and cannot be executed outside of a draw call");
};
const sampleTypeToVecType = {
	f32: vec4f,
	i32: vec4i,
	u32: vec4u
};
const textureGather = dualImpl({
	name: "textureGather",
	normalImpl: textureGatherCpu,
	codegenImpl: (_ctx, args) => stitch`textureGather(${args})`,
	signature: (...args) => {
		if (args[0].type.startsWith("texture")) {
			const [texture$1, sampler$1, coords$1, _$1, ...rest$1] = args;
			return {
				argTypes: texture$1.type === "texture_depth_2d_array" || texture$1.type === "texture_depth_cube_array" ? [
					texture$1,
					sampler$1,
					coords$1,
					[u32, i32],
					...rest$1
				] : args,
				returnType: vec4f
			};
		}
		const [_, texture, sampler, coords, ...rest] = args;
		return {
			argTypes: texture.type === "texture_2d_array" || texture.type === "texture_cube_array" ? [
				[u32, i32],
				texture,
				sampler,
				coords,
				[u32, i32],
				...rest
			] : [
				[u32, i32],
				texture,
				sampler,
				coords,
				...rest
			],
			returnType: sampleTypeToVecType[texture.sampleType.type]
		};
	}
});
function textureSampleCompareCpu(_texture, _sampler, _coords, _depthRefOrArrayIndex, _depthRefOrOffset, _maybeOffset) {
	throw new MissingCpuImplError("Texture comparison sampling relies on GPU resources and cannot be executed outside of a draw call");
}
const textureSampleCompare = dualImpl({
	name: "textureSampleCompare",
	normalImpl: textureSampleCompareCpu,
	codegenImpl: (_ctx, args) => stitch`textureSampleCompare(${args})`,
	signature: (...args) => ({
		argTypes: args,
		returnType: f32
	})
});
function textureSampleCompareLevelCpu(_texture, _sampler, _coords, _depthRefOrArrayIndex, _depthRefOrOffset, _maybeOffset) {
	throw new MissingCpuImplError("Texture comparison sampling with level relies on GPU resources and cannot be executed outside of a draw call");
}
const textureSampleCompareLevel = dualImpl({
	name: "textureSampleCompareLevel",
	normalImpl: textureSampleCompareLevelCpu,
	codegenImpl: (ctx, args) => stitch`textureSampleCompareLevel(${args})`,
	signature: (...args) => ({
		argTypes: args,
		returnType: f32
	})
});
function textureSampleBaseClampToEdgeCpu(_texture, _sampler, _coords) {
	throw new MissingCpuImplError("Texture sampling with base clamp to edge is not supported outside of GPU mode.");
}
function sampleGradCpu(_texture, _sampler, _coords, _ddxOrArrayIndex, _ddyOrDdx, _offsetOrDdy, _maybeOffset) {
	throw new MissingCpuImplError("Texture sampling with gradients relies on GPU resources and cannot be executed outside of a draw call");
}
const textureSampleGrad = dualImpl({
	name: "textureSampleGrad",
	normalImpl: sampleGradCpu,
	codegenImpl: (_ctx, args) => stitch`textureSampleGrad(${args})`,
	signature: (...args) => ({
		argTypes: args,
		returnType: vec4f
	})
});
const textureSampleBaseClampToEdge = dualImpl({
	name: "textureSampleBaseClampToEdge",
	normalImpl: textureSampleBaseClampToEdgeCpu,
	codegenImpl: (_ctx, args) => stitch`textureSampleBaseClampToEdge(${args})`,
	signature: (...args) => ({
		argTypes: args,
		returnType: vec4f
	})
});

//#endregion
export { textureDimensions, textureGather, textureLoad, textureSample, textureSampleBaseClampToEdge, textureSampleBias, textureSampleCompare, textureSampleCompareLevel, textureSampleGrad, textureSampleLevel, textureStore };
