// tsover-runtime/dist/index.js
//#region src/index.ts
function polyfillSymbol(name) {
	if (typeof Symbol[name] === "undefined") Object.defineProperty(Symbol, name, {
		value: Symbol.for(`Symbol.${name}`),
		writable: false,
		enumerable: false,
		configurable: false
	});
}
/**
* Checks if a value has the [T] property
*/
function hasOperator(value, operator) {
	return typeof value?.[operator] === "function";
}
function binOp(a, b, opSymbol, fallback) {
	let result = Operator.deferOperation;
	if (hasOperator(a, opSymbol)) result = a[opSymbol](a, b);
	if (result === Operator.deferOperation && hasOperator(b, opSymbol)) result = b[opSymbol](a, b);
	if (result === Operator.deferOperation) return fallback(a, b);
	return result;
}
const normalAdd = (a, b) => a + b;
/**
* Performs the + operation with support for operator overloading
* If either operand has Symbol.operatorPlus, uses that operator
* Otherwise falls back to standard JavaScript + behavior
*/
function add(a, b) {
	if ((typeof a === "number" || typeof a === "string") && (typeof b === "number" || typeof b === "string")) return a + b;
	return binOp(a, b, Operator.plus, normalAdd);
}
const normalSub = (a, b) => a - b;
/**
* Performs the - operation with support for operator overloading
* If either operand has Symbol.operatorMinus, uses that operator
* Otherwise falls back to standard JavaScript - behavior
*/
function sub(a, b) {
	if (typeof a === "number" && typeof b === "number") return a - b;
	return binOp(a, b, Operator.minus, normalSub);
}
const normalMul = (a, b) => a * b;
/**
* Performs the * operation with support for operator overloading
* If either operand has Symbol.operatorStar, uses that operator
* Otherwise falls back to standard JavaScript * behavior
*/
function mul(a, b) {
	if (typeof a === "number" && typeof b === "number") return a * b;
	return binOp(a, b, Operator.star, normalMul);
}
const normalDiv = (a, b) => a / b;
/**
* Performs the / operation with support for operator overloading
* If either operand has Symbol.operatorSlash, uses that operator
* Otherwise falls back to standard JavaScript / behavior
*/
function div(a, b) {
	if (typeof a === "number" && typeof b === "number") return a / b;
	return binOp(a, b, Operator.slash, normalDiv);
}
const normalPow = (a, b) => a ** b;
/**
* Performs the ** operation with support for operator overloading
* If either operand has Symbol.operatorStarStar, uses that operator
* Otherwise falls back to standard JavaScript ** behavior
*/
function pow(a, b) {
	if (typeof a === "number" && typeof b === "number") return a ** b;
	return binOp(a, b, Operator.starStar, normalPow);
}
const normalMod = (a, b) => a % b;
/**
* Performs the % operation with support for operator overloading
* If either operand has Symbol.operatorPercent, uses that operator
* Otherwise falls back to standard JavaScript % behavior
*/
function mod(a, b) {
	if (typeof a === "number" && typeof b === "number") return a % b;
	return binOp(a, b, Operator.percent, normalMod);
}
const Operator = /* @__PURE__ */ (() => {
	polyfillSymbol("deferOperation");
	polyfillSymbol("operatorPlus");
	polyfillSymbol("operatorMinus");
	polyfillSymbol("operatorStar");
	polyfillSymbol("operatorSlash");
	polyfillSymbol("operatorStarStar");
	polyfillSymbol("operatorPercent");
	polyfillSymbol("operatorEqEq");
	polyfillSymbol("operatorPrePlusPlus");
	polyfillSymbol("operatorPreMinusMinus");
	polyfillSymbol("operatorPostPlusPlus");
	polyfillSymbol("operatorPostMinusMinus");
	polyfillSymbol("operatorPreMinus");
	globalThis.__tsover_add = add;
	globalThis.__tsover_sub = sub;
	globalThis.__tsover_mul = mul;
	globalThis.__tsover_div = div;
	globalThis.__tsover_pow = pow;
	globalThis.__tsover_mod = mod;
	return {
		deferOperation: Symbol.deferOperation,
		plus: Symbol.operatorPlus,
		minus: Symbol.operatorMinus,
		star: Symbol.operatorStar,
		slash: Symbol.operatorSlash,
		starStar: Symbol.operatorStarStar,
		percent: Symbol.operatorPercent,
		eqEq: Symbol.operatorEqEq,
		prePlusPlus: Symbol.operatorPrePlusPlus,
		preMinusMinus: Symbol.operatorPreMinusMinus,
		postPlusPlus: Symbol.operatorPostPlusPlus,
		postMinusMinus: Symbol.operatorPostMinusMinus,
		preMinus: Symbol.operatorPreMinus
	};
})();
//#endregion
export { Operator };
